
<!DOCTYPE html>


<html lang="en" data-content_root="../" >

  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="viewport" content="width=device-width, initial-scale=1" />

    <title>46. Advanced integration &#8212; Single-cell best practices</title>
  
  
  
  <script data-cfasync="false">
    document.documentElement.dataset.mode = localStorage.getItem("mode") || "";
    document.documentElement.dataset.theme = localStorage.getItem("theme") || "";
  </script>
  
  <!-- Loaded before other Sphinx assets -->
  <link href="../_static/styles/theme.css?digest=dfe6caa3a7d634c4db9b" rel="stylesheet" />
<link href="../_static/styles/bootstrap.css?digest=dfe6caa3a7d634c4db9b" rel="stylesheet" />
<link href="../_static/styles/pydata-sphinx-theme.css?digest=dfe6caa3a7d634c4db9b" rel="stylesheet" />

  
  <link href="../_static/vendor/fontawesome/6.5.2/css/all.min.css?digest=dfe6caa3a7d634c4db9b" rel="stylesheet" />
  <link rel="preload" as="font" type="font/woff2" crossorigin href="../_static/vendor/fontawesome/6.5.2/webfonts/fa-solid-900.woff2" />
<link rel="preload" as="font" type="font/woff2" crossorigin href="../_static/vendor/fontawesome/6.5.2/webfonts/fa-brands-400.woff2" />
<link rel="preload" as="font" type="font/woff2" crossorigin href="../_static/vendor/fontawesome/6.5.2/webfonts/fa-regular-400.woff2" />

    <link rel="stylesheet" type="text/css" href="../_static/pygments.css?v=03e43079" />
    <link rel="stylesheet" type="text/css" href="../_static/styles/sphinx-book-theme.css?v=eba8b062" />
    <link rel="stylesheet" type="text/css" href="../_static/togglebutton.css?v=13237357" />
    <link rel="stylesheet" type="text/css" href="../_static/copybutton.css?v=76b2166b" />
    <link rel="stylesheet" type="text/css" href="../_static/mystnb.8ecb98da25f57f5357bf6f572d296f466b2cfe2517ffebfabe82451661e28f02.css" />
    <link rel="stylesheet" type="text/css" href="../_static/sphinx-thebe.css?v=4fa983c6" />
    <link rel="stylesheet" type="text/css" href="../_static/sphinx-design.min.css?v=95c83b7e" />
    <link rel="stylesheet" type="text/css" href="../_static/book.css?v=148c11e2" />
  
  <!-- Pre-loaded scripts that we'll load fully later -->
  <link rel="preload" as="script" href="../_static/scripts/bootstrap.js?digest=dfe6caa3a7d634c4db9b" />
<link rel="preload" as="script" href="../_static/scripts/pydata-sphinx-theme.js?digest=dfe6caa3a7d634c4db9b" />
  <script src="../_static/vendor/fontawesome/6.5.2/js/all.min.js?digest=dfe6caa3a7d634c4db9b"></script>

    <script src="../_static/documentation_options.js?v=9eb32ce0"></script>
    <script src="../_static/doctools.js?v=9a2dae69"></script>
    <script src="../_static/sphinx_highlight.js?v=dc90522c"></script>
    <script src="../_static/clipboard.min.js?v=a7894cd8"></script>
    <script src="../_static/copybutton.js?v=f281be69"></script>
    <script src="../_static/scripts/sphinx-book-theme.js?v=887ef09a"></script>
    <script>let toggleHintShow = 'Click to show';</script>
    <script>let toggleHintHide = 'Click to hide';</script>
    <script>let toggleOpenOnPrint = 'true';</script>
    <script src="../_static/togglebutton.js?v=4a39c7ea"></script>
    <script>var togglebuttonSelector = '.toggle, .admonition.dropdown';</script>
    <script src="../_static/design-tabs.js?v=f930bc37"></script>
    <script>const THEBE_JS_URL = "https://unpkg.com/thebe@0.8.2/lib/index.js"; const thebe_selector = ".thebe,.cell"; const thebe_selector_input = "pre"; const thebe_selector_output = ".output, .cell_output"</script>
    <script async="async" src="../_static/sphinx-thebe.js?v=c100c467"></script>
    <script>var togglebuttonSelector = '.toggle, .admonition.dropdown';</script>
    <script>const THEBE_JS_URL = "https://unpkg.com/thebe@0.8.2/lib/index.js"; const thebe_selector = ".thebe,.cell"; const thebe_selector_input = "pre"; const thebe_selector_output = ".output, .cell_output"</script>
    <script>DOCUMENTATION_OPTIONS.pagename = 'multimodal_integration/advanced_integration';</script>
    <script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
    <script src="../_static/book.js?v=57d67cd3"></script>
    <link rel="icon" href="../_static/favicon.ico"/>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="47. Outlook" href="../outlook.html" />
    <link rel="prev" title="45. Paired integration" href="paired_integration.html" />
  <meta name="viewport" content="width=device-width, initial-scale=1"/>
  <meta name="docsearch:language" content="en"/>
  </head>
  
  
  <body data-bs-spy="scroll" data-bs-target=".bd-toc-nav" data-offset="180" data-bs-root-margin="0px 0px -60%" data-default-mode="">

  
  
  <div id="pst-skip-link" class="skip-link d-print-none"><a href="#main-content">Skip to main content</a></div>
  
  <div id="pst-scroll-pixel-helper"></div>
  
  <button type="button" class="btn rounded-pill" id="pst-back-to-top">
    <i class="fa-solid fa-arrow-up"></i>Back to top</button>

  
  <input type="checkbox"
          class="sidebar-toggle"
          id="pst-primary-sidebar-checkbox"/>
  <label class="overlay overlay-primary" for="pst-primary-sidebar-checkbox"></label>
  
  <input type="checkbox"
          class="sidebar-toggle"
          id="pst-secondary-sidebar-checkbox"/>
  <label class="overlay overlay-secondary" for="pst-secondary-sidebar-checkbox"></label>
  
  <div class="search-button__wrapper">
    <div class="search-button__overlay"></div>
    <div class="search-button__search-container">
<form class="bd-search d-flex align-items-center"
      action="../search.html"
      method="get">
  <i class="fa-solid fa-magnifying-glass"></i>
  <input type="search"
         class="form-control"
         name="q"
         id="search-input"
         placeholder="Search this book..."
         aria-label="Search this book..."
         autocomplete="off"
         autocorrect="off"
         autocapitalize="off"
         spellcheck="false"/>
  <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd>K</kbd></span>
</form></div>
  </div>

  <div class="pst-async-banner-revealer d-none">
  <aside id="bd-header-version-warning" class="d-none d-print-none" aria-label="Version warning"></aside>
</div>

  
    <header class="bd-header navbar navbar-expand-lg bd-navbar d-print-none">
    </header>
  

  <div class="bd-container">
    <div class="bd-container__inner bd-page-width">
      
      
      
      <div class="bd-sidebar-primary bd-sidebar">
        

  
  <div class="sidebar-header-items sidebar-primary__section">
    
    
    
    
  </div>
  
    <div class="sidebar-primary-items__start sidebar-primary__section">
        <div class="sidebar-primary-item">

  
    
  

<a class="navbar-brand logo" href="../preamble.html">
  
  
  
  
  
    
    
      
    
    
    <img src="../_static/logo.png" class="logo__image only-light" alt="Single-cell best practices - Home"/>
    <script>document.write(`<img src="../_static/logo.png" class="logo__image only-dark" alt="Single-cell best practices - Home"/>`);</script>
  
  
</a></div>
        <div class="sidebar-primary-item">

 <script>
 document.write(`
   <button class="btn search-button-field search-button__button" title="Search" aria-label="Search" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="fa-solid fa-magnifying-glass"></i>
    <span class="search-button__default-text">Search</span>
    <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd class="kbd-shortcut__modifier">K</kbd></span>
   </button>
 `);
 </script></div>
        <div class="sidebar-primary-item"><nav class="bd-links bd-docs-nav" aria-label="Main">
    <div class="bd-toc-item navbar-nav active">
        <p aria-level="2" class="caption" role="heading"><span class="caption-text">Introduction</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../introduction/prior_art.html">1. Prior art</a></li>
<li class="toctree-l1"><a class="reference internal" href="../introduction/scrna_seq.html">2. Single-cell RNA sequencing</a></li>
<li class="toctree-l1"><a class="reference internal" href="../introduction/raw_data_processing.html">3. Raw data processing</a></li>
<li class="toctree-l1"><a class="reference internal" href="../introduction/analysis_tools.html">4. Analysis frameworks and tools</a></li>
<li class="toctree-l1"><a class="reference internal" href="../introduction/interoperability.html">5. Interoperability</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Preprocessing and visualization</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../preprocessing_visualization/quality_control.html">6. Quality Control</a></li>
<li class="toctree-l1"><a class="reference internal" href="../preprocessing_visualization/normalization.html">7. Normalization</a></li>
<li class="toctree-l1"><a class="reference internal" href="../preprocessing_visualization/feature_selection.html">8. Feature selection</a></li>
<li class="toctree-l1"><a class="reference internal" href="../preprocessing_visualization/dimensionality_reduction.html">9. Dimensionality Reduction</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Identifying cellular structure</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../cellular_structure/clustering.html">10. Clustering</a></li>
<li class="toctree-l1"><a class="reference internal" href="../cellular_structure/annotation.html">11. Annotation</a></li>
<li class="toctree-l1"><a class="reference internal" href="../cellular_structure/integration.html">12. Data integration</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Inferring trajectories</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../trajectories/pseudotemporal.html">13. Pseudotemporal ordering</a></li>
<li class="toctree-l1"><a class="reference internal" href="../trajectories/rna_velocity.html">14. RNA velocity</a></li>
<li class="toctree-l1"><a class="reference internal" href="../trajectories/lineage_tracing.html">15. Lineage tracing</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Dealing with conditions</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../conditions/differential_gene_expression.html">16. Differential gene expression analysis</a></li>
<li class="toctree-l1"><a class="reference internal" href="../conditions/compositional.html">17. Compositional analysis</a></li>
<li class="toctree-l1"><a class="reference internal" href="../conditions/gsea_pathway.html">18. Gene set enrichment and pathway analysis</a></li>
<li class="toctree-l1"><a class="reference internal" href="../conditions/perturbation_modeling.html">19. Perturbation modeling</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Modeling mechanisms</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../mechanisms/gene_regulatory_networks.html">20. Gene regulatory networks</a></li>
<li class="toctree-l1"><a class="reference internal" href="../mechanisms/cell_cell_communication.html">21. Cell-cell communication</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Deconvolution</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../deconvolution/bulk_deconvolution.html">22. Bulk deconvolution</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Chromatin Accessibility</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../chromatin_accessibility/introduction.html">23. Single-cell ATAC sequencing</a></li>
<li class="toctree-l1"><a class="reference internal" href="../chromatin_accessibility/quality_control.html">24. Quality Control</a></li>
<li class="toctree-l1"><a class="reference internal" href="../chromatin_accessibility/gene_regulatory_networks_atac.html">25. Gene regulatory networks using chromatin accessibility</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Spatial omics</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../spatial/introduction.html">26. Single-cell data resolved in space</a></li>
<li class="toctree-l1"><a class="reference internal" href="../spatial/neighborhood.html">27. Neighborhood analysis</a></li>
<li class="toctree-l1"><a class="reference internal" href="../spatial/domains.html">28. Spatial domains</a></li>
<li class="toctree-l1"><a class="reference internal" href="../spatial/spatially_variable_genes.html">29. Spatially variable genes</a></li>
<li class="toctree-l1"><a class="reference internal" href="../spatial/deconvolution.html">30. Spatial deconvolution</a></li>
<li class="toctree-l1"><a class="reference internal" href="../spatial/imputation.html">31. Imputation</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Surface protein</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../surface_protein/quality_control.html">32. Quality control</a></li>
<li class="toctree-l1"><a class="reference internal" href="../surface_protein/normalization.html">33. Normalization</a></li>
<li class="toctree-l1"><a class="reference internal" href="../surface_protein/doublet_detection.html">34. Doublet detection</a></li>
<li class="toctree-l1"><a class="reference internal" href="../surface_protein/dimensionality_reduction.html">35. Dimensionality Reduction</a></li>
<li class="toctree-l1"><a class="reference internal" href="../surface_protein/batch_correction.html">36. Batch correction</a></li>
<li class="toctree-l1"><a class="reference internal" href="../surface_protein/annotation.html">37. Annotation</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Adaptive immune receptor repertoire</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../air_repertoire/ir_profiling.html">38. Immune Receptor Profiling</a></li>
<li class="toctree-l1"><a class="reference internal" href="../air_repertoire/clonotype.html">39. Clonotype analysis</a></li>



<li class="toctree-l1"><a class="reference internal" href="../air_repertoire/specificity.html">43. Specificity analysis</a></li>
<li class="toctree-l1"><a class="reference internal" href="../air_repertoire/multimodal_integration.html">44. Integrating AIR and transcriptomics</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Multimodal integration</span></p>
<ul class="current nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="paired_integration.html">45. Paired integration</a></li>
<li class="toctree-l1 current active"><a class="current reference internal" href="#">46. Advanced integration</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Outlook</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../outlook.html">47. Outlook</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Acknowledgements</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../acknowledgements.html">48. Acknowledgements</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Glossary</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../glossary.html">49. Glossary</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Changelog</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../CHANGELOG.html">Changelog</a></li>
</ul>

    </div>
</nav></div>
    </div>
  
  
  <div class="sidebar-primary-items__end sidebar-primary__section">
  </div>
  
  <div id="rtd-footer-container"></div>


      </div>
      
      <main id="main-content" class="bd-main" role="main">
        
        

<div class="sbt-scroll-pixel-helper"></div>

          <div class="bd-content">
            <div class="bd-article-container">
              
              <div class="bd-header-article d-print-none">
<div class="header-article-items header-article__inner">
  
    <div class="header-article-items__start">
      
        <div class="header-article-item"><button class="sidebar-toggle primary-toggle btn btn-sm" title="Toggle primary sidebar" data-bs-placement="bottom" data-bs-toggle="tooltip">
  <span class="fa-solid fa-bars"></span>
</button></div>
      
    </div>
  
  
    <div class="header-article-items__end">
      
        <div class="header-article-item">

<div class="article-header-buttons">





<div class="dropdown dropdown-source-buttons">
  <button class="btn dropdown-toggle" type="button" data-bs-toggle="dropdown" aria-expanded="false" aria-label="Source repositories">
    <i class="fab fa-github"></i>
  </button>
  <ul class="dropdown-menu">
      
      
      
      <li><a href="https://github.com/theislab/single-cell-best-practices" target="_blank"
   class="btn btn-sm btn-source-repository-button dropdown-item"
   title="Source repository"
   data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fab fa-github"></i>
  </span>
<span class="btn__text-container">Repository</span>
</a>
</li>
      
      
      
      
      <li><a href="https://github.com/theislab/single-cell-best-practices/edit/main/jupyter-book/multimodal_integration/advanced_integration.ipynb" target="_blank"
   class="btn btn-sm btn-source-edit-button dropdown-item"
   title="Suggest edit"
   data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-pencil-alt"></i>
  </span>
<span class="btn__text-container">Suggest edit</span>
</a>
</li>
      
      
      
      
      <li><a href="https://github.com/theislab/single-cell-best-practices/issues/new?title=Issue%20on%20page%20%2Fmultimodal_integration/advanced_integration.html&body=Your%20issue%20content%20here." target="_blank"
   class="btn btn-sm btn-source-issues-button dropdown-item"
   title="Open an issue"
   data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-lightbulb"></i>
  </span>
<span class="btn__text-container">Open issue</span>
</a>
</li>
      
  </ul>
</div>






<div class="dropdown dropdown-download-buttons">
  <button class="btn dropdown-toggle" type="button" data-bs-toggle="dropdown" aria-expanded="false" aria-label="Download this page">
    <i class="fas fa-download"></i>
  </button>
  <ul class="dropdown-menu">
      
      
      
      <li><a href="../_sources/multimodal_integration/advanced_integration.ipynb" target="_blank"
   class="btn btn-sm btn-download-source-button dropdown-item"
   title="Download source file"
   data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-file"></i>
  </span>
<span class="btn__text-container">.ipynb</span>
</a>
</li>
      
      
      
      
      <li>
<button onclick="window.print()"
  class="btn btn-sm btn-download-pdf-button dropdown-item"
  title="Print to PDF"
  data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-file-pdf"></i>
  </span>
<span class="btn__text-container">.pdf</span>
</button>
</li>
      
  </ul>
</div>




<button onclick="toggleFullScreen()"
  class="btn btn-sm btn-fullscreen-button"
  title="Fullscreen mode"
  data-bs-placement="bottom" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-expand"></i>
  </span>

</button>



<script>
document.write(`
  <button class="btn btn-sm nav-link pst-navbar-icon theme-switch-button" title="light/dark" aria-label="light/dark" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="theme-switch fa-solid fa-sun fa-lg" data-mode="light"></i>
    <i class="theme-switch fa-solid fa-moon fa-lg" data-mode="dark"></i>
    <i class="theme-switch fa-solid fa-circle-half-stroke fa-lg" data-mode="auto"></i>
  </button>
`);
</script>


<script>
document.write(`
  <button class="btn btn-sm pst-navbar-icon search-button search-button__button" title="Search" aria-label="Search" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="fa-solid fa-magnifying-glass fa-lg"></i>
  </button>
`);
</script>
<button class="sidebar-toggle secondary-toggle btn btn-sm" title="Toggle secondary sidebar" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <span class="fa-solid fa-list"></span>
</button>
</div></div>
      
    </div>
  
</div>
</div>
              
              

<div id="jb-print-docs-body" class="onlyprint">
    <h1>Advanced integration</h1>
    <!-- Table of contents -->
    <div id="print-main-content">
        <div id="jb-print-toc">
            
            <div>
                <h2> Contents </h2>
            </div>
            <nav aria-label="Page">
                <ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#motivation">46.1. Motivation</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#environment-setup">46.2. Environment setup</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#prepare-data">46.3. Prepare data</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#unpaired-integration-with-graph-linked-unified-embedding-glue">46.4. Unpaired integration with Graph Linked Unified Embedding (GLUE)</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#graph-construction">46.4.1. Graph construction</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#configure-data">46.4.2. Configure data</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#integration-with-partially-overlapping-data-and-query-to-reference-mapping-with-totalvi">46.5. Integration with partially overlapping data and query-to-reference mapping with totalVI</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#query-to-reference-mapping">46.5.1. Query-to-reference mapping</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#query-mapping-and-imputation-with-seurat-s-wnn">46.6. Query mapping and imputation with Seurat’s WNN</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#mapping-adt-onto-an-rna-atlas-with-bridge-integration">46.7. Mapping ADT onto an RNA atlas with bridge integration</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#trimodal-integration-and-query-to-reference-mapping-with-multigrate">46.8. Trimodal integration and query-to-reference mapping with multigrate</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#building-the-reference">46.8.1. Building the reference</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#querying-the-trimodal-reference">46.8.2. Querying the trimodal reference</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#session-info">46.9. Session info</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#references">46.10. References</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#contributors">46.11. Contributors</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#authors">46.11.1. Authors</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#reviewers">46.11.2. Reviewers</a></li>
</ul>
</li>
</ul>
            </nav>
        </div>
    </div>
</div>

              
                
<div id="searchbox"></div>
                <article class="bd-article">
                  
  <section class="tex2jax_ignore mathjax_ignore" id="advanced-integration">
<span id="multimodal-integration-advanced-integration"></span><h1><span class="section-number">46. </span>Advanced integration<a class="headerlink" href="#advanced-integration" title="Link to this heading">#</a></h1>
<details class="sd-sphinx-override sd-dropdown sd-card sd-mb-3">
<summary class="sd-summary-title sd-card-header">
<span class="sd-summary-text"><i class="fas fa-brain"></i>   Key takeaways</span><span class="sd-summary-state-marker sd-summary-chevron-right"><svg version="1.1" width="1.5em" height="1.5em" class="sd-octicon sd-octicon-chevron-right" viewBox="0 0 24 24" aria-hidden="true"><path d="M8.72 18.78a.75.75 0 0 1 0-1.06L14.44 12 8.72 6.28a.751.751 0 0 1 .018-1.042.751.751 0 0 1 1.042-.018l6.25 6.25a.75.75 0 0 1 0 1.06l-6.25 6.25a.75.75 0 0 1-1.06 0Z"></path></svg></span></summary><div class="sd-summary-content sd-card-body docutils">
<div class="sd-card sd-sphinx-override sd-mb-3 sd-shadow-sm sd-card-hover docutils">
<div class="sd-card-body docutils">
<p class="sd-card-text">Advanced multimodal integration techniques techniques are particularly useful when data from different modalities (e.g., RNA-seq, ATAC-seq, and ADT) originate from separate experiments and are not jointly measured per cell.</p>
</div>
<a class="sd-stretched-link sd-hide-link-text reference internal" href="#multimodal-integration-advanced-integration-key-takeaway-1"><span class="std std-ref">Motivation</span></a></div>
<div class="sd-card sd-sphinx-override sd-mb-3 sd-shadow-sm sd-card-hover docutils">
<div class="sd-card-body docutils">
<p class="sd-card-text">Query-to-reference mapping allows for the integration of new data into a pre-existing reference framework, enabling tasks like cell type prediction and imputation of missing modalities (e.g., predicting protein abundance from RNA-seq data)</p>
</div>
<a class="sd-stretched-link sd-hide-link-text reference internal" href="#multimodal-integration-advanced-integration-key-takeaway-2"><span class="std std-ref">Integration with partially overlapping data and query-to-reference mapping with totalVI</span></a></div>
</div>
</details><details class="sd-sphinx-override sd-dropdown sd-card sd-mb-3">
<summary class="sd-summary-title sd-card-header">
<span class="sd-summary-text"><i class="fa-solid fa-gear"></i>   Environment setup</span><span class="sd-summary-state-marker sd-summary-chevron-right"><svg version="1.1" width="1.5em" height="1.5em" class="sd-octicon sd-octicon-chevron-right" viewBox="0 0 24 24" aria-hidden="true"><path d="M8.72 18.78a.75.75 0 0 1 0-1.06L14.44 12 8.72 6.28a.751.751 0 0 1 .018-1.042.751.751 0 0 1 1.042-.018l6.25 6.25a.75.75 0 0 1 0 1.06l-6.25 6.25a.75.75 0 0 1-1.06 0Z"></path></svg></span></summary><div class="sd-summary-content sd-card-body docutils">
<div class="sd-tab-set docutils">
<input checked="checked" id="sd-tab-item-0" name="sd-tab-set-0" type="radio">
<label class="sd-tab-label" for="sd-tab-item-0">
Steps</label><div class="sd-tab-content docutils">
<ol class="arabic">
<li><p class="sd-card-text"><strong>Install conda</strong>:</p>
<ul class="simple">
<li><p class="sd-card-text">Before creating the environment, ensure that <a class="reference external" href="https://docs.conda.io/projects/conda/en/latest/user-guide/install/index.html">conda</a> is installed on your system.</p></li>
</ul>
</li>
<li><p class="sd-card-text"><strong>Save the yml content</strong>:</p>
<ul class="simple">
<li><p class="sd-card-text">Copy the content from the yml tab into a file named <code class="docutils literal notranslate"><span class="pre">environment.yml</span></code>.</p></li>
</ul>
</li>
<li><p class="sd-card-text"><strong>Create the environment</strong>:</p>
<ul>
<li><p class="sd-card-text">Open a terminal or command prompt.</p></li>
<li><p class="sd-card-text">Run the following command:</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>conda<span class="w"> </span>env<span class="w"> </span>create<span class="w"> </span>-f<span class="w"> </span>environment.yml
</pre></div>
</div>
</li>
</ul>
</li>
<li><p class="sd-card-text"><strong>Activate the environment</strong>:</p>
<ul>
<li><p class="sd-card-text">After the environment is created, activate it using:</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>conda<span class="w"> </span>activate<span class="w"> </span>&lt;environment_name&gt;
</pre></div>
</div>
</li>
<li><p class="sd-card-text">Replace <code class="docutils literal notranslate"><span class="pre">&lt;environment_name&gt;</span></code> with the name specified in the <code class="docutils literal notranslate"><span class="pre">environment.yml</span></code> file. In the yml file it will look like this:</p>
<div class="highlight-yaml notranslate"><div class="highlight"><pre><span></span><span class="nt">name</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">&lt;environment_name&gt;</span>
</pre></div>
</div>
</li>
</ul>
</li>
<li><p class="sd-card-text"><strong>Verify the installation</strong>:</p>
<ul>
<li><p class="sd-card-text">Check that the environment was created successfully by running:</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>conda<span class="w"> </span>env<span class="w"> </span>list
</pre></div>
</div>
</li>
</ul>
</li>
</ol>
</div>
<input id="sd-tab-item-1" name="sd-tab-set-0" type="radio">
<label class="sd-tab-label" for="sd-tab-item-1">
yml</label><div class="sd-tab-content docutils">
<div class="highlight-yaml notranslate"><div class="highlight"><pre><span></span><span class="nt">name</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">advanced-integration</span>
<span class="nt">channels</span><span class="p">:</span>
<span class="w">  </span><span class="p p-Indicator">-</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">conda-forge</span>
<span class="w">  </span><span class="p p-Indicator">-</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">bioconda</span>
<span class="w">  </span><span class="p p-Indicator">-</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">defaults</span>
<span class="nt">dependencies</span><span class="p">:</span>
<span class="w">  </span><span class="p p-Indicator">-</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">conda-forge::python=3.12.11</span>
<span class="w">  </span><span class="p p-Indicator">-</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">conda-forge::scanpy=1.11.5</span>
<span class="w">  </span><span class="p p-Indicator">-</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">conda-forge::r-base=4.4</span>
<span class="w">  </span><span class="p p-Indicator">-</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">conda-forge::r-seurat=5.1</span>
<span class="w">  </span><span class="p p-Indicator">-</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">conda-forge::r-remotes</span>
<span class="w">  </span><span class="p p-Indicator">-</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">conda-forge::r-devtools</span>
<span class="w">  </span><span class="p p-Indicator">-</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">bioconda::bioconductor-singlecellexperiment</span>
<span class="w">  </span><span class="p p-Indicator">-</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">conda-forge::rpy2=3.6</span>
<span class="w">  </span><span class="p p-Indicator">-</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">bioconda::anndata2ri=1.3</span>
<span class="w">  </span><span class="p p-Indicator">-</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">conda-forge::r-spatstat</span>
<span class="w">  </span><span class="p p-Indicator">-</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">conda-forge::pip</span>
<span class="w">  </span><span class="p p-Indicator">-</span><span class="w"> </span><span class="nt">pip</span><span class="p">:</span>
<span class="w">      </span><span class="p p-Indicator">-</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">scvi-tools==1.4.1</span>
<span class="w">      </span><span class="p p-Indicator">-</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">git+https://github.com/theislab/multigrate.git@main</span>
<span class="w">      </span><span class="p p-Indicator">-</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">scglue==0.4.0</span>
</pre></div>
</div>
</div>
</div>
</div>
</details><!-- END DROPDOWNS -->
<section id="motivation">
<span id="multimodal-integration-advanced-integration-key-takeaway-1"></span><h2><span class="section-number">46.1. </span>Motivation<a class="headerlink" href="#motivation" title="Link to this heading">#</a></h2>
<p>In this notebook, we showcase more advanced methods and techniques for multimodal integration. Examples of such advanced techniques are unpaired integration, integration of partially overlapping data or multimodal query-to-reference mapping. These are especially useful when the measurements of various modalities were not done jointly per cell, but originate from different experiments.</p>
<p>We discuss each of these cases in more detail in the corresponding sections below.</p>
</section>
<section id="environment-setup">
<h2><span class="section-number">46.2. </span>Environment setup<a class="headerlink" href="#environment-setup" title="Link to this heading">#</a></h2>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span><span class="w"> </span><span class="nn">logging</span>

<span class="kn">import</span><span class="w"> </span><span class="nn">anndata</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">anndata2ri</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">multigrate</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">mtg</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">networkx</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">nx</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">numpy</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">np</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">pandas</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">pd</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">rpy2.rinterface_lib.callbacks</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">scanpy</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">sc</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">scglue</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">rpy2.robjects</span><span class="w"> </span><span class="kn">import</span> <span class="n">pandas2ri</span><span class="p">,</span> <span class="n">r</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">scvi.model</span><span class="w"> </span><span class="kn">import</span> <span class="n">TOTALVI</span>

<span class="n">rpy2</span><span class="o">.</span><span class="n">rinterface_lib</span><span class="o">.</span><span class="n">callbacks</span><span class="o">.</span><span class="n">logger</span><span class="o">.</span><span class="n">setLevel</span><span class="p">(</span><span class="n">logging</span><span class="o">.</span><span class="n">ERROR</span><span class="p">)</span>

<span class="n">pandas2ri</span><span class="o">.</span><span class="n">activate</span><span class="p">()</span>
<span class="n">anndata2ri</span><span class="o">.</span><span class="n">activate</span><span class="p">()</span>

<span class="o">%</span><span class="k">load_ext</span> rpy2.ipython

<span class="kn">import</span><span class="w"> </span><span class="nn">warnings</span>

<span class="n">warnings</span><span class="o">.</span><span class="n">filterwarnings</span><span class="p">(</span><span class="s2">&quot;ignore&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Global seed set to 0
/lustre/groups/ml01/workspace/anastasia.litinetskaya/miniconda3/envs/advanced-integration/lib/python3.9/site-packages/pytorch_lightning/utilities/warnings.py:53: LightningDeprecationWarning: pytorch_lightning.utilities.warnings.rank_zero_deprecation has been deprecated in v1.6 and will be removed in v1.8. Use the equivalent function from the pytorch_lightning.utilities.rank_zero module instead.
  new_rank_zero_deprecation(
/lustre/groups/ml01/workspace/anastasia.litinetskaya/miniconda3/envs/advanced-integration/lib/python3.9/site-packages/pytorch_lightning/utilities/warnings.py:58: LightningDeprecationWarning: The `pytorch_lightning.loggers.base.rank_zero_experiment` is deprecated in v1.7 and will be removed in v1.9. Please use `pytorch_lightning.loggers.logger.rank_zero_experiment` instead.
  return new_rank_zero_deprecation(*args, **kwargs)
During startup - Warning message:
Setting LC_CTYPE failed, using &quot;C&quot; 
</pre></div>
</div>
</div>
</div>
<p>If you want to run Seurat bridge integration, please additionally install the developmental version of Seurat package with separately using <code class="docutils literal notranslate"><span class="pre">remotes::install_github(&quot;satijalab/seurat&quot;,</span> <span class="pre">&quot;feat/dictionary&quot;,</span> <span class="pre">quiet</span> <span class="pre">=</span> <span class="pre">TRUE)</span></code> or <code class="docutils literal notranslate"><span class="pre">devtools::install_github(&quot;https://github.com/satijalab/seurat/tree/feat/dictionary&quot;)</span></code> command.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="o">%%</span><span class="k">R</span>
suppressPackageStartupMessages({
    library(SingleCellExperiment)
    library(Seurat)
})
set.seed(123)
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>    WARNING: The R package &quot;reticulate&quot; only fixed recently
    an issue that caused a segfault when used with rpy2:
    https://github.com/rstudio/reticulate/pull/1188
    Make sure that you use a version of that package that includes
    the fix.
    
</pre></div>
</div>
</div>
</div>
</section>
<section id="prepare-data">
<h2><span class="section-number">46.3. </span>Prepare data<a class="headerlink" href="#prepare-data" title="Link to this heading">#</a></h2>
<p>We continue using Multiome and CITE-seq data from the NeurIPS 2021 single cell competition <span id="id1">[<a class="reference internal" href="#id550" title="Malte D Luecken, Daniel Bernard Burkhardt, Robrecht Cannoodt, Christopher Lance, Aditi Agrawal, Hananeh Aliee, Ann T Chen, Louise Deconinck, Angela M Detweiler, Alejandro A Granados, Shelly Huynh, Laura Isacco, Yang Joon Kim, Dominik Klein, BONY DE KUMAR, Sunil Kuppasani, Heiko Lickert, Aaron McGeever, Honey Mekonen, Joaquin Caceres Melgarejo, Maurizio Morri, Michaela Müller, Norma Neff, Sheryl Paul, Bastian Rieck, Kaylie Schneider, Scott Steelman, Michael Sterr, Daniel J. Treacy, Alexander Tong, Alexandra-Chloe Villani, Guilin Wang, Jia Yan, Ce Zhang, Angela Oliveira Pisco, Smita Krishnaswamy, Fabian J Theis, and Jonathan M. Bloom. A sandbox for prediction and integration of dna, rna, and proteins in single cells. In Thirty-fifth Conference on Neural Information Processing Systems Datasets and Benchmarks Track (Round 2). 2021. URL: https://openreview.net/forum?id=gN35BGa1Rt.">Luecken <em>et al.</em>, 2021</a>]</span>.</p>
<p>We will need:</p>
<ul class="simple">
<li><p>RNA-seq part of the multiome and ADT from the CITE-seq data for unpaired integration with <a class="reference internal" href="#glue"><span class="xref myst">GLUE</span></a></p></li>
<li><p>paired gene expression and protein from the CITE-seq data for query-to-reference mapping with <a class="reference internal" href="#totalvi"><span class="xref myst">totalVI</span></a></p></li>
<li><p>RNA-seq query from the CITE-seq data for query mapping with <a class="reference internal" href="#seurat"><span class="xref myst">Seurat v4</span></a></p></li>
<li><p>CITE-seq data for <a class="reference internal" href="#bridge"><span class="xref myst">bridge</span></a> mapping</p></li>
<li><p>NeurIPS multiome and NeurIPS CITE-seq for trimodal integration and query-to-reference mapping with <a class="reference internal" href="#multigrate"><span class="xref myst">multigrate</span></a></p></li>
</ul>
<p>Since we showcase query-to-reference mapping functionality of some methods, we split the data into a reference and a query. For the methods that we only use for integration, we also use the reference batches.</p>
<p>We note that the batch names for Multiome and CITE-seq data are the same, but they are actually not the same cells.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">cite_reference_batches</span> <span class="o">=</span> <span class="p">[</span>
    <span class="s2">&quot;s1d1&quot;</span><span class="p">,</span>
    <span class="s2">&quot;s1d2&quot;</span><span class="p">,</span>
    <span class="s2">&quot;s1d3&quot;</span><span class="p">,</span>
<span class="p">]</span>  <span class="c1"># need for totalVI, multigrate and bridge (RNA part)</span>
<span class="n">multiome_reference_batches</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;s1d1&quot;</span><span class="p">,</span> <span class="s2">&quot;s1d2&quot;</span><span class="p">,</span> <span class="s2">&quot;s1d3&quot;</span><span class="p">]</span>  <span class="c1"># need for GLUE and multigrate</span>
<span class="n">cite_query_batches</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;s2d1&quot;</span><span class="p">,</span> <span class="s2">&quot;s2d4&quot;</span><span class="p">]</span>  <span class="c1"># need for totalVI, multigrate and bridge</span>
<span class="n">multiome_query_batches</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;s2d1&quot;</span><span class="p">,</span> <span class="s2">&quot;s2d4&quot;</span><span class="p">]</span>  <span class="c1"># need for multigrate</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">rna_multiome</span> <span class="o">=</span> <span class="n">sc</span><span class="o">.</span><span class="n">read</span><span class="p">(</span>
    <span class="s2">&quot;/lustre/groups/ml01/workspace/anastasia.litinetskaya/data/neurips-multiome/rna_hvg.h5ad&quot;</span>
<span class="p">)</span>
<span class="n">rna_multiome</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>AnnData object with n_obs × n_vars = 69249 × 4000
    obs: &#39;GEX_pct_counts_mt&#39;, &#39;GEX_n_counts&#39;, &#39;GEX_n_genes&#39;, &#39;GEX_size_factors&#39;, &#39;GEX_phase&#39;, &#39;ATAC_nCount_peaks&#39;, &#39;ATAC_atac_fragments&#39;, &#39;ATAC_reads_in_peaks_frac&#39;, &#39;ATAC_blacklist_fraction&#39;, &#39;ATAC_nucleosome_signal&#39;, &#39;cell_type&#39;, &#39;batch&#39;, &#39;ATAC_pseudotime_order&#39;, &#39;GEX_pseudotime_order&#39;, &#39;Samplename&#39;, &#39;Site&#39;, &#39;DonorNumber&#39;, &#39;Modality&#39;, &#39;VendorLot&#39;, &#39;DonorID&#39;, &#39;DonorAge&#39;, &#39;DonorBMI&#39;, &#39;DonorBloodType&#39;, &#39;DonorRace&#39;, &#39;Ethnicity&#39;, &#39;DonorGender&#39;, &#39;QCMeds&#39;, &#39;DonorSmoker&#39;
    var: &#39;feature_types&#39;, &#39;gene_id&#39;, &#39;highly_variable&#39;, &#39;means&#39;, &#39;dispersions&#39;, &#39;dispersions_norm&#39;
    uns: &#39;ATAC_gene_activity_var_names&#39;, &#39;Site_colors&#39;, &#39;batch_colors&#39;, &#39;cell_type_colors&#39;, &#39;dataset_id&#39;, &#39;genome&#39;, &#39;hvg&#39;, &#39;organism&#39;
    obsm: &#39;ATAC_gene_activity&#39;, &#39;ATAC_lsi_full&#39;, &#39;ATAC_lsi_red&#39;, &#39;ATAC_umap&#39;, &#39;GEX_X_pca&#39;, &#39;GEX_X_umap&#39;, &#39;X_umap&#39;
    layers: &#39;counts&#39;
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">atac_multiome</span> <span class="o">=</span> <span class="n">sc</span><span class="o">.</span><span class="n">read</span><span class="p">(</span>
    <span class="s2">&quot;/lustre/groups/ml01/workspace/anastasia.litinetskaya/data/trimodal_neurips/atac_hvf_muon.h5ad&quot;</span>
<span class="p">)</span>
<span class="n">atac_multiome</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>AnnData object with n_obs × n_vars = 69249 × 20000
    obs: &#39;GEX_pct_counts_mt&#39;, &#39;GEX_n_counts&#39;, &#39;GEX_n_genes&#39;, &#39;GEX_size_factors&#39;, &#39;GEX_phase&#39;, &#39;ATAC_nCount_peaks&#39;, &#39;ATAC_atac_fragments&#39;, &#39;ATAC_reads_in_peaks_frac&#39;, &#39;ATAC_blacklist_fraction&#39;, &#39;ATAC_nucleosome_signal&#39;, &#39;cell_type&#39;, &#39;batch&#39;, &#39;ATAC_pseudotime_order&#39;, &#39;GEX_pseudotime_order&#39;, &#39;Samplename&#39;, &#39;Site&#39;, &#39;DonorNumber&#39;, &#39;Modality&#39;, &#39;VendorLot&#39;, &#39;DonorID&#39;, &#39;DonorAge&#39;, &#39;DonorBMI&#39;, &#39;DonorBloodType&#39;, &#39;DonorRace&#39;, &#39;Ethnicity&#39;, &#39;DonorGender&#39;, &#39;QCMeds&#39;, &#39;DonorSmoker&#39;, &#39;technology&#39;, &#39;cell_type_l2&#39;, &#39;cell_type_l1&#39;, &#39;cell_type_l3&#39;, &#39;assay&#39;, &#39;split&#39;
    var: &#39;feature_types&#39;, &#39;gene_id&#39;, &#39;highly_variable&#39;, &#39;means&#39;, &#39;dispersions&#39;, &#39;dispersions_norm&#39;
    uns: &#39;ATAC_gene_activity_var_names&#39;, &#39;dataset_id&#39;, &#39;genome&#39;, &#39;hvg&#39;, &#39;log1p&#39;, &#39;organism&#39;
    obsm: &#39;ATAC_gene_activity&#39;, &#39;ATAC_lsi_full&#39;, &#39;ATAC_lsi_red&#39;, &#39;ATAC_umap&#39;, &#39;GEX_X_pca&#39;, &#39;GEX_X_umap&#39;
    layers: &#39;binary&#39;, &#39;counts&#39;, &#39;cpm&#39;, &#39;tf-idf-binary&#39;, &#39;tf-idf-counts&#39;
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">rna_cite</span> <span class="o">=</span> <span class="n">sc</span><span class="o">.</span><span class="n">read</span><span class="p">(</span>
    <span class="s2">&quot;/lustre/groups/ml01/workspace/anastasia.litinetskaya/data/neurips-cite/rna_hvg.h5ad&quot;</span>
<span class="p">)</span>
<span class="n">rna_cite</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>AnnData object with n_obs × n_vars = 90261 × 4000
    obs: &#39;GEX_n_genes_by_counts&#39;, &#39;GEX_pct_counts_mt&#39;, &#39;GEX_size_factors&#39;, &#39;GEX_phase&#39;, &#39;ADT_n_antibodies_by_counts&#39;, &#39;ADT_total_counts&#39;, &#39;ADT_iso_count&#39;, &#39;cell_type&#39;, &#39;batch&#39;, &#39;ADT_pseudotime_order&#39;, &#39;GEX_pseudotime_order&#39;, &#39;Samplename&#39;, &#39;Site&#39;, &#39;DonorNumber&#39;, &#39;Modality&#39;, &#39;VendorLot&#39;, &#39;DonorID&#39;, &#39;DonorAge&#39;, &#39;DonorBMI&#39;, &#39;DonorBloodType&#39;, &#39;DonorRace&#39;, &#39;Ethnicity&#39;, &#39;DonorGender&#39;, &#39;QCMeds&#39;, &#39;DonorSmoker&#39;, &#39;is_train&#39;
    var: &#39;feature_types&#39;, &#39;gene_id&#39;, &#39;highly_variable&#39;, &#39;means&#39;, &#39;dispersions&#39;, &#39;dispersions_norm&#39;
    uns: &#39;dataset_id&#39;, &#39;genome&#39;, &#39;hvg&#39;, &#39;organism&#39;
    obsm: &#39;ADT_X_pca&#39;, &#39;ADT_X_umap&#39;, &#39;ADT_isotype_controls&#39;, &#39;GEX_X_pca&#39;, &#39;GEX_X_umap&#39;
    layers: &#39;counts&#39;
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">adt_cite</span> <span class="o">=</span> <span class="n">sc</span><span class="o">.</span><span class="n">read</span><span class="p">(</span><span class="s2">&quot;/lustre/groups/ml01/workspace/daniel.strobl/neurips_cite_pp.h5ad&quot;</span><span class="p">)</span>
<span class="n">adt_cite</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>AnnData object with n_obs × n_vars = 120502 × 136
    obs: &#39;donor&#39;, &#39;batch&#39;, &#39;n_genes_by_counts&#39;, &#39;log1p_n_genes_by_counts&#39;, &#39;total_counts&#39;, &#39;log1p_total_counts&#39;, &#39;n_counts&#39;, &#39;outliers&#39;
    var: &#39;gene_ids&#39;, &#39;feature_types&#39;, &#39;n_cells_by_counts&#39;, &#39;mean_counts&#39;, &#39;log1p_mean_counts&#39;, &#39;pct_dropout_by_counts&#39;, &#39;total_counts&#39;, &#39;log1p_total_counts&#39;
    uns: &#39;batch_colors&#39;, &#39;donor_colors&#39;, &#39;neighbors&#39;, &#39;pca&#39;, &#39;umap&#39;
    obsm: &#39;X_isotypes&#39;, &#39;X_pca&#39;, &#39;X_pcahm&#39;, &#39;X_umap&#39;
    varm: &#39;PCs&#39;
    layers: &#39;counts&#39;
    obsp: &#39;connectivities&#39;, &#39;distances&#39;
</pre></div>
</div>
</div>
</div>
<p>There are some differences in <code class="docutils literal notranslate"><span class="pre">.obs_names</span></code> of RNA and ADT of CITE-seq data, so we update them to make sure they align between modalities.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">rna_cite</span><span class="o">.</span><span class="n">obs_names</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Index([&#39;GCATTAGCATAAGCGG-1-s1d1&#39;, &#39;TACAGGTGTTAGAGTA-1-s1d1&#39;,
       &#39;AGGATCTAGGTCTACT-1-s1d1&#39;, &#39;GTAGAAAGTGACACAG-1-s1d1&#39;,
       &#39;TCCGAAAAGGATCATA-1-s1d1&#39;, &#39;CTCCCAATCCATTGGA-1-s1d1&#39;,
       &#39;GACCAATCAATTTCGG-1-s1d1&#39;, &#39;TTCCGGTAGTTGTAAG-1-s1d1&#39;,
       &#39;ACCTGTCAGGACTGGT-1-s1d1&#39;, &#39;TTCGATTTCAGGACAG-1-s1d1&#39;,
       ...
       &#39;TCTTCCTAGCCAACCC-1-s4d9&#39;, &#39;TTCCACGGTTGAGGAC-1-s4d9&#39;,
       &#39;ATTCCTAGTCCAAGAG-1-s4d9&#39;, &#39;GCGGAAAGTACGCGTC-1-s4d9&#39;,
       &#39;TAACTTCAGATACAGT-1-s4d9&#39;, &#39;GAATCACCACGGAAGT-1-s4d9&#39;,
       &#39;GCTGGGTGTACGGATG-1-s4d9&#39;, &#39;TCGAAGTGTGACAGGT-1-s4d9&#39;,
       &#39;GCAGGCTGTTGCATAC-1-s4d9&#39;, &#39;ACGTAACAGGTCTACT-1-s4d9&#39;],
      dtype=&#39;object&#39;, length=90261)
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">adt_cite</span><span class="o">.</span><span class="n">obs_names</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Index([&#39;AAACCCAAGGATGGCT-1-0-0-0-0-0-0-0-0-0-0-0&#39;,
       &#39;AAACCCAAGGCCTAGA-1-0-0-0-0-0-0-0-0-0-0-0&#39;,
       &#39;AAACCCAAGTGAGTGC-1-0-0-0-0-0-0-0-0-0-0-0&#39;,
       &#39;AAACCCACAAGAGGCT-1-0-0-0-0-0-0-0-0-0-0-0&#39;,
       &#39;AAACCCACATCGTGGC-1-0-0-0-0-0-0-0-0-0-0-0&#39;,
       &#39;AAACCCACATTCTCTA-1-0-0-0-0-0-0-0-0-0-0-0&#39;,
       &#39;AAACCCAGTCCGCAGT-1-0-0-0-0-0-0-0-0-0-0-0&#39;,
       &#39;AAACCCAGTGCATACT-1-0-0-0-0-0-0-0-0-0-0-0&#39;,
       &#39;AAACCCAGTTGACGGA-1-0-0-0-0-0-0-0-0-0-0-0&#39;,
       &#39;AAACCCATCGATACTG-1-0-0-0-0-0-0-0-0-0-0-0&#39;,
       ...
       &#39;TTTGGTTGTCGTACTA-1-1&#39;, &#39;TTTGGTTTCACCTCGT-1-1&#39;, &#39;TTTGGTTTCGAGAACG-1-1&#39;,
       &#39;TTTGGTTTCGATACGT-1-1&#39;, &#39;TTTGTTGGTAGCTTTG-1-1&#39;, &#39;TTTGTTGGTCCAATCA-1-1&#39;,
       &#39;TTTGTTGGTGACTAAA-1-1&#39;, &#39;TTTGTTGTCCCTCTCC-1-1&#39;, &#39;TTTGTTGTCTAGAGCT-1-1&#39;,
       &#39;TTTGTTGTCTCTGCCA-1-1&#39;],
      dtype=&#39;object&#39;, length=120502)
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">adt_cite</span><span class="o">.</span><span class="n">obs_names</span> <span class="o">=</span> <span class="p">[</span>
    <span class="n">name</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;-&quot;</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span> <span class="o">+</span> <span class="s2">&quot;-&quot;</span> <span class="o">+</span> <span class="n">name</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;-&quot;</span><span class="p">)[</span><span class="mi">1</span><span class="p">]</span> <span class="o">+</span> <span class="s2">&quot;-&quot;</span> <span class="o">+</span> <span class="n">batch</span>
    <span class="k">for</span> <span class="n">batch</span><span class="p">,</span> <span class="n">name</span> <span class="ow">in</span> <span class="nb">zip</span><span class="p">(</span><span class="n">adt_cite</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s2">&quot;donor&quot;</span><span class="p">],</span> <span class="n">adt_cite</span><span class="o">.</span><span class="n">obs_names</span><span class="p">,</span> <span class="n">strict</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
<span class="p">]</span>
</pre></div>
</div>
</div>
</div>
<p>We subset to the cells that are present in both modalities.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">common_idx</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="nb">set</span><span class="p">(</span><span class="n">rna_cite</span><span class="o">.</span><span class="n">obs_names</span><span class="p">)</span><span class="o">.</span><span class="n">intersection</span><span class="p">(</span><span class="nb">set</span><span class="p">(</span><span class="n">adt_cite</span><span class="o">.</span><span class="n">obs_names</span><span class="p">)))</span>
<span class="n">rna_cite</span> <span class="o">=</span> <span class="n">rna_cite</span><span class="p">[</span><span class="n">common_idx</span><span class="p">]</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
<span class="n">adt_cite</span> <span class="o">=</span> <span class="n">adt_cite</span><span class="p">[</span><span class="n">common_idx</span><span class="p">]</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
</pre></div>
</div>
</div>
</div>
<p>We also copy some of the metadata from RNA to ADT.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">adt_cite</span><span class="o">.</span><span class="n">obs</span> <span class="o">=</span> <span class="n">adt_cite</span><span class="o">.</span><span class="n">obs</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">rna_cite</span><span class="o">.</span><span class="n">obs</span><span class="p">[[</span><span class="s2">&quot;Samplename&quot;</span><span class="p">,</span> <span class="s2">&quot;cell_type&quot;</span><span class="p">]])</span>
</pre></div>
</div>
</div>
</div>
<p>Also we make sure that cells are the same in multiome data.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="k">assert</span> <span class="n">np</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">rna_multiome</span><span class="o">.</span><span class="n">obs_names</span> <span class="o">!=</span> <span class="n">atac_multiome</span><span class="o">.</span><span class="n">obs_names</span><span class="p">)</span> <span class="o">==</span> <span class="mi">0</span>
</pre></div>
</div>
</div>
</div>
<p>Because batch names are the same for different modalities and datasets, we will need to update a bunch of batch columns in <code class="docutils literal notranslate"><span class="pre">.obs</span></code> so we create a helper function for this.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="k">def</span><span class="w"> </span><span class="nf">update_obs_column</span><span class="p">(</span>
    <span class="n">adata</span><span class="p">,</span> <span class="n">obs_column_name</span><span class="p">,</span> <span class="n">suffix</span><span class="p">,</span> <span class="n">how</span><span class="o">=</span><span class="s2">&quot;right&quot;</span><span class="p">,</span> <span class="n">new_obs_column_name</span><span class="o">=</span><span class="kc">None</span>
<span class="p">):</span>
    <span class="k">if</span> <span class="n">new_obs_column_name</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
        <span class="n">new_obs_column_name</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;new_</span><span class="si">{</span><span class="n">obs_column_name</span><span class="si">}</span><span class="s2">&quot;</span>
    <span class="c1"># otherwise can&#39;t use + operator with categorical columns</span>
    <span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="n">obs_column_name</span><span class="p">]</span> <span class="o">=</span> <span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="n">obs_column_name</span><span class="p">]</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="s2">&quot;str&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
    <span class="c1"># create new column in .obs</span>
    <span class="k">if</span> <span class="n">how</span> <span class="o">==</span> <span class="s2">&quot;right&quot;</span><span class="p">:</span>
        <span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="n">new_obs_column_name</span><span class="p">]</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;_</span><span class="si">{</span><span class="n">suffix</span><span class="si">}</span><span class="s2">&quot;</span>
        <span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="n">new_obs_column_name</span><span class="p">]</span> <span class="o">=</span> <span class="p">(</span>
            <span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="n">obs_column_name</span><span class="p">]</span> <span class="o">+</span> <span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="n">new_obs_column_name</span><span class="p">]</span>
        <span class="p">)</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="n">new_obs_column_name</span><span class="p">]</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">suffix</span><span class="si">}</span><span class="s2">_&quot;</span>
        <span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="n">new_obs_column_name</span><span class="p">]</span> <span class="o">=</span> <span class="p">(</span>
            <span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="n">new_obs_column_name</span><span class="p">]</span> <span class="o">+</span> <span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="n">obs_column_name</span><span class="p">]</span>
        <span class="p">)</span>
    <span class="k">return</span> <span class="n">adata</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">rna_multiome</span> <span class="o">=</span> <span class="n">update_obs_column</span><span class="p">(</span><span class="n">rna_multiome</span><span class="p">,</span> <span class="s2">&quot;batch&quot;</span><span class="p">,</span> <span class="s2">&quot;_rna_multiome&quot;</span><span class="p">)</span>
<span class="n">atac_multiome</span> <span class="o">=</span> <span class="n">update_obs_column</span><span class="p">(</span><span class="n">atac_multiome</span><span class="p">,</span> <span class="s2">&quot;batch&quot;</span><span class="p">,</span> <span class="s2">&quot;_atac_multiome&quot;</span><span class="p">)</span>
<span class="n">rna_cite</span> <span class="o">=</span> <span class="n">update_obs_column</span><span class="p">(</span><span class="n">rna_cite</span><span class="p">,</span> <span class="s2">&quot;batch&quot;</span><span class="p">,</span> <span class="s2">&quot;_rna_cite&quot;</span><span class="p">)</span>
<span class="n">adt_cite</span> <span class="o">=</span> <span class="n">update_obs_column</span><span class="p">(</span><span class="n">adt_cite</span><span class="p">,</span> <span class="s2">&quot;donor&quot;</span><span class="p">,</span> <span class="s2">&quot;_adt_cite&quot;</span><span class="p">,</span> <span class="s2">&quot;batch&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<p>Finally we subset to the correct batches the reference datasets and the query.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># query</span>
<span class="n">rna_multiome_query</span> <span class="o">=</span> <span class="n">rna_multiome</span><span class="p">[</span>
    <span class="n">rna_multiome</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s2">&quot;batch&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">isin</span><span class="p">(</span><span class="n">multiome_query_batches</span><span class="p">)</span>
<span class="p">]</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
<span class="n">atac_multiome_query</span> <span class="o">=</span> <span class="n">atac_multiome</span><span class="p">[</span>
    <span class="n">atac_multiome</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s2">&quot;batch&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">isin</span><span class="p">(</span><span class="n">multiome_query_batches</span><span class="p">)</span>
<span class="p">]</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
<span class="n">rna_cite_query</span> <span class="o">=</span> <span class="n">rna_cite</span><span class="p">[</span><span class="n">rna_cite</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s2">&quot;batch&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">isin</span><span class="p">(</span><span class="n">cite_query_batches</span><span class="p">)]</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
<span class="n">adt_cite_query</span> <span class="o">=</span> <span class="n">adt_cite</span><span class="p">[</span><span class="n">adt_cite</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s2">&quot;donor&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">isin</span><span class="p">(</span><span class="n">cite_query_batches</span><span class="p">)]</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
<span class="c1"># reference</span>
<span class="n">rna_multiome</span> <span class="o">=</span> <span class="n">rna_multiome</span><span class="p">[</span>
    <span class="n">rna_multiome</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s2">&quot;batch&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">isin</span><span class="p">(</span><span class="n">multiome_reference_batches</span><span class="p">)</span>
<span class="p">]</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
<span class="n">atac_multiome</span> <span class="o">=</span> <span class="n">atac_multiome</span><span class="p">[</span>
    <span class="n">atac_multiome</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s2">&quot;batch&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">isin</span><span class="p">(</span><span class="n">multiome_reference_batches</span><span class="p">)</span>
<span class="p">]</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
<span class="n">rna_cite</span> <span class="o">=</span> <span class="n">rna_cite</span><span class="p">[</span><span class="n">rna_cite</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s2">&quot;batch&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">isin</span><span class="p">(</span><span class="n">cite_reference_batches</span><span class="p">)]</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
<span class="n">adt_cite</span> <span class="o">=</span> <span class="n">adt_cite</span><span class="p">[</span><span class="n">adt_cite</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s2">&quot;donor&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">isin</span><span class="p">(</span><span class="n">cite_reference_batches</span><span class="p">)]</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
</pre></div>
</div>
</div>
</div>
<p><a id='glue'></a></p>
</section>
<section id="unpaired-integration-with-graph-linked-unified-embedding-glue">
<h2><span class="section-number">46.4. </span>Unpaired integration with Graph Linked Unified Embedding (GLUE)<a class="headerlink" href="#unpaired-integration-with-graph-linked-unified-embedding-glue" title="Link to this heading">#</a></h2>
<p>Opposed to paired integration that we demonstrated before, it is also possible to perform completely unpaired integration. In this case, there is no intersection of cell barcodes or features. Hence, we need some prior knowledge to connect different modalities.</p>
<p>GLUE <span id="id2">[<a class="reference internal" href="#id551" title="Zhi-Jie Cao and Ge Gao. Multi-omics single-cell data integration and regulatory inference with graph-linked embedding. Nature Biotechnology, May 2022. URL: https://doi.org/10.1038/s41587-022-01284-4, doi:10.1038/s41587-022-01284-4.">Cao and Gao, 2022</a>]</span> is a deep learning model for unpaired integration which makes use of a regulatory graph helping connect features from different modalities. The model is based on conditional variational autoencoders, where the model learns to reconstruct while simultaneously allowing for batch correction. To guide the integration, GLUE learns an embedding for each modality for each feature by utilizing a prior knowledge graph. We demonstrate how to use GLUE to integrate unpaired RNA and ADT data using RNA part of Multiome data and ADT part of CITE-seq data from the NeurIPS competition (<a class="reference external" href="https://openproblems.bio/neurips_2021/">https://openproblems.bio/neurips_2021/</a>). To construct the graph, we connect nodes from RNA modality to nodes from ADT modality if and only if the RNA node is a protein encoding gene of a given protein from ADT modality. The output of the GLUE model is a representation of each cell in a shared latent space.</p>
<p>We refer the reader to GLUE tutorial to see how one can integrate unpaired RNA and ATAC with GLUE and for more details about the model <a class="reference external" href="https://scglue.readthedocs.io/en/latest/tutorials.html">https://scglue.readthedocs.io/en/latest/tutorials.html</a>.</p>
<p>We follow the GLUE tutorial on how to preprocess RNA-seq. We log-normalize the raw counts, scale and calculate PCA.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">rna_multiome</span><span class="o">.</span><span class="n">X</span> <span class="o">=</span> <span class="n">rna_multiome</span><span class="o">.</span><span class="n">layers</span><span class="p">[</span><span class="s2">&quot;counts&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
<span class="n">sc</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">normalize_total</span><span class="p">(</span><span class="n">rna_multiome</span><span class="p">)</span>
<span class="n">sc</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">log1p</span><span class="p">(</span><span class="n">rna_multiome</span><span class="p">)</span>
<span class="n">sc</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">scale</span><span class="p">(</span><span class="n">rna_multiome</span><span class="p">)</span>
<span class="n">sc</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">pca</span><span class="p">(</span><span class="n">rna_multiome</span><span class="p">,</span> <span class="n">n_comps</span><span class="o">=</span><span class="mi">100</span><span class="p">,</span> <span class="n">svd_solver</span><span class="o">=</span><span class="s2">&quot;auto&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<p>For ADT counts, we use CLR-normalized counts from <code class="docutils literal notranslate"><span class="pre">.X</span></code> and PCA calculated on the normalized values.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">np</span><span class="o">.</span><span class="n">max</span><span class="p">(</span><span class="n">adt_cite</span><span class="o">.</span><span class="n">X</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>69.27085
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">sc</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">pca</span><span class="p">(</span><span class="n">adt_cite</span><span class="p">,</span> <span class="n">n_comps</span><span class="o">=</span><span class="mi">100</span><span class="p">,</span> <span class="n">svd_solver</span><span class="o">=</span><span class="s2">&quot;auto&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<p>We need to make sure that protein encoding gene names and the corresponding protein names align in both data objects. We will need this later for the graph construction.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">rename_proteins</span> <span class="o">=</span> <span class="p">{</span>
    <span class="s2">&quot;CD103&quot;</span><span class="p">:</span> <span class="s2">&quot;ITGAE&quot;</span><span class="p">,</span>
    <span class="s2">&quot;CD11b&quot;</span><span class="p">:</span> <span class="s2">&quot;ITGAM&quot;</span><span class="p">,</span>
    <span class="s2">&quot;CD11c&quot;</span><span class="p">:</span> <span class="s2">&quot;ITGAX&quot;</span><span class="p">,</span>
    <span class="s2">&quot;CD122&quot;</span><span class="p">:</span> <span class="s2">&quot;IL2RB&quot;</span><span class="p">,</span>
    <span class="s2">&quot;CD124&quot;</span><span class="p">:</span> <span class="s2">&quot;IL4R&quot;</span><span class="p">,</span>
    <span class="s2">&quot;CD127&quot;</span><span class="p">:</span> <span class="s2">&quot;IL7R&quot;</span><span class="p">,</span>
    <span class="s2">&quot;CD13&quot;</span><span class="p">:</span> <span class="s2">&quot;ANPEP&quot;</span><span class="p">,</span>
    <span class="s2">&quot;CD137&quot;</span><span class="p">:</span> <span class="s2">&quot;TNFRSF9&quot;</span><span class="p">,</span>
    <span class="s2">&quot;CD152&quot;</span><span class="p">:</span> <span class="s2">&quot;CTLA4&quot;</span><span class="p">,</span>
    <span class="s2">&quot;CD154&quot;</span><span class="p">:</span> <span class="s2">&quot;CD40LG&quot;</span><span class="p">,</span>
    <span class="s2">&quot;CD16&quot;</span><span class="p">:</span> <span class="s2">&quot;FCGR3A&quot;</span><span class="p">,</span>
    <span class="s2">&quot;CD161&quot;</span><span class="p">:</span> <span class="s2">&quot;KLRB1&quot;</span><span class="p">,</span>
    <span class="s2">&quot;CD185&quot;</span><span class="p">:</span> <span class="s2">&quot;CXCR5&quot;</span><span class="p">,</span>
    <span class="s2">&quot;CD194&quot;</span><span class="p">:</span> <span class="s2">&quot;CCR4&quot;</span><span class="p">,</span>
    <span class="s2">&quot;CD196&quot;</span><span class="p">:</span> <span class="s2">&quot;CCR6&quot;</span><span class="p">,</span>
    <span class="s2">&quot;CD20&quot;</span><span class="p">:</span> <span class="s2">&quot;MS4A1&quot;</span><span class="p">,</span>
    <span class="s2">&quot;CD21&quot;</span><span class="p">:</span> <span class="s2">&quot;CR2&quot;</span><span class="p">,</span>
    <span class="s2">&quot;CD23&quot;</span><span class="p">:</span> <span class="s2">&quot;FCER2&quot;</span><span class="p">,</span>
    <span class="s2">&quot;CD25&quot;</span><span class="p">:</span> <span class="s2">&quot;IL2RA&quot;</span><span class="p">,</span>
    <span class="s2">&quot;CD26&quot;</span><span class="p">:</span> <span class="s2">&quot;DPP4&quot;</span><span class="p">,</span>
    <span class="s2">&quot;CD268&quot;</span><span class="p">:</span> <span class="s2">&quot;TNFRSF13C&quot;</span><span class="p">,</span>
    <span class="s2">&quot;CD272&quot;</span><span class="p">:</span> <span class="s2">&quot;BTLA&quot;</span><span class="p">,</span>
    <span class="s2">&quot;CD278&quot;</span><span class="p">:</span> <span class="s2">&quot;ICOS&quot;</span><span class="p">,</span>
    <span class="s2">&quot;CD29&quot;</span><span class="p">:</span> <span class="s2">&quot;ITGB1&quot;</span><span class="p">,</span>
    <span class="s2">&quot;CD3&quot;</span><span class="p">:</span> <span class="s2">&quot;CD3G&quot;</span><span class="p">,</span>
    <span class="s2">&quot;CD303&quot;</span><span class="p">:</span> <span class="s2">&quot;CLEC4C&quot;</span><span class="p">,</span>
    <span class="s2">&quot;CD304&quot;</span><span class="p">:</span> <span class="s2">&quot;NRP1&quot;</span><span class="p">,</span>
    <span class="s2">&quot;CD314&quot;</span><span class="p">:</span> <span class="s2">&quot;KLRK1&quot;</span><span class="p">,</span>
    <span class="s2">&quot;CD319&quot;</span><span class="p">:</span> <span class="s2">&quot;SLAMF7&quot;</span><span class="p">,</span>
    <span class="s2">&quot;CD335&quot;</span><span class="p">:</span> <span class="s2">&quot;NCR1&quot;</span><span class="p">,</span>
    <span class="s2">&quot;CD35&quot;</span><span class="p">:</span> <span class="s2">&quot;CR1&quot;</span><span class="p">,</span>
    <span class="s2">&quot;CD352&quot;</span><span class="p">:</span> <span class="s2">&quot;SLAMF6&quot;</span><span class="p">,</span>
    <span class="s2">&quot;CD39&quot;</span><span class="p">:</span> <span class="s2">&quot;ENTPD1&quot;</span><span class="p">,</span>
    <span class="s2">&quot;CD49a&quot;</span><span class="p">:</span> <span class="s2">&quot;ITGA1&quot;</span><span class="p">,</span>
    <span class="s2">&quot;CD49f&quot;</span><span class="p">:</span> <span class="s2">&quot;ITGA6&quot;</span><span class="p">,</span>
    <span class="s2">&quot;CD54&quot;</span><span class="p">:</span> <span class="s2">&quot;ICAM1&quot;</span><span class="p">,</span>
    <span class="s2">&quot;CD56&quot;</span><span class="p">:</span> <span class="s2">&quot;NCAM1&quot;</span><span class="p">,</span>
    <span class="s2">&quot;CD62L&quot;</span><span class="p">:</span> <span class="s2">&quot;SELL&quot;</span><span class="p">,</span>
    <span class="s2">&quot;CD71&quot;</span><span class="p">:</span> <span class="s2">&quot;TFRC&quot;</span><span class="p">,</span>
    <span class="s2">&quot;CD73&quot;</span><span class="p">:</span> <span class="s2">&quot;NT5E&quot;</span><span class="p">,</span>
    <span class="s2">&quot;CD79b&quot;</span><span class="p">:</span> <span class="s2">&quot;CD79B&quot;</span><span class="p">,</span>
    <span class="s2">&quot;CD8&quot;</span><span class="p">:</span> <span class="s2">&quot;CD8A&quot;</span><span class="p">,</span>
    <span class="s2">&quot;CD85j&quot;</span><span class="p">:</span> <span class="s2">&quot;LILRB1&quot;</span><span class="p">,</span>
    <span class="s2">&quot;CD88&quot;</span><span class="p">:</span> <span class="s2">&quot;C5AR1&quot;</span><span class="p">,</span>
    <span class="s2">&quot;CD94&quot;</span><span class="p">:</span> <span class="s2">&quot;KLRD1&quot;</span><span class="p">,</span>
    <span class="s2">&quot;CD95&quot;</span><span class="p">:</span> <span class="s2">&quot;FAS&quot;</span><span class="p">,</span>
    <span class="s2">&quot;HLA-DR&quot;</span><span class="p">:</span> <span class="s2">&quot;HLA-DRA&quot;</span><span class="p">,</span>
    <span class="s2">&quot;IgD&quot;</span><span class="p">:</span> <span class="s2">&quot;IGHD&quot;</span><span class="p">,</span>
    <span class="s2">&quot;IgM&quot;</span><span class="p">:</span> <span class="s2">&quot;IGHM&quot;</span><span class="p">,</span>
    <span class="s2">&quot;TCR&quot;</span><span class="p">:</span> <span class="s2">&quot;TRAC&quot;</span><span class="p">,</span>
<span class="p">}</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">adt_cite</span><span class="o">.</span><span class="n">var_names</span> <span class="o">=</span> <span class="p">[</span>
    <span class="n">rename_proteins</span><span class="p">[</span><span class="n">name</span><span class="p">]</span> <span class="k">if</span> <span class="n">name</span> <span class="ow">in</span> <span class="n">rename_proteins</span> <span class="k">else</span> <span class="n">name</span>
    <span class="k">for</span> <span class="n">name</span> <span class="ow">in</span> <span class="n">adt_cite</span><span class="o">.</span><span class="n">var_names</span>
<span class="p">]</span>
</pre></div>
</div>
</div>
</div>
<section id="graph-construction">
<h3><span class="section-number">46.4.1. </span>Graph construction<a class="headerlink" href="#graph-construction" title="Link to this heading">#</a></h3>
<p>Before we can run GLUE integration, we need to construct a graph with prior knowledge on how feature from the two modalities are connected. The node set has to be the union of the features from all the modalities, and the edge weights have to between 0 and 1. To integrate ADT with RNA, we construct the graph the following way: we set the edge weight to 1 if the protein name is the same as the gene name and 0 for all the other edges.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">p</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">adt_cite</span><span class="o">.</span><span class="n">var_names</span><span class="p">)</span>
<span class="n">r</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">rna_multiome</span><span class="o">.</span><span class="n">var_names</span><span class="p">)</span>  <span class="c1"># noqa F811</span>
<span class="c1"># mask entries are set to 1 where protein name is the same as gene name</span>
<span class="n">mask</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">repeat</span><span class="p">(</span><span class="n">p</span><span class="o">.</span><span class="n">reshape</span><span class="p">(</span><span class="o">-</span><span class="mi">1</span><span class="p">,</span> <span class="mi">1</span><span class="p">),</span> <span class="n">r</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span> <span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span> <span class="o">==</span> <span class="n">r</span>
<span class="n">mask</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">mask</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<p>We rename the features so they differ for ADT and RNA modalities.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">rna_vars</span> <span class="o">=</span> <span class="p">[</span><span class="n">v</span> <span class="o">+</span> <span class="s2">&quot;_rna&quot;</span> <span class="k">for</span> <span class="n">v</span> <span class="ow">in</span> <span class="n">rna_multiome</span><span class="o">.</span><span class="n">var_names</span><span class="p">]</span>
<span class="n">prot_vars</span> <span class="o">=</span> <span class="p">[</span><span class="n">v</span> <span class="o">+</span> <span class="s2">&quot;_prot&quot;</span> <span class="k">for</span> <span class="n">v</span> <span class="ow">in</span> <span class="n">adt_cite</span><span class="o">.</span><span class="n">var_names</span><span class="p">]</span>
<span class="n">rna_multiome</span><span class="o">.</span><span class="n">var_names</span> <span class="o">=</span> <span class="n">rna_vars</span>
<span class="n">adt_cite</span><span class="o">.</span><span class="n">var_names</span> <span class="o">=</span> <span class="n">prot_vars</span>
</pre></div>
</div>
</div>
</div>
<p>GLUE also requires each node to have an self-loop with weight 1, so we add these here too.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">adj</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">mask</span><span class="p">,</span> <span class="n">index</span><span class="o">=</span><span class="n">prot_vars</span><span class="p">,</span> <span class="n">columns</span><span class="o">=</span><span class="n">rna_vars</span><span class="p">)</span>
<span class="n">diag_edges</span> <span class="o">=</span> <span class="n">adj</span><span class="p">[</span><span class="n">adj</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">stack</span><span class="p">()</span><span class="o">.</span><span class="n">index</span><span class="o">.</span><span class="n">tolist</span><span class="p">()</span>
<span class="n">diag_edges</span> <span class="o">=</span> <span class="p">[(</span><span class="n">n1</span><span class="p">,</span> <span class="n">n2</span><span class="p">,</span> <span class="p">{</span><span class="s2">&quot;weight&quot;</span><span class="p">:</span> <span class="mf">1.0</span><span class="p">,</span> <span class="s2">&quot;sign&quot;</span><span class="p">:</span> <span class="mi">1</span><span class="p">})</span> <span class="k">for</span> <span class="n">n1</span><span class="p">,</span> <span class="n">n2</span> <span class="ow">in</span> <span class="n">diag_edges</span><span class="p">]</span>
<span class="n">self_loop_rna</span> <span class="o">=</span> <span class="p">[(</span><span class="n">g</span><span class="p">,</span> <span class="n">g</span><span class="p">,</span> <span class="p">{</span><span class="s2">&quot;weight&quot;</span><span class="p">:</span> <span class="mf">1.0</span><span class="p">,</span> <span class="s2">&quot;sign&quot;</span><span class="p">:</span> <span class="mi">1</span><span class="p">})</span> <span class="k">for</span> <span class="n">g</span> <span class="ow">in</span> <span class="n">rna_vars</span><span class="p">]</span>
<span class="n">self_loop_prot</span> <span class="o">=</span> <span class="p">[(</span><span class="n">g</span><span class="p">,</span> <span class="n">g</span><span class="p">,</span> <span class="p">{</span><span class="s2">&quot;weight&quot;</span><span class="p">:</span> <span class="mf">1.0</span><span class="p">,</span> <span class="s2">&quot;sign&quot;</span><span class="p">:</span> <span class="mi">1</span><span class="p">})</span> <span class="k">for</span> <span class="n">g</span> <span class="ow">in</span> <span class="n">prot_vars</span><span class="p">]</span>
</pre></div>
</div>
</div>
</div>
<p>Next we construct the actual graph object.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">graph</span> <span class="o">=</span> <span class="n">nx</span><span class="o">.</span><span class="n">Graph</span><span class="p">()</span>
<span class="n">graph</span><span class="o">.</span><span class="n">add_nodes_from</span><span class="p">(</span><span class="n">rna_vars</span><span class="p">)</span>
<span class="n">graph</span><span class="o">.</span><span class="n">add_nodes_from</span><span class="p">(</span><span class="n">prot_vars</span><span class="p">)</span>
<span class="n">graph</span><span class="o">.</span><span class="n">add_edges_from</span><span class="p">(</span><span class="n">diag_edges</span><span class="p">)</span>
<span class="n">graph</span><span class="o">.</span><span class="n">add_edges_from</span><span class="p">(</span><span class="n">self_loop_prot</span><span class="p">)</span>
<span class="n">graph</span><span class="o">.</span><span class="n">add_edges_from</span><span class="p">(</span><span class="n">self_loop_rna</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<p>We have a graph with 4136 nodes which corresponds to 4000 genes and 136 proteins. We also have 4186 edges with non-zero weights: 4136 self-loops and 50 connections between genes and proteins with the same name.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">graph</span><span class="o">.</span><span class="n">number_of_nodes</span><span class="p">(),</span> <span class="n">graph</span><span class="o">.</span><span class="n">number_of_edges</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>(4136, 4186)
</pre></div>
</div>
</div>
</div>
</section>
<section id="configure-data">
<h3><span class="section-number">46.4.2. </span>Configure data<a class="headerlink" href="#configure-data" title="Link to this heading">#</a></h3>
<p>In this section we again follow the GLUE tutorial to configure the model. First, we set up the RNA encoder-decoder pair to learn to reconstruct raw counts that we assume follow NB distribution. We specify PCA embedding to be used with in the encoder.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">scglue</span><span class="o">.</span><span class="n">models</span><span class="o">.</span><span class="n">configure_dataset</span><span class="p">(</span>
    <span class="n">rna_multiome</span><span class="p">,</span>
    <span class="s2">&quot;NB&quot;</span><span class="p">,</span>
    <span class="n">use_highly_variable</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
    <span class="n">use_batch</span><span class="o">=</span><span class="s2">&quot;Samplename&quot;</span><span class="p">,</span>
    <span class="n">use_layer</span><span class="o">=</span><span class="s2">&quot;counts&quot;</span><span class="p">,</span>
    <span class="n">use_rep</span><span class="o">=</span><span class="s2">&quot;X_pca&quot;</span><span class="p">,</span>
<span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<p>Next, we set up the ADT encoder and decoder to reconstruct normalized counts following the normal distribution and  to use PCA embeddings.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">scglue</span><span class="o">.</span><span class="n">models</span><span class="o">.</span><span class="n">configure_dataset</span><span class="p">(</span>
    <span class="n">adt_cite</span><span class="p">,</span>
    <span class="s2">&quot;Normal&quot;</span><span class="p">,</span>
    <span class="n">use_highly_variable</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
    <span class="n">use_batch</span><span class="o">=</span><span class="s2">&quot;Samplename&quot;</span><span class="p">,</span>
    <span class="n">use_rep</span><span class="o">=</span><span class="s2">&quot;X_pca&quot;</span><span class="p">,</span>
<span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<p>We initialize and train the final model.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">glue</span> <span class="o">=</span> <span class="n">scglue</span><span class="o">.</span><span class="n">models</span><span class="o">.</span><span class="n">fit_SCGLUE</span><span class="p">(</span>
    <span class="p">{</span><span class="s2">&quot;rna&quot;</span><span class="p">:</span> <span class="n">rna_multiome</span><span class="p">,</span> <span class="s2">&quot;adt&quot;</span><span class="p">:</span> <span class="n">adt_cite</span><span class="p">},</span>
    <span class="n">graph</span><span class="p">,</span>
<span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>[INFO] fit_SCGLUE: Pretraining SCGLUE model...
[INFO] autodevice: Using GPU 0 as computation device.
[INFO] check_graph: Checking variable coverage...
[INFO] check_graph: Checking edge attributes...
[INFO] check_graph: Checking self-loops...
[INFO] check_graph: Checking graph symmetry...
[INFO] check_graph: All checks passed!
[INFO] SCGLUEModel: Setting `graph_batch_size` = 1457
[INFO] SCGLUEModel: Setting `max_epochs` = 102
[INFO] SCGLUEModel: Setting `patience` = 9
[INFO] SCGLUEModel: Setting `reduce_lr_patience` = 5
[INFO] SCGLUETrainer: Using training directory: &quot;/tmp/GLUETMPx6ds9f5b&quot;
[INFO] SCGLUETrainer: [Epoch 10] train={&#39;g_nll&#39;: 0.199, &#39;g_kl&#39;: 0.039, &#39;g_elbo&#39;: 0.238, &#39;x_rna_nll&#39;: 0.241, &#39;x_rna_kl&#39;: 0.006, &#39;x_rna_elbo&#39;: 0.247, &#39;x_adt_nll&#39;: 1.934, &#39;x_adt_kl&#39;: 0.21, &#39;x_adt_elbo&#39;: 2.144, &#39;dsc_loss&#39;: 0.636, &#39;vae_loss&#39;: 2.401, &#39;gen_loss&#39;: 2.369}, val={&#39;g_nll&#39;: 0.193, &#39;g_kl&#39;: 0.04, &#39;g_elbo&#39;: 0.233, &#39;x_rna_nll&#39;: 0.243, &#39;x_rna_kl&#39;: 0.006, &#39;x_rna_elbo&#39;: 0.249, &#39;x_adt_nll&#39;: 1.862, &#39;x_adt_kl&#39;: 0.199, &#39;x_adt_elbo&#39;: 2.061, &#39;dsc_loss&#39;: 0.63, &#39;vae_loss&#39;: 2.32, &#39;gen_loss&#39;: 2.288}, 4.8s elapsed
[INFO] SCGLUETrainer: [Epoch 20] train={&#39;g_nll&#39;: 0.196, &#39;g_kl&#39;: 0.041, &#39;g_elbo&#39;: 0.237, &#39;x_rna_nll&#39;: 0.237, &#39;x_rna_kl&#39;: 0.006, &#39;x_rna_elbo&#39;: 0.242, &#39;x_adt_nll&#39;: 1.845, &#39;x_adt_kl&#39;: 0.175, &#39;x_adt_elbo&#39;: 2.02, &#39;dsc_loss&#39;: 0.674, &#39;vae_loss&#39;: 2.272, &#39;gen_loss&#39;: 2.238}, val={&#39;g_nll&#39;: 0.194, &#39;g_kl&#39;: 0.041, &#39;g_elbo&#39;: 0.235, &#39;x_rna_nll&#39;: 0.236, &#39;x_rna_kl&#39;: 0.005, &#39;x_rna_elbo&#39;: 0.242, &#39;x_adt_nll&#39;: 1.798, &#39;x_adt_kl&#39;: 0.173, &#39;x_adt_elbo&#39;: 1.97, &#39;dsc_loss&#39;: 0.67, &#39;vae_loss&#39;: 2.222, &#39;gen_loss&#39;: 2.188}, 5.9s elapsed
[INFO] SCGLUETrainer: [Epoch 30] train={&#39;g_nll&#39;: 0.182, &#39;g_kl&#39;: 0.041, &#39;g_elbo&#39;: 0.224, &#39;x_rna_nll&#39;: 0.236, &#39;x_rna_kl&#39;: 0.005, &#39;x_rna_elbo&#39;: 0.241, &#39;x_adt_nll&#39;: 1.835, &#39;x_adt_kl&#39;: 0.172, &#39;x_adt_elbo&#39;: 2.007, &#39;dsc_loss&#39;: 0.676, &#39;vae_loss&#39;: 2.257, &#39;gen_loss&#39;: 2.223}, val={&#39;g_nll&#39;: 0.155, &#39;g_kl&#39;: 0.041, &#39;g_elbo&#39;: 0.196, &#39;x_rna_nll&#39;: 0.237, &#39;x_rna_kl&#39;: 0.005, &#39;x_rna_elbo&#39;: 0.242, &#39;x_adt_nll&#39;: 1.782, &#39;x_adt_kl&#39;: 0.166, &#39;x_adt_elbo&#39;: 1.948, &#39;dsc_loss&#39;: 0.669, &#39;vae_loss&#39;: 2.198, &#39;gen_loss&#39;: 2.165}, 12.0s elapsed
[INFO] SCGLUETrainer: [Epoch 40] train={&#39;g_nll&#39;: 0.147, &#39;g_kl&#39;: 0.042, &#39;g_elbo&#39;: 0.189, &#39;x_rna_nll&#39;: 0.236, &#39;x_rna_kl&#39;: 0.005, &#39;x_rna_elbo&#39;: 0.241, &#39;x_adt_nll&#39;: 1.833, &#39;x_adt_kl&#39;: 0.169, &#39;x_adt_elbo&#39;: 2.002, &#39;dsc_loss&#39;: 0.677, &#39;vae_loss&#39;: 2.251, &#39;gen_loss&#39;: 2.217}, val={&#39;g_nll&#39;: 0.159, &#39;g_kl&#39;: 0.042, &#39;g_elbo&#39;: 0.2, &#39;x_rna_nll&#39;: 0.238, &#39;x_rna_kl&#39;: 0.005, &#39;x_rna_elbo&#39;: 0.243, &#39;x_adt_nll&#39;: 1.781, &#39;x_adt_kl&#39;: 0.166, &#39;x_adt_elbo&#39;: 1.947, &#39;dsc_loss&#39;: 0.674, &#39;vae_loss&#39;: 2.199, &#39;gen_loss&#39;: 2.165}, 12.1s elapsed
Epoch 00044: reducing learning rate of group 0 to 2.0000e-04.
Epoch 00044: reducing learning rate of group 0 to 2.0000e-04.
[INFO] LRScheduler: Learning rate reduction: step 1
[INFO] SCGLUETrainer: [Epoch 50] train={&#39;g_nll&#39;: 0.146, &#39;g_kl&#39;: 0.042, &#39;g_elbo&#39;: 0.188, &#39;x_rna_nll&#39;: 0.235, &#39;x_rna_kl&#39;: 0.005, &#39;x_rna_elbo&#39;: 0.24, &#39;x_adt_nll&#39;: 1.824, &#39;x_adt_kl&#39;: 0.17, &#39;x_adt_elbo&#39;: 1.993, &#39;dsc_loss&#39;: 0.676, &#39;vae_loss&#39;: 2.241, &#39;gen_loss&#39;: 2.207}, val={&#39;g_nll&#39;: 0.142, &#39;g_kl&#39;: 0.042, &#39;g_elbo&#39;: 0.183, &#39;x_rna_nll&#39;: 0.235, &#39;x_rna_kl&#39;: 0.005, &#39;x_rna_elbo&#39;: 0.241, &#39;x_adt_nll&#39;: 1.771, &#39;x_adt_kl&#39;: 0.166, &#39;x_adt_elbo&#39;: 1.938, &#39;dsc_loss&#39;: 0.669, &#39;vae_loss&#39;: 2.185, &#39;gen_loss&#39;: 2.152}, 14.4s elapsed
Epoch 00058: reducing learning rate of group 0 to 2.0000e-05.
Epoch 00058: reducing learning rate of group 0 to 2.0000e-05.
[INFO] LRScheduler: Learning rate reduction: step 2
[INFO] SCGLUETrainer: [Epoch 60] train={&#39;g_nll&#39;: 0.147, &#39;g_kl&#39;: 0.042, &#39;g_elbo&#39;: 0.188, &#39;x_rna_nll&#39;: 0.235, &#39;x_rna_kl&#39;: 0.005, &#39;x_rna_elbo&#39;: 0.24, &#39;x_adt_nll&#39;: 1.818, &#39;x_adt_kl&#39;: 0.17, &#39;x_adt_elbo&#39;: 1.989, &#39;dsc_loss&#39;: 0.677, &#39;vae_loss&#39;: 2.237, &#39;gen_loss&#39;: 2.203}, val={&#39;g_nll&#39;: 0.159, &#39;g_kl&#39;: 0.042, &#39;g_elbo&#39;: 0.201, &#39;x_rna_nll&#39;: 0.239, &#39;x_rna_kl&#39;: 0.005, &#39;x_rna_elbo&#39;: 0.244, &#39;x_adt_nll&#39;: 1.77, &#39;x_adt_kl&#39;: 0.167, &#39;x_adt_elbo&#39;: 1.936, &#39;dsc_loss&#39;: 0.668, &#39;vae_loss&#39;: 2.188, &#39;gen_loss&#39;: 2.155}, 6.4s elapsed
</pre></div>
</div>
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>2023-01-04 16:22:27,927 ignite.handlers.early_stopping.EarlyStopping INFO: EarlyStopping: Stop training
</pre></div>
</div>
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>[INFO] EarlyStopping: Restoring checkpoint &quot;59&quot;...
[INFO] EarlyStopping: Restoring checkpoint &quot;59&quot;...
[INFO] fit_SCGLUE: Estimating balancing weight...
[INFO] estimate_balancing_weight: Clustering cells...
[INFO] estimate_balancing_weight: Matching clusters...
[INFO] estimate_balancing_weight: Matching array shape = (24, 25)...
[INFO] estimate_balancing_weight: Estimating balancing weight...
[INFO] fit_SCGLUE: Fine-tuning SCGLUE model...
[INFO] check_graph: Checking variable coverage...
[INFO] check_graph: Checking edge attributes...
[INFO] check_graph: Checking self-loops...
[INFO] check_graph: Checking graph symmetry...
[INFO] check_graph: All checks passed!
[INFO] SCGLUEModel: Setting `graph_batch_size` = 1457
[INFO] SCGLUEModel: Setting `align_burnin` = 17
[INFO] SCGLUEModel: Setting `max_epochs` = 102
[INFO] SCGLUEModel: Setting `patience` = 9
[INFO] SCGLUEModel: Setting `reduce_lr_patience` = 5
[INFO] SCGLUETrainer: Using training directory: &quot;/tmp/GLUETMPe8ok569r&quot;
[INFO] SCGLUETrainer: [Epoch 10] train={&#39;g_nll&#39;: 0.136, &#39;g_kl&#39;: 0.042, &#39;g_elbo&#39;: 0.178, &#39;x_rna_nll&#39;: 0.237, &#39;x_rna_kl&#39;: 0.005, &#39;x_rna_elbo&#39;: 0.242, &#39;x_adt_nll&#39;: 1.831, &#39;x_adt_kl&#39;: 0.168, &#39;x_adt_elbo&#39;: 2.0, &#39;dsc_loss&#39;: 0.662, &#39;vae_loss&#39;: 2.249, &#39;gen_loss&#39;: 2.216}, val={&#39;g_nll&#39;: 0.131, &#39;g_kl&#39;: 0.042, &#39;g_elbo&#39;: 0.173, &#39;x_rna_nll&#39;: 0.237, &#39;x_rna_kl&#39;: 0.005, &#39;x_rna_elbo&#39;: 0.243, &#39;x_adt_nll&#39;: 1.787, &#39;x_adt_kl&#39;: 0.166, &#39;x_adt_elbo&#39;: 1.952, &#39;dsc_loss&#39;: 0.659, &#39;vae_loss&#39;: 2.202, &#39;gen_loss&#39;: 2.169}, 8.9s elapsed
[INFO] SCGLUETrainer: [Epoch 20] train={&#39;g_nll&#39;: 0.126, &#39;g_kl&#39;: 0.042, &#39;g_elbo&#39;: 0.167, &#39;x_rna_nll&#39;: 0.238, &#39;x_rna_kl&#39;: 0.005, &#39;x_rna_elbo&#39;: 0.244, &#39;x_adt_nll&#39;: 1.829, &#39;x_adt_kl&#39;: 0.168, &#39;x_adt_elbo&#39;: 1.997, &#39;dsc_loss&#39;: 0.641, &#39;vae_loss&#39;: 2.247, &#39;gen_loss&#39;: 2.215}, val={&#39;g_nll&#39;: 0.135, &#39;g_kl&#39;: 0.042, &#39;g_elbo&#39;: 0.177, &#39;x_rna_nll&#39;: 0.239, &#39;x_rna_kl&#39;: 0.005, &#39;x_rna_elbo&#39;: 0.244, &#39;x_adt_nll&#39;: 1.786, &#39;x_adt_kl&#39;: 0.164, &#39;x_adt_elbo&#39;: 1.95, &#39;dsc_loss&#39;: 0.635, &#39;vae_loss&#39;: 2.201, &#39;gen_loss&#39;: 2.169}, 5.9s elapsed
[INFO] SCGLUETrainer: [Epoch 30] train={&#39;g_nll&#39;: 0.126, &#39;g_kl&#39;: 0.041, &#39;g_elbo&#39;: 0.168, &#39;x_rna_nll&#39;: 0.238, &#39;x_rna_kl&#39;: 0.005, &#39;x_rna_elbo&#39;: 0.243, &#39;x_adt_nll&#39;: 1.828, &#39;x_adt_kl&#39;: 0.168, &#39;x_adt_elbo&#39;: 1.995, &#39;dsc_loss&#39;: 0.647, &#39;vae_loss&#39;: 2.245, &#39;gen_loss&#39;: 2.213}, val={&#39;g_nll&#39;: 0.129, &#39;g_kl&#39;: 0.042, &#39;g_elbo&#39;: 0.171, &#39;x_rna_nll&#39;: 0.239, &#39;x_rna_kl&#39;: 0.005, &#39;x_rna_elbo&#39;: 0.244, &#39;x_adt_nll&#39;: 1.779, &#39;x_adt_kl&#39;: 0.164, &#39;x_adt_elbo&#39;: 1.943, &#39;dsc_loss&#39;: 0.634, &#39;vae_loss&#39;: 2.194, &#39;gen_loss&#39;: 2.163}, 4.4s elapsed
Epoch 00036: reducing learning rate of group 0 to 2.0000e-04.
Epoch 00036: reducing learning rate of group 0 to 2.0000e-04.
[INFO] LRScheduler: Learning rate reduction: step 1
[INFO] SCGLUETrainer: [Epoch 40] train={&#39;g_nll&#39;: 0.114, &#39;g_kl&#39;: 0.041, &#39;g_elbo&#39;: 0.155, &#39;x_rna_nll&#39;: 0.238, &#39;x_rna_kl&#39;: 0.005, &#39;x_rna_elbo&#39;: 0.243, &#39;x_adt_nll&#39;: 1.815, &#39;x_adt_kl&#39;: 0.168, &#39;x_adt_elbo&#39;: 1.982, &#39;dsc_loss&#39;: 0.667, &#39;vae_loss&#39;: 2.232, &#39;gen_loss&#39;: 2.198}, val={&#39;g_nll&#39;: 0.132, &#39;g_kl&#39;: 0.041, &#39;g_elbo&#39;: 0.173, &#39;x_rna_nll&#39;: 0.239, &#39;x_rna_kl&#39;: 0.005, &#39;x_rna_elbo&#39;: 0.244, &#39;x_adt_nll&#39;: 1.777, &#39;x_adt_kl&#39;: 0.165, &#39;x_adt_elbo&#39;: 1.942, &#39;dsc_loss&#39;: 0.632, &#39;vae_loss&#39;: 2.192, &#39;gen_loss&#39;: 2.161}, 10.3s elapsed
[INFO] SCGLUETrainer: [Epoch 50] train={&#39;g_nll&#39;: 0.117, &#39;g_kl&#39;: 0.041, &#39;g_elbo&#39;: 0.158, &#39;x_rna_nll&#39;: 0.238, &#39;x_rna_kl&#39;: 0.005, &#39;x_rna_elbo&#39;: 0.243, &#39;x_adt_nll&#39;: 1.814, &#39;x_adt_kl&#39;: 0.169, &#39;x_adt_elbo&#39;: 1.983, &#39;dsc_loss&#39;: 0.671, &#39;vae_loss&#39;: 2.232, &#39;gen_loss&#39;: 2.198}, val={&#39;g_nll&#39;: 0.115, &#39;g_kl&#39;: 0.041, &#39;g_elbo&#39;: 0.156, &#39;x_rna_nll&#39;: 0.24, &#39;x_rna_kl&#39;: 0.005, &#39;x_rna_elbo&#39;: 0.245, &#39;x_adt_nll&#39;: 1.77, &#39;x_adt_kl&#39;: 0.166, &#39;x_adt_elbo&#39;: 1.936, &#39;dsc_loss&#39;: 0.632, &#39;vae_loss&#39;: 2.188, &#39;gen_loss&#39;: 2.156}, 9.2s elapsed
Epoch 00050: reducing learning rate of group 0 to 2.0000e-05.
Epoch 00050: reducing learning rate of group 0 to 2.0000e-05.
[INFO] LRScheduler: Learning rate reduction: step 2
Epoch 00058: reducing learning rate of group 0 to 2.0000e-06.
Epoch 00058: reducing learning rate of group 0 to 2.0000e-06.
[INFO] LRScheduler: Learning rate reduction: step 3
[INFO] SCGLUETrainer: [Epoch 60] train={&#39;g_nll&#39;: 0.118, &#39;g_kl&#39;: 0.041, &#39;g_elbo&#39;: 0.159, &#39;x_rna_nll&#39;: 0.237, &#39;x_rna_kl&#39;: 0.005, &#39;x_rna_elbo&#39;: 0.242, &#39;x_adt_nll&#39;: 1.815, &#39;x_adt_kl&#39;: 0.169, &#39;x_adt_elbo&#39;: 1.983, &#39;dsc_loss&#39;: 0.668, &#39;vae_loss&#39;: 2.232, &#39;gen_loss&#39;: 2.199}, val={&#39;g_nll&#39;: 0.113, &#39;g_kl&#39;: 0.041, &#39;g_elbo&#39;: 0.154, &#39;x_rna_nll&#39;: 0.239, &#39;x_rna_kl&#39;: 0.005, &#39;x_rna_elbo&#39;: 0.244, &#39;x_adt_nll&#39;: 1.771, &#39;x_adt_kl&#39;: 0.166, &#39;x_adt_elbo&#39;: 1.937, &#39;dsc_loss&#39;: 0.643, &#39;vae_loss&#39;: 2.187, &#39;gen_loss&#39;: 2.155}, 14.5s elapsed
Epoch 00065: reducing learning rate of group 0 to 2.0000e-07.
Epoch 00065: reducing learning rate of group 0 to 2.0000e-07.
[INFO] LRScheduler: Learning rate reduction: step 4
</pre></div>
</div>
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>2023-01-04 16:32:13,598 ignite.handlers.early_stopping.EarlyStopping INFO: EarlyStopping: Stop training
</pre></div>
</div>
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>[INFO] EarlyStopping: Restoring checkpoint &quot;59&quot;...
[INFO] EarlyStopping: Restoring checkpoint &quot;59&quot;...
</pre></div>
</div>
</div>
</div>
<p>Now we can obtain the latent representation for both modalities and concatenate them into one AnnData object for later visualization.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">rna_multiome</span><span class="o">.</span><span class="n">obsm</span><span class="p">[</span><span class="s2">&quot;X_glue&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">glue</span><span class="o">.</span><span class="n">encode_data</span><span class="p">(</span><span class="s2">&quot;rna&quot;</span><span class="p">,</span> <span class="n">rna_multiome</span><span class="p">)</span>
<span class="n">adt_cite</span><span class="o">.</span><span class="n">obsm</span><span class="p">[</span><span class="s2">&quot;X_glue&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">glue</span><span class="o">.</span><span class="n">encode_data</span><span class="p">(</span><span class="s2">&quot;adt&quot;</span><span class="p">,</span> <span class="n">adt_cite</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">adt_cite</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s2">&quot;modality&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="s2">&quot;ADT&quot;</span>
<span class="n">rna_multiome</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s2">&quot;modality&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="s2">&quot;RNA&quot;</span>

<span class="n">combined</span> <span class="o">=</span> <span class="n">anndata</span><span class="o">.</span><span class="n">concat</span><span class="p">([</span><span class="n">rna_multiome</span><span class="p">,</span> <span class="n">adt_cite</span><span class="p">])</span>
</pre></div>
</div>
</div>
</div>
<p>Finally, we visualize the integrated latent space.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">sc</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">neighbors</span><span class="p">(</span><span class="n">combined</span><span class="p">,</span> <span class="n">use_rep</span><span class="o">=</span><span class="s2">&quot;X_glue&quot;</span><span class="p">,</span> <span class="n">metric</span><span class="o">=</span><span class="s2">&quot;cosine&quot;</span><span class="p">)</span>
<span class="n">sc</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">umap</span><span class="p">(</span><span class="n">combined</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">sc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">umap</span><span class="p">(</span>
    <span class="n">combined</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;cell_type&quot;</span><span class="p">,</span> <span class="s2">&quot;modality&quot;</span><span class="p">,</span> <span class="s2">&quot;Samplename&quot;</span><span class="p">],</span> <span class="n">ncols</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span> <span class="n">frameon</span><span class="o">=</span><span class="kc">False</span>
<span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="../_images/f7c135a0267b13a1076695c909c2a63b6586ec1f1bc7f66818dbf55565cf22c7.png" src="../_images/f7c135a0267b13a1076695c909c2a63b6586ec1f1bc7f66818dbf55565cf22c7.png" />
</div>
</div>
<p>We see that the batches were well integrated as well as the two modalities. Yet again, scIB could be used to evaluate the integration.</p>
<p><a id='totalvi'></a></p>
</section>
</section>
<section id="integration-with-partially-overlapping-data-and-query-to-reference-mapping-with-totalvi">
<span id="multimodal-integration-advanced-integration-key-takeaway-2"></span><h2><span class="section-number">46.5. </span>Integration with partially overlapping data and query-to-reference mapping with totalVI<a class="headerlink" href="#integration-with-partially-overlapping-data-and-query-to-reference-mapping-with-totalvi" title="Link to this heading">#</a></h2>
<p>Another multimodal data integration scenario is when we have paired and unpaired data available at the same time, for instance, a CITE-seq and an RNA-seq datasets. This scenario is also known as mosaic integration. We might be interested in mapping the datasets into a shared latent space to make use of all the data available to us. Another use case would be to impute missing modalities (i.e. proteins abundance for RNA-seq dataset). We show how to perform these two tasks, partially overlapping integration and imputation, using totalVI <span id="id3">[<a class="reference internal" href="paired_integration.html#id544" title="Adam Gayoso, Zoë Steier, Romain Lopez, Jeffrey Regier, Kristopher L. Nazor, Aaron Streets, and Nir Yosef. Joint probabilistic modeling of single-cell multi-omic data with totalvi. Nature Methods, 18(3):272-282, Mar 2021. URL: https://doi.org/10.1038/s41592-020-01050-x, doi:10.1038/s41592-020-01050-x.">Gayoso <em>et al.</em>, 2021</a>]</span>.</p>
<p>We set the protein counts of one of the batches to zeros to model the missing proteins. This helps later with mapping of the new RNA-only queries.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">adata</span> <span class="o">=</span> <span class="n">rna_cite</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
<span class="n">adata</span><span class="o">.</span><span class="n">obsm</span><span class="p">[</span><span class="s2">&quot;protein_counts&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">adt_cite</span><span class="o">.</span><span class="n">layers</span><span class="p">[</span><span class="s2">&quot;counts&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">A</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
<span class="n">adata</span><span class="o">.</span><span class="n">obsm</span><span class="p">[</span><span class="s2">&quot;protein_counts&quot;</span><span class="p">][</span><span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s2">&quot;batch&quot;</span><span class="p">]</span> <span class="o">==</span> <span class="s2">&quot;s1d3&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="mf">0.0</span>
<span class="n">adata</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>AnnData object with n_obs × n_vars = 16294 × 4000
    obs: &#39;GEX_n_genes_by_counts&#39;, &#39;GEX_pct_counts_mt&#39;, &#39;GEX_size_factors&#39;, &#39;GEX_phase&#39;, &#39;ADT_n_antibodies_by_counts&#39;, &#39;ADT_total_counts&#39;, &#39;ADT_iso_count&#39;, &#39;cell_type&#39;, &#39;batch&#39;, &#39;ADT_pseudotime_order&#39;, &#39;GEX_pseudotime_order&#39;, &#39;Samplename&#39;, &#39;Site&#39;, &#39;DonorNumber&#39;, &#39;Modality&#39;, &#39;VendorLot&#39;, &#39;DonorID&#39;, &#39;DonorAge&#39;, &#39;DonorBMI&#39;, &#39;DonorBloodType&#39;, &#39;DonorRace&#39;, &#39;Ethnicity&#39;, &#39;DonorGender&#39;, &#39;QCMeds&#39;, &#39;DonorSmoker&#39;, &#39;is_train&#39;, &#39;new_batch&#39;
    var: &#39;feature_types&#39;, &#39;gene_id&#39;, &#39;highly_variable&#39;, &#39;means&#39;, &#39;dispersions&#39;, &#39;dispersions_norm&#39;
    uns: &#39;dataset_id&#39;, &#39;genome&#39;, &#39;hvg&#39;, &#39;organism&#39;
    obsm: &#39;ADT_X_pca&#39;, &#39;ADT_X_umap&#39;, &#39;ADT_isotype_controls&#39;, &#39;GEX_X_pca&#39;, &#39;GEX_X_umap&#39;, &#39;protein_counts&#39;
    layers: &#39;counts&#39;
</pre></div>
</div>
</div>
</div>
<p>Next, we setup the AnnData, specify the model parameters and train the model.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">TOTALVI</span><span class="o">.</span><span class="n">setup_anndata</span><span class="p">(</span>
    <span class="n">adata</span><span class="p">,</span>
    <span class="n">layer</span><span class="o">=</span><span class="s2">&quot;counts&quot;</span><span class="p">,</span>
    <span class="n">batch_key</span><span class="o">=</span><span class="s2">&quot;batch&quot;</span><span class="p">,</span>
    <span class="n">protein_expression_obsm_key</span><span class="o">=</span><span class="s2">&quot;protein_counts&quot;</span><span class="p">,</span>
<span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>No GPU/TPU found, falling back to CPU. (Set TF_CPP_MIN_LOG_LEVEL=0 and rerun for more info.)
</pre></div>
</div>
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span><span class=" -Color -Color-Blue">INFO    </span> Generating sequential column names                                                                        
<span class=" -Color -Color-Blue">INFO    </span> Found batches with missing protein expression                                                             
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">arches_params</span> <span class="o">=</span> <span class="p">{</span>
    <span class="s2">&quot;use_layer_norm&quot;</span><span class="p">:</span> <span class="s2">&quot;both&quot;</span><span class="p">,</span>
    <span class="s2">&quot;use_batch_norm&quot;</span><span class="p">:</span> <span class="s2">&quot;none&quot;</span><span class="p">,</span>
    <span class="s2">&quot;n_layers_decoder&quot;</span><span class="p">:</span> <span class="mi">2</span><span class="p">,</span>
    <span class="s2">&quot;n_layers_encoder&quot;</span><span class="p">:</span> <span class="mi">2</span><span class="p">,</span>
<span class="p">}</span>

<span class="n">vae</span> <span class="o">=</span> <span class="n">TOTALVI</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="o">**</span><span class="n">arches_params</span><span class="p">)</span>
<span class="n">vae</span><span class="o">.</span><span class="n">train</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span><span class=" -Color -Color-Blue">INFO    </span> Computing empirical prior initialization for protein background.                                          
</pre></div>
</div>
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>GPU available: True (cuda), used: True
TPU available: False, using: 0 TPU cores
IPU available: False, using: 0 IPUs
HPU available: False, using: 0 HPUs
LOCAL_RANK: 0 - CUDA_VISIBLE_DEVICES: [0]
</pre></div>
</div>
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Epoch 258/400:  64%|██████▍   | 258/400 [06:23&lt;03:23,  1.43s/it, loss=709, v_num=1]Epoch 00258: reducing learning rate of group 0 to 2.4000e-03.
Epoch 381/400:  95%|█████████▌| 381/400 [09:23&lt;00:27,  1.44s/it, loss=703, v_num=1]Epoch 00381: reducing learning rate of group 0 to 1.4400e-03.
Epoch 400/400: 100%|██████████| 400/400 [09:51&lt;00:00,  1.44s/it, loss=710, v_num=1]
</pre></div>
</div>
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>`Trainer.fit` stopped: `max_epochs=400` reached.
</pre></div>
</div>
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Epoch 400/400: 100%|██████████| 400/400 [09:51&lt;00:00,  1.48s/it, loss=710, v_num=1]
</pre></div>
</div>
</div>
</div>
<p>Now we obtain the latent representation and visualize it.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">adata</span><span class="o">.</span><span class="n">obsm</span><span class="p">[</span><span class="s2">&quot;X_totalvi&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">vae</span><span class="o">.</span><span class="n">get_latent_representation</span><span class="p">()</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">sc</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">neighbors</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">use_rep</span><span class="o">=</span><span class="s2">&quot;X_totalvi&quot;</span><span class="p">)</span>
<span class="n">sc</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">umap</span><span class="p">(</span><span class="n">adata</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">sc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">umap</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;cell_type&quot;</span><span class="p">,</span> <span class="s2">&quot;batch&quot;</span><span class="p">],</span> <span class="n">ncols</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span> <span class="n">frameon</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="../_images/36ff0739f5053dcb41e037ebc08d66698c7cd8fb21da0b22ed0abd9f7c9771a2.png" src="../_images/36ff0739f5053dcb41e037ebc08d66698c7cd8fb21da0b22ed0abd9f7c9771a2.png" />
</div>
</div>
<p>We see that batches as well as paired/RNA-only data are well integrated.</p>
<section id="query-to-reference-mapping">
<h3><span class="section-number">46.5.1. </span>Query-to-reference mapping<a class="headerlink" href="#query-to-reference-mapping" title="Link to this heading">#</a></h3>
<p>Now we demonstrate how to map new unimodal (RNA-only) and multimodal query (CITE-seq) onto the above reference.</p>
<p>We mimic RNA-only query by setting the protein counts of one of the two batches to zero.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">query</span> <span class="o">=</span> <span class="n">rna_cite_query</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
<span class="n">query</span><span class="o">.</span><span class="n">obsm</span><span class="p">[</span><span class="s2">&quot;protein_counts&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">adt_cite_query</span><span class="o">.</span><span class="n">layers</span><span class="p">[</span><span class="s2">&quot;counts&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">A</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
<span class="n">query</span><span class="o">.</span><span class="n">obsm</span><span class="p">[</span><span class="s2">&quot;protein_counts&quot;</span><span class="p">][</span><span class="n">query</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s2">&quot;batch&quot;</span><span class="p">]</span> <span class="o">==</span> <span class="s2">&quot;s2d4&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="mf">0.0</span>
</pre></div>
</div>
</div>
</div>
<p>We create some additional columns in <code class="docutils literal notranslate"><span class="pre">.obs</span></code> to help with the visualization later.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s2">&quot;dataset_name&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="s2">&quot;Reference&quot;</span>
<span class="n">query</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s2">&quot;dataset_name&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="s2">&quot;Query&quot;</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s2">&quot;dataset_name_fine&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="s2">&quot;CITE reference&quot;</span>
<span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s2">&quot;dataset_name_fine&quot;</span><span class="p">][</span><span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s2">&quot;batch&quot;</span><span class="p">]</span> <span class="o">==</span> <span class="s2">&quot;s1d3&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="s2">&quot;RNA reference&quot;</span>
<span class="n">query</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s2">&quot;dataset_name_fine&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="s2">&quot;CITE query&quot;</span>
<span class="n">query</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s2">&quot;dataset_name_fine&quot;</span><span class="p">][</span><span class="n">query</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s2">&quot;batch&quot;</span><span class="p">]</span> <span class="o">==</span> <span class="s2">&quot;s2d4&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="s2">&quot;RNA query&quot;</span>
</pre></div>
</div>
</div>
</div>
<p>Next, we update and fine-tune the model.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">vae_q</span> <span class="o">=</span> <span class="n">TOTALVI</span><span class="o">.</span><span class="n">load_query_data</span><span class="p">(</span>
    <span class="n">query</span><span class="p">,</span>
    <span class="n">vae</span><span class="p">,</span>
<span class="p">)</span>
<span class="n">vae_q</span><span class="o">.</span><span class="n">train</span><span class="p">(</span>
    <span class="n">plan_kwargs</span><span class="o">=</span><span class="p">{</span><span class="s2">&quot;weight_decay&quot;</span><span class="p">:</span> <span class="mf">0.0</span><span class="p">,</span> <span class="s2">&quot;scale_adversarial_loss&quot;</span><span class="p">:</span> <span class="mf">0.0</span><span class="p">},</span>
<span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span><span class=" -Color -Color-Blue">INFO    </span> Found batches with missing protein expression                                                             
<span class=" -Color -Color-Blue">INFO    </span> Computing empirical prior initialization for protein background.                                          
</pre></div>
</div>
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>GPU available: True (cuda), used: True
TPU available: False, using: 0 TPU cores
IPU available: False, using: 0 IPUs
HPU available: False, using: 0 HPUs
LOCAL_RANK: 0 - CUDA_VISIBLE_DEVICES: [0]
</pre></div>
</div>
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Epoch 233/400:  58%|█████▊    | 233/400 [05:32&lt;03:54,  1.40s/it, loss=546, v_num=1]Epoch 00233: reducing learning rate of group 0 to 2.4000e-03.
Epoch 273/400:  68%|██████▊   | 273/400 [06:29&lt;02:58,  1.40s/it, loss=548, v_num=1]Epoch 00273: reducing learning rate of group 0 to 1.4400e-03.
Epoch 287/400:  72%|███████▏  | 287/400 [06:49&lt;02:41,  1.43s/it, loss=542, v_num=1]
Monitored metric elbo_validation did not improve in the last 45 records. Best score: 1102.225. Signaling Trainer to stop.
</pre></div>
</div>
</div>
</div>
<p>Finally, we obtain the latent representation for the query and visualize it together with the reference. For this we concatenate reference and the query and recalculate the neighors and the UMAP.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">query</span><span class="o">.</span><span class="n">obsm</span><span class="p">[</span><span class="s2">&quot;X_totalvi_scarches&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">vae_q</span><span class="o">.</span><span class="n">get_latent_representation</span><span class="p">(</span><span class="n">query</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">adata</span><span class="o">.</span><span class="n">obsm</span><span class="p">[</span><span class="s2">&quot;X_totalvi_scarches&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">adata</span><span class="o">.</span><span class="n">obsm</span><span class="p">[</span><span class="s2">&quot;X_totalvi&quot;</span><span class="p">]</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">full_data</span> <span class="o">=</span> <span class="n">adata</span><span class="o">.</span><span class="n">concatenate</span><span class="p">(</span><span class="n">query</span><span class="p">,</span> <span class="n">batch_key</span><span class="o">=</span><span class="s2">&quot;concat_batch&quot;</span><span class="p">)</span>
<span class="n">full_data</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>AnnData object with n_obs × n_vars = 32340 × 4000
    obs: &#39;GEX_n_genes_by_counts&#39;, &#39;GEX_pct_counts_mt&#39;, &#39;GEX_size_factors&#39;, &#39;GEX_phase&#39;, &#39;ADT_n_antibodies_by_counts&#39;, &#39;ADT_total_counts&#39;, &#39;ADT_iso_count&#39;, &#39;cell_type&#39;, &#39;batch&#39;, &#39;ADT_pseudotime_order&#39;, &#39;GEX_pseudotime_order&#39;, &#39;Samplename&#39;, &#39;Site&#39;, &#39;DonorNumber&#39;, &#39;Modality&#39;, &#39;VendorLot&#39;, &#39;DonorID&#39;, &#39;DonorAge&#39;, &#39;DonorBMI&#39;, &#39;DonorBloodType&#39;, &#39;DonorRace&#39;, &#39;Ethnicity&#39;, &#39;DonorGender&#39;, &#39;QCMeds&#39;, &#39;DonorSmoker&#39;, &#39;is_train&#39;, &#39;new_batch&#39;, &#39;_scvi_labels&#39;, &#39;_scvi_batch&#39;, &#39;dataset_name&#39;, &#39;dataset_name_fine&#39;, &#39;concat_batch&#39;
    var: &#39;feature_types&#39;, &#39;gene_id&#39;, &#39;highly_variable&#39;, &#39;means&#39;, &#39;dispersions&#39;, &#39;dispersions_norm&#39;
    obsm: &#39;ADT_X_pca&#39;, &#39;ADT_X_umap&#39;, &#39;ADT_isotype_controls&#39;, &#39;GEX_X_pca&#39;, &#39;GEX_X_umap&#39;, &#39;protein_counts&#39;, &#39;X_totalvi_scarches&#39;
    layers: &#39;counts&#39;
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">sc</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">neighbors</span><span class="p">(</span><span class="n">full_data</span><span class="p">,</span> <span class="n">use_rep</span><span class="o">=</span><span class="s2">&quot;X_totalvi_scarches&quot;</span><span class="p">)</span>
<span class="n">sc</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">umap</span><span class="p">(</span><span class="n">full_data</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">sc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">umap</span><span class="p">(</span>
    <span class="n">full_data</span><span class="p">,</span>
    <span class="n">color</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;cell_type&quot;</span><span class="p">,</span> <span class="s2">&quot;batch&quot;</span><span class="p">,</span> <span class="s2">&quot;dataset_name&quot;</span><span class="p">,</span> <span class="s2">&quot;dataset_name_fine&quot;</span><span class="p">],</span>
    <span class="n">ncols</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span>
    <span class="n">frameon</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
<span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="../_images/54c14a0afd4f383a27c2a8026bfc0f94f4e1ab081c3faddbfac9bb95c120ac22.png" src="../_images/54c14a0afd4f383a27c2a8026bfc0f94f4e1ab081c3faddbfac9bb95c120ac22.png" />
</div>
</div>
<p><a id='seurat'></a></p>
</section>
</section>
<section id="query-mapping-and-imputation-with-seurat-s-wnn">
<h2><span class="section-number">46.6. </span>Query mapping and imputation with Seurat’s WNN<a class="headerlink" href="#query-mapping-and-imputation-with-seurat-s-wnn" title="Link to this heading">#</a></h2>
<p>Seurat v4 allows mapping of new RNA-seq queries onto a multimodal reference integrated with weighted-nearest-neighbor (WNN) <span id="id4">[<a class="reference internal" href="paired_integration.html#id542" title="Yuhan Hao, Stephanie Hao, Erica Andersen-Nissen, William M. Mauck, Shiwei Zheng, Andrew Butler, Maddie J. Lee, Aaron J. Wilk, Charlotte Darby, Michael Zager, Paul Hoffman, Marlon Stoeckius, Efthymia Papalexi, Eleni P. Mimitou, Jaison Jain, Avi Srivastava, Tim Stuart, Lamar M. Fleming, Bertrand Yeung, Angela J. Rogers, Juliana M. McElrath, Catherine A. Blish, Raphael Gottardo, Peter Smibert, and Rahul Satija. Integrated analysis of multimodal single-cell data. Cell, 184(13):3573-3587.e29, 2021. URL: https://www.sciencedirect.com/science/article/pii/S0092867421005833, doi:https://doi.org/10.1016/j.cell.2021.04.048.">Hao <em>et al.</em>, 2021</a>]</span>. We use the reference we built in the paired integration notebook.</p>
<p>The mapping is based on finding anchors between the query and the reference. This approach also allows for missing information transfer from the reference to the query. Here we show how to predict cell types and how to impute missing modalities (protein abundance in this case).</p>
<p>We note that since Seurat v4 in an R library we will need to work with <code class="docutils literal notranslate"><span class="pre">anndata2ri</span></code> package (<a class="github reference external" href="https://github.com/theislab/anndata2ri">theislab/anndata2ri</a>) to move all of our data from Python to R.</p>
<p>First, let’s take a look at our reference and the query.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="o">%%</span><span class="k">R</span>
ref &lt;- readRDS(cite, file = &quot;wnn_ref.rds&quot;)
ref
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>An object of class Seurat 
4136 features across 16294 samples within 2 assays 
Active assay: RNA (4000 features, 4000 variable features)
 1 other assay present: ADT
 4 dimensional reductions calculated: pca, harmony_pca, spca, wnn.umap
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">rna_cite_query</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>AnnData object with n_obs × n_vars = 16046 × 4000
    obs: &#39;GEX_n_genes_by_counts&#39;, &#39;GEX_pct_counts_mt&#39;, &#39;GEX_size_factors&#39;, &#39;GEX_phase&#39;, &#39;ADT_n_antibodies_by_counts&#39;, &#39;ADT_total_counts&#39;, &#39;ADT_iso_count&#39;, &#39;cell_type&#39;, &#39;batch&#39;, &#39;ADT_pseudotime_order&#39;, &#39;GEX_pseudotime_order&#39;, &#39;Samplename&#39;, &#39;Site&#39;, &#39;DonorNumber&#39;, &#39;Modality&#39;, &#39;VendorLot&#39;, &#39;DonorID&#39;, &#39;DonorAge&#39;, &#39;DonorBMI&#39;, &#39;DonorBloodType&#39;, &#39;DonorRace&#39;, &#39;Ethnicity&#39;, &#39;DonorGender&#39;, &#39;QCMeds&#39;, &#39;DonorSmoker&#39;, &#39;is_train&#39;, &#39;new_batch&#39;
    var: &#39;feature_types&#39;, &#39;gene_id&#39;, &#39;highly_variable&#39;, &#39;means&#39;, &#39;dispersions&#39;, &#39;dispersions_norm&#39;
    uns: &#39;dataset_id&#39;, &#39;genome&#39;, &#39;hvg&#39;, &#39;organism&#39;
    obsm: &#39;ADT_X_pca&#39;, &#39;ADT_X_umap&#39;, &#39;ADT_isotype_controls&#39;, &#39;GEX_X_pca&#39;, &#39;GEX_X_umap&#39;
    layers: &#39;counts&#39;
</pre></div>
</div>
</div>
</div>
<p>We need to preprocess the query first, so we use CLR-normalized counts to calculate PCA.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">sc</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">pca</span><span class="p">(</span><span class="n">rna_cite_query</span><span class="p">)</span>
<span class="n">rna_cite_query</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>AnnData object with n_obs × n_vars = 16046 × 4000
    obs: &#39;GEX_n_genes_by_counts&#39;, &#39;GEX_pct_counts_mt&#39;, &#39;GEX_size_factors&#39;, &#39;GEX_phase&#39;, &#39;ADT_n_antibodies_by_counts&#39;, &#39;ADT_total_counts&#39;, &#39;ADT_iso_count&#39;, &#39;cell_type&#39;, &#39;batch&#39;, &#39;ADT_pseudotime_order&#39;, &#39;GEX_pseudotime_order&#39;, &#39;Samplename&#39;, &#39;Site&#39;, &#39;DonorNumber&#39;, &#39;Modality&#39;, &#39;VendorLot&#39;, &#39;DonorID&#39;, &#39;DonorAge&#39;, &#39;DonorBMI&#39;, &#39;DonorBloodType&#39;, &#39;DonorRace&#39;, &#39;Ethnicity&#39;, &#39;DonorGender&#39;, &#39;QCMeds&#39;, &#39;DonorSmoker&#39;, &#39;is_train&#39;, &#39;new_batch&#39;
    var: &#39;feature_types&#39;, &#39;gene_id&#39;, &#39;highly_variable&#39;, &#39;means&#39;, &#39;dispersions&#39;, &#39;dispersions_norm&#39;
    uns: &#39;dataset_id&#39;, &#39;genome&#39;, &#39;hvg&#39;, &#39;organism&#39;, &#39;pca&#39;
    obsm: &#39;ADT_X_pca&#39;, &#39;ADT_X_umap&#39;, &#39;ADT_isotype_controls&#39;, &#39;GEX_X_pca&#39;, &#39;GEX_X_umap&#39;, &#39;X_pca&#39;
    varm: &#39;PCs&#39;
    layers: &#39;counts&#39;
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">np</span><span class="o">.</span><span class="n">max</span><span class="p">(</span><span class="n">rna_cite_query</span><span class="o">.</span><span class="n">X</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>9.061238
</pre></div>
</div>
</div>
</div>
<p>Next, we move the query from Python to R.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">adata_</span> <span class="o">=</span> <span class="n">sc</span><span class="o">.</span><span class="n">AnnData</span><span class="p">(</span><span class="n">rna_cite_query</span><span class="o">.</span><span class="n">X</span><span class="o">.</span><span class="n">copy</span><span class="p">())</span>
<span class="n">adata_</span><span class="o">.</span><span class="n">obs_names</span> <span class="o">=</span> <span class="n">rna_cite_query</span><span class="o">.</span><span class="n">obs_names</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
<span class="n">adata_</span><span class="o">.</span><span class="n">var_names</span> <span class="o">=</span> <span class="n">rna_cite_query</span><span class="o">.</span><span class="n">var_names</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
<span class="n">adata_</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s2">&quot;cell_type&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">rna_cite_query</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s2">&quot;cell_type&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
<span class="n">adata_</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s2">&quot;batch&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">rna_cite_query</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s2">&quot;batch&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
<span class="n">adata_</span><span class="o">.</span><span class="n">obsm</span><span class="p">[</span><span class="s2">&quot;X_pca&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">rna_cite_query</span><span class="o">.</span><span class="n">obsm</span><span class="p">[</span><span class="s2">&quot;X_pca&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="o">%%</span><span class="k">R</span> -i adata_
query = as.Seurat(adata_, data=&#39;X&#39;, counts=NULL)
query
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>An object of class Seurat 
4000 features across 16046 samples within 1 assay 
Active assay: originalexp (4000 features, 0 variable features)
 1 dimensional reduction calculated: PCA
</pre></div>
</div>
</div>
</div>
<p>Now we need to find anchors between the reference and the query. We specify that we want to use SPCA dimensionality reduction from the reference.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="o">%%</span><span class="k">R</span>
anchors &lt;- FindTransferAnchors(
  reference = ref,
  query = query,
  reference.reduction = &quot;spca&quot;,
  dims = 1:20
)
</pre></div>
</div>
</div>
</div>
<p>Next, we map query onto the reference using the saved UMAP model to make sure that the UMAP projections stay the same for the reference and the query is mapped onto it. We additionally specify what information we want to transfer from the reference to the query with <code class="docutils literal notranslate"><span class="pre">refdata</span></code> parameter: in our case we want to predict the cell type and the protein counts.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="o">%%</span><span class="k">R</span>
query &lt;- MapQuery(
  anchorset = anchors,
  query = query,
  reference = ref,
  refdata = list(
    cell_type = &quot;cell_type&quot;,
    predicted_ADT = &quot;ADT&quot;
  ),
  reference.reduction = &quot;spca&quot;,
  reduction.model = &quot;wnn.umap&quot;,
  verbose=FALSE
)
</pre></div>
</div>
</div>
</div>
<p>Let’s visualize the reference and the query together and take a look at predicted vs true cell labels.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="o">%%</span><span class="k">R</span>
ref$id &lt;- &#39;reference&#39;
query$id &lt;- &#39;query&#39;
refquery &lt;-  merge(ref, query)
refquery[[&quot;umap&quot;]] &lt;- merge(ref[[&quot;wnn.umap&quot;]], query[[&quot;ref.umap&quot;]])
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="o">%%</span><span class="k">R</span>
p1 &lt;- DimPlot(refquery, reduction = &quot;umap&quot;, group.by = &quot;cell_type&quot;, label = TRUE, label.size = 3, repel = TRUE) + NoLegend()
p2 &lt;- DimPlot(refquery, reduction = &quot;umap&quot;, group.by = &quot;predicted.cell_type&quot;, label = TRUE, label.size = 3, repel = TRUE) + NoLegend()
p1 + p2
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Fontconfig warning: ignoring UTF-8: not a valid region tag
</pre></div>
</div>
<img alt="../_images/9f69a8cadfd460d4e6ad48dc730b6b3caddc21922dbf78b86f75d0516cf44d79.png" src="../_images/9f69a8cadfd460d4e6ad48dc730b6b3caddc21922dbf78b86f75d0516cf44d79.png" />
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="o">%%</span><span class="k">R</span>
DimPlot(refquery, reduction = &quot;umap&quot;, group.by = &quot;id&quot;, label = TRUE, label.size = 3, repel = TRUE) + NoLegend()
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="../_images/1b0a6ae40e2cf8b5c0689f4d7c928a362e0138b330759148c8f7a998f38c4a14.png" src="../_images/1b0a6ae40e2cf8b5c0689f4d7c928a362e0138b330759148c8f7a998f38c4a14.png" />
</div>
</div>
<p><a id='bridge'></a></p>
</section>
<section id="mapping-adt-onto-an-rna-atlas-with-bridge-integration">
<h2><span class="section-number">46.7. </span>Mapping ADT onto an RNA atlas with bridge integration<a class="headerlink" href="#mapping-adt-onto-an-rna-atlas-with-bridge-integration" title="Link to this heading">#</a></h2>
<p>Bridge <span id="id5">[<a class="reference internal" href="#id552" title="Yuhan Hao, Tim Stuart, Madeline Kowalski, Saket Choudhary, Paul Hoffman, Austin Hartman, Avi Srivastava, Gesmira Molla, Shaista Madad, Carlos Fernandez-Granda, and Rahul Satija. Dictionary learning for integrative, multimodal, and scalable single-cell analysis. bioRxiv, 2022. URL: https://www.biorxiv.org/content/early/2022/02/26/2022.02.24.481684, arXiv:https://www.biorxiv.org/content/early/2022/02/26/2022.02.24.481684.full.pdf, doi:10.1101/2022.02.24.481684.">Hao <em>et al.</em>, 2022</a>]</span> is a method for mapping new modalities onto an RNA-seq reference. To do so we need a so-called “bridge” dataset that contains paired data (RNA-seq and the new modality we want to map).</p>
<p>Since we want to showcase the bridging functionality, we assume that our reference consist of only one batch. We refer the reader to the integration vignette if there is a need to integrate batches in the reference (<a class="reference external" href="https://satijalab.org/seurat/articles/integration_introduction.html">https://satijalab.org/seurat/articles/integration_introduction.html</a>). As the bridge dataset we use one batch of the CITE-seq query, and as the actual ADT query we use ADT modality of the second batch in the CITE-seq query.</p>
<p>First, we subset the datasets to reference, bridge and query as described above.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">rna_cite_ref</span> <span class="o">=</span> <span class="n">rna_cite</span><span class="p">[</span><span class="n">rna_cite</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s2">&quot;batch&quot;</span><span class="p">]</span> <span class="o">==</span> <span class="s2">&quot;s1d1&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
<span class="n">rna_cite_bridge</span> <span class="o">=</span> <span class="n">rna_cite_query</span><span class="p">[</span><span class="n">rna_cite_query</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s2">&quot;batch&quot;</span><span class="p">]</span> <span class="o">==</span> <span class="s2">&quot;s2d1&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">adt_cite_bridge</span> <span class="o">=</span> <span class="n">adt_cite_query</span><span class="p">[</span><span class="n">adt_cite_query</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s2">&quot;donor&quot;</span><span class="p">]</span> <span class="o">==</span> <span class="s2">&quot;s2d1&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
<span class="n">adt_cite_query_bridge</span> <span class="o">=</span> <span class="n">adt_cite_query</span><span class="p">[</span><span class="n">adt_cite_query</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s2">&quot;donor&quot;</span><span class="p">]</span> <span class="o">==</span> <span class="s2">&quot;s2d4&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
</pre></div>
</div>
</div>
</div>
<p>Next, we prepare the reference. We move the object from Python to R, normalize the data with SCTransform and calculate PCA. We also calculate the UMAP (and keep the model) that we will later use to map query onto.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">adata_</span> <span class="o">=</span> <span class="n">sc</span><span class="o">.</span><span class="n">AnnData</span><span class="p">(</span><span class="n">rna_cite_ref</span><span class="o">.</span><span class="n">layers</span><span class="p">[</span><span class="s2">&quot;counts&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">A</span><span class="o">.</span><span class="n">copy</span><span class="p">())</span>
<span class="n">adata_</span><span class="o">.</span><span class="n">obs_names</span> <span class="o">=</span> <span class="n">rna_cite_ref</span><span class="o">.</span><span class="n">obs_names</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
<span class="n">adata_</span><span class="o">.</span><span class="n">var_names</span> <span class="o">=</span> <span class="n">rna_cite_ref</span><span class="o">.</span><span class="n">var_names</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
<span class="n">adata_</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s2">&quot;cell_type&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">rna_cite_ref</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s2">&quot;cell_type&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="o">%%</span><span class="k">R</span> -i adata_
ref = as.Seurat(adata_, data=NULL, counts=&#39;X&#39;)
ref
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>An object of class Seurat 
4000 features across 5219 samples within 1 assay 
Active assay: originalexp (4000 features, 0 variable features)
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="o">%%</span><span class="k">R</span>
ref &lt;- RenameAssays(object = ref, originalexp = &quot;RNA&quot;, verbose=FALSE) 
ref &lt;- SCTransform(ref, verbose=FALSE)
ref &lt;- RunPCA(ref, verbose=FALSE)
ref &lt;- RunUMAP(ref, dims = 1:50, return.model=TRUE, verbose=FALSE)
</pre></div>
</div>
</div>
</div>
<p>Next, we prepare the bridge dataset. We move both RNA and ADT objects to R, follow the same normalization procedure as above for RNA, perform CLR-normalization and calculate PCA for ADT counts.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">adata_</span> <span class="o">=</span> <span class="n">sc</span><span class="o">.</span><span class="n">AnnData</span><span class="p">(</span><span class="n">adt_cite_bridge</span><span class="o">.</span><span class="n">layers</span><span class="p">[</span><span class="s2">&quot;counts&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">A</span><span class="o">.</span><span class="n">copy</span><span class="p">())</span>
<span class="n">adata_</span><span class="o">.</span><span class="n">obs_names</span> <span class="o">=</span> <span class="n">adt_cite_bridge</span><span class="o">.</span><span class="n">obs_names</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
<span class="n">adata_</span><span class="o">.</span><span class="n">var_names</span> <span class="o">=</span> <span class="n">adt_cite_bridge</span><span class="o">.</span><span class="n">var_names</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
<span class="n">adata_</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s2">&quot;cell_type&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">adt_cite_bridge</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s2">&quot;cell_type&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">adata_</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>AnnData object with n_obs × n_vars = 10465 × 136
    obs: &#39;cell_type&#39;
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="o">%%</span><span class="k">R</span> -i adata_
adt &lt;- as.Seurat(adata_, data=NULL, counts=&#39;X&#39;)
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">adata_</span> <span class="o">=</span> <span class="n">sc</span><span class="o">.</span><span class="n">AnnData</span><span class="p">(</span><span class="n">rna_cite_bridge</span><span class="o">.</span><span class="n">layers</span><span class="p">[</span><span class="s2">&quot;counts&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">A</span><span class="o">.</span><span class="n">copy</span><span class="p">())</span>
<span class="n">adata_</span><span class="o">.</span><span class="n">obs_names</span> <span class="o">=</span> <span class="n">rna_cite_bridge</span><span class="o">.</span><span class="n">obs_names</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
<span class="n">adata_</span><span class="o">.</span><span class="n">var_names</span> <span class="o">=</span> <span class="n">rna_cite_bridge</span><span class="o">.</span><span class="n">var_names</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
<span class="n">adata_</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s2">&quot;cell_type&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">rna_cite_bridge</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s2">&quot;cell_type&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="o">%%</span><span class="k">R</span> -i adata_
rna = as.Seurat(adata_, data=NULL, counts=&#39;X&#39;)
cite &lt;- rna
cite &lt;- RenameAssays(object = cite, originalexp = &quot;RNA&quot;) 
cite[[&quot;ADT&quot;]] &lt;- CreateAssayObject(counts = adt@assays$originalexp@data)
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="o">%%</span><span class="k">R</span>
DefaultAssay(cite) &lt;- &quot;RNA&quot;
VariableFeatures(cite) &lt;- rownames(cite)
cite &lt;- SCTransform(cite, verbose = FALSE) 
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="o">%%</span><span class="k">R</span>
DefaultAssay(cite) &lt;- &quot;ADT&quot;
VariableFeatures(cite) &lt;- rownames(cite)
cite &lt;- NormalizeData(cite, normalization.method = &#39;CLR&#39;, margin = 2, verbose=FALSE)
cite &lt;- ScaleData(cite, verbose=FALSE)
</pre></div>
</div>
</div>
</div>
<p>Finally, we prepare the ADT query the same way as the bridge ADT above.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">adata_</span> <span class="o">=</span> <span class="n">sc</span><span class="o">.</span><span class="n">AnnData</span><span class="p">(</span><span class="n">adt_cite_query_bridge</span><span class="o">.</span><span class="n">layers</span><span class="p">[</span><span class="s2">&quot;counts&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">A</span><span class="o">.</span><span class="n">copy</span><span class="p">())</span>
<span class="n">adata_</span><span class="o">.</span><span class="n">obs_names</span> <span class="o">=</span> <span class="n">adt_cite_query_bridge</span><span class="o">.</span><span class="n">obs_names</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
<span class="n">adata_</span><span class="o">.</span><span class="n">var_names</span> <span class="o">=</span> <span class="n">adt_cite_query_bridge</span><span class="o">.</span><span class="n">var_names</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
<span class="n">adata_</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s2">&quot;cell_type&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">adt_cite_query_bridge</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s2">&quot;cell_type&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="o">%%</span><span class="k">R</span> -i adata_
query &lt;- as.Seurat(adata_, data=NULL, counts=&#39;X&#39;)
query &lt;- RenameAssays(object = query, originalexp = &quot;ADT&quot;, verbose=FALSE) 
VariableFeatures(query) &lt;- rownames(query)
query &lt;- NormalizeData(query, normalization.method = &#39;CLR&#39;, margin = 2, verbose=FALSE)
query &lt;- ScaleData(query, verbose=FALSE)
query &lt;- RunPCA(query, verbose=FALSE, reduction.name = &#39;apca&#39;)
</pre></div>
</div>
</div>
</div>
<p>Now we can perform the bridge integration. The final step of query mapping is the same as with Seurat v4 query mapping so we can also predict cell types for the query data.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="o">%%</span><span class="k">R</span>
dims.adt &lt;- 1:50
dims.rna &lt;- 1:50

DefaultAssay(cite) &lt;-  &quot;RNA&quot;
DefaultAssay(ref) &lt;- &quot;SCT&quot;
obj.rna.ext &lt;- PrepareBridgeReference(
    reference = ref,
    bridge = cite,
    bridge.query.assay = &quot;ADT&quot;,
    reference.reduction = &quot;pca&quot;,
    reference.dims = dims.rna,
    normalization.method = &quot;SCT&quot;,
    verbose = FALSE
)
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>  |                                                  | 0 % ~calculating   |++++++++++++++++++++++++++++++++++++++++++++++++++| 100% elapsed=00s  
  |                                                  | 0 % ~calculating   |++++++++++++++++++++++++++++++++++++++++++++++++++| 100% elapsed=00s  
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="o">%%</span><span class="k">R</span>
bridge.anchor &lt;- FindBridgeTransferAnchors(
    extended.reference = obj.rna.ext, 
    query = query,
    reduction = &quot;pcaproject&quot;,
    dims = dims.adt,
    verbose = FALSE
)
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>  |                                                  | 0 % ~calculating   |++++++++++++++++++++++++++++++++++++++++++++++++++| 100% elapsed=00s  
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="o">%%</span><span class="k">R</span>
query &lt;- MapQuery(
    anchorset = bridge.anchor, 
    reference = ref, 
    query = query, 
    refdata = list(
        cell_type = &quot;cell_type&quot;
    ),
    reduction.model = &quot;umap&quot;,
    verbose = FALSE
)
</pre></div>
</div>
</div>
</div>
<p>Let’s visualize the result.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="o">%%</span><span class="k">R</span>
p1 &lt;- DimPlot(query, group.by = &quot;predicted.cell_type&quot;, reduction = &quot;ref.umap&quot;, label = TRUE) + NoLegend()
p2 &lt;- DimPlot(query, group.by = &quot;cell_type&quot;, reduction = &quot;ref.umap&quot;, label = TRUE) + NoLegend()
p1 + p2
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="../_images/ee78feaf113f5bfc7f22bedddfa0d49b3fe434f6e980ee231a74ac1673edac26.png" src="../_images/ee78feaf113f5bfc7f22bedddfa0d49b3fe434f6e980ee231a74ac1673edac26.png" />
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="o">%%</span><span class="k">R</span>
ref$id &lt;- &#39;reference&#39;
query$id &lt;- &#39;query&#39;
refquery &lt;-  merge(ref, query)
refquery[[&quot;umap&quot;]] &lt;- merge(ref[[&quot;umap&quot;]], query[[&quot;ref.umap&quot;]])
DimPlot(refquery, group.by = &#39;id&#39;, shuffle = TRUE)
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="../_images/ade34f9d96a236e3894c7ab060511f9577ba1f02e47b70bd6bdc2ae77a067df1.png" src="../_images/ade34f9d96a236e3894c7ab060511f9577ba1f02e47b70bd6bdc2ae77a067df1.png" />
</div>
</div>
<p><a id='multigrate'></a></p>
</section>
<section id="trimodal-integration-and-query-to-reference-mapping-with-multigrate">
<h2><span class="section-number">46.8. </span>Trimodal integration and query-to-reference mapping with multigrate<a class="headerlink" href="#trimodal-integration-and-query-to-reference-mapping-with-multigrate" title="Link to this heading">#</a></h2>
<p>In our final example, we will build a trimodal reference atlas using RNA, ATAC and ADT as three modalities. Multigrate <span id="id6">[]</span> is a deep-learning method based on conditional autoencoders. Data from each modality is fed into a separate encoder which outputs the parameters of the marginal distribution, then the parameters of the joint distribution are calculated with product of experts (PoE) <span id="id7">[]</span>. The conditional decoders then learn to reconstruct the original input data while simultaneously correction for batch effects. PoE allows the integration of data with missing measurements for some modalities by setting the corresponding marginals to a non-informative distribution. Multigrate also follows scArches <span id="id8">[]</span> framework to allow query-to-reference mapping of unimodal as well as multimodal queries.</p>
<section id="building-the-reference">
<h3><span class="section-number">46.8.1. </span>Building the reference<a class="headerlink" href="#building-the-reference" title="Link to this heading">#</a></h3>
<p>First, we load data that has 4000 highly variable genes that are common for the scRNA-seq and scRNA-seq from the CITE-seq and the multiome experiment respectively.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">rna1</span> <span class="o">=</span> <span class="n">sc</span><span class="o">.</span><span class="n">read</span><span class="p">(</span>
    <span class="s2">&quot;/lustre/groups/ml01/workspace/anastasia.litinetskaya/data/trimodal_neurips/rna_hvg_cite.h5ad&quot;</span>
<span class="p">)</span>
<span class="n">rna1_ref</span> <span class="o">=</span> <span class="n">rna1</span><span class="p">[</span><span class="n">adt_cite</span><span class="o">.</span><span class="n">obs_names</span><span class="p">]</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
<span class="n">rna1_query</span> <span class="o">=</span> <span class="n">rna1</span><span class="p">[</span><span class="n">adt_cite_query</span><span class="o">.</span><span class="n">obs_names</span><span class="p">]</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
<span class="n">rna1_ref</span><span class="o">.</span><span class="n">shape</span><span class="p">,</span> <span class="n">rna1_query</span><span class="o">.</span><span class="n">shape</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>((16294, 4000), (16046, 4000))
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">rna2</span> <span class="o">=</span> <span class="n">sc</span><span class="o">.</span><span class="n">read</span><span class="p">(</span>
    <span class="s2">&quot;/lustre/groups/ml01/workspace/anastasia.litinetskaya/data/trimodal_neurips/rna_hvg_multiome.h5ad&quot;</span>
<span class="p">)</span>
<span class="n">rna2_ref</span> <span class="o">=</span> <span class="n">rna2</span><span class="p">[</span><span class="n">atac_multiome</span><span class="o">.</span><span class="n">obs_names</span><span class="p">]</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
<span class="n">rna2_query</span> <span class="o">=</span> <span class="n">rna2</span><span class="p">[</span><span class="n">atac_multiome_query</span><span class="o">.</span><span class="n">obs_names</span><span class="p">]</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
<span class="n">rna2_ref</span><span class="o">.</span><span class="n">shape</span><span class="p">,</span> <span class="n">rna2_query</span><span class="o">.</span><span class="n">shape</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>((17243, 4000), (10331, 4000))
</pre></div>
</div>
</div>
</div>
<p>Next, we concatenate all the data into one AnnData object specifying how the data is paired and which layers to use.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">adata</span> <span class="o">=</span> <span class="n">mtg</span><span class="o">.</span><span class="n">data</span><span class="o">.</span><span class="n">organize_multiome_anndatas</span><span class="p">(</span>
    <span class="n">adatas</span><span class="o">=</span><span class="p">[[</span><span class="n">rna1_ref</span><span class="p">,</span> <span class="n">rna2_ref</span><span class="p">],</span> <span class="p">[</span><span class="kc">None</span><span class="p">,</span> <span class="n">atac_multiome</span><span class="p">],</span> <span class="p">[</span><span class="n">adt_cite</span><span class="p">,</span> <span class="kc">None</span><span class="p">]],</span>
    <span class="n">layers</span><span class="o">=</span><span class="p">[[</span><span class="s2">&quot;counts&quot;</span><span class="p">,</span> <span class="s2">&quot;counts&quot;</span><span class="p">],</span> <span class="p">[</span><span class="kc">None</span><span class="p">,</span> <span class="s2">&quot;cpm&quot;</span><span class="p">],</span> <span class="p">[</span><span class="kc">None</span><span class="p">,</span> <span class="kc">None</span><span class="p">]],</span>
<span class="p">)</span>
<span class="n">adata</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>AnnData object with n_obs × n_vars = 33537 × 24136
    obs: &#39;GEX_n_genes_by_counts&#39;, &#39;GEX_pct_counts_mt&#39;, &#39;GEX_size_factors&#39;, &#39;GEX_phase&#39;, &#39;ADT_n_antibodies_by_counts&#39;, &#39;ADT_total_counts&#39;, &#39;ADT_iso_count&#39;, &#39;cell_type&#39;, &#39;batch&#39;, &#39;ADT_pseudotime_order&#39;, &#39;GEX_pseudotime_order&#39;, &#39;Samplename&#39;, &#39;Site&#39;, &#39;DonorNumber&#39;, &#39;Modality&#39;, &#39;VendorLot&#39;, &#39;DonorID&#39;, &#39;DonorAge&#39;, &#39;DonorBMI&#39;, &#39;DonorBloodType&#39;, &#39;DonorRace&#39;, &#39;Ethnicity&#39;, &#39;DonorGender&#39;, &#39;QCMeds&#39;, &#39;DonorSmoker&#39;, &#39;is_train&#39;, &#39;GEX_n_counts&#39;, &#39;GEX_n_genes&#39;, &#39;ATAC_nCount_peaks&#39;, &#39;ATAC_atac_fragments&#39;, &#39;ATAC_reads_in_peaks_frac&#39;, &#39;ATAC_blacklist_fraction&#39;, &#39;ATAC_nucleosome_signal&#39;, &#39;ATAC_pseudotime_order&#39;, &#39;technology&#39;, &#39;cell_type_l2&#39;, &#39;cell_type_l1&#39;, &#39;cell_type_l3&#39;, &#39;assay&#39;, &#39;split&#39;, &#39;group&#39;, &#39;balancing_weight&#39;, &#39;donor&#39;, &#39;log1p_n_genes_by_counts&#39;, &#39;log1p_total_counts&#39;, &#39;modality&#39;, &#39;n_counts&#39;, &#39;n_genes_by_counts&#39;, &#39;new_donor&#39;, &#39;outliers&#39;, &#39;total_counts&#39;, &#39;new_batch&#39;
    var: &#39;modality&#39;
    uns: &#39;modality_lengths&#39;
    layers: &#39;counts&#39;
</pre></div>
</div>
</div>
</div>
<p>Next, we register covariates that the model will correct for in the latent space.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">mtg</span><span class="o">.</span><span class="n">model</span><span class="o">.</span><span class="n">MultiVAE</span><span class="o">.</span><span class="n">setup_anndata</span><span class="p">(</span>
    <span class="n">adata</span><span class="p">,</span>
    <span class="n">categorical_covariate_keys</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;Modality&quot;</span><span class="p">,</span> <span class="s2">&quot;Samplename&quot;</span><span class="p">],</span>
    <span class="n">rna_indices_end</span><span class="o">=</span><span class="mi">4000</span><span class="p">,</span>
<span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<p>Next, we initialize the model and train it.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">model</span> <span class="o">=</span> <span class="n">mtg</span><span class="o">.</span><span class="n">model</span><span class="o">.</span><span class="n">MultiVAE</span><span class="p">(</span>
    <span class="n">adata</span><span class="p">,</span>
    <span class="n">losses</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;nb&quot;</span><span class="p">,</span> <span class="s2">&quot;mse&quot;</span><span class="p">,</span> <span class="s2">&quot;mse&quot;</span><span class="p">],</span>
    <span class="n">loss_coefs</span><span class="o">=</span><span class="p">{</span>
        <span class="s2">&quot;kl&quot;</span><span class="p">:</span> <span class="mf">1e-3</span><span class="p">,</span>
        <span class="s2">&quot;integ&quot;</span><span class="p">:</span> <span class="mi">10000</span><span class="p">,</span>
    <span class="p">},</span>
    <span class="n">integrate_on</span><span class="o">=</span><span class="s2">&quot;Modality&quot;</span><span class="p">,</span>
    <span class="n">mmd</span><span class="o">=</span><span class="s2">&quot;marginal&quot;</span><span class="p">,</span>
<span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">model</span><span class="o">.</span><span class="n">train</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>GPU available: True (cuda), used: True
TPU available: False, using: 0 TPU cores
IPU available: False, using: 0 IPUs
HPU available: False, using: 0 HPUs
LOCAL_RANK: 0 - CUDA_VISIBLE_DEVICES: [0]
</pre></div>
</div>
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Epoch 200/200: 100%|██████████| 200/200 [28:53&lt;00:00,  8.43s/it, loss=2.34e+03, v_num=1]
</pre></div>
</div>
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>`Trainer.fit` stopped: `max_epochs=200` reached.
</pre></div>
</div>
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Epoch 200/200: 100%|██████████| 200/200 [28:53&lt;00:00,  8.67s/it, loss=2.34e+03, v_num=1]
</pre></div>
</div>
</div>
</div>
<p>Finally, we obtain the latent representation, save in explicitly in <code class="docutils literal notranslate"><span class="pre">.obsm['latent_ref']</span></code> as we will overwrite <code class="docutils literal notranslate"><span class="pre">.obsm['latent']</span></code> later when we work with fine-tuned model after query mapping and visualize the result.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">model</span><span class="o">.</span><span class="n">get_latent_representation</span><span class="p">()</span>
<span class="n">adata</span><span class="o">.</span><span class="n">obsm</span><span class="p">[</span><span class="s2">&quot;latent_ref&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">adata</span><span class="o">.</span><span class="n">obsm</span><span class="p">[</span><span class="s2">&quot;latent&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
<span class="n">adata</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>AnnData object with n_obs × n_vars = 33537 × 24136
    obs: &#39;GEX_n_genes_by_counts&#39;, &#39;GEX_pct_counts_mt&#39;, &#39;GEX_size_factors&#39;, &#39;GEX_phase&#39;, &#39;ADT_n_antibodies_by_counts&#39;, &#39;ADT_total_counts&#39;, &#39;ADT_iso_count&#39;, &#39;cell_type&#39;, &#39;batch&#39;, &#39;ADT_pseudotime_order&#39;, &#39;GEX_pseudotime_order&#39;, &#39;Samplename&#39;, &#39;Site&#39;, &#39;DonorNumber&#39;, &#39;Modality&#39;, &#39;VendorLot&#39;, &#39;DonorID&#39;, &#39;DonorAge&#39;, &#39;DonorBMI&#39;, &#39;DonorBloodType&#39;, &#39;DonorRace&#39;, &#39;Ethnicity&#39;, &#39;DonorGender&#39;, &#39;QCMeds&#39;, &#39;DonorSmoker&#39;, &#39;is_train&#39;, &#39;GEX_n_counts&#39;, &#39;GEX_n_genes&#39;, &#39;ATAC_nCount_peaks&#39;, &#39;ATAC_atac_fragments&#39;, &#39;ATAC_reads_in_peaks_frac&#39;, &#39;ATAC_blacklist_fraction&#39;, &#39;ATAC_nucleosome_signal&#39;, &#39;ATAC_pseudotime_order&#39;, &#39;technology&#39;, &#39;cell_type_l2&#39;, &#39;cell_type_l1&#39;, &#39;cell_type_l3&#39;, &#39;assay&#39;, &#39;split&#39;, &#39;group&#39;, &#39;balancing_weight&#39;, &#39;donor&#39;, &#39;log1p_n_genes_by_counts&#39;, &#39;log1p_total_counts&#39;, &#39;modality&#39;, &#39;n_counts&#39;, &#39;n_genes_by_counts&#39;, &#39;new_donor&#39;, &#39;outliers&#39;, &#39;total_counts&#39;, &#39;new_batch&#39;, &#39;size_factors&#39;, &#39;_scvi_batch&#39;
    var: &#39;modality&#39;
    uns: &#39;modality_lengths&#39;, &#39;_scvi_uuid&#39;, &#39;_scvi_manager_uuid&#39;
    obsm: &#39;_scvi_extra_categorical_covs&#39;, &#39;latent&#39;, &#39;latent_ref&#39;
    layers: &#39;counts&#39;
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">sc</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">neighbors</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">use_rep</span><span class="o">=</span><span class="s2">&quot;latent&quot;</span><span class="p">)</span>
<span class="n">sc</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">umap</span><span class="p">(</span><span class="n">adata</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">sc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">umap</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;cell_type&quot;</span><span class="p">,</span> <span class="s2">&quot;Modality&quot;</span><span class="p">,</span> <span class="s2">&quot;Samplename&quot;</span><span class="p">],</span> <span class="n">ncols</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span> <span class="n">frameon</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="../_images/45af6ed67d447f4c32b57372c28293aee6c844ecdb234bd6ccdce1c625dd0e76.png" src="../_images/45af6ed67d447f4c32b57372c28293aee6c844ecdb234bd6ccdce1c625dd0e76.png" />
</div>
</div>
</section>
<section id="querying-the-trimodal-reference">
<h3><span class="section-number">46.8.2. </span>Querying the trimodal reference<a class="headerlink" href="#querying-the-trimodal-reference" title="Link to this heading">#</a></h3>
<p>We repeat the same steps for the setting up the query as for the reference before.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">query</span> <span class="o">=</span> <span class="n">mtg</span><span class="o">.</span><span class="n">data</span><span class="o">.</span><span class="n">organize_multiome_anndatas</span><span class="p">(</span>
    <span class="n">adatas</span><span class="o">=</span><span class="p">[</span>
        <span class="p">[</span><span class="n">rna1_query</span><span class="p">,</span> <span class="n">rna2_query</span><span class="p">],</span>
        <span class="p">[</span><span class="kc">None</span><span class="p">,</span> <span class="n">atac_multiome_query</span><span class="p">],</span>
        <span class="p">[</span><span class="n">adt_cite_query</span><span class="p">,</span> <span class="kc">None</span><span class="p">],</span>
    <span class="p">],</span>
    <span class="n">layers</span><span class="o">=</span><span class="p">[[</span><span class="s2">&quot;counts&quot;</span><span class="p">,</span> <span class="s2">&quot;counts&quot;</span><span class="p">],</span> <span class="p">[</span><span class="kc">None</span><span class="p">,</span> <span class="s2">&quot;cpm&quot;</span><span class="p">],</span> <span class="p">[</span><span class="kc">None</span><span class="p">,</span> <span class="kc">None</span><span class="p">]],</span>
<span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">mtg</span><span class="o">.</span><span class="n">model</span><span class="o">.</span><span class="n">MultiVAE</span><span class="o">.</span><span class="n">setup_anndata</span><span class="p">(</span>
    <span class="n">query</span><span class="p">,</span>
    <span class="n">categorical_covariate_keys</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;Modality&quot;</span><span class="p">,</span> <span class="s2">&quot;Samplename&quot;</span><span class="p">],</span>
    <span class="n">rna_indices_end</span><span class="o">=</span><span class="mi">4000</span><span class="p">,</span>
<span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<p>Let’s imitate unimodal queries by masking with zeros the snRNA part of one multiome batch and ADT part of one CITE-seq batch in the query. We get one ATAC-only query batch and one scRNA-only query batch respectively.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">idx_atac_query</span> <span class="o">=</span> <span class="n">query</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s2">&quot;Samplename&quot;</span><span class="p">]</span> <span class="o">==</span> <span class="s2">&quot;site2_donor4_multiome&quot;</span>
<span class="n">idx_scrna_query</span> <span class="o">=</span> <span class="n">query</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s2">&quot;Samplename&quot;</span><span class="p">]</span> <span class="o">==</span> <span class="s2">&quot;site2_donor1_cite&quot;</span>

<span class="n">idx_mutiome_query</span> <span class="o">=</span> <span class="n">query</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s2">&quot;Samplename&quot;</span><span class="p">]</span> <span class="o">==</span> <span class="s2">&quot;site2_donor1_multiome&quot;</span>
<span class="n">idx_cite_query</span> <span class="o">=</span> <span class="n">query</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s2">&quot;Samplename&quot;</span><span class="p">]</span> <span class="o">==</span> <span class="s2">&quot;site2_donor4_cite&quot;</span>

<span class="p">(</span>
    <span class="n">np</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">idx_atac_query</span><span class="p">),</span>
    <span class="n">np</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">idx_scrna_query</span><span class="p">),</span>
    <span class="n">np</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">idx_mutiome_query</span><span class="p">),</span>
    <span class="n">np</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">idx_cite_query</span><span class="p">),</span>
<span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>(6111, 10465, 4220, 5581)
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">query</span><span class="p">[</span><span class="n">idx_atac_query</span><span class="p">,</span> <span class="p">:</span><span class="mi">4000</span><span class="p">]</span><span class="o">.</span><span class="n">X</span> <span class="o">=</span> <span class="mi">0</span>
<span class="n">query</span><span class="p">[</span><span class="n">idx_scrna_query</span><span class="p">,</span> <span class="mi">4000</span><span class="p">:]</span><span class="o">.</span><span class="n">X</span> <span class="o">=</span> <span class="mi">0</span>
</pre></div>
</div>
</div>
</div>
<p>We update the model by adding new weights to the new batches in the query and fine-tune those weights.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">q_model</span> <span class="o">=</span> <span class="n">mtg</span><span class="o">.</span><span class="n">model</span><span class="o">.</span><span class="n">MultiVAE</span><span class="o">.</span><span class="n">load_query_data</span><span class="p">(</span><span class="n">query</span><span class="p">,</span> <span class="n">model</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">q_model</span><span class="o">.</span><span class="n">train</span><span class="p">(</span><span class="n">weight_decay</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>GPU available: True (cuda), used: True
TPU available: False, using: 0 TPU cores
IPU available: False, using: 0 IPUs
HPU available: False, using: 0 HPUs
LOCAL_RANK: 0 - CUDA_VISIBLE_DEVICES: [0]
</pre></div>
</div>
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Epoch 200/200: 100%|██████████| 200/200 [16:13&lt;00:00,  4.78s/it, loss=2.1e+03, v_num=1] 
</pre></div>
</div>
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>`Trainer.fit` stopped: `max_epochs=200` reached.
</pre></div>
</div>
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Epoch 200/200: 100%|██████████| 200/200 [16:13&lt;00:00,  4.87s/it, loss=2.1e+03, v_num=1]
</pre></div>
</div>
</div>
</div>
<p>We obtain the latent representation for query and the reference from the updated model. Note that the representation of the reference is the same as before up to some sampling noise.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">q_model</span><span class="o">.</span><span class="n">get_latent_representation</span><span class="p">(</span><span class="n">adata</span><span class="o">=</span><span class="n">query</span><span class="p">)</span>
<span class="n">q_model</span><span class="o">.</span><span class="n">get_latent_representation</span><span class="p">(</span><span class="n">adata</span><span class="o">=</span><span class="n">adata</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span><span class=" -Color -Color-Blue">INFO    </span> Input AnnData not setup with scvi-tools. attempting to transfer AnnData setup                             
</pre></div>
</div>
</div>
</div>
<p>Finally, we concatenate the reference and the query, and visualize both on a UMAP.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s2">&quot;reference&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="s2">&quot;reference&quot;</span>
<span class="n">query</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s2">&quot;reference&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="s2">&quot;query&quot;</span>

<span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s2">&quot;type_of_query&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="s2">&quot;reference&quot;</span>
<span class="n">query</span><span class="o">.</span><span class="n">obs</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">idx_atac_query</span><span class="p">,</span> <span class="s2">&quot;type_of_query&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="s2">&quot;ATAC query&quot;</span>
<span class="n">query</span><span class="o">.</span><span class="n">obs</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">idx_scrna_query</span><span class="p">,</span> <span class="s2">&quot;type_of_query&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="s2">&quot;scRNA query&quot;</span>
<span class="n">query</span><span class="o">.</span><span class="n">obs</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">idx_mutiome_query</span><span class="p">,</span> <span class="s2">&quot;type_of_query&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="s2">&quot;multiome query&quot;</span>
<span class="n">query</span><span class="o">.</span><span class="n">obs</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">idx_cite_query</span><span class="p">,</span> <span class="s2">&quot;type_of_query&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="s2">&quot;CITE-seq query&quot;</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">adata_both</span> <span class="o">=</span> <span class="n">anndata</span><span class="o">.</span><span class="n">concat</span><span class="p">([</span><span class="n">adata</span><span class="p">,</span> <span class="n">query</span><span class="p">])</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">sc</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">neighbors</span><span class="p">(</span><span class="n">adata_both</span><span class="p">,</span> <span class="n">use_rep</span><span class="o">=</span><span class="s2">&quot;latent&quot;</span><span class="p">)</span>
<span class="n">sc</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">umap</span><span class="p">(</span><span class="n">adata_both</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">sc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">umap</span><span class="p">(</span>
    <span class="n">adata_both</span><span class="p">,</span>
    <span class="n">color</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;cell_type&quot;</span><span class="p">,</span> <span class="s2">&quot;Modality&quot;</span><span class="p">,</span> <span class="s2">&quot;Samplename&quot;</span><span class="p">,</span> <span class="s2">&quot;reference&quot;</span><span class="p">],</span>
    <span class="n">ncols</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span>
    <span class="n">frameon</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
<span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="../_images/ce4dfb019c090906752c400a51829aa4d0069a436638597230e3f475bd934276.png" src="../_images/ce4dfb019c090906752c400a51829aa4d0069a436638597230e3f475bd934276.png" />
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">sc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">umap</span><span class="p">(</span>
    <span class="n">adata_both</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="s2">&quot;type_of_query&quot;</span><span class="p">,</span> <span class="n">ncols</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span> <span class="n">frameon</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">groups</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;ATAC query&quot;</span><span class="p">]</span>
<span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="../_images/32d9ccb2c504c8f3dfbb1e2e6077ac7179e5ac524efb2688a48b9d9484827a66.png" src="../_images/32d9ccb2c504c8f3dfbb1e2e6077ac7179e5ac524efb2688a48b9d9484827a66.png" />
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">sc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">umap</span><span class="p">(</span>
    <span class="n">adata_both</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="s2">&quot;type_of_query&quot;</span><span class="p">,</span> <span class="n">ncols</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span> <span class="n">frameon</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">groups</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;CITE-seq query&quot;</span><span class="p">]</span>
<span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="../_images/797523771cbb6cda6d4ee668a38e9f263de2ae774f04de76ed149d6cc5ccf078.png" src="../_images/797523771cbb6cda6d4ee668a38e9f263de2ae774f04de76ed149d6cc5ccf078.png" />
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">sc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">umap</span><span class="p">(</span>
    <span class="n">adata_both</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="s2">&quot;type_of_query&quot;</span><span class="p">,</span> <span class="n">ncols</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span> <span class="n">frameon</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">groups</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;multiome query&quot;</span><span class="p">]</span>
<span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="../_images/957b51236aca8bd549628434d17854900a2be683dc658d410000150f5cc388f2.png" src="../_images/957b51236aca8bd549628434d17854900a2be683dc658d410000150f5cc388f2.png" />
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">sc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">umap</span><span class="p">(</span>
    <span class="n">adata_both</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="s2">&quot;type_of_query&quot;</span><span class="p">,</span> <span class="n">ncols</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span> <span class="n">frameon</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">groups</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;scRNA query&quot;</span><span class="p">]</span>
<span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="../_images/2e3ce8e96190133c2771784fb978b351b309a397000a718928cee7dfa3e560f4.png" src="../_images/2e3ce8e96190133c2771784fb978b351b309a397000a718928cee7dfa3e560f4.png" />
</div>
</div>
<p>We observe that the query cell types are mapped to the corresponding cell types in the reference. We also note that unimodal query mapping is possible with Multigrate even though there was no unimodal data in the reference.</p>
</section>
</section>
<section id="session-info">
<h2><span class="section-number">46.9. </span>Session info<a class="headerlink" href="#session-info" title="Link to this heading">#</a></h2>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="o">%%</span><span class="k">R</span>
sessionInfo()
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>R version 4.1.3 (2022-03-10)
Platform: x86_64-conda-linux-gnu (64-bit)
Running under: CentOS Linux 7 (Core)

Matrix products: default
BLAS/LAPACK: /lustre/groups/ml01/workspace/anastasia.litinetskaya/miniconda3/envs/advanced-integration/lib/libopenblasp-r0.3.21.so

locale:
 [1] LC_CTYPE=C                 LC_NUMERIC=C              
 [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    
 [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   
 [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 
 [9] LC_ADDRESS=C               LC_TELEPHONE=C            
[11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       

attached base packages:
[1] stats4    tools     stats     graphics  grDevices utils     datasets 
[8] methods   base     

other attached packages:
 [1] Matrix_1.5-3                SeuratObject_4.1.3         
 [3] Seurat_4.1.1.9001           SingleCellExperiment_1.16.0
 [5] SummarizedExperiment_1.24.0 Biobase_2.54.0             
 [7] GenomicRanges_1.46.1        GenomeInfoDb_1.30.1        
 [9] IRanges_2.28.0              S4Vectors_0.32.4           
[11] BiocGenerics_0.40.0         MatrixGenerics_1.6.0       
[13] matrixStats_0.63.0         

loaded via a namespace (and not attached):
  [1] Rtsne_0.16             colorspace_2.0-3       deldir_1.0-6          
  [4] ellipsis_0.3.2         ggridges_0.5.4         XVector_0.34.0        
  [7] RcppHNSW_0.4.1         spatstat.data_3.0-0    farver_2.1.1          
 [10] leiden_0.4.3           listenv_0.9.0          ggrepel_0.9.2         
 [13] RSpectra_0.16-1        fansi_1.0.3            codetools_0.2-18      
 [16] splines_4.1.3          polyclip_1.10-4        jsonlite_1.8.4        
 [19] ica_1.0-3              cluster_2.1.4          png_0.1-8             
 [22] uwot_0.1.14            spatstat.sparse_3.0-0  shiny_1.7.4           
 [25] sctransform_0.3.5      compiler_4.1.3         httr_1.4.4            
 [28] fastmap_1.1.0          lazyeval_0.2.2         cli_3.5.0             
 [31] later_1.3.0            htmltools_0.5.4        igraph_1.3.5          
 [34] gtable_0.3.1           glue_1.6.2             GenomeInfoDbData_1.2.7
 [37] RANN_2.6.1             reshape2_1.4.4         dplyr_1.0.10          
 [40] Rcpp_1.0.9             scattermore_0.8        vctrs_0.5.1           
 [43] nlme_3.1-161           progressr_0.12.0       lmtest_0.9-40         
 [46] spatstat.random_3.0-1  stringr_1.5.0          globals_0.16.2        
 [49] mime_0.12              miniUI_0.1.1.1         lifecycle_1.0.3       
 [52] irlba_2.3.5.1          goftest_1.2-3          future_1.30.0         
 [55] zlibbioc_1.40.0        MASS_7.3-58.1          zoo_1.8-11            
 [58] scales_1.2.1           spatstat.core_2.4-4    promises_1.2.0.1      
 [61] spatstat.utils_3.0-1   parallel_4.1.3         RColorBrewer_1.1-3    
 [64] reticulate_1.26        pbapply_1.6-0          gridExtra_2.3         
 [67] ggplot2_3.4.0          rpart_4.1.19           stringi_1.7.8         
 [70] fastDummies_1.6.3      rlang_1.0.6            pkgconfig_2.0.3       
 [73] bitops_1.0-7           lattice_0.20-45        tensor_1.5            
 [76] ROCR_1.0-11            purrr_1.0.0            labeling_0.4.2        
 [79] patchwork_1.1.2        htmlwidgets_1.6.0      cowplot_1.1.1         
 [82] tidyselect_1.2.0       parallelly_1.33.0      RcppAnnoy_0.0.20      
 [85] plyr_1.8.8             magrittr_2.0.3         R6_2.5.1              
 [88] generics_0.1.3         DelayedArray_0.20.0    withr_2.5.0           
 [91] mgcv_1.8-41            pillar_1.8.1           fitdistrplus_1.1-8    
 [94] abind_1.4-5            survival_3.4-0         RCurl_1.98-1.9        
 [97] sp_1.5-1               tibble_3.1.8           future.apply_1.10.0   
[100] crayon_1.5.2           KernSmooth_2.23-20     utf8_1.2.2            
[103] spatstat.geom_3.0-3    plotly_4.10.1          grid_4.1.3            
[106] data.table_1.14.6      digest_0.6.31          xtable_1.8-4          
[109] tidyr_1.2.1            httpuv_1.6.7           munsell_0.5.0         
[112] viridisLite_0.4.1     
</pre></div>
</div>
</div>
</div>
</section>
<section id="references">
<h2><span class="section-number">46.10. </span>References<a class="headerlink" href="#references" title="Link to this heading">#</a></h2>
<div class="docutils container" id="id9">
<div role="list" class="citation-list">
<div class="citation" id="id551" role="doc-biblioentry">
<span class="label"><span class="fn-bracket">[</span><a role="doc-backlink" href="#id2">CG22</a><span class="fn-bracket">]</span></span>
<p>Zhi-Jie Cao and Ge Gao. Multi-omics single-cell data integration and regulatory inference with graph-linked embedding. <em>Nature Biotechnology</em>, May 2022. URL: <a class="reference external" href="https://doi.org/10.1038/s41587-022-01284-4">https://doi.org/10.1038/s41587-022-01284-4</a>, <a class="reference external" href="https://doi.org/10.1038/s41587-022-01284-4">doi:10.1038/s41587-022-01284-4</a>.</p>
</div>
<div class="citation" id="id543" role="doc-biblioentry">
<span class="label"><span class="fn-bracket">[</span><a role="doc-backlink" href="#id3">GSL+21</a><span class="fn-bracket">]</span></span>
<p>Adam Gayoso, Zoë Steier, Romain Lopez, Jeffrey Regier, Kristopher L. Nazor, Aaron Streets, and Nir Yosef. Joint probabilistic modeling of single-cell multi-omic data with totalvi. <em>Nature Methods</em>, 18(3):272–282, Mar 2021. URL: <a class="reference external" href="https://doi.org/10.1038/s41592-020-01050-x">https://doi.org/10.1038/s41592-020-01050-x</a>, <a class="reference external" href="https://doi.org/10.1038/s41592-020-01050-x">doi:10.1038/s41592-020-01050-x</a>.</p>
</div>
<div class="citation" id="id541" role="doc-biblioentry">
<span class="label"><span class="fn-bracket">[</span><a role="doc-backlink" href="#id4">HHAN+21</a><span class="fn-bracket">]</span></span>
<p>Yuhan Hao, Stephanie Hao, Erica Andersen-Nissen, William M. Mauck, Shiwei Zheng, Andrew Butler, Maddie J. Lee, Aaron J. Wilk, Charlotte Darby, Michael Zager, Paul Hoffman, Marlon Stoeckius, Efthymia Papalexi, Eleni P. Mimitou, Jaison Jain, Avi Srivastava, Tim Stuart, Lamar M. Fleming, Bertrand Yeung, Angela J. Rogers, Juliana M. McElrath, Catherine A. Blish, Raphael Gottardo, Peter Smibert, and Rahul Satija. Integrated analysis of multimodal single-cell data. <em>Cell</em>, 184(13):3573–3587.e29, 2021. URL: <a class="reference external" href="https://www.sciencedirect.com/science/article/pii/S0092867421005833">https://www.sciencedirect.com/science/article/pii/S0092867421005833</a>, <a class="reference external" href="https://doi.org/https://doi.org/10.1016/j.cell.2021.04.048">doi:https://doi.org/10.1016/j.cell.2021.04.048</a>.</p>
</div>
<div class="citation" id="id552" role="doc-biblioentry">
<span class="label"><span class="fn-bracket">[</span><a role="doc-backlink" href="#id5">HSK+22</a><span class="fn-bracket">]</span></span>
<p>Yuhan Hao, Tim Stuart, Madeline Kowalski, Saket Choudhary, Paul Hoffman, Austin Hartman, Avi Srivastava, Gesmira Molla, Shaista Madad, Carlos Fernandez-Granda, and Rahul Satija. Dictionary learning for integrative, multimodal, and scalable single-cell analysis. <em>bioRxiv</em>, 2022. URL: <a class="reference external" href="https://www.biorxiv.org/content/early/2022/02/26/2022.02.24.481684">https://www.biorxiv.org/content/early/2022/02/26/2022.02.24.481684</a>, <a class="reference external" href="https://arxiv.org/abs/https://www.biorxiv.org/content/early/2022/02/26/2022.02.24.481684.full.pdf">arXiv:https://www.biorxiv.org/content/early/2022/02/26/2022.02.24.481684.full.pdf</a>, <a class="reference external" href="https://doi.org/10.1101/2022.02.24.481684">doi:10.1101/2022.02.24.481684</a>.</p>
</div>
<div class="citation" id="id550" role="doc-biblioentry">
<span class="label"><span class="fn-bracket">[</span><a role="doc-backlink" href="#id1">LBC+21</a><span class="fn-bracket">]</span></span>
<p>Malte D Luecken, Daniel Bernard Burkhardt, Robrecht Cannoodt, Christopher Lance, Aditi Agrawal, Hananeh Aliee, Ann T Chen, Louise Deconinck, Angela M Detweiler, Alejandro A Granados, Shelly Huynh, Laura Isacco, Yang Joon Kim, Dominik Klein, BONY DE KUMAR, Sunil Kuppasani, Heiko Lickert, Aaron McGeever, Honey Mekonen, Joaquin Caceres Melgarejo, Maurizio Morri, Michaela Müller, Norma Neff, Sheryl Paul, Bastian Rieck, Kaylie Schneider, Scott Steelman, Michael Sterr, Daniel J. Treacy, Alexander Tong, Alexandra-Chloe Villani, Guilin Wang, Jia Yan, Ce Zhang, Angela Oliveira Pisco, Smita Krishnaswamy, Fabian J Theis, and Jonathan M. Bloom. A sandbox for prediction and integration of dna, rna, and proteins in single cells. In <em>Thirty-fifth Conference on Neural Information Processing Systems Datasets and Benchmarks Track (Round 2)</em>. 2021. URL: <a class="reference external" href="https://openreview.net/forum?id=gN35BGa1Rt">https://openreview.net/forum?id=gN35BGa1Rt</a>.</p>
</div>
</div>
</div>
</section>
<section id="contributors">
<h2><span class="section-number">46.11. </span>Contributors<a class="headerlink" href="#contributors" title="Link to this heading">#</a></h2>
<p>We gratefully acknowledge the contributions of:</p>
<section id="authors">
<h3><span class="section-number">46.11.1. </span>Authors<a class="headerlink" href="#authors" title="Link to this heading">#</a></h3>
<ul class="simple">
<li><p>Anastasia Litinetskaya</p></li>
</ul>
</section>
<section id="reviewers">
<h3><span class="section-number">46.11.2. </span>Reviewers<a class="headerlink" href="#reviewers" title="Link to this heading">#</a></h3>
<ul class="simple">
<li><p>Lukas Heumos</p></li>
</ul>
</section>
</section>
</section>

    <script type="text/x-thebe-config">
    {
        requestKernel: true,
        binderOptions: {
            repo: "binder-examples/jupyter-stacks-datascience",
            ref: "master",
        },
        codeMirrorConfig: {
            theme: "abcdef",
            mode: "python"
        },
        kernelOptions: {
            name: "python3",
            path: "./multimodal_integration"
        },
        predefinedOutput: true
    }
    </script>
    <script>kernelName = 'python3'</script>

                </article>
              

              
              
              
              
                <footer class="prev-next-footer d-print-none">
                  
<div class="prev-next-area">
    <a class="left-prev"
       href="paired_integration.html"
       title="previous page">
      <i class="fa-solid fa-angle-left"></i>
      <div class="prev-next-info">
        <p class="prev-next-subtitle">previous</p>
        <p class="prev-next-title"><span class="section-number">45. </span>Paired integration</p>
      </div>
    </a>
    <a class="right-next"
       href="../outlook.html"
       title="next page">
      <div class="prev-next-info">
        <p class="prev-next-subtitle">next</p>
        <p class="prev-next-title"><span class="section-number">47. </span>Outlook</p>
      </div>
      <i class="fa-solid fa-angle-right"></i>
    </a>
</div>
                </footer>
              
            </div>
            
            
              
                <div class="bd-sidebar-secondary bd-toc"><div class="sidebar-secondary-items sidebar-secondary__inner">


  <div class="sidebar-secondary-item">
  <div class="page-toc tocsection onthispage">
    <i class="fa-solid fa-list"></i> Contents
  </div>
  <nav class="bd-toc-nav page-toc">
    <ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#motivation">46.1. Motivation</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#environment-setup">46.2. Environment setup</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#prepare-data">46.3. Prepare data</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#unpaired-integration-with-graph-linked-unified-embedding-glue">46.4. Unpaired integration with Graph Linked Unified Embedding (GLUE)</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#graph-construction">46.4.1. Graph construction</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#configure-data">46.4.2. Configure data</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#integration-with-partially-overlapping-data-and-query-to-reference-mapping-with-totalvi">46.5. Integration with partially overlapping data and query-to-reference mapping with totalVI</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#query-to-reference-mapping">46.5.1. Query-to-reference mapping</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#query-mapping-and-imputation-with-seurat-s-wnn">46.6. Query mapping and imputation with Seurat’s WNN</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#mapping-adt-onto-an-rna-atlas-with-bridge-integration">46.7. Mapping ADT onto an RNA atlas with bridge integration</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#trimodal-integration-and-query-to-reference-mapping-with-multigrate">46.8. Trimodal integration and query-to-reference mapping with multigrate</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#building-the-reference">46.8.1. Building the reference</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#querying-the-trimodal-reference">46.8.2. Querying the trimodal reference</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#session-info">46.9. Session info</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#references">46.10. References</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#contributors">46.11. Contributors</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#authors">46.11.1. Authors</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#reviewers">46.11.2. Reviewers</a></li>
</ul>
</li>
</ul>
  </nav></div>

</div></div>
              
            
          </div>
          <footer class="bd-footer-content">
            
<div class="bd-footer-content__inner container">
  
  <div class="footer-item">
    
<p class="component-author">
By Lukas Heumos, Anna Schaar, single-cell best practices consortium
</p>

  </div>
  
  <div class="footer-item">
    

  <p class="copyright">
    
      © Copyright 2023.
      <br/>
    
  </p>

  </div>
  
  <div class="footer-item">
    
  </div>
  
  <div class="footer-item">
    
<div class="extra_footer">
  <div>
  Brought to you by
  <a href="https://www.github.com/theislab">Theislab</a>, with many thanks to the single-cell community as a whole!
</div>

</div>
  </div>
  
</div>
          </footer>
        

      </main>
    </div>
  </div>
  
  <!-- Scripts loaded after <body> so the DOM is not blocked -->
  <script src="../_static/scripts/bootstrap.js?digest=dfe6caa3a7d634c4db9b"></script>
<script src="../_static/scripts/pydata-sphinx-theme.js?digest=dfe6caa3a7d634c4db9b"></script>

  <footer class="bd-footer">
  </footer>
  </body>
</html>