
<!DOCTYPE html>


<html lang="en" data-content_root="../" >

  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="viewport" content="width=device-width, initial-scale=1" />

    <title>12. Data integration &#8212; Single-cell best practices</title>
  
  
  
  <script data-cfasync="false">
    document.documentElement.dataset.mode = localStorage.getItem("mode") || "";
    document.documentElement.dataset.theme = localStorage.getItem("theme") || "";
  </script>
  
  <!-- Loaded before other Sphinx assets -->
  <link href="../_static/styles/theme.css?digest=dfe6caa3a7d634c4db9b" rel="stylesheet" />
<link href="../_static/styles/bootstrap.css?digest=dfe6caa3a7d634c4db9b" rel="stylesheet" />
<link href="../_static/styles/pydata-sphinx-theme.css?digest=dfe6caa3a7d634c4db9b" rel="stylesheet" />

  
  <link href="../_static/vendor/fontawesome/6.5.2/css/all.min.css?digest=dfe6caa3a7d634c4db9b" rel="stylesheet" />
  <link rel="preload" as="font" type="font/woff2" crossorigin href="../_static/vendor/fontawesome/6.5.2/webfonts/fa-solid-900.woff2" />
<link rel="preload" as="font" type="font/woff2" crossorigin href="../_static/vendor/fontawesome/6.5.2/webfonts/fa-brands-400.woff2" />
<link rel="preload" as="font" type="font/woff2" crossorigin href="../_static/vendor/fontawesome/6.5.2/webfonts/fa-regular-400.woff2" />

    <link rel="stylesheet" type="text/css" href="../_static/pygments.css?v=03e43079" />
    <link rel="stylesheet" type="text/css" href="../_static/styles/sphinx-book-theme.css?v=eba8b062" />
    <link rel="stylesheet" type="text/css" href="../_static/togglebutton.css?v=13237357" />
    <link rel="stylesheet" type="text/css" href="../_static/copybutton.css?v=76b2166b" />
    <link rel="stylesheet" type="text/css" href="../_static/mystnb.4510f1fc1dee50b3e5859aac5469c37c29e427902b24a333a5f9fcb2f0b3ac41.css" />
    <link rel="stylesheet" type="text/css" href="../_static/sphinx-thebe.css?v=4fa983c6" />
    <link rel="stylesheet" type="text/css" href="../_static/sphinx-design.min.css?v=95c83b7e" />
    <link rel="stylesheet" type="text/css" href="../_static/book.css?v=148c11e2" />
  
  <!-- Pre-loaded scripts that we'll load fully later -->
  <link rel="preload" as="script" href="../_static/scripts/bootstrap.js?digest=dfe6caa3a7d634c4db9b" />
<link rel="preload" as="script" href="../_static/scripts/pydata-sphinx-theme.js?digest=dfe6caa3a7d634c4db9b" />
  <script src="../_static/vendor/fontawesome/6.5.2/js/all.min.js?digest=dfe6caa3a7d634c4db9b"></script>

    <script src="../_static/documentation_options.js?v=9eb32ce0"></script>
    <script src="../_static/doctools.js?v=9a2dae69"></script>
    <script src="../_static/sphinx_highlight.js?v=dc90522c"></script>
    <script src="../_static/clipboard.min.js?v=a7894cd8"></script>
    <script src="../_static/copybutton.js?v=f281be69"></script>
    <script src="../_static/scripts/sphinx-book-theme.js?v=887ef09a"></script>
    <script>let toggleHintShow = 'Click to show';</script>
    <script>let toggleHintHide = 'Click to hide';</script>
    <script>let toggleOpenOnPrint = 'true';</script>
    <script src="../_static/togglebutton.js?v=4a39c7ea"></script>
    <script>var togglebuttonSelector = '.toggle, .admonition.dropdown';</script>
    <script src="../_static/design-tabs.js?v=f930bc37"></script>
    <script>const THEBE_JS_URL = "https://unpkg.com/thebe@0.8.2/lib/index.js"; const thebe_selector = ".thebe,.cell"; const thebe_selector_input = "pre"; const thebe_selector_output = ".output, .cell_output"</script>
    <script async="async" src="../_static/sphinx-thebe.js?v=c100c467"></script>
    <script>var togglebuttonSelector = '.toggle, .admonition.dropdown';</script>
    <script>const THEBE_JS_URL = "https://unpkg.com/thebe@0.8.2/lib/index.js"; const thebe_selector = ".thebe,.cell"; const thebe_selector_input = "pre"; const thebe_selector_output = ".output, .cell_output"</script>
    <script>DOCUMENTATION_OPTIONS.pagename = 'cellular_structure/integration';</script>
    <script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
    <script src="../_static/book.js?v=dcd3d8bf"></script>
    <link rel="icon" href="../_static/favicon.ico"/>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="13. Pseudotemporal ordering" href="../trajectories/pseudotemporal.html" />
    <link rel="prev" title="11. Annotation" href="annotation.html" />
  <meta name="viewport" content="width=device-width, initial-scale=1"/>
  <meta name="docsearch:language" content="en"/>
  </head>
  
  
  <body data-bs-spy="scroll" data-bs-target=".bd-toc-nav" data-offset="180" data-bs-root-margin="0px 0px -60%" data-default-mode="">

  
  
  <div id="pst-skip-link" class="skip-link d-print-none"><a href="#main-content">Skip to main content</a></div>
  
  <div id="pst-scroll-pixel-helper"></div>
  
  <button type="button" class="btn rounded-pill" id="pst-back-to-top">
    <i class="fa-solid fa-arrow-up"></i>Back to top</button>

  
  <input type="checkbox"
          class="sidebar-toggle"
          id="pst-primary-sidebar-checkbox"/>
  <label class="overlay overlay-primary" for="pst-primary-sidebar-checkbox"></label>
  
  <input type="checkbox"
          class="sidebar-toggle"
          id="pst-secondary-sidebar-checkbox"/>
  <label class="overlay overlay-secondary" for="pst-secondary-sidebar-checkbox"></label>
  
  <div class="search-button__wrapper">
    <div class="search-button__overlay"></div>
    <div class="search-button__search-container">
<form class="bd-search d-flex align-items-center"
      action="../search.html"
      method="get">
  <i class="fa-solid fa-magnifying-glass"></i>
  <input type="search"
         class="form-control"
         name="q"
         id="search-input"
         placeholder="Search this book..."
         aria-label="Search this book..."
         autocomplete="off"
         autocorrect="off"
         autocapitalize="off"
         spellcheck="false"/>
  <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd>K</kbd></span>
</form></div>
  </div>

  <div class="pst-async-banner-revealer d-none">
  <aside id="bd-header-version-warning" class="d-none d-print-none" aria-label="Version warning"></aside>
</div>

  
    <header class="bd-header navbar navbar-expand-lg bd-navbar d-print-none">
    </header>
  

  <div class="bd-container">
    <div class="bd-container__inner bd-page-width">
      
      
      
      <div class="bd-sidebar-primary bd-sidebar">
        

  
  <div class="sidebar-header-items sidebar-primary__section">
    
    
    
    
  </div>
  
    <div class="sidebar-primary-items__start sidebar-primary__section">
        <div class="sidebar-primary-item">

  
    
  

<a class="navbar-brand logo" href="../preamble.html">
  
  
  
  
  
    
    
      
    
    
    <img src="../_static/logo.png" class="logo__image only-light" alt="Single-cell best practices - Home"/>
    <script>document.write(`<img src="../_static/logo.png" class="logo__image only-dark" alt="Single-cell best practices - Home"/>`);</script>
  
  
</a></div>
        <div class="sidebar-primary-item">

 <script>
 document.write(`
   <button class="btn search-button-field search-button__button" title="Search" aria-label="Search" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="fa-solid fa-magnifying-glass"></i>
    <span class="search-button__default-text">Search</span>
    <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd class="kbd-shortcut__modifier">K</kbd></span>
   </button>
 `);
 </script></div>
        <div class="sidebar-primary-item"><nav class="bd-links bd-docs-nav" aria-label="Main">
    <div class="bd-toc-item navbar-nav active">
        <p aria-level="2" class="caption" role="heading"><span class="caption-text">Introduction</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../introduction/prior_art.html">1. Prior art</a></li>
<li class="toctree-l1"><a class="reference internal" href="../introduction/scrna_seq.html">2. Single-cell RNA sequencing</a></li>
<li class="toctree-l1"><a class="reference internal" href="../introduction/raw_data_processing.html">3. Raw data processing</a></li>
<li class="toctree-l1"><a class="reference internal" href="../introduction/analysis_tools.html">4. Analysis frameworks and tools</a></li>
<li class="toctree-l1"><a class="reference internal" href="../introduction/interoperability.html">5. Interoperability</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Preprocessing and visualization</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../preprocessing_visualization/quality_control.html">6. Quality Control</a></li>
<li class="toctree-l1"><a class="reference internal" href="../preprocessing_visualization/normalization.html">7. Normalization</a></li>
<li class="toctree-l1"><a class="reference internal" href="../preprocessing_visualization/feature_selection.html">8. Feature selection</a></li>
<li class="toctree-l1"><a class="reference internal" href="../preprocessing_visualization/dimensionality_reduction.html">9. Dimensionality Reduction</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Identifying cellular structure</span></p>
<ul class="current nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="clustering.html">10. Clustering</a></li>
<li class="toctree-l1"><a class="reference internal" href="annotation.html">11. Annotation</a></li>
<li class="toctree-l1 current active"><a class="current reference internal" href="#">12. Data integration</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Inferring trajectories</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../trajectories/pseudotemporal.html">13. Pseudotemporal ordering</a></li>
<li class="toctree-l1"><a class="reference internal" href="../trajectories/rna_velocity.html">14. RNA velocity</a></li>
<li class="toctree-l1"><a class="reference internal" href="../trajectories/lineage_tracing.html">15. Lineage tracing</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Dealing with conditions</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../conditions/differential_gene_expression.html">16. Differential gene expression analysis</a></li>
<li class="toctree-l1"><a class="reference internal" href="../conditions/compositional.html">17. Compositional analysis</a></li>
<li class="toctree-l1"><a class="reference internal" href="../conditions/gsea_pathway.html">18. Gene set enrichment and pathway analysis</a></li>
<li class="toctree-l1"><a class="reference internal" href="../conditions/perturbation_modeling.html">19. Perturbation modeling</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Modeling mechanisms</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../mechanisms/gene_regulatory_networks.html">20. Gene regulatory networks</a></li>
<li class="toctree-l1"><a class="reference internal" href="../mechanisms/cell_cell_communication.html">21. Cell-cell communication</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Deconvolution</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../deconvolution/bulk_deconvolution.html">22. Bulk deconvolution</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Chromatin Accessibility</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../chromatin_accessibility/introduction.html">23. Single-cell ATAC sequencing</a></li>
<li class="toctree-l1"><a class="reference internal" href="../chromatin_accessibility/quality_control.html">24. Quality Control</a></li>
<li class="toctree-l1"><a class="reference internal" href="../chromatin_accessibility/gene_regulatory_networks_atac.html">25. Gene regulatory networks using chromatin accessibility</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Spatial omics</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../spatial/introduction.html">26. Single-cell data resolved in space</a></li>
<li class="toctree-l1"><a class="reference internal" href="../spatial/neighborhood.html">27. Neighborhood analysis</a></li>
<li class="toctree-l1"><a class="reference internal" href="../spatial/domains.html">28. Spatial domains</a></li>
<li class="toctree-l1"><a class="reference internal" href="../spatial/spatially_variable_genes.html">29. Spatially variable genes</a></li>
<li class="toctree-l1"><a class="reference internal" href="../spatial/deconvolution.html">30. Spatial deconvolution</a></li>
<li class="toctree-l1"><a class="reference internal" href="../spatial/imputation.html">31. Imputation</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Surface protein</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../surface_protein/quality_control.html">32. Quality control</a></li>
<li class="toctree-l1"><a class="reference internal" href="../surface_protein/normalization.html">33. Normalization</a></li>
<li class="toctree-l1"><a class="reference internal" href="../surface_protein/doublet_detection.html">34. Doublet detection</a></li>
<li class="toctree-l1"><a class="reference internal" href="../surface_protein/dimensionality_reduction.html">35. Dimensionality Reduction</a></li>
<li class="toctree-l1"><a class="reference internal" href="../surface_protein/batch_correction.html">36. Batch correction</a></li>
<li class="toctree-l1"><a class="reference internal" href="../surface_protein/annotation.html">37. Annotation</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Adaptive immune receptor repertoire</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../air_repertoire/ir_profiling.html">38. Immune Receptor Profiling</a></li>
<li class="toctree-l1"><a class="reference internal" href="../air_repertoire/clonotype.html">39. Clonotype analysis</a></li>



<li class="toctree-l1"><a class="reference internal" href="../air_repertoire/specificity.html">43. Specificity analysis</a></li>
<li class="toctree-l1"><a class="reference internal" href="../air_repertoire/multimodal_integration.html">44. Integrating AIR and transcriptomics</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Multimodal integration</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../multimodal_integration/paired_integration.html">45. Paired integration</a></li>
<li class="toctree-l1"><a class="reference internal" href="../multimodal_integration/advanced_integration.html">46. Advanced integration</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Reproducibility</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../reproducibility/introduction.html">47. Reproducibility</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Outlook</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../outlook.html">48. Outlook</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Acknowledgements</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../acknowledgements.html">49. Acknowledgements</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Glossary</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../glossary.html">50. Glossary</a></li>
</ul>

    </div>
</nav></div>
    </div>
  
  
  <div class="sidebar-primary-items__end sidebar-primary__section">
  </div>
  
  <div id="rtd-footer-container"></div>


      </div>
      
      <main id="main-content" class="bd-main" role="main">
        
        

<div class="sbt-scroll-pixel-helper"></div>

          <div class="bd-content">
            <div class="bd-article-container">
              
              <div class="bd-header-article d-print-none">
<div class="header-article-items header-article__inner">
  
    <div class="header-article-items__start">
      
        <div class="header-article-item"><button class="sidebar-toggle primary-toggle btn btn-sm" title="Toggle primary sidebar" data-bs-placement="bottom" data-bs-toggle="tooltip">
  <span class="fa-solid fa-bars"></span>
</button></div>
      
    </div>
  
  
    <div class="header-article-items__end">
      
        <div class="header-article-item">

<div class="article-header-buttons">





<div class="dropdown dropdown-source-buttons">
  <button class="btn dropdown-toggle" type="button" data-bs-toggle="dropdown" aria-expanded="false" aria-label="Source repositories">
    <i class="fab fa-github"></i>
  </button>
  <ul class="dropdown-menu">
      
      
      
      <li><a href="https://github.com/theislab/single-cell-best-practices" target="_blank"
   class="btn btn-sm btn-source-repository-button dropdown-item"
   title="Source repository"
   data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fab fa-github"></i>
  </span>
<span class="btn__text-container">Repository</span>
</a>
</li>
      
      
      
      
      <li><a href="https://github.com/theislab/single-cell-best-practices/edit/master/jupyter-book/cellular_structure/integration.ipynb" target="_blank"
   class="btn btn-sm btn-source-edit-button dropdown-item"
   title="Suggest edit"
   data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-pencil-alt"></i>
  </span>
<span class="btn__text-container">Suggest edit</span>
</a>
</li>
      
      
      
      
      <li><a href="https://github.com/theislab/single-cell-best-practices/issues/new?title=Issue%20on%20page%20%2Fcellular_structure/integration.html&body=Your%20issue%20content%20here." target="_blank"
   class="btn btn-sm btn-source-issues-button dropdown-item"
   title="Open an issue"
   data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-lightbulb"></i>
  </span>
<span class="btn__text-container">Open issue</span>
</a>
</li>
      
  </ul>
</div>






<div class="dropdown dropdown-download-buttons">
  <button class="btn dropdown-toggle" type="button" data-bs-toggle="dropdown" aria-expanded="false" aria-label="Download this page">
    <i class="fas fa-download"></i>
  </button>
  <ul class="dropdown-menu">
      
      
      
      <li><a href="../_sources/cellular_structure/integration.ipynb" target="_blank"
   class="btn btn-sm btn-download-source-button dropdown-item"
   title="Download source file"
   data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-file"></i>
  </span>
<span class="btn__text-container">.ipynb</span>
</a>
</li>
      
      
      
      
      <li>
<button onclick="window.print()"
  class="btn btn-sm btn-download-pdf-button dropdown-item"
  title="Print to PDF"
  data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-file-pdf"></i>
  </span>
<span class="btn__text-container">.pdf</span>
</button>
</li>
      
  </ul>
</div>




<button onclick="toggleFullScreen()"
  class="btn btn-sm btn-fullscreen-button"
  title="Fullscreen mode"
  data-bs-placement="bottom" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-expand"></i>
  </span>

</button>



<script>
document.write(`
  <button class="btn btn-sm nav-link pst-navbar-icon theme-switch-button" title="light/dark" aria-label="light/dark" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="theme-switch fa-solid fa-sun fa-lg" data-mode="light"></i>
    <i class="theme-switch fa-solid fa-moon fa-lg" data-mode="dark"></i>
    <i class="theme-switch fa-solid fa-circle-half-stroke fa-lg" data-mode="auto"></i>
  </button>
`);
</script>


<script>
document.write(`
  <button class="btn btn-sm pst-navbar-icon search-button search-button__button" title="Search" aria-label="Search" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="fa-solid fa-magnifying-glass fa-lg"></i>
  </button>
`);
</script>
<button class="sidebar-toggle secondary-toggle btn btn-sm" title="Toggle secondary sidebar" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <span class="fa-solid fa-list"></span>
</button>
</div></div>
      
    </div>
  
</div>
</div>
              
              

<div id="jb-print-docs-body" class="onlyprint">
    <h1>Data integration</h1>
    <!-- Table of contents -->
    <div id="print-main-content">
        <div id="jb-print-toc">
            
            <div>
                <h2> Contents </h2>
            </div>
            <nav aria-label="Page">
                <ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#motivation">12.1. Motivation</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#types-of-integration-models">12.1.1. Types of integration models</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#batch-removal-complexity">12.1.2. Batch removal complexity</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#comparisons-of-data-integration-methods">12.1.3. Comparisons of data integration methods</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#choosing-an-integration-method">12.1.4. Choosing an integration method</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#dataset">12.2. Dataset</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#unintegrated-data">12.3. Unintegrated data</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#batch-aware-feature-selection">12.4. Batch-aware feature selection</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#variational-autoencoder-vae-based-integration">12.5. Variational autoencoder (VAE) based integration</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#data-preparation">12.5.1. Data preparation</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#building-the-model">12.5.2. Building the model</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#training-the-model">12.5.3. Training the model</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#extracting-the-embedding">12.5.4. Extracting the embedding</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#calculate-a-batch-corrected-umap">12.5.5. Calculate a batch-corrected UMAP</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#vae-integration-using-cell-labels">12.6. VAE integration using cell labels</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#graph-based-integration">12.7. Graph-based integration</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#linear-embedding-integration-using-mutual-nearest-neighbors-mnn">12.8. Linear embedding integration using Mutual Nearest Neighbors (MNN)</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#benchmarking-your-own-integration">12.9. Benchmarking your own integration</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#quiz">12.10. Quiz</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#session-information">12.11. Session information</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#python">12.11.1. Python</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#r">12.11.2. R</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#references">12.12. References</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#contributors">12.13. Contributors</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#authors">12.13.1. Authors</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#reviewers">12.13.2. Reviewers</a></li>
</ul>
</li>
</ul>
            </nav>
        </div>
    </div>
</div>

              
                
<div id="searchbox"></div>
                <article class="bd-article">
                  
  <section class="tex2jax_ignore mathjax_ignore" id="data-integration">
<span id="cellular-structure-data-integration"></span><h1><span class="section-number">12. </span>Data integration<a class="headerlink" href="#data-integration" title="Link to this heading">#</a></h1>
<details class="sd-sphinx-override sd-dropdown sd-card sd-mb-3">
<summary class="sd-summary-title sd-card-header">
<span class="sd-summary-text"><i class="fas fa-brain"></i>   Key takeaways</span><span class="sd-summary-state-marker sd-summary-chevron-right"><svg version="1.1" width="1.5em" height="1.5em" class="sd-octicon sd-octicon-chevron-right" viewBox="0 0 24 24" aria-hidden="true"><path d="M8.72 18.78a.75.75 0 0 1 0-1.06L14.44 12 8.72 6.28a.751.751 0 0 1 .018-1.042.751.751 0 0 1 1.042-.018l6.25 6.25a.75.75 0 0 1 0 1.06l-6.25 6.25a.75.75 0 0 1-1.06 0Z"></path></svg></span></summary><div class="sd-summary-content sd-card-body docutils">
<div class="sd-card sd-sphinx-override sd-mb-3 sd-shadow-sm sd-card-hover docutils">
<div class="sd-card-body docutils">
<p class="sd-card-text">Visualize your data before attempting to correct for batch effects to assess the extent of the issue.
Batch effect correction is not always required and it might mask the biological variation of interest.</p>
</div>
<a class="sd-stretched-link sd-hide-link-text reference internal" href="#cellular-structure-integration-key-takeaway-1"><span class="std std-ref">Motivation</span></a></div>
<div class="sd-card sd-sphinx-override sd-mb-3 sd-shadow-sm sd-card-hover docutils">
<div class="sd-card-body docutils">
<p class="sd-card-text">If cell labels are available and biological variation is the most important, the usage of methods that can use these labels (such as scANVI) is advised.</p>
</div>
<a class="sd-stretched-link sd-hide-link-text reference internal" href="#cellular-structure-integration-key-takeaway-2"><span class="std std-ref">Types of integration models</span></a></div>
<div class="sd-card sd-sphinx-override sd-mb-3 sd-shadow-sm sd-card-hover docutils">
<div class="sd-card-body docutils">
<p class="sd-card-text">Consider running several integration methods on your dataset and evaluating them with the <strong>scIB</strong> metrics to use the integration that is most robust for your use case.</p>
</div>
<a class="sd-stretched-link sd-hide-link-text reference internal" href="#cellular-structure-integration-key-takeaway-3"><span class="std std-ref">Comparisons of data integration methods</span></a></div>
</div>
</details><details class="sd-sphinx-override sd-dropdown sd-card sd-mb-3">
<summary class="sd-summary-title sd-card-header">
<span class="sd-summary-text"><i class="fa-solid fa-gear"></i>   Environment setup</span><span class="sd-summary-state-marker sd-summary-chevron-right"><svg version="1.1" width="1.5em" height="1.5em" class="sd-octicon sd-octicon-chevron-right" viewBox="0 0 24 24" aria-hidden="true"><path d="M8.72 18.78a.75.75 0 0 1 0-1.06L14.44 12 8.72 6.28a.751.751 0 0 1 .018-1.042.751.751 0 0 1 1.042-.018l6.25 6.25a.75.75 0 0 1 0 1.06l-6.25 6.25a.75.75 0 0 1-1.06 0Z"></path></svg></span></summary><div class="sd-summary-content sd-card-body docutils">
<div class="sd-tab-set docutils">
<input checked="checked" id="sd-tab-item-0" name="sd-tab-set-0" type="radio">
<label class="sd-tab-label" for="sd-tab-item-0">
Steps</label><div class="sd-tab-content docutils">
<ol class="arabic">
<li><p class="sd-card-text"><strong>Install conda</strong>:</p>
<ul class="simple">
<li><p class="sd-card-text">Before creating the environment, ensure that <a class="reference external" href="https://docs.conda.io/projects/conda/en/latest/user-guide/install/index.html">conda</a> is installed on your system.</p></li>
</ul>
</li>
<li><p class="sd-card-text"><strong>Save the yml content</strong>:</p>
<ul class="simple">
<li><p class="sd-card-text">Copy the content from the yml tab into a file named <code class="docutils literal notranslate"><span class="pre">environment.yml</span></code>.</p></li>
</ul>
</li>
<li><p class="sd-card-text"><strong>Create the environment</strong>:</p>
<ul>
<li><p class="sd-card-text">Open a terminal or command prompt.</p></li>
<li><p class="sd-card-text">Run the following command:</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>conda<span class="w"> </span>env<span class="w"> </span>create<span class="w"> </span>-f<span class="w"> </span>environment.yml
</pre></div>
</div>
</li>
</ul>
</li>
<li><p class="sd-card-text"><strong>Activate the environment</strong>:</p>
<ul>
<li><p class="sd-card-text">After the environment is created, activate it using:</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>conda<span class="w"> </span>activate<span class="w"> </span>&lt;environment_name&gt;
</pre></div>
</div>
</li>
<li><p class="sd-card-text">Replace <code class="docutils literal notranslate"><span class="pre">&lt;environment_name&gt;</span></code> with the name specified in the <code class="docutils literal notranslate"><span class="pre">environment.yml</span></code> file. In the yml file it will look like this:</p>
<div class="highlight-yaml notranslate"><div class="highlight"><pre><span></span><span class="nt">name</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">&lt;environment_name&gt;</span>
</pre></div>
</div>
</li>
</ul>
</li>
<li><p class="sd-card-text"><strong>Verify the installation</strong>:</p>
<ul>
<li><p class="sd-card-text">Check that the environment was created successfully by running:</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>conda<span class="w"> </span>env<span class="w"> </span>list
</pre></div>
</div>
</li>
</ul>
</li>
</ol>
</div>
<input id="sd-tab-item-1" name="sd-tab-set-0" type="radio">
<label class="sd-tab-label" for="sd-tab-item-1">
yml</label><div class="sd-tab-content docutils">
<div class="highlight-yaml notranslate"><div class="highlight"><pre><span></span><span class="nt">name</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">integration</span>
<span class="nt">channels</span><span class="p">:</span>
<span class="w">  </span><span class="p p-Indicator">-</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">conda-forge</span>
<span class="w">  </span><span class="p p-Indicator">-</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">bioconda</span>
<span class="w">  </span><span class="p p-Indicator">-</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">defaults</span>
<span class="nt">dependencies</span><span class="p">:</span>
<span class="w">  </span><span class="p p-Indicator">-</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">conda-forge::python=3.9.15</span>
<span class="w">  </span><span class="p p-Indicator">-</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">conda-forge::jupyterlab=3.5.0</span>
<span class="w">  </span><span class="p p-Indicator">-</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">conda-forge::scanpy=1.9.1</span>
<span class="w">  </span><span class="p p-Indicator">-</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">anndata2ri=1.1</span>
<span class="w">  </span><span class="p p-Indicator">-</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">bbknn=1.5.1</span>
<span class="w">  </span><span class="p p-Indicator">-</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">bioconductor-singlecellexperiment=1.16.0</span>
<span class="w">  </span><span class="p p-Indicator">-</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">igraph=0.10.2</span>
<span class="w">  </span><span class="p p-Indicator">-</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">jupyterlab=3.5.0</span>
<span class="w">  </span><span class="p p-Indicator">-</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">pip=22.3</span>
<span class="w">  </span><span class="p p-Indicator">-</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">python-igraph=0.10.2</span>
<span class="w">  </span><span class="p p-Indicator">-</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">conda-forge::r-base=4.1.3</span>
<span class="w">  </span><span class="p p-Indicator">-</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">r-sessioninfo=1.2.2</span>
<span class="w">  </span><span class="p p-Indicator">-</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">r-seurat=4.2.0</span>
<span class="w">  </span><span class="p p-Indicator">-</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">r-spatstat=2.3_4</span>
<span class="w">  </span><span class="p p-Indicator">-</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">rpy2=3.5.1</span>
<span class="w">  </span><span class="p p-Indicator">-</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">scikit-learn=1.1.3</span>
<span class="w">  </span><span class="p p-Indicator">-</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">scikit-misc=0.1.4</span>
<span class="w">  </span><span class="p p-Indicator">-</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">scipy=1.9.3</span>
<span class="w">  </span><span class="p p-Indicator">-</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">scvi-tools=0.18.0</span>
<span class="w">  </span><span class="p p-Indicator">-</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">session-info=1.0.0</span>
<span class="w">  </span><span class="p p-Indicator">-</span><span class="w"> </span><span class="nt">pip</span><span class="p">:</span>
<span class="w">      </span><span class="p p-Indicator">-</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">scib==1.0.4</span>
</pre></div>
</div>
</div>
</div>
</div>
</details><section id="motivation">
<span id="cellular-structure-integration-key-takeaway-1"></span><h2><span class="section-number">12.1. </span>Motivation<a class="headerlink" href="#motivation" title="Link to this heading">#</a></h2>
<p>A central challenge in most scRNA-seq data analyses is presented by batch effects. Batch effects are changes in measured expression levels that are the result of handling cells in distinct groups or “batches”. For example, a batch effect can arise if two labs have taken samples from the same cohort, but these samples are dissociated differently. If Lab A optimizes its dissociation protocol to dissociate cells in the sample while minimizing the stress on them, and Lab B does not, then it is likely that the cells in the data from the group B will express more stress-linked genes (JUN, JUNB, FOS, etc. see <span id="id1">[<a class="reference internal" href="#id229" title="Susanne C van den Brink, Fanny Sage, Ábel Vértesy, Bastiaan Spanjaard, Josi Peterson-Maduro, Chloé S Baron, Catherine Robin, and Alexander van Oudenaarden. Single-cell sequencing reveals dissociation-induced gene expression in tissue subpopulations. Nature methods, 14(10):935–936, September 2017. doi:10.1038/nmeth.4437.">van den Brink <em>et al.</em>, 2017</a>]</span>) even if the cells had the same profile in the original tissue. In general, the origins of batch effects are diverse and difficult to pin down. Some batch effect sources might be technical such as differences in sample handling, experimental protocols, or sequencing depths, but biological effects such as donor variation, tissue, or sampling location are also often interpreted as a batch effect <span id="id2">[<a class="reference internal" href="#id220" title="Malte D Luecken, M Büttner, K Chaichoompu, A Danese, M Interlandi, M F Mueller, D C Strobl, L Zappia, M Dugas, M Colomé-Tatché, and Fabian J Theis. Benchmarking atlas-level data integration in single-cell genomics. Nature methods, December 2021. doi:10.1038/s41592-021-01336-8.">Luecken <em>et al.</em>, 2021</a>]</span>. Whether or not biological factors should be considered batch effects can depend on the experimental design and the question being asked. Removing batch effects is crucial to enable joint analysis that can focus on finding common structure in the data across batches and enable us to perform queries across datasets. Often it is only after removing these effects that rare cell populations can be identified that were previously obscured by differences between batches. Enabling queries across datasets allows us to ask questions that could not be answered by analysing individual datasets, such as <em>Which cell types express SARS-CoV-2 entry factors and how does this expression differ between individuals?</em> <span id="id3">[<a class="reference internal" href="#id230" title="Christoph Muus, Malte D Luecken, Gökcen Eraslan, Lisa Sikkema, Avinash Waghray, Graham Heimberg, Yoshihiko Kobayashi, Eeshit Dhaval Vaishnav, Ayshwarya Subramanian, Christopher Smillie, Karthik A Jagadeesh, Elizabeth Thu Duong, Evgenij Fiskin, Elena Torlai Triglia, Meshal Ansari, Peiwen Cai, Brian Lin, Justin Buchanan, Sijia Chen, Jian Shu, Adam L Haber, Hattie Chung, Daniel T Montoro, Taylor Adams, Hananeh Aliee, Samuel J Allon, Zaneta Andrusivova, Ilias Angelidis, Orr Ashenberg, Kevin Bassler, Christophe Bécavin, Inbal Benhar, Joseph Bergenstråhle, Ludvig Bergenstråhle, Liam Bolt, Emelie Braun, Linh T Bui, Steven Callori, Mark Chaffin, Evgeny Chichelnitskiy, Joshua Chiou, Thomas M Conlon, Michael S Cuoco, Anna S E Cuomo, Marie Deprez, Grant Duclos, Denise Fine, David S Fischer, Shila Ghazanfar, Astrid Gillich, Bruno Giotti, Joshua Gould, Minzhe Guo, Austin J Gutierrez, Arun C Habermann, Tyler Harvey, Peng He, Xiaomeng Hou, Lijuan Hu, Yan Hu, Alok Jaiswal, Lu Ji, Peiyong Jiang, Theodoros S Kapellos, Christin S Kuo, Ludvig Larsson, Michael A Leney-Greene, Kyungtae Lim, Monika Litviňuková, Leif S Ludwig, Soeren Lukassen, Wendy Luo, Henrike Maatz, Elo Madissoon, Lira Mamanova, Kasidet Manakongtreecheep, Sylvie Leroy, Christoph H Mayr, Ian M Mbano, Alexi M McAdams, Ahmad N Nabhan, Sarah K Nyquist, Lolita Penland, Olivier B Poirion, Sergio Poli, Cancan Qi, Rachel Queen, Daniel Reichart, Ivan Rosas, Jonas C Schupp, Conor V Shea, Xingyi Shi, Rahul Sinha, Rene V Sit, Kamil Slowikowski, Michal Slyper, Neal P Smith, Alex Sountoulidis, Maximilian Strunz, Travis B Sullivan, Dawei Sun, Carlos Talavera-López, Peng Tan, Jessica Tantivit, Kyle J Travaglini, Nathan R Tucker, Katherine A Vernon, Marc H Wadsworth, Julia Waldman, Xiuting Wang, Ke Xu, Wenjun Yan, William Zhao, Carly G K Ziegler, NHLBI LungMap Consortium, and Human Cell Atlas Lung Biological Network. Single-cell meta-analysis of SARS-CoV-2 entry genes across tissues and demographics. Nature medicine, 27(3):546–559, March 2021. doi:10.1038/s41591-020-01227-z.">Muus <em>et al.</em>, 2021</a>]</span>.</p>
<p>When removing batch effects from omics data, one must make two central choices: (1) the method and parameterization, and (2) the batch covariate. As batch effects can arise between groupings of cells at different levels (i.e., samples, donors, datasets etc.), the choice of batch covariate indicates which level of variation should be retained and which level removed. The finer the batch resolution, the more effects will be removed. However, fine batch variation is also more likely to be confounded with biologically meaningful signals. For example, samples typically come from different individuals or different locations in the tissue. These effects may be meaningful to inspect. Thus, the choice of batch covariate will depend on the goal of your integration task. Do you want to see differences between individuals, or are you focused on common cell type variation? An approach to batch covariate selection based on quantitative analyses was pioneered in a recent effort to build an integrated atlas of the human lung, where the variance attributable to different technical covariates was used to make this choice <span id="id4">[<a class="reference internal" href="#id223" title="Lisa Sikkema, Daniel C Strobl, Luke Zappia, Elo Madissoon, Nikolay S Markov, Laure-Emmanuelle Zaragosi, Meshal Ansari, Marie-Jeanne Arguel, Leonie Apperloo, Christophe Becavin, Marijn Berg, Evgeny Chichelnitskiy, Mei-I Chung, Antoine Collin, Aurore C A Gay, Baharak Hooshiar Kashani, Manu Jain, Theodore Kapellos, Tessa M Kole, Christoph H Mayr, Von Michael Papen, Lance Peter, Ciro Ramirez-Suastegui, Janine Schniering, Chase J Taylor, Thomas Walzthoeni, Chuan Xu, Linh T Bui, Carlo de Donno, Leander Dony, Minzhe Guo, Austin J Gutierrez, Lukas Heumos, Ni Huang, Ignacio L Ibarra, Nathan D Jackson, Preetish Kadur Lakshminarasimha Murthy, Mohammad Lotfollahi, Tracy Tabib, Carlos Talavera-Lopez, Kyle J Travaglini, Anna Wilbrey-Clark, Kaylee B Worlock, Masahiro Yoshida, Tushar J Desai, Oliver Eickelberg, Christine Falk, Naftali Kaminski, Mark A Krasnow, Robert Lafyatis, Marko Z Nikoli, Joseph E Powell, Jayaraj Rajagopal, Orit Rozenblatt-Rosen, Max A Seibold, Dean Sheppard, Douglas P Shepherd, Sarah A Teichmann, Alexander M Tsankov, Jeffrey Whitsett, Yan Xu, Nicholas E Banovich, Pascal Barbry, Thu E Duong, Kerstin B Meyer, Jonathan A Kropski, Dana Pe'er, Herbert B Schiller, Purushothama Rao Tata, Joachim L Schultze, Alexander V Misharin, Martijn C Nawijn, Malte D Luecken, and Fabian J Theis. An integrated cell atlas of the human lung in health and disease. bioRxiv, pages 2022.03.10.483747, March 2022. doi:10.1101/2022.03.10.483747.">Sikkema <em>et al.</em>, 2022</a>]</span>.</p>
<section id="types-of-integration-models">
<span id="cellular-structure-integration-key-takeaway-2"></span><h3><span class="section-number">12.1.1. </span>Types of integration models<a class="headerlink" href="#types-of-integration-models" title="Link to this heading">#</a></h3>
<p>Methods that remove batch effects in scRNA-seq are typically composed of (up to) three steps:</p>
<ol class="arabic simple">
<li><p>Dimensionality reduction</p></li>
<li><p>Modeling and removing the batch effect</p></li>
<li><p>Projection back into a high-dimensional space</p></li>
</ol>
<p>While modeling and removing the batch effect (Step 2) is the central part of any batch removal method, many methods first project the data to a lower dimensional space (Step 1) to improve the signal-to-noise ratio (see the <a class="reference internal" href="../preprocessing_visualization/dimensionality_reduction.html#pre-processing-dimensionality-reduction"><span class="std std-ref">dimensionality reduction chapter</span></a>) and perform batch correction in that space to achieve better performance (see <span id="id5">[<a class="reference internal" href="#id220" title="Malte D Luecken, M Büttner, K Chaichoompu, A Danese, M Interlandi, M F Mueller, D C Strobl, L Zappia, M Dugas, M Colomé-Tatché, and Fabian J Theis. Benchmarking atlas-level data integration in single-cell genomics. Nature methods, December 2021. doi:10.1038/s41592-021-01336-8.">Luecken <em>et al.</em>, 2021</a>]</span>). In the third step, a method may project the data back into the original high-dimensional feature space after removing the fitted batch effect to output a batch-corrected gene expression matrix.</p>
<p>Batch-effect removal methods can vary in each of these three steps. They may use various linear or non-linear dimensionality reduction approaches, linear or non-linear batch effect models, and they may output different formats of batch-corrected data. Overall, we can divide methods for batch effect removal into 4 categories. In their order of development, these are global models, linear embedding models, graph-based methods, and deep learning approaches (Fig I1).</p>
<p><strong>Global models</strong> originate from bulk transcriptomics and model the batch effect as a consistent (additive and/or multiplicative) effect across all cells. A common example is ComBat <span id="id6">[<a class="reference internal" href="#id231" title="W Evan Johnson, Cheng Li, and Ariel Rabinovic. Adjusting batch effects in microarray expression data using empirical Bayes methods. Biostatistics, 8(1):118–127, January 2007. doi:10.1093/biostatistics/kxj037.">Johnson <em>et al.</em>, 2007</a>]</span>.</p>
<p><strong>Linear embedding models</strong> were the first single-cell-specific batch removal methods. These approaches often use a variant of singular value decomposition (SVD) to embed the data, then look for local neighborhoods of similar cells across batches in the embedding, which they use to correct the batch effect in a locally adaptive (non-linear) manner. Methods often project the data back into gene expression space using the SVD loadings, but may also only output a corrected embedding. This is the most common group of methods and prominent examples include the pioneering mutual nearest neighbors (MNN) method <span id="id7">[<a class="reference internal" href="#id215" title="Laleh Haghverdi, Aaron T L Lun, Michael D Morgan, and John C Marioni. Batch effects in single-cell RNA-sequencing data are corrected by matching mutual nearest neighbors. Nature biotechnology, April 2018. doi:10.1038/nbt.4091.">Haghverdi <em>et al.</em>, 2018</a>]</span> (which does not perform any dimensionality reduction), Seurat integration <span id="id8">[<a class="reference internal" href="#id211" title="Andrew Butler, Paul Hoffman, Peter Smibert, Efthymia Papalexi, and Rahul Satija. Integrating single-cell transcriptomic data across different conditions, technologies, and species. Nature biotechnology, April 2018. doi:10.1038/nbt.4096.">Butler <em>et al.</em>, 2018</a>, <a class="reference internal" href="#id208" title="Tim Stuart, Andrew Butler, Paul Hoffman, Christoph Hafemeister, Efthymia Papalexi, William M Mauck, 3rd, Yuhan Hao, Marlon Stoeckius, Peter Smibert, and Rahul Satija. Comprehensive Integration of Single-Cell Data. Cell, 177(7):1888–1902.e21, June 2019. doi:10.1016/j.cell.2019.05.031.">Stuart <em>et al.</em>, 2019</a>]</span>, Scanorama <span id="id9">[<a class="reference internal" href="#id214" title="Brian Hie, Bryan Bryson, and Bonnie Berger. Efficient integration of heterogeneous single-cell transcriptomes using Scanorama. Nature biotechnology, May 2019. doi:10.1038/s41587-019-0113-3.">Hie <em>et al.</em>, 2019</a>]</span>, FastMNN <span id="id10">[<a class="reference internal" href="#id215" title="Laleh Haghverdi, Aaron T L Lun, Michael D Morgan, and John C Marioni. Batch effects in single-cell RNA-sequencing data are corrected by matching mutual nearest neighbors. Nature biotechnology, April 2018. doi:10.1038/nbt.4091.">Haghverdi <em>et al.</em>, 2018</a>]</span>, and Harmony <span id="id11">[<a class="reference internal" href="#id210" title="Ilya Korsunsky, Nghia Millard, Jean Fan, Kamil Slowikowski, Fan Zhang, Kevin Wei, Yuriy Baglaenko, Michael Brenner, Po-Ru Loh, and Soumya Raychaudhuri. Fast, sensitive and accurate integration of single-cell data with Harmony. Nature methods, November 2019. doi:10.1038/s41592-019-0619-0.">Korsunsky <em>et al.</em>, 2019</a>]</span>.</p>
<p><strong>Graph-based methods</strong> are typically the fastest methods to run. These approaches use a nearest-neighbor graph to represent the data from each batch. Batch effects are corrected by forcing connections between cells from different batches and then allowing for differences in cell type compositions by pruning the forced edges. The most prominent example of these approaches is the Batch-Balanced <em>k</em>-Nearest Neighbor (BBKNN) method <span id="id12">[<a class="reference internal" href="#id228" title="Krzysztof Polański, Jong-Eun Park, Matthew D Young, Zhichao Miao, Kerstin B Meyer, and Sarah A Teichmann. BBKNN: Fast Batch Alignment of Single Cell Transcriptomes. Bioinformatics, August 2019. doi:10.1093/bioinformatics/btz625.">Polański <em>et al.</em>, 2019</a>]</span>.</p>
<p><strong>Deep learning (DL) approaches</strong> are the most recent, and most complex methods for batch effect removal that typically require the most data for good performance. Most deep learning integration methods are based on autoencoder networks, and either condition the dimensionality reduction on the batch covariate in a conditional variational autoencoder (CVAE) or fit a locally linear correction in the embedded space. Prominent examples of DL methods are <strong>scVI</strong> <span id="id13">[<a class="reference internal" href="#id226" title="Romain Lopez, Jeffrey Regier, Michael B Cole, Michael I Jordan, and Nir Yosef. Deep generative modeling for single-cell transcriptomics. Nature methods, 15(12):1053–1058, November 2018. doi:10.1038/s41592-018-0229-2.">Lopez <em>et al.</em>, 2018</a>]</span>, <strong>scANVI</strong> <span id="id14">[<a class="reference internal" href="#id227" title="Chenling Xu, Romain Lopez, Edouard Mehlman, Jeffrey Regier, Michael I Jordan, and Nir Yosef. Probabilistic harmonization and annotation of single-cell transcriptomics data with deep generative models. Molecular systems biology, 17(1):e9620, January 2021. doi:10.15252/msb.20209620.">Xu <em>et al.</em>, 2021</a>]</span>, and <strong>scGen</strong> <span id="id15">[<a class="reference internal" href="#id221" title="Mohammad Lotfollahi, F Alexander Wolf, and Fabian J Theis. scGen predicts single-cell perturbation responses. Nature methods, 16(8):715–721, August 2019. doi:10.1038/s41592-019-0494-8.">Lotfollahi <em>et al.</em>, 2019</a>]</span>.</p>
<p>Some methods can use cell identity labels to provide the method with a reference for what biological variation should not be removed as a batch effect. As batch-effect removal is typically a preprocessing task, such approaches may not be applicable to many integration scenarios as labels are generally not yet available at this stage.</p>
<p>More detailed overviews of batch-effect removal methods can be found in <span id="id16">[<a class="reference internal" href="#id219" title="Ricard Argelaguet, Anna S E Cuomo, Oliver Stegle, and John C Marioni. Computational principles and challenges in single-cell data integration. Nature biotechnology, pages 1–14, May 2021. doi:10.1038/s41587-021-00895-7.">Argelaguet <em>et al.</em>, 2021</a>]</span> and <span id="id17">[<a class="reference internal" href="#id220" title="Malte D Luecken, M Büttner, K Chaichoompu, A Danese, M Interlandi, M F Mueller, D C Strobl, L Zappia, M Dugas, M Colomé-Tatché, and Fabian J Theis. Benchmarking atlas-level data integration in single-cell genomics. Nature methods, December 2021. doi:10.1038/s41592-021-01336-8.">Luecken <em>et al.</em>, 2021</a>]</span>.</p>
<p><img alt="Overview_fig" src="../_images/integration_overview_figure.jpeg" />
Fig. I1: Overview of different types of integration methods with examples.</p>
</section>
<section id="batch-removal-complexity">
<h3><span class="section-number">12.1.2. </span>Batch removal complexity<a class="headerlink" href="#batch-removal-complexity" title="Link to this heading">#</a></h3>
<p>The removal of batch effects in scRNA-seq data has previously been divided into two subtasks: batch correction and data integration <span id="id18">[<a class="reference internal" href="#id218" title="Malte D Luecken and Fabian J Theis. Current best practices in single‐cell RNA‐seq analysis: a tutorial. Molecular systems biology, June 2019. doi:10.15252/msb.20188746.">Luecken and Theis, 2019</a>]</span>. These subtasks differ in the complexity of the batch effect that must be removed. Batch correction methods deal with batch effects between samples in the same experiment where cell identity compositions are consistent, and the effect is often quasi-linear. In contrast, data integration methods deal with complex, often nested, batch effects between datasets that may be generated with different protocols and where cell identities may not be shared across batches. While we use this distinction here we should note that these terms are often used interchangeably in general use. Given the differences in complexity, it is not surprising that different methods have been benchmarked as being optimal for these two subtasks.</p>
</section>
<section id="comparisons-of-data-integration-methods">
<span id="cellular-structure-integration-key-takeaway-3"></span><h3><span class="section-number">12.1.3. </span>Comparisons of data integration methods<a class="headerlink" href="#comparisons-of-data-integration-methods" title="Link to this heading">#</a></h3>
<p>Several benchmarks have previously evaluated the performance of methods for batch correction and data integration. When removing batch effects, methods may overcorrect and remove meaningful biological variation in addition to the batch effect. For this reason, it is important that integration performance is evaluated by considering both batch effect removal and the conservation of biological variation.</p>
<p>The <em>k</em>-nearest-neighbor Batch-Effect Test (kBET) was the first metric for quantifying batch correction of scRNA-seq data  <span id="id19">[<a class="reference internal" href="#id212" title="Maren Büttner, Zhichao Miao, F Alexander Wolf, Sarah A Teichmann, and Fabian J Theis. A test metric for assessing single-cell RNA-seq batch correction. Nature methods, 16(1):43–49, January 2019. doi:10.1038/s41592-018-0254-1.">Büttner <em>et al.</em>, 2019</a>]</span>. Using kBET, the authors found that <strong>ComBat</strong> outperformed other approaches for batch correction while comparing predominantly global models. Building on this, two recent benchmarks <span id="id20">[<a class="reference internal" href="#id209" title="Hoa Thi Nhu Tran, Kok Siong Ang, Marion Chevrier, Xiaomeng Zhang, Nicole Yee Shin Lee, Michelle Goh, and Jinmiao Chen. A benchmark of batch-effect correction methods for single-cell RNA sequencing data. Genome biology, 21(1):12, January 2020. doi:10.1186/s13059-019-1850-9.">Tran <em>et al.</em>, 2020</a>]</span> and <span id="id21">[<a class="reference internal" href="#id232" title="Ruben Chazarra-Gil, Stijn van Dongen, Vladimir Yu Kiselev, and Martin Hemberg. Flexible comparison of batch correction methods for single-cell RNA-seq using BatchBench. Nucleic acids research, 49(7):e42, April 2021. doi:10.1093/nar/gkab004.">Chazarra-Gil <em>et al.</em>, 2021</a>]</span> also benchmarked linear-embedding and deep-learning models on batch correction tasks with few batches or low biological complexity. These studies found that the linear-embedding models <strong>Seurat</strong> <span id="id22">[<a class="reference internal" href="#id211" title="Andrew Butler, Paul Hoffman, Peter Smibert, Efthymia Papalexi, and Rahul Satija. Integrating single-cell transcriptomic data across different conditions, technologies, and species. Nature biotechnology, April 2018. doi:10.1038/nbt.4096.">Butler <em>et al.</em>, 2018</a>, <a class="reference internal" href="#id208" title="Tim Stuart, Andrew Butler, Paul Hoffman, Christoph Hafemeister, Efthymia Papalexi, William M Mauck, 3rd, Yuhan Hao, Marlon Stoeckius, Peter Smibert, and Rahul Satija. Comprehensive Integration of Single-Cell Data. Cell, 177(7):1888–1902.e21, June 2019. doi:10.1016/j.cell.2019.05.031.">Stuart <em>et al.</em>, 2019</a>]</span> and <strong>Harmony</strong> <span id="id23">[<a class="reference internal" href="#id210" title="Ilya Korsunsky, Nghia Millard, Jean Fan, Kamil Slowikowski, Fan Zhang, Kevin Wei, Yuriy Baglaenko, Michael Brenner, Po-Ru Loh, and Soumya Raychaudhuri. Fast, sensitive and accurate integration of single-cell data with Harmony. Nature methods, November 2019. doi:10.1038/s41592-019-0619-0.">Korsunsky <em>et al.</em>, 2019</a>]</span> performed well for simple batch correction tasks.</p>
<p>Benchmarking complex integration tasks poses additional challenges due to both the size and number of datasets as well as the diversity of scenarios. Recently, a large study used 14 metrics to benchmark 16 methods across integration method classes on five RNA tasks and two simulations <span id="id24">[<a class="reference internal" href="#id220" title="Malte D Luecken, M Büttner, K Chaichoompu, A Danese, M Interlandi, M F Mueller, D C Strobl, L Zappia, M Dugas, M Colomé-Tatché, and Fabian J Theis. Benchmarking atlas-level data integration in single-cell genomics. Nature methods, December 2021. doi:10.1038/s41592-021-01336-8.">Luecken <em>et al.</em>, 2021</a>]</span>. While top-performing methods per task differed, approaches that use cell type labels performed better across tasks. Furthermore, deep learning approaches <strong>scANVI</strong> (with labels), <strong>scVI</strong>, and <strong>scGen</strong> (with labels), as well as the linear embedding model <strong>Scanorama,</strong> performed best, particularly on complex tasks, while <strong>Harmony</strong> performed well on less complex tasks. A similar benchmark performed for the specific purpose of integrating retina datasets to build an <em>ocular mega-atlas</em> also found that <strong>scVI</strong> outperformed other methods <span id="id25">[<a class="reference internal" href="#id233" title="Vinay S Swamy, Temesgen D Fufa, Robert B Hufnagel, and David M McGaughey. Building the mega single-cell transcriptome ocular meta-atlas. GigaScience, October 2021. doi:10.1093/gigascience/giab061.">Swamy <em>et al.</em>, 2021</a>]</span>.</p>
</section>
<section id="choosing-an-integration-method">
<h3><span class="section-number">12.1.4. </span>Choosing an integration method<a class="headerlink" href="#choosing-an-integration-method" title="Link to this heading">#</a></h3>
<p>While integration methods have now been extensively benchmarked, an optimal method for all scenarios does not exist. Packages of integration performance metrics and evaluation pipelines like <a class="reference external" href="https://github.com/theislab/scib"><strong>scIB</strong></a> and <a class="reference external" href="https://github.com/cellgeni/batchbench"><strong>batchbench</strong></a> can be used to evaluate integration performance on your own data. However many metrics (particularly those that measure the conservation of biological variation) require ground-truth cell identity labels. Parameter optimization may tune many methods to work for particular tasks, yet in general, one can say that <strong>Harmony</strong> and <strong>Seurat</strong> consistently perform well for simple batch correction tasks, and <strong>scVI</strong>, <strong>scGen</strong>, <strong>scANVI</strong>, and <strong>Scanorama</strong> perform well for more complex data integration tasks. When choosing a method, we would recommend looking into these options first. Additionally, integration method choice may be guided by the required output data format  (i.e., do you need corrected gene expression data or does an integrated embedding suffice?). It would be prudent to test multiple methods and evaluate the outputs on the basis of quantitative definitions of success before selecting one. Extensive guidelines for data integration method choice can be found in <span id="id26">[<a class="reference internal" href="#id220" title="Malte D Luecken, M Büttner, K Chaichoompu, A Danese, M Interlandi, M F Mueller, D C Strobl, L Zappia, M Dugas, M Colomé-Tatché, and Fabian J Theis. Benchmarking atlas-level data integration in single-cell genomics. Nature methods, December 2021. doi:10.1038/s41592-021-01336-8.">Luecken <em>et al.</em>, 2021</a>]</span>.</p>
<p>In the rest of this chapter, we demonstrate some of the best-performing methods and quickly demonstrate how integration performance can be evaluated.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Python packages</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">anndata2ri</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">bbknn</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">matplotlib.pyplot</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">plt</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">numpy</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">np</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">pandas</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">pd</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">scanpy</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">sc</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">scib</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">scvi</span>

<span class="c1"># R interface</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">rpy2.robjects</span><span class="w"> </span><span class="kn">import</span> <span class="n">pandas2ri</span>

<span class="n">pandas2ri</span><span class="o">.</span><span class="n">activate</span><span class="p">()</span>
<span class="n">anndata2ri</span><span class="o">.</span><span class="n">activate</span><span class="p">()</span>

<span class="o">%</span><span class="k">load_ext</span> rpy2.ipython
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Global seed set to 0
/Users/luke.zappia/miniconda/envs/bp-integration/lib/python3.9/site-packages/pytorch_lightning/utilities/warnings.py:53: LightningDeprecationWarning: pytorch_lightning.utilities.warnings.rank_zero_deprecation has been deprecated in v1.6 and will be removed in v1.8. Use the equivalent function from the pytorch_lightning.utilities.rank_zero module instead.
  new_rank_zero_deprecation(
/Users/luke.zappia/miniconda/envs/bp-integration/lib/python3.9/site-packages/pytorch_lightning/utilities/warnings.py:58: LightningDeprecationWarning: The `pytorch_lightning.loggers.base.rank_zero_experiment` is deprecated in v1.7 and will be removed in v1.9. Please use `pytorch_lightning.loggers.logger.rank_zero_experiment` instead.
  return new_rank_zero_deprecation(*args, **kwargs)
/Users/luke.zappia/miniconda/envs/bp-integration/lib/python3.9/site-packages/rpy2/robjects/pandas2ri.py:263: DeprecationWarning: The global conversion available with activate() is deprecated and will be removed in the next major release. Use a local converter.
  warnings.warn(&#39;The global conversion available with activate() &#39;
/Users/luke.zappia/miniconda/envs/bp-integration/lib/python3.9/site-packages/rpy2/robjects/numpy2ri.py:205: DeprecationWarning: The global conversion available with activate() is deprecated and will be removed in the next major release. Use a local converter.
  warnings.warn(&#39;The global conversion available with activate() &#39;
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="o">%%</span><span class="k">R</span>
# R packages
library(Seurat)
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>    WARNING: The R package &quot;reticulate&quot; does not
    consider that it could be called from a Python process. This
    results in a quasi-obligatory segfault when rpy2 is evaluating
    R code using it. On the hand, rpy2 is accounting for the
    fact that it might already be running embedded in a Python
    process. This is why:
    - Python -&gt; rpy2 -&gt; R -&gt; reticulate: crashes
    - R -&gt; reticulate -&gt; Python -&gt; rpy2: works

    The issue with reticulate is tracked here:
    https://github.com/rstudio/reticulate/issues/208
    
</pre></div>
</div>
</div>
</div>
</section>
</section>
<section id="dataset">
<h2><span class="section-number">12.2. </span>Dataset<a class="headerlink" href="#dataset" title="Link to this heading">#</a></h2>
<p>The dataset we will use to demonstrate data integration contains several samples of bone marrow mononuclear cells. These samples were originally created for the Open Problems in Single-Cell Analysis <a class="reference external" href="https://openproblems.bio/neurips_2021/">NeurIPS Competition 2021</a> <span id="id27">[<a class="reference internal" href="#id225" title="Christopher Lance, Malte D Luecken, Daniel B Burkhardt, Robrecht Cannoodt, Pia Rautenstrauch, Anna Laddach, Aidyn Ubingazhibov, Zhi-Jie Cao, Kaiwen Deng, Sumeer Khan, Qiao Liu, Nikolay Russkikh, Gleb Ryazantsev, Uwe Ohler, Neurips 2021 Multimodal Data integration competition participants, Angela Oliveira Pisco, Jonathan Bloom, Smita Krishnaswamy, and Fabian J Theis. Multimodal single cell data integration challenge: Results and lessons learned. In Douwe Kiela, Marco Ciccone, and Barbara Caputo, editors, Proceedings of the NeurIPS 2021 Competitions and Demonstrations Track, volume 176 of Proceedings of Machine Learning Research, 162–176. PMLR, 2022.">Lance <em>et al.</em>, 2022</a>, <a class="reference internal" href="#id224" title="Malte D Luecken, Daniel Bernard Burkhardt, Robrecht Cannoodt, Christopher Lance, Aditi Agrawal, Hananeh Aliee, Ann T Chen, Louise Deconinck, Angela M Detweiler, Alejandro A Granados, Shelly Huynh, Laura Isacco, Yang Joon Kim, Dominik Klein, Bony de Kumar, Sunil Kuppasani, Heiko Lickert, Aaron McGeever, Honey Mekonen, Joaquin Caceres Melgarejo, Maurizio Morri, Michaela Müller, Norma Neff, Sheryl Paul, Bastian Rieck, Kaylie Schneider, Scott Steelman, Michael Sterr, Daniel J Treacy, Alexander Tong, Alexandra-Chloe Villani, Guilin Wang, Jia Yan, Ce Zhang, Angela Oliveira Pisco, Smita Krishnaswamy, Fabian J Theis, and Jonathan M Bloom. A sandbox for prediction and integration of DNA, RNA, and proteins in single cells. In Thirty-fifth Conference on Neural Information Processing Systems Datasets and Benchmarks Track (Round 2). January 2022.">Luecken <em>et al.</em>, 2022</a>]</span>. The <a class="reference external" href="https://www.10xgenomics.com/products/single-cell-multiome-atac-plus-gene-expression">10x Multiome</a> protocol was used which measures both RNA expression (scRNA-seq) and chromatin accessibility (scATAC-seq) in the same cells. The version of the data we use here was already pre-processed to remove low-quality cells.</p>
<p>Let’s read in the dataset using <strong>scanpy</strong> to get an <code class="docutils literal notranslate"><span class="pre">AnnData</span></code> object.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="o">!</span>wget<span class="w"> </span>https://figshare.com/ndownloader/files/45452260<span class="w"> </span>-O<span class="w"> </span>openproblems_bmmc_multiome_genes_filtered.h5ad
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">adata_raw</span> <span class="o">=</span> <span class="n">sc</span><span class="o">.</span><span class="n">read_h5ad</span><span class="p">(</span><span class="s2">&quot;openproblems_bmmc_multiome_genes_filtered.h5ad&quot;</span><span class="p">)</span>
<span class="n">adata_raw</span><span class="o">.</span><span class="n">layers</span><span class="p">[</span><span class="s2">&quot;logcounts&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">adata_raw</span><span class="o">.</span><span class="n">X</span>
<span class="n">adata_raw</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>/Users/luke.zappia/miniconda/envs/bp-integration/lib/python3.9/site-packages/anndata/_io/h5ad.py:238: OldFormatWarning: Element &#39;/layers&#39; was written without encoding metadata.
  d[k] = read_elem(f[k])
/Users/luke.zappia/miniconda/envs/bp-integration/lib/python3.9/site-packages/anndata/_io/specs/methods.py:590: OldFormatWarning: Element &#39;/obs/GEX_pct_counts_mt&#39; was written without encoding metadata.
  return read_elem(dataset)
/Users/luke.zappia/miniconda/envs/bp-integration/lib/python3.9/site-packages/anndata/_io/specs/methods.py:590: OldFormatWarning: Element &#39;/obs/GEX_n_counts&#39; was written without encoding metadata.
  return read_elem(dataset)
/Users/luke.zappia/miniconda/envs/bp-integration/lib/python3.9/site-packages/anndata/_io/specs/methods.py:590: OldFormatWarning: Element &#39;/obs/GEX_n_genes&#39; was written without encoding metadata.
  return read_elem(dataset)
/Users/luke.zappia/miniconda/envs/bp-integration/lib/python3.9/site-packages/anndata/_io/specs/methods.py:590: OldFormatWarning: Element &#39;/obs/GEX_size_factors&#39; was written without encoding metadata.
  return read_elem(dataset)
/Users/luke.zappia/miniconda/envs/bp-integration/lib/python3.9/site-packages/anndata/_io/specs/methods.py:584: OldFormatWarning: Element &#39;/obs/__categories/GEX_phase&#39; was written without encoding metadata.
  categories = read_elem(categories_dset)
/Users/luke.zappia/miniconda/envs/bp-integration/lib/python3.9/site-packages/anndata/_io/specs/methods.py:587: OldFormatWarning: Element &#39;/obs/GEX_phase&#39; was written without encoding metadata.
  read_elem(dataset), categories, ordered=ordered
/Users/luke.zappia/miniconda/envs/bp-integration/lib/python3.9/site-packages/anndata/_io/specs/methods.py:590: OldFormatWarning: Element &#39;/obs/ATAC_nCount_peaks&#39; was written without encoding metadata.
  return read_elem(dataset)
/Users/luke.zappia/miniconda/envs/bp-integration/lib/python3.9/site-packages/anndata/_io/specs/methods.py:590: OldFormatWarning: Element &#39;/obs/ATAC_atac_fragments&#39; was written without encoding metadata.
  return read_elem(dataset)
/Users/luke.zappia/miniconda/envs/bp-integration/lib/python3.9/site-packages/anndata/_io/specs/methods.py:590: OldFormatWarning: Element &#39;/obs/ATAC_reads_in_peaks_frac&#39; was written without encoding metadata.
  return read_elem(dataset)
/Users/luke.zappia/miniconda/envs/bp-integration/lib/python3.9/site-packages/anndata/_io/specs/methods.py:590: OldFormatWarning: Element &#39;/obs/ATAC_blacklist_fraction&#39; was written without encoding metadata.
  return read_elem(dataset)
/Users/luke.zappia/miniconda/envs/bp-integration/lib/python3.9/site-packages/anndata/_io/specs/methods.py:590: OldFormatWarning: Element &#39;/obs/ATAC_nucleosome_signal&#39; was written without encoding metadata.
  return read_elem(dataset)
/Users/luke.zappia/miniconda/envs/bp-integration/lib/python3.9/site-packages/anndata/_io/specs/methods.py:584: OldFormatWarning: Element &#39;/obs/__categories/cell_type&#39; was written without encoding metadata.
  categories = read_elem(categories_dset)
/Users/luke.zappia/miniconda/envs/bp-integration/lib/python3.9/site-packages/anndata/_io/specs/methods.py:587: OldFormatWarning: Element &#39;/obs/cell_type&#39; was written without encoding metadata.
  read_elem(dataset), categories, ordered=ordered
/Users/luke.zappia/miniconda/envs/bp-integration/lib/python3.9/site-packages/anndata/_io/specs/methods.py:584: OldFormatWarning: Element &#39;/obs/__categories/batch&#39; was written without encoding metadata.
  categories = read_elem(categories_dset)
/Users/luke.zappia/miniconda/envs/bp-integration/lib/python3.9/site-packages/anndata/_io/specs/methods.py:587: OldFormatWarning: Element &#39;/obs/batch&#39; was written without encoding metadata.
  read_elem(dataset), categories, ordered=ordered
/Users/luke.zappia/miniconda/envs/bp-integration/lib/python3.9/site-packages/anndata/_io/specs/methods.py:590: OldFormatWarning: Element &#39;/obs/ATAC_pseudotime_order&#39; was written without encoding metadata.
  return read_elem(dataset)
/Users/luke.zappia/miniconda/envs/bp-integration/lib/python3.9/site-packages/anndata/_io/specs/methods.py:590: OldFormatWarning: Element &#39;/obs/GEX_pseudotime_order&#39; was written without encoding metadata.
  return read_elem(dataset)
/Users/luke.zappia/miniconda/envs/bp-integration/lib/python3.9/site-packages/anndata/_io/specs/methods.py:584: OldFormatWarning: Element &#39;/obs/__categories/Samplename&#39; was written without encoding metadata.
  categories = read_elem(categories_dset)
/Users/luke.zappia/miniconda/envs/bp-integration/lib/python3.9/site-packages/anndata/_io/specs/methods.py:587: OldFormatWarning: Element &#39;/obs/Samplename&#39; was written without encoding metadata.
  read_elem(dataset), categories, ordered=ordered
/Users/luke.zappia/miniconda/envs/bp-integration/lib/python3.9/site-packages/anndata/_io/specs/methods.py:584: OldFormatWarning: Element &#39;/obs/__categories/Site&#39; was written without encoding metadata.
  categories = read_elem(categories_dset)
/Users/luke.zappia/miniconda/envs/bp-integration/lib/python3.9/site-packages/anndata/_io/specs/methods.py:587: OldFormatWarning: Element &#39;/obs/Site&#39; was written without encoding metadata.
  read_elem(dataset), categories, ordered=ordered
/Users/luke.zappia/miniconda/envs/bp-integration/lib/python3.9/site-packages/anndata/_io/specs/methods.py:584: OldFormatWarning: Element &#39;/obs/__categories/DonorNumber&#39; was written without encoding metadata.
  categories = read_elem(categories_dset)
/Users/luke.zappia/miniconda/envs/bp-integration/lib/python3.9/site-packages/anndata/_io/specs/methods.py:587: OldFormatWarning: Element &#39;/obs/DonorNumber&#39; was written without encoding metadata.
  read_elem(dataset), categories, ordered=ordered
/Users/luke.zappia/miniconda/envs/bp-integration/lib/python3.9/site-packages/anndata/_io/specs/methods.py:584: OldFormatWarning: Element &#39;/obs/__categories/Modality&#39; was written without encoding metadata.
  categories = read_elem(categories_dset)
/Users/luke.zappia/miniconda/envs/bp-integration/lib/python3.9/site-packages/anndata/_io/specs/methods.py:587: OldFormatWarning: Element &#39;/obs/Modality&#39; was written without encoding metadata.
  read_elem(dataset), categories, ordered=ordered
/Users/luke.zappia/miniconda/envs/bp-integration/lib/python3.9/site-packages/anndata/_io/specs/methods.py:590: OldFormatWarning: Element &#39;/obs/VendorLot&#39; was written without encoding metadata.
  return read_elem(dataset)
/Users/luke.zappia/miniconda/envs/bp-integration/lib/python3.9/site-packages/anndata/_io/specs/methods.py:590: OldFormatWarning: Element &#39;/obs/DonorID&#39; was written without encoding metadata.
  return read_elem(dataset)
/Users/luke.zappia/miniconda/envs/bp-integration/lib/python3.9/site-packages/anndata/_io/specs/methods.py:590: OldFormatWarning: Element &#39;/obs/DonorAge&#39; was written without encoding metadata.
  return read_elem(dataset)
/Users/luke.zappia/miniconda/envs/bp-integration/lib/python3.9/site-packages/anndata/_io/specs/methods.py:590: OldFormatWarning: Element &#39;/obs/DonorBMI&#39; was written without encoding metadata.
  return read_elem(dataset)
/Users/luke.zappia/miniconda/envs/bp-integration/lib/python3.9/site-packages/anndata/_io/specs/methods.py:584: OldFormatWarning: Element &#39;/obs/__categories/DonorBloodType&#39; was written without encoding metadata.
  categories = read_elem(categories_dset)
/Users/luke.zappia/miniconda/envs/bp-integration/lib/python3.9/site-packages/anndata/_io/specs/methods.py:587: OldFormatWarning: Element &#39;/obs/DonorBloodType&#39; was written without encoding metadata.
  read_elem(dataset), categories, ordered=ordered
/Users/luke.zappia/miniconda/envs/bp-integration/lib/python3.9/site-packages/anndata/_io/specs/methods.py:584: OldFormatWarning: Element &#39;/obs/__categories/DonorRace&#39; was written without encoding metadata.
  categories = read_elem(categories_dset)
/Users/luke.zappia/miniconda/envs/bp-integration/lib/python3.9/site-packages/anndata/_io/specs/methods.py:587: OldFormatWarning: Element &#39;/obs/DonorRace&#39; was written without encoding metadata.
  read_elem(dataset), categories, ordered=ordered
/Users/luke.zappia/miniconda/envs/bp-integration/lib/python3.9/site-packages/anndata/_io/specs/methods.py:584: OldFormatWarning: Element &#39;/obs/__categories/Ethnicity&#39; was written without encoding metadata.
  categories = read_elem(categories_dset)
/Users/luke.zappia/miniconda/envs/bp-integration/lib/python3.9/site-packages/anndata/_io/specs/methods.py:587: OldFormatWarning: Element &#39;/obs/Ethnicity&#39; was written without encoding metadata.
  read_elem(dataset), categories, ordered=ordered
/Users/luke.zappia/miniconda/envs/bp-integration/lib/python3.9/site-packages/anndata/_io/specs/methods.py:584: OldFormatWarning: Element &#39;/obs/__categories/DonorGender&#39; was written without encoding metadata.
  categories = read_elem(categories_dset)
/Users/luke.zappia/miniconda/envs/bp-integration/lib/python3.9/site-packages/anndata/_io/specs/methods.py:587: OldFormatWarning: Element &#39;/obs/DonorGender&#39; was written without encoding metadata.
  read_elem(dataset), categories, ordered=ordered
/Users/luke.zappia/miniconda/envs/bp-integration/lib/python3.9/site-packages/anndata/_io/specs/methods.py:584: OldFormatWarning: Element &#39;/obs/__categories/QCMeds&#39; was written without encoding metadata.
  categories = read_elem(categories_dset)
/Users/luke.zappia/miniconda/envs/bp-integration/lib/python3.9/site-packages/anndata/_io/specs/methods.py:587: OldFormatWarning: Element &#39;/obs/QCMeds&#39; was written without encoding metadata.
  read_elem(dataset), categories, ordered=ordered
/Users/luke.zappia/miniconda/envs/bp-integration/lib/python3.9/site-packages/anndata/_io/specs/methods.py:584: OldFormatWarning: Element &#39;/obs/__categories/DonorSmoker&#39; was written without encoding metadata.
  categories = read_elem(categories_dset)
/Users/luke.zappia/miniconda/envs/bp-integration/lib/python3.9/site-packages/anndata/_io/specs/methods.py:587: OldFormatWarning: Element &#39;/obs/DonorSmoker&#39; was written without encoding metadata.
  read_elem(dataset), categories, ordered=ordered
/Users/luke.zappia/miniconda/envs/bp-integration/lib/python3.9/site-packages/anndata/_io/specs/methods.py:590: OldFormatWarning: Element &#39;/obs/_index&#39; was written without encoding metadata.
  return read_elem(dataset)
/Users/luke.zappia/miniconda/envs/bp-integration/lib/python3.9/site-packages/anndata/_io/h5ad.py:238: OldFormatWarning: Element &#39;/obsm&#39; was written without encoding metadata.
  d[k] = read_elem(f[k])
/Users/luke.zappia/miniconda/envs/bp-integration/lib/python3.9/site-packages/anndata/_io/specs/methods.py:92: OldFormatWarning: Element &#39;/obsm/ATAC_lsi_full&#39; was written without encoding metadata.
  return {k: read_elem(v) for k, v in elem.items()}
/Users/luke.zappia/miniconda/envs/bp-integration/lib/python3.9/site-packages/anndata/_io/specs/methods.py:92: OldFormatWarning: Element &#39;/obsm/ATAC_lsi_red&#39; was written without encoding metadata.
  return {k: read_elem(v) for k, v in elem.items()}
/Users/luke.zappia/miniconda/envs/bp-integration/lib/python3.9/site-packages/anndata/_io/specs/methods.py:92: OldFormatWarning: Element &#39;/obsm/ATAC_umap&#39; was written without encoding metadata.
  return {k: read_elem(v) for k, v in elem.items()}
/Users/luke.zappia/miniconda/envs/bp-integration/lib/python3.9/site-packages/anndata/_io/specs/methods.py:92: OldFormatWarning: Element &#39;/obsm/GEX_X_pca&#39; was written without encoding metadata.
  return {k: read_elem(v) for k, v in elem.items()}
/Users/luke.zappia/miniconda/envs/bp-integration/lib/python3.9/site-packages/anndata/_io/specs/methods.py:92: OldFormatWarning: Element &#39;/obsm/GEX_X_umap&#39; was written without encoding metadata.
  return {k: read_elem(v) for k, v in elem.items()}
/Users/luke.zappia/miniconda/envs/bp-integration/lib/python3.9/site-packages/anndata/_io/h5ad.py:238: OldFormatWarning: Element &#39;/uns&#39; was written without encoding metadata.
  d[k] = read_elem(f[k])
/Users/luke.zappia/miniconda/envs/bp-integration/lib/python3.9/site-packages/anndata/_io/specs/methods.py:92: OldFormatWarning: Element &#39;/uns/ATAC_gene_activity_var_names&#39; was written without encoding metadata.
  return {k: read_elem(v) for k, v in elem.items()}
/Users/luke.zappia/miniconda/envs/bp-integration/lib/python3.9/site-packages/anndata/_io/specs/methods.py:92: OldFormatWarning: Element &#39;/uns/dataset_id&#39; was written without encoding metadata.
  return {k: read_elem(v) for k, v in elem.items()}
/Users/luke.zappia/miniconda/envs/bp-integration/lib/python3.9/site-packages/anndata/_io/specs/methods.py:92: OldFormatWarning: Element &#39;/uns/genome&#39; was written without encoding metadata.
  return {k: read_elem(v) for k, v in elem.items()}
/Users/luke.zappia/miniconda/envs/bp-integration/lib/python3.9/site-packages/anndata/_io/specs/methods.py:92: OldFormatWarning: Element &#39;/uns/organism&#39; was written without encoding metadata.
  return {k: read_elem(v) for k, v in elem.items()}
/Users/luke.zappia/miniconda/envs/bp-integration/lib/python3.9/site-packages/anndata/_io/specs/methods.py:584: OldFormatWarning: Element &#39;/var/__categories/feature_types&#39; was written without encoding metadata.
  categories = read_elem(categories_dset)
/Users/luke.zappia/miniconda/envs/bp-integration/lib/python3.9/site-packages/anndata/_io/specs/methods.py:587: OldFormatWarning: Element &#39;/var/feature_types&#39; was written without encoding metadata.
  read_elem(dataset), categories, ordered=ordered
/Users/luke.zappia/miniconda/envs/bp-integration/lib/python3.9/site-packages/anndata/_io/specs/methods.py:584: OldFormatWarning: Element &#39;/var/__categories/gene_id&#39; was written without encoding metadata.
  categories = read_elem(categories_dset)
/Users/luke.zappia/miniconda/envs/bp-integration/lib/python3.9/site-packages/anndata/_io/specs/methods.py:587: OldFormatWarning: Element &#39;/var/gene_id&#39; was written without encoding metadata.
  read_elem(dataset), categories, ordered=ordered
/Users/luke.zappia/miniconda/envs/bp-integration/lib/python3.9/site-packages/anndata/_io/specs/methods.py:590: OldFormatWarning: Element &#39;/var/_index&#39; was written without encoding metadata.
  return read_elem(dataset)
</pre></div>
</div>
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>AnnData object with n_obs × n_vars = 69249 × 129921
    obs: &#39;GEX_pct_counts_mt&#39;, &#39;GEX_n_counts&#39;, &#39;GEX_n_genes&#39;, &#39;GEX_size_factors&#39;, &#39;GEX_phase&#39;, &#39;ATAC_nCount_peaks&#39;, &#39;ATAC_atac_fragments&#39;, &#39;ATAC_reads_in_peaks_frac&#39;, &#39;ATAC_blacklist_fraction&#39;, &#39;ATAC_nucleosome_signal&#39;, &#39;cell_type&#39;, &#39;batch&#39;, &#39;ATAC_pseudotime_order&#39;, &#39;GEX_pseudotime_order&#39;, &#39;Samplename&#39;, &#39;Site&#39;, &#39;DonorNumber&#39;, &#39;Modality&#39;, &#39;VendorLot&#39;, &#39;DonorID&#39;, &#39;DonorAge&#39;, &#39;DonorBMI&#39;, &#39;DonorBloodType&#39;, &#39;DonorRace&#39;, &#39;Ethnicity&#39;, &#39;DonorGender&#39;, &#39;QCMeds&#39;, &#39;DonorSmoker&#39;
    var: &#39;feature_types&#39;, &#39;gene_id&#39;
    uns: &#39;ATAC_gene_activity_var_names&#39;, &#39;dataset_id&#39;, &#39;genome&#39;, &#39;organism&#39;
    obsm: &#39;ATAC_gene_activity&#39;, &#39;ATAC_lsi_full&#39;, &#39;ATAC_lsi_red&#39;, &#39;ATAC_umap&#39;, &#39;GEX_X_pca&#39;, &#39;GEX_X_umap&#39;
    layers: &#39;counts&#39;, &#39;logcounts&#39;
</pre></div>
</div>
</div>
</div>
<p>The full dataset contains 69,249 cells and measurements for 129,921 features. There are two versions of the expression matrix, <code class="docutils literal notranslate"><span class="pre">counts</span></code> which contains the raw count values and <code class="docutils literal notranslate"><span class="pre">logcounts</span></code> which contains normalised log counts (these values are also stored in <code class="docutils literal notranslate"><span class="pre">adata.X</span></code>).</p>
<p>The <code class="docutils literal notranslate"><span class="pre">obs</span></code> slot contains several variables, some of which were calculated during pre-processing (for quality control) and others that contain metadata about the samples. The ones we are interested in here are:</p>
<ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">cell_type</span></code> - The annotated label for each cell</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">batch</span></code> - The sequencing batch for each cell</p></li>
</ul>
<p>For a real analysis it would be important to consider more variables but to keep it simple here we will only look at these.</p>
<p>We define variables to hold these names so that it is clear how we are using them in the code. This also helps with reproducibility because if we decided to change one of them for whatever reason we can be sure it has changed in the whole notebook.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">label_key</span> <span class="o">=</span> <span class="s2">&quot;cell_type&quot;</span>
<span class="n">batch_key</span> <span class="o">=</span> <span class="s2">&quot;batch&quot;</span>
</pre></div>
</div>
</div>
</div>
<div class="admonition-what-to-use-as-the-batch-label admonition">
<p class="admonition-title">What to use as the batch label?</p>
<p>As mentioned above, deciding what to use as a “batch” for data integration is one of the central choices when integrating your data. The most common approach is to define each sample as a batch (as we have here) which generally produces the strongest batch correction. However, samples are usually confounded with biological factors that you may want to preserve. For example, imagine an experiment that took samples from two locations in a tissue. If samples are considered as batches then data integration methods will attempt to remove differences between them and therefore differences between the locations. In this case, it may be more appropriate to use the donor as the batch to remove differences between individuals but not between locations. The planned analysis should also be considered. In cases where you are integrating many datasets and want to capture differences between individuals, the dataset may be a useful batch covariate. In our example, it may be better to have consistent cell type labels for the two locations and then test for differences between them than to have separate clusters for each location which need to be annotated separately and then matched.</p>
<p>The issue of confounding between samples and biological factors can be mitigated through careful experimental design that minimizes the batch effect overall by using multiplexing techniques that allow biological samples to be combined into a single sequencing sample. However, this is not always possible and requires both extra processing in the lab and as well as extra computational steps.</p>
</div>
<p>Let’s have a look at the different batches and how many cells we have for each.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">adata_raw</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="n">batch_key</span><span class="p">]</span><span class="o">.</span><span class="n">value_counts</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>s4d8     9876
s4d1     8023
s3d10    6781
s1d2     6740
s1d1     6224
s2d4     6111
s2d5     4895
s3d3     4325
s4d9     4325
s1d3     4279
s2d1     4220
s3d7     1771
s3d6     1679
Name: batch, dtype: int64
</pre></div>
</div>
</div>
</div>
<p>There are 13 different batches in the dataset. During this experiment, multiple samples were taken from a set of donors and sequenced at different facilities so the names here are a combination of the sample number (eg. “s1”) and the donor (eg. “d2”). For simplicity, and to reduce computational time, we will select three samples to use.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">keep_batches</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;s1d3&quot;</span><span class="p">,</span> <span class="s2">&quot;s2d1&quot;</span><span class="p">,</span> <span class="s2">&quot;s3d7&quot;</span><span class="p">]</span>
<span class="n">adata</span> <span class="o">=</span> <span class="n">adata_raw</span><span class="p">[</span><span class="n">adata_raw</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="n">batch_key</span><span class="p">]</span><span class="o">.</span><span class="n">isin</span><span class="p">(</span><span class="n">keep_batches</span><span class="p">)]</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
<span class="n">adata</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>AnnData object with n_obs × n_vars = 10270 × 129921
    obs: &#39;GEX_pct_counts_mt&#39;, &#39;GEX_n_counts&#39;, &#39;GEX_n_genes&#39;, &#39;GEX_size_factors&#39;, &#39;GEX_phase&#39;, &#39;ATAC_nCount_peaks&#39;, &#39;ATAC_atac_fragments&#39;, &#39;ATAC_reads_in_peaks_frac&#39;, &#39;ATAC_blacklist_fraction&#39;, &#39;ATAC_nucleosome_signal&#39;, &#39;cell_type&#39;, &#39;batch&#39;, &#39;ATAC_pseudotime_order&#39;, &#39;GEX_pseudotime_order&#39;, &#39;Samplename&#39;, &#39;Site&#39;, &#39;DonorNumber&#39;, &#39;Modality&#39;, &#39;VendorLot&#39;, &#39;DonorID&#39;, &#39;DonorAge&#39;, &#39;DonorBMI&#39;, &#39;DonorBloodType&#39;, &#39;DonorRace&#39;, &#39;Ethnicity&#39;, &#39;DonorGender&#39;, &#39;QCMeds&#39;, &#39;DonorSmoker&#39;
    var: &#39;feature_types&#39;, &#39;gene_id&#39;
    uns: &#39;ATAC_gene_activity_var_names&#39;, &#39;dataset_id&#39;, &#39;genome&#39;, &#39;organism&#39;
    obsm: &#39;ATAC_gene_activity&#39;, &#39;ATAC_lsi_full&#39;, &#39;ATAC_lsi_red&#39;, &#39;ATAC_umap&#39;, &#39;GEX_X_pca&#39;, &#39;GEX_X_umap&#39;
    layers: &#39;counts&#39;, &#39;logcounts&#39;
</pre></div>
</div>
</div>
</div>
<p>After subsetting to select these batches we are left with 10,270 cells.</p>
<p>We have two annotations for the features stored in <code class="docutils literal notranslate"><span class="pre">var</span></code>:</p>
<ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">feature_types</span></code> - The type of each feature (RNA or ATAC)</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">gene_id</span></code> - The gene associated with each feature</p></li>
</ul>
<p>Let’s have a look at the feature types.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">adata</span><span class="o">.</span><span class="n">var</span><span class="p">[</span><span class="s2">&quot;feature_types&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">value_counts</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>ATAC    116490
GEX      13431
Name: feature_types, dtype: int64
</pre></div>
</div>
</div>
</div>
<p>We can see that there are over 100,000 ATAC features but only around 13,000 gene expression (“GEX”) features. Integration of multiple modalities is a complex problem that will be described in the <a class="reference internal" href="../multimodal_integration/advanced_integration.html#multimodal-integration-advanced-integration"><span class="std std-ref">multimodal integration chapter</span></a>, so for now we will subset to only the gene expression features. We also perform simple filtering to make sure we have no features with zero counts (this is necessary because by selecting a subset of samples we may have removed all the cells which expressed a particular feature).</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">adata</span> <span class="o">=</span> <span class="n">adata</span><span class="p">[:,</span> <span class="n">adata</span><span class="o">.</span><span class="n">var</span><span class="p">[</span><span class="s2">&quot;feature_types&quot;</span><span class="p">]</span> <span class="o">==</span> <span class="s2">&quot;GEX&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
<span class="n">sc</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">filter_genes</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">min_cells</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
<span class="n">adata</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>AnnData object with n_obs × n_vars = 10270 × 13431
    obs: &#39;GEX_pct_counts_mt&#39;, &#39;GEX_n_counts&#39;, &#39;GEX_n_genes&#39;, &#39;GEX_size_factors&#39;, &#39;GEX_phase&#39;, &#39;ATAC_nCount_peaks&#39;, &#39;ATAC_atac_fragments&#39;, &#39;ATAC_reads_in_peaks_frac&#39;, &#39;ATAC_blacklist_fraction&#39;, &#39;ATAC_nucleosome_signal&#39;, &#39;cell_type&#39;, &#39;batch&#39;, &#39;ATAC_pseudotime_order&#39;, &#39;GEX_pseudotime_order&#39;, &#39;Samplename&#39;, &#39;Site&#39;, &#39;DonorNumber&#39;, &#39;Modality&#39;, &#39;VendorLot&#39;, &#39;DonorID&#39;, &#39;DonorAge&#39;, &#39;DonorBMI&#39;, &#39;DonorBloodType&#39;, &#39;DonorRace&#39;, &#39;Ethnicity&#39;, &#39;DonorGender&#39;, &#39;QCMeds&#39;, &#39;DonorSmoker&#39;
    var: &#39;feature_types&#39;, &#39;gene_id&#39;, &#39;n_cells&#39;
    uns: &#39;ATAC_gene_activity_var_names&#39;, &#39;dataset_id&#39;, &#39;genome&#39;, &#39;organism&#39;
    obsm: &#39;ATAC_gene_activity&#39;, &#39;ATAC_lsi_full&#39;, &#39;ATAC_lsi_red&#39;, &#39;ATAC_umap&#39;, &#39;GEX_X_pca&#39;, &#39;GEX_X_umap&#39;
    layers: &#39;counts&#39;, &#39;logcounts&#39;
</pre></div>
</div>
</div>
</div>
<p>Because of the subsetting we also need to re-normalise the data. Here we just normalise using global scaling by the total counts per cell.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">adata</span><span class="o">.</span><span class="n">X</span> <span class="o">=</span> <span class="n">adata</span><span class="o">.</span><span class="n">layers</span><span class="p">[</span><span class="s2">&quot;counts&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
<span class="n">sc</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">normalize_total</span><span class="p">(</span><span class="n">adata</span><span class="p">)</span>
<span class="n">sc</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">log1p</span><span class="p">(</span><span class="n">adata</span><span class="p">)</span>
<span class="n">adata</span><span class="o">.</span><span class="n">layers</span><span class="p">[</span><span class="s2">&quot;logcounts&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">adata</span><span class="o">.</span><span class="n">X</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
</pre></div>
</div>
</div>
</div>
<p>We will use this dataset to demonstrate integration.</p>
<p>Most integration methods require a single object containing all the samples and a batch variable (like we have here). If instead, you have separate objects for each of your samples you can join them using the <strong>anndata</strong> <code class="docutils literal notranslate"><span class="pre">concat()</span></code> function. See the <a class="reference external" href="https://anndata.readthedocs.io/en/stable/concatenation.html">concatenation tutorial</a> for more details. Similar functionality exists in other ecosystems.</p>
<div class="admonition-integrating-umi-and-full-length-data admonition">
<p class="admonition-title">Integrating UMI and full-length data</p>
<p>Integrating samples from UMI and full-length protocols can present additional challenges. This is because full-length protocols are affected by gene-length bias (longer genes will be more highly expressed) while UMI data is not <span id="id28">[<a class="reference internal" href="#id216" title="Belinda Phipson, Luke Zappia, and Alicia Oshlack. Gene length and detection bias in single cell RNA sequencing protocols. F1000Research, April 2017. doi:10.12688/f1000research.11290.1.">Phipson <em>et al.</em>, 2017</a>]</span>. Because of this, it is generally recommended to transform counts for full-length samples into a unit that corrects for gene-length (such as transcripts per million (TPM) <span id="id29">[<a class="reference internal" href="#id217" title="Günter P Wagner, Koryu Kin, and Vincent J Lynch. Measurement of mRNA abundance using RNA-seq data: RPKM measure is inconsistent among samples. Theory in Biosciences, 131(4):281–285, December 2012. doi:10.1007/s12064-012-0162-3.">Wagner <em>et al.</em>, 2012</a>]</span>) before attempting integration. This isn’t necessary however if all the samples being integrated used a full-length protocol.</p>
</div>
</section>
<section id="unintegrated-data">
<h2><span class="section-number">12.3. </span>Unintegrated data<a class="headerlink" href="#unintegrated-data" title="Link to this heading">#</a></h2>
<p>It is always recommended to look at the raw data before performing any integration. This can give some indication of how big any batch effects are and what might be causing them (and therefore which variables to consider as the batch label). For some experiments, it might even suggest that integration is not required if samples already overlap. This is not uncommon for mouse or cell line studies from a single lab for example, where most of the variables which contribute to batch effects can be controlled (i.e. the batch correction setting).</p>
<p>We will perform highly variable gene (HVG) selection, PCA and UMAP dimensionality reduction as we have seen in previous chapters.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">sc</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">highly_variable_genes</span><span class="p">(</span><span class="n">adata</span><span class="p">)</span>
<span class="n">sc</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">pca</span><span class="p">(</span><span class="n">adata</span><span class="p">)</span>
<span class="n">sc</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">neighbors</span><span class="p">(</span><span class="n">adata</span><span class="p">)</span>
<span class="n">sc</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">umap</span><span class="p">(</span><span class="n">adata</span><span class="p">)</span>
<span class="n">adata</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>AnnData object with n_obs × n_vars = 10270 × 13431
    obs: &#39;GEX_pct_counts_mt&#39;, &#39;GEX_n_counts&#39;, &#39;GEX_n_genes&#39;, &#39;GEX_size_factors&#39;, &#39;GEX_phase&#39;, &#39;ATAC_nCount_peaks&#39;, &#39;ATAC_atac_fragments&#39;, &#39;ATAC_reads_in_peaks_frac&#39;, &#39;ATAC_blacklist_fraction&#39;, &#39;ATAC_nucleosome_signal&#39;, &#39;cell_type&#39;, &#39;batch&#39;, &#39;ATAC_pseudotime_order&#39;, &#39;GEX_pseudotime_order&#39;, &#39;Samplename&#39;, &#39;Site&#39;, &#39;DonorNumber&#39;, &#39;Modality&#39;, &#39;VendorLot&#39;, &#39;DonorID&#39;, &#39;DonorAge&#39;, &#39;DonorBMI&#39;, &#39;DonorBloodType&#39;, &#39;DonorRace&#39;, &#39;Ethnicity&#39;, &#39;DonorGender&#39;, &#39;QCMeds&#39;, &#39;DonorSmoker&#39;
    var: &#39;feature_types&#39;, &#39;gene_id&#39;, &#39;n_cells&#39;, &#39;highly_variable&#39;, &#39;means&#39;, &#39;dispersions&#39;, &#39;dispersions_norm&#39;
    uns: &#39;ATAC_gene_activity_var_names&#39;, &#39;dataset_id&#39;, &#39;genome&#39;, &#39;organism&#39;, &#39;log1p&#39;, &#39;hvg&#39;, &#39;pca&#39;, &#39;neighbors&#39;, &#39;umap&#39;
    obsm: &#39;ATAC_gene_activity&#39;, &#39;ATAC_lsi_full&#39;, &#39;ATAC_lsi_red&#39;, &#39;ATAC_umap&#39;, &#39;GEX_X_pca&#39;, &#39;GEX_X_umap&#39;, &#39;X_pca&#39;, &#39;X_umap&#39;
    varm: &#39;PCs&#39;
    layers: &#39;counts&#39;, &#39;logcounts&#39;
    obsp: &#39;distances&#39;, &#39;connectivities&#39;
</pre></div>
</div>
</div>
</div>
<p>This adds several new items to our AnnData object. The <code class="docutils literal notranslate"><span class="pre">var</span></code> slot now includes means, dispersions and the selected variable genes. In the <code class="docutils literal notranslate"><span class="pre">obsp</span></code> slot we have distances and connectivities for our KNN graph and in <code class="docutils literal notranslate"><span class="pre">obsm</span></code> are the PCA and UMAP embeddings.</p>
<p>Let’s plot the UMAP, colouring the points by cell identity and batch labels. If the dataset had not already been labelled (which is often the case) we would only be able to consider the batch labels.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">adata</span><span class="o">.</span><span class="n">uns</span><span class="p">[</span><span class="n">batch_key</span> <span class="o">+</span> <span class="s2">&quot;_colors&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="p">[</span>
    <span class="s2">&quot;#1b9e77&quot;</span><span class="p">,</span>
    <span class="s2">&quot;#d95f02&quot;</span><span class="p">,</span>
    <span class="s2">&quot;#7570b3&quot;</span><span class="p">,</span>
<span class="p">]</span>  <span class="c1"># Set custom colours for batches</span>
<span class="n">sc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">umap</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="p">[</span><span class="n">label_key</span><span class="p">,</span> <span class="n">batch_key</span><span class="p">],</span> <span class="n">wspace</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>/Users/luke.zappia/miniconda/envs/bp-integration/lib/python3.9/site-packages/scanpy/plotting/_tools/scatterplots.py:163: PendingDeprecationWarning: The get_cmap function will be deprecated in a future version. Use ``matplotlib.colormaps[name]`` or ``matplotlib.colormaps.get_cmap(obj)`` instead.
  cmap = copy(get_cmap(cmap))
/Users/luke.zappia/miniconda/envs/bp-integration/lib/python3.9/site-packages/scanpy/plotting/_tools/scatterplots.py:392: UserWarning: No data for colormapping provided via &#39;c&#39;. Parameters &#39;cmap&#39; will be ignored
  cax = scatter(
/Users/luke.zappia/miniconda/envs/bp-integration/lib/python3.9/site-packages/scanpy/plotting/_tools/scatterplots.py:392: UserWarning: No data for colormapping provided via &#39;c&#39;. Parameters &#39;cmap&#39; will be ignored
  cax = scatter(
</pre></div>
</div>
<img alt="../_images/754e7fe740ab1d359b5bc5245d857a5c5ccc2675f7acc1e41e2679d9c72fc1f9.png" src="../_images/754e7fe740ab1d359b5bc5245d857a5c5ccc2675f7acc1e41e2679d9c72fc1f9.png" />
</div>
</div>
<p>Often when looking at these plots you will see a clear separation between batches. In this case, what we see is more subtle and while cells from the same label are generally near each other there is a shift between batches. If we were to perform a clustering analysis using this raw data we would probably end up with some clusters containing a single batch which would be difficult to interpret at the annotation stage. We are also likely to overlook rare cell types which are not common enough in any single sample to produce their own cluster. While UMAPs can often display batch effects, as always when considering these 2D representations it is important not to overinterpret them. For a real analysis, you should confirm the integration in other ways such as by inspecting the distribution of marker genes. In the “Benchmarking your own integration” section below we discuss metrics for quantifying the quality of an integration.</p>
<p>Now that we have confirmed there are batch effects to correct we can move on to the different integration methods. If the batches perfectly overlaid each other or we could discover meaningful cell clusters without correction then there would be no need to perform integration.</p>
</section>
<section id="batch-aware-feature-selection">
<h2><span class="section-number">12.4. </span>Batch-aware feature selection<a class="headerlink" href="#batch-aware-feature-selection" title="Link to this heading">#</a></h2>
<p>As shown in <a class="reference internal" href="../preprocessing_visualization/feature_selection.html#pre-processing-feature-selection"><span class="std std-ref">previous chapters</span></a> we often select a subset of genes to use for our analysis in order to reduce noise and processing time. We do the same thing when we have multiple samples, however, it is important that gene selection is performed in a batch-aware way. This is because genes that are variable across the whole dataset could be capturing batch effects rather than the biological signals we are interested in. It also helps to select genes relevant to rare cell identities, for example, if an identity is only present in one sample then markers for it may not be variable across all the samples but should be in that one sample.</p>
<p>We can perform batch-aware highly variable gene selection by setting the <code class="docutils literal notranslate"><span class="pre">batch_key</span></code> argument in the <strong>scanpy</strong> <code class="docutils literal notranslate"><span class="pre">highly_variable_genes()</span></code> function. <strong>scanpy</strong> will then calculate HVGs for each batch separately and combine the results by selecting those genes that are highly variable in the highest number of batches. We use the <strong>scanpy</strong> function here because it has this batch awareness built in. For other methods, we would have to run them on each batch individually and then manually combine the results.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">sc</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">highly_variable_genes</span><span class="p">(</span>
    <span class="n">adata</span><span class="p">,</span> <span class="n">n_top_genes</span><span class="o">=</span><span class="mi">2000</span><span class="p">,</span> <span class="n">flavor</span><span class="o">=</span><span class="s2">&quot;cell_ranger&quot;</span><span class="p">,</span> <span class="n">batch_key</span><span class="o">=</span><span class="n">batch_key</span>
<span class="p">)</span>
<span class="n">adata</span><span class="o">.</span><span class="n">var</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>/Users/luke.zappia/miniconda/envs/bp-integration/lib/python3.9/site-packages/scanpy/preprocessing/_highly_variable_genes.py:478: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.
  hvg = hvg.append(missing_hvg, ignore_index=True)
/Users/luke.zappia/miniconda/envs/bp-integration/lib/python3.9/site-packages/scanpy/preprocessing/_highly_variable_genes.py:478: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.
  hvg = hvg.append(missing_hvg, ignore_index=True)
/Users/luke.zappia/miniconda/envs/bp-integration/lib/python3.9/site-packages/scanpy/preprocessing/_highly_variable_genes.py:478: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.
  hvg = hvg.append(missing_hvg, ignore_index=True)
</pre></div>
</div>
<div class="output text_html"><div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>feature_types</th>
      <th>gene_id</th>
      <th>n_cells</th>
      <th>highly_variable</th>
      <th>means</th>
      <th>dispersions</th>
      <th>dispersions_norm</th>
      <th>highly_variable_nbatches</th>
      <th>highly_variable_intersection</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>AL627309.5</th>
      <td>GEX</td>
      <td>ENSG00000241860</td>
      <td>112</td>
      <td>False</td>
      <td>0.006533</td>
      <td>0.775289</td>
      <td>0.357973</td>
      <td>1</td>
      <td>False</td>
    </tr>
    <tr>
      <th>LINC01409</th>
      <td>GEX</td>
      <td>ENSG00000237491</td>
      <td>422</td>
      <td>False</td>
      <td>0.024462</td>
      <td>0.716935</td>
      <td>-0.126119</td>
      <td>0</td>
      <td>False</td>
    </tr>
    <tr>
      <th>LINC01128</th>
      <td>GEX</td>
      <td>ENSG00000228794</td>
      <td>569</td>
      <td>False</td>
      <td>0.030714</td>
      <td>0.709340</td>
      <td>-0.296701</td>
      <td>0</td>
      <td>False</td>
    </tr>
    <tr>
      <th>NOC2L</th>
      <td>GEX</td>
      <td>ENSG00000188976</td>
      <td>675</td>
      <td>False</td>
      <td>0.037059</td>
      <td>0.704363</td>
      <td>-0.494025</td>
      <td>0</td>
      <td>False</td>
    </tr>
    <tr>
      <th>KLHL17</th>
      <td>GEX</td>
      <td>ENSG00000187961</td>
      <td>88</td>
      <td>False</td>
      <td>0.005295</td>
      <td>0.721757</td>
      <td>-0.028456</td>
      <td>0</td>
      <td>False</td>
    </tr>
    <tr>
      <th>...</th>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
    </tr>
    <tr>
      <th>MT-ND5</th>
      <td>GEX</td>
      <td>ENSG00000198786</td>
      <td>4056</td>
      <td>False</td>
      <td>0.269375</td>
      <td>0.645837</td>
      <td>-0.457491</td>
      <td>0</td>
      <td>False</td>
    </tr>
    <tr>
      <th>MT-ND6</th>
      <td>GEX</td>
      <td>ENSG00000198695</td>
      <td>1102</td>
      <td>False</td>
      <td>0.051506</td>
      <td>0.697710</td>
      <td>-0.248421</td>
      <td>0</td>
      <td>False</td>
    </tr>
    <tr>
      <th>MT-CYB</th>
      <td>GEX</td>
      <td>ENSG00000198727</td>
      <td>5647</td>
      <td>False</td>
      <td>0.520368</td>
      <td>0.613233</td>
      <td>-0.441362</td>
      <td>0</td>
      <td>False</td>
    </tr>
    <tr>
      <th>AL592183.1</th>
      <td>GEX</td>
      <td>ENSG00000273748</td>
      <td>732</td>
      <td>False</td>
      <td>0.047486</td>
      <td>0.753417</td>
      <td>0.413699</td>
      <td>0</td>
      <td>False</td>
    </tr>
    <tr>
      <th>AC240274.1</th>
      <td>GEX</td>
      <td>ENSG00000271254</td>
      <td>43</td>
      <td>False</td>
      <td>0.001871</td>
      <td>0.413772</td>
      <td>-0.386635</td>
      <td>0</td>
      <td>False</td>
    </tr>
  </tbody>
</table>
<p>13431 rows × 9 columns</p>
</div></div></div>
</div>
<p>We can see there are now some additional columns in <code class="docutils literal notranslate"><span class="pre">var</span></code>:</p>
<ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">highly_variable_nbatches</span></code> - The number of batches where each gene was found to be highly variable</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">highly_variable_intersection</span></code> - Whether each gene was highly variable in every batch</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">highly_variable</span></code> - Whether each gene was selected as highly variable after combining the results from each batch</p></li>
</ul>
<p>Let’s check how many batches each gene was variable in:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">n_batches</span> <span class="o">=</span> <span class="n">adata</span><span class="o">.</span><span class="n">var</span><span class="p">[</span><span class="s2">&quot;highly_variable_nbatches&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">value_counts</span><span class="p">()</span>
<span class="n">ax</span> <span class="o">=</span> <span class="n">n_batches</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">kind</span><span class="o">=</span><span class="s2">&quot;bar&quot;</span><span class="p">)</span>
<span class="n">n_batches</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>0    9931
1    1824
2     852
3     824
Name: highly_variable_nbatches, dtype: int64
</pre></div>
</div>
<img alt="../_images/41d44028f52454e07da80ad3afdf65b386f5951c3dba3f5e92afea0c8803af23.png" src="../_images/41d44028f52454e07da80ad3afdf65b386f5951c3dba3f5e92afea0c8803af23.png" />
</div>
</div>
<p>The first thing we notice is that most genes are not highly variable. This is typically the case but it can depend on how different the samples we are trying to integrate are. The overlap then decreases as we add more samples, with relatively few genes being highly variable in all three batches. By selecting the top 2000 genes we have selected all HVGs that are present in two or three batches and most of those that are present in one batch.</p>
<div class="admonition-how-many-genes-to-use admonition">
<p class="admonition-title">How many genes to use?</p>
<p>This is a question that doesn’t have a clear answer. The authors of the <strong>scvi-tools</strong> package which we use below recommend between 1000 and 10000 genes but how many depends on the context including the complexity of the dataset and the number of batches. A survey from a previous best practices paper <span id="id30">[<a class="reference internal" href="#id218" title="Malte D Luecken and Fabian J Theis. Current best practices in single‐cell RNA‐seq analysis: a tutorial. Molecular systems biology, June 2019. doi:10.15252/msb.20188746.">Luecken and Theis, 2019</a>]</span> indicated people typically use between 1000 and 6000 HVGs in a standard analysis. While selecting fewer genes can aid in the removal of batch effects <span id="id31">[<a class="reference internal" href="#id220" title="Malte D Luecken, M Büttner, K Chaichoompu, A Danese, M Interlandi, M F Mueller, D C Strobl, L Zappia, M Dugas, M Colomé-Tatché, and Fabian J Theis. Benchmarking atlas-level data integration in single-cell genomics. Nature methods, December 2021. doi:10.1038/s41592-021-01336-8.">Luecken <em>et al.</em>, 2021</a>]</span> (the most highly-variable genes often describe only dominant biological variation), we recommend selecting slightly too many genes rather than selecting too few and risk removing genes that are important for a rare cell type or a pathway of interest. It should however be noted that more genes will also increase the time required to run the integration methods.</p>
</div>
<p>We will create an object with just the selected genes to use for integration.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">adata_hvg</span> <span class="o">=</span> <span class="n">adata</span><span class="p">[:,</span> <span class="n">adata</span><span class="o">.</span><span class="n">var</span><span class="p">[</span><span class="s2">&quot;highly_variable&quot;</span><span class="p">]]</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
<span class="n">adata_hvg</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>AnnData object with n_obs × n_vars = 10270 × 2000
    obs: &#39;GEX_pct_counts_mt&#39;, &#39;GEX_n_counts&#39;, &#39;GEX_n_genes&#39;, &#39;GEX_size_factors&#39;, &#39;GEX_phase&#39;, &#39;ATAC_nCount_peaks&#39;, &#39;ATAC_atac_fragments&#39;, &#39;ATAC_reads_in_peaks_frac&#39;, &#39;ATAC_blacklist_fraction&#39;, &#39;ATAC_nucleosome_signal&#39;, &#39;cell_type&#39;, &#39;batch&#39;, &#39;ATAC_pseudotime_order&#39;, &#39;GEX_pseudotime_order&#39;, &#39;Samplename&#39;, &#39;Site&#39;, &#39;DonorNumber&#39;, &#39;Modality&#39;, &#39;VendorLot&#39;, &#39;DonorID&#39;, &#39;DonorAge&#39;, &#39;DonorBMI&#39;, &#39;DonorBloodType&#39;, &#39;DonorRace&#39;, &#39;Ethnicity&#39;, &#39;DonorGender&#39;, &#39;QCMeds&#39;, &#39;DonorSmoker&#39;
    var: &#39;feature_types&#39;, &#39;gene_id&#39;, &#39;n_cells&#39;, &#39;highly_variable&#39;, &#39;means&#39;, &#39;dispersions&#39;, &#39;dispersions_norm&#39;, &#39;highly_variable_nbatches&#39;, &#39;highly_variable_intersection&#39;
    uns: &#39;ATAC_gene_activity_var_names&#39;, &#39;dataset_id&#39;, &#39;genome&#39;, &#39;organism&#39;, &#39;log1p&#39;, &#39;hvg&#39;, &#39;pca&#39;, &#39;neighbors&#39;, &#39;umap&#39;, &#39;batch_colors&#39;, &#39;cell_type_colors&#39;
    obsm: &#39;ATAC_gene_activity&#39;, &#39;ATAC_lsi_full&#39;, &#39;ATAC_lsi_red&#39;, &#39;ATAC_umap&#39;, &#39;GEX_X_pca&#39;, &#39;GEX_X_umap&#39;, &#39;X_pca&#39;, &#39;X_umap&#39;
    varm: &#39;PCs&#39;
    layers: &#39;counts&#39;, &#39;logcounts&#39;
    obsp: &#39;distances&#39;, &#39;connectivities&#39;
</pre></div>
</div>
</div>
</div>
</section>
<section id="variational-autoencoder-vae-based-integration">
<h2><span class="section-number">12.5. </span>Variational autoencoder (VAE) based integration<a class="headerlink" href="#variational-autoencoder-vae-based-integration" title="Link to this heading">#</a></h2>
<p>The first integration method we will use is <strong>scVI</strong> (single-cell Variational Inference), a method based on a conditional variational autoencoder <span id="id32">[<a class="reference internal" href="#id226" title="Romain Lopez, Jeffrey Regier, Michael B Cole, Michael I Jordan, and Nir Yosef. Deep generative modeling for single-cell transcriptomics. Nature methods, 15(12):1053–1058, November 2018. doi:10.1038/s41592-018-0229-2.">Lopez <em>et al.</em>, 2018</a>]</span> available in the <strong>scvi-tools</strong> package <span id="id33">[<a class="reference internal" href="#id222" title="Adam Gayoso, Romain Lopez, Galen Xing, Pierre Boyeau, Valeh Valiollah Pour Amiri, Justin Hong, Katherine Wu, Michael Jayasuriya, Edouard Mehlman, Maxime Langevin, Yining Liu, Jules Samaran, Gabriel Misrachi, Achille Nazaret, Oscar Clivio, Chenling Xu, Tal Ashuach, Mariano Gabitto, Mohammad Lotfollahi, Valentine Svensson, Eduardo da Veiga Beltrame, Vitalii Kleshchevnikov, Carlos Talavera-López, Lior Pachter, Fabian J Theis, Aaron Streets, Michael I Jordan, Jeffrey Regier, and Nir Yosef. A Python library for probabilistic analysis of single-cell omics data. Nature biotechnology, February 2022. doi:10.1038/s41587-021-01206-w.">Gayoso <em>et al.</em>, 2022</a>]</span>. A <a class="reference external" href="https://towardsdatascience.com/understanding-variational-autoencoders-vaes-f70510919f73">variational autoencoder</a> is a type of artificial neural network which attempts to reduce the dimensionality of a dataset. The conditional part refers to conditioning this dimensionality reduction process on a particular covariate (in this case batches) such that the covariate does not affect the low-dimensional representation. In benchmarking studies <strong>scVI</strong> has been shown to perform well across a range of datasets with a good balance of batch correction while conserving biological variability <span id="id34">[<a class="reference internal" href="#id220" title="Malte D Luecken, M Büttner, K Chaichoompu, A Danese, M Interlandi, M F Mueller, D C Strobl, L Zappia, M Dugas, M Colomé-Tatché, and Fabian J Theis. Benchmarking atlas-level data integration in single-cell genomics. Nature methods, December 2021. doi:10.1038/s41592-021-01336-8.">Luecken <em>et al.</em>, 2021</a>]</span>. <strong>scVI</strong> models raw counts directly, so it is important that we provide it with a count matrix rather than a normalized expression matrix.</p>
<p>First, let’s make a copy of our dataset to use for this integration. Normally it is not necessary to do this but as we will demonstrate multiple integration methods making a copy makes it easier to show what has been added by each method.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">adata_scvi</span> <span class="o">=</span> <span class="n">adata_hvg</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
</pre></div>
</div>
</div>
</div>
<section id="data-preparation">
<h3><span class="section-number">12.5.1. </span>Data preparation<a class="headerlink" href="#data-preparation" title="Link to this heading">#</a></h3>
<p>The first step in using <strong>scVI</strong> is to prepare our AnnData object. This step stores some information required by <strong>scVI</strong> such as which expression matrix to use and what the batch key is.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">scvi</span><span class="o">.</span><span class="n">model</span><span class="o">.</span><span class="n">SCVI</span><span class="o">.</span><span class="n">setup_anndata</span><span class="p">(</span><span class="n">adata_scvi</span><span class="p">,</span> <span class="n">layer</span><span class="o">=</span><span class="s2">&quot;counts&quot;</span><span class="p">,</span> <span class="n">batch_key</span><span class="o">=</span><span class="n">batch_key</span><span class="p">)</span>
<span class="n">adata_scvi</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>AnnData object with n_obs × n_vars = 10270 × 2000
    obs: &#39;GEX_pct_counts_mt&#39;, &#39;GEX_n_counts&#39;, &#39;GEX_n_genes&#39;, &#39;GEX_size_factors&#39;, &#39;GEX_phase&#39;, &#39;ATAC_nCount_peaks&#39;, &#39;ATAC_atac_fragments&#39;, &#39;ATAC_reads_in_peaks_frac&#39;, &#39;ATAC_blacklist_fraction&#39;, &#39;ATAC_nucleosome_signal&#39;, &#39;cell_type&#39;, &#39;batch&#39;, &#39;ATAC_pseudotime_order&#39;, &#39;GEX_pseudotime_order&#39;, &#39;Samplename&#39;, &#39;Site&#39;, &#39;DonorNumber&#39;, &#39;Modality&#39;, &#39;VendorLot&#39;, &#39;DonorID&#39;, &#39;DonorAge&#39;, &#39;DonorBMI&#39;, &#39;DonorBloodType&#39;, &#39;DonorRace&#39;, &#39;Ethnicity&#39;, &#39;DonorGender&#39;, &#39;QCMeds&#39;, &#39;DonorSmoker&#39;, &#39;_scvi_batch&#39;, &#39;_scvi_labels&#39;
    var: &#39;feature_types&#39;, &#39;gene_id&#39;, &#39;n_cells&#39;, &#39;highly_variable&#39;, &#39;means&#39;, &#39;dispersions&#39;, &#39;dispersions_norm&#39;, &#39;highly_variable_nbatches&#39;, &#39;highly_variable_intersection&#39;
    uns: &#39;ATAC_gene_activity_var_names&#39;, &#39;dataset_id&#39;, &#39;genome&#39;, &#39;organism&#39;, &#39;log1p&#39;, &#39;hvg&#39;, &#39;pca&#39;, &#39;neighbors&#39;, &#39;umap&#39;, &#39;batch_colors&#39;, &#39;cell_type_colors&#39;, &#39;_scvi_uuid&#39;, &#39;_scvi_manager_uuid&#39;
    obsm: &#39;ATAC_gene_activity&#39;, &#39;ATAC_lsi_full&#39;, &#39;ATAC_lsi_red&#39;, &#39;ATAC_umap&#39;, &#39;GEX_X_pca&#39;, &#39;GEX_X_umap&#39;, &#39;X_pca&#39;, &#39;X_umap&#39;
    varm: &#39;PCs&#39;
    layers: &#39;counts&#39;, &#39;logcounts&#39;
    obsp: &#39;distances&#39;, &#39;connectivities&#39;
</pre></div>
</div>
</div>
</div>
<p>The fields created by <strong>scVI</strong> are prefixed with <code class="docutils literal notranslate"><span class="pre">_scvi</span></code>. These are designed for internal use and should not be manually modified. The general advice from the <strong>scvi-tools</strong> authors is to not make any changes to our object until after the model is trained. On other datasets, you may see a warning about the input expression matrix containing unnormalised count data. This usually means you should check that the layer provided to the setup function does actually contain count values but it can also happen if you have values from performing gene length correction on data from a full-length protocol or from another quantification method that does not produce integer counts.</p>
</section>
<section id="building-the-model">
<h3><span class="section-number">12.5.2. </span>Building the model<a class="headerlink" href="#building-the-model" title="Link to this heading">#</a></h3>
<p>We can now construct an <strong>scVI</strong> model object. As well as the <strong>scVI</strong> model we use here, the <strong>scvi-tools</strong> package contains various other models (we will use the <strong>scANVI</strong> model below).</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">model_scvi</span> <span class="o">=</span> <span class="n">scvi</span><span class="o">.</span><span class="n">model</span><span class="o">.</span><span class="n">SCVI</span><span class="p">(</span><span class="n">adata_scvi</span><span class="p">)</span>
<span class="n">model_scvi</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><pre style="white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace">SCVI Model with the following params: 
n_hidden: <span style="color: #008080; text-decoration-color: #008080; font-weight: bold">128</span>, n_latent: <span style="color: #008080; text-decoration-color: #008080; font-weight: bold">10</span>, n_layers: <span style="color: #008080; text-decoration-color: #008080; font-weight: bold">1</span>, dropout_rate: <span style="color: #008080; text-decoration-color: #008080; font-weight: bold">0.1</span>, dispersion: gene, gene_likelihood: zinb, 
latent_distribution: normal
Training status: Not Trained
</pre>
</div><div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>
</pre></div>
</div>
</div>
</div>
<p>The <strong>scVI</strong> model object contains the provided AnnData object as well as the neural network for the model itself. You can see that currently the model is not trained. If we wanted to modify the structure of the network we could provide additional arguments to the model construction function but here we just use the defaults.</p>
<p>We can also print a more detailed description of the model that shows us where things are stored in the associated AnnData object.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">model_scvi</span><span class="o">.</span><span class="n">view_anndata_setup</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><pre style="white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace">Anndata setup with scvi-tools version <span style="color: #008080; text-decoration-color: #008080; font-weight: bold">0.18</span>.<span style="color: #008080; text-decoration-color: #008080; font-weight: bold">0</span>.
</pre>
</div><div class="output text_html"><pre style="white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace">
</pre>
</div><div class="output text_html"><pre style="white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace">Setup via `SCVI.setup_anndata` with arguments:
</pre>
</div><div class="output text_html"><pre style="white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace"><span style="font-weight: bold">{</span>
<span style="color: #7fbf7f; text-decoration-color: #7fbf7f">│   </span><span style="color: #008000; text-decoration-color: #008000">'layer'</span>: <span style="color: #008000; text-decoration-color: #008000">'counts'</span>,
<span style="color: #7fbf7f; text-decoration-color: #7fbf7f">│   </span><span style="color: #008000; text-decoration-color: #008000">'batch_key'</span>: <span style="color: #008000; text-decoration-color: #008000">'batch'</span>,
<span style="color: #7fbf7f; text-decoration-color: #7fbf7f">│   </span><span style="color: #008000; text-decoration-color: #008000">'labels_key'</span>: <span style="color: #800080; text-decoration-color: #800080; font-style: italic">None</span>,
<span style="color: #7fbf7f; text-decoration-color: #7fbf7f">│   </span><span style="color: #008000; text-decoration-color: #008000">'size_factor_key'</span>: <span style="color: #800080; text-decoration-color: #800080; font-style: italic">None</span>,
<span style="color: #7fbf7f; text-decoration-color: #7fbf7f">│   </span><span style="color: #008000; text-decoration-color: #008000">'categorical_covariate_keys'</span>: <span style="color: #800080; text-decoration-color: #800080; font-style: italic">None</span>,
<span style="color: #7fbf7f; text-decoration-color: #7fbf7f">│   </span><span style="color: #008000; text-decoration-color: #008000">'continuous_covariate_keys'</span>: <span style="color: #800080; text-decoration-color: #800080; font-style: italic">None</span>
<span style="font-weight: bold">}</span>
</pre>
</div><div class="output text_html"><pre style="white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace">
</pre>
</div><div class="output text_html"><pre style="white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace"><span style="font-style: italic">         Summary Statistics         </span>
┏━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━┓
┃<span style="font-weight: bold">     Summary Stat Key     </span>┃<span style="font-weight: bold"> Value </span>┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━┩
│<span style="color: #0087ff; text-decoration-color: #0087ff">         n_batch          </span>│<span style="color: #af00d7; text-decoration-color: #af00d7">   3   </span>│
│<span style="color: #0087ff; text-decoration-color: #0087ff">         n_cells          </span>│<span style="color: #af00d7; text-decoration-color: #af00d7"> 10270 </span>│
│<span style="color: #0087ff; text-decoration-color: #0087ff"> n_extra_categorical_covs </span>│<span style="color: #af00d7; text-decoration-color: #af00d7">   0   </span>│
│<span style="color: #0087ff; text-decoration-color: #0087ff"> n_extra_continuous_covs  </span>│<span style="color: #af00d7; text-decoration-color: #af00d7">   0   </span>│
│<span style="color: #0087ff; text-decoration-color: #0087ff">         n_labels         </span>│<span style="color: #af00d7; text-decoration-color: #af00d7">   1   </span>│
│<span style="color: #0087ff; text-decoration-color: #0087ff">          n_vars          </span>│<span style="color: #af00d7; text-decoration-color: #af00d7"> 2000  </span>│
└──────────────────────────┴───────┘
</pre>
</div><div class="output text_html"><pre style="white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace"><span style="font-style: italic">               Data Registry                </span>
┏━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━━━━┓
┃<span style="font-weight: bold"> Registry Key </span>┃<span style="font-weight: bold">    scvi-tools Location    </span>┃
┡━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━━━━┩
│<span style="color: #0087ff; text-decoration-color: #0087ff">      X       </span>│<span style="color: #af00d7; text-decoration-color: #af00d7">  adata.layers['counts']   </span>│
│<span style="color: #0087ff; text-decoration-color: #0087ff">    batch     </span>│<span style="color: #af00d7; text-decoration-color: #af00d7"> adata.obs['_scvi_batch']  </span>│
│<span style="color: #0087ff; text-decoration-color: #0087ff">    labels    </span>│<span style="color: #af00d7; text-decoration-color: #af00d7"> adata.obs['_scvi_labels'] </span>│
└──────────────┴───────────────────────────┘
</pre>
</div><div class="output text_html"><pre style="white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace"><span style="font-style: italic">                  batch State Registry                   </span>
┏━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━┓
┃<span style="font-weight: bold">  Source Location   </span>┃<span style="font-weight: bold"> Categories </span>┃<span style="font-weight: bold"> scvi-tools Encoding </span>┃
┡━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━┩
│<span style="color: #0087ff; text-decoration-color: #0087ff"> adata.obs['batch'] </span>│<span style="color: #008000; text-decoration-color: #008000">    s1d3    </span>│<span style="color: #af00d7; text-decoration-color: #af00d7">          0          </span>│
│<span style="color: #0087ff; text-decoration-color: #0087ff">                    </span>│<span style="color: #008000; text-decoration-color: #008000">    s2d1    </span>│<span style="color: #af00d7; text-decoration-color: #af00d7">          1          </span>│
│<span style="color: #0087ff; text-decoration-color: #0087ff">                    </span>│<span style="color: #008000; text-decoration-color: #008000">    s3d7    </span>│<span style="color: #af00d7; text-decoration-color: #af00d7">          2          </span>│
└────────────────────┴────────────┴─────────────────────┘
</pre>
</div><div class="output text_html"><pre style="white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace"><span style="font-style: italic">                     labels State Registry                      </span>
┏━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━┓
┃<span style="font-weight: bold">      Source Location      </span>┃<span style="font-weight: bold"> Categories </span>┃<span style="font-weight: bold"> scvi-tools Encoding </span>┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━┩
│<span style="color: #0087ff; text-decoration-color: #0087ff"> adata.obs['_scvi_labels'] </span>│<span style="color: #008000; text-decoration-color: #008000">     0      </span>│<span style="color: #af00d7; text-decoration-color: #af00d7">          0          </span>│
└───────────────────────────┴────────────┴─────────────────────┘
</pre>
</div></div>
</div>
<p>Here we can see exactly what information has been assigned by <strong>scVI</strong>, including details like how each different batch is encoded in the model.</p>
</section>
<section id="training-the-model">
<h3><span class="section-number">12.5.3. </span>Training the model<a class="headerlink" href="#training-the-model" title="Link to this heading">#</a></h3>
<p>The model will be trained for a given number of <em>epochs</em>, a training iteration where every cell is passed through the network. By default <strong>scVI</strong> uses the following heuristic to set the number of epochs. For datasets with fewer than 20,000 cells, 400 epochs will be used and as the number of cells grows above 20,000 the number of epochs is continuously reduced. The reasoning behind this is that as the network sees more cells during each epoch it can learn the same amount of information as it would from more epochs with fewer cells.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">max_epochs_scvi</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">min</span><span class="p">([</span><span class="nb">round</span><span class="p">((</span><span class="mi">20000</span> <span class="o">/</span> <span class="n">adata</span><span class="o">.</span><span class="n">n_obs</span><span class="p">)</span> <span class="o">*</span> <span class="mi">400</span><span class="p">),</span> <span class="mi">400</span><span class="p">])</span>
<span class="n">max_epochs_scvi</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>400
</pre></div>
</div>
</div>
</div>
<p>We now train the model for the selected number of epochs (this will take ~20-40 minutes depending on the computer you are using).</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">model_scvi</span><span class="o">.</span><span class="n">train</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>GPU available: False, used: False
TPU available: False, using: 0 TPU cores
IPU available: False, using: 0 IPUs
HPU available: False, using: 0 HPUs
</pre></div>
</div>
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Epoch 400/400: 100%|██████████| 400/400 [15:49&lt;00:00,  2.27s/it, loss=648, v_num=1]
</pre></div>
</div>
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>`Trainer.fit` stopped: `max_epochs=400` reached.
</pre></div>
</div>
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Epoch 400/400: 100%|██████████| 400/400 [15:49&lt;00:00,  2.37s/it, loss=648, v_num=1]
</pre></div>
</div>
</div>
</div>
<div class="admonition-early-stopping admonition">
<p class="admonition-title">Early stopping</p>
<p>Additionally to setting a target number of epochs, it is possible to also set <code class="docutils literal notranslate"><span class="pre">early_stopping=True</span></code> in the training function. This will let <strong>scVI</strong> decide to stop training early depending on the convergence of the model. The exact conditions for stopping can be controlled by other parameters.</p>
</div>
</section>
<section id="extracting-the-embedding">
<h3><span class="section-number">12.5.4. </span>Extracting the embedding<a class="headerlink" href="#extracting-the-embedding" title="Link to this heading">#</a></h3>
<p>The main result we want to extract from the trained model is the latent representation for each cell. This is a multi-dimensional embedding where the batch effects have been removed that can be used in a similar way to how we use PCA dimensions when analysing a single dataset. We store this in <code class="docutils literal notranslate"><span class="pre">obsm</span></code> with the key <code class="docutils literal notranslate"><span class="pre">X_scvi</span></code>.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">adata_scvi</span><span class="o">.</span><span class="n">obsm</span><span class="p">[</span><span class="s2">&quot;X_scVI&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">model_scvi</span><span class="o">.</span><span class="n">get_latent_representation</span><span class="p">()</span>
</pre></div>
</div>
</div>
</div>
</section>
<section id="calculate-a-batch-corrected-umap">
<h3><span class="section-number">12.5.5. </span>Calculate a batch-corrected UMAP<a class="headerlink" href="#calculate-a-batch-corrected-umap" title="Link to this heading">#</a></h3>
<p>We will now visualise the data as we did before integration. We calculate a new UMAP embedding but instead of finding nearest neighbors in PCA space, we start with the corrected representation from <strong>scVI</strong>.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">sc</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">neighbors</span><span class="p">(</span><span class="n">adata_scvi</span><span class="p">,</span> <span class="n">use_rep</span><span class="o">=</span><span class="s2">&quot;X_scVI&quot;</span><span class="p">)</span>
<span class="n">sc</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">umap</span><span class="p">(</span><span class="n">adata_scvi</span><span class="p">)</span>
<span class="n">adata_scvi</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>AnnData object with n_obs × n_vars = 10270 × 2000
    obs: &#39;GEX_pct_counts_mt&#39;, &#39;GEX_n_counts&#39;, &#39;GEX_n_genes&#39;, &#39;GEX_size_factors&#39;, &#39;GEX_phase&#39;, &#39;ATAC_nCount_peaks&#39;, &#39;ATAC_atac_fragments&#39;, &#39;ATAC_reads_in_peaks_frac&#39;, &#39;ATAC_blacklist_fraction&#39;, &#39;ATAC_nucleosome_signal&#39;, &#39;cell_type&#39;, &#39;batch&#39;, &#39;ATAC_pseudotime_order&#39;, &#39;GEX_pseudotime_order&#39;, &#39;Samplename&#39;, &#39;Site&#39;, &#39;DonorNumber&#39;, &#39;Modality&#39;, &#39;VendorLot&#39;, &#39;DonorID&#39;, &#39;DonorAge&#39;, &#39;DonorBMI&#39;, &#39;DonorBloodType&#39;, &#39;DonorRace&#39;, &#39;Ethnicity&#39;, &#39;DonorGender&#39;, &#39;QCMeds&#39;, &#39;DonorSmoker&#39;, &#39;_scvi_batch&#39;, &#39;_scvi_labels&#39;
    var: &#39;feature_types&#39;, &#39;gene_id&#39;, &#39;n_cells&#39;, &#39;highly_variable&#39;, &#39;means&#39;, &#39;dispersions&#39;, &#39;dispersions_norm&#39;, &#39;highly_variable_nbatches&#39;, &#39;highly_variable_intersection&#39;
    uns: &#39;ATAC_gene_activity_var_names&#39;, &#39;dataset_id&#39;, &#39;genome&#39;, &#39;organism&#39;, &#39;log1p&#39;, &#39;hvg&#39;, &#39;pca&#39;, &#39;neighbors&#39;, &#39;umap&#39;, &#39;batch_colors&#39;, &#39;cell_type_colors&#39;, &#39;_scvi_uuid&#39;, &#39;_scvi_manager_uuid&#39;
    obsm: &#39;ATAC_gene_activity&#39;, &#39;ATAC_lsi_full&#39;, &#39;ATAC_lsi_red&#39;, &#39;ATAC_umap&#39;, &#39;GEX_X_pca&#39;, &#39;GEX_X_umap&#39;, &#39;X_pca&#39;, &#39;X_umap&#39;, &#39;X_scVI&#39;
    varm: &#39;PCs&#39;
    layers: &#39;counts&#39;, &#39;logcounts&#39;
    obsp: &#39;distances&#39;, &#39;connectivities&#39;
</pre></div>
</div>
</div>
</div>
<p>Once we have the new UMAP representation we can plot it coloured by batch and identity labels as before.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">sc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">umap</span><span class="p">(</span><span class="n">adata_scvi</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="p">[</span><span class="n">label_key</span><span class="p">,</span> <span class="n">batch_key</span><span class="p">],</span> <span class="n">wspace</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>/Users/luke.zappia/miniconda/envs/bp-integration/lib/python3.9/site-packages/scanpy/plotting/_tools/scatterplots.py:163: PendingDeprecationWarning: The get_cmap function will be deprecated in a future version. Use ``matplotlib.colormaps[name]`` or ``matplotlib.colormaps.get_cmap(obj)`` instead.
  cmap = copy(get_cmap(cmap))
/Users/luke.zappia/miniconda/envs/bp-integration/lib/python3.9/site-packages/scanpy/plotting/_tools/scatterplots.py:392: UserWarning: No data for colormapping provided via &#39;c&#39;. Parameters &#39;cmap&#39; will be ignored
  cax = scatter(
/Users/luke.zappia/miniconda/envs/bp-integration/lib/python3.9/site-packages/scanpy/plotting/_tools/scatterplots.py:392: UserWarning: No data for colormapping provided via &#39;c&#39;. Parameters &#39;cmap&#39; will be ignored
  cax = scatter(
</pre></div>
</div>
<img alt="../_images/8e5313bb4f33a892aaef9b713a3343aa7328fce5af7be5ce85460c48a64cc163.png" src="../_images/8e5313bb4f33a892aaef9b713a3343aa7328fce5af7be5ce85460c48a64cc163.png" />
</div>
</div>
<p>This looks better! Before, the various batches were shifted apart from each other. Now, the batches overlap more and we have a single blob for each cell identity label.</p>
<p>In many cases, we would not already have identity labels so from this stage we would continue with clustering, annotation and further analysis as described in other chapters.</p>
</section>
</section>
<section id="vae-integration-using-cell-labels">
<h2><span class="section-number">12.6. </span>VAE integration using cell labels<a class="headerlink" href="#vae-integration-using-cell-labels" title="Link to this heading">#</a></h2>
<p>When performing integration with <strong>scVI</strong> we pretended that we didn’t already have any cell labels (although we showed them in plots). While this scenario is common there are some cases where we do know something about cell identity in advance. Most often this is when we want to combine one or more publicly available datasets with data from a new study. When we have labels for at least some of the cells we can use <strong>scANVI</strong> (single-cell ANnotation using Variational Inference) <span id="id35">[<a class="reference internal" href="#id227" title="Chenling Xu, Romain Lopez, Edouard Mehlman, Jeffrey Regier, Michael I Jordan, and Nir Yosef. Probabilistic harmonization and annotation of single-cell transcriptomics data with deep generative models. Molecular systems biology, 17(1):e9620, January 2021. doi:10.15252/msb.20209620.">Xu <em>et al.</em>, 2021</a>]</span>. This is an extension of the <strong>scVI</strong> model that can incorporate cell identity label information as well as batch information. Because it has this extra information it can try to keep the differences between cell labels while removing batch effects. Benchmarking suggests that <strong>scANVI</strong> tends to better preserve biological signals compared to <strong>scVI</strong> but sometimes it is not as effective at removing batch effects <span id="id36">[<a class="reference internal" href="#id220" title="Malte D Luecken, M Büttner, K Chaichoompu, A Danese, M Interlandi, M F Mueller, D C Strobl, L Zappia, M Dugas, M Colomé-Tatché, and Fabian J Theis. Benchmarking atlas-level data integration in single-cell genomics. Nature methods, December 2021. doi:10.1038/s41592-021-01336-8.">Luecken <em>et al.</em>, 2021</a>]</span>. While we have labels for all cells here it is also possible to use <strong>scANVI</strong> in a semi-supervised manner where labels are only provided for some cells.</p>
<div class="admonition-label-harmonization admonition">
<p class="admonition-title">Label harmonization</p>
<p>If you are using <strong>scANVI</strong> to integrate multiple datasets for which you already have labels it is important to first perform <em>label harmonization</em>. This refers to a process of checking that labels are consistent across the datasets that are being integrated. For example, a cell may be annotated as a “T cell” in one dataset, but a cell of the same type could have been given the label “CD8+ T cell” in another dataset. How best to harmonize labels is an open question but often requires input from subject-matter experts.</p>
</div>
<p>We start by creating a <strong>scANVI</strong> model object. Note that because <strong>scANVI</strong> refines an already trained <strong>scVI</strong> model, we provide the scVI model rather than an AnnData object. If we had not already trained an <strong>scVI</strong> model we would need to do that first. We also provide a key for the column of <code class="docutils literal notranslate"><span class="pre">adata.obs</span></code> which contains our cell labels as well as the label which corresponds to unlabelled cells. In this case all of our cells are labelled so we just provide a dummy value. In most cases, it is important to check that this is set correctly so that <strong>scANVI</strong> knows which label to ignore during training.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Normally we would need to run scVI first but we have already done that here</span>
<span class="c1"># model_scvi = scvi.model.SCVI(adata_scvi) etc.</span>
<span class="n">model_scanvi</span> <span class="o">=</span> <span class="n">scvi</span><span class="o">.</span><span class="n">model</span><span class="o">.</span><span class="n">SCANVI</span><span class="o">.</span><span class="n">from_scvi_model</span><span class="p">(</span>
    <span class="n">model_scvi</span><span class="p">,</span> <span class="n">labels_key</span><span class="o">=</span><span class="n">label_key</span><span class="p">,</span> <span class="n">unlabeled_category</span><span class="o">=</span><span class="s2">&quot;unlabelled&quot;</span>
<span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="n">model_scanvi</span><span class="p">)</span>
<span class="n">model_scanvi</span><span class="o">.</span><span class="n">view_anndata_setup</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><pre style="white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace">ScanVI Model with the following params: 
unlabeled_category: unlabelled, n_hidden: <span style="color: #008080; text-decoration-color: #008080; font-weight: bold">128</span>, n_latent: <span style="color: #008080; text-decoration-color: #008080; font-weight: bold">10</span>, n_layers: <span style="color: #008080; text-decoration-color: #008080; font-weight: bold">1</span>, dropout_rate: <span style="color: #008080; text-decoration-color: #008080; font-weight: bold">0.1</span>, dispersion: gene, 
gene_likelihood: zinb
Training status: Not Trained
</pre>
</div><div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>
</pre></div>
</div>
<div class="output text_html"><pre style="white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace">Anndata setup with scvi-tools version <span style="color: #008080; text-decoration-color: #008080; font-weight: bold">0.18</span>.<span style="color: #008080; text-decoration-color: #008080; font-weight: bold">0</span>.
</pre>
</div><div class="output text_html"><pre style="white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace">
</pre>
</div><div class="output text_html"><pre style="white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace">Setup via `SCANVI.setup_anndata` with arguments:
</pre>
</div><div class="output text_html"><pre style="white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace"><span style="font-weight: bold">{</span>
<span style="color: #7fbf7f; text-decoration-color: #7fbf7f">│   </span><span style="color: #008000; text-decoration-color: #008000">'labels_key'</span>: <span style="color: #008000; text-decoration-color: #008000">'cell_type'</span>,
<span style="color: #7fbf7f; text-decoration-color: #7fbf7f">│   </span><span style="color: #008000; text-decoration-color: #008000">'unlabeled_category'</span>: <span style="color: #008000; text-decoration-color: #008000">'unlabelled'</span>,
<span style="color: #7fbf7f; text-decoration-color: #7fbf7f">│   </span><span style="color: #008000; text-decoration-color: #008000">'layer'</span>: <span style="color: #008000; text-decoration-color: #008000">'counts'</span>,
<span style="color: #7fbf7f; text-decoration-color: #7fbf7f">│   </span><span style="color: #008000; text-decoration-color: #008000">'batch_key'</span>: <span style="color: #008000; text-decoration-color: #008000">'batch'</span>,
<span style="color: #7fbf7f; text-decoration-color: #7fbf7f">│   </span><span style="color: #008000; text-decoration-color: #008000">'size_factor_key'</span>: <span style="color: #800080; text-decoration-color: #800080; font-style: italic">None</span>,
<span style="color: #7fbf7f; text-decoration-color: #7fbf7f">│   </span><span style="color: #008000; text-decoration-color: #008000">'categorical_covariate_keys'</span>: <span style="color: #800080; text-decoration-color: #800080; font-style: italic">None</span>,
<span style="color: #7fbf7f; text-decoration-color: #7fbf7f">│   </span><span style="color: #008000; text-decoration-color: #008000">'continuous_covariate_keys'</span>: <span style="color: #800080; text-decoration-color: #800080; font-style: italic">None</span>
<span style="font-weight: bold">}</span>
</pre>
</div><div class="output text_html"><pre style="white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace">
</pre>
</div><div class="output text_html"><pre style="white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace"><span style="font-style: italic">         Summary Statistics         </span>
┏━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━┓
┃<span style="font-weight: bold">     Summary Stat Key     </span>┃<span style="font-weight: bold"> Value </span>┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━┩
│<span style="color: #0087ff; text-decoration-color: #0087ff">         n_batch          </span>│<span style="color: #af00d7; text-decoration-color: #af00d7">   3   </span>│
│<span style="color: #0087ff; text-decoration-color: #0087ff">         n_cells          </span>│<span style="color: #af00d7; text-decoration-color: #af00d7"> 10270 </span>│
│<span style="color: #0087ff; text-decoration-color: #0087ff"> n_extra_categorical_covs </span>│<span style="color: #af00d7; text-decoration-color: #af00d7">   0   </span>│
│<span style="color: #0087ff; text-decoration-color: #0087ff"> n_extra_continuous_covs  </span>│<span style="color: #af00d7; text-decoration-color: #af00d7">   0   </span>│
│<span style="color: #0087ff; text-decoration-color: #0087ff">         n_labels         </span>│<span style="color: #af00d7; text-decoration-color: #af00d7">  22   </span>│
│<span style="color: #0087ff; text-decoration-color: #0087ff">          n_vars          </span>│<span style="color: #af00d7; text-decoration-color: #af00d7"> 2000  </span>│
└──────────────────────────┴───────┘
</pre>
</div><div class="output text_html"><pre style="white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace"><span style="font-style: italic">               Data Registry                </span>
┏━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━━━━┓
┃<span style="font-weight: bold"> Registry Key </span>┃<span style="font-weight: bold">    scvi-tools Location    </span>┃
┡━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━━━━┩
│<span style="color: #0087ff; text-decoration-color: #0087ff">      X       </span>│<span style="color: #af00d7; text-decoration-color: #af00d7">  adata.layers['counts']   </span>│
│<span style="color: #0087ff; text-decoration-color: #0087ff">    batch     </span>│<span style="color: #af00d7; text-decoration-color: #af00d7"> adata.obs['_scvi_batch']  </span>│
│<span style="color: #0087ff; text-decoration-color: #0087ff">    labels    </span>│<span style="color: #af00d7; text-decoration-color: #af00d7"> adata.obs['_scvi_labels'] </span>│
└──────────────┴───────────────────────────┘
</pre>
</div><div class="output text_html"><pre style="white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace"><span style="font-style: italic">                  batch State Registry                   </span>
┏━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━┓
┃<span style="font-weight: bold">  Source Location   </span>┃<span style="font-weight: bold"> Categories </span>┃<span style="font-weight: bold"> scvi-tools Encoding </span>┃
┡━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━┩
│<span style="color: #0087ff; text-decoration-color: #0087ff"> adata.obs['batch'] </span>│<span style="color: #008000; text-decoration-color: #008000">    s1d3    </span>│<span style="color: #af00d7; text-decoration-color: #af00d7">          0          </span>│
│<span style="color: #0087ff; text-decoration-color: #0087ff">                    </span>│<span style="color: #008000; text-decoration-color: #008000">    s2d1    </span>│<span style="color: #af00d7; text-decoration-color: #af00d7">          1          </span>│
│<span style="color: #0087ff; text-decoration-color: #0087ff">                    </span>│<span style="color: #008000; text-decoration-color: #008000">    s3d7    </span>│<span style="color: #af00d7; text-decoration-color: #af00d7">          2          </span>│
└────────────────────┴────────────┴─────────────────────┘
</pre>
</div><div class="output text_html"><pre style="white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace"><span style="font-style: italic">                       labels State Registry                       </span>
┏━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━┓
┃<span style="font-weight: bold">    Source Location     </span>┃<span style="font-weight: bold">    Categories    </span>┃<span style="font-weight: bold"> scvi-tools Encoding </span>┃
┡━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━┩
│<span style="color: #0087ff; text-decoration-color: #0087ff"> adata.obs['cell_type'] </span>│<span style="color: #008000; text-decoration-color: #008000">       B1 B       </span>│<span style="color: #af00d7; text-decoration-color: #af00d7">          0          </span>│
│<span style="color: #0087ff; text-decoration-color: #0087ff">                        </span>│<span style="color: #008000; text-decoration-color: #008000"> CD4+ T activated </span>│<span style="color: #af00d7; text-decoration-color: #af00d7">          1          </span>│
│<span style="color: #0087ff; text-decoration-color: #0087ff">                        </span>│<span style="color: #008000; text-decoration-color: #008000">   CD4+ T naive   </span>│<span style="color: #af00d7; text-decoration-color: #af00d7">          2          </span>│
│<span style="color: #0087ff; text-decoration-color: #0087ff">                        </span>│<span style="color: #008000; text-decoration-color: #008000">      CD8+ T      </span>│<span style="color: #af00d7; text-decoration-color: #af00d7">          3          </span>│
│<span style="color: #0087ff; text-decoration-color: #0087ff">                        </span>│<span style="color: #008000; text-decoration-color: #008000">   CD8+ T naive   </span>│<span style="color: #af00d7; text-decoration-color: #af00d7">          4          </span>│
│<span style="color: #0087ff; text-decoration-color: #0087ff">                        </span>│<span style="color: #008000; text-decoration-color: #008000">    CD14+ Mono    </span>│<span style="color: #af00d7; text-decoration-color: #af00d7">          5          </span>│
│<span style="color: #0087ff; text-decoration-color: #0087ff">                        </span>│<span style="color: #008000; text-decoration-color: #008000">    CD16+ Mono    </span>│<span style="color: #af00d7; text-decoration-color: #af00d7">          6          </span>│
│<span style="color: #0087ff; text-decoration-color: #0087ff">                        </span>│<span style="color: #008000; text-decoration-color: #008000">   Erythroblast   </span>│<span style="color: #af00d7; text-decoration-color: #af00d7">          7          </span>│
│<span style="color: #0087ff; text-decoration-color: #0087ff">                        </span>│<span style="color: #008000; text-decoration-color: #008000">     G/M prog     </span>│<span style="color: #af00d7; text-decoration-color: #af00d7">          8          </span>│
│<span style="color: #0087ff; text-decoration-color: #0087ff">                        </span>│<span style="color: #008000; text-decoration-color: #008000">       HSC        </span>│<span style="color: #af00d7; text-decoration-color: #af00d7">          9          </span>│
│<span style="color: #0087ff; text-decoration-color: #0087ff">                        </span>│<span style="color: #008000; text-decoration-color: #008000">       ILC        </span>│<span style="color: #af00d7; text-decoration-color: #af00d7">         10          </span>│
│<span style="color: #0087ff; text-decoration-color: #0087ff">                        </span>│<span style="color: #008000; text-decoration-color: #008000">    Lymph prog    </span>│<span style="color: #af00d7; text-decoration-color: #af00d7">         11          </span>│
│<span style="color: #0087ff; text-decoration-color: #0087ff">                        </span>│<span style="color: #008000; text-decoration-color: #008000">    MK/E prog     </span>│<span style="color: #af00d7; text-decoration-color: #af00d7">         12          </span>│
│<span style="color: #0087ff; text-decoration-color: #0087ff">                        </span>│<span style="color: #008000; text-decoration-color: #008000">        NK        </span>│<span style="color: #af00d7; text-decoration-color: #af00d7">         13          </span>│
│<span style="color: #0087ff; text-decoration-color: #0087ff">                        </span>│<span style="color: #008000; text-decoration-color: #008000">  Naive CD20+ B   </span>│<span style="color: #af00d7; text-decoration-color: #af00d7">         14          </span>│
│<span style="color: #0087ff; text-decoration-color: #0087ff">                        </span>│<span style="color: #008000; text-decoration-color: #008000">    Normoblast    </span>│<span style="color: #af00d7; text-decoration-color: #af00d7">         15          </span>│
│<span style="color: #0087ff; text-decoration-color: #0087ff">                        </span>│<span style="color: #008000; text-decoration-color: #008000">   Plasma cell    </span>│<span style="color: #af00d7; text-decoration-color: #af00d7">         16          </span>│
│<span style="color: #0087ff; text-decoration-color: #0087ff">                        </span>│<span style="color: #008000; text-decoration-color: #008000"> Proerythroblast  </span>│<span style="color: #af00d7; text-decoration-color: #af00d7">         17          </span>│
│<span style="color: #0087ff; text-decoration-color: #0087ff">                        </span>│<span style="color: #008000; text-decoration-color: #008000">  Transitional B  </span>│<span style="color: #af00d7; text-decoration-color: #af00d7">         18          </span>│
│<span style="color: #0087ff; text-decoration-color: #0087ff">                        </span>│<span style="color: #008000; text-decoration-color: #008000">       cDC2       </span>│<span style="color: #af00d7; text-decoration-color: #af00d7">         19          </span>│
│<span style="color: #0087ff; text-decoration-color: #0087ff">                        </span>│<span style="color: #008000; text-decoration-color: #008000">       pDC        </span>│<span style="color: #af00d7; text-decoration-color: #af00d7">         20          </span>│
│<span style="color: #0087ff; text-decoration-color: #0087ff">                        </span>│<span style="color: #008000; text-decoration-color: #008000">    unlabelled    </span>│<span style="color: #af00d7; text-decoration-color: #af00d7">         21          </span>│
└────────────────────────┴──────────────────┴─────────────────────┘
</pre>
</div></div>
</div>
<p>This <strong>scANVI</strong> model object is very similar to what we saw before for <strong>scVI</strong>. As mentioned previously, we could modify the structure of the model network but here we just use the default parameters.</p>
<p>Again, we have a heuristic for selecting the number of training epochs. Note that this is much fewer than before as we are just refining the <strong>scVI</strong> model, rather than training a whole network from scratch.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">max_epochs_scanvi</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">min</span><span class="p">([</span><span class="mi">10</span><span class="p">,</span> <span class="n">np</span><span class="o">.</span><span class="n">max</span><span class="p">([</span><span class="mi">2</span><span class="p">,</span> <span class="nb">round</span><span class="p">(</span><span class="n">max_epochs_scvi</span> <span class="o">/</span> <span class="mf">3.0</span><span class="p">)])]))</span>
<span class="n">model_scanvi</span><span class="o">.</span><span class="n">train</span><span class="p">(</span><span class="n">max_epochs</span><span class="o">=</span><span class="n">max_epochs_scanvi</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span><span class=" -Color -Color-Blue">INFO    </span> Training for <span class=" -Color -Color-Bold -Color-Bold-Cyan">10</span> epochs.                                                                                   
</pre></div>
</div>
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>GPU available: False, used: False
TPU available: False, using: 0 TPU cores
IPU available: False, using: 0 IPUs
HPU available: False, using: 0 HPUs
</pre></div>
</div>
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Epoch 10/10: 100%|██████████| 10/10 [00:35&lt;00:00,  3.69s/it, loss=747, v_num=1]
</pre></div>
</div>
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>`Trainer.fit` stopped: `max_epochs=10` reached.
</pre></div>
</div>
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Epoch 10/10: 100%|██████████| 10/10 [00:35&lt;00:00,  3.56s/it, loss=747, v_num=1]
</pre></div>
</div>
</div>
</div>
<p>We can extract the new latent representation from the model and create a new UMAP embedding as we did for <strong>scVI</strong>.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">adata_scanvi</span> <span class="o">=</span> <span class="n">adata_scvi</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
<span class="n">adata_scanvi</span><span class="o">.</span><span class="n">obsm</span><span class="p">[</span><span class="s2">&quot;X_scANVI&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">model_scanvi</span><span class="o">.</span><span class="n">get_latent_representation</span><span class="p">()</span>
<span class="n">sc</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">neighbors</span><span class="p">(</span><span class="n">adata_scanvi</span><span class="p">,</span> <span class="n">use_rep</span><span class="o">=</span><span class="s2">&quot;X_scANVI&quot;</span><span class="p">)</span>
<span class="n">sc</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">umap</span><span class="p">(</span><span class="n">adata_scanvi</span><span class="p">)</span>
<span class="n">sc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">umap</span><span class="p">(</span><span class="n">adata_scanvi</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="p">[</span><span class="n">label_key</span><span class="p">,</span> <span class="n">batch_key</span><span class="p">],</span> <span class="n">wspace</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>/Users/luke.zappia/miniconda/envs/bp-integration/lib/python3.9/site-packages/scanpy/plotting/_tools/scatterplots.py:163: PendingDeprecationWarning: The get_cmap function will be deprecated in a future version. Use ``matplotlib.colormaps[name]`` or ``matplotlib.colormaps.get_cmap(obj)`` instead.
  cmap = copy(get_cmap(cmap))
/Users/luke.zappia/miniconda/envs/bp-integration/lib/python3.9/site-packages/scanpy/plotting/_tools/scatterplots.py:392: UserWarning: No data for colormapping provided via &#39;c&#39;. Parameters &#39;cmap&#39; will be ignored
  cax = scatter(
/Users/luke.zappia/miniconda/envs/bp-integration/lib/python3.9/site-packages/scanpy/plotting/_tools/scatterplots.py:392: UserWarning: No data for colormapping provided via &#39;c&#39;. Parameters &#39;cmap&#39; will be ignored
  cax = scatter(
</pre></div>
</div>
<img alt="../_images/55f6b0b00ec8bef5e16f261b03f8a355d603bb935fa2f505cf1cc16b3e4e2f77.png" src="../_images/55f6b0b00ec8bef5e16f261b03f8a355d603bb935fa2f505cf1cc16b3e4e2f77.png" />
</div>
</div>
<p>By looking at the UMAP representation it is difficult to tell the difference between <strong>scANVI</strong> and <strong>scVI</strong> but as we will see below there are differences in metric scores when the quality of the integrations is quantified. This is a reminder that we shouldn’t overinterpret these two-dimensional representations, especially when it comes to comparing methods.</p>
</section>
<section id="graph-based-integration">
<h2><span class="section-number">12.7. </span>Graph-based integration<a class="headerlink" href="#graph-based-integration" title="Link to this heading">#</a></h2>
<p>The next method we will look at is <strong>BBKNN</strong> or “Batch Balanced KNN” <span id="id37">[<a class="reference internal" href="#id228" title="Krzysztof Polański, Jong-Eun Park, Matthew D Young, Zhichao Miao, Kerstin B Meyer, and Sarah A Teichmann. BBKNN: Fast Batch Alignment of Single Cell Transcriptomes. Bioinformatics, August 2019. doi:10.1093/bioinformatics/btz625.">Polański <em>et al.</em>, 2019</a>]</span>. This is a very different approach to <strong>scVI</strong>, which rather than using a neural network to embed cells in a batch corrected space, instead modifies how the <em>k</em>-nearest neighbor (KNN) graph used for clustering and embedding is constructed. As we have seen in <a class="reference internal" href="clustering.html#cellular-structure-clustering"><span class="std std-ref">previous chapters</span></a> the normal KNN procedure connects cells to the most similar cells across the whole dataset. The change that <strong>BBKNN</strong> makes is to enforce that cells are connected to cells from other batches. While this is a simple modification it can be quite effective, particularly when there are very strong batch effects. However, as the output is an integrated graph it can have limited downstream uses as few packages will accept this as an input.</p>
<p>An important parameter for <strong>BBKNN</strong> is the number of neighbors per batch. A suggested heuristic for this is to use 25 if there are more than 100,000 cells or the default of 3 if there are fewer than 100,000.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">neighbors_within_batch</span> <span class="o">=</span> <span class="mi">25</span> <span class="k">if</span> <span class="n">adata_hvg</span><span class="o">.</span><span class="n">n_obs</span> <span class="o">&gt;</span> <span class="mi">100000</span> <span class="k">else</span> <span class="mi">3</span>
<span class="n">neighbors_within_batch</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>3
</pre></div>
</div>
</div>
</div>
<p>Before using <strong>BBKNN</strong> we first perform a PCA as we would before building a normal KNN graph. Unlike <strong>scVI</strong> which models raw counts here, we start with the log-normalised expression matrix.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">adata_bbknn</span> <span class="o">=</span> <span class="n">adata_hvg</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
<span class="n">adata_bbknn</span><span class="o">.</span><span class="n">X</span> <span class="o">=</span> <span class="n">adata_bbknn</span><span class="o">.</span><span class="n">layers</span><span class="p">[</span><span class="s2">&quot;logcounts&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
<span class="n">sc</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">pca</span><span class="p">(</span><span class="n">adata_bbknn</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<p>We can now run <strong>BBKNN</strong>, replacing the call to the <strong>scanpy</strong> <code class="docutils literal notranslate"><span class="pre">neighbors()</span></code> function in a standard workflow. An important difference is to make sure the <code class="docutils literal notranslate"><span class="pre">batch_key</span></code> argument is set which specifies a column in <code class="docutils literal notranslate"><span class="pre">adata_hvg.obs</span></code> that contains batch labels.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">bbknn</span><span class="o">.</span><span class="n">bbknn</span><span class="p">(</span>
    <span class="n">adata_bbknn</span><span class="p">,</span> <span class="n">batch_key</span><span class="o">=</span><span class="n">batch_key</span><span class="p">,</span> <span class="n">neighbors_within_batch</span><span class="o">=</span><span class="n">neighbors_within_batch</span>
<span class="p">)</span>
<span class="n">adata_bbknn</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>AnnData object with n_obs × n_vars = 10270 × 2000
    obs: &#39;GEX_pct_counts_mt&#39;, &#39;GEX_n_counts&#39;, &#39;GEX_n_genes&#39;, &#39;GEX_size_factors&#39;, &#39;GEX_phase&#39;, &#39;ATAC_nCount_peaks&#39;, &#39;ATAC_atac_fragments&#39;, &#39;ATAC_reads_in_peaks_frac&#39;, &#39;ATAC_blacklist_fraction&#39;, &#39;ATAC_nucleosome_signal&#39;, &#39;cell_type&#39;, &#39;batch&#39;, &#39;ATAC_pseudotime_order&#39;, &#39;GEX_pseudotime_order&#39;, &#39;Samplename&#39;, &#39;Site&#39;, &#39;DonorNumber&#39;, &#39;Modality&#39;, &#39;VendorLot&#39;, &#39;DonorID&#39;, &#39;DonorAge&#39;, &#39;DonorBMI&#39;, &#39;DonorBloodType&#39;, &#39;DonorRace&#39;, &#39;Ethnicity&#39;, &#39;DonorGender&#39;, &#39;QCMeds&#39;, &#39;DonorSmoker&#39;
    var: &#39;feature_types&#39;, &#39;gene_id&#39;, &#39;n_cells&#39;, &#39;highly_variable&#39;, &#39;means&#39;, &#39;dispersions&#39;, &#39;dispersions_norm&#39;, &#39;highly_variable_nbatches&#39;, &#39;highly_variable_intersection&#39;
    uns: &#39;ATAC_gene_activity_var_names&#39;, &#39;dataset_id&#39;, &#39;genome&#39;, &#39;organism&#39;, &#39;log1p&#39;, &#39;hvg&#39;, &#39;pca&#39;, &#39;neighbors&#39;, &#39;umap&#39;, &#39;batch_colors&#39;, &#39;cell_type_colors&#39;
    obsm: &#39;ATAC_gene_activity&#39;, &#39;ATAC_lsi_full&#39;, &#39;ATAC_lsi_red&#39;, &#39;ATAC_umap&#39;, &#39;GEX_X_pca&#39;, &#39;GEX_X_umap&#39;, &#39;X_pca&#39;, &#39;X_umap&#39;
    varm: &#39;PCs&#39;
    layers: &#39;counts&#39;, &#39;logcounts&#39;
    obsp: &#39;distances&#39;, &#39;connectivities&#39;
</pre></div>
</div>
</div>
</div>
<p>Unlike the default <strong>scanpy</strong> function, <strong>BBKNN</strong> does not allow specifying a key for storing results so they are always stored under the default “neighbors” key.</p>
<p>We can use this new integrated graph just like we would a normal KNN graph to construct a UMAP embedding.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">sc</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">umap</span><span class="p">(</span><span class="n">adata_bbknn</span><span class="p">)</span>
<span class="n">sc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">umap</span><span class="p">(</span><span class="n">adata_bbknn</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="p">[</span><span class="n">label_key</span><span class="p">,</span> <span class="n">batch_key</span><span class="p">],</span> <span class="n">wspace</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>/Users/luke.zappia/miniconda/envs/bp-integration/lib/python3.9/site-packages/scanpy/plotting/_tools/scatterplots.py:163: PendingDeprecationWarning: The get_cmap function will be deprecated in a future version. Use ``matplotlib.colormaps[name]`` or ``matplotlib.colormaps.get_cmap(obj)`` instead.
  cmap = copy(get_cmap(cmap))
/Users/luke.zappia/miniconda/envs/bp-integration/lib/python3.9/site-packages/scanpy/plotting/_tools/scatterplots.py:392: UserWarning: No data for colormapping provided via &#39;c&#39;. Parameters &#39;cmap&#39; will be ignored
  cax = scatter(
/Users/luke.zappia/miniconda/envs/bp-integration/lib/python3.9/site-packages/scanpy/plotting/_tools/scatterplots.py:392: UserWarning: No data for colormapping provided via &#39;c&#39;. Parameters &#39;cmap&#39; will be ignored
  cax = scatter(
</pre></div>
</div>
<img alt="../_images/cf3c4dddae9af800a23218887dbd9684fc4639658d2066de0fff02b5262f9e47.png" src="../_images/cf3c4dddae9af800a23218887dbd9684fc4639658d2066de0fff02b5262f9e47.png" />
</div>
</div>
<p>This integration is also improved compared to the unintegrated data with cell identities grouped together but we still see some shifts between batches.</p>
</section>
<section id="linear-embedding-integration-using-mutual-nearest-neighbors-mnn">
<h2><span class="section-number">12.8. </span>Linear embedding integration using Mutual Nearest Neighbors (MNN)<a class="headerlink" href="#linear-embedding-integration-using-mutual-nearest-neighbors-mnn" title="Link to this heading">#</a></h2>
<p>Some downstream applications cannot accept an integrated embedding or neighborhood graph and require a corrected expression matrix. One approach that can produce this output is the integration method in <strong>Seurat</strong> <span id="id38">[<a class="reference internal" href="#id211" title="Andrew Butler, Paul Hoffman, Peter Smibert, Efthymia Papalexi, and Rahul Satija. Integrating single-cell transcriptomic data across different conditions, technologies, and species. Nature biotechnology, April 2018. doi:10.1038/nbt.4096.">Butler <em>et al.</em>, 2018</a>, <a class="reference internal" href="#id213" title="Rahul Satija, Jeffrey A Farrell, David Gennert, Alexander F Schier, and Aviv Regev. Spatial reconstruction of single-cell gene expression data. Nature Biotechnology, 33(5):495–502, April 2015. doi:10.1038/nbt.3192.">Satija <em>et al.</em>, 2015</a>, <a class="reference internal" href="#id208" title="Tim Stuart, Andrew Butler, Paul Hoffman, Christoph Hafemeister, Efthymia Papalexi, William M Mauck, 3rd, Yuhan Hao, Marlon Stoeckius, Peter Smibert, and Rahul Satija. Comprehensive Integration of Single-Cell Data. Cell, 177(7):1888–1902.e21, June 2019. doi:10.1016/j.cell.2019.05.031.">Stuart <em>et al.</em>, 2019</a>]</span>. The <strong>Seurat</strong> integration method belongs to a class of <em>linear embedding models</em> that make use of the idea of <em>mutual nearest neighbors</em> (which <strong>Seurat</strong> calls <em>anchors</em>) to correct batch effects <span id="id39">[<a class="reference internal" href="#id215" title="Laleh Haghverdi, Aaron T L Lun, Michael D Morgan, and John C Marioni. Batch effects in single-cell RNA-sequencing data are corrected by matching mutual nearest neighbors. Nature biotechnology, April 2018. doi:10.1038/nbt.4091.">Haghverdi <em>et al.</em>, 2018</a>]</span>. Mutual nearest neighbors are pairs of cells from two different datasets which are in the neighborhood of each other when the datasets are placed in the same (latent) space. After finding these cells they can be used to align the two datasets and correct the differences between them. <strong>Seurat</strong> has also been found to be one of the top mixing methods in some evaluations <span id="id40">[<a class="reference internal" href="#id209" title="Hoa Thi Nhu Tran, Kok Siong Ang, Marion Chevrier, Xiaomeng Zhang, Nicole Yee Shin Lee, Michelle Goh, and Jinmiao Chen. A benchmark of batch-effect correction methods for single-cell RNA sequencing data. Genome biology, 21(1):12, January 2020. doi:10.1186/s13059-019-1850-9.">Tran <em>et al.</em>, 2020</a>]</span>.</p>
<p>As <strong>Seurat</strong> is an R package we must transfer our data from Python to R. Here we prepare the AnnData to convert so that it can be handled by <strong>rpy2</strong> and <strong>anndata2ri</strong>.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">adata_seurat</span> <span class="o">=</span> <span class="n">adata_hvg</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
<span class="c1"># Convert categorical columns to strings</span>
<span class="n">adata_seurat</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="n">batch_key</span><span class="p">]</span> <span class="o">=</span> <span class="n">adata_seurat</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="n">batch_key</span><span class="p">]</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="nb">str</span><span class="p">)</span>
<span class="n">adata_seurat</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="n">label_key</span><span class="p">]</span> <span class="o">=</span> <span class="n">adata_seurat</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="n">label_key</span><span class="p">]</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="nb">str</span><span class="p">)</span>
<span class="c1"># Delete uns as this can contain arbitrary objects which are difficult to convert</span>
<span class="k">del</span> <span class="n">adata_seurat</span><span class="o">.</span><span class="n">uns</span>
<span class="n">adata_seurat</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>AnnData object with n_obs × n_vars = 10270 × 2000
    obs: &#39;GEX_pct_counts_mt&#39;, &#39;GEX_n_counts&#39;, &#39;GEX_n_genes&#39;, &#39;GEX_size_factors&#39;, &#39;GEX_phase&#39;, &#39;ATAC_nCount_peaks&#39;, &#39;ATAC_atac_fragments&#39;, &#39;ATAC_reads_in_peaks_frac&#39;, &#39;ATAC_blacklist_fraction&#39;, &#39;ATAC_nucleosome_signal&#39;, &#39;cell_type&#39;, &#39;batch&#39;, &#39;ATAC_pseudotime_order&#39;, &#39;GEX_pseudotime_order&#39;, &#39;Samplename&#39;, &#39;Site&#39;, &#39;DonorNumber&#39;, &#39;Modality&#39;, &#39;VendorLot&#39;, &#39;DonorID&#39;, &#39;DonorAge&#39;, &#39;DonorBMI&#39;, &#39;DonorBloodType&#39;, &#39;DonorRace&#39;, &#39;Ethnicity&#39;, &#39;DonorGender&#39;, &#39;QCMeds&#39;, &#39;DonorSmoker&#39;
    var: &#39;feature_types&#39;, &#39;gene_id&#39;, &#39;n_cells&#39;, &#39;highly_variable&#39;, &#39;means&#39;, &#39;dispersions&#39;, &#39;dispersions_norm&#39;, &#39;highly_variable_nbatches&#39;, &#39;highly_variable_intersection&#39;
    obsm: &#39;ATAC_gene_activity&#39;, &#39;ATAC_lsi_full&#39;, &#39;ATAC_lsi_red&#39;, &#39;ATAC_umap&#39;, &#39;GEX_X_pca&#39;, &#39;GEX_X_umap&#39;, &#39;X_pca&#39;, &#39;X_umap&#39;
    varm: &#39;PCs&#39;
    layers: &#39;counts&#39;, &#39;logcounts&#39;
    obsp: &#39;distances&#39;, &#39;connectivities&#39;
</pre></div>
</div>
</div>
</div>
<p>The prepared AnnData is now available in R as a SingleCellExperiment object thanks to <strong>anndata2ri</strong>. Note that this is transposed compared to an AnnData object so our observations (cells) are now the columns and our variables (genes) are now the rows.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="o">%%</span><span class="k">R</span> -i adata_seurat
adata_seurat
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>/Users/luke.zappia/miniconda/envs/bp-integration/lib/python3.9/site-packages/rpy2/robjects/pandas2ri.py:263: DeprecationWarning: The global conversion available with activate() is deprecated and will be removed in the next major release. Use a local converter.
  warnings.warn(&#39;The global conversion available with activate() &#39;
/Users/luke.zappia/miniconda/envs/bp-integration/lib/python3.9/site-packages/rpy2/robjects/numpy2ri.py:205: DeprecationWarning: The global conversion available with activate() is deprecated and will be removed in the next major release. Use a local converter.
  warnings.warn(&#39;The global conversion available with activate() &#39;
/Users/luke.zappia/miniconda/envs/bp-integration/lib/python3.9/site-packages/rpy2/robjects/pandas2ri.py:263: DeprecationWarning: The global conversion available with activate() is deprecated and will be removed in the next major release. Use a local converter.
  warnings.warn(&#39;The global conversion available with activate() &#39;
/Users/luke.zappia/miniconda/envs/bp-integration/lib/python3.9/site-packages/rpy2/robjects/numpy2ri.py:205: DeprecationWarning: The global conversion available with activate() is deprecated and will be removed in the next major release. Use a local converter.
  warnings.warn(&#39;The global conversion available with activate() &#39;
/Users/luke.zappia/miniconda/envs/bp-integration/lib/python3.9/site-packages/anndata2ri/r2py.py:106: FutureWarning: X.dtype being converted to np.float32 from float64. In the next version of anndata (0.9) conversion will not be automatic. Pass dtype explicitly to avoid this warning. Pass `AnnData(X, dtype=X.dtype, ...)` to get the future behavour.
  return AnnData(exprs, obs, var, uns, obsm or None, layers=layers)
/Users/luke.zappia/miniconda/envs/bp-integration/lib/python3.9/site-packages/rpy2/robjects/pandas2ri.py:263: DeprecationWarning: The global conversion available with activate() is deprecated and will be removed in the next major release. Use a local converter.
  warnings.warn(&#39;The global conversion available with activate() &#39;
/Users/luke.zappia/miniconda/envs/bp-integration/lib/python3.9/site-packages/rpy2/robjects/numpy2ri.py:205: DeprecationWarning: The global conversion available with activate() is deprecated and will be removed in the next major release. Use a local converter.
  warnings.warn(&#39;The global conversion available with activate() &#39;
/Users/luke.zappia/miniconda/envs/bp-integration/lib/python3.9/site-packages/anndata2ri/r2py.py:106: FutureWarning: X.dtype being converted to np.float32 from float64. In the next version of anndata (0.9) conversion will not be automatic. Pass dtype explicitly to avoid this warning. Pass `AnnData(X, dtype=X.dtype, ...)` to get the future behavour.
  return AnnData(exprs, obs, var, uns, obsm or None, layers=layers)
</pre></div>
</div>
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>class: SingleCellExperiment 
dim: 2000 10270 
metadata(0):
assays(3): X counts logcounts
rownames(2000): GPR153 TNFRSF25 ... TMLHE-AS1 MT-ND3
rowData names(9): feature_types gene_id ... highly_variable_nbatches
  highly_variable_intersection
colnames(10270): TCACCTGGTTAGGTTG-3-s1d3 CGTTAACAGGTGTCCA-3-s1d3 ...
  AGCAGGTAGGCTATGT-12-s3d7 GCCATGATCCCTTGCG-12-s3d7
colData names(28): GEX_pct_counts_mt GEX_n_counts ... QCMeds
  DonorSmoker
reducedDimNames(8): ATAC_gene_activity ATAC_lsi_full ... PCA UMAP
mainExpName: NULL
altExpNames(0):
</pre></div>
</div>
</div>
</div>
<p><strong>Seurat</strong> uses its own object to store data. Helpfully the authors provide a function to convert from SingleCellExperiment. We just provide the SingleCellExperiment object and tell <strong>Seurat</strong> which assays (layers in our AnnData object) contain raw counts and normalised expression (which <strong>Seurat</strong> stores in a slot called “data”).</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="o">%%</span><span class="k">R</span> -i adata_seurat
seurat &lt;- as.Seurat(adata_seurat, counts = &quot;counts&quot;, data = &quot;logcounts&quot;)
seurat
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>/Users/luke.zappia/miniconda/envs/bp-integration/lib/python3.9/site-packages/rpy2/robjects/pandas2ri.py:263: DeprecationWarning: The global conversion available with activate() is deprecated and will be removed in the next major release. Use a local converter.
  warnings.warn(&#39;The global conversion available with activate() &#39;
/Users/luke.zappia/miniconda/envs/bp-integration/lib/python3.9/site-packages/rpy2/robjects/numpy2ri.py:205: DeprecationWarning: The global conversion available with activate() is deprecated and will be removed in the next major release. Use a local converter.
  warnings.warn(&#39;The global conversion available with activate() &#39;
/Users/luke.zappia/miniconda/envs/bp-integration/lib/python3.9/site-packages/anndata2ri/r2py.py:106: FutureWarning: X.dtype being converted to np.float32 from float64. In the next version of anndata (0.9) conversion will not be automatic. Pass dtype explicitly to avoid this warning. Pass `AnnData(X, dtype=X.dtype, ...)` to get the future behavour.
  return AnnData(exprs, obs, var, uns, obsm or None, layers=layers)
</pre></div>
</div>
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>An object of class Seurat 
2000 features across 10270 samples within 1 assay 
Active assay: originalexp (2000 features, 0 variable features)
 8 dimensional reductions calculated: ATAC_gene_activity, ATAC_lsi_full, ATAC_lsi_red, ATAC_umap, GEX_X_pca, GEX_X_umap, PCA, UMAP
</pre></div>
</div>
</div>
</div>
<p>Unlike some of the other methods, we have seen which take a single object and a batch key, the <strong>Seurat</strong> integration functions require a list of objects. We create this using the <code class="docutils literal notranslate"><span class="pre">SplitObject()</span></code> function.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="o">%%</span><span class="k">R</span> -i batch_key
batch_list &lt;- SplitObject(seurat, split.by = batch_key)
batch_list
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>$s1d3
An object of class Seurat 
2000 features across 4279 samples within 1 assay 
Active assay: originalexp (2000 features, 0 variable features)
 8 dimensional reductions calculated: ATAC_gene_activity, ATAC_lsi_full, ATAC_lsi_red, ATAC_umap, GEX_X_pca, GEX_X_umap, PCA, UMAP

$s2d1
An object of class Seurat 
2000 features across 4220 samples within 1 assay 
Active assay: originalexp (2000 features, 0 variable features)
 8 dimensional reductions calculated: ATAC_gene_activity, ATAC_lsi_full, ATAC_lsi_red, ATAC_umap, GEX_X_pca, GEX_X_umap, PCA, UMAP

$s3d7
An object of class Seurat 
2000 features across 1771 samples within 1 assay 
Active assay: originalexp (2000 features, 0 variable features)
 8 dimensional reductions calculated: ATAC_gene_activity, ATAC_lsi_full, ATAC_lsi_red, ATAC_umap, GEX_X_pca, GEX_X_umap, PCA, UMAP
</pre></div>
</div>
</div>
</div>
<p>We can now use this list to find anchors for each pair of datasets. Usually, you would identify batch-aware highly variable genes first (using the <code class="docutils literal notranslate"><span class="pre">FindVariableFeatures()</span></code> and <code class="docutils literal notranslate"><span class="pre">SelectIntegrationFeatures()</span></code> functions) but as we have done that already we tell <strong>Seurat</strong> to use all the features in the object.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="o">%%</span><span class="k">R</span>
anchors &lt;- FindIntegrationAnchors(batch_list, anchor.features = rownames(seurat))
anchors
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>  |                                                  | 0 % ~calculating   |+++++++++++++++++                                 | 33% ~03s           |++++++++++++++++++++++++++++++++++                | 67% ~01s           |++++++++++++++++++++++++++++++++++++++++++++++++++| 100% elapsed=02s  
  |                                                  | 0 % ~calculating   |+++++++++++++++++                                 | 33% ~01m 39s       |++++++++++++++++++++++++++++++++++                | 67% ~41s           |++++++++++++++++++++++++++++++++++++++++++++++++++| 100% elapsed=01m 56s
An AnchorSet object containing 25352 anchors between 3 Seurat objects 
 This can be used as input to IntegrateData.
</pre></div>
</div>
</div>
</div>
<p><strong>Seurat</strong> can then use the anchors to compute a transformation that maps one dataset onto another. This is done in a pairwise way until all the datasets are merged. By default <strong>Seurat</strong> will determine a merge order so that more similar datasets are merged together first but it is also possible to define this order.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="o">%%</span><span class="k">R</span>
integrated &lt;- IntegrateData(anchors)
integrated
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>An object of class Seurat 
4000 features across 10270 samples within 2 assays 
Active assay: integrated (2000 features, 2000 variable features)
 1 other assay present: originalexp
</pre></div>
</div>
</div>
</div>
<p>The result is another Seurat object, but notice now that the active assay is called “integrated”. This contains the corrected expression matrix which is the final output of the integration.</p>
<p>Here we extract that matrix and prepare it for transfer back to Python.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="o">%%</span><span class="k">R</span> -o integrated_expr
# Extract the integrated expression matrix
integrated_expr &lt;- GetAssayData(integrated)
# Make sure the rows and columns are in the same order as the original object
integrated_expr &lt;- integrated_expr[rownames(seurat), colnames(seurat)]
# Transpose the matrix to AnnData format
integrated_expr &lt;- t(integrated_expr)
print(integrated_expr[1:10, 1:10])
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>10 x 10 sparse Matrix of class &quot;dgCMatrix&quot;
                                                                               
TCACCTGGTTAGGTTG-3-s1d3  .            -0.0005365199  1.032812e-02 -2.653187e-02
CGTTAACAGGTGTCCA-3-s1d3  0.0001382038 -0.1809919733 -1.454901e-02  3.608087e-03
ATTCGTTTCAGTATTG-3-s1d3 -0.0121073040 -0.0634131457  .             2.144075e-02
GGACCGAAGTGAGGTA-3-s1d3  .             .             2.972292e-04  .           
ATGAAGCCAGGGAGCT-3-s1d3 -0.0139047071 -0.0313151274  .             2.239855e-02
AGTGCGGAGTAAGGGC-3-s1d3 -0.0004299227 -0.0002657828  .            -1.871410e-03
CTACCTCAGACACCGC-3-s1d3 -0.0055208620 -0.0398862297  7.182253e-06  8.240406e-03
CTTCAATTCACGAATC-3-s1d3  .            -0.0109928448  .             1.935677e-04
CCATTGTGTAGACAAA-3-s1d3  .             0.0171909615  .             5.711311e-05
CCGTTACTCAATGTGC-3-s1d3  0.0139905515  0.0007981117  2.303346e-03  1.356206e-02
                                                                          
TCACCTGGTTAGGTTG-3-s1d3 -0.023237586  0.031938502 -0.003196878  0.01777767
CGTTAACAGGTGTCCA-3-s1d3  0.114149766 -0.013183394  0.038076743  0.80491293
ATTCGTTTCAGTATTG-3-s1d3 -0.054419895  0.010955783 -0.005951634  0.37223306
GGACCGAAGTGAGGTA-3-s1d3  0.002305526  0.011544715  0.011133475  0.02366670
ATGAAGCCAGGGAGCT-3-s1d3 -0.123505733 -0.009382408  0.002153635 -0.07013584
AGTGCGGAGTAAGGGC-3-s1d3  0.035848768  0.013858992 -0.000379393  0.05617137
CTACCTCAGACACCGC-3-s1d3 -0.003837952  0.082027580 -0.001109391 -0.06307772
CTTCAATTCACGAATC-3-s1d3  0.052970708  0.153601547  0.247920321 -0.01143158
CCATTGTGTAGACAAA-3-s1d3 -0.015445186  0.025763467 -0.003632830  0.02040172
CCGTTACTCAATGTGC-3-s1d3 -0.018025385  0.022560136  0.005755797  0.61496228
                                                   
TCACCTGGTTAGGTTG-3-s1d3 -6.661643e-03 -0.0183198213
CGTTAACAGGTGTCCA-3-s1d3  5.079864e-02  0.0394717101
ATTCGTTTCAGTATTG-3-s1d3  6.600434e-02  0.0009021682
GGACCGAAGTGAGGTA-3-s1d3  7.172740e-04  0.0095352521
ATGAAGCCAGGGAGCT-3-s1d3  1.226039e-01  0.0063816143
AGTGCGGAGTAAGGGC-3-s1d3 -1.830964e-03  0.0008381943
CTACCTCAGACACCGC-3-s1d3  8.559358e-01  0.0102285085
CTTCAATTCACGAATC-3-s1d3 -5.318167e-06 -0.0341661907
CCATTGTGTAGACAAA-3-s1d3 -6.362298e-03  0.0232357011
CCGTTACTCAATGTGC-3-s1d3 -4.910886e-02  0.0317359576
</pre></div>
</div>
</div>
</div>
<p>We will now store the corrected expression matrix as a layer in our AnnData object. We also set <code class="docutils literal notranslate"><span class="pre">adata.X</span></code> to use this matrix.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">adata_seurat</span><span class="o">.</span><span class="n">X</span> <span class="o">=</span> <span class="n">integrated_expr</span>
<span class="n">adata_seurat</span><span class="o">.</span><span class="n">layers</span><span class="p">[</span><span class="s2">&quot;seurat&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">integrated_expr</span>
<span class="nb">print</span><span class="p">(</span><span class="n">adata_seurat</span><span class="p">)</span>
<span class="n">adata</span><span class="o">.</span><span class="n">X</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>AnnData object with n_obs × n_vars = 10270 × 2000
    obs: &#39;GEX_pct_counts_mt&#39;, &#39;GEX_n_counts&#39;, &#39;GEX_n_genes&#39;, &#39;GEX_size_factors&#39;, &#39;GEX_phase&#39;, &#39;ATAC_nCount_peaks&#39;, &#39;ATAC_atac_fragments&#39;, &#39;ATAC_reads_in_peaks_frac&#39;, &#39;ATAC_blacklist_fraction&#39;, &#39;ATAC_nucleosome_signal&#39;, &#39;cell_type&#39;, &#39;batch&#39;, &#39;ATAC_pseudotime_order&#39;, &#39;GEX_pseudotime_order&#39;, &#39;Samplename&#39;, &#39;Site&#39;, &#39;DonorNumber&#39;, &#39;Modality&#39;, &#39;VendorLot&#39;, &#39;DonorID&#39;, &#39;DonorAge&#39;, &#39;DonorBMI&#39;, &#39;DonorBloodType&#39;, &#39;DonorRace&#39;, &#39;Ethnicity&#39;, &#39;DonorGender&#39;, &#39;QCMeds&#39;, &#39;DonorSmoker&#39;
    var: &#39;feature_types&#39;, &#39;gene_id&#39;, &#39;n_cells&#39;, &#39;highly_variable&#39;, &#39;means&#39;, &#39;dispersions&#39;, &#39;dispersions_norm&#39;, &#39;highly_variable_nbatches&#39;, &#39;highly_variable_intersection&#39;
    obsm: &#39;ATAC_gene_activity&#39;, &#39;ATAC_lsi_full&#39;, &#39;ATAC_lsi_red&#39;, &#39;ATAC_umap&#39;, &#39;GEX_X_pca&#39;, &#39;GEX_X_umap&#39;, &#39;X_pca&#39;, &#39;X_umap&#39;
    varm: &#39;PCs&#39;
    layers: &#39;counts&#39;, &#39;logcounts&#39;, &#39;seurat&#39;
    obsp: &#39;distances&#39;, &#39;connectivities&#39;
</pre></div>
</div>
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>&lt;10270x13431 sparse matrix of type &#39;&lt;class &#39;numpy.float32&#39;&gt;&#39;
	with 14348115 stored elements in Compressed Sparse Row format&gt;
</pre></div>
</div>
</div>
</div>
<p>Now that we have the results of our integration we can calculate a UMAP and plot it as we have for the other methods (we could also have done this in R).</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Reset the batch colours because we deleted them earlier</span>
<span class="n">adata_seurat</span><span class="o">.</span><span class="n">uns</span><span class="p">[</span><span class="n">batch_key</span> <span class="o">+</span> <span class="s2">&quot;_colors&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="p">[</span>
    <span class="s2">&quot;#1b9e77&quot;</span><span class="p">,</span>
    <span class="s2">&quot;#d95f02&quot;</span><span class="p">,</span>
    <span class="s2">&quot;#7570b3&quot;</span><span class="p">,</span>
<span class="p">]</span>
<span class="n">sc</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">pca</span><span class="p">(</span><span class="n">adata_seurat</span><span class="p">)</span>
<span class="n">sc</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">neighbors</span><span class="p">(</span><span class="n">adata_seurat</span><span class="p">)</span>
<span class="n">sc</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">umap</span><span class="p">(</span><span class="n">adata_seurat</span><span class="p">)</span>
<span class="n">sc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">umap</span><span class="p">(</span><span class="n">adata_seurat</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="p">[</span><span class="n">label_key</span><span class="p">,</span> <span class="n">batch_key</span><span class="p">],</span> <span class="n">wspace</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>/Users/luke.zappia/miniconda/envs/bp-integration/lib/python3.9/site-packages/scanpy/plotting/_tools/scatterplots.py:163: PendingDeprecationWarning: The get_cmap function will be deprecated in a future version. Use ``matplotlib.colormaps[name]`` or ``matplotlib.colormaps.get_cmap(obj)`` instead.
  cmap = copy(get_cmap(cmap))
/Users/luke.zappia/miniconda/envs/bp-integration/lib/python3.9/site-packages/scanpy/plotting/_tools/scatterplots.py:392: UserWarning: No data for colormapping provided via &#39;c&#39;. Parameters &#39;cmap&#39; will be ignored
  cax = scatter(
/Users/luke.zappia/miniconda/envs/bp-integration/lib/python3.9/site-packages/scanpy/plotting/_tools/scatterplots.py:392: UserWarning: No data for colormapping provided via &#39;c&#39;. Parameters &#39;cmap&#39; will be ignored
  cax = scatter(
</pre></div>
</div>
<img alt="../_images/31c555d9e6f54b51ed9e6522efb720e2253e0e4696088c4f59f2e4a8969604d4.png" src="../_images/31c555d9e6f54b51ed9e6522efb720e2253e0e4696088c4f59f2e4a8969604d4.png" />
</div>
</div>
<p>As we have previously seen, the batches are mixed while the labels are separated. It is tempting to select an integration based on the UMAPs but this does not fully represent the quality of an integration. In the next section, we present some approaches to more rigorously evaluate integration methods.</p>
<div class="admonition-a-note-on-scalability admonition">
<p class="admonition-title">A note on scalability</p>
<p>As you ran the different integration methods you may have noticed that <strong>scVI</strong> took the most time. While this is true for small datasets like the example shown here, <a class="reference external" href="https://www.nature.com/articles/s41592-021-01336-8/figures/13">benchmarks have shown</a> that <strong>scVI</strong> scales well for larger datasets. This is largely because the number of training epochs is adjusted for larger dataset sizes. MNN methods typically don’t scale as well, partly because they perform several pairwise integrations, so if you have 20 batches you are performing 20 integrations while other methods can consider all batches at once.</p>
</div>
</section>
<section id="benchmarking-your-own-integration">
<h2><span class="section-number">12.9. </span>Benchmarking your own integration<a class="headerlink" href="#benchmarking-your-own-integration" title="Link to this heading">#</a></h2>
<p>The methods demonstrated here are selected based on results from benchmarking experiments including the <a class="reference external" href="https://theislab.github.io/scib-reproducibility/">single-cell integration benchmarking project</a> <span id="id41">[<a class="reference internal" href="#id220" title="Malte D Luecken, M Büttner, K Chaichoompu, A Danese, M Interlandi, M F Mueller, D C Strobl, L Zappia, M Dugas, M Colomé-Tatché, and Fabian J Theis. Benchmarking atlas-level data integration in single-cell genomics. Nature methods, December 2021. doi:10.1038/s41592-021-01336-8.">Luecken <em>et al.</em>, 2021</a>]</span>. This project also produced a software package called <a class="reference external" href="https://www.github.com/theislab/scib"><strong>scib</strong></a> that can be used to run a range of integration methods as well as the metrics that were used for evaluation. In this section, we show how to use this package to evaluate the quality of an integration.</p>
<div class="admonition-what-is-the-ground-truth admonition">
<p class="admonition-title">What is the ground truth?</p>
<p>Some of these metrics, particularly those that evaluate the conservation of biological variation, require a known ground truth to compare to. Usually, this is a cell identity label but can sometimes be other information such as known trajectories. Because of this requirement, it is difficult to evaluate integration for a completely new dataset where it is unclear what biological signal should be preserved.</p>
</div>
<p>The <strong>scib</strong> metrics can be run individually but there are also wrappers for running multiple metrics at once. Here we run a subset of the metrics which are quick to compute using the <code class="docutils literal notranslate"><span class="pre">metrics_fast()</span></code> function. This function takes a few arguments: the original unintegrated dataset, the integrated dataset, a batch key and a label key. Depending on the output of the integration method we might also need to supply additional arguments, for example here we specify the embedding to use for <strong>scVI</strong> and <strong>scANVI</strong> with the <code class="docutils literal notranslate"><span class="pre">embed</span></code> argument. You can also control how some metrics are run with additional arguments. Also note that you may need to check that objects are formatted properly so that <strong>scIB</strong> can find the required information.</p>
<p>Let’s run the metrics for each of the integrations we have performed above, as well as the unintegrated data (after highly variable gene selection).</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">metrics_scvi</span> <span class="o">=</span> <span class="n">scib</span><span class="o">.</span><span class="n">metrics</span><span class="o">.</span><span class="n">metrics_fast</span><span class="p">(</span>
    <span class="n">adata</span><span class="p">,</span> <span class="n">adata_scvi</span><span class="p">,</span> <span class="n">batch_key</span><span class="p">,</span> <span class="n">label_key</span><span class="p">,</span> <span class="n">embed</span><span class="o">=</span><span class="s2">&quot;X_scVI&quot;</span>
<span class="p">)</span>
<span class="n">metrics_scanvi</span> <span class="o">=</span> <span class="n">scib</span><span class="o">.</span><span class="n">metrics</span><span class="o">.</span><span class="n">metrics_fast</span><span class="p">(</span>
    <span class="n">adata</span><span class="p">,</span> <span class="n">adata_scanvi</span><span class="p">,</span> <span class="n">batch_key</span><span class="p">,</span> <span class="n">label_key</span><span class="p">,</span> <span class="n">embed</span><span class="o">=</span><span class="s2">&quot;X_scANVI&quot;</span>
<span class="p">)</span>
<span class="n">metrics_bbknn</span> <span class="o">=</span> <span class="n">scib</span><span class="o">.</span><span class="n">metrics</span><span class="o">.</span><span class="n">metrics_fast</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">adata_bbknn</span><span class="p">,</span> <span class="n">batch_key</span><span class="p">,</span> <span class="n">label_key</span><span class="p">)</span>
<span class="n">metrics_seurat</span> <span class="o">=</span> <span class="n">scib</span><span class="o">.</span><span class="n">metrics</span><span class="o">.</span><span class="n">metrics_fast</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">adata_seurat</span><span class="p">,</span> <span class="n">batch_key</span><span class="p">,</span> <span class="n">label_key</span><span class="p">)</span>
<span class="n">metrics_hvg</span> <span class="o">=</span> <span class="n">scib</span><span class="o">.</span><span class="n">metrics</span><span class="o">.</span><span class="n">metrics_fast</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">adata_hvg</span><span class="p">,</span> <span class="n">batch_key</span><span class="p">,</span> <span class="n">label_key</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Silhouette score...
PC regression...
Isolated labels ASW...
Graph connectivity...
Silhouette score...
PC regression...
Isolated labels ASW...
Graph connectivity...
Silhouette score...
PC regression...
Isolated labels ASW...
Graph connectivity...
Silhouette score...
PC regression...
Isolated labels ASW...
Graph connectivity...
Silhouette score...
PC regression...
Isolated labels ASW...
Graph connectivity...
</pre></div>
</div>
</div>
</div>
<p>Here is an example of what one of the metrics results looks like for a single integration:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">metrics_hvg</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>0</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>NMI_cluster/label</th>
      <td>NaN</td>
    </tr>
    <tr>
      <th>ARI_cluster/label</th>
      <td>NaN</td>
    </tr>
    <tr>
      <th>ASW_label</th>
      <td>0.555775</td>
    </tr>
    <tr>
      <th>ASW_label/batch</th>
      <td>0.833656</td>
    </tr>
    <tr>
      <th>PCR_batch</th>
      <td>0.537850</td>
    </tr>
    <tr>
      <th>cell_cycle_conservation</th>
      <td>NaN</td>
    </tr>
    <tr>
      <th>isolated_label_F1</th>
      <td>NaN</td>
    </tr>
    <tr>
      <th>isolated_label_silhouette</th>
      <td>0.487975</td>
    </tr>
    <tr>
      <th>graph_conn</th>
      <td>0.985533</td>
    </tr>
    <tr>
      <th>kBET</th>
      <td>NaN</td>
    </tr>
    <tr>
      <th>iLISI</th>
      <td>NaN</td>
    </tr>
    <tr>
      <th>cLISI</th>
      <td>NaN</td>
    </tr>
    <tr>
      <th>hvg_overlap</th>
      <td>1.000000</td>
    </tr>
    <tr>
      <th>trajectory</th>
      <td>NaN</td>
    </tr>
  </tbody>
</table>
</div></div></div>
</div>
<p>Each row is a different metric and the values show the score for that metric. Scores are between 0 and 1, where 1 is a good performance and 0 is a poor performance (<strong>scib</strong> can also return unscaled scores for some metrics if required). Because we have only run the fast metrics here, some of the metrics have <code class="docutils literal notranslate"><span class="pre">NaN</span></code> scores. Also, note that some metrics cannot be used with some output formats which can also be a reason for <code class="docutils literal notranslate"><span class="pre">NaN</span></code> values being returned.</p>
<p>To compare the methods it is useful to have all the metrics results in one table. This code combines them and tidies them into a more convenient format.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Concatenate metrics results</span>
<span class="n">metrics</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">concat</span><span class="p">(</span>
    <span class="p">[</span><span class="n">metrics_scvi</span><span class="p">,</span> <span class="n">metrics_scanvi</span><span class="p">,</span> <span class="n">metrics_bbknn</span><span class="p">,</span> <span class="n">metrics_seurat</span><span class="p">,</span> <span class="n">metrics_hvg</span><span class="p">],</span>
    <span class="n">axis</span><span class="o">=</span><span class="s2">&quot;columns&quot;</span><span class="p">,</span>
<span class="p">)</span>
<span class="c1"># Set methods as column names</span>
<span class="n">metrics</span> <span class="o">=</span> <span class="n">metrics</span><span class="o">.</span><span class="n">set_axis</span><span class="p">(</span>
    <span class="p">[</span><span class="s2">&quot;scVI&quot;</span><span class="p">,</span> <span class="s2">&quot;scANVI&quot;</span><span class="p">,</span> <span class="s2">&quot;BBKNN&quot;</span><span class="p">,</span> <span class="s2">&quot;Seurat&quot;</span><span class="p">,</span> <span class="s2">&quot;Unintegrated&quot;</span><span class="p">],</span> <span class="n">axis</span><span class="o">=</span><span class="s2">&quot;columns&quot;</span>
<span class="p">)</span>
<span class="c1"># Select only the fast metrics</span>
<span class="n">metrics</span> <span class="o">=</span> <span class="n">metrics</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span>
    <span class="p">[</span>
        <span class="s2">&quot;ASW_label&quot;</span><span class="p">,</span>
        <span class="s2">&quot;ASW_label/batch&quot;</span><span class="p">,</span>
        <span class="s2">&quot;PCR_batch&quot;</span><span class="p">,</span>
        <span class="s2">&quot;isolated_label_silhouette&quot;</span><span class="p">,</span>
        <span class="s2">&quot;graph_conn&quot;</span><span class="p">,</span>
        <span class="s2">&quot;hvg_overlap&quot;</span><span class="p">,</span>
    <span class="p">],</span>
    <span class="p">:,</span>
<span class="p">]</span>
<span class="c1"># Transpose so that metrics are columns and methods are rows</span>
<span class="n">metrics</span> <span class="o">=</span> <span class="n">metrics</span><span class="o">.</span><span class="n">T</span>
<span class="c1"># Remove the HVG overlap metric because it&#39;s not relevant to embedding outputs</span>
<span class="n">metrics</span> <span class="o">=</span> <span class="n">metrics</span><span class="o">.</span><span class="n">drop</span><span class="p">(</span><span class="n">columns</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;hvg_overlap&quot;</span><span class="p">])</span>
<span class="n">metrics</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>ASW_label</th>
      <th>ASW_label/batch</th>
      <th>PCR_batch</th>
      <th>isolated_label_silhouette</th>
      <th>graph_conn</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>scVI</th>
      <td>0.570953</td>
      <td>0.907165</td>
      <td>0.913474</td>
      <td>0.560875</td>
      <td>0.982940</td>
    </tr>
    <tr>
      <th>scANVI</th>
      <td>0.587934</td>
      <td>0.905402</td>
      <td>0.887735</td>
      <td>0.556232</td>
      <td>0.983729</td>
    </tr>
    <tr>
      <th>BBKNN</th>
      <td>0.555469</td>
      <td>0.847772</td>
      <td>0.537850</td>
      <td>0.481913</td>
      <td>0.969514</td>
    </tr>
    <tr>
      <th>Seurat</th>
      <td>0.571231</td>
      <td>0.915237</td>
      <td>0.788894</td>
      <td>0.488585</td>
      <td>0.989366</td>
    </tr>
    <tr>
      <th>Unintegrated</th>
      <td>0.555775</td>
      <td>0.833656</td>
      <td>0.537850</td>
      <td>0.487975</td>
      <td>0.985533</td>
    </tr>
  </tbody>
</table>
</div></div></div>
</div>
<p>We now have all the scores in one table with metrics as columns and methods as rows. Styling the table with a gradient can make it easier to see the differences between scores.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">metrics</span><span class="o">.</span><span class="n">style</span><span class="o">.</span><span class="n">background_gradient</span><span class="p">(</span><span class="n">cmap</span><span class="o">=</span><span class="s2">&quot;Blues&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>/Users/luke.zappia/miniconda/envs/bp-integration/lib/python3.9/site-packages/pandas/io/formats/style.py:3930: PendingDeprecationWarning: The get_cmap function will be deprecated in a future version. Use ``matplotlib.colormaps[name]`` or ``matplotlib.colormaps.get_cmap(obj)`` instead.
  rgbas = plt.cm.get_cmap(cmap)(norm(gmap))
</pre></div>
</div>
<div class="output text_html"><style type="text/css">
#T_6f2b1_row0_col0 {
  background-color: #74b3d8;
  color: #000000;
}
#T_6f2b1_row0_col1 {
  background-color: #084a91;
  color: #f1f1f1;
}
#T_6f2b1_row0_col2, #T_6f2b1_row0_col3, #T_6f2b1_row1_col0, #T_6f2b1_row3_col1, #T_6f2b1_row3_col4 {
  background-color: #08306b;
  color: #f1f1f1;
}
#T_6f2b1_row0_col4 {
  background-color: #3484bf;
  color: #f1f1f1;
}
#T_6f2b1_row1_col1 {
  background-color: #084f99;
  color: #f1f1f1;
}
#T_6f2b1_row1_col2 {
  background-color: #084285;
  color: #f1f1f1;
}
#T_6f2b1_row1_col3 {
  background-color: #084082;
  color: #f1f1f1;
}
#T_6f2b1_row1_col4 {
  background-color: #2a7ab9;
  color: #f1f1f1;
}
#T_6f2b1_row2_col0, #T_6f2b1_row2_col2, #T_6f2b1_row2_col3, #T_6f2b1_row2_col4, #T_6f2b1_row4_col1, #T_6f2b1_row4_col2 {
  background-color: #f7fbff;
  color: #000000;
}
#T_6f2b1_row2_col1 {
  background-color: #d5e5f4;
  color: #000000;
}
#T_6f2b1_row3_col0 {
  background-color: #71b1d7;
  color: #f1f1f1;
}
#T_6f2b1_row3_col2 {
  background-color: #3686c0;
  color: #f1f1f1;
}
#T_6f2b1_row3_col3 {
  background-color: #e7f0fa;
  color: #000000;
}
#T_6f2b1_row4_col0 {
  background-color: #f5fafe;
  color: #000000;
}
#T_6f2b1_row4_col3 {
  background-color: #e8f1fa;
  color: #000000;
}
#T_6f2b1_row4_col4 {
  background-color: #1562a9;
  color: #f1f1f1;
}
</style>
<table id="T_6f2b1">
  <thead>
    <tr>
      <th class="blank level0" >&nbsp;</th>
      <th id="T_6f2b1_level0_col0" class="col_heading level0 col0" >ASW_label</th>
      <th id="T_6f2b1_level0_col1" class="col_heading level0 col1" >ASW_label/batch</th>
      <th id="T_6f2b1_level0_col2" class="col_heading level0 col2" >PCR_batch</th>
      <th id="T_6f2b1_level0_col3" class="col_heading level0 col3" >isolated_label_silhouette</th>
      <th id="T_6f2b1_level0_col4" class="col_heading level0 col4" >graph_conn</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th id="T_6f2b1_level0_row0" class="row_heading level0 row0" >scVI</th>
      <td id="T_6f2b1_row0_col0" class="data row0 col0" >0.570953</td>
      <td id="T_6f2b1_row0_col1" class="data row0 col1" >0.907165</td>
      <td id="T_6f2b1_row0_col2" class="data row0 col2" >0.913474</td>
      <td id="T_6f2b1_row0_col3" class="data row0 col3" >0.560875</td>
      <td id="T_6f2b1_row0_col4" class="data row0 col4" >0.982940</td>
    </tr>
    <tr>
      <th id="T_6f2b1_level0_row1" class="row_heading level0 row1" >scANVI</th>
      <td id="T_6f2b1_row1_col0" class="data row1 col0" >0.587934</td>
      <td id="T_6f2b1_row1_col1" class="data row1 col1" >0.905402</td>
      <td id="T_6f2b1_row1_col2" class="data row1 col2" >0.887735</td>
      <td id="T_6f2b1_row1_col3" class="data row1 col3" >0.556232</td>
      <td id="T_6f2b1_row1_col4" class="data row1 col4" >0.983729</td>
    </tr>
    <tr>
      <th id="T_6f2b1_level0_row2" class="row_heading level0 row2" >BBKNN</th>
      <td id="T_6f2b1_row2_col0" class="data row2 col0" >0.555469</td>
      <td id="T_6f2b1_row2_col1" class="data row2 col1" >0.847772</td>
      <td id="T_6f2b1_row2_col2" class="data row2 col2" >0.537850</td>
      <td id="T_6f2b1_row2_col3" class="data row2 col3" >0.481913</td>
      <td id="T_6f2b1_row2_col4" class="data row2 col4" >0.969514</td>
    </tr>
    <tr>
      <th id="T_6f2b1_level0_row3" class="row_heading level0 row3" >Seurat</th>
      <td id="T_6f2b1_row3_col0" class="data row3 col0" >0.571231</td>
      <td id="T_6f2b1_row3_col1" class="data row3 col1" >0.915237</td>
      <td id="T_6f2b1_row3_col2" class="data row3 col2" >0.788894</td>
      <td id="T_6f2b1_row3_col3" class="data row3 col3" >0.488585</td>
      <td id="T_6f2b1_row3_col4" class="data row3 col4" >0.989366</td>
    </tr>
    <tr>
      <th id="T_6f2b1_level0_row4" class="row_heading level0 row4" >Unintegrated</th>
      <td id="T_6f2b1_row4_col0" class="data row4 col0" >0.555775</td>
      <td id="T_6f2b1_row4_col1" class="data row4 col1" >0.833656</td>
      <td id="T_6f2b1_row4_col2" class="data row4 col2" >0.537850</td>
      <td id="T_6f2b1_row4_col3" class="data row4 col3" >0.487975</td>
      <td id="T_6f2b1_row4_col4" class="data row4 col4" >0.985533</td>
    </tr>
  </tbody>
</table>
</div></div>
</div>
<p>For some metrics, the scores tend to be in a relatively small range. To emphasise the differences between methods and place each metric on the same scale, we scale them so that the worst performer gets a score of 0, the best performer gets a score of 1 and the others are somewhere in between.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">metrics_scaled</span> <span class="o">=</span> <span class="p">(</span><span class="n">metrics</span> <span class="o">-</span> <span class="n">metrics</span><span class="o">.</span><span class="n">min</span><span class="p">())</span> <span class="o">/</span> <span class="p">(</span><span class="n">metrics</span><span class="o">.</span><span class="n">max</span><span class="p">()</span> <span class="o">-</span> <span class="n">metrics</span><span class="o">.</span><span class="n">min</span><span class="p">())</span>
<span class="n">metrics_scaled</span><span class="o">.</span><span class="n">style</span><span class="o">.</span><span class="n">background_gradient</span><span class="p">(</span><span class="n">cmap</span><span class="o">=</span><span class="s2">&quot;Blues&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><style type="text/css">
#T_de919_row0_col0 {
  background-color: #74b3d8;
  color: #000000;
}
#T_de919_row0_col1 {
  background-color: #084a91;
  color: #f1f1f1;
}
#T_de919_row0_col2, #T_de919_row0_col3, #T_de919_row1_col0, #T_de919_row3_col1, #T_de919_row3_col4 {
  background-color: #08306b;
  color: #f1f1f1;
}
#T_de919_row0_col4 {
  background-color: #3484bf;
  color: #f1f1f1;
}
#T_de919_row1_col1 {
  background-color: #084f99;
  color: #f1f1f1;
}
#T_de919_row1_col2 {
  background-color: #084285;
  color: #f1f1f1;
}
#T_de919_row1_col3 {
  background-color: #084082;
  color: #f1f1f1;
}
#T_de919_row1_col4 {
  background-color: #2a7ab9;
  color: #f1f1f1;
}
#T_de919_row2_col0, #T_de919_row2_col2, #T_de919_row2_col3, #T_de919_row2_col4, #T_de919_row4_col1, #T_de919_row4_col2 {
  background-color: #f7fbff;
  color: #000000;
}
#T_de919_row2_col1 {
  background-color: #d5e5f4;
  color: #000000;
}
#T_de919_row3_col0 {
  background-color: #71b1d7;
  color: #f1f1f1;
}
#T_de919_row3_col2 {
  background-color: #3686c0;
  color: #f1f1f1;
}
#T_de919_row3_col3 {
  background-color: #e7f0fa;
  color: #000000;
}
#T_de919_row4_col0 {
  background-color: #f5fafe;
  color: #000000;
}
#T_de919_row4_col3 {
  background-color: #e8f1fa;
  color: #000000;
}
#T_de919_row4_col4 {
  background-color: #1562a9;
  color: #f1f1f1;
}
</style>
<table id="T_de919">
  <thead>
    <tr>
      <th class="blank level0" >&nbsp;</th>
      <th id="T_de919_level0_col0" class="col_heading level0 col0" >ASW_label</th>
      <th id="T_de919_level0_col1" class="col_heading level0 col1" >ASW_label/batch</th>
      <th id="T_de919_level0_col2" class="col_heading level0 col2" >PCR_batch</th>
      <th id="T_de919_level0_col3" class="col_heading level0 col3" >isolated_label_silhouette</th>
      <th id="T_de919_level0_col4" class="col_heading level0 col4" >graph_conn</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th id="T_de919_level0_row0" class="row_heading level0 row0" >scVI</th>
      <td id="T_de919_row0_col0" class="data row0 col0" >0.476942</td>
      <td id="T_de919_row0_col1" class="data row0 col1" >0.901060</td>
      <td id="T_de919_row0_col2" class="data row0 col2" >1.000000</td>
      <td id="T_de919_row0_col3" class="data row0 col3" >1.000000</td>
      <td id="T_de919_row0_col4" class="data row0 col4" >0.676270</td>
    </tr>
    <tr>
      <th id="T_de919_level0_row1" class="row_heading level0 row1" >scANVI</th>
      <td id="T_de919_row1_col0" class="data row1 col0" >1.000000</td>
      <td id="T_de919_row1_col1" class="data row1 col1" >0.879454</td>
      <td id="T_de919_row1_col2" class="data row1 col2" >0.931475</td>
      <td id="T_de919_row1_col3" class="data row1 col3" >0.941194</td>
      <td id="T_de919_row1_col4" class="data row1 col4" >0.716018</td>
    </tr>
    <tr>
      <th id="T_de919_level0_row2" class="row_heading level0 row2" >BBKNN</th>
      <td id="T_de919_row2_col0" class="data row2 col0" >0.000000</td>
      <td id="T_de919_row2_col1" class="data row2 col1" >0.173036</td>
      <td id="T_de919_row2_col2" class="data row2 col2" >0.000000</td>
      <td id="T_de919_row2_col3" class="data row2 col3" >0.000000</td>
      <td id="T_de919_row2_col4" class="data row2 col4" >0.000000</td>
    </tr>
    <tr>
      <th id="T_de919_level0_row3" class="row_heading level0 row3" >Seurat</th>
      <td id="T_de919_row3_col0" class="data row3 col0" >0.485521</td>
      <td id="T_de919_row3_col1" class="data row3 col1" >1.000000</td>
      <td id="T_de919_row3_col2" class="data row3 col2" >0.668338</td>
      <td id="T_de919_row3_col3" class="data row3 col3" >0.084500</td>
      <td id="T_de919_row3_col4" class="data row3 col4" >1.000000</td>
    </tr>
    <tr>
      <th id="T_de919_level0_row4" class="row_heading level0 row4" >Unintegrated</th>
      <td id="T_de919_row4_col0" class="data row4 col0" >0.009437</td>
      <td id="T_de919_row4_col1" class="data row4 col1" >0.000000</td>
      <td id="T_de919_row4_col2" class="data row4 col2" >0.000000</td>
      <td id="T_de919_row4_col3" class="data row4 col3" >0.076776</td>
      <td id="T_de919_row4_col4" class="data row4 col4" >0.806917</td>
    </tr>
  </tbody>
</table>
</div></div>
</div>
<p>The values now better represent the differences between methods (and better match the colour scale). However, it is important to note that the scaled scores can only be used to compare the relative performance of this specific set of integrations. If we wanted to add another method we would need to perform the scaling again. We also can’t say that an integration is definitively “good”, only that it is better than the other methods we have tried. This scaling emphasises differences between methods. For example, if we had metric scores of 0.92, 0.94 and 0.96 these would be scaled to 0, 0.5 and 1.0. This makes the first method appear to score much worse, even though it is only slightly lower than the other two and still got a very high score. This effect is bigger when comparing a few methods and when they get similar raw scores. Whether you look at raw or scaled scores depends on whether you want to focus on absolute performance or the difference in performance between methods.</p>
<p>The evaluation metrics can be grouped into two categories, those that measure the removal of batch effects and those that measure the conservation of biological variation. We can calculate summary scores for each of these categories by taking the mean of the scaled values for each group. This kind of summary score wouldn’t make sense with raw values as some metrics consistently produce higher scores than others (and therefore have a greater effect on the mean).</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">metrics_scaled</span><span class="p">[</span><span class="s2">&quot;Batch&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">metrics_scaled</span><span class="p">[</span>
    <span class="p">[</span><span class="s2">&quot;ASW_label/batch&quot;</span><span class="p">,</span> <span class="s2">&quot;PCR_batch&quot;</span><span class="p">,</span> <span class="s2">&quot;graph_conn&quot;</span><span class="p">]</span>
<span class="p">]</span><span class="o">.</span><span class="n">mean</span><span class="p">(</span><span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
<span class="n">metrics_scaled</span><span class="p">[</span><span class="s2">&quot;Bio&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">metrics_scaled</span><span class="p">[[</span><span class="s2">&quot;ASW_label&quot;</span><span class="p">,</span> <span class="s2">&quot;isolated_label_silhouette&quot;</span><span class="p">]]</span><span class="o">.</span><span class="n">mean</span><span class="p">(</span>
    <span class="n">axis</span><span class="o">=</span><span class="mi">1</span>
<span class="p">)</span>
<span class="n">metrics_scaled</span><span class="o">.</span><span class="n">style</span><span class="o">.</span><span class="n">background_gradient</span><span class="p">(</span><span class="n">cmap</span><span class="o">=</span><span class="s2">&quot;Blues&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>/Users/luke.zappia/miniconda/envs/bp-integration/lib/python3.9/site-packages/pandas/io/formats/style.py:3930: PendingDeprecationWarning: The get_cmap function will be deprecated in a future version. Use ``matplotlib.colormaps[name]`` or ``matplotlib.colormaps.get_cmap(obj)`` instead.
  rgbas = plt.cm.get_cmap(cmap)(norm(gmap))
</pre></div>
</div>
<div class="output text_html"><style type="text/css">
#T_0e733_row0_col0 {
  background-color: #74b3d8;
  color: #000000;
}
#T_0e733_row0_col1 {
  background-color: #084a91;
  color: #f1f1f1;
}
#T_0e733_row0_col2, #T_0e733_row0_col3, #T_0e733_row1_col0, #T_0e733_row1_col6, #T_0e733_row3_col1, #T_0e733_row3_col4, #T_0e733_row3_col5 {
  background-color: #08306b;
  color: #f1f1f1;
}
#T_0e733_row0_col4 {
  background-color: #3484bf;
  color: #f1f1f1;
}
#T_0e733_row0_col5 {
  background-color: #083979;
  color: #f1f1f1;
}
#T_0e733_row0_col6 {
  background-color: #1f6eb3;
  color: #f1f1f1;
}
#T_0e733_row1_col1 {
  background-color: #084f99;
  color: #f1f1f1;
}
#T_0e733_row1_col2 {
  background-color: #084285;
  color: #f1f1f1;
}
#T_0e733_row1_col3 {
  background-color: #084082;
  color: #f1f1f1;
}
#T_0e733_row1_col4 {
  background-color: #2a7ab9;
  color: #f1f1f1;
}
#T_0e733_row1_col5 {
  background-color: #083e81;
  color: #f1f1f1;
}
#T_0e733_row2_col0, #T_0e733_row2_col2, #T_0e733_row2_col3, #T_0e733_row2_col4, #T_0e733_row2_col5, #T_0e733_row2_col6, #T_0e733_row4_col1, #T_0e733_row4_col2 {
  background-color: #f7fbff;
  color: #000000;
}
#T_0e733_row2_col1 {
  background-color: #d5e5f4;
  color: #000000;
}
#T_0e733_row3_col0 {
  background-color: #71b1d7;
  color: #f1f1f1;
}
#T_0e733_row3_col2 {
  background-color: #3686c0;
  color: #f1f1f1;
}
#T_0e733_row3_col3 {
  background-color: #e7f0fa;
  color: #000000;
}
#T_0e733_row3_col6 {
  background-color: #b8d5ea;
  color: #000000;
}
#T_0e733_row4_col0 {
  background-color: #f5fafe;
  color: #000000;
}
#T_0e733_row4_col3 {
  background-color: #e8f1fa;
  color: #000000;
}
#T_0e733_row4_col4 {
  background-color: #1562a9;
  color: #f1f1f1;
}
#T_0e733_row4_col5 {
  background-color: #c4daee;
  color: #000000;
}
#T_0e733_row4_col6 {
  background-color: #eef5fc;
  color: #000000;
}
</style>
<table id="T_0e733">
  <thead>
    <tr>
      <th class="blank level0" >&nbsp;</th>
      <th id="T_0e733_level0_col0" class="col_heading level0 col0" >ASW_label</th>
      <th id="T_0e733_level0_col1" class="col_heading level0 col1" >ASW_label/batch</th>
      <th id="T_0e733_level0_col2" class="col_heading level0 col2" >PCR_batch</th>
      <th id="T_0e733_level0_col3" class="col_heading level0 col3" >isolated_label_silhouette</th>
      <th id="T_0e733_level0_col4" class="col_heading level0 col4" >graph_conn</th>
      <th id="T_0e733_level0_col5" class="col_heading level0 col5" >Batch</th>
      <th id="T_0e733_level0_col6" class="col_heading level0 col6" >Bio</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th id="T_0e733_level0_row0" class="row_heading level0 row0" >scVI</th>
      <td id="T_0e733_row0_col0" class="data row0 col0" >0.476942</td>
      <td id="T_0e733_row0_col1" class="data row0 col1" >0.901060</td>
      <td id="T_0e733_row0_col2" class="data row0 col2" >1.000000</td>
      <td id="T_0e733_row0_col3" class="data row0 col3" >1.000000</td>
      <td id="T_0e733_row0_col4" class="data row0 col4" >0.676270</td>
      <td id="T_0e733_row0_col5" class="data row0 col5" >0.859110</td>
      <td id="T_0e733_row0_col6" class="data row0 col6" >0.738471</td>
    </tr>
    <tr>
      <th id="T_0e733_level0_row1" class="row_heading level0 row1" >scANVI</th>
      <td id="T_0e733_row1_col0" class="data row1 col0" >1.000000</td>
      <td id="T_0e733_row1_col1" class="data row1 col1" >0.879454</td>
      <td id="T_0e733_row1_col2" class="data row1 col2" >0.931475</td>
      <td id="T_0e733_row1_col3" class="data row1 col3" >0.941194</td>
      <td id="T_0e733_row1_col4" class="data row1 col4" >0.716018</td>
      <td id="T_0e733_row1_col5" class="data row1 col5" >0.842316</td>
      <td id="T_0e733_row1_col6" class="data row1 col6" >0.970597</td>
    </tr>
    <tr>
      <th id="T_0e733_level0_row2" class="row_heading level0 row2" >BBKNN</th>
      <td id="T_0e733_row2_col0" class="data row2 col0" >0.000000</td>
      <td id="T_0e733_row2_col1" class="data row2 col1" >0.173036</td>
      <td id="T_0e733_row2_col2" class="data row2 col2" >0.000000</td>
      <td id="T_0e733_row2_col3" class="data row2 col3" >0.000000</td>
      <td id="T_0e733_row2_col4" class="data row2 col4" >0.000000</td>
      <td id="T_0e733_row2_col5" class="data row2 col5" >0.057679</td>
      <td id="T_0e733_row2_col6" class="data row2 col6" >0.000000</td>
    </tr>
    <tr>
      <th id="T_0e733_level0_row3" class="row_heading level0 row3" >Seurat</th>
      <td id="T_0e733_row3_col0" class="data row3 col0" >0.485521</td>
      <td id="T_0e733_row3_col1" class="data row3 col1" >1.000000</td>
      <td id="T_0e733_row3_col2" class="data row3 col2" >0.668338</td>
      <td id="T_0e733_row3_col3" class="data row3 col3" >0.084500</td>
      <td id="T_0e733_row3_col4" class="data row3 col4" >1.000000</td>
      <td id="T_0e733_row3_col5" class="data row3 col5" >0.889446</td>
      <td id="T_0e733_row3_col6" class="data row3 col6" >0.285010</td>
    </tr>
    <tr>
      <th id="T_0e733_level0_row4" class="row_heading level0 row4" >Unintegrated</th>
      <td id="T_0e733_row4_col0" class="data row4 col0" >0.009437</td>
      <td id="T_0e733_row4_col1" class="data row4 col1" >0.000000</td>
      <td id="T_0e733_row4_col2" class="data row4 col2" >0.000000</td>
      <td id="T_0e733_row4_col3" class="data row4 col3" >0.076776</td>
      <td id="T_0e733_row4_col4" class="data row4 col4" >0.806917</td>
      <td id="T_0e733_row4_col5" class="data row4 col5" >0.268972</td>
      <td id="T_0e733_row4_col6" class="data row4 col6" >0.043106</td>
    </tr>
  </tbody>
</table>
</div></div>
</div>
<p>Plotting the two summary scores against each other gives an indication of the priorities of each method. Some will be biased towards batch correction while others will favour retaining biological variation.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">fig</span><span class="p">,</span> <span class="n">ax</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">subplots</span><span class="p">()</span>
<span class="n">ax</span><span class="o">.</span><span class="n">set_xlim</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="mi">1</span><span class="p">)</span>
<span class="n">ax</span><span class="o">.</span><span class="n">set_ylim</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="mi">1</span><span class="p">)</span>
<span class="n">metrics_scaled</span><span class="o">.</span><span class="n">plot</span><span class="o">.</span><span class="n">scatter</span><span class="p">(</span>
    <span class="n">x</span><span class="o">=</span><span class="s2">&quot;Batch&quot;</span><span class="p">,</span>
    <span class="n">y</span><span class="o">=</span><span class="s2">&quot;Bio&quot;</span><span class="p">,</span>
    <span class="n">c</span><span class="o">=</span><span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">metrics_scaled</span><span class="p">)),</span>
    <span class="n">ax</span><span class="o">=</span><span class="n">ax</span><span class="p">,</span>
<span class="p">)</span>

<span class="k">for</span> <span class="n">k</span><span class="p">,</span> <span class="n">v</span> <span class="ow">in</span> <span class="n">metrics_scaled</span><span class="p">[[</span><span class="s2">&quot;Batch&quot;</span><span class="p">,</span> <span class="s2">&quot;Bio&quot;</span><span class="p">]]</span><span class="o">.</span><span class="n">iterrows</span><span class="p">():</span>
    <span class="n">ax</span><span class="o">.</span><span class="n">annotate</span><span class="p">(</span>
        <span class="n">k</span><span class="p">,</span>
        <span class="n">v</span><span class="p">,</span>
        <span class="n">xytext</span><span class="o">=</span><span class="p">(</span><span class="mi">6</span><span class="p">,</span> <span class="o">-</span><span class="mi">3</span><span class="p">),</span>
        <span class="n">textcoords</span><span class="o">=</span><span class="s2">&quot;offset points&quot;</span><span class="p">,</span>
        <span class="n">family</span><span class="o">=</span><span class="s2">&quot;sans-serif&quot;</span><span class="p">,</span>
        <span class="n">fontsize</span><span class="o">=</span><span class="mi">12</span><span class="p">,</span>
    <span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="../_images/83f33a364c9b09be9f8ac6299b638db499d0fb06de66d991190b4155c9e0757d.png" src="../_images/83f33a364c9b09be9f8ac6299b638db499d0fb06de66d991190b4155c9e0757d.png" />
</div>
</div>
<p>In our small example scenario <strong>BBKNN</strong> is clearly the worst performer, getting the lowest scores for both batch removal and biological conservation. The other three methods have similar batch correction scores with <strong>scANVI</strong> scoring highest for biological conservation followed by <strong>scVI</strong> and <strong>Seurat</strong>.</p>
<p>To get an overall score for each method we can combine the two summary scores. The <strong>scIB</strong> paper suggests a weighting of 40% batch correction and 60% biological conservation but you may prefer to weight things differently depending on the priorities for your dataset.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">metrics_scaled</span><span class="p">[</span><span class="s2">&quot;Overall&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="mf">0.4</span> <span class="o">*</span> <span class="n">metrics_scaled</span><span class="p">[</span><span class="s2">&quot;Batch&quot;</span><span class="p">]</span> <span class="o">+</span> <span class="mf">0.6</span> <span class="o">*</span> <span class="n">metrics_scaled</span><span class="p">[</span><span class="s2">&quot;Bio&quot;</span><span class="p">]</span>
<span class="n">metrics_scaled</span><span class="o">.</span><span class="n">style</span><span class="o">.</span><span class="n">background_gradient</span><span class="p">(</span><span class="n">cmap</span><span class="o">=</span><span class="s2">&quot;Blues&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>/Users/luke.zappia/miniconda/envs/bp-integration/lib/python3.9/site-packages/pandas/io/formats/style.py:3930: PendingDeprecationWarning: The get_cmap function will be deprecated in a future version. Use ``matplotlib.colormaps[name]`` or ``matplotlib.colormaps.get_cmap(obj)`` instead.
  rgbas = plt.cm.get_cmap(cmap)(norm(gmap))
</pre></div>
</div>
<div class="output text_html"><style type="text/css">
#T_f3d3d_row0_col0 {
  background-color: #74b3d8;
  color: #000000;
}
#T_f3d3d_row0_col1 {
  background-color: #084a91;
  color: #f1f1f1;
}
#T_f3d3d_row0_col2, #T_f3d3d_row0_col3, #T_f3d3d_row1_col0, #T_f3d3d_row1_col6, #T_f3d3d_row1_col7, #T_f3d3d_row3_col1, #T_f3d3d_row3_col4, #T_f3d3d_row3_col5 {
  background-color: #08306b;
  color: #f1f1f1;
}
#T_f3d3d_row0_col4 {
  background-color: #3484bf;
  color: #f1f1f1;
}
#T_f3d3d_row0_col5 {
  background-color: #083979;
  color: #f1f1f1;
}
#T_f3d3d_row0_col6 {
  background-color: #1f6eb3;
  color: #f1f1f1;
}
#T_f3d3d_row0_col7 {
  background-color: #0c56a0;
  color: #f1f1f1;
}
#T_f3d3d_row1_col1 {
  background-color: #084f99;
  color: #f1f1f1;
}
#T_f3d3d_row1_col2 {
  background-color: #084285;
  color: #f1f1f1;
}
#T_f3d3d_row1_col3 {
  background-color: #084082;
  color: #f1f1f1;
}
#T_f3d3d_row1_col4 {
  background-color: #2a7ab9;
  color: #f1f1f1;
}
#T_f3d3d_row1_col5 {
  background-color: #083e81;
  color: #f1f1f1;
}
#T_f3d3d_row2_col0, #T_f3d3d_row2_col2, #T_f3d3d_row2_col3, #T_f3d3d_row2_col4, #T_f3d3d_row2_col5, #T_f3d3d_row2_col6, #T_f3d3d_row2_col7, #T_f3d3d_row4_col1, #T_f3d3d_row4_col2 {
  background-color: #f7fbff;
  color: #000000;
}
#T_f3d3d_row2_col1 {
  background-color: #d5e5f4;
  color: #000000;
}
#T_f3d3d_row3_col0 {
  background-color: #71b1d7;
  color: #f1f1f1;
}
#T_f3d3d_row3_col2 {
  background-color: #3686c0;
  color: #f1f1f1;
}
#T_f3d3d_row3_col3 {
  background-color: #e7f0fa;
  color: #000000;
}
#T_f3d3d_row3_col6 {
  background-color: #b8d5ea;
  color: #000000;
}
#T_f3d3d_row3_col7 {
  background-color: #57a0ce;
  color: #f1f1f1;
}
#T_f3d3d_row4_col0 {
  background-color: #f5fafe;
  color: #000000;
}
#T_f3d3d_row4_col3 {
  background-color: #e8f1fa;
  color: #000000;
}
#T_f3d3d_row4_col4 {
  background-color: #1562a9;
  color: #f1f1f1;
}
#T_f3d3d_row4_col5 {
  background-color: #c4daee;
  color: #000000;
}
#T_f3d3d_row4_col6 {
  background-color: #eef5fc;
  color: #000000;
}
#T_f3d3d_row4_col7 {
  background-color: #dfebf7;
  color: #000000;
}
</style>
<table id="T_f3d3d">
  <thead>
    <tr>
      <th class="blank level0" >&nbsp;</th>
      <th id="T_f3d3d_level0_col0" class="col_heading level0 col0" >ASW_label</th>
      <th id="T_f3d3d_level0_col1" class="col_heading level0 col1" >ASW_label/batch</th>
      <th id="T_f3d3d_level0_col2" class="col_heading level0 col2" >PCR_batch</th>
      <th id="T_f3d3d_level0_col3" class="col_heading level0 col3" >isolated_label_silhouette</th>
      <th id="T_f3d3d_level0_col4" class="col_heading level0 col4" >graph_conn</th>
      <th id="T_f3d3d_level0_col5" class="col_heading level0 col5" >Batch</th>
      <th id="T_f3d3d_level0_col6" class="col_heading level0 col6" >Bio</th>
      <th id="T_f3d3d_level0_col7" class="col_heading level0 col7" >Overall</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th id="T_f3d3d_level0_row0" class="row_heading level0 row0" >scVI</th>
      <td id="T_f3d3d_row0_col0" class="data row0 col0" >0.476942</td>
      <td id="T_f3d3d_row0_col1" class="data row0 col1" >0.901060</td>
      <td id="T_f3d3d_row0_col2" class="data row0 col2" >1.000000</td>
      <td id="T_f3d3d_row0_col3" class="data row0 col3" >1.000000</td>
      <td id="T_f3d3d_row0_col4" class="data row0 col4" >0.676270</td>
      <td id="T_f3d3d_row0_col5" class="data row0 col5" >0.859110</td>
      <td id="T_f3d3d_row0_col6" class="data row0 col6" >0.738471</td>
      <td id="T_f3d3d_row0_col7" class="data row0 col7" >0.786726</td>
    </tr>
    <tr>
      <th id="T_f3d3d_level0_row1" class="row_heading level0 row1" >scANVI</th>
      <td id="T_f3d3d_row1_col0" class="data row1 col0" >1.000000</td>
      <td id="T_f3d3d_row1_col1" class="data row1 col1" >0.879454</td>
      <td id="T_f3d3d_row1_col2" class="data row1 col2" >0.931475</td>
      <td id="T_f3d3d_row1_col3" class="data row1 col3" >0.941194</td>
      <td id="T_f3d3d_row1_col4" class="data row1 col4" >0.716018</td>
      <td id="T_f3d3d_row1_col5" class="data row1 col5" >0.842316</td>
      <td id="T_f3d3d_row1_col6" class="data row1 col6" >0.970597</td>
      <td id="T_f3d3d_row1_col7" class="data row1 col7" >0.919285</td>
    </tr>
    <tr>
      <th id="T_f3d3d_level0_row2" class="row_heading level0 row2" >BBKNN</th>
      <td id="T_f3d3d_row2_col0" class="data row2 col0" >0.000000</td>
      <td id="T_f3d3d_row2_col1" class="data row2 col1" >0.173036</td>
      <td id="T_f3d3d_row2_col2" class="data row2 col2" >0.000000</td>
      <td id="T_f3d3d_row2_col3" class="data row2 col3" >0.000000</td>
      <td id="T_f3d3d_row2_col4" class="data row2 col4" >0.000000</td>
      <td id="T_f3d3d_row2_col5" class="data row2 col5" >0.057679</td>
      <td id="T_f3d3d_row2_col6" class="data row2 col6" >0.000000</td>
      <td id="T_f3d3d_row2_col7" class="data row2 col7" >0.023071</td>
    </tr>
    <tr>
      <th id="T_f3d3d_level0_row3" class="row_heading level0 row3" >Seurat</th>
      <td id="T_f3d3d_row3_col0" class="data row3 col0" >0.485521</td>
      <td id="T_f3d3d_row3_col1" class="data row3 col1" >1.000000</td>
      <td id="T_f3d3d_row3_col2" class="data row3 col2" >0.668338</td>
      <td id="T_f3d3d_row3_col3" class="data row3 col3" >0.084500</td>
      <td id="T_f3d3d_row3_col4" class="data row3 col4" >1.000000</td>
      <td id="T_f3d3d_row3_col5" class="data row3 col5" >0.889446</td>
      <td id="T_f3d3d_row3_col6" class="data row3 col6" >0.285010</td>
      <td id="T_f3d3d_row3_col7" class="data row3 col7" >0.526785</td>
    </tr>
    <tr>
      <th id="T_f3d3d_level0_row4" class="row_heading level0 row4" >Unintegrated</th>
      <td id="T_f3d3d_row4_col0" class="data row4 col0" >0.009437</td>
      <td id="T_f3d3d_row4_col1" class="data row4 col1" >0.000000</td>
      <td id="T_f3d3d_row4_col2" class="data row4 col2" >0.000000</td>
      <td id="T_f3d3d_row4_col3" class="data row4 col3" >0.076776</td>
      <td id="T_f3d3d_row4_col4" class="data row4 col4" >0.806917</td>
      <td id="T_f3d3d_row4_col5" class="data row4 col5" >0.268972</td>
      <td id="T_f3d3d_row4_col6" class="data row4 col6" >0.043106</td>
      <td id="T_f3d3d_row4_col7" class="data row4 col7" >0.133453</td>
    </tr>
  </tbody>
</table>
</div></div>
</div>
<p>Let’s make a quick bar chart to visualise the overall performance.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">metrics_scaled</span><span class="o">.</span><span class="n">plot</span><span class="o">.</span><span class="n">bar</span><span class="p">(</span><span class="n">y</span><span class="o">=</span><span class="s2">&quot;Overall&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>&lt;AxesSubplot: &gt;
</pre></div>
</div>
<img alt="../_images/a301c0818aba8302479b965e75b1249516c0fd6979b2a7c1a3833bbf04b01613.png" src="../_images/a301c0818aba8302479b965e75b1249516c0fd6979b2a7c1a3833bbf04b01613.png" />
</div>
</div>
<p>As we have already seen <strong>scVI</strong> and <strong>scANVI</strong> are the best performers with <strong>scANVI</strong> scoring slightly higher. It is important to note that this is just an example of how to run these metrics for this specific dataset, not a proper evaluation of these methods. For that, you should refer to existing benchmarking publications. In particular, we have only run a small selection of high-performing methods and a subset of metrics. Also, remember that scores are relative to the methods used so even if the methods perform almost equally well and small differences will be exaggerated.</p>
<p>Existing benchmarks have suggested methods that generally perform well, but performance can also be quite variable across scenarios. For some analyses, it may be worthwhile performing your own evaluation of integration. The <strong>scib</strong> package makes this process easier, but it can still be a significant undertaking, relying on a good knowledge of the ground truth and interpretation of the metrics.</p>
</section>
<section id="quiz">
<h2><span class="section-number">12.10. </span>Quiz<a class="headerlink" href="#quiz" title="Link to this heading">#</a></h2>
<div class="cell tag_remove_input docutils container">
<div class="cell_output docutils container">
<div class="output text_html">
    <style>
        .output.text_html {
            --pst-color-text-base: transparent !important;
        }
        .flip-card-q1 {
            background-color: transparent;
            width: 350px;
            height: 200px;
            perspective: 1000px;
            display: inline-block;
            margin: 10px;
            text-align: center;
            padding: 10px;
        }
        .flip-card-inner-q1 {
            position: relative;
            width: 100%;
            height: 100%;
            text-align: center;
            transition: transform 0.6s;
            transform-style: preserve-3d;
            transform-origin: center;
        }
        .flip-card-q1:hover .flip-card-inner-q1 {
            transform: rotateY(180deg);
        }
        .flip-card-front-q1, .flip-card-back-q1 {
            position: absolute;
            width: 100%;
            height: 100%;
            min-height: 200px;
            backface-visibility: hidden;
            display: flex;
            align-items: center;
            justify-content: center;
            font-weight: bold;
            color: white;
            padding: 20px;
            border-radius: 15px;
            box-sizing: border-box;
            overflow: hidden;
            text-align: center;
        }
        .flip-card-front-q1 {
            background-color: #3965a3;
            font-size: 20px;
        }
        .flip-card-back-q1 {
            background-color: #a8d480;
            transform: rotateY(180deg);
            font-size: 20px;
        }_b
    </style>

    <div class="flip-card-q1">
        <div class="flip-card-inner-q1">
            <div class="flip-card-front-q1">
                What are the sources of batch effects?
            </div>
            <div class="flip-card-back-q1">
                e.g. Differences in sample handling, experimental protocols, factors such as donor differences, tissue heterogeneity
            </div>
        </div>
    </div>
    <noscript>
        <p><strong>Q:</strong> What are the sources of batch effects?</p>
        <p><strong>Answer:</strong> e.g. Differences in sample handling, experimental protocols, factors such as donor differences, tissue heterogeneity</p>
    </noscript>
    </div><div class="output text_html">
    <style>
        .output.text_html {
            --pst-color-text-base: transparent !important;
        }
        .flip-card-q2 {
            background-color: transparent;
            width: 350px;
            height: 200px;
            perspective: 1000px;
            display: inline-block;
            margin: 10px;
            text-align: center;
            padding: 10px;
        }
        .flip-card-inner-q2 {
            position: relative;
            width: 100%;
            height: 100%;
            text-align: center;
            transition: transform 0.6s;
            transform-style: preserve-3d;
            transform-origin: center;
        }
        .flip-card-q2:hover .flip-card-inner-q2 {
            transform: rotateY(180deg);
        }
        .flip-card-front-q2, .flip-card-back-q2 {
            position: absolute;
            width: 100%;
            height: 100%;
            min-height: 200px;
            backface-visibility: hidden;
            display: flex;
            align-items: center;
            justify-content: center;
            font-weight: bold;
            color: white;
            padding: 20px;
            border-radius: 15px;
            box-sizing: border-box;
            overflow: hidden;
            text-align: center;
        }
        .flip-card-front-q2 {
            background-color: #3965a3;
            font-size: 20px;
        }
        .flip-card-back-q2 {
            background-color: #a8d480;
            transform: rotateY(180deg);
            font-size: 15px;
        }_b
    </style>

    <div class="flip-card-q2">
        <div class="flip-card-inner-q2">
            <div class="flip-card-front-q2">
                What is the difference between technical and biological variation?
            </div>
            <div class="flip-card-back-q2">
                Technical variation refers to inconsistencies arising from the experimental process itself. In contrast, biological variation encompasses intrinsic differences between samples.
            </div>
        </div>
    </div>
    <noscript>
        <p><strong>Q:</strong> What is the difference between technical and biological variation?</p>
        <p><strong>Answer:</strong> Technical variation refers to inconsistencies arising from the experimental process itself. In contrast, biological variation encompasses intrinsic differences between samples.</p>
    </noscript>
    </div><div class="output text_html">
    <style>
        .output.text_html {
            --pst-color-text-base: transparent !important;
        }
        .flip-card-q3 {
            background-color: transparent;
            width: 350px;
            height: 200px;
            perspective: 1000px;
            display: inline-block;
            margin: 10px;
            text-align: center;
            padding: 10px;
        }
        .flip-card-inner-q3 {
            position: relative;
            width: 100%;
            height: 100%;
            text-align: center;
            transition: transform 0.6s;
            transform-style: preserve-3d;
            transform-origin: center;
        }
        .flip-card-q3:hover .flip-card-inner-q3 {
            transform: rotateY(180deg);
        }
        .flip-card-front-q3, .flip-card-back-q3 {
            position: absolute;
            width: 100%;
            height: 100%;
            min-height: 200px;
            backface-visibility: hidden;
            display: flex;
            align-items: center;
            justify-content: center;
            font-weight: bold;
            color: white;
            padding: 20px;
            border-radius: 15px;
            box-sizing: border-box;
            overflow: hidden;
            text-align: center;
        }
        .flip-card-front-q3 {
            background-color: #3965a3;
            font-size: 20px;
        }
        .flip-card-back-q3 {
            background-color: #a8d480;
            transform: rotateY(180deg);
            font-size: 13px;
        }_b
    </style>

    <div class="flip-card-q3">
        <div class="flip-card-inner-q3">
            <div class="flip-card-front-q3">
                How does one evaluate whether the integration worked well or not? What are useful metrics for this purpose?
            </div>
            <div class="flip-card-back-q3">
                To evaluate integration effectiveness, visualization techniques like UMAP plots help assess whether cells from different batches mix well. Quantitative metrics include graph connectivity, batch entropy mixing, and silhouette scores. Combining these methods ensures that batch effects are minimized while preserving true biological signals.
            </div>
        </div>
    </div>
    <noscript>
        <p><strong>Q:</strong> How does one evaluate whether the integration worked well or not? What are useful metrics for this purpose?</p>
        <p><strong>Answer:</strong> To evaluate integration effectiveness, visualization techniques like UMAP plots help assess whether cells from different batches mix well. Quantitative metrics include graph connectivity, batch entropy mixing, and silhouette scores. Combining these methods ensures that batch effects are minimized while preserving true biological signals.</p>
    </noscript>
    </div></div>
</div>
</section>
<section id="session-information">
<h2><span class="section-number">12.11. </span>Session information<a class="headerlink" href="#session-information" title="Link to this heading">#</a></h2>
<section id="python">
<h3><span class="section-number">12.11.1. </span>Python<a class="headerlink" href="#python" title="Link to this heading">#</a></h3>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span><span class="w"> </span><span class="nn">session_info</span>

<span class="n">session_info</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><details>
<summary>Click to view session information</summary>
<pre>
-----
anndata             0.8.0
anndata2ri          1.1
bbknn               NA
matplotlib          3.6.1
numpy               1.23.4
pandas              1.5.1
rpy2                3.5.1
scanpy              1.9.1
scib                1.0.4
scipy               1.9.3
scvi                0.18.0
session_info        1.0.0
-----
</pre>
<details>
<summary>Click to view modules imported as dependencies</summary>
<pre>
PIL                         9.2.0
absl                        NA
annoy                       NA
appnope                     0.1.3
asttokens                   NA
attr                        22.1.0
backcall                    0.2.0
beta_ufunc                  NA
binom_ufunc                 NA
cffi                        1.15.1
chex                        0.1.5
colorama                    0.4.6
contextlib2                 NA
cycler                      0.10.0
cython_runtime              NA
dateutil                    2.8.2
debugpy                     1.6.3
decorator                   5.1.1
defusedxml                  0.7.1
deprecate                   0.3.2
deprecated                  1.2.13
docrep                      0.3.2
entrypoints                 0.4
etils                       0.8.0
executing                   1.1.1
flax                        0.5.0
fsspec                      2022.10.0
google                      NA
h5py                        3.7.0
hypergeom_ufunc             NA
igraph                      0.10.2
importlib_metadata          NA
ipykernel                   6.16.2
ipython_genutils            0.2.0
ipywidgets                  8.0.2
jax                         0.3.23
jaxlib                      0.3.22
jedi                        0.18.1
jinja2                      3.1.2
joblib                      1.2.0
jupyter_server              1.21.0
kiwisolver                  1.4.4
leidenalg                   0.9.0
llvmlite                    0.39.1
louvain                     0.8.0
markupsafe                  2.1.1
matplotlib_inline           0.1.6
ml_collections              NA
mpl_toolkits                NA
msgpack                     1.0.4
mudata                      0.2.0
multipledispatch            0.6.0
natsort                     8.2.0
nbinom_ufunc                NA
ncf_ufunc                   NA
numba                       0.56.3
numpyro                     0.10.1
opt_einsum                  v3.3.0
optax                       0.1.3
packaging                   21.3
parso                       0.8.3
pexpect                     4.8.0
pickleshare                 0.7.5
pkg_resources               NA
prompt_toolkit              3.0.31
psutil                      5.9.3
ptyprocess                  0.7.0
pure_eval                   0.2.2
pycparser                   2.21
pydev_ipython               NA
pydevconsole                NA
pydevd                      2.8.0
pydevd_file_utils           NA
pydevd_plugins              NA
pydevd_tracing              NA
pygments                    2.13.0
pynndescent                 0.5.7
pyparsing                   3.0.9
pyro                        1.8.2+2e3bd02
pytorch_lightning           1.7.7
pytz                        2022.5
pytz_deprecation_shim       NA
rich                        NA
seaborn                     0.12.0
setuptools                  65.5.0
six                         1.16.0
sklearn                     1.1.3
sphinxcontrib               NA
stack_data                  0.5.1
statsmodels                 0.13.2
tensorboard                 2.10.1
texttable                   1.6.4
threadpoolctl               3.1.0
toolz                       0.12.0
torch                       1.12.1
torchmetrics                0.10.1
tornado                     6.2
tqdm                        4.64.1
traitlets                   5.5.0
tree                        0.1.7
typing_extensions           NA
tzlocal                     NA
umap                        0.5.3
wcwidth                     0.2.5
wrapt                       1.14.1
yaml                        6.0
zipp                        NA
zmq                         24.0.1
zoneinfo                    NA
</pre>
</details> <!-- seems like this ends pre, so might as well be explicit -->
<pre>
-----
IPython             8.5.0
jupyter_client      7.4.4
jupyter_core        4.11.1
jupyterlab          3.5.0
notebook            6.5.1
-----
Python 3.9.13 | packaged by conda-forge | (main, May 27 2022, 17:01:00) [Clang 13.0.1 ]
macOS-10.15.7-x86_64-i386-64bit
-----
Session information updated at 2022-11-10 12:53
</pre>
</details></div></div>
</div>
</section>
<section id="r">
<h3><span class="section-number">12.11.2. </span>R<a class="headerlink" href="#r" title="Link to this heading">#</a></h3>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="o">%%</span><span class="k">R</span>
sessioninfo::session_info()
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>─ Session info ───────────────────────────────────────────────────────────────
 setting  value
 version  R version 4.1.3 (2022-03-10)
 os       macOS Catalina 10.15.7
 system   x86_64, darwin13.4.0
 ui       unknown
 language (EN)
 collate  C
 ctype    UTF-8
 tz       Europe/Berlin
 date     2022-11-10
 pandoc   2.19.2 @ /Users/luke.zappia/miniconda/envs/bp-integration/bin/pandoc

─ Packages ───────────────────────────────────────────────────────────────────
 package              * version  date (UTC) lib source
 abind                  1.4-5    2016-07-21 [1] CRAN (R 4.1.3)
 Biobase              * 2.54.0   2021-10-26 [1] Bioconductor
 BiocGenerics         * 0.40.0   2021-10-26 [1] Bioconductor
 bitops                 1.0-7    2021-04-24 [1] CRAN (R 4.1.3)
 cli                    3.4.1    2022-09-23 [1] CRAN (R 4.1.3)
 cluster                2.1.4    2022-08-22 [1] CRAN (R 4.1.3)
 codetools              0.2-18   2020-11-04 [1] CRAN (R 4.1.3)
 colorspace             2.0-3    2022-02-21 [1] CRAN (R 4.1.3)
 cowplot                1.1.1    2020-12-30 [1] CRAN (R 4.1.3)
 data.table             1.14.4   2022-10-17 [1] CRAN (R 4.1.3)
 DelayedArray           0.20.0   2021-10-26 [1] Bioconductor
 deldir                 1.0-6    2021-10-23 [1] CRAN (R 4.1.3)
 digest                 0.6.30   2022-10-18 [1] CRAN (R 4.1.3)
 dplyr                  1.0.10   2022-09-01 [1] CRAN (R 4.1.3)
 ellipsis               0.3.2    2021-04-29 [1] CRAN (R 4.1.3)
 fansi                  1.0.3    2022-03-24 [1] CRAN (R 4.1.3)
 fastmap                1.1.0    2021-01-25 [1] CRAN (R 4.1.3)
 fitdistrplus           1.1-8    2022-03-10 [1] CRAN (R 4.1.3)
 future                 1.28.0   2022-09-02 [1] CRAN (R 4.1.3)
 future.apply           1.9.1    2022-09-07 [1] CRAN (R 4.1.3)
 generics               0.1.3    2022-07-05 [1] CRAN (R 4.1.3)
 GenomeInfoDb         * 1.30.1   2022-01-30 [1] Bioconductor
 GenomeInfoDbData       1.2.7    2022-10-28 [1] Bioconductor
 GenomicRanges        * 1.46.1   2021-11-18 [1] Bioconductor
 ggplot2                3.3.6    2022-05-03 [1] CRAN (R 4.1.3)
 ggrepel                0.9.1    2021-01-15 [1] CRAN (R 4.1.3)
 ggridges               0.5.4    2022-09-26 [1] CRAN (R 4.1.3)
 globals                0.16.1   2022-08-28 [1] CRAN (R 4.1.3)
 glue                   1.6.2    2022-02-24 [1] CRAN (R 4.1.3)
 goftest                1.2-3    2021-10-07 [1] CRAN (R 4.1.3)
 gridExtra              2.3      2017-09-09 [1] CRAN (R 4.1.3)
 gtable                 0.3.1    2022-09-01 [1] CRAN (R 4.1.3)
 htmltools              0.5.3    2022-07-18 [1] CRAN (R 4.1.3)
 htmlwidgets            1.5.4    2021-09-08 [1] CRAN (R 4.1.3)
 httpuv                 1.6.6    2022-09-08 [1] CRAN (R 4.1.3)
 httr                   1.4.4    2022-08-17 [1] CRAN (R 4.1.3)
 ica                    1.0-3    2022-07-08 [1] CRAN (R 4.1.3)
 igraph                 1.3.5    2022-09-22 [1] CRAN (R 4.1.3)
 IRanges              * 2.28.0   2021-10-26 [1] Bioconductor
 irlba                  2.3.5.1  2022-10-03 [1] CRAN (R 4.1.3)
 jsonlite               1.8.3    2022-10-21 [1] CRAN (R 4.1.3)
 KernSmooth             2.23-20  2021-05-03 [1] CRAN (R 4.1.3)
 later                  1.2.0    2021-04-23 [1] CRAN (R 4.1.3)
 lattice                0.20-45  2021-09-22 [1] CRAN (R 4.1.3)
 lazyeval               0.2.2    2019-03-15 [1] CRAN (R 4.1.3)
 leiden                 0.4.3    2022-09-10 [1] CRAN (R 4.1.3)
 lifecycle              1.0.3    2022-10-07 [1] CRAN (R 4.1.3)
 listenv                0.8.0    2019-12-05 [1] CRAN (R 4.1.3)
 lmtest                 0.9-40   2022-03-21 [1] CRAN (R 4.1.3)
 magrittr               2.0.3    2022-03-30 [1] CRAN (R 4.1.3)
 MASS                   7.3-58.1 2022-08-03 [1] CRAN (R 4.1.3)
 Matrix               * 1.5-1    2022-09-13 [1] CRAN (R 4.1.3)
 MatrixGenerics       * 1.6.0    2021-10-26 [1] Bioconductor
 matrixStats          * 0.62.0   2022-04-19 [1] CRAN (R 4.1.3)
 mgcv                   1.8-41   2022-10-21 [1] CRAN (R 4.1.3)
 mime                   0.12     2021-09-28 [1] CRAN (R 4.1.3)
 miniUI                 0.1.1.1  2018-05-18 [1] CRAN (R 4.1.3)
 munsell                0.5.0    2018-06-12 [1] CRAN (R 4.1.3)
 nlme                   3.1-160  2022-10-10 [1] CRAN (R 4.1.3)
 parallelly             1.32.1   2022-07-21 [1] CRAN (R 4.1.3)
 patchwork              1.1.2    2022-08-19 [1] CRAN (R 4.1.3)
 pbapply                1.5-0    2021-09-16 [1] CRAN (R 4.1.3)
 pillar                 1.8.1    2022-08-19 [1] CRAN (R 4.1.3)
 pkgconfig              2.0.3    2019-09-22 [1] CRAN (R 4.1.3)
 plotly                 4.10.0   2021-10-09 [1] CRAN (R 4.1.3)
 plyr                   1.8.7    2022-03-24 [1] CRAN (R 4.1.3)
 png                    0.1-7    2013-12-03 [1] CRAN (R 4.1.3)
 polyclip               1.10-4   2022-10-20 [1] CRAN (R 4.1.3)
 progressr              0.11.0   2022-09-02 [1] CRAN (R 4.1.3)
 promises               1.2.0.1  2021-02-11 [1] CRAN (R 4.1.3)
 purrr                  0.3.5    2022-10-06 [1] CRAN (R 4.1.3)
 R6                     2.5.1    2021-08-19 [1] CRAN (R 4.1.3)
 RANN                   2.6.1    2019-01-08 [1] CRAN (R 4.1.3)
 RColorBrewer           1.1-3    2022-04-03 [1] CRAN (R 4.1.3)
 Rcpp                   1.0.9    2022-07-08 [1] CRAN (R 4.1.3)
 RcppAnnoy              0.0.19   2021-07-30 [1] CRAN (R 4.1.3)
 RCurl                  1.98-1.9 2022-10-03 [1] CRAN (R 4.1.3)
 reshape2               1.4.4    2020-04-09 [1] CRAN (R 4.1.3)
 reticulate             1.26     2022-08-31 [1] CRAN (R 4.1.3)
 rgeos                  0.5-9    2021-12-15 [1] CRAN (R 4.1.3)
 rlang                  1.0.6    2022-09-24 [1] CRAN (R 4.1.3)
 ROCR                   1.0-11   2020-05-02 [1] CRAN (R 4.1.3)
 rpart                  4.1.19   2022-10-21 [1] CRAN (R 4.1.3)
 Rtsne                  0.16     2022-04-17 [1] CRAN (R 4.1.3)
 S4Vectors            * 0.32.4   2022-03-24 [1] Bioconductor
 scales                 1.2.1    2022-08-20 [1] CRAN (R 4.1.3)
 scattermore            0.8      2022-02-14 [1] CRAN (R 4.1.3)
 sctransform            0.3.5    2022-09-21 [1] CRAN (R 4.1.3)
 sessioninfo            1.2.2    2021-12-06 [1] CRAN (R 4.1.3)
 Seurat               * 4.2.0    2022-09-21 [1] CRAN (R 4.1.3)
 SeuratObject         * 4.1.2    2022-09-20 [1] CRAN (R 4.1.3)
 shiny                  1.7.3    2022-10-25 [1] CRAN (R 4.1.3)
 SingleCellExperiment * 1.16.0   2021-10-26 [1] Bioconductor
 sp                   * 1.5-0    2022-06-05 [1] CRAN (R 4.1.3)
 spatstat.core          2.4-4    2022-05-18 [1] CRAN (R 4.1.3)
 spatstat.data          3.0-0    2022-10-21 [1] CRAN (R 4.1.3)
 spatstat.geom          3.0-3    2022-10-25 [1] CRAN (R 4.1.3)
 spatstat.random        2.2-0    2022-03-30 [1] CRAN (R 4.1.3)
 spatstat.sparse        3.0-0    2022-10-21 [1] CRAN (R 4.1.3)
 spatstat.utils         3.0-1    2022-10-19 [1] CRAN (R 4.1.3)
 stringi                1.7.8    2022-07-11 [1] CRAN (R 4.1.3)
 stringr                1.4.1    2022-08-20 [1] CRAN (R 4.1.3)
 SummarizedExperiment * 1.24.0   2021-10-26 [1] Bioconductor
 survival               3.4-0    2022-08-09 [1] CRAN (R 4.1.3)
 tensor                 1.5      2012-05-05 [1] CRAN (R 4.1.3)
 tibble                 3.1.8    2022-07-22 [1] CRAN (R 4.1.3)
 tidyr                  1.2.1    2022-09-08 [1] CRAN (R 4.1.3)
 tidyselect             1.2.0    2022-10-10 [1] CRAN (R 4.1.3)
 utf8                   1.2.2    2021-07-24 [1] CRAN (R 4.1.3)
 uwot                   0.1.14   2022-08-22 [1] CRAN (R 4.1.3)
 vctrs                  0.5.0    2022-10-22 [1] CRAN (R 4.1.3)
 viridisLite            0.4.1    2022-08-22 [1] CRAN (R 4.1.3)
 xtable                 1.8-4    2019-04-21 [1] CRAN (R 4.1.3)
 XVector                0.34.0   2021-10-26 [1] Bioconductor
 zlibbioc               1.40.0   2021-10-26 [1] Bioconductor
 zoo                    1.8-11   2022-09-17 [1] CRAN (R 4.1.3)

 [1] /Users/luke.zappia/miniconda/envs/bp-integration/lib/R/library

──────────────────────────────────────────────────────────────────────────────
</pre></div>
</div>
</div>
</div>
</section>
</section>
<section id="references">
<h2><span class="section-number">12.12. </span>References<a class="headerlink" href="#references" title="Link to this heading">#</a></h2>
<div class="docutils container" id="id42">
<div role="list" class="citation-list">
<div class="citation" id="id219" role="doc-biblioentry">
<span class="label"><span class="fn-bracket">[</span><a role="doc-backlink" href="#id16">intACSM21</a><span class="fn-bracket">]</span></span>
<p>Ricard Argelaguet, Anna S E Cuomo, Oliver Stegle, and John C Marioni. Computational principles and challenges in single-cell data integration. <em>Nature biotechnology</em>, pages 1–14, May 2021. <a class="reference external" href="https://doi.org/10.1038/s41587-021-00895-7">doi:10.1038/s41587-021-00895-7</a>.</p>
</div>
<div class="citation" id="id211" role="doc-biblioentry">
<span class="label"><span class="fn-bracket">[</span>intBHS+18<span class="fn-bracket">]</span></span>
<span class="backrefs">(<a role="doc-backlink" href="#id8">1</a>,<a role="doc-backlink" href="#id22">2</a>,<a role="doc-backlink" href="#id38">3</a>)</span>
<p>Andrew Butler, Paul Hoffman, Peter Smibert, Efthymia Papalexi, and Rahul Satija. Integrating single-cell transcriptomic data across different conditions, technologies, and species. <em>Nature biotechnology</em>, April 2018. <a class="reference external" href="https://doi.org/10.1038/nbt.4096">doi:10.1038/nbt.4096</a>.</p>
</div>
<div class="citation" id="id212" role="doc-biblioentry">
<span class="label"><span class="fn-bracket">[</span><a role="doc-backlink" href="#id19">intButtnerMW+19</a><span class="fn-bracket">]</span></span>
<p>Maren Büttner, Zhichao Miao, F Alexander Wolf, Sarah A Teichmann, and Fabian J Theis. A test metric for assessing single-cell RNA-seq batch correction. <em>Nature methods</em>, 16(1):43–49, January 2019. <a class="reference external" href="https://doi.org/10.1038/s41592-018-0254-1">doi:10.1038/s41592-018-0254-1</a>.</p>
</div>
<div class="citation" id="id232" role="doc-biblioentry">
<span class="label"><span class="fn-bracket">[</span><a role="doc-backlink" href="#id21">intCGvDKH21</a><span class="fn-bracket">]</span></span>
<p>Ruben Chazarra-Gil, Stijn van Dongen, Vladimir Yu Kiselev, and Martin Hemberg. Flexible comparison of batch correction methods for single-cell RNA-seq using BatchBench. <em>Nucleic acids research</em>, 49(7):e42, April 2021. <a class="reference external" href="https://doi.org/10.1093/nar/gkab004">doi:10.1093/nar/gkab004</a>.</p>
</div>
<div class="citation" id="id222" role="doc-biblioentry">
<span class="label"><span class="fn-bracket">[</span><a role="doc-backlink" href="#id33">intGLX+22</a><span class="fn-bracket">]</span></span>
<p>Adam Gayoso, Romain Lopez, Galen Xing, Pierre Boyeau, Valeh Valiollah Pour Amiri, Justin Hong, Katherine Wu, Michael Jayasuriya, Edouard Mehlman, Maxime Langevin, Yining Liu, Jules Samaran, Gabriel Misrachi, Achille Nazaret, Oscar Clivio, Chenling Xu, Tal Ashuach, Mariano Gabitto, Mohammad Lotfollahi, Valentine Svensson, Eduardo da Veiga Beltrame, Vitalii Kleshchevnikov, Carlos Talavera-López, Lior Pachter, Fabian J Theis, Aaron Streets, Michael I Jordan, Jeffrey Regier, and Nir Yosef. A Python library for probabilistic analysis of single-cell omics data. <em>Nature biotechnology</em>, February 2022. <a class="reference external" href="https://doi.org/10.1038/s41587-021-01206-w">doi:10.1038/s41587-021-01206-w</a>.</p>
</div>
<div class="citation" id="id215" role="doc-biblioentry">
<span class="label"><span class="fn-bracket">[</span>intHLMM18<span class="fn-bracket">]</span></span>
<span class="backrefs">(<a role="doc-backlink" href="#id7">1</a>,<a role="doc-backlink" href="#id10">2</a>,<a role="doc-backlink" href="#id39">3</a>)</span>
<p>Laleh Haghverdi, Aaron T L Lun, Michael D Morgan, and John C Marioni. Batch effects in single-cell RNA-sequencing data are corrected by matching mutual nearest neighbors. <em>Nature biotechnology</em>, April 2018. <a class="reference external" href="https://doi.org/10.1038/nbt.4091">doi:10.1038/nbt.4091</a>.</p>
</div>
<div class="citation" id="id214" role="doc-biblioentry">
<span class="label"><span class="fn-bracket">[</span><a role="doc-backlink" href="#id9">intHBB19</a><span class="fn-bracket">]</span></span>
<p>Brian Hie, Bryan Bryson, and Bonnie Berger. Efficient integration of heterogeneous single-cell transcriptomes using Scanorama. <em>Nature biotechnology</em>, May 2019. <a class="reference external" href="https://doi.org/10.1038/s41587-019-0113-3">doi:10.1038/s41587-019-0113-3</a>.</p>
</div>
<div class="citation" id="id231" role="doc-biblioentry">
<span class="label"><span class="fn-bracket">[</span><a role="doc-backlink" href="#id6">intJLR07</a><span class="fn-bracket">]</span></span>
<p>W Evan Johnson, Cheng Li, and Ariel Rabinovic. Adjusting batch effects in microarray expression data using empirical Bayes methods. <em>Biostatistics</em>, 8(1):118–127, January 2007. <a class="reference external" href="https://doi.org/10.1093/biostatistics/kxj037">doi:10.1093/biostatistics/kxj037</a>.</p>
</div>
<div class="citation" id="id210" role="doc-biblioentry">
<span class="label"><span class="fn-bracket">[</span>intKMF+19<span class="fn-bracket">]</span></span>
<span class="backrefs">(<a role="doc-backlink" href="#id11">1</a>,<a role="doc-backlink" href="#id23">2</a>)</span>
<p>Ilya Korsunsky, Nghia Millard, Jean Fan, Kamil Slowikowski, Fan Zhang, Kevin Wei, Yuriy Baglaenko, Michael Brenner, Po-Ru Loh, and Soumya Raychaudhuri. Fast, sensitive and accurate integration of single-cell data with Harmony. <em>Nature methods</em>, November 2019. <a class="reference external" href="https://doi.org/10.1038/s41592-019-0619-0">doi:10.1038/s41592-019-0619-0</a>.</p>
</div>
<div class="citation" id="id225" role="doc-biblioentry">
<span class="label"><span class="fn-bracket">[</span><a role="doc-backlink" href="#id27">intLLB+22</a><span class="fn-bracket">]</span></span>
<p>Christopher Lance, Malte D Luecken, Daniel B Burkhardt, Robrecht Cannoodt, Pia Rautenstrauch, Anna Laddach, Aidyn Ubingazhibov, Zhi-Jie Cao, Kaiwen Deng, Sumeer Khan, Qiao Liu, Nikolay Russkikh, Gleb Ryazantsev, Uwe Ohler, Neurips 2021 Multimodal Data integration competition participants, Angela Oliveira Pisco, Jonathan Bloom, Smita Krishnaswamy, and Fabian J Theis. Multimodal single cell data integration challenge: Results and lessons learned. In Douwe Kiela, Marco Ciccone, and Barbara Caputo, editors, <em>Proceedings of the NeurIPS 2021 Competitions and Demonstrations Track</em>, volume 176 of Proceedings of Machine Learning Research, 162–176. PMLR, 2022.</p>
</div>
<div class="citation" id="id226" role="doc-biblioentry">
<span class="label"><span class="fn-bracket">[</span>intLRC+18<span class="fn-bracket">]</span></span>
<span class="backrefs">(<a role="doc-backlink" href="#id13">1</a>,<a role="doc-backlink" href="#id32">2</a>)</span>
<p>Romain Lopez, Jeffrey Regier, Michael B Cole, Michael I Jordan, and Nir Yosef. Deep generative modeling for single-cell transcriptomics. <em>Nature methods</em>, 15(12):1053–1058, November 2018. <a class="reference external" href="https://doi.org/10.1038/s41592-018-0229-2">doi:10.1038/s41592-018-0229-2</a>.</p>
</div>
<div class="citation" id="id221" role="doc-biblioentry">
<span class="label"><span class="fn-bracket">[</span><a role="doc-backlink" href="#id15">intLWT19</a><span class="fn-bracket">]</span></span>
<p>Mohammad Lotfollahi, F Alexander Wolf, and Fabian J Theis. scGen predicts single-cell perturbation responses. <em>Nature methods</em>, 16(8):715–721, August 2019. <a class="reference external" href="https://doi.org/10.1038/s41592-019-0494-8">doi:10.1038/s41592-019-0494-8</a>.</p>
</div>
<div class="citation" id="id224" role="doc-biblioentry">
<span class="label"><span class="fn-bracket">[</span><a role="doc-backlink" href="#id27">intLBC+22</a><span class="fn-bracket">]</span></span>
<p>Malte D Luecken, Daniel Bernard Burkhardt, Robrecht Cannoodt, Christopher Lance, Aditi Agrawal, Hananeh Aliee, Ann T Chen, Louise Deconinck, Angela M Detweiler, Alejandro A Granados, Shelly Huynh, Laura Isacco, Yang Joon Kim, Dominik Klein, Bony de Kumar, Sunil Kuppasani, Heiko Lickert, Aaron McGeever, Honey Mekonen, Joaquin Caceres Melgarejo, Maurizio Morri, Michaela Müller, Norma Neff, Sheryl Paul, Bastian Rieck, Kaylie Schneider, Scott Steelman, Michael Sterr, Daniel J Treacy, Alexander Tong, Alexandra-Chloe Villani, Guilin Wang, Jia Yan, Ce Zhang, Angela Oliveira Pisco, Smita Krishnaswamy, Fabian J Theis, and Jonathan M Bloom. A sandbox for prediction and integration of DNA, RNA, and proteins in single cells. In <em>Thirty-fifth Conference on Neural Information Processing Systems Datasets and Benchmarks Track (Round 2)</em>. January 2022.</p>
</div>
<div class="citation" id="id220" role="doc-biblioentry">
<span class="label"><span class="fn-bracket">[</span>intLButtnerC+21<span class="fn-bracket">]</span></span>
<span class="backrefs">(<a role="doc-backlink" href="#id2">1</a>,<a role="doc-backlink" href="#id5">2</a>,<a role="doc-backlink" href="#id17">3</a>,<a role="doc-backlink" href="#id24">4</a>,<a role="doc-backlink" href="#id26">5</a>,<a role="doc-backlink" href="#id31">6</a>,<a role="doc-backlink" href="#id34">7</a>,<a role="doc-backlink" href="#id36">8</a>,<a role="doc-backlink" href="#id41">9</a>)</span>
<p>Malte D Luecken, M Büttner, K Chaichoompu, A Danese, M Interlandi, M F Mueller, D C Strobl, L Zappia, M Dugas, M Colomé-Tatché, and Fabian J Theis. Benchmarking atlas-level data integration in single-cell genomics. <em>Nature methods</em>, December 2021. <a class="reference external" href="https://doi.org/10.1038/s41592-021-01336-8">doi:10.1038/s41592-021-01336-8</a>.</p>
</div>
<div class="citation" id="id218" role="doc-biblioentry">
<span class="label"><span class="fn-bracket">[</span>intLT19<span class="fn-bracket">]</span></span>
<span class="backrefs">(<a role="doc-backlink" href="#id18">1</a>,<a role="doc-backlink" href="#id30">2</a>)</span>
<p>Malte D Luecken and Fabian J Theis. Current best practices in single‐cell RNA‐seq analysis: a tutorial. <em>Molecular systems biology</em>, June 2019. <a class="reference external" href="https://doi.org/10.15252/msb.20188746">doi:10.15252/msb.20188746</a>.</p>
</div>
<div class="citation" id="id230" role="doc-biblioentry">
<span class="label"><span class="fn-bracket">[</span><a role="doc-backlink" href="#id3">intMLE+21</a><span class="fn-bracket">]</span></span>
<p>Christoph Muus, Malte D Luecken, Gökcen Eraslan, Lisa Sikkema, Avinash Waghray, Graham Heimberg, Yoshihiko Kobayashi, Eeshit Dhaval Vaishnav, Ayshwarya Subramanian, Christopher Smillie, Karthik A Jagadeesh, Elizabeth Thu Duong, Evgenij Fiskin, Elena Torlai Triglia, Meshal Ansari, Peiwen Cai, Brian Lin, Justin Buchanan, Sijia Chen, Jian Shu, Adam L Haber, Hattie Chung, Daniel T Montoro, Taylor Adams, Hananeh Aliee, Samuel J Allon, Zaneta Andrusivova, Ilias Angelidis, Orr Ashenberg, Kevin Bassler, Christophe Bécavin, Inbal Benhar, Joseph Bergenstråhle, Ludvig Bergenstråhle, Liam Bolt, Emelie Braun, Linh T Bui, Steven Callori, Mark Chaffin, Evgeny Chichelnitskiy, Joshua Chiou, Thomas M Conlon, Michael S Cuoco, Anna S E Cuomo, Marie Deprez, Grant Duclos, Denise Fine, David S Fischer, Shila Ghazanfar, Astrid Gillich, Bruno Giotti, Joshua Gould, Minzhe Guo, Austin J Gutierrez, Arun C Habermann, Tyler Harvey, Peng He, Xiaomeng Hou, Lijuan Hu, Yan Hu, Alok Jaiswal, Lu Ji, Peiyong Jiang, Theodoros S Kapellos, Christin S Kuo, Ludvig Larsson, Michael A Leney-Greene, Kyungtae Lim, Monika Litviňuková, Leif S Ludwig, Soeren Lukassen, Wendy Luo, Henrike Maatz, Elo Madissoon, Lira Mamanova, Kasidet Manakongtreecheep, Sylvie Leroy, Christoph H Mayr, Ian M Mbano, Alexi M McAdams, Ahmad N Nabhan, Sarah K Nyquist, Lolita Penland, Olivier B Poirion, Sergio Poli, Cancan Qi, Rachel Queen, Daniel Reichart, Ivan Rosas, Jonas C Schupp, Conor V Shea, Xingyi Shi, Rahul Sinha, Rene V Sit, Kamil Slowikowski, Michal Slyper, Neal P Smith, Alex Sountoulidis, Maximilian Strunz, Travis B Sullivan, Dawei Sun, Carlos Talavera-López, Peng Tan, Jessica Tantivit, Kyle J Travaglini, Nathan R Tucker, Katherine A Vernon, Marc H Wadsworth, Julia Waldman, Xiuting Wang, Ke Xu, Wenjun Yan, William Zhao, Carly G K Ziegler, NHLBI LungMap Consortium, and Human Cell Atlas Lung Biological Network. Single-cell meta-analysis of SARS-CoV-2 entry genes across tissues and demographics. <em>Nature medicine</em>, 27(3):546–559, March 2021. <a class="reference external" href="https://doi.org/10.1038/s41591-020-01227-z">doi:10.1038/s41591-020-01227-z</a>.</p>
</div>
<div class="citation" id="id216" role="doc-biblioentry">
<span class="label"><span class="fn-bracket">[</span><a role="doc-backlink" href="#id28">intPZO17</a><span class="fn-bracket">]</span></span>
<p>Belinda Phipson, Luke Zappia, and Alicia Oshlack. Gene length and detection bias in single cell RNA sequencing protocols. <em>F1000Research</em>, April 2017. <a class="reference external" href="https://doi.org/10.12688/f1000research.11290.1">doi:10.12688/f1000research.11290.1</a>.</p>
</div>
<div class="citation" id="id228" role="doc-biblioentry">
<span class="label"><span class="fn-bracket">[</span>intPolanskiPY+19<span class="fn-bracket">]</span></span>
<span class="backrefs">(<a role="doc-backlink" href="#id12">1</a>,<a role="doc-backlink" href="#id37">2</a>)</span>
<p>Krzysztof Polański, Jong-Eun Park, Matthew D Young, Zhichao Miao, Kerstin B Meyer, and Sarah A Teichmann. BBKNN: Fast Batch Alignment of Single Cell Transcriptomes. <em>Bioinformatics</em>, August 2019. <a class="reference external" href="https://doi.org/10.1093/bioinformatics/btz625">doi:10.1093/bioinformatics/btz625</a>.</p>
</div>
<div class="citation" id="id213" role="doc-biblioentry">
<span class="label"><span class="fn-bracket">[</span><a role="doc-backlink" href="#id38">intSFG+15</a><span class="fn-bracket">]</span></span>
<p>Rahul Satija, Jeffrey A Farrell, David Gennert, Alexander F Schier, and Aviv Regev. Spatial reconstruction of single-cell gene expression data. <em>Nature Biotechnology</em>, 33(5):495–502, April 2015. <a class="reference external" href="https://doi.org/10.1038/nbt.3192">doi:10.1038/nbt.3192</a>.</p>
</div>
<div class="citation" id="id223" role="doc-biblioentry">
<span class="label"><span class="fn-bracket">[</span><a role="doc-backlink" href="#id4">intSSZ+22</a><span class="fn-bracket">]</span></span>
<p>Lisa Sikkema, Daniel C Strobl, Luke Zappia, Elo Madissoon, Nikolay S Markov, Laure-Emmanuelle Zaragosi, Meshal Ansari, Marie-Jeanne Arguel, Leonie Apperloo, Christophe Becavin, Marijn Berg, Evgeny Chichelnitskiy, Mei-I Chung, Antoine Collin, Aurore C A Gay, Baharak Hooshiar Kashani, Manu Jain, Theodore Kapellos, Tessa M Kole, Christoph H Mayr, Von Michael Papen, Lance Peter, Ciro Ramirez-Suastegui, Janine Schniering, Chase J Taylor, Thomas Walzthoeni, Chuan Xu, Linh T Bui, Carlo de Donno, Leander Dony, Minzhe Guo, Austin J Gutierrez, Lukas Heumos, Ni Huang, Ignacio L Ibarra, Nathan D Jackson, Preetish Kadur Lakshminarasimha Murthy, Mohammad Lotfollahi, Tracy Tabib, Carlos Talavera-Lopez, Kyle J Travaglini, Anna Wilbrey-Clark, Kaylee B Worlock, Masahiro Yoshida, Tushar J Desai, Oliver Eickelberg, Christine Falk, Naftali Kaminski, Mark A Krasnow, Robert Lafyatis, Marko Z Nikoli, Joseph E Powell, Jayaraj Rajagopal, Orit Rozenblatt-Rosen, Max A Seibold, Dean Sheppard, Douglas P Shepherd, Sarah A Teichmann, Alexander M Tsankov, Jeffrey Whitsett, Yan Xu, Nicholas E Banovich, Pascal Barbry, Thu E Duong, Kerstin B Meyer, Jonathan A Kropski, Dana Pe'er, Herbert B Schiller, Purushothama Rao Tata, Joachim L Schultze, Alexander V Misharin, Martijn C Nawijn, Malte D Luecken, and Fabian J Theis. An integrated cell atlas of the human lung in health and disease. <em>bioRxiv</em>, pages 2022.03.10.483747, March 2022. <a class="reference external" href="https://doi.org/10.1101/2022.03.10.483747">doi:10.1101/2022.03.10.483747</a>.</p>
</div>
<div class="citation" id="id208" role="doc-biblioentry">
<span class="label"><span class="fn-bracket">[</span>intSBH+19<span class="fn-bracket">]</span></span>
<span class="backrefs">(<a role="doc-backlink" href="#id8">1</a>,<a role="doc-backlink" href="#id22">2</a>,<a role="doc-backlink" href="#id38">3</a>)</span>
<p>Tim Stuart, Andrew Butler, Paul Hoffman, Christoph Hafemeister, Efthymia Papalexi, William M Mauck, 3rd, Yuhan Hao, Marlon Stoeckius, Peter Smibert, and Rahul Satija. Comprehensive Integration of Single-Cell Data. <em>Cell</em>, 177(7):1888–1902.e21, June 2019. <a class="reference external" href="https://doi.org/10.1016/j.cell.2019.05.031">doi:10.1016/j.cell.2019.05.031</a>.</p>
</div>
<div class="citation" id="id233" role="doc-biblioentry">
<span class="label"><span class="fn-bracket">[</span><a role="doc-backlink" href="#id25">intSFHM21</a><span class="fn-bracket">]</span></span>
<p>Vinay S Swamy, Temesgen D Fufa, Robert B Hufnagel, and David M McGaughey. Building the mega single-cell transcriptome ocular meta-atlas. <em>GigaScience</em>, October 2021. <a class="reference external" href="https://doi.org/10.1093/gigascience/giab061">doi:10.1093/gigascience/giab061</a>.</p>
</div>
<div class="citation" id="id209" role="doc-biblioentry">
<span class="label"><span class="fn-bracket">[</span>intTAC+20<span class="fn-bracket">]</span></span>
<span class="backrefs">(<a role="doc-backlink" href="#id20">1</a>,<a role="doc-backlink" href="#id40">2</a>)</span>
<p>Hoa Thi Nhu Tran, Kok Siong Ang, Marion Chevrier, Xiaomeng Zhang, Nicole Yee Shin Lee, Michelle Goh, and Jinmiao Chen. A benchmark of batch-effect correction methods for single-cell RNA sequencing data. <em>Genome biology</em>, 21(1):12, January 2020. <a class="reference external" href="https://doi.org/10.1186/s13059-019-1850-9">doi:10.1186/s13059-019-1850-9</a>.</p>
</div>
<div class="citation" id="id229" role="doc-biblioentry">
<span class="label"><span class="fn-bracket">[</span><a role="doc-backlink" href="#id1">intvdBSVertesy+17</a><span class="fn-bracket">]</span></span>
<p>Susanne C van den Brink, Fanny Sage, Ábel Vértesy, Bastiaan Spanjaard, Josi Peterson-Maduro, Chloé S Baron, Catherine Robin, and Alexander van Oudenaarden. Single-cell sequencing reveals dissociation-induced gene expression in tissue subpopulations. <em>Nature methods</em>, 14(10):935–936, September 2017. <a class="reference external" href="https://doi.org/10.1038/nmeth.4437">doi:10.1038/nmeth.4437</a>.</p>
</div>
<div class="citation" id="id217" role="doc-biblioentry">
<span class="label"><span class="fn-bracket">[</span><a role="doc-backlink" href="#id29">intWKL12</a><span class="fn-bracket">]</span></span>
<p>Günter P Wagner, Koryu Kin, and Vincent J Lynch. Measurement of mRNA abundance using RNA-seq data: RPKM measure is inconsistent among samples. <em>Theory in Biosciences</em>, 131(4):281–285, December 2012. <a class="reference external" href="https://doi.org/10.1007/s12064-012-0162-3">doi:10.1007/s12064-012-0162-3</a>.</p>
</div>
<div class="citation" id="id227" role="doc-biblioentry">
<span class="label"><span class="fn-bracket">[</span>intXLM+21<span class="fn-bracket">]</span></span>
<span class="backrefs">(<a role="doc-backlink" href="#id14">1</a>,<a role="doc-backlink" href="#id35">2</a>)</span>
<p>Chenling Xu, Romain Lopez, Edouard Mehlman, Jeffrey Regier, Michael I Jordan, and Nir Yosef. Probabilistic harmonization and annotation of single-cell transcriptomics data with deep generative models. <em>Molecular systems biology</em>, 17(1):e9620, January 2021. <a class="reference external" href="https://doi.org/10.15252/msb.20209620">doi:10.15252/msb.20209620</a>.</p>
</div>
</div>
</div>
</section>
<section id="contributors">
<h2><span class="section-number">12.13. </span>Contributors<a class="headerlink" href="#contributors" title="Link to this heading">#</a></h2>
<p>We gratefully acknowledge the contributions of:</p>
<section id="authors">
<h3><span class="section-number">12.13.1. </span>Authors<a class="headerlink" href="#authors" title="Link to this heading">#</a></h3>
<ul class="simple">
<li><p>Luke Zappia</p></li>
<li><p>Malte Lücken</p></li>
</ul>
</section>
<section id="reviewers">
<h3><span class="section-number">12.13.2. </span>Reviewers<a class="headerlink" href="#reviewers" title="Link to this heading">#</a></h3>
<ul class="simple">
<li><p>Lukas Heumos</p></li>
</ul>
</section>
</section>
</section>

    <script type="text/x-thebe-config">
    {
        requestKernel: true,
        binderOptions: {
            repo: "binder-examples/jupyter-stacks-datascience",
            ref: "master",
        },
        codeMirrorConfig: {
            theme: "abcdef",
            mode: "python"
        },
        kernelOptions: {
            name: "python3",
            path: "./cellular_structure"
        },
        predefinedOutput: true
    }
    </script>
    <script>kernelName = 'python3'</script>

                </article>
              

              
              
              
              
                <footer class="prev-next-footer d-print-none">
                  
<div class="prev-next-area">
    <a class="left-prev"
       href="annotation.html"
       title="previous page">
      <i class="fa-solid fa-angle-left"></i>
      <div class="prev-next-info">
        <p class="prev-next-subtitle">previous</p>
        <p class="prev-next-title"><span class="section-number">11. </span>Annotation</p>
      </div>
    </a>
    <a class="right-next"
       href="../trajectories/pseudotemporal.html"
       title="next page">
      <div class="prev-next-info">
        <p class="prev-next-subtitle">next</p>
        <p class="prev-next-title"><span class="section-number">13. </span>Pseudotemporal ordering</p>
      </div>
      <i class="fa-solid fa-angle-right"></i>
    </a>
</div>
                </footer>
              
            </div>
            
            
              
                <div class="bd-sidebar-secondary bd-toc"><div class="sidebar-secondary-items sidebar-secondary__inner">


  <div class="sidebar-secondary-item">
  <div class="page-toc tocsection onthispage">
    <i class="fa-solid fa-list"></i> Contents
  </div>
  <nav class="bd-toc-nav page-toc">
    <ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#motivation">12.1. Motivation</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#types-of-integration-models">12.1.1. Types of integration models</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#batch-removal-complexity">12.1.2. Batch removal complexity</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#comparisons-of-data-integration-methods">12.1.3. Comparisons of data integration methods</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#choosing-an-integration-method">12.1.4. Choosing an integration method</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#dataset">12.2. Dataset</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#unintegrated-data">12.3. Unintegrated data</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#batch-aware-feature-selection">12.4. Batch-aware feature selection</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#variational-autoencoder-vae-based-integration">12.5. Variational autoencoder (VAE) based integration</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#data-preparation">12.5.1. Data preparation</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#building-the-model">12.5.2. Building the model</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#training-the-model">12.5.3. Training the model</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#extracting-the-embedding">12.5.4. Extracting the embedding</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#calculate-a-batch-corrected-umap">12.5.5. Calculate a batch-corrected UMAP</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#vae-integration-using-cell-labels">12.6. VAE integration using cell labels</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#graph-based-integration">12.7. Graph-based integration</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#linear-embedding-integration-using-mutual-nearest-neighbors-mnn">12.8. Linear embedding integration using Mutual Nearest Neighbors (MNN)</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#benchmarking-your-own-integration">12.9. Benchmarking your own integration</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#quiz">12.10. Quiz</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#session-information">12.11. Session information</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#python">12.11.1. Python</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#r">12.11.2. R</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#references">12.12. References</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#contributors">12.13. Contributors</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#authors">12.13.1. Authors</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#reviewers">12.13.2. Reviewers</a></li>
</ul>
</li>
</ul>
  </nav></div>

</div></div>
              
            
          </div>
          <footer class="bd-footer-content">
            
<div class="bd-footer-content__inner container">
  
  <div class="footer-item">
    
<p class="component-author">
By Lukas Heumos, Anna Schaar, single-cell best practices consortium
</p>

  </div>
  
  <div class="footer-item">
    

  <p class="copyright">
    
      © Copyright 2023.
      <br/>
    
  </p>

  </div>
  
  <div class="footer-item">
    
  </div>
  
  <div class="footer-item">
    
<div class="extra_footer">
  <div>
  Brought to you by
  <a href="https://www.github.com/theislab">Theislab</a>, with many thanks to the single-cell community as a whole!
</div>

</div>
  </div>
  
</div>
          </footer>
        

      </main>
    </div>
  </div>
  
  <!-- Scripts loaded after <body> so the DOM is not blocked -->
  <script src="../_static/scripts/bootstrap.js?digest=dfe6caa3a7d634c4db9b"></script>
<script src="../_static/scripts/pydata-sphinx-theme.js?digest=dfe6caa3a7d634c4db9b"></script>

  <footer class="bd-footer">
  </footer>
  </body>
</html>