{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "eceea5f9-1817-43b8-9e2f-3eb2bf8a1688",
   "metadata": {},
   "source": [
    "# Preprocessing & visualization introduction\n",
    "\n",
    "This chapter will guide you through the steps of data preprocessing and visualization. Pre-processing aims to address the following analysis steps: quality control, normalization, feature selection, and dimensionality reduction. \n",
    "\n",
    "This chapter will additionally introduce the advanced topics of ambient RNA, single nuclei and interactive visualization. \n",
    "\n",
    "The starting point of this notebook are raw sequencing data generated by sequencing machines which are processed and aligned to obtain matrices of molecular counts, so called count matrices, or read counts (read matrices). The difference between cound and read matrices depends on whether unique molecular identifiers (UMIs) were included in the single-cell library construction protocol. \n",
    "\n",
    "Read and count matrices have the dimension number of barcodes x number of transcripts. It is important to note that the term \"barcode\" is used here instead of \"cell\" as a barcode might wrongly have tagged multiple cells (doublet) or might have not tagged any cell (empty droplet/well). We will elaborate more on this in the next section \"Doublet detraction & quality control\". \n",
    "\n",
    "### Imports\n",
    "Before we introduce the first pre-processing step of doublet detection and quality control, we import the respective packages from python and R. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "a12410e9-deff-4de1-b118-eac694753f9e",
   "metadata": {},
   "outputs": [],
   "source": [
    "import warnings\n",
    "warnings.filterwarnings('ignore')\n",
    "\n",
    "import scanpy as sc\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "\n",
    "import anndata2ri\n",
    "import logging\n",
    "\n",
    "import rpy2.rinterface_lib.callbacks as rcb\n",
    "import rpy2.robjects as ro\n",
    "\n",
    "sc.settings.verbosity = 0\n",
    "rcb.logger.setLevel(logging.ERROR)\n",
    "\n",
    "ro.pandas2ri.activate()\n",
    "anndata2ri.activate()\n",
    "\n",
    "%load_ext rpy2.ipython"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "53094abd-90dd-4dea-84e4-b99c6c179104",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%R\n",
    "library(Seurat)\n",
    "library(scDblFinder)\n",
    "library(scater)\n",
    "library(BiocParallel)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c9df7e52-432a-4021-b108-58ed7c453d96",
   "metadata": {},
   "source": [
    "### Dataset"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "2b79a2bf-2227-4575-8961-e6994edba05b",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Variable names are not unique. To make them unique, call `.var_names_make_unique`.\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "AnnData object with n_obs × n_vars = 131515 × 36601\n",
       "    obs: 'site', 'donor'\n",
       "    var: 'gene_ids', 'feature_types', 'genome'"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata = sc.read('cellranger_neurips21_bmmc.h5ad')\n",
    "adata.var_names_make_unique()\n",
    "adata"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bdecd730-93d3-40d2-a287-7556c5d1973d",
   "metadata": {},
   "source": [
    "The dataset has the shape `n_obs` 131,515 x `n_vars` 36,601, so barcodes x number of transcripts. We additionally inspect some further information in `.obs` and `.var` which we will neglect for now. "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "cc3210ab-1187-40ec-acfd-a602fe9dd377",
   "metadata": {},
   "source": [
    "# 2.1 Doublet Detection\n",
    "\n",
    "Doublets are defined as two cells that are sequenced under the same cellular barcode, so for example if they were captured in the same droplet. We therefore used so far the term \"barcode\" instead if \"cell\". A doublet is called homotypic if it is formed by the same cell type (but from different individuals) and heterotypic otherwise. Homotypic doublets are not necessarily identifiable from count matrices and are often considered innocuous as they can be identified with cell hashing or SNPs. Hence, their identification is not the main goal of the doublet detection methods. \n",
    "\n",
    "Doublets formed from different cell types or states are called heterotypic. Their identification is crucial as they are most likedly misclassified and can lead to distorted downstream analysis steps. Hence, doublet detection and removal is typically an initial preprocessing step. Doublets can be either identified through their high number of reads and detected features, or with methods that create atrifical doublets and comparing these with each cell. Doublet detection methods are computationally efficient and several software packages exist for this analysis them. To mention a few: *scDblFinder*,  *scds*, *DoubletFinder* or *scran’s doubletCells*. These methods are not yet benchmarked  in an independent comparison, but compared in a pipeline setting by {cite}`germain_pipecomp_2020` that collects several preprocessung steps. scDblFinder showed to enhance downstream clustering results. \n",
    "\n",
    "#### scDblFinder\n",
    "\n",
    "It proved to be benefical and computationally efficient with software packages such as *scDblFinder* or *scds*. We will demonstrate the usage of scDblFinder which is a Bioconductor package working on raw count matrices. \n",
    "\n",
    "The scDblFinder R package collects several methods for the detection of doublets in scRNA-seq data and identifies heterotypic doublets which are in many cases also the most critical ones as stated above. \n",
    "\n",
    "scDblFinder can be used for datasets with multiple samples. It will use this information to generate artificial doublets and the nearest neighbors are identifies for each sample separatly. The overall scoring is then performed globally and sample-specific doublet rates will be considered for the threshold.\n",
    "\n",
    "scDblFinde expects an `SingleCellExperiment` object of shape number of transcripts x barcodes which does not contain empty drops, but other than that has not been filtered beore. We will now apply the method to the transposed raw count matrix `.X`. To use scDblFinder on multiple samples we additionally provide a vector of the sample ids and allow multithreading using the BPPARAM parameter. Both parameters can be ignored if your dataset only consists of one sample."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "26e28d2a-fb3e-4d02-93b9-9e6259d1407d",
   "metadata": {},
   "outputs": [],
   "source": [
    "data_mat = adata.X.T\n",
    "sample_ids = pd.factorize(adata.obs['donor'])[0]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "974e5ea0-9e93-45b8-8577-71e34a5f98df",
   "metadata": {},
   "source": [
    "We can now launch the doublet detection by using `data_mat` as input to scDblFinder within a SingleCellExperiment.\n",
    "\n",
    "scBblFinder adds several columns to the colData of sce, three of them are in particular intersting for the user:\n",
    "\n",
    "* `sce$scDblFinder.score` : the final doublet score (the higher the more likely that the cell is a doublet)\n",
    "\n",
    "* `sce$scDblFinder.ratio` : the ratio of artificial doublets in the cell's neighborhood\n",
    "\n",
    "* `sce$scDblFinder.class` : the classification (doublet or singlet)\n",
    "\n",
    "We will only output the `class` argument. The other arguments can be added to the anndata object in a similar way. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "d3dbbf06-0d63-448b-82bc-c98c51d2cbdb",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%R -i data_mat -i sample_ids -o droplet_class\n",
    "\n",
    "sce = scDblFinder(\n",
    "    SingleCellExperiment(\n",
    "        list(counts=data_mat),\n",
    "    ), \n",
    "    samples=sample_ids, \n",
    "    BPPARAM=MulticoreParam(3), \n",
    "    verbose=TRUE\n",
    ")\n",
    "droplet_class = sce$scDblFinder.class"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9b9559a6-c118-491f-a527-0c2319b4192e",
   "metadata": {},
   "source": [
    "We now store the result in `.obs` and show the value counts for both classes singlet and doublet in the dataset."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "1174ae23-7f77-46aa-9207-d3411bbcb6ea",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "singlet    108183\n",
       "doublet     23332\n",
       "Name: scDblFinder_class, dtype: int64"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata.obs['scDblFinder_class'] = np.expand_dims(droplet_class, axis=1)\n",
    "adata.obs['scDblFinder_class'].value_counts()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "74e18ca0-da68-4330-ad61-8d0d3f972e50",
   "metadata": {},
   "source": [
    "The resulting classification into singlets and doublets is added to our anndata object. In our dataset, scDblFinder identified 58,264 droplets as doublets. Those are removed from the data."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "dc6b624b-fa52-472c-9a89-0a280e70a24e",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total number of cells: 131515\n",
      "Number of cells after filtering of doublets: 108183\n"
     ]
    }
   ],
   "source": [
    "print('Total number of cells: {:d}'.format(adata.n_obs))\n",
    "adata = adata[adata.obs['scDblFinder_class']=='singlet'].copy()\n",
    "\n",
    "print('Number of cells after filtering of doublets: {:d}'.format(adata.n_obs))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "24f10385-90c1-4167-8df2-a981c7e4313d",
   "metadata": {},
   "source": [
    "## 2.2 Quality Control\n",
    "\n",
    "Apart from doublets, the dataset might contain low-quality cells which are filtered by cell quality control (QC). Cell QC is typically performed on the following three QC covariates: \n",
    "\n",
    "* number of counts per barcode (count depth),\n",
    "\n",
    "* number of genes per barcode, and\n",
    "\n",
    "* fraction of counts from mitochondrial genes per barcode. \n",
    "\n",
    "In cell QC these covariates are filtered via thresholding as outlier barcodes might correspond to dying cells, so cells with broken membranes whose cytoplasmic mRNA has leaked out and therefore only the mRNA in the mitochondria is still present. These cells might then show a low count depth, few detected genes and a high fraction of mitochondrial reads.\n",
    "\n",
    "However, it is important to consider the three QC covariates jointly as otherwise this might lead to misinterpretation of cellular signal. One example are cells with a relativly high fraction of mitochondrial counts which might be involved in respiratory processes and should not be filtered out. Another example are cells with low or high counts which might correspond to quiescent cell populations or cells larger in size. Therefore, it is prefered to consider multiple covariates when univariate thesholding decisions are made. In general, it is advised to exclude fewer cells and be as permissive as possible to avoid filtering out viable cell populations or small sub-populations. \n",
    "\n",
    "QC on datasets is often performed in a manual fashion by looking at the distribution of different QC covariates. However, as these datasets grow in size it might be worth to consider automatic thresholding via MAD (median absolute deviations). {cite}`germain_pipecomp_2020` proposed in their study on different preprocessing steps to mark cells as outliers if they differ by 5 MADs which relects a relatively permissive filtering strategy. \n",
    "\n",
    "In the analysis workflow, one first needs to calculate the QC covariates (QC metrics). We use the scanpy function `sc.pp.calculate_qc_metrics()` for this manner and additionally compute the log1p transformed values. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "98d1b455-fdb6-4ed7-aaaf-990f2b4152e2",
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.pp.calculate_qc_metrics(adata, inplace=True, percent_top=[20], log1p=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "dae9df50-4b01-4954-8625-9143b81e2e1b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "AnnData object with n_obs × n_vars = 108183 × 36601\n",
       "    obs: 'site', 'donor', 'scDblFinder_class', 'n_genes_by_counts', 'log1p_n_genes_by_counts', 'total_counts', 'log1p_total_counts', 'pct_counts_in_top_20_genes'\n",
       "    var: 'gene_ids', 'feature_types', 'genome', 'n_cells_by_counts', 'mean_counts', 'log1p_mean_counts', 'pct_dropout_by_counts', 'total_counts', 'log1p_total_counts'"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2fbc43d1-ad47-4ec2-81dd-ceae93645558",
   "metadata": {},
   "source": [
    "Next, we define a function that allows automatic filtering based on MAD. This function takes a `metric`, i.e. a column in `.obs` and the number of MADs (nmad) that is still permissive within the filtering strategy. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "4220cf26-0b1c-4e30-89a9-1405f9a22e22",
   "metadata": {},
   "outputs": [],
   "source": [
    "def isOutlier(metric, nmads):\n",
    "    M = adata.obs[metric]\n",
    "    outlier = (M < np.median(M) - nmads * M.mad()) | (np.median(M) + nmads * M.mad() < M)\n",
    "    return outlier"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "44980e02-ad05-4868-9df2-fafaca0c66ba",
   "metadata": {},
   "source": [
    "We now apply this function to the `log1p_total_counts`, `log1p_n_genes_by_counts` and `pct_counts_in_top_20_genes` QC covariates each with a threshold of 5 MADs."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "18bc03fe-919f-456c-927a-4222f77e5d7b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "False    104883\n",
       "True       3067\n",
       "Name: outlier, dtype: int64"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata.obs['outlier'] = isOutlier('log1p_total_counts', 5) | isOutlier('log1p_n_genes_by_counts', 5) | isOutlier('pct_counts_in_top_20_genes', 5) \n",
    "adata.obs['outlier'].value_counts()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "612ca405-8dcf-4231-8602-750e36561f56",
   "metadata": {},
   "source": [
    "So we classify 3,067 cells as outlier with respect to these three metrics. \n",
    "\n",
    "As a next step, we calculate the percentage of mitochondrial counts as they are often associated with cell degradation. It is important to note that mitochondrial counts are annotated either with the prefix \"mt-\" or \"MT-\" depending on the species considered in the dataset. The dataset considered in this notebook is .... ?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "392fef30-c8e4-440e-8293-fdf5c37e789e",
   "metadata": {},
   "outputs": [],
   "source": [
    "mt_gene_mask = [gene.startswith('MT-') for gene in adata.var_names]\n",
    "adata.obs['pct_counts_Mt'] = adata.X.toarray()[:, mt_gene_mask].sum(1)/adata.obs['total_counts']"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2373e14e-2f2d-4133-814b-60ab3154d676",
   "metadata": {},
   "source": [
    "`pct_counts_Mt` is filtered with 3 MADs. Additionally, cells with a percentage of mitochondrial counts exceding 8 % are filtered out."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "2d1c5137-1003-497c-82ac-8b1ac3807dae",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "False    101043\n",
       "True       6907\n",
       "Name: mt_outlier, dtype: int64"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata.obs['mt_outlier'] = isOutlier('pct_counts_Mt', 3) | (adata.obs['pct_counts_Mt'] > 0.08)\n",
    "adata.obs['mt_outlier'].value_counts()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f8985d6d-22b2-4c90-a35a-795d0a06f39e",
   "metadata": {},
   "source": [
    "So we classify 6,907 cells as outlier with respect to mitochonrial reads.\n",
    "\n",
    "We now filter our anndata object based on these two additional columns. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "204a7195-001b-428a-9151-dd28322e9bc0",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata = adata[(~adata.obs['outlier']) & (~adata.obs['mt_outlier'])].copy()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "d93388ab-ec0a-4eb0-9e39-c665a0614376",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total number of genes: 36601\n",
      "Number of genes after cell filter: 29437\n"
     ]
    }
   ],
   "source": [
    "#Filter genes:\n",
    "print('Total number of genes: {:d}'.format(adata.n_vars))\n",
    "\n",
    "# Min 5 cells - filters out 0 count genes\n",
    "sc.pp.filter_genes(adata, min_cells=5)\n",
    "print('Number of genes after cell filter: {:d}'.format(adata.n_vars))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1acfd6f9-cba2-4228-9122-5d5b3cdf8fd5",
   "metadata": {},
   "source": [
    "## 2.3 Normalization"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9a6b48de-1a7c-4786-b941-4a048537a6be",
   "metadata": {},
   "source": [
    "### 2.3.1 *sctransform* normalization"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ac7ceb44-99bc-4a6d-a41c-e545ea4a8675",
   "metadata": {},
   "source": [
    "### 2.3.2 *scran* normalization"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "28855f58-cda0-4280-b8e9-23bfbfe4be72",
   "metadata": {},
   "source": [
    "## References\n",
    "\n",
    "```{bibliography}\n",
    ":filter: docname in docnames\n",
    "```"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "b3b5053a-74e2-4a26-ba92-f58e49263d1b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<details>\n",
       "<summary>Click to view session information</summary>\n",
       "<pre>\n",
       "-----\n",
       "anndata             0.7.6\n",
       "anndata2ri          1.0.6\n",
       "numpy               1.20.3\n",
       "pandas              1.3.3\n",
       "rpy2                3.4.2\n",
       "scanpy              1.8.1\n",
       "scipy               1.7.1\n",
       "session_info        1.0.0\n",
       "-----\n",
       "</pre>\n",
       "<details>\n",
       "<summary>Click to view modules imported as dependencies</summary>\n",
       "<pre>\n",
       "PIL                         8.3.2\n",
       "anyio                       NA\n",
       "appnope                     0.1.2\n",
       "attr                        21.2.0\n",
       "babel                       2.9.1\n",
       "backcall                    0.2.0\n",
       "backports                   NA\n",
       "beta_ufunc                  NA\n",
       "binom_ufunc                 NA\n",
       "certifi                     2021.05.30\n",
       "cffi                        1.14.6\n",
       "charset_normalizer          2.0.6\n",
       "cycler                      0.10.0\n",
       "cython_runtime              NA\n",
       "dateutil                    2.8.2\n",
       "debugpy                     1.4.3\n",
       "decorator                   5.1.0\n",
       "defusedxml                  0.7.1\n",
       "dunamai                     1.6.0\n",
       "entrypoints                 0.3\n",
       "get_version                 3.5\n",
       "h5py                        3.4.0\n",
       "idna                        3.2\n",
       "igraph                      0.9.8\n",
       "ipykernel                   6.4.1\n",
       "ipython_genutils            0.2.0\n",
       "jedi                        0.18.0\n",
       "jinja2                      3.0.1\n",
       "joblib                      1.0.1\n",
       "json5                       NA\n",
       "jsonschema                  3.2.0\n",
       "jupyter_server              1.11.0\n",
       "jupyterlab_server           2.8.1\n",
       "kiwisolver                  1.3.2\n",
       "leidenalg                   0.8.8\n",
       "llvmlite                    0.37.0\n",
       "markupsafe                  2.0.1\n",
       "matplotlib                  3.3.3\n",
       "matplotlib_inline           NA\n",
       "mpl_toolkits                NA\n",
       "natsort                     7.1.1\n",
       "nbclassic                   NA\n",
       "nbformat                    5.1.3\n",
       "nbinom_ufunc                NA\n",
       "numba                       0.54.0\n",
       "numexpr                     2.7.3\n",
       "packaging                   21.0\n",
       "parso                       0.8.2\n",
       "pexpect                     4.8.0\n",
       "pickleshare                 0.7.5\n",
       "pkg_resources               NA\n",
       "prometheus_client           NA\n",
       "prompt_toolkit              3.0.20\n",
       "ptyprocess                  0.7.0\n",
       "pvectorc                    NA\n",
       "pycparser                   2.20\n",
       "pydev_ipython               NA\n",
       "pydevconsole                NA\n",
       "pydevd                      2.4.1\n",
       "pydevd_concurrency_analyser NA\n",
       "pydevd_file_utils           NA\n",
       "pydevd_plugins              NA\n",
       "pydevd_tracing              NA\n",
       "pygments                    2.10.0\n",
       "pyparsing                   2.4.7\n",
       "pyrsistent                  NA\n",
       "pytz                        2021.1\n",
       "requests                    2.26.0\n",
       "send2trash                  NA\n",
       "sinfo                       0.3.4\n",
       "six                         1.16.0\n",
       "sklearn                     0.24.2\n",
       "sniffio                     1.2.0\n",
       "storemagic                  NA\n",
       "tables                      3.6.1\n",
       "terminado                   0.12.1\n",
       "texttable                   1.6.4\n",
       "tornado                     6.1\n",
       "traitlets                   5.1.0\n",
       "typing_extensions           NA\n",
       "tzlocal                     NA\n",
       "urllib3                     1.26.6\n",
       "wcwidth                     0.2.5\n",
       "websocket                   1.2.1\n",
       "zmq                         22.3.0\n",
       "</pre>\n",
       "</details> <!-- seems like this ends pre, so might as well be explicit -->\n",
       "<pre>\n",
       "-----\n",
       "IPython             7.27.0\n",
       "jupyter_client      7.0.3\n",
       "jupyter_core        4.8.1\n",
       "jupyterlab          3.1.12\n",
       "notebook            6.4.4\n",
       "-----\n",
       "Python 3.8.11 (default, Aug  6 2021, 08:56:27) [Clang 10.0.0 ]\n",
       "macOS-10.16-x86_64-i386-64bit\n",
       "-----\n",
       "Session information updated at 2022-05-05 09:59\n",
       "</pre>\n",
       "</details>"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import session_info\n",
    "session_info.show()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.11"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
