
<!DOCTYPE html>


<html lang="en" data-content_root="../" >

  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="viewport" content="width=device-width, initial-scale=1" />

    <title>16. Differential gene expression analysis &#8212; Single-cell best practices</title>
  
  
  
  <script data-cfasync="false">
    document.documentElement.dataset.mode = localStorage.getItem("mode") || "";
    document.documentElement.dataset.theme = localStorage.getItem("theme") || "";
  </script>
  
  <!-- Loaded before other Sphinx assets -->
  <link href="../_static/styles/theme.css?digest=dfe6caa3a7d634c4db9b" rel="stylesheet" />
<link href="../_static/styles/bootstrap.css?digest=dfe6caa3a7d634c4db9b" rel="stylesheet" />
<link href="../_static/styles/pydata-sphinx-theme.css?digest=dfe6caa3a7d634c4db9b" rel="stylesheet" />

  
  <link href="../_static/vendor/fontawesome/6.5.2/css/all.min.css?digest=dfe6caa3a7d634c4db9b" rel="stylesheet" />
  <link rel="preload" as="font" type="font/woff2" crossorigin href="../_static/vendor/fontawesome/6.5.2/webfonts/fa-solid-900.woff2" />
<link rel="preload" as="font" type="font/woff2" crossorigin href="../_static/vendor/fontawesome/6.5.2/webfonts/fa-brands-400.woff2" />
<link rel="preload" as="font" type="font/woff2" crossorigin href="../_static/vendor/fontawesome/6.5.2/webfonts/fa-regular-400.woff2" />

    <link rel="stylesheet" type="text/css" href="../_static/pygments.css?v=03e43079" />
    <link rel="stylesheet" type="text/css" href="../_static/styles/sphinx-book-theme.css?v=eba8b062" />
    <link rel="stylesheet" type="text/css" href="../_static/togglebutton.css?v=13237357" />
    <link rel="stylesheet" type="text/css" href="../_static/copybutton.css?v=76b2166b" />
    <link rel="stylesheet" type="text/css" href="../_static/mystnb.4510f1fc1dee50b3e5859aac5469c37c29e427902b24a333a5f9fcb2f0b3ac41.css" />
    <link rel="stylesheet" type="text/css" href="../_static/sphinx-thebe.css?v=4fa983c6" />
    <link rel="stylesheet" type="text/css" href="../_static/sphinx-design.min.css?v=95c83b7e" />
    <link rel="stylesheet" type="text/css" href="../_static/book.css?v=148c11e2" />
  
  <!-- Pre-loaded scripts that we'll load fully later -->
  <link rel="preload" as="script" href="../_static/scripts/bootstrap.js?digest=dfe6caa3a7d634c4db9b" />
<link rel="preload" as="script" href="../_static/scripts/pydata-sphinx-theme.js?digest=dfe6caa3a7d634c4db9b" />
  <script src="../_static/vendor/fontawesome/6.5.2/js/all.min.js?digest=dfe6caa3a7d634c4db9b"></script>

    <script src="../_static/documentation_options.js?v=9eb32ce0"></script>
    <script src="../_static/doctools.js?v=9a2dae69"></script>
    <script src="../_static/sphinx_highlight.js?v=dc90522c"></script>
    <script src="../_static/clipboard.min.js?v=a7894cd8"></script>
    <script src="../_static/copybutton.js?v=f281be69"></script>
    <script src="../_static/scripts/sphinx-book-theme.js?v=887ef09a"></script>
    <script>let toggleHintShow = 'Click to show';</script>
    <script>let toggleHintHide = 'Click to hide';</script>
    <script>let toggleOpenOnPrint = 'true';</script>
    <script src="../_static/togglebutton.js?v=4a39c7ea"></script>
    <script>var togglebuttonSelector = '.toggle, .admonition.dropdown';</script>
    <script src="../_static/design-tabs.js?v=f930bc37"></script>
    <script>const THEBE_JS_URL = "https://unpkg.com/thebe@0.8.2/lib/index.js"; const thebe_selector = ".thebe,.cell"; const thebe_selector_input = "pre"; const thebe_selector_output = ".output, .cell_output"</script>
    <script async="async" src="../_static/sphinx-thebe.js?v=c100c467"></script>
    <script>var togglebuttonSelector = '.toggle, .admonition.dropdown';</script>
    <script>const THEBE_JS_URL = "https://unpkg.com/thebe@0.8.2/lib/index.js"; const thebe_selector = ".thebe,.cell"; const thebe_selector_input = "pre"; const thebe_selector_output = ".output, .cell_output"</script>
    <script>DOCUMENTATION_OPTIONS.pagename = 'conditions/differential_gene_expression';</script>
    <script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
    <script src="../_static/book.js?v=57d67cd3"></script>
    <link rel="icon" href="../_static/favicon.ico"/>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="17. Compositional analysis" href="compositional.html" />
    <link rel="prev" title="15. Lineage tracing" href="../trajectories/lineage_tracing.html" />
  <meta name="viewport" content="width=device-width, initial-scale=1"/>
  <meta name="docsearch:language" content="en"/>
  </head>
  
  
  <body data-bs-spy="scroll" data-bs-target=".bd-toc-nav" data-offset="180" data-bs-root-margin="0px 0px -60%" data-default-mode="">

  
  
  <div id="pst-skip-link" class="skip-link d-print-none"><a href="#main-content">Skip to main content</a></div>
  
  <div id="pst-scroll-pixel-helper"></div>
  
  <button type="button" class="btn rounded-pill" id="pst-back-to-top">
    <i class="fa-solid fa-arrow-up"></i>Back to top</button>

  
  <input type="checkbox"
          class="sidebar-toggle"
          id="pst-primary-sidebar-checkbox"/>
  <label class="overlay overlay-primary" for="pst-primary-sidebar-checkbox"></label>
  
  <input type="checkbox"
          class="sidebar-toggle"
          id="pst-secondary-sidebar-checkbox"/>
  <label class="overlay overlay-secondary" for="pst-secondary-sidebar-checkbox"></label>
  
  <div class="search-button__wrapper">
    <div class="search-button__overlay"></div>
    <div class="search-button__search-container">
<form class="bd-search d-flex align-items-center"
      action="../search.html"
      method="get">
  <i class="fa-solid fa-magnifying-glass"></i>
  <input type="search"
         class="form-control"
         name="q"
         id="search-input"
         placeholder="Search this book..."
         aria-label="Search this book..."
         autocomplete="off"
         autocorrect="off"
         autocapitalize="off"
         spellcheck="false"/>
  <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd>K</kbd></span>
</form></div>
  </div>

  <div class="pst-async-banner-revealer d-none">
  <aside id="bd-header-version-warning" class="d-none d-print-none" aria-label="Version warning"></aside>
</div>

  
    <header class="bd-header navbar navbar-expand-lg bd-navbar d-print-none">
    </header>
  

  <div class="bd-container">
    <div class="bd-container__inner bd-page-width">
      
      
      
      <div class="bd-sidebar-primary bd-sidebar">
        

  
  <div class="sidebar-header-items sidebar-primary__section">
    
    
    
    
  </div>
  
    <div class="sidebar-primary-items__start sidebar-primary__section">
        <div class="sidebar-primary-item">

  
    
  

<a class="navbar-brand logo" href="../preamble.html">
  
  
  
  
  
    
    
      
    
    
    <img src="../_static/logo.png" class="logo__image only-light" alt="Single-cell best practices - Home"/>
    <script>document.write(`<img src="../_static/logo.png" class="logo__image only-dark" alt="Single-cell best practices - Home"/>`);</script>
  
  
</a></div>
        <div class="sidebar-primary-item">

 <script>
 document.write(`
   <button class="btn search-button-field search-button__button" title="Search" aria-label="Search" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="fa-solid fa-magnifying-glass"></i>
    <span class="search-button__default-text">Search</span>
    <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd class="kbd-shortcut__modifier">K</kbd></span>
   </button>
 `);
 </script></div>
        <div class="sidebar-primary-item"><nav class="bd-links bd-docs-nav" aria-label="Main">
    <div class="bd-toc-item navbar-nav active">
        <p aria-level="2" class="caption" role="heading"><span class="caption-text">Introduction</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../introduction/prior_art.html">1. Prior art</a></li>
<li class="toctree-l1"><a class="reference internal" href="../introduction/scrna_seq.html">2. Single-cell RNA sequencing</a></li>
<li class="toctree-l1"><a class="reference internal" href="../introduction/raw_data_processing.html">3. Raw data processing</a></li>
<li class="toctree-l1"><a class="reference internal" href="../introduction/analysis_tools.html">4. Analysis frameworks and tools</a></li>
<li class="toctree-l1"><a class="reference internal" href="../introduction/interoperability.html">5. Interoperability</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Preprocessing and visualization</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../preprocessing_visualization/quality_control.html">6. Quality Control</a></li>
<li class="toctree-l1"><a class="reference internal" href="../preprocessing_visualization/normalization.html">7. Normalization</a></li>
<li class="toctree-l1"><a class="reference internal" href="../preprocessing_visualization/feature_selection.html">8. Feature selection</a></li>
<li class="toctree-l1"><a class="reference internal" href="../preprocessing_visualization/dimensionality_reduction.html">9. Dimensionality Reduction</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Identifying cellular structure</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../cellular_structure/clustering.html">10. Clustering</a></li>
<li class="toctree-l1"><a class="reference internal" href="../cellular_structure/annotation.html">11. Annotation</a></li>
<li class="toctree-l1"><a class="reference internal" href="../cellular_structure/integration.html">12. Data integration</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Inferring trajectories</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../trajectories/pseudotemporal.html">13. Pseudotemporal ordering</a></li>
<li class="toctree-l1"><a class="reference internal" href="../trajectories/rna_velocity.html">14. RNA velocity</a></li>
<li class="toctree-l1"><a class="reference internal" href="../trajectories/lineage_tracing.html">15. Lineage tracing</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Dealing with conditions</span></p>
<ul class="current nav bd-sidenav">
<li class="toctree-l1 current active"><a class="current reference internal" href="#">16. Differential gene expression analysis</a></li>
<li class="toctree-l1"><a class="reference internal" href="compositional.html">17. Compositional analysis</a></li>
<li class="toctree-l1"><a class="reference internal" href="gsea_pathway.html">18. Gene set enrichment and pathway analysis</a></li>
<li class="toctree-l1"><a class="reference internal" href="perturbation_modeling.html">19. Perturbation modeling</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Modeling mechanisms</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../mechanisms/gene_regulatory_networks.html">20. Gene regulatory networks</a></li>
<li class="toctree-l1"><a class="reference internal" href="../mechanisms/cell_cell_communication.html">21. Cell-cell communication</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Deconvolution</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../deconvolution/bulk_deconvolution.html">22. Bulk deconvolution</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Chromatin Accessibility</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../chromatin_accessibility/introduction.html">23. Single-cell ATAC sequencing</a></li>
<li class="toctree-l1"><a class="reference internal" href="../chromatin_accessibility/quality_control.html">24. Quality Control</a></li>
<li class="toctree-l1"><a class="reference internal" href="../chromatin_accessibility/gene_regulatory_networks_atac.html">25. Gene regulatory networks using chromatin accessibility</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Spatial omics</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../spatial/introduction.html">26. Single-cell data resolved in space</a></li>
<li class="toctree-l1"><a class="reference internal" href="../spatial/neighborhood.html">27. Neighborhood analysis</a></li>
<li class="toctree-l1"><a class="reference internal" href="../spatial/domains.html">28. Spatial domains</a></li>
<li class="toctree-l1"><a class="reference internal" href="../spatial/spatially_variable_genes.html">29. Spatially variable genes</a></li>
<li class="toctree-l1"><a class="reference internal" href="../spatial/deconvolution.html">30. Spatial deconvolution</a></li>
<li class="toctree-l1"><a class="reference internal" href="../spatial/imputation.html">31. Imputation</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Surface protein</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../surface_protein/quality_control.html">32. Quality control</a></li>
<li class="toctree-l1"><a class="reference internal" href="../surface_protein/normalization.html">33. Normalization</a></li>
<li class="toctree-l1"><a class="reference internal" href="../surface_protein/doublet_detection.html">34. Doublet detection</a></li>
<li class="toctree-l1"><a class="reference internal" href="../surface_protein/dimensionality_reduction.html">35. Dimensionality Reduction</a></li>
<li class="toctree-l1"><a class="reference internal" href="../surface_protein/batch_correction.html">36. Batch correction</a></li>
<li class="toctree-l1"><a class="reference internal" href="../surface_protein/annotation.html">37. Annotation</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Adaptive immune receptor repertoire</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../air_repertoire/ir_profiling.html">38. Immune Receptor Profiling</a></li>
<li class="toctree-l1"><a class="reference internal" href="../air_repertoire/clonotype.html">39. Clonotype analysis</a></li>



<li class="toctree-l1"><a class="reference internal" href="../air_repertoire/specificity.html">43. Specificity analysis</a></li>
<li class="toctree-l1"><a class="reference internal" href="../air_repertoire/multimodal_integration.html">44. Integrating AIR and transcriptomics</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Multimodal integration</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../multimodal_integration/paired_integration.html">45. Paired integration</a></li>
<li class="toctree-l1"><a class="reference internal" href="../multimodal_integration/advanced_integration.html">46. Advanced integration</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Outlook</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../outlook.html">47. Outlook</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Acknowledgements</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../acknowledgements.html">48. Acknowledgements</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Glossary</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../glossary.html">49. Glossary</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Changelog</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../CHANGELOG.html">Changelog</a></li>
</ul>

    </div>
</nav></div>
    </div>
  
  
  <div class="sidebar-primary-items__end sidebar-primary__section">
  </div>
  
  <div id="rtd-footer-container"></div>


      </div>
      
      <main id="main-content" class="bd-main" role="main">
        
        

<div class="sbt-scroll-pixel-helper"></div>

          <div class="bd-content">
            <div class="bd-article-container">
              
              <div class="bd-header-article d-print-none">
<div class="header-article-items header-article__inner">
  
    <div class="header-article-items__start">
      
        <div class="header-article-item"><button class="sidebar-toggle primary-toggle btn btn-sm" title="Toggle primary sidebar" data-bs-placement="bottom" data-bs-toggle="tooltip">
  <span class="fa-solid fa-bars"></span>
</button></div>
      
    </div>
  
  
    <div class="header-article-items__end">
      
        <div class="header-article-item">

<div class="article-header-buttons">





<div class="dropdown dropdown-source-buttons">
  <button class="btn dropdown-toggle" type="button" data-bs-toggle="dropdown" aria-expanded="false" aria-label="Source repositories">
    <i class="fab fa-github"></i>
  </button>
  <ul class="dropdown-menu">
      
      
      
      <li><a href="https://github.com/theislab/single-cell-best-practices" target="_blank"
   class="btn btn-sm btn-source-repository-button dropdown-item"
   title="Source repository"
   data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fab fa-github"></i>
  </span>
<span class="btn__text-container">Repository</span>
</a>
</li>
      
      
      
      
      <li><a href="https://github.com/theislab/single-cell-best-practices/edit/master/jupyter-book/conditions/differential_gene_expression.ipynb" target="_blank"
   class="btn btn-sm btn-source-edit-button dropdown-item"
   title="Suggest edit"
   data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-pencil-alt"></i>
  </span>
<span class="btn__text-container">Suggest edit</span>
</a>
</li>
      
      
      
      
      <li><a href="https://github.com/theislab/single-cell-best-practices/issues/new?title=Issue%20on%20page%20%2Fconditions/differential_gene_expression.html&body=Your%20issue%20content%20here." target="_blank"
   class="btn btn-sm btn-source-issues-button dropdown-item"
   title="Open an issue"
   data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-lightbulb"></i>
  </span>
<span class="btn__text-container">Open issue</span>
</a>
</li>
      
  </ul>
</div>






<div class="dropdown dropdown-download-buttons">
  <button class="btn dropdown-toggle" type="button" data-bs-toggle="dropdown" aria-expanded="false" aria-label="Download this page">
    <i class="fas fa-download"></i>
  </button>
  <ul class="dropdown-menu">
      
      
      
      <li><a href="../_sources/conditions/differential_gene_expression.ipynb" target="_blank"
   class="btn btn-sm btn-download-source-button dropdown-item"
   title="Download source file"
   data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-file"></i>
  </span>
<span class="btn__text-container">.ipynb</span>
</a>
</li>
      
      
      
      
      <li>
<button onclick="window.print()"
  class="btn btn-sm btn-download-pdf-button dropdown-item"
  title="Print to PDF"
  data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-file-pdf"></i>
  </span>
<span class="btn__text-container">.pdf</span>
</button>
</li>
      
  </ul>
</div>




<button onclick="toggleFullScreen()"
  class="btn btn-sm btn-fullscreen-button"
  title="Fullscreen mode"
  data-bs-placement="bottom" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-expand"></i>
  </span>

</button>



<script>
document.write(`
  <button class="btn btn-sm nav-link pst-navbar-icon theme-switch-button" title="light/dark" aria-label="light/dark" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="theme-switch fa-solid fa-sun fa-lg" data-mode="light"></i>
    <i class="theme-switch fa-solid fa-moon fa-lg" data-mode="dark"></i>
    <i class="theme-switch fa-solid fa-circle-half-stroke fa-lg" data-mode="auto"></i>
  </button>
`);
</script>


<script>
document.write(`
  <button class="btn btn-sm pst-navbar-icon search-button search-button__button" title="Search" aria-label="Search" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="fa-solid fa-magnifying-glass fa-lg"></i>
  </button>
`);
</script>
<button class="sidebar-toggle secondary-toggle btn btn-sm" title="Toggle secondary sidebar" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <span class="fa-solid fa-list"></span>
</button>
</div></div>
      
    </div>
  
</div>
</div>
              
              

<div id="jb-print-docs-body" class="onlyprint">
    <h1>Differential gene expression analysis</h1>
    <!-- Table of contents -->
    <div id="print-main-content">
        <div id="jb-print-toc">
            
            <div>
                <h2> Contents </h2>
            </div>
            <nav aria-label="Page">
                <ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#motivation">16.1. Motivation</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#environment-setup">16.2. Environment setup</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#preparing-the-dataset">16.3. Preparing the dataset</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#pseudobulk">16.4. Pseudobulk</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#one-group">16.4.1. One group</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#multiple-groups">16.4.2. Multiple groups</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#notes-on-edger">16.4.3. Notes on edgeR</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#single-cell-specific">16.5. Single-cell specific</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#id21">16.5.1. One group</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#visualization">16.6. Visualization</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#quiz">16.7. Quiz</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#references">16.8. References</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#contributors">16.9. Contributors</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#authors">16.9.1. Authors</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#reviewers">16.9.2. Reviewers</a></li>
</ul>
</li>
</ul>
            </nav>
        </div>
    </div>
</div>

              
                
<div id="searchbox"></div>
                <article class="bd-article">
                  
  <section class="tex2jax_ignore mathjax_ignore" id="differential-gene-expression-analysis">
<span id="differential-analysis"></span><h1><span class="section-number">16. </span>Differential gene expression analysis<a class="headerlink" href="#differential-gene-expression-analysis" title="Link to this heading">#</a></h1>
<details class="sd-sphinx-override sd-dropdown sd-card sd-mb-3">
<summary class="sd-summary-title sd-card-header">
<span class="sd-summary-text"><i class="fas fa-brain"></i>   Key takeaways</span><span class="sd-summary-state-marker sd-summary-chevron-right"><svg version="1.1" width="1.5em" height="1.5em" class="sd-octicon sd-octicon-chevron-right" viewBox="0 0 24 24" aria-hidden="true"><path d="M8.72 18.78a.75.75 0 0 1 0-1.06L14.44 12 8.72 6.28a.751.751 0 0 1 .018-1.042.751.751 0 0 1 1.042-.018l6.25 6.25a.75.75 0 0 1 0 1.06l-6.25 6.25a.75.75 0 0 1-1.06 0Z"></path></svg></span></summary><div class="sd-summary-content sd-card-body docutils">
<div class="sd-card sd-sphinx-override sd-mb-3 sd-shadow-sm sd-card-hover docutils">
<div class="sd-card-body docutils">
<p class="sd-card-text">Repeated measurements (cells) from the same experimental subject in scRNA-seq introduce correlations between measurements, which need to be adjusted by modeling the experimental units as Random Effect to mitigate the pseudoreplication issue.
Account for them through a sum or mean aggregation (pseudobulk methods - by fixed- or random effect terms) or by accounting for individual as a random effect (single-cell methods) to mitigate the pseudoreplication issue.</p>
</div>
<a class="sd-stretched-link sd-hide-link-text reference internal" href="#conditions-differential-gene-expression-key-takeaway-1"><span class="std std-ref">Motivation</span></a></div>
<div class="sd-card sd-sphinx-override sd-mb-3 sd-shadow-sm sd-card-hover docutils">
<div class="sd-card-body docutils">
<p class="sd-card-text">Statistical power is best increased with more samples in the experimental design.</p>
</div>
<a class="sd-stretched-link sd-hide-link-text reference internal" href="#conditions-differential-gene-expression-key-takeaway-2"><span class="std std-ref">Pseudobulk</span></a></div>
</div>
</details><details class="sd-sphinx-override sd-dropdown sd-card sd-mb-3">
<summary class="sd-summary-title sd-card-header">
<span class="sd-summary-text"><i class="fa-solid fa-gear"></i>   Environment setup</span><span class="sd-summary-state-marker sd-summary-chevron-right"><svg version="1.1" width="1.5em" height="1.5em" class="sd-octicon sd-octicon-chevron-right" viewBox="0 0 24 24" aria-hidden="true"><path d="M8.72 18.78a.75.75 0 0 1 0-1.06L14.44 12 8.72 6.28a.751.751 0 0 1 .018-1.042.751.751 0 0 1 1.042-.018l6.25 6.25a.75.75 0 0 1 0 1.06l-6.25 6.25a.75.75 0 0 1-1.06 0Z"></path></svg></span></summary><div class="sd-summary-content sd-card-body docutils">
<div class="sd-tab-set docutils">
<input checked="checked" id="sd-tab-item-0" name="sd-tab-set-0" type="radio">
<label class="sd-tab-label" for="sd-tab-item-0">
Steps</label><div class="sd-tab-content docutils">
<ol class="arabic">
<li><p class="sd-card-text"><strong>Install conda</strong>:</p>
<ul class="simple">
<li><p class="sd-card-text">Before creating the environment, ensure that <a class="reference external" href="https://docs.conda.io/projects/conda/en/latest/user-guide/install/index.html">conda</a> is installed on your system.</p></li>
</ul>
</li>
<li><p class="sd-card-text"><strong>Save the yml content</strong>:</p>
<ul class="simple">
<li><p class="sd-card-text">Copy the content from the yml tab into a file named <code class="docutils literal notranslate"><span class="pre">environment.yml</span></code>.</p></li>
</ul>
</li>
<li><p class="sd-card-text"><strong>Create the environment</strong>:</p>
<ul>
<li><p class="sd-card-text">Open a terminal or command prompt.</p></li>
<li><p class="sd-card-text">Run the following command:</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>conda<span class="w"> </span>env<span class="w"> </span>create<span class="w"> </span>-f<span class="w"> </span>environment.yml
</pre></div>
</div>
</li>
</ul>
</li>
<li><p class="sd-card-text"><strong>Activate the environment</strong>:</p>
<ul>
<li><p class="sd-card-text">After the environment is created, activate it using:</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>conda<span class="w"> </span>activate<span class="w"> </span>&lt;environment_name&gt;
</pre></div>
</div>
</li>
<li><p class="sd-card-text">Replace <code class="docutils literal notranslate"><span class="pre">&lt;environment_name&gt;</span></code> with the name specified in the <code class="docutils literal notranslate"><span class="pre">environment.yml</span></code> file. In the yml file it will look like this:</p>
<div class="highlight-yaml notranslate"><div class="highlight"><pre><span></span><span class="nt">name</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">&lt;environment_name&gt;</span>
</pre></div>
</div>
</li>
</ul>
</li>
<li><p class="sd-card-text"><strong>Verify the installation</strong>:</p>
<ul>
<li><p class="sd-card-text">Check that the environment was created successfully by running:</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>conda<span class="w"> </span>env<span class="w"> </span>list
</pre></div>
</div>
</li>
</ul>
</li>
</ol>
</div>
<input id="sd-tab-item-1" name="sd-tab-set-0" type="radio">
<label class="sd-tab-label" for="sd-tab-item-1">
yml</label><div class="sd-tab-content docutils">
<div class="highlight-yaml notranslate"><div class="highlight"><pre><span></span><span class="nt">name</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">differential-gene-expression</span>
<span class="nt">channels</span><span class="p">:</span>
<span class="w">  </span><span class="p p-Indicator">-</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">conda-forge</span>
<span class="w">  </span><span class="p p-Indicator">-</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">bioconda</span>
<span class="w">  </span><span class="p p-Indicator">-</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">defaults</span>
<span class="nt">dependencies</span><span class="p">:</span>
<span class="w">  </span><span class="p p-Indicator">-</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">conda-forge::python=3.9.15</span>
<span class="w">  </span><span class="p p-Indicator">-</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">conda-forge::scanpy=1.9.1</span>
<span class="w">  </span><span class="p p-Indicator">-</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">conda-forge::r-base=4.2.2</span>
<span class="w">  </span><span class="p p-Indicator">-</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">bioconda::bioconductor-mast=1.24.0</span>
<span class="w">  </span><span class="p p-Indicator">-</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">bioconda::bioconductor-edger=3.40.0</span>
<span class="w">  </span><span class="p p-Indicator">-</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">bioconda::bioconductor-singlecellexperiment=1.20.0</span>
<span class="w">  </span><span class="p p-Indicator">-</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">conda-forge::r-lme4</span>
<span class="w">  </span><span class="p p-Indicator">-</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">conda-forge::notebook</span>
<span class="w">  </span><span class="p p-Indicator">-</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">conda-forge::rpy2=3.5.6</span>
<span class="w">  </span><span class="p p-Indicator">-</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">bioconda::anndata2ri=1.1</span>
<span class="w">  </span><span class="p p-Indicator">-</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">conda-forge::pip</span>
<span class="w">  </span><span class="p p-Indicator">-</span><span class="w"> </span><span class="nt">pip</span><span class="p">:</span>
<span class="w">      </span><span class="p p-Indicator">-</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">sc-toolbox</span>
<span class="w">      </span><span class="p p-Indicator">-</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">pertpy==0.5.0</span>
</pre></div>
</div>
</div>
</div>
</div>
</details><!-- END DROPDOWNS -->
<section id="motivation">
<span id="conditions-differential-gene-expression-key-takeaway-1"></span><h2><span class="section-number">16.1. </span>Motivation<a class="headerlink" href="#motivation" title="Link to this heading">#</a></h2>
<p>This chapter is a more detailed continuation of the Annotation subchapter which already introduced differential gene expression (DGE) as a tool to annotate clusters with cell types.
Here, we focus on more advanced use-cases of differential gene expression testing on more complex experimental designs which involve one or more conditions such as diseases, genetic knockouts or drugs.
In such cases we are commonly interested in the magnitude and significance of differences in gene expression patterns between the condition of interest and a reference. This reference can be everything but is commonly a healthy sample. This statistical test can be applied to arbitrary groups, but in the case of single-cell RNA-Seq is commonly applied on the cell type level.</p>
<figure class="align-default" id="differential-gene-expression-overview">
<a class="bg-primary mb-1 reference internal image-reference" href="../_images/differential_gene_expression.jpg"><img alt="DGE analysis overview" class="bg-primary mb-1" src="../_images/differential_gene_expression.jpg" style="width: 800px;" />
</a>
<figcaption>
<p><span class="caption-number">Fig. 16.1 </span><span class="caption-text">Differential gene expression analysis attempts to infer genes that are statistically significantly over- or underexpressed between any compared groups (commonly between healthy and condition per cell type).</span><a class="headerlink" href="#differential-gene-expression-overview" title="Link to this image">#</a></p>
</figcaption>
</figure>
<p>The outcome of such an analysis could be genesets which effect and potentially explain any observed phenotypes. These genesets can then be examined more closely with respect to, for example, affected pathways or induced cell-cell communication changes.</p>
<p>A differential gene expression test usually returns the log2 fold-change and the adjusted p-value per compared genes per compared conditions. This list can then be sorted by p-value and investigated in more detail.</p>
<p>The popular student’s t-test is one way of conducting such a test. However, it fails to take several single-cell RNA-seq peculiarities into account such as the excess number of zeros originating from dropouts or the need for complex experimental designs. More specifically, very rarely does one have sufficient sample numbers to accurately estimate the variance without pooling information across genes. Moreover, raw counts are never an absolute measurement of expression for a specific gene within a given sample. The actual read number per gene depends on the efficiency of the library preparation, the amount of contamination from non-coding transcripts and the sequencing depth. It does therefore lack in both, sensitivity and specificity for single-cell RNA-seq, let alone experimental design flexibility.</p>
<p>As a result, differential gene expression testing is a classic bioinformatics problem which has been tackled by many tools already. Generally, the problem is currently being approached from two views, the sample-level view where expression is aggregated to create “pseudobulks” and then analysed with methods originally designed for bulk expression samples such as edgeR<span id="id1">[<a class="reference internal" href="#id317" title="Mark D. Robinson, Davis J. McCarthy, and Gordon K. Smyth. Edger: a bioconductor package for differential expression analysis of digital gene expression data. Bioinformatics (Oxford, England), 26(1):139-140, Jan 2010. btp616[PII]. URL: https://doi.org/10.1093/bioinformatics/btp616, doi:10.1093/bioinformatics/btp616.">Robinson <em>et al.</em>, 2010</a>]</span> or DEseq2<span id="id2">[<a class="reference internal" href="#id318" title="Michael I. Love, Wolfgang Huber, and Simon Anders. Moderated estimation of fold change and dispersion for term`rna`-seq data with deseq2. Genome Biology, 15(12):550, Dec 2014. URL: https://doi.org/10.1186/s13059-014-0550-8, doi:10.1186/s13059-014-0550-8.">Love <em>et al.</em>, 2014</a>]</span> and the cell-level view where cells are modeled individually using generalized mixed effect models such as MAST<span id="id3">[<a class="reference internal" href="#id322" title="Greg Finak, Andrew McDavid, Masanao Yajima, Jingyuan Deng, Vivian Gersuk, Alex K. Shalek, Chloe K. Slichter, Hannah W. Miller, M. Juliana McElrath, Martin Prlic, Peter S. Linsley, and Raphael Gottardo. Mast: a flexible statistical framework for assessing transcriptional changes and characterizing heterogeneity in single-cell term`rna` sequencing data. Genome Biology, 16(1):278, Dec 2015. URL: https://doi.org/10.1186/s13059-015-0844-5, doi:10.1186/s13059-015-0844-5.">Finak <em>et al.</em>, 2015</a>]</span> or glmmTMB<span id="id4">[<a class="reference internal" href="#id323" title="Mollie E. Brooks, Kasper Kristensen, Koen J. van Benthem, Arni Magnusson, Casper W. Berg, Anders Nielsen, Hans J. Skaug, Martin Mächler, and Benjamin M. Bolker. glmmTMB Balances Speed and Flexibility Among Packages for Zero-inflated Generalized Linear Mixed Modeling. The R Journal, 9(2):378–400, 2017. URL: https://doi.org/10.32614/RJ-2017-066, doi:10.32614/RJ-2017-066.">Brooks <em>et al.</em>, 2017</a>]</span>. The consensus and robustness across datasets for DGE tools is low <span id="id5">[<a class="reference internal" href="#id324" title="Samarendra Das, Anil Rai, Michael L Merchant, Matthew C Cave, and Shesh N Rai. A comprehensive survey of statistical approaches for differential expression analysis in single-cell term`RNA` sequencing studies. Genes (Basel), 12(12):1947, December 2021.">Das <em>et al.</em>, 2021</a>, <a class="reference internal" href="#id319" title="Tianyu Wang, Boyang Li, Craig E. Nelson, and Sheida Nabavi. Comparative analysis of differential gene expression analysis tools for single-cell term`rna` sequencing data. BMC Bioinformatics, 20(1):40, Jan 2019. URL: https://doi.org/10.1186/s12859-019-2599-6, doi:10.1186/s12859-019-2599-6.">Wang <em>et al.</em>, 2019</a>]</span>. As previously described, although single-cell data contains technical noise artifacts such as dropout, zero-inflation and high cell-to-cell variability <span id="id6">[<a class="reference internal" href="#id320" title="Stephanie C Hicks, F William Townes, Mingxiang Teng, and Rafael A Irizarry. Missing data and technical variability in single-cell term`RNA`-sequencing experiments. Biostatistics, 19(4):562-578, 11 2017. URL: https://doi.org/10.1093/biostatistics/kxx053, arXiv:https://academic.oup.com/biostatistics/article-pdf/19/4/562/26346801/kxx053.pdf, doi:10.1093/biostatistics/kxx053.">Hicks <em>et al.</em>, 2017</a>, <a class="reference internal" href="#id338" title="Malte D Luecken and Fabian J Theis. Current best practices in single-cell term`rna`-seq analysis: a tutorial. Molecular Systems Biology, 15(6):e8746, 2019. URL: https://www.embopress.org/doi/abs/10.15252/msb.20188746, arXiv:https://www.embopress.org/doi/pdf/10.15252/msb.20188746, doi:https://doi.org/10.15252/msb.20188746.">Luecken and Theis, 2019</a>, <a class="reference internal" href="#id321" title="Catalina A. Vallejos, Davide Risso, Antonio Scialdone, Sandrine Dudoit, and John C. Marioni. Normalizing single-cell term`rna` sequencing data: challenges and opportunities. Nature Methods, 14(6):565-571, Jun 2017. URL: https://doi.org/10.1038/nmeth.4292, doi:10.1038/nmeth.4292.">Vallejos <em>et al.</em>, 2017</a>]</span>, methods designed for bulk <a class="reference internal" href="../glossary.html#term-RNA"><span class="xref std std-term">RNA</span></a>-seq data performed favorably compared to methods explicitly designed for scRNA-seq data<span id="id7">[<a class="reference internal" href="#id324" title="Samarendra Das, Anil Rai, Michael L Merchant, Matthew C Cave, and Shesh N Rai. A comprehensive survey of statistical approaches for differential expression analysis in single-cell term`RNA` sequencing studies. Genes (Basel), 12(12):1947, December 2021.">Das <em>et al.</em>, 2021</a>, <a class="reference internal" href="#id326" title="Maria K Jaakkola, Fatemeh Seyeterm`DNA`srollah, Arfa Mehmood, and Laura L Elo. Comparison of methods to detect differentially expressed genes between single-cell populations. Briefings in Bioinformatics, 18(5):735-743, 07 2016. URL: https://doi.org/10.1093/bib/bbw057, arXiv:https://academic.oup.com/bib/article-pdf/18/5/735/25581122/bbw057.pdf, doi:10.1093/bib/bbw057.">Jaakkola <em>et al.</em>, 2016</a>, <a class="reference internal" href="#id325" title="Charlotte Soneson and Mark D. Robinson. Bias, robustness and scalability in single-cell differential expression analysis. Nature Methods, 15(4):255-261, Apr 2018. URL: https://doi.org/10.1038/nmeth.4612, doi:10.1038/nmeth.4612.">Soneson and Robinson, 2018</a>, <a class="reference internal" href="#id327" title="Jordan W. Squair, Matthieu Gautier, Claudia Kathe, Mark A. Anderson, Nicholas D. James, Thomas H. Hutson, Rémi Hudelle, Taha Qaiser, Kaya J. E. Matson, Quentin Barraud, Ariel J. Levine, Gioele La Manno, Michael A. Skinnider, and Grégoire Courtine. Confronting false discoveries in single-cell differential expression. Nature Communications, 12(1):5692, Sep 2021. URL: https://doi.org/10.1038/s41467-021-25960-2, doi:10.1038/s41467-021-25960-2.">Squair <em>et al.</em>, 2021</a>]</span>. Single-cell specific methods were found to be especially prone to wrongly labeling highly expressed genes as differentially expressed.</p>
<p>A recent study highlighted the issue of pseudoreplication where inferential statistics is applied to biological replicates which are not statistically independent. Failing to account for the inherent correlation of replicates (cells from the same individual) inflates the false discovery rate (FDR)<span id="id8">[<a class="reference internal" href="#id331" title="Sini Junttila, Johannes Smolander, and Laura L Elo. Benchmarking methods for detecting differential states between conditions from multi-subject single-cell term`rna`-seq data. bioRxiv, 2022. URL: https://www.biorxiv.org/content/early/2022/02/19/2022.02.16.480662, arXiv:https://www.biorxiv.org/content/early/2022/02/19/2022.02.16.480662.full.pdf, doi:10.1101/2022.02.16.480662.">Junttila <em>et al.</em>, 2022</a>, <a class="reference internal" href="#id327" title="Jordan W. Squair, Matthieu Gautier, Claudia Kathe, Mark A. Anderson, Nicholas D. James, Thomas H. Hutson, Rémi Hudelle, Taha Qaiser, Kaya J. E. Matson, Quentin Barraud, Ariel J. Levine, Gioele La Manno, Michael A. Skinnider, and Grégoire Courtine. Confronting false discoveries in single-cell differential expression. Nature Communications, 12(1):5692, Sep 2021. URL: https://doi.org/10.1038/s41467-021-25960-2, doi:10.1038/s41467-021-25960-2.">Squair <em>et al.</em>, 2021</a>, <a class="reference internal" href="#id329" title="Kip D. Zimmerman, Mark A. Espeland, and Carl D. Langefeld. A practical solution to pseudoreplication bias in single-cell studies. Nature Communications, 12(1):738, Feb 2021. URL: https://doi.org/10.1038/s41467-021-21038-1, doi:10.1038/s41467-021-21038-1.">Zimmerman <em>et al.</em>, 2021</a>]</span>. Therefore, batch effect correction or the aggregation of cell-type-specific expression values within an individual through either a sum, mean or random effect per individual, that is pseudobulk generation, should be applied prior to DGE analysis to account for within-sample correlations <span id="id9">[<a class="reference internal" href="#id329" title="Kip D. Zimmerman, Mark A. Espeland, and Carl D. Langefeld. A practical solution to pseudoreplication bias in single-cell studies. Nature Communications, 12(1):738, Feb 2021. URL: https://doi.org/10.1038/s41467-021-21038-1, doi:10.1038/s41467-021-21038-1.">Zimmerman <em>et al.</em>, 2021</a>]</span>. Generally, both, pseudobulk methods with sum aggregation such as edgeR, DESeq2, or Limma<span id="id10">[<a class="reference internal" href="#id332" title="Matthew E. Ritchie, Belinda Phipson, Di Wu, Yifang Hu, Charity W. Law, Wei Shi, and Gordon K. Smyth. Limma powers differential expression analyses for term`rna`-sequencing and microarray studies. Nucleic acids research, 43(7):e47-e47, Apr 2015. gkv007[PII]. URL: https://doi.org/10.1093/nar/gkv007, doi:10.1093/nar/gkv007.">Ritchie <em>et al.</em>, 2015</a>]</span> and mixed models such as MAST with random effect setting were found to be superior compared to naive methods, such as the popular Wilcoxon rank-sum test or Seurat’s <span id="id11">[<a class="reference internal" href="#id339" title="Yuhan Hao, Stephanie Hao, Erica Andersen-Nissen, William M. Mauck III, Shiwei Zheng, Andrew Butler, Maddie J. Lee, Aaron J. Wilk, Charlotte Darby, Michael Zagar, Paul Hoffman, Marlon Stoeckius, Efthymia Papalexi, Eleni P. Mimitou, Jaison Jain, Avi Srivastava, Tim Stuart, Lamar B. Fleming, Bertrand Yeung, Angela J. Rogers, Juliana M. McElrath, Catherine A. Blish, Raphael Gottardo, Peter Smibert, and Rahul Satija. Integrated analysis of multimodal single-cell data. Cell, 2021. URL: https://doi.org/10.1016/j.cell.2021.04.048, doi:10.1016/j.cell.2021.04.048.">Hao <em>et al.</em>, 2021</a>]</span> latent models, which do not account for them<span id="id12">[<a class="reference internal" href="#id331" title="Sini Junttila, Johannes Smolander, and Laura L Elo. Benchmarking methods for detecting differential states between conditions from multi-subject single-cell term`rna`-seq data. bioRxiv, 2022. URL: https://www.biorxiv.org/content/early/2022/02/19/2022.02.16.480662, arXiv:https://www.biorxiv.org/content/early/2022/02/19/2022.02.16.480662.full.pdf, doi:10.1101/2022.02.16.480662.">Junttila <em>et al.</em>, 2022</a>]</span>.</p>
<p>In matters arising from the Zimmerman paper, Murphy et al. critically examined the Zimmerman benchmarking strategy and improved it <span id="id13">[<a class="reference internal" href="#id330" title="Alan E. Murphy and Nathan G. Skene. A balanced measure shows superior performance of pseudobulk methods in single-cell RNA-sequencing analysis. Nat Commun, dec 2022. URL: https://doi.org/10.1038%2Fs41467-022-35519-4, doi:10.1038/s41467-022-35519-4.">Murphy and Skene, 2022</a>]</span>. They came to the conclusion that pseudobulk methods perform best but whether sum or mean aggregation works better requires further investigation.</p>
<p>Hence, in this notebook, we demonstrate how to use two tools for DE analysis: edgeR with a quasi-likelihood test and MAST with random effects. Since both edgeR and MAST are implemented in R, we use the anndata2ri package to be able to simultaneously work with AnnData objects in Python and SingeCellExperiment objects in R.</p>
<p>For each of the two methods, we show two use cases: how to run the analysis on full data and perform testing on multiple cell types and on one specific cell type.</p>
</section>
<section id="environment-setup">
<h2><span class="section-number">16.2. </span>Environment setup<a class="headerlink" href="#environment-setup" title="Link to this heading">#</a></h2>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span><span class="w"> </span><span class="nn">warnings</span>

<span class="n">warnings</span><span class="o">.</span><span class="n">filterwarnings</span><span class="p">(</span><span class="s2">&quot;ignore&quot;</span><span class="p">)</span>

<span class="kn">import</span><span class="w"> </span><span class="nn">logging</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">random</span>

<span class="kn">import</span><span class="w"> </span><span class="nn">anndata2ri</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">matplotlib.pyplot</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">plt</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">numpy</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">np</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">pandas</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">pd</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">pertpy</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">rpy2.rinterface_lib.callbacks</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">sc_toolbox</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">scanpy</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">sc</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">seaborn</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">sns</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">rpy2.robjects</span><span class="w"> </span><span class="kn">import</span> <span class="n">pandas2ri</span>

<span class="n">sc</span><span class="o">.</span><span class="n">settings</span><span class="o">.</span><span class="n">verbosity</span> <span class="o">=</span> <span class="mi">0</span>
<span class="n">rpy2</span><span class="o">.</span><span class="n">rinterface_lib</span><span class="o">.</span><span class="n">callbacks</span><span class="o">.</span><span class="n">logger</span><span class="o">.</span><span class="n">setLevel</span><span class="p">(</span><span class="n">logging</span><span class="o">.</span><span class="n">ERROR</span><span class="p">)</span>

<span class="n">pandas2ri</span><span class="o">.</span><span class="n">activate</span><span class="p">()</span>
<span class="n">anndata2ri</span><span class="o">.</span><span class="n">activate</span><span class="p">()</span>

<span class="o">%</span><span class="k">load_ext</span> rpy2.ipython
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><pre style="white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace"><span style="color: #808000; text-decoration-color: #808000; font-weight: bold">To use sccoda or tasccoda please install ete3 with </span><span style="color: #008000; text-decoration-color: #008000; font-weight: bold">pip install ete3</span>
</pre>
</div></div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="o">%%</span><span class="k">R</span>
library(edgeR)
library(MAST)
</pre></div>
</div>
</div>
</div>
</section>
<section id="preparing-the-dataset">
<h2><span class="section-number">16.3. </span>Preparing the dataset<a class="headerlink" href="#preparing-the-dataset" title="Link to this heading">#</a></h2>
<p>We will use the Kang dataset, which is a 10x droplet-based scRNA-seq peripheral blood mononuclear cell (PBMC) data from 8 Lupus patients before and after 6h-treatment with INF-β (16 samples in total)<span id="id14">[<a class="reference internal" href="#id337" title="Hyun Min Kang, Meena Subramaniam, Sasha Targ, Michelle Nguyen, Lenka Maliskova, Elizabeth McCarthy, Eunice Wan, Simon Wong, Lauren Byrnes, Cristina M Lanata, and others. Multiplexed droplet single-cell term`rna`-sequencing using natural genetic variation. Nature biotechnology, 36(1):89–94, 2018.">Kang <em>et al.</em>, 2018</a>]</span>. Interferon beta is used in the form of natural fibroblast or recombinant preparations (interferon beta-1a and interferon beta-1b) and exerts antiviral and antiproliferative properties similar to those of interferon alpha. Interferon beta has been approved for the treatment of relapsing–remitting multiple sclerosis and secondary progressive multiple sclerosis.</p>
<p>First, we load the full dataset.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">adata</span> <span class="o">=</span> <span class="n">pertpy</span><span class="o">.</span><span class="n">data</span><span class="o">.</span><span class="n">kang_2018</span><span class="p">()</span>
<span class="n">adata</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>AnnData object with n_obs × n_vars = 24673 × 15706
    obs: &#39;nCount_RNA&#39;, &#39;nFeature_RNA&#39;, &#39;tsne1&#39;, &#39;tsne2&#39;, &#39;label&#39;, &#39;cluster&#39;, &#39;cell_type&#39;, &#39;replicate&#39;, &#39;nCount_SCT&#39;, &#39;nFeature_SCT&#39;, &#39;integrated_snn_res.0.4&#39;, &#39;seurat_clusters&#39;
    var: &#39;name&#39;
    obsm: &#39;X_pca&#39;, &#39;X_umap&#39;
</pre></div>
</div>
</div>
</div>
<p>We will need <code class="docutils literal notranslate"><span class="pre">label</span></code> (which contains the condition label), <code class="docutils literal notranslate"><span class="pre">replicate</span></code> and <code class="docutils literal notranslate"><span class="pre">cell_type</span></code> columns of the <code class="docutils literal notranslate"><span class="pre">.obs</span></code>.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[:</span><span class="mi">5</span><span class="p">]</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>nCount_RNA</th>
      <th>nFeature_RNA</th>
      <th>tsne1</th>
      <th>tsne2</th>
      <th>label</th>
      <th>cluster</th>
      <th>cell_type</th>
      <th>replicate</th>
      <th>nCount_SCT</th>
      <th>nFeature_SCT</th>
      <th>integrated_snn_res.0.4</th>
      <th>seurat_clusters</th>
    </tr>
    <tr>
      <th>index</th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>AAACATACATTTCC-1</th>
      <td>3017.0</td>
      <td>877</td>
      <td>-27.640373</td>
      <td>14.966629</td>
      <td>ctrl</td>
      <td>9</td>
      <td>CD14+ Monocytes</td>
      <td>patient_1016</td>
      <td>1704.0</td>
      <td>711</td>
      <td>1</td>
      <td>1</td>
    </tr>
    <tr>
      <th>AAACATACCAGAAA-1</th>
      <td>2481.0</td>
      <td>713</td>
      <td>-27.493646</td>
      <td>28.924885</td>
      <td>ctrl</td>
      <td>9</td>
      <td>CD14+ Monocytes</td>
      <td>patient_1256</td>
      <td>1614.0</td>
      <td>662</td>
      <td>1</td>
      <td>1</td>
    </tr>
    <tr>
      <th>AAACATACCATGCA-1</th>
      <td>703.0</td>
      <td>337</td>
      <td>-10.468194</td>
      <td>-5.984389</td>
      <td>ctrl</td>
      <td>3</td>
      <td>CD4 T cells</td>
      <td>patient_1488</td>
      <td>908.0</td>
      <td>337</td>
      <td>6</td>
      <td>6</td>
    </tr>
    <tr>
      <th>AAACATACCTCGCT-1</th>
      <td>3420.0</td>
      <td>850</td>
      <td>-24.367997</td>
      <td>20.429285</td>
      <td>ctrl</td>
      <td>9</td>
      <td>CD14+ Monocytes</td>
      <td>patient_1256</td>
      <td>1738.0</td>
      <td>653</td>
      <td>1</td>
      <td>1</td>
    </tr>
    <tr>
      <th>AAACATACCTGGTA-1</th>
      <td>3158.0</td>
      <td>1111</td>
      <td>27.952170</td>
      <td>24.159738</td>
      <td>ctrl</td>
      <td>4</td>
      <td>Dendritic cells</td>
      <td>patient_1039</td>
      <td>1857.0</td>
      <td>928</td>
      <td>12</td>
      <td>12</td>
    </tr>
  </tbody>
</table>
</div></div></div>
</div>
<p>We will need to work with raw counts so we check that <code class="docutils literal notranslate"><span class="pre">.X</span></code> indeed contains raw counts and put them into the <code class="docutils literal notranslate"><span class="pre">counts</span></code> layer of our AnnData object.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">np</span><span class="o">.</span><span class="n">max</span><span class="p">(</span><span class="n">adata</span><span class="o">.</span><span class="n">X</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>3828.0
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">adata</span><span class="o">.</span><span class="n">layers</span><span class="p">[</span><span class="s2">&quot;counts&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">adata</span><span class="o">.</span><span class="n">X</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
</pre></div>
</div>
</div>
</div>
<p>We have 8 control and 8 disease patients.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="nb">print</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">adata</span><span class="p">[</span><span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s2">&quot;label&quot;</span><span class="p">]</span> <span class="o">==</span> <span class="s2">&quot;ctrl&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s2">&quot;replicate&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">cat</span><span class="o">.</span><span class="n">categories</span><span class="p">))</span>
<span class="nb">print</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">adata</span><span class="p">[</span><span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s2">&quot;label&quot;</span><span class="p">]</span> <span class="o">==</span> <span class="s2">&quot;stim&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s2">&quot;replicate&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">cat</span><span class="o">.</span><span class="n">categories</span><span class="p">))</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>8
8
</pre></div>
</div>
</div>
</div>
<p>We filter cells which have less than 200 genes and genes which were found in less than 3 cells for a rudimentary quality control.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">sc</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">filter_cells</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">min_genes</span><span class="o">=</span><span class="mi">200</span><span class="p">)</span>
<span class="n">sc</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">filter_genes</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">min_cells</span><span class="o">=</span><span class="mi">3</span><span class="p">)</span>
<span class="n">adata</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>AnnData object with n_obs × n_vars = 24562 × 15701
    obs: &#39;nCount_RNA&#39;, &#39;nFeature_RNA&#39;, &#39;tsne1&#39;, &#39;tsne2&#39;, &#39;label&#39;, &#39;cluster&#39;, &#39;cell_type&#39;, &#39;replicate&#39;, &#39;nCount_SCT&#39;, &#39;nFeature_SCT&#39;, &#39;integrated_snn_res.0.4&#39;, &#39;seurat_clusters&#39;, &#39;n_genes&#39;
    var: &#39;name&#39;, &#39;n_cells&#39;
    obsm: &#39;X_pca&#39;, &#39;X_umap&#39;
    layers: &#39;counts&#39;
</pre></div>
</div>
</div>
</div>
</section>
<section id="pseudobulk">
<span id="conditions-differential-gene-expression-key-takeaway-2"></span><h2><span class="section-number">16.4. </span>Pseudobulk<a class="headerlink" href="#pseudobulk" title="Link to this heading">#</a></h2>
<p>edgeR is a differential gene expression testing tool implemented in R which was initially designed for bulk gene expression data. It implements a wide range of statistical methodologies based on the negative binomial distribution, empirical Bayes estimation, exact tests, generalized linear models (GLMs) and quasi-likelihood tests.</p>
<p>For more details please read the original publication<span id="id15">[<a class="reference internal" href="#id317" title="Mark D. Robinson, Davis J. McCarthy, and Gordon K. Smyth. Edger: a bioconductor package for differential expression analysis of digital gene expression data. Bioinformatics (Oxford, England), 26(1):139-140, Jan 2010. btp616[PII]. URL: https://doi.org/10.1093/bioinformatics/btp616, doi:10.1093/bioinformatics/btp616.">Robinson <em>et al.</em>, 2010</a>]</span>.</p>
<p>Here, we will be using the quasi-likelihood test because it accounts for the uncertainty of the dispersion estimates. In contrast, the exact test assumes that the estimated dispersion is the true value, which can result in some inaccuracy. Additionally, the quasi-likelihood GLMs are more flexible when it comes to experimental design.</p>
<p>Since edgeR was introduced as a method for DE analysis for bulk data, we first need to create pseudobulk samples from our single-cell dataset. For each patient we create 1 pseudobulk sample per cell type by aggregating the cell from each subpopulation and taking the mean gene expression within that subpopulation.</p>
<p>If the data you are working with does not have replicates, it could be beneficial to create multiple (e.g. 2-3) pseudobulks per patient to account for patient variability. Here we chose to create 1 pseudobulk per patient because for each patient we have 2 replicates, one control and one stimulated.</p>
<p>We strongly recommend to read this guide <a class="reference external" href="https://f1000research.com/articles/9-1444">https://f1000research.com/articles/9-1444</a> on design matrices.</p>
<p>Regardless of whether we want to run the analysis only on a few cell subpopulations and fit a model for each one of them separately or fit one model for all of them, we first need to prepare the data, define a function to create pseudobulks and run the edgeR pipeline. First, let’s prepare the data.</p>
<p>Since we need to create pseudobulks for each patient-condition combination, we first need to create such a column by concatenating <code class="docutils literal notranslate"><span class="pre">replicate</span></code> and <code class="docutils literal notranslate"><span class="pre">label</span></code>.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s2">&quot;sample&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="p">[</span>
    <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">rep</span><span class="si">}</span><span class="s2">_</span><span class="si">{</span><span class="n">l</span><span class="si">}</span><span class="s2">&quot;</span> <span class="k">for</span> <span class="n">rep</span><span class="p">,</span> <span class="n">l</span> <span class="ow">in</span> <span class="nb">zip</span><span class="p">(</span><span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s2">&quot;replicate&quot;</span><span class="p">],</span> <span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s2">&quot;label&quot;</span><span class="p">])</span>
<span class="p">]</span>
</pre></div>
</div>
</div>
</div>
<p>We need to clean up the cell type names, i.e. replace spaces with underscores and remove plus-signs, to avoid Python to R conversion issues.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s2">&quot;cell_type&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="p">[</span><span class="n">ct</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s2">&quot; &quot;</span><span class="p">,</span> <span class="s2">&quot;_&quot;</span><span class="p">)</span> <span class="k">for</span> <span class="n">ct</span> <span class="ow">in</span> <span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s2">&quot;cell_type&quot;</span><span class="p">]]</span>
<span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s2">&quot;cell_type&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="p">[</span><span class="n">ct</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s2">&quot;+&quot;</span><span class="p">,</span> <span class="s2">&quot;&quot;</span><span class="p">)</span> <span class="k">for</span> <span class="n">ct</span> <span class="ow">in</span> <span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s2">&quot;cell_type&quot;</span><span class="p">]]</span>
</pre></div>
</div>
</div>
</div>
<p>We need to set categorical metadata to be indeed categorical to create pseudobulks.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s2">&quot;replicate&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s2">&quot;replicate&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="s2">&quot;category&quot;</span><span class="p">)</span>
<span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s2">&quot;label&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s2">&quot;label&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="s2">&quot;category&quot;</span><span class="p">)</span>
<span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s2">&quot;sample&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s2">&quot;sample&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="s2">&quot;category&quot;</span><span class="p">)</span>
<span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s2">&quot;cell_type&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s2">&quot;cell_type&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="s2">&quot;category&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<p>Now, let’s define the function we need to aggregate single cells into pseudo-replicates:</p>
<ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">aggregate_and_filter</span></code> is a function that creates an AnnData object with one pseudo-replicate for each donor for a specified subpopulation from the original single-cell AnnData object. Here, we also filter out donors that have fewer than 30 cells for the specified population.</p></li>
<li><p>by changing the <code class="docutils literal notranslate"><span class="pre">replicates_per_patient</span></code> parameter, several (n) pseudo-replicates can be created for each sample; cells are then split into n subsets of roughly equal sizes.</p></li>
</ul>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">NUM_OF_CELL_PER_DONOR</span> <span class="o">=</span> <span class="mi">30</span>


<span class="k">def</span><span class="w"> </span><span class="nf">aggregate_and_filter</span><span class="p">(</span>
    <span class="n">adata</span><span class="p">,</span>
    <span class="n">cell_identity</span><span class="p">,</span>
    <span class="n">donor_key</span><span class="o">=</span><span class="s2">&quot;sample&quot;</span><span class="p">,</span>
    <span class="n">condition_key</span><span class="o">=</span><span class="s2">&quot;label&quot;</span><span class="p">,</span>
    <span class="n">cell_identity_key</span><span class="o">=</span><span class="s2">&quot;cell_type&quot;</span><span class="p">,</span>
    <span class="n">obs_to_keep</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>  <span class="c1"># which additional metadata to keep, e.g. gender, age, etc.</span>
    <span class="n">replicates_per_patient</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span>
<span class="p">):</span>
    <span class="c1"># subset adata to the given cell identity</span>
    <span class="k">if</span> <span class="n">obs_to_keep</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
        <span class="n">obs_to_keep</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="n">adata_cell_pop</span> <span class="o">=</span> <span class="n">adata</span><span class="p">[</span><span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="n">cell_identity_key</span><span class="p">]</span> <span class="o">==</span> <span class="n">cell_identity</span><span class="p">]</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
    <span class="c1"># check which donors to keep according to the number of cells specified with NUM_OF_CELL_PER_DONOR</span>
    <span class="n">size_by_donor</span> <span class="o">=</span> <span class="n">adata_cell_pop</span><span class="o">.</span><span class="n">obs</span><span class="o">.</span><span class="n">groupby</span><span class="p">([</span><span class="n">donor_key</span><span class="p">])</span><span class="o">.</span><span class="n">size</span><span class="p">()</span>
    <span class="n">donors_to_drop</span> <span class="o">=</span> <span class="p">[</span>
        <span class="n">donor</span>
        <span class="k">for</span> <span class="n">donor</span> <span class="ow">in</span> <span class="n">size_by_donor</span><span class="o">.</span><span class="n">index</span>
        <span class="k">if</span> <span class="n">size_by_donor</span><span class="p">[</span><span class="n">donor</span><span class="p">]</span> <span class="o">&lt;=</span> <span class="n">NUM_OF_CELL_PER_DONOR</span>
    <span class="p">]</span>
    <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">donors_to_drop</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">:</span>
        <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Dropping the following samples:&quot;</span><span class="p">)</span>
        <span class="nb">print</span><span class="p">(</span><span class="n">donors_to_drop</span><span class="p">)</span>
    <span class="n">df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">columns</span><span class="o">=</span><span class="p">[</span><span class="o">*</span><span class="n">adata_cell_pop</span><span class="o">.</span><span class="n">var_names</span><span class="p">,</span> <span class="o">*</span><span class="n">obs_to_keep</span><span class="p">])</span>

    <span class="n">adata_cell_pop</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="n">donor_key</span><span class="p">]</span> <span class="o">=</span> <span class="n">adata_cell_pop</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="n">donor_key</span><span class="p">]</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="s2">&quot;category&quot;</span><span class="p">)</span>
    <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">donor</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">donors</span> <span class="o">:=</span> <span class="n">adata_cell_pop</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="n">donor_key</span><span class="p">]</span><span class="o">.</span><span class="n">cat</span><span class="o">.</span><span class="n">categories</span><span class="p">):</span>
        <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">Processing donor </span><span class="si">{</span><span class="n">i</span><span class="o">+</span><span class="mi">1</span><span class="si">}</span><span class="s2"> out of </span><span class="si">{</span><span class="nb">len</span><span class="p">(</span><span class="n">donors</span><span class="p">)</span><span class="si">}</span><span class="s2">...&quot;</span><span class="p">,</span> <span class="n">end</span><span class="o">=</span><span class="s2">&quot;</span><span class="se">\r</span><span class="s2">&quot;</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">donor</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">donors_to_drop</span><span class="p">:</span>
            <span class="n">adata_donor</span> <span class="o">=</span> <span class="n">adata_cell_pop</span><span class="p">[</span><span class="n">adata_cell_pop</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="n">donor_key</span><span class="p">]</span> <span class="o">==</span> <span class="n">donor</span><span class="p">]</span>
            <span class="c1"># create replicates for each donor</span>
            <span class="n">indices</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">adata_donor</span><span class="o">.</span><span class="n">obs_names</span><span class="p">)</span>
            <span class="n">random</span><span class="o">.</span><span class="n">shuffle</span><span class="p">(</span><span class="n">indices</span><span class="p">)</span>
            <span class="n">indices</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array_split</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">indices</span><span class="p">),</span> <span class="n">replicates_per_patient</span><span class="p">)</span>
            <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">rep_idx</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">indices</span><span class="p">):</span>
                <span class="n">adata_replicate</span> <span class="o">=</span> <span class="n">adata_donor</span><span class="p">[</span><span class="n">rep_idx</span><span class="p">]</span>
                <span class="c1"># specify how to aggregate: sum gene expression for each gene for each donor and also keep the condition information</span>
                <span class="n">agg_dict</span> <span class="o">=</span> <span class="p">{</span><span class="n">gene</span><span class="p">:</span> <span class="s2">&quot;sum&quot;</span> <span class="k">for</span> <span class="n">gene</span> <span class="ow">in</span> <span class="n">adata_replicate</span><span class="o">.</span><span class="n">var_names</span><span class="p">}</span>
                <span class="k">for</span> <span class="n">obs</span> <span class="ow">in</span> <span class="n">obs_to_keep</span><span class="p">:</span>
                    <span class="n">agg_dict</span><span class="p">[</span><span class="n">obs</span><span class="p">]</span> <span class="o">=</span> <span class="s2">&quot;first&quot;</span>
                <span class="c1"># create a df with all genes, donor and condition info</span>
                <span class="n">df_donor</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">adata_replicate</span><span class="o">.</span><span class="n">X</span><span class="o">.</span><span class="n">A</span><span class="p">)</span>
                <span class="n">df_donor</span><span class="o">.</span><span class="n">index</span> <span class="o">=</span> <span class="n">adata_replicate</span><span class="o">.</span><span class="n">obs_names</span>
                <span class="n">df_donor</span><span class="o">.</span><span class="n">columns</span> <span class="o">=</span> <span class="n">adata_replicate</span><span class="o">.</span><span class="n">var_names</span>
                <span class="n">df_donor</span> <span class="o">=</span> <span class="n">df_donor</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">adata_replicate</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="n">obs_to_keep</span><span class="p">])</span>
                <span class="c1"># aggregate</span>
                <span class="n">df_donor</span> <span class="o">=</span> <span class="n">df_donor</span><span class="o">.</span><span class="n">groupby</span><span class="p">(</span><span class="n">donor_key</span><span class="p">)</span><span class="o">.</span><span class="n">agg</span><span class="p">(</span><span class="n">agg_dict</span><span class="p">)</span>
                <span class="n">df_donor</span><span class="p">[</span><span class="n">donor_key</span><span class="p">]</span> <span class="o">=</span> <span class="n">donor</span>
                <span class="n">df</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="sa">f</span><span class="s2">&quot;donor_</span><span class="si">{</span><span class="n">donor</span><span class="si">}</span><span class="s2">_</span><span class="si">{</span><span class="n">i</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">df_donor</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">donor</span><span class="p">]</span>
    <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;</span><span class="se">\n</span><span class="s2">&quot;</span><span class="p">)</span>
    <span class="c1"># create AnnData object from the df</span>
    <span class="n">adata_cell_pop</span> <span class="o">=</span> <span class="n">sc</span><span class="o">.</span><span class="n">AnnData</span><span class="p">(</span>
        <span class="n">df</span><span class="p">[</span><span class="n">adata_cell_pop</span><span class="o">.</span><span class="n">var_names</span><span class="p">],</span> <span class="n">obs</span><span class="o">=</span><span class="n">df</span><span class="o">.</span><span class="n">drop</span><span class="p">(</span><span class="n">columns</span><span class="o">=</span><span class="n">adata_cell_pop</span><span class="o">.</span><span class="n">var_names</span><span class="p">)</span>
    <span class="p">)</span>
    <span class="k">return</span> <span class="n">adata_cell_pop</span>
</pre></div>
</div>
</div>
</div>
<p>We also need to define a separate function to fit an edgeR GLM:</p>
<ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">fit_model</span></code> takes a SingleCellExperiment object as input, creates the design matrix and outputs the fitted GLM. We also output the edgeR object of class DGEList to do some exploratory data analysis (EDA).</p></li>
</ul>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="o">%%</span><span class="k">R</span>
fit_model &lt;- function(adata_){
    # create an edgeR object with counts and grouping factor
    y &lt;- DGEList(assay(adata_, &quot;X&quot;), group = colData(adata_)$label)
    # filter out genes with low counts
    print(&quot;Dimensions before subsetting:&quot;)
    print(dim(y))
    print(&quot;&quot;)
    keep &lt;- filterByExpr(y)
    y &lt;- y[keep, , keep.lib.sizes=FALSE]
    print(&quot;Dimensions after subsetting:&quot;)
    print(dim(y))
    print(&quot;&quot;)
    # normalize
    y &lt;- calcNormFactors(y)
    # create a vector that is a concatenation of condition and cell type that we will later use with contrasts
    group &lt;- paste0(colData(adata_)$label, &quot;.&quot;, colData(adata_)$cell_type)
    replicate &lt;- colData(adata_)$replicate
    # create a design matrix: here we have multiple donors so also consider that in the design matrix
    design &lt;- model.matrix(~ 0 + group + replicate)
    # estimate dispersion
    y &lt;- estimateDisp(y, design = design)
    # fit the model
    fit &lt;- glmQLFit(y, design)
    return(list(&quot;fit&quot;=fit, &quot;design&quot;=design, &quot;y&quot;=y))
}
</pre></div>
</div>
</div>
</div>
<p>Now we have defined all the functions we need, so we can proceed with creating pseudobulks. We might want to look at available metadata later and therefore keep it in the AnnData object.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">obs_to_keep</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;label&quot;</span><span class="p">,</span> <span class="s2">&quot;cell_type&quot;</span><span class="p">,</span> <span class="s2">&quot;replicate&quot;</span><span class="p">,</span> <span class="s2">&quot;sample&quot;</span><span class="p">]</span>
</pre></div>
</div>
</div>
</div>
<p>We need to pass the raw counts to edgeR. Hence, we set <code class="docutils literal notranslate"><span class="pre">.X</span></code> to the <code class="docutils literal notranslate"><span class="pre">counts</span></code> layer to ensure the pseudo-replicates are created for raw counts.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">adata</span><span class="o">.</span><span class="n">X</span> <span class="o">=</span> <span class="n">adata</span><span class="o">.</span><span class="n">layers</span><span class="p">[</span><span class="s2">&quot;counts&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
</pre></div>
</div>
</div>
</div>
<p>Next, we create the AnnData object with pseudobulks.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># process first cell type separately...</span>
<span class="n">cell_type</span> <span class="o">=</span> <span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s2">&quot;cell_type&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">cat</span><span class="o">.</span><span class="n">categories</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
<span class="nb">print</span><span class="p">(</span>
    <span class="sa">f</span><span class="s1">&#39;Processing </span><span class="si">{</span><span class="n">cell_type</span><span class="si">}</span><span class="s1"> (1 out of </span><span class="si">{</span><span class="nb">len</span><span class="p">(</span><span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s2">&quot;cell_type&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">cat</span><span class="o">.</span><span class="n">categories</span><span class="p">)</span><span class="si">}</span><span class="s1">)...&#39;</span>
<span class="p">)</span>
<span class="n">adata_pb</span> <span class="o">=</span> <span class="n">aggregate_and_filter</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">cell_type</span><span class="p">,</span> <span class="n">obs_to_keep</span><span class="o">=</span><span class="n">obs_to_keep</span><span class="p">)</span>
<span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">cell_type</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s2">&quot;cell_type&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">cat</span><span class="o">.</span><span class="n">categories</span><span class="p">[</span><span class="mi">1</span><span class="p">:]):</span>
    <span class="nb">print</span><span class="p">(</span>
        <span class="sa">f</span><span class="s1">&#39;Processing </span><span class="si">{</span><span class="n">cell_type</span><span class="si">}</span><span class="s1"> (</span><span class="si">{</span><span class="n">i</span><span class="o">+</span><span class="mi">2</span><span class="si">}</span><span class="s1"> out of </span><span class="si">{</span><span class="nb">len</span><span class="p">(</span><span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s2">&quot;cell_type&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">cat</span><span class="o">.</span><span class="n">categories</span><span class="p">)</span><span class="si">}</span><span class="s1">)...&#39;</span>
    <span class="p">)</span>
    <span class="n">adata_cell_type</span> <span class="o">=</span> <span class="n">aggregate_and_filter</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">cell_type</span><span class="p">,</span> <span class="n">obs_to_keep</span><span class="o">=</span><span class="n">obs_to_keep</span><span class="p">)</span>
    <span class="n">adata_pb</span> <span class="o">=</span> <span class="n">adata_pb</span><span class="o">.</span><span class="n">concatenate</span><span class="p">(</span><span class="n">adata_cell_type</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Processing B_cells (1 out of 8)...
Dropping the following samples:
[&#39;patient_1039_ctrl&#39;]
	Processing donor 16 out of 16...

Processing CD14_Monocytes (2 out of 8)...
	Processing donor 16 out of 16...

Processing CD4_T_cells (3 out of 8)...
	Processing donor 16 out of 16...

Processing CD8_T_cells (4 out of 8)...
Dropping the following samples:
[&#39;patient_101_ctrl&#39;, &#39;patient_1039_ctrl&#39;, &#39;patient_1039_stim&#39;, &#39;patient_107_ctrl&#39;, &#39;patient_107_stim&#39;, &#39;patient_1244_ctrl&#39;, &#39;patient_1244_stim&#39;]
	Processing donor 16 out of 16...

Processing Dendritic_cells (5 out of 8)...
Dropping the following samples:
[&#39;patient_1016_ctrl&#39;, &#39;patient_1016_stim&#39;, &#39;patient_101_ctrl&#39;, &#39;patient_1039_ctrl&#39;, &#39;patient_1039_stim&#39;, &#39;patient_107_ctrl&#39;, &#39;patient_107_stim&#39;]
	Processing donor 16 out of 16...

Processing FCGR3A_Monocytes (6 out of 8)...
Dropping the following samples:
[&#39;patient_1039_ctrl&#39;, &#39;patient_1039_stim&#39;, &#39;patient_107_ctrl&#39;, &#39;patient_107_stim&#39;, &#39;patient_1244_stim&#39;]
	Processing donor 16 out of 16...

Processing Megakaryocytes (7 out of 8)...
Dropping the following samples:
[&#39;patient_1015_ctrl&#39;, &#39;patient_1015_stim&#39;, &#39;patient_1016_ctrl&#39;, &#39;patient_101_ctrl&#39;, &#39;patient_1039_stim&#39;, &#39;patient_107_stim&#39;, &#39;patient_1244_ctrl&#39;, &#39;patient_1244_stim&#39;, &#39;patient_1256_ctrl&#39;, &#39;patient_1256_stim&#39;, &#39;patient_1488_ctrl&#39;]
	Processing donor 11 out of 11...

Processing NK_cells (8 out of 8)...
Dropping the following samples:
[&#39;patient_1039_ctrl&#39;, &#39;patient_1039_stim&#39;]
	Processing donor 16 out of 16...
</pre></div>
</div>
</div>
</div>
<p>The validity of differential gene expression results highly depends on the capture of the major axis of variations in the statistical model. Intermediate data exploration steps such as principal component analysis (PCA) or multidimensional scaling (MDS) on pseudobulk samples allow for the identification of the sources of variation and thus can guide the construction of corresponding design and contrast matrices that model the data<span id="id16">[<a class="reference internal" href="#id334" title="Charity W. Law, Kathleen Zeglinski, Xueyi Dong, Monther Alhamdoosh, Gordon K. Smyth, and Matthew E. Ritchie. A guide to creating design matrices for gene expression experiments. F1000Research, 9:1444-1444, Dec 2020. 33604029[pmid]. URL: https://pubmed.ncbi.nlm.nih.gov/33604029.">Law <em>et al.</em>, 2020</a>]</span>.</p>
<p>Failing to account for multiple sources of biological variability for experiments which include biological replicates will inflate the FDR<span id="id17">[<a class="reference internal" href="#id333" title="Andrew L Thurman, Jason A Ratcliff, Michael S Chimenti, and Alejandro A Pezzulo. Differential gene expression analysis for multi-subject single cell term`RNA` sequencing studies with aggregateBioVar. Bioinformatics, 37(19):3243–3251, May 2021.">Thurman <em>et al.</em>, 2021</a>]</span>,<span id="id18">[<a class="reference internal" href="#id335" title="David Lähnemann, Johannes Köster, Ewa Szczurek, Davis J. McCarthy, Stephanie C. Hicks, Mark D. Robinson, Catalina A. Vallejos, Kieran R. Campbell, Niko Beerenwinkel, Ahmed Mahfouz, Luca Pinello, Pavel Skums, Alexandros Stamatakis, Camille Stephan-Otto Attolini, Samuel Aparicio, Jasmijn Baaijens, Marleen Balvert, Buys de Barbanson, Antonio Cappuccio, Giacomo Corleone, Bas E. Dutilh, Maria Florescu, Victor Guryev, Rens Holmer, Katharina Jahn, Thamar Jessurun Lobo, Emma M. Keizer, Indu Khatri, Szymon M. Kielbasa, Jan O. Korbel, Alexey M. Kozlov, Tzu-Hao Kuo, Boudewijn P.F. Lelieveldt, Ion I. Mandoiu, John C. Marioni, Tobias Marschall, Felix Mölder, Amir Niknejad, Lukasz Raczkowski, Marcel Reinders, Jeroen de Ridder, Antoine-Emmanuel Saliba, Antonios Somarakis, Oliver Stegle, Fabian J. Theis, Huan Yang, Alex Zelikovsky, Alice C. McHardy, Benjamin J. Raphael, Sohrab P. Shah, and Alexander Schönhuth. Eleven grand challenges in single-cell data science. Genome Biology, 21(1):31, Feb 2020. URL: https://doi.org/10.1186/s13059-020-1926-6, doi:10.1186/s13059-020-1926-6.">Lähnemann <em>et al.</em>, 2020</a>]</span>. While increasing the number of cells per individual increases the precision, it has a limited effect on the power for the detection of differences across individuals. Therefore, the best way to increase statistical power is to increase the number of independent experimental samples<span id="id19">[<a class="reference internal" href="#id329" title="Kip D. Zimmerman, Mark A. Espeland, and Carl D. Langefeld. A practical solution to pseudoreplication bias in single-cell studies. Nature Communications, 12(1):738, Feb 2021. URL: https://doi.org/10.1038/s41467-021-21038-1, doi:10.1038/s41467-021-21038-1.">Zimmerman <em>et al.</em>, 2021</a>]</span>.</p>
<p>Since our data has already been generated, we cannot further increase the number of independent experimental samples. Nevertheless, we will now explore our data to determine the major axes of variation to properly generate our design matrices.</p>
<p>We perform very basic EDA on the created pseudo-replicates to check if some patients/pseudobulks are outliers that we need to exclude so as not to bias the DE results. We save the raw counts in the <code class="docutils literal notranslate"><span class="pre">'counts'</span></code> layer, then normalize the counts and calculate the PCA coordinates for the normalized pseudobulk counts.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">adata_pb</span><span class="o">.</span><span class="n">layers</span><span class="p">[</span><span class="s1">&#39;counts&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">adata_pb</span><span class="o">.</span><span class="n">X</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">sc</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">normalize_total</span><span class="p">(</span><span class="n">adata_pb</span><span class="p">,</span> <span class="n">target_sum</span><span class="o">=</span><span class="mf">1e6</span><span class="p">)</span>
<span class="n">sc</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">log1p</span><span class="p">(</span><span class="n">adata_pb</span><span class="p">)</span>
<span class="n">sc</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">pca</span><span class="p">(</span><span class="n">adata_pb</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<p>Next, we look at created pseudo-replicates on a PCA plot and color by all the available metadata to see if there are any confounding factors that we might want to include in the design matrix. We also add a <code class="docutils literal notranslate"><span class="pre">lib_size</span></code> and <code class="docutils literal notranslate"><span class="pre">log_lib_size</span></code> columns to check if there is a correlation between library size and PC components.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">adata_pb</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s2">&quot;lib_size&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">adata_pb</span><span class="o">.</span><span class="n">layers</span><span class="p">[</span><span class="s2">&quot;counts&quot;</span><span class="p">],</span> <span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
<span class="n">adata_pb</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s2">&quot;log_lib_size&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">log</span><span class="p">(</span><span class="n">adata_pb</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s2">&quot;lib_size&quot;</span><span class="p">])</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">sc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">pca</span><span class="p">(</span><span class="n">adata_pb</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="n">adata_pb</span><span class="o">.</span><span class="n">obs</span><span class="p">,</span> <span class="n">ncols</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span> <span class="n">size</span><span class="o">=</span><span class="mi">300</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="../_images/ed653d6a112e7b0bad511c363c79275dd271a560b95b3487a285f7907ad3a2a6.png" src="../_images/ed653d6a112e7b0bad511c363c79275dd271a560b95b3487a285f7907ad3a2a6.png" />
</div>
</div>
<p>We observe separation of cell types on the PCA plots as well as the separation into stimulated and unstimulated cells. Other covariates (batch) do not seem to be clearly correlated with the PCA components so we do not include any of them in our design matrix.</p>
<p>As mentioned above, edgeR takes raw counts as input, so we put counts back into the <code class="docutils literal notranslate"><span class="pre">.X</span></code> field before we proceed.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">adata_pb</span><span class="o">.</span><span class="n">X</span> <span class="o">=</span> <span class="n">adata_pb</span><span class="o">.</span><span class="n">layers</span><span class="p">[</span><span class="s1">&#39;counts&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
</pre></div>
</div>
</div>
</div>
<section id="one-group">
<h3><span class="section-number">16.4.1. </span>One group<a class="headerlink" href="#one-group" title="Link to this heading">#</a></h3>
<p>First, we show how to prepare the data, construct the design matrix and perform the DE testing for one specific cell type.</p>
<p>We run the pipeline on CD14+ Monocytes subset of the data, as it was shown in the paper that the highest number of DE genes was identified in this subpopulation.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">adata_mono</span> <span class="o">=</span> <span class="n">adata_pb</span><span class="p">[</span><span class="n">adata_pb</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s2">&quot;cell_type&quot;</span><span class="p">]</span> <span class="o">==</span> <span class="s2">&quot;CD14_Monocytes&quot;</span><span class="p">]</span>
<span class="n">adata_mono</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>View of AnnData object with n_obs × n_vars = 16 × 15701
    obs: &#39;label&#39;, &#39;cell_type&#39;, &#39;replicate&#39;, &#39;sample&#39;, &#39;batch&#39;, &#39;lib_size&#39;, &#39;log_lib_size&#39;
    uns: &#39;log1p&#39;, &#39;pca&#39;, &#39;label_colors&#39;, &#39;cell_type_colors&#39;, &#39;replicate_colors&#39;, &#39;sample_colors&#39;, &#39;batch_colors&#39;
    obsm: &#39;X_pca&#39;
    varm: &#39;PCs&#39;
    layers: &#39;counts&#39;
</pre></div>
</div>
</div>
</div>
<p>Clean the sample names to make plots less crowded.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">adata_mono</span><span class="o">.</span><span class="n">obs_names</span> <span class="o">=</span> <span class="p">[</span>
    <span class="n">name</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;_&quot;</span><span class="p">)[</span><span class="mi">2</span><span class="p">]</span> <span class="o">+</span> <span class="s2">&quot;_&quot;</span> <span class="o">+</span> <span class="n">name</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;_&quot;</span><span class="p">)[</span><span class="mi">3</span><span class="p">]</span> <span class="k">for</span> <span class="n">name</span> <span class="ow">in</span> <span class="n">adata_mono</span><span class="o">.</span><span class="n">obs_names</span>
<span class="p">]</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="o">%%time</span>
<span class="o">%%</span><span class="n">R</span> <span class="o">-</span><span class="n">i</span> <span class="n">adata_mono</span>
<span class="n">outs</span> <span class="o">&lt;-</span><span class="n">fit_model</span><span class="p">(</span><span class="n">adata_mono</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>[1] &quot;Dimensions before subsetting:&quot;
[1] 15701    16
[1] &quot;&quot;
[1] &quot;Dimensions after subsetting:&quot;
[1] 3709   16
[1] &quot;&quot;
CPU times: user 2.95 s, sys: 224 ms, total: 3.18 s
Wall time: 4.27 s
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="o">%%</span><span class="k">R</span>
fit &lt;- outs$fit
y &lt;- outs$y
</pre></div>
</div>
</div>
</div>
<p>Since we did not enter our analysis with a prior assumption that a specific gene will be up or downregulated, we need visualizations to make sense of the DGE results. MDS plots allow for a high level overview. Commonly, we expect a separation between samples from different conditions as can be seen in the following plot for our results.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="o">%%</span><span class="k">R</span>
plotMDS(y, col=ifelse(y$samples$group == &quot;stim&quot;, &quot;red&quot;, &quot;blue&quot;))
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Fontconfig warning: ignoring UTF-8: not a valid region tag
</pre></div>
</div>
<img alt="../_images/d776dd82b691a480b78991126ea38648ef0d15ff4a92efb332268df711c76d6e.png" src="../_images/d776dd82b691a480b78991126ea38648ef0d15ff4a92efb332268df711c76d6e.png" />
</div>
</div>
<p>Biological Coefficient of Variation (BCV) plots show the variability of every gene on average across biological groups as a function of mean expression. For example, bcv of 0.3 indicates that there is on average 30% percent variability in the expression of genes across groups.</p>
<p>Genes with low abundance generally exhibit larger BCV as read count measurements are more uncertain for low abundance genes. On the other hand, genes with high average expression are quantified more reliably, thus they generally have lower variability and hence lower BCV. Use this plot to detect outlier genes or pinpoint other experimental factors that may need to be reflected in the design matrix. For example, a group of genes with high average expression and high BCV appearing in the top right corner of the plot can flag experimental stress, contamination etc, particularly if they belong to similar gene families.</p>
<p>BCV is the square root of dispersion. The distance between tagwise and common bcv trend indicates if tagwise (gene-wise) dispersion estimates are highly variable (i.e. heterogenous gene expression). If tag-wise dispersion values are very heterogeneous, less moderation is applied to capture heterogeneity. The distance between any two points on those curves reflects how much the dispersion for a gene was shrunken towards the common trend. That is, it captures the amount of dispersion moderation applied.</p>
<p>In the BCV plot below, we see some low abundance genes with high bcv, but no high abundance genes with high bcv, which indicates that there should not be any further experimental considerations modelled into the design matrix.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="o">%%</span><span class="k">R</span>
plotBCV(y)
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="../_images/c7e11a09e30b1110e12103cf487e53d3baab2c61760493e750ff315a6008246a.png" src="../_images/c7e11a09e30b1110e12103cf487e53d3baab2c61760493e750ff315a6008246a.png" />
</div>
</div>
<p>Next, we perform the quasi-likelihood test to find DE genes between control and stimulated conditions. Let’s check what columns our design matrix has to specify the correct columns to test on.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="o">%%</span><span class="k">R</span>
colnames(y$design)
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>[1] &quot;groupctrl.CD14_Monocytes&quot; &quot;groupstim.CD14_Monocytes&quot;
[3] &quot;replicatepatient_107&quot;     &quot;replicatepatient_1015&quot;   
[5] &quot;replicatepatient_1016&quot;    &quot;replicatepatient_1039&quot;   
[7] &quot;replicatepatient_1244&quot;    &quot;replicatepatient_1256&quot;   
[9] &quot;replicatepatient_1488&quot;   
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="o">%%</span><span class="k">R</span> -o tt
myContrast &lt;- makeContrasts(&#39;groupstim.CD14_Monocytes-groupctrl.CD14_Monocytes&#39;, levels = y$design)
qlf &lt;- glmQLFTest(fit, contrast=myContrast)
# get all of the DE genes and calculate Benjamini-Hochberg adjusted FDR
tt &lt;- topTags(qlf, n = Inf)
tt &lt;- tt$table
</pre></div>
</div>
</div>
</div>
<p>Let’s inspect the table: for each of the genes that were not filtered out by edgeR (3709 in our case), the table contains the results of the DE testing. The table can be saved as a <code class="docutils literal notranslate"><span class="pre">.csv</span></code> file to work with later and used for visualization. We will show how to use volcano plots at the end of the section after we run the analysis on full data.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">tt</span><span class="o">.</span><span class="n">shape</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>(3709, 5)
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">tt</span><span class="p">[:</span><span class="mi">5</span><span class="p">]</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>logFC</th>
      <th>logCPM</th>
      <th>F</th>
      <th>PValue</th>
      <th>FDR</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>HESX1</th>
      <td>8.345536</td>
      <td>6.773420</td>
      <td>1281.013295</td>
      <td>1.837373e-15</td>
      <td>2.766927e-12</td>
    </tr>
    <tr>
      <th>CD38</th>
      <td>7.126846</td>
      <td>7.420668</td>
      <td>1243.793133</td>
      <td>2.266164e-15</td>
      <td>2.766927e-12</td>
    </tr>
    <tr>
      <th>NT5C3A</th>
      <td>5.657050</td>
      <td>8.327003</td>
      <td>1218.102628</td>
      <td>2.628780e-15</td>
      <td>2.766927e-12</td>
    </tr>
    <tr>
      <th>SOCS1</th>
      <td>4.388247</td>
      <td>6.943768</td>
      <td>1191.289806</td>
      <td>3.079524e-15</td>
      <td>2.766927e-12</td>
    </tr>
    <tr>
      <th>GMPR</th>
      <td>6.943484</td>
      <td>7.031832</td>
      <td>1159.601183</td>
      <td>3.730018e-15</td>
      <td>2.766927e-12</td>
    </tr>
  </tbody>
</table>
</div></div></div>
</div>
<p>As a side note, one can also test for genes that are differentially expressed for a provided coefficient or contrast relative to a specified fold-change using <code class="docutils literal notranslate"><span class="pre">glmTreat</span></code>.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="o">%%</span><span class="k">R</span>
tr &lt;- glmTreat(fit, contrast=myContrast, lfc=1.5)
print(head(topTags(tr)))
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Coefficient:  -1*groupctrl.CD14_Monocytes 1*groupstim.CD14_Monocytes 
          logFC unshrunk.logFC    logCPM       PValue          FDR
HESX1  8.345536       8.702486  6.773420 2.108737e-14 3.535749e-11
CD38   7.126846       7.219184  7.420668 2.292459e-14 3.535749e-11
NT5C3A 5.657050       5.674640  8.327003 3.149544e-14 3.535749e-11
IL1RN  6.588583       6.596787 10.358946 4.249159e-14 3.535749e-11
GMPR   6.943484       7.047504  7.031832 4.766446e-14 3.535749e-11
DEFB1  6.654201       6.696759  8.067159 7.670147e-14 4.741429e-11
</pre></div>
</div>
</div>
</div>
<p>And finally, we can see how many genes we have with FDR-corrected values of less than 0.01.</p>
<p>The smear plot (also known as MA, M-values vs A-values, plot) shows the log fold-change of the genes as a function of their mean abundance. We generally observe higher logFC at low abundance ranges as read counts are more variable at low abundance resulting in large logFC estimates. If we fit a loess curve to logFC and Average logCPM values, the trend should center around zero. Any deviations from this can indicate that data has not been properly normalised. Genes with large mean expression and large logFC in absolute values can flag biologically interesting genes for investigation and follow-up.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="o">%%</span><span class="k">R</span>
plotSmear(qlf, de.tags = rownames(tt)[which(tt$FDR&lt;0.01)])
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="../_images/3827ecb1083670ac4d3084906b6f49eace169259f0a28316a7a58f81ea741209.png" src="../_images/3827ecb1083670ac4d3084906b6f49eace169259f0a28316a7a58f81ea741209.png" />
</div>
</div>
</section>
<section id="multiple-groups">
<h3><span class="section-number">16.4.2. </span>Multiple groups<a class="headerlink" href="#multiple-groups" title="Link to this heading">#</a></h3>
<p>Next, we show how to prepare the data, construct the design matrix and perform the DE testing using contrasts on the full dataset for all available cell types.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="o">%%time</span>
<span class="o">%%</span><span class="n">R</span> <span class="o">-</span><span class="n">i</span> <span class="n">adata_pb</span>
<span class="n">outs</span> <span class="o">&lt;-</span><span class="n">fit_model</span><span class="p">(</span><span class="n">adata_pb</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>[1] &quot;Dimensions before subsetting:&quot;
[1] 15701    90
[1] &quot;&quot;
[1] &quot;Dimensions after subsetting:&quot;
[1] 2358   90
[1] &quot;&quot;
CPU times: user 11.9 s, sys: 39.7 ms, total: 12 s
Wall time: 12 s
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="o">%%</span><span class="k">R</span>
fit &lt;- outs$fit
y &lt;- outs$y
</pre></div>
</div>
</div>
</div>
<p>Now we use contrasts to perform a quasi-likelihood test for each of our cell types. Because there is no straightforward way to move tables from R to Python from within an R loop we get the results manually for each cell type.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="o">%%</span><span class="k">R</span> -i adata_pb -o de_per_cell_type
de_per_cell_type &lt;- list()
for (cell_type in unique(colData(adata_pb)$cell_type)) {
    print(cell_type)
    # create contrast for this cell type
    myContrast &lt;- makeContrasts(paste0(&quot;groupstim.&quot;, cell_type, &quot;-groupctrl.&quot;, cell_type), levels = y$design)
    # perform QLF test
    qlf &lt;- glmQLFTest(fit, contrast=myContrast)
    # get all of the DE genes and calculate Benjamini-Hochberg adjusted FDR
    tt &lt;- topTags(qlf, n = Inf)
    # save in the list with the results for all the cell types
    de_per_cell_type[[cell_type]] &lt;- tt$table
}
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>[1] &quot;B_cells&quot;
[1] &quot;CD14_Monocytes&quot;
[1] &quot;CD4_T_cells&quot;
[1] &quot;CD8_T_cells&quot;
[1] &quot;Dendritic_cells&quot;
[1] &quot;FCGR3A_Monocytes&quot;
[1] &quot;NK_cells&quot;
</pre></div>
</div>
</div>
</div>
<p>Now let’s and save it in <code class="docutils literal notranslate"><span class="pre">.uns</span></code> of our original <code class="docutils literal notranslate"><span class="pre">adata</span></code> object using <code class="docutils literal notranslate"><span class="pre">sc_toolbox.tools.de_res_to_anndata</span></code> from <code class="docutils literal notranslate"><span class="pre">sc_toolbox</span></code> package (<a class="github reference external" href="https://github.com/schillerlab/sc-toolbox">schillerlab/sc-toolbox</a>), which saves the results as if they were created with the scanpy’s <code class="docutils literal notranslate"><span class="pre">rank_genes_groups()</span></code> function. The advantage of storing the DE tables like this is that we could now also use standard scanpy plotting functions. We save the DEG tables as <code class="docutils literal notranslate"><span class="pre">.csv</span></code> for each cell type as we will need them later in the cell-to-cell communication chapter.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># get cell types that we ran the analysis for</span>
<span class="n">cell_types</span> <span class="o">=</span> <span class="n">de_per_cell_type</span><span class="o">.</span><span class="n">keys</span><span class="p">()</span>
<span class="c1"># add the table to .uns for each cell type</span>
<span class="k">for</span> <span class="n">cell_type</span> <span class="ow">in</span> <span class="n">cell_types</span><span class="p">:</span>
    <span class="n">df</span> <span class="o">=</span> <span class="n">de_per_cell_type</span><span class="p">[</span><span class="n">cell_type</span><span class="p">]</span>
    <span class="n">df</span><span class="p">[</span><span class="s2">&quot;gene_symbol&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">df</span><span class="o">.</span><span class="n">index</span>
    <span class="n">df</span><span class="p">[</span><span class="s2">&quot;cell_type&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">cell_type</span>
    <span class="n">sc_toolbox</span><span class="o">.</span><span class="n">tools</span><span class="o">.</span><span class="n">de_res_to_anndata</span><span class="p">(</span>
        <span class="n">adata</span><span class="p">,</span>
        <span class="n">df</span><span class="p">,</span>
        <span class="n">groupby</span><span class="o">=</span><span class="s2">&quot;cell_type&quot;</span><span class="p">,</span>
        <span class="n">score_col</span><span class="o">=</span><span class="s2">&quot;logCPM&quot;</span><span class="p">,</span>
        <span class="n">pval_col</span><span class="o">=</span><span class="s2">&quot;PValue&quot;</span><span class="p">,</span>
        <span class="n">pval_adj_col</span><span class="o">=</span><span class="s2">&quot;FDR&quot;</span><span class="p">,</span>
        <span class="n">lfc_col</span><span class="o">=</span><span class="s2">&quot;logFC&quot;</span><span class="p">,</span>
        <span class="n">key_added</span><span class="o">=</span><span class="s2">&quot;edgeR_&quot;</span> <span class="o">+</span> <span class="n">cell_type</span><span class="p">,</span>
    <span class="p">)</span>
    <span class="n">df</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;de_edgeR_</span><span class="si">{</span><span class="n">cell_type</span><span class="si">}</span><span class="s2">.csv&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<p>To get the table as a <code class="docutils literal notranslate"><span class="pre">pandas</span></code> DataFrame again, we use <code class="docutils literal notranslate"><span class="pre">sc.get.rank_genes_groups_df</span></code> function.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">sc</span><span class="o">.</span><span class="n">get</span><span class="o">.</span><span class="n">rank_genes_groups_df</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">group</span><span class="o">=</span><span class="s2">&quot;CD14_Monocytes&quot;</span><span class="p">,</span> <span class="n">key</span><span class="o">=</span><span class="s2">&quot;edgeR_CD14_Monocytes&quot;</span><span class="p">)[</span>
    <span class="p">:</span><span class="mi">5</span>
<span class="p">]</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>names</th>
      <th>scores</th>
      <th>logfoldchanges</th>
      <th>pvals</th>
      <th>pvals_adj</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>FTH1</td>
      <td>16.186098</td>
      <td>-0.572272</td>
      <td>0.001589</td>
      <td>0.002712</td>
    </tr>
    <tr>
      <th>1</th>
      <td>MALAT1</td>
      <td>16.025682</td>
      <td>0.010049</td>
      <td>0.877078</td>
      <td>0.897245</td>
    </tr>
    <tr>
      <th>2</th>
      <td>B2M</td>
      <td>15.468524</td>
      <td>0.677266</td>
      <td>0.0</td>
      <td>0.0</td>
    </tr>
    <tr>
      <th>3</th>
      <td>TMSB4X</td>
      <td>15.054697</td>
      <td>-0.217131</td>
      <td>0.004663</td>
      <td>0.007331</td>
    </tr>
    <tr>
      <th>4</th>
      <td>FTL</td>
      <td>14.984937</td>
      <td>-0.041832</td>
      <td>0.669951</td>
      <td>0.710956</td>
    </tr>
  </tbody>
</table>
</div></div></div>
</div>
</section>
<section id="notes-on-edger">
<h3><span class="section-number">16.4.3. </span>Notes on edgeR<a class="headerlink" href="#notes-on-edger" title="Link to this heading">#</a></h3>
<ul class="simple">
<li><p>Requires raw counts as input</p></li>
<li><p>Requires pseudobulks from a single-cell experiment</p></li>
<li><p>If there are several donors in the single-cell experiment and the user wants to account for the patient variability, we recommend creating 2 or 3 pseudo-replicates for each patient and including patient information into the design matrix</p></li>
</ul>
</section>
</section>
<section id="single-cell-specific">
<h2><span class="section-number">16.5. </span>Single-cell specific<a class="headerlink" href="#single-cell-specific" title="Link to this heading">#</a></h2>
<p>The MAST framework models single-cell gene expression using a two-part generalized linear model. One component of MAST models the discrete expression rate of each gene across cells, while the other component models the conditional continuous expression level (conditional on the gene being expressed).
For more details please read the original publication<span id="id20">[<a class="reference internal" href="#id322" title="Greg Finak, Andrew McDavid, Masanao Yajima, Jingyuan Deng, Vivian Gersuk, Alex K. Shalek, Chloe K. Slichter, Hannah W. Miller, M. Juliana McElrath, Martin Prlic, Peter S. Linsley, and Raphael Gottardo. Mast: a flexible statistical framework for assessing transcriptional changes and characterizing heterogeneity in single-cell term`rna` sequencing data. Genome Biology, 16(1):278, Dec 2015. URL: https://doi.org/10.1186/s13059-015-0844-5, doi:10.1186/s13059-015-0844-5.">Finak <em>et al.</em>, 2015</a>]</span>.</p>
<p>MAST takes normalized counts as input, so we first take the ‘counts’ layer and then perform the normalization step.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">adata</span><span class="o">.</span><span class="n">X</span> <span class="o">=</span> <span class="n">adata</span><span class="o">.</span><span class="n">layers</span><span class="p">[</span><span class="s2">&quot;counts&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">sc</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">normalize_total</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">target_sum</span><span class="o">=</span><span class="mf">1e6</span><span class="p">)</span>
<span class="n">sc</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">log1p</span><span class="p">(</span><span class="n">adata</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<p>We define a small helper function that takes care of some object type conversion issue between R and Python. We also need to filter out genes that are expressed in a small number of cells (3 in this case) for each subpopulation as the model needs to be able to estimate the variance for each gene.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="k">def</span><span class="w"> </span><span class="nf">prep_anndata</span><span class="p">(</span><span class="n">adata_</span><span class="p">):</span>
    <span class="k">def</span><span class="w"> </span><span class="nf">fix_dtypes</span><span class="p">(</span><span class="n">adata_</span><span class="p">):</span>
        <span class="n">df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">adata_</span><span class="o">.</span><span class="n">X</span><span class="o">.</span><span class="n">A</span><span class="p">,</span> <span class="n">index</span><span class="o">=</span><span class="n">adata_</span><span class="o">.</span><span class="n">obs_names</span><span class="p">,</span> <span class="n">columns</span><span class="o">=</span><span class="n">adata_</span><span class="o">.</span><span class="n">var_names</span><span class="p">)</span>
        <span class="n">df</span> <span class="o">=</span> <span class="n">df</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">adata_</span><span class="o">.</span><span class="n">obs</span><span class="p">)</span>
        <span class="k">return</span> <span class="n">sc</span><span class="o">.</span><span class="n">AnnData</span><span class="p">(</span><span class="n">df</span><span class="p">[</span><span class="n">adata_</span><span class="o">.</span><span class="n">var_names</span><span class="p">],</span> <span class="n">obs</span><span class="o">=</span><span class="n">df</span><span class="o">.</span><span class="n">drop</span><span class="p">(</span><span class="n">columns</span><span class="o">=</span><span class="n">adata_</span><span class="o">.</span><span class="n">var_names</span><span class="p">))</span>

    <span class="n">adata_</span> <span class="o">=</span> <span class="n">fix_dtypes</span><span class="p">(</span><span class="n">adata_</span><span class="p">)</span>
    <span class="n">sc</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">filter_genes</span><span class="p">(</span><span class="n">adata_</span><span class="p">,</span> <span class="n">min_cells</span><span class="o">=</span><span class="mi">3</span><span class="p">)</span>
    <span class="k">return</span> <span class="n">adata_</span>
</pre></div>
</div>
</div>
</div>
<section id="id21">
<h3><span class="section-number">16.5.1. </span>One group<a class="headerlink" href="#id21" title="Link to this heading">#</a></h3>
<p>As with edgeR, it is possible to fit the model using the full dataset, and then use contrasts to perform the testing for each cell type of interest, but in this notebook we only show the MAST-RE pipeline for one cell type, namely CD14 Monocytes, to shorten the runtime.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">adata_mono</span> <span class="o">=</span> <span class="n">adata</span><span class="p">[</span><span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s2">&quot;cell_type&quot;</span><span class="p">]</span> <span class="o">==</span> <span class="s2">&quot;CD14_Monocytes&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
<span class="n">adata_mono</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>AnnData object with n_obs × n_vars = 5696 × 15701
    obs: &#39;nCount_RNA&#39;, &#39;nFeature_RNA&#39;, &#39;tsne1&#39;, &#39;tsne2&#39;, &#39;label&#39;, &#39;cluster&#39;, &#39;cell_type&#39;, &#39;replicate&#39;, &#39;nCount_SCT&#39;, &#39;nFeature_SCT&#39;, &#39;integrated_snn_res.0.4&#39;, &#39;seurat_clusters&#39;, &#39;n_genes&#39;, &#39;sample&#39;
    var: &#39;name&#39;, &#39;n_cells&#39;
    uns: &#39;edgeR_B_cells&#39;, &#39;edgeR_CD14_Monocytes&#39;, &#39;edgeR_CD4_T_cells&#39;, &#39;edgeR_CD8_T_cells&#39;, &#39;edgeR_Dendritic_cells&#39;, &#39;edgeR_FCGR3A_Monocytes&#39;, &#39;edgeR_NK_cells&#39;, &#39;log1p&#39;
    obsm: &#39;X_pca&#39;, &#39;X_umap&#39;
    layers: &#39;counts&#39;
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">sc</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">filter_genes</span><span class="p">(</span><span class="n">adata_mono</span><span class="p">,</span> <span class="n">min_cells</span><span class="o">=</span><span class="mi">3</span><span class="p">)</span>
<span class="n">adata_mono</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>AnnData object with n_obs × n_vars = 5696 × 12268
    obs: &#39;nCount_RNA&#39;, &#39;nFeature_RNA&#39;, &#39;tsne1&#39;, &#39;tsne2&#39;, &#39;label&#39;, &#39;cluster&#39;, &#39;cell_type&#39;, &#39;replicate&#39;, &#39;nCount_SCT&#39;, &#39;nFeature_SCT&#39;, &#39;integrated_snn_res.0.4&#39;, &#39;seurat_clusters&#39;, &#39;n_genes&#39;, &#39;sample&#39;
    var: &#39;name&#39;, &#39;n_cells&#39;
    uns: &#39;edgeR_B_cells&#39;, &#39;edgeR_CD14_Monocytes&#39;, &#39;edgeR_CD4_T_cells&#39;, &#39;edgeR_CD8_T_cells&#39;, &#39;edgeR_Dendritic_cells&#39;, &#39;edgeR_FCGR3A_Monocytes&#39;, &#39;edgeR_NK_cells&#39;, &#39;log1p&#39;
    obsm: &#39;X_pca&#39;, &#39;X_umap&#39;
    layers: &#39;counts&#39;
</pre></div>
</div>
</div>
</div>
<p>Next we filter both objects as mentioned above.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">adata_mono</span> <span class="o">=</span> <span class="n">prep_anndata</span><span class="p">(</span><span class="n">adata_mono</span><span class="p">)</span>
<span class="n">adata_mono</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>AnnData object with n_obs × n_vars = 5696 × 12268
    obs: &#39;nCount_RNA&#39;, &#39;nFeature_RNA&#39;, &#39;tsne1&#39;, &#39;tsne2&#39;, &#39;label&#39;, &#39;cluster&#39;, &#39;cell_type&#39;, &#39;replicate&#39;, &#39;nCount_SCT&#39;, &#39;nFeature_SCT&#39;, &#39;integrated_snn_res.0.4&#39;, &#39;seurat_clusters&#39;, &#39;n_genes&#39;, &#39;sample&#39;
    var: &#39;n_cells&#39;
</pre></div>
</div>
</div>
</div>
<p>As is the case whenever multiple statistical tests are performed, the obtained p-values for DGE tests over conditions must be corrected for multiple testing using, for example, a Benjamini-Hochberg correction<span id="id22">[<a class="reference internal" href="#id338" title="Malte D Luecken and Fabian J Theis. Current best practices in single-cell term`rna`-seq analysis: a tutorial. Molecular Systems Biology, 15(6):e8746, 2019. URL: https://www.embopress.org/doi/abs/10.15252/msb.20188746, arXiv:https://www.embopress.org/doi/pdf/10.15252/msb.20188746, doi:https://doi.org/10.15252/msb.20188746.">Luecken and Theis, 2019</a>]</span>,<span id="id23">[<a class="reference internal" href="#id336" title="Yoav Benjamini and Yosef Hochberg. Controlling the false discovery rate: a practical and powerful approach to multiple testing. Journal of the Royal Statistical Society: Series B (Methodological), 57(1):289-300, 1995. URL: https://rss.onlinelibrary.wiley.com/doi/abs/10.1111/j.2517-6161.1995.tb02031.x, arXiv:https://rss.onlinelibrary.wiley.com/doi/pdf/10.1111/j.2517-6161.1995.tb02031.x, doi:https://doi.org/10.1111/j.2517-6161.1995.tb02031.x.">Benjamini and Hochberg, 1995</a>]</span>.</p>
<p>Similarly to edgeR analysis, we define a separate function that we use for the analysis:</p>
<ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">find_de_MAST_RE</span></code> takes a <code class="docutils literal notranslate"><span class="pre">SingleCellExperiment</span></code> object as input and runs MAST with RE pipeline. The output of the function is a table (pandas DataFrame in Python) which contains results of the analysis, e.g. log-fold change, p-value and FDR-corrected value for each gene.</p></li>
</ul>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">adata_mono</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s2">&quot;cell_type&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="p">[</span>
    <span class="n">ct</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s2">&quot; &quot;</span><span class="p">,</span> <span class="s2">&quot;_&quot;</span><span class="p">)</span> <span class="k">for</span> <span class="n">ct</span> <span class="ow">in</span> <span class="n">adata_mono</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s2">&quot;cell_type&quot;</span><span class="p">]</span>
<span class="p">]</span>
<span class="n">adata_mono</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s2">&quot;cell_type&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="p">[</span>
    <span class="n">ct</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s2">&quot;+&quot;</span><span class="p">,</span> <span class="s2">&quot;&quot;</span><span class="p">)</span> <span class="k">for</span> <span class="n">ct</span> <span class="ow">in</span> <span class="n">adata_mono</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s2">&quot;cell_type&quot;</span><span class="p">]</span>
<span class="p">]</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="o">%%</span><span class="k">R</span>
find_de_MAST_RE &lt;- function(adata_){
    # create a MAST object
    sca &lt;- SceToSingleCellAssay(adata_, class = &quot;SingleCellAssay&quot;)
    print(&quot;Dimensions before subsetting:&quot;)
    print(dim(sca))
    print(&quot;&quot;)
    # keep genes that are expressed in more than 10% of all cells
    sca &lt;- sca[freq(sca)&gt;0.1,]
    print(&quot;Dimensions after subsetting:&quot;)
    print(dim(sca))
    print(&quot;&quot;)
    # add a column to the data which contains scaled number of genes that are expressed in each cell
    cdr2 &lt;- colSums(assay(sca)&gt;0)
    colData(sca)$ngeneson &lt;- scale(cdr2)
    # store the columns that we are interested in as factors
    label &lt;- factor(colData(sca)$label)
    # set the reference level
    label &lt;- relevel(label,&quot;ctrl&quot;)
    colData(sca)$label &lt;- label
    celltype &lt;- factor(colData(sca)$cell_type)
    colData(sca)$celltype &lt;- celltype
    # same for donors (which we need to model random effects)
    replicate &lt;- factor(colData(sca)$replicate)
    colData(sca)$replicate &lt;- replicate
    # create a group per condition-celltype combination
    colData(sca)$group &lt;- paste0(colData(adata_)$label, &quot;.&quot;, colData(adata_)$cell_type)
    colData(sca)$group &lt;- factor(colData(sca)$group)
    # define and fit the model
    zlmCond &lt;- zlm(formula = ~ngeneson + group + (1 | replicate), 
                   sca=sca, 
                   method=&#39;glmer&#39;, 
                   ebayes=F, 
                   strictConvergence=F,
                   fitArgsD=list(nAGQ = 0)) # to speed up calculations
    
    # perform likelihood-ratio test for the condition that we are interested in    
    summaryCond &lt;- summary(zlmCond, doLRT=&#39;groupstim.CD14_Monocytes&#39;)
    # get the table with log-fold changes and p-values
    summaryDt &lt;- summaryCond$datatable
    result &lt;- merge(summaryDt[contrast==&#39;groupstim.CD14_Monocytes&#39; &amp; component==&#39;H&#39;,.(primerid, `Pr(&gt;Chisq)`)], # p-values
                     summaryDt[contrast==&#39;groupstim.CD14_Monocytes&#39; &amp; component==&#39;logFC&#39;, .(primerid, coef)],
                     by=&#39;primerid&#39;) # logFC coefficients
    # MAST uses natural logarithm so we convert the coefficients to log2 base to be comparable to edgeR
    result[,coef:=result[,coef]/log(2)]
    # do multiple testing correction
    result[,FDR:=p.adjust(`Pr(&gt;Chisq)`, &#39;fdr&#39;)]
    result = result[result$FDR&lt;0.01,, drop=F]

    result &lt;- stats::na.omit(as.data.frame(result))
    return(result)
}
</pre></div>
</div>
</div>
</div>
<p>We run the pipeline for monocytes.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="o">%%time</span>
<span class="o">%%</span><span class="n">R</span> <span class="o">-</span><span class="n">i</span> <span class="n">adata_mono</span> <span class="o">-</span><span class="n">o</span> <span class="n">res</span>
<span class="n">res</span> <span class="o">&lt;-</span><span class="n">find_de_MAST_RE</span><span class="p">(</span><span class="n">adata_mono</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>[1] &quot;Dimensions before subsetting:&quot;
[1] 12268  5696
[1] &quot;&quot;
[1] &quot;Dimensions after subsetting:&quot;
[1] 1676 5696
[1] &quot;&quot;
CPU times: user 9min 35s, sys: 2.96 s, total: 9min 38s
Wall time: 9min 43s
</pre></div>
</div>
</div>
</div>
<p>Let’s take a look at the results.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">res</span><span class="p">[:</span><span class="mi">5</span><span class="p">]</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>primerid</th>
      <th>Pr(&gt;Chisq)</th>
      <th>coef</th>
      <th>FDR</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>1</th>
      <td>AAED1</td>
      <td>8.410341e-19</td>
      <td>0.709836</td>
      <td>1.597106e-18</td>
    </tr>
    <tr>
      <th>2</th>
      <td>ABI1</td>
      <td>1.401688e-05</td>
      <td>0.312797</td>
      <td>1.824921e-05</td>
    </tr>
    <tr>
      <th>3</th>
      <td>ABRACL</td>
      <td>3.920183e-06</td>
      <td>0.418028</td>
      <td>5.201004e-06</td>
    </tr>
    <tr>
      <th>4</th>
      <td>ACADVL</td>
      <td>2.121110e-26</td>
      <td>-0.751305</td>
      <td>4.656979e-26</td>
    </tr>
    <tr>
      <th>5</th>
      <td>ACOT9</td>
      <td>3.424174e-151</td>
      <td>2.921133</td>
      <td>2.689503e-150</td>
    </tr>
  </tbody>
</table>
</div></div></div>
</div>
<p>We store the result in <code class="docutils literal notranslate"><span class="pre">.uns</span></code> as before. Note that we won’t need the <code class="docutils literal notranslate"><span class="pre">score</span></code> column so we just pass logFC there.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">res</span><span class="p">[</span><span class="s2">&quot;gene_symbol&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">res</span><span class="p">[</span><span class="s2">&quot;primerid&quot;</span><span class="p">]</span>
<span class="n">res</span><span class="p">[</span><span class="s2">&quot;cell_type&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="s2">&quot;CD14_Monocytes&quot;</span>
<span class="n">sc_toolbox</span><span class="o">.</span><span class="n">tools</span><span class="o">.</span><span class="n">de_res_to_anndata</span><span class="p">(</span>
    <span class="n">adata</span><span class="p">,</span>
    <span class="n">res</span><span class="p">,</span>
    <span class="n">groupby</span><span class="o">=</span><span class="s2">&quot;cell_type&quot;</span><span class="p">,</span>
    <span class="n">score_col</span><span class="o">=</span><span class="s2">&quot;coef&quot;</span><span class="p">,</span>
    <span class="n">pval_col</span><span class="o">=</span><span class="s2">&quot;Pr(&gt;Chisq)&quot;</span><span class="p">,</span>
    <span class="n">pval_adj_col</span><span class="o">=</span><span class="s2">&quot;FDR&quot;</span><span class="p">,</span>
    <span class="n">lfc_col</span><span class="o">=</span><span class="s2">&quot;coef&quot;</span><span class="p">,</span>
    <span class="n">key_added</span><span class="o">=</span><span class="s2">&quot;MAST_CD14_Monocytes&quot;</span><span class="p">,</span>
<span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">adata_copy</span> <span class="o">=</span> <span class="n">adata</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
</pre></div>
</div>
</div>
</div>
</section>
</section>
<section id="visualization">
<h2><span class="section-number">16.6. </span>Visualization<a class="headerlink" href="#visualization" title="Link to this heading">#</a></h2>
<p>We can visualize the results with a heatmap and a volcano plot. Each row of a heatmap corresponds to a gene and each column to a single-cell. The brighter the color the higher the expression of that gene in a particular cell. Since we only plot DE genes, we would like to see clear differences in expression between the two conditions. Volcano plots are often used to visualize the results of statistical testing, and they show the change in expression on the x-axis (log-fold change) and statistical significance on the y-axis (FDR-corrected p-values). We color code the genes that have FDR-corrected p-value under 0.01 and log-fold change of over 1.5</p>
<p>We normalize the data before plotting the heatmaps to see the differences in expression between two conditions better.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">adata</span><span class="o">.</span><span class="n">X</span> <span class="o">=</span> <span class="n">adata</span><span class="o">.</span><span class="n">layers</span><span class="p">[</span><span class="s2">&quot;counts&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
<span class="n">sc</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">normalize_total</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">target_sum</span><span class="o">=</span><span class="mf">1e6</span><span class="p">)</span>
<span class="n">sc</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">log1p</span><span class="p">(</span><span class="n">adata</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<p>Next, we define a helper plotting function for the heatmaps.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">FDR</span> <span class="o">=</span> <span class="mf">0.01</span>
<span class="n">LOG_FOLD_CHANGE</span> <span class="o">=</span> <span class="mf">1.5</span>


<span class="k">def</span><span class="w"> </span><span class="nf">plot_heatmap</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">group_key</span><span class="p">,</span> <span class="n">group_name</span><span class="o">=</span><span class="s2">&quot;cell_type&quot;</span><span class="p">,</span> <span class="n">groupby</span><span class="o">=</span><span class="s2">&quot;label&quot;</span><span class="p">):</span>
    <span class="n">cell_type</span> <span class="o">=</span> <span class="s2">&quot;_&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">group_key</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;_&quot;</span><span class="p">)[</span><span class="mi">1</span><span class="p">:])</span>
    <span class="n">res</span> <span class="o">=</span> <span class="n">sc</span><span class="o">.</span><span class="n">get</span><span class="o">.</span><span class="n">rank_genes_groups_df</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">group</span><span class="o">=</span><span class="n">cell_type</span><span class="p">,</span> <span class="n">key</span><span class="o">=</span><span class="n">group_key</span><span class="p">)</span>
    <span class="n">res</span><span class="o">.</span><span class="n">index</span> <span class="o">=</span> <span class="n">res</span><span class="p">[</span><span class="s2">&quot;names&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">values</span>
    <span class="n">res</span> <span class="o">=</span> <span class="n">res</span><span class="p">[</span>
        <span class="p">(</span><span class="n">res</span><span class="p">[</span><span class="s2">&quot;pvals_adj&quot;</span><span class="p">]</span> <span class="o">&lt;</span> <span class="n">FDR</span><span class="p">)</span> <span class="o">&amp;</span> <span class="p">(</span><span class="nb">abs</span><span class="p">(</span><span class="n">res</span><span class="p">[</span><span class="s2">&quot;logfoldchanges&quot;</span><span class="p">])</span> <span class="o">&gt;</span> <span class="n">LOG_FOLD_CHANGE</span><span class="p">)</span>
    <span class="p">]</span><span class="o">.</span><span class="n">sort_values</span><span class="p">(</span><span class="n">by</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;logfoldchanges&quot;</span><span class="p">])</span>
    <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Plotting </span><span class="si">{</span><span class="nb">len</span><span class="p">(</span><span class="n">res</span><span class="p">)</span><span class="si">}</span><span class="s2"> genes...&quot;</span><span class="p">)</span>
    <span class="n">markers</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">res</span><span class="o">.</span><span class="n">index</span><span class="p">)</span>
    <span class="n">sc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">heatmap</span><span class="p">(</span>
        <span class="n">adata</span><span class="p">[</span><span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="n">group_name</span><span class="p">]</span> <span class="o">==</span> <span class="n">cell_type</span><span class="p">]</span><span class="o">.</span><span class="n">copy</span><span class="p">(),</span>
        <span class="n">markers</span><span class="p">,</span>
        <span class="n">groupby</span><span class="o">=</span><span class="n">groupby</span><span class="p">,</span>
        <span class="n">swap_axes</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
    <span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<p>And finally we can plot the heatmaps with DEG for CD14+ Monocytes for both edgeR and MAST with RE.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">plot_heatmap</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="s2">&quot;edgeR_CD14_Monocytes&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Plotting 303 genes...
</pre></div>
</div>
<img alt="../_images/8187960e3422145ad21f53845ded1f806737c58ad5b02dd4b4ab22c943c4a6ba.png" src="../_images/8187960e3422145ad21f53845ded1f806737c58ad5b02dd4b4ab22c943c4a6ba.png" />
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">plot_heatmap</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="s2">&quot;MAST_CD14_Monocytes&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Plotting 436 genes...
</pre></div>
</div>
<img alt="../_images/c0529832b01a7536e63d7656c9b845b5d573a62eb462421166aa645f30ccabf2.png" src="../_images/c0529832b01a7536e63d7656c9b845b5d573a62eb462421166aa645f30ccabf2.png" />
</div>
</div>
<p>We observe that MAST identified 436 DEG with our given cut-offs for adjusted p-values and logfold change, while edgeR identified 303 genes.</p>
<p>Next, we define the helper plotting function for the volcano plots.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">FDR</span> <span class="o">=</span> <span class="mf">0.01</span>
<span class="n">LOG_FOLD_CHANGE</span> <span class="o">=</span> <span class="mf">1.5</span>


<span class="k">def</span><span class="w"> </span><span class="nf">volcano_plot</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">group_key</span><span class="p">,</span> <span class="n">group_name</span><span class="o">=</span><span class="s2">&quot;cell_type&quot;</span><span class="p">,</span> <span class="n">groupby</span><span class="o">=</span><span class="s2">&quot;label&quot;</span><span class="p">,</span> <span class="n">title</span><span class="o">=</span><span class="kc">None</span><span class="p">):</span>
    <span class="n">cell_type</span> <span class="o">=</span> <span class="s2">&quot;_&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">group_key</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;_&quot;</span><span class="p">)[</span><span class="mi">1</span><span class="p">:])</span>
    <span class="n">result</span> <span class="o">=</span> <span class="n">sc</span><span class="o">.</span><span class="n">get</span><span class="o">.</span><span class="n">rank_genes_groups_df</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">group</span><span class="o">=</span><span class="n">cell_type</span><span class="p">,</span> <span class="n">key</span><span class="o">=</span><span class="n">group_key</span><span class="p">)</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
    <span class="n">result</span><span class="p">[</span><span class="s2">&quot;-logQ&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="o">-</span><span class="n">np</span><span class="o">.</span><span class="n">log</span><span class="p">(</span><span class="n">result</span><span class="p">[</span><span class="s2">&quot;pvals&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="s2">&quot;float&quot;</span><span class="p">))</span>
    <span class="n">lowqval_de</span> <span class="o">=</span> <span class="n">result</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="nb">abs</span><span class="p">(</span><span class="n">result</span><span class="p">[</span><span class="s2">&quot;logfoldchanges&quot;</span><span class="p">])</span> <span class="o">&gt;</span> <span class="n">LOG_FOLD_CHANGE</span><span class="p">]</span>
    <span class="n">other_de</span> <span class="o">=</span> <span class="n">result</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="nb">abs</span><span class="p">(</span><span class="n">result</span><span class="p">[</span><span class="s2">&quot;logfoldchanges&quot;</span><span class="p">])</span> <span class="o">&lt;=</span> <span class="n">LOG_FOLD_CHANGE</span><span class="p">]</span>

    <span class="n">fig</span><span class="p">,</span> <span class="n">ax</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">subplots</span><span class="p">()</span>
    <span class="n">sns</span><span class="o">.</span><span class="n">regplot</span><span class="p">(</span>
        <span class="n">x</span><span class="o">=</span><span class="n">other_de</span><span class="p">[</span><span class="s2">&quot;logfoldchanges&quot;</span><span class="p">],</span>
        <span class="n">y</span><span class="o">=</span><span class="n">other_de</span><span class="p">[</span><span class="s2">&quot;-logQ&quot;</span><span class="p">],</span>
        <span class="n">fit_reg</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
        <span class="n">scatter_kws</span><span class="o">=</span><span class="p">{</span><span class="s2">&quot;s&quot;</span><span class="p">:</span> <span class="mi">6</span><span class="p">},</span>
    <span class="p">)</span>
    <span class="n">sns</span><span class="o">.</span><span class="n">regplot</span><span class="p">(</span>
        <span class="n">x</span><span class="o">=</span><span class="n">lowqval_de</span><span class="p">[</span><span class="s2">&quot;logfoldchanges&quot;</span><span class="p">],</span>
        <span class="n">y</span><span class="o">=</span><span class="n">lowqval_de</span><span class="p">[</span><span class="s2">&quot;-logQ&quot;</span><span class="p">],</span>
        <span class="n">fit_reg</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
        <span class="n">scatter_kws</span><span class="o">=</span><span class="p">{</span><span class="s2">&quot;s&quot;</span><span class="p">:</span> <span class="mi">6</span><span class="p">},</span>
    <span class="p">)</span>
    <span class="n">ax</span><span class="o">.</span><span class="n">set_xlabel</span><span class="p">(</span><span class="s2">&quot;log2 FC&quot;</span><span class="p">)</span>
    <span class="n">ax</span><span class="o">.</span><span class="n">set_ylabel</span><span class="p">(</span><span class="s2">&quot;-log Q-value&quot;</span><span class="p">)</span>

    <span class="k">if</span> <span class="n">title</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
        <span class="n">title</span> <span class="o">=</span> <span class="n">group_key</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s2">&quot;_&quot;</span><span class="p">,</span> <span class="s2">&quot; &quot;</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">title</span><span class="p">(</span><span class="n">title</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">volcano_plot</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="s2">&quot;MAST_CD14_Monocytes&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="../_images/f3012fc09bcab99baba50ea46b82811be540031578ef2ac76455c278716629f3.png" src="../_images/f3012fc09bcab99baba50ea46b82811be540031578ef2ac76455c278716629f3.png" />
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">volcano_plot</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="s2">&quot;edgeR_CD14_Monocytes&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="../_images/f469e6905f58bee7375427c05e6ce7d6b69a48afe0af66c51e17928e072c93de.png" src="../_images/f469e6905f58bee7375427c05e6ce7d6b69a48afe0af66c51e17928e072c93de.png" />
</div>
</div>
<p>From the heatmaps and especially from the volcano plots, one can see that edgeR identified more up-regulated than down-regulated genes (in stimulated vs control) in contrast to MAST which identified similar number of up- and down-regulated genes.</p>
</section>
<section id="quiz">
<h2><span class="section-number">16.7. </span>Quiz<a class="headerlink" href="#quiz" title="Link to this heading">#</a></h2>
<div class="cell tag_remove_input docutils container">
<div class="cell_output docutils container">
<div class="output text_html">
    <style>
        .output.text_html {
            --pst-color-text-base: transparent !important;
        }
        .flip-card-q1 {
            background-color: transparent;
            width: 350px;
            height: 200px;
            perspective: 1000px;
            display: inline-block;
            margin: 10px;
            text-align: center;
            padding: 10px;
        }
        .flip-card-inner-q1 {
            position: relative;
            width: 100%;
            height: 100%;
            text-align: center;
            transition: transform 0.6s;
            transform-style: preserve-3d;
            transform-origin: center;
        }
        .flip-card-q1:hover .flip-card-inner-q1 {
            transform: rotateY(180deg);
        }
        .flip-card-front-q1, .flip-card-back-q1 {
            position: absolute;
            width: 100%;
            height: 100%;
            min-height: 200px;
            backface-visibility: hidden;
            display: flex;
            align-items: center;
            justify-content: center;
            font-weight: bold;
            color: white;
            padding: 20px;
            border-radius: 15px;
            box-sizing: border-box;
            overflow: hidden;
            text-align: center;
        }
        .flip-card-front-q1 {
            background-color: #3965a3;
            font-size: 20px;
        }
        .flip-card-back-q1 {
            background-color: #a8d480;
            transform: rotateY(180deg);
            font-size: 13px;
        }_b
    </style>

    <div class="flip-card-q1">
        <div class="flip-card-inner-q1">
            <div class="flip-card-front-q1">
                What is differential gene expression and in which cases are we interested in testing for it?
            </div>
            <div class="flip-card-back-q1">
                DGE analysis identifies genes with significant expression differences between conditions, such as healthy versus diseased states, to elucidate underlying biological mechanisms.
            </div>
        </div>
    </div>
    <noscript>
        <p><strong>Q:</strong> What is differential gene expression and in which cases are we interested in testing for it?</p>
        <p><strong>Answer:</strong> DGE analysis identifies genes with significant expression differences between conditions, such as healthy versus diseased states, to elucidate underlying biological mechanisms.</p>
    </noscript>
    </div><div class="output text_html">
    <style>
        .output.text_html {
            --pst-color-text-base: transparent !important;
        }
        .flip-card-q2 {
            background-color: transparent;
            width: 350px;
            height: 200px;
            perspective: 1000px;
            display: inline-block;
            margin: 10px;
            text-align: center;
            padding: 10px;
        }
        .flip-card-inner-q2 {
            position: relative;
            width: 100%;
            height: 100%;
            text-align: center;
            transition: transform 0.6s;
            transform-style: preserve-3d;
            transform-origin: center;
        }
        .flip-card-q2:hover .flip-card-inner-q2 {
            transform: rotateY(180deg);
        }
        .flip-card-front-q2, .flip-card-back-q2 {
            position: absolute;
            width: 100%;
            height: 100%;
            min-height: 200px;
            backface-visibility: hidden;
            display: flex;
            align-items: center;
            justify-content: center;
            font-weight: bold;
            color: white;
            padding: 20px;
            border-radius: 15px;
            box-sizing: border-box;
            overflow: hidden;
            text-align: center;
        }
        .flip-card-front-q2 {
            background-color: #3965a3;
            font-size: 20px;
        }
        .flip-card-back-q2 {
            background-color: #a8d480;
            transform: rotateY(180deg);
            font-size: 13px;
        }_b
    </style>

    <div class="flip-card-q2">
        <div class="flip-card-inner-q2">
            <div class="flip-card-front-q2">
                What is the 'pseudoreplication' problem and how can it be circumvented?
            </div>
            <div class="flip-card-back-q2">
                In single-cell RNA sequencing, treating individual cells from the same subject as independent samples can inflate false discovery rates. This issue can be mitigated by aggregating data into 'pseudobulks' per subject or modeling subjects as random effects.
            </div>
        </div>
    </div>
    <noscript>
        <p><strong>Q:</strong> What is the 'pseudoreplication' problem and how can it be circumvented?</p>
        <p><strong>Answer:</strong> In single-cell RNA sequencing, treating individual cells from the same subject as independent samples can inflate false discovery rates. This issue can be mitigated by aggregating data into 'pseudobulks' per subject or modeling subjects as random effects.</p>
    </noscript>
    </div><div class="output text_html">
    <style>
        .output.text_html {
            --pst-color-text-base: transparent !important;
        }
        .flip-card-q3 {
            background-color: transparent;
            width: 350px;
            height: 200px;
            perspective: 1000px;
            display: inline-block;
            margin: 10px;
            text-align: center;
            padding: 10px;
        }
        .flip-card-inner-q3 {
            position: relative;
            width: 100%;
            height: 100%;
            text-align: center;
            transition: transform 0.6s;
            transform-style: preserve-3d;
            transform-origin: center;
        }
        .flip-card-q3:hover .flip-card-inner-q3 {
            transform: rotateY(180deg);
        }
        .flip-card-front-q3, .flip-card-back-q3 {
            position: absolute;
            width: 100%;
            height: 100%;
            min-height: 200px;
            backface-visibility: hidden;
            display: flex;
            align-items: center;
            justify-content: center;
            font-weight: bold;
            color: white;
            padding: 20px;
            border-radius: 15px;
            box-sizing: border-box;
            overflow: hidden;
            text-align: center;
        }
        .flip-card-front-q3 {
            background-color: #3965a3;
            font-size: 20px;
        }
        .flip-card-back-q3 {
            background-color: #a8d480;
            transform: rotateY(180deg);
            font-size: 15px;
        }_b
    </style>

    <div class="flip-card-q3">
        <div class="flip-card-inner-q3">
            <div class="flip-card-front-q3">
                What is the 'multiple testig' problem and how can it be eluded?
            </div>
            <div class="flip-card-back-q3">
                Testing thousands of genes simultaneously increases the likelihood of false positives. To address this, corrections like the Benjamini-Hochberg procedure are applied to control the false discovery rate..
            </div>
        </div>
    </div>
    <noscript>
        <p><strong>Q:</strong> What is the 'multiple testig' problem and how can it be eluded?</p>
        <p><strong>Answer:</strong> Testing thousands of genes simultaneously increases the likelihood of false positives. To address this, corrections like the Benjamini-Hochberg procedure are applied to control the false discovery rate..</p>
    </noscript>
    </div></div>
</div>
</section>
<section id="references">
<h2><span class="section-number">16.8. </span>References<a class="headerlink" href="#references" title="Link to this heading">#</a></h2>
<div class="docutils container" id="id24">
<div role="list" class="citation-list">
<div class="citation" id="id336" role="doc-biblioentry">
<span class="label"><span class="fn-bracket">[</span><a role="doc-backlink" href="#id23">deBH95</a><span class="fn-bracket">]</span></span>
<p>Yoav Benjamini and Yosef Hochberg. Controlling the false discovery rate: a practical and powerful approach to multiple testing. <em>Journal of the Royal Statistical Society: Series B (Methodological)</em>, 57(1):289–300, 1995. URL: <a class="reference external" href="https://rss.onlinelibrary.wiley.com/doi/abs/10.1111/j.2517-6161.1995.tb02031.x">https://rss.onlinelibrary.wiley.com/doi/abs/10.1111/j.2517-6161.1995.tb02031.x</a>, <a class="reference external" href="https://arxiv.org/abs/https://rss.onlinelibrary.wiley.com/doi/pdf/10.1111/j.2517-6161.1995.tb02031.x">arXiv:https://rss.onlinelibrary.wiley.com/doi/pdf/10.1111/j.2517-6161.1995.tb02031.x</a>, <a class="reference external" href="https://doi.org/https://doi.org/10.1111/j.2517-6161.1995.tb02031.x">doi:https://doi.org/10.1111/j.2517-6161.1995.tb02031.x</a>.</p>
</div>
<div class="citation" id="id323" role="doc-biblioentry">
<span class="label"><span class="fn-bracket">[</span><a role="doc-backlink" href="#id4">deBKvB+17</a><span class="fn-bracket">]</span></span>
<p>Mollie E. Brooks, Kasper Kristensen, Koen J. van Benthem, Arni Magnusson, Casper W. Berg, Anders Nielsen, Hans J. Skaug, Martin Mächler, and Benjamin M. Bolker. glmmTMB Balances Speed and Flexibility Among Packages for Zero-inflated Generalized Linear Mixed Modeling. <em>The R Journal</em>, 9(2):378–400, 2017. URL: <a class="reference external" href="https://doi.org/10.32614/RJ-2017-066">https://doi.org/10.32614/RJ-2017-066</a>, <a class="reference external" href="https://doi.org/10.32614/RJ-2017-066">doi:10.32614/RJ-2017-066</a>.</p>
</div>
<div class="citation" id="id324" role="doc-biblioentry">
<span class="label"><span class="fn-bracket">[</span>deDRM+21<span class="fn-bracket">]</span></span>
<span class="backrefs">(<a role="doc-backlink" href="#id5">1</a>,<a role="doc-backlink" href="#id7">2</a>)</span>
<p>Samarendra Das, Anil Rai, Michael L Merchant, Matthew C Cave, and Shesh N Rai. A comprehensive survey of statistical approaches for differential expression analysis in single-cell term`RNA` sequencing studies. <em>Genes (Basel)</em>, 12(12):1947, December 2021.</p>
</div>
<div class="citation" id="id322" role="doc-biblioentry">
<span class="label"><span class="fn-bracket">[</span>deFMY+15<span class="fn-bracket">]</span></span>
<span class="backrefs">(<a role="doc-backlink" href="#id3">1</a>,<a role="doc-backlink" href="#id20">2</a>)</span>
<p>Greg Finak, Andrew McDavid, Masanao Yajima, Jingyuan Deng, Vivian Gersuk, Alex K. Shalek, Chloe K. Slichter, Hannah W. Miller, M. Juliana McElrath, Martin Prlic, Peter S. Linsley, and Raphael Gottardo. Mast: a flexible statistical framework for assessing transcriptional changes and characterizing heterogeneity in single-cell term`rna` sequencing data. <em>Genome Biology</em>, 16(1):278, Dec 2015. URL: <a class="reference external" href="https://doi.org/10.1186/s13059-015-0844-5">https://doi.org/10.1186/s13059-015-0844-5</a>, <a class="reference external" href="https://doi.org/10.1186/s13059-015-0844-5">doi:10.1186/s13059-015-0844-5</a>.</p>
</div>
<div class="citation" id="id339" role="doc-biblioentry">
<span class="label"><span class="fn-bracket">[</span><a role="doc-backlink" href="#id11">deHHAN+21</a><span class="fn-bracket">]</span></span>
<p>Yuhan Hao, Stephanie Hao, Erica Andersen-Nissen, William M. Mauck III, Shiwei Zheng, Andrew Butler, Maddie J. Lee, Aaron J. Wilk, Charlotte Darby, Michael Zagar, Paul Hoffman, Marlon Stoeckius, Efthymia Papalexi, Eleni P. Mimitou, Jaison Jain, Avi Srivastava, Tim Stuart, Lamar B. Fleming, Bertrand Yeung, Angela J. Rogers, Juliana M. McElrath, Catherine A. Blish, Raphael Gottardo, Peter Smibert, and Rahul Satija. Integrated analysis of multimodal single-cell data. <em>Cell</em>, 2021. URL: <a class="reference external" href="https://doi.org/10.1016/j.cell.2021.04.048">https://doi.org/10.1016/j.cell.2021.04.048</a>, <a class="reference external" href="https://doi.org/10.1016/j.cell.2021.04.048">doi:10.1016/j.cell.2021.04.048</a>.</p>
</div>
<div class="citation" id="id320" role="doc-biblioentry">
<span class="label"><span class="fn-bracket">[</span><a role="doc-backlink" href="#id6">deHTTI17</a><span class="fn-bracket">]</span></span>
<p>Stephanie C Hicks, F William Townes, Mingxiang Teng, and Rafael A Irizarry. Missing data and technical variability in single-cell term`RNA`-sequencing experiments. <em>Biostatistics</em>, 19(4):562–578, 11 2017. URL: <a class="reference external" href="https://doi.org/10.1093/biostatistics/kxx053">https://doi.org/10.1093/biostatistics/kxx053</a>, <a class="reference external" href="https://arxiv.org/abs/https://academic.oup.com/biostatistics/article-pdf/19/4/562/26346801/kxx053.pdf">arXiv:https://academic.oup.com/biostatistics/article-pdf/19/4/562/26346801/kxx053.pdf</a>, <a class="reference external" href="https://doi.org/10.1093/biostatistics/kxx053">doi:10.1093/biostatistics/kxx053</a>.</p>
</div>
<div class="citation" id="id326" role="doc-biblioentry">
<span class="label"><span class="fn-bracket">[</span><a role="doc-backlink" href="#id7">deJSeyetermDNAsrollahME16</a><span class="fn-bracket">]</span></span>
<p>Maria K Jaakkola, Fatemeh Seyeterm`DNA`srollah, Arfa Mehmood, and Laura L Elo. Comparison of methods to detect differentially expressed genes between single-cell populations. <em>Briefings in Bioinformatics</em>, 18(5):735–743, 07 2016. URL: <a class="reference external" href="https://doi.org/10.1093/bib/bbw057">https://doi.org/10.1093/bib/bbw057</a>, <a class="reference external" href="https://arxiv.org/abs/https://academic.oup.com/bib/article-pdf/18/5/735/25581122/bbw057.pdf">arXiv:https://academic.oup.com/bib/article-pdf/18/5/735/25581122/bbw057.pdf</a>, <a class="reference external" href="https://doi.org/10.1093/bib/bbw057">doi:10.1093/bib/bbw057</a>.</p>
</div>
<div class="citation" id="id331" role="doc-biblioentry">
<span class="label"><span class="fn-bracket">[</span>deJSE22<span class="fn-bracket">]</span></span>
<span class="backrefs">(<a role="doc-backlink" href="#id8">1</a>,<a role="doc-backlink" href="#id12">2</a>)</span>
<p>Sini Junttila, Johannes Smolander, and Laura L Elo. Benchmarking methods for detecting differential states between conditions from multi-subject single-cell term`rna`-seq data. <em>bioRxiv</em>, 2022. URL: <a class="reference external" href="https://www.biorxiv.org/content/early/2022/02/19/2022.02.16.480662">https://www.biorxiv.org/content/early/2022/02/19/2022.02.16.480662</a>, <a class="reference external" href="https://arxiv.org/abs/https://www.biorxiv.org/content/early/2022/02/19/2022.02.16.480662.full.pdf">arXiv:https://www.biorxiv.org/content/early/2022/02/19/2022.02.16.480662.full.pdf</a>, <a class="reference external" href="https://doi.org/10.1101/2022.02.16.480662">doi:10.1101/2022.02.16.480662</a>.</p>
</div>
<div class="citation" id="id337" role="doc-biblioentry">
<span class="label"><span class="fn-bracket">[</span><a role="doc-backlink" href="#id14">deKST+18</a><span class="fn-bracket">]</span></span>
<p>Hyun Min Kang, Meena Subramaniam, Sasha Targ, Michelle Nguyen, Lenka Maliskova, Elizabeth McCarthy, Eunice Wan, Simon Wong, Lauren Byrnes, Cristina M Lanata, and others. Multiplexed droplet single-cell term`rna`-sequencing using natural genetic variation. <em>Nature biotechnology</em>, 36(1):89–94, 2018.</p>
</div>
<div class="citation" id="id334" role="doc-biblioentry">
<span class="label"><span class="fn-bracket">[</span><a role="doc-backlink" href="#id16">deLZD+20</a><span class="fn-bracket">]</span></span>
<p>Charity W. Law, Kathleen Zeglinski, Xueyi Dong, Monther Alhamdoosh, Gordon K. Smyth, and Matthew E. Ritchie. A guide to creating design matrices for gene expression experiments. <em>F1000Research</em>, 9:1444–1444, Dec 2020. 33604029[pmid]. URL: <a class="reference external" href="https://pubmed.ncbi.nlm.nih.gov/33604029">https://pubmed.ncbi.nlm.nih.gov/33604029</a>.</p>
</div>
<div class="citation" id="id318" role="doc-biblioentry">
<span class="label"><span class="fn-bracket">[</span><a role="doc-backlink" href="#id2">deLHA14</a><span class="fn-bracket">]</span></span>
<p>Michael I. Love, Wolfgang Huber, and Simon Anders. Moderated estimation of fold change and dispersion for term`rna`-seq data with deseq2. <em>Genome Biology</em>, 15(12):550, Dec 2014. URL: <a class="reference external" href="https://doi.org/10.1186/s13059-014-0550-8">https://doi.org/10.1186/s13059-014-0550-8</a>, <a class="reference external" href="https://doi.org/10.1186/s13059-014-0550-8">doi:10.1186/s13059-014-0550-8</a>.</p>
</div>
<div class="citation" id="id338" role="doc-biblioentry">
<span class="label"><span class="fn-bracket">[</span>deLT19<span class="fn-bracket">]</span></span>
<span class="backrefs">(<a role="doc-backlink" href="#id6">1</a>,<a role="doc-backlink" href="#id22">2</a>)</span>
<p>Malte D Luecken and Fabian J Theis. Current best practices in single-cell term`rna`-seq analysis: a tutorial. <em>Molecular Systems Biology</em>, 15(6):e8746, 2019. URL: <a class="reference external" href="https://www.embopress.org/doi/abs/10.15252/msb.20188746">https://www.embopress.org/doi/abs/10.15252/msb.20188746</a>, <a class="reference external" href="https://arxiv.org/abs/https://www.embopress.org/doi/pdf/10.15252/msb.20188746">arXiv:https://www.embopress.org/doi/pdf/10.15252/msb.20188746</a>, <a class="reference external" href="https://doi.org/https://doi.org/10.15252/msb.20188746">doi:https://doi.org/10.15252/msb.20188746</a>.</p>
</div>
<div class="citation" id="id335" role="doc-biblioentry">
<span class="label"><span class="fn-bracket">[</span><a role="doc-backlink" href="#id18">deLahnemannKosterS+20</a><span class="fn-bracket">]</span></span>
<p>David Lähnemann, Johannes Köster, Ewa Szczurek, Davis J. McCarthy, Stephanie C. Hicks, Mark D. Robinson, Catalina A. Vallejos, Kieran R. Campbell, Niko Beerenwinkel, Ahmed Mahfouz, Luca Pinello, Pavel Skums, Alexandros Stamatakis, Camille Stephan-Otto Attolini, Samuel Aparicio, Jasmijn Baaijens, Marleen Balvert, Buys de Barbanson, Antonio Cappuccio, Giacomo Corleone, Bas E. Dutilh, Maria Florescu, Victor Guryev, Rens Holmer, Katharina Jahn, Thamar Jessurun Lobo, Emma M. Keizer, Indu Khatri, Szymon M. Kielbasa, Jan O. Korbel, Alexey M. Kozlov, Tzu-Hao Kuo, Boudewijn P.F. Lelieveldt, Ion I. Mandoiu, John C. Marioni, Tobias Marschall, Felix Mölder, Amir Niknejad, Lukasz Raczkowski, Marcel Reinders, Jeroen de Ridder, Antoine-Emmanuel Saliba, Antonios Somarakis, Oliver Stegle, Fabian J. Theis, Huan Yang, Alex Zelikovsky, Alice C. McHardy, Benjamin J. Raphael, Sohrab P. Shah, and Alexander Schönhuth. Eleven grand challenges in single-cell data science. <em>Genome Biology</em>, 21(1):31, Feb 2020. URL: <a class="reference external" href="https://doi.org/10.1186/s13059-020-1926-6">https://doi.org/10.1186/s13059-020-1926-6</a>, <a class="reference external" href="https://doi.org/10.1186/s13059-020-1926-6">doi:10.1186/s13059-020-1926-6</a>.</p>
</div>
<div class="citation" id="id330" role="doc-biblioentry">
<span class="label"><span class="fn-bracket">[</span><a role="doc-backlink" href="#id13">deMS22</a><span class="fn-bracket">]</span></span>
<p>Alan E. Murphy and Nathan G. Skene. A balanced measure shows superior performance of pseudobulk methods in single-cell RNA-sequencing analysis. <em>Nat Commun</em>, dec 2022. URL: <a class="reference external" href="https://doi.org/10.1038%2Fs41467-022-35519-4">https://doi.org/10.1038%2Fs41467-022-35519-4</a>, <a class="reference external" href="https://doi.org/10.1038/s41467-022-35519-4">doi:10.1038/s41467-022-35519-4</a>.</p>
</div>
<div class="citation" id="id332" role="doc-biblioentry">
<span class="label"><span class="fn-bracket">[</span><a role="doc-backlink" href="#id10">deRPW+15</a><span class="fn-bracket">]</span></span>
<p>Matthew E. Ritchie, Belinda Phipson, Di Wu, Yifang Hu, Charity W. Law, Wei Shi, and Gordon K. Smyth. Limma powers differential expression analyses for term`rna`-sequencing and microarray studies. <em>Nucleic acids research</em>, 43(7):e47–e47, Apr 2015. gkv007[PII]. URL: <a class="reference external" href="https://doi.org/10.1093/nar/gkv007">https://doi.org/10.1093/nar/gkv007</a>, <a class="reference external" href="https://doi.org/10.1093/nar/gkv007">doi:10.1093/nar/gkv007</a>.</p>
</div>
<div class="citation" id="id317" role="doc-biblioentry">
<span class="label"><span class="fn-bracket">[</span>deRMS10<span class="fn-bracket">]</span></span>
<span class="backrefs">(<a role="doc-backlink" href="#id1">1</a>,<a role="doc-backlink" href="#id15">2</a>)</span>
<p>Mark D. Robinson, Davis J. McCarthy, and Gordon K. Smyth. Edger: a bioconductor package for differential expression analysis of digital gene expression data. <em>Bioinformatics (Oxford, England)</em>, 26(1):139–140, Jan 2010. btp616[PII]. URL: <a class="reference external" href="https://doi.org/10.1093/bioinformatics/btp616">https://doi.org/10.1093/bioinformatics/btp616</a>, <a class="reference external" href="https://doi.org/10.1093/bioinformatics/btp616">doi:10.1093/bioinformatics/btp616</a>.</p>
</div>
<div class="citation" id="id325" role="doc-biblioentry">
<span class="label"><span class="fn-bracket">[</span><a role="doc-backlink" href="#id7">deSR18</a><span class="fn-bracket">]</span></span>
<p>Charlotte Soneson and Mark D. Robinson. Bias, robustness and scalability in single-cell differential expression analysis. <em>Nature Methods</em>, 15(4):255–261, Apr 2018. URL: <a class="reference external" href="https://doi.org/10.1038/nmeth.4612">https://doi.org/10.1038/nmeth.4612</a>, <a class="reference external" href="https://doi.org/10.1038/nmeth.4612">doi:10.1038/nmeth.4612</a>.</p>
</div>
<div class="citation" id="id327" role="doc-biblioentry">
<span class="label"><span class="fn-bracket">[</span>deSGK+21<span class="fn-bracket">]</span></span>
<span class="backrefs">(<a role="doc-backlink" href="#id7">1</a>,<a role="doc-backlink" href="#id8">2</a>)</span>
<p>Jordan W. Squair, Matthieu Gautier, Claudia Kathe, Mark A. Anderson, Nicholas D. James, Thomas H. Hutson, Rémi Hudelle, Taha Qaiser, Kaya J. E. Matson, Quentin Barraud, Ariel J. Levine, Gioele La Manno, Michael A. Skinnider, and Grégoire Courtine. Confronting false discoveries in single-cell differential expression. <em>Nature Communications</em>, 12(1):5692, Sep 2021. URL: <a class="reference external" href="https://doi.org/10.1038/s41467-021-25960-2">https://doi.org/10.1038/s41467-021-25960-2</a>, <a class="reference external" href="https://doi.org/10.1038/s41467-021-25960-2">doi:10.1038/s41467-021-25960-2</a>.</p>
</div>
<div class="citation" id="id333" role="doc-biblioentry">
<span class="label"><span class="fn-bracket">[</span><a role="doc-backlink" href="#id17">deTRCP21</a><span class="fn-bracket">]</span></span>
<p>Andrew L Thurman, Jason A Ratcliff, Michael S Chimenti, and Alejandro A Pezzulo. Differential gene expression analysis for multi-subject single cell term`RNA` sequencing studies with aggregateBioVar. <em>Bioinformatics</em>, 37(19):3243–3251, May 2021.</p>
</div>
<div class="citation" id="id321" role="doc-biblioentry">
<span class="label"><span class="fn-bracket">[</span><a role="doc-backlink" href="#id6">deVRS+17</a><span class="fn-bracket">]</span></span>
<p>Catalina A. Vallejos, Davide Risso, Antonio Scialdone, Sandrine Dudoit, and John C. Marioni. Normalizing single-cell term`rna` sequencing data: challenges and opportunities. <em>Nature Methods</em>, 14(6):565–571, Jun 2017. URL: <a class="reference external" href="https://doi.org/10.1038/nmeth.4292">https://doi.org/10.1038/nmeth.4292</a>, <a class="reference external" href="https://doi.org/10.1038/nmeth.4292">doi:10.1038/nmeth.4292</a>.</p>
</div>
<div class="citation" id="id319" role="doc-biblioentry">
<span class="label"><span class="fn-bracket">[</span><a role="doc-backlink" href="#id5">deWLNN19</a><span class="fn-bracket">]</span></span>
<p>Tianyu Wang, Boyang Li, Craig E. Nelson, and Sheida Nabavi. Comparative analysis of differential gene expression analysis tools for single-cell term`rna` sequencing data. <em>BMC Bioinformatics</em>, 20(1):40, Jan 2019. URL: <a class="reference external" href="https://doi.org/10.1186/s12859-019-2599-6">https://doi.org/10.1186/s12859-019-2599-6</a>, <a class="reference external" href="https://doi.org/10.1186/s12859-019-2599-6">doi:10.1186/s12859-019-2599-6</a>.</p>
</div>
<div class="citation" id="id329" role="doc-biblioentry">
<span class="label"><span class="fn-bracket">[</span>deZEL21<span class="fn-bracket">]</span></span>
<span class="backrefs">(<a role="doc-backlink" href="#id8">1</a>,<a role="doc-backlink" href="#id9">2</a>,<a role="doc-backlink" href="#id19">3</a>)</span>
<p>Kip D. Zimmerman, Mark A. Espeland, and Carl D. Langefeld. A practical solution to pseudoreplication bias in single-cell studies. <em>Nature Communications</em>, 12(1):738, Feb 2021. URL: <a class="reference external" href="https://doi.org/10.1038/s41467-021-21038-1">https://doi.org/10.1038/s41467-021-21038-1</a>, <a class="reference external" href="https://doi.org/10.1038/s41467-021-21038-1">doi:10.1038/s41467-021-21038-1</a>.</p>
</div>
</div>
</div>
</section>
<section id="contributors">
<h2><span class="section-number">16.9. </span>Contributors<a class="headerlink" href="#contributors" title="Link to this heading">#</a></h2>
<p>We gratefully acknowledge the contributions of:</p>
<section id="authors">
<h3><span class="section-number">16.9.1. </span>Authors<a class="headerlink" href="#authors" title="Link to this heading">#</a></h3>
<ul class="simple">
<li><p>Lukas Heumos</p></li>
<li><p>Anastasia Litinetskaya</p></li>
<li><p>Soroor Hediyeh-Zadeh</p></li>
</ul>
</section>
<section id="reviewers">
<h3><span class="section-number">16.9.2. </span>Reviewers<a class="headerlink" href="#reviewers" title="Link to this heading">#</a></h3>
</section>
</section>
</section>

    <script type="text/x-thebe-config">
    {
        requestKernel: true,
        binderOptions: {
            repo: "binder-examples/jupyter-stacks-datascience",
            ref: "master",
        },
        codeMirrorConfig: {
            theme: "abcdef",
            mode: "python"
        },
        kernelOptions: {
            name: "python3",
            path: "./conditions"
        },
        predefinedOutput: true
    }
    </script>
    <script>kernelName = 'python3'</script>

                </article>
              

              
              
              
              
                <footer class="prev-next-footer d-print-none">
                  
<div class="prev-next-area">
    <a class="left-prev"
       href="../trajectories/lineage_tracing.html"
       title="previous page">
      <i class="fa-solid fa-angle-left"></i>
      <div class="prev-next-info">
        <p class="prev-next-subtitle">previous</p>
        <p class="prev-next-title"><span class="section-number">15. </span>Lineage tracing</p>
      </div>
    </a>
    <a class="right-next"
       href="compositional.html"
       title="next page">
      <div class="prev-next-info">
        <p class="prev-next-subtitle">next</p>
        <p class="prev-next-title"><span class="section-number">17. </span>Compositional analysis</p>
      </div>
      <i class="fa-solid fa-angle-right"></i>
    </a>
</div>
                </footer>
              
            </div>
            
            
              
                <div class="bd-sidebar-secondary bd-toc"><div class="sidebar-secondary-items sidebar-secondary__inner">


  <div class="sidebar-secondary-item">
  <div class="page-toc tocsection onthispage">
    <i class="fa-solid fa-list"></i> Contents
  </div>
  <nav class="bd-toc-nav page-toc">
    <ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#motivation">16.1. Motivation</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#environment-setup">16.2. Environment setup</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#preparing-the-dataset">16.3. Preparing the dataset</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#pseudobulk">16.4. Pseudobulk</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#one-group">16.4.1. One group</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#multiple-groups">16.4.2. Multiple groups</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#notes-on-edger">16.4.3. Notes on edgeR</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#single-cell-specific">16.5. Single-cell specific</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#id21">16.5.1. One group</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#visualization">16.6. Visualization</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#quiz">16.7. Quiz</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#references">16.8. References</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#contributors">16.9. Contributors</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#authors">16.9.1. Authors</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#reviewers">16.9.2. Reviewers</a></li>
</ul>
</li>
</ul>
  </nav></div>

</div></div>
              
            
          </div>
          <footer class="bd-footer-content">
            
<div class="bd-footer-content__inner container">
  
  <div class="footer-item">
    
<p class="component-author">
By Lukas Heumos, Anna Schaar, single-cell best practices consortium
</p>

  </div>
  
  <div class="footer-item">
    

  <p class="copyright">
    
      © Copyright 2023.
      <br/>
    
  </p>

  </div>
  
  <div class="footer-item">
    
  </div>
  
  <div class="footer-item">
    
<div class="extra_footer">
  <div>
  Brought to you by
  <a href="https://www.github.com/theislab">Theislab</a>, with many thanks to the single-cell community as a whole!
</div>

</div>
  </div>
  
</div>
          </footer>
        

      </main>
    </div>
  </div>
  
  <!-- Scripts loaded after <body> so the DOM is not blocked -->
  <script src="../_static/scripts/bootstrap.js?digest=dfe6caa3a7d634c4db9b"></script>
<script src="../_static/scripts/pydata-sphinx-theme.js?digest=dfe6caa3a7d634c4db9b"></script>

  <footer class="bd-footer">
  </footer>
  </body>
</html>