# Release notes

<!-- towncrier release notes start -->

## [v1.0.0](https://github.com/theislab/single-cell-best-practices/releases/tag/1.0.0) <sub>2021-08-17</sub>

### Changed

- Fix typos ([#164](https://github.com/theislab/single-cell-best-practices/pull/164)) <sub>@mjstrumillo</sub>
- Correct 2 spelling mistakes ([#155](https://github.com/theislab/single-cell-best-practices/pull/155)) <sub>@namsaraeva</sub>
- Fix typos in chapter Normalization ([#118](https://github.com/theislab/single-cell-best-practices/pull/118)) <sub>@galicae</sub>
- Update `RNA velocity` chapter ([#100](https://github.com/theislab/single-cell-best-practices/pull/100)) <sub>@WeilerP</sub>
- Update `pseudotime` chapter ([#148](https://github.com/theislab/single-cell-best-practices/pull/148)) <sub>@WeilerP</sub>
- Update multimodal immune receptor chapters ([#110](https://github.com/theislab/single-cell-best-practices/pull/110)) <sub>@drEast</sub>
- Simplify `GSEA` chapter and improve performance ([#119](https://github.com/theislab/single-cell-best-practices/pull/119)) <sub>@ivirshup</sub>
- Improve `integration` chapter ([#109](https://github.com/theislab/single-cell-best-practices/pull/109)) <sub>@lazappi</sub>
- Update B-cell clonotype analysis ([#105](https://github.com/theislab/single-cell-best-practices/pull/105)) <sub>@jdhenaos</sub>
- Citeseq proofread ([#102](https://github.com/theislab/single-cell-best-practices/pull/102)) <sub>@cramsuig</sub>
- Review of analysis frameworks ([#98](https://github.com/theislab/single-cell-best-practices/pull/98)) <sub>@ivirshup</sub>
- Citeseq pp notebook ([#60](https://github.com/theislab/single-cell-best-practices/pull/60)) <sub>@danielStrobl</sub>
- Update `perturbation modeling` chapter ([#91](https://github.com/theislab/single-cell-best-practices/pull/91)) <sub>@yugeji</sub>
- Update `compositional analysis` chapter ([#76](https://github.com/theislab/single-cell-best-practices/pull/76)) <sub>@johannesostner</sub>
- Update `raw data processing` chapter ([#77](https://github.com/theislab/single-cell-best-practices/pull/77)) <sub>@DongzeHE</sub>
- Update `differential gene expression` chapter ([#80](https://github.com/theislab/single-cell-best-practices/pull/80)) <sub>@alitinet</sub>
- Update `preprocessing` section ([#139](https://github.com/theislab/single-cell-best-practices/pull/139)) <sub>@AnnaChristina</sub>
- Update `compositional analysis` chapter ([#149](https://github.com/theislab/single-cell-best-practices/pull/149)) <sub>@xinyuejohn</sub>
- Update `paired integration` and `advanced integration` chapter ([#135](https://github.com/theislab/single-cell-best-practices/pull/135)) <sub>@alitinet</sub>
- Improve pseudobulk plots for `differential gene expression` chapter ([#131](https://github.com/theislab/single-cell-best-practices/pull/131)) <sub>@alitinet</sub>
- Update `advanced integration` chapter ([#112](https://github.com/theislab/single-cell-best-practices/pull/112)) <sub>@alitinet</sub>
- Fix typos ([#166](https://github.com/theislab/single-cell-best-practices/pull/166)) <sub>@alitinet</sub>
- Update `compositional analysis` chapter according to changes of `pertpy` ([#152](https://github.com/theislab/single-cell-best-practices/pull/152)) <sub>@xinyuejohn</sub>
- Update link in `README.md` ([#107](https://github.com/theislab/single-cell-best-practices/pull/107)) <sub>@WeilerP</sub>
- Update `spatial deconvolution` chapter ([#93](https://github.com/theislab/single-cell-best-practices/pull/93)) <sub>@amitfrish</sub>
- Update `experimental data collection` chapter ([#71](https://github.com/theislab/single-cell-best-practices/pull/71)) <sub>@AnnaChristina</sub>
- Update `introduction` section ([#70](https://github.com/theislab/single-cell-best-practices/pull/70)) <sub>@AnnaChristina</sub>
- Update `preamble` ([#69](https://github.com/theislab/single-cell-best-practices/pull/69)) <sub>@AnnaChristina</sub>

### Added

- Add simpleaf example to the raw data processing section ([#116](https://github.com/theislab/single-cell-best-practices/pull/116)) <sub>@DongzeHE</sub>
- Add `inferring trajectories` section ([#101](https://github.com/theislab/single-cell-best-practices/pull/101)) <sub>@WeilerP</sub>
- Add `perturbation modelling` chapter ([#79](https://github.com/theislab/single-cell-best-practices/pull/79)) <sub>@Zethson</sub>
- Add Annotation ([#85](https://github.com/theislab/single-cell-best-practices/pull/85)) <sub>@LisaSikkema</sub>
- Add `gsea pathway` chapter ([#62](https://github.com/theislab/single-cell-best-practices/pull/62)) <sub>@soroorh</sub>
- Add `neighborhood analysis`, `spatial domains`, `spacially variable genes` chapters ([#86](https://github.com/theislab/single-cell-best-practices/pull/86)) <sub>@giovp</sub>
- Add `cell-cell communication` chapter ([#140](https://github.com/theislab/single-cell-best-practices/pull/140)) <sub>@dbdimitrov</sub>
- Add `paired integration` chapter ([#103](https://github.com/theislab/single-cell-best-practices/pull/103)) <sub>@alitinet</sub>
- Add multiple multimodal immunce receptor chapters ([#63](https://github.com/theislab/single-cell-best-practices/pull/63)) <sub>@drEast</sub>
- Add `spatial deconvolution` and `imputation` chapter ([#142](https://github.com/theislab/single-cell-best-practices/pull/142)) <sub>@AnnaChristina</sub>
- Add feedback from Robin Browaeys to `cell-cell comunication` chapter ([#157](https://github.com/theislab/single-cell-best-practices/pull/157)) <sub>@dbdimitrov</sub>
- Add docker container & improve environments ([#124](https://github.com/theislab/single-cell-best-practices/pull/124)) <sub>@Zethson</sub>
- Add `compisitional analysis` chapter ([#89](https://github.com/theislab/single-cell-best-practices/pull/89)) <sub>@Zethson</sub>
- Add `lineage tracing` chapter ([#88](https://github.com/theislab/single-cell-best-practices/pull/88)) <sub>@mattjones315</sub>
- Add `preprocessing and visualizaton` section ([#59](https://github.com/theislab/single-cell-best-practices/pull/59)) <sub>@AnnaChristina</sub>
- Add `clustering` chapter ([#72](https://github.com/theislab/single-cell-best-practices/pull/72)) <sub>@AnnaChristina</sub>
- Add PDF job ([#65](https://github.com/theislab/single-cell-best-practices/pull/65)) <sub>@Zethson</sub>
- Add `prior art` chapter ([#58](https://github.com/theislab/single-cell-best-practices/pull/58)) <sub>@Zethson</sub>
- Add new book CI ([#57](https://github.com/theislab/single-cell-best-practices/pull/57)) <sub>@Zethson</sub>
- Add `preamble` chapter ([#53](https://github.com/theislab/single-cell-best-practices/pull/53)) <sub>@Zethson</sub>
- Add scATAC part ([#162](https://github.com/theislab/single-cell-best-practices/pull/162)) <sub>@lauradmartens</sub>
- Add `modelling mechanism` section ([#96](https://github.com/theislab/single-cell-best-practices/pull/96)) <sub>@ilibarra</sub>
- Add `Integration` chapter ([#17](https://github.com/theislab/single-cell-best-practices/pull/17)) <sub>@lazappi</sub>
- Add new book CI ([#57](https://github.com/theislab/single-cell-best-practices/pull/57)) <sub>@Zethson</sub>

### Fixed

- Fix broken link on edit button ([#115](https://github.com/theislab/single-cell-best-practices/pull/115)) <sub>@emiller88</sub>
- Fix environment CI ([#158](https://github.com/theislab/single-cell-best-practices/pull/158)) <sub>@Zethson</sub>
- Fix surface protein environment and small bugs ([#154](https://github.com/theislab/single-cell-best-practices/pulls/154)) <sub>@danielStrobl</sub>
- Add missing links to other chapters in `Compositional analysis` chapter ([#145](https://github.com/theislab/single-cell-best-practices/pull/145)) <sub>@dbdimitrov</sub>
- Fix environment CI ([#158](https://github.com/theislab/single-cell-best-practices/pull/158)) <sub>@Zethson</sub>
- Fix environments for `surface protein` section and small bugs ([#154](https://github.com/theislab/single-cell-best-practices/pull/154)) <sub>@danielStrobl</sub>
