
<!DOCTYPE html>


<html lang="en" data-content_root="../" >

  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="viewport" content="width=device-width, initial-scale=1" />

    <title>24. Quality Control &#8212; Single-cell best practices</title>
  
  
  
  <script data-cfasync="false">
    document.documentElement.dataset.mode = localStorage.getItem("mode") || "";
    document.documentElement.dataset.theme = localStorage.getItem("theme") || "";
  </script>
  
  <!-- Loaded before other Sphinx assets -->
  <link href="../_static/styles/theme.css?digest=dfe6caa3a7d634c4db9b" rel="stylesheet" />
<link href="../_static/styles/bootstrap.css?digest=dfe6caa3a7d634c4db9b" rel="stylesheet" />
<link href="../_static/styles/pydata-sphinx-theme.css?digest=dfe6caa3a7d634c4db9b" rel="stylesheet" />

  
  <link href="../_static/vendor/fontawesome/6.5.2/css/all.min.css?digest=dfe6caa3a7d634c4db9b" rel="stylesheet" />
  <link rel="preload" as="font" type="font/woff2" crossorigin href="../_static/vendor/fontawesome/6.5.2/webfonts/fa-solid-900.woff2" />
<link rel="preload" as="font" type="font/woff2" crossorigin href="../_static/vendor/fontawesome/6.5.2/webfonts/fa-brands-400.woff2" />
<link rel="preload" as="font" type="font/woff2" crossorigin href="../_static/vendor/fontawesome/6.5.2/webfonts/fa-regular-400.woff2" />

    <link rel="stylesheet" type="text/css" href="../_static/pygments.css?v=03e43079" />
    <link rel="stylesheet" type="text/css" href="../_static/styles/sphinx-book-theme.css?v=eba8b062" />
    <link rel="stylesheet" type="text/css" href="../_static/togglebutton.css?v=13237357" />
    <link rel="stylesheet" type="text/css" href="../_static/copybutton.css?v=76b2166b" />
    <link rel="stylesheet" type="text/css" href="../_static/mystnb.4510f1fc1dee50b3e5859aac5469c37c29e427902b24a333a5f9fcb2f0b3ac41.css" />
    <link rel="stylesheet" type="text/css" href="../_static/sphinx-thebe.css?v=4fa983c6" />
    <link rel="stylesheet" type="text/css" href="../_static/sphinx-design.min.css?v=95c83b7e" />
    <link rel="stylesheet" type="text/css" href="../_static/book.css?v=148c11e2" />
  
  <!-- Pre-loaded scripts that we'll load fully later -->
  <link rel="preload" as="script" href="../_static/scripts/bootstrap.js?digest=dfe6caa3a7d634c4db9b" />
<link rel="preload" as="script" href="../_static/scripts/pydata-sphinx-theme.js?digest=dfe6caa3a7d634c4db9b" />
  <script src="../_static/vendor/fontawesome/6.5.2/js/all.min.js?digest=dfe6caa3a7d634c4db9b"></script>

    <script src="../_static/documentation_options.js?v=9eb32ce0"></script>
    <script src="../_static/doctools.js?v=9a2dae69"></script>
    <script src="../_static/sphinx_highlight.js?v=dc90522c"></script>
    <script src="../_static/clipboard.min.js?v=a7894cd8"></script>
    <script src="../_static/copybutton.js?v=f281be69"></script>
    <script src="../_static/scripts/sphinx-book-theme.js?v=887ef09a"></script>
    <script>let toggleHintShow = 'Click to show';</script>
    <script>let toggleHintHide = 'Click to hide';</script>
    <script>let toggleOpenOnPrint = 'true';</script>
    <script src="../_static/togglebutton.js?v=4a39c7ea"></script>
    <script>var togglebuttonSelector = '.toggle, .admonition.dropdown';</script>
    <script src="../_static/design-tabs.js?v=f930bc37"></script>
    <script>const THEBE_JS_URL = "https://unpkg.com/thebe@0.8.2/lib/index.js"; const thebe_selector = ".thebe,.cell"; const thebe_selector_input = "pre"; const thebe_selector_output = ".output, .cell_output"</script>
    <script async="async" src="../_static/sphinx-thebe.js?v=c100c467"></script>
    <script>var togglebuttonSelector = '.toggle, .admonition.dropdown';</script>
    <script>const THEBE_JS_URL = "https://unpkg.com/thebe@0.8.2/lib/index.js"; const thebe_selector = ".thebe,.cell"; const thebe_selector_input = "pre"; const thebe_selector_output = ".output, .cell_output"</script>
    <script>DOCUMENTATION_OPTIONS.pagename = 'chromatin_accessibility/quality_control';</script>
    <script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
    <script src="../_static/book.js?v=dcd3d8bf"></script>
    <link rel="icon" href="../_static/favicon.ico"/>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="25. Gene regulatory networks using chromatin accessibility" href="gene_regulatory_networks_atac.html" />
    <link rel="prev" title="23. Single-cell ATAC sequencing" href="introduction.html" />
  <meta name="viewport" content="width=device-width, initial-scale=1"/>
  <meta name="docsearch:language" content="en"/>
  </head>
  
  
  <body data-bs-spy="scroll" data-bs-target=".bd-toc-nav" data-offset="180" data-bs-root-margin="0px 0px -60%" data-default-mode="">

  
  
  <div id="pst-skip-link" class="skip-link d-print-none"><a href="#main-content">Skip to main content</a></div>
  
  <div id="pst-scroll-pixel-helper"></div>
  
  <button type="button" class="btn rounded-pill" id="pst-back-to-top">
    <i class="fa-solid fa-arrow-up"></i>Back to top</button>

  
  <input type="checkbox"
          class="sidebar-toggle"
          id="pst-primary-sidebar-checkbox"/>
  <label class="overlay overlay-primary" for="pst-primary-sidebar-checkbox"></label>
  
  <input type="checkbox"
          class="sidebar-toggle"
          id="pst-secondary-sidebar-checkbox"/>
  <label class="overlay overlay-secondary" for="pst-secondary-sidebar-checkbox"></label>
  
  <div class="search-button__wrapper">
    <div class="search-button__overlay"></div>
    <div class="search-button__search-container">
<form class="bd-search d-flex align-items-center"
      action="../search.html"
      method="get">
  <i class="fa-solid fa-magnifying-glass"></i>
  <input type="search"
         class="form-control"
         name="q"
         id="search-input"
         placeholder="Search this book..."
         aria-label="Search this book..."
         autocomplete="off"
         autocorrect="off"
         autocapitalize="off"
         spellcheck="false"/>
  <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd>K</kbd></span>
</form></div>
  </div>

  <div class="pst-async-banner-revealer d-none">
  <aside id="bd-header-version-warning" class="d-none d-print-none" aria-label="Version warning"></aside>
</div>

  
    <header class="bd-header navbar navbar-expand-lg bd-navbar d-print-none">
    </header>
  

  <div class="bd-container">
    <div class="bd-container__inner bd-page-width">
      
      
      
      <div class="bd-sidebar-primary bd-sidebar">
        

  
  <div class="sidebar-header-items sidebar-primary__section">
    
    
    
    
  </div>
  
    <div class="sidebar-primary-items__start sidebar-primary__section">
        <div class="sidebar-primary-item">

  
    
  

<a class="navbar-brand logo" href="../preamble.html">
  
  
  
  
  
    
    
      
    
    
    <img src="../_static/logo.png" class="logo__image only-light" alt="Single-cell best practices - Home"/>
    <script>document.write(`<img src="../_static/logo.png" class="logo__image only-dark" alt="Single-cell best practices - Home"/>`);</script>
  
  
</a></div>
        <div class="sidebar-primary-item">

 <script>
 document.write(`
   <button class="btn search-button-field search-button__button" title="Search" aria-label="Search" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="fa-solid fa-magnifying-glass"></i>
    <span class="search-button__default-text">Search</span>
    <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd class="kbd-shortcut__modifier">K</kbd></span>
   </button>
 `);
 </script></div>
        <div class="sidebar-primary-item"><nav class="bd-links bd-docs-nav" aria-label="Main">
    <div class="bd-toc-item navbar-nav active">
        <p aria-level="2" class="caption" role="heading"><span class="caption-text">Introduction</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../introduction/prior_art.html">1. Prior art</a></li>
<li class="toctree-l1"><a class="reference internal" href="../introduction/scrna_seq.html">2. Single-cell RNA sequencing</a></li>
<li class="toctree-l1"><a class="reference internal" href="../introduction/raw_data_processing.html">3. Raw data processing</a></li>
<li class="toctree-l1"><a class="reference internal" href="../introduction/analysis_tools.html">4. Analysis frameworks and tools</a></li>
<li class="toctree-l1"><a class="reference internal" href="../introduction/interoperability.html">5. Interoperability</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Preprocessing and visualization</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../preprocessing_visualization/quality_control.html">6. Quality Control</a></li>
<li class="toctree-l1"><a class="reference internal" href="../preprocessing_visualization/normalization.html">7. Normalization</a></li>
<li class="toctree-l1"><a class="reference internal" href="../preprocessing_visualization/feature_selection.html">8. Feature selection</a></li>
<li class="toctree-l1"><a class="reference internal" href="../preprocessing_visualization/dimensionality_reduction.html">9. Dimensionality Reduction</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Identifying cellular structure</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../cellular_structure/clustering.html">10. Clustering</a></li>
<li class="toctree-l1"><a class="reference internal" href="../cellular_structure/annotation.html">11. Annotation</a></li>
<li class="toctree-l1"><a class="reference internal" href="../cellular_structure/integration.html">12. Data integration</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Inferring trajectories</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../trajectories/pseudotemporal.html">13. Pseudotemporal ordering</a></li>
<li class="toctree-l1"><a class="reference internal" href="../trajectories/rna_velocity.html">14. RNA velocity</a></li>
<li class="toctree-l1"><a class="reference internal" href="../trajectories/lineage_tracing.html">15. Lineage tracing</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Dealing with conditions</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../conditions/differential_gene_expression.html">16. Differential gene expression analysis</a></li>
<li class="toctree-l1"><a class="reference internal" href="../conditions/compositional.html">17. Compositional analysis</a></li>
<li class="toctree-l1"><a class="reference internal" href="../conditions/gsea_pathway.html">18. Gene set enrichment and pathway analysis</a></li>
<li class="toctree-l1"><a class="reference internal" href="../conditions/perturbation_modeling.html">19. Perturbation modeling</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Modeling mechanisms</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../mechanisms/gene_regulatory_networks.html">20. Gene regulatory networks</a></li>
<li class="toctree-l1"><a class="reference internal" href="../mechanisms/cell_cell_communication.html">21. Cell-cell communication</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Deconvolution</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../deconvolution/bulk_deconvolution.html">22. Bulk deconvolution</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Chromatin Accessibility</span></p>
<ul class="current nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="introduction.html">23. Single-cell ATAC sequencing</a></li>
<li class="toctree-l1 current active"><a class="current reference internal" href="#">24. Quality Control</a></li>
<li class="toctree-l1"><a class="reference internal" href="gene_regulatory_networks_atac.html">25. Gene regulatory networks using chromatin accessibility</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Spatial omics</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../spatial/introduction.html">26. Single-cell data resolved in space</a></li>
<li class="toctree-l1"><a class="reference internal" href="../spatial/neighborhood.html">27. Neighborhood analysis</a></li>
<li class="toctree-l1"><a class="reference internal" href="../spatial/domains.html">28. Spatial domains</a></li>
<li class="toctree-l1"><a class="reference internal" href="../spatial/spatially_variable_genes.html">29. Spatially variable genes</a></li>
<li class="toctree-l1"><a class="reference internal" href="../spatial/deconvolution.html">30. Spatial deconvolution</a></li>
<li class="toctree-l1"><a class="reference internal" href="../spatial/imputation.html">31. Imputation</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Surface protein</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../surface_protein/quality_control.html">32. Quality control</a></li>
<li class="toctree-l1"><a class="reference internal" href="../surface_protein/normalization.html">33. Normalization</a></li>
<li class="toctree-l1"><a class="reference internal" href="../surface_protein/doublet_detection.html">34. Doublet detection</a></li>
<li class="toctree-l1"><a class="reference internal" href="../surface_protein/dimensionality_reduction.html">35. Dimensionality Reduction</a></li>
<li class="toctree-l1"><a class="reference internal" href="../surface_protein/batch_correction.html">36. Batch correction</a></li>
<li class="toctree-l1"><a class="reference internal" href="../surface_protein/annotation.html">37. Annotation</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Adaptive immune receptor repertoire</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../air_repertoire/ir_profiling.html">38. Immune Receptor Profiling</a></li>
<li class="toctree-l1"><a class="reference internal" href="../air_repertoire/clonotype.html">39. Clonotype analysis</a></li>



<li class="toctree-l1"><a class="reference internal" href="../air_repertoire/specificity.html">43. Specificity analysis</a></li>
<li class="toctree-l1"><a class="reference internal" href="../air_repertoire/multimodal_integration.html">44. Integrating AIR and transcriptomics</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Multimodal integration</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../multimodal_integration/paired_integration.html">45. Paired integration</a></li>
<li class="toctree-l1"><a class="reference internal" href="../multimodal_integration/advanced_integration.html">46. Advanced integration</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Reproducibility</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../reproducibility/introduction.html">47. Reproducibility</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Outlook</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../outlook.html">48. Outlook</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Acknowledgements</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../acknowledgements.html">49. Acknowledgements</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Glossary</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../glossary.html">50. Glossary</a></li>
</ul>

    </div>
</nav></div>
    </div>
  
  
  <div class="sidebar-primary-items__end sidebar-primary__section">
  </div>
  
  <div id="rtd-footer-container"></div>


      </div>
      
      <main id="main-content" class="bd-main" role="main">
        
        

<div class="sbt-scroll-pixel-helper"></div>

          <div class="bd-content">
            <div class="bd-article-container">
              
              <div class="bd-header-article d-print-none">
<div class="header-article-items header-article__inner">
  
    <div class="header-article-items__start">
      
        <div class="header-article-item"><button class="sidebar-toggle primary-toggle btn btn-sm" title="Toggle primary sidebar" data-bs-placement="bottom" data-bs-toggle="tooltip">
  <span class="fa-solid fa-bars"></span>
</button></div>
      
    </div>
  
  
    <div class="header-article-items__end">
      
        <div class="header-article-item">

<div class="article-header-buttons">





<div class="dropdown dropdown-source-buttons">
  <button class="btn dropdown-toggle" type="button" data-bs-toggle="dropdown" aria-expanded="false" aria-label="Source repositories">
    <i class="fab fa-github"></i>
  </button>
  <ul class="dropdown-menu">
      
      
      
      <li><a href="https://github.com/theislab/single-cell-best-practices" target="_blank"
   class="btn btn-sm btn-source-repository-button dropdown-item"
   title="Source repository"
   data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fab fa-github"></i>
  </span>
<span class="btn__text-container">Repository</span>
</a>
</li>
      
      
      
      
      <li><a href="https://github.com/theislab/single-cell-best-practices/edit/master/jupyter-book/chromatin_accessibility/quality_control.ipynb" target="_blank"
   class="btn btn-sm btn-source-edit-button dropdown-item"
   title="Suggest edit"
   data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-pencil-alt"></i>
  </span>
<span class="btn__text-container">Suggest edit</span>
</a>
</li>
      
      
      
      
      <li><a href="https://github.com/theislab/single-cell-best-practices/issues/new?title=Issue%20on%20page%20%2Fchromatin_accessibility/quality_control.html&body=Your%20issue%20content%20here." target="_blank"
   class="btn btn-sm btn-source-issues-button dropdown-item"
   title="Open an issue"
   data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-lightbulb"></i>
  </span>
<span class="btn__text-container">Open issue</span>
</a>
</li>
      
  </ul>
</div>






<div class="dropdown dropdown-download-buttons">
  <button class="btn dropdown-toggle" type="button" data-bs-toggle="dropdown" aria-expanded="false" aria-label="Download this page">
    <i class="fas fa-download"></i>
  </button>
  <ul class="dropdown-menu">
      
      
      
      <li><a href="../_sources/chromatin_accessibility/quality_control.ipynb" target="_blank"
   class="btn btn-sm btn-download-source-button dropdown-item"
   title="Download source file"
   data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-file"></i>
  </span>
<span class="btn__text-container">.ipynb</span>
</a>
</li>
      
      
      
      
      <li>
<button onclick="window.print()"
  class="btn btn-sm btn-download-pdf-button dropdown-item"
  title="Print to PDF"
  data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-file-pdf"></i>
  </span>
<span class="btn__text-container">.pdf</span>
</button>
</li>
      
  </ul>
</div>




<button onclick="toggleFullScreen()"
  class="btn btn-sm btn-fullscreen-button"
  title="Fullscreen mode"
  data-bs-placement="bottom" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-expand"></i>
  </span>

</button>



<script>
document.write(`
  <button class="btn btn-sm nav-link pst-navbar-icon theme-switch-button" title="light/dark" aria-label="light/dark" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="theme-switch fa-solid fa-sun fa-lg" data-mode="light"></i>
    <i class="theme-switch fa-solid fa-moon fa-lg" data-mode="dark"></i>
    <i class="theme-switch fa-solid fa-circle-half-stroke fa-lg" data-mode="auto"></i>
  </button>
`);
</script>


<script>
document.write(`
  <button class="btn btn-sm pst-navbar-icon search-button search-button__button" title="Search" aria-label="Search" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="fa-solid fa-magnifying-glass fa-lg"></i>
  </button>
`);
</script>
<button class="sidebar-toggle secondary-toggle btn btn-sm" title="Toggle secondary sidebar" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <span class="fa-solid fa-list"></span>
</button>
</div></div>
      
    </div>
  
</div>
</div>
              
              

<div id="jb-print-docs-body" class="onlyprint">
    <h1>Quality Control</h1>
    <!-- Table of contents -->
    <div id="print-main-content">
        <div id="jb-print-toc">
            
            <div>
                <h2> Contents </h2>
            </div>
            <nav aria-label="Page">
                <ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#motivation">24.1. Motivation</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#dataset">24.2. Dataset</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#doublet-detection">24.3. Doublet detection</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#count-distribution-based-doublet-scoring">24.3.1. Count distribution-based doublet scoring</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#coverage-based-doublet-scoring">24.3.2. Coverage-based doublet scoring</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#calculating-qc-metrics">24.4. Calculating QC metrics</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#total-fragment-count-and-number-of-features">24.4.1. Total fragment count and number of features</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#nucleosome-signal">24.4.2. Nucleosome signal</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#tss-enrichment">24.4.3. TSS enrichment</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#filtering-cells">24.5. Filtering cells</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#upper-bound-qc-thresholds">24.5.1. Upper bound QC thresholds</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#lower-bound-qc-thresholds">24.5.2. Lower bound QC thresholds</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#perform-filtering">24.5.3. Perform filtering</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#filtering-features">24.6. Filtering features</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#references">24.7. References</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#contributors">24.8. Contributors</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#authors">24.8.1. Authors</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#reviewers">24.8.2. Reviewers</a></li>
</ul>
</li>
</ul>
            </nav>
        </div>
    </div>
</div>

              
                
<div id="searchbox"></div>
                <article class="bd-article">
                  
  <section class="tex2jax_ignore mathjax_ignore" id="quality-control">
<h1><span class="section-number">24. </span>Quality Control<a class="headerlink" href="#quality-control" title="Link to this heading">#</a></h1>
<details class="sd-sphinx-override sd-dropdown sd-card sd-mb-3">
<summary class="sd-summary-title sd-card-header">
<span class="sd-summary-text"><i class="fas fa-brain"></i>   Key takeaways</span><span class="sd-summary-state-marker sd-summary-chevron-right"><svg version="1.1" width="1.5em" height="1.5em" class="sd-octicon sd-octicon-chevron-right" viewBox="0 0 24 24" aria-hidden="true"><path d="M8.72 18.78a.75.75 0 0 1 0-1.06L14.44 12 8.72 6.28a.751.751 0 0 1 .018-1.042.751.751 0 0 1 1.042-.018l6.25 6.25a.75.75 0 0 1 0 1.06l-6.25 6.25a.75.75 0 0 1-1.06 0Z"></path></svg></span></summary><div class="sd-summary-content sd-card-body docutils">
<div class="sd-card sd-sphinx-override sd-mb-3 sd-shadow-sm sd-card-hover docutils">
<div class="sd-card-body docutils">
<p class="sd-card-text">Use two orthogonal methods for robust doublet scoring leveraging the count distribution of simulated doublets as well as the number of genomic positions with more than two counts.
Keep in mind that the latter requires a sufficient sequencing depth (&gt; 10-15k reads per cell).</p>
</div>
<a class="sd-stretched-link sd-hide-link-text reference internal" href="#chromatin-accessibility-quality-control-takeaway-1"><span class="std std-ref">Doublet detection</span></a></div>
<div class="sd-card sd-sphinx-override sd-mb-3 sd-shadow-sm sd-card-hover docutils">
<div class="sd-card-body docutils">
<p class="sd-card-text">Use the total number of fragments, the number of features and scATAC- specific metrics like the transcription start site (TSS) enrichment score and the nucleosome signal to identify low quality cells.
Remove barcodes with extreme values in any of the QC metrics (mostly upper bound thresholds).
Remove the majority of low-quality cells mostly represented by a low total number of fragments combined with a low TSS enrichment score per cell.</p>
</div>
<a class="sd-stretched-link sd-hide-link-text reference internal" href="#chromatin-accessibility-quality-control-takeaway-2"><span class="std std-ref">Filtering cells</span></a></div>
<div class="sd-card sd-sphinx-override sd-mb-3 sd-shadow-sm sd-card-hover docutils">
<div class="sd-card-body docutils">
<p class="sd-card-text">Filter features considering the total number of cells in the data set and the objective of the analysis.
If no rare cell states are of interest, features can be filtered to be present in at least 1% of the cells. If rare cell populations are of interest consider using a minimal number of cells that features should be detected in.</p>
</div>
<a class="sd-stretched-link sd-hide-link-text reference internal" href="#chromatin-accessibility-quality-control-takeaway-3"><span class="std std-ref">Filtering features</span></a></div>
</div>
</details><details class="sd-sphinx-override sd-dropdown sd-card sd-mb-3">
<summary class="sd-summary-title sd-card-header">
<span class="sd-summary-text"><i class="fa-solid fa-gear"></i>   Environment setup</span><span class="sd-summary-state-marker sd-summary-chevron-right"><svg version="1.1" width="1.5em" height="1.5em" class="sd-octicon sd-octicon-chevron-right" viewBox="0 0 24 24" aria-hidden="true"><path d="M8.72 18.78a.75.75 0 0 1 0-1.06L14.44 12 8.72 6.28a.751.751 0 0 1 .018-1.042.751.751 0 0 1 1.042-.018l6.25 6.25a.75.75 0 0 1 0 1.06l-6.25 6.25a.75.75 0 0 1-1.06 0Z"></path></svg></span></summary><div class="sd-summary-content sd-card-body docutils">
<div class="sd-tab-set docutils">
<input checked="checked" id="sd-tab-item-0" name="sd-tab-set-0" type="radio">
<label class="sd-tab-label" for="sd-tab-item-0">
Steps</label><div class="sd-tab-content docutils">
<ol class="arabic">
<li><p class="sd-card-text"><strong>Install conda</strong>:</p>
<ul class="simple">
<li><p class="sd-card-text">Before creating the environment, ensure that <a class="reference external" href="https://docs.conda.io/projects/conda/en/latest/user-guide/install/index.html">conda</a> is installed on your system.</p></li>
</ul>
</li>
<li><p class="sd-card-text"><strong>Save the yml content</strong>:</p>
<ul class="simple">
<li><p class="sd-card-text">Copy the content from the yml tab into a file named <code class="docutils literal notranslate"><span class="pre">environment.yml</span></code>.</p></li>
</ul>
</li>
<li><p class="sd-card-text"><strong>Create the environment</strong>:</p>
<ul>
<li><p class="sd-card-text">Open a terminal or command prompt.</p></li>
<li><p class="sd-card-text">Run the following command:</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>conda<span class="w"> </span>env<span class="w"> </span>create<span class="w"> </span>-f<span class="w"> </span>environment.yml
</pre></div>
</div>
</li>
</ul>
</li>
<li><p class="sd-card-text"><strong>Activate the environment</strong>:</p>
<ul>
<li><p class="sd-card-text">After the environment is created, activate it using:</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>conda<span class="w"> </span>activate<span class="w"> </span>&lt;environment_name&gt;
</pre></div>
</div>
</li>
<li><p class="sd-card-text">Replace <code class="docutils literal notranslate"><span class="pre">&lt;environment_name&gt;</span></code> with the name specified in the <code class="docutils literal notranslate"><span class="pre">environment.yml</span></code> file. In the yml file it will look like this:</p>
<div class="highlight-yaml notranslate"><div class="highlight"><pre><span></span><span class="nt">name</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">&lt;environment_name&gt;</span>
</pre></div>
</div>
</li>
</ul>
</li>
<li><p class="sd-card-text"><strong>Verify the installation</strong>:</p>
<ul>
<li><p class="sd-card-text">Check that the environment was created successfully by running:</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>conda<span class="w"> </span>env<span class="w"> </span>list
</pre></div>
</div>
</li>
</ul>
</li>
</ol>
</div>
<input id="sd-tab-item-1" name="sd-tab-set-0" type="radio">
<label class="sd-tab-label" for="sd-tab-item-1">
yml</label><div class="sd-tab-content docutils">
<div class="highlight-yaml notranslate"><div class="highlight"><pre><span></span><span class="nt">name</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">gene-regulatory-networks-atac</span>

<span class="nt">channels</span><span class="p">:</span>
<span class="w">  </span><span class="p p-Indicator">-</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">conda-forge</span>
<span class="w">  </span><span class="p p-Indicator">-</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">anaconda</span>
<span class="w">  </span><span class="p p-Indicator">-</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">bioconda</span>
<span class="nt">dependencies</span><span class="p">:</span>
<span class="w">  </span><span class="p p-Indicator">-</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">conda-forge::python=3.8</span><span class="w"> </span><span class="c1"># this package version is required for figR</span>
<span class="w">  </span><span class="p p-Indicator">-</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">conda-forge::r-base=4.2.2</span>
<span class="w">  </span><span class="p p-Indicator">-</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">conda-forge::r-umap</span>
<span class="w">  </span><span class="p p-Indicator">-</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">conda-forge::r-devtools</span>
<span class="w">  </span><span class="p p-Indicator">-</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">conda-forge::r-cairo</span>
<span class="w">  </span><span class="p p-Indicator">-</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">conda-forge::r-ggrastr</span>
<span class="w">  </span><span class="p p-Indicator">-</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">conda-forge::r-patchwork</span>
<span class="w">  </span><span class="p p-Indicator">-</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">conda-forge::r-gplots</span>
<span class="w">  </span><span class="p p-Indicator">-</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">conda-forge::r-networkd3</span><span class="w"> </span><span class="c1"># network viz</span>
<span class="w">  </span><span class="p p-Indicator">-</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">conda-forge::r-r2d3</span><span class="w"> </span><span class="c1"># network viz</span>
<span class="w">  </span><span class="p p-Indicator">-</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">conda-forge::r-webshot</span><span class="w"> </span><span class="c1"># network viz</span>
<span class="w">  </span><span class="p p-Indicator">-</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">conda-forge::phantomjs</span><span class="w"> </span><span class="c1"># network viz</span>
<span class="w">  </span><span class="p p-Indicator">-</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">conda-forge::r-imager</span><span class="w"> </span><span class="c1"># network viz</span>
<span class="w">  </span><span class="p p-Indicator">-</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">conda-forge::jupyterlab</span>
<span class="w">  </span><span class="p p-Indicator">-</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">conda-forge::jupyter_server=1.18.1</span>
<span class="w">  </span><span class="p p-Indicator">-</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">conda-forge::r-rmpfr</span>
<span class="w">  </span><span class="p p-Indicator">-</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">conda-forge::r-irkernel</span>
<span class="w">  </span><span class="p p-Indicator">-</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">conda-forge::r-dplyr</span>
<span class="w">  </span><span class="p p-Indicator">-</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">bioconda::bioconductor-zellkonverter&gt;=1.8</span>
<span class="w">  </span><span class="c1"># - bioconda::bioconductor-bsgenome.hsapiens.ucsc.hg38=1.4.4 # this package will have to be installed while running</span>
<span class="w">  </span><span class="p p-Indicator">-</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">bioconda::bioconductor-complexheatmap</span>
<span class="w">  </span><span class="p p-Indicator">-</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">bioconda::bioconductor-motifmatchr</span>
<span class="w">  </span><span class="p p-Indicator">-</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">bioconda::bioconductor-rcistarget</span>
<span class="w">  </span><span class="p p-Indicator">-</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">bioconda::bioconductor-chromvar</span>
</pre></div>
</div>
</div>
</div>
</div>
</details><section id="motivation">
<h2><span class="section-number">24.1. </span>Motivation<a class="headerlink" href="#motivation" title="Link to this heading">#</a></h2>
<p>A crucial part of every single-cell analysis is the removal of low-quality cells which can distort downstream analyses outcomes. scATAC-seq data, having only 1-10% of open chromatin regions detected in each cell, is even more sparse than scRNA-seq data <span id="id1">[<a class="reference internal" href="#id444" title="Huidong Chen, Caleb Lareau, Tommaso Andreani, Michael E. Vinyard, Sara P. Garcia, Kendell Clement, Miguel A. Andrade-Navarro, Jason D. Buenrostro, and Luca Pinello. Assessment of computational methods for the analysis of single-cell ATAC-seq data. Genome Biology, 20(1):241, 2019. URL: https://genomebiology.biomedcentral.com/articles/10.1186/s13059-019-1854-5, doi:10.1186/s13059-019-1854-5.">Chen <em>et al.</em>, 2019</a>]</span>. Therefore, quality issues such as low sequencing depth per cell or poor signal-to-noise ratio can lead to uninformative observations (cells). Another challenge is the detection of multiplets (two or more cells profiled together). In the following chapter, we will introduce metrics for quality control (QC) of scATAC-seq data and introduce doublet detection methods.</p>
</section>
<section id="dataset">
<h2><span class="section-number">24.2. </span>Dataset<a class="headerlink" href="#dataset" title="Link to this heading">#</a></h2>
<p>To showcase the processing of scATAC-seq data, we use a 10x Multiome data set generated for the single cell data integration challenge at the NeurIPS conference 2021 <span id="id2">[<a class="reference internal" href="#id447" title="Malte D Luecken, Daniel Bernard Burkhardt, Robrecht Cannoodt, Christopher Lance, Aditi Agrawal, Hananeh Aliee, Ann T Chen, Louise Deconinck, Angela M Detweiler, Alejandro A Granados, Shelly Huynh, Laura Isacco, Yang Joon Kim, Dominik Klein, BONY DE KUMAR, Sunil Kuppasani, Heiko Lickert, Aaron McGeever, Honey Mekonen, Joaquin Caceres Melgarejo, Maurizio Morri, Michaela Müller, Norma Neff, Sheryl Paul, Bastian Rieck, Kaylie Schneider, Scott Steelman, Michael Sterr, Daniel J. Treacy, Alexander Tong, Alexandra-Chloe Villani, Guilin Wang, Jia Yan, Ce Zhang, Angela Oliveira Pisco, Smita Krishnaswamy, Fabian J Theis, and Jonathan M. Bloom. A sandbox for prediction and integration of dna, rna, and proteins in single cells. In Thirty-fifth Conference on Neural Information Processing Systems Datasets and Benchmarks Track (Round 2). 2021. URL: https://openreview.net/forum?id=gN35BGa1Rt.">Luecken <em>et al.</em>, 2021</a>]</span>. Note that this data set contains multiple samples, making feature harmonization and integration important to consider before analysing them jointly (discussed in later chapters). However, the most unbiased quality assessment can be derived by examining each sample individually. Therefore, we describe the preprocessing of one selected sample in this notebook.</p>
<p>Our starting point is the output of <code class="docutils literal notranslate"><span class="pre">cellranger-arc</span></code>, the software solution of 10x to perform alignment, peak calling and initial QC of their 10x Multiome assay. By default, the output files contain the snRNA-seq and the scATAC-seq data. Since the preprocessing of scRNA-seq or snRNA-seq data has been described extensively in previous chapters, here, we only discuss the processing of the chromatin accessibility data (which are also applicable to data from a unimodal scATAC-seq assay).</p>
<p>The primary file we load is the <code class="docutils literal notranslate"><span class="pre">filtered_feature_bc_matrix.h5</span></code> file containing the cell-by-peak count matrix. During loading, we leverage the functionality of <code class="docutils literal notranslate"><span class="pre">moun</span></code> to automatically detect the corresponding fragment and peak annotation files within the same directory.</p>
<p>To get started, we import all required Python packages.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Single-cell packages</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">anndata2ri</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">matplotlib.pyplot</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">plt</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">muon</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">mu</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">numpy</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">np</span>

<span class="c1"># General helpful packages for data analysis and visualization</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">pandas</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">pd</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">scanpy</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">sc</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">seaborn</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">sns</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">muon</span><span class="w"> </span><span class="kn">import</span> <span class="n">atac</span> <span class="k">as</span> <span class="n">ac</span>  <span class="c1"># the module containing function for scATAC data processing</span>

<span class="c1"># Packages enabling to run R code</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">rpy2.robjects</span><span class="w"> </span><span class="kn">import</span> <span class="n">pandas2ri</span>

<span class="n">pandas2ri</span><span class="o">.</span><span class="n">activate</span><span class="p">()</span>  <span class="c1"># Automatically convert rpy2 outputs to pandas DataFrames</span>
<span class="n">anndata2ri</span><span class="o">.</span><span class="n">activate</span><span class="p">()</span>
<span class="o">%</span><span class="k">load_ext</span> rpy2.ipython


<span class="c1"># Setting figure parameters</span>
<span class="n">sc</span><span class="o">.</span><span class="n">settings</span><span class="o">.</span><span class="n">verbosity</span> <span class="o">=</span> <span class="mi">0</span>
<span class="n">sns</span><span class="o">.</span><span class="n">set</span><span class="p">(</span><span class="n">rc</span><span class="o">=</span><span class="p">{</span><span class="s2">&quot;figure.figsize&quot;</span><span class="p">:</span> <span class="p">(</span><span class="mi">4</span><span class="p">,</span> <span class="mf">3.5</span><span class="p">),</span> <span class="s2">&quot;figure.dpi&quot;</span><span class="p">:</span> <span class="mi">100</span><span class="p">})</span>
<span class="n">sns</span><span class="o">.</span><span class="n">set_style</span><span class="p">(</span><span class="s2">&quot;whitegrid&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>The rpy2.ipython extension is already loaded. To reload it, use:
  %reload_ext rpy2.ipython
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">mdata</span> <span class="o">=</span> <span class="n">mu</span><span class="o">.</span><span class="n">read_10x_h5</span><span class="p">(</span><span class="s2">&quot;cellranger_out/filtered_feature_bc_matrix.h5&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>/Users/christopher.lance/mambaforge/envs/scatac_pp/lib/python3.9/site-packages/anndata/_core/anndata.py:1830: UserWarning: Variable names are not unique. To make them unique, call `.var_names_make_unique`.
  utils.warn_names_duplicates(&quot;var&quot;)
</pre></div>
</div>
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Added `interval` annotation for features from resources/cellranger_out/filtered_feature_bc_matrix.h5
</pre></div>
</div>
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>/Users/christopher.lance/mambaforge/envs/scatac_pp/lib/python3.9/site-packages/anndata/_core/anndata.py:1830: UserWarning: Variable names are not unique. To make them unique, call `.var_names_make_unique`.
  utils.warn_names_duplicates(&quot;var&quot;)
/Users/christopher.lance/mambaforge/envs/scatac_pp/lib/python3.9/site-packages/mudata/_core/mudata.py:446: UserWarning: var_names are not unique. To make them unique, call `.var_names_make_unique`.
  warnings.warn(
</pre></div>
</div>
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Added peak annotation from resources/cellranger_out/atac_peak_annotation.tsv to .uns[&#39;atac&#39;][&#39;peak_annotation&#39;]
Added gene names to peak annotation in .uns[&#39;atac&#39;][&#39;peak_annotation&#39;]
Located fragments file: resources/cellranger_out/atac_fragments.tsv.gz
</pre></div>
</div>
</div>
</div>
<p>As the warning messages indicate, peak identifiers are not unique. To prevent issues in downstream analysis steps, we make them unique using the <code class="docutils literal notranslate"><span class="pre">.var_names_make_unique()</span></code> method.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">mdata</span><span class="o">.</span><span class="n">var_names_make_unique</span><span class="p">()</span>
</pre></div>
</div>
</div>
</div>
<p>Let’s take a look into the <code class="docutils literal notranslate"><span class="pre">MuData</span></code> object we just generated.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">mdata</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><pre>MuData object with n_obs × n_vars = 16934 × 190628
  var:	&#x27;gene_ids&#x27;, &#x27;feature_types&#x27;, &#x27;genome&#x27;, &#x27;interval&#x27;
  2 modalities
    rna:	16934 x 36601
      var:	&#x27;gene_ids&#x27;, &#x27;feature_types&#x27;, &#x27;genome&#x27;, &#x27;interval&#x27;
    atac:	16934 x 154027
      var:	&#x27;gene_ids&#x27;, &#x27;feature_types&#x27;, &#x27;genome&#x27;, &#x27;interval&#x27;
      uns:	&#x27;atac&#x27;, &#x27;files&#x27;</pre></div></div>
</div>
<p>Both modalities (RNA and ATAC) have been loaded as two <code class="docutils literal notranslate"><span class="pre">AnnData</span></code> objects. In total 16934 cells were loaded and 36601 genes and 154027 peaks for RNA and ATAC, respectively. The ATAC modality additionally contains two entries in the unstructured data slot <code class="docutils literal notranslate"><span class="pre">atac.uns</span></code> that contain the peak annotation and the path to the fragment file automatically located by <code class="docutils literal notranslate"><span class="pre">muon</span></code>.</p>
<p>Let’s take a look at the <code class="docutils literal notranslate"><span class="pre">atac.uns</span></code> slot, printing the peak annotation and path to the fragment file.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">mdata</span><span class="o">.</span><span class="n">mod</span><span class="p">[</span><span class="s2">&quot;atac&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">uns</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>OverloadedDict, wrapping:
	OrderedDict([(&#39;atac&#39;, {&#39;peak_annotation&#39;:                                peak  distance peak_type
gene_name                                              
MIR1302-2HG         chr1:9763-10648    -18906    distal
AL627309.1       chr1:115270-116168      4764    distal
AL627309.5       chr1:181107-181768     -7246    distal
AL627309.5       chr1:183973-184845    -10112    distal
AL627309.5       chr1:191237-192111    -17376    distal
...                             ...       ...       ...
AC213203.2   KI270713.1:21361-22256     10272    distal
AC213203.2   KI270713.1:29645-30508      2020    distal
AC213203.2   KI270713.1:32401-33088         0  promoter
AC213203.1   KI270713.1:34369-35136      -271  promoter
AC213203.1   KI270713.1:36970-37884      1564    distal

[196633 rows x 3 columns]}), (&#39;files&#39;, {&#39;fragments&#39;: &#39;resources/cellranger_out/atac_fragments.tsv.gz&#39;})])
With overloaded keys:
	[&#39;neighbors&#39;].
</pre></div>
</div>
</div>
</div>
<p>The table generated from the peak annotation file gives a first idea which genes might be associated with a given peak. However, this is purely based on the distance to the closest gene and should be considered with care.</p>
<p>For subsequent processing we extract the <code class="docutils literal notranslate"><span class="pre">Anndata</span></code> object corresponding to the atac modality:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">atac</span> <span class="o">=</span> <span class="n">mdata</span><span class="o">.</span><span class="n">mod</span><span class="p">[</span><span class="s2">&quot;atac&quot;</span><span class="p">]</span>
</pre></div>
</div>
</div>
</div>
</section>
<section id="doublet-detection">
<span id="chromatin-accessibility-quality-control-takeaway-1"></span><h2><span class="section-number">24.3. </span>Doublet detection<a class="headerlink" href="#doublet-detection" title="Link to this heading">#</a></h2>
<p>The sparsity of scATAC data is considerably higher compared to scRNA-seq data. Therefore, it is not recommended to directly apply methods developed for scRNA-seq on scATAC-seq data. Here, we describe the use of two orthogonal methods for doublet scoring which can be used to detect clusters of cells containing doublets later on. Both approaches have been implemented in the R package scDoubletFinder <span id="id3">[<a class="reference internal" href="#id445" title="Pierre-Luc Germain, Aaron Lun, Carlos Garcia Meixide, Will Macnair, and Mark D Robinson. Doublet identification in single-cell sequencing data using scdblfinder. F1000Research, 2021.">Germain <em>et al.</em>, 2021</a>]</span> and the following steps were adapted from <a class="reference external" href="https://www.bioconductor.org/packages/devel/bioc/vignettes/scDblFinder/inst/doc/scATAC.html">this tutorial</a>.</p>
<ul class="simple">
<li><p><strong>Approach 1: Doublet Scoring based on simulated doublets:</strong> The native scDoubletFinder method is leveraging simulated doublets to assign doublet scores (see <a class="reference internal" href="../preprocessing_visualization/quality_control.html#rna-doublet-detection"><span class="std std-ref">Doublet Detection</span></a>). Prior to generating doublets, highly correlated features are aggregated to reduce sparsity and the number of features. As with methods for scRNA-seq data, this method detects heterotypic doublets (mix of different cell types).</p></li>
<li><p><strong>Approach 2: Coverage-based doublet scoring using AMULET</strong> <span id="id4">[<a class="reference internal" href="#id446" title="Asa Thibodeau, Alper Eroglu, Christopher S. McGinnis, Nathan Lawlor, Djamel Nehar-Belaid, Romy Kursawe, Radu Marches, Daniel N. Conrad, George A. Kuchel, Zev J. Gartner, Jacques Banchereau, Michael L. Stitzel, A. Ercument Cicek, and Duygu Ucar. AMULET: a novel read count-based method for effective multiplet detection from single nucleus ATAC-seq data. Genome Biology, 22(1):252, September 2021. URL: https://doi.org/10.1186/s13059-021-02469-x (visited on 2022-10-25), doi:10.1186/s13059-021-02469-x.">Thibodeau <em>et al.</em>, 2021</a>]</span>: Since DNA is only present as two copies in a diploid organism, one can expect a maximum count of two for any given position in the genome. AMULET makes use of this characteristic and evaluates the number of instances with more than two overlapping fragments for a given position (see figure below). An unexpected high number of such instances indicates a doublet (a small number of positions with more than two overlapping fragments is expected due to errors like duplication errors, sequencing and alignment errors or repetitive sequence content). AMULET performs best with a sufficient sequencing depth (&gt; 10-15k reads per cell) and can capture heterotypic and homotypic doublets (identical cell type).</p></li>
</ul>
<figure class="align-default" id="amulet">
<a class="bg-primary mb-1 reference internal image-reference" href="../_images/amulet.png"><img alt="Amulet" class="bg-primary mb-1" src="../_images/amulet.png" style="width: 800px;" />
</a>
<figcaption>
<p><span class="caption-number">Fig. 24.1 </span><span class="caption-text">Overview of AMULET approach.</span><a class="headerlink" href="#amulet" title="Link to this image">#</a></p>
</figcaption>
</figure>
<p>Before we run the doublet detection, let’s make sure R uses the correct path to our installed environment.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="o">%%</span><span class="k">R</span>
.libPaths()
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>[1] &quot;/Users/christopher.lance/mambaforge/envs/scatac_pp/lib/R/library&quot;
</pre></div>
</div>
</div>
</div>
<p>Load the <code class="docutils literal notranslate"><span class="pre">scDblFinder</span></code> and <code class="docutils literal notranslate"><span class="pre">SingleCellExperiment</span></code> package.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="o">%%</span><span class="k">R</span>
suppressPackageStartupMessages(library(scDblFinder))
suppressPackageStartupMessages(library(SingleCellExperiment))
</pre></div>
</div>
</div>
</div>
<section id="count-distribution-based-doublet-scoring">
<h3><span class="section-number">24.3.1. </span>Count distribution-based doublet scoring<a class="headerlink" href="#count-distribution-based-doublet-scoring" title="Link to this heading">#</a></h3>
<p>Since the doublet scoring can take some time we specify a path and sample identifier to save the output.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Set output paths</span>
<span class="n">save_path_dir</span> <span class="o">=</span> <span class="s2">&quot;output/doublet_scores/&quot;</span>
<span class="n">sample_ident</span> <span class="o">=</span> <span class="s2">&quot;s4d8&quot;</span>
</pre></div>
</div>
</div>
</div>
<p>To enable the transfer of our data matrix to an <code class="docutils literal notranslate"><span class="pre">SingleCellExperiment</span></code> R object, we transpose (<code class="docutils literal notranslate"><span class="pre">.T</span></code>) and convert the sparse matrix in the <code class="docutils literal notranslate"><span class="pre">atac.X</span></code> slot into an array (<code class="docutils literal notranslate"><span class="pre">.A</span></code>). Additionally we save the barcodes as a separate list.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">barcodes</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">atac</span><span class="o">.</span><span class="n">obs_names</span><span class="p">)</span>
<span class="n">data_mat</span> <span class="o">=</span> <span class="n">atac</span><span class="o">.</span><span class="n">X</span><span class="o">.</span><span class="n">T</span><span class="o">.</span><span class="n">A</span>
</pre></div>
</div>
</div>
</div>
<p>The <code class="docutils literal notranslate"><span class="pre">scDblFinder</span></code> methods works on a <code class="docutils literal notranslate"><span class="pre">SingleCellExperiment</span></code> object that we create from the our <code class="docutils literal notranslate"><span class="pre">data_mat</span></code> object. The user has to decide, whether doublet detection should happen cluster based (by using the <code class="docutils literal notranslate"><span class="pre">clusters</span></code> arguments) or using random cells. Cluster-based scoring is recommended if well-distinguishable cell populations are expected (e.g.PBMCs), while it is not recommended if continuous trajectories are expected. You can either provide pre-annotated clusters by specifying a vector of cluster labels for each cell, a metadata column of the <code class="docutils literal notranslate"><span class="pre">SingleCellExperiment</span></code> containing such labels, or <code class="docutils literal notranslate"><span class="pre">TRUE</span></code>. To not use clusters and randomly sample, one can set <code class="docutils literal notranslate"><span class="pre">clusters=FALSE</span></code> or <code class="docutils literal notranslate"><span class="pre">clusters=NULL</span></code>.
As working with all features (peaks or bins) is computationally very expensive, the authors suggest to aggregate correlated features into a small set <span id="id5">[<a class="reference internal" href="#id445" title="Pierre-Luc Germain, Aaron Lun, Carlos Garcia Meixide, Will Macnair, and Mark D Robinson. Doublet identification in single-cell sequencing data using scdblfinder. F1000Research, 2021.">Germain <em>et al.</em>, 2021</a>]</span>. This is enforced by setting <code class="docutils literal notranslate"><span class="pre">aggregateFeatures=TRUE</span></code>. This will aggregate peak or bin counts into <code class="docutils literal notranslate"><span class="pre">nfeatures</span></code> meta features. It is recommended to use a small number of features, we use 25. In a last step, the aggregated features are normalized <code class="docutils literal notranslate"><span class="pre">processing=&quot;normFeatures&quot;</span></code> before calculation of the doublet scores.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="o">%</span><span class="k">R</span> -i data_mat -o dbl_score sce &lt;- scDblFinder(SingleCellExperiment(list(counts=data_mat)), \
                                               <span class="n">clusters</span><span class="o">=</span><span class="n">TRUE</span><span class="p">,</span> <span class="n">aggregateFeatures</span><span class="o">=</span><span class="n">TRUE</span><span class="p">,</span> <span class="n">nfeatures</span><span class="o">=</span><span class="mi">25</span><span class="p">,</span> \
                                               <span class="n">processing</span><span class="o">=</span><span class="s2">&quot;normFeatures&quot;</span><span class="p">);</span> <span class="n">dbl_score</span> <span class="o">&lt;-</span> <span class="n">sce</span><span class="err">$</span><span class="n">scDblFinder</span><span class="o">.</span><span class="n">score</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>R[write to console]: dimnames(.) &lt;- NULL translated to
dimnames(.) &lt;- list(NULL,NULL)

R[write to console]: Aggregating features...

R[write to console]: Clustering cells...

R[write to console]: Warnung in (function (A, nv = 5, nu = nv, maxit = 1000, work = nv + 7, reorth = TRUE, 
R[write to console]: 
 
R[write to console]:  You&#39;re computing too large a percentage of total singular values, use a standard svd instead.

R[write to console]: 8 clusters

R[write to console]: Creating ~13548 artificial doublets...

R[write to console]: Dimensional reduction

R[write to console]: Evaluating kNN...

R[write to console]: Training model...

R[write to console]: iter=0, 2617 cells excluded from training.

R[write to console]: iter=1, 2725 cells excluded from training.

R[write to console]: iter=2, 2756 cells excluded from training.

R[write to console]: Threshold found:0.466

R[write to console]: 3038 (17.9%) doublets called

/Users/christopher.lance/mambaforge/envs/scatac_pp/lib/python3.9/site-packages/anndata2ri/r2py.py:106: FutureWarning: X.dtype being converted to np.float32 from float64. In the next version of anndata (0.9) conversion will not be automatic. Pass dtype explicitly to avoid this warning. Pass `AnnData(X, dtype=X.dtype, ...)` to get the future behavour.
  return AnnData(exprs, obs, var, uns, obsm or None, layers=layers)
</pre></div>
</div>
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>array([0.02819141, 0.00082514, 0.06341238, ..., 0.15534039, 0.02116382,
       0.12427134])
</pre></div>
</div>
</div>
</div>
<p>Based on the output messages about 18% of cells are suggested to be doublets. This might sound like a high fraction, but for the number of cells loaded for this sample and since isolated nuclei tend to stick together, this is reasonable.</p>
<p>Once the calculation is done, let’s create a <code class="docutils literal notranslate"><span class="pre">pandas.DataFrame</span></code> containing the scores for all barcodes and save it as a file.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">scDbl_result</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">({</span><span class="s2">&quot;barcodes&quot;</span><span class="p">:</span> <span class="n">barcodes</span><span class="p">,</span> <span class="s2">&quot;scDblFinder_score&quot;</span><span class="p">:</span> <span class="n">dbl_score</span><span class="p">})</span>
<span class="n">scDbl_result</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="n">save_path_dir</span> <span class="o">+</span> <span class="s2">&quot;/scDblFinder_scores_&quot;</span> <span class="o">+</span> <span class="n">sample_ident</span> <span class="o">+</span> <span class="s2">&quot;.csv&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">scDbl_result</span><span class="o">.</span><span class="n">head</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>barcodes</th>
      <th>scDblFinder_score</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>AAACAGCCAAGCTTAT-1</td>
      <td>0.028191</td>
    </tr>
    <tr>
      <th>1</th>
      <td>AAACAGCCATAGCTTG-1</td>
      <td>0.000825</td>
    </tr>
    <tr>
      <th>2</th>
      <td>AAACAGCCATGAAATG-1</td>
      <td>0.063412</td>
    </tr>
    <tr>
      <th>3</th>
      <td>AAACAGCCATGTTTGG-1</td>
      <td>0.026449</td>
    </tr>
    <tr>
      <th>4</th>
      <td>AAACATGCAACGTGCT-1</td>
      <td>0.019928</td>
    </tr>
  </tbody>
</table>
</div></div></div>
</div>
<p>To add the scores to our atac <code class="docutils literal notranslate"><span class="pre">AnnData</span></code> object, we set the barcodes as the index of the dataframe, which then matches the index of <code class="docutils literal notranslate"><span class="pre">adata.obs</span></code>.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">scDbl_result</span> <span class="o">=</span> <span class="n">scDbl_result</span><span class="o">.</span><span class="n">set_index</span><span class="p">(</span><span class="s2">&quot;barcodes&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<p>Now we can add the column to <code class="docutils literal notranslate"><span class="pre">adata.obs</span></code>.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">atac</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s2">&quot;scDblFinder_score&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">scDbl_result</span><span class="p">[</span><span class="s2">&quot;scDblFinder_score&quot;</span><span class="p">]</span>
</pre></div>
</div>
</div>
</div>
</section>
<section id="coverage-based-doublet-scoring">
<h3><span class="section-number">24.3.2. </span>Coverage-based doublet scoring<a class="headerlink" href="#coverage-based-doublet-scoring" title="Link to this heading">#</a></h3>
<p>As described before, AMULET estimates doublet scores based on the number of genomic positions with a fragment coverage higher two. Since this assumption is based on a diploid genome, we exclude mitochondrial genes (a cell can have many copies of the mitochondrial DNA) and sex chromosomes, as suggested by <span id="id6">[<a class="reference internal" href="#id446" title="Asa Thibodeau, Alper Eroglu, Christopher S. McGinnis, Nathan Lawlor, Djamel Nehar-Belaid, Romy Kursawe, Radu Marches, Daniel N. Conrad, George A. Kuchel, Zev J. Gartner, Jacques Banchereau, Michael L. Stitzel, A. Ercument Cicek, and Duygu Ucar. AMULET: a novel read count-based method for effective multiplet detection from single nucleus ATAC-seq data. Genome Biology, 22(1):252, September 2021. URL: https://doi.org/10.1186/s13059-021-02469-x (visited on 2022-10-25), doi:10.1186/s13059-021-02469-x.">Thibodeau <em>et al.</em>, 2021</a>]</span>. In addition, we want to mask repetitive sequences in the genome (e.g. interspersed nuclear elements or satellite repeats). To do so, we downloaded the <code class="docutils literal notranslate"><span class="pre">.bed</span></code> file provided with the AMULET publication available on <a class="reference external" href="https://zenodo.org/record/5189588#.ZDlrwexBxqs">Zenodo</a> and load it with the R package <code class="docutils literal notranslate"><span class="pre">rtracklayer</span></code>, which directly creates a <code class="docutils literal notranslate"><span class="pre">GRanges</span></code> R object. Note that for this sample, with about 16k cells, AMULET takes about 4 hours to run on a personal computer (the <code class="docutils literal notranslate"><span class="pre">.bed</span></code> file of repeat elements and the AMULET output is also provided on figshare).</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="o">%%</span><span class="k">R</span>

# Set up a GRanges objects of repeat elements, mitochondrial genes and sex chromosomes we want to exclude
suppressPackageStartupMessages(library(GenomicRanges))
suppressPackageStartupMessages(library(rtracklayer))

repeats =  import(&#39;resources/blacklist_repeats_segdups_rmsk_hg38.bed&#39;)
otherChroms &lt;- GRanges(c(&quot;chrM&quot;,&quot;chrX&quot;,&quot;chrY&quot;,&quot;MT&quot;),IRanges(1L,width=10^8)) # check which chromosome notation you are using c(&quot;M&quot;, &quot;X&quot;, &quot;Y&quot;, &quot;MT&quot;)
toExclude &lt;- suppressWarnings(c(repeats, otherChroms))
</pre></div>
</div>
</div>
</div>
<p>Furthermore, we need to define the path to the fragment file.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">frag_path</span> <span class="o">=</span> <span class="n">atac</span><span class="o">.</span><span class="n">uns</span><span class="p">[</span><span class="s2">&quot;files&quot;</span><span class="p">][</span><span class="s2">&quot;fragments&quot;</span><span class="p">]</span>
<span class="n">frag_path</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>&#39;resources/cellranger_out/atac_fragments.tsv.gz&#39;
</pre></div>
</div>
</div>
</div>
<p>Now we are ready to run AMULET. Note that we input objects from our python environment and define outputs we use afterwards with the <code class="docutils literal notranslate"><span class="pre">-i</span></code> and <code class="docutils literal notranslate"><span class="pre">-o</span></code> tag following the <code class="docutils literal notranslate"><span class="pre">%R</span></code> magic, while the <code class="docutils literal notranslate"><span class="pre">toExclude</span></code> object is already available to the <code class="docutils literal notranslate"><span class="pre">amulet</span></code> function, since we generated it inside the R environment above. (In case you have a larger memory available, check out the <code class="docutils literal notranslate"><span class="pre">fullInMemory=TRUE</span></code> option for faster runtime.)</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Run AMULET</span>
<span class="o">%</span><span class="k">R</span> -i frag_path -o amulet_result amulet_result &lt;- amulet(frag_path, regionsToExclude=toExclude)

<span class="c1"># Save output</span>
<span class="n">amulet_result</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="n">save_path_dir</span> <span class="o">+</span> <span class="s2">&quot;/AMULET_scores_&quot;</span> <span class="o">+</span> <span class="n">sample_ident</span> <span class="o">+</span> <span class="s2">&quot;.csv&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>R[write to console]: 18:01:17 - Reading Tabix-indexed fragment file and computing overlaps
</pre></div>
</div>
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>chr1, chr10, chr11, chr12, chr13, chr14, chr15, chr16, chr17, chr18, chr19, chr2, chr20, chr21, chr22, chr3, chr4, chr5, chr6, chr7, chr8, chr9, chrX, chrY, KI270728.1, KI270727.1, GL000009.2, GL000194.1, GL000205.2, GL000195.1, GL000219.1, KI270734.1, GL000213.1, GL000218.1, KI270731.1, KI270721.1, KI270726.1, KI270711.1, KI270713.1, 
</pre></div>
</div>
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>R[write to console]: 22:20:44 - Merging
</pre></div>
</div>
</div>
</div>
<p>Let’s have.a quick look at the output of AMULET.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">amulet_result</span><span class="o">.</span><span class="n">head</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>nFrags</th>
      <th>uniqFrags</th>
      <th>nAbove2</th>
      <th>total.nAbove2</th>
      <th>p.value</th>
      <th>q.value</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>AAACAGCCAAGCTTAT-1</th>
      <td>1696.0</td>
      <td>1696.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.994851</td>
      <td>0.994851</td>
    </tr>
    <tr>
      <th>AAACAGCCATGAAATG-1</th>
      <td>2536.0</td>
      <td>2536.0</td>
      <td>0.0</td>
      <td>1.0</td>
      <td>0.994851</td>
      <td>0.994851</td>
    </tr>
    <tr>
      <th>AAACAGCCATGTTTGG-1</th>
      <td>3464.0</td>
      <td>3464.0</td>
      <td>1.0</td>
      <td>6.0</td>
      <td>0.967723</td>
      <td>0.994851</td>
    </tr>
    <tr>
      <th>AAACATGCAACGTGCT-1</th>
      <td>3602.0</td>
      <td>3602.0</td>
      <td>0.0</td>
      <td>2.0</td>
      <td>0.994851</td>
      <td>0.994851</td>
    </tr>
    <tr>
      <th>AAACATGCAATATAGG-1</th>
      <td>3893.0</td>
      <td>3893.0</td>
      <td>1.0</td>
      <td>3.0</td>
      <td>0.967723</td>
      <td>0.994851</td>
    </tr>
  </tbody>
</table>
</div></div></div>
</div>
<p>For each barcode, we got the number of fragments, the number of sites covered by more than two fragments and the p- and q- values testing for the null hypothesis that a cell is a singlet. The lower the <code class="docutils literal notranslate"><span class="pre">q-value</span></code> the more likely the corresponding cell is a doublet. One could pick a common significance threshold to automatically define doublets, but we prefer to visualize the AMULET score later and look for populations of doublets manually.</p>
<p>We now add the scores to the <code class="docutils literal notranslate"><span class="pre">atac</span></code> object as well and transform them for nicer plotting.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">atac</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s2">&quot;AMULET_pVal&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">amulet_result</span><span class="p">[</span><span class="s2">&quot;p.value&quot;</span><span class="p">]</span>
<span class="n">atac</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s2">&quot;AMULET_qVal&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">amulet_result</span><span class="p">[</span><span class="s2">&quot;q.value&quot;</span><span class="p">]</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Transform q-values for nicer plotting</span>
<span class="n">atac</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s2">&quot;AMULET_negLog10qVal&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="o">-</span><span class="mi">1</span> <span class="o">*</span> <span class="n">np</span><span class="o">.</span><span class="n">log10</span><span class="p">(</span><span class="n">amulet_result</span><span class="p">[</span><span class="s2">&quot;q.value&quot;</span><span class="p">])</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>/Users/christopher.lance/mambaforge/envs/scatac_pp/lib/python3.9/site-packages/pandas/core/arraylike.py:402: RuntimeWarning: divide by zero encountered in log10
  result = getattr(ufunc, method)(*inputs, **kwargs)
</pre></div>
</div>
</div>
</div>
<p>Let us now plot the scores we derived from the two approaches.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">atac</span><span class="o">.</span><span class="n">obs</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">x</span><span class="o">=</span><span class="s2">&quot;scDblFinder_score&quot;</span><span class="p">,</span> <span class="n">y</span><span class="o">=</span><span class="s2">&quot;AMULET_negLog10qVal&quot;</span><span class="p">,</span> <span class="n">kind</span><span class="o">=</span><span class="s2">&quot;scatter&quot;</span><span class="p">)</span>

<span class="n">plt</span><span class="o">.</span><span class="n">title</span><span class="p">(</span><span class="s2">&quot;Association of doublet scores&quot;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="../_images/3003d95267fcb78a73d6ce701b5546c7cb97e5c83ec2f2785ea2d288255de348.png" src="../_images/3003d95267fcb78a73d6ce701b5546c7cb97e5c83ec2f2785ea2d288255de348.png" />
</div>
</div>
<p>The scatter plot shows a comparison of the adjusted P-values from AMULET and the scDblFinder scores for each barcode. Points rising towards the top right quadrant of the plot represent barcodes where both methods score the cells as doublets. Points in the bottom left quadrant of the plot indicate barcodes where both AMULET and scDblFinder scores indicate valid cells. In the bottom right quadrant, points correspond to barcodes identified as doublets only by scDblFinder based on the count distribution. There are few dots higher up in the left quadrant, which are called doublets by AMULET but not scDblFinder which may correspond to homotypic doublets (only detected by AMULET but not with count-based methods). We plan to utilize both scores in a later step to identify clusters of doublets.</p>
</section>
</section>
<section id="calculating-qc-metrics">
<h2><span class="section-number">24.4. </span>Calculating QC metrics<a class="headerlink" href="#calculating-qc-metrics" title="Link to this heading">#</a></h2>
<p>To detect low quality cells, we need to define metrics, that allow us to separate high quality from low-quality cells. Here is an overview of the primary QC metrics used across multiple scATAC-seq processing pipelines:</p>
<ul class="simple">
<li><p><strong>total_fragment_counts</strong>: Total number of fragments per cell representing cellular sequencing depth. This metric is analogous to the number of total counts in scRNA-seq data.</p></li>
<li><p><strong>tss_enrichment</strong>: Transcription start site (TSS) enrichment score, which is the ratio of fragments centered at the TSS to fragments in TSS-flanking regions. This metric can be interpreted as a signal-to noise ratio of each cell.</p></li>
<li><p><strong>n_features_per_cell</strong>: The number of peaks with non-zero counts in each cell. This metric is analogous to the number of genes detected in scRNA-seq data.</p></li>
<li><p><strong>nucleosome_signal</strong>: The nucleosome signal refers to the ratio of mono-nucleosomal to nucleosome-free fragments and can also be interpreted as a signal-to-noise ratio in each cell (more details below).</p></li>
</ul>
<p>Additional metrics that can be considered:</p>
<ul class="simple">
<li><p><strong>reads_in_peaks_frac:</strong> The fraction of fragments in peak regions versus fragments outside of peaks. Similar to the TSS score, this is an indicator for the signal-to-noise ratio.</p></li>
<li><p><strong>blacklist_fraction:</strong> The ratio of fragments in genomic blacklist regions, which have been associated with artefactual signal (defined by ENCODE). Benchmarking studies on processing of single-cell ATAC-seq data <span id="id7">[<a class="reference internal" href="#id444" title="Huidong Chen, Caleb Lareau, Tommaso Andreani, Michael E. Vinyard, Sara P. Garcia, Kendell Clement, Miguel A. Andrade-Navarro, Jason D. Buenrostro, and Luca Pinello. Assessment of computational methods for the analysis of single-cell ATAC-seq data. Genome Biology, 20(1):241, 2019. URL: https://genomebiology.biomedcentral.com/articles/10.1186/s13059-019-1854-5, doi:10.1186/s13059-019-1854-5.">Chen <em>et al.</em>, 2019</a>]</span> and based on our own experience, often the mapping of reads to blacklist regions is not a major issue single-cell data.</p></li>
</ul>
<p>Besides cell-level QC metrics, we also recommend to assess sample-level quality, for instance, by assessing the number of detected cells per sample and comparing the distributions of cell-level metrics. For the later, a violin plots grouped by sample is very informative.</p>
<section id="total-fragment-count-and-number-of-features">
<h3><span class="section-number">24.4.1. </span>Total fragment count and number of features<a class="headerlink" href="#total-fragment-count-and-number-of-features" title="Link to this heading">#</a></h3>
<p>To get the first metrics, we make use of the <code class="docutils literal notranslate"><span class="pre">calculate_qc_metrics</span></code> function in scanpy to calculate the total number of fragments and number of features per cell. Since scanpy labels the metrics according to RNA-seq data, we modify the variable names to make them fit to scATAC-seq data, i.e. <code class="docutils literal notranslate"><span class="pre">total_counts</span></code> becomes <code class="docutils literal notranslate"><span class="pre">total_fragment_counts</span></code> and <code class="docutils literal notranslate"><span class="pre">n_genes_by_counts</span></code> becomes the more general <code class="docutils literal notranslate"><span class="pre">n_features_per_cell</span></code>. Additionally, we log-transform the <code class="docutils literal notranslate"><span class="pre">total_fragment_counts</span></code> which is often helpful for plotting.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Calculate general qc metrics using scanpy</span>
<span class="n">sc</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">calculate_qc_metrics</span><span class="p">(</span><span class="n">atac</span><span class="p">,</span> <span class="n">percent_top</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">log1p</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">inplace</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>

<span class="c1"># Rename columns</span>
<span class="n">atac</span><span class="o">.</span><span class="n">obs</span><span class="o">.</span><span class="n">rename</span><span class="p">(</span>
    <span class="n">columns</span><span class="o">=</span><span class="p">{</span>
        <span class="s2">&quot;n_genes_by_counts&quot;</span><span class="p">:</span> <span class="s2">&quot;n_features_per_cell&quot;</span><span class="p">,</span>
        <span class="s2">&quot;total_counts&quot;</span><span class="p">:</span> <span class="s2">&quot;total_fragment_counts&quot;</span><span class="p">,</span>
    <span class="p">},</span>
    <span class="n">inplace</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
<span class="p">)</span>

<span class="c1"># log-transform total counts and add as column</span>
<span class="n">atac</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s2">&quot;log_total_fragment_counts&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">log10</span><span class="p">(</span><span class="n">atac</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s2">&quot;total_fragment_counts&quot;</span><span class="p">])</span>
</pre></div>
</div>
</div>
</div>
</section>
<section id="nucleosome-signal">
<h3><span class="section-number">24.4.2. </span>Nucleosome signal<a class="headerlink" href="#nucleosome-signal" title="Link to this heading">#</a></h3>
<p>Next, we calculate scATAC specific QC metrics, the nucleosome signal and TTS enrichment score.</p>
<p>For the nucleosome signal, the default number of fragments <code class="docutils literal notranslate"><span class="pre">n</span></code> per cell used to calculate the metric is 10e4*n_cells. Since this takes long to calculate, we reduce it by a factor of 10, which still returns a good estimate. However, in a production pipeline we would still recommend to keep the default.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Calculate the nucleosome signal across cells</span>
<span class="c1"># set n=10e3*atac.n_obs for rough estimate but faster run time</span>
<span class="n">ac</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">nucleosome_signal</span><span class="p">(</span><span class="n">atac</span><span class="p">,</span> <span class="n">n</span><span class="o">=</span><span class="mf">10e3</span> <span class="o">*</span> <span class="n">atac</span><span class="o">.</span><span class="n">n_obs</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Reading Fragments: 100%|██████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 169340000/169340000 [08:54&lt;00:00, 317005.72it/s]
</pre></div>
</div>
</div>
</div>
<p>To better understand the metric, let’s take a look at the overall distribution first.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">sns</span><span class="o">.</span><span class="n">histplot</span><span class="p">(</span><span class="n">atac</span><span class="o">.</span><span class="n">obs</span><span class="p">,</span> <span class="n">x</span><span class="o">=</span><span class="s2">&quot;nucleosome_signal&quot;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">title</span><span class="p">(</span><span class="s2">&quot;Distribution of the nucleome signal&quot;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>

<span class="c1"># Alternatively as a violin plot (uncomment to plot)</span>
<span class="c1"># sc.pl.violin(atac, &quot;nucleosome_signal&quot;)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="../_images/cfb3f82490532ba957662d770934ab00f87ce23d23d1f0637c0cbf34a0fc0717.png" src="../_images/cfb3f82490532ba957662d770934ab00f87ce23d23d1f0637c0cbf34a0fc0717.png" />
</div>
</div>
<p>The obtained scores in this data set range from 0 to 3. As a rule of thumb, previous analysis projects chose a threshold between 2 and 4 to label low quality cells. Let’s now take a closer look at cells with a high versus low nucleosome signal metric. To do so, we add a column to atac.obs, which contains a category for the two classes.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Add group labels for above and below the nucleosome signal threshold</span>
<span class="n">nuc_signal_threshold</span> <span class="o">=</span> <span class="mi">2</span>
<span class="n">atac</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s2">&quot;nuc_signal_filter&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="p">[</span>
    <span class="s2">&quot;NS_FAIL&quot;</span> <span class="k">if</span> <span class="n">ns</span> <span class="o">&gt;</span> <span class="n">nuc_signal_threshold</span> <span class="k">else</span> <span class="s2">&quot;NS_PASS&quot;</span>
    <span class="k">for</span> <span class="n">ns</span> <span class="ow">in</span> <span class="n">atac</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s2">&quot;nucleosome_signal&quot;</span><span class="p">]</span>
<span class="p">]</span>

<span class="c1"># Print number cells not passing nucleosome signal threshold</span>
<span class="n">atac</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s2">&quot;nuc_signal_filter&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">value_counts</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>NS_PASS    16877
NS_FAIL       57
Name: nuc_signal_filter, dtype: int64
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">atac</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s2">&quot;nuc_signal_filter&quot;</span><span class="p">]</span>  <span class="c1"># = atac.obs[&quot;nuc_signal_filter&quot;].astype(&#39;category&#39;)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>AAACAGCCAAGCTTAT-1    NS_PASS
AAACAGCCATAGCTTG-1    NS_PASS
AAACAGCCATGAAATG-1    NS_PASS
AAACAGCCATGTTTGG-1    NS_PASS
AAACATGCAACGTGCT-1    NS_PASS
                       ...   
TTTGTTGGTGGCTTCC-1    NS_PASS
TTTGTTGGTTCTTTAG-1    NS_PASS
TTTGTTGGTTGGCCGA-1    NS_PASS
TTTGTTGGTTTACTTG-1    NS_PASS
TTTGTTGGTTTGTGGA-1    NS_PASS
Name: nuc_signal_filter, Length: 16934, dtype: object
</pre></div>
</div>
</div>
</div>
<p>We can see, this sample only contains 57 barcodes with a higher nucleosome signal than 2.</p>
<p>Next, we plot the distribution of fragment lengths per group of cells. To speed up the generation of the plotting, we analyze fragments mapping to a certain region on chromosome 1 by setting <code class="docutils literal notranslate"><span class="pre">region=&quot;chr1:1-2000000&quot;</span></code>.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Plot fragment size distribution</span>
<span class="n">p1</span> <span class="o">=</span> <span class="n">ac</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">fragment_histogram</span><span class="p">(</span>
    <span class="n">atac</span><span class="p">[</span><span class="n">atac</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s2">&quot;nuc_signal_filter&quot;</span><span class="p">]</span> <span class="o">==</span> <span class="s2">&quot;NS_PASS&quot;</span><span class="p">],</span> <span class="n">region</span><span class="o">=</span><span class="s2">&quot;chr1:1-2000000&quot;</span>
<span class="p">)</span>

<span class="n">p2</span> <span class="o">=</span> <span class="n">ac</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">fragment_histogram</span><span class="p">(</span>
    <span class="n">atac</span><span class="p">[</span><span class="n">atac</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s2">&quot;nuc_signal_filter&quot;</span><span class="p">]</span> <span class="o">==</span> <span class="s2">&quot;NS_FAIL&quot;</span><span class="p">],</span> <span class="n">region</span><span class="o">=</span><span class="s2">&quot;chr1:1-2000000&quot;</span>
<span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Fetching Regions...: 100%|████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 1/1 [00:01&lt;00:00,  1.57s/it]
</pre></div>
</div>
<img alt="../_images/8e77745eb3a3ed81b906fff5976f6ee52f763be6af870f13f6c46fc54df25ad2.png" src="../_images/8e77745eb3a3ed81b906fff5976f6ee52f763be6af870f13f6c46fc54df25ad2.png" />
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Fetching Regions...: 100%|████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 1/1 [00:01&lt;00:00,  1.53s/it]
</pre></div>
</div>
<img alt="../_images/ac34cee7eb2963db7da3a6534556b6099acec5ff1cda959ba5dccd957e998257.png" src="../_images/ac34cee7eb2963db7da3a6534556b6099acec5ff1cda959ba5dccd957e998257.png" />
</div>
</div>
<p>In the first histograms, we see periodically decreasing peaks corresponding to fragment lengths in nucleosome-free regions (&lt;100bp), mono-, di- and a small tri-nucleosomal length. In high quality ATAC-seq data we expect an enrichment of fragments from nucleosome-free regions compared to fragments with mono- and multi-nucleosomal lengths (this ratio is calculated as the nucleosome signal). In the second plot this ratio sub-optimal, and we exclude these cells. However, the low number of cells with a high nucleosome signal indicate that we are working with data with overall good quality.</p>
</section>
<section id="tss-enrichment">
<h3><span class="section-number">24.4.3. </span>TSS enrichment<a class="headerlink" href="#tss-enrichment" title="Link to this heading">#</a></h3>
<p>The next QC metric we assess is the signal-to-noise ratio in each cell. To do so, one can calculate the fraction of reads mapped to peak regions or the enrichment of fragments around transcription start sites (TSS). Here we use the later for assessing cell quality. To obtain a robust TSS score estimates while keeping compute time reasonably low, we set the number of randomly selected TSS to calculate the score to <code class="docutils literal notranslate"><span class="pre">n_tss=3000</span></code> (compared to the default of 2000).</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">tss</span> <span class="o">=</span> <span class="n">ac</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">tss_enrichment</span><span class="p">(</span><span class="n">mdata</span><span class="p">,</span> <span class="n">n_tss</span><span class="o">=</span><span class="mi">3000</span><span class="p">,</span> <span class="n">random_state</span><span class="o">=</span><span class="mi">666</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Fetching Regions...: 100%|██████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 3000/3000 [00:44&lt;00:00, 67.37it/s]
</pre></div>
</div>
</div>
</div>
<p>This creates a new <code class="docutils literal notranslate"><span class="pre">AnnData</span></code> object <code class="docutils literal notranslate"><span class="pre">tss</span></code>, which contains positions from -1000 to +1000 around each TSS as features. Additionally, a <code class="docutils literal notranslate"><span class="pre">tss_score</span></code> column is added to the <code class="docutils literal notranslate"><span class="pre">atac.obs</span></code>  data frame.</p>
<p>Let’s check the distribution of the full range of scores and with extreme outliers removed (up to the 99.5% percentile).</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">fig</span><span class="p">,</span> <span class="n">axs</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">subplots</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">7</span><span class="p">,</span> <span class="mf">3.5</span><span class="p">))</span>

<span class="n">p1</span> <span class="o">=</span> <span class="n">sns</span><span class="o">.</span><span class="n">histplot</span><span class="p">(</span><span class="n">atac</span><span class="o">.</span><span class="n">obs</span><span class="p">,</span> <span class="n">x</span><span class="o">=</span><span class="s2">&quot;tss_score&quot;</span><span class="p">,</span> <span class="n">ax</span><span class="o">=</span><span class="n">axs</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span>
<span class="n">p1</span><span class="o">.</span><span class="n">set_title</span><span class="p">(</span><span class="s2">&quot;Full range&quot;</span><span class="p">)</span>

<span class="n">p2</span> <span class="o">=</span> <span class="n">sns</span><span class="o">.</span><span class="n">histplot</span><span class="p">(</span>
    <span class="n">atac</span><span class="o">.</span><span class="n">obs</span><span class="p">,</span>
    <span class="n">x</span><span class="o">=</span><span class="s2">&quot;tss_score&quot;</span><span class="p">,</span>
    <span class="n">binrange</span><span class="o">=</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="n">atac</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s2">&quot;tss_score&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">quantile</span><span class="p">(</span><span class="mf">0.995</span><span class="p">)),</span>
    <span class="n">ax</span><span class="o">=</span><span class="n">axs</span><span class="p">[</span><span class="mi">1</span><span class="p">],</span>
<span class="p">)</span>
<span class="n">p2</span><span class="o">.</span><span class="n">set_title</span><span class="p">(</span><span class="s2">&quot;Up to 99.5% percentile&quot;</span><span class="p">)</span>

<span class="n">plt</span><span class="o">.</span><span class="n">suptitle</span><span class="p">(</span><span class="s2">&quot;Distribution of the TSS score&quot;</span><span class="p">)</span>

<span class="n">plt</span><span class="o">.</span><span class="n">tight_layout</span><span class="p">()</span>
<span class="n">plt</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="../_images/4daf8f02897b1daca95808de1f82e6fed82522a9546d435399897be542ac1bce.png" src="../_images/4daf8f02897b1daca95808de1f82e6fed82522a9546d435399897be542ac1bce.png" />
</div>
</div>
<p>The histograms show that there are a few outliers with extremely high TSS scores and a small peak of barcodes with low TSS scores. Both are observations we would like to filter out.</p>
<p>To understand the TSS score better, we plot the number of transposition events (fragment ends) up- and downstream of transcription start sites. We make use of the plotting function <code class="docutils literal notranslate"><span class="pre">ac.pl.tss_enrichment</span></code> available in muon and input the newly created <code class="docutils literal notranslate"><span class="pre">tss</span></code> object. To compare the enrichment profile corresponding to high and low TSS scores, we add a new column to <code class="docutils literal notranslate"><span class="pre">tss.obs</span></code> containing  <code class="docutils literal notranslate"><span class="pre">&quot;TSS_PASS&quot;</span></code> or <code class="docutils literal notranslate"><span class="pre">&quot;TSS_FAIL&quot;</span></code> depending on a threshold that separates observations in the small peak of low scores we identified above.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">tss_threshold</span> <span class="o">=</span> <span class="mf">1.5</span>
<span class="n">tss</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s2">&quot;tss_filter&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="p">[</span>
    <span class="s2">&quot;TSS_FAIL&quot;</span> <span class="k">if</span> <span class="n">score</span> <span class="o">&lt;</span> <span class="n">tss_threshold</span> <span class="k">else</span> <span class="s2">&quot;TSS_PASS&quot;</span>
    <span class="k">for</span> <span class="n">score</span> <span class="ow">in</span> <span class="n">atac</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s2">&quot;tss_score&quot;</span><span class="p">]</span>
<span class="p">]</span>

<span class="c1"># Print number cells not passing nucleosome signal threshold</span>
<span class="n">tss</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s2">&quot;tss_filter&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">value_counts</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>TSS_PASS    16483
TSS_FAIL      451
Name: tss_filter, dtype: int64
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Temporarily set different color palette</span>
<span class="n">sns</span><span class="o">.</span><span class="n">set_palette</span><span class="p">(</span><span class="n">palette</span><span class="o">=</span><span class="s2">&quot;Set1&quot;</span><span class="p">)</span>
<span class="n">ac</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">tss_enrichment</span><span class="p">(</span><span class="n">tss</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="s2">&quot;tss_filter&quot;</span><span class="p">)</span>
<span class="c1"># reset color palette</span>
<span class="n">sns</span><span class="o">.</span><span class="n">set_palette</span><span class="p">(</span><span class="n">palette</span><span class="o">=</span><span class="s2">&quot;tab10&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>/Users/christopher.lance/mambaforge/envs/scatac_pp/lib/python3.9/site-packages/muon/_atac/plot.py:282: FutureWarning: In a future version of pandas, a length 1 tuple will be returned when iterating over a groupby with a grouper equal to a list of length 1. Don&#39;t supply a list with a single grouper to avoid this warning.
  for name, group in groups:
</pre></div>
</div>
<img alt="../_images/46c53e91befa03072caa73d6f3d4ce50cf72b481685126d7ced5b22f333c4595.png" src="../_images/46c53e91befa03072caa73d6f3d4ce50cf72b481685126d7ced5b22f333c4595.png" />
</div>
</div>
<p>We can see, how the 451 cells with low TSS scores show a much reduced enrichment of fragments around the TSS compared to high quality cells indicating a lower signal to noise ratio.</p>
<p>Let’s print the object we created so far and double check that all QC scores were added correctly.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">atac</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>AnnData object with n_obs × n_vars = 16934 × 154027
    obs: &#39;scDblFinder_score&#39;, &#39;AMULET_pVal&#39;, &#39;AMULET_qVal&#39;, &#39;AMULET_negLog10qVal&#39;, &#39;n_features_per_cell&#39;, &#39;total_fragment_counts&#39;, &#39;log_total_fragment_counts&#39;, &#39;nucleosome_signal&#39;, &#39;nuc_signal_filter&#39;, &#39;tss_score&#39;
    var: &#39;gene_ids&#39;, &#39;feature_types&#39;, &#39;genome&#39;, &#39;interval&#39;, &#39;n_cells_by_counts&#39;, &#39;mean_counts&#39;, &#39;pct_dropout_by_counts&#39;, &#39;total_counts&#39;
    uns: &#39;atac&#39;, &#39;files&#39;
</pre></div>
</div>
</div>
</div>
<p>Now, let’s save the object before we filter out low quality cells.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># save after calculation of QC metrics</span>
<span class="n">atac</span><span class="o">.</span><span class="n">write_h5ad</span><span class="p">(</span><span class="s2">&quot;output/atac_qc_metrics.h5ad&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
</section>
</section>
<section id="filtering-cells">
<span id="chromatin-accessibility-quality-control-takeaway-2"></span><h2><span class="section-number">24.5. </span>Filtering cells<a class="headerlink" href="#filtering-cells" title="Link to this heading">#</a></h2>
<p>The value range of QC metrics can vary from sample to sample. Therefore, we recommend plotting quality metrics in multiple ways to get a good feeling for sensible thresholds to filter out outliers and low quality cells.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Reload from file if needed</span>
<span class="c1"># atac = sc.read_h5ad(&quot;output/atac_qc_metrics.h5ad&quot;)</span>
</pre></div>
</div>
</div>
</div>
<section id="upper-bound-qc-thresholds">
<h3><span class="section-number">24.5.1. </span>Upper bound QC thresholds<a class="headerlink" href="#upper-bound-qc-thresholds" title="Link to this heading">#</a></h3>
<p>First, we plot the full range of QC metrics to identify barcodes with unrealistically high total fragment counts or TSS scores or undesired high nucleosome signals. We already added lines for suitable thresholds in the following plots which we identified by looking at the output before.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Set thresholds for upper boundaries.</span>
<span class="c1"># These were identified by looking at the plots in this code cell before.</span>
<span class="n">total_count_upper</span> <span class="o">=</span> <span class="mi">100000</span>
<span class="n">tss_upper</span> <span class="o">=</span> <span class="mi">50</span>
<span class="n">nucleosome_signal_upper</span> <span class="o">=</span> <span class="mi">2</span>


<span class="c1"># Plot total counts of fragments &amp; features colored by TSS score</span>
<span class="n">p1</span> <span class="o">=</span> <span class="n">sc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">scatter</span><span class="p">(</span>
    <span class="n">atac</span><span class="p">,</span>
    <span class="n">x</span><span class="o">=</span><span class="s2">&quot;total_fragment_counts&quot;</span><span class="p">,</span>
    <span class="n">y</span><span class="o">=</span><span class="s2">&quot;n_features_per_cell&quot;</span><span class="p">,</span>
    <span class="n">size</span><span class="o">=</span><span class="mi">40</span><span class="p">,</span>
    <span class="n">color</span><span class="o">=</span><span class="s2">&quot;tss_score&quot;</span><span class="p">,</span>
    <span class="n">show</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>  <span class="c1"># so that funstion output axis object where threshold line can be drawn.</span>
<span class="p">)</span>
<span class="n">p1</span><span class="o">.</span><span class="n">axvline</span><span class="p">(</span><span class="n">x</span><span class="o">=</span><span class="n">total_count_upper</span><span class="p">,</span> <span class="n">c</span><span class="o">=</span><span class="s2">&quot;red&quot;</span><span class="p">)</span>  <span class="c1"># Add vertical line</span>

<span class="c1"># tss.score</span>
<span class="n">p2</span> <span class="o">=</span> <span class="n">sc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">violin</span><span class="p">(</span><span class="n">atac</span><span class="p">,</span> <span class="s2">&quot;tss_score&quot;</span><span class="p">,</span> <span class="n">show</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
<span class="n">p2</span><span class="o">.</span><span class="n">set_ylim</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="mi">200</span><span class="p">)</span>  <span class="c1"># zooming in a little to</span>
<span class="n">p2</span><span class="o">.</span><span class="n">axhline</span><span class="p">(</span><span class="n">y</span><span class="o">=</span><span class="n">tss_upper</span><span class="p">,</span> <span class="n">c</span><span class="o">=</span><span class="s2">&quot;red&quot;</span><span class="p">)</span>  <span class="c1"># Add horizontal line</span>

<span class="c1"># nucleosome signal</span>
<span class="n">p3</span> <span class="o">=</span> <span class="n">sc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">violin</span><span class="p">(</span><span class="n">atac</span><span class="p">,</span> <span class="s2">&quot;nucleosome_signal&quot;</span><span class="p">,</span> <span class="n">show</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
<span class="n">p3</span><span class="o">.</span><span class="n">axhline</span><span class="p">(</span><span class="n">y</span><span class="o">=</span><span class="n">nucleosome_signal_upper</span><span class="p">,</span> <span class="n">c</span><span class="o">=</span><span class="s2">&quot;red&quot;</span><span class="p">)</span>

<span class="n">plt</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="../_images/dc2e65a65cc9b2af3fa2750882523b325150412187bcb5258173e42b467091b6.png" src="../_images/dc2e65a65cc9b2af3fa2750882523b325150412187bcb5258173e42b467091b6.png" />
<img alt="../_images/378cfb2810a008000e6de0a30ed55a69a01e1d2d3800b02d4bf4beda89c497f4.png" src="../_images/378cfb2810a008000e6de0a30ed55a69a01e1d2d3800b02d4bf4beda89c497f4.png" />
<img alt="../_images/3b074512868d2e83fe5fd8d5dabce2f6f882c6ae4f44d29c1c6318dc5d14fadc.png" src="../_images/3b074512868d2e83fe5fd8d5dabce2f6f882c6ae4f44d29c1c6318dc5d14fadc.png" />
</div>
</div>
<p>We define upper boundaries, such that only extreme outliers are excluded:</p>
<ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">total_count_upper</span> <span class="pre">=</span> <span class="pre">100000</span></code></p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">tss_upper</span> <span class="pre">=</span> <span class="pre">50</span></code></p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">nucleosome_signal_upper</span> <span class="pre">=</span> <span class="pre">2</span></code></p></li>
</ul>
<p>Note that the number of features is closely related by the number of total fragment counts and therefore extreme values for this QC metric excluded as well. Also, in principle a high TSS score is positive, but extremely high values are likely artifacts. This is also supported by the first scatter plot, where the highest TSS score corresponds to a barcode with very low fragment count and number of features.</p>
</section>
<section id="lower-bound-qc-thresholds">
<h3><span class="section-number">24.5.2. </span>Lower bound QC thresholds<a class="headerlink" href="#lower-bound-qc-thresholds" title="Link to this heading">#</a></h3>
<p>To identify the lower bounds of QC metrics, we create one of the most common visualizations in scATAC data to identify empty droplets and/or low quality cells. This is a scatter plot showing the TSS score over the log of the total counts.</p>
<p>We also added lines of suitable thresholds that we identified by looking at this and additional plots shown below.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># upper TSS score boundary for plotting</span>
<span class="n">plot_tss_max</span> <span class="o">=</span> <span class="mi">20</span>

<span class="c1"># Suggested thresholds (before log transform)</span>
<span class="n">count_cutoff_lower</span> <span class="o">=</span> <span class="mi">1500</span>
<span class="n">lcount_cutoff_upper</span> <span class="o">=</span> <span class="mi">100000</span>
<span class="n">tss_cutoff_lower</span> <span class="o">=</span> <span class="mf">1.5</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Scatter plot &amp; histograms</span>
<span class="n">g</span> <span class="o">=</span> <span class="n">sns</span><span class="o">.</span><span class="n">jointplot</span><span class="p">(</span>
    <span class="n">data</span><span class="o">=</span><span class="n">atac</span><span class="p">[(</span><span class="n">atac</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s2">&quot;tss_score&quot;</span><span class="p">]</span> <span class="o">&lt;</span> <span class="n">plot_tss_max</span><span class="p">)]</span><span class="o">.</span><span class="n">obs</span><span class="p">,</span>
    <span class="n">x</span><span class="o">=</span><span class="s2">&quot;log_total_fragment_counts&quot;</span><span class="p">,</span>
    <span class="n">y</span><span class="o">=</span><span class="s2">&quot;tss_score&quot;</span><span class="p">,</span>
    <span class="n">color</span><span class="o">=</span><span class="s2">&quot;black&quot;</span><span class="p">,</span>
    <span class="n">marker</span><span class="o">=</span><span class="s2">&quot;.&quot;</span><span class="p">,</span>
<span class="p">)</span>
<span class="c1"># Density plot including lines</span>
<span class="n">g</span><span class="o">.</span><span class="n">plot_joint</span><span class="p">(</span><span class="n">sns</span><span class="o">.</span><span class="n">kdeplot</span><span class="p">,</span> <span class="n">fill</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">cmap</span><span class="o">=</span><span class="s2">&quot;Blues&quot;</span><span class="p">,</span> <span class="n">zorder</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span> <span class="n">alpha</span><span class="o">=</span><span class="mf">0.75</span><span class="p">)</span>
<span class="n">g</span><span class="o">.</span><span class="n">plot_joint</span><span class="p">(</span><span class="n">sns</span><span class="o">.</span><span class="n">kdeplot</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="s2">&quot;black&quot;</span><span class="p">,</span> <span class="n">zorder</span><span class="o">=</span><span class="mi">2</span><span class="p">,</span> <span class="n">alpha</span><span class="o">=</span><span class="mf">0.75</span><span class="p">)</span>

<span class="c1"># Lines thresholds</span>
<span class="n">plt</span><span class="o">.</span><span class="n">axvline</span><span class="p">(</span><span class="n">x</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">log10</span><span class="p">(</span><span class="n">count_cutoff_lower</span><span class="p">),</span> <span class="n">c</span><span class="o">=</span><span class="s2">&quot;red&quot;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">axvline</span><span class="p">(</span><span class="n">x</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">log10</span><span class="p">(</span><span class="n">lcount_cutoff_upper</span><span class="p">),</span> <span class="n">c</span><span class="o">=</span><span class="s2">&quot;red&quot;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">axhline</span><span class="p">(</span><span class="n">y</span><span class="o">=</span><span class="n">tss_cutoff_lower</span><span class="p">,</span> <span class="n">c</span><span class="o">=</span><span class="s2">&quot;red&quot;</span><span class="p">)</span>

<span class="n">plt</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="../_images/6c862141cbec8d396b89e65e114ed46e27c22d35dca4db68408cbfe178f2ea5e.png" src="../_images/6c862141cbec8d396b89e65e114ed46e27c22d35dca4db68408cbfe178f2ea5e.png" />
</div>
</div>
<p>There is a distinct cluster of low quality cells in the bottom left corner that we filter out through the TSS score threshold of 1.5. Also, we cut the long left tail of the distribution of total counts per cell. Identifying an ideal threshold for this is sometimes not straightforward. Therefore, we suggest to generate additional plots that can help to find a suitable value.</p>
<p>As with scRNA data, we might need to optimize thresholds in an iterative process, after we went through downstream processing steps and observe too many low quality cells still being present or a cell type with an expected low number of counts/features missing.</p>
<p>For now, we generate additional histograms of the raw total count values and zoom into the lower tail of the distribution.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">fig</span><span class="p">,</span> <span class="n">axs</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">subplots</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">7</span><span class="p">,</span> <span class="mf">3.5</span><span class="p">))</span>

<span class="n">p1</span> <span class="o">=</span> <span class="n">sns</span><span class="o">.</span><span class="n">histplot</span><span class="p">(</span>
    <span class="n">atac</span><span class="o">.</span><span class="n">obs</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">atac</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s2">&quot;total_fragment_counts&quot;</span><span class="p">]</span> <span class="o">&lt;</span> <span class="mi">15000</span><span class="p">],</span>
    <span class="n">x</span><span class="o">=</span><span class="s2">&quot;total_fragment_counts&quot;</span><span class="p">,</span>
    <span class="n">bins</span><span class="o">=</span><span class="mi">40</span><span class="p">,</span>
    <span class="n">ax</span><span class="o">=</span><span class="n">axs</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span>
<span class="p">)</span>
<span class="n">p1</span><span class="o">.</span><span class="n">set_title</span><span class="p">(</span><span class="s2">&quot;&lt; 15000&quot;</span><span class="p">)</span>

<span class="n">p2</span> <span class="o">=</span> <span class="n">sns</span><span class="o">.</span><span class="n">histplot</span><span class="p">(</span>
    <span class="n">atac</span><span class="o">.</span><span class="n">obs</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">atac</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s2">&quot;total_fragment_counts&quot;</span><span class="p">]</span> <span class="o">&lt;</span> <span class="mi">3500</span><span class="p">],</span>
    <span class="n">x</span><span class="o">=</span><span class="s2">&quot;total_fragment_counts&quot;</span><span class="p">,</span>
    <span class="n">bins</span><span class="o">=</span><span class="mi">40</span><span class="p">,</span>
    <span class="n">ax</span><span class="o">=</span><span class="n">axs</span><span class="p">[</span><span class="mi">1</span><span class="p">],</span>
<span class="p">)</span>
<span class="n">p2</span><span class="o">.</span><span class="n">set_title</span><span class="p">(</span><span class="s2">&quot;&lt; 3500&quot;</span><span class="p">)</span>
<span class="n">p2</span><span class="o">.</span><span class="n">axvline</span><span class="p">(</span><span class="n">x</span><span class="o">=</span><span class="mi">1250</span><span class="p">,</span> <span class="n">c</span><span class="o">=</span><span class="s2">&quot;black&quot;</span><span class="p">,</span> <span class="n">linestyle</span><span class="o">=</span><span class="s2">&quot;--&quot;</span><span class="p">)</span>
<span class="n">p2</span><span class="o">.</span><span class="n">axvline</span><span class="p">(</span><span class="n">x</span><span class="o">=</span><span class="mi">1750</span><span class="p">,</span> <span class="n">c</span><span class="o">=</span><span class="s2">&quot;black&quot;</span><span class="p">,</span> <span class="n">linestyle</span><span class="o">=</span><span class="s2">&quot;--&quot;</span><span class="p">)</span>

<span class="n">plt</span><span class="o">.</span><span class="n">suptitle</span><span class="p">(</span><span class="s2">&quot;Total fragment count per cell&quot;</span><span class="p">)</span>

<span class="n">plt</span><span class="o">.</span><span class="n">tight_layout</span><span class="p">()</span>
<span class="n">plt</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="../_images/980138b20eaf37e8ddebba1615b11025d2347f4d304596d911c5fdfe25c3c7bd.png" src="../_images/980138b20eaf37e8ddebba1615b11025d2347f4d304596d911c5fdfe25c3c7bd.png" />
</div>
</div>
<p>We observe a peak of barcodes with a low number of total counts that likely represent low quality barcodes. Based on the histogram to the right we could choose a threshold of about 1250 to 1750 to remove this peak.</p>
<p>Let’s also get an idea for the number of features that barcodes with a total number of fragments in that range have. We added lines for a total fragment count of 1500, which represents an intermediate value of the suggested thresholds above.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Scatter plot total fragment count by number of features</span>

<span class="n">n_feature_cutoff</span> <span class="o">=</span> <span class="mi">750</span>  <span class="c1"># added after looking at this plot</span>

<span class="n">p2</span> <span class="o">=</span> <span class="n">sc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">scatter</span><span class="p">(</span>
    <span class="n">atac</span><span class="p">[</span><span class="n">atac</span><span class="o">.</span><span class="n">obs</span><span class="o">.</span><span class="n">total_fragment_counts</span> <span class="o">&lt;</span> <span class="mi">3500</span><span class="p">],</span>
    <span class="n">x</span><span class="o">=</span><span class="s2">&quot;total_fragment_counts&quot;</span><span class="p">,</span>
    <span class="n">y</span><span class="o">=</span><span class="s2">&quot;n_features_per_cell&quot;</span><span class="p">,</span>
    <span class="n">size</span><span class="o">=</span><span class="mi">100</span><span class="p">,</span>
    <span class="n">color</span><span class="o">=</span><span class="s2">&quot;tss_score&quot;</span><span class="p">,</span>
    <span class="n">show</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
<span class="p">)</span>
<span class="n">p2</span><span class="o">.</span><span class="n">axvline</span><span class="p">(</span><span class="n">x</span><span class="o">=</span><span class="n">count_cutoff_lower</span><span class="p">,</span> <span class="n">c</span><span class="o">=</span><span class="s2">&quot;red&quot;</span><span class="p">)</span>
<span class="n">p2</span><span class="o">.</span><span class="n">axhline</span><span class="p">(</span><span class="n">y</span><span class="o">=</span><span class="n">n_feature_cutoff</span><span class="p">,</span> <span class="n">c</span><span class="o">=</span><span class="s2">&quot;red&quot;</span><span class="p">)</span>

<span class="n">plt</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="../_images/1e4fb95537df3e14d221da7047f874ebbf16ececefc020548289e8a5ca9be051.png" src="../_images/1e4fb95537df3e14d221da7047f874ebbf16ececefc020548289e8a5ca9be051.png" />
</div>
</div>
<p>This scatter plot shows a nearly perfect linear relationship of the total fragment count with the number of features per cell. Considering the total number of features in our data set of more than 150000, 750 features seems like a permissive lower bound for filtering not risking to remove too many cells of interest. Taking all this together we decide on a lower total count threshold of 1500.</p>
</section>
<section id="perform-filtering">
<h3><span class="section-number">24.5.3. </span>Perform filtering<a class="headerlink" href="#perform-filtering" title="Link to this heading">#</a></h3>
<p>To summarize the outcome of our analysis thresholds:</p>
<ul class="simple">
<li><p>Total number of fragment counts per cell: &gt; 1500 and &lt; 100000</p></li>
<li><p>Number of features per cell: &gt; 750 (corresponding to the total fragment counts of 1500)</p></li>
<li><p>TSS score: &gt; 1.5 and &lt; 50</p></li>
<li><p>Nucleosome signal: &lt; 2</p></li>
</ul>
<p>Muon provides useful functionality to filter on any column in the <code class="docutils literal notranslate"><span class="pre">.obs</span></code> slot.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Total number of cells: </span><span class="si">{</span><span class="n">atac</span><span class="o">.</span><span class="n">n_obs</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
<span class="n">mu</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">filter_obs</span><span class="p">(</span>
    <span class="n">atac</span><span class="p">,</span>
    <span class="s2">&quot;total_fragment_counts&quot;</span><span class="p">,</span>
    <span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="p">(</span><span class="n">x</span> <span class="o">&gt;=</span> <span class="mi">1500</span><span class="p">)</span> <span class="o">&amp;</span> <span class="p">(</span><span class="n">x</span> <span class="o">&lt;=</span> <span class="mi">100000</span><span class="p">),</span>
<span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Number of cells after filtering on total_fragment_counts: </span><span class="si">{</span><span class="n">atac</span><span class="o">.</span><span class="n">n_obs</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
<span class="n">mu</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">filter_obs</span><span class="p">(</span><span class="n">atac</span><span class="p">,</span> <span class="s2">&quot;n_features_per_cell&quot;</span><span class="p">,</span> <span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">x</span> <span class="o">&gt;=</span> <span class="mi">750</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Number of cells after filtering on n_features_per_cell: </span><span class="si">{</span><span class="n">atac</span><span class="o">.</span><span class="n">n_obs</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Total number of cells: 16934
Number of cells after filtering on total_fragment_counts: 15386
Number of cells after filtering on n_features_per_cell: 15377
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">mu</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">filter_obs</span><span class="p">(</span>
    <span class="n">atac</span><span class="p">,</span>
    <span class="s2">&quot;tss_score&quot;</span><span class="p">,</span>
    <span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="p">(</span><span class="n">x</span> <span class="o">&gt;=</span> <span class="mf">1.5</span><span class="p">)</span> <span class="o">&amp;</span> <span class="p">(</span><span class="n">x</span> <span class="o">&lt;=</span> <span class="mi">50</span><span class="p">),</span>
<span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Number of cells after filtering on tss_score: </span><span class="si">{</span><span class="n">atac</span><span class="o">.</span><span class="n">n_obs</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
<span class="n">mu</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">filter_obs</span><span class="p">(</span><span class="n">atac</span><span class="p">,</span> <span class="s2">&quot;nucleosome_signal&quot;</span><span class="p">,</span> <span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">x</span> <span class="o">&lt;=</span> <span class="mi">2</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Number of cells after filtering on nucleosome_signal: </span><span class="si">{</span><span class="n">atac</span><span class="o">.</span><span class="n">n_obs</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Number of cells after filtering on tss_score: 15339
Number of cells after filtering on nucleosome_signal: 15286
</pre></div>
</div>
</div>
</div>
</section>
</section>
<section id="filtering-features">
<span id="chromatin-accessibility-quality-control-takeaway-3"></span><h2><span class="section-number">24.6. </span>Filtering features<a class="headerlink" href="#filtering-features" title="Link to this heading">#</a></h2>
<p>Lastly, we filter out features present in only a very low number of cells. These mostly add noise and computational resources needed in our downstream analysis.</p>
<p>Depending on the scATAC processing workflow filtering of features is done either on a minimum absolute number of cells, a percentage of cells or a total number of features (e.g. the most accessible ones).</p>
<p>Let us do a little thought experiment to decide on a suitable threshold of a minimum number of cells a feature should be present. Let’s say we would like to analyze a rare cell state, that makes up about 2% of our entire cell population. Considering that scATAC data has a high level of random drop outs, we assume features specific to this cell state to be detected in only 5% of the time. Therefore, on average we assume to detect this feature in about 5% of the 2% of the total number of cells. In our case with a data set of about 15000 cells, that would mean we expect cell state relevant features to be present in only 15 cells.</p>
<p>This threshold is also in the range of a general minimum of cell where a feature should be present and there are enough observations to allow downstream methods to pick up meaningful signals.</p>
<p>Here we choose this threshold, but keep in mind in case you are not interested in rare cell types or states, it might also improve your analysis to reduce the number of features to e.g. being present in at least 1% of the cells.</p>
<p>As with filtering cells, muon provides functionality to filter features as well.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">mu</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">filter_var</span><span class="p">(</span><span class="n">atac</span><span class="p">,</span> <span class="s2">&quot;n_cells_by_counts&quot;</span><span class="p">,</span> <span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">x</span> <span class="o">&gt;=</span> <span class="mi">15</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<p>To ensure we keep a version of the raw counts we save them as a separate layer.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">atac</span><span class="o">.</span><span class="n">layers</span><span class="p">[</span><span class="s2">&quot;counts&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">atac</span><span class="o">.</span><span class="n">X</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">atac</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>AnnData object with n_obs × n_vars = 15286 × 154015
    obs: &#39;scDblFinder_score&#39;, &#39;AMULET_pVal&#39;, &#39;AMULET_qVal&#39;, &#39;AMULET_negLog10qVal&#39;, &#39;n_features_per_cell&#39;, &#39;total_fragment_counts&#39;, &#39;log_total_fragment_counts&#39;, &#39;nucleosome_signal&#39;, &#39;nuc_signal_filter&#39;, &#39;tss_score&#39;
    var: &#39;gene_ids&#39;, &#39;feature_types&#39;, &#39;genome&#39;, &#39;interval&#39;, &#39;n_cells_by_counts&#39;, &#39;mean_counts&#39;, &#39;pct_dropout_by_counts&#39;, &#39;total_counts&#39;
    uns: &#39;atac&#39;, &#39;files&#39;
    layers: &#39;counts&#39;
</pre></div>
</div>
</div>
</div>
<p>Lastly, we save the filtered adata object for dimensionality reduction described in the next chapter.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">atac</span><span class="o">.</span><span class="n">write_h5ad</span><span class="p">(</span><span class="s2">&quot;output/atac_qc_filtered.h5ad&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
</section>
<section id="references">
<h2><span class="section-number">24.7. </span>References<a class="headerlink" href="#references" title="Link to this heading">#</a></h2>
<div class="docutils container" id="id8">
<div role="list" class="citation-list">
<div class="citation" id="id444" role="doc-biblioentry">
<span class="label"><span class="fn-bracket">[</span>atacCLA+19<span class="fn-bracket">]</span></span>
<span class="backrefs">(<a role="doc-backlink" href="#id1">1</a>,<a role="doc-backlink" href="#id7">2</a>)</span>
<p>Huidong Chen, Caleb Lareau, Tommaso Andreani, Michael E. Vinyard, Sara P. Garcia, Kendell Clement, Miguel A. Andrade-Navarro, Jason D. Buenrostro, and Luca Pinello. Assessment of computational methods for the analysis of single-cell ATAC-seq data. <em>Genome Biology</em>, 20(1):241, 2019. URL: <a class="reference external" href="https://genomebiology.biomedcentral.com/articles/10.1186/s13059-019-1854-5">https://genomebiology.biomedcentral.com/articles/10.1186/s13059-019-1854-5</a>, <a class="reference external" href="https://doi.org/10.1186/s13059-019-1854-5">doi:10.1186/s13059-019-1854-5</a>.</p>
</div>
<div class="citation" id="id445" role="doc-biblioentry">
<span class="label"><span class="fn-bracket">[</span>atacGLM+21<span class="fn-bracket">]</span></span>
<span class="backrefs">(<a role="doc-backlink" href="#id3">1</a>,<a role="doc-backlink" href="#id5">2</a>)</span>
<p>Pierre-Luc Germain, Aaron Lun, Carlos Garcia Meixide, Will Macnair, and Mark D Robinson. Doublet identification in single-cell sequencing data using scdblfinder. <em>F1000Research</em>, 2021.</p>
</div>
<div class="citation" id="id447" role="doc-biblioentry">
<span class="label"><span class="fn-bracket">[</span><a role="doc-backlink" href="#id2">atacLBC+21</a><span class="fn-bracket">]</span></span>
<p>Malte D Luecken, Daniel Bernard Burkhardt, Robrecht Cannoodt, Christopher Lance, Aditi Agrawal, Hananeh Aliee, Ann T Chen, Louise Deconinck, Angela M Detweiler, Alejandro A Granados, Shelly Huynh, Laura Isacco, Yang Joon Kim, Dominik Klein, BONY DE KUMAR, Sunil Kuppasani, Heiko Lickert, Aaron McGeever, Honey Mekonen, Joaquin Caceres Melgarejo, Maurizio Morri, Michaela Müller, Norma Neff, Sheryl Paul, Bastian Rieck, Kaylie Schneider, Scott Steelman, Michael Sterr, Daniel J. Treacy, Alexander Tong, Alexandra-Chloe Villani, Guilin Wang, Jia Yan, Ce Zhang, Angela Oliveira Pisco, Smita Krishnaswamy, Fabian J Theis, and Jonathan M. Bloom. A sandbox for prediction and integration of dna, rna, and proteins in single cells. In <em>Thirty-fifth Conference on Neural Information Processing Systems Datasets and Benchmarks Track (Round 2)</em>. 2021. URL: <a class="reference external" href="https://openreview.net/forum?id=gN35BGa1Rt">https://openreview.net/forum?id=gN35BGa1Rt</a>.</p>
</div>
<div class="citation" id="id446" role="doc-biblioentry">
<span class="label"><span class="fn-bracket">[</span>atacTEM+21<span class="fn-bracket">]</span></span>
<span class="backrefs">(<a role="doc-backlink" href="#id4">1</a>,<a role="doc-backlink" href="#id6">2</a>)</span>
<p>Asa Thibodeau, Alper Eroglu, Christopher S. McGinnis, Nathan Lawlor, Djamel Nehar-Belaid, Romy Kursawe, Radu Marches, Daniel N. Conrad, George A. Kuchel, Zev J. Gartner, Jacques Banchereau, Michael L. Stitzel, A. Ercument Cicek, and Duygu Ucar. AMULET: a novel read count-based method for effective multiplet detection from single nucleus ATAC-seq data. <em>Genome Biology</em>, 22(1):252, September 2021. URL: <a class="reference external" href="https://doi.org/10.1186/s13059-021-02469-x">https://doi.org/10.1186/s13059-021-02469-x</a> (visited on 2022-10-25), <a class="reference external" href="https://doi.org/10.1186/s13059-021-02469-x">doi:10.1186/s13059-021-02469-x</a>.</p>
</div>
</div>
</div>
</section>
<section id="contributors">
<h2><span class="section-number">24.8. </span>Contributors<a class="headerlink" href="#contributors" title="Link to this heading">#</a></h2>
<p>We gratefully acknowledge the contributions of:</p>
<section id="authors">
<h3><span class="section-number">24.8.1. </span>Authors<a class="headerlink" href="#authors" title="Link to this heading">#</a></h3>
<ul class="simple">
<li><p>Christopher Lance</p></li>
<li><p>Laura Martens</p></li>
</ul>
</section>
<section id="reviewers">
<h3><span class="section-number">24.8.2. </span>Reviewers<a class="headerlink" href="#reviewers" title="Link to this heading">#</a></h3>
<ul class="simple">
<li><p>Lukas Heumos</p></li>
<li><p>Anna Schaar</p></li>
</ul>
</section>
</section>
</section>

    <script type="text/x-thebe-config">
    {
        requestKernel: true,
        binderOptions: {
            repo: "binder-examples/jupyter-stacks-datascience",
            ref: "master",
        },
        codeMirrorConfig: {
            theme: "abcdef",
            mode: "python"
        },
        kernelOptions: {
            name: "python3",
            path: "./chromatin_accessibility"
        },
        predefinedOutput: true
    }
    </script>
    <script>kernelName = 'python3'</script>

                </article>
              

              
              
              
              
                <footer class="prev-next-footer d-print-none">
                  
<div class="prev-next-area">
    <a class="left-prev"
       href="introduction.html"
       title="previous page">
      <i class="fa-solid fa-angle-left"></i>
      <div class="prev-next-info">
        <p class="prev-next-subtitle">previous</p>
        <p class="prev-next-title"><span class="section-number">23. </span>Single-cell ATAC sequencing</p>
      </div>
    </a>
    <a class="right-next"
       href="gene_regulatory_networks_atac.html"
       title="next page">
      <div class="prev-next-info">
        <p class="prev-next-subtitle">next</p>
        <p class="prev-next-title"><span class="section-number">25. </span>Gene regulatory networks using chromatin accessibility</p>
      </div>
      <i class="fa-solid fa-angle-right"></i>
    </a>
</div>
                </footer>
              
            </div>
            
            
              
                <div class="bd-sidebar-secondary bd-toc"><div class="sidebar-secondary-items sidebar-secondary__inner">


  <div class="sidebar-secondary-item">
  <div class="page-toc tocsection onthispage">
    <i class="fa-solid fa-list"></i> Contents
  </div>
  <nav class="bd-toc-nav page-toc">
    <ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#motivation">24.1. Motivation</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#dataset">24.2. Dataset</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#doublet-detection">24.3. Doublet detection</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#count-distribution-based-doublet-scoring">24.3.1. Count distribution-based doublet scoring</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#coverage-based-doublet-scoring">24.3.2. Coverage-based doublet scoring</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#calculating-qc-metrics">24.4. Calculating QC metrics</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#total-fragment-count-and-number-of-features">24.4.1. Total fragment count and number of features</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#nucleosome-signal">24.4.2. Nucleosome signal</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#tss-enrichment">24.4.3. TSS enrichment</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#filtering-cells">24.5. Filtering cells</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#upper-bound-qc-thresholds">24.5.1. Upper bound QC thresholds</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#lower-bound-qc-thresholds">24.5.2. Lower bound QC thresholds</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#perform-filtering">24.5.3. Perform filtering</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#filtering-features">24.6. Filtering features</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#references">24.7. References</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#contributors">24.8. Contributors</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#authors">24.8.1. Authors</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#reviewers">24.8.2. Reviewers</a></li>
</ul>
</li>
</ul>
  </nav></div>

</div></div>
              
            
          </div>
          <footer class="bd-footer-content">
            
<div class="bd-footer-content__inner container">
  
  <div class="footer-item">
    
<p class="component-author">
By Lukas Heumos, Anna Schaar, single-cell best practices consortium
</p>

  </div>
  
  <div class="footer-item">
    

  <p class="copyright">
    
      © Copyright 2023.
      <br/>
    
  </p>

  </div>
  
  <div class="footer-item">
    
  </div>
  
  <div class="footer-item">
    
<div class="extra_footer">
  <div>
  Brought to you by
  <a href="https://www.github.com/theislab">Theislab</a>, with many thanks to the single-cell community as a whole!
</div>

</div>
  </div>
  
</div>
          </footer>
        

      </main>
    </div>
  </div>
  
  <!-- Scripts loaded after <body> so the DOM is not blocked -->
  <script src="../_static/scripts/bootstrap.js?digest=dfe6caa3a7d634c4db9b"></script>
<script src="../_static/scripts/pydata-sphinx-theme.js?digest=dfe6caa3a7d634c4db9b"></script>

  <footer class="bd-footer">
  </footer>
  </body>
</html>