
<!DOCTYPE html>


<html lang="en" data-content_root="../" >

  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="viewport" content="width=device-width, initial-scale=1" />

    <title>39. Clonotype analysis &#8212; Single-cell best practices</title>
  
  
  
  <script data-cfasync="false">
    document.documentElement.dataset.mode = localStorage.getItem("mode") || "";
    document.documentElement.dataset.theme = localStorage.getItem("theme") || "";
  </script>
  
  <!-- Loaded before other Sphinx assets -->
  <link href="../_static/styles/theme.css?digest=dfe6caa3a7d634c4db9b" rel="stylesheet" />
<link href="../_static/styles/bootstrap.css?digest=dfe6caa3a7d634c4db9b" rel="stylesheet" />
<link href="../_static/styles/pydata-sphinx-theme.css?digest=dfe6caa3a7d634c4db9b" rel="stylesheet" />

  
  <link href="../_static/vendor/fontawesome/6.5.2/css/all.min.css?digest=dfe6caa3a7d634c4db9b" rel="stylesheet" />
  <link rel="preload" as="font" type="font/woff2" crossorigin href="../_static/vendor/fontawesome/6.5.2/webfonts/fa-solid-900.woff2" />
<link rel="preload" as="font" type="font/woff2" crossorigin href="../_static/vendor/fontawesome/6.5.2/webfonts/fa-brands-400.woff2" />
<link rel="preload" as="font" type="font/woff2" crossorigin href="../_static/vendor/fontawesome/6.5.2/webfonts/fa-regular-400.woff2" />

    <link rel="stylesheet" type="text/css" href="../_static/pygments.css?v=03e43079" />
    <link rel="stylesheet" type="text/css" href="../_static/styles/sphinx-book-theme.css?v=eba8b062" />
    <link rel="stylesheet" type="text/css" href="../_static/togglebutton.css?v=13237357" />
    <link rel="stylesheet" type="text/css" href="../_static/copybutton.css?v=76b2166b" />
    <link rel="stylesheet" type="text/css" href="../_static/mystnb.4510f1fc1dee50b3e5859aac5469c37c29e427902b24a333a5f9fcb2f0b3ac41.css" />
    <link rel="stylesheet" type="text/css" href="../_static/sphinx-thebe.css?v=4fa983c6" />
    <link rel="stylesheet" type="text/css" href="../_static/sphinx-design.min.css?v=95c83b7e" />
    <link rel="stylesheet" type="text/css" href="../_static/book.css?v=148c11e2" />
  
  <!-- Pre-loaded scripts that we'll load fully later -->
  <link rel="preload" as="script" href="../_static/scripts/bootstrap.js?digest=dfe6caa3a7d634c4db9b" />
<link rel="preload" as="script" href="../_static/scripts/pydata-sphinx-theme.js?digest=dfe6caa3a7d634c4db9b" />
  <script src="../_static/vendor/fontawesome/6.5.2/js/all.min.js?digest=dfe6caa3a7d634c4db9b"></script>

    <script src="../_static/documentation_options.js?v=9eb32ce0"></script>
    <script src="../_static/doctools.js?v=9a2dae69"></script>
    <script src="../_static/sphinx_highlight.js?v=dc90522c"></script>
    <script src="../_static/clipboard.min.js?v=a7894cd8"></script>
    <script src="../_static/copybutton.js?v=f281be69"></script>
    <script src="../_static/scripts/sphinx-book-theme.js?v=887ef09a"></script>
    <script>let toggleHintShow = 'Click to show';</script>
    <script>let toggleHintHide = 'Click to hide';</script>
    <script>let toggleOpenOnPrint = 'true';</script>
    <script src="../_static/togglebutton.js?v=4a39c7ea"></script>
    <script>var togglebuttonSelector = '.toggle, .admonition.dropdown';</script>
    <script src="../_static/design-tabs.js?v=f930bc37"></script>
    <script>const THEBE_JS_URL = "https://unpkg.com/thebe@0.8.2/lib/index.js"; const thebe_selector = ".thebe,.cell"; const thebe_selector_input = "pre"; const thebe_selector_output = ".output, .cell_output"</script>
    <script async="async" src="../_static/sphinx-thebe.js?v=c100c467"></script>
    <script>var togglebuttonSelector = '.toggle, .admonition.dropdown';</script>
    <script>const THEBE_JS_URL = "https://unpkg.com/thebe@0.8.2/lib/index.js"; const thebe_selector = ".thebe,.cell"; const thebe_selector_input = "pre"; const thebe_selector_output = ".output, .cell_output"</script>
    <script>DOCUMENTATION_OPTIONS.pagename = 'air_repertoire/clonotype';</script>
    <script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
    <script src="../_static/book.js?v=57d67cd3"></script>
    <link rel="icon" href="../_static/favicon.ico"/>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="43. Specificity analysis" href="specificity.html" />
    <link rel="prev" title="38. Immune Receptor Profiling" href="ir_profiling.html" />
  <meta name="viewport" content="width=device-width, initial-scale=1"/>
  <meta name="docsearch:language" content="en"/>
  </head>
  
  
  <body data-bs-spy="scroll" data-bs-target=".bd-toc-nav" data-offset="180" data-bs-root-margin="0px 0px -60%" data-default-mode="">

  
  
  <div id="pst-skip-link" class="skip-link d-print-none"><a href="#main-content">Skip to main content</a></div>
  
  <div id="pst-scroll-pixel-helper"></div>
  
  <button type="button" class="btn rounded-pill" id="pst-back-to-top">
    <i class="fa-solid fa-arrow-up"></i>Back to top</button>

  
  <input type="checkbox"
          class="sidebar-toggle"
          id="pst-primary-sidebar-checkbox"/>
  <label class="overlay overlay-primary" for="pst-primary-sidebar-checkbox"></label>
  
  <input type="checkbox"
          class="sidebar-toggle"
          id="pst-secondary-sidebar-checkbox"/>
  <label class="overlay overlay-secondary" for="pst-secondary-sidebar-checkbox"></label>
  
  <div class="search-button__wrapper">
    <div class="search-button__overlay"></div>
    <div class="search-button__search-container">
<form class="bd-search d-flex align-items-center"
      action="../search.html"
      method="get">
  <i class="fa-solid fa-magnifying-glass"></i>
  <input type="search"
         class="form-control"
         name="q"
         id="search-input"
         placeholder="Search this book..."
         aria-label="Search this book..."
         autocomplete="off"
         autocorrect="off"
         autocapitalize="off"
         spellcheck="false"/>
  <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd>K</kbd></span>
</form></div>
  </div>

  <div class="pst-async-banner-revealer d-none">
  <aside id="bd-header-version-warning" class="d-none d-print-none" aria-label="Version warning"></aside>
</div>

  
    <header class="bd-header navbar navbar-expand-lg bd-navbar d-print-none">
    </header>
  

  <div class="bd-container">
    <div class="bd-container__inner bd-page-width">
      
      
      
      <div class="bd-sidebar-primary bd-sidebar">
        

  
  <div class="sidebar-header-items sidebar-primary__section">
    
    
    
    
  </div>
  
    <div class="sidebar-primary-items__start sidebar-primary__section">
        <div class="sidebar-primary-item">

  
    
  

<a class="navbar-brand logo" href="../preamble.html">
  
  
  
  
  
    
    
      
    
    
    <img src="../_static/logo.png" class="logo__image only-light" alt="Single-cell best practices - Home"/>
    <script>document.write(`<img src="../_static/logo.png" class="logo__image only-dark" alt="Single-cell best practices - Home"/>`);</script>
  
  
</a></div>
        <div class="sidebar-primary-item">

 <script>
 document.write(`
   <button class="btn search-button-field search-button__button" title="Search" aria-label="Search" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="fa-solid fa-magnifying-glass"></i>
    <span class="search-button__default-text">Search</span>
    <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd class="kbd-shortcut__modifier">K</kbd></span>
   </button>
 `);
 </script></div>
        <div class="sidebar-primary-item"><nav class="bd-links bd-docs-nav" aria-label="Main">
    <div class="bd-toc-item navbar-nav active">
        <p aria-level="2" class="caption" role="heading"><span class="caption-text">Introduction</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../introduction/prior_art.html">1. Prior art</a></li>
<li class="toctree-l1"><a class="reference internal" href="../introduction/scrna_seq.html">2. Single-cell RNA sequencing</a></li>
<li class="toctree-l1"><a class="reference internal" href="../introduction/raw_data_processing.html">3. Raw data processing</a></li>
<li class="toctree-l1"><a class="reference internal" href="../introduction/analysis_tools.html">4. Analysis frameworks and tools</a></li>
<li class="toctree-l1"><a class="reference internal" href="../introduction/interoperability.html">5. Interoperability</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Preprocessing and visualization</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../preprocessing_visualization/quality_control.html">6. Quality Control</a></li>
<li class="toctree-l1"><a class="reference internal" href="../preprocessing_visualization/normalization.html">7. Normalization</a></li>
<li class="toctree-l1"><a class="reference internal" href="../preprocessing_visualization/feature_selection.html">8. Feature selection</a></li>
<li class="toctree-l1"><a class="reference internal" href="../preprocessing_visualization/dimensionality_reduction.html">9. Dimensionality Reduction</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Identifying cellular structure</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../cellular_structure/clustering.html">10. Clustering</a></li>
<li class="toctree-l1"><a class="reference internal" href="../cellular_structure/annotation.html">11. Annotation</a></li>
<li class="toctree-l1"><a class="reference internal" href="../cellular_structure/integration.html">12. Data integration</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Inferring trajectories</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../trajectories/pseudotemporal.html">13. Pseudotemporal ordering</a></li>
<li class="toctree-l1"><a class="reference internal" href="../trajectories/rna_velocity.html">14. RNA velocity</a></li>
<li class="toctree-l1"><a class="reference internal" href="../trajectories/lineage_tracing.html">15. Lineage tracing</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Dealing with conditions</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../conditions/differential_gene_expression.html">16. Differential gene expression analysis</a></li>
<li class="toctree-l1"><a class="reference internal" href="../conditions/compositional.html">17. Compositional analysis</a></li>
<li class="toctree-l1"><a class="reference internal" href="../conditions/gsea_pathway.html">18. Gene set enrichment and pathway analysis</a></li>
<li class="toctree-l1"><a class="reference internal" href="../conditions/perturbation_modeling.html">19. Perturbation modeling</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Modeling mechanisms</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../mechanisms/gene_regulatory_networks.html">20. Gene regulatory networks</a></li>
<li class="toctree-l1"><a class="reference internal" href="../mechanisms/cell_cell_communication.html">21. Cell-cell communication</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Deconvolution</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../deconvolution/bulk_deconvolution.html">22. Bulk deconvolution</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Chromatin Accessibility</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../chromatin_accessibility/introduction.html">23. Single-cell ATAC sequencing</a></li>
<li class="toctree-l1"><a class="reference internal" href="../chromatin_accessibility/quality_control.html">24. Quality Control</a></li>
<li class="toctree-l1"><a class="reference internal" href="../chromatin_accessibility/gene_regulatory_networks_atac.html">25. Gene regulatory networks using chromatin accessibility</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Spatial omics</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../spatial/introduction.html">26. Single-cell data resolved in space</a></li>
<li class="toctree-l1"><a class="reference internal" href="../spatial/neighborhood.html">27. Neighborhood analysis</a></li>
<li class="toctree-l1"><a class="reference internal" href="../spatial/domains.html">28. Spatial domains</a></li>
<li class="toctree-l1"><a class="reference internal" href="../spatial/spatially_variable_genes.html">29. Spatially variable genes</a></li>
<li class="toctree-l1"><a class="reference internal" href="../spatial/deconvolution.html">30. Spatial deconvolution</a></li>
<li class="toctree-l1"><a class="reference internal" href="../spatial/imputation.html">31. Imputation</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Surface protein</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../surface_protein/quality_control.html">32. Quality control</a></li>
<li class="toctree-l1"><a class="reference internal" href="../surface_protein/normalization.html">33. Normalization</a></li>
<li class="toctree-l1"><a class="reference internal" href="../surface_protein/doublet_detection.html">34. Doublet detection</a></li>
<li class="toctree-l1"><a class="reference internal" href="../surface_protein/dimensionality_reduction.html">35. Dimensionality Reduction</a></li>
<li class="toctree-l1"><a class="reference internal" href="../surface_protein/batch_correction.html">36. Batch correction</a></li>
<li class="toctree-l1"><a class="reference internal" href="../surface_protein/annotation.html">37. Annotation</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Adaptive immune receptor repertoire</span></p>
<ul class="current nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="ir_profiling.html">38. Immune Receptor Profiling</a></li>
<li class="toctree-l1 current active"><a class="current reference internal" href="#">39. Clonotype analysis</a></li>



<li class="toctree-l1"><a class="reference internal" href="specificity.html">43. Specificity analysis</a></li>
<li class="toctree-l1"><a class="reference internal" href="multimodal_integration.html">44. Integrating AIR and transcriptomics</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Multimodal integration</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../multimodal_integration/paired_integration.html">45. Paired integration</a></li>
<li class="toctree-l1"><a class="reference internal" href="../multimodal_integration/advanced_integration.html">46. Advanced integration</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Outlook</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../outlook.html">47. Outlook</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Acknowledgements</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../acknowledgements.html">48. Acknowledgements</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Glossary</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../glossary.html">49. Glossary</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Changelog</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../CHANGELOG.html">Changelog</a></li>
</ul>

    </div>
</nav></div>
    </div>
  
  
  <div class="sidebar-primary-items__end sidebar-primary__section">
  </div>
  
  <div id="rtd-footer-container"></div>


      </div>
      
      <main id="main-content" class="bd-main" role="main">
        
        

<div class="sbt-scroll-pixel-helper"></div>

          <div class="bd-content">
            <div class="bd-article-container">
              
              <div class="bd-header-article d-print-none">
<div class="header-article-items header-article__inner">
  
    <div class="header-article-items__start">
      
        <div class="header-article-item"><button class="sidebar-toggle primary-toggle btn btn-sm" title="Toggle primary sidebar" data-bs-placement="bottom" data-bs-toggle="tooltip">
  <span class="fa-solid fa-bars"></span>
</button></div>
      
    </div>
  
  
    <div class="header-article-items__end">
      
        <div class="header-article-item">

<div class="article-header-buttons">





<div class="dropdown dropdown-source-buttons">
  <button class="btn dropdown-toggle" type="button" data-bs-toggle="dropdown" aria-expanded="false" aria-label="Source repositories">
    <i class="fab fa-github"></i>
  </button>
  <ul class="dropdown-menu">
      
      
      
      <li><a href="https://github.com/theislab/single-cell-best-practices" target="_blank"
   class="btn btn-sm btn-source-repository-button dropdown-item"
   title="Source repository"
   data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fab fa-github"></i>
  </span>
<span class="btn__text-container">Repository</span>
</a>
</li>
      
      
      
      
      <li><a href="https://github.com/theislab/single-cell-best-practices/edit/master/jupyter-book/air_repertoire/clonotype.ipynb" target="_blank"
   class="btn btn-sm btn-source-edit-button dropdown-item"
   title="Suggest edit"
   data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-pencil-alt"></i>
  </span>
<span class="btn__text-container">Suggest edit</span>
</a>
</li>
      
      
      
      
      <li><a href="https://github.com/theislab/single-cell-best-practices/issues/new?title=Issue%20on%20page%20%2Fair_repertoire/clonotype.html&body=Your%20issue%20content%20here." target="_blank"
   class="btn btn-sm btn-source-issues-button dropdown-item"
   title="Open an issue"
   data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-lightbulb"></i>
  </span>
<span class="btn__text-container">Open issue</span>
</a>
</li>
      
  </ul>
</div>






<div class="dropdown dropdown-download-buttons">
  <button class="btn dropdown-toggle" type="button" data-bs-toggle="dropdown" aria-expanded="false" aria-label="Download this page">
    <i class="fas fa-download"></i>
  </button>
  <ul class="dropdown-menu">
      
      
      
      <li><a href="../_sources/air_repertoire/clonotype.ipynb" target="_blank"
   class="btn btn-sm btn-download-source-button dropdown-item"
   title="Download source file"
   data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-file"></i>
  </span>
<span class="btn__text-container">.ipynb</span>
</a>
</li>
      
      
      
      
      <li>
<button onclick="window.print()"
  class="btn btn-sm btn-download-pdf-button dropdown-item"
  title="Print to PDF"
  data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-file-pdf"></i>
  </span>
<span class="btn__text-container">.pdf</span>
</button>
</li>
      
  </ul>
</div>




<button onclick="toggleFullScreen()"
  class="btn btn-sm btn-fullscreen-button"
  title="Fullscreen mode"
  data-bs-placement="bottom" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-expand"></i>
  </span>

</button>



<script>
document.write(`
  <button class="btn btn-sm nav-link pst-navbar-icon theme-switch-button" title="light/dark" aria-label="light/dark" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="theme-switch fa-solid fa-sun fa-lg" data-mode="light"></i>
    <i class="theme-switch fa-solid fa-moon fa-lg" data-mode="dark"></i>
    <i class="theme-switch fa-solid fa-circle-half-stroke fa-lg" data-mode="auto"></i>
  </button>
`);
</script>


<script>
document.write(`
  <button class="btn btn-sm pst-navbar-icon search-button search-button__button" title="Search" aria-label="Search" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="fa-solid fa-magnifying-glass fa-lg"></i>
  </button>
`);
</script>
<button class="sidebar-toggle secondary-toggle btn btn-sm" title="Toggle secondary sidebar" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <span class="fa-solid fa-list"></span>
</button>
</div></div>
      
    </div>
  
</div>
</div>
              
              

<div id="jb-print-docs-body" class="onlyprint">
    <h1>Clonotype analysis</h1>
    <!-- Table of contents -->
    <div id="print-main-content">
        <div id="jb-print-toc">
            
            <div>
                <h2> Contents </h2>
            </div>
            <nav aria-label="Page">
                <ul class="visible nav section-nav flex-column">
<li class="toc-h1 nav-item toc-entry"><a class="reference internal nav-link" href="#">39. Clonotype analysis</a><ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#clonal-expansion-diversity-and-abundance">39.1. Clonal expansion: diversity and abundance</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#gene-segment-usage-and-spectratype">39.2. Gene segment usage and spectratype</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#tcr-data-preparation">39.3. TCR data preparation</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#clonotype-definition">39.4. Clonotype definition</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#clonal-expansion">39.5. Clonal expansion</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#clonotype-abundance">39.6. Clonotype abundance</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#gene-usage">39.7. Gene usage</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#spectratype-analysis">39.8. Spectratype analysis</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#motif-sequence-analysis">39.9. Motif sequence analysis</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#repertoire-comparison">39.10. Repertoire comparison</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#bcr-data-analysis-with-dandelion">39.11. BCR Data Analysis with Dandelion</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#dandelion-interoperability">39.12. Dandelion interoperability</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#id13">39.13. Clonotype definition</a></li>
</ul>
</li>
<li class="toc-h1 nav-item toc-entry"><a class="reference internal nav-link" href="#gene-segment-usage">40. Gene segment usage</a></li>
<li class="toc-h1 nav-item toc-entry"><a class="reference internal nav-link" href="#spectratype">41. Spectratype</a></li>
<li class="toc-h1 nav-item toc-entry"><a class="reference internal nav-link" href="#id23">42. Motif sequence analysis</a><ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#references">42.1. References</a></li>
</ul>
</li>
</ul>

            </nav>
        </div>
    </div>
</div>

              
                
<div id="searchbox"></div>
                <article class="bd-article">
                  
  <section class="tex2jax_ignore mathjax_ignore" id="clonotype-analysis">
<span id="air-sequence"></span><h1><span class="section-number">39. </span>Clonotype analysis<a class="headerlink" href="#clonotype-analysis" title="Link to this heading">#</a></h1>
<details class="sd-sphinx-override sd-dropdown sd-card sd-mb-3">
<summary class="sd-summary-title sd-card-header">
<span class="sd-summary-text"><i class="fas fa-brain"></i>   Key takeaways</span><span class="sd-summary-state-marker sd-summary-chevron-right"><svg version="1.1" width="1.5em" height="1.5em" class="sd-octicon sd-octicon-chevron-right" viewBox="0 0 24 24" aria-hidden="true"><path d="M8.72 18.78a.75.75 0 0 1 0-1.06L14.44 12 8.72 6.28a.751.751 0 0 1 .018-1.042.751.751 0 0 1 1.042-.018l6.25 6.25a.75.75 0 0 1 0 1.06l-6.25 6.25a.75.75 0 0 1-1.06 0Z"></path></svg></span></summary><div class="sd-summary-content sd-card-body docutils">
<div class="sd-card sd-sphinx-override sd-mb-3 sd-shadow-sm sd-card-hover docutils">
<div class="sd-card-body docutils">
<p class="sd-card-text">Clonal Expansion: Specific lymphocytes proliferate in response to immune stimuli, reducing clonal diversity but increasing the abundance of expanded clones.
This process is critical for understanding immune responses, as it reflects the transition from naive to effector and memory cells.</p>
</div>
<a class="sd-stretched-link sd-hide-link-text reference internal" href="#air-repertoire-clonotype-key-takeaway-1"><span class="std std-ref">Clonal expansion: diversity and abundance</span></a></div>
<div class="sd-card sd-sphinx-override sd-mb-3 sd-shadow-sm sd-card-hover docutils">
<div class="sd-card-body docutils">
<p class="sd-card-text">Gene Segment Usage &amp; Spectratype: The rearrangement of V(D)J gene segments shows preferential usage, and spectratype analysis measures CDR3 length diversity.
Together, they help identify immunodominant clonotypes and characterize immune repertoire patterns.</p>
</div>
<a class="sd-stretched-link sd-hide-link-text reference internal" href="#air-repertoire-clonotype-key-takeaway-2"><span class="std std-ref">Gene segment usage and spectratype</span></a></div>
</div>
</details><details class="sd-sphinx-override sd-dropdown sd-card sd-mb-3">
<summary class="sd-summary-title sd-card-header">
<span class="sd-summary-text"><i class="fa-solid fa-gear"></i>   Environment setup</span><span class="sd-summary-state-marker sd-summary-chevron-right"><svg version="1.1" width="1.5em" height="1.5em" class="sd-octicon sd-octicon-chevron-right" viewBox="0 0 24 24" aria-hidden="true"><path d="M8.72 18.78a.75.75 0 0 1 0-1.06L14.44 12 8.72 6.28a.751.751 0 0 1 .018-1.042.751.751 0 0 1 1.042-.018l6.25 6.25a.75.75 0 0 1 0 1.06l-6.25 6.25a.75.75 0 0 1-1.06 0Z"></path></svg></span></summary><div class="sd-summary-content sd-card-body docutils">
<div class="sd-tab-set docutils">
<input checked="checked" id="sd-tab-item-0" name="sd-tab-set-0" type="radio">
<label class="sd-tab-label" for="sd-tab-item-0">
Steps</label><div class="sd-tab-content docutils">
<ol class="arabic">
<li><p class="sd-card-text"><strong>Install conda</strong>:</p>
<ul class="simple">
<li><p class="sd-card-text">Before creating the environment, ensure that <a class="reference external" href="https://docs.conda.io/projects/conda/en/latest/user-guide/install/index.html">conda</a> is installed on your system.</p></li>
</ul>
</li>
<li><p class="sd-card-text"><strong>Save the yml content</strong>:</p>
<ul class="simple">
<li><p class="sd-card-text">Copy the content from the yml tab into a file named <code class="docutils literal notranslate"><span class="pre">environment.yml</span></code>.</p></li>
</ul>
</li>
<li><p class="sd-card-text"><strong>Create the environment</strong>:</p>
<ul>
<li><p class="sd-card-text">Open a terminal or command prompt.</p></li>
<li><p class="sd-card-text">Run the following command:</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>conda<span class="w"> </span>env<span class="w"> </span>create<span class="w"> </span>-f<span class="w"> </span>environment.yml
</pre></div>
</div>
</li>
</ul>
</li>
<li><p class="sd-card-text"><strong>Activate the environment</strong>:</p>
<ul>
<li><p class="sd-card-text">After the environment is created, activate it using:</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>conda<span class="w"> </span>activate<span class="w"> </span>&lt;environment_name&gt;
</pre></div>
</div>
</li>
<li><p class="sd-card-text">Replace <code class="docutils literal notranslate"><span class="pre">&lt;environment_name&gt;</span></code> with the name specified in the <code class="docutils literal notranslate"><span class="pre">environment.yml</span></code> file. In the yml file it will look like this:</p>
<div class="highlight-yaml notranslate"><div class="highlight"><pre><span></span><span class="nt">name</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">&lt;environment_name&gt;</span>
</pre></div>
</div>
</li>
</ul>
</li>
<li><p class="sd-card-text"><strong>Verify the installation</strong>:</p>
<ul>
<li><p class="sd-card-text">Check that the environment was created successfully by running:</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>conda<span class="w"> </span>env<span class="w"> </span>list
</pre></div>
</div>
</li>
</ul>
</li>
</ol>
</div>
<input id="sd-tab-item-1" name="sd-tab-set-0" type="radio">
<label class="sd-tab-label" for="sd-tab-item-1">
yml</label><div class="sd-tab-content docutils">
</div>
</div>
</div>
</details><!-- END DROPDOWNS -->
<section id="clonal-expansion-diversity-and-abundance">
<span id="air-repertoire-clonotype-key-takeaway-1"></span><h2><span class="section-number">39.1. </span>Clonal expansion: diversity and abundance<a class="headerlink" href="#clonal-expansion-diversity-and-abundance" title="Link to this heading">#</a></h2>
<p>In general, lymphocytes are in a dormant state until receiving an external signal (epitope recognition of foreign agent) or stimulation from autocrine agents (signaling from the same organism as a response from the innate immune system). As a consequence, the specific cells proliferate dramatically to fulfill the defense response they are programmed to perform in a process known as clonal expansion <span id="id1">[<a class="reference internal" href="#id514" title="Michal Polonsky, Benjamin Chain, and Nir Friedman. Clonal expansion under the microscope: studying lymphocyte activation and differentiation using live-cell imaging. Immunology and Cell Biology, 94(3):242–249, 2016.">Polonsky <em>et al.</em>, 2016</a>]</span>. This refers to the recognition of the proliferation of specific cells given the high number of the same IR through many, different cells (expanded clones). This expansion provides hints of differentiation from naive lymphocytes to mature effector and memory lymphocytes, helping in the interpretation and expected results regarding previous cell annotation <span id="id2">[<a class="reference internal" href="#id514" title="Michal Polonsky, Benjamin Chain, and Nir Friedman. Clonal expansion under the microscope: studying lymphocyte activation and differentiation using live-cell imaging. Immunology and Cell Biology, 94(3):242–249, 2016.">Polonsky <em>et al.</em>, 2016</a>]</span>. On the other hand, the analysis of expanded clones should consider derivative processes such as clonal competitions (two or more clones in expansion competing for the same space), clonal dominance (one single clonal expanded cell outnumbering the rest of the clonal cells), and bystander activation (activation of T-cells by cytokines but not for T-cell receptor coupling) <span id="id3">[<a class="reference internal" href="#id513" title="Kamila Naxerova. Clonal competition in a confined space. Nature Genetics, 52(6):553–554, 2020.">Naxerova, 2020</a>]</span><span id="id4">[<a class="reference internal" href="#id512" title="Peter Ashcroft, Markus G Manz, and Sebastian Bonhoeffer. Clonal dominance and transplantation dynamics in hematopoietic stem cell compartments. PLoS computational biology, 13(10):e1005803, 2017.">Ashcroft <em>et al.</em>, 2017</a>]</span><span id="id5">[<a class="reference internal" href="#id511" title="Tae-Shin Kim and Eui-Cheol Shin. The activation of bystander cd8+ t cells and their roles in viral infection. Experimental &amp; molecular medicine, 51(12):1–9, 2019.">Kim and Shin, 2019</a>]</span>.</p>
<p>The dynamical changes in terms of the number of cells per clonotype in a given space allow applying concepts from population biology such as diversity and abundance. Diversity is defined as number of species and their amount in an area or community, whereas abundance is the number, or frequency of individuals of the same species <span id="id6">[<a class="reference internal" href="#id515" title="Ilias S Travlos, Nikolina Cheimona, Ioannis Roussis, and Dimitrios J Bilalis. Weed-species abundance and diversity indices in relation to tillage systems and fertilization. Frontiers in Environmental Science, 6:11, 2018.">Travlos <em>et al.</em>, 2018</a>]</span>. Here, we can replace the term species with clones to make clear their relevance in single-cell IRs analysis. If a high clonal expansion has been detected in a specific cell type, e.g., effector CD8+ T-cells, the number of clones is expected to reduce because the expanded clones are taken space from the total available sacrificing receptor alternatives in the process. Therefore, we could expect a reduction in the diversity for this cell-type. On the other hand, we could expect an increase in abundance regarding the expanded clones given they have increased the number of individuals (cells) belonging to this specific clone (observed by the number of cells per clone ID).</p>
</section>
<section id="gene-segment-usage-and-spectratype">
<span id="air-repertoire-clonotype-key-takeaway-2"></span><h2><span class="section-number">39.2. </span>Gene segment usage and spectratype<a class="headerlink" href="#gene-segment-usage-and-spectratype" title="Link to this heading">#</a></h2>
<p>The process shaping a T-cell or B-cell receptor by rearrangement of the V(D)J segments is thinking to generate random sequences and, in consequence, the distribution of V(D)J sequences should follow a uniform distribution. Nevertheless, it has been observed that V(D)J gene usage frequency is largely consistent across different individuals, which suggests a preference selection in terms of the V(D)J gene segments used <span id="id7">[<a class="reference internal" href="#id516" title="Yuval Elhanati, Anand Murugan, Curtis G Callan Jr, Thierry Mora, and Aleksandra M Walczak. Quantifying selection in immune receptor repertoires. Proceedings of the National Academy of Sciences, 111(27):9875–9880, 2014.">Elhanati <em>et al.</em>, 2014</a>]</span>. That allows the analysis of gene segment usage in terms of abundance of most used gene segments per cell type and frequency of most abundant segment per cell type per individual <span id="id8">[<a class="reference internal" href="#id517" title="Mark Chernyshev, Mateusz Kaduk, Martin Corcoran, and Gunilla B Karlsson Hedestam. Vdj gene usage in igm repertoires of rhesus and cynomolgus macaques. Frontiers in immunology, 2021.">Chernyshev <em>et al.</em>, 2021</a>]</span>. Likewise, considering we know the amino acid composition of the immune receptors for each cell, it is possible to identify the exact combinations of V(D)J segments of interest.</p>
<p>On the other hand, the recombination of V(D)J gene segments and the imprecise junction of V and J segments produce CDR3 regions with variable lengths. Spectratype analysis is seen as the measurement of the heterogeneity of CDR3 regions by their length diversity across the different cell types <span id="id9">[<a class="reference internal" href="#id518" title="Stanca M Ciupe, Blythe H Devlin, Mary Louise Markert, and Thomas B Kepler. Quantification of total t-cell receptor diversity by flow cytometry and spectratyping. BMC immunology, 14(1):1–12, 2013.">Ciupe <em>et al.</em>, 2013</a>]</span>. This measurement, in combination with clonal expansion and gene segment usage provides pieces of evidence to define well-described immunodominant clonotypes.</p>
</section>
<section id="tcr-data-preparation">
<h2><span class="section-number">39.3. </span>TCR data preparation<a class="headerlink" href="#tcr-data-preparation" title="Link to this heading">#</a></h2>
<p>Here, as well as in the pre-processing step, we will use the utilities from the <em>Scirpy</em> library to perform the analysis and locate the results in the <em>AnnData</em> object.</p>
<div class="admonition warning">
<p class="admonition-title">Warning</p>
<p>Scirpy changed the format of <a class="reference external" href="https://scirpy.scverse.org/en/latest/data-structure.html#storing-airr-rearrangement-data-in-anndata">its data structure</a>
with v0.13. While the overall analysis workflow has not changed, some outputs shown in this chapter might not be accurate anymore.</p>
<p>See <a class="reference external" href="https://scirpy.scverse.org/en/latest/changelog.html#v0-13-0-new-data-structure-based-on-awkward-arrays">the scirpy release notes</a> for more details about this change.
Until we update this chapter, please also refer to the <a class="reference external" href="https://scirpy.scverse.org">official scirpy documentation</a>.</p>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython2 notranslate"><div class="highlight"><pre><span></span>import warnings

warnings.filterwarnings(
    &quot;ignore&quot;,
    &quot;.*IProgress not found*&quot;,
)
warnings.simplefilter(action=&quot;ignore&quot;, category=FutureWarning)

import numpy as np
import pandas as pd
import scanpy as sc
import scirpy as ir
from palmotif import compute_motif, svg_logo

warnings.simplefilter(action=&quot;ignore&quot;, category=pd.errors.DtypeWarning)
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython2 notranslate"><div class="highlight"><pre><span></span>sc.logging.print_versions()
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>WARNING: If you miss a compact list, please try `print_header`!
-----
anndata     0.8.0
scanpy      1.6.1
sinfo       0.3.1
-----
Levenshtein                 NA
PIL                         8.1.0
adjustText                  NA
airr                        1.3.1
anndata                     0.8.0
anyio                       NA
attr                        20.3.0
babel                       2.9.0
backcall                    0.2.0
brotli                      1.0.9
certifi                     2020.12.05
cffi                        1.14.4
chardet                     4.0.0
cloudpickle                 1.6.0
constants                   NA
cycler                      0.10.0
cython_runtime              NA
dask                        2020.12.0
dateutil                    2.8.1
decorator                   4.4.2
future_fstrings             NA
get_version                 2.1
google                      NA
h5py                        3.7.0
highs_wrapper               NA
idna                        2.10
igraph                      0.8.3
ipykernel                   5.4.3
ipython_genutils            0.2.0
ipywidgets                  7.6.3
jedi                        0.18.0
jinja2                      2.11.2
joblib                      1.0.0
json5                       NA
jsonschema                  3.2.0
jupyter_server              1.2.1
jupyterlab_server           2.1.2
kiwisolver                  1.3.1
legacy_api_wrap             1.2
leidenalg                   0.8.3
llvmlite                    0.35.0
louvain                     0.7.0
lxml                        4.6.2
markupsafe                  1.1.1
matplotlib                  3.3.3
mpl_toolkits                NA
natsort                     7.1.0
nbclassic                   NA
nbformat                    5.1.1
networkx                    2.5
numba                       0.52.0
numexpr                     2.7.2
numpy                       1.19.5
packaging                   20.8
pandas                      1.2.0
parasail                    1.2.4
parso                       0.8.1
pexpect                     4.8.0
pickleshare                 0.7.5
pkg_resources               NA
prometheus_client           NA
prompt_toolkit              3.0.10
ptyprocess                  0.7.0
pvectorc                    NA
pygments                    2.7.4
pyparsing                   2.4.7
pyrsistent                  NA
pytz                        2020.5
requests                    2.25.1
scanpy                      1.6.1
scipy                       1.6.0
scirpy                      0.11.0
seaborn                     0.11.1
send2trash                  NA
setuptools_scm              NA
sinfo                       0.3.1
six                         1.15.0
sklearn                     0.24.0
sniffio                     1.2.0
sparse                      0.11.2
statsmodels                 0.12.1
storemagic                  NA
tables                      3.6.1
texttable                   1.6.3
tlz                         0.11.1
toolz                       0.11.1
tornado                     6.1
tqdm                        4.56.0
tracerlib                   NA
traitlets                   5.0.5
typing_extensions           NA
urllib3                     1.26.2
wcwidth                     0.2.5
yaml                        5.3.1
yamlordereddictloader       NA
zmq                         21.0.0
-----
IPython             7.19.0
jupyter_client      6.1.11
jupyter_core        4.7.0
jupyterlab          3.0.4
notebook            6.2.0
-----
Python 3.8.6 (default, Jan 14 2021, 17:39:54) [GCC 8.3.0]
Linux-3.10.0-1160.25.1.el7.x86_64-x86_64-with-glibc2.28
48 logical CPU cores
-----
Session information updated at 2022-08-12 11:05
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython2 notranslate"><div class="highlight"><pre><span></span>path_data = &quot;/home/icb/juan.henao/BestPracticeStart/data&quot;

path_gex = f&quot;{path_data}/TCR_filtered.h5ad&quot;
adata = sc.read(path_gex)
</pre></div>
</div>
</div>
</div>
<p><strong>Note</strong></p>
<p>Before performing any analysis, it is necessary to reduce the size of our data to make this tutorial computationally less expensive and make it timely suitable.</p>
<p>To do so, we have chosen six samples ensuring the different centers and medical statuses were represented.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython2 notranslate"><div class="highlight"><pre><span></span>adata = adata[
    adata.obs[&quot;patient_id&quot;].isin(
        [&quot;COVID-014&quot;, &quot;CV0902&quot;, &quot;AP6&quot;, &quot;COVID-045&quot;, &quot;COVID-066&quot;, &quot;COVID-067&quot;]
    )
]
# adata[adata.obs[&#39;Status&#39;] == &#39;Healthy&#39;].obs[&#39;patient_id&#39;]
_ = ir.pl.group_abundance(
    adata,
    groupby=&quot;patient_id&quot;,
    target_col=&quot;chain_pairing&quot;,
    normalize=True,
    figsize=[10, 10],
)
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>/home/juan.henao/.local/lib/python3.8/site-packages/anndata/compat/_overloaded_dict.py:106: ImplicitModificationWarning: Trying to modify attribute `._uns` of view, initializing view as actual.
  self.data[key] = value
</pre></div>
</div>
<img alt="../_images/15471dbdbcebc91c7fcb86e9ed7f04c4e19ee3b3e4f1a528cdd62fae81086ad6.png" src="../_images/15471dbdbcebc91c7fcb86e9ed7f04c4e19ee3b3e4f1a528cdd62fae81086ad6.png" />
</div>
</div>
</section>
<section id="clonotype-definition">
<h2><span class="section-number">39.4. </span>Clonotype definition<a class="headerlink" href="#clonotype-definition" title="Link to this heading">#</a></h2>
<p>One way to define the clonotypes is by detecting all identical sequences for VJ CDR3 and VDJ CDR3. It is the most used one due to allowing to define trustable cell lineages. However, another way to deal with this problem is finding a distance between sequences, this method is restrictive, and it could be useful to test some hypotheses regarding complex immunological phenotypes.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython2 notranslate"><div class="highlight"><pre><span></span>ir.pp.ir_dist(adata, sequence=&quot;aa&quot;)
</pre></div>
</div>
</div>
</div>
<p>Once the identity between T-cells is obtained for V(D)J CDR3, it is time to define the cluster of cells corresponding to one specific clonotype. A clonotype will be a set of cells with identical sequences, considering the parameters used in the previous step. However, it is possible to define clonotypes as a set of cells with just identical VJ or just identical VDJ sequences. Furthermore, it is possible defining the clonotypes by comparing either or both pairs of VJ or VDJ sequences.</p>
<p>The set of parameters to define clonotypes should be the same as used previously. In our case, the sequences of amino acids must be compared using identity as a metric. In addition, we are setting the additional parameters to define clonotypes if the V(D)J are identical using the most abundant pair as the target sequence.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython2 notranslate"><div class="highlight"><pre><span></span>ir.tl.define_clonotype_clusters(
    adata, sequence=&quot;aa&quot;, receptor_arms=&quot;all&quot;, dual_ir=&quot;primary_only&quot;
)
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<script type="application/vnd.jupyter.widget-view+json">{"model_id": "e7159ee89ce947b18b58af4fedb6ec77", "version_major": 2, "version_minor": 0}</script></div>
</div>
<p>The way to visualize the results is through a network where each node represents a clonotype (cluster of cells), and its size represents the number of cells detected in that cluster. They are labeled with a numerical ID, however, the order is given randomly, and it is not showing any additional information beyond to identify clonotypes of interest.</p>
<p>To generate the network, it is necessary to establish the layout to be plotted afterward. This parameter should be one of the igraph library layouts. Furthermore, it is recommended to set at least <em>min_cells</em> &gt;=2 to avoid overcrowding the plot with singletons (clonotypes with only one cell as a member). Here, this parameter is established as &gt;= 50 to show just the biggest clonotypes and make the observation of the expected result easier.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython2 notranslate"><div class="highlight"><pre><span></span>ir.tl.clonotype_network(adata, min_cells=50, sequence=&quot;aa&quot;)
</pre></div>
</div>
</div>
</div>
<p>Now it is possible to plot the network. The result is just like the one you can observe below. As we said previously, each node (circle) represents a clonotype with a unique number as ID. Furthermore, the size represents the number of cells belonging to each specific clonotype.</p>
<p>On the other hand, we set the color according to the samples to observe if a clonotype appears in two or more samples, those clonotypes are called <em>public clonotypes</em> and are of high interest as they represent shared immunological responses, and therefore they are candidates to explain general response over the disease/phenotype under study. Otherwise, there are <em>private clonotypes</em> which represent patient/sample specific clonal response, and it could be interesting for analysis regarding personalized medicine. As you can see below, the highest clonotypes are composed just of private clonotypes.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython2 notranslate"><div class="highlight"><pre><span></span>_ = ir.pl.clonotype_network(
    adata,
    color=&quot;patient_id&quot;,
    base_size=10,
    label_fontsize=9,
    panel_size=(10, 10),
    legend_fontsize=15,
)
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>... storing &#39;cc_aa_identity&#39; as categorical
</pre></div>
</div>
<img alt="../_images/692f9cc1c983d1e3cd51644e08778aa6dde66f146d19d1a562f5c04c0f7ec122.png" src="../_images/692f9cc1c983d1e3cd51644e08778aa6dde66f146d19d1a562f5c04c0f7ec122.png" />
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython2 notranslate"><div class="highlight"><pre><span></span>adata.obs[&quot;cc_aa_identity&quot;] = adata.obs[&quot;cc_aa_identity&quot;].astype(&quot;str&quot;)
</pre></div>
</div>
</div>
</div>
<p>As mentioned previously, the detected clonotypes are labeled with a number we can use to extract more information. For instance, we are extracting the immune sequences for clonotype number 0.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython2 notranslate"><div class="highlight"><pre><span></span>adata.obs.loc[adata.obs[&quot;cc_aa_identity&quot;] == &quot;0&quot;, :].groupby(
    [
        &quot;IR_VJ_1_junction_aa&quot;,
        &quot;IR_VDJ_1_junction_aa&quot;,
        &quot;receptor_subtype&quot;,
    ],
    observed=True,
).size().reset_index(name=&quot;n_cells&quot;)
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>IR_VJ_1_junction_aa</th>
      <th>IR_VDJ_1_junction_aa</th>
      <th>receptor_subtype</th>
      <th>n_cells</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>CAVSVVRNNNARLMF</td>
      <td>CASSARGASGERTDTQYF</td>
      <td>TRA+TRB</td>
      <td>71</td>
    </tr>
  </tbody>
</table>
</div></div></div>
</div>
</section>
<section id="clonal-expansion">
<h2><span class="section-number">39.5. </span>Clonal expansion<a class="headerlink" href="#clonal-expansion" title="Link to this heading">#</a></h2>
<p>The positive selection of immune cells, e.g., for immune response activation, causes their expansion (division) reflected in the representation of clonotypes in one or more cells. The first step is to identify the clonal expansion and add a column in the .obs table.</p>
<p>Once the clonal expansion has been identified, it is easy to observe that, plotting the number of cells per cell type corresponding to expanded clones as a stacked bar plot.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython2 notranslate"><div class="highlight"><pre><span></span>ir.tl.clonal_expansion(adata, target_col=&quot;cc_aa_identity&quot;)

_ = ir.pl.clonal_expansion(
    adata,
    groupby=&quot;initial_clustering&quot;,
    target_col=&quot;cc_aa_identity&quot;,
    clip_at=4,
    normalize=False,
    figsize=[10, 10],
)
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="../_images/a931a8d7795f5a866ca39029afd46f55f498cc0211999e8ff4120f55443ca3a2.png" src="../_images/a931a8d7795f5a866ca39029afd46f55f498cc0211999e8ff4120f55443ca3a2.png" />
</div>
</div>
<p>The plot above tells us that CD4+ is the most abundant cell type. However, CD8+ shows the highest number of expanded clonotypes, which could be possible by the positive selection of CD8+ effector cells.</p>
<p>Another way to observe clonal expansion is by normalizing the size of cell-type clusters. Here, the clonal expansion differences between CD4+ and CD8+ are easier to observe.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython2 notranslate"><div class="highlight"><pre><span></span>_ = ir.pl.clonal_expansion(
    adata, &quot;initial_clustering&quot;, target_col=&quot;cc_aa_identity&quot;, figsize=[10, 10]
)
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="../_images/059ef7e63f956528b9bebd6190fd677164f10c3c11b2b6c5d6119081324ae730.png" src="../_images/059ef7e63f956528b9bebd6190fd677164f10c3c11b2b6c5d6119081324ae730.png" />
</div>
</div>
<p>Another way to observe the clonal expansion phenomenon is through the loss of alpha diversity. The expanded clones are less diverse because there are more individuals (cells) from the same type (clone) inside a specific population (cluster of cells). There are different types of diversity and different ways to calculate this according to different assumptions. However, in TCR analysis, we are interested in correlate this concept with the clonality expansion per cluster as the reduction of variability in a specific set of features (alpha diversity).</p>
<p>The plot below reflects this assumption, the alpha diversity of CD8+ is lower than CD4+ due to the possible clonal expansion from CD8+ effector cells reducing their diversity in consideration of the total cells conforming to the CD8+ cluster. The same observation is repeated for clusters NK_16hi and gdT-cell, which are highly expanded, as the clonal expansion plot demonstrates (see above).</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython2 notranslate"><div class="highlight"><pre><span></span>_ = ir.pl.alpha_diversity(
    adata, groupby=&quot;initial_clustering&quot;, target_col=&quot;cc_aa_identity&quot;, figsize=[10, 10]
)
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="../_images/af577d10d387da4db5535e1c3417babdbd04fb4bc0902f9a6b9d6e70df613924.png" src="../_images/af577d10d387da4db5535e1c3417babdbd04fb4bc0902f9a6b9d6e70df613924.png" />
</div>
</div>
</section>
<section id="clonotype-abundance">
<h2><span class="section-number">39.6. </span>Clonotype abundance<a class="headerlink" href="#clonotype-abundance" title="Link to this heading">#</a></h2>
<p>Contrary to diversity, which is expected to be reduced when clonal expansion is gained, regarding a cluster of cells, the abundance is expected to be gained when clonal expansion occurred due to the increase in the number of cells from the same clone.</p>
<p>The next function helps to plot the most abundant clonotypes showing the ID, the number of cells and the cluster (cell type) the specific clonotype belongs to. Here, we highlight the ten most abundant clonotypes.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython2 notranslate"><div class="highlight"><pre><span></span>_ = ir.pl.group_abundance(
    adata,
    groupby=&quot;cc_aa_identity&quot;,
    target_col=&quot;initial_clustering&quot;,
    max_cols=10,
    figsize=[10, 10],
)
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>... storing &#39;cc_aa_identity&#39; as categorical
... storing &#39;clonal_expansion&#39; as categorical
Using categorical units to plot a list of strings that are all parsable as floats or dates. If these strings should be plotted as numbers, cast to the appropriate data type before plotting.
Using categorical units to plot a list of strings that are all parsable as floats or dates. If these strings should be plotted as numbers, cast to the appropriate data type before plotting.
Using categorical units to plot a list of strings that are all parsable as floats or dates. If these strings should be plotted as numbers, cast to the appropriate data type before plotting.
Using categorical units to plot a list of strings that are all parsable as floats or dates. If these strings should be plotted as numbers, cast to the appropriate data type before plotting.
Using categorical units to plot a list of strings that are all parsable as floats or dates. If these strings should be plotted as numbers, cast to the appropriate data type before plotting.
Using categorical units to plot a list of strings that are all parsable as floats or dates. If these strings should be plotted as numbers, cast to the appropriate data type before plotting.
Using categorical units to plot a list of strings that are all parsable as floats or dates. If these strings should be plotted as numbers, cast to the appropriate data type before plotting.
Using categorical units to plot a list of strings that are all parsable as floats or dates. If these strings should be plotted as numbers, cast to the appropriate data type before plotting.
Using categorical units to plot a list of strings that are all parsable as floats or dates. If these strings should be plotted as numbers, cast to the appropriate data type before plotting.
Using categorical units to plot a list of strings that are all parsable as floats or dates. If these strings should be plotted as numbers, cast to the appropriate data type before plotting.
Using categorical units to plot a list of strings that are all parsable as floats or dates. If these strings should be plotted as numbers, cast to the appropriate data type before plotting.
Using categorical units to plot a list of strings that are all parsable as floats or dates. If these strings should be plotted as numbers, cast to the appropriate data type before plotting.
Using categorical units to plot a list of strings that are all parsable as floats or dates. If these strings should be plotted as numbers, cast to the appropriate data type before plotting.
Using categorical units to plot a list of strings that are all parsable as floats or dates. If these strings should be plotted as numbers, cast to the appropriate data type before plotting.
Using categorical units to plot a list of strings that are all parsable as floats or dates. If these strings should be plotted as numbers, cast to the appropriate data type before plotting.
Using categorical units to plot a list of strings that are all parsable as floats or dates. If these strings should be plotted as numbers, cast to the appropriate data type before plotting.
Using categorical units to plot a list of strings that are all parsable as floats or dates. If these strings should be plotted as numbers, cast to the appropriate data type before plotting.
Using categorical units to plot a list of strings that are all parsable as floats or dates. If these strings should be plotted as numbers, cast to the appropriate data type before plotting.
Using categorical units to plot a list of strings that are all parsable as floats or dates. If these strings should be plotted as numbers, cast to the appropriate data type before plotting.
Using categorical units to plot a list of strings that are all parsable as floats or dates. If these strings should be plotted as numbers, cast to the appropriate data type before plotting.
Using categorical units to plot a list of strings that are all parsable as floats or dates. If these strings should be plotted as numbers, cast to the appropriate data type before plotting.
Using categorical units to plot a list of strings that are all parsable as floats or dates. If these strings should be plotted as numbers, cast to the appropriate data type before plotting.
Using categorical units to plot a list of strings that are all parsable as floats or dates. If these strings should be plotted as numbers, cast to the appropriate data type before plotting.
Using categorical units to plot a list of strings that are all parsable as floats or dates. If these strings should be plotted as numbers, cast to the appropriate data type before plotting.
Using categorical units to plot a list of strings that are all parsable as floats or dates. If these strings should be plotted as numbers, cast to the appropriate data type before plotting.
Using categorical units to plot a list of strings that are all parsable as floats or dates. If these strings should be plotted as numbers, cast to the appropriate data type before plotting.
Using categorical units to plot a list of strings that are all parsable as floats or dates. If these strings should be plotted as numbers, cast to the appropriate data type before plotting.
Using categorical units to plot a list of strings that are all parsable as floats or dates. If these strings should be plotted as numbers, cast to the appropriate data type before plotting.
Using categorical units to plot a list of strings that are all parsable as floats or dates. If these strings should be plotted as numbers, cast to the appropriate data type before plotting.
Using categorical units to plot a list of strings that are all parsable as floats or dates. If these strings should be plotted as numbers, cast to the appropriate data type before plotting.
Using categorical units to plot a list of strings that are all parsable as floats or dates. If these strings should be plotted as numbers, cast to the appropriate data type before plotting.
Using categorical units to plot a list of strings that are all parsable as floats or dates. If these strings should be plotted as numbers, cast to the appropriate data type before plotting.
Using categorical units to plot a list of strings that are all parsable as floats or dates. If these strings should be plotted as numbers, cast to the appropriate data type before plotting.
Using categorical units to plot a list of strings that are all parsable as floats or dates. If these strings should be plotted as numbers, cast to the appropriate data type before plotting.
</pre></div>
</div>
<img alt="../_images/1c975a8ab57c2f703f3c3e45dd2893a77cc646af1de2d7bea357325e2417d3bc.png" src="../_images/1c975a8ab57c2f703f3c3e45dd2893a77cc646af1de2d7bea357325e2417d3bc.png" />
</div>
</div>
<p>The plot above showed that clonotype ID 11501 was the most expanded and is present in more than one cell cluster. However, according to the previous results, it is expected this clonotype corresponds to the same sample and therefore is involved in one specific condition (COVID or healthy). Hence, the repetition of this plotting is worth it to check this assumption.</p>
<p>Furthermore, the clonal expansion is expected to occur in response to an immune event, which means it is expected to the total, or at least the majority of the most expanded clones belong to the COVID set of patients.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython2 notranslate"><div class="highlight"><pre><span></span># By condition

_ = ir.pl.group_abundance(
    adata,
    groupby=&quot;cc_aa_identity&quot;,
    target_col=&quot;Status&quot;,
    max_cols=15,
    fig_kws={&quot;dpi&quot;: 100},
    figsize=[10, 10],
)

# By sample

_ = ir.pl.group_abundance(
    adata,
    groupby=&quot;cc_aa_identity&quot;,
    target_col=&quot;patient_id&quot;,
    max_cols=15,
    fig_kws={&quot;dpi&quot;: 100},
    figsize=[10, 10],
)
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Using categorical units to plot a list of strings that are all parsable as floats or dates. If these strings should be plotted as numbers, cast to the appropriate data type before plotting.
Using categorical units to plot a list of strings that are all parsable as floats or dates. If these strings should be plotted as numbers, cast to the appropriate data type before plotting.
Using categorical units to plot a list of strings that are all parsable as floats or dates. If these strings should be plotted as numbers, cast to the appropriate data type before plotting.
Using categorical units to plot a list of strings that are all parsable as floats or dates. If these strings should be plotted as numbers, cast to the appropriate data type before plotting.
Using categorical units to plot a list of strings that are all parsable as floats or dates. If these strings should be plotted as numbers, cast to the appropriate data type before plotting.
Using categorical units to plot a list of strings that are all parsable as floats or dates. If these strings should be plotted as numbers, cast to the appropriate data type before plotting.
Using categorical units to plot a list of strings that are all parsable as floats or dates. If these strings should be plotted as numbers, cast to the appropriate data type before plotting.
Using categorical units to plot a list of strings that are all parsable as floats or dates. If these strings should be plotted as numbers, cast to the appropriate data type before plotting.
Using categorical units to plot a list of strings that are all parsable as floats or dates. If these strings should be plotted as numbers, cast to the appropriate data type before plotting.
Using categorical units to plot a list of strings that are all parsable as floats or dates. If these strings should be plotted as numbers, cast to the appropriate data type before plotting.
Using categorical units to plot a list of strings that are all parsable as floats or dates. If these strings should be plotted as numbers, cast to the appropriate data type before plotting.
Using categorical units to plot a list of strings that are all parsable as floats or dates. If these strings should be plotted as numbers, cast to the appropriate data type before plotting.
Using categorical units to plot a list of strings that are all parsable as floats or dates. If these strings should be plotted as numbers, cast to the appropriate data type before plotting.
Using categorical units to plot a list of strings that are all parsable as floats or dates. If these strings should be plotted as numbers, cast to the appropriate data type before plotting.
Using categorical units to plot a list of strings that are all parsable as floats or dates. If these strings should be plotted as numbers, cast to the appropriate data type before plotting.
Using categorical units to plot a list of strings that are all parsable as floats or dates. If these strings should be plotted as numbers, cast to the appropriate data type before plotting.
</pre></div>
</div>
<img alt="../_images/6c830ed5eeb519cda0b1a030a7be3bebe87824dfe7b4d1e3e09ed034721d4931.png" src="../_images/6c830ed5eeb519cda0b1a030a7be3bebe87824dfe7b4d1e3e09ed034721d4931.png" />
<img alt="../_images/bde53cbb12840e74229bd28ae14efa0e46970488b20c5069de1e18389f0270e2.png" src="../_images/bde53cbb12840e74229bd28ae14efa0e46970488b20c5069de1e18389f0270e2.png" />
</div>
</div>
</section>
<section id="gene-usage">
<h2><span class="section-number">39.7. </span>Gene usage<a class="headerlink" href="#gene-usage" title="Link to this heading">#</a></h2>
<p>In the previous steps, we analyzed the data regarding clones, their expansion, and their diversity and abundance. However, it is possible to analyze the data to describe the specific V(D)J gene segments, their abundances across the different cell clusters, and their specific combination resulting in the immune receptors.</p>
<p>As described in the introduction, there are shreds of evidence of selective selection of gene segments privileging the use of some gene segments over the rest of them. Therefore, the first step is to detect those segments privileged for this phenomenon via abundance analysis, the assumption behind is straightforward, if one gene segment is highly abundant, it means it was selectively chosen in IR final arrangement.</p>
<p>We calculated the abundances for the V segment in the VJ chain using the same function to clone abundance calculation. The sequence with ID TRAV19 was the most abundant V gene, and it was present in CD8+ cells mainly. Otherwise, the second most abundant V segment (TRAV29/DV5) was mostly selected for the CD4+ cell cluster.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython2 notranslate"><div class="highlight"><pre><span></span>_ = ir.pl.group_abundance(
    adata,
    groupby=&quot;IR_VJ_1_v_call&quot;,
    target_col=&quot;initial_clustering&quot;,
    normalize=False,
    max_cols=10,
    figsize=[10, 10],
)

# Normalized abundances

_ = ir.pl.group_abundance(
    adata,
    groupby=&quot;IR_VJ_1_v_call&quot;,
    target_col=&quot;initial_clustering&quot;,
    normalize=True,
    max_cols=10,
    figsize=[10, 10],
)
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="../_images/e0cd9db9fc912ff0428e71e4c3f83812a6554ab146977b478aee62382670d4f2.png" src="../_images/e0cd9db9fc912ff0428e71e4c3f83812a6554ab146977b478aee62382670d4f2.png" />
<img alt="../_images/53912819f4bf48062b42b329868615c306093f15044e3571a90bdeb6f1733ffa.png" src="../_images/53912819f4bf48062b42b329868615c306093f15044e3571a90bdeb6f1733ffa.png" />
</div>
</div>
<p>On the other hand, it is possible to choose the list of segments of interest and detect the fraction of cells for which those segments are represented across the different clusters of cells.</p>
<p>Here, we are showing a list of four V segments for the VDJ chain. From them, the TRBV18 segment was the most abundant in the majority of cell clusters.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython2 notranslate"><div class="highlight"><pre><span></span>_ = ir.pl.group_abundance(
    adata[
        adata.obs[&quot;IR_VDJ_1_v_call&quot;].isin(
            [&quot;TRBV19&quot;, &quot;TRBV10-1&quot;, &quot;TRBV11-1&quot;, &quot;TRBV7-9&quot;]
        ),
        :,
    ],
    groupby=&quot;initial_clustering&quot;,
    target_col=&quot;IR_VDJ_1_v_call&quot;,
    normalize=True,
    figsize=[10, 10],
)
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>/home/juan.henao/.local/lib/python3.8/site-packages/anndata/compat/_overloaded_dict.py:106: ImplicitModificationWarning: Trying to modify attribute `._uns` of view, initializing view as actual.
  self.data[key] = value
</pre></div>
</div>
<img alt="../_images/a5b7dd236f9a261ca425135671c41be8168954ac9cef0c0de218fc5e0a17c982.png" src="../_images/a5b7dd236f9a261ca425135671c41be8168954ac9cef0c0de218fc5e0a17c982.png" />
</div>
</div>
<p>Beyond the individual analysis and specific V(D)J segments according to abundances, it is possible to visualize a specific combination of V, D, and J segments for the total number of cells (see below). It provides valuable information regarding how the different privileged segments are combined to create different alpha and beta IRs sequences and how they are related to the rest of the low-selected segments.</p>
<p>One assumption for this graphic would be the most abundant segments are combined expecting a kind of linear representation. However, those segments are independent of each other in the random somatic recombination, therefore, this assumption is not true.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython2 notranslate"><div class="highlight"><pre><span></span>_ = ir.pl.vdj_usage(
    adata,
    full_combination=False,
    max_segments=None,
    max_ribbons=30,
    fig_kws={&quot;figsize&quot;: [10, 10]},
)
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="../_images/57cfb6e3c80c0821f787cd8f98a49a945e56ec85fed6a4e39ecb4c27bd6fb5e1.png" src="../_images/57cfb6e3c80c0821f787cd8f98a49a945e56ec85fed6a4e39ecb4c27bd6fb5e1.png" />
</div>
</div>
<p>The segment combination plot showed that there was just one D segment (TBRD2) for the VDJ chains detected for the set of cells and patients used here from the original experiment.</p>
<p>As well as the previous plots, it is possible to plot the combination of segments for a set of specific observations. In this case, we focused attention checking the combination of V(D)J segments for a set of clonotypes with positive detection of the TRBD2 segment highlighted above. In the next example, we used the first five clonotype IDs.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython2 notranslate"><div class="highlight"><pre><span></span>adata.obs[adata.obs[&quot;IR_VDJ_1_d_call&quot;] == &quot;TRBD2&quot;].cc_aa_identity.value_counts()
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>5150     1
1815     1
4427     1
3500     1
4078     1
        ..
4734     0
4735     0
4736     0
4737     0
14180    0
Name: cc_aa_identity, Length: 14181, dtype: int64
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython2 notranslate"><div class="highlight"><pre><span></span>_ = ir.pl.vdj_usage(
    adata[
        adata.obs[&quot;cc_aa_identity&quot;].isin([&quot;5150&quot;, &quot;1815&quot;, &quot;4427&quot;, &quot;3500&quot;, &quot;4078&quot;]), :
    ],
    max_ribbons=None,
    max_segments=100,
    fig_kws={&quot;figsize&quot;: [10, 10]},
)
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="../_images/0a6af364201ea55cec8cef82f462320b0de8e3f39bc28dfd27087ef6f5b3c74b.png" src="../_images/0a6af364201ea55cec8cef82f462320b0de8e3f39bc28dfd27087ef6f5b3c74b.png" />
</div>
</div>
</section>
<section id="spectratype-analysis">
<h2><span class="section-number">39.8. </span>Spectratype analysis<a class="headerlink" href="#spectratype-analysis" title="Link to this heading">#</a></h2>
<p>Spectratype analysis provides more information about V(D)J sequences heterogeneity. Not all IR sequences have the same number of amino acids, given the pseudo-random gene segment cleavage during somatic recombination. In addition, spectratype is another way to define immunodominance based on the most abundant sequence length. If the majority of V(D)J sequences share the same length, it means the functional chains should have the same or a similar number of amino acids.</p>
<p>Below, we plotted the abundances for VDJ sequences and the most common length was 15, closely followed by 14 amino acids detected in a variety of cell clusters, mainly CD4+ and CD8+. We highlight the last cell type, which was previously detected as highly clonal expanded. However, we found even VDJ sequences with lengths of 10 or 21 amino acids. However, they were strongly less represented in the general set of VDJ sequences.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython2 notranslate"><div class="highlight"><pre><span></span>_ = ir.pl.spectratype(
    adata,
    cdr3_col=&quot;IR_VDJ_1_junction_aa&quot;,
    color=&quot;initial_clustering&quot;,
    viztype=&quot;bar&quot;,
    fig_kws={&quot;dpi&quot;: 120},
    figsize=[10, 10],
)
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="../_images/9fffb9b0110e35ec0a14818ebd25d252b35c92a70137f45ed9003d0cee75c0c2.png" src="../_images/9fffb9b0110e35ec0a14818ebd25d252b35c92a70137f45ed9003d0cee75c0c2.png" />
</div>
</div>
<p>Furthermore, it is possible to visualize the V(D)J sequence length distributions individually for each cluster of cells, which provides a better perspective of those clusters closer to the most abundant sequence length. The plot below shows how the distribution of CD4+ and CD8+ is related to the previous analysis for general length abundances and how heterogenic the distribution per cluster is, for example, in the case of gdT-cells.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython2 notranslate"><div class="highlight"><pre><span></span>_ = ir.pl.spectratype(
    adata,
    cdr3_col=&quot;IR_VDJ_1_junction_aa&quot;,
    color=&quot;initial_clustering&quot;,
    viztype=&quot;curve&quot;,
    curve_layout=&quot;shifted&quot;,
    fig_kws={&quot;figsize&quot;: [10, 10]},
    kde_kws={&quot;kde_norm&quot;: False},
)
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>/home/juan.henao/.local/lib/python3.8/site-packages/scirpy/pl/base.py:262: UserWarning: FixedFormatter should only be used together with FixedLocator
  ax.set_yticklabels(order)
</pre></div>
</div>
<img alt="../_images/c51a8a265910d15a7e39cab00ff53f6528170a315491035ad3457451b16eb823.png" src="../_images/c51a8a265910d15a7e39cab00ff53f6528170a315491035ad3457451b16eb823.png" />
</div>
</div>
<p>By exploiting the commodities provided by the AnnData object, we can extract more specific information from spectratype analysis easily. For instance, we can visualize the distribution of sequence lengths regarding a specific V(D)J gene segment according to their fraction in the cluster of cells. Here, we selected the four V segments from V(D)J chain illustrated in the last gene usage plot. Once again, they were represented mainly by sequences of length 15. TRBV5-1 was the most abundant, followed by TRBV11-3, which was represented mainly by this specific length.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython2 notranslate"><div class="highlight"><pre><span></span>_ = ir.pl.spectratype(
    adata[
        adata.obs[&quot;IR_VDJ_1_v_call&quot;].isin(
            [&quot;TRBV5-1&quot;, &quot;TRBV11-2&quot;, &quot;TRBV7-2&quot;, &quot;TRBV11-3&quot;]
        ),
        :,
    ],
    cdr3_col=&quot;IR_VDJ_1_junction_aa&quot;,
    color=&quot;IR_VDJ_1_v_call&quot;,
    normalize=&quot;initial_clustering&quot;,
    fig_kws={&quot;dpi&quot;: 120},
    figsize=[10, 10],
)
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>/home/juan.henao/.local/lib/python3.8/site-packages/anndata/compat/_overloaded_dict.py:106: ImplicitModificationWarning: Trying to modify attribute `._uns` of view, initializing view as actual.
  self.data[key] = value
</pre></div>
</div>
<img alt="../_images/59a66dc973451c35647295cccedd4ee454a2ab0eaec2881af1623891e30cb787.png" src="../_images/59a66dc973451c35647295cccedd4ee454a2ab0eaec2881af1623891e30cb787.png" />
</div>
</div>
</section>
<section id="motif-sequence-analysis">
<h2><span class="section-number">39.9. </span>Motif sequence analysis<a class="headerlink" href="#motif-sequence-analysis" title="Link to this heading">#</a></h2>
<p>So far, we have analyzed different properties of the IR sequences, which provides us with information we can use to establish similarities. However, it is necessary to compare the sequences per se and to be able to detect those similarities at the amino acid-specific position level. One widely used tool to do that is the generation of logo plots. They show the amino acids per position, and the size of every letter indicates how much a specific letter is repeated across all the sequences under analysis.</p>
<p>Considering the information we have, we can create a logo plot for all the V(D)J sequences of length 15 with one of the five V segments analyzed in the previous step.</p>
<p>To perform this plot, we used the <em>palmotif</em> python library, this is implemented in other libraries for TCR analysis, such as <em>TCRdist3</em>. This library uses as input a list of sequences and saves the logo plot directly to a designated file path.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython2 notranslate"><div class="highlight"><pre><span></span>motif = compute_motif(
    adata[
        (
            adata.obs[&quot;IR_VDJ_1_v_call&quot;].isin(
                [&quot;TRBV5-1&quot;, &quot;TRBV11-2&quot;, &quot;TRBV7-2&quot;, &quot;TRBV11-3&quot;]
            )
        )
        &amp; (adata.obs[&quot;IR_VDJ_1_junction_aa&quot;].str.len() == 15),
        :,
    ]
    .obs[&quot;IR_VDJ_1_junction_aa&quot;]
    .to_list()
)
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython2 notranslate"><div class="highlight"><pre><span></span>_ = svg_logo(
    motif, &quot;../_static/images/air_repertoire/logo_motif.svg&quot;, color_scheme=&quot;taylor&quot;
)
</pre></div>
</div>
</div>
</div>
<p>The result was saved as <em>logo_motif.svg</em>, and the result is the plot you can see below.</p>
<p><img alt="" src="../_images/logo_motif.svg" /></p>
<p>As you can see, C-A-S are the three first amino acids for all the sequences analyzed in this example, as well F is the last amino acid for all the sequences. In addition, position 14 could be a Y, or an F, or an H, or an T. The fourth amino acid is most probable to be an S more than R or T. Those types of characterization are useful in protein profile discoveries, which are relevant in further experimental performance such as protein design.</p>
</section>
<section id="repertoire-comparison">
<h2><span class="section-number">39.10. </span>Repertoire comparison<a class="headerlink" href="#repertoire-comparison" title="Link to this heading">#</a></h2>
<p>The IR repertoires can be used to identify similarities between samples, which could help to understand the general response regarding the experimental perturbation. <em>Scirpy</em> allows the comparison between samples through the construction of a matrix with the abundances of clonotypes per sample (<code class="docutils literal notranslate"><span class="pre">df</span></code>), a Jaccard distance between samples (<code class="docutils literal notranslate"><span class="pre">dst</span></code>), and a linkage for hierarchical clustering (<code class="docutils literal notranslate"><span class="pre">lk</span></code>).</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython2 notranslate"><div class="highlight"><pre><span></span>df, dst, lk = ir.tl.repertoire_overlap(
    adata, &quot;patient_id&quot;, target_col=&quot;cc_aa_identity&quot;, inplace=False
)
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython2 notranslate"><div class="highlight"><pre><span></span>df
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th>cc_aa_identity</th>
      <th>0</th>
      <th>1</th>
      <th>2</th>
      <th>3</th>
      <th>4</th>
      <th>5</th>
      <th>6</th>
      <th>7</th>
      <th>8</th>
      <th>9</th>
      <th>...</th>
      <th>14171</th>
      <th>14172</th>
      <th>14173</th>
      <th>14174</th>
      <th>14175</th>
      <th>14176</th>
      <th>14177</th>
      <th>14178</th>
      <th>14179</th>
      <th>14180</th>
    </tr>
    <tr>
      <th>patient_id</th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>AP6</th>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>...</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
    </tr>
    <tr>
      <th>COVID-014</th>
      <td>71.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>1.0</td>
      <td>4.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>2.0</td>
      <td>0.0</td>
      <td>1.0</td>
      <td>...</td>
      <td>1.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>1.0</td>
      <td>1.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
    </tr>
    <tr>
      <th>COVID-045</th>
      <td>0.0</td>
      <td>1.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>1.0</td>
      <td>0.0</td>
      <td>1.0</td>
      <td>0.0</td>
      <td>...</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>1.0</td>
      <td>1.0</td>
      <td>1.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>1.0</td>
      <td>0.0</td>
    </tr>
    <tr>
      <th>COVID-066</th>
      <td>0.0</td>
      <td>0.0</td>
      <td>3.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>...</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
    </tr>
    <tr>
      <th>COVID-067</th>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>32.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>...</td>
      <td>0.0</td>
      <td>1.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>1.0</td>
      <td>0.0</td>
      <td>1.0</td>
    </tr>
    <tr>
      <th>CV0902</th>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>...</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
    </tr>
  </tbody>
</table>
<p>6 rows × 14181 columns</p>
</div></div></div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython2 notranslate"><div class="highlight"><pre><span></span>dst
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>array([1.        , 1.        , 1.        , 1.        , 1.        ,
       1.        , 1.        , 1.        , 1.        , 1.        ,
       1.        , 1.        , 1.        , 0.99982818, 1.        ])
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython2 notranslate"><div class="highlight"><pre><span></span>lk
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>array([[3.        , 5.        , 0.99982818, 2.        ],
       [0.        , 1.        , 1.        , 2.        ],
       [2.        , 7.        , 1.        , 3.        ],
       [6.        , 8.        , 1.        , 5.        ],
       [4.        , 9.        , 1.        , 6.        ]])
</pre></div>
</div>
</div>
</div>
<p>The previous data can be displayed as a heatmap which facilitates the interpretation of the results. For instance, the patients <em>COVID-066</em> and <em>CV0902</em> are the most similar, although with a low distance between them. Curiously both samples come from different centers, <em>COVID-066</em> from Newcastle and <em>CV0902</em> from Cambridge.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython2 notranslate"><div class="highlight"><pre><span></span>ir.pl.repertoire_overlap(
    adata, &quot;patient_id&quot;, target_col=&quot;cc_aa_identity&quot;, heatmap_cats=[&quot;Centre&quot;]
)
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>&lt;seaborn.matrix.ClusterGrid at 0x7ff53cc2e2b0&gt;
</pre></div>
</div>
<img alt="../_images/e366b2619e1b0cd9afadfad07179594eb45f6f66635ab4d34da2588390544f05.png" src="../_images/e366b2619e1b0cd9afadfad07179594eb45f6f66635ab4d34da2588390544f05.png" />
</div>
</div>
<p>Once sample similarities are detected, it is possible to analyze the pair of samples of interest. One way to do it is by comparing the clonotype sizes (number of cells) and the number of clonotypes (IDs) sharing a specific size. Visually, a scatter plot allows easy interpretations of the comparison result. Here, we compared the samples <em>COVID-067</em> versus <em>CV0902</em> in consideration of the heatmap interpretation above.</p>
<p>The scatterplot below shows that both samples are characterized by a high number of clonotypes with small size. Specifically, <em>COVID-067</em> presents a few clonotypes with high number of cells (size) than <em>CV0902</em>.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython2 notranslate"><div class="highlight"><pre><span></span>_ = ir.pl.repertoire_overlap(
    adata,
    &quot;patient_id&quot;,
    pair_to_plot=[&quot;COVID-067&quot;, &quot;CV0902&quot;],
    fig_kws={&quot;figsize&quot;: [10, 10]},
)
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>No handles with labels found to put in legend.
</pre></div>
</div>
<img alt="../_images/46c373923f4d5d157aa8b0e1dc91c93714959336f3160463b86ab43ef5f4adbf.png" src="../_images/46c373923f4d5d157aa8b0e1dc91c93714959336f3160463b86ab43ef5f4adbf.png" />
</div>
</div>
</section>
<section id="bcr-data-analysis-with-dandelion">
<h2><span class="section-number">39.11. </span>BCR Data Analysis with Dandelion<a class="headerlink" href="#bcr-data-analysis-with-dandelion" title="Link to this heading">#</a></h2>
<p>So far, we have shown all the analyses you can perform to characterize the T-cell receptor repertoire, including the cell clones’ identification and expansion. Besides, the representation in both, cell clusters and biological samples. Furthermore, the sequence motif for V(D)J gene segments, which is highlighted by interpretation of gene usage and spectratype results.</p>
<p>Those methods could be applied to characterize B-cell receptors as well <span id="id10">[<a class="reference internal" href="#id530" title="Namita T Gupta, Jason A Vander Heiden, Mohamed Uduman, Daniel Gadala-Maria, Gur Yaari, and Steven H Kleinstein. Change-o: a toolkit for analyzing large-scale b cell immunoglobulin repertoire sequencing data. Bioinformatics, 31(20):3356–3358, 2015.">Gupta <em>et al.</em>, 2015</a>]</span>. However, over the lifetime of B-cells mutual mutations occur in the V gene segment helping the low-affinity receptors to acquire a high affinity phenotype. This process is known as <strong>affinity maturation</strong>, and the high rate of mutual mutations (~10000 more than germline cells) is called <strong>somatic hypermutation</strong> <span id="id11">[<a class="reference internal" href="#id529" title="F Nina Papavasiliou and David G Schatz. Somatic hypermutation of immunoglobulin genes: merging mechanisms for genetic diversity. Cell, 109(2):S35–S44, 2002.">Papavasiliou and Schatz, 2002</a>]</span>. Therefore, the clonotype definition for B-cells should take this phenomenon into account. One way to deal with this is through distance-based clonotype analysis.</p>
<p>Here, we use <strong>Dandelion</strong>, a python library focused on BCR analysis which interoperates with <em>Scanpy</em> and <em>Scirpy</em> and provides a BCR distance-based method for clone definition, which is explained below in more detail <span id="id12">[<a class="reference internal" href="ir_profiling.html#id479" title="Emily Stephenson, Gary Reynolds, Rachel A Botting, Fernando J Calero-Nieto, Michael D Morgan, Zewen Kelvin Tuong, Karsten Bach, Waradon Sungnak, Kaylee B Worlock, Masahiro Yoshida, and others. Single-cell multi-omics analysis of the immune response in covid-19. Nature medicine, 27(5):904–916, 2021.">Stephenson <em>et al.</em>, 2021</a>]</span>.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython2 notranslate"><div class="highlight"><pre><span></span>import warnings

warnings.filterwarnings(
    &quot;ignore&quot;,
    &quot;.*IProgress not found*&quot;,
)
warnings.simplefilter(action=&quot;ignore&quot;, category=FutureWarning)

import dandelion as ddl
import matplotlib as mpl
import matplotlib.pyplot as plt
import pandas as pd
import scanpy as sc
import scirpy as ir
from palmotif import compute_motif, svg_logo

warnings.simplefilter(action=&quot;ignore&quot;, category=pd.errors.DtypeWarning)
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython2 notranslate"><div class="highlight"><pre><span></span>sc.logging.print_versions()
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>WARNING: If you miss a compact list, please try `print_header`!
The `sinfo` package has changed name and is now called `session_info` to become more discoverable and self-explanatory. The `sinfo` PyPI package will be kept around to avoid breaking old installs and you can downgrade to 0.3.2 if you want to use it without seeing this message. For the latest features and bug fixes, please install `session_info` instead. The usage and defaults also changed slightly, so please review the latest README at https://gitlab.com/joelostblom/session_info.
-----
anndata     0.8.0
scanpy      1.8.2
sinfo       0.3.4
-----
Bio                         1.79
Levenshtein                 NA
PIL                         8.4.0
adjustText                  NA
airr                        1.4.1
anyio                       NA
asciitree                   NA
attr                        21.2.0
babel                       2.9.1
backcall                    0.2.0
beta_ufunc                  NA
binom_ufunc                 NA
brotli                      1.0.9
certifi                     2021.10.08
cffi                        1.15.0
changeo                     1.2.0
chardet                     4.0.0
charset_normalizer          2.0.8
cloudpickle                 2.0.0
colorama                    0.4.4
cycler                      0.10.0
cython_runtime              NA
dandelion                   0.2.4
dask                        2021.11.2
dateutil                    2.8.2
debugpy                     1.5.1
decorator                   5.1.0
defusedxml                  0.7.1
distance                    NA
entrypoints                 0.3
fasteners                   NA
fontTools                   4.28.2
fsspec                      0.7.4
google                      NA
h5py                        3.6.0
idna                        3.3
igraph                      0.9.8
importlib_resources         NA
ipykernel                   6.5.1
ipython_genutils            0.2.0
ipywidgets                  7.6.5
jedi                        0.18.1
jinja2                      3.0.3
joblib                      1.1.0
json5                       NA
jsonschema                  4.2.1
jupyter_server              1.12.1
jupyterlab_server           2.8.2
kiwisolver                  1.3.2
leidenalg                   0.8.8
llvmlite                    0.36.0
louvain                     0.7.0
lxml                        4.6.4
markupsafe                  2.0.1
matplotlib                  3.5.0
matplotlib_inline           NA
mizani                      0.7.4
mpl_toolkits                NA
natsort                     8.0.0
nbclassic                   NA
nbformat                    5.1.3
nbinom_ufunc                NA
networkx                    2.6.3
numba                       0.53.1
numcodecs                   0.9.1
numexpr                     2.7.3
numpy                       1.21.4
packaging                   21.3
palettable                  3.3.0
palmotif                    NA
pandas                      1.5.0
parasail                    1.3.3
parso                       0.8.2
patsy                       0.5.2
pexpect                     4.8.0
pickleshare                 0.7.5
pkg_resources               NA
plotnine                    0.9.0
polyleven                   NA
presto                      0.7.0
prometheus_client           NA
prompt_toolkit              3.0.23
ptyprocess                  0.7.0
pvectorc                    NA
pyarrow                     6.0.1
pydev_ipython               NA
pydevconsole                NA
pydevd                      2.6.0
pydevd_concurrency_analyser NA
pydevd_file_utils           NA
pydevd_plugins              NA
pydevd_tracing              NA
pygments                    2.10.0
pyparsing                   3.0.6
pyrsistent                  NA
pytoml                      NA
pytz                        2021.3
requests                    2.26.0
scipy                       1.7.3
scirpy                      0.11.1
seaborn                     0.11.2
send2trash                  NA
setuptools_scm              NA
six                         1.16.0
sklearn                     1.1.2
sniffio                     1.2.0
sparse                      0.13.0
statsmodels                 0.13.2
storemagic                  NA
svgwrite                    1.4.3
tables                      3.6.1
terminado                   0.12.1
texttable                   1.6.4
threadpoolctl               3.0.0
tlz                         0.11.2
toolz                       0.11.2
tornado                     6.1
tqdm                        4.62.3
tracerlib                   NA
traitlets                   5.1.1
typing_extensions           NA
urllib3                     1.26.7
wcwidth                     0.2.5
websocket                   1.2.1
yaml                        6.0
yamlordereddictloader       NA
zarr                        2.10.3
zipp                        NA
zmq                         22.3.0
-----
IPython             7.30.0
jupyter_client      7.1.0
jupyter_core        4.9.1
jupyterlab          3.2.4
notebook            6.4.6
-----
Python 3.8.12 (default, Nov 26 2021, 20:28:57) [GCC 10.2.1 20210110]
Linux-3.10.0-1160.25.1.el7.x86_64-x86_64-with-glibc2.29
48 logical CPU cores
-----
Session information updated at 2022-09-20 21:27
</pre></div>
</div>
</div>
</div>
<p>As well as the TCR analysis described previously, we need to load the data as <em>.h5ad</em> format obtained after the pre-processing step. Besides, to improve the runtime of this tutorial, we will work with a subset of samples that allow us to explore the BCR receptors properly and perform the necessary comparisons between conditions.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython2 notranslate"><div class="highlight"><pre><span></span>path_data = &quot;/home/icb/juan.henao/BestPracticeStart/data&quot;

path_gex = f&quot;{path_data}/BCR_filtered.h5ad&quot;
adata_bcr = sc.read(path_gex)
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython2 notranslate"><div class="highlight"><pre><span></span>adata = adata_bcr[adata_bcr.obs[&quot;patient_id&quot;].isin([&quot;COVID-064&quot;, &quot;COVID-014&quot;])].copy()
</pre></div>
</div>
</div>
</div>
</section>
<section id="dandelion-interoperability">
<h2><span class="section-number">39.12. </span>Dandelion interoperability<a class="headerlink" href="#dandelion-interoperability" title="Link to this heading">#</a></h2>
<p>Dandelion uses its own object to locate the BCR data. Therefore, this package provides a function to translate to annData and be able to use the functionalities from Scanpy and Scirpy.</p>
<p>Previously, we loaded our pre-processed data into an annData object. Therefore, we should relocate the information into the correct format to use Dandelion. The next piece of code shows how easy is the interoperability between both objects.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython2 notranslate"><div class="highlight"><pre><span></span>vdjx = ddl.from_scirpy(adata)
vdjx
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>WARNING: Non-standard locus name ignored: Multi 
</pre></div>
</div>
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Dandelion class object with n_obs = 6366 and n_contigs = 16953
    data: &#39;sequence_id&#39;, &#39;sequence&#39;, &#39;rev_comp&#39;, &#39;productive&#39;, &#39;v_call&#39;, &#39;d_call&#39;, &#39;j_call&#39;, &#39;sequence_alignment&#39;, &#39;germline_alignment&#39;, &#39;junction&#39;, &#39;junction_aa&#39;, &#39;v_cigar&#39;, &#39;d_cigar&#39;, &#39;j_cigar&#39;, &#39;c_call&#39;, &#39;consensus_count&#39;, &#39;duplicate_count&#39;, &#39;locus&#39;, &#39;cell_id&#39;, &#39;multi_chain&#39;, &#39;receptor_subtype&#39;, &#39;chain_pairing&#39;, &#39;receptor_type&#39;, &#39;is_cell&#39;, &#39;high_confidence&#39;, &#39;patient_id&#39;, &#39;rearrangement_status&#39;
    metadata: &#39;locus_VDJ&#39;, &#39;locus_VJ&#39;, &#39;productive_VDJ&#39;, &#39;productive_VJ&#39;, &#39;v_call_VDJ&#39;, &#39;d_call_VDJ&#39;, &#39;j_call_VDJ&#39;, &#39;v_call_VJ&#39;, &#39;j_call_VJ&#39;, &#39;c_call_VDJ&#39;, &#39;c_call_VJ&#39;, &#39;junction_VDJ&#39;, &#39;junction_VJ&#39;, &#39;junction_aa_VDJ&#39;, &#39;junction_aa_VJ&#39;, &#39;v_call_B_VDJ&#39;, &#39;d_call_B_VDJ&#39;, &#39;j_call_B_VDJ&#39;, &#39;v_call_B_VJ&#39;, &#39;j_call_B_VJ&#39;, &#39;c_call_B_VDJ&#39;, &#39;c_call_B_VJ&#39;, &#39;productive_B_VDJ&#39;, &#39;productive_B_VJ&#39;, &#39;duplicate_count_B_VDJ&#39;, &#39;duplicate_count_B_VJ&#39;, &#39;isotype&#39;, &#39;isotype_status&#39;, &#39;locus_status&#39;, &#39;chain_status&#39;, &#39;rearrangement_status_VDJ&#39;, &#39;rearrangement_status_VJ&#39;
</pre></div>
</div>
</div>
</div>
</section>
<section id="id13">
<h2><span class="section-number">39.13. </span>Clonotype definition<a class="headerlink" href="#id13" title="Link to this heading">#</a></h2>
<p>Before the clonotype definition for our BCR data, the empty cells (<em>NA data</em>) should be removed, and receptor lengths must be calculated.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython2 notranslate"><div class="highlight"><pre><span></span>vdjx.data[&quot;v_call&quot;].replace(&quot;&quot;, np.nan, inplace=True)
vdjx.data.dropna(subset=[&quot;v_call&quot;], inplace=True)

vdjx.data[&quot;j_call&quot;].replace(&quot;&quot;, np.nan, inplace=True)
vdjx.data.dropna(subset=[&quot;j_call&quot;], inplace=True)

vdjx.data[&quot;junction_aa&quot;].replace(&quot;&quot;, np.nan, inplace=True)
vdjx.data.dropna(subset=[&quot;junction_aa&quot;], inplace=True)

vdjx.data[&quot;junction_length&quot;] = [len(a) for a in vdjx.data[&quot;junction_aa&quot;]]
</pre></div>
</div>
</div>
</div>
<p>Dandelion defines clonotypes using a substitution model based on distances. It was created specifically to deal with the problem of somatic hypermutation in B-cells <span id="id14">[<a class="reference internal" href="#id532" title="Gur Yaari, Jason A Vander Heiden, Mohamed Uduman, Daniel Gadala-Maria, Namita Gupta, Joel NH Stern, Kevin C O’Connor, David A Hafler, Uri Laserson, Francois Vigneault, and others. Models of somatic hypermutation targeting and substitution based on synonymous mutations from high-throughput immunoglobulin sequencing data. Frontiers in immunology, 4:358, 2013.">Yaari <em>et al.</em>, 2013</a>]</span> <span id="id15">[<a class="reference internal" href="#id533" title="Ang Cui, Roberto Di Niro, Jason A Vander Heiden, Adrian W Briggs, Kris Adams, Tamara Gilbert, Kevin C O’Connor, Francois Vigneault, Mark J Shlomchik, and Steven H Kleinstein. A model of somatic hypermutation targeting in mice based on high-throughput ig sequencing data. The Journal of Immunology, 197(9):3566–3574, 2016.">Cui <em>et al.</em>, 2016</a>]</span>. This model was available in the <strong>Immcantation</strong> suite as an R package <span id="id16">[<a class="reference internal" href="#id530" title="Namita T Gupta, Jason A Vander Heiden, Mohamed Uduman, Daniel Gadala-Maria, Gur Yaari, and Steven H Kleinstein. Change-o: a toolkit for analyzing large-scale b cell immunoglobulin repertoire sequencing data. Bioinformatics, 31(20):3356–3358, 2015.">Gupta <em>et al.</em>, 2015</a>]</span> <span id="id17">[<a class="reference internal" href="#id531" title="Jason A Vander Heiden, Gur Yaari, Mohamed Uduman, Joel NH Stern, Kevin C O’Connor, David A Hafler, Francois Vigneault, and Steven H Kleinstein. Presto: a toolkit for processing high-throughput sequencing raw reads of lymphocyte receptor repertoires. Bioinformatics, 30(13):1930–1932, 2014.">Vander Heiden <em>et al.</em>, 2014</a>]</span>. However, Dandelion makes possible to use it, avoiding the complication of moving between code languages and keeping the interoperability with <em>Scanpy</em> and <em>Scirpy</em>.</p>
<p>The model was created based on the probability of a punctual nucleotide change, considering the influence of the immediate two down- and upstream nucleotides <span id="id18">[<a class="reference internal" href="#id532" title="Gur Yaari, Jason A Vander Heiden, Mohamed Uduman, Daniel Gadala-Maria, Namita Gupta, Joel NH Stern, Kevin C O’Connor, David A Hafler, Uri Laserson, Francois Vigneault, and others. Models of somatic hypermutation targeting and substitution based on synonymous mutations from high-throughput immunoglobulin sequencing data. Frontiers in immunology, 4:358, 2013.">Yaari <em>et al.</em>, 2013</a>]</span>. This methodology considered all the possible different 5-mers combinations just for the synonym mutation cases, i.e., those changes where the amino acid represented by the <a class="reference internal" href="../glossary.html#term-Codon"><span class="xref std std-term">codon</span></a> is not modified <span id="id19">[<a class="reference internal" href="#id532" title="Gur Yaari, Jason A Vander Heiden, Mohamed Uduman, Daniel Gadala-Maria, Namita Gupta, Joel NH Stern, Kevin C O’Connor, David A Hafler, Uri Laserson, Francois Vigneault, and others. Models of somatic hypermutation targeting and substitution based on synonymous mutations from high-throughput immunoglobulin sequencing data. Frontiers in immunology, 4:358, 2013.">Yaari <em>et al.</em>, 2013</a>]</span>.</p>
<p>Furthermore, Dandelion considers a model of substitution rates for single nucleotide instead of the 5-mer model. Therefore, all the substitutions are not changing, and they are displayed in the table below:</p>
<ul class="simple">
<li><p>Human substitution model <span id="id20">[<a class="reference internal" href="#id532" title="Gur Yaari, Jason A Vander Heiden, Mohamed Uduman, Daniel Gadala-Maria, Namita Gupta, Joel NH Stern, Kevin C O’Connor, David A Hafler, Uri Laserson, Francois Vigneault, and others. Models of somatic hypermutation targeting and substitution based on synonymous mutations from high-throughput immunoglobulin sequencing data. Frontiers in immunology, 4:358, 2013.">Yaari <em>et al.</em>, 2013</a>]</span>:</p></li>
</ul>
<div class="pst-scrollable-table-container"><table class="table">
<thead>
<tr class="row-odd"><th class="head"><p>Nucleotide</p></th>
<th class="head"><p>A</p></th>
<th class="head"><p>C</p></th>
<th class="head"><p>G</p></th>
<th class="head"><p>T</p></th>
<th class="head"><p>N</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>A</p></td>
<td><p>0</p></td>
<td><p>1.21</p></td>
<td><p>0.64</p></td>
<td><p>1.16</p></td>
<td><p>0</p></td>
</tr>
<tr class="row-odd"><td><p>C</p></td>
<td><p>1.21</p></td>
<td><p>0</p></td>
<td><p>1.16</p></td>
<td><p>0.64</p></td>
<td><p>0</p></td>
</tr>
<tr class="row-even"><td><p>G</p></td>
<td><p>0.64</p></td>
<td><p>1.16</p></td>
<td><p>0</p></td>
<td><p>1.21</p></td>
<td><p>0</p></td>
</tr>
<tr class="row-odd"><td><p>T</p></td>
<td><p>1.16</p></td>
<td><p>0.64</p></td>
<td><p>1.21</p></td>
<td><p>0</p></td>
<td><p>0</p></td>
</tr>
<tr class="row-even"><td><p>N</p></td>
<td><p>0</p></td>
<td><p>0</p></td>
<td><p>0</p></td>
<td><p>0</p></td>
<td><p>0</p></td>
</tr>
</tbody>
</table>
</div>
<ul class="simple">
<li><p>Mice substitution model <span id="id21">[<a class="reference internal" href="#id533" title="Ang Cui, Roberto Di Niro, Jason A Vander Heiden, Adrian W Briggs, Kris Adams, Tamara Gilbert, Kevin C O’Connor, Francois Vigneault, Mark J Shlomchik, and Steven H Kleinstein. A model of somatic hypermutation targeting in mice based on high-throughput ig sequencing data. The Journal of Immunology, 197(9):3566–3574, 2016.">Cui <em>et al.</em>, 2016</a>]</span>:</p></li>
</ul>
<div class="pst-scrollable-table-container"><table class="table">
<thead>
<tr class="row-odd"><th class="head"><p>Nucleotide</p></th>
<th class="head"><p>A</p></th>
<th class="head"><p>C</p></th>
<th class="head"><p>G</p></th>
<th class="head"><p>T</p></th>
<th class="head"><p>N</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>A</p></td>
<td><p>0</p></td>
<td><p>1.51</p></td>
<td><p>0.32</p></td>
<td><p>1.17</p></td>
<td><p>0</p></td>
</tr>
<tr class="row-odd"><td><p>C</p></td>
<td><p>1.51</p></td>
<td><p>0</p></td>
<td><p>1.17</p></td>
<td><p>0.32</p></td>
<td><p>0</p></td>
</tr>
<tr class="row-even"><td><p>G</p></td>
<td><p>0.32</p></td>
<td><p>1.17</p></td>
<td><p>0</p></td>
<td><p>1.51</p></td>
<td><p>0</p></td>
</tr>
<tr class="row-odd"><td><p>T</p></td>
<td><p>1.17</p></td>
<td><p>0.32</p></td>
<td><p>1.51</p></td>
<td><p>0</p></td>
<td><p>0</p></td>
</tr>
<tr class="row-even"><td><p>N</p></td>
<td><p>0</p></td>
<td><p>0</p></td>
<td><p>0</p></td>
<td><p>0</p></td>
<td><p>0</p></td>
</tr>
</tbody>
</table>
</div>
<p>The clonotype definition based on distance requires a threshold value definition to allow the separation into close related receptor sequences, which are interpreted as clonal related. Dandelion wrapped the functionalities from <strong>Immcantation</strong> to choose this value based on distance calculation for those receptor sequences with the same V and J segment and the same length. The distribution of the distances is expected to be bimodal, being the first mode composed by closed-related sequences, and the second mode by singletons <span id="id22">[<a class="reference internal" href="#id530" title="Namita T Gupta, Jason A Vander Heiden, Mohamed Uduman, Daniel Gadala-Maria, Gur Yaari, and Steven H Kleinstein. Change-o: a toolkit for analyzing large-scale b cell immunoglobulin repertoire sequencing data. Bioinformatics, 31(20):3356–3358, 2015.">Gupta <em>et al.</em>, 2015</a>]</span>. The threshold is selected as the breakpoint between both modes.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython2 notranslate"><div class="highlight"><pre><span></span>ddl.pp.calculate_threshold(vdjx, model=&quot;hh_s5f&quot;, plot=False)
vdjx.threshold
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>/opt/python/lib/python3.8/site-packages/rpy2/robjects/pandas2ri.py:59: UserWarning: Error while trying to convert the column &quot;productive&quot;. Fall back to string conversion. The error is: Series can only be of one type, or None (and here we have &lt;class &#39;str&#39;&gt; and &lt;class &#39;bool&#39;&gt;).
R[write to console]: Error in (function (db, sequenceColumn = &quot;junction&quot;, vCallColumn = &quot;v_call&quot;,  : 
  The locus column contains invalid loci annotations.
</pre></div>
</div>
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>0.13319079238387194
</pre></div>
</div>
</div>
</div>
<p>We have already chosen the model to calculate the distances between receptors and performed the threshold value selection. Therefore, the next step is the clonotype definition by itself.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython2 notranslate"><div class="highlight"><pre><span></span>ddl.tl.define_clones(vdjx, key_added=&quot;changeo_clone_id&quot;, model=&quot;hh_s5f&quot;)
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>       START&gt; DefineClones
        FILE&gt; dandelion_define_clones_heavy-clone.tsv
   SEQ_FIELD&gt; junction
     V_FIELD&gt; v_call
     J_FIELD&gt; j_call
 MAX_MISSING&gt; 0
GROUP_FIELDS&gt; None
      ACTION&gt; set
        MODE&gt; gene
    DISTANCE&gt; 0.13319079238387194
     LINKAGE&gt; single
       MODEL&gt; hh_s5f
        NORM&gt; len
         SYM&gt; avg
       NPROC&gt; 1

PROGRESS&gt; [Grouping sequences] 21:28:26 (6828) 0.0 min

PROGRESS&gt; [Assigning clones] 21:28:31 |####################| 100% (6,828) 0.1 min

 OUTPUT&gt; dandelion_define_clones_heavy-clone.tsv
 CLONES&gt; 4741
RECORDS&gt; 6828
   PASS&gt; 6828
   FAIL&gt; 0
    END&gt; DefineClones
</pre></div>
</div>
</div>
</div>
<p>Once the clonotypes are defined, we need to create the network layout for visualization. Dandelion, as well as <em>Scirpy</em>, use <em>igraph</em> layouts to visualize clonotype networks. Here, we used <strong>sfdp</strong> because it is computationally less expensive than the rest of the methods available.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython2 notranslate"><div class="highlight"><pre><span></span>ddl.tl.generate_network(
    vdjx, key=&quot;sequence_alignment&quot;, layout_method=&quot;sfdp&quot;, clone_key=&quot;changeo_clone_id&quot;
)
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Setting up data: 14530it [00:01, 10889.53it/s]
Calculating distances... : 100%|██████████| 4720/4720 [00:02&lt;00:00, 1758.62it/s]                                                                     
Generating edge list :   0%|          | 0/220 [00:00&lt;?, ?it/s]                                                                                       /home/juan.henao/.local/lib/python3.8/site-packages/dandelion/tools/_network.py:259: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
Generating edge list : 100%|██████████| 220/220 [00:00&lt;00:00, 852.52it/s]                                                                            
Computing overlap : 100%|██████████| 4721/4721 [00:05&lt;00:00, 846.28it/s]                                                                             
Linking edges : 100%|██████████| 4450/4450 [00:02&lt;00:00, 1500.32it/s]                                                                                
</pre></div>
</div>
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>To benefit from faster layout computation, please install graph-tool: conda install -c conda-forge graph-tool
</pre></div>
</div>
</div>
</div>
<p>To visualize the results, <em>Dandelion</em> takes advantage of the interoperability with <em>Scanpy</em>. Therefore, it is necessary to do the translation from the Dandelion into an annData object to perform the visualization. Additionally, considering that the clonotype definition was based on a distance between the different BCR sequences, it is useful to add an additional feature related to the weight (distance) between BCRs (nodes), improving the final plot observation.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython2 notranslate"><div class="highlight"><pre><span></span>ddl.tl.transfer(adata, vdjx, clone_key=&quot;changeo_clone_id&quot;, expanded_only=True)

edgeweights = [
    1 / (e + 1) for e in ddl.tl.extract_edge_weights(vdjx)
]  # invert and add 1 to each edge weight (e) so that distance of 0 becomes the thickest edge
# therefore, the thicker the line, the shorter the edit distance.

sc.set_figure_params(figsize=[10, 10])
_ = ddl.pl.clone_network(
    adata,
    color=[&quot;isotype_status&quot;],
    legend_fontoutline=3,
    edges_width=edgeweights,
    size=50,
)
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>... storing &#39;changeo_clone_id&#39; as categorical
... storing &#39;locus_VDJ&#39; as categorical
... storing &#39;locus_VJ&#39; as categorical
... storing &#39;productive_VDJ&#39; as categorical
... storing &#39;productive_VJ&#39; as categorical
... storing &#39;v_call_VDJ&#39; as categorical
... storing &#39;d_call_VDJ&#39; as categorical
... storing &#39;j_call_VDJ&#39; as categorical
... storing &#39;v_call_VJ&#39; as categorical
... storing &#39;j_call_VJ&#39; as categorical
... storing &#39;c_call_VDJ&#39; as categorical
... storing &#39;c_call_VJ&#39; as categorical
... storing &#39;junction_VDJ&#39; as categorical
... storing &#39;junction_VJ&#39; as categorical
... storing &#39;junction_aa_VDJ&#39; as categorical
... storing &#39;junction_aa_VJ&#39; as categorical
... storing &#39;v_call_B_VDJ&#39; as categorical
... storing &#39;d_call_B_VDJ&#39; as categorical
... storing &#39;j_call_B_VDJ&#39; as categorical
... storing &#39;v_call_B_VJ&#39; as categorical
... storing &#39;j_call_B_VJ&#39; as categorical
... storing &#39;c_call_B_VDJ&#39; as categorical
... storing &#39;c_call_B_VJ&#39; as categorical
... storing &#39;productive_B_VDJ&#39; as categorical
... storing &#39;productive_B_VJ&#39; as categorical
... storing &#39;isotype&#39; as categorical
... storing &#39;isotype_status&#39; as categorical
... storing &#39;locus_status&#39; as categorical
... storing &#39;chain_status&#39; as categorical
... storing &#39;rearrangement_status_VDJ&#39; as categorical
... storing &#39;rearrangement_status_VJ&#39; as categorical
</pre></div>
</div>
<a class="reference internal image-reference" href="../_images/7fc779371d1c59e463865e913b61065b782c7650b86b64ae1a3551eb0326380d.png"><img alt="../_images/7fc779371d1c59e463865e913b61065b782c7650b86b64ae1a3551eb0326380d.png" src="../_images/7fc779371d1c59e463865e913b61065b782c7650b86b64ae1a3551eb0326380d.png" style="width: 745px; height: 661px;" />
</a>
</div>
</div>
<p>Compared to <em>Scirpy</em>, the clonotypes visualization in <em>Dandelion</em> does not show their sizes (number of cells). This process should be done separately, i.e., first, it is necessary to calculate the size of the clones and transfer this information to the <em>annData</em> object to perform the visualization via <em>Scanpy</em>.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython2 notranslate"><div class="highlight"><pre><span></span>ddl.tl.clone_size(vdjx, clone_key=&quot;changeo_clone_id&quot;)
ddl.tl.transfer(adata, vdjx, clone_key=&quot;changeo_clone_id&quot;)
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython2 notranslate"><div class="highlight"><pre><span></span>sc.set_figure_params(figsize=[10, 10])

_ = ddl.pl.clone_network(
    adata,
    color=[&quot;changeo_clone_id_size&quot;],
    legend_loc=&quot;none&quot;,
    legend_fontoutline=3,
    edges_width=1,
    size=10,
)
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<a class="reference internal image-reference" href="../_images/060ac2389258dc9340416fea6426e6691e4dcdbdd4a2737d81283c36064d7497.png"><img alt="../_images/060ac2389258dc9340416fea6426e6691e4dcdbdd4a2737d81283c36064d7497.png" src="../_images/060ac2389258dc9340416fea6426e6691e4dcdbdd4a2737d81283c36064d7497.png" style="width: 666px; height: 661px;" />
</a>
</div>
</div>
<p>Additionally, it is possible to fix a maximum size to establish a threshold to get a grouped visualization of the clonotypes.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython2 notranslate"><div class="highlight"><pre><span></span>ddl.tl.clone_size(vdjx, clone_key=&quot;changeo_clone_id&quot;, max_size=50)
ddl.tl.transfer(adata, vdjx, clone_key=&quot;changeo_clone_id&quot;)
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython2 notranslate"><div class="highlight"><pre><span></span>_ = ddl.pl.clone_network(
    adata,
    color=[&quot;changeo_clone_id_size_max_50&quot;],
    ncols=2,
    legend_fontoutline=3,
    edges_width=1,
    palette=&quot;tab20c&quot;,
    size=20,
)
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<a class="reference internal image-reference" href="../_images/ea4b1468f31330c35b18631a6ef007b33c37fa557dcf349e98b2bef2a62ed331.png"><img alt="../_images/ea4b1468f31330c35b18631a6ef007b33c37fa557dcf349e98b2bef2a62ed331.png" src="../_images/ea4b1468f31330c35b18631a6ef007b33c37fa557dcf349e98b2bef2a62ed331.png" style="width: 796px; height: 661px;" />
</a>
</div>
</div>
</section>
</section>
<section class="tex2jax_ignore mathjax_ignore" id="gene-segment-usage">
<h1><span class="section-number">40. </span>Gene segment usage<a class="headerlink" href="#gene-segment-usage" title="Link to this heading">#</a></h1>
<p>As well as TCR analysis, we can observe the gene segment preference usage. <em>Dandelion</em> allows analyzing this phenomenon by visualizing the abundance of specific segments and chains.</p>
<p>Let us take the V gene segment in the heavy chain (V(D)J) for the clonotypes composed of more than 50 cells. Notice we are ignoring multi-chain, given they are not viable receptors. <em>Dandelion</em> by default, do not remove them. It could be a source of misinterpretations.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython2 notranslate"><div class="highlight"><pre><span></span>mpl.rcParams.update(mpl.rcParamsDefault)
_ = ddl.pl.barplot(
    vdjx[vdjx.metadata.isotype_status != &quot;Multi&quot;],
    clone_key=&quot;changeo_clone_id&quot;,
    color=&quot;v_call_VDJ&quot;,
    min_clone_size=50,
    figsize=[15, 8],
)
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<a class="reference internal image-reference" href="../_images/07403c5c961e54c35586c4eaf165f7ad75554eab07964b926cc56396911ff657.png"><img alt="../_images/07403c5c961e54c35586c4eaf165f7ad75554eab07964b926cc56396911ff657.png" src="../_images/07403c5c961e54c35586c4eaf165f7ad75554eab07964b926cc56396911ff657.png" style="width: 1233px; height: 875px;" />
</a>
</div>
</div>
<p>As you can appreciate before, <strong>IGHV3-48</strong> and <strong>IGHV1-18</strong> were the gene segments consistently more abundant in comparison to the rest of the segments in the plot, providing evidence of strong V gene preferential usage for the samples analyzed here.</p>
<p>The previous analysis can be improved by just adding information for the visualization. For example, let us see if those privilege V segments are shared between isotypes.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython2 notranslate"><div class="highlight"><pre><span></span>_ = (
    ddl.pl.stackedbarplot(
        vdjx[vdjx.metadata.isotype_status != &quot;Multi&quot;],
        clone_key=&quot;changeo_clone_id&quot;,
        color=&quot;v_call_VDJ&quot;,
        groupby=&quot;isotype_status&quot;,
        min_clone_size=50,
        xtick_rotation=90,
        figsize=(18, 8),
        normalize=True,
    ),
)

_ = plt.legend(bbox_to_anchor=(1, 1), loc=&quot;upper left&quot;, frameon=False)
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<a class="reference internal image-reference" href="../_images/fabde315ee8f0e39f9de6c1938849fc9558d5e49b9ccb334e5c2a8d681ac5d9f.png"><img alt="../_images/fabde315ee8f0e39f9de6c1938849fc9558d5e49b9ccb334e5c2a8d681ac5d9f.png" src="../_images/fabde315ee8f0e39f9de6c1938849fc9558d5e49b9ccb334e5c2a8d681ac5d9f.png" style="width: 1576px; height: 875px;" />
</a>
</div>
</div>
<p>The plot above showed that IgG mostly used the V segments <strong>IGHV3-48</strong> and <strong>IGHV1-18</strong>. It could be due to IgG being the isotype most expanded. Therefore, it masked the gene segment usage from the rest of the isotypes.</p>
<p>This fact was supported by the plot showed below. The abundance of isotypes in the biggest clonotypes (size &gt; 50 cells) is represented by IgG almost entirely. However, this plot provided additional evidence about the immune response triggered by the samples.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython2 notranslate"><div class="highlight"><pre><span></span>_ = ddl.pl.barplot(
    vdjx[vdjx.metadata.isotype_status != &quot;Multi&quot;],
    clone_key=&quot;changeo_clone_id&quot;,
    color=&quot;isotype_status&quot;,
    min_clone_size=50,
    figsize=[10, 10],
)
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<a class="reference internal image-reference" href="../_images/31ea8330f023c1ba0496af6deb89e4f2546b53131a6ee60cef0ca15b5d654be2.png"><img alt="../_images/31ea8330f023c1ba0496af6deb89e4f2546b53131a6ee60cef0ca15b5d654be2.png" src="../_images/31ea8330f023c1ba0496af6deb89e4f2546b53131a6ee60cef0ca15b5d654be2.png" style="width: 846px; height: 874px;" />
</a>
</div>
</div>
</section>
<section class="tex2jax_ignore mathjax_ignore" id="spectratype">
<h1><span class="section-number">41. </span>Spectratype<a class="headerlink" href="#spectratype" title="Link to this heading">#</a></h1>
<p>We have identified key expanded clonotypes and the isotype they represented. In addition, we can explore spectratype to observe the dominance in terms of sequence length. As well as in the previous analysis, we discarded the multi-chain cells, and we conserved those clonotypes whose sizes were higher than 50 cells to keep the analysis consistency.</p>
<p>The plot below shows an interesting behaviour, despite the clear spectratype dominance reflected in our previous TCR analysis. Here, two sequence lengths raised, the first and the most dominant conformed by sequences of 23 amino acids, and the second one composed by 15 amino acids.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython2 notranslate"><div class="highlight"><pre><span></span>_ = ddl.pl.spectratype(
    vdjx[
        (vdjx.metadata.isotype_status != &quot;Multi&quot;)
        &amp; (vdjx.metadata.changeo_clone_id_size &gt; 50.0)
    ],
    color=&quot;junction_length&quot;,
    groupby=&quot;v_call&quot;,
    locus=&quot;IGH&quot;,
    figsize=(10, 10),
)
_ = plt.legend(bbox_to_anchor=(1, 1), loc=&quot;upper left&quot;, frameon=False)
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<a class="reference internal image-reference" href="../_images/927abe134191bd3287882a21481f4232c4f95a2d7676f3bd049720dd26b12e0b.png"><img alt="../_images/927abe134191bd3287882a21481f4232c4f95a2d7676f3bd049720dd26b12e0b.png" src="../_images/927abe134191bd3287882a21481f4232c4f95a2d7676f3bd049720dd26b12e0b.png" style="width: 1008px; height: 833px;" />
</a>
</div>
</div>
</section>
<section class="tex2jax_ignore mathjax_ignore" id="id23">
<h1><span class="section-number">42. </span>Motif sequence analysis<a class="headerlink" href="#id23" title="Link to this heading">#</a></h1>
<p>The results from spectratype gave us a hint about which sequences to compare, looking for a possible amino acid motif. In this case, we analyzed the V segment for <strong>IGHV3-48</strong> and <strong>IGHV1-18</strong> with a sequence length of 15 and 23 amino acids. Let us start with the first comparison (length = 15).</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython2 notranslate"><div class="highlight"><pre><span></span>ddl.tl.transfer(adata, vdjx, clone_key=&quot;changeo_clone_id&quot;)
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython2 notranslate"><div class="highlight"><pre><span></span>motif = compute_motif(
    adata[
        (adata.obs[&quot;IR_VDJ_1_v_call&quot;].isin([&quot;IGHV3-48&quot;, &quot;IGHV1-18&quot;]))
        &amp; (adata.obs[&quot;junction_aa_VDJ&quot;].str.len() == 15),
        :,
    ]
    .obs[&quot;junction_aa_VDJ&quot;]
    .to_list()
)
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython2 notranslate"><div class="highlight"><pre><span></span>_ = svg_logo(
    motif, &quot;../_static/images/air_repertoire/bcr_logo_motif.svg&quot;, color_scheme=&quot;taylor&quot;
)
</pre></div>
</div>
</div>
</div>
<p>We observe a clear dominance of certain amino acids at most positions, except for a few where additional amino acids contribute to the sequence motif landscape. Nonetheless, the sequence composition for the V segments of interest with a length of 15 remains relatively stable.</p>
<p><img alt="" src="../_images/bcr_logo_motif.svg" /></p>
<p>On the other hand, we analyzed the same V gene segments for the V(D)J chain but with a sequence length of 23 aminoacids.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython2 notranslate"><div class="highlight"><pre><span></span>motif = compute_motif(
    adata[
        (adata.obs[&quot;IR_VDJ_1_v_call&quot;].isin([&quot;IGHV3-48&quot;, &quot;IGHV1-18&quot;]))
        &amp; (adata.obs[&quot;junction_aa_VDJ&quot;].str.len() == 23),
        :,
    ]
    .obs[&quot;junction_aa_VDJ&quot;]
    .to_list()
)
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython2 notranslate"><div class="highlight"><pre><span></span>_ = svg_logo(
    motif, &quot;../_static/images/air_repertoire/bcr2_logo_motif.svg&quot;, color_scheme=&quot;taylor&quot;
)
</pre></div>
</div>
</div>
</div>
<p><img alt="" src="../_images/bcr2_logo_motif.svg" /></p>
<section id="references">
<h2><span class="section-number">42.1. </span>References<a class="headerlink" href="#references" title="Link to this heading">#</a></h2>
<div class="docutils container" id="id24">
<div role="list" class="citation-list">
<div class="citation" id="id512" role="doc-biblioentry">
<span class="label"><span class="fn-bracket">[</span><a role="doc-backlink" href="#id4">AMB17</a><span class="fn-bracket">]</span></span>
<p>Peter Ashcroft, Markus G Manz, and Sebastian Bonhoeffer. Clonal dominance and transplantation dynamics in hematopoietic stem cell compartments. <em>PLoS computational biology</em>, 13(10):e1005803, 2017.</p>
</div>
<div class="citation" id="id517" role="doc-biblioentry">
<span class="label"><span class="fn-bracket">[</span><a role="doc-backlink" href="#id8">CKCH21</a><span class="fn-bracket">]</span></span>
<p>Mark Chernyshev, Mateusz Kaduk, Martin Corcoran, and Gunilla B Karlsson Hedestam. Vdj gene usage in igm repertoires of rhesus and cynomolgus macaques. <em>Frontiers in immunology</em>, 2021.</p>
</div>
<div class="citation" id="id518" role="doc-biblioentry">
<span class="label"><span class="fn-bracket">[</span><a role="doc-backlink" href="#id9">CDMK13</a><span class="fn-bracket">]</span></span>
<p>Stanca M Ciupe, Blythe H Devlin, Mary Louise Markert, and Thomas B Kepler. Quantification of total t-cell receptor diversity by flow cytometry and spectratyping. <em>BMC immunology</em>, 14(1):1–12, 2013.</p>
</div>
<div class="citation" id="id533" role="doc-biblioentry">
<span class="label"><span class="fn-bracket">[</span>CDNVH+16<span class="fn-bracket">]</span></span>
<span class="backrefs">(<a role="doc-backlink" href="#id15">1</a>,<a role="doc-backlink" href="#id21">2</a>)</span>
<p>Ang Cui, Roberto Di Niro, Jason A Vander Heiden, Adrian W Briggs, Kris Adams, Tamara Gilbert, Kevin C O’Connor, Francois Vigneault, Mark J Shlomchik, and Steven H Kleinstein. A model of somatic hypermutation targeting in mice based on high-throughput ig sequencing data. <em>The Journal of Immunology</em>, 197(9):3566–3574, 2016.</p>
</div>
<div class="citation" id="id516" role="doc-biblioentry">
<span class="label"><span class="fn-bracket">[</span><a role="doc-backlink" href="#id7">EMCJ+14</a><span class="fn-bracket">]</span></span>
<p>Yuval Elhanati, Anand Murugan, Curtis G Callan Jr, Thierry Mora, and Aleksandra M Walczak. Quantifying selection in immune receptor repertoires. <em>Proceedings of the National Academy of Sciences</em>, 111(27):9875–9880, 2014.</p>
</div>
<div class="citation" id="id530" role="doc-biblioentry">
<span class="label"><span class="fn-bracket">[</span>GVHU+15<span class="fn-bracket">]</span></span>
<span class="backrefs">(<a role="doc-backlink" href="#id10">1</a>,<a role="doc-backlink" href="#id16">2</a>,<a role="doc-backlink" href="#id22">3</a>)</span>
<p>Namita T Gupta, Jason A Vander Heiden, Mohamed Uduman, Daniel Gadala-Maria, Gur Yaari, and Steven H Kleinstein. Change-o: a toolkit for analyzing large-scale b cell immunoglobulin repertoire sequencing data. <em>Bioinformatics</em>, 31(20):3356–3358, 2015.</p>
</div>
<div class="citation" id="id511" role="doc-biblioentry">
<span class="label"><span class="fn-bracket">[</span><a role="doc-backlink" href="#id5">KS19</a><span class="fn-bracket">]</span></span>
<p>Tae-Shin Kim and Eui-Cheol Shin. The activation of bystander cd8+ t cells and their roles in viral infection. <em>Experimental &amp; molecular medicine</em>, 51(12):1–9, 2019.</p>
</div>
<div class="citation" id="id513" role="doc-biblioentry">
<span class="label"><span class="fn-bracket">[</span><a role="doc-backlink" href="#id3">Nax20</a><span class="fn-bracket">]</span></span>
<p>Kamila Naxerova. Clonal competition in a confined space. <em>Nature Genetics</em>, 52(6):553–554, 2020.</p>
</div>
<div class="citation" id="id529" role="doc-biblioentry">
<span class="label"><span class="fn-bracket">[</span><a role="doc-backlink" href="#id11">PS02</a><span class="fn-bracket">]</span></span>
<p>F Nina Papavasiliou and David G Schatz. Somatic hypermutation of immunoglobulin genes: merging mechanisms for genetic diversity. <em>Cell</em>, 109(2):S35–S44, 2002.</p>
</div>
<div class="citation" id="id514" role="doc-biblioentry">
<span class="label"><span class="fn-bracket">[</span>PCF16<span class="fn-bracket">]</span></span>
<span class="backrefs">(<a role="doc-backlink" href="#id1">1</a>,<a role="doc-backlink" href="#id2">2</a>)</span>
<p>Michal Polonsky, Benjamin Chain, and Nir Friedman. Clonal expansion under the microscope: studying lymphocyte activation and differentiation using live-cell imaging. <em>Immunology and Cell Biology</em>, 94(3):242–249, 2016.</p>
</div>
<div class="citation" id="id484" role="doc-biblioentry">
<span class="label"><span class="fn-bracket">[</span><a role="doc-backlink" href="#id12">SRB+21</a><span class="fn-bracket">]</span></span>
<p>Emily Stephenson, Gary Reynolds, Rachel A Botting, Fernando J Calero-Nieto, Michael D Morgan, Zewen Kelvin Tuong, Karsten Bach, Waradon Sungnak, Kaylee B Worlock, Masahiro Yoshida, and others. Single-cell multi-omics analysis of the immune response in covid-19. <em>Nature medicine</em>, 27(5):904–916, 2021.</p>
</div>
<div class="citation" id="id515" role="doc-biblioentry">
<span class="label"><span class="fn-bracket">[</span><a role="doc-backlink" href="#id6">TCRB18</a><span class="fn-bracket">]</span></span>
<p>Ilias S Travlos, Nikolina Cheimona, Ioannis Roussis, and Dimitrios J Bilalis. Weed-species abundance and diversity indices in relation to tillage systems and fertilization. <em>Frontiers in Environmental Science</em>, 6:11, 2018.</p>
</div>
<div class="citation" id="id531" role="doc-biblioentry">
<span class="label"><span class="fn-bracket">[</span><a role="doc-backlink" href="#id17">VHYU+14</a><span class="fn-bracket">]</span></span>
<p>Jason A Vander Heiden, Gur Yaari, Mohamed Uduman, Joel NH Stern, Kevin C O’Connor, David A Hafler, Francois Vigneault, and Steven H Kleinstein. Presto: a toolkit for processing high-throughput sequencing raw reads of lymphocyte receptor repertoires. <em>Bioinformatics</em>, 30(13):1930–1932, 2014.</p>
</div>
<div class="citation" id="id532" role="doc-biblioentry">
<span class="label"><span class="fn-bracket">[</span>YVHU+13<span class="fn-bracket">]</span></span>
<span class="backrefs">(<a role="doc-backlink" href="#id14">1</a>,<a role="doc-backlink" href="#id18">2</a>,<a role="doc-backlink" href="#id19">3</a>,<a role="doc-backlink" href="#id20">4</a>)</span>
<p>Gur Yaari, Jason A Vander Heiden, Mohamed Uduman, Daniel Gadala-Maria, Namita Gupta, Joel NH Stern, Kevin C O’Connor, David A Hafler, Uri Laserson, Francois Vigneault, and others. Models of somatic hypermutation targeting and substitution based on synonymous mutations from high-throughput immunoglobulin sequencing data. <em>Frontiers in immunology</em>, 4:358, 2013.</p>
</div>
</div>
</div>
</section>
</section>

    <script type="text/x-thebe-config">
    {
        requestKernel: true,
        binderOptions: {
            repo: "binder-examples/jupyter-stacks-datascience",
            ref: "master",
        },
        codeMirrorConfig: {
            theme: "abcdef",
            mode: "python"
        },
        kernelOptions: {
            name: "python3",
            path: "./air_repertoire"
        },
        predefinedOutput: true
    }
    </script>
    <script>kernelName = 'python3'</script>

                </article>
              

              
              
              
              
                <footer class="prev-next-footer d-print-none">
                  
<div class="prev-next-area">
    <a class="left-prev"
       href="ir_profiling.html"
       title="previous page">
      <i class="fa-solid fa-angle-left"></i>
      <div class="prev-next-info">
        <p class="prev-next-subtitle">previous</p>
        <p class="prev-next-title"><span class="section-number">38. </span>Immune Receptor Profiling</p>
      </div>
    </a>
    <a class="right-next"
       href="specificity.html"
       title="next page">
      <div class="prev-next-info">
        <p class="prev-next-subtitle">next</p>
        <p class="prev-next-title"><span class="section-number">43. </span>Specificity analysis</p>
      </div>
      <i class="fa-solid fa-angle-right"></i>
    </a>
</div>
                </footer>
              
            </div>
            
            
              
                <div class="bd-sidebar-secondary bd-toc"><div class="sidebar-secondary-items sidebar-secondary__inner">


  <div class="sidebar-secondary-item">
  <div class="page-toc tocsection onthispage">
    <i class="fa-solid fa-list"></i> Contents
  </div>
  <nav class="bd-toc-nav page-toc">
    <ul class="visible nav section-nav flex-column">
<li class="toc-h1 nav-item toc-entry"><a class="reference internal nav-link" href="#">39. Clonotype analysis</a><ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#clonal-expansion-diversity-and-abundance">39.1. Clonal expansion: diversity and abundance</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#gene-segment-usage-and-spectratype">39.2. Gene segment usage and spectratype</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#tcr-data-preparation">39.3. TCR data preparation</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#clonotype-definition">39.4. Clonotype definition</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#clonal-expansion">39.5. Clonal expansion</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#clonotype-abundance">39.6. Clonotype abundance</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#gene-usage">39.7. Gene usage</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#spectratype-analysis">39.8. Spectratype analysis</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#motif-sequence-analysis">39.9. Motif sequence analysis</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#repertoire-comparison">39.10. Repertoire comparison</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#bcr-data-analysis-with-dandelion">39.11. BCR Data Analysis with Dandelion</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#dandelion-interoperability">39.12. Dandelion interoperability</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#id13">39.13. Clonotype definition</a></li>
</ul>
</li>
<li class="toc-h1 nav-item toc-entry"><a class="reference internal nav-link" href="#gene-segment-usage">40. Gene segment usage</a></li>
<li class="toc-h1 nav-item toc-entry"><a class="reference internal nav-link" href="#spectratype">41. Spectratype</a></li>
<li class="toc-h1 nav-item toc-entry"><a class="reference internal nav-link" href="#id23">42. Motif sequence analysis</a><ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#references">42.1. References</a></li>
</ul>
</li>
</ul>

  </nav></div>

</div></div>
              
            
          </div>
          <footer class="bd-footer-content">
            
<div class="bd-footer-content__inner container">
  
  <div class="footer-item">
    
<p class="component-author">
By Lukas Heumos, Anna Schaar, single-cell best practices consortium
</p>

  </div>
  
  <div class="footer-item">
    

  <p class="copyright">
    
      © Copyright 2023.
      <br/>
    
  </p>

  </div>
  
  <div class="footer-item">
    
  </div>
  
  <div class="footer-item">
    
<div class="extra_footer">
  <div>
  Brought to you by
  <a href="https://www.github.com/theislab">Theislab</a>, with many thanks to the single-cell community as a whole!
</div>

</div>
  </div>
  
</div>
          </footer>
        

      </main>
    </div>
  </div>
  
  <!-- Scripts loaded after <body> so the DOM is not blocked -->
  <script src="../_static/scripts/bootstrap.js?digest=dfe6caa3a7d634c4db9b"></script>
<script src="../_static/scripts/pydata-sphinx-theme.js?digest=dfe6caa3a7d634c4db9b"></script>

  <footer class="bd-footer">
  </footer>
  </body>
</html>