
<!DOCTYPE html>


<html lang="en" data-content_root="../" >

  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="viewport" content="width=device-width, initial-scale=1" />

    <title>Conversion of muon objects to Seurat objects to use in downstream analysis &#8212; Single-cell best practices</title>
  
  
  
  <script data-cfasync="false">
    document.documentElement.dataset.mode = localStorage.getItem("mode") || "";
    document.documentElement.dataset.theme = localStorage.getItem("theme") || "";
  </script>
  <!--
    this give us a css class that will be invisible only if js is disabled
  -->
  <noscript>
    <style>
      .pst-js-only { display: none !important; }

    </style>
  </noscript>
  
  <!-- Loaded before other Sphinx assets -->
  <link href="../_static/styles/theme.css?digest=8878045cc6db502f8baf" rel="stylesheet" />
<link href="../_static/styles/pydata-sphinx-theme.css?digest=8878045cc6db502f8baf" rel="stylesheet" />

    <link rel="stylesheet" type="text/css" href="../_static/pygments.css?v=03e43079" />
    <link rel="stylesheet" type="text/css" href="../_static/styles/sphinx-book-theme.css?v=a3416100" />
    <link rel="stylesheet" type="text/css" href="../_static/togglebutton.css?v=13237357" />
    <link rel="stylesheet" type="text/css" href="../_static/copybutton.css?v=76b2166b" />
    <link rel="stylesheet" type="text/css" href="../_static/mystnb.4510f1fc1dee50b3e5859aac5469c37c29e427902b24a333a5f9fcb2f0b3ac41.css" />
    <link rel="stylesheet" type="text/css" href="../_static/sphinx-thebe.css?v=4fa983c6" />
    <link rel="stylesheet" type="text/css" href="../_static/sphinx-design.min.css?v=95c83b7e" />
    <link rel="stylesheet" type="text/css" href="../_static/book.css?v=148c11e2" />
  
  <!-- So that users can add custom icons -->
  <script src="../_static/scripts/fontawesome.js?digest=8878045cc6db502f8baf"></script>
  <!-- Pre-loaded scripts that we'll load fully later -->
  <link rel="preload" as="script" href="../_static/scripts/bootstrap.js?digest=8878045cc6db502f8baf" />
<link rel="preload" as="script" href="../_static/scripts/pydata-sphinx-theme.js?digest=8878045cc6db502f8baf" />

    <script src="../_static/documentation_options.js?v=9eb32ce0"></script>
    <script src="../_static/doctools.js?v=9a2dae69"></script>
    <script src="../_static/sphinx_highlight.js?v=dc90522c"></script>
    <script src="../_static/clipboard.min.js?v=a7894cd8"></script>
    <script src="../_static/copybutton.js?v=f281be69"></script>
    <script src="../_static/scripts/sphinx-book-theme.js?v=887ef09a"></script>
    <script>let toggleHintShow = 'Click to show';</script>
    <script>let toggleHintHide = 'Click to hide';</script>
    <script>let toggleOpenOnPrint = 'true';</script>
    <script src="../_static/togglebutton.js?v=4a39c7ea"></script>
    <script>var togglebuttonSelector = '.toggle, .admonition.dropdown';</script>
    <script src="../_static/design-tabs.js?v=f930bc37"></script>
    <script>const THEBE_JS_URL = "https://unpkg.com/thebe@0.8.2/lib/index.js"; const thebe_selector = ".thebe,.cell"; const thebe_selector_input = "pre"; const thebe_selector_output = ".output, .cell_output"</script>
    <script async="async" src="../_static/sphinx-thebe.js?v=c100c467"></script>
    <script>var togglebuttonSelector = '.toggle, .admonition.dropdown';</script>
    <script>const THEBE_JS_URL = "https://unpkg.com/thebe@0.8.2/lib/index.js"; const thebe_selector = ".thebe,.cell"; const thebe_selector_input = "pre"; const thebe_selector_output = ".output, .cell_output"</script>
    <script>window.MathJax = {"options": {"processHtmlClass": "tex2jax_process|mathjax_process|math|output_area"}}</script>
    <script defer="defer" src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-mml-chtml.js"></script>
    <script>DOCUMENTATION_OPTIONS.pagename = 'chromatin_accessibility/muon_to_seurat';</script>
    <script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
    <script src="../_static/book.js?v=dcd3d8bf"></script>
    <link rel="icon" href="../_static/favicon.ico"/>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
  <meta name="viewport" content="width=device-width, initial-scale=1"/>
  <meta name="docsearch:language" content="en"/>
  <meta name="docsearch:version" content="" />
  </head>
  
  
  <body data-bs-spy="scroll" data-bs-target=".bd-toc-nav" data-offset="180" data-bs-root-margin="0px 0px -60%" data-default-mode="">

  
  
  <div id="pst-skip-link" class="skip-link d-print-none"><a href="#main-content">Skip to main content</a></div>
  
  <div id="pst-scroll-pixel-helper"></div>
  
  <button type="button" class="btn rounded-pill" id="pst-back-to-top">
    <i class="fa-solid fa-arrow-up"></i>Back to top</button>

  
  <dialog id="pst-search-dialog">
    
<form class="bd-search d-flex align-items-center"
      action="../search.html"
      method="get">
  <i class="fa-solid fa-magnifying-glass"></i>
  <input type="search"
         class="form-control"
         name="q"
         placeholder="Search this book..."
         aria-label="Search this book..."
         autocomplete="off"
         autocorrect="off"
         autocapitalize="off"
         spellcheck="false"/>
  <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd>K</kbd></span>
</form>
  </dialog>

  <div class="pst-async-banner-revealer d-none">
  <aside id="bd-header-version-warning" class="d-none d-print-none" aria-label="Version warning"></aside>
</div>

  
    <header class="bd-header navbar navbar-expand-lg bd-navbar d-print-none">
    </header>
  

  <div class="bd-container">
    <div class="bd-container__inner bd-page-width">
      
      
      
        
      
      <dialog id="pst-primary-sidebar-modal"></dialog>
      <div id="pst-primary-sidebar" class="bd-sidebar-primary bd-sidebar">
        

  
  <div class="sidebar-header-items sidebar-primary__section">
    
    
    
    
  </div>
  
    <div class="sidebar-primary-items__start sidebar-primary__section">
        <div class="sidebar-primary-item">

  
    
  

<a class="navbar-brand logo" href="../preamble.html">
  
  
  
  
  
    
    
      
    
    
    <img src="../_static/logo.png" class="logo__image only-light" alt="Single-cell best practices - Home"/>
    <img src="../_static/logo.png" class="logo__image only-dark pst-js-only" alt="Single-cell best practices - Home"/>
  
  
</a></div>
        <div class="sidebar-primary-item">

<button class="btn search-button-field search-button__button pst-js-only" title="Search" aria-label="Search" data-bs-placement="bottom" data-bs-toggle="tooltip">
 <i class="fa-solid fa-magnifying-glass"></i>
 <span class="search-button__default-text">Search</span>
 <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd class="kbd-shortcut__modifier">K</kbd></span>
</button></div>
        <div class="sidebar-primary-item"><nav class="bd-links bd-docs-nav" aria-label="Main">
    <div class="bd-toc-item navbar-nav active">
        <p aria-level="2" class="caption" role="heading"><span class="caption-text">Introduction</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../introduction/prior_art.html">1. Prior art</a></li>
<li class="toctree-l1"><a class="reference internal" href="../introduction/scrna_seq.html">2. Single-cell RNA sequencing</a></li>
<li class="toctree-l1"><a class="reference internal" href="../introduction/raw_data_processing.html">3. Raw data processing</a></li>
<li class="toctree-l1"><a class="reference internal" href="../introduction/analysis_tools.html">4. Analysis frameworks and tools</a></li>
<li class="toctree-l1"><a class="reference internal" href="../introduction/interoperability.html">5. Interoperability</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Preprocessing and visualization</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../preprocessing_visualization/quality_control.html">6. Quality Control</a></li>
<li class="toctree-l1"><a class="reference internal" href="../preprocessing_visualization/normalization.html">7. Normalization</a></li>
<li class="toctree-l1"><a class="reference internal" href="../preprocessing_visualization/feature_selection.html">8. Feature selection</a></li>
<li class="toctree-l1"><a class="reference internal" href="../preprocessing_visualization/dimensionality_reduction.html">9. Dimensionality Reduction</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Identifying cellular structure</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../cellular_structure/clustering.html">10. Clustering</a></li>
<li class="toctree-l1"><a class="reference internal" href="../cellular_structure/annotation.html">11. Annotation</a></li>
<li class="toctree-l1"><a class="reference internal" href="../cellular_structure/integration.html">12. Data integration</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Inferring trajectories</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../trajectories/pseudotemporal.html">13. Pseudotemporal ordering</a></li>
<li class="toctree-l1"><a class="reference internal" href="../trajectories/rna_velocity.html">14. RNA velocity</a></li>
<li class="toctree-l1"><a class="reference internal" href="../trajectories/lineage_tracing.html">15. Lineage tracing</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Dealing with conditions</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../conditions/differential_gene_expression.html">16. Differential gene expression analysis</a></li>
<li class="toctree-l1"><a class="reference internal" href="../conditions/compositional.html">17. Compositional analysis</a></li>
<li class="toctree-l1"><a class="reference internal" href="../conditions/gsea_pathway.html">18. Gene set enrichment and pathway analysis</a></li>
<li class="toctree-l1"><a class="reference internal" href="../conditions/perturbation_modeling.html">19. Perturbation modeling</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Modeling mechanisms</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../mechanisms/gene_regulatory_networks.html">20. Gene regulatory networks</a></li>
<li class="toctree-l1"><a class="reference internal" href="../mechanisms/cell_cell_communication.html">21. Cell-cell communication</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Deconvolution</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../deconvolution/bulk_deconvolution.html">22. Bulk deconvolution</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Chromatin Accessibility</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="introduction.html">23. Single-cell ATAC sequencing</a></li>
<li class="toctree-l1"><a class="reference internal" href="quality_control.html">24. Quality Control</a></li>
<li class="toctree-l1"><a class="reference internal" href="gene_regulatory_networks_atac.html">25. Gene regulatory networks using chromatin accessibility</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Spatial omics</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../spatial/introduction.html">26. Single-cell data resolved in space</a></li>
<li class="toctree-l1"><a class="reference internal" href="../spatial/neighborhood.html">27. Neighborhood analysis</a></li>
<li class="toctree-l1"><a class="reference internal" href="../spatial/domains.html">28. Spatial domains</a></li>
<li class="toctree-l1"><a class="reference internal" href="../spatial/spatially_variable_genes.html">29. Spatially variable genes</a></li>
<li class="toctree-l1"><a class="reference internal" href="../spatial/deconvolution.html">30. Spatial deconvolution</a></li>
<li class="toctree-l1"><a class="reference internal" href="../spatial/imputation.html">31. Imputation</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Surface protein</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../surface_protein/quality_control.html">32. Quality control</a></li>
<li class="toctree-l1"><a class="reference internal" href="../surface_protein/normalization.html">33. Normalization</a></li>
<li class="toctree-l1"><a class="reference internal" href="../surface_protein/doublet_detection.html">34. Doublet detection</a></li>
<li class="toctree-l1"><a class="reference internal" href="../surface_protein/dimensionality_reduction.html">35. Dimensionality Reduction</a></li>
<li class="toctree-l1"><a class="reference internal" href="../surface_protein/batch_correction.html">36. Batch correction</a></li>
<li class="toctree-l1"><a class="reference internal" href="../surface_protein/annotation.html">37. Annotation</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Adaptive immune receptor repertoire</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../air_repertoire/ir_profiling.html">38. Immune Receptor Profiling</a></li>
<li class="toctree-l1"><a class="reference internal" href="../air_repertoire/clonotype.html">39. Clonotype analysis</a></li>



<li class="toctree-l1"><a class="reference internal" href="../air_repertoire/specificity.html">43. Specificity analysis</a></li>
<li class="toctree-l1"><a class="reference internal" href="../air_repertoire/multimodal_integration.html">44. Integrating AIR and transcriptomics</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Multimodal integration</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../multimodal_integration/paired_integration.html">45. Paired integration</a></li>
<li class="toctree-l1"><a class="reference internal" href="../multimodal_integration/advanced_integration.html">46. Advanced integration</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Reproducibility</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../reproducibility/introduction.html">47. Reproducibility</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Outlook</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../outlook.html">48. Outlook</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Acknowledgements</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../acknowledgements.html">49. Acknowledgements</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Glossary</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../glossary.html">50. Glossary</a></li>
</ul>

    </div>
</nav></div>
    </div>
  
  
  <div class="sidebar-primary-items__end sidebar-primary__section">
      <div class="sidebar-primary-item">
<div id="ethical-ad-placement"
      class="flat"
      data-ea-publisher="readthedocs"
      data-ea-type="readthedocs-sidebar"
      data-ea-manual="true">
</div></div>
  </div>


      </div>
      
      <main id="main-content" class="bd-main" role="main">
        
        

<div class="sbt-scroll-pixel-helper"></div>

          <div class="bd-content">
            <div class="bd-article-container">
              
              <div class="bd-header-article d-print-none">
<div class="header-article-items header-article__inner">
  
    <div class="header-article-items__start">
      
        <div class="header-article-item"><button class="sidebar-toggle primary-toggle btn btn-sm" title="Toggle primary sidebar" data-bs-placement="bottom" data-bs-toggle="tooltip">
  <span class="fa-solid fa-bars"></span>
</button></div>
      
    </div>
  
  
    <div class="header-article-items__end">
      
        <div class="header-article-item">

<div class="article-header-buttons">





<div class="dropdown dropdown-source-buttons">
  <button class="btn dropdown-toggle" type="button" data-bs-toggle="dropdown" aria-expanded="false" aria-label="Source repositories">
    <i class="fab fa-github"></i>
  </button>
  <ul class="dropdown-menu">
      
      
      
      <li><a href="https://github.com/theislab/single-cell-best-practices" target="_blank"
   class="btn btn-sm btn-source-repository-button dropdown-item"
   title="Source repository"
   data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fab fa-github"></i>
  </span>
<span class="btn__text-container">Repository</span>
</a>
</li>
      
      
      
      
      <li><a href="https://github.com/theislab/single-cell-best-practices/edit/master/jupyter-book/chromatin_accessibility/muon_to_seurat.ipynb" target="_blank"
   class="btn btn-sm btn-source-edit-button dropdown-item"
   title="Suggest edit"
   data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-pencil-alt"></i>
  </span>
<span class="btn__text-container">Suggest edit</span>
</a>
</li>
      
      
      
      
      <li><a href="https://github.com/theislab/single-cell-best-practices/issues/new?title=Issue%20on%20page%20%2Fchromatin_accessibility/muon_to_seurat.html&body=Your%20issue%20content%20here." target="_blank"
   class="btn btn-sm btn-source-issues-button dropdown-item"
   title="Open an issue"
   data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-lightbulb"></i>
  </span>
<span class="btn__text-container">Open issue</span>
</a>
</li>
      
  </ul>
</div>






<div class="dropdown dropdown-download-buttons">
  <button class="btn dropdown-toggle" type="button" data-bs-toggle="dropdown" aria-expanded="false" aria-label="Download this page">
    <i class="fas fa-download"></i>
  </button>
  <ul class="dropdown-menu">
      
      
      
      <li><a href="../_sources/chromatin_accessibility/muon_to_seurat.ipynb" target="_blank"
   class="btn btn-sm btn-download-source-button dropdown-item"
   title="Download source file"
   data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-file"></i>
  </span>
<span class="btn__text-container">.ipynb</span>
</a>
</li>
      
      
      
      
      <li>
<button onclick="window.print()"
  class="btn btn-sm btn-download-pdf-button dropdown-item"
  title="Print to PDF"
  data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-file-pdf"></i>
  </span>
<span class="btn__text-container">.pdf</span>
</button>
</li>
      
  </ul>
</div>




<button onclick="toggleFullScreen()"
  class="btn btn-sm btn-fullscreen-button"
  title="Fullscreen mode"
  data-bs-placement="bottom" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-expand"></i>
  </span>

</button>



<button class="btn btn-sm nav-link pst-navbar-icon theme-switch-button pst-js-only" aria-label="Color mode" data-bs-title="Color mode"  data-bs-placement="bottom" data-bs-toggle="tooltip">
  <i class="theme-switch fa-solid fa-sun                fa-lg" data-mode="light" title="Light"></i>
  <i class="theme-switch fa-solid fa-moon               fa-lg" data-mode="dark"  title="Dark"></i>
  <i class="theme-switch fa-solid fa-circle-half-stroke fa-lg" data-mode="auto"  title="System Settings"></i>
</button>


<button class="btn btn-sm pst-navbar-icon search-button search-button__button pst-js-only" title="Search" aria-label="Search" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="fa-solid fa-magnifying-glass fa-lg"></i>
</button>
<button class="sidebar-toggle secondary-toggle btn btn-sm" title="Toggle secondary sidebar" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <span class="fa-solid fa-list"></span>
</button>
</div></div>
      
    </div>
  
</div>
</div>
              
              

<div id="jb-print-docs-body" class="onlyprint">
    <h1>Conversion of muon objects to Seurat objects to use in downstream analysis</h1>
    <!-- Table of contents -->
    <div id="print-main-content">
        <div id="jb-print-toc">
            
            <div>
                <h2> Contents </h2>
            </div>
            <nav aria-label="Page">
                <ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#environment-setup">Environment setup</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#load-data">Load data</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#read-in-mudata-object">Read in mudata object</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#convert-seurat-assay-to-chromatin-assay">Convert Seurat assay to chromatin assay</a></li>
</ul>
</li>
</ul>
            </nav>
        </div>
    </div>
</div>

              
                
<div id="searchbox"></div>
                <article class="bd-article">
                  
  <section class="tex2jax_ignore mathjax_ignore" id="conversion-of-muon-objects-to-seurat-objects-to-use-in-downstream-analysis">
<h1>Conversion of muon objects to Seurat objects to use in downstream analysis<a class="headerlink" href="#conversion-of-muon-objects-to-seurat-objects-to-use-in-downstream-analysis" title="Link to this heading">#</a></h1>
<p>As we saw in the introduction, there are some downstream analyses that are currently only available in R. In this notebook, we therefore want to showcase how you can convert a muon object into a Seurat object for downstream analysis using Signac.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="nf">.libPaths</span><span class="p">(</span><span class="nf">c</span><span class="p">(</span><span class="s">&quot;/home/icb/laura.martens/miniconda3/envs/best_practices/lib/R/library&quot;</span><span class="p">,</span><span class="w"> </span><span class="s">&quot;/home/icb/laura.martens/miniconda3/envs/signac/lib/R/library&quot;</span><span class="p">))</span>
</pre></div>
</div>
</div>
</div>
<section id="environment-setup">
<h2>Environment setup<a class="headerlink" href="#environment-setup" title="Link to this heading">#</a></h2>
<p>We load the R libraries <code class="docutils literal notranslate"><span class="pre">Signac</span></code> and <code class="docutils literal notranslate"><span class="pre">Seurat</span></code> to handle scATAC-seq data in R. Furthermore, we need to load the annotation for our reference genome. In this example, we load the annotation for human and the <code class="docutils literal notranslate"><span class="pre">hg38</span></code> reference genome.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="nf">library</span><span class="p">(</span><span class="n">MuDataSeurat</span><span class="p">)</span><span class="c1"># to read in muon object and convert to Seurat</span>
<span class="nf">library</span><span class="p">(</span><span class="n">Signac</span><span class="p">)</span><span class="w"> </span>
<span class="nf">library</span><span class="p">(</span><span class="n">Seurat</span><span class="p">)</span>
<span class="nf">library</span><span class="p">(</span><span class="n">stringr</span><span class="p">)</span>

<span class="nf">library</span><span class="p">(</span><span class="n">EnsDb.Hsapiens.v86</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Warning message:
&quot;package &#39;stringr&#39; was built under R version 4.2.2&quot;
</pre></div>
</div>
</div>
</div>
</section>
<section id="load-data">
<h2>Load data<a class="headerlink" href="#load-data" title="Link to this heading">#</a></h2>
<p>With Signac, we have the possibility to visualize the coverage of the features or perform other analysis like footprinting. For some of these downstream analysis, it is necessary to add the fragment file to the Seurat object which we specify here with <code class="docutils literal notranslate"><span class="pre">fragmen_file</span></code></p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="n">fragment_file</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="s">&quot;/lustre/groups/ml01/datasets/projects/20220323_neurips21_bmmc_christopher.lance/multiome/aggr_donors/atac_fragments.tsv.gz&quot;</span>
<span class="n">adata_file</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="s">&quot;/lustre/groups/ml01/code/laura.martens/scATAC_pipelines/src/best_practices/multiome_donor_s4d8_pp_test.h5mu&quot;</span><span class="w"> </span><span class="c1">#&quot;/lustre/groups/ml01/workspace/christopher.lance/mm_best/src/scATAC/muon_data/neurips_qc_filtered_allsamp.h5ad&quot;</span>
</pre></div>
</div>
</div>
</div>
<section id="read-in-mudata-object">
<h3>Read in mudata object<a class="headerlink" href="#read-in-mudata-object" title="Link to this heading">#</a></h3>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="n">atac</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">MuDataSeurat</span><span class="o">::</span><span class="nf">ReadH5MU</span><span class="p">(</span><span class="n">adata_file</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Warning message in missing_on_read(&quot;/var&quot;, paste0(&quot;global variables metadata (&quot;, :
&quot;Missing on read: /var. Seurat does not support global variables metadata (gene_ids, feature_types, genome, interval).&quot;
Warning message in read_layers_to_assay(h5[[&quot;mod&quot;]][[mod]], mod):
&quot;The var_names from modality gene_activity have been renamed as feature names cannot contain &#39;_&#39;. E.g. gene_activity:MIR1302-2HG -&gt; gene-activity:MIR1302-2HG.&quot;
as(&lt;dsCMatrix&gt;, &quot;dgCMatrix&quot;) is deprecated since Matrix 1.5-0; do as(., &quot;generalMatrix&quot;) instead

Warning message in missing_on_read(paste0(&quot;/mod&quot;, mod, &quot;/varp&quot;), &quot;pairwise annotation of variables&quot;):
&quot;Missing on read: /modrna/varp. Seurat does not support pairwise annotation of variables.&quot;
Warning message in missing_on_read(paste0(&quot;/mod&quot;, mod, &quot;/varp&quot;), &quot;pairwise annotation of variables&quot;):
&quot;Missing on read: /modatac/varp. Seurat does not support pairwise annotation of variables.&quot;
Warning message in missing_on_read(paste0(&quot;/mod&quot;, mod, &quot;/varp&quot;), &quot;pairwise annotation of variables&quot;):
&quot;Missing on read: /modgene_activity/varp. Seurat does not support pairwise annotation of variables.&quot;
Warning message:
&quot;Keys should be one or more alphanumeric characters followed by an underscore, setting key from rna to rna_&quot;
Warning message:
&quot;Keys should be one or more alphanumeric characters followed by an underscore, setting key from gene_activity_ to geneactivity_&quot;
Warning message:
&quot;No columnames present in cell embeddings, setting to &#39;LSI_1:49&#39;&quot;
Warning message:
&quot;No columnames present in cell embeddings, setting to &#39;UMAP_1:2&#39;&quot;
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="n">atac</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>An object of class Seurat 
220291 features across 14710 samples within 3 assays 
Active assay: rna (36601 features, 0 variable features)
 2 other assays present: atac, gene_activity
 2 dimensional reductions calculated: LSI, UMAP
</pre></div>
</div>
</div>
</div>
<p>We have successfully loaded the muon object into R and it has been converted to a Seurat object with three assays <code class="docutils literal notranslate"><span class="pre">atac</span></code>, <code class="docutils literal notranslate"><span class="pre">rna</span></code> and <code class="docutils literal notranslate"><span class="pre">gene_activity</span></code>. The dimensionality reductions and metadata have also been transferred.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="nf">head</span><span class="p">(</span><span class="n">atac</span><span class="p">[[]])</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><table class="dataframe">
<caption>A data.frame: 6 × 27</caption>
<thead>
	<tr><th></th><th scope=col>orig.ident</th><th scope=col>nCount_rna</th><th scope=col>nFeature_rna</th><th scope=col>nCount_atac</th><th scope=col>nFeature_atac</th><th scope=col>nCount_gene_activity</th><th scope=col>nFeature_gene_activity</th><th scope=col>n_features_per_cell</th><th scope=col>total_fragment_counts</th><th scope=col>log_total_fragment_counts</th><th scope=col>⋯</th><th scope=col>n_features_per_cell.1</th><th scope=col>total_fragment_counts.1</th><th scope=col>log_total_fragment_counts.1</th><th scope=col>nucleosome_signal.1</th><th scope=col>nuc_signal_filter.1</th><th scope=col>tss_score.1</th><th scope=col>site.1</th><th scope=col>donor.1</th><th scope=col>sample.1</th><th scope=col>leiden.1</th></tr>
	<tr><th></th><th scope=col>&lt;fct&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>⋯</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;fct&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;fct&gt;</th><th scope=col>&lt;fct&gt;</th><th scope=col>&lt;fct&gt;</th><th scope=col>&lt;fct&gt;</th></tr>
</thead>
<tbody>
	<tr><th scope=row>AAACAGCCAAGCTTAT-14</th><td>SeuratProject</td><td>1707</td><td>1128</td><td>3588</td><td>1797</td><td>2738</td><td>1409</td><td>1799</td><td>3591</td><td>3.555215</td><td>⋯</td><td>1799</td><td>3591</td><td>3.555215</td><td>0.6643357</td><td>NS_PASS</td><td> 3.521329</td><td>site4</td><td>donor8</td><td>s4d8</td><td>4 </td></tr>
	<tr><th scope=row>AAACAGCCATGAAATG-14</th><td>SeuratProject</td><td>1409</td><td> 479</td><td>5056</td><td>2552</td><td>3965</td><td>1957</td><td>2556</td><td>5062</td><td>3.704322</td><td>⋯</td><td>2556</td><td>5062</td><td>3.704322</td><td>0.6542056</td><td>NS_PASS</td><td> 2.296477</td><td>site4</td><td>donor8</td><td>s4d8</td><td>10</td></tr>
	<tr><th scope=row>AAACAGCCATGTTTGG-14</th><td>SeuratProject</td><td>1666</td><td>1077</td><td>7110</td><td>3360</td><td>5105</td><td>2495</td><td>3361</td><td>7112</td><td>3.851992</td><td>⋯</td><td>3361</td><td>7112</td><td>3.851992</td><td>0.6789298</td><td>NS_PASS</td><td>12.890983</td><td>site4</td><td>donor8</td><td>s4d8</td><td>3 </td></tr>
	<tr><th scope=row>AAACATGCAACGTGCT-14</th><td>SeuratProject</td><td> 882</td><td> 641</td><td>7569</td><td>3586</td><td>5397</td><td>2565</td><td>3591</td><td>7578</td><td>3.879555</td><td>⋯</td><td>3591</td><td>7578</td><td>3.879555</td><td>1.0609319</td><td>NS_PASS</td><td> 4.689176</td><td>site4</td><td>donor8</td><td>s4d8</td><td>6 </td></tr>
	<tr><th scope=row>AAACATGCAATATAGG-14</th><td>SeuratProject</td><td> 325</td><td> 277</td><td>8550</td><td>4005</td><td>6500</td><td>2932</td><td>4006</td><td>8551</td><td>3.932017</td><td>⋯</td><td>4006</td><td>8551</td><td>3.932017</td><td>0.9504644</td><td>NS_PASS</td><td> 4.600777</td><td>site4</td><td>donor8</td><td>s4d8</td><td>3 </td></tr>
	<tr><th scope=row>AAACATGCAATTAACC-14</th><td>SeuratProject</td><td>1704</td><td>1083</td><td>6993</td><td>3268</td><td>5061</td><td>2450</td><td>3269</td><td>6995</td><td>3.844788</td><td>⋯</td><td>3269</td><td>6995</td><td>3.844788</td><td>0.7665505</td><td>NS_PASS</td><td> 6.894277</td><td>site4</td><td>donor8</td><td>s4d8</td><td>0 </td></tr>
</tbody>
</table>
</div></div>
</div>
<p>We verify that the correct embedding and clustering has been added using <code class="docutils literal notranslate"><span class="pre">DimPlot</span></code>.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="nf">DimPlot</span><span class="p">(</span><span class="n">atac</span><span class="p">,</span><span class="w"> </span><span class="n">group.by</span><span class="o">=</span><span class="s">&#39;leiden&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<a class="reference internal image-reference" href="../_images/fd2d523a39ae24a5b77809de04dc78055d87a333af9fdde5a46ec5de4c705c1e.png"><img alt="../_images/fd2d523a39ae24a5b77809de04dc78055d87a333af9fdde5a46ec5de4c705c1e.png" src="../_images/fd2d523a39ae24a5b77809de04dc78055d87a333af9fdde5a46ec5de4c705c1e.png" style="width: 420px; height: 420px;" /></a>
</div>
</div>
<p>Seurat by default groups its cells by <code class="docutils literal notranslate"><span class="pre">idents</span></code> which we can set to our cell types or clustering</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="nf">Idents</span><span class="p">(</span><span class="n">atac</span><span class="p">)</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">atac</span><span class="p">[[]]</span><span class="o">$</span><span class="n">leiden</span>
</pre></div>
</div>
</div>
</div>
</section>
<section id="convert-seurat-assay-to-chromatin-assay">
<h3>Convert Seurat assay to chromatin assay<a class="headerlink" href="#convert-seurat-assay-to-chromatin-assay" title="Link to this heading">#</a></h3>
<p>Loading using <code class="docutils literal notranslate"><span class="pre">MuDataSeurat</span></code> does not support features with the same name in different assays (the same gene name for <code class="docutils literal notranslate"><span class="pre">rna</span></code> and <code class="docutils literal notranslate"><span class="pre">gene_activity</span></code>), which is why we had to add the assay name to the feature with <code class="docutils literal notranslate"><span class="pre">assay:feature</span></code>. Additionally, the <code class="docutils literal notranslate"><span class="pre">atac</span></code> assay is still a regular Seurat <code class="docutils literal notranslate"><span class="pre">Assay</span></code> and not a Signac <code class="docutils literal notranslate"><span class="pre">ChromatinAssay</span></code> which we need for chromatin specific downstream analyses.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="n">split_assay_names</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="kr">function</span><span class="p">(</span><span class="n">assay</span><span class="p">,</span><span class="w"> </span><span class="n">sep</span><span class="o">=</span><span class="s">&#39;:&#39;</span><span class="p">){</span>
<span class="w">    </span><span class="c1">#Function to split the assay name from the feature name</span>
<span class="w">    </span><span class="n">new_names</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">str_split</span><span class="p">(</span><span class="nf">rownames</span><span class="p">(</span><span class="n">assay</span><span class="p">),</span><span class="w"> </span><span class="n">sep</span><span class="p">,</span><span class="w"> </span><span class="n">simplify</span><span class="o">=</span><span class="kc">TRUE</span><span class="p">,</span><span class="w"> </span><span class="n">n</span><span class="o">=</span><span class="m">2</span><span class="p">)[,</span><span class="w"> </span><span class="m">2</span><span class="p">]</span>
<span class="w">    </span><span class="nf">rownames</span><span class="p">(</span><span class="n">assay</span><span class="p">)</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">new_names</span>
<span class="w">    </span><span class="kr">return</span><span class="p">(</span><span class="n">assay</span><span class="p">)</span>
<span class="p">}</span>
</pre></div>
</div>
</div>
</div>
<p>We will add gene annotation to our <code class="docutils literal notranslate"><span class="pre">ChromatinAssay</span></code>. You have to make sure that the chromosome names in your annotation match the names in your fragment files (so either <code class="docutils literal notranslate"><span class="pre">chr1</span></code> or <code class="docutils literal notranslate"><span class="pre">1</span></code>). See <code class="docutils literal notranslate"><span class="pre">?seqlevelsStyle</span></code> to see the different options.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="c1"># extract gene annotations from EnsDb</span>
<span class="n">annotations</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">GetGRangesFromEnsDb</span><span class="p">(</span><span class="n">ensdb</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">EnsDb.Hsapiens.v86</span><span class="p">)</span>
<span class="nf">seqlevelsStyle</span><span class="p">(</span><span class="n">annotations</span><span class="p">)</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="s">&#39;UCSC&#39;</span>
<span class="nf">genome</span><span class="p">(</span><span class="n">annotations</span><span class="p">)</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="s">&quot;hg38&quot;</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="n">data</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">split_assay_names</span><span class="p">(</span><span class="n">atac</span><span class="p">[[</span><span class="s">&#39;atac&#39;</span><span class="p">]]</span><span class="o">@</span><span class="n">data</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="n">chrom_assay</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">CreateChromatinAssay</span><span class="p">(</span>
<span class="w">  </span><span class="n">data</span><span class="o">=</span><span class="n">data</span><span class="p">,</span>
<span class="w">  </span><span class="n">sep</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="s">&quot;:&quot;</span><span class="p">,</span><span class="w"> </span><span class="s">&quot;-&quot;</span><span class="p">),</span><span class="w"> </span><span class="c1">#seperator of adata$var_names</span>
<span class="w">  </span><span class="n">genome</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&#39;hg38&#39;</span><span class="p">,</span><span class="w"> </span><span class="c1">#used reference genome</span>
<span class="w">  </span><span class="n">fragments</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">fragment_file</span><span class="p">,</span><span class="w"> </span>
<span class="w">  </span><span class="n">annotation</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">annotations</span>
<span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Computing hash
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="n">atac</span><span class="p">[[</span><span class="s">&#39;atac&#39;</span><span class="p">]]</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">chrom_assay</span>
</pre></div>
</div>
</div>
</div>
<p>Finally, we can check if we can plot all features in all assays</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="nf">DefaultAssay</span><span class="p">(</span><span class="n">atac</span><span class="p">)</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="s">&#39;atac&#39;</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="n">cov_plot</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">CoveragePlot</span><span class="p">(</span>
<span class="w">  </span><span class="n">object</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">atac</span><span class="p">,</span>
<span class="w">  </span><span class="n">region</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;PAX5&quot;</span><span class="p">,</span>
<span class="w">  </span><span class="n">annotation</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="kc">TRUE</span><span class="p">,</span>
<span class="w">  </span><span class="n">peaks</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="kc">TRUE</span><span class="p">,</span><span class="w"> </span>
<span class="p">)</span>
<span class="n">cov_plot</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Warning message:
&quot;Removed 1 rows containing missing values (`geom_segment()`).&quot;
</pre></div>
</div>
<a class="reference internal image-reference" href="../_images/1647ed19ce62da6b9f79dc23fb97457b07b1267198f4cafc518c9a7cdcc789b5.png"><img alt="../_images/1647ed19ce62da6b9f79dc23fb97457b07b1267198f4cafc518c9a7cdcc789b5.png" src="../_images/1647ed19ce62da6b9f79dc23fb97457b07b1267198f4cafc518c9a7cdcc789b5.png" style="width: 420px; height: 420px;" /></a>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="nf">DefaultAssay</span><span class="p">(</span><span class="n">atac</span><span class="p">)</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="s">&#39;rna&#39;</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="nf">FeaturePlot</span><span class="p">(</span><span class="n">atac</span><span class="p">,</span><span class="w"> </span><span class="s">&#39;rna:PAX5&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<a class="reference internal image-reference" href="../_images/9673b436f723e7368fac51f02f3a5657dba35978f69b60512fc92a615f77f192.png"><img alt="../_images/9673b436f723e7368fac51f02f3a5657dba35978f69b60512fc92a615f77f192.png" src="../_images/9673b436f723e7368fac51f02f3a5657dba35978f69b60512fc92a615f77f192.png" style="width: 420px; height: 420px;" /></a>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="nf">DefaultAssay</span><span class="p">(</span><span class="n">atac</span><span class="p">)</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="s">&#39;gene_activity&#39;</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="nf">FeaturePlot</span><span class="p">(</span><span class="n">atac</span><span class="p">,</span><span class="w"> </span><span class="s">&#39;gene-activity:PAX5&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<a class="reference internal image-reference" href="../_images/f350993f06e622cd8c6509a39edaa7e4b7120e56bf1e54cb377c0c406897a48f.png"><img alt="../_images/f350993f06e622cd8c6509a39edaa7e4b7120e56bf1e54cb377c0c406897a48f.png" src="../_images/f350993f06e622cd8c6509a39edaa7e4b7120e56bf1e54cb377c0c406897a48f.png" style="width: 420px; height: 420px;" /></a>
</div>
</div>
</section>
</section>
</section>

    <script type="text/x-thebe-config">
    {
        requestKernel: true,
        binderOptions: {
            repo: "binder-examples/jupyter-stacks-datascience",
            ref: "master",
        },
        codeMirrorConfig: {
            theme: "abcdef",
            mode: "r"
        },
        kernelOptions: {
            name: "ir",
            path: "./chromatin_accessibility"
        },
        predefinedOutput: true
    }
    </script>
    <script>kernelName = 'ir'</script>

                </article>
              

              
              
              
              
                <footer class="prev-next-footer d-print-none">
                  
<div class="prev-next-area">
</div>
                </footer>
              
            </div>
            
            
              
                <dialog id="pst-secondary-sidebar-modal"></dialog>
                <div id="pst-secondary-sidebar" class="bd-sidebar-secondary bd-toc"><div class="sidebar-secondary-items sidebar-secondary__inner">


  <div class="sidebar-secondary-item">
  <div class="page-toc tocsection onthispage">
    <i class="fa-solid fa-list"></i> Contents
  </div>
  <nav class="bd-toc-nav page-toc">
    <ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#environment-setup">Environment setup</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#load-data">Load data</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#read-in-mudata-object">Read in mudata object</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#convert-seurat-assay-to-chromatin-assay">Convert Seurat assay to chromatin assay</a></li>
</ul>
</li>
</ul>
  </nav></div>

</div></div>
              
            
          </div>
          <footer class="bd-footer-content">
            
<div class="bd-footer-content__inner container">
  
  <div class="footer-item">
    
<p class="component-author">
By Lukas Heumos, Anna Schaar, single-cell best practices consortium
</p>

  </div>
  
  <div class="footer-item">
    

  <p class="copyright">
    
      © Copyright 2023.
      <br/>
    
  </p>

  </div>
  
  <div class="footer-item">
    
  </div>
  
  <div class="footer-item">
    
<div class="extra_footer">
  <div>
  Brought to you by
  <a href="https://www.github.com/theislab">Theislab</a>, with many thanks to the single-cell community as a whole!
</div>

</div>
  </div>
  
</div>
          </footer>
        

      </main>
    </div>
  </div>
  
  <!-- Scripts loaded after <body> so the DOM is not blocked -->
  <script defer src="../_static/scripts/bootstrap.js?digest=8878045cc6db502f8baf"></script>
<script defer src="../_static/scripts/pydata-sphinx-theme.js?digest=8878045cc6db502f8baf"></script>

  <footer class="bd-footer">
  </footer>
  </body>
</html>