
<!DOCTYPE html>


<html lang="en" data-content_root="../" >

  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="viewport" content="width=device-width, initial-scale=1" />

    <title>4. Analysis frameworks and tools &#8212; Single-cell best practices</title>
  
  
  
  <script data-cfasync="false">
    document.documentElement.dataset.mode = localStorage.getItem("mode") || "";
    document.documentElement.dataset.theme = localStorage.getItem("theme") || "";
  </script>
  
  <!-- Loaded before other Sphinx assets -->
  <link href="../_static/styles/theme.css?digest=dfe6caa3a7d634c4db9b" rel="stylesheet" />
<link href="../_static/styles/bootstrap.css?digest=dfe6caa3a7d634c4db9b" rel="stylesheet" />
<link href="../_static/styles/pydata-sphinx-theme.css?digest=dfe6caa3a7d634c4db9b" rel="stylesheet" />

  
  <link href="../_static/vendor/fontawesome/6.5.2/css/all.min.css?digest=dfe6caa3a7d634c4db9b" rel="stylesheet" />
  <link rel="preload" as="font" type="font/woff2" crossorigin href="../_static/vendor/fontawesome/6.5.2/webfonts/fa-solid-900.woff2" />
<link rel="preload" as="font" type="font/woff2" crossorigin href="../_static/vendor/fontawesome/6.5.2/webfonts/fa-brands-400.woff2" />
<link rel="preload" as="font" type="font/woff2" crossorigin href="../_static/vendor/fontawesome/6.5.2/webfonts/fa-regular-400.woff2" />

    <link rel="stylesheet" type="text/css" href="../_static/pygments.css?v=03e43079" />
    <link rel="stylesheet" type="text/css" href="../_static/styles/sphinx-book-theme.css?v=eba8b062" />
    <link rel="stylesheet" type="text/css" href="../_static/togglebutton.css?v=13237357" />
    <link rel="stylesheet" type="text/css" href="../_static/copybutton.css?v=76b2166b" />
    <link rel="stylesheet" type="text/css" href="../_static/mystnb.4510f1fc1dee50b3e5859aac5469c37c29e427902b24a333a5f9fcb2f0b3ac41.css" />
    <link rel="stylesheet" type="text/css" href="../_static/sphinx-thebe.css?v=4fa983c6" />
    <link rel="stylesheet" type="text/css" href="../_static/sphinx-design.min.css?v=95c83b7e" />
    <link rel="stylesheet" type="text/css" href="../_static/book.css?v=148c11e2" />
  
  <!-- Pre-loaded scripts that we'll load fully later -->
  <link rel="preload" as="script" href="../_static/scripts/bootstrap.js?digest=dfe6caa3a7d634c4db9b" />
<link rel="preload" as="script" href="../_static/scripts/pydata-sphinx-theme.js?digest=dfe6caa3a7d634c4db9b" />
  <script src="../_static/vendor/fontawesome/6.5.2/js/all.min.js?digest=dfe6caa3a7d634c4db9b"></script>

    <script src="../_static/documentation_options.js?v=9eb32ce0"></script>
    <script src="../_static/doctools.js?v=9a2dae69"></script>
    <script src="../_static/sphinx_highlight.js?v=dc90522c"></script>
    <script src="../_static/clipboard.min.js?v=a7894cd8"></script>
    <script src="../_static/copybutton.js?v=f281be69"></script>
    <script src="../_static/scripts/sphinx-book-theme.js?v=887ef09a"></script>
    <script>let toggleHintShow = 'Click to show';</script>
    <script>let toggleHintHide = 'Click to hide';</script>
    <script>let toggleOpenOnPrint = 'true';</script>
    <script src="../_static/togglebutton.js?v=4a39c7ea"></script>
    <script>var togglebuttonSelector = '.toggle, .admonition.dropdown';</script>
    <script src="../_static/design-tabs.js?v=f930bc37"></script>
    <script>const THEBE_JS_URL = "https://unpkg.com/thebe@0.8.2/lib/index.js"; const thebe_selector = ".thebe,.cell"; const thebe_selector_input = "pre"; const thebe_selector_output = ".output, .cell_output"</script>
    <script async="async" src="../_static/sphinx-thebe.js?v=c100c467"></script>
    <script>var togglebuttonSelector = '.toggle, .admonition.dropdown';</script>
    <script>const THEBE_JS_URL = "https://unpkg.com/thebe@0.8.2/lib/index.js"; const thebe_selector = ".thebe,.cell"; const thebe_selector_input = "pre"; const thebe_selector_output = ".output, .cell_output"</script>
    <script>DOCUMENTATION_OPTIONS.pagename = 'introduction/analysis_tools';</script>
    <script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
    <script src="../_static/book.js?v=57d67cd3"></script>
    <link rel="icon" href="../_static/favicon.ico"/>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="5. Interoperability" href="interoperability.html" />
    <link rel="prev" title="3. Raw data processing" href="raw_data_processing.html" />
  <meta name="viewport" content="width=device-width, initial-scale=1"/>
  <meta name="docsearch:language" content="en"/>
  </head>
  
  
  <body data-bs-spy="scroll" data-bs-target=".bd-toc-nav" data-offset="180" data-bs-root-margin="0px 0px -60%" data-default-mode="">

  
  
  <div id="pst-skip-link" class="skip-link d-print-none"><a href="#main-content">Skip to main content</a></div>
  
  <div id="pst-scroll-pixel-helper"></div>
  
  <button type="button" class="btn rounded-pill" id="pst-back-to-top">
    <i class="fa-solid fa-arrow-up"></i>Back to top</button>

  
  <input type="checkbox"
          class="sidebar-toggle"
          id="pst-primary-sidebar-checkbox"/>
  <label class="overlay overlay-primary" for="pst-primary-sidebar-checkbox"></label>
  
  <input type="checkbox"
          class="sidebar-toggle"
          id="pst-secondary-sidebar-checkbox"/>
  <label class="overlay overlay-secondary" for="pst-secondary-sidebar-checkbox"></label>
  
  <div class="search-button__wrapper">
    <div class="search-button__overlay"></div>
    <div class="search-button__search-container">
<form class="bd-search d-flex align-items-center"
      action="../search.html"
      method="get">
  <i class="fa-solid fa-magnifying-glass"></i>
  <input type="search"
         class="form-control"
         name="q"
         id="search-input"
         placeholder="Search this book..."
         aria-label="Search this book..."
         autocomplete="off"
         autocorrect="off"
         autocapitalize="off"
         spellcheck="false"/>
  <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd>K</kbd></span>
</form></div>
  </div>

  <div class="pst-async-banner-revealer d-none">
  <aside id="bd-header-version-warning" class="d-none d-print-none" aria-label="Version warning"></aside>
</div>

  
    <header class="bd-header navbar navbar-expand-lg bd-navbar d-print-none">
    </header>
  

  <div class="bd-container">
    <div class="bd-container__inner bd-page-width">
      
      
      
      <div class="bd-sidebar-primary bd-sidebar">
        

  
  <div class="sidebar-header-items sidebar-primary__section">
    
    
    
    
  </div>
  
    <div class="sidebar-primary-items__start sidebar-primary__section">
        <div class="sidebar-primary-item">

  
    
  

<a class="navbar-brand logo" href="../preamble.html">
  
  
  
  
  
    
    
      
    
    
    <img src="../_static/logo.png" class="logo__image only-light" alt="Single-cell best practices - Home"/>
    <script>document.write(`<img src="../_static/logo.png" class="logo__image only-dark" alt="Single-cell best practices - Home"/>`);</script>
  
  
</a></div>
        <div class="sidebar-primary-item">

 <script>
 document.write(`
   <button class="btn search-button-field search-button__button" title="Search" aria-label="Search" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="fa-solid fa-magnifying-glass"></i>
    <span class="search-button__default-text">Search</span>
    <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd class="kbd-shortcut__modifier">K</kbd></span>
   </button>
 `);
 </script></div>
        <div class="sidebar-primary-item"><nav class="bd-links bd-docs-nav" aria-label="Main">
    <div class="bd-toc-item navbar-nav active">
        <p aria-level="2" class="caption" role="heading"><span class="caption-text">Introduction</span></p>
<ul class="current nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="prior_art.html">1. Prior art</a></li>
<li class="toctree-l1"><a class="reference internal" href="scrna_seq.html">2. Single-cell RNA sequencing</a></li>
<li class="toctree-l1"><a class="reference internal" href="raw_data_processing.html">3. Raw data processing</a></li>
<li class="toctree-l1 current active"><a class="current reference internal" href="#">4. Analysis frameworks and tools</a></li>
<li class="toctree-l1"><a class="reference internal" href="interoperability.html">5. Interoperability</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Preprocessing and visualization</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../preprocessing_visualization/quality_control.html">6. Quality Control</a></li>
<li class="toctree-l1"><a class="reference internal" href="../preprocessing_visualization/normalization.html">7. Normalization</a></li>
<li class="toctree-l1"><a class="reference internal" href="../preprocessing_visualization/feature_selection.html">8. Feature selection</a></li>
<li class="toctree-l1"><a class="reference internal" href="../preprocessing_visualization/dimensionality_reduction.html">9. Dimensionality Reduction</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Identifying cellular structure</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../cellular_structure/clustering.html">10. Clustering</a></li>
<li class="toctree-l1"><a class="reference internal" href="../cellular_structure/annotation.html">11. Annotation</a></li>
<li class="toctree-l1"><a class="reference internal" href="../cellular_structure/integration.html">12. Data integration</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Inferring trajectories</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../trajectories/pseudotemporal.html">13. Pseudotemporal ordering</a></li>
<li class="toctree-l1"><a class="reference internal" href="../trajectories/rna_velocity.html">14. RNA velocity</a></li>
<li class="toctree-l1"><a class="reference internal" href="../trajectories/lineage_tracing.html">15. Lineage tracing</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Dealing with conditions</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../conditions/differential_gene_expression.html">16. Differential gene expression analysis</a></li>
<li class="toctree-l1"><a class="reference internal" href="../conditions/compositional.html">17. Compositional analysis</a></li>
<li class="toctree-l1"><a class="reference internal" href="../conditions/gsea_pathway.html">18. Gene set enrichment and pathway analysis</a></li>
<li class="toctree-l1"><a class="reference internal" href="../conditions/perturbation_modeling.html">19. Perturbation modeling</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Modeling mechanisms</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../mechanisms/gene_regulatory_networks.html">20. Gene regulatory networks</a></li>
<li class="toctree-l1"><a class="reference internal" href="../mechanisms/cell_cell_communication.html">21. Cell-cell communication</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Deconvolution</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../deconvolution/bulk_deconvolution.html">22. Bulk deconvolution</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Chromatin Accessibility</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../chromatin_accessibility/introduction.html">23. Single-cell ATAC sequencing</a></li>
<li class="toctree-l1"><a class="reference internal" href="../chromatin_accessibility/quality_control.html">24. Quality Control</a></li>
<li class="toctree-l1"><a class="reference internal" href="../chromatin_accessibility/gene_regulatory_networks_atac.html">25. Gene regulatory networks using chromatin accessibility</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Spatial omics</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../spatial/introduction.html">26. Single-cell data resolved in space</a></li>
<li class="toctree-l1"><a class="reference internal" href="../spatial/neighborhood.html">27. Neighborhood analysis</a></li>
<li class="toctree-l1"><a class="reference internal" href="../spatial/domains.html">28. Spatial domains</a></li>
<li class="toctree-l1"><a class="reference internal" href="../spatial/spatially_variable_genes.html">29. Spatially variable genes</a></li>
<li class="toctree-l1"><a class="reference internal" href="../spatial/deconvolution.html">30. Spatial deconvolution</a></li>
<li class="toctree-l1"><a class="reference internal" href="../spatial/imputation.html">31. Imputation</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Surface protein</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../surface_protein/quality_control.html">32. Quality control</a></li>
<li class="toctree-l1"><a class="reference internal" href="../surface_protein/normalization.html">33. Normalization</a></li>
<li class="toctree-l1"><a class="reference internal" href="../surface_protein/doublet_detection.html">34. Doublet detection</a></li>
<li class="toctree-l1"><a class="reference internal" href="../surface_protein/dimensionality_reduction.html">35. Dimensionality Reduction</a></li>
<li class="toctree-l1"><a class="reference internal" href="../surface_protein/batch_correction.html">36. Batch correction</a></li>
<li class="toctree-l1"><a class="reference internal" href="../surface_protein/annotation.html">37. Annotation</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Adaptive immune receptor repertoire</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../air_repertoire/ir_profiling.html">38. Immune Receptor Profiling</a></li>
<li class="toctree-l1"><a class="reference internal" href="../air_repertoire/clonotype.html">39. Clonotype analysis</a></li>



<li class="toctree-l1"><a class="reference internal" href="../air_repertoire/specificity.html">43. Specificity analysis</a></li>
<li class="toctree-l1"><a class="reference internal" href="../air_repertoire/multimodal_integration.html">44. Integrating AIR and transcriptomics</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Multimodal integration</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../multimodal_integration/paired_integration.html">45. Paired integration</a></li>
<li class="toctree-l1"><a class="reference internal" href="../multimodal_integration/advanced_integration.html">46. Advanced integration</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Outlook</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../outlook.html">47. Outlook</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Acknowledgements</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../acknowledgements.html">48. Acknowledgements</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Glossary</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../glossary.html">49. Glossary</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Changelog</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../CHANGELOG.html">Changelog</a></li>
</ul>

    </div>
</nav></div>
    </div>
  
  
  <div class="sidebar-primary-items__end sidebar-primary__section">
  </div>
  
  <div id="rtd-footer-container"></div>


      </div>
      
      <main id="main-content" class="bd-main" role="main">
        
        

<div class="sbt-scroll-pixel-helper"></div>

          <div class="bd-content">
            <div class="bd-article-container">
              
              <div class="bd-header-article d-print-none">
<div class="header-article-items header-article__inner">
  
    <div class="header-article-items__start">
      
        <div class="header-article-item"><button class="sidebar-toggle primary-toggle btn btn-sm" title="Toggle primary sidebar" data-bs-placement="bottom" data-bs-toggle="tooltip">
  <span class="fa-solid fa-bars"></span>
</button></div>
      
    </div>
  
  
    <div class="header-article-items__end">
      
        <div class="header-article-item">

<div class="article-header-buttons">





<div class="dropdown dropdown-source-buttons">
  <button class="btn dropdown-toggle" type="button" data-bs-toggle="dropdown" aria-expanded="false" aria-label="Source repositories">
    <i class="fab fa-github"></i>
  </button>
  <ul class="dropdown-menu">
      
      
      
      <li><a href="https://github.com/theislab/single-cell-best-practices" target="_blank"
   class="btn btn-sm btn-source-repository-button dropdown-item"
   title="Source repository"
   data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fab fa-github"></i>
  </span>
<span class="btn__text-container">Repository</span>
</a>
</li>
      
      
      
      
      <li><a href="https://github.com/theislab/single-cell-best-practices/edit/master/jupyter-book/introduction/analysis_tools.ipynb" target="_blank"
   class="btn btn-sm btn-source-edit-button dropdown-item"
   title="Suggest edit"
   data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-pencil-alt"></i>
  </span>
<span class="btn__text-container">Suggest edit</span>
</a>
</li>
      
      
      
      
      <li><a href="https://github.com/theislab/single-cell-best-practices/issues/new?title=Issue%20on%20page%20%2Fintroduction/analysis_tools.html&body=Your%20issue%20content%20here." target="_blank"
   class="btn btn-sm btn-source-issues-button dropdown-item"
   title="Open an issue"
   data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-lightbulb"></i>
  </span>
<span class="btn__text-container">Open issue</span>
</a>
</li>
      
  </ul>
</div>






<div class="dropdown dropdown-download-buttons">
  <button class="btn dropdown-toggle" type="button" data-bs-toggle="dropdown" aria-expanded="false" aria-label="Download this page">
    <i class="fas fa-download"></i>
  </button>
  <ul class="dropdown-menu">
      
      
      
      <li><a href="../_sources/introduction/analysis_tools.ipynb" target="_blank"
   class="btn btn-sm btn-download-source-button dropdown-item"
   title="Download source file"
   data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-file"></i>
  </span>
<span class="btn__text-container">.ipynb</span>
</a>
</li>
      
      
      
      
      <li>
<button onclick="window.print()"
  class="btn btn-sm btn-download-pdf-button dropdown-item"
  title="Print to PDF"
  data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-file-pdf"></i>
  </span>
<span class="btn__text-container">.pdf</span>
</button>
</li>
      
  </ul>
</div>




<button onclick="toggleFullScreen()"
  class="btn btn-sm btn-fullscreen-button"
  title="Fullscreen mode"
  data-bs-placement="bottom" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-expand"></i>
  </span>

</button>



<script>
document.write(`
  <button class="btn btn-sm nav-link pst-navbar-icon theme-switch-button" title="light/dark" aria-label="light/dark" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="theme-switch fa-solid fa-sun fa-lg" data-mode="light"></i>
    <i class="theme-switch fa-solid fa-moon fa-lg" data-mode="dark"></i>
    <i class="theme-switch fa-solid fa-circle-half-stroke fa-lg" data-mode="auto"></i>
  </button>
`);
</script>


<script>
document.write(`
  <button class="btn btn-sm pst-navbar-icon search-button search-button__button" title="Search" aria-label="Search" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="fa-solid fa-magnifying-glass fa-lg"></i>
  </button>
`);
</script>
<button class="sidebar-toggle secondary-toggle btn btn-sm" title="Toggle secondary sidebar" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <span class="fa-solid fa-list"></span>
</button>
</div></div>
      
    </div>
  
</div>
</div>
              
              

<div id="jb-print-docs-body" class="onlyprint">
    <h1>Analysis frameworks and tools</h1>
    <!-- Table of contents -->
    <div id="print-main-content">
        <div id="jb-print-toc">
            
            <div>
                <h2> Contents </h2>
            </div>
            <nav aria-label="Page">
                <ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#single-cell-analysis-frameworks-and-consortia">4.1. Single-cell analysis frameworks and consortia</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#storing-unimodal-data-with-anndata">4.2. Storing unimodal data with AnnData</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#installation">4.2.1. Installation</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#initializing-an-anndata-object">4.2.2. Initializing an AnnData object</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#adding-aligned-metadata">4.2.3. Adding aligned metadata</a><ul class="nav section-nav flex-column">
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#observational-or-variable-level">4.2.3.1. Observational or Variable level</a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#subsetting-using-metadata">4.2.3.2. Subsetting using metadata</a></li>
</ul>
</li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#observation-variable-level-matrices">4.2.4. Observation/variable-level matrices</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#unstructured-metadata">4.2.5. Unstructured metadata</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#layers">4.2.6. Layers</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#conversion-to-dataframes">4.2.7. Conversion to DataFrames</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#reading-and-writing-of-anndata-objects">4.2.8. Reading and writing of AnnData objects</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#efficient-data-access">4.2.9. Efficient data access</a><ul class="nav section-nav flex-column">
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#view-and-copies">4.2.9.1. View and copies</a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#partial-reading-of-large-data">4.2.9.2. Partial reading of large data</a></li>
</ul>
</li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#unimodal-data-analysis-with-scanpy">4.3. Unimodal data analysis with scanpy</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#id12">4.3.1. Installation</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#scanpy-api-design">4.3.2. Scanpy API design</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#scanpy-example">4.3.3. Scanpy example</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#advanced-using-mudata-to-store-multimodal-data">4.4. Advanced: Using MuData to store multimodal data</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#id18">4.4.1. Installation</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#initializing-a-mudata-object">4.4.2. Initializing a MuData object</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#mudata-attributes">4.4.3. MuData attributes</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#variable-mappings">4.4.4. Variable mappings</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#mudata-views">4.4.5. MuData views</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#common-observations">4.4.6. Common observations</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#reading-and-writing-of-mudata-objects">4.4.7. Reading and Writing of MuData objects</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#multimodal-methods">4.4.8. Multimodal methods</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#advanced-multimodal-data-analysis-with-muon">4.5. Advanced: Multimodal data analysis with muon</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#id20">4.5.1. Installation</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#api-overview">4.5.2. API overview</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#muon-api-demo">4.5.3. muon API demo</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#references">4.6. References</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#contributors">4.7. Contributors</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#authors">4.7.1. Authors</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#reviewers">4.7.2. Reviewers</a></li>
</ul>
</li>
</ul>
            </nav>
        </div>
    </div>
</div>

              
                
<div id="searchbox"></div>
                <article class="bd-article">
                  
  <section class="tex2jax_ignore mathjax_ignore" id="analysis-frameworks-and-tools">
<span id="introduction-analysis-frameworks"></span><h1><span class="section-number">4. </span>Analysis frameworks and tools<a class="headerlink" href="#analysis-frameworks-and-tools" title="Link to this heading">#</a></h1>
<details class="sd-sphinx-override sd-dropdown sd-card sd-mb-3">
<summary class="sd-summary-title sd-card-header">
<span class="sd-summary-text"><i class="fas fa-brain"></i>   Key takeaways</span><span class="sd-summary-state-marker sd-summary-chevron-right"><svg version="1.1" width="1.5em" height="1.5em" class="sd-octicon sd-octicon-chevron-right" viewBox="0 0 24 24" aria-hidden="true"><path d="M8.72 18.78a.75.75 0 0 1 0-1.06L14.44 12 8.72 6.28a.751.751 0 0 1 .018-1.042.751.751 0 0 1 1.042-.018l6.25 6.25a.75.75 0 0 1 0 1.06l-6.25 6.25a.75.75 0 0 1-1.06 0Z"></path></svg></span></summary><div class="sd-summary-content sd-card-body docutils">
<div class="sd-card sd-sphinx-override sd-mb-3 sd-shadow-sm sd-card-hover docutils">
<div class="sd-card-body docutils">
<p class="sd-card-text">The scverse ecosystem is preferred for single-cell analysis due to its scalability, Python interoperability, and consistent data structures.</p>
</div>
<a class="sd-stretched-link sd-hide-link-text reference internal" href="#introduction-analysis-tools-key-takeaway-1"><span class="std std-ref">Single-cell analysis frameworks and consortia</span></a></div>
</div>
</details><details class="sd-sphinx-override sd-dropdown sd-card sd-mb-3">
<summary class="sd-summary-title sd-card-header">
<span class="sd-summary-text"><i class="fa-solid fa-gear"></i>   Environment setup</span><span class="sd-summary-state-marker sd-summary-chevron-right"><svg version="1.1" width="1.5em" height="1.5em" class="sd-octicon sd-octicon-chevron-right" viewBox="0 0 24 24" aria-hidden="true"><path d="M8.72 18.78a.75.75 0 0 1 0-1.06L14.44 12 8.72 6.28a.751.751 0 0 1 .018-1.042.751.751 0 0 1 1.042-.018l6.25 6.25a.75.75 0 0 1 0 1.06l-6.25 6.25a.75.75 0 0 1-1.06 0Z"></path></svg></span></summary><div class="sd-summary-content sd-card-body docutils">
<div class="sd-tab-set docutils">
<input checked="checked" id="sd-tab-item-0" name="sd-tab-set-0" type="radio">
<label class="sd-tab-label" for="sd-tab-item-0">
Steps</label><div class="sd-tab-content docutils">
<ol class="arabic">
<li><p class="sd-card-text"><strong>Install conda</strong>:</p>
<ul class="simple">
<li><p class="sd-card-text">Before creating the environment, ensure that <a class="reference external" href="https://docs.conda.io/projects/conda/en/latest/user-guide/install/index.html">conda</a> is installed on your system.</p></li>
</ul>
</li>
<li><p class="sd-card-text"><strong>Save the yml content</strong>:</p>
<ul class="simple">
<li><p class="sd-card-text">Copy the content from the yml tab into a file named <code class="docutils literal notranslate"><span class="pre">environment.yml</span></code>.</p></li>
</ul>
</li>
<li><p class="sd-card-text"><strong>Create the environment</strong>:</p>
<ul>
<li><p class="sd-card-text">Open a terminal or command prompt.</p></li>
<li><p class="sd-card-text">Run the following command:</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>conda<span class="w"> </span>env<span class="w"> </span>create<span class="w"> </span>-f<span class="w"> </span>environment.yml
</pre></div>
</div>
</li>
</ul>
</li>
<li><p class="sd-card-text"><strong>Activate the environment</strong>:</p>
<ul>
<li><p class="sd-card-text">After the environment is created, activate it using:</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>conda<span class="w"> </span>activate<span class="w"> </span>&lt;environment_name&gt;
</pre></div>
</div>
</li>
<li><p class="sd-card-text">Replace <code class="docutils literal notranslate"><span class="pre">&lt;environment_name&gt;</span></code> with the name specified in the <code class="docutils literal notranslate"><span class="pre">environment.yml</span></code> file. In the yml file it will look like this:</p>
<div class="highlight-yaml notranslate"><div class="highlight"><pre><span></span><span class="nt">name</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">&lt;environment_name&gt;</span>
</pre></div>
</div>
</li>
</ul>
</li>
<li><p class="sd-card-text"><strong>Verify the installation</strong>:</p>
<ul>
<li><p class="sd-card-text">Check that the environment was created successfully by running:</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>conda<span class="w"> </span>env<span class="w"> </span>list
</pre></div>
</div>
</li>
</ul>
</li>
</ol>
</div>
<input id="sd-tab-item-1" name="sd-tab-set-0" type="radio">
<label class="sd-tab-label" for="sd-tab-item-1">
yml</label><div class="sd-tab-content docutils">
<div class="highlight-yaml notranslate"><div class="highlight"><pre><span></span><span class="nt">name</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">analysis-tools</span>
<span class="nt">channels</span><span class="p">:</span>
<span class="w">  </span><span class="p p-Indicator">-</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">conda-forge</span>
<span class="w">  </span><span class="p p-Indicator">-</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">bioconda</span>
<span class="nt">dependencies</span><span class="p">:</span>
<span class="w">  </span><span class="p p-Indicator">-</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">conda-forge::scanpy=1.9.1</span>
<span class="w">  </span><span class="p p-Indicator">-</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">conda-forge::muon=0.1.2</span>
</pre></div>
</div>
</div>
</div>
</div>
</details><!-- END DROPDOWNS -->
<section id="single-cell-analysis-frameworks-and-consortia">
<span id="introduction-analysis-tools-key-takeaway-1"></span><h2><span class="section-number">4.1. </span>Single-cell analysis frameworks and consortia<a class="headerlink" href="#single-cell-analysis-frameworks-and-consortia" title="Link to this heading">#</a></h2>
<p>After having obtained the count matrices, as described earlier, the exploratory data analysis phase begins. Due to the size and complexity of the data, specialized tooling is required. While in the early days of single-cell analysis people used to analyze their data with custom scripts, frameworks for specifically this purpose now exist. The three most popular options are the R based Bioconductor<span id="id1">[<a class="reference internal" href="#id159" title="Wolfgang Huber, Vincent J. Carey, Robert Gentleman, Simon Anders, Marc Carlson, Benilton S. Carvalho, Hector Corrada Bravo, Sean Davis, Laurent Gatto, Thomas Girke, Raphael Gottardo, Florian Hahne, Kasper D. Hansen, Rafael A. Irizarry, Michael Lawrence, Michael I. Love, James MacDonald, Valerie Obenchain, Andrzej K. Oleś, Hervé Pagès, Alejandro Reyes, Paul Shannon, Gordon K. Smyth, Dan Tenenbaum, Levi Waldron, and Martin Morgan. Orchestrating high-throughput genomic analysis with bioconductor. Nature Methods, 12(2):115-121, Feb 2015. URL: https://doi.org/10.1038/nmeth.3252, doi:10.1038/nmeth.3252.">Huber <em>et al.</em>, 2015</a>]</span> and Seurat<span id="id2">[<a class="reference internal" href="#id133" title="Yuhan Hao, Stephanie Hao, Erica Andersen-Nissen, William M. Mauck III, Shiwei Zheng, Andrew Butler, Maddie J. Lee, Aaron J. Wilk, Charlotte Darby, Michael Zagar, Paul Hoffman, Marlon Stoeckius, Efthymia Papalexi, Eleni P. Mimitou, Jaison Jain, Avi Srivastava, Tim Stuart, Lamar B. Fleming, Bertrand Yeung, Angela J. Rogers, Juliana M. McElrath, Catherine A. Blish, Raphael Gottardo, Peter Smibert, and Rahul Satija. Integrated analysis of multimodal single-cell data. Cell, 2021. URL: https://doi.org/10.1016/j.cell.2021.04.048, doi:10.1016/j.cell.2021.04.048.">Hao <em>et al.</em>, 2021</a>]</span> ecosystems and the Python based scverse<span id="id3">[<a class="reference internal" href="#id134" title="scverse. Scverse. https://scverse.org, 2022. Accessed: 2022-04-21.">scverse, 2022</a>]</span> ecosystem. These differ not only in the used programming languages, but also in the underlying data structures and available specialized analysis tools.</p>
<p>Bioconductor is a project which develops, supports and shares free open source software with a focus on rigorous and reproducible analysis of data for many different biological assays including single-cell. A homogeneous developer and user experience and extensive documentation with user friendly vignettes are the biggest strengths of Bioconductor. Seurat is a well regarded R package specifically designed for the analysis of single-cell data. It offers tooling for all steps of the analysis including multimodal and spatial data. The well written vignettes and the large user-base is what Seurat is known for.
However, both R options can run into scalability issues for extremely large datasets (more than half a million cells) which motivated the Python based community to develop the scverse ecosystem. scverse is an organization and an ecosystem dedicated to foundational tools in the life sciences with an initial focus on single-cell. Scalability, extendability and strong interoperability with the existing Python data and machine learning tooling are some of the advantages of the scverse ecosystem.</p>
<p>All three ecosystems are involved in many efforts to allow for interoperability of the involved frameworks. This will be discussed in the “Interoperability” chapter. This book always focuses on the best tools for the corresponding question and will therefore use a mix of the above-mentioned ecosystems for specific issues. However, the basis of all analyses will be the scverse ecosystem for two reasons:</p>
<ol class="arabic simple">
<li><p>While we will regularly switch ecosystems and even programming languages throughout this book, a consistent use of data structures and tooling helps readers to focus on the concepts and not on implementation details.</p></li>
<li><p><a class="reference external" href="https://bioconductor.org/books/release/OSCA/">A great book on exclusively the Bioconductor ecosystem</a> already exists. We encourage users that only want to learn about single-cell analysis with Bioconductor to read it.</p></li>
</ol>
<p>In the following sections the scverse ecosystem will be introduced in more detail and the most important concepts will be explained with a focus on the most important data structures. This introduction cannot cover all aspects of the data structures and frameworks. It is not in scope to introduce all available analysis functions. We therefore refer to the respective frameworks’ tutorials and documentation where required.</p>
</section>
<section id="storing-unimodal-data-with-anndata">
<h2><span class="section-number">4.2. </span>Storing unimodal data with AnnData<a class="headerlink" href="#storing-unimodal-data-with-anndata" title="Link to this heading">#</a></h2>
<p>As previously discussed, after alignment and gene annotation, genomics data is typically summarized to a feature matrix. This matrix will be of the shape <code class="docutils literal notranslate"><span class="pre">number_observations</span> <span class="pre">x</span> <span class="pre">number_variables</span></code> – where for scRNA-seq observations are cellular barcodes and the variables are annotated genes. Over the course of an analysis the observations and variables of this matrix are annotated with computationally derived measurements (e.g. quality control metrics, or latent space embeddings) and prior knowledge (e.g. source donor or alternative gene identifier). In the scverse ecosystem, AnnData<span id="id4">[<a class="reference internal" href="#id135" title="Isaac Virshup, Sergei Rybakov, Fabian J. Theis, Philipp Angerer, and F. Alexander Wolf. Anndata: annotated data. bioRxiv, 2021. URL: https://www.biorxiv.org/content/early/2021/12/19/2021.12.16.473007, arXiv:https://www.biorxiv.org/content/early/2021/12/19/2021.12.16.473007.full.pdf, doi:10.1101/2021.12.16.473007.">Virshup <em>et al.</em>, 2021</a>]</span> is used to associate the data matrix with these annotations. To allow for fast and memory efficient transformations, AnnData also supports sparse matrices and partial reading.</p>
<p>While AnnData is broadly similar to data structures from the R ecosystems (e.g. <a class="reference external" href="https://bioconductor.org/packages/release/bioc/html/SummarizedExperiment.html">Bioconductor’s SummarizedExperiment</a> or <a class="reference external" href="https://mojaveazure.github.io/seurat-object/">Seurat’s object</a>), R packages use a transposed feature matrix.</p>
<p>At its core, an AnnData object stores a sparse or dense matrix (the count matrix in the case of scRNA-Seq) in <code class="docutils literal notranslate"><span class="pre">X</span></code>. This matrix has the dimensions of <code class="docutils literal notranslate"><span class="pre">obs_names</span> <span class="pre">x</span> <span class="pre">var_names</span></code> where the obs (=observations) correspond to the cells’ barcodes and the var (=variables) correspond to the gene identifiers. This matrix <code class="docutils literal notranslate"><span class="pre">X</span></code> is surrounded by Pandas DataFrames <code class="docutils literal notranslate"><span class="pre">obs</span></code> and <code class="docutils literal notranslate"><span class="pre">var</span></code> which save annotations of cells and genes respectively. Further, AnnData saves whole matrices of calculations for the observations (<code class="docutils literal notranslate"><span class="pre">obsm</span></code>) or variables (<code class="docutils literal notranslate"><span class="pre">varm</span></code>) with the corresponding dimensions. Graph like structures which associate cells with cells or genes with genes are usually saved in <code class="docutils literal notranslate"><span class="pre">obsp</span></code> and <code class="docutils literal notranslate"><span class="pre">varp</span></code>. Any other unstructured data which does not fit any other slot is saved as unstructured data in <code class="docutils literal notranslate"><span class="pre">uns</span></code>. It is further possible to store more values of <code class="docutils literal notranslate"><span class="pre">X</span></code> in <code class="docutils literal notranslate"><span class="pre">layers</span></code>. Use cases for this are for example the storage of raw, unnormalized count data in a <code class="docutils literal notranslate"><span class="pre">counts</span></code> layer and the normalized data in the unnamed default layer.<br />
AnnData is primarily designed for unimodal (for example just scRNA-Seq) data. However, extensions of AnnData such as MuData, which is covered later in this chapter, allow for the efficient storage and access of multimodal data.</p>
<figure class="align-default" id="anndata-fig">
<a class="bg-primary mb-1 reference internal image-reference" href="../_images/anndata.jpg"><img alt="AnnData Overview" class="bg-primary mb-1" src="../_images/anndata.jpg" style="width: 800px;" />
</a>
<figcaption>
<p><span class="caption-number">Fig. 4.1 </span><span class="caption-text">AnnData overview. Image obtained from <span id="id5">[<a class="reference internal" href="#id135" title="Isaac Virshup, Sergei Rybakov, Fabian J. Theis, Philipp Angerer, and F. Alexander Wolf. Anndata: annotated data. bioRxiv, 2021. URL: https://www.biorxiv.org/content/early/2021/12/19/2021.12.16.473007, arXiv:https://www.biorxiv.org/content/early/2021/12/19/2021.12.16.473007.full.pdf, doi:10.1101/2021.12.16.473007.">Virshup <em>et al.</em>, 2021</a>]</span>.</span><a class="headerlink" href="#anndata-fig" title="Link to this image">#</a></p>
</figcaption>
</figure>
<section id="installation">
<h3><span class="section-number">4.2.1. </span>Installation<a class="headerlink" href="#installation" title="Link to this heading">#</a></h3>
<p>AnnData is available on PyPI and Conda can be installed with using either:</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>pip<span class="w"> </span>install<span class="w"> </span>anndata
conda<span class="w"> </span>install<span class="w"> </span>-c<span class="w"> </span>conda-forge<span class="w"> </span>anndata
</pre></div>
</div>
</section>
<section id="initializing-an-anndata-object">
<h3><span class="section-number">4.2.2. </span>Initializing an AnnData object<a class="headerlink" href="#initializing-an-anndata-object" title="Link to this heading">#</a></h3>
<p>This section is inspired by AnnData’s “getting started” tutorial: <a class="reference external" href="https://anndata.readthedocs.io/en/latest/tutorials/notebooks/getting-started.html">https://anndata.readthedocs.io/en/latest/tutorials/notebooks/getting-started.html</a></p>
<p>Let us create a simple AnnData object with sparse count information which may for example represent gene expression counts. First, we import the required packages.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span><span class="w"> </span><span class="nn">anndata</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">ad</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">numpy</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">np</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">pandas</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">pd</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">scipy.sparse</span><span class="w"> </span><span class="kn">import</span> <span class="n">csr_matrix</span>
</pre></div>
</div>
</div>
</div>
<p>As a next step we initialize an AnnData object with random poisson distributed data. It is an unwritten rule to name the primary AnnData object of the analysis <code class="docutils literal notranslate"><span class="pre">adata</span></code>.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">counts</span> <span class="o">=</span> <span class="n">csr_matrix</span><span class="p">(</span>
    <span class="n">np</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">default_rng</span><span class="p">()</span><span class="o">.</span><span class="n">poisson</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="n">size</span><span class="o">=</span><span class="p">(</span><span class="mi">100</span><span class="p">,</span> <span class="mi">2000</span><span class="p">)),</span> <span class="n">dtype</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">float32</span>
<span class="p">)</span>
<span class="n">adata</span> <span class="o">=</span> <span class="n">ad</span><span class="o">.</span><span class="n">AnnData</span><span class="p">(</span><span class="n">counts</span><span class="p">)</span>
<span class="n">adata</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>AnnData object with n_obs × n_vars = 100 × 2000
</pre></div>
</div>
</div>
</div>
<p>The obtained AnnData object has 100 observations and 2000 variables. This would correspond to 100 cells with 2000 genes. The initial data we passed are accessible as a sparse matrix using <code class="docutils literal notranslate"><span class="pre">adata.X</span></code>.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">adata</span><span class="o">.</span><span class="n">X</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>&lt;100x2000 sparse matrix of type &#39;&lt;class &#39;numpy.float32&#39;&gt;&#39;
	with 126413 stored elements in Compressed Sparse Row format&gt;
</pre></div>
</div>
</div>
</div>
<p>Now, we provide the index to both the <code class="docutils literal notranslate"><span class="pre">obs</span></code> and <code class="docutils literal notranslate"><span class="pre">var</span></code> axes using <code class="docutils literal notranslate"><span class="pre">.obs_names</span></code> and <code class="docutils literal notranslate"><span class="pre">.var_names</span></code> respectively.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">adata</span><span class="o">.</span><span class="n">obs_names</span> <span class="o">=</span> <span class="p">[</span><span class="sa">f</span><span class="s2">&quot;Cell_</span><span class="si">{</span><span class="n">i</span><span class="si">:</span><span class="s2">d</span><span class="si">}</span><span class="s2">&quot;</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">adata</span><span class="o">.</span><span class="n">n_obs</span><span class="p">)]</span>
<span class="n">adata</span><span class="o">.</span><span class="n">var_names</span> <span class="o">=</span> <span class="p">[</span><span class="sa">f</span><span class="s2">&quot;Gene_</span><span class="si">{</span><span class="n">i</span><span class="si">:</span><span class="s2">d</span><span class="si">}</span><span class="s2">&quot;</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">adata</span><span class="o">.</span><span class="n">n_vars</span><span class="p">)]</span>
<span class="nb">print</span><span class="p">(</span><span class="n">adata</span><span class="o">.</span><span class="n">obs_names</span><span class="p">[:</span><span class="mi">10</span><span class="p">])</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Index([&#39;Cell_0&#39;, &#39;Cell_1&#39;, &#39;Cell_2&#39;, &#39;Cell_3&#39;, &#39;Cell_4&#39;, &#39;Cell_5&#39;, &#39;Cell_6&#39;,
       &#39;Cell_7&#39;, &#39;Cell_8&#39;, &#39;Cell_9&#39;],
      dtype=&#39;object&#39;)
</pre></div>
</div>
</div>
</div>
</section>
<section id="adding-aligned-metadata">
<h3><span class="section-number">4.2.3. </span>Adding aligned metadata<a class="headerlink" href="#adding-aligned-metadata" title="Link to this heading">#</a></h3>
<section id="observational-or-variable-level">
<h4><span class="section-number">4.2.3.1. </span>Observational or Variable level<a class="headerlink" href="#observational-or-variable-level" title="Link to this heading">#</a></h4>
<p>The core of our AnnData object is now in place. As a next step we add metadata at both, the observational and variable levels. Remember, we store such annotations in the <code class="docutils literal notranslate"><span class="pre">.obs</span></code> and <code class="docutils literal notranslate"><span class="pre">.var</span></code> slots of the AnnData object for cell and gene annotations respectively.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">ct</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">default_rng</span><span class="p">()</span><span class="o">.</span><span class="n">choice</span><span class="p">([</span><span class="s2">&quot;B&quot;</span><span class="p">,</span> <span class="s2">&quot;T&quot;</span><span class="p">,</span> <span class="s2">&quot;Monocyte&quot;</span><span class="p">],</span> <span class="n">size</span><span class="o">=</span><span class="p">(</span><span class="n">adata</span><span class="o">.</span><span class="n">n_obs</span><span class="p">,))</span>
<span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s2">&quot;cell_type&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">Categorical</span><span class="p">(</span><span class="n">ct</span><span class="p">)</span>  <span class="c1"># Categoricals are preferred for efficiency</span>
<span class="n">adata</span><span class="o">.</span><span class="n">obs</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>cell_type</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>Cell_0</th>
      <td>B</td>
    </tr>
    <tr>
      <th>Cell_1</th>
      <td>T</td>
    </tr>
    <tr>
      <th>Cell_2</th>
      <td>B</td>
    </tr>
    <tr>
      <th>Cell_3</th>
      <td>T</td>
    </tr>
    <tr>
      <th>Cell_4</th>
      <td>Monocyte</td>
    </tr>
    <tr>
      <th>...</th>
      <td>...</td>
    </tr>
    <tr>
      <th>Cell_95</th>
      <td>T</td>
    </tr>
    <tr>
      <th>Cell_96</th>
      <td>T</td>
    </tr>
    <tr>
      <th>Cell_97</th>
      <td>B</td>
    </tr>
    <tr>
      <th>Cell_98</th>
      <td>T</td>
    </tr>
    <tr>
      <th>Cell_99</th>
      <td>Monocyte</td>
    </tr>
  </tbody>
</table>
<p>100 rows × 1 columns</p>
</div></div></div>
</div>
<p>If we examine the representation of the AnnData object again now, we will notice that it got updated as well with the <code class="docutils literal notranslate"><span class="pre">cell_type</span></code> information in <code class="docutils literal notranslate"><span class="pre">obs</span></code>.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">adata</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>AnnData object with n_obs × n_vars = 100 × 2000
    obs: &#39;cell_type&#39;
</pre></div>
</div>
</div>
</div>
</section>
<section id="subsetting-using-metadata">
<h4><span class="section-number">4.2.3.2. </span>Subsetting using metadata<a class="headerlink" href="#subsetting-using-metadata" title="Link to this heading">#</a></h4>
<p>We can also subset the AnnData object with the randomly generated cell types. The slicing and masking of the AnnData object behaves similarly to the access of data in Pandas DataFrames or R matrices. More details on this can be found below.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">bdata</span> <span class="o">=</span> <span class="n">adata</span><span class="p">[</span><span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="o">.</span><span class="n">cell_type</span> <span class="o">==</span> <span class="s2">&quot;B&quot;</span><span class="p">]</span>
<span class="n">bdata</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>View of AnnData object with n_obs × n_vars = 26 × 2000
    obs: &#39;cell_type&#39;
</pre></div>
</div>
</div>
</div>
</section>
</section>
<section id="observation-variable-level-matrices">
<h3><span class="section-number">4.2.4. </span>Observation/variable-level matrices<a class="headerlink" href="#observation-variable-level-matrices" title="Link to this heading">#</a></h3>
<p>We might also have metadata at either level that has many dimensions to it, such as a UMAP embedding of the data. For this type of metadata, AnnData has the <code class="docutils literal notranslate"><span class="pre">.obsm/.varm</span></code> attributes. We use keys to identify the different matrices we insert. The restriction of <code class="docutils literal notranslate"><span class="pre">.obsm/.varm</span></code> are that <code class="docutils literal notranslate"><span class="pre">.obsm</span></code> matrices must have a length equal to the number of observations as <code class="docutils literal notranslate"><span class="pre">.n_obs</span></code> and <code class="docutils literal notranslate"><span class="pre">.varm</span></code> matrices must have a length equal to <code class="docutils literal notranslate"><span class="pre">.n_vars</span></code>. They can each independently have different number of dimensions.</p>
<p>Let us start with a randomly generated matrix that we can interpret as a UMAP embedding of the data we would like to store, as well as some random gene-level metadata.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">adata</span><span class="o">.</span><span class="n">obsm</span><span class="p">[</span><span class="s2">&quot;X_umap&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">default_generator</span><span class="p">()</span><span class="o">.</span><span class="n">normal</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="n">size</span><span class="o">=</span><span class="p">(</span><span class="n">adata</span><span class="o">.</span><span class="n">n_obs</span><span class="p">,</span> <span class="mi">2</span><span class="p">))</span>
<span class="n">adata</span><span class="o">.</span><span class="n">varm</span><span class="p">[</span><span class="s2">&quot;gene_stuff&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">default_generator</span><span class="p">()</span><span class="o">.</span><span class="n">normal</span><span class="p">(</span>
    <span class="mi">0</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="n">size</span><span class="o">=</span><span class="p">(</span><span class="n">adata</span><span class="o">.</span><span class="n">n_vars</span><span class="p">,</span> <span class="mi">5</span><span class="p">)</span>
<span class="p">)</span>
<span class="n">adata</span><span class="o">.</span><span class="n">obsm</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>AxisArrays with keys: X_umap
</pre></div>
</div>
</div>
</div>
<p>Again, the AnnData representation is updated.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">adata</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>AnnData object with n_obs × n_vars = 100 × 2000
    obs: &#39;cell_type&#39;
    obsm: &#39;X_umap&#39;
    varm: &#39;gene_stuff&#39;
</pre></div>
</div>
</div>
</div>
<p>A few more notes about <code class="docutils literal notranslate"><span class="pre">.obsm/.varm</span></code>:</p>
<ol class="arabic simple">
<li><p>The “array-like” metadata can originate from a Pandas DataFrame, scipy sparse matrix, or numpy dense array.</p></li>
<li><p>When using scanpy, their values (columns) are not easily plotted, where instead items from <code class="docutils literal notranslate"><span class="pre">.obs</span></code> are easily plotted on, e.g., UMAP plots.</p></li>
</ol>
</section>
<section id="unstructured-metadata">
<h3><span class="section-number">4.2.5. </span>Unstructured metadata<a class="headerlink" href="#unstructured-metadata" title="Link to this heading">#</a></h3>
<p>As mentioned above, AnnData has <code class="docutils literal notranslate"><span class="pre">.uns</span></code>, which allows for any unstructured metadata. This can be anything, like a list or a dictionary with some general information that was useful in the analysis of our data. Try to only use this slot for data which cannot be efficiently stored in the other slots.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">adata</span><span class="o">.</span><span class="n">uns</span><span class="p">[</span><span class="s2">&quot;random&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="p">[</span><span class="mi">1</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="mi">3</span><span class="p">]</span>
<span class="n">adata</span><span class="o">.</span><span class="n">uns</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>OverloadedDict, wrapping:
	OrderedDict([(&#39;random&#39;, [1, 2, 3])])
With overloaded keys:
	[&#39;neighbors&#39;].
</pre></div>
</div>
</div>
</div>
</section>
<section id="layers">
<h3><span class="section-number">4.2.6. </span>Layers<a class="headerlink" href="#layers" title="Link to this heading">#</a></h3>
<p>Finally, we may have different forms of our original core data, perhaps one that is normalized and one that is not. These can be stored in different layers in AnnData. For example, let us log transform the original data and store it in a layer.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">adata</span><span class="o">.</span><span class="n">layers</span><span class="p">[</span><span class="s2">&quot;log_transformed&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">log1p</span><span class="p">(</span><span class="n">adata</span><span class="o">.</span><span class="n">X</span><span class="p">)</span>
<span class="n">adata</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>AnnData object with n_obs × n_vars = 100 × 2000
    obs: &#39;cell_type&#39;
    uns: &#39;random&#39;
    obsm: &#39;X_umap&#39;
    varm: &#39;gene_stuff&#39;
    layers: &#39;log_transformed&#39;
</pre></div>
</div>
</div>
</div>
<p>Our original matrix <code class="docutils literal notranslate"><span class="pre">X</span></code> was not modified and is still accessible. We can verify this by comparing the original <code class="docutils literal notranslate"><span class="pre">X</span></code> to the new layer.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="p">(</span><span class="n">adata</span><span class="o">.</span><span class="n">X</span> <span class="o">!=</span> <span class="n">adata</span><span class="o">.</span><span class="n">layers</span><span class="p">[</span><span class="s2">&quot;log_transformed&quot;</span><span class="p">])</span><span class="o">.</span><span class="n">nnz</span> <span class="o">==</span> <span class="mi">0</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>False
</pre></div>
</div>
</div>
</div>
</section>
<section id="conversion-to-dataframes">
<h3><span class="section-number">4.2.7. </span>Conversion to DataFrames<a class="headerlink" href="#conversion-to-dataframes" title="Link to this heading">#</a></h3>
<p>It is possible to obtain a Pandas DataFrame from one of the layers.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">adata</span><span class="o">.</span><span class="n">to_df</span><span class="p">(</span><span class="n">layer</span><span class="o">=</span><span class="s2">&quot;log_transformed&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>Gene_0</th>
      <th>Gene_1</th>
      <th>Gene_2</th>
      <th>Gene_3</th>
      <th>Gene_4</th>
      <th>Gene_5</th>
      <th>Gene_6</th>
      <th>Gene_7</th>
      <th>Gene_8</th>
      <th>Gene_9</th>
      <th>...</th>
      <th>Gene_1990</th>
      <th>Gene_1991</th>
      <th>Gene_1992</th>
      <th>Gene_1993</th>
      <th>Gene_1994</th>
      <th>Gene_1995</th>
      <th>Gene_1996</th>
      <th>Gene_1997</th>
      <th>Gene_1998</th>
      <th>Gene_1999</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>Cell_0</th>
      <td>0.693147</td>
      <td>0.000000</td>
      <td>1.386294</td>
      <td>1.386294</td>
      <td>0.693147</td>
      <td>0.000000</td>
      <td>0.693147</td>
      <td>1.098612</td>
      <td>0.000000</td>
      <td>1.098612</td>
      <td>...</td>
      <td>0.693147</td>
      <td>0.693147</td>
      <td>0.000000</td>
      <td>0.693147</td>
      <td>1.098612</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>1.098612</td>
      <td>0.693147</td>
      <td>0.000000</td>
    </tr>
    <tr>
      <th>Cell_1</th>
      <td>1.098612</td>
      <td>1.098612</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.693147</td>
      <td>0.000000</td>
      <td>0.693147</td>
      <td>1.386294</td>
      <td>0.000000</td>
      <td>...</td>
      <td>1.098612</td>
      <td>1.386294</td>
      <td>1.098612</td>
      <td>0.000000</td>
      <td>0.693147</td>
      <td>0.693147</td>
      <td>1.386294</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.693147</td>
    </tr>
    <tr>
      <th>Cell_2</th>
      <td>0.693147</td>
      <td>0.693147</td>
      <td>1.098612</td>
      <td>0.693147</td>
      <td>1.386294</td>
      <td>1.386294</td>
      <td>0.693147</td>
      <td>0.693147</td>
      <td>0.000000</td>
      <td>1.098612</td>
      <td>...</td>
      <td>0.000000</td>
      <td>0.693147</td>
      <td>0.693147</td>
      <td>0.000000</td>
      <td>1.386294</td>
      <td>0.693147</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.693147</td>
      <td>0.693147</td>
    </tr>
    <tr>
      <th>Cell_3</th>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.693147</td>
      <td>0.693147</td>
      <td>0.000000</td>
      <td>1.386294</td>
      <td>1.098612</td>
      <td>0.693147</td>
      <td>0.693147</td>
      <td>...</td>
      <td>0.693147</td>
      <td>0.693147</td>
      <td>0.000000</td>
      <td>0.693147</td>
      <td>0.693147</td>
      <td>0.693147</td>
      <td>0.693147</td>
      <td>0.000000</td>
      <td>1.386294</td>
      <td>0.693147</td>
    </tr>
    <tr>
      <th>Cell_4</th>
      <td>0.000000</td>
      <td>0.693147</td>
      <td>0.000000</td>
      <td>1.609438</td>
      <td>1.098612</td>
      <td>0.693147</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.693147</td>
      <td>0.000000</td>
      <td>...</td>
      <td>1.098612</td>
      <td>1.098612</td>
      <td>0.000000</td>
      <td>1.386294</td>
      <td>1.791759</td>
      <td>1.098612</td>
      <td>1.609438</td>
      <td>0.693147</td>
      <td>0.000000</td>
      <td>0.693147</td>
    </tr>
    <tr>
      <th>...</th>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
    </tr>
    <tr>
      <th>Cell_95</th>
      <td>0.693147</td>
      <td>1.098612</td>
      <td>0.693147</td>
      <td>1.386294</td>
      <td>1.098612</td>
      <td>0.693147</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.693147</td>
      <td>...</td>
      <td>0.000000</td>
      <td>0.693147</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.693147</td>
      <td>0.000000</td>
      <td>0.693147</td>
      <td>1.386294</td>
      <td>0.693147</td>
    </tr>
    <tr>
      <th>Cell_96</th>
      <td>0.693147</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>1.098612</td>
      <td>0.000000</td>
      <td>0.693147</td>
      <td>0.693147</td>
      <td>1.098612</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>...</td>
      <td>1.386294</td>
      <td>0.693147</td>
      <td>1.098612</td>
      <td>0.693147</td>
      <td>0.000000</td>
      <td>0.693147</td>
      <td>1.098612</td>
      <td>1.386294</td>
      <td>0.693147</td>
      <td>0.000000</td>
    </tr>
    <tr>
      <th>Cell_97</th>
      <td>0.693147</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>1.098612</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.693147</td>
      <td>...</td>
      <td>0.000000</td>
      <td>0.693147</td>
      <td>1.386294</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.693147</td>
      <td>0.693147</td>
      <td>1.609438</td>
      <td>0.693147</td>
    </tr>
    <tr>
      <th>Cell_98</th>
      <td>0.693147</td>
      <td>0.693147</td>
      <td>0.000000</td>
      <td>1.098612</td>
      <td>0.000000</td>
      <td>0.693147</td>
      <td>0.693147</td>
      <td>1.098612</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>...</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.693147</td>
      <td>0.693147</td>
      <td>0.000000</td>
      <td>0.693147</td>
      <td>1.386294</td>
      <td>0.693147</td>
      <td>1.098612</td>
      <td>1.386294</td>
    </tr>
    <tr>
      <th>Cell_99</th>
      <td>1.098612</td>
      <td>1.791759</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>1.098612</td>
      <td>1.609438</td>
      <td>0.693147</td>
      <td>0.693147</td>
      <td>...</td>
      <td>0.693147</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.693147</td>
      <td>0.693147</td>
      <td>0.693147</td>
      <td>0.693147</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>1.386294</td>
    </tr>
  </tbody>
</table>
<p>100 rows × 2000 columns</p>
</div></div></div>
</div>
</section>
<section id="reading-and-writing-of-anndata-objects">
<h3><span class="section-number">4.2.8. </span>Reading and writing of AnnData objects<a class="headerlink" href="#reading-and-writing-of-anndata-objects" title="Link to this heading">#</a></h3>
<p>AnnData objects can be saved on disk to hierarchical array stores like <a class="reference external" href="https://en.wikipedia.org/wiki/Hierarchical_Data_Format">HDF5</a> or <a class="reference external" href="https://zarr.readthedocs.io/en/stable/index.html">Zarr</a> to enable similar structures in disk and on memory. AnnData comes with its own persistent HDF5-based file format: <code class="docutils literal notranslate"><span class="pre">h5ad</span></code>. If string columns with small number of categories are not yet categoricals, AnnData will auto-transform them to categoricals. We will now save our AnnData object in <code class="docutils literal notranslate"><span class="pre">h5ad</span></code> format.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">adata</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s2">&quot;my_results.h5ad&quot;</span><span class="p">,</span> <span class="n">compression</span><span class="o">=</span><span class="s2">&quot;gzip&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="o">!</span>h5ls<span class="w"> </span><span class="s1">&#39;my_results.h5ad&#39;</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>X                        Group
layers                   Group
obs                      Group
obsm                     Group
uns                      Group
var                      Group
varm                     Group
</pre></div>
</div>
</div>
</div>
<p>… and read it back in.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">adata_new</span> <span class="o">=</span> <span class="n">ad</span><span class="o">.</span><span class="n">read_h5ad</span><span class="p">(</span><span class="s2">&quot;my_results.h5ad&quot;</span><span class="p">)</span>
<span class="n">adata_new</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>AnnData object with n_obs × n_vars = 100 × 2000
    obs: &#39;cell_type&#39;
    uns: &#39;random&#39;
    obsm: &#39;X_umap&#39;
    varm: &#39;gene_stuff&#39;
    layers: &#39;log_transformed&#39;
</pre></div>
</div>
</div>
</div>
</section>
<section id="efficient-data-access">
<h3><span class="section-number">4.2.9. </span>Efficient data access<a class="headerlink" href="#efficient-data-access" title="Link to this heading">#</a></h3>
<section id="view-and-copies">
<h4><span class="section-number">4.2.9.1. </span>View and copies<a class="headerlink" href="#view-and-copies" title="Link to this heading">#</a></h4>
<p>For the fun of it, let us look at another metadata use case. Imagine that the observations come from instruments characterizing 10 readouts in a multi-year study with samples taken from different subjects at different sites. We would typically get that information in some format and then store it in a DataFrame:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">obs_meta</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span>
    <span class="p">{</span>
        <span class="s2">&quot;time_yr&quot;</span><span class="p">:</span> <span class="n">np</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">default_rng</span><span class="p">()</span><span class="o">.</span><span class="n">choice</span><span class="p">([</span><span class="mi">0</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span> <span class="mi">8</span><span class="p">],</span> <span class="n">adata</span><span class="o">.</span><span class="n">n_obs</span><span class="p">),</span>
        <span class="s2">&quot;subject_id&quot;</span><span class="p">:</span> <span class="n">np</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">default_rng</span><span class="p">()</span><span class="o">.</span><span class="n">choice</span><span class="p">(</span>
            <span class="p">[</span><span class="s2">&quot;subject 1&quot;</span><span class="p">,</span> <span class="s2">&quot;subject 2&quot;</span><span class="p">,</span> <span class="s2">&quot;subject 4&quot;</span><span class="p">,</span> <span class="s2">&quot;subject 8&quot;</span><span class="p">],</span> <span class="n">adata</span><span class="o">.</span><span class="n">n_obs</span>
        <span class="p">),</span>
        <span class="s2">&quot;instrument_type&quot;</span><span class="p">:</span> <span class="n">np</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">default_rng</span><span class="p">()</span><span class="o">.</span><span class="n">choice</span><span class="p">(</span>
            <span class="p">[</span><span class="s2">&quot;type a&quot;</span><span class="p">,</span> <span class="s2">&quot;type b&quot;</span><span class="p">],</span> <span class="n">adata</span><span class="o">.</span><span class="n">n_obs</span>
        <span class="p">),</span>
        <span class="s2">&quot;site&quot;</span><span class="p">:</span> <span class="n">np</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">default_rng</span><span class="p">()</span><span class="o">.</span><span class="n">choice</span><span class="p">([</span><span class="s2">&quot;site x&quot;</span><span class="p">,</span> <span class="s2">&quot;site y&quot;</span><span class="p">],</span> <span class="n">adata</span><span class="o">.</span><span class="n">n_obs</span><span class="p">),</span>
    <span class="p">},</span>
    <span class="n">index</span><span class="o">=</span><span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="o">.</span><span class="n">index</span><span class="p">,</span>  <span class="c1"># these are the same IDs of observations as above!</span>
<span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<p>This is how we join the readout data with the metadata. Of course, the first argument of the following call for <em>X</em> could also just be a DataFrame. This will result in a single data container which tracks everything.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">adata</span> <span class="o">=</span> <span class="n">ad</span><span class="o">.</span><span class="n">AnnData</span><span class="p">(</span><span class="n">adata</span><span class="o">.</span><span class="n">X</span><span class="p">,</span> <span class="n">obs</span><span class="o">=</span><span class="n">obs_meta</span><span class="p">,</span> <span class="n">var</span><span class="o">=</span><span class="n">adata</span><span class="o">.</span><span class="n">var</span><span class="p">)</span>
<span class="n">adata</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>AnnData object with n_obs × n_vars = 100 × 2000
    obs: &#39;time_yr&#39;, &#39;subject_id&#39;, &#39;instrument_type&#39;, &#39;site&#39;
</pre></div>
</div>
</div>
</div>
<p>Subsetting the joint data matrix can be important to focus on subsets of variables or observations, or to define train-test splits for a machine learning model.</p>
<p><strong>Note</strong> Similar to numpy arrays, AnnData objects can either hold actual data or reference another <code class="docutils literal notranslate"><span class="pre">AnnData</span></code> object. In the later case, they are referred to as “view”. Subsetting AnnData objects always returns views, which has two advantages:</p>
<ul class="simple">
<li><p>no new memory is allocated</p></li>
<li><p>it is possible to modify the underlying AnnData object.</p></li>
</ul>
<p>You can get an actual AnnData object from a view by calling <code class="docutils literal notranslate"><span class="pre">.copy()</span></code> on the view. Usually, this is not necessary, as any modification of elements of a view (calling <code class="docutils literal notranslate"><span class="pre">.[]</span></code> on an attribute of the view) internally calls <code class="docutils literal notranslate"><span class="pre">.copy()</span></code> and makes the view an AnnData object that holds actual data. See the example below.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">adata</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>AnnData object with n_obs × n_vars = 100 × 2000
    obs: &#39;time_yr&#39;, &#39;subject_id&#39;, &#39;instrument_type&#39;, &#39;site&#39;
</pre></div>
</div>
</div>
</div>
<p><strong>Note</strong> Indexing into AnnData will assume that integer arguments to <code class="docutils literal notranslate"><span class="pre">[]</span></code> behave like <code class="docutils literal notranslate"><span class="pre">.iloc</span></code> in pandas, whereas string arguments behave like <code class="docutils literal notranslate"><span class="pre">.loc</span></code>. <code class="docutils literal notranslate"><span class="pre">AnnData</span></code> always assumes string indices.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">adata_view</span> <span class="o">=</span> <span class="n">adata</span><span class="p">[:</span><span class="mi">5</span><span class="p">,</span> <span class="p">[</span><span class="s2">&quot;Gene_1&quot;</span><span class="p">,</span> <span class="s2">&quot;Gene_3&quot;</span><span class="p">]]</span>
<span class="n">adata_view</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>View of AnnData object with n_obs × n_vars = 5 × 2
    obs: &#39;time_yr&#39;, &#39;subject_id&#39;, &#39;instrument_type&#39;, &#39;site&#39;
</pre></div>
</div>
</div>
</div>
<p>This is a view! This can be verified by examining the AnnData object again.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">adata</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>AnnData object with n_obs × n_vars = 100 × 2000
    obs: &#39;time_yr&#39;, &#39;subject_id&#39;, &#39;instrument_type&#39;, &#39;site&#39;
</pre></div>
</div>
</div>
</div>
<p>The dimensions of the AnnData object have not changed it still contains the same data. If we want an AnnData that holds the data in memory we have to call <code class="docutils literal notranslate"><span class="pre">.copy()</span></code>.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">adata_subset</span> <span class="o">=</span> <span class="n">adata</span><span class="p">[:</span><span class="mi">5</span><span class="p">,</span> <span class="p">[</span><span class="s2">&quot;Gene_1&quot;</span><span class="p">,</span> <span class="s2">&quot;Gene_3&quot;</span><span class="p">]]</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
</pre></div>
</div>
</div>
</div>
<p>For a view, we can also set the first 3 elements of a column.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="nb">print</span><span class="p">(</span><span class="n">adata</span><span class="p">[:</span><span class="mi">3</span><span class="p">,</span> <span class="s2">&quot;Gene_1&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">X</span><span class="o">.</span><span class="n">toarray</span><span class="p">()</span><span class="o">.</span><span class="n">tolist</span><span class="p">())</span>
<span class="n">adata</span><span class="p">[:</span><span class="mi">3</span><span class="p">,</span> <span class="s2">&quot;Gene_1&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">X</span> <span class="o">=</span> <span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="mi">0</span><span class="p">]</span>
<span class="nb">print</span><span class="p">(</span><span class="n">adata</span><span class="p">[:</span><span class="mi">3</span><span class="p">,</span> <span class="s2">&quot;Gene_1&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">X</span><span class="o">.</span><span class="n">toarray</span><span class="p">()</span><span class="o">.</span><span class="n">tolist</span><span class="p">())</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>[[0.0], [2.0], [1.0]]
[[0.0], [0.0], [0.0]]
</pre></div>
</div>
</div>
</div>
<p>If you try to access parts of a view of an AnnData, the content will be auto-copied and a data-storing object will be generated.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">adata_subset</span> <span class="o">=</span> <span class="n">adata</span><span class="p">[:</span><span class="mi">3</span><span class="p">,</span> <span class="p">[</span><span class="s2">&quot;Gene_1&quot;</span><span class="p">,</span> <span class="s2">&quot;Gene_2&quot;</span><span class="p">]]</span>
<span class="n">adata_subset</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>View of AnnData object with n_obs × n_vars = 3 × 2
    obs: &#39;time_yr&#39;, &#39;subject_id&#39;, &#39;instrument_type&#39;, &#39;site&#39;
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">adata_subset</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s2">&quot;foo&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="nb">range</span><span class="p">(</span><span class="mi">3</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Trying to set attribute `.obs` of view, copying.
</pre></div>
</div>
</div>
</div>
<p>Now <code class="docutils literal notranslate"><span class="pre">adata_subset</span></code> stores the actual data and is no longer just a reference to adata.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">adata_subset</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>AnnData object with n_obs × n_vars = 3 × 2
    obs: &#39;time_yr&#39;, &#39;subject_id&#39;, &#39;instrument_type&#39;, &#39;site&#39;, &#39;foo&#39;
</pre></div>
</div>
</div>
</div>
<p>Evidently, you can use all of pandas to slice with sequences or boolean indices.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">adata</span><span class="p">[</span><span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="o">.</span><span class="n">time_yr</span><span class="o">.</span><span class="n">isin</span><span class="p">([</span><span class="mi">2</span><span class="p">,</span> <span class="mi">4</span><span class="p">])]</span><span class="o">.</span><span class="n">obs</span><span class="o">.</span><span class="n">head</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>time_yr</th>
      <th>subject_id</th>
      <th>instrument_type</th>
      <th>site</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>Cell_0</th>
      <td>4</td>
      <td>subject 8</td>
      <td>type b</td>
      <td>site x</td>
    </tr>
    <tr>
      <th>Cell_2</th>
      <td>2</td>
      <td>subject 2</td>
      <td>type b</td>
      <td>site y</td>
    </tr>
    <tr>
      <th>Cell_3</th>
      <td>4</td>
      <td>subject 8</td>
      <td>type a</td>
      <td>site y</td>
    </tr>
    <tr>
      <th>Cell_5</th>
      <td>2</td>
      <td>subject 2</td>
      <td>type a</td>
      <td>site y</td>
    </tr>
    <tr>
      <th>Cell_10</th>
      <td>2</td>
      <td>subject 8</td>
      <td>type b</td>
      <td>site x</td>
    </tr>
  </tbody>
</table>
</div></div></div>
</div>
</section>
<section id="partial-reading-of-large-data">
<h4><span class="section-number">4.2.9.2. </span>Partial reading of large data<a class="headerlink" href="#partial-reading-of-large-data" title="Link to this heading">#</a></h4>
<p>If a single <code class="docutils literal notranslate"><span class="pre">h5ad</span></code> file is very large, you can partially read it into memory by using backed mode or with the currently experimental <a class="reference external" href="https://anndata.readthedocs.io/en/latest/generated/anndata.experimental.read_elem.html">read_elem</a> API.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">adata</span> <span class="o">=</span> <span class="n">ad</span><span class="o">.</span><span class="n">read</span><span class="p">(</span><span class="s2">&quot;my_results.h5ad&quot;</span><span class="p">,</span> <span class="n">backed</span><span class="o">=</span><span class="s2">&quot;r&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">adata</span><span class="o">.</span><span class="n">isbacked</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>True
</pre></div>
</div>
</div>
</div>
<p>If you do this, you will need to remember that the AnnData object has an open connection to the file used for reading.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">adata</span><span class="o">.</span><span class="n">filename</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>PosixPath(&#39;my_results.h5ad&#39;)
</pre></div>
</div>
</div>
</div>
<p>As we are using it in read-only mode, we cannot damage anything. To proceed with this tutorial, we still need to explicitly close it.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">adata</span><span class="o">.</span><span class="n">file</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>
</pre></div>
</div>
</div>
</div>
</section>
</section>
</section>
<section id="unimodal-data-analysis-with-scanpy">
<h2><span class="section-number">4.3. </span>Unimodal data analysis with scanpy<a class="headerlink" href="#unimodal-data-analysis-with-scanpy" title="Link to this heading">#</a></h2>
<p>Now that we understood the fundamental data structure of unimodal single-cell analysis, the question remains how we can actually analyze the stored data. In the scverse ecosystem several tools for the analysis of specific omics data exist. For example, scanpy<span id="id6">[<a class="reference internal" href="#id136" title="F. Alexander Wolf, Philipp Angerer, and Fabian J. Theis. Scanpy: large-scale single-cell gene expression data analysis. Genome Biology, 19(1):15, Feb 2018. URL: https://doi.org/10.1186/s13059-017-1382-0, doi:10.1186/s13059-017-1382-0.">Wolf <em>et al.</em>, 2018</a>]</span> provides tooling for general RNA-Seq focused analysis, squidpy<span id="id7">[<a class="reference internal" href="../spatial/spatially_variable_genes.html#id126" title="Giovanni Palla, Hannah Spitzer, Michal Klein, David Fischer, Anna Christina Schaar, Louis Benedikt Kuemmerle, Sergei Rybakov, Ignacio L. Ibarra, Olle Holmberg, Isaac Virshup, Mohammad Lotfollahi, Sabrina Richter, and Fabian J. Theis. Squidpy: a scalable framework for spatial omics analysis. Nature Methods, 19(2):171-178, Feb 2022. URL: https://doi.org/10.1038/s41592-021-01358-2, doi:10.1038/s41592-021-01358-2.">Palla <em>et al.</em>, 2022</a>]</span> focuses on spatial transcriptomics and scirpy<span id="id8">[<a class="reference internal" href="#id138" title="Gregor Sturm, Tamas Szabo, Georgios Fotakis, Marlene Haider, Dietmar Rieder, Zlatko Trajanoski, and Francesca Finotello. Scirpy: a Scanpy extension for analyzing single-cell T-cell receptor-sequencing data. Bioinformatics, 36(18):4817-4818, 07 2020. URL: https://doi.org/10.1093/bioinformatics/btaa611, arXiv:https://academic.oup.com/bioinformatics/article-pdf/36/18/4817/34560298/btaa611.pdf, doi:10.1093/bioinformatics/btaa611.">Sturm <em>et al.</em>, 2020</a>]</span> provides tooling for the analysis of T-cell receptor (TCR) and B-cell receptor (BCR) data. Even though many scverse extensions for various data modalities exist, they usually use some of scanpy’s preprocessing and visualization capabilities to some extent.</p>
<p>More specifically, scanpy is a Python package which builds on top of AnnData to facilitate the analysis of single-cell gene expression data. Several methods for preprocessing, embedding, visualization, clustering, differential gene expression testing, pseudotime and trajectory inference, and simulation of gene regulatory networks are accessible through scanpy. The efficient implementation based on the Python data science and machine learning libraries allows scanpy to scale to millions of cells.<br />
Generally, best-practice single-cell data analysis is an interactive process. Many of the decisions and analysis steps depend on the results of previous steps and potentially input of experimental partners. Pipelines such as scflow<span id="id9">[<a class="reference internal" href="#id141" title="Combiz Khozoie, Nurun Fancy, Mahdi M. Marjaneh, Alan E. Murphy, Paul M. Matthews, and Nathan Skene. Scflow: a scalable and reproducible analysis pipeline for single-cell term`rna` sequencing data. bioRxiv, 2021. URL: https://www.biorxiv.org/content/early/2021/08/19/2021.08.16.456499.1, arXiv:https://www.biorxiv.org/content/early/2021/08/19/2021.08.16.456499.1.full.pdf, doi:10.1101/2021.08.16.456499.">Khozoie <em>et al.</em>, 2021</a>]</span>, which fully automate some downstream analysis steps, started to appear only recently. These pipelines have to make assumptions and simplifications which may not result in the most robust analysis.
scanpy is therefore designed to be used for interactive analyses with for example Jupyter Notebooks<span id="id10">[<a class="reference internal" href="#id142" title="Project Jupyter. Jupyter. https://jupyter.org/, 2022. Accessed: 2022-04-21.">Jupyter, 2022</a>]</span>.</p>
<figure class="align-default" id="scanpy-fig">
<a class="bg-primary mb-1 reference internal image-reference" href="../_images/scanpy.jpg"><img alt="scanpy Overview" class="bg-primary mb-1" src="../_images/scanpy.jpg" style="width: 800px;" />
</a>
<figcaption>
<p><span class="caption-number">Fig. 4.2 </span><span class="caption-text">scanpy overview. Image obtained from <span id="id11">[<a class="reference internal" href="#id136" title="F. Alexander Wolf, Philipp Angerer, and Fabian J. Theis. Scanpy: large-scale single-cell gene expression data analysis. Genome Biology, 19(1):15, Feb 2018. URL: https://doi.org/10.1186/s13059-017-1382-0, doi:10.1186/s13059-017-1382-0.">Wolf <em>et al.</em>, 2018</a>]</span>.</span><a class="headerlink" href="#scanpy-fig" title="Link to this image">#</a></p>
</figcaption>
</figure>
<section id="id12">
<h3><span class="section-number">4.3.1. </span>Installation<a class="headerlink" href="#id12" title="Link to this heading">#</a></h3>
<p>scanpy is available on PyPI and Conda can be installed with using either:</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>pip<span class="w"> </span>install<span class="w"> </span>scanpy
conda<span class="w"> </span>install<span class="w"> </span>-c<span class="w"> </span>conda-forge<span class="w"> </span>scanpy
</pre></div>
</div>
</section>
<section id="scanpy-api-design">
<h3><span class="section-number">4.3.2. </span>Scanpy API design<a class="headerlink" href="#scanpy-api-design" title="Link to this heading">#</a></h3>
<p>The scanpy framework is designed in a way that functions belonging to the same step are grouped in corresponding modules. For example, all preprocessing functions are available in the <code class="docutils literal notranslate"><span class="pre">scanpy.preprocessing</span></code> module, all transformation of a data matrix which are not preprocessing are available in <code class="docutils literal notranslate"><span class="pre">scanpy.tools</span></code> and all visualizations are available in <code class="docutils literal notranslate"><span class="pre">scanpy.plot</span></code>. These modules are commonly accessed after having imported scanpy like <code class="docutils literal notranslate"><span class="pre">import</span> <span class="pre">scanpy</span> <span class="pre">as</span> <span class="pre">sc</span></code> with the corresponding abbreviations <code class="docutils literal notranslate"><span class="pre">sc.pp</span></code> for preprocessing, <code class="docutils literal notranslate"><span class="pre">sc.tl</span></code> for tools and <code class="docutils literal notranslate"><span class="pre">sc.pl</span></code> for plots. All modules which read or write data are directly accessed. Further, a module for various datasets is available as <code class="docutils literal notranslate"><span class="pre">sc.datasets</span></code>.
All functions with their corresponding parameters and potentially example plots are documented in the scanpy API documentation<span id="id13">[<a class="reference internal" href="#id143" title="scverse scanpy. Scanpy api. https://scanpy.readthedocs.io/en/stable/api.html#, 2022. Accessed: 2022-04-21.">scverse scanpy, 2022</a>]</span>.</p>
<p>Note that this tutorial only covers a very small subset of scanpy’s features and options. Readers are strongly encouraged to examine <a class="reference external" href="https://scanpy.readthedocs.io/">scanpy’s documentation</a> for more details.</p>
<figure class="align-default" id="scanpy-api">
<a class="bg-primary mb-1 reference internal image-reference" href="../_images/scanpy_api.png"><img alt="scanpy API" class="bg-primary mb-1" src="../_images/scanpy_api.png" style="width: 800px;" />
</a>
<figcaption>
<p><span class="caption-number">Fig. 4.3 </span><span class="caption-text">scanpy API overview. The API is divided into datasets, preprocessing (pp), tools (tl) and corresponding plotting (pl) functions.</span><a class="headerlink" href="#scanpy-api" title="Link to this image">#</a></p>
</figcaption>
</figure>
</section>
<section id="scanpy-example">
<h3><span class="section-number">4.3.3. </span>Scanpy example<a class="headerlink" href="#scanpy-example" title="Link to this heading">#</a></h3>
<p>In the following cells we will shortly demonstrate the workflow of an analysis with scanpy. We explicitly do not conduct a full analysis because the specific analysis steps are covered in the corresponding chapters.</p>
<p>The dataset of choice is a dataset of 2700 peripheral blood mononuclear cells of a healthy donor which were sequenced on the Illumina NextSeq 500.</p>
<p>As a first step we import scanpy and define defaults for our following quick scanpy demo. We use scanpy’s setting object to  set the Matplotlib plotting defaults for all of scanpy’s plots and finally print scanpy’s header. This header contains the versions of all relevant Python packages in the current environment including Scanpy and AnnData. This output is especially useful when reporting bugs to the scverse team and for reproducibility reasons.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span><span class="w"> </span><span class="nn">scanpy</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">sc</span>

<span class="n">sc</span><span class="o">.</span><span class="n">settings</span><span class="o">.</span><span class="n">set_figure_params</span><span class="p">(</span><span class="n">dpi</span><span class="o">=</span><span class="mi">80</span><span class="p">,</span> <span class="n">facecolor</span><span class="o">=</span><span class="s2">&quot;white&quot;</span><span class="p">)</span>
<span class="n">sc</span><span class="o">.</span><span class="n">logging</span><span class="o">.</span><span class="n">print_header</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>scanpy==1.9.1 anndata==0.7.8 umap==0.5.2 numpy==1.20.3 scipy==1.7.2 pandas==1.4.3 scikit-learn==1.0.1 statsmodels==0.13.1 python-igraph==0.9.10 pynndescent==0.5.5
</pre></div>
</div>
</div>
</div>
<p>Next, we get the 3k PBMC dataset with scanpy as an AnnData object. The dataset is automatically downloaded and saved to a path which can be also specified with scanpy’s settings object. By default it is the folder <code class="docutils literal notranslate"><span class="pre">data</span></code>.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">adata</span> <span class="o">=</span> <span class="n">sc</span><span class="o">.</span><span class="n">datasets</span><span class="o">.</span><span class="n">pbmc3k</span><span class="p">()</span>
<span class="n">adata</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>100%|██████████| 5.58M/5.58M [00:01&lt;00:00, 3.25MB/s]
</pre></div>
</div>
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>AnnData object with n_obs × n_vars = 2700 × 32738
    var: &#39;gene_ids&#39;
</pre></div>
</div>
</div>
</div>
<p>The returned AnnData object has 2700 cells with 32738 genes. The <code class="docutils literal notranslate"><span class="pre">var</span></code> slot further contains the gene IDs.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">adata</span><span class="o">.</span><span class="n">var</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>gene_ids</th>
    </tr>
    <tr>
      <th>index</th>
      <th></th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>MIR1302-10</th>
      <td>ENSG00000243485</td>
    </tr>
    <tr>
      <th>FAM138A</th>
      <td>ENSG00000237613</td>
    </tr>
    <tr>
      <th>OR4F5</th>
      <td>ENSG00000186092</td>
    </tr>
    <tr>
      <th>RP11-34P13.7</th>
      <td>ENSG00000238009</td>
    </tr>
    <tr>
      <th>RP11-34P13.8</th>
      <td>ENSG00000239945</td>
    </tr>
    <tr>
      <th>...</th>
      <td>...</td>
    </tr>
    <tr>
      <th>AC145205.1</th>
      <td>ENSG00000215635</td>
    </tr>
    <tr>
      <th>BAGE5</th>
      <td>ENSG00000268590</td>
    </tr>
    <tr>
      <th>CU459201.1</th>
      <td>ENSG00000251180</td>
    </tr>
    <tr>
      <th>AC002321.2</th>
      <td>ENSG00000215616</td>
    </tr>
    <tr>
      <th>AC002321.1</th>
      <td>ENSG00000215611</td>
    </tr>
  </tbody>
</table>
<p>32738 rows × 1 columns</p>
</div></div></div>
</div>
<p>As mentioned above, all of scanpy’s analysis functions are accessible via <code class="docutils literal notranslate"><span class="pre">sc.[pp,</span> <span class="pre">tl,</span> <span class="pre">pl]</span></code>. As a first step to get an overview over our data, we use scanpy to show those genes that yield the highest fraction of counts in each single cell, across all cells. We simply call the <code class="docutils literal notranslate"><span class="pre">sc.pl.highest_expr_genes</span></code> function, pass the AnnData object which is in pretty much all cases the first parameter of any scanpy function, and specify that we want the top 20 expressed genes to be shown.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">sc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">highest_expr_genes</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">n_top</span><span class="o">=</span><span class="mi">20</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<a class="reference internal image-reference" href="../_images/c0fb293bb4fa88a2d4d6fb9899ed3a3529134f9475e96ba5e1f1f08dbbfd88b9.png"><img alt="../_images/c0fb293bb4fa88a2d4d6fb9899ed3a3529134f9475e96ba5e1f1f08dbbfd88b9.png" src="../_images/c0fb293bb4fa88a2d4d6fb9899ed3a3529134f9475e96ba5e1f1f08dbbfd88b9.png" style="width: 418px; height: 391px;" />
</a>
</div>
</div>
<p>Apparently, MALAT1 is the most expressed gene which is frequently detected in poly-A captured scRNA-Seq data, independent of protocol. This gene has been shown to have an inverse correlation with cell health. Especially dead/dying cells have a higher expression of MALAT1.</p>
<p>We now filter cells with less than 200 detected genes and genes which were found in less than 3 cells for a rough quality threshold with scanpy’s preprocessing module.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">sc</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">filter_cells</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">min_genes</span><span class="o">=</span><span class="mi">200</span><span class="p">)</span>
<span class="n">sc</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">filter_genes</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">min_cells</span><span class="o">=</span><span class="mi">3</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<p>A common step in single-cell RNA-Seq analysis is dimensionality reduction with for example principal component analysis (PCA) to unveil the main axes of variation. This also denoises the data. scanpy offers PCA as a <code class="docutils literal notranslate"><span class="pre">preprocessing</span></code> or <code class="docutils literal notranslate"><span class="pre">tools</span></code> function. These are equivalent. Here, we use the version in <code class="docutils literal notranslate"><span class="pre">tools</span></code> for no particular reason.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">sc</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">pca</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">svd_solver</span><span class="o">=</span><span class="s2">&quot;arpack&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<p>The corresponding plotting function allows us to pass genes to the color argument. The corresponding values are automatically extracted from the AnnData object.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">sc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">pca</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="s2">&quot;CST3&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<a class="reference internal image-reference" href="../_images/6a951bc7f5c43678cb8758b0bdd977f1299d3fb9c3071ff7a42a14a9804f406e.png"><img alt="../_images/6a951bc7f5c43678cb8758b0bdd977f1299d3fb9c3071ff7a42a14a9804f406e.png" src="../_images/6a951bc7f5c43678cb8758b0bdd977f1299d3fb9c3071ff7a42a14a9804f406e.png" style="width: 310px; height: 296px;" />
</a>
</div>
</div>
<p>A fundamental step for any advanced embedding and downstream calculations is the calculating of the neighborhood graph using the PCA representation of the data matrix. It is automatically used for other tools that require it such as the calculation of a UMAP.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">sc</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">neighbors</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">n_neighbors</span><span class="o">=</span><span class="mi">10</span><span class="p">,</span> <span class="n">n_pcs</span><span class="o">=</span><span class="mi">40</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<p>We now use the calculating neighborhood graph to embed the cells with a UMAP, one of many advanced dimension reduction <span class="xref std std-term">algorithms</span> implemented in scanpy.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">sc</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">umap</span><span class="p">(</span><span class="n">adata</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">sc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">umap</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;CST3&quot;</span><span class="p">,</span> <span class="s2">&quot;NKG7&quot;</span><span class="p">,</span> <span class="s2">&quot;PPBP&quot;</span><span class="p">])</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<a class="reference internal image-reference" href="../_images/90f4f980c81308de5a7a796b38f77c3083980eedeff91e0c3647136003104860.png"><img alt="../_images/90f4f980c81308de5a7a796b38f77c3083980eedeff91e0c3647136003104860.png" src="../_images/90f4f980c81308de5a7a796b38f77c3083980eedeff91e0c3647136003104860.png" style="width: 972px; height: 299px;" />
</a>
</div>
</div>
<p>scanpy’s documentation also provides <a class="reference external" href="https://scanpy.readthedocs.io/en/stable/tutorials.html">tutorials</a> which we recommend to all readers who need a refresher of scanpy or are new to scanpy. Video tutorials are available on the <a class="reference external" href="https://www.youtube.com/channel/UCpsvsIAW3R5OdftJKKuLNMA">scverse youtube channel</a>.</p>
</section>
</section>
<section id="advanced-using-mudata-to-store-multimodal-data">
<span id="at-mudata"></span><h2><span class="section-number">4.4. </span>Advanced: Using MuData to store multimodal data<a class="headerlink" href="#advanced-using-mudata-to-store-multimodal-data" title="Link to this heading">#</a></h2>
<p>AnnData is primarily designed for storing and manipulating unimodal data. However, multimodal assays such as CITE-Seq generate multimodal data by simultaneously measuring RNA and surface proteins. This data requires more advanced ways of storing, which is where MuData comes into play. MuData builds on top of AnnData to store and manipulate multimodal data. muon<span id="id14">[<a class="reference internal" href="#id146" title="Danila Bredikhin, Ilia Kats, and Oliver Stegle. Muon: multimodal omics analysis framework. Genome Biology, 23(1):42, Feb 2022. URL: https://doi.org/10.1186/s13059-021-02577-8, doi:10.1186/s13059-021-02577-8.">Bredikhin <em>et al.</em>, 2022</a>]</span>, a “scanpy equivalent” and core package of scverse, can then be used to analyze the multimodal omics data.
The following section is based on the MuData Quickstart tutorial<span id="id15">[<a class="reference internal" href="#id147" title="scverse mudata. Mudata quickstart. https://mudata.readthedocs.io/en/latest/notebooks/quickstart_mudata.html, 2022. Accessed: 2022-04-21.">scverse mudata, 2022</a>]</span> and the Multimodal data objects tutorial<span id="id16">[<a class="reference internal" href="#id148" title="scverse mudata. Mudata quickstart. https://mudata.readthedocs.io/en/latest/io/mudata.html, 2022. Accessed: 2022-04-21.">scverse mudata, 2022</a>]</span>.</p>
<figure class="align-default" id="mudata-fig">
<a class="bg-primary mb-1 reference internal image-reference" href="../_images/mudata.png"><img alt="MuData Overview" class="bg-primary mb-1" src="../_images/mudata.png" style="width: 800px;" />
</a>
<figcaption>
<p><span class="caption-number">Fig. 4.4 </span><span class="caption-text">MuData overview. Image obtained from <span id="id17">[<a class="reference internal" href="#id146" title="Danila Bredikhin, Ilia Kats, and Oliver Stegle. Muon: multimodal omics analysis framework. Genome Biology, 23(1):42, Feb 2022. URL: https://doi.org/10.1186/s13059-021-02577-8, doi:10.1186/s13059-021-02577-8.">Bredikhin <em>et al.</em>, 2022</a>]</span>.</span><a class="headerlink" href="#mudata-fig" title="Link to this image">#</a></p>
</figcaption>
</figure>
<section id="id18">
<h3><span class="section-number">4.4.1. </span>Installation<a class="headerlink" href="#id18" title="Link to this heading">#</a></h3>
<p>MuData is available on PyPI and Conda can be installed with using either:</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>pip<span class="w"> </span>install<span class="w"> </span>mudata
conda<span class="w"> </span>install<span class="w"> </span>-c<span class="w"> </span>conda-forge<span class="w"> </span>mudata
</pre></div>
</div>
<p>The main idea behind MuData is that the MuData object contains references to the single AnnData objects of the unimodal data, but the MuData object itself also stores multimodal annotations. It is therefore possible to directly access the AnnData objects to perform unimodal data transformations which store their results in the corresponding AnnData annotations, but also to aggregate the modalities for joint calculations whose results can be stored in the global MuData object. Technically, this is realized by MuData objects comprising a dictionary with AnnData objects, one per modality, in their <code class="docutils literal notranslate"><span class="pre">.mod</span></code> (=modality) attribute. Just as AnnData objects themselves, they also contain attributes like <code class="docutils literal notranslate"><span class="pre">.obs</span></code> with annotation of observations (samples or cells), <code class="docutils literal notranslate"><span class="pre">.obsm</span></code> with their multidimensional annotations such as embeddings.</p>
</section>
<section id="initializing-a-mudata-object">
<h3><span class="section-number">4.4.2. </span>Initializing a MuData object<a class="headerlink" href="#initializing-a-mudata-object" title="Link to this heading">#</a></h3>
<p>We will start by importing MuData from the mudata package.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span><span class="w"> </span><span class="nn">mudata</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">md</span>
</pre></div>
</div>
</div>
</div>
<p>To create an example MuData object we require simulated data.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">n</span><span class="p">,</span> <span class="n">d</span><span class="p">,</span> <span class="n">k</span> <span class="o">=</span> <span class="mi">1000</span><span class="p">,</span> <span class="mi">100</span><span class="p">,</span> <span class="mi">10</span>

<span class="n">z</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">default_generator</span><span class="p">()</span><span class="o">.</span><span class="n">normal</span><span class="p">(</span>
    <span class="n">loc</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">arange</span><span class="p">(</span><span class="n">k</span><span class="p">),</span> <span class="n">scale</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">arange</span><span class="p">(</span><span class="n">k</span><span class="p">)</span> <span class="o">*</span> <span class="mi">2</span><span class="p">,</span> <span class="n">size</span><span class="o">=</span><span class="p">(</span><span class="n">n</span><span class="p">,</span> <span class="n">k</span><span class="p">)</span>
<span class="p">)</span>
<span class="n">w</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">default_generator</span><span class="p">()</span><span class="o">.</span><span class="n">normal</span><span class="p">(</span><span class="n">size</span><span class="o">=</span><span class="p">(</span><span class="n">d</span><span class="p">,</span> <span class="n">k</span><span class="p">))</span>
<span class="n">y</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">dot</span><span class="p">(</span><span class="n">z</span><span class="p">,</span> <span class="n">w</span><span class="o">.</span><span class="n">T</span><span class="p">)</span>
<span class="n">y</span><span class="o">.</span><span class="n">shape</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>(1000, 100)
</pre></div>
</div>
</div>
</div>
<p>To create a MuData object we require the unimodal AnnData objects first. Therefore, we create two AnnData objects with <em>data for the same observations</em>, but for <em>different variables</em>.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">adata</span> <span class="o">=</span> <span class="n">ad</span><span class="o">.</span><span class="n">AnnData</span><span class="p">(</span><span class="n">y</span><span class="p">)</span>
<span class="n">adata</span><span class="o">.</span><span class="n">obs_names</span> <span class="o">=</span> <span class="p">[</span><span class="sa">f</span><span class="s2">&quot;obs_</span><span class="si">{</span><span class="n">i</span><span class="w"> </span><span class="o">+</span><span class="w"> </span><span class="mi">1</span><span class="si">}</span><span class="s2">&quot;</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">n</span><span class="p">)]</span>
<span class="n">adata</span><span class="o">.</span><span class="n">var_names</span> <span class="o">=</span> <span class="p">[</span><span class="sa">f</span><span class="s2">&quot;var_</span><span class="si">{</span><span class="n">j</span><span class="w"> </span><span class="o">+</span><span class="w"> </span><span class="mi">1</span><span class="si">}</span><span class="s2">&quot;</span> <span class="k">for</span> <span class="n">j</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">d</span><span class="p">)]</span>
<span class="n">adata</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>AnnData object with n_obs × n_vars = 1000 × 100
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">d2</span> <span class="o">=</span> <span class="mi">50</span>
<span class="n">w2</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">default_generator</span><span class="p">()</span><span class="o">.</span><span class="n">normal</span><span class="p">(</span><span class="n">size</span><span class="o">=</span><span class="p">(</span><span class="n">d2</span><span class="p">,</span> <span class="n">k</span><span class="p">))</span>
<span class="n">y2</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">dot</span><span class="p">(</span><span class="n">z</span><span class="p">,</span> <span class="n">w2</span><span class="o">.</span><span class="n">T</span><span class="p">)</span>

<span class="n">adata2</span> <span class="o">=</span> <span class="n">ad</span><span class="o">.</span><span class="n">AnnData</span><span class="p">(</span><span class="n">y2</span><span class="p">)</span>
<span class="n">adata2</span><span class="o">.</span><span class="n">obs_names</span> <span class="o">=</span> <span class="p">[</span><span class="sa">f</span><span class="s2">&quot;obs_</span><span class="si">{</span><span class="n">i</span><span class="w"> </span><span class="o">+</span><span class="w"> </span><span class="mi">1</span><span class="si">}</span><span class="s2">&quot;</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">n</span><span class="p">)]</span>
<span class="n">adata2</span><span class="o">.</span><span class="n">var_names</span> <span class="o">=</span> <span class="p">[</span><span class="sa">f</span><span class="s2">&quot;var2_</span><span class="si">{</span><span class="n">j</span><span class="w"> </span><span class="o">+</span><span class="w"> </span><span class="mi">1</span><span class="si">}</span><span class="s2">&quot;</span> <span class="k">for</span> <span class="n">j</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">d2</span><span class="p">)]</span>
<span class="n">adata2</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>AnnData object with n_obs × n_vars = 1000 × 50
</pre></div>
</div>
</div>
</div>
<p>These two AnnData objects (two “modalities”) can then be wrapped into a single MuData object. Here, we name modality one <code class="docutils literal notranslate"><span class="pre">A</span></code> and modality two <code class="docutils literal notranslate"><span class="pre">B</span></code>.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">mdata</span> <span class="o">=</span> <span class="n">md</span><span class="o">.</span><span class="n">MuData</span><span class="p">({</span><span class="s2">&quot;A&quot;</span><span class="p">:</span> <span class="n">adata</span><span class="p">,</span> <span class="s2">&quot;B&quot;</span><span class="p">:</span> <span class="n">adata2</span><span class="p">})</span>
<span class="n">mdata</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><pre>MuData object with n_obs × n_vars = 1000 × 150
  2 modalities
    A:	1000 x 100
    B:	1000 x 50</pre></div></div>
</div>
<p>Observations and variables of the MuData object are global, which means that observations with the identical name (<code class="docutils literal notranslate"><span class="pre">.obs_names</span></code>) in different modalities are considered to be the same observation. This also means variable names (<code class="docutils literal notranslate"><span class="pre">.var_names</span></code>) should be unique. This is reflected in the object description above: <code class="docutils literal notranslate"><span class="pre">mdata</span></code> has 1000 observations and 150 = 100+50 variables.</p>
</section>
<section id="mudata-attributes">
<h3><span class="section-number">4.4.3. </span>MuData attributes<a class="headerlink" href="#mudata-attributes" title="Link to this heading">#</a></h3>
<p>MuData objects consist of annotations as earlier described for AnnData objects like <code class="docutils literal notranslate"><span class="pre">.obs</span></code> or <code class="docutils literal notranslate"><span class="pre">.var</span></code>, but extend this behavior with <code class="docutils literal notranslate"><span class="pre">.mod</span></code> which serves as an accessor to the individual modalities.</p>
<p>Modalities are stored in a collection accessible via the <code class="docutils literal notranslate"><span class="pre">.mod</span></code> attribute of the MuData object with names of modalities as keys and AnnData objects as values.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="nb">list</span><span class="p">(</span><span class="n">mdata</span><span class="o">.</span><span class="n">mod</span><span class="o">.</span><span class="n">keys</span><span class="p">())</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>[&#39;A&#39;, &#39;B&#39;]
</pre></div>
</div>
</div>
</div>
<p>Individual modalities can be accessed with their names via the <code class="docutils literal notranslate"><span class="pre">.mod</span></code> attribute or via the MuData object itself as a shorthand.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="nb">print</span><span class="p">(</span><span class="n">mdata</span><span class="o">.</span><span class="n">mod</span><span class="p">[</span><span class="s2">&quot;A&quot;</span><span class="p">])</span>
<span class="nb">print</span><span class="p">(</span><span class="n">mdata</span><span class="p">[</span><span class="s2">&quot;A&quot;</span><span class="p">])</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>AnnData object with n_obs × n_vars = 1000 × 100
AnnData object with n_obs × n_vars = 1000 × 100
</pre></div>
</div>
</div>
</div>
<p>Samples (cells) annotation is accessible via the <code class="docutils literal notranslate"><span class="pre">.obs</span></code> attribute and by default includes copies of columns from <code class="docutils literal notranslate"><span class="pre">.obs</span></code> data frames of individual modalities. The same goes for <code class="docutils literal notranslate"><span class="pre">.var</span></code>, which contains annotation of variables (features). Observations columns copied from individual modalities contain modality name as their prefix, e.g. rna:n_genes. This is also true for variables columns. However if there are columns with identical names in <code class="docutils literal notranslate"><span class="pre">.var</span></code> of multiple modalities — e.g. n_cells, — these columns are merged across modalities and no prefix is added. When those slots are changed in AnnData objects of modalities, e.g. new columns are added or samples (cells) are filtered out, the changes have to be fetched with the <code class="docutils literal notranslate"><span class="pre">.update()</span></code> method (see below).</p>
<p>Multidimensional annotations of samples (cells) are accessible in the <code class="docutils literal notranslate"><span class="pre">.obsm</span></code> attribute. For instance, that can be UMAP coordinates that were learnt jointly on all modalities.</p>
<p>The MuData object’s shape is represented by two numbers calculated as a sum of the shapes of individual modalities — one for the number of observations and one for the number of variables.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="nb">print</span><span class="p">(</span><span class="n">mdata</span><span class="o">.</span><span class="n">shape</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="n">mdata</span><span class="o">.</span><span class="n">n_obs</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="n">mdata</span><span class="o">.</span><span class="n">n_vars</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>(1000, 150)
1000
150
</pre></div>
</div>
</div>
</div>
<p>By default, variables are always counted as belonging uniquely to a single modality while observations with the same name are counted as the same observation, which has variables across multiple modalities measured for.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="p">[</span><span class="n">adata</span><span class="o">.</span><span class="n">shape</span> <span class="k">for</span> <span class="n">adata</span> <span class="ow">in</span> <span class="n">mdata</span><span class="o">.</span><span class="n">mod</span><span class="o">.</span><span class="n">values</span><span class="p">()]</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>[(1000, 100), (1000, 50)]
</pre></div>
</div>
</div>
</div>
<p>If the shape of a modality is changed, <code class="docutils literal notranslate"><span class="pre">MuData.update()</span></code> has to be run to bring the respective updates to the MuData object.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">adata2</span><span class="o">.</span><span class="n">var_names</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;var_ad2_&quot;</span> <span class="o">+</span> <span class="n">e</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;_&quot;</span><span class="p">)[</span><span class="mi">1</span><span class="p">]</span> <span class="k">for</span> <span class="n">e</span> <span class="ow">in</span> <span class="n">adata2</span><span class="o">.</span><span class="n">var_names</span><span class="p">]</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Outdated variables names: ...,&quot;</span><span class="p">,</span> <span class="s2">&quot;, &quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">mdata</span><span class="o">.</span><span class="n">var_names</span><span class="p">[</span><span class="o">-</span><span class="mi">3</span><span class="p">:]))</span>
<span class="n">mdata</span><span class="o">.</span><span class="n">update</span><span class="p">()</span>
<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Updated variables names: ...,&quot;</span><span class="p">,</span> <span class="s2">&quot;, &quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">mdata</span><span class="o">.</span><span class="n">var_names</span><span class="p">[</span><span class="o">-</span><span class="mi">3</span><span class="p">:]))</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Outdated variables names: ..., var2_48, var2_49, var2_50
Updated variables names: ..., var_ad2_48, var_ad2_49, var_ad2_50
</pre></div>
</div>
</div>
</div>
<p>Importantly, individual modalities are stored as references to the original objects. Hence, if the original AnnData is changed the change will also be reflected in the MuData object.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Add some unstructured data to the original object</span>
<span class="n">adata</span><span class="o">.</span><span class="n">uns</span><span class="p">[</span><span class="s2">&quot;misc&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="p">{</span><span class="s2">&quot;adata&quot;</span><span class="p">:</span> <span class="kc">True</span><span class="p">}</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Access modality A via the .mod attribute</span>
<span class="n">mdata</span><span class="o">.</span><span class="n">mod</span><span class="p">[</span><span class="s2">&quot;A&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">uns</span><span class="p">[</span><span class="s2">&quot;misc&quot;</span><span class="p">]</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>{&#39;adata&#39;: True}
</pre></div>
</div>
</div>
</div>
</section>
<section id="variable-mappings">
<h3><span class="section-number">4.4.4. </span>Variable mappings<a class="headerlink" href="#variable-mappings" title="Link to this heading">#</a></h3>
<p>Upon construction of a MuData object, a global binary mapping between observations and individual modalities is created as well as between variables and modalities. Since all the observations are the same across modalities in <code class="docutils literal notranslate"><span class="pre">mdata</span></code>, all the values in the observations mappings are set to <code class="docutils literal notranslate"><span class="pre">True</span></code>.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">np</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">mdata</span><span class="o">.</span><span class="n">obsm</span><span class="p">[</span><span class="s2">&quot;A&quot;</span><span class="p">])</span> <span class="o">==</span> <span class="n">np</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">mdata</span><span class="o">.</span><span class="n">obsm</span><span class="p">[</span><span class="s2">&quot;B&quot;</span><span class="p">])</span> <span class="o">==</span> <span class="n">n</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>True
</pre></div>
</div>
</div>
</div>
<p>For variables however, those are 150-long vectors. The <code class="docutils literal notranslate"><span class="pre">A</span></code> modality has 100 <code class="docutils literal notranslate"><span class="pre">True</span></code> values followed by 50 <code class="docutils literal notranslate"><span class="pre">False</span></code> values.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">mdata</span><span class="o">.</span><span class="n">varm</span><span class="p">[</span><span class="s2">&quot;A&quot;</span><span class="p">]</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>array([ True,  True,  True,  True,  True,  True,  True,  True,  True,
        True,  True,  True,  True,  True,  True,  True,  True,  True,
        True,  True,  True,  True,  True,  True,  True,  True,  True,
        True,  True,  True,  True,  True,  True,  True,  True,  True,
        True,  True,  True,  True,  True,  True,  True,  True,  True,
        True,  True,  True,  True,  True,  True,  True,  True,  True,
        True,  True,  True,  True,  True,  True,  True,  True,  True,
        True,  True,  True,  True,  True,  True,  True,  True,  True,
        True,  True,  True,  True,  True,  True,  True,  True,  True,
        True,  True,  True,  True,  True,  True,  True,  True,  True,
        True,  True,  True,  True,  True,  True,  True,  True,  True,
        True, False, False, False, False, False, False, False, False,
       False, False, False, False, False, False, False, False, False,
       False, False, False, False, False, False, False, False, False,
       False, False, False, False, False, False, False, False, False,
       False, False, False, False, False, False, False, False, False,
       False, False, False, False, False, False])
</pre></div>
</div>
</div>
</div>
</section>
<section id="mudata-views">
<h3><span class="section-number">4.4.5. </span>MuData views<a class="headerlink" href="#mudata-views" title="Link to this heading">#</a></h3>
<p>Analogous to the behavior of AnnData objects, slicing MuData objects returns views of the original data.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">view</span> <span class="o">=</span> <span class="n">mdata</span><span class="p">[:</span><span class="mi">100</span><span class="p">,</span> <span class="p">:</span><span class="mi">1000</span><span class="p">]</span>
<span class="nb">print</span><span class="p">(</span><span class="n">view</span><span class="o">.</span><span class="n">is_view</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="n">view</span><span class="p">[</span><span class="s2">&quot;A&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">is_view</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>True
True
</pre></div>
</div>
</div>
</div>
<p>Subsetting MuData objects is special since it slices them across modalities. I.e. the slicing operation for a set of <code class="docutils literal notranslate"><span class="pre">obs_names</span></code> and/or <code class="docutils literal notranslate"><span class="pre">var_names</span></code> will be performed for each modality and not only for the global multimodal annotation. This behavior makes workflows memory-efficient, which is especially important when working with large datasets. If the object is to be modified however, a copy of it should be created, which is not a view anymore and has no dependence on the original object.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">mdata_sub</span> <span class="o">=</span> <span class="n">view</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
<span class="n">mdata_sub</span><span class="o">.</span><span class="n">is_view</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>False
</pre></div>
</div>
</div>
</div>
</section>
<section id="common-observations">
<h3><span class="section-number">4.4.6. </span>Common observations<a class="headerlink" href="#common-observations" title="Link to this heading">#</a></h3>
<p>While a MuData object is comprised of the same observations for both modalities, it is not always the case in the real world where some data might be missing. By design, MuData accounts for these scenarios since there’s no guarantee observations are the same — or even intersecting — for a MuData instance. It’s worth noting that other tools might provide convenience functions for some common scenarios of dealing with missing data, such as <code class="docutils literal notranslate"><span class="pre">intersect_obs()</span></code> implemented in muon.</p>
</section>
<section id="reading-and-writing-of-mudata-objects">
<h3><span class="section-number">4.4.7. </span>Reading and Writing of MuData objects<a class="headerlink" href="#reading-and-writing-of-mudata-objects" title="Link to this heading">#</a></h3>
<p>Similarly to AnnData objects, MuData objects were designed to be serialized into HDF5 based <code class="docutils literal notranslate"><span class="pre">.h5mu</span></code> files. All modalities are stored under their respective names in the <code class="docutils literal notranslate"><span class="pre">/mod</span></code> HDF5 group of the <code class="docutils literal notranslate"><span class="pre">.h5mu</span></code> file. Each individual modality, e.g. <code class="docutils literal notranslate"><span class="pre">/mod/A</span></code>, is stored in the same way as it would be stored in the <code class="docutils literal notranslate"><span class="pre">.h5ad</span></code> file. MuData objects can be read and written as follows:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">mdata</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s2">&quot;my_mudata.h5mu&quot;</span><span class="p">)</span>
<span class="n">mdata_r</span> <span class="o">=</span> <span class="n">md</span><span class="o">.</span><span class="n">read</span><span class="p">(</span><span class="s2">&quot;my_mudata.h5mu&quot;</span><span class="p">,</span> <span class="n">backed</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="n">mdata_r</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><pre>MuData object with n_obs × n_vars = 1000 × 150 backed at &#x27;my_mudata.h5mu&#x27;
  2 modalities
    A:	1000 x 100
      uns:	&#x27;misc&#x27;
    B:	1000 x 50</pre></div></div>
</div>
<p>Individual modalities are backed as well inside the <code class="docutils literal notranslate"><span class="pre">.h5mu</span></code> file.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">mdata_r</span><span class="p">[</span><span class="s2">&quot;A&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">isbacked</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>True
</pre></div>
</div>
</div>
</div>
<p>If the original object is backed, the filename has to be provided to the <code class="docutils literal notranslate"><span class="pre">.copy()</span></code> call, and the resulting object will be backed at a new location.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">mdata_sub</span> <span class="o">=</span> <span class="n">mdata_r</span><span class="o">.</span><span class="n">copy</span><span class="p">(</span><span class="s2">&quot;mdata_sub.h5mu&quot;</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="n">mdata_sub</span><span class="o">.</span><span class="n">is_view</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="n">mdata_sub</span><span class="o">.</span><span class="n">isbacked</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>False
True
</pre></div>
</div>
</div>
</div>
</section>
<section id="multimodal-methods">
<h3><span class="section-number">4.4.8. </span>Multimodal methods<a class="headerlink" href="#multimodal-methods" title="Link to this heading">#</a></h3>
<p>When the MuData object is prepared, it is up to multimodal methods to be used to make sense of the data. The most simple and naïve approach is to concatenate matrices from multiple modalities to perform for example dimensionality reduction.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">x</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">hstack</span><span class="p">([</span><span class="n">mdata</span><span class="o">.</span><span class="n">mod</span><span class="p">[</span><span class="s2">&quot;A&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">X</span><span class="p">,</span> <span class="n">mdata</span><span class="o">.</span><span class="n">mod</span><span class="p">[</span><span class="s2">&quot;B&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">X</span><span class="p">])</span>
<span class="n">x</span><span class="o">.</span><span class="n">shape</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>(1000, 150)
</pre></div>
</div>
</div>
</div>
<p>We can write a simple function to run principal component analysis on such a concatenated matrix. MuData object provides a place to store multimodal embeddings: MuData<code class="docutils literal notranslate"><span class="pre">.obsm</span></code>. It is similar to how the embeddings generated on individual modalities are stored, only this time it is saved inside the MuData object rather than in AnnData<code class="docutils literal notranslate"><span class="pre">.obsm</span></code>.</p>
<p>To calculate for example a principal component analysis (PCA) for the joint values of the modalities, we horizontally stack the values stored in the individual modalities and then perform the PCA on the stacked matrix. This is possible because the number of observations matches across modalities (remember, the number of features per modality does not have to match).</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="k">def</span><span class="w"> </span><span class="nf">simple_pca</span><span class="p">(</span><span class="n">mdata</span><span class="p">):</span>
    <span class="kn">from</span><span class="w"> </span><span class="nn">sklearn</span><span class="w"> </span><span class="kn">import</span> <span class="n">decomposition</span>

    <span class="n">x</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">hstack</span><span class="p">([</span><span class="n">m</span><span class="o">.</span><span class="n">X</span> <span class="k">for</span> <span class="n">m</span> <span class="ow">in</span> <span class="n">mdata</span><span class="o">.</span><span class="n">mod</span><span class="o">.</span><span class="n">values</span><span class="p">()])</span>

    <span class="n">pca</span> <span class="o">=</span> <span class="n">decomposition</span><span class="o">.</span><span class="n">PCA</span><span class="p">(</span><span class="n">n_components</span><span class="o">=</span><span class="mi">2</span><span class="p">)</span>
    <span class="n">components</span> <span class="o">=</span> <span class="n">pca</span><span class="o">.</span><span class="n">fit_transform</span><span class="p">(</span><span class="n">x</span><span class="p">)</span>

    <span class="c1"># By default, methods operate in-place and embeddings are stored in the .obsm slot</span>
    <span class="n">mdata</span><span class="o">.</span><span class="n">obsm</span><span class="p">[</span><span class="s2">&quot;X_pca&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">components</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">simple_pca</span><span class="p">(</span><span class="n">mdata</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="n">mdata</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>MuData object with n_obs × n_vars = 1000 × 150
  obsm:	&#39;X_pca&#39;
  2 modalities
    A:	1000 x 100
      uns:	&#39;misc&#39;
    B:	1000 x 50
</pre></div>
</div>
</div>
</div>
<p>Our calculated principal components are now stored in the MuData object itself and accessible for further multimodal transformations.</p>
<p>In reality, however, having different modalities often means that the features between them come from different generative processes and are not comparable. This is where special multimodal integration methods come into play. For omics technologies, these methods are frequently addressed as multi-omics integration methods. In the following section we will introduce muon which provides many tools to preprocess unimodal data beyond RNA-Seq and multi-omics integration methods.</p>
</section>
</section>
<section id="advanced-multimodal-data-analysis-with-muon">
<h2><span class="section-number">4.5. </span>Advanced: Multimodal data analysis with muon<a class="headerlink" href="#advanced-multimodal-data-analysis-with-muon" title="Link to this heading">#</a></h2>
<p>Although scanpy provides generally applicable tools such as PCA, UMAP and various visualizations, it is primarily designed for the analysis of RNA-Seq data. muon fills this gap by providing preprocessing functions for other omics such as chromatin accessibility (ATAC) or protein (CITE) data. As mentioned above, muon further provides algorithms to run multi-omics algorithms which infer knowledge from the joint modalities. For example, users may run a PCA on a single modality, but muon further provides multi-omics factor analysis algorithms which take several modalities as input.</p>
<figure class="align-default" id="muon-fig">
<a class="bg-primary mb-1 reference internal image-reference" href="../_images/muon.png"><img alt="muon Overview" class="bg-primary mb-1" src="../_images/muon.png" style="width: 800px;" />
</a>
<figcaption>
<p><span class="caption-number">Fig. 4.5 </span><span class="caption-text">muon overview. Image obtained from <span id="id19">[<a class="reference internal" href="#id146" title="Danila Bredikhin, Ilia Kats, and Oliver Stegle. Muon: multimodal omics analysis framework. Genome Biology, 23(1):42, Feb 2022. URL: https://doi.org/10.1186/s13059-021-02577-8, doi:10.1186/s13059-021-02577-8.">Bredikhin <em>et al.</em>, 2022</a>]</span>.</span><a class="headerlink" href="#muon-fig" title="Link to this image">#</a></p>
</figcaption>
</figure>
<section id="id20">
<h3><span class="section-number">4.5.1. </span>Installation<a class="headerlink" href="#id20" title="Link to this heading">#</a></h3>
<p>muon is available on PyPI can be installed using:</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>pip<span class="w"> </span>install<span class="w"> </span>muon
</pre></div>
</div>
</section>
<section id="api-overview">
<h3><span class="section-number">4.5.2. </span>API overview<a class="headerlink" href="#api-overview" title="Link to this heading">#</a></h3>
<p>For the sake of introducing muon we will examine the ATAC data of a multimodal dataset. Analogously to the scanpy chapter, this chapter solely serves as a quick demo and overview over muon and does not analyze a dataset completely, let alone providing best-practice multi-omics analysis. Please read the corresponding chapters to learn how to properly conduct such analyses.</p>
<p>muon separates its modules in two ways. First, analogously to scanpy, general and multimodal functions are grouped in preprocessing (<code class="docutils literal notranslate"><span class="pre">muon.pp</span></code>), tools (<code class="docutils literal notranslate"><span class="pre">muon.tl</span></code>) and plots (<code class="docutils literal notranslate"><span class="pre">muon.pl</span></code>). Second, unimodal tools are available from the corresponding of muon, which are itself again separated into preprocessing, tools and plots. For example, all ATAC preprocessing functions are grouped into <code class="docutils literal notranslate"><span class="pre">muon.atac.pp</span></code>. This also applies to CITE-Seq preprocessing functions (<code class="docutils literal notranslate"><span class="pre">muon.prot.pp</span></code>).</p>
<figure class="align-default" id="muon-api">
<a class="bg-primary mb-1 reference internal image-reference" href="../_images/muon_api.png"><img alt="muon API Overview" class="bg-primary mb-1" src="../_images/muon_api.png" style="width: 800px;" />
</a>
<figcaption>
<p><span class="caption-number">Fig. 4.6 </span><span class="caption-text">muon API overview. Modality specific functions are provided in correspondingly named modules. Functions for the joint analysis of modalities are available via the muon module directly.</span><a class="headerlink" href="#muon-api" title="Link to this image">#</a></p>
</figcaption>
</figure>
</section>
<section id="muon-api-demo">
<h3><span class="section-number">4.5.3. </span>muon API demo<a class="headerlink" href="#muon-api-demo" title="Link to this heading">#</a></h3>
<p>The dataset for our demo is a publicly available 10x Genomics Multiome dataset for human PBMCs. Let us import muon and get the dataset with mudatasets.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span><span class="w"> </span><span class="nn">mudatasets</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">mds</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">muon</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">mu</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">mdata</span><span class="p">,</span> <span class="n">_</span> <span class="o">=</span> <span class="n">mds</span><span class="o">.</span><span class="n">load</span><span class="p">(</span><span class="s2">&quot;pbmc10k_multiome&quot;</span><span class="p">,</span> <span class="n">with_info</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">full</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="n">mdata</span><span class="o">.</span><span class="n">var_names_make_unique</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>■ File filtered_feature_bc_matrix.h5 from pbmc10k_multiome has been found at /home/zeth/mudatasets/pbmc10k_multiome/filtered_feature_bc_matrix.h5
■ Checksum is validated (md5) for filtered_feature_bc_matrix.h5
■ File atac_fragments.tsv.gz from pbmc10k_multiome has been found at /home/zeth/mudatasets/pbmc10k_multiome/atac_fragments.tsv.gz
■ Checksum is validated (md5) for atac_fragments.tsv.gz
■ File atac_fragments.tsv.gz.tbi from pbmc10k_multiome has been found at /home/zeth/mudatasets/pbmc10k_multiome/atac_fragments.tsv.gz.tbi
■ Checksum is validated (md5) for atac_fragments.tsv.gz.tbi
■ File atac_peaks.bed from pbmc10k_multiome has been found at /home/zeth/mudatasets/pbmc10k_multiome/atac_peaks.bed
■ Checksum is validated (md5) for atac_peaks.bed
■ File atac_peak_annotation.tsv from pbmc10k_multiome has been found at /home/zeth/mudatasets/pbmc10k_multiome/atac_peak_annotation.tsv
■ Checksum is validated (md5) for atac_peak_annotation.tsv
■ Loading filtered_feature_bc_matrix.h5...
</pre></div>
</div>
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>/home/zeth/miniconda3/envs/bp2/lib/python3.9/site-packages/mudatasets/core.py:203: UserWarning: Dataset is in the 10X .h5 format and can&#39;t be loaded as backed.
  warn(&quot;Dataset is in the 10X .h5 format and can&#39;t be loaded as backed.&quot;)
Variable names are not unique. To make them unique, call `.var_names_make_unique`.
</pre></div>
</div>
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Added `interval` annotation for features from /home/zeth/mudatasets/pbmc10k_multiome/filtered_feature_bc_matrix.h5
</pre></div>
</div>
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Variable names are not unique. To make them unique, call `.var_names_make_unique`.
</pre></div>
</div>
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Added peak annotation from /home/zeth/mudatasets/pbmc10k_multiome/atac_peak_annotation.tsv to .uns[&#39;atac&#39;][&#39;peak_annotation&#39;]
Added gene names to peak annotation in .uns[&#39;atac&#39;][&#39;peak_annotation&#39;]
Located fragments file: /home/zeth/mudatasets/pbmc10k_multiome/atac_fragments.tsv.gz
</pre></div>
</div>
</div>
</div>
<p>As a first step we subset to the ATAC modality.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">atac</span> <span class="o">=</span> <span class="n">mdata</span><span class="o">.</span><span class="n">mod</span><span class="p">[</span><span class="s2">&quot;atac&quot;</span><span class="p">]</span>
</pre></div>
</div>
</div>
</div>
<p>Although, we are now not working with RNA-Seq, it is possible to use some of scanpy’s preprocessing functions which can also be used on ATAC data. This is possible due to similar distribution and quality issues of both modalities. The only thing to bear in mind here that a gene would mean a peak in the context of the AnnData object with ATAC-seq data. Afterwards, ATAC specific preprocessing can be conducted with the ATAC module of muon.</p>
<p>Let us start with some quality control by filtering out cells with too few peaks and peaks detected in too few cells. For now, we will filter out cells that do not pass QC.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">sc</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">calculate_qc_metrics</span><span class="p">(</span><span class="n">atac</span><span class="p">,</span> <span class="n">percent_top</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">log1p</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">inplace</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">sc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">violin</span><span class="p">(</span><span class="n">atac</span><span class="p">,</span> <span class="p">[</span><span class="s2">&quot;total_counts&quot;</span><span class="p">,</span> <span class="s2">&quot;n_genes_by_counts&quot;</span><span class="p">],</span> <span class="n">jitter</span><span class="o">=</span><span class="mf">0.4</span><span class="p">,</span> <span class="n">multi_panel</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>/home/zeth/miniconda3/envs/bp2/lib/python3.9/site-packages/anndata/_core/anndata.py:1228: FutureWarning: The `inplace` parameter in pandas.Categorical.reorder_categories is deprecated and will be removed in a future version. Reordering categories will always return a new Categorical object.
  c.reorder_categories(natsorted(c.categories), inplace=True)
... storing &#39;feature_types&#39; as categorical
/home/zeth/miniconda3/envs/bp2/lib/python3.9/site-packages/anndata/_core/anndata.py:1228: FutureWarning: The `inplace` parameter in pandas.Categorical.reorder_categories is deprecated and will be removed in a future version. Reordering categories will always return a new Categorical object.
  c.reorder_categories(natsorted(c.categories), inplace=True)
... storing &#39;genome&#39; as categorical
</pre></div>
</div>
<a class="reference internal image-reference" href="../_images/af874602e780c9e686a842b4d95c729ef75d817a01e4e613f1973ddb2d92370b.png"><img alt="../_images/af874602e780c9e686a842b4d95c729ef75d817a01e4e613f1973ddb2d92370b.png" src="../_images/af874602e780c9e686a842b4d95c729ef75d817a01e4e613f1973ddb2d92370b.png" style="width: 782px; height: 381px;" />
</a>
</div>
</div>
<p>Filter peaks whose expression is not detected.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">mu</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">filter_var</span><span class="p">(</span><span class="n">atac</span><span class="p">,</span> <span class="s2">&quot;n_cells_by_counts&quot;</span><span class="p">,</span> <span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">x</span> <span class="o">&gt;=</span> <span class="mi">10</span><span class="p">)</span>
<span class="c1"># This is analogous to</span>
<span class="c1">#   sc.pp.filter_genes(rna, min_cells=10)</span>
<span class="c1"># but does in-place filtering and avoids copying the object</span>
</pre></div>
</div>
</div>
</div>
<p>We also filter the cells.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">mu</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">filter_obs</span><span class="p">(</span><span class="n">atac</span><span class="p">,</span> <span class="s2">&quot;n_genes_by_counts&quot;</span><span class="p">,</span> <span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="p">(</span><span class="n">x</span> <span class="o">&gt;=</span> <span class="mi">2000</span><span class="p">)</span> <span class="o">&amp;</span> <span class="p">(</span><span class="n">x</span> <span class="o">&lt;=</span> <span class="mi">15000</span><span class="p">))</span>
<span class="c1"># This is analogous to</span>
<span class="c1">#   sc.pp.filter_cells(atac, max_genes=15000)</span>
<span class="c1">#   sc.pp.filter_cells(atac, min_genes=2000)</span>
<span class="c1"># but does in-place filtering avoiding copying the object</span>

<span class="n">mu</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">filter_obs</span><span class="p">(</span><span class="n">atac</span><span class="p">,</span> <span class="s2">&quot;total_counts&quot;</span><span class="p">,</span> <span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="p">(</span><span class="n">x</span> <span class="o">&gt;=</span> <span class="mi">4000</span><span class="p">)</span> <span class="o">&amp;</span> <span class="p">(</span><span class="n">x</span> <span class="o">&lt;=</span> <span class="mi">40000</span><span class="p">))</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">sc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">violin</span><span class="p">(</span><span class="n">atac</span><span class="p">,</span> <span class="p">[</span><span class="s2">&quot;n_genes_by_counts&quot;</span><span class="p">,</span> <span class="s2">&quot;total_counts&quot;</span><span class="p">],</span> <span class="n">jitter</span><span class="o">=</span><span class="mf">0.4</span><span class="p">,</span> <span class="n">multi_panel</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<a class="reference internal image-reference" href="../_images/03e1c22ce484a20491d297966cb3f717ce5bf756df418bcca5453152410a458c.png"><img alt="../_images/03e1c22ce484a20491d297966cb3f717ce5bf756df418bcca5453152410a458c.png" src="../_images/03e1c22ce484a20491d297966cb3f717ce5bf756df418bcca5453152410a458c.png" style="width: 782px; height: 381px;" />
</a>
</div>
</div>
<p>muon also provides histograms which allows for a different view on the metrics.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">mu</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">histogram</span><span class="p">(</span><span class="n">atac</span><span class="p">,</span> <span class="p">[</span><span class="s2">&quot;n_genes_by_counts&quot;</span><span class="p">,</span> <span class="s2">&quot;total_counts&quot;</span><span class="p">])</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<a class="reference internal image-reference" href="../_images/750ad0deb64d8e0242a27beccf8b02431853a720d5cc0d67b1c8f1a206febc30.png"><img alt="../_images/750ad0deb64d8e0242a27beccf8b02431853a720d5cc0d67b1c8f1a206febc30.png" src="../_images/750ad0deb64d8e0242a27beccf8b02431853a720d5cc0d67b1c8f1a206febc30.png" style="width: 463px; height: 203px;" />
</a>
</div>
</div>
<p>Now that we rudimentary filtered out cells with too few peaks and peaks detected in too few cells, we can start with ATAC specific quality control with muon. muon has modality specific preprocessing functions in corresponding modules. We import the ATAC module to access the ATAC specific preprocessing functions.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span><span class="w"> </span><span class="nn">muon</span><span class="w"> </span><span class="kn">import</span> <span class="n">atac</span> <span class="k">as</span> <span class="n">ac</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Perform rudimentary quality control with muon&#39;s ATAC module</span>
<span class="n">atac</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s2">&quot;NS&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="mi">1</span>
<span class="n">ac</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">nucleosome_signal</span><span class="p">(</span><span class="n">atac</span><span class="p">,</span> <span class="n">n</span><span class="o">=</span><span class="mf">1e6</span><span class="p">)</span>
<span class="n">ac</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">get_gene_annotation_from_rna</span><span class="p">(</span><span class="n">mdata</span><span class="p">[</span><span class="s2">&quot;rna&quot;</span><span class="p">])</span><span class="o">.</span><span class="n">head</span><span class="p">(</span><span class="mi">3</span><span class="p">)</span>
<span class="n">tss</span> <span class="o">=</span> <span class="n">ac</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">tss_enrichment</span><span class="p">(</span><span class="n">mdata</span><span class="p">,</span> <span class="n">n_tss</span><span class="o">=</span><span class="mi">1000</span><span class="p">)</span>
<span class="n">ac</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">tss_enrichment</span><span class="p">(</span><span class="n">tss</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Reading Fragments: 100%|██████████| 1000000/1000000 [00:05&lt;00:00, 185917.24it/s]
Fetching Regions...: 100%|██████████| 1000/1000 [00:19&lt;00:00, 50.03it/s]
</pre></div>
</div>
<a class="reference internal image-reference" href="../_images/55eb6874ce8bf52d43523d5aa4299bed653f9aceb75c2bdf65a51812ecfac499.png"><img alt="../_images/55eb6874ce8bf52d43523d5aa4299bed653f9aceb75c2bdf65a51812ecfac499.png" src="../_images/55eb6874ce8bf52d43523d5aa4299bed653f9aceb75c2bdf65a51812ecfac499.png" style="width: 310px; height: 318px;" />
</a>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Save original counts, normalize data and select highly variable genes with scanpy</span>
<span class="n">atac</span><span class="o">.</span><span class="n">layers</span><span class="p">[</span><span class="s2">&quot;counts&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">atac</span><span class="o">.</span><span class="n">X</span>
<span class="n">sc</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">normalize_per_cell</span><span class="p">(</span><span class="n">atac</span><span class="p">,</span> <span class="n">counts_per_cell_after</span><span class="o">=</span><span class="mf">1e4</span><span class="p">)</span>
<span class="n">sc</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">log1p</span><span class="p">(</span><span class="n">atac</span><span class="p">)</span>
<span class="n">sc</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">highly_variable_genes</span><span class="p">(</span><span class="n">atac</span><span class="p">,</span> <span class="n">min_mean</span><span class="o">=</span><span class="mf">0.05</span><span class="p">,</span> <span class="n">max_mean</span><span class="o">=</span><span class="mf">1.5</span><span class="p">,</span> <span class="n">min_disp</span><span class="o">=</span><span class="mf">0.5</span><span class="p">)</span>
<span class="n">atac</span><span class="o">.</span><span class="n">raw</span> <span class="o">=</span> <span class="n">atac</span>
</pre></div>
</div>
</div>
</div>
<p>Although PCA is also commonly used for ATAC data, latent semantic indexing (LSI) is another popular option. It is implemented in muon’s ATAC module.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">ac</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">lsi</span><span class="p">(</span><span class="n">atac</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">sc</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">neighbors</span><span class="p">(</span><span class="n">atac</span><span class="p">,</span> <span class="n">use_rep</span><span class="o">=</span><span class="s2">&quot;X_lsi&quot;</span><span class="p">,</span> <span class="n">n_neighbors</span><span class="o">=</span><span class="mi">10</span><span class="p">,</span> <span class="n">n_pcs</span><span class="o">=</span><span class="mi">30</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">sc</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">leiden</span><span class="p">(</span><span class="n">atac</span><span class="p">,</span> <span class="n">resolution</span><span class="o">=</span><span class="mf">0.5</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">sc</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">umap</span><span class="p">(</span><span class="n">atac</span><span class="p">,</span> <span class="n">spread</span><span class="o">=</span><span class="mf">1.5</span><span class="p">,</span> <span class="n">min_dist</span><span class="o">=</span><span class="mf">0.5</span><span class="p">,</span> <span class="n">random_state</span><span class="o">=</span><span class="mi">20</span><span class="p">)</span>
<span class="n">sc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">umap</span><span class="p">(</span><span class="n">atac</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;leiden&quot;</span><span class="p">,</span> <span class="s2">&quot;n_genes_by_counts&quot;</span><span class="p">],</span> <span class="n">legend_loc</span><span class="o">=</span><span class="s2">&quot;on data&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<a class="reference internal image-reference" href="../_images/c805123a0db86b8944bcbab0b09263875ddaa4dfd88127632b1966b06a6d15c4.png"><img alt="../_images/c805123a0db86b8944bcbab0b09263875ddaa4dfd88127632b1966b06a6d15c4.png" src="../_images/c805123a0db86b8944bcbab0b09263875ddaa4dfd88127632b1966b06a6d15c4.png" style="width: 692px; height: 299px;" />
</a>
</div>
</div>
<p>We can use the functionality of the ATAC module in muon to color plots by cut values in peaks corresponding to a certain gene.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">ac</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">umap</span><span class="p">(</span><span class="n">atac</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;KLF4&quot;</span><span class="p">],</span> <span class="n">average</span><span class="o">=</span><span class="s2">&quot;peak_type&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<a class="reference internal image-reference" href="../_images/0d8d7f7863bfa9193357387e7e542d3a8c1a59323687d3270e30e58a84a78cef.png"><img alt="../_images/0d8d7f7863bfa9193357387e7e542d3a8c1a59323687d3270e30e58a84a78cef.png" src="../_images/0d8d7f7863bfa9193357387e7e542d3a8c1a59323687d3270e30e58a84a78cef.png" style="width: 667px; height: 299px;" />
</a>
</div>
</div>
<p>For more details on all available functions of muon, please read the muon API reference at <a class="reference external" href="https://muon.readthedocs.io/en/latest/api/index.html">https://muon.readthedocs.io/en/latest/api/index.html</a> and the muon tutorials at <a class="reference external" href="https://muon-tutorials.readthedocs.io/en/latest/">https://muon-tutorials.readthedocs.io/en/latest/</a>.</p>
</section>
</section>
<section id="references">
<h2><span class="section-number">4.6. </span>References<a class="headerlink" href="#references" title="Link to this heading">#</a></h2>
<div class="docutils container" id="id21">
<div role="list" class="citation-list">
<div class="citation" id="id146" role="doc-biblioentry">
<span class="label"><span class="fn-bracket">[</span>atBKS22<span class="fn-bracket">]</span></span>
<span class="backrefs">(<a role="doc-backlink" href="#id14">1</a>,<a role="doc-backlink" href="#id17">2</a>,<a role="doc-backlink" href="#id19">3</a>)</span>
<p>Danila Bredikhin, Ilia Kats, and Oliver Stegle. Muon: multimodal omics analysis framework. <em>Genome Biology</em>, 23(1):42, Feb 2022. URL: <a class="reference external" href="https://doi.org/10.1186/s13059-021-02577-8">https://doi.org/10.1186/s13059-021-02577-8</a>, <a class="reference external" href="https://doi.org/10.1186/s13059-021-02577-8">doi:10.1186/s13059-021-02577-8</a>.</p>
</div>
<div class="citation" id="id133" role="doc-biblioentry">
<span class="label"><span class="fn-bracket">[</span><a role="doc-backlink" href="#id2">atHHAN+21</a><span class="fn-bracket">]</span></span>
<p>Yuhan Hao, Stephanie Hao, Erica Andersen-Nissen, William M. Mauck III, Shiwei Zheng, Andrew Butler, Maddie J. Lee, Aaron J. Wilk, Charlotte Darby, Michael Zagar, Paul Hoffman, Marlon Stoeckius, Efthymia Papalexi, Eleni P. Mimitou, Jaison Jain, Avi Srivastava, Tim Stuart, Lamar B. Fleming, Bertrand Yeung, Angela J. Rogers, Juliana M. McElrath, Catherine A. Blish, Raphael Gottardo, Peter Smibert, and Rahul Satija. Integrated analysis of multimodal single-cell data. <em>Cell</em>, 2021. URL: <a class="reference external" href="https://doi.org/10.1016/j.cell.2021.04.048">https://doi.org/10.1016/j.cell.2021.04.048</a>, <a class="reference external" href="https://doi.org/10.1016/j.cell.2021.04.048">doi:10.1016/j.cell.2021.04.048</a>.</p>
</div>
<div class="citation" id="id159" role="doc-biblioentry">
<span class="label"><span class="fn-bracket">[</span><a role="doc-backlink" href="#id1">atHCG+15</a><span class="fn-bracket">]</span></span>
<p>Wolfgang Huber, Vincent J. Carey, Robert Gentleman, Simon Anders, Marc Carlson, Benilton S. Carvalho, Hector Corrada Bravo, Sean Davis, Laurent Gatto, Thomas Girke, Raphael Gottardo, Florian Hahne, Kasper D. Hansen, Rafael A. Irizarry, Michael Lawrence, Michael I. Love, James MacDonald, Valerie Obenchain, Andrzej K. Oleś, Hervé Pagès, Alejandro Reyes, Paul Shannon, Gordon K. Smyth, Dan Tenenbaum, Levi Waldron, and Martin Morgan. Orchestrating high-throughput genomic analysis with bioconductor. <em>Nature Methods</em>, 12(2):115–121, Feb 2015. URL: <a class="reference external" href="https://doi.org/10.1038/nmeth.3252">https://doi.org/10.1038/nmeth.3252</a>, <a class="reference external" href="https://doi.org/10.1038/nmeth.3252">doi:10.1038/nmeth.3252</a>.</p>
</div>
<div class="citation" id="id142" role="doc-biblioentry">
<span class="label"><span class="fn-bracket">[</span><a role="doc-backlink" href="#id10">atJup22</a><span class="fn-bracket">]</span></span>
<p>Project Jupyter. Jupyter. <a class="reference external" href="https://jupyter.org/">https://jupyter.org/</a>, 2022. Accessed: 2022-04-21.</p>
</div>
<div class="citation" id="id141" role="doc-biblioentry">
<span class="label"><span class="fn-bracket">[</span><a role="doc-backlink" href="#id9">atKFM+21</a><span class="fn-bracket">]</span></span>
<p>Combiz Khozoie, Nurun Fancy, Mahdi M. Marjaneh, Alan E. Murphy, Paul M. Matthews, and Nathan Skene. Scflow: a scalable and reproducible analysis pipeline for single-cell term`rna` sequencing data. <em>bioRxiv</em>, 2021. URL: <a class="reference external" href="https://www.biorxiv.org/content/early/2021/08/19/2021.08.16.456499.1">https://www.biorxiv.org/content/early/2021/08/19/2021.08.16.456499.1</a>, <a class="reference external" href="https://arxiv.org/abs/https://www.biorxiv.org/content/early/2021/08/19/2021.08.16.456499.1.full.pdf">arXiv:https://www.biorxiv.org/content/early/2021/08/19/2021.08.16.456499.1.full.pdf</a>, <a class="reference external" href="https://doi.org/10.1101/2021.08.16.456499">doi:10.1101/2021.08.16.456499</a>.</p>
</div>
<div class="citation" id="id137" role="doc-biblioentry">
<span class="label"><span class="fn-bracket">[</span><a role="doc-backlink" href="#id7">atPSK+22</a><span class="fn-bracket">]</span></span>
<p>Giovanni Palla, Hannah Spitzer, Michal Klein, David Fischer, Anna Christina Schaar, Louis Benedikt Kuemmerle, Sergei Rybakov, Ignacio L. Ibarra, Olle Holmberg, Isaac Virshup, Mohammad Lotfollahi, Sabrina Richter, and Fabian J. Theis. Squidpy: a scalable framework for spatial omics analysis. <em>Nature Methods</em>, 19(2):171–178, Feb 2022. URL: <a class="reference external" href="https://doi.org/10.1038/s41592-021-01358-2">https://doi.org/10.1038/s41592-021-01358-2</a>, <a class="reference external" href="https://doi.org/10.1038/s41592-021-01358-2">doi:10.1038/s41592-021-01358-2</a>.</p>
</div>
<div class="citation" id="id134" role="doc-biblioentry">
<span class="label"><span class="fn-bracket">[</span><a role="doc-backlink" href="#id3">atscv22</a><span class="fn-bracket">]</span></span>
<p>scverse. Scverse. <a class="reference external" href="https://scverse.org">https://scverse.org</a>, 2022. Accessed: 2022-04-21.</p>
</div>
<div class="citation" id="id147" role="doc-biblioentry">
<span class="label"><span class="fn-bracket">[</span><a role="doc-backlink" href="#id15">atsm22a</a><span class="fn-bracket">]</span></span>
<p>scverse mudata. Mudata quickstart. <a class="reference external" href="https://mudata.readthedocs.io/en/latest/notebooks/quickstart_mudata.html">https://mudata.readthedocs.io/en/latest/notebooks/quickstart_mudata.html</a>, 2022. Accessed: 2022-04-21.</p>
</div>
<div class="citation" id="id148" role="doc-biblioentry">
<span class="label"><span class="fn-bracket">[</span><a role="doc-backlink" href="#id16">atsm22b</a><span class="fn-bracket">]</span></span>
<p>scverse mudata. Mudata quickstart. <a class="reference external" href="https://mudata.readthedocs.io/en/latest/io/mudata.html">https://mudata.readthedocs.io/en/latest/io/mudata.html</a>, 2022. Accessed: 2022-04-21.</p>
</div>
<div class="citation" id="id143" role="doc-biblioentry">
<span class="label"><span class="fn-bracket">[</span><a role="doc-backlink" href="#id13">atss22</a><span class="fn-bracket">]</span></span>
<p>scverse scanpy. Scanpy api. <a class="reference external" href="https://scanpy.readthedocs.io/en/stable/api.html#">https://scanpy.readthedocs.io/en/stable/api.html#</a>, 2022. Accessed: 2022-04-21.</p>
</div>
<div class="citation" id="id138" role="doc-biblioentry">
<span class="label"><span class="fn-bracket">[</span><a role="doc-backlink" href="#id8">atSSF+20</a><span class="fn-bracket">]</span></span>
<p>Gregor Sturm, Tamas Szabo, Georgios Fotakis, Marlene Haider, Dietmar Rieder, Zlatko Trajanoski, and Francesca Finotello. Scirpy: a Scanpy extension for analyzing single-cell T-cell receptor-sequencing data. <em>Bioinformatics</em>, 36(18):4817–4818, 07 2020. URL: <a class="reference external" href="https://doi.org/10.1093/bioinformatics/btaa611">https://doi.org/10.1093/bioinformatics/btaa611</a>, <a class="reference external" href="https://arxiv.org/abs/https://academic.oup.com/bioinformatics/article-pdf/36/18/4817/34560298/btaa611.pdf">arXiv:https://academic.oup.com/bioinformatics/article-pdf/36/18/4817/34560298/btaa611.pdf</a>, <a class="reference external" href="https://doi.org/10.1093/bioinformatics/btaa611">doi:10.1093/bioinformatics/btaa611</a>.</p>
</div>
<div class="citation" id="id135" role="doc-biblioentry">
<span class="label"><span class="fn-bracket">[</span>atVRT+21<span class="fn-bracket">]</span></span>
<span class="backrefs">(<a role="doc-backlink" href="#id4">1</a>,<a role="doc-backlink" href="#id5">2</a>)</span>
<p>Isaac Virshup, Sergei Rybakov, Fabian J. Theis, Philipp Angerer, and F. Alexander Wolf. Anndata: annotated data. <em>bioRxiv</em>, 2021. URL: <a class="reference external" href="https://www.biorxiv.org/content/early/2021/12/19/2021.12.16.473007">https://www.biorxiv.org/content/early/2021/12/19/2021.12.16.473007</a>, <a class="reference external" href="https://arxiv.org/abs/https://www.biorxiv.org/content/early/2021/12/19/2021.12.16.473007.full.pdf">arXiv:https://www.biorxiv.org/content/early/2021/12/19/2021.12.16.473007.full.pdf</a>, <a class="reference external" href="https://doi.org/10.1101/2021.12.16.473007">doi:10.1101/2021.12.16.473007</a>.</p>
</div>
<div class="citation" id="id136" role="doc-biblioentry">
<span class="label"><span class="fn-bracket">[</span>atWAT18<span class="fn-bracket">]</span></span>
<span class="backrefs">(<a role="doc-backlink" href="#id6">1</a>,<a role="doc-backlink" href="#id11">2</a>)</span>
<p>F. Alexander Wolf, Philipp Angerer, and Fabian J. Theis. Scanpy: large-scale single-cell gene expression data analysis. <em>Genome Biology</em>, 19(1):15, Feb 2018. URL: <a class="reference external" href="https://doi.org/10.1186/s13059-017-1382-0">https://doi.org/10.1186/s13059-017-1382-0</a>, <a class="reference external" href="https://doi.org/10.1186/s13059-017-1382-0">doi:10.1186/s13059-017-1382-0</a>.</p>
</div>
</div>
</div>
</section>
<section id="contributors">
<h2><span class="section-number">4.7. </span>Contributors<a class="headerlink" href="#contributors" title="Link to this heading">#</a></h2>
<p>We gratefully acknowledge the contributions of:</p>
<section id="authors">
<h3><span class="section-number">4.7.1. </span>Authors<a class="headerlink" href="#authors" title="Link to this heading">#</a></h3>
<ul class="simple">
<li><p>Lukas Heumos</p></li>
</ul>
</section>
<section id="reviewers">
<h3><span class="section-number">4.7.2. </span>Reviewers<a class="headerlink" href="#reviewers" title="Link to this heading">#</a></h3>
<ul class="simple">
<li><p>Isaac Virshup</p></li>
</ul>
</section>
</section>
</section>

    <script type="text/x-thebe-config">
    {
        requestKernel: true,
        binderOptions: {
            repo: "binder-examples/jupyter-stacks-datascience",
            ref: "master",
        },
        codeMirrorConfig: {
            theme: "abcdef",
            mode: "python"
        },
        kernelOptions: {
            name: "python3",
            path: "./introduction"
        },
        predefinedOutput: true
    }
    </script>
    <script>kernelName = 'python3'</script>

                </article>
              

              
              
              
              
                <footer class="prev-next-footer d-print-none">
                  
<div class="prev-next-area">
    <a class="left-prev"
       href="raw_data_processing.html"
       title="previous page">
      <i class="fa-solid fa-angle-left"></i>
      <div class="prev-next-info">
        <p class="prev-next-subtitle">previous</p>
        <p class="prev-next-title"><span class="section-number">3. </span>Raw data processing</p>
      </div>
    </a>
    <a class="right-next"
       href="interoperability.html"
       title="next page">
      <div class="prev-next-info">
        <p class="prev-next-subtitle">next</p>
        <p class="prev-next-title"><span class="section-number">5. </span>Interoperability</p>
      </div>
      <i class="fa-solid fa-angle-right"></i>
    </a>
</div>
                </footer>
              
            </div>
            
            
              
                <div class="bd-sidebar-secondary bd-toc"><div class="sidebar-secondary-items sidebar-secondary__inner">


  <div class="sidebar-secondary-item">
  <div class="page-toc tocsection onthispage">
    <i class="fa-solid fa-list"></i> Contents
  </div>
  <nav class="bd-toc-nav page-toc">
    <ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#single-cell-analysis-frameworks-and-consortia">4.1. Single-cell analysis frameworks and consortia</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#storing-unimodal-data-with-anndata">4.2. Storing unimodal data with AnnData</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#installation">4.2.1. Installation</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#initializing-an-anndata-object">4.2.2. Initializing an AnnData object</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#adding-aligned-metadata">4.2.3. Adding aligned metadata</a><ul class="nav section-nav flex-column">
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#observational-or-variable-level">4.2.3.1. Observational or Variable level</a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#subsetting-using-metadata">4.2.3.2. Subsetting using metadata</a></li>
</ul>
</li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#observation-variable-level-matrices">4.2.4. Observation/variable-level matrices</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#unstructured-metadata">4.2.5. Unstructured metadata</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#layers">4.2.6. Layers</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#conversion-to-dataframes">4.2.7. Conversion to DataFrames</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#reading-and-writing-of-anndata-objects">4.2.8. Reading and writing of AnnData objects</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#efficient-data-access">4.2.9. Efficient data access</a><ul class="nav section-nav flex-column">
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#view-and-copies">4.2.9.1. View and copies</a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#partial-reading-of-large-data">4.2.9.2. Partial reading of large data</a></li>
</ul>
</li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#unimodal-data-analysis-with-scanpy">4.3. Unimodal data analysis with scanpy</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#id12">4.3.1. Installation</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#scanpy-api-design">4.3.2. Scanpy API design</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#scanpy-example">4.3.3. Scanpy example</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#advanced-using-mudata-to-store-multimodal-data">4.4. Advanced: Using MuData to store multimodal data</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#id18">4.4.1. Installation</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#initializing-a-mudata-object">4.4.2. Initializing a MuData object</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#mudata-attributes">4.4.3. MuData attributes</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#variable-mappings">4.4.4. Variable mappings</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#mudata-views">4.4.5. MuData views</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#common-observations">4.4.6. Common observations</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#reading-and-writing-of-mudata-objects">4.4.7. Reading and Writing of MuData objects</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#multimodal-methods">4.4.8. Multimodal methods</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#advanced-multimodal-data-analysis-with-muon">4.5. Advanced: Multimodal data analysis with muon</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#id20">4.5.1. Installation</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#api-overview">4.5.2. API overview</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#muon-api-demo">4.5.3. muon API demo</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#references">4.6. References</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#contributors">4.7. Contributors</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#authors">4.7.1. Authors</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#reviewers">4.7.2. Reviewers</a></li>
</ul>
</li>
</ul>
  </nav></div>

</div></div>
              
            
          </div>
          <footer class="bd-footer-content">
            
<div class="bd-footer-content__inner container">
  
  <div class="footer-item">
    
<p class="component-author">
By Lukas Heumos, Anna Schaar, single-cell best practices consortium
</p>

  </div>
  
  <div class="footer-item">
    

  <p class="copyright">
    
      © Copyright 2023.
      <br/>
    
  </p>

  </div>
  
  <div class="footer-item">
    
  </div>
  
  <div class="footer-item">
    
<div class="extra_footer">
  <div>
  Brought to you by
  <a href="https://www.github.com/theislab">Theislab</a>, with many thanks to the single-cell community as a whole!
</div>

</div>
  </div>
  
</div>
          </footer>
        

      </main>
    </div>
  </div>
  
  <!-- Scripts loaded after <body> so the DOM is not blocked -->
  <script src="../_static/scripts/bootstrap.js?digest=dfe6caa3a7d634c4db9b"></script>
<script src="../_static/scripts/pydata-sphinx-theme.js?digest=dfe6caa3a7d634c4db9b"></script>

  <footer class="bd-footer">
  </footer>
  </body>
</html>