# Experimental data collection

Sources:

- https://www.nature.com/articles/s41576-019-0093-7
- The paper that helped me understand experimental protocols most, but only very old methods: (Ziegenhain et al. 2017)
- Short, and broad overview of now somewhat old methods: (Svensson et al. 2017)
- More recent comparison of methods that are probably more relevant: (Mereu et al. 2020)
- Guidelines for the experimental design of single-cell RNA sequencing studies: (Lafzi et al. 2018)

Aimed length: 7-10 pages
Chapter is not aimed at experimentalists, but only at computational biologists to learn the biases of the assays and to get an overview about how it generally works.

Content:
1. Mention that we only cover RNA-Seq assays here. Multimodal and spatial assays are covered in their own advanced chapters.
2. Brief history about sequencing (0.5 pages)
3. Maybe show the overview figure which shows that we are now able to sequence many things and the number of assays is only growing (I Forgot where this figure is from, but think Mo showed it in his defense?)
4. Platelet vs Droplet etc high level overview
5. Now always introduce one assay and highlight strengths and weaknesses (esp based on https://www.nature.com/articles/s41587-020-0469-4)
6. Drop-Seq
7. Chromium
8. Smart-Seq2
9. What else ?!? InDrop? MARS-Seq? Cel-Seq2? Quartz-seq2