# Experimental data collection

This chapter, which is aimed at bioinformatians, provides a short overview about the most widely used single-cell RNA sequencing assays. Multimodal or spatial assays are not covered here, but instead in the respective advanced chapters. All sequencing assays have individual strengths and limitations which must be known by data analysis to be aware of possible biases in the data. The count distribution and gene coverage for example is heavily influenced by the chemistry of the assay.

:::{figure-md} markdown-fig
<img src="../_static/images/assays.png" alt="Overview of sequencing assays" class="bg-primary mb-1" width="800px">

Sequencing assays overview. Image obtained from <https://www.nature.com/articles/s41592-019-0691-5> WE HAVE NO RIGHTS AND THIS NEEDS TO BE REDONE OR RECONSIDERED.
:::

## A brief history of single-cell RNA-sequencing

### First generation sequencing

Although DNA was already first isolated in 1869 by Friedrich Mietscher it took the scientific community more than 100 years to develop high throughput sequencing technologies. In 1953 Watson, Crick and Franklin discovered the structure of DNA and in 1965 Robert Holley sequenced the first tRNA. Seven years later, in 1972, Walter Fiers was the first to sequence the DNA of a complete gene (MS2) using RNAses to digest the virus RNA and isolate oligonucleotides and finally separating them with electrophoresis and chromatography{cite}`JOU1972`. In parallel, Friedrich Sanger developed a DNA sequencing method that uses radiolabelled partially digested fragments termed "chain termination method", which is more commonly known as "Sanger Sequencing". Although, Sanger Sequencing is still used even today, it suffered from several short comings: It lacked automation and was time consuming. In 1987 Leroy Hood and Michael Hunkapiller developed the ABI 370, an instrument that automates the Sanger Sequencing process. The most important innovation to accomplish this was the automatic labelling of DNA fragments with fluorescent dies instead of radioactive molecules. This allowed for computers to analyze the acquired data{cite}`Hood1987`.

Strengths:

* Sanger sequencing is simple and does not require expensive machines.
* If done well, the error rate is very low.

Limitations:

* Sanger methods can only sequence short pieces of DNA - about 300 to 1000 base pairs.
* The quality of a Sanger sequence is often not very good in the first 15 to 40 bases because that is where the primer binds.
* Sequencing degrades after 700 to 900 bases.
* If the DNA fragment being sequenced has been cloned, some of the cloning vector sequence may find its way into the final sequence.
* Sanger sequencing is more expensive than second or third generation sequencing.

### Second generation sequencing

Second generation sequencing is also known as next-generation sequencing (NGS) which was primarily made possible by automation in the lab, the usage of computers and the miniaturisation of reactions. Nine years later, in 1996, Mostafa Ronaghi, Mathias Uhlen and Pȧl Nyŕen introduced a new DNA sequencing technique called pyrosequencing introducing the age of second generation sequencing. Pyrosequencing measures luminescence that is generated by pyrophosphate synthesis during sequencing. This process is also commonly known as "sequencing-by-synthesis"). Two years later, Shankar Balasubramanian and David Klenerman, developed adapted the sequencing-by-synthesis process for a new method which utilises fluorescent dyes at the company Solexa. The Roche 454 sequencer developed in 2005, was the first sequencer to fully automate the pyrosequencing process in a single, automated machine. Although, many other platforms such as SOLiD systems' "sequencing-by-ligation" (2007) and Life Technologies' Ion Torrent (2011) that uses "sequencing-by-synthesis" to detect hydrogen ions when new DNA is synthesized were introduced, Illumina's sequencers which are based on Solexa's technology after a takeover, dominate the market.

Strengths:

* Second generation generation sequencing is often times the cheapest option concerning the required chemicals.
* Sparse material can still be used as input.
* High sensitivity to detect low-frequency variants and comprehensive genome coverage.
* High capacity with sample multiplexing.
* Ability to sequence thousands of genes simultaneously,

Limitations:

* Illumina sequencers are expensive and often times need to be shared with colleagues.
* Second generation sequencers are big, stationary machines and not designed for field work.
* Generally, second generation sequencing results in many short sequencing fragments (reads) which are hard to use for novel genomes.

### Third generation sequencing

The third generation of sequencing, which is nowadays also known as next-generation sequencing depending on whom you ask, brought two innovations to the market. First, long read sequencing which describes the ability to obtain nucleotide fragments of longer lengths than the usual Illumina short read sequencers generate (order of 75 to 300 base pairs depending on the sequencer). This is especially important for the assembly of novel genomes where no reference genome is known. Second, the ability to sequence in real time is another major advancement of third generation sequencing. Combined with portable sequencers which are small in size and do not require further complicated machines for the chemistry, sequencing is now "field-ready" and can be used even far away from society to collect samples.

Pacific Biosciences' (PacBio) introduced zero-mode waveguide (ZMW) sequencing in 2010. ZMW uses so called nanoholes containing a single DNA polymerase. This allows for the incorporation of any single nucleotide to be directly observed which is used by detecters that are attached below the ZMW. Every nucleotide is labelled with a different fluorescent dye that emit fluorescent signals during the incorporation process which is subsequently detected. Reads obtained from PacBio sequencers are usually of 8 to 15 kilobases with possibilities for up to 70kb making them several times longer than Illumina reads.

However, PacBio sequencers are hardly portable. Oxford Nanopore Technologies' introduced the GridION in 2012. The GridION and its successors MinIO and Flongle are portable sequencers for DNA and RNA sequencing which produces reads of more than 2 Mb. Notably, such a sequencing device even fits into a single human hand. Oxford Nanopore's sequencers observes changes in electrical conductivity that occur when the DNA strands pass through biological nanopores to identify the nucleotide sequence{cite}`Jain2016`.

Strengths:

* Long reads which allow for the assembly of large novel genomes.
* Sequencers are portable allowing for field work.
* Possibility to detect epigenetic modifications.
* Speed. Third generation sequencers are fast.

Limitations:

* Some third generation sequencers exhibit higher error rates than second generation sequencers.
* The chemistry is generally more expensive than second generation sequencing.

### Overview of the NGS process

Even though a variety of NGS technologies exist, the general steps to sequence DNA (and therefore reverse transcribed RNA) are largely the same. The differences lie primarily in the chemistry of the respective sequencing technologies.

1. **Sample and library preparation**: As a first step a so called library is prepared by fragmenting the DNA samples and ligating them with adapter molecules. They act in the hybridisation of the library fragments to the matrix and provide a priming site.

2. **Amplification and sequencing**: In the second step, the library gets converted into single strand molecules. During an amplification step (such as polymerase chain reaction), clusters of DNA molecules are being created. All of the clusters perform individual reactions during a single sequencing run.

3. **Data output and analysis**: The output of a sequencing experiment depends on the sequencing technology and chemistry. Some sequencers generate fluorescence signals which are stored in specific output files and others may generate electric signals which are stored in corresponding file formats. Generally, the amount of generated data, the raw data, is very large. Such data requires complex and computationally heavy processing. This is further discussed in the raw data processing chapter.

### Single-cell sequencing & platelet vs droplet

## Drop-Seq

## 10x Chromium

## Smart-Seq

## Other assays

InDrop? MARS-Seq? Cel-Seq2? Quartz-seq2

## single-cell vs single-nuclei

Sources:

* <https://www.nature.com/articles/s41576-019-0093-7>
* The paper that helped me understand experimental protocols most, but only very old methods: (Ziegenhain et al. 2017)
* Short, and broad overview of now somewhat old methods: (Svensson et al. 2017) -> <https://www.nature.com/articles/nmeth.4220>
* More recent comparison of methods that are probably more relevant: (Mereu et al. 2020) -> <https://www.nature.com/articles/s41587-020-0469-4>
* Guidelines for the experimental design of single-cell RNA sequencing studies: (Lafzi et al. 2018)
* 10x vs SMART comparison -> <https://www.sciencedirect.com/science/article/pii/S1672022921000486>

## References

```{bibliography}
:filter: docname in docnames
```
