
<!DOCTYPE html>

<html lang="en">
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="generator" content="Docutils 0.17.1: http://docutils.sourceforge.net/" />

    <title>16. Lineage tracing &#8212; Multimodal single-cell analysis</title>
    
  <!-- Loaded before other Sphinx assets -->
  <link href="../_static/styles/theme.css?digest=1999514e3f237ded88cf" rel="stylesheet">
<link href="../_static/styles/pydata-sphinx-theme.css?digest=1999514e3f237ded88cf" rel="stylesheet">

    
  <link rel="stylesheet"
    href="../_static/vendor/fontawesome/5.13.0/css/all.min.css">
  <link rel="preload" as="font" type="font/woff2" crossorigin
    href="../_static/vendor/fontawesome/5.13.0/webfonts/fa-solid-900.woff2">
  <link rel="preload" as="font" type="font/woff2" crossorigin
    href="../_static/vendor/fontawesome/5.13.0/webfonts/fa-brands-400.woff2">

    <link rel="stylesheet" type="text/css" href="../_static/pygments.css" />
    <link rel="stylesheet" href="../_static/styles/sphinx-book-theme.css?digest=5115cc725059bd94278eecd172e13a965bf8f5a9" type="text/css" />
    <link rel="stylesheet" type="text/css" href="../_static/togglebutton.css" />
    <link rel="stylesheet" type="text/css" href="../_static/copybutton.css" />
    <link rel="stylesheet" type="text/css" href="../_static/mystnb.css" />
    <link rel="stylesheet" type="text/css" href="../_static/sphinx-thebe.css" />
    <link rel="stylesheet" type="text/css" href="../_static/book.css" />
    <link rel="stylesheet" type="text/css" href="../_static/design-style.b7bb847fb20b106c3d81b95245e65545.min.css" />
    
  <!-- Pre-loaded scripts that we'll load fully later -->
  <link rel="preload" as="script" href="../_static/scripts/pydata-sphinx-theme.js?digest=1999514e3f237ded88cf">

    <script data-url_root="../" id="documentation_options" src="../_static/documentation_options.js"></script>
    <script src="../_static/jquery.js"></script>
    <script src="../_static/underscore.js"></script>
    <script src="../_static/doctools.js"></script>
    <script src="../_static/clipboard.min.js"></script>
    <script src="../_static/copybutton.js"></script>
    <script src="../_static/scripts/sphinx-book-theme.js?digest=9c920249402e914e316237a7dbc6769907cce411"></script>
    <script>let toggleHintShow = 'Click to show';</script>
    <script>let toggleHintHide = 'Click to hide';</script>
    <script>let toggleOpenOnPrint = 'true';</script>
    <script src="../_static/togglebutton.js"></script>
    <script async="async" kind="hypothesis" src="https://hypothes.is/embed.js"></script>
    <script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
    <script src="../_static/book.js"></script>
    <script>var togglebuttonSelector = '.toggle, .admonition.dropdown, .tag_hide_input div.cell_input, .tag_hide-input div.cell_input, .tag_hide_output div.cell_output, .tag_hide-output div.cell_output, .tag_hide_cell.cell, .tag_hide-cell.cell';</script>
    <script src="../_static/design-tabs.js"></script>
    <script>const THEBE_JS_URL = "https://unpkg.com/thebe@0.8.2/lib/index.js"
const thebe_selector = ".thebe,.cell"
const thebe_selector_input = "pre"
const thebe_selector_output = ".output, .cell_output"
</script>
    <script async="async" src="../_static/sphinx-thebe.js"></script>
    <script>window.MathJax = {"options": {"processHtmlClass": "tex2jax_process|mathjax_process|math|output_area"}}</script>
    <script defer="defer" src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-mml-chtml.js"></script>
    <link rel="shortcut icon" href="../_static/favicon.ico"/>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="17. Differential gene expression analysis" href="../conditions/differential_gene_expression.html" />
    <link rel="prev" title="15. Fate mapping" href="fate_mapping.html" />
    <meta name="viewport" content="width=device-width, initial-scale=1" />
    <meta name="docsearch:language" content="en">
    

    <!-- Google Analytics -->
    
  </head>
  <body data-spy="scroll" data-target="#bd-toc-nav" data-offset="60">
<!-- Checkboxes to toggle the left sidebar -->
<input type="checkbox" class="sidebar-toggle" name="__navigation" id="__navigation" aria-label="Toggle navigation sidebar">
<label class="overlay overlay-navbar" for="__navigation">
    <div class="visually-hidden">Toggle navigation sidebar</div>
</label>
<!-- Checkboxes to toggle the in-page toc -->
<input type="checkbox" class="sidebar-toggle" name="__page-toc" id="__page-toc" aria-label="Toggle in-page Table of Contents">
<label class="overlay overlay-pagetoc" for="__page-toc">
    <div class="visually-hidden">Toggle in-page Table of Contents</div>
</label>
<!-- Headers at the top -->
<div class="announcement header-item noprint"></div>
<div class="header header-item noprint"></div>

    
    <div class="container-fluid" id="banner"></div>

    

    <div class="container-xl">
      <div class="row">
          
<!-- Sidebar -->
<div class="bd-sidebar noprint" id="site-navigation">
    <div class="bd-sidebar__content">
        <div class="bd-sidebar__top"><div class="navbar-brand-box">
    <a class="navbar-brand text-wrap" href="../index.html">
      
        <!-- `logo` is deprecated in Sphinx 4.0, so remove this when we stop supporting 3 -->
        
      
      
      <img src="../_static/logo.png" class="logo" alt="logo">
      
      
      <h1 class="site-logo" id="site-title">Multimodal single-cell analysis</h1>
      
    </a>
</div><form class="bd-search d-flex align-items-center" action="../search.html" method="get">
  <i class="icon fas fa-search"></i>
  <input type="search" class="form-control" name="q" id="search-input" placeholder="Search this book..." aria-label="Search this book..." autocomplete="off" >
</form><nav class="bd-links" id="bd-docs-nav" aria-label="Main">
    <div class="bd-toc-item active">
        <p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Introduction
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../introduction/prior_art.html">
   1. Prior art
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../introduction/scrna_seq.html">
   2. Single-cell RNA sequencing
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../introduction/raw_data_processing.html">
   3. Raw data processing
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../introduction/analysis_tools.html">
   4. Analysis frameworks and tools
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../introduction/data_infrastructure.html">
   5. Data infrastructure
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../introduction/interoperability.html">
   6. Interoperability
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Preprocessing and visualization
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../preprocessing_visualization/quality_control.html">
   7. Quality Control
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../preprocessing_visualization/normalization.html">
   8. Normalization
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../preprocessing_visualization/feature_selection.html">
   9. Feature selection
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../preprocessing_visualization/dimensionality_reduction.html">
   10. Dimensionality Reduction
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Identifying cellular structure
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../cellular_structure/clustering.html">
   11. Clustering
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../cellular_structure/integration.html">
   12. Data integration
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Inferring trajectories
 </span>
</p>
<ul class="current nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="pseudotemporal.html">
   13. Pseudotemporal ordering
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="rna_velocity.html">
   14. RNA velocity
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="fate_mapping.html">
   15. Fate mapping
  </a>
 </li>
 <li class="toctree-l1 current active">
  <a class="current reference internal" href="#">
   16. Lineage tracing
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Dealing with conditions
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../conditions/differential_gene_expression.html">
   17. Differential gene expression analysis
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../conditions/compositional.html">
   18. Compositional analysis
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../conditions/gsea_pathway.html">
   19. Gene set enrichment and pathway analysis
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../conditions/perturbation_modeling.html">
   20. Perturbation modeling
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Modeling mechanisms
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../mechanisms/grns.html">
   21. Gene regulatory networks
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../mechanisms/cell_cell_communication.html">
   22. Cell-cell communication
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Deconvolution
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../deconvolution/motivation.html">
   23. Motivation
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Surface protein
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../surface_protein/quality_control.html">
   24. Quality control
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../surface_protein/normalization.html">
   25. Normalization
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../surface_protein/doublet_detection.html">
   26. Doublet detection
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../surface_protein/dimensionality_reduction.html">
   27. Dimensionality Reduction
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../surface_protein/batch_correction.html">
   28. Batch correction
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../surface_protein/annotation.html">
   29. Annotation
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  AIR repertoire
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../air_repertoire/ir_profiling.html">
   30. Immune Receptor Profiling
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../air_repertoire/clonotype.html">
   31. Clonotype analysis
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../air_repertoire/specificity.html">
   32. Specificity analysis
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../air_repertoire/multimodal_integration.html">
   33. Integrating AIR and transcriptomics
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Reproducibility
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../reproducibility/introduction.html">
   34. Reproducibility
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Outlook
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../outlook.html">
   35. Outlook
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Acknowledgements
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../acknowledgements.html">
   36. Acknowledgements
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Glossary
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../glossary.html">
   37. Glossary
  </a>
 </li>
</ul>

    </div>
</nav></div>
        <div class="bd-sidebar__bottom">
             <!-- To handle the deprecated key -->
            
            <div class="navbar_extra_footer">
            Powered by <a href="https://jupyterbook.org">Jupyter Book</a>
            </div>
            
        </div>
    </div>
    <div id="rtd-footer-container"></div>
</div>


          


          
<!-- A tiny helper pixel to detect if we've scrolled -->
<div class="sbt-scroll-pixel-helper"></div>
<!-- Main content -->
<div class="col py-0 content-container">
    
    <div class="header-article row sticky-top noprint">
        



<div class="col py-1 d-flex header-article-main">
    <div class="header-article__left">
        
        <label for="__navigation"
  class="headerbtn"
  data-toggle="tooltip"
data-placement="right"
title="Toggle navigation"
>
  

<span class="headerbtn__icon-container">
  <i class="fas fa-bars"></i>
  </span>

</label>

        
    </div>
    <div class="header-article__right">
<div class="menu-dropdown menu-dropdown-launch-buttons">
  <button class="headerbtn menu-dropdown__trigger"
      aria-label="Launch interactive content">
      <i class="fas fa-rocket"></i>
  </button>
  <div class="menu-dropdown__content">
    <ul>
      <li>
        <a href="https://mybinder.org/v2/gh/theislab/multimodal-single-cell-best-practices/master?urlpath=tree/trajectories/lineage_tracing.ipynb"
   class="headerbtn"
   data-toggle="tooltip"
data-placement="left"
title="Launch on Binder"
>
  

<span class="headerbtn__icon-container">
  
    <img src="../_static/images/logo_binder.svg">
  </span>
<span class="headerbtn__text-container">Binder</span>
</a>

      </li>
      
    </ul>
  </div>
</div>

<button onclick="toggleFullScreen()"
  class="headerbtn"
  data-toggle="tooltip"
data-placement="bottom"
title="Fullscreen mode"
>
  

<span class="headerbtn__icon-container">
  <i class="fas fa-expand"></i>
  </span>

</button>

<div class="menu-dropdown menu-dropdown-repository-buttons">
  <button class="headerbtn menu-dropdown__trigger"
      aria-label="Source repositories">
      <i class="fab fa-github"></i>
  </button>
  <div class="menu-dropdown__content">
    <ul>
      <li>
        <a href="https://github.com/theislab/multimodal-single-cell-best-practices"
   class="headerbtn"
   data-toggle="tooltip"
data-placement="left"
title="Source repository"
>
  

<span class="headerbtn__icon-container">
  <i class="fab fa-github"></i>
  </span>
<span class="headerbtn__text-container">repository</span>
</a>

      </li>
      
      <li>
        <a href="https://github.com/theislab/multimodal-single-cell-best-practices/issues/new?title=Issue%20on%20page%20%2Ftrajectories/lineage_tracing.html&body=Your%20issue%20content%20here."
   class="headerbtn"
   data-toggle="tooltip"
data-placement="left"
title="Open an issue"
>
  

<span class="headerbtn__icon-container">
  <i class="fas fa-lightbulb"></i>
  </span>
<span class="headerbtn__text-container">open issue</span>
</a>

      </li>
      
      <li>
        <a href="https://github.com/theislab/multimodal-single-cell-best-practices/edit/master/trajectories/lineage_tracing.ipynb"
   class="headerbtn"
   data-toggle="tooltip"
data-placement="left"
title="Edit this page"
>
  

<span class="headerbtn__icon-container">
  <i class="fas fa-pencil-alt"></i>
  </span>
<span class="headerbtn__text-container">suggest edit</span>
</a>

      </li>
      
    </ul>
  </div>
</div>

<div class="menu-dropdown menu-dropdown-download-buttons">
  <button class="headerbtn menu-dropdown__trigger"
      aria-label="Download this page">
      <i class="fas fa-download"></i>
  </button>
  <div class="menu-dropdown__content">
    <ul>
      <li>
        <a href="../_sources/trajectories/lineage_tracing.ipynb"
   class="headerbtn"
   data-toggle="tooltip"
data-placement="left"
title="Download source file"
>
  

<span class="headerbtn__icon-container">
  <i class="fas fa-file"></i>
  </span>
<span class="headerbtn__text-container">.ipynb</span>
</a>

      </li>
      
      <li>
        
<button onclick="printPdf(this)"
  class="headerbtn"
  data-toggle="tooltip"
data-placement="left"
title="Print to PDF"
>
  

<span class="headerbtn__icon-container">
  <i class="fas fa-file-pdf"></i>
  </span>
<span class="headerbtn__text-container">.pdf</span>
</button>

      </li>
      
    </ul>
  </div>
</div>
<label for="__page-toc"
  class="headerbtn headerbtn-page-toc"
  
>
  

<span class="headerbtn__icon-container">
  <i class="fas fa-list"></i>
  </span>

</label>

    </div>
</div>

<!-- Table of contents -->
<div class="col-md-3 bd-toc show noprint">
    <div class="tocsection onthispage pt-5 pb-3">
        <i class="fas fa-list"></i> Contents
    </div>
    <nav id="bd-toc-nav" aria-label="Page">
        <ul class="visible nav section-nav flex-column">
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#motivation">
   16.1. Motivation
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#lineage-tracing-technologies">
   16.2. Lineage tracing technologies
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#overview-of-evolving-lineage-tracing-data-analysis-pipelines">
   16.3. Overview of evolving lineage tracing data analysis pipelines
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#cassiopeia-for-lineage-tracing-analysis">
   16.4.
   <code class="docutils literal notranslate">
    <span class="pre">
     Cassiopeia
    </span>
   </code>
   for lineage tracing analysis
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#tracing-tumor-development-in-a-mouse-model-of-lung-cancer">
   16.5. Tracing tumor development in a mouse model of lung cancer
  </a>
  <ul class="nav section-nav flex-column">
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#downloading-the-data">
     16.5.1. Downloading the data
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#environment-setup">
     16.5.2. Environment setup.
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#examine-the-data">
     16.5.3. Examine the data
    </a>
    <ul class="nav section-nav flex-column">
     <li class="toc-h4 nav-item toc-entry">
      <a class="reference internal nav-link" href="#number-of-intbcs-per-tumor">
       16.5.3.1. Number of intBCs per Tumor
      </a>
     </li>
     <li class="toc-h4 nav-item toc-entry">
      <a class="reference internal nav-link" href="#size-of-each-tumor">
       16.5.3.2. Size of each tumor
      </a>
     </li>
    </ul>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#preparing-the-data-for-lineage-reconstruction">
     16.5.4. Preparing the data for lineage reconstruction
    </a>
    <ul class="nav section-nav flex-column">
     <li class="toc-h4 nav-item toc-entry">
      <a class="reference internal nav-link" href="#filtering-out-low-quality-tumors">
       16.5.4.1. Filtering out low-quality tumors
      </a>
     </li>
     <li class="toc-h4 nav-item toc-entry">
      <a class="reference internal nav-link" href="#calculate-and-visualize-tumors-statistics">
       16.5.4.2. Calculate and visualize tumors’ statistics
      </a>
     </li>
     <li class="toc-h4 nav-item toc-entry">
      <a class="reference internal nav-link" href="#reconstruction-of-a-selected-tumor-3726-nt-t1">
       16.5.4.3. Reconstruction of a selected tumor (
       <code class="docutils literal notranslate">
        <span class="pre">
         3726_NT_T1
        </span>
       </code>
       )
      </a>
     </li>
    </ul>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#reconstructing-the-lineage">
     16.5.5. Reconstructing the lineage
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#quantifying-dynamic-properties-from-the-tree">
     16.5.6. Quantifying dynamic properties from the tree
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#inferring-expansion-events">
     16.5.7. Inferring expansion events
    </a>
    <ul class="nav section-nav flex-column">
     <li class="toc-h4 nav-item toc-entry">
      <a class="reference internal nav-link" href="#inferring-tree-plasticity">
       16.5.7.1. Inferring tree plasticity
      </a>
     </li>
    </ul>
   </li>
  </ul>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#conclusions">
   16.6. Conclusions
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#new-directions">
   16.7. New directions
  </a>
  <ul class="nav section-nav flex-column">
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#new-phylogenetic-inference-algorithms">
     16.7.1. New phylogenetic inference algorithms
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#computational-tools-to-interpret-time-course-lineage-tracing-data">
     16.7.2. Computational tools to interpret time-course lineage tracing data
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#resources-for-developing-new-computational-methods">
     16.7.3. Resources for developing new computational methods
    </a>
   </li>
  </ul>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#key-takeaways">
   16.8. Key takeaways
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#references">
   16.9. References
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#contributors">
   16.10. Contributors
  </a>
  <ul class="nav section-nav flex-column">
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#authors">
     16.10.1. Authors
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#reviewers">
     16.10.2. Reviewers
    </a>
   </li>
  </ul>
 </li>
</ul>

    </nav>
</div>
    </div>
    <div class="article row">
        <div class="col pl-md-3 pl-lg-5 content-container">
            <!-- Table of contents that is only displayed when printing the page -->
            <div id="jb-print-docs-body" class="onlyprint">
                <h1>Lineage tracing</h1>
                <!-- Table of contents -->
                <div id="print-main-content">
                    <div id="jb-print-toc">
                        
                        <div>
                            <h2> Contents </h2>
                        </div>
                        <nav aria-label="Page">
                            <ul class="visible nav section-nav flex-column">
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#motivation">
   16.1. Motivation
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#lineage-tracing-technologies">
   16.2. Lineage tracing technologies
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#overview-of-evolving-lineage-tracing-data-analysis-pipelines">
   16.3. Overview of evolving lineage tracing data analysis pipelines
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#cassiopeia-for-lineage-tracing-analysis">
   16.4.
   <code class="docutils literal notranslate">
    <span class="pre">
     Cassiopeia
    </span>
   </code>
   for lineage tracing analysis
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#tracing-tumor-development-in-a-mouse-model-of-lung-cancer">
   16.5. Tracing tumor development in a mouse model of lung cancer
  </a>
  <ul class="nav section-nav flex-column">
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#downloading-the-data">
     16.5.1. Downloading the data
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#environment-setup">
     16.5.2. Environment setup.
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#examine-the-data">
     16.5.3. Examine the data
    </a>
    <ul class="nav section-nav flex-column">
     <li class="toc-h4 nav-item toc-entry">
      <a class="reference internal nav-link" href="#number-of-intbcs-per-tumor">
       16.5.3.1. Number of intBCs per Tumor
      </a>
     </li>
     <li class="toc-h4 nav-item toc-entry">
      <a class="reference internal nav-link" href="#size-of-each-tumor">
       16.5.3.2. Size of each tumor
      </a>
     </li>
    </ul>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#preparing-the-data-for-lineage-reconstruction">
     16.5.4. Preparing the data for lineage reconstruction
    </a>
    <ul class="nav section-nav flex-column">
     <li class="toc-h4 nav-item toc-entry">
      <a class="reference internal nav-link" href="#filtering-out-low-quality-tumors">
       16.5.4.1. Filtering out low-quality tumors
      </a>
     </li>
     <li class="toc-h4 nav-item toc-entry">
      <a class="reference internal nav-link" href="#calculate-and-visualize-tumors-statistics">
       16.5.4.2. Calculate and visualize tumors’ statistics
      </a>
     </li>
     <li class="toc-h4 nav-item toc-entry">
      <a class="reference internal nav-link" href="#reconstruction-of-a-selected-tumor-3726-nt-t1">
       16.5.4.3. Reconstruction of a selected tumor (
       <code class="docutils literal notranslate">
        <span class="pre">
         3726_NT_T1
        </span>
       </code>
       )
      </a>
     </li>
    </ul>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#reconstructing-the-lineage">
     16.5.5. Reconstructing the lineage
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#quantifying-dynamic-properties-from-the-tree">
     16.5.6. Quantifying dynamic properties from the tree
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#inferring-expansion-events">
     16.5.7. Inferring expansion events
    </a>
    <ul class="nav section-nav flex-column">
     <li class="toc-h4 nav-item toc-entry">
      <a class="reference internal nav-link" href="#inferring-tree-plasticity">
       16.5.7.1. Inferring tree plasticity
      </a>
     </li>
    </ul>
   </li>
  </ul>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#conclusions">
   16.6. Conclusions
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#new-directions">
   16.7. New directions
  </a>
  <ul class="nav section-nav flex-column">
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#new-phylogenetic-inference-algorithms">
     16.7.1. New phylogenetic inference algorithms
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#computational-tools-to-interpret-time-course-lineage-tracing-data">
     16.7.2. Computational tools to interpret time-course lineage tracing data
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#resources-for-developing-new-computational-methods">
     16.7.3. Resources for developing new computational methods
    </a>
   </li>
  </ul>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#key-takeaways">
   16.8. Key takeaways
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#references">
   16.9. References
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#contributors">
   16.10. Contributors
  </a>
  <ul class="nav section-nav flex-column">
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#authors">
     16.10.1. Authors
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#reviewers">
     16.10.2. Reviewers
    </a>
   </li>
  </ul>
 </li>
</ul>

                        </nav>
                    </div>
                </div>
            </div>
            <main id="main-content" role="main">
                
              <div>
                
  <section class="tex2jax_ignore mathjax_ignore" id="lineage-tracing">
<h1><span class="section-number">16. </span>Lineage tracing<a class="headerlink" href="#lineage-tracing" title="Permalink to this headline">#</a></h1>
<p><em>TL;DR we provide a brief overview assays providing measurments of both cell state and lineage history and on the available computational pipelines using a leading example, tracing tumor development in a mouse model of lung cancer.</em></p>
<section id="motivation">
<h2><span class="section-number">16.1. </span>Motivation<a class="headerlink" href="#motivation" title="Permalink to this headline">#</a></h2>
<p>Cellular lineages are ubiquitious in biology. Perhaps the most famous example is that of embyrogenesis: the process by which an organism like a human being is generated from from a single cell, the fertilized egg. During this process, subsequent cell divisions give rise to daughter cells and over time entire “lineages” that take on specialized roles within the developing embryo. The amazing complexity of this process has captured the imagination of scientists for centuries, and over the past century and a half our understanding of this process has been bolstered by the development of high-throughput sequencing assays and new “lineage tracing” technologies for visualizing and characterizing this process <span id="id1">[<a class="reference internal" href="#id236" title="Mollie B Woodworth, Kelly M Girskis, and Christopher A Walsh. Building a lineage from single cells: genetic techniques for cell lineage tracking. Nature Reviews Genetics, 18(4):230–244, 2017.">Woodworth <em>et al.</em>, 2017</a>]</span>. Amongst the most exciting of these methods allow investigators to link measurements of cell state with models of their history, thus providing a window into how differentiation trajectories might have unfolded.</p>
<p>The marriage of single-cell assays and lineage tracing approaches has yielded an exponential growth in the complexity of datasets, requiring the development of new computational methodology for their analysis. As such, there has been a strong need in developing new computational methodology for processing these datasets <span id="id2">[<a class="reference internal" href="#id237" title="Wuming Gong, Alejandro A Granados, Jingyuan Hu, Matthew G Jones, Ofir Raz, Irepan Salvador-Martínez, Hanrui Zhang, Ke-Huan K Chow, Il-Youp Kwak, Renata Retkute, and others. Benchmarked approaches for reconstruction of in vitro cell lineages and in silico models of c. elegans and m. musculus developmental trees. Cell systems, 12(8):810–826, 2021.">Gong <em>et al.</em>, 2021</a>]</span>. Sourcing heavily from population genetics literature, the past half decade has witnessed an exciting confluence of traditional concepts in evolutionary biology with cutting-edge genome engineering techniques.</p>
<p>In this chapter, we provide a brief overview of these new technologies and focus on the available computational pipelines for the analysis of their output and extraction of biological insight. To note, we pay special focus with our example to the CRISPR/Cas9-based “evolving” lineage tracing setting. For a more complete view of other useful experimental alternatives, we refer the interested reader to the excellent reviews by McKenna &amp; Gagnon <span id="id3">[<a class="reference internal" href="#id239" title="Aaron McKenna and James A Gagnon. Recording development with single cell dynamic lineage tracing. Development, 146(12):dev169730, 2019.">McKenna and Gagnon, 2019</a>]</span>, Wagner &amp; Klein <span id="id4">[<a class="reference internal" href="#id238" title="Daniel E Wagner and Allon M Klein. Lineage tracing meets single-cell omics: opportunities and challenges. Nature Reviews Genetics, 21(7):410–427, 2020.">Wagner and Klein, 2020</a>]</span>, and VanHorn &amp; Morris <span id="id5">[<a class="reference internal" href="#id240" title="Sadie VanHorn and Samantha A Morris. Next-generation lineage tracing and fate mapping to interrogate development. Developmental cell, 56(1):7–21, 2021.">VanHorn and Morris, 2021</a>]</span>.</p>
</section>
<section id="lineage-tracing-technologies">
<h2><span class="section-number">16.2. </span>Lineage tracing technologies<a class="headerlink" href="#lineage-tracing-technologies" title="Permalink to this headline">#</a></h2>
<p>The goal of lineage tracing techniques is to infer lineage, or ancestry relationships between observed cells. In this, there are two major variables to consider: scale and resolution. Classical approaches relied heavily on visual observation: for example, in the 1970s Sulston and colleagues derived the first developmental lineage of the nematode <em>C. elegans</em> by meticulously watching cell divisions under a microscope <span id="id6">[<a class="reference internal" href="#id241" title="John E Sulston, Einhard Schierenberg, John G White, and J Nichol Thomson. The embryonic cell lineage of the nematode caenorhabditis elegans. Developmental biology, 100(1):64–119, 1983.">Sulston <em>et al.</em>, 1983</a>]</span>. While playing an imperative role in the progression of the field, such approaches cannot scale to complex organisms with more stochastic developmental lineages.</p>
<p>Over the past two decades, the development of revolutionary sequencing assays and microfluidic devices has contributed to the development of new and diverse lineage tracing methodology <span id="id7">[<a class="reference internal" href="#id238" title="Daniel E Wagner and Allon M Klein. Lineage tracing meets single-cell omics: opportunities and challenges. Nature Reviews Genetics, 21(7):410–427, 2020.">Wagner and Klein, 2020</a>]</span>. To digest the plethora of techniques, it is helpful to classify approaches as “<em>prospective</em>” or “<em>retrospective</em>”:</p>
<ul class="simple">
<li><p><strong>Prospective lineage tracing</strong> approaches enable investigators to trace the descendants of a single cell (i.e., a “<em>clone”</em> or “<em>clonal population</em>”). Typically, this is done by introducing a heritable marker into a cell (termed a “<em>clonal progenitor</em>”) that is passed on from generation to generation.</p></li>
<li><p><strong>Retrospective lineage tracing</strong> approaches use variability observed in cells - such as naturally occurring genetic mutations - to infer a model of their lineage (or “<em>phylogeny</em>”) summarizing the cell division history in a clonal population.</p></li>
</ul>
<p>Several approaches have been developed for prospectively tracing a clonal population: for example, recombinases under a tissue-specific promoter can be used to activate fluorescent markers that act as heritable marks for a specific tissue lineage <span id="id8">[<a class="reference internal" href="#id242" title="Tamily A Weissman and Y Albert Pan. Brainbow: new resources and emerging biological applications for multicolor genetic labeling and analysis. Genetics, 199(2):293–306, 2015.">Weissman and Pan, 2015</a>]</span>, <span id="id9">[<a class="reference internal" href="#id243" title="Andras Nagy. Cre recombinase: the universal reagent for genome tailoring. genesis, 26(2):99–109, 2000.">Nagy, 2000</a>]</span>, <span id="id10">[<a class="reference internal" href="#id244" title="Kuo Liu, Hengwei Jin, and Bin Zhou. Genetic lineage tracing with multiple term`dna` recombinases: a user's guide for conducting more precise cell fate mapping studies. Journal of Biological Chemistry, 295(19):6413–6424, 2020.">Liu <em>et al.</em>, 2020</a>]</span>, <span id="id11">[<a class="reference internal" href="#id245" title="Kuo Liu, Muxue Tang, Hengwei Jin, Qiaozhen Liu, Lingjuan He, Huan Zhu, Xiuxiu Liu, Ximeng Han, Yan Li, Libo Zhang, and others. Triple-cell lineage tracing by a dual reporter on a single allele. Journal of Biological Chemistry, 295(3):690–700, 2020.">Liu <em>et al.</em>, 2020</a>]</span>, <span id="id12">[<a class="reference internal" href="#id246" title="Lingjuan He, Yan Li, Yi Li, Wenjuan Pu, Xiuzhen Huang, Xueying Tian, Yue Wang, Hui Zhang, Qiaozhen Liu, Libo Zhang, and others. Enhancing the precision of genetic lineage tracing using dual recombinases. Nature medicine, 23(12):1488–1498, 2017.">He <em>et al.</em>, 2017</a>]</span>. Alternatively, lentiviral transduction can be used to integrate random DNA barcodes into cellular genomes to provide a heritable mark that can be used to deconvolve clonal identities with a sequencing readout <span id="id13">[<a class="reference internal" href="#id247" title="Alice Gerrits, Brad Dykstra, Olga J Kalmykowa, Karin Klauke, Evgenia Verovskaya, Mathilde JC Broekhuis, Gerald de Haan, and Leonid V Bystrykh. Cellular barcoding tool for clonal analysis in the hematopoietic system. Blood, The Journal of the American Society of Hematology, 115(13):2610–2618, 2010.">Gerrits <em>et al.</em>, 2010</a>]</span>, <span id="id14">[<a class="reference internal" href="#id248" title="Brent A Biddy, Wenjun Kong, Kenji Kamimoto, Chuner Guo, Sarah E Waye, Tao Sun, and Samantha A Morris. Single-cell mapping of lineage and identity in direct reprogramming. Nature, 564(7735):219–224, 2018.">Biddy <em>et al.</em>, 2018</a>]</span>, <span id="id15">[<a class="reference internal" href="#id249" title="Caleb Weinreb, Alejo Rodriguez-Fraticelli, Fernando D Camargo, and Allon M Klein. Lineage tracing on transcriptional landscapes links state to fate during differentiation. Science, 367(6479):eaaw3381, 2020.">Weinreb <em>et al.</em>, 2020</a>]</span>, <span id="id16">[<a class="reference internal" href="#id250" title="Zizhen Yao, John K Mich, Sherman Ku, Vilas Menon, Anne-Rachel Krostag, Refugio A Martinez, Leon Furchtgott, Heather Mulholland, Susan Bort, Margaret A Fuqua, and others. A single-cell roadmap of lineage bifurcation in human esc models of embryonic brain development. Cell stem cell, 20(1):120–134, 2017.">Yao <em>et al.</em>, 2017</a>]</span>. Though these approaches are highly scalable and often do not require heavy genome engineering, they can only report properties at the clonal-level such as clone size and composition.</p>
<p>Retrospective lineage tracers overcome these limitations and provide an additional advantage over prospective tracers by reporting properties about the <em>sub</em>clonal dynamics of clones. Traditionally, this has been done by leveraging natural variation between cells to reconstruct a cell division history, such as single-nucleotide variation <span id="id17">[<a class="reference internal" href="#id251" title="Bert Vogelstein, Nickolas Papadopoulos, Victor E Velculescu, Shibin Zhou, Luis A Diaz Jr, and Kenneth W Kinzler. Cancer genome landscapes. science, 339(6127):1546–1558, 2013.">Vogelstein <em>et al.</em>, 2013</a>]</span>, <span id="id18">[<a class="reference internal" href="#id252" title="Samra Turajlic and Charles Swanton. Metastasis as an evolutionary process. Science, 352(6282):169–175, 2016.">Turajlic and Swanton, 2016</a>]</span>, <span id="id19">[<a class="reference internal" href="#id254" title="Chris Bailey, James RM Black, James L Reading, Kevin Litchfield, Samra Turajlic, Nicholas McGranahan, Mariam Jamal-Hanjani, and Charles Swanton. Tracking cancer evolution through the disease course. Cancer discovery, 11(4):916–932, 2021.">Bailey <em>et al.</em>, 2021</a>]</span>, <span id="id20">[<a class="reference internal" href="#id255" title="Moritz Gerstung, Clemency Jolly, Ignaty Leshchiner, Stefan C Dentro, Santiago Gonzalez, Daniel Rosebrock, Thomas J Mitchell, Yulia Rubanova, Pavana Anur, Kaixian Yu, and others. The evolutionary history of 2,658 cancers. Nature, 578(7793):122–128, 2020.">Gerstung <em>et al.</em>, 2020</a>]</span>, <span id="id21">[<a class="reference internal" href="#id256" title="Alexej Abyzov and Flora M Vaccarino. Cell lineage tracing and cellular diversity in humans. Annual Review of Genomics and Human Genetics, 21:101–116, 2020.">Abyzov and Vaccarino, 2020</a>]</span>, <span id="id22">[<a class="reference internal" href="#id276" title="Sara Bizzotto, Yanmei Dou, Javier Ganz, Ryan N Doan, Minseok Kwon, Craig L Bohrson, Sonia N Kim, Taejeong Bae, Alexej Abyzov, NIMH Brain Somatic Mosaicism Network†, and others. Landmarks of human embryonic development inscribed in somatic mutations. Science, 371(6535):1249–1253, 2021.">Bizzotto <em>et al.</em>, 2021</a>]</span>, <span id="id23">[<a class="reference internal" href="#id277" title="Young Seok Ju, Inigo Martincorena, Moritz Gerstung, Mia Petljak, Ludmil B Alexandrov, Raheleh Rahbari, David C Wedge, Helen R Davies, Manasa Ramakrishna, Anthony Fullam, and others. Somatic mutations reveal asymmetric cellular dynamics in the early human embryo. Nature, 543(7647):714–718, 2017.">Ju <em>et al.</em>, 2017</a>]</span> or copy-number-variation <span id="id24">[<a class="reference internal" href="#id275" title="Anoop P Patel, Itay Tirosh, John J Trombetta, Alex K Shalek, Shawn M Gillespie, Hiroaki Wakimoto, Daniel P Cahill, Brian V Nahed, William T Curry, Robert L Martuza, and others. Single-cell rna-seq highlights intratumoral heterogeneity in primary glioblastoma. Science, 344(6190):1396–1401, 2014.">Patel <em>et al.</em>, 2014</a>]</span>, <span id="id25">[<a class="reference internal" href="#id274" title="Ruli Gao, Shanshan Bai, Ying C Henderson, Yiyun Lin, Aislyn Schalck, Yun Yan, Tapsi Kumar, Min Hu, Emi Sei, Alexander Davis, and others. Delineating copy number and clonal substructure in human tumors from single-cell transcriptomes. Nature biotechnology, 39(5):599–608, 2021.">Gao <em>et al.</em>, 2021</a>]</span>. While this approach is still widely and successfully used to study human tumors or tissue developmental histories, experimentalists have little or no control over how often or where mutations occur. In experimental models, there are opportunities to recapitulate the advantages of retrospective tracers while improving on the caveats by engineering evolvable lineage tracers. Such evolving tracers typically consist of engineering cells with a “scratchpad” (or, synonymously, “target site”) that can acquire mutations <span id="id26">[<a class="reference internal" href="#id238" title="Daniel E Wagner and Allon M Klein. Lineage tracing meets single-cell omics: opportunities and challenges. Nature Reviews Genetics, 21(7):410–427, 2020.">Wagner and Klein, 2020</a>]</span>, <span id="id27">[<a class="reference internal" href="#id239" title="Aaron McKenna and James A Gagnon. Recording development with single cell dynamic lineage tracing. Development, 146(12):dev169730, 2019.">McKenna and Gagnon, 2019</a>]</span>. For example, a popular approach that this chapter focuses on uses Cas9 to introduce insertions and deletions (i.e., “indels”) at the target site <span id="id28">[<a class="reference internal" href="#id257" title="Aaron McKenna, Gregory M Findlay, James A Gagnon, Marshall S Horwitz, Alexander F Schier, and Jay Shendure. Whole-organism lineage tracing by combinatorial and cumulative genome editing. Science, 353(6298):aaf7907, 2016.">McKenna <em>et al.</em>, 2016</a>]</span>, <span id="id29">[<a class="reference internal" href="#id259" title="Bastiaan Spanjaard, Bo Hu, Nina Mitic, Pedro Olivares-Chauvet, Sharan Janjuha, Nikolay Ninov, and Jan Philipp Junker. Simultaneous lineage tracing and cell-type identification using crispr–cas9-induced genetic scars. Nature biotechnology, 36(5):469–473, 2018.">Spanjaard <em>et al.</em>, 2018</a>]</span>, <span id="id30">[<a class="reference internal" href="#id260" title="Michelle M Chan, Zachary D Smith, Stefanie Grosswendt, Helene Kretzmer, Thomas M Norman, Britt Adamson, Marco Jost, Jeffrey J Quinn, Dian Yang, Matthew G Jones, and others. Molecular recording of mammalian embryogenesis. Nature, 570(7759):77–82, 2019.">Chan <em>et al.</em>, 2019</a>]</span>, <span id="id31">[<a class="reference internal" href="#id261" title="Kirsten L Frieda, James M Linton, Sahand Hormoz, Joonhyuk Choi, Ke-Huan K Chow, Zakary S Singer, Mark W Budde, Michael B Elowitz, and Long Cai. Synthetic recording and in situ readout of lineage information in single cells. Nature, 541(7635):107–111, 2017.">Frieda <em>et al.</em>, 2017</a>]</span>, <span id="id32">[<a class="reference internal" href="#id262" title="Reza Kalhor, Kian Kalhor, Leo Mejia, Kathleen Leeper, Amanda Graveline, Prashant Mali, and George M Church. Developmental barcoding of whole mouse via homing crispr. Science, 361(6405):eaat9804, 2018.">Kalhor <em>et al.</em>, 2018</a>]</span>, <span id="id33">[<a class="reference internal" href="#id263" title="Anna Alemany, Maria Florescu, Chloé S Baron, Josi Peterson-Maduro, and Alexander Van Oudenaarden. Whole-organism clone tracing using single-cell sequencing. Nature, 556(7699):108–112, 2018.">Alemany <em>et al.</em>, 2018</a>]</span>. In this way, cellular lineages acquire heritable mutations over time that can be subsequently read out with high-throughput sequencing platforms and used to infer phylogenies representing a model of the cell lineage. Building on this work, additional technologies have emerged that improve the interpretability of this data: for example, approaches like peCHYRON <span id="id34">[<a class="reference internal" href="#id278" title="Theresa B. Loveless, Joseph H. Grotts, Mason W. Schechter, Elmira Forouzmand, Courtney K. Carlson, Bijan S. Agahi, Guohao Liang, Michelle Ficht, Beide Liu, Xiaohui Xie, and Chang C. Liu. Lineage tracing and analog recording in mammalian cells by single-site term`dna` writing. Nature Chemical Biology, 17(6):739-747, Jun 2021. URL: https://doi.org/10.1038/s41589-021-00769-8, doi:10.1038/s41589-021-00769-8.">Loveless <em>et al.</em>, 2021</a>]</span> and DNA Typewriter <span id="id35">[<a class="reference internal" href="#id279" title="Junhong Choi, Wei Chen, Anna Minkina, Florence M. Chardon, Chase C. Suiter, Samuel G. Regalado, Silvia Domcke, Nobuhiko Hamazaki, Choli Lee, Beth Martin, Riza M. Daza, and Jay Shendure. A time-resolved, multi-symbol molecular recorder via sequential genome editing. Nature, 608(7921):98-107, Aug 2022. URL: https://doi.org/10.1038/s41586-022-04922-8, doi:10.1038/s41586-022-04922-8.">Choi <em>et al.</em>, 2022</a>]</span> introduce ordered, sequential edits to allow for more confident assessment of lineage histories.</p>
<p>Both classes of lineage-tracing approaches can take advantage of adjacent advances in single-cell multiomic profiling. For example, investigators have routinely used single-cell RNA-seq (scRNA-seq) to read out simultaneously the functional state of single cells and their lineage relationships <span id="id36">[<a class="reference internal" href="#id258" title="Bushra Raj, James A Gagnon, and Alexander F Schier. Large-scale reconstruction of cell lineages using single-cell readout of transcriptomes and crispr–cas9 barcodes by scgestalt. Nature protocols, 13(11):2685–2713, 2018.">Raj <em>et al.</em>, 2018</a>]</span>, <span id="id37">[<a class="reference internal" href="#id260" title="Michelle M Chan, Zachary D Smith, Stefanie Grosswendt, Helene Kretzmer, Thomas M Norman, Britt Adamson, Marco Jost, Jeffrey J Quinn, Dian Yang, Matthew G Jones, and others. Molecular recording of mammalian embryogenesis. Nature, 570(7759):77–82, 2019.">Chan <em>et al.</em>, 2019</a>]</span>, <span id="id38">[<a class="reference internal" href="#id249" title="Caleb Weinreb, Alejo Rodriguez-Fraticelli, Fernando D Camargo, and Allon M Klein. Lineage tracing on transcriptional landscapes links state to fate during differentiation. Science, 367(6479):eaaw3381, 2020.">Weinreb <em>et al.</em>, 2020</a>]</span>, <span id="id39">[<a class="reference internal" href="#id238" title="Daniel E Wagner and Allon M Klein. Lineage tracing meets single-cell omics: opportunities and challenges. Nature Reviews Genetics, 21(7):410–427, 2020.">Wagner and Klein, 2020</a>]</span>, <span id="id40">[<a class="reference internal" href="#id259" title="Bastiaan Spanjaard, Bo Hu, Nina Mitic, Pedro Olivares-Chauvet, Sharan Janjuha, Nikolay Ninov, and Jan Philipp Junker. Simultaneous lineage tracing and cell-type identification using crispr–cas9-induced genetic scars. Nature biotechnology, 36(5):469–473, 2018.">Spanjaard <em>et al.</em>, 2018</a>]</span>. This multimodal readout has created opportunities for new computational methodologies, which we detail below.</p>
<p>As stated above, in this chapter we provide a <strong>detailed walkthrough on the analysis of data from evolving CRISPR/Cas9-based lineage tracers.</strong></p>
<div class="alert alert-block alert-warning" style="color:black;">
<b>Important Definitions</b> 
<br/br>
    We've used a few terms above that can be confusing to first-time readers. Briefly, we offer some definitions before moving on:
<ul>
    <li> <b>clonal population (or, clone)</b>: The full set of descendants of a specific cell progenitor.</li>
    <li> <b>clonal progenitor</b>: The original single-cell giving rise to a clonal population.</li>
    <li> <b>subclonal resolution</b>: Insights into the relationships between subsets of cells within a clonal population. </li>
    <li> <b>phylogeny</b>: A model of the cell division history of a clonal population, represented as a tree.</li>
    <li> <b>scratchpad (or, target site)</b>: A synthetic, exogeneous region capable of accumulating targeted variation in evolving lineage tracing technologies.</li>
</ul>
</div></section>
<section id="overview-of-evolving-lineage-tracing-data-analysis-pipelines">
<h2><span class="section-number">16.3. </span>Overview of evolving lineage tracing data analysis pipelines<a class="headerlink" href="#overview-of-evolving-lineage-tracing-data-analysis-pipelines" title="Permalink to this headline">#</a></h2>
<p>Before delving into the analysis of our example dataset, we will provide an overview of the computational pipeline for the analysis of data generated by evolving tracers (based on <span id="id41">[<a class="reference internal" href="#id264" title="Matthew G Jones, Alex Khodaverdian, Jeffrey J Quinn, Michelle M Chan, Jeffrey A Hussmann, Robert Wang, Chenling Xu, Jonathan S Weissman, and Nir Yosef. Inference of single-cell phylogenies from lineage tracing data using cassiopeia. Genome biology, 21(1):1–27, 2020.">Jones <em>et al.</em>, 2020</a>]</span>). In general, with these systems, analysis will begin with raw sequencing data of an <em>amplicon</em> library of target sites (often derived from a conventional scRNA-seq platform like 10X Chromium). Depending on the technology at hand, each sequenced amplicon will be between 150-300bp long; in the case of CRISPR/Cas9-based evolving tracers, each read will contain one more Cas9 cut sites. Within the preprocessing of this data, analysts are tasked with aligning the reads to a reference sequence and identifying any mutations (e.g., indels).</p>
<p>While the preprocessing of these datasets is a critical step, in the interest of space, we focus on analysis pipelines that receive preprocessed sequencing data as input and refer the reader to an external preprocessing tutorial which can be found <a class="reference external" href="https://cassiopeia-lineage.readthedocs.io/en/latest/notebooks/preprocess.html">here</a>.</p>
<p><img alt="SegmentLocal" src="../_images/CRISPR-Cas9-based.gif" /></p>
<p>In most analysis frameworks, the preprocessing of the raw sequencing reads produces a data structure called the <strong>character matrix</strong> that summarizes the observed mutations in each cell across the target sites. In this data structure, each row is a cell (or “sample”), each column is a target site (or “character”), and the (row, column) values are categorical variables representing the identity of the indel observed in that cell at that particular cut site (or “character-state”). Depending on the technology at hand, these character matrices can report on anywhere between 100 and 10,000 samples across up to 100 characters.</p>
<p>At this point, this data structure abstracts away the technicalities of the evolving lineage tracing assay and opens up the opportunity to computationally infer a <strong>phylogenetic tree</strong> over the cells. Specifically, the goal is to learn a hierarchical tree structure over each of the cells in our character matrix. In this tree, each node represents a sample and each edge represents a lineage relationship. Importantly, we often have only observed the <em>leaves</em> of the tree and we refer to any of the unobserved set of internal nodes as <em>ancestral</em> nodes. (To note, we tend to use the term “phylogenetic tree” loosely as this has a precise definition. In reality, we often infer a cladogram summarizing the relationships between cells.)</p>
<p>There are many algorithmic choices for inferring the phylogenetic tree from the character matrix, which can be generally broken up into “character-based” and “distance-based” approaches:</p>
<ul class="simple">
<li><p><strong>Character-based</strong>: perform a combinatorial search through all possible tree topologies while seeking to optimize a function over the characters (e.g., the likelihood of the evolutionary history given the mutations observed in the characters).</p>
<ul>
<li><p>Maximum Parsimony <span id="id42">[<a class="reference internal" href="#id235" title="LL Cavalli-Sforza and AWF Edwards. The reconstruction of evolution. Ann. Hum. Genet, 27:105–106, 1963.">Cavalli-Sforza and Edwards, 1963</a>]</span>: Find a tree with the minimum number of mutations.</p></li>
<li><p>Maximum Likelihood <span id="id43">[<a class="reference internal" href="#id234" title="Joseph Felsenstein. Evolutionary trees from term`dna` sequences: a maximum likelihood approach. Journal of molecular evolution, 17(6):368–376, 1981.">Felsenstein, 1981</a>]</span>: Find a tree with the most likely mutation history (for a lineage tracing-specific algorithm, we refer the reader to GAPML<span id="id44">[<a class="reference internal" href="#id228" title="Jean Feng, William S DeWitt III, Aaron McKenna, Noah Simon, Amy D Willis, and Frederick A Matsen IV. Estimation of cell lineage trees by maximum-likelihood phylogenetics. The Annals of Applied Statistics, 15(1):343–362, 2021.">Feng <em>et al.</em>, 2021</a>]</span>).</p></li>
<li><p>Bayesian Phylogenetic Inference <span id="id45">[<a class="reference internal" href="#id233" title="John P Huelsenbeck, Fredrik Ronquist, Rasmus Nielsen, and Jonathan P Bollback. Bayesian inference of phylogeny and its impact on evolutionary biology. science, 294(5550):2310–2314, 2001.">Huelsenbeck <em>et al.</em>, 2001</a>]</span>: Find a tree that maximizes the posterior probability of the evolutionary history given observed mutations.</p></li>
</ul>
</li>
<li><p><strong>Distance-based</strong>: use a notion of cell-cell distances (such as the number of edits that they share, denoted by <span class="math notranslate nohighlight">\(\delta\)</span>) to infer a phylogenetic tree and typically run in polynomial time. While distance-based approaches can perform much faster, they require one to iteratively find the best cell-cell dissimilarity function which can be equally time consuming.</p>
<ul>
<li><p>Neighbor-Joining <span id="id46">[<a class="reference internal" href="#id232" title="Naruya Saitou and Masatoshi Nei. The neighbor-joining method: a new method for reconstructing phylogenetic trees. Molecular biology and evolution, 4(4):406–425, 1987.">Saitou and Nei, 1987</a>]</span>: Produces a tree from a given cell-cell dissimilarity matrix by iteratively finding pairs that minimize a branch length according to specific criteria.</p></li>
<li><p>UPGMA <span id="id47">[<a class="reference internal" href="#id231" title="Robert R Sokal. A statistical method for evaluating systematic relationships. Univ. Kansas, Sci. Bull., 38:1409–1438, 1958.">Sokal, 1958</a>]</span>: Produces a tree from a given cell-cell dissimilarity matrix using a faster algorithm than Neighbor-Joining. However, it has more strict requirements to produce an accurate tree.</p></li>
</ul>
</li>
</ul>
<p>Traditionally, the task of inferring phylogenies from this type of data is challenging and often requires the application of combinatorial algorithms with poor scalability. In applying such algorithms to the single-cell evolving lineage-tracers previously described, these problems become exacerbated as the number of cells sampled is routinely an order or magnitude larger than the number of species assessed in traditional phylogenetic studies. These challenges are further compounded with non-random missing data and non-uniform editing distributions. Fortunately, there have been recent algorithmic developments for addressing the modeling challenges for evolving lineage tracers <span id="id48">[<a class="reference internal" href="#id264" title="Matthew G Jones, Alex Khodaverdian, Jeffrey J Quinn, Michelle M Chan, Jeffrey A Hussmann, Robert Wang, Chenling Xu, Jonathan S Weissman, and Nir Yosef. Inference of single-cell phylogenies from lineage tracing data using cassiopeia. Genome biology, 21(1):1–27, 2020.">Jones <em>et al.</em>, 2020</a>]</span>, <span id="id49">[<a class="reference internal" href="#id228" title="Jean Feng, William S DeWitt III, Aaron McKenna, Noah Simon, Amy D Willis, and Frederick A Matsen IV. Estimation of cell lineage trees by maximum-likelihood phylogenetics. The Annals of Applied Statistics, 15(1):343–362, 2021.">Feng <em>et al.</em>, 2021</a>]</span>, <span id="id50">[<a class="reference internal" href="#id237" title="Wuming Gong, Alejandro A Granados, Jingyuan Hu, Matthew G Jones, Ofir Raz, Irepan Salvador-Martínez, Hanrui Zhang, Ke-Huan K Chow, Il-Youp Kwak, Renata Retkute, and others. Benchmarked approaches for reconstruction of in vitro cell lineages and in silico models of c. elegans and m. musculus developmental trees. Cell systems, 12(8):810–826, 2021.">Gong <em>et al.</em>, 2021</a>]</span> and for scaling inference to extremely large trees with deep distributed computing <span id="id51">[<a class="reference internal" href="#id227" title="Naoki Konno, Yusuke Kijima, Keito Watano, Soh Ishiguro, Keiichiro Ono, Mamoru Tanaka, Hideto Mori, Nanami Masuyama, Dexter Pratt, Trey Ideker, and others. Deep distributed computing to reconstruct extremely large lineage trees. Nature Biotechnology, 40(4):566–575, 2022.">Konno <em>et al.</em>, 2022</a>]</span>. In practice, the majority of applications have used variants of maximum-parsimony approaches or distance-based approaches like Neighbor-Joining for their tree inference. However, this is currently an active area of development and for a more in-depth discussion, we refer the reader to the Conclusions.</p>
<p>Following phylogenetic tree reconstruction, there are several options for <em>downstream analysis</em>. For example, one can learn about the rates of cell state changes across the developmental history or the relative propensities of cells to divide in a population. Below, we will demonstrate via code examples how these different components fit together to gain fundamental insights into the dynamic processes underlying cellular lineages.</p>
</section>
<section id="cassiopeia-for-lineage-tracing-analysis">
<h2><span class="section-number">16.4. </span><code class="docutils literal notranslate"><span class="pre">Cassiopeia</span></code> for lineage tracing analysis<a class="headerlink" href="#cassiopeia-for-lineage-tracing-analysis" title="Permalink to this headline">#</a></h2>
<p>In this tutorial, we will primarily make use of <code class="docutils literal notranslate"><span class="pre">Cassiopeia</span></code>, one of the few packages for lineage tracing analysis <span id="id52">[<a class="reference internal" href="#id264" title="Matthew G Jones, Alex Khodaverdian, Jeffrey J Quinn, Michelle M Chan, Jeffrey A Hussmann, Robert Wang, Chenling Xu, Jonathan S Weissman, and Nir Yosef. Inference of single-cell phylogenies from lineage tracing data using cassiopeia. Genome biology, 21(1):1–27, 2020.">Jones <em>et al.</em>, 2020</a>]</span>.</p>
<p>Generally, <code class="docutils literal notranslate"><span class="pre">Cassiopeia</span></code> is a software suite for processing and analyzing evolving lineage tracing sequencing data. The software contains four major modules:</p>
<ol class="simple">
<li><p>preprocessing sequencing reads from evolving lineage tracing datasets (e.g., CRISPR/Cas9-based lineage tracers) into character matrices.</p></li>
<li><p>reconstructing phylogenies from these character matrices with one of several algorithms available in the codebase</p></li>
<li><p>providing analysis tools for deriving insights from reconstructed phylogenies.</p></li>
<li><p>simulating realistic phylogenies and lineage tracing data for benchmarking purposes</p></li>
</ol>
<p>Though these modules fit together in the <code class="docutils literal notranslate"><span class="pre">Cassiopeia</span></code> pipeline, they are additionally independent of one another. For example, a user might use a different software suite for lineage inference but use tools in the <code class="docutils literal notranslate"><span class="pre">Cassiopeia</span></code> codebase for post-reconstruction analysis.</p>
<p>For more installation guidelines and additional documentation, we refer the reader <a class="reference external" href="https://cassiopeia-lineage.readthedocs.io/en/latest/index.html">here</a>.</p>
</section>
<section id="tracing-tumor-development-in-a-mouse-model-of-lung-cancer">
<h2><span class="section-number">16.5. </span>Tracing tumor development in a mouse model of lung cancer<a class="headerlink" href="#tracing-tumor-development-in-a-mouse-model-of-lung-cancer" title="Permalink to this headline">#</a></h2>
<p>In this case study, we will make use of the recent study presented in <span id="id53">[<a class="reference internal" href="#id265" title="Dian Yang, Matthew G Jones, Santiago Naranjo, William M Rideout III, Kyung Hoi Joseph Min, Raymond Ho, Wei Wu, Joseph M Replogle, Jennifer L Page, Jeffrey J Quinn, and others. Lineage tracing reveals the phylodynamics, plasticity, and paths of tumor evolution. Cell, 185(11):1905–1923, 2022.">Yang <em>et al.</em>, 2022</a>]</span>. Briefly, in this study the authors integrated a evolving CRISPR/Cas9-based lineage tracer into the KP mouse model of non-small-cell lung cancer <span id="id54">[<a class="reference internal" href="#id266" title="Michel DuPage, Alison L Dooley, and Tyler Jacks. Conditional mouse lung cancer models using adenoviral or lentiviral delivery of cre recombinase. Nature protocols, 4(7):1064–1072, 2009.">DuPage <em>et al.</em>, 2009</a>]</span>. Specifically, this mouse model carries oncogenic <em>Kras</em> and <em>Tp53</em> mutations that under native conditions are not expressed. However, upon the introduction of Cre recombinase via lentiviral inhalation, these oncogenic mutations are activated in single cells of the lung airway epithelium inducing tumors. In this study, the CRISPR/Cas9-based lineage tracer is under similar control and thus becomes active simultaneously upon tumor induction.</p>
<div>
<img src="../_static/images/trajectories/tumor_tracing.png" width="1000"/>
</div>
<p>With this system, the authors followed tumors from their single-cell origins over the course of approximately 4.5-6 months at which point they harvested aggressive, metastatic tumors. After dissociation of the tumors, the authors profiled both the lineage tracing target sites and the RNA content of single cells. This resulted in a large dataset of more than 70,000 cells across more than 100 tumors with both lineage and scRNA-seq information.</p>
<p>In this tutorial, we will demonstrate how a user can take processed target site data to learn interesting dynamic properties of lineages. Throughout this case study, each lineage will correspond to a single primary tumor sampled from a mouse lung.</p>
<div class="alert alert-block alert-warning" style="color:black;">
<b>Overview of input data:</b> 
<br/br>
    Before moving on, make sure you understand what data types we'll be utilizing. As discussed above, we are equipped with scRNA-seq count matrices and, separately, lineage tracing data (this can come in many forms; below, we begin with the <b>allele table</b> which summarizes the indels observed on each target within a cell. This is described more below). These two data will be processed separately. The lineage tracing data will be used to infer lineages, and the scRNA-seq data will be used to interpret interesting behaviors of these lineages.
</div>
<div class="alert alert-block alert-warning" style="color:black;">
<b>Structure of the lineage tracer:</b> 
<br/br>
    For the dataset analyzed below, each cell is engineered with approximately 10 target sites, each of which carrying 3 Cas9 cut sites in tandem. Each target site is labeled with a unique integration barcode ("intBC" for short). All together, we can expect to observe 3*(# target sites) characters in each cell that can accumulate mutations and be used for lineage inference. For more information on the structure of the sequencing cassette, please refer to <a href="https://www.nature.com/articles/s41586-019-1184-5">Chan, Smith et al. Molecular recording of mammalian embryogenesis. Nature 2019</a>.
</div>
<section id="downloading-the-data">
<h3><span class="section-number">16.5.1. </span>Downloading the data<a class="headerlink" href="#downloading-the-data" title="Permalink to this headline">#</a></h3>
<p>This data is publicly hosted on <a class="reference external" href="https://zenodo.org/record/5847462#.YrFDKuzMI6A">Zenodo</a> and we can download the data as follows:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="o">!</span>wget <span class="s2">&quot;https://zenodo.org/record/5847462/files/KPTracer-Data.tar.gz?download=1&quot;</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>--2022-06-30 10:28:22--  https://zenodo.org/record/5847462/files/KPTracer-Data.tar.gz?download=1
Resolving zenodo.org (zenodo.org)... 137.138.76.77
Connecting to zenodo.org (zenodo.org)|137.138.76.77|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 1304975216 (1.2G) [application/octet-stream]
Saving to: ‘KPTracer-Data.tar.gz?download=1’

KPTracer-Data.tar.g 100%[===================&gt;]   1.21G  1.49MB/s    in 7m 3s   

2022-06-30 10:35:27 (2.94 MB/s) - ‘KPTracer-Data.tar.gz?download=1’ saved [1304975216/1304975216]
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="o">!</span>tar -xvzf KPTracer-Data.tar.gz?download<span class="o">=</span><span class="m">1</span>
</pre></div>
</div>
</div>
</div>
</section>
<section id="environment-setup">
<h3><span class="section-number">16.5.2. </span>Environment setup.<a class="headerlink" href="#environment-setup" title="Permalink to this headline">#</a></h3>
<p>Before we enter this notebook’s analysis, let’s set up our environment.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">cassiopeia</span> <span class="k">as</span> <span class="nn">cas</span>
<span class="kn">import</span> <span class="nn">matplotlib.pyplot</span> <span class="k">as</span> <span class="nn">plt</span>
<span class="kn">import</span> <span class="nn">numpy</span> <span class="k">as</span> <span class="nn">np</span>
<span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span>
<span class="kn">import</span> <span class="nn">scanpy</span> <span class="k">as</span> <span class="nn">sc</span>

<span class="kn">from</span> <span class="nn">cassiopeia.preprocess</span> <span class="kn">import</span> <span class="n">lineage_utils</span>
</pre></div>
</div>
</div>
</div>
</section>
<section id="examine-the-data">
<h3><span class="section-number">16.5.3. </span>Examine the data<a class="headerlink" href="#examine-the-data" title="Permalink to this headline">#</a></h3>
<p>After preprocessing of the target site sequencing data, <code class="docutils literal notranslate"><span class="pre">Cassiopeia</span></code> uses an <code class="docutils literal notranslate"><span class="pre">allele_table</span></code> to summarize the mutations observed on each target site integration.</p>
<p>As described above, the system at hand should have approximately ~10 target sites per cell where each target site contains three Cas9 cut sites. Columns labeled <code class="docutils literal notranslate"><span class="pre">intBC</span></code> indicate the target site and columns labeled <code class="docutils literal notranslate"><span class="pre">r*</span></code> indicate the Cas9 cut sites on that target site. Mutations in each cut site are stored as CIGAR strings, indicating the size and type of indel observed.</p>
<p>Other metadata can also be stored in this table such as the total number of UMIs and reads associated with the target site molecule, which cell the molecule came from, and which tumor that cell belongs to.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">allele_table</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span>
    <span class="s2">&quot;KPTracer-Data/KPTracer.alleleTable.FINAL.txt&quot;</span><span class="p">,</span> <span class="n">sep</span><span class="o">=</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">,</span> <span class="n">index_col</span><span class="o">=</span><span class="mi">0</span>
<span class="p">)</span>

<span class="n">allele_table</span><span class="o">.</span><span class="n">head</span><span class="p">(</span><span class="mi">5</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>/var/folders/kt/wk3kbll507v4h18f8nmbzkl80000gq/T/ipykernel_65366/3031588044.py:1: DtypeWarning: Columns (12) have mixed types. Specify dtype option on import or set low_memory=False.
  allele_table = pd.read_csv(&quot;KPTracer-Data/KPTracer.alleleTable.FINAL.txt&quot;, sep=&#39;\t&#39;, index_col = 0)
</pre></div>
</div>
<div class="output text_html"><div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>cellBC</th>
      <th>intBC</th>
      <th>r1</th>
      <th>r2</th>
      <th>r3</th>
      <th>allele</th>
      <th>sampleID</th>
      <th>UMI</th>
      <th>readCount</th>
      <th>Tumor</th>
      <th>MetFamily</th>
      <th>ES_clone</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>L9.TTTGTCATCTGTCAAG-1</td>
      <td>TTCCCTATTTGCTA</td>
      <td>CGCCG[111:2D]AAATG</td>
      <td>GATAT[None]CTCTG</td>
      <td>AATTC[220:1I]GGCGGA</td>
      <td>CGCCG[111:2D]AAATGGATAT[None]CTCTGAATTC[220:1I...</td>
      <td>L9</td>
      <td>54</td>
      <td>796</td>
      <td>3724_NT_T1</td>
      <td>3724_NT_T1</td>
      <td>2E1</td>
    </tr>
    <tr>
      <th>1</th>
      <td>L9.TTTGTCATCTGTCAAG-1</td>
      <td>TGTTTTTGTCTGCA</td>
      <td>CGCCG[111:1I]AAAAAA</td>
      <td>CATGT[151:19D]TGGTT</td>
      <td>TTAAT[218:2D]GCGGA</td>
      <td>CGCCG[111:1I]AAAAAACATGT[151:19D]TGGTTTTAAT[21...</td>
      <td>L9</td>
      <td>13</td>
      <td>209</td>
      <td>3724_NT_T1</td>
      <td>3724_NT_T1</td>
      <td>2E1</td>
    </tr>
    <tr>
      <th>2</th>
      <td>L9.TTTGTCATCTGTCAAG-1</td>
      <td>TGTGAAGGTCAATA</td>
      <td>CCGAA[113:49D]GATAT</td>
      <td>CCGAA[113:49D]GATAT</td>
      <td>AATTC[220:5D]GGACA</td>
      <td>CCGAA[113:49D]GATATCCGAA[113:49D]GATATAATTC[22...</td>
      <td>L9</td>
      <td>68</td>
      <td>1193</td>
      <td>3724_NT_T1</td>
      <td>3724_NT_T1</td>
      <td>2E1</td>
    </tr>
    <tr>
      <th>3</th>
      <td>L9.TTTGTCATCTGTCAAG-1</td>
      <td>TCAGGCGATGCGAA</td>
      <td>CGCCG[111:1I]AAAAAA</td>
      <td>CGATA[166:1I]TTCTCT</td>
      <td>TAATT[219:2D]CGGAG</td>
      <td>CGCCG[111:1I]AAAAAACGATA[166:1I]TTCTCTTAATT[21...</td>
      <td>L9</td>
      <td>43</td>
      <td>745</td>
      <td>3724_NT_T1</td>
      <td>3724_NT_T1</td>
      <td>2E1</td>
    </tr>
    <tr>
      <th>4</th>
      <td>L9.TTTGTCATCTGTCAAG-1</td>
      <td>TATGATTAGTCGCG</td>
      <td>CGCCG[111:1D]AAAAT</td>
      <td>GATAT[167:54D]CGGAG</td>
      <td>GATAT[167:54D]CGGAG</td>
      <td>CGCCG[111:1D]AAAATGATAT[167:54D]CGGAGGATAT[167...</td>
      <td>L9</td>
      <td>27</td>
      <td>530</td>
      <td>3724_NT_T1</td>
      <td>3724_NT_T1</td>
      <td>2E1</td>
    </tr>
  </tbody>
</table>
</div></div></div>
</div>
<p>We’ll focus on the data from KP tumors without any additional perturbations:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">all_tumors</span> <span class="o">=</span> <span class="n">allele_table</span><span class="p">[</span><span class="s2">&quot;Tumor&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">unique</span><span class="p">()</span>
<span class="n">primary_nt_tumors</span> <span class="o">=</span> <span class="p">[</span>
    <span class="n">tumor</span>
    <span class="k">for</span> <span class="n">tumor</span> <span class="ow">in</span> <span class="n">all_tumors</span>
    <span class="k">if</span> <span class="s2">&quot;NT&quot;</span> <span class="ow">in</span> <span class="n">tumor</span> <span class="ow">and</span> <span class="n">tumor</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;_&quot;</span><span class="p">)[</span><span class="mi">2</span><span class="p">]</span><span class="o">.</span><span class="n">startswith</span><span class="p">(</span><span class="s2">&quot;T&quot;</span><span class="p">)</span>
<span class="p">]</span>

<span class="n">primary_nt_allele_table</span> <span class="o">=</span> <span class="n">allele_table</span><span class="p">[</span><span class="n">allele_table</span><span class="p">[</span><span class="s2">&quot;Tumor&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">isin</span><span class="p">(</span><span class="n">primary_nt_tumors</span><span class="p">)]</span>
</pre></div>
</div>
</div>
</div>
<p>As described above, we expect to see around 10 unique intBCs (i.e., target sites) per tumor population. Below, we summarize some key statistics for the dataset.</p>
<section id="number-of-intbcs-per-tumor">
<h4><span class="section-number">16.5.3.1. </span>Number of intBCs per Tumor<a class="headerlink" href="#number-of-intbcs-per-tumor" title="Permalink to this headline">#</a></h4>
<p>One statistic of interest to analysts is the <strong>number of intBCs per tumor</strong> as this informs on the information capacity for each clonal population.</p>
<p>In recent applications (<span id="id55">[<a class="reference internal" href="#id265" title="Dian Yang, Matthew G Jones, Santiago Naranjo, William M Rideout III, Kyung Hoi Joseph Min, Raymond Ho, Wei Wu, Joseph M Replogle, Jennifer L Page, Jeffrey J Quinn, and others. Lineage tracing reveals the phylodynamics, plasticity, and paths of tumor evolution. Cell, 185(11):1905–1923, 2022.">Yang <em>et al.</em>, 2022</a>]</span>), we find that high-quality clones typically have between 5 and 30 intBCs (corresponding to 15-90 characters). In plotting the number of intBCs per tumor, we can make sure there are not outliers tumors that we’d like to filter out before reconstruction.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">primary_nt_allele_table</span><span class="o">.</span><span class="n">groupby</span><span class="p">([</span><span class="s2">&quot;Tumor&quot;</span><span class="p">])</span><span class="o">.</span><span class="n">agg</span><span class="p">({</span><span class="s2">&quot;intBC&quot;</span><span class="p">:</span> <span class="s2">&quot;nunique&quot;</span><span class="p">})</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">kind</span><span class="o">=</span><span class="s2">&quot;bar&quot;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">ylabel</span><span class="p">(</span><span class="s2">&quot;Number of unique intBCs&quot;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">title</span><span class="p">(</span><span class="s2">&quot;Number of intBCs per Tumor&quot;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="../_images/lineage_tracing_18_0.png" src="../_images/lineage_tracing_18_0.png" />
</div>
</div>
<p>As expected, we observe that the majority of tumors have around 10 intBCs while one tumor reports few observed intBCs and a few report more than 25 intBCs.  Next we will discuss strategies for filtering out low-quality tumors.</p>
</section>
<section id="size-of-each-tumor">
<h4><span class="section-number">16.5.3.2. </span>Size of each tumor<a class="headerlink" href="#size-of-each-tumor" title="Permalink to this headline">#</a></h4>
<p>We are also interested in the size of the tumors that will be reconstructed. It’s typical to observe clones between 100 and 10,000 cells. Once again, we’ll want to make sure there are no clear outliers - specifically, clones that are too small for reconstruction (usually fewer than 100 cells).</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">primary_nt_allele_table</span><span class="o">.</span><span class="n">groupby</span><span class="p">([</span><span class="s2">&quot;Tumor&quot;</span><span class="p">])</span><span class="o">.</span><span class="n">agg</span><span class="p">({</span><span class="s2">&quot;cellBC&quot;</span><span class="p">:</span> <span class="s2">&quot;nunique&quot;</span><span class="p">})</span><span class="o">.</span><span class="n">sort_values</span><span class="p">(</span>
    <span class="n">by</span><span class="o">=</span><span class="s2">&quot;cellBC&quot;</span><span class="p">,</span> <span class="n">ascending</span><span class="o">=</span><span class="kc">False</span>
<span class="p">)</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">kind</span><span class="o">=</span><span class="s2">&quot;bar&quot;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">yscale</span><span class="p">(</span><span class="s2">&quot;log&quot;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">ylabel</span><span class="p">(</span><span class="s2">&quot;Number of cells (log)&quot;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">title</span><span class="p">(</span><span class="s2">&quot;Size of each tumor&quot;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="../_images/lineage_tracing_21_0.png" src="../_images/lineage_tracing_21_0.png" />
</div>
</div>
<p>We see that we have one very large clone (with ~30,000 cells) and the rest of the clones are in the expected range, reporting between 1,000 and 5,000 cells.</p>
</section>
</section>
<section id="preparing-the-data-for-lineage-reconstruction">
<h3><span class="section-number">16.5.4. </span>Preparing the data for lineage reconstruction<a class="headerlink" href="#preparing-the-data-for-lineage-reconstruction" title="Permalink to this headline">#</a></h3>
<p>A general feature of Cas9-based tracers is that certain insertions and deletions are much more common than others. This can be an important modeling concern as high-probability edits can occur several times independently of one another and could potentially lead to an analyst incorrectly believing that two cells are related to one another.</p>
<p>Cassiopeia has a built-in utility for estimating the prior probabilities of insertions and deletions by counting the number of times that an indel appeared in unrelated tumors or intBCs.</p>
<p>In observing the frequencies of observed indels, we see that a few indels (small deletions or insertions) appear quite often in clones. On the other hand, we also observe several edits observed in a single clone. Because of the difference in frequencies, an analyst should try to incorporate these observed biases into downstream analyses.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">indel_priors</span> <span class="o">=</span> <span class="n">cas</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">compute_empirical_indel_priors</span><span class="p">(</span>
    <span class="n">allele_table</span><span class="p">,</span> <span class="n">grouping_variables</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;intBC&quot;</span><span class="p">,</span> <span class="s2">&quot;MetFamily&quot;</span><span class="p">]</span>
<span class="p">)</span>

<span class="n">indel_priors</span><span class="o">.</span><span class="n">sort_values</span><span class="p">(</span><span class="n">by</span><span class="o">=</span><span class="s2">&quot;count&quot;</span><span class="p">,</span> <span class="n">ascending</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span><span class="o">.</span><span class="n">head</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>count</th>
      <th>freq</th>
    </tr>
    <tr>
      <th>indel</th>
      <th></th>
      <th></th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>TAATT[219:2D]CGGAG</th>
      <td>664.0</td>
      <td>0.764977</td>
    </tr>
    <tr>
      <th>CGCCG[111:1I]AAAAAA</th>
      <td>584.0</td>
      <td>0.672811</td>
    </tr>
    <tr>
      <th>CGCCG[111:1D]AAAAT</th>
      <td>504.0</td>
      <td>0.580645</td>
    </tr>
    <tr>
      <th>CGCCG[111:2D]AAATG</th>
      <td>401.0</td>
      <td>0.461982</td>
    </tr>
    <tr>
      <th>AATTC[220:3D]GAGGA</th>
      <td>395.0</td>
      <td>0.455069</td>
    </tr>
  </tbody>
</table>
</div></div></div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">indel_priors</span><span class="o">.</span><span class="n">sort_values</span><span class="p">(</span><span class="n">by</span><span class="o">=</span><span class="s2">&quot;count&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">head</span><span class="p">(</span><span class="mi">5</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>count</th>
      <th>freq</th>
    </tr>
    <tr>
      <th>indel</th>
      <th></th>
      <th></th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>ATATC[168:49I]GTTGTGGCCCAACATGGCAGCGTGCCGTAGCTTAGTTGTCAGGCCATTTGCTGG</th>
      <td>1.0</td>
      <td>0.001152</td>
    </tr>
    <tr>
      <th>ACGCC[110:2D]AGAAT</th>
      <td>1.0</td>
      <td>0.001152</td>
    </tr>
    <tr>
      <th>CATCT[101:15D]TGGCC</th>
      <td>1.0</td>
      <td>0.001152</td>
    </tr>
    <tr>
      <th>CCCGG[111:3D]ATTGG</th>
      <td>1.0</td>
      <td>0.001152</td>
    </tr>
    <tr>
      <th>CCGAA[113:1I]GGAATG</th>
      <td>1.0</td>
      <td>0.001152</td>
    </tr>
  </tbody>
</table>
</div></div></div>
</div>
<section id="filtering-out-low-quality-tumors">
<h4><span class="section-number">16.5.4.1. </span>Filtering out low-quality tumors<a class="headerlink" href="#filtering-out-low-quality-tumors" title="Permalink to this headline">#</a></h4>
<p>Before tumor reconstruction, we’ll filter out tumors with (i) poor lineage tracing kinetics or (ii) that are too small to consider for analysis.</p>
<p>In previous publications (<span id="id56">[<a class="reference internal" href="#id230" title="Jeffrey J Quinn, Matthew G Jones, Ross A Okimoto, Shigeki Nanjo, Michelle M Chan, Nir Yosef, Trever G Bivona, and Jonathan S Weissman. Single-cell lineages reveal the rates, routes, and drivers of metastasis in cancer xenografts. Science, 371(6532):eabc1944, 2021.">Quinn <em>et al.</em>, 2021</a>]</span> and <span id="id57">[<a class="reference internal" href="#id265" title="Dian Yang, Matthew G Jones, Santiago Naranjo, William M Rideout III, Kyung Hoi Joseph Min, Raymond Ho, Wei Wu, Joseph M Replogle, Jennifer L Page, Jeffrey J Quinn, and others. Lineage tracing reveals the phylodynamics, plasticity, and paths of tumor evolution. Cell, 185(11):1905–1923, 2022.">Yang <em>et al.</em>, 2022</a>]</span>), we have found that looking at the following statistics are useful for determining the quality of the tumor:</p>
<ul class="simple">
<li><p><strong>percent unique</strong>: This is the percentage of lineage states that are unique in a tumor.</p></li>
<li><p><strong>percent cut</strong>: This is the percentage of Cas9 targets that are mutated in a population of cells.</p></li>
<li><p><strong>percent exhausted</strong>: This is the fraction of target sites that are identical across cells (i.e., are exhausted)</p></li>
<li><p><strong>size of tumor</strong>: The number of cells in a tumor.</p></li>
</ul>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># utility functions for computing summary statistics</span>


<span class="k">def</span> <span class="nf">compute_percent_indels</span><span class="p">(</span><span class="n">character_matrix</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Computes the percentage of sites carrying indels in a character matrix.</span>

<span class="sd">    Args:</span>
<span class="sd">        character_matrix: A pandas Dataframe summarizing the mutation status of each cell.</span>

<span class="sd">    Returns:</span>
<span class="sd">        A percentage of sites in cells that contain an edit.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">all_vals</span> <span class="o">=</span> <span class="n">character_matrix</span><span class="o">.</span><span class="n">values</span><span class="o">.</span><span class="n">ravel</span><span class="p">()</span>
    <span class="n">num_not_missing</span> <span class="o">=</span> <span class="nb">len</span><span class="p">([</span><span class="n">n</span> <span class="k">for</span> <span class="n">n</span> <span class="ow">in</span> <span class="n">all_vals</span> <span class="k">if</span> <span class="n">n</span> <span class="o">!=</span> <span class="o">-</span><span class="mi">1</span><span class="p">])</span>
    <span class="n">num_uncut</span> <span class="o">=</span> <span class="nb">len</span><span class="p">([</span><span class="n">n</span> <span class="k">for</span> <span class="n">n</span> <span class="ow">in</span> <span class="n">all_vals</span> <span class="k">if</span> <span class="n">n</span> <span class="o">==</span> <span class="mi">0</span><span class="p">])</span>

    <span class="k">return</span> <span class="mf">1.0</span> <span class="o">-</span> <span class="p">(</span><span class="n">num_uncut</span> <span class="o">/</span> <span class="n">num_not_missing</span><span class="p">)</span>


<span class="k">def</span> <span class="nf">compute_percent_uncut</span><span class="p">(</span><span class="n">cell</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Computes the percentage of sites uncut in a cell.</span>

<span class="sd">    Args:</span>
<span class="sd">        A vector containing the edited sites for a particular cell.</span>

<span class="sd">    Returns:</span>
<span class="sd">        The number of sites uncut in a cell.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">uncut</span> <span class="o">=</span> <span class="mi">0</span>
    <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">cell</span><span class="p">:</span>
        <span class="k">if</span> <span class="n">i</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
            <span class="n">uncut</span> <span class="o">+=</span> <span class="mi">1</span>
    <span class="k">return</span> <span class="n">uncut</span> <span class="o">/</span> <span class="nb">max</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="nb">len</span><span class="p">([</span><span class="n">i</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">cell</span> <span class="k">if</span> <span class="n">i</span> <span class="o">!=</span> <span class="o">-</span><span class="mi">1</span><span class="p">]))</span>


<span class="k">def</span> <span class="nf">summarize_tumor_quality</span><span class="p">(</span>
    <span class="n">allele_table</span><span class="p">,</span>
    <span class="n">minimum_intbc_thresh</span><span class="o">=</span><span class="mf">0.2</span><span class="p">,</span>
    <span class="n">minimum_number_of_cells</span><span class="o">=</span><span class="mi">2</span><span class="p">,</span>
    <span class="n">maximum_percent_uncut_in_cell</span><span class="o">=</span><span class="mf">0.8</span><span class="p">,</span>
    <span class="n">allele_rep_thresh</span><span class="o">=</span><span class="mf">0.98</span><span class="p">,</span>
<span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Compute QC statistics for each tumor.</span>

<span class="sd">    Computes statistics for each clone that will be used for filtering tumors for downstream lineage reconstruction.</span>

<span class="sd">    Args:</span>
<span class="sd">        allele_table: A Cassipoeia allele table summarizing the indels on each molecule in each cell.</span>
<span class="sd">        min_intbc_thresh: The minimum proportion of cells that an intBC must appear in to be considered</span>
<span class="sd">            for downstream reconstruction.</span>
<span class="sd">        minimum_number_of_cells: Minimum number of cells in a tumor to be processed in this QC pipeline.</span>
<span class="sd">        maximum_percent_uncut_in_cell: The maximum percentage of sites allowed to be uncut in a cell. If</span>
<span class="sd">            a cell exceeds this threshold, it is filtered out.</span>
<span class="sd">        allele_rep_thresh: Maximum allele representation in a single cut site allowed. If a character has</span>
<span class="sd">            less diversity than allowed, it is filtered out.</span>

<span class="sd">    Returns:</span>
<span class="sd">        A pandas Dataframe summarizing the quality-control information for each tumor.</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="n">tumor_statistics</span> <span class="o">=</span> <span class="p">{}</span>
    <span class="n">NUMBER_OF_SITES_PER_INTBC</span> <span class="o">=</span> <span class="mi">3</span>

    <span class="c1"># iterate through Tumors and compute summary statistics</span>
    <span class="k">for</span> <span class="n">tumor_name</span><span class="p">,</span> <span class="n">tumor_allele_table</span> <span class="ow">in</span> <span class="n">allele_table</span><span class="o">.</span><span class="n">groupby</span><span class="p">(</span><span class="s2">&quot;Tumor&quot;</span><span class="p">):</span>

        <span class="k">if</span> <span class="n">tumor_allele_table</span><span class="p">[</span><span class="s2">&quot;cellBC&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">nunique</span><span class="p">()</span> <span class="o">&lt;</span> <span class="n">minimum_number_of_cells</span><span class="p">:</span>
            <span class="k">continue</span>

        <span class="n">tumor_allele_table</span> <span class="o">=</span> <span class="n">allele_table</span><span class="p">[</span><span class="n">allele_table</span><span class="p">[</span><span class="s2">&quot;Tumor&quot;</span><span class="p">]</span> <span class="o">==</span> <span class="n">tumor_name</span><span class="p">]</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
        <span class="n">tumor_allele_table</span><span class="p">[</span><span class="s2">&quot;lineageGrp&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">tumor_allele_table</span><span class="p">[</span><span class="s2">&quot;Tumor&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
        <span class="n">lineage_group</span> <span class="o">=</span> <span class="n">lineage_utils</span><span class="o">.</span><span class="n">filter_intbcs_final_lineages</span><span class="p">(</span>
            <span class="n">tumor_allele_table</span><span class="p">,</span> <span class="n">min_intbc_thresh</span><span class="o">=</span><span class="n">minimum_intbc_thresh</span>
        <span class="p">)[</span><span class="mi">0</span><span class="p">]</span>

        <span class="n">number_of_cutsites</span> <span class="o">=</span> <span class="p">(</span>
            <span class="nb">len</span><span class="p">(</span><span class="n">lineage_group</span><span class="p">[</span><span class="s2">&quot;intBC&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">unique</span><span class="p">())</span> <span class="o">*</span> <span class="n">NUMBER_OF_SITES_PER_INTBC</span>
        <span class="p">)</span>

        <span class="n">character_matrix</span><span class="p">,</span> <span class="n">_</span><span class="p">,</span> <span class="n">_</span> <span class="o">=</span> <span class="n">cas</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">convert_alleletable_to_character_matrix</span><span class="p">(</span>
            <span class="n">lineage_group</span><span class="p">,</span> <span class="n">allele_rep_thresh</span><span class="o">=</span><span class="n">allele_rep_thresh</span>
        <span class="p">)</span>

        <span class="c1"># We&#39;ll hit this if we filter out all characters with the specified allele_rep_thresh</span>
        <span class="k">if</span> <span class="n">character_matrix</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
            <span class="n">character_matrix</span><span class="p">,</span> <span class="n">_</span><span class="p">,</span> <span class="n">_</span> <span class="o">=</span> <span class="n">cas</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">convert_alleletable_to_character_matrix</span><span class="p">(</span>
                <span class="n">lineage_group</span><span class="p">,</span> <span class="n">allele_rep_thresh</span><span class="o">=</span><span class="mf">1.0</span>
            <span class="p">)</span>

        <span class="n">number_dropped_intbcs</span> <span class="o">=</span> <span class="n">number_of_cutsites</span> <span class="o">-</span> <span class="n">character_matrix</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span>
        <span class="n">percent_uncut</span> <span class="o">=</span> <span class="n">character_matrix</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span>
            <span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">compute_percent_uncut</span><span class="p">(</span><span class="n">x</span><span class="o">.</span><span class="n">values</span><span class="p">),</span> <span class="n">axis</span><span class="o">=</span><span class="mi">1</span>
        <span class="p">)</span>

        <span class="c1"># drop normal cells from lineage (cells without editing)</span>
        <span class="n">character_matrix_filtered</span> <span class="o">=</span> <span class="n">character_matrix</span><span class="p">[</span>
            <span class="n">percent_uncut</span> <span class="o">&lt;</span> <span class="n">maximum_percent_uncut_in_cell</span>
        <span class="p">]</span>

        <span class="n">percent_unique</span> <span class="o">=</span> <span class="p">(</span>
            <span class="n">character_matrix_filtered</span><span class="o">.</span><span class="n">drop_duplicates</span><span class="p">()</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
            <span class="o">/</span> <span class="n">character_matrix_filtered</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
        <span class="p">)</span>
        <span class="n">tumor_statistics</span><span class="p">[</span><span class="n">tumor_name</span><span class="p">]</span> <span class="o">=</span> <span class="p">(</span>
            <span class="n">percent_unique</span><span class="p">,</span>
            <span class="n">compute_percent_indels</span><span class="p">(</span><span class="n">character_matrix_filtered</span><span class="p">),</span>
            <span class="n">number_dropped_intbcs</span><span class="p">,</span>
            <span class="mf">1.0</span> <span class="o">-</span> <span class="p">(</span><span class="n">number_dropped_intbcs</span> <span class="o">/</span> <span class="n">number_of_cutsites</span><span class="p">),</span>
            <span class="n">character_matrix_filtered</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span>
        <span class="p">)</span>

    <span class="n">tumor_clone_statistics</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="o">.</span><span class="n">from_dict</span><span class="p">(</span>
        <span class="n">tumor_statistics</span><span class="p">,</span>
        <span class="n">orient</span><span class="o">=</span><span class="s2">&quot;index&quot;</span><span class="p">,</span>
        <span class="n">columns</span><span class="o">=</span><span class="p">[</span>
            <span class="s2">&quot;PercentUnique&quot;</span><span class="p">,</span>
            <span class="s2">&quot;CutRate&quot;</span><span class="p">,</span>
            <span class="s2">&quot;NumSaturatedTargets&quot;</span><span class="p">,</span>
            <span class="s2">&quot;PercentUnsaturatedTargets&quot;</span><span class="p">,</span>
            <span class="s2">&quot;NumCells&quot;</span><span class="p">,</span>
        <span class="p">],</span>
    <span class="p">)</span>

    <span class="k">return</span> <span class="n">tumor_clone_statistics</span>
</pre></div>
</div>
</div>
</div>
</section>
<section id="calculate-and-visualize-tumors-statistics">
<h4><span class="section-number">16.5.4.2. </span>Calculate and visualize tumors’ statistics<a class="headerlink" href="#calculate-and-visualize-tumors-statistics" title="Permalink to this headline">#</a></h4>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">tumor_clone_statistics</span> <span class="o">=</span> <span class="n">summarize_tumor_quality</span><span class="p">(</span><span class="n">primary_nt_allele_table</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">NUM_CELLS_THRESH</span> <span class="o">=</span> <span class="mi">100</span>
<span class="n">PERCENT_UNIQUE_THRESH</span> <span class="o">=</span> <span class="mf">0.05</span>
<span class="n">PERCENT_UNSATURATED_TARGETS_THRESH</span> <span class="o">=</span> <span class="mf">0.2</span>

<span class="n">low_qc</span> <span class="o">=</span> <span class="n">tumor_clone_statistics</span><span class="p">[</span>
    <span class="p">(</span><span class="n">tumor_clone_statistics</span><span class="p">[</span><span class="s2">&quot;PercentUnique&quot;</span><span class="p">]</span> <span class="o">&lt;=</span> <span class="n">PERCENT_UNIQUE_THRESH</span><span class="p">)</span>
    <span class="o">|</span> <span class="p">(</span>
        <span class="n">tumor_clone_statistics</span><span class="p">[</span><span class="s2">&quot;PercentUnsaturatedTargets&quot;</span><span class="p">]</span>
        <span class="o">&lt;=</span> <span class="n">PERCENT_UNSATURATED_TARGETS_THRESH</span>
    <span class="p">)</span>
<span class="p">]</span><span class="o">.</span><span class="n">index</span>
<span class="n">small</span> <span class="o">=</span> <span class="n">tumor_clone_statistics</span><span class="p">[</span>
    <span class="p">(</span><span class="n">tumor_clone_statistics</span><span class="p">[</span><span class="s2">&quot;NumCells&quot;</span><span class="p">]</span> <span class="o">&lt;</span> <span class="n">NUM_CELLS_THRESH</span><span class="p">)</span>
<span class="p">]</span><span class="o">.</span><span class="n">index</span>

<span class="n">unfiltered</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">setdiff1d</span><span class="p">(</span><span class="n">tumor_clone_statistics</span><span class="o">.</span><span class="n">index</span><span class="p">,</span> <span class="n">np</span><span class="o">.</span><span class="n">union1d</span><span class="p">(</span><span class="n">low_qc</span><span class="p">,</span> <span class="n">small</span><span class="p">))</span>

<span class="n">h</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">figure</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">6</span><span class="p">,</span> <span class="mi">6</span><span class="p">))</span>
<span class="n">plt</span><span class="o">.</span><span class="n">scatter</span><span class="p">(</span>
    <span class="n">tumor_clone_statistics</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">unfiltered</span><span class="p">,</span> <span class="s2">&quot;PercentUnsaturatedTargets&quot;</span><span class="p">],</span>
    <span class="n">tumor_clone_statistics</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">unfiltered</span><span class="p">,</span> <span class="s2">&quot;PercentUnique&quot;</span><span class="p">],</span>
    <span class="n">color</span><span class="o">=</span><span class="s2">&quot;black&quot;</span><span class="p">,</span>
<span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">scatter</span><span class="p">(</span>
    <span class="n">tumor_clone_statistics</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">low_qc</span><span class="p">,</span> <span class="s2">&quot;PercentUnsaturatedTargets&quot;</span><span class="p">],</span>
    <span class="n">tumor_clone_statistics</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">low_qc</span><span class="p">,</span> <span class="s2">&quot;PercentUnique&quot;</span><span class="p">],</span>
    <span class="n">color</span><span class="o">=</span><span class="s2">&quot;red&quot;</span><span class="p">,</span>
    <span class="n">label</span><span class="o">=</span><span class="s2">&quot;Poor QC&quot;</span><span class="p">,</span>
<span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">scatter</span><span class="p">(</span>
    <span class="n">tumor_clone_statistics</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">small</span><span class="p">,</span> <span class="s2">&quot;PercentUnsaturatedTargets&quot;</span><span class="p">],</span>
    <span class="n">tumor_clone_statistics</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">small</span><span class="p">,</span> <span class="s2">&quot;PercentUnique&quot;</span><span class="p">],</span>
    <span class="n">color</span><span class="o">=</span><span class="s2">&quot;orange&quot;</span><span class="p">,</span>
    <span class="n">label</span><span class="o">=</span><span class="s2">&quot;Small lineages&quot;</span><span class="p">,</span>
<span class="p">)</span>


<span class="n">plt</span><span class="o">.</span><span class="n">axhline</span><span class="p">(</span><span class="n">y</span><span class="o">=</span><span class="n">PERCENT_UNIQUE_THRESH</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="s2">&quot;red&quot;</span><span class="p">,</span> <span class="n">alpha</span><span class="o">=</span><span class="mf">0.5</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">axvline</span><span class="p">(</span><span class="n">x</span><span class="o">=</span><span class="n">PERCENT_UNSATURATED_TARGETS_THRESH</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="s2">&quot;red&quot;</span><span class="p">,</span> <span class="n">alpha</span><span class="o">=</span><span class="mf">0.5</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">xlabel</span><span class="p">(</span><span class="s2">&quot;Percent Unsaturated&quot;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">ylabel</span><span class="p">(</span><span class="s2">&quot;Percent Unique&quot;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">title</span><span class="p">(</span><span class="s2">&quot;Summary statistics for Tumor lineages&quot;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">legend</span><span class="p">(</span><span class="n">loc</span><span class="o">=</span><span class="s2">&quot;lower right&quot;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="../_images/lineage_tracing_30_0.png" src="../_images/lineage_tracing_30_0.png" />
</div>
</div>
<p>This plot summarizes our quality-control filtering efforts. Each point corresponds to a tumor and its color indicates its filter status:</p>
<ul class="simple">
<li><p>Tumors colored in <b><font color='red'>red</font></b> are filtered because either they have too few unique states or too few characters that can be used for reconstruction</p></li>
<li><p>Tumors colored in <b><font color='orange'>orange</font></b> are filtered out because they are too small for reconstruction (we use a 100 cell filter)</p></li>
<li><p>Tumors in <strong>black</strong> have passed our quality-control filters and will be considered for reconstruction.</p></li>
</ul>
<p>The tumor <code class="docutils literal notranslate"><span class="pre">3726_NT_T1</span></code> appears to be a lineage with satisfactory tracing data quality and we will reconstruct the tumor lineage using Cassiopeia.</p>
</section>
<section id="reconstruction-of-a-selected-tumor-3726-nt-t1">
<h4><span class="section-number">16.5.4.3. </span>Reconstruction of a selected tumor (<code class="docutils literal notranslate"><span class="pre">3726_NT_T1</span></code>)<a class="headerlink" href="#reconstruction-of-a-selected-tumor-3726-nt-t1" title="Permalink to this headline">#</a></h4>
<p>In order to reconstruct lineages, we’ll have to convert the allele table for a particular tumor into a “character matrix”. As defined above, these data structures summarize the mutations observed in each target site in each cell. By default, uncut sites are indicated as 0 and missing sites are indicated as -1. All other values in the matrix correspond to unique indels.</p>
<p>The tumor <code class="docutils literal notranslate"><span class="pre">3726_NT_T1</span></code> appears to be a lineage with satisfactory tracing data quality and we will reconstruct the tumor lineage using Cassiopeia.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">tumor</span> <span class="o">=</span> <span class="s2">&quot;3726_NT_T1&quot;</span>

<span class="n">tumor_allele_table</span> <span class="o">=</span> <span class="n">primary_nt_allele_table</span><span class="p">[</span><span class="n">primary_nt_allele_table</span><span class="p">[</span><span class="s2">&quot;Tumor&quot;</span><span class="p">]</span> <span class="o">==</span> <span class="n">tumor</span><span class="p">]</span>

<span class="n">n_cells</span> <span class="o">=</span> <span class="n">tumor_allele_table</span><span class="p">[</span><span class="s2">&quot;cellBC&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">nunique</span><span class="p">()</span>
<span class="n">n_intbc</span> <span class="o">=</span> <span class="n">tumor_allele_table</span><span class="p">[</span><span class="s2">&quot;intBC&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">nunique</span><span class="p">()</span>

<span class="nb">print</span><span class="p">(</span>
    <span class="sa">f</span><span class="s2">&quot;Tumor population </span><span class="si">{</span><span class="n">tumor</span><span class="si">}</span><span class="s2"> has </span><span class="si">{</span><span class="n">n_cells</span><span class="si">}</span><span class="s2"> cells and </span><span class="si">{</span><span class="n">n_intbc</span><span class="si">}</span><span class="s2"> intBCs (</span><span class="si">{</span><span class="n">n_intbc</span> <span class="o">*</span> <span class="mi">3</span><span class="si">}</span><span class="s2"> characters).&quot;</span>
<span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Tumor population 3726_NT_T1 has 772 cells and 10 intBCs (30) characters.
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="p">(</span>
    <span class="n">character_matrix</span><span class="p">,</span>
    <span class="n">priors</span><span class="p">,</span>
    <span class="n">state_to_indel</span><span class="p">,</span>
<span class="p">)</span> <span class="o">=</span> <span class="n">cas</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">convert_alleletable_to_character_matrix</span><span class="p">(</span>
    <span class="n">tumor_allele_table</span><span class="p">,</span> <span class="n">allele_rep_thresh</span><span class="o">=</span><span class="mf">0.9</span><span class="p">,</span> <span class="n">mutation_priors</span><span class="o">=</span><span class="n">indel_priors</span>
<span class="p">)</span>

<span class="n">character_matrix</span><span class="o">.</span><span class="n">head</span><span class="p">(</span><span class="mi">5</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Dropping the following intBCs due to lack of diversity with threshold 0.9: [&#39;ACTCTGCTCCAGATr2&#39;, &#39;ACTCTGCTCCAGATr3&#39;, &#39;GCCTACTTAAGTCCr1&#39;, &#39;GTTTATTTCCGTATr3&#39;, &#39;TATGATTAGTCGCGr1&#39;, &#39;TATGATTAGTCGCGr2&#39;, &#39;TGATATAAATCTTTr2&#39;, &#39;TTCCCTATTTGCTAr2&#39;, &#39;TGTTTTTGTCTGCAr1&#39;, &#39;ACAGGTGCTCAAATr1&#39;, &#39;ACAGGTGCTCAAATr2&#39;, &#39;ACAGGTGCTCAAATr3&#39;]
</pre></div>
</div>
<script type="application/vnd.jupyter.widget-view+json">
{"model_id": "228f295010bb41ddab67c47de393ef8a", "version_major": 2, "version_minor": 0}
</script><div class="output text_html"><div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>r1</th>
      <th>r2</th>
      <th>r3</th>
      <th>r4</th>
      <th>r5</th>
      <th>r6</th>
      <th>r7</th>
      <th>r8</th>
      <th>r9</th>
      <th>r10</th>
      <th>r11</th>
      <th>r12</th>
      <th>r13</th>
      <th>r14</th>
      <th>r15</th>
      <th>r16</th>
      <th>r17</th>
      <th>r18</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>L6.TTTGTCACACATCCAA-1</th>
      <td>1</td>
      <td>1</td>
      <td>1</td>
      <td>1</td>
      <td>1</td>
      <td>1</td>
      <td>1</td>
      <td>1</td>
      <td>1</td>
      <td>1</td>
      <td>1</td>
      <td>1</td>
      <td>1</td>
      <td>1</td>
      <td>1</td>
      <td>1</td>
      <td>-1</td>
      <td>-1</td>
    </tr>
    <tr>
      <th>L6.TTTGGTTTCTGAGTGT-1</th>
      <td>1</td>
      <td>1</td>
      <td>1</td>
      <td>1</td>
      <td>1</td>
      <td>1</td>
      <td>1</td>
      <td>1</td>
      <td>1</td>
      <td>1</td>
      <td>1</td>
      <td>1</td>
      <td>1</td>
      <td>1</td>
      <td>1</td>
      <td>1</td>
      <td>1</td>
      <td>1</td>
    </tr>
    <tr>
      <th>L6.TTTGGTTCATGTAAGA-1</th>
      <td>1</td>
      <td>1</td>
      <td>2</td>
      <td>2</td>
      <td>2</td>
      <td>2</td>
      <td>2</td>
      <td>2</td>
      <td>2</td>
      <td>2</td>
      <td>2</td>
      <td>2</td>
      <td>2</td>
      <td>2</td>
      <td>2</td>
      <td>2</td>
      <td>2</td>
      <td>2</td>
    </tr>
    <tr>
      <th>L6.TTTGCGCAGCTCCTCT-1</th>
      <td>1</td>
      <td>2</td>
      <td>3</td>
      <td>3</td>
      <td>3</td>
      <td>3</td>
      <td>1</td>
      <td>3</td>
      <td>3</td>
      <td>3</td>
      <td>1</td>
      <td>1</td>
      <td>3</td>
      <td>3</td>
      <td>3</td>
      <td>3</td>
      <td>3</td>
      <td>0</td>
    </tr>
    <tr>
      <th>L6.TTTATGCTCGCCGTGA-1</th>
      <td>1</td>
      <td>1</td>
      <td>1</td>
      <td>1</td>
      <td>1</td>
      <td>1</td>
      <td>1</td>
      <td>1</td>
      <td>1</td>
      <td>1</td>
      <td>1</td>
      <td>1</td>
      <td>1</td>
      <td>1</td>
      <td>1</td>
      <td>1</td>
      <td>4</td>
      <td>0</td>
    </tr>
  </tbody>
</table>
</div></div></div>
</div>
</section>
</section>
<section id="reconstructing-the-lineage">
<h3><span class="section-number">16.5.5. </span>Reconstructing the lineage<a class="headerlink" href="#reconstructing-the-lineage" title="Permalink to this headline">#</a></h3>
<p>There exist several algorithms for inferring phylogenies, many of which can be employed from the general character matrix format that Cassiopeia uses. Because of this, Cassiopeia has implemented several algorithms that can be used for lineage reconstruction, and many others can be implemented using the general <code class="docutils literal notranslate"><span class="pre">CassiopeiaSolver</span></code> API. Among the most popular algorithms are:</p>
<ul class="simple">
<li><p><strong>VanillaGreedy</strong>: A simple and efficient heuristic-based algorithm that can be good for a first-pass at your lineage reconstruction. Based on the landmark Gusfield algorithm <span id="id58">[<a class="reference internal" href="#id229" title="Dan Gusfield. Efficient algorithms for inferring evolutionary trees. Networks, 21(1):19–28, 1991.">Gusfield, 1991</a>]</span> and described in <span id="id59">[<a class="reference internal" href="#id264" title="Matthew G Jones, Alex Khodaverdian, Jeffrey J Quinn, Michelle M Chan, Jeffrey A Hussmann, Robert Wang, Chenling Xu, Jonathan S Weissman, and Nir Yosef. Inference of single-cell phylogenies from lineage tracing data using cassiopeia. Genome biology, 21(1):1–27, 2020.">Jones <em>et al.</em>, 2020</a>]</span>.</p></li>
<li><p><strong>Neighbor-Joining</strong>: A classic distance-based algorithm, as described in <span id="id60">[<a class="reference internal" href="#id232" title="Naruya Saitou and Masatoshi Nei. The neighbor-joining method: a new method for reconstructing phylogenetic trees. Molecular biology and evolution, 4(4):406–425, 1987.">Saitou and Nei, 1987</a>]</span>.</p></li>
<li><p><strong>UPGMA</strong>: An efficient distance-based algorithm that has special assumptions about how the samples are related to one another. Originally described in <span id="id61">[<a class="reference internal" href="#id231" title="Robert R Sokal. A statistical method for evaluating systematic relationships. Univ. Kansas, Sci. Bull., 38:1409–1438, 1958.">Sokal, 1958</a>]</span>.</p></li>
<li><p><strong>ILPSolver</strong>: An implementation of a Steiner-Tree optimization inference algorithm, as described in <span id="id62">[<a class="reference internal" href="#id264" title="Matthew G Jones, Alex Khodaverdian, Jeffrey J Quinn, Michelle M Chan, Jeffrey A Hussmann, Robert Wang, Chenling Xu, Jonathan S Weissman, and Nir Yosef. Inference of single-cell phylogenies from lineage tracing data using cassiopeia. Genome biology, 21(1):1–27, 2020.">Jones <em>et al.</em>, 2020</a>]</span>. This algorithm is slow (and will often not scale beyond around 1.5k cells) but precise.</p></li>
<li><p><strong>HybridSolver</strong>: A divide-and-conquer hybrid algorithm that uses a Greedy top-down algorithm to divide the data with a precise algorithm to conquer the subproblems (i.e., the ILPSolver). Described in <span id="id63">[<a class="reference internal" href="#id264" title="Matthew G Jones, Alex Khodaverdian, Jeffrey J Quinn, Michelle M Chan, Jeffrey A Hussmann, Robert Wang, Chenling Xu, Jonathan S Weissman, and Nir Yosef. Inference of single-cell phylogenies from lineage tracing data using cassiopeia. Genome biology, 21(1):1–27, 2020.">Jones <em>et al.</em>, 2020</a>]</span>.</p></li>
</ul>
<p>For the tutorial at hand, we will proceed with the VanillaGreedySolver. Additional solvers from the <a class="reference external" href="https://cassiopeia-lineage.readthedocs.io/en/latest/api/solver.html">Cassiopeia codebase</a> can be plugged in as alternatives.</p>
<p>To set up inference, we’ll instantiate a <code class="docutils literal notranslate"><span class="pre">CassiopeiaTree</span></code> which is a Cassiopeia-specific data structure for storing the character matrix and lineage meta data, as well as providing specific utilities for tree manipulation. We’ll also instantiate a <code class="docutils literal notranslate"><span class="pre">VanillaGreedySolver</span></code> which will populate the <code class="docutils literal notranslate"><span class="pre">tree</span></code> field in the <code class="docutils literal notranslate"><span class="pre">CassiopeiaTree</span></code> object.</p>
<div class="alert alert-block alert-warning" style="color:black;">
<b>Choosing an algorithm:</b> 
<br/br>
    Having so many algorithms at your disposal, it might seem daunting to choose the "right" one. For this reason, we often suggest that a user perform inference with a few algorithms for comparison. In addition to this, users will want to consider the size of their dataset and prior benchmarking studies as criteria for selecting the algorithms to use. As a rule of thumb, prior experience has found that the <b>HybridSolver</b> is a good balance between scalability and precision, and <b>Neighbor-Joining</b> is a convenient alternative for comparison (Neighbor-Joining also has the advantage of being a different class of algorithm [distance-based versus character-based], so it might highlight different parts of the lineage). For the tutorial purposes here, we will proceed with the GreedySolver as it is extremely efficient and often can serve as a good first-pass of the data before applying more precise (but slower) algorithms.
</div><div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">tree</span> <span class="o">=</span> <span class="n">cas</span><span class="o">.</span><span class="n">data</span><span class="o">.</span><span class="n">CassiopeiaTree</span><span class="p">(</span><span class="n">character_matrix</span><span class="o">=</span><span class="n">character_matrix</span><span class="p">,</span> <span class="n">priors</span><span class="o">=</span><span class="n">priors</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">greedy_solver</span> <span class="o">=</span> <span class="n">cas</span><span class="o">.</span><span class="n">solver</span><span class="o">.</span><span class="n">VanillaGreedySolver</span><span class="p">()</span>
</pre></div>
</div>
</div>
</div>
<p>Running <code class="docutils literal notranslate"><span class="pre">greedy_solver.solve</span></code> will perform tree inference and populate the <code class="docutils literal notranslate"><span class="pre">tree</span></code> field in the <code class="docutils literal notranslate"><span class="pre">CassiopeiaTree</span></code> object.</p>
<p>After inference, we can take advantage of the Cassiopeia visualization library to study the resulting tree structure. We’ll observe that there are some mistakes in the inference, but overall large groups of cells seem to be correctly placed. As described above, this can be a nice first-pass of the data before more time-consuming and sophisticated algorithms are deployed.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">greedy_solver</span><span class="o">.</span><span class="n">solve</span><span class="p">(</span><span class="n">tree</span><span class="p">)</span>

<span class="n">cas</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">plot_matplotlib</span><span class="p">(</span><span class="n">tree</span><span class="p">,</span> <span class="n">orient</span><span class="o">=</span><span class="s2">&quot;right&quot;</span><span class="p">,</span> <span class="n">allele_table</span><span class="o">=</span><span class="n">tumor_allele_table</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<script type="application/vnd.jupyter.widget-view+json">
{"model_id": "469e36b62b6f45999d9a7380cdde4578", "version_major": 2, "version_minor": 0}
</script><div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>100%|██████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 30/30 [00:11&lt;00:00,  2.71it/s]
</pre></div>
</div>
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>(&lt;Figure size 504x504 with 1 Axes&gt;, &lt;AxesSubplot:&gt;)
</pre></div>
</div>
<img alt="../_images/lineage_tracing_39_3.png" src="../_images/lineage_tracing_39_3.png" />
</div>
</div>
<p>This representation of the reconstruction can be very useful for assessing algorithmic performance. Assessing the performance of these reconstructions can be quite difficult, but becomes easier with practice. A good tip for starting to assess the performance is to select a column and see if a given indel appears more than once in separate groups of unrelated cells. This might indicate an algorithmic error.</p>
<p>As one can see, the majority of cells (rows in the heatmap) are in neighborhoods of cells with similar edit states. One consequence of this is that there are specific indels in cut sites that are grouped together cleanly (e.g., the pink allele in the 4th column from the left). We can also be confident in the reconstruction if several alleles group cells together – for example, the pink allele in the 4th column, the light-purple allele in the 6th column, and so forth.</p>
<p>There are a few mistakes that can be gleaned from this reconstruction, however. For example, in the fourth column two groups of cells are separated that both share a purple allele. There are other alleles in the heatmap that argue that these should be grouped together, indicating that this is an algorithmic error.</p>
<p>As we discussed above, since there are several algorithmic options available to an analyst, <strong>it can often be useful to apply a few algorithms for comparison.</strong></p>
</section>
<section id="quantifying-dynamic-properties-from-the-tree">
<h3><span class="section-number">16.5.6. </span>Quantifying dynamic properties from the tree<a class="headerlink" href="#quantifying-dynamic-properties-from-the-tree" title="Permalink to this headline">#</a></h3>
<p>From the tree structure learning properties about the population would be of biological interest. Some common properties would be:</p>
<ul class="simple">
<li><p>The <em>timing</em> and <em>location</em> of “expansion” events (i.e., the emergence of a population that can outgrow their neighbors)</p></li>
<li><p>The <em>fitness</em> (i.e., relative growth rate) of individual cells</p></li>
<li><p>The number of <em>cell state</em> changes in a population (i.e., plasticity)</p></li>
<li><p>Common <em>differentiation programs</em> in a population of cells</p></li>
</ul>
<p>There exist several other downstream analytical tasks, and this is growing area of research.</p>
<p>Below we provide examples of how <code class="docutils literal notranslate"><span class="pre">Cassiopeia</span></code> can be used to infer some of these parameters.</p>
</section>
<section id="inferring-expansion-events">
<h3><span class="section-number">16.5.7. </span>Inferring expansion events<a class="headerlink" href="#inferring-expansion-events" title="Permalink to this headline">#</a></h3>
<p>We’ll use the <code class="docutils literal notranslate"><span class="pre">compute_expansion_pvalues</span></code> to compute the probability that a certain internal node in the tree gave rise to a faster-growing population. This procedure, introduced in <span id="id64">[<a class="reference internal" href="#id265" title="Dian Yang, Matthew G Jones, Santiago Naranjo, William M Rideout III, Kyung Hoi Joseph Min, Raymond Ho, Wei Wu, Joseph M Replogle, Jennifer L Page, Jeffrey J Quinn, and others. Lineage tracing reveals the phylodynamics, plasticity, and paths of tumor evolution. Cell, 185(11):1905–1923, 2022.">Yang <em>et al.</em>, 2022</a>]</span>, performs a depth-first-search over the tree and annotates each internal node with a probability that the number of descendants would have been generated under a neutral model of evolution. The procedure takes in a few hyperparameters of interest:</p>
<ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">min_clade_size</span></code>: The minimum size of a clade (i.e., number of leaves below an internal node) for consideration. It is often the case that clades that are too uninformative.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">min_depth</span></code>: The minimum depth for an expansion to start. A <code class="docutils literal notranslate"><span class="pre">min_depth</span></code> of 0 indicates that the root could be considered as expanding; a more reasonable depth would be 1 (one level below the root).</p></li>
</ul>
<p>From this procedure, we can sift through the nodes and highlight where expansions are occurring. For example, in <span id="id65">[<a class="reference internal" href="#id265" title="Dian Yang, Matthew G Jones, Santiago Naranjo, William M Rideout III, Kyung Hoi Joseph Min, Raymond Ho, Wei Wu, Joseph M Replogle, Jennifer L Page, Jeffrey J Quinn, and others. Lineage tracing reveals the phylodynamics, plasticity, and paths of tumor evolution. Cell, 185(11):1905–1923, 2022.">Yang <em>et al.</em>, 2022</a>]</span>, the authors present a strategy for highlighting the most important expansion event. Below, we highlight an interesting expansion event in red.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">cas</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">compute_expansion_pvalues</span><span class="p">(</span><span class="n">tree</span><span class="p">,</span> <span class="n">min_clade_size</span><span class="o">=</span><span class="p">(</span><span class="mf">0.15</span> <span class="o">*</span> <span class="n">tree</span><span class="o">.</span><span class="n">n_cell</span><span class="p">),</span> <span class="n">min_depth</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># this specifies a p-value for identifying expansions unlikely to have occurred</span>
<span class="c1"># in a neutral model of evolution</span>
<span class="n">probability_threshold</span> <span class="o">=</span> <span class="mf">0.01</span>

<span class="n">expanding_nodes</span> <span class="o">=</span> <span class="p">[]</span>
<span class="k">for</span> <span class="n">node</span> <span class="ow">in</span> <span class="n">tree</span><span class="o">.</span><span class="n">depth_first_traverse_nodes</span><span class="p">():</span>
    <span class="k">if</span> <span class="n">tree</span><span class="o">.</span><span class="n">get_attribute</span><span class="p">(</span><span class="n">node</span><span class="p">,</span> <span class="s2">&quot;expansion_pvalue&quot;</span><span class="p">)</span> <span class="o">&lt;</span> <span class="n">probability_threshold</span><span class="p">:</span>
        <span class="n">expanding_nodes</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">node</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">cas</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">plot_matplotlib</span><span class="p">(</span><span class="n">tree</span><span class="p">,</span> <span class="n">clade_colors</span><span class="o">=</span><span class="p">{</span><span class="n">expanding_nodes</span><span class="p">[</span><span class="mi">6</span><span class="p">]:</span> <span class="s2">&quot;red&quot;</span><span class="p">})</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>(&lt;Figure size 504x504 with 1 Axes&gt;, &lt;AxesSubplot:&gt;)
</pre></div>
</div>
<img alt="../_images/lineage_tracing_45_1.png" src="../_images/lineage_tracing_45_1.png" />
</div>
</div>
<p>In highlighting where expansions occurred, we can partition the tree into groups of more- and less-aggressive subpopulations. We highlight in red one such population that is likely more aggressive as it is a detected expansion.</p>
<p>In the original study, the authors found that <strong>distinct transcriptional patterns</strong> are associated with <strong>expanding regions</strong> that likely confer a fitness advantage.</p>
<section id="inferring-tree-plasticity">
<h4><span class="section-number">16.5.7.1. </span>Inferring tree plasticity<a class="headerlink" href="#inferring-tree-plasticity" title="Permalink to this headline">#</a></h4>
<p>Tumors from this model consist of cells from various transcriptional states. The heterogeneity of this dataset can be observed on a low-dimensional projection of this dataset, such as those computed with UMAP:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">kptracer_adata</span> <span class="o">=</span> <span class="n">sc</span><span class="o">.</span><span class="n">read_h5ad</span><span class="p">(</span><span class="s2">&quot;KPTracer-Data/expression/adata_processed.nt.h5ad&quot;</span><span class="p">)</span>
<span class="n">sc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">umap</span><span class="p">(</span>
    <span class="n">kptracer_adata</span><span class="p">,</span>
    <span class="n">color</span><span class="o">=</span><span class="s2">&quot;Cluster-Name&quot;</span><span class="p">,</span>
    <span class="n">show</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
    <span class="n">title</span><span class="o">=</span><span class="s2">&quot;Cluster Annotations, full dataset&quot;</span><span class="p">,</span>
<span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>

<span class="c1"># plot only tumor of interest</span>
<span class="n">fig</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">figure</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">10</span><span class="p">,</span> <span class="mi">6</span><span class="p">))</span>
<span class="n">ax</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">gca</span><span class="p">()</span>
<span class="n">sc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">umap</span><span class="p">(</span><span class="n">kptracer_adata</span><span class="p">[</span><span class="n">tree</span><span class="o">.</span><span class="n">leaves</span><span class="p">,</span> <span class="p">:],</span> <span class="n">color</span><span class="o">=</span><span class="s2">&quot;Cluster-Name&quot;</span><span class="p">,</span> <span class="n">show</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">ax</span><span class="o">=</span><span class="n">ax</span><span class="p">)</span>
<span class="n">sc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">umap</span><span class="p">(</span>
    <span class="n">kptracer_adata</span><span class="p">[</span><span class="n">np</span><span class="o">.</span><span class="n">setdiff1d</span><span class="p">(</span><span class="n">kptracer_adata</span><span class="o">.</span><span class="n">obs_names</span><span class="p">,</span> <span class="n">tree</span><span class="o">.</span><span class="n">leaves</span><span class="p">),</span> <span class="p">:],</span>
    <span class="n">show</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
    <span class="n">ax</span><span class="o">=</span><span class="n">ax</span><span class="p">,</span>
    <span class="n">title</span><span class="o">=</span><span class="sa">f</span><span class="s2">&quot;Cluster Annotations, </span><span class="si">{</span><span class="n">tumor</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">,</span>
<span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="../_images/lineage_tracing_48_0.png" src="../_images/lineage_tracing_48_0.png" />
<img alt="../_images/lineage_tracing_48_1.png" src="../_images/lineage_tracing_48_1.png" />
</div>
</div>
<p>The visualization of single-cell transcriptome clusters on the UMAP reveals a <strong>continuum of cellular states</strong> that can populate a tumor. As the authors had expected, these states had been described in previous work which suggested that the CRISPR/Cas9-based recording did not noticeably perturb tumor development.</p>
<p>We can also just subset to the cells that are in our tumor of interest and observe that even in this <strong>single tumor</strong>, cells <strong>occupy disparate transcriptional states.</strong></p>
<p>Separately, we can overlay the cellular state assignments onto the hierarchical structure of the tree and evaluate the heritability of cellular states. In visualizing, we can see that certain parts of the lineage are more mixed than others.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">cas</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">plot_matplotlib</span><span class="p">(</span><span class="n">tree</span><span class="p">,</span> <span class="n">meta_data</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;Cluster-Name&quot;</span><span class="p">])</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>(&lt;Figure size 504x504 with 1 Axes&gt;, &lt;AxesSubplot:&gt;)
</pre></div>
</div>
<img alt="../_images/lineage_tracing_50_1.png" src="../_images/lineage_tracing_50_1.png" />
</div>
</div>
<p>In the visualization above, the color bar on the outside of the tree represents the cluster annotation of each cell. If states were very stable across generations, we’d expect to see strong coherence of state assignments between adjacent cells. However, above, we see that cells often are in a different state than their closest neighbor (especially on the right-hand side of the phylogeny).</p>
<p>The <strong>instability</strong> of these cellular states has been referred to as <strong>“effective plasticity”</strong> and several algorithms can be used to quantify it. One approach is to use the Fitch-Hartigan maximum-parsimony algorithm to infer the minimum number of times that cellular states had to have changed to give rise to the observed pattern. This function is implemented in Cassiopeia and can be utilized as below:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">tree</span><span class="o">.</span><span class="n">cell_meta</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span>
    <span class="n">kptracer_adata</span><span class="o">.</span><span class="n">obs</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">tree</span><span class="o">.</span><span class="n">leaves</span><span class="p">,</span> <span class="s2">&quot;Cluster-Name&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="nb">str</span><span class="p">)</span>
<span class="p">)</span>

<span class="n">parsimony</span> <span class="o">=</span> <span class="n">cas</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">score_small_parsimony</span><span class="p">(</span><span class="n">tree</span><span class="p">,</span> <span class="n">meta_item</span><span class="o">=</span><span class="s2">&quot;Cluster-Name&quot;</span><span class="p">)</span>

<span class="n">plasticity</span> <span class="o">=</span> <span class="n">parsimony</span> <span class="o">/</span> <span class="nb">len</span><span class="p">(</span><span class="n">tree</span><span class="o">.</span><span class="n">nodes</span><span class="p">)</span>

<span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Observed effective plasticity score of </span><span class="si">{</span><span class="n">plasticity</span><span class="si">}</span><span class="s2">.&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Observed effective plasticity score of 0.2356902356902357.
</pre></div>
</div>
</div>
</div>
<p>This effective plasticity score is a score between 0 and 1 indicating how mixed the lineage is overall. A score of 0 would indicate no mixture at all and 1 would indicate that every cell has a different state than its sisters.</p>
<p>Often, we are more concerned with single-cell measurements of effective plasticity. We can compute a single-cell plasticity score as below:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># compute plasticities for each node in the tree</span>
<span class="k">for</span> <span class="n">node</span> <span class="ow">in</span> <span class="n">tree</span><span class="o">.</span><span class="n">depth_first_traverse_nodes</span><span class="p">():</span>
    <span class="n">effective_plasticity</span> <span class="o">=</span> <span class="n">cas</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">score_small_parsimony</span><span class="p">(</span>
        <span class="n">tree</span><span class="p">,</span> <span class="n">meta_item</span><span class="o">=</span><span class="s2">&quot;Cluster-Name&quot;</span><span class="p">,</span> <span class="n">root</span><span class="o">=</span><span class="n">node</span>
    <span class="p">)</span>
    <span class="n">size_of_subtree</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">tree</span><span class="o">.</span><span class="n">leaves_in_subtree</span><span class="p">(</span><span class="n">node</span><span class="p">))</span>
    <span class="n">tree</span><span class="o">.</span><span class="n">set_attribute</span><span class="p">(</span>
        <span class="n">node</span><span class="p">,</span> <span class="s2">&quot;effective_plasticity&quot;</span><span class="p">,</span> <span class="n">effective_plasticity</span> <span class="o">/</span> <span class="n">size_of_subtree</span>
    <span class="p">)</span>

<span class="n">tree</span><span class="o">.</span><span class="n">cell_meta</span><span class="p">[</span><span class="s2">&quot;scPlasticity&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="mi">0</span>
<span class="k">for</span> <span class="n">leaf</span> <span class="ow">in</span> <span class="n">tree</span><span class="o">.</span><span class="n">leaves</span><span class="p">:</span>
    <span class="n">plasticities</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="n">parent</span> <span class="o">=</span> <span class="n">tree</span><span class="o">.</span><span class="n">parent</span><span class="p">(</span><span class="n">leaf</span><span class="p">)</span>
    <span class="k">while</span> <span class="kc">True</span><span class="p">:</span>
        <span class="n">plasticities</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">tree</span><span class="o">.</span><span class="n">get_attribute</span><span class="p">(</span><span class="n">parent</span><span class="p">,</span> <span class="s2">&quot;effective_plasticity&quot;</span><span class="p">))</span>
        <span class="k">if</span> <span class="n">parent</span> <span class="o">==</span> <span class="n">tree</span><span class="o">.</span><span class="n">root</span><span class="p">:</span>
            <span class="k">break</span>
        <span class="n">parent</span> <span class="o">=</span> <span class="n">tree</span><span class="o">.</span><span class="n">parent</span><span class="p">(</span><span class="n">parent</span><span class="p">)</span>

    <span class="n">tree</span><span class="o">.</span><span class="n">cell_meta</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">leaf</span><span class="p">,</span> <span class="s2">&quot;scPlasticity&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">mean</span><span class="p">(</span><span class="n">plasticities</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">cas</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">plot_matplotlib</span><span class="p">(</span><span class="n">tree</span><span class="p">,</span> <span class="n">meta_data</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;scPlasticity&quot;</span><span class="p">])</span>

<span class="n">kptracer_adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s2">&quot;scPlasticity&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">nan</span>
<span class="n">kptracer_adata</span><span class="o">.</span><span class="n">obs</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">tree</span><span class="o">.</span><span class="n">leaves</span><span class="p">,</span> <span class="s2">&quot;scPlasticity&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">tree</span><span class="o">.</span><span class="n">cell_meta</span><span class="p">[</span><span class="s2">&quot;scPlasticity&quot;</span><span class="p">]</span>

<span class="c1"># plot only tumor of interest</span>
<span class="n">fig</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">figure</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">10</span><span class="p">,</span> <span class="mi">6</span><span class="p">))</span>
<span class="n">ax</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">gca</span><span class="p">()</span>
<span class="n">sc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">umap</span><span class="p">(</span><span class="n">kptracer_adata</span><span class="p">[</span><span class="n">tree</span><span class="o">.</span><span class="n">leaves</span><span class="p">,</span> <span class="p">:],</span> <span class="n">color</span><span class="o">=</span><span class="s2">&quot;scPlasticity&quot;</span><span class="p">,</span> <span class="n">show</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">ax</span><span class="o">=</span><span class="n">ax</span><span class="p">)</span>
<span class="n">sc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">umap</span><span class="p">(</span>
    <span class="n">kptracer_adata</span><span class="p">[</span><span class="n">np</span><span class="o">.</span><span class="n">setdiff1d</span><span class="p">(</span><span class="n">kptracer_adata</span><span class="o">.</span><span class="n">obs_names</span><span class="p">,</span> <span class="n">tree</span><span class="o">.</span><span class="n">leaves</span><span class="p">),</span> <span class="p">:],</span>
    <span class="n">show</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
    <span class="n">ax</span><span class="o">=</span><span class="n">ax</span><span class="p">,</span>
<span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">title</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Single-cell Effective Plasticity, </span><span class="si">{</span><span class="n">tumor</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="../_images/lineage_tracing_55_0.png" src="../_images/lineage_tracing_55_0.png" />
<img alt="../_images/lineage_tracing_55_1.png" src="../_images/lineage_tracing_55_1.png" />
</div>
</div>
<p>In comparing the patterns of <strong>inferred effective plasticity</strong> on the tree and the low-dimensional visualization, we can observe that as expected regions with more mixing between transcriptional states have higher effective plasticity. We can also observe that the effective plasticity appears to be enriched in “mid-stage” clusters like the “AT1-like” and “High-Plasticity” states.</p>
<p><strong>For additional discussion into these patterns, we refer the reader to the original study <span id="id66">[<a class="reference internal" href="#id265" title="Dian Yang, Matthew G Jones, Santiago Naranjo, William M Rideout III, Kyung Hoi Joseph Min, Raymond Ho, Wei Wu, Joseph M Replogle, Jennifer L Page, Jeffrey J Quinn, and others. Lineage tracing reveals the phylodynamics, plasticity, and paths of tumor evolution. Cell, 185(11):1905–1923, 2022.">Yang <em>et al.</em>, 2022</a>]</span>.</strong></p>
</section>
</section>
</section>
<section id="conclusions">
<h2><span class="section-number">16.6. </span>Conclusions<a class="headerlink" href="#conclusions" title="Permalink to this headline">#</a></h2>
<p>In this chapter we have provided an overview of lineage tracing technologies and presented a case study for CRISPR/Cas9-based lineage tracing analysis using a recent dataset. Here, in closing, we point to additional <strong>resources</strong> for lineage tracing analysis and tool development as well as highlight the <strong>key takeaways</strong> for new users.</p>
<p>As this is an emerging field we expect to see a rise in the number of large-scale time-series lineage traced datasets which and that novel computational tools tailored for the analysis of this data will be developed, <span id="id67">[<a class="reference internal" href="#id273" title="Alejo Rodriguez-Fraticelli and Samantha A Morris. In preprints: the fast-paced field of single-cell lineage tracing. Development, 149(11):dev200877, 2022.">Rodriguez-Fraticelli and Morris, 2022</a>]</span> and <span id="id68">[<a class="reference internal" href="#id267" title="Madhura Mukhopadhyay. Tracing cell relationships. Nature Methods, 19(1):27–27, 2022.">Mukhopadhyay, 2022</a>]</span>. We anticipate that the main focus will be put into methods which integrate gene-expression with lineage information hence providing a more complete biological picture of individuals. Further, following the new-directions detailed below we expect that methods will focus on trajectory inference from time series data.</p>
</section>
<section id="new-directions">
<h2><span class="section-number">16.7. </span>New directions<a class="headerlink" href="#new-directions" title="Permalink to this headline">#</a></h2>
<section id="new-phylogenetic-inference-algorithms">
<h3><span class="section-number">16.7.1. </span>New phylogenetic inference algorithms<a class="headerlink" href="#new-phylogenetic-inference-algorithms" title="Permalink to this headline">#</a></h3>
<p>In terms of new lineage inference algorithms, there are many promising directions:</p>
<ul class="simple">
<li><p><strong>Scalable bayesian inference</strong>: Mirroring the trends of more traditional phylogenetic algorithms, one potential direction is that of scaling Bayesian approaches to larger inputs. While most Bayesian algorithms have leveraged Markov chain Monte Carlo (MCMC) to estimate the posterior distribution <span id="id69">[<a class="reference internal" href="#id233" title="John P Huelsenbeck, Fredrik Ronquist, Rasmus Nielsen, and Jonathan P Bollback. Bayesian inference of phylogeny and its impact on evolutionary biology. science, 294(5550):2310–2314, 2001.">Huelsenbeck <em>et al.</em>, 2001</a>]</span>, advances in variational inference would greatly improve the scalability of Bayesian algorithms <span id="id70">[<a class="reference internal" href="#id226" title="Cheng Zhang and Frederick A Matsen IV. Variational bayesian phylogenetic inference. In International Conference on Learning Representations. 2018.">Zhang and Matsen IV, 2018</a>]</span>. The probablistic nature of such an advance would support high-throughput uncertainty estimation of the tree as well as fit naturally with other single-cell transcriptomic Bayesian approaches, like scVI <span id="id71">[<a class="reference internal" href="#id280" title="Romain Lopez, Jeffrey Regier, Michael B. Cole, Michael I. Jordan, and Nir Yosef. Deep generative modeling for single-cell transcriptomics. Nature Methods, 15(12):1053-1058, Dec 2018. URL: https://doi.org/10.1038/s41592-018-0229-2, doi:10.1038/s41592-018-0229-2.">Lopez <em>et al.</em>, 2018</a>]</span>.</p></li>
<li><p><strong>Improved distance-based algorithms</strong>: A fundamental aspect of distance-based algorithms is the estimation of dissimilarities between samples based on their mutation data and known properties of the dataset. With this in mind, a promising direction is that of developing more statistically robust and consistent dissimilarity functions for evolving lineage tracers by taking into account properites of how mutations arise and the priors on the likelihoods of specific mutations. Already, this has proven successful with DCLEAR <span id="id72">[<a class="reference internal" href="#id237" title="Wuming Gong, Alejandro A Granados, Jingyuan Hu, Matthew G Jones, Ofir Raz, Irepan Salvador-Martínez, Hanrui Zhang, Ke-Huan K Chow, Il-Youp Kwak, Renata Retkute, and others. Benchmarked approaches for reconstruction of in vitro cell lineages and in silico models of c. elegans and m. musculus developmental trees. Cell systems, 12(8):810–826, 2021.">Gong <em>et al.</em>, 2021</a>]</span> and <span id="id73">[<a class="reference internal" href="#id225" title="Weixiang Fang, Claire M Bell, Abel Sapirstein, Soichiro Asami, Kathleen Leeper, Donald J Zack, Hongkai Ji, and Reza Kalhor. Quantitative fate mapping: reconstructing progenitor field dynamics via retrospective lineage barcoding. bioRxiv, 2022.">Fang <em>et al.</em>, 2022</a>]</span>. Continuing advances in this realm would be greatly enabling as distance-based algorithms can run in polynomial time and produce very accurate trees with an adequate dissimilarity function.</p></li>
</ul>
</section>
<section id="computational-tools-to-interpret-time-course-lineage-tracing-data">
<h3><span class="section-number">16.7.2. </span>Computational tools to interpret time-course lineage tracing data<a class="headerlink" href="#computational-tools-to-interpret-time-course-lineage-tracing-data" title="Permalink to this headline">#</a></h3>
<p>The increasing complexity of lineage tracing studies which include time-course information must be accompanied by computational methods which extend the analysis beyond the construction of phylogenetic trees. That is, methods that integrate multi-omics measurements with lineage tracing and temporal information to enable recovery of programs governing cell state, differentiation and behavior <span id="id74">[<a class="reference internal" href="#id267" title="Madhura Mukhopadhyay. Tracing cell relationships. Nature Methods, 19(1):27–27, 2022.">Mukhopadhyay, 2022</a>]</span>.</p>
<p>As this field is at its beginning the number of available tools is still limited, yet it is worth highlighting leading approaches:</p>
<ul class="simple">
<li><p><strong><code class="docutils literal notranslate"><span class="pre">LineageOT</span></code> for evolving CRISPR/Cas9-based setting</strong> <span id="id75">[<a class="reference internal" href="#id268" title="Aden Forrow and Geoffrey Schiebinger. Lineageot is a unified framework for lineage tracing and trajectory inference. Nature communications, 12(1):1–10, 2021.">Forrow and Schiebinger, 2021</a>]</span>: A general-purpose method for inferring developmental trajectories from scRNA-seq time courses equipped with lineage information each time point applicable for evolving CRISPR/Cas9-based setting. The method was suggested as an extension of the Waddington OT <span id="id76">[<a class="reference internal" href="#id269" title="Geoffrey Schiebinger, Jian Shu, Marcin Tabaka, Brian Cleary, Vidya Subramanian, Aryeh Solomon, Joshua Gould, Siyan Liu, Stacie Lin, Peter Berube, and others. Optimal-transport analysis of single-cell gene expression identifies developmental trajectories in reprogramming. Cell, 176(4):928–943, 2019.">Schiebinger <em>et al.</em>, 2019</a>]</span> algorithm to take the lineage relationships into account when mapping cells from earlier to later time-points. When computing transition matrices across pairs of time-points, LineageOT corrects expression profiles in the later time-point based on their lineage similarity. This method has been successfully applied for the reconstruction of time-course of <em>C. elegans</em> development. Further, through simulation it was proven that <code class="docutils literal notranslate"><span class="pre">LineageOT</span></code> can accurately recover complex trajectory structures that are impossible to recover from measurements of cell state alone. For more details and tutorials we refer the reader to <a class="reference external" href="https://lineageot.readthedocs.io">https://lineageot.readthedocs.io</a>.</p></li>
<li><p><strong><code class="docutils literal notranslate"><span class="pre">CoSpar</span></code> for static barcoding lineage tracing data</strong> <span id="id77">[<a class="reference internal" href="#id270" title="Shou-Wen Wang, Michael J Herriges, Kilian Hurley, Darrell N Kotton, and Allon M Klein. Cospar identifies early cell fate biases from single-cell transcriptomic and lineage information. Nature Biotechnology, pages 1–9, 2022.">Wang <em>et al.</em>, 2022</a>]</span>: A computational approach to infer cell dynamics from single-cell transcriptomics integrated with static barcoding lineage tracing data. The method relies on two basic assumptions on the nature of biological dynamics: (i) cells in similar states behave similarly and (ii) cells limit their possible dynamics to give sparse transitions. The application of <code class="docutils literal notranslate"><span class="pre">CoSpar</span></code> is demonstrated on hematopoiesis, reprogramming and directed differentiation datasets. These examples show that <code class="docutils literal notranslate"><span class="pre">CoSpar</span></code> can identify early fate biases not previously detected, predicting transcription factors and receptors implicated in fate choice. Documentation and detailed examples can be found at <a class="reference external" href="https://cospar.readthedocs.io/">https://cospar.readthedocs.io/</a>.</p></li>
</ul>
</section>
<section id="resources-for-developing-new-computational-methods">
<h3><span class="section-number">16.7.3. </span>Resources for developing new computational methods<a class="headerlink" href="#resources-for-developing-new-computational-methods" title="Permalink to this headline">#</a></h3>
<p>While the current repertoire of analytical tools are quite powerful, there is a great opportunity for further development. To facilitate future efforts, we highlight a few tools for benchmarking algorithms:</p>
<ul class="simple">
<li><p>The recent Allen Institute Lineage Reconstruction DREAM challenge <span id="id78">[<a class="reference internal" href="#id237" title="Wuming Gong, Alejandro A Granados, Jingyuan Hu, Matthew G Jones, Ofir Raz, Irepan Salvador-Martínez, Hanrui Zhang, Ke-Huan K Chow, Il-Youp Kwak, Renata Retkute, and others. Benchmarked approaches for reconstruction of in vitro cell lineages and in silico models of c. elegans and m. musculus developmental trees. Cell systems, 12(8):810–826, 2021.">Gong <em>et al.</em>, 2021</a>]</span> generated three benchmarking datasets that were used to evaluate new algorithms. Two of these datasets were synthetic (i.e., simulated) whereas one was generated with the intMEMOIR <span id="id79">[<a class="reference internal" href="#id271" title="Ke-Huan K Chow, Mark W Budde, Alejandro A Granados, Maria Cabrera, Shinae Yoon, Soomin Cho, Ting-hao Huang, Noushin Koulena, Kirsten L Frieda, Long Cai, and others. Imaging cell lineage with a synthetic digital recording system. Science, 372(6538):eabb3099, 2021.">Chow <em>et al.</em>, 2021</a>]</span> technology. The recently published results offer insights into the behavior of successful algorithms as well as provides access to these benchmarking datasets <span id="id80">[<a class="reference internal" href="#id237" title="Wuming Gong, Alejandro A Granados, Jingyuan Hu, Matthew G Jones, Ofir Raz, Irepan Salvador-Martínez, Hanrui Zhang, Ke-Huan K Chow, Il-Youp Kwak, Renata Retkute, and others. Benchmarked approaches for reconstruction of in vitro cell lineages and in silico models of c. elegans and m. musculus developmental trees. Cell systems, 12(8):810–826, 2021.">Gong <em>et al.</em>, 2021</a>]</span>.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">Cassiopeia-benchmark</span></code> is a module within Cassiopeia that allows users to efficiently generate simulated lineage data for benchmarking new lineage reconstruction algorithms. While there is a strong focus on CRISPR/Cas9-based evolving tracer simulations, the general simulator API can be extended for additional technological considerations. The authors have provided a detailed walkthrough of using this benchmarking suite on <a class="reference external" href="https://cassiopeia-lineage.readthedocs.io/en/latest/notebooks/benchmark.html">their website</a>.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">TedSim</span></code> is a simulation framework for not only simulating barcoding data but also transcriptomic data on top of a lineage <span id="id81">[<a class="reference internal" href="#id272" title="Xinhai Pan, Hechen Li, and Xiuwei Zhang. Tedsim: temporal dynamics simulation of single-cell term`rna` sequencing data and cell division history. Nucleic acids research, 50(8):4272–4288, 2022.">Pan <em>et al.</em>, 2022</a>]</span>. This simulation framework can be very useful for testing downstream analytical tools, such as those designed to infer developmental trajectories from lineage tracing data.</p></li>
</ul>
</section>
</section>
<section id="key-takeaways">
<h2><span class="section-number">16.8. </span>Key takeaways<a class="headerlink" href="#key-takeaways" title="Permalink to this headline">#</a></h2>
<ul class="simple">
<li><p>Remember, the <strong>resolution of experimental data</strong> is still limited. <strong>Statistical analysis</strong> and <strong>assessment of the quality</strong> of lineage information shall be performed in order to identify individuals for which lineage reconstruction can be applied. As demonstrated above there are dedicated tools which easily provide the relevant statistics.</p></li>
<li><p>Often it is difficult to presume <strong>which algorithm</strong> will perform best for reconstruction, and even algorithms with good performances might highlight different part of lineages. It is recommended that you <strong>apply a few algorithms</strong> for reconstruction for comparisons.</p></li>
<li><p>As with any analysis without a certain ground truth, any <strong>conclusions must be carefully drawn</strong>. First and foremost, it is always good practice to visualize lineages along with the indel heatmap to verify that no obvious reconstruction errors were made (see the section entitled “Reconstructing the lineage” for more tips on how to evaluate reconstruction accuracy by eye). Additionally, an effective strategy for drawing stronger conclusions is to compare downstream analyses on the same lineage inferred with different algorithms. Of course, as with most matters concerning computational conclusions, we often suggest that conclusions are validated in an experimental system.</p></li>
</ul>
</section>
<section id="references">
<h2><span class="section-number">16.9. </span>References<a class="headerlink" href="#references" title="Permalink to this headline">#</a></h2>
<div class="docutils container" id="id82">
<dl class="citation">
<dt class="label" id="id256"><span class="brackets"><a class="fn-backref" href="#id21">AV20</a></span></dt>
<dd><p>Alexej Abyzov and Flora M Vaccarino. Cell lineage tracing and cellular diversity in humans. <em>Annual Review of Genomics and Human Genetics</em>, 21:101–116, 2020.</p>
</dd>
<dt class="label" id="id263"><span class="brackets"><a class="fn-backref" href="#id33">AFB+18</a></span></dt>
<dd><p>Anna Alemany, Maria Florescu, Chloé S Baron, Josi Peterson-Maduro, and Alexander Van Oudenaarden. Whole-organism clone tracing using single-cell sequencing. <em>Nature</em>, 556(7699):108–112, 2018.</p>
</dd>
<dt class="label" id="id254"><span class="brackets"><a class="fn-backref" href="#id19">BBR+21</a></span></dt>
<dd><p>Chris Bailey, James RM Black, James L Reading, Kevin Litchfield, Samra Turajlic, Nicholas McGranahan, Mariam Jamal-Hanjani, and Charles Swanton. Tracking cancer evolution through the disease course. <em>Cancer discovery</em>, 11(4):916–932, 2021.</p>
</dd>
<dt class="label" id="id248"><span class="brackets"><a class="fn-backref" href="#id14">BKK+18</a></span></dt>
<dd><p>Brent A Biddy, Wenjun Kong, Kenji Kamimoto, Chuner Guo, Sarah E Waye, Tao Sun, and Samantha A Morris. Single-cell mapping of lineage and identity in direct reprogramming. <em>Nature</em>, 564(7735):219–224, 2018.</p>
</dd>
<dt class="label" id="id276"><span class="brackets"><a class="fn-backref" href="#id22">BDG+21</a></span></dt>
<dd><p>Sara Bizzotto, Yanmei Dou, Javier Ganz, Ryan N Doan, Minseok Kwon, Craig L Bohrson, Sonia N Kim, Taejeong Bae, Alexej Abyzov, NIMH Brain Somatic Mosaicism Network†, and others. Landmarks of human embryonic development inscribed in somatic mutations. <em>Science</em>, 371(6535):1249–1253, 2021.</p>
</dd>
<dt class="label" id="id235"><span class="brackets"><a class="fn-backref" href="#id42">CSE63</a></span></dt>
<dd><p>LL Cavalli-Sforza and AWF Edwards. The reconstruction of evolution. <em>Ann. Hum. Genet</em>, 27:105–106, 1963.</p>
</dd>
<dt class="label" id="id260"><span class="brackets">CSG+19</span><span class="fn-backref">(<a href="#id30">1</a>,<a href="#id37">2</a>)</span></dt>
<dd><p>Michelle M Chan, Zachary D Smith, Stefanie Grosswendt, Helene Kretzmer, Thomas M Norman, Britt Adamson, Marco Jost, Jeffrey J Quinn, Dian Yang, Matthew G Jones, and others. Molecular recording of mammalian embryogenesis. <em>Nature</em>, 570(7759):77–82, 2019.</p>
</dd>
<dt class="label" id="id279"><span class="brackets"><a class="fn-backref" href="#id35">CCM+22</a></span></dt>
<dd><p>Junhong Choi, Wei Chen, Anna Minkina, Florence M. Chardon, Chase C. Suiter, Samuel G. Regalado, Silvia Domcke, Nobuhiko Hamazaki, Choli Lee, Beth Martin, Riza M. Daza, and Jay Shendure. A time-resolved, multi-symbol molecular recorder via sequential genome editing. <em>Nature</em>, 608(7921):98–107, Aug 2022. URL: <a class="reference external" href="https://doi.org/10.1038/s41586-022-04922-8">https://doi.org/10.1038/s41586-022-04922-8</a>, <a class="reference external" href="https://doi.org/10.1038/s41586-022-04922-8">doi:10.1038/s41586-022-04922-8</a>.</p>
</dd>
<dt class="label" id="id271"><span class="brackets"><a class="fn-backref" href="#id79">CBG+21</a></span></dt>
<dd><p>Ke-Huan K Chow, Mark W Budde, Alejandro A Granados, Maria Cabrera, Shinae Yoon, Soomin Cho, Ting-hao Huang, Noushin Koulena, Kirsten L Frieda, Long Cai, and others. Imaging cell lineage with a synthetic digital recording system. <em>Science</em>, 372(6538):eabb3099, 2021.</p>
</dd>
<dt class="label" id="id266"><span class="brackets"><a class="fn-backref" href="#id54">DDJ09</a></span></dt>
<dd><p>Michel DuPage, Alison L Dooley, and Tyler Jacks. Conditional mouse lung cancer models using adenoviral or lentiviral delivery of cre recombinase. <em>Nature protocols</em>, 4(7):1064–1072, 2009.</p>
</dd>
<dt class="label" id="id225"><span class="brackets"><a class="fn-backref" href="#id73">FBS+22</a></span></dt>
<dd><p>Weixiang Fang, Claire M Bell, Abel Sapirstein, Soichiro Asami, Kathleen Leeper, Donald J Zack, Hongkai Ji, and Reza Kalhor. Quantitative fate mapping: reconstructing progenitor field dynamics via retrospective lineage barcoding. <em>bioRxiv</em>, 2022.</p>
</dd>
<dt class="label" id="id234"><span class="brackets"><a class="fn-backref" href="#id43">Fel81</a></span></dt>
<dd><p>Joseph Felsenstein. Evolutionary trees from term`dna` sequences: a maximum likelihood approach. <em>Journal of molecular evolution</em>, 17(6):368–376, 1981.</p>
</dd>
<dt class="label" id="id228"><span class="brackets">FDIM+21</span><span class="fn-backref">(<a href="#id44">1</a>,<a href="#id49">2</a>)</span></dt>
<dd><p>Jean Feng, William S DeWitt III, Aaron McKenna, Noah Simon, Amy D Willis, and Frederick A Matsen IV. Estimation of cell lineage trees by maximum-likelihood phylogenetics. <em>The Annals of Applied Statistics</em>, 15(1):343–362, 2021.</p>
</dd>
<dt class="label" id="id268"><span class="brackets"><a class="fn-backref" href="#id75">FS21</a></span></dt>
<dd><p>Aden Forrow and Geoffrey Schiebinger. Lineageot is a unified framework for lineage tracing and trajectory inference. <em>Nature communications</em>, 12(1):1–10, 2021.</p>
</dd>
<dt class="label" id="id261"><span class="brackets"><a class="fn-backref" href="#id31">FLH+17</a></span></dt>
<dd><p>Kirsten L Frieda, James M Linton, Sahand Hormoz, Joonhyuk Choi, Ke-Huan K Chow, Zakary S Singer, Mark W Budde, Michael B Elowitz, and Long Cai. Synthetic recording and in situ readout of lineage information in single cells. <em>Nature</em>, 541(7635):107–111, 2017.</p>
</dd>
<dt class="label" id="id274"><span class="brackets"><a class="fn-backref" href="#id25">GBH+21</a></span></dt>
<dd><p>Ruli Gao, Shanshan Bai, Ying C Henderson, Yiyun Lin, Aislyn Schalck, Yun Yan, Tapsi Kumar, Min Hu, Emi Sei, Alexander Davis, and others. Delineating copy number and clonal substructure in human tumors from single-cell transcriptomes. <em>Nature biotechnology</em>, 39(5):599–608, 2021.</p>
</dd>
<dt class="label" id="id247"><span class="brackets"><a class="fn-backref" href="#id13">GDK+10</a></span></dt>
<dd><p>Alice Gerrits, Brad Dykstra, Olga J Kalmykowa, Karin Klauke, Evgenia Verovskaya, Mathilde JC Broekhuis, Gerald de Haan, and Leonid V Bystrykh. Cellular barcoding tool for clonal analysis in the hematopoietic system. <em>Blood, The Journal of the American Society of Hematology</em>, 115(13):2610–2618, 2010.</p>
</dd>
<dt class="label" id="id255"><span class="brackets"><a class="fn-backref" href="#id20">GJL+20</a></span></dt>
<dd><p>Moritz Gerstung, Clemency Jolly, Ignaty Leshchiner, Stefan C Dentro, Santiago Gonzalez, Daniel Rosebrock, Thomas J Mitchell, Yulia Rubanova, Pavana Anur, Kaixian Yu, and others. The evolutionary history of 2,658 cancers. <em>Nature</em>, 578(7793):122–128, 2020.</p>
</dd>
<dt class="label" id="id237"><span class="brackets">GGH+21</span><span class="fn-backref">(<a href="#id2">1</a>,<a href="#id50">2</a>,<a href="#id72">3</a>,<a href="#id78">4</a>,<a href="#id80">5</a>)</span></dt>
<dd><p>Wuming Gong, Alejandro A Granados, Jingyuan Hu, Matthew G Jones, Ofir Raz, Irepan Salvador-Martínez, Hanrui Zhang, Ke-Huan K Chow, Il-Youp Kwak, Renata Retkute, and others. Benchmarked approaches for reconstruction of in vitro cell lineages and in silico models of c. elegans and m. musculus developmental trees. <em>Cell systems</em>, 12(8):810–826, 2021.</p>
</dd>
<dt class="label" id="id229"><span class="brackets"><a class="fn-backref" href="#id58">Gus91</a></span></dt>
<dd><p>Dan Gusfield. Efficient algorithms for inferring evolutionary trees. <em>Networks</em>, 21(1):19–28, 1991.</p>
</dd>
<dt class="label" id="id246"><span class="brackets"><a class="fn-backref" href="#id12">HLL+17</a></span></dt>
<dd><p>Lingjuan He, Yan Li, Yi Li, Wenjuan Pu, Xiuzhen Huang, Xueying Tian, Yue Wang, Hui Zhang, Qiaozhen Liu, Libo Zhang, and others. Enhancing the precision of genetic lineage tracing using dual recombinases. <em>Nature medicine</em>, 23(12):1488–1498, 2017.</p>
</dd>
<dt class="label" id="id233"><span class="brackets">HRNB01</span><span class="fn-backref">(<a href="#id45">1</a>,<a href="#id69">2</a>)</span></dt>
<dd><p>John P Huelsenbeck, Fredrik Ronquist, Rasmus Nielsen, and Jonathan P Bollback. Bayesian inference of phylogeny and its impact on evolutionary biology. <em>science</em>, 294(5550):2310–2314, 2001.</p>
</dd>
<dt class="label" id="id264"><span class="brackets">JKQ+20</span><span class="fn-backref">(<a href="#id41">1</a>,<a href="#id48">2</a>,<a href="#id52">3</a>,<a href="#id59">4</a>,<a href="#id62">5</a>,<a href="#id63">6</a>)</span></dt>
<dd><p>Matthew G Jones, Alex Khodaverdian, Jeffrey J Quinn, Michelle M Chan, Jeffrey A Hussmann, Robert Wang, Chenling Xu, Jonathan S Weissman, and Nir Yosef. Inference of single-cell phylogenies from lineage tracing data using cassiopeia. <em>Genome biology</em>, 21(1):1–27, 2020.</p>
</dd>
<dt class="label" id="id277"><span class="brackets"><a class="fn-backref" href="#id23">JMG+17</a></span></dt>
<dd><p>Young Seok Ju, Inigo Martincorena, Moritz Gerstung, Mia Petljak, Ludmil B Alexandrov, Raheleh Rahbari, David C Wedge, Helen R Davies, Manasa Ramakrishna, Anthony Fullam, and others. Somatic mutations reveal asymmetric cellular dynamics in the early human embryo. <em>Nature</em>, 543(7647):714–718, 2017.</p>
</dd>
<dt class="label" id="id262"><span class="brackets"><a class="fn-backref" href="#id32">KKM+18</a></span></dt>
<dd><p>Reza Kalhor, Kian Kalhor, Leo Mejia, Kathleen Leeper, Amanda Graveline, Prashant Mali, and George M Church. Developmental barcoding of whole mouse via homing crispr. <em>Science</em>, 361(6405):eaat9804, 2018.</p>
</dd>
<dt class="label" id="id227"><span class="brackets"><a class="fn-backref" href="#id51">KKW+22</a></span></dt>
<dd><p>Naoki Konno, Yusuke Kijima, Keito Watano, Soh Ishiguro, Keiichiro Ono, Mamoru Tanaka, Hideto Mori, Nanami Masuyama, Dexter Pratt, Trey Ideker, and others. Deep distributed computing to reconstruct extremely large lineage trees. <em>Nature Biotechnology</em>, 40(4):566–575, 2022.</p>
</dd>
<dt class="label" id="id244"><span class="brackets"><a class="fn-backref" href="#id10">LJZ20</a></span></dt>
<dd><p>Kuo Liu, Hengwei Jin, and Bin Zhou. Genetic lineage tracing with multiple term`dna` recombinases: a user's guide for conducting more precise cell fate mapping studies. <em>Journal of Biological Chemistry</em>, 295(19):6413–6424, 2020.</p>
</dd>
<dt class="label" id="id245"><span class="brackets"><a class="fn-backref" href="#id11">LTJ+20</a></span></dt>
<dd><p>Kuo Liu, Muxue Tang, Hengwei Jin, Qiaozhen Liu, Lingjuan He, Huan Zhu, Xiuxiu Liu, Ximeng Han, Yan Li, Libo Zhang, and others. Triple-cell lineage tracing by a dual reporter on a single allele. <em>Journal of Biological Chemistry</em>, 295(3):690–700, 2020.</p>
</dd>
<dt class="label" id="id280"><span class="brackets"><a class="fn-backref" href="#id71">LRC+18</a></span></dt>
<dd><p>Romain Lopez, Jeffrey Regier, Michael B. Cole, Michael I. Jordan, and Nir Yosef. Deep generative modeling for single-cell transcriptomics. <em>Nature Methods</em>, 15(12):1053–1058, Dec 2018. URL: <a class="reference external" href="https://doi.org/10.1038/s41592-018-0229-2">https://doi.org/10.1038/s41592-018-0229-2</a>, <a class="reference external" href="https://doi.org/10.1038/s41592-018-0229-2">doi:10.1038/s41592-018-0229-2</a>.</p>
</dd>
<dt class="label" id="id278"><span class="brackets"><a class="fn-backref" href="#id34">LGS+21</a></span></dt>
<dd><p>Theresa B. Loveless, Joseph H. Grotts, Mason W. Schechter, Elmira Forouzmand, Courtney K. Carlson, Bijan S. Agahi, Guohao Liang, Michelle Ficht, Beide Liu, Xiaohui Xie, and Chang C. Liu. Lineage tracing and analog recording in mammalian cells by single-site term`dna` writing. <em>Nature Chemical Biology</em>, 17(6):739–747, Jun 2021. URL: <a class="reference external" href="https://doi.org/10.1038/s41589-021-00769-8">https://doi.org/10.1038/s41589-021-00769-8</a>, <a class="reference external" href="https://doi.org/10.1038/s41589-021-00769-8">doi:10.1038/s41589-021-00769-8</a>.</p>
</dd>
<dt class="label" id="id257"><span class="brackets"><a class="fn-backref" href="#id28">MFG+16</a></span></dt>
<dd><p>Aaron McKenna, Gregory M Findlay, James A Gagnon, Marshall S Horwitz, Alexander F Schier, and Jay Shendure. Whole-organism lineage tracing by combinatorial and cumulative genome editing. <em>Science</em>, 353(6298):aaf7907, 2016.</p>
</dd>
<dt class="label" id="id239"><span class="brackets">MG19</span><span class="fn-backref">(<a href="#id3">1</a>,<a href="#id27">2</a>)</span></dt>
<dd><p>Aaron McKenna and James A Gagnon. Recording development with single cell dynamic lineage tracing. <em>Development</em>, 146(12):dev169730, 2019.</p>
</dd>
<dt class="label" id="id267"><span class="brackets">Muk22</span><span class="fn-backref">(<a href="#id68">1</a>,<a href="#id74">2</a>)</span></dt>
<dd><p>Madhura Mukhopadhyay. Tracing cell relationships. <em>Nature Methods</em>, 19(1):27–27, 2022.</p>
</dd>
<dt class="label" id="id243"><span class="brackets"><a class="fn-backref" href="#id9">Nag00</a></span></dt>
<dd><p>Andras Nagy. Cre recombinase: the universal reagent for genome tailoring. <em>genesis</em>, 26(2):99–109, 2000.</p>
</dd>
<dt class="label" id="id272"><span class="brackets"><a class="fn-backref" href="#id81">PLZ22</a></span></dt>
<dd><p>Xinhai Pan, Hechen Li, and Xiuwei Zhang. Tedsim: temporal dynamics simulation of single-cell term`rna` sequencing data and cell division history. <em>Nucleic acids research</em>, 50(8):4272–4288, 2022.</p>
</dd>
<dt class="label" id="id275"><span class="brackets"><a class="fn-backref" href="#id24">PTT+14</a></span></dt>
<dd><p>Anoop P Patel, Itay Tirosh, John J Trombetta, Alex K Shalek, Shawn M Gillespie, Hiroaki Wakimoto, Daniel P Cahill, Brian V Nahed, William T Curry, Robert L Martuza, and others. Single-cell rna-seq highlights intratumoral heterogeneity in primary glioblastoma. <em>Science</em>, 344(6190):1396–1401, 2014.</p>
</dd>
<dt class="label" id="id230"><span class="brackets"><a class="fn-backref" href="#id56">QJO+21</a></span></dt>
<dd><p>Jeffrey J Quinn, Matthew G Jones, Ross A Okimoto, Shigeki Nanjo, Michelle M Chan, Nir Yosef, Trever G Bivona, and Jonathan S Weissman. Single-cell lineages reveal the rates, routes, and drivers of metastasis in cancer xenografts. <em>Science</em>, 371(6532):eabc1944, 2021.</p>
</dd>
<dt class="label" id="id258"><span class="brackets"><a class="fn-backref" href="#id36">RGS18</a></span></dt>
<dd><p>Bushra Raj, James A Gagnon, and Alexander F Schier. Large-scale reconstruction of cell lineages using single-cell readout of transcriptomes and crispr–cas9 barcodes by scgestalt. <em>Nature protocols</em>, 13(11):2685–2713, 2018.</p>
</dd>
<dt class="label" id="id273"><span class="brackets"><a class="fn-backref" href="#id67">RFM22</a></span></dt>
<dd><p>Alejo Rodriguez-Fraticelli and Samantha A Morris. In preprints: the fast-paced field of single-cell lineage tracing. <em>Development</em>, 149(11):dev200877, 2022.</p>
</dd>
<dt class="label" id="id232"><span class="brackets">SN87</span><span class="fn-backref">(<a href="#id46">1</a>,<a href="#id60">2</a>)</span></dt>
<dd><p>Naruya Saitou and Masatoshi Nei. The neighbor-joining method: a new method for reconstructing phylogenetic trees. <em>Molecular biology and evolution</em>, 4(4):406–425, 1987.</p>
</dd>
<dt class="label" id="id269"><span class="brackets"><a class="fn-backref" href="#id76">SST+19</a></span></dt>
<dd><p>Geoffrey Schiebinger, Jian Shu, Marcin Tabaka, Brian Cleary, Vidya Subramanian, Aryeh Solomon, Joshua Gould, Siyan Liu, Stacie Lin, Peter Berube, and others. Optimal-transport analysis of single-cell gene expression identifies developmental trajectories in reprogramming. <em>Cell</em>, 176(4):928–943, 2019.</p>
</dd>
<dt class="label" id="id231"><span class="brackets">Sok58</span><span class="fn-backref">(<a href="#id47">1</a>,<a href="#id61">2</a>)</span></dt>
<dd><p>Robert R Sokal. A statistical method for evaluating systematic relationships. <em>Univ. Kansas, Sci. Bull.</em>, 38:1409–1438, 1958.</p>
</dd>
<dt class="label" id="id259"><span class="brackets">SHM+18</span><span class="fn-backref">(<a href="#id29">1</a>,<a href="#id40">2</a>)</span></dt>
<dd><p>Bastiaan Spanjaard, Bo Hu, Nina Mitic, Pedro Olivares-Chauvet, Sharan Janjuha, Nikolay Ninov, and Jan Philipp Junker. Simultaneous lineage tracing and cell-type identification using crispr–cas9-induced genetic scars. <em>Nature biotechnology</em>, 36(5):469–473, 2018.</p>
</dd>
<dt class="label" id="id241"><span class="brackets"><a class="fn-backref" href="#id6">SSWT83</a></span></dt>
<dd><p>John E Sulston, Einhard Schierenberg, John G White, and J Nichol Thomson. The embryonic cell lineage of the nematode caenorhabditis elegans. <em>Developmental biology</em>, 100(1):64–119, 1983.</p>
</dd>
<dt class="label" id="id252"><span class="brackets"><a class="fn-backref" href="#id18">TS16</a></span></dt>
<dd><p>Samra Turajlic and Charles Swanton. Metastasis as an evolutionary process. <em>Science</em>, 352(6282):169–175, 2016.</p>
</dd>
<dt class="label" id="id240"><span class="brackets"><a class="fn-backref" href="#id5">VM21</a></span></dt>
<dd><p>Sadie VanHorn and Samantha A Morris. Next-generation lineage tracing and fate mapping to interrogate development. <em>Developmental cell</em>, 56(1):7–21, 2021.</p>
</dd>
<dt class="label" id="id251"><span class="brackets"><a class="fn-backref" href="#id17">VPV+13</a></span></dt>
<dd><p>Bert Vogelstein, Nickolas Papadopoulos, Victor E Velculescu, Shibin Zhou, Luis A Diaz Jr, and Kenneth W Kinzler. Cancer genome landscapes. <em>science</em>, 339(6127):1546–1558, 2013.</p>
</dd>
<dt class="label" id="id238"><span class="brackets">WK20</span><span class="fn-backref">(<a href="#id4">1</a>,<a href="#id7">2</a>,<a href="#id26">3</a>,<a href="#id39">4</a>)</span></dt>
<dd><p>Daniel E Wagner and Allon M Klein. Lineage tracing meets single-cell omics: opportunities and challenges. <em>Nature Reviews Genetics</em>, 21(7):410–427, 2020.</p>
</dd>
<dt class="label" id="id270"><span class="brackets"><a class="fn-backref" href="#id77">WHH+22</a></span></dt>
<dd><p>Shou-Wen Wang, Michael J Herriges, Kilian Hurley, Darrell N Kotton, and Allon M Klein. Cospar identifies early cell fate biases from single-cell transcriptomic and lineage information. <em>Nature Biotechnology</em>, pages 1–9, 2022.</p>
</dd>
<dt class="label" id="id249"><span class="brackets">WRFCK20</span><span class="fn-backref">(<a href="#id15">1</a>,<a href="#id38">2</a>)</span></dt>
<dd><p>Caleb Weinreb, Alejo Rodriguez-Fraticelli, Fernando D Camargo, and Allon M Klein. Lineage tracing on transcriptional landscapes links state to fate during differentiation. <em>Science</em>, 367(6479):eaaw3381, 2020.</p>
</dd>
<dt class="label" id="id242"><span class="brackets"><a class="fn-backref" href="#id8">WP15</a></span></dt>
<dd><p>Tamily A Weissman and Y Albert Pan. Brainbow: new resources and emerging biological applications for multicolor genetic labeling and analysis. <em>Genetics</em>, 199(2):293–306, 2015.</p>
</dd>
<dt class="label" id="id236"><span class="brackets"><a class="fn-backref" href="#id1">WGW17</a></span></dt>
<dd><p>Mollie B Woodworth, Kelly M Girskis, and Christopher A Walsh. Building a lineage from single cells: genetic techniques for cell lineage tracking. <em>Nature Reviews Genetics</em>, 18(4):230–244, 2017.</p>
</dd>
<dt class="label" id="id265"><span class="brackets">YJN+22</span><span class="fn-backref">(<a href="#id53">1</a>,<a href="#id55">2</a>,<a href="#id57">3</a>,<a href="#id64">4</a>,<a href="#id65">5</a>,<a href="#id66">6</a>)</span></dt>
<dd><p>Dian Yang, Matthew G Jones, Santiago Naranjo, William M Rideout III, Kyung Hoi Joseph Min, Raymond Ho, Wei Wu, Joseph M Replogle, Jennifer L Page, Jeffrey J Quinn, and others. Lineage tracing reveals the phylodynamics, plasticity, and paths of tumor evolution. <em>Cell</em>, 185(11):1905–1923, 2022.</p>
</dd>
<dt class="label" id="id250"><span class="brackets"><a class="fn-backref" href="#id16">YMK+17</a></span></dt>
<dd><p>Zizhen Yao, John K Mich, Sherman Ku, Vilas Menon, Anne-Rachel Krostag, Refugio A Martinez, Leon Furchtgott, Heather Mulholland, Susan Bort, Margaret A Fuqua, and others. A single-cell roadmap of lineage bifurcation in human esc models of embryonic brain development. <em>Cell stem cell</em>, 20(1):120–134, 2017.</p>
</dd>
<dt class="label" id="id226"><span class="brackets"><a class="fn-backref" href="#id70">ZMI18</a></span></dt>
<dd><p>Cheng Zhang and Frederick A Matsen IV. Variational bayesian phylogenetic inference. In <em>International Conference on Learning Representations</em>. 2018.</p>
</dd>
</dl>
</div>
</section>
<section id="contributors">
<h2><span class="section-number">16.10. </span>Contributors<a class="headerlink" href="#contributors" title="Permalink to this headline">#</a></h2>
<p>We gratefully acknowledge the contributions of:</p>
<section id="authors">
<h3><span class="section-number">16.10.1. </span>Authors<a class="headerlink" href="#authors" title="Permalink to this headline">#</a></h3>
<ul class="simple">
<li><p>Matthew Gregory Jones</p></li>
<li><p>Zoe Piran</p></li>
<li><p>Aaron McKenna</p></li>
</ul>
</section>
<section id="reviewers">
<h3><span class="section-number">16.10.2. </span>Reviewers<a class="headerlink" href="#reviewers" title="Permalink to this headline">#</a></h3>
<ul class="simple">
<li><p>Lukas Heumos</p></li>
</ul>
</section>
</section>
</section>

    <script type="text/x-thebe-config">
    {
        requestKernel: true,
        binderOptions: {
            repo: "binder-examples/jupyter-stacks-datascience",
            ref: "master",
        },
        codeMirrorConfig: {
            theme: "abcdef",
            mode: "python"
        },
        kernelOptions: {
            kernelName: "python3",
            path: "./trajectories"
        },
        predefinedOutput: true
    }
    </script>
    <script>kernelName = 'python3'</script>

              </div>
              
            </main>
            <footer class="footer-article noprint">
                
    <!-- Previous / next buttons -->
<div class='prev-next-area'>
    <a class='left-prev' id="prev-link" href="fate_mapping.html" title="previous page">
        <i class="fas fa-angle-left"></i>
        <div class="prev-next-info">
            <p class="prev-next-subtitle">previous</p>
            <p class="prev-next-title"><span class="section-number">15. </span>Fate mapping</p>
        </div>
    </a>
    <a class='right-next' id="next-link" href="../conditions/differential_gene_expression.html" title="next page">
    <div class="prev-next-info">
        <p class="prev-next-subtitle">next</p>
        <p class="prev-next-title"><span class="section-number">17. </span>Differential gene expression analysis</p>
    </div>
    <i class="fas fa-angle-right"></i>
    </a>
</div>
            </footer>
        </div>
    </div>
    <div class="footer-content row">
        <footer class="col footer"><p>
  
    By Lukas Heumos, Anna Schaar<br/>
  
      &copy; Copyright 2022.<br/>
    <div class="extra_footer">
      <div>
  Brought to you by
  <a href="https://www.github.com/theislab">Theislab</a>, with many thanks to the single-cell community as a whole!
</div>

    </div>
</p>
        </footer>
    </div>
    
</div>


      </div>
    </div>
  
  <!-- Scripts loaded after <body> so the DOM is not blocked -->
  <script src="../_static/scripts/pydata-sphinx-theme.js?digest=1999514e3f237ded88cf"></script>


  </body>
</html>