
<!DOCTYPE html>


<html lang="en" data-content_root="./" >

  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="viewport" content="width=device-width, initial-scale=1" />

    <title>Single-cell best practices &#8212; Single-cell best practices</title>
  
  
  
  <script data-cfasync="false">
    document.documentElement.dataset.mode = localStorage.getItem("mode") || "";
    document.documentElement.dataset.theme = localStorage.getItem("theme") || "";
  </script>
  
  <!-- Loaded before other Sphinx assets -->
  <link href="_static/styles/theme.css?digest=dfe6caa3a7d634c4db9b" rel="stylesheet" />
<link href="_static/styles/bootstrap.css?digest=dfe6caa3a7d634c4db9b" rel="stylesheet" />
<link href="_static/styles/pydata-sphinx-theme.css?digest=dfe6caa3a7d634c4db9b" rel="stylesheet" />

  
  <link href="_static/vendor/fontawesome/6.5.2/css/all.min.css?digest=dfe6caa3a7d634c4db9b" rel="stylesheet" />
  <link rel="preload" as="font" type="font/woff2" crossorigin href="_static/vendor/fontawesome/6.5.2/webfonts/fa-solid-900.woff2" />
<link rel="preload" as="font" type="font/woff2" crossorigin href="_static/vendor/fontawesome/6.5.2/webfonts/fa-brands-400.woff2" />
<link rel="preload" as="font" type="font/woff2" crossorigin href="_static/vendor/fontawesome/6.5.2/webfonts/fa-regular-400.woff2" />

    <link rel="stylesheet" type="text/css" href="_static/pygments.css?v=03e43079" />
    <link rel="stylesheet" type="text/css" href="_static/styles/sphinx-book-theme.css?v=eba8b062" />
    <link rel="stylesheet" type="text/css" href="_static/togglebutton.css?v=13237357" />
    <link rel="stylesheet" type="text/css" href="_static/copybutton.css?v=76b2166b" />
    <link rel="stylesheet" type="text/css" href="_static/mystnb.4510f1fc1dee50b3e5859aac5469c37c29e427902b24a333a5f9fcb2f0b3ac41.css" />
    <link rel="stylesheet" type="text/css" href="_static/sphinx-thebe.css?v=4fa983c6" />
    <link rel="stylesheet" type="text/css" href="_static/sphinx-design.min.css?v=95c83b7e" />
    <link rel="stylesheet" type="text/css" href="_static/book.css?v=148c11e2" />
  
  <!-- Pre-loaded scripts that we'll load fully later -->
  <link rel="preload" as="script" href="_static/scripts/bootstrap.js?digest=dfe6caa3a7d634c4db9b" />
<link rel="preload" as="script" href="_static/scripts/pydata-sphinx-theme.js?digest=dfe6caa3a7d634c4db9b" />
  <script src="_static/vendor/fontawesome/6.5.2/js/all.min.js?digest=dfe6caa3a7d634c4db9b"></script>

    <script src="_static/documentation_options.js?v=9eb32ce0"></script>
    <script src="_static/doctools.js?v=9a2dae69"></script>
    <script src="_static/sphinx_highlight.js?v=dc90522c"></script>
    <script src="_static/clipboard.min.js?v=a7894cd8"></script>
    <script src="_static/copybutton.js?v=f281be69"></script>
    <script src="_static/scripts/sphinx-book-theme.js?v=887ef09a"></script>
    <script>let toggleHintShow = 'Click to show';</script>
    <script>let toggleHintHide = 'Click to hide';</script>
    <script>let toggleOpenOnPrint = 'true';</script>
    <script src="_static/togglebutton.js?v=4a39c7ea"></script>
    <script>var togglebuttonSelector = '.toggle, .admonition.dropdown';</script>
    <script src="_static/design-tabs.js?v=f930bc37"></script>
    <script>const THEBE_JS_URL = "https://unpkg.com/thebe@0.8.2/lib/index.js"; const thebe_selector = ".thebe,.cell"; const thebe_selector_input = "pre"; const thebe_selector_output = ".output, .cell_output"</script>
    <script async="async" src="_static/sphinx-thebe.js?v=c100c467"></script>
    <script>var togglebuttonSelector = '.toggle, .admonition.dropdown';</script>
    <script>const THEBE_JS_URL = "https://unpkg.com/thebe@0.8.2/lib/index.js"; const thebe_selector = ".thebe,.cell"; const thebe_selector_input = "pre"; const thebe_selector_output = ".output, .cell_output"</script>
    <script>window.MathJax = {"options": {"processHtmlClass": "tex2jax_process|mathjax_process|math|output_area"}}</script>
    <script defer="defer" src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-mml-chtml.js"></script>
    <script>DOCUMENTATION_OPTIONS.pagename = 'preamble';</script>
    <script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
    <script src="_static/book.js?v=dcd3d8bf"></script>
    <link rel="icon" href="_static/favicon.ico"/>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="1. Prior art" href="introduction/prior_art.html" />
  <meta name="viewport" content="width=device-width, initial-scale=1"/>
  <meta name="docsearch:language" content="en"/>
  </head>
  
  
  <body data-bs-spy="scroll" data-bs-target=".bd-toc-nav" data-offset="180" data-bs-root-margin="0px 0px -60%" data-default-mode="">

  
  
  <div id="pst-skip-link" class="skip-link d-print-none"><a href="#main-content">Skip to main content</a></div>
  
  <div id="pst-scroll-pixel-helper"></div>
  
  <button type="button" class="btn rounded-pill" id="pst-back-to-top">
    <i class="fa-solid fa-arrow-up"></i>Back to top</button>

  
  <input type="checkbox"
          class="sidebar-toggle"
          id="pst-primary-sidebar-checkbox"/>
  <label class="overlay overlay-primary" for="pst-primary-sidebar-checkbox"></label>
  
  <input type="checkbox"
          class="sidebar-toggle"
          id="pst-secondary-sidebar-checkbox"/>
  <label class="overlay overlay-secondary" for="pst-secondary-sidebar-checkbox"></label>
  
  <div class="search-button__wrapper">
    <div class="search-button__overlay"></div>
    <div class="search-button__search-container">
<form class="bd-search d-flex align-items-center"
      action="search.html"
      method="get">
  <i class="fa-solid fa-magnifying-glass"></i>
  <input type="search"
         class="form-control"
         name="q"
         id="search-input"
         placeholder="Search this book..."
         aria-label="Search this book..."
         autocomplete="off"
         autocorrect="off"
         autocapitalize="off"
         spellcheck="false"/>
  <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd>K</kbd></span>
</form></div>
  </div>

  <div class="pst-async-banner-revealer d-none">
  <aside id="bd-header-version-warning" class="d-none d-print-none" aria-label="Version warning"></aside>
</div>

  
    <header class="bd-header navbar navbar-expand-lg bd-navbar d-print-none">
    </header>
  

  <div class="bd-container">
    <div class="bd-container__inner bd-page-width">
      
      
      
        
      
      <div class="bd-sidebar-primary bd-sidebar">
        

  
  <div class="sidebar-header-items sidebar-primary__section">
    
    
    
    
  </div>
  
    <div class="sidebar-primary-items__start sidebar-primary__section">
        <div class="sidebar-primary-item">

  
    
  

<a class="navbar-brand logo" href="#">
  
  
  
  
  
    
    
      
    
    
    <img src="_static/logo.png" class="logo__image only-light" alt="Single-cell best practices - Home"/>
    <script>document.write(`<img src="_static/logo.png" class="logo__image only-dark" alt="Single-cell best practices - Home"/>`);</script>
  
  
</a></div>
        <div class="sidebar-primary-item">

 <script>
 document.write(`
   <button class="btn search-button-field search-button__button" title="Search" aria-label="Search" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="fa-solid fa-magnifying-glass"></i>
    <span class="search-button__default-text">Search</span>
    <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd class="kbd-shortcut__modifier">K</kbd></span>
   </button>
 `);
 </script></div>
        <div class="sidebar-primary-item"><nav class="bd-links bd-docs-nav" aria-label="Main">
    <div class="bd-toc-item navbar-nav active">
        <p aria-level="2" class="caption" role="heading"><span class="caption-text">Introduction</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="introduction/prior_art.html">1. Prior art</a></li>
<li class="toctree-l1"><a class="reference internal" href="introduction/scrna_seq.html">2. Single-cell RNA sequencing</a></li>
<li class="toctree-l1"><a class="reference internal" href="introduction/raw_data_processing.html">3. Raw data processing</a></li>
<li class="toctree-l1"><a class="reference internal" href="introduction/analysis_tools.html">4. Analysis frameworks and tools</a></li>
<li class="toctree-l1"><a class="reference internal" href="introduction/interoperability.html">5. Interoperability</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Preprocessing and visualization</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="preprocessing_visualization/quality_control.html">6. Quality Control</a></li>
<li class="toctree-l1"><a class="reference internal" href="preprocessing_visualization/normalization.html">7. Normalization</a></li>
<li class="toctree-l1"><a class="reference internal" href="preprocessing_visualization/feature_selection.html">8. Feature selection</a></li>
<li class="toctree-l1"><a class="reference internal" href="preprocessing_visualization/dimensionality_reduction.html">9. Dimensionality Reduction</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Identifying cellular structure</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="cellular_structure/clustering.html">10. Clustering</a></li>
<li class="toctree-l1"><a class="reference internal" href="cellular_structure/annotation.html">11. Annotation</a></li>
<li class="toctree-l1"><a class="reference internal" href="cellular_structure/integration.html">12. Data integration</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Inferring trajectories</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="trajectories/pseudotemporal.html">13. Pseudotemporal ordering</a></li>
<li class="toctree-l1"><a class="reference internal" href="trajectories/rna_velocity.html">14. RNA velocity</a></li>
<li class="toctree-l1"><a class="reference internal" href="trajectories/lineage_tracing.html">15. Lineage tracing</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Dealing with conditions</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="conditions/differential_gene_expression.html">16. Differential gene expression analysis</a></li>
<li class="toctree-l1"><a class="reference internal" href="conditions/compositional.html">17. Compositional analysis</a></li>
<li class="toctree-l1"><a class="reference internal" href="conditions/gsea_pathway.html">18. Gene set enrichment and pathway analysis</a></li>
<li class="toctree-l1"><a class="reference internal" href="conditions/perturbation_modeling.html">19. Perturbation modeling</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Modeling mechanisms</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="mechanisms/gene_regulatory_networks.html">20. Gene regulatory networks</a></li>
<li class="toctree-l1"><a class="reference internal" href="mechanisms/cell_cell_communication.html">21. Cell-cell communication</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Deconvolution</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="deconvolution/bulk_deconvolution.html">22. Bulk deconvolution</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Chromatin Accessibility</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="chromatin_accessibility/introduction.html">23. Single-cell ATAC sequencing</a></li>
<li class="toctree-l1"><a class="reference internal" href="chromatin_accessibility/quality_control.html">24. Quality Control</a></li>
<li class="toctree-l1"><a class="reference internal" href="chromatin_accessibility/gene_regulatory_networks_atac.html">25. Gene regulatory networks using chromatin accessibility</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Spatial omics</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="spatial/introduction.html">26. Single-cell data resolved in space</a></li>
<li class="toctree-l1"><a class="reference internal" href="spatial/neighborhood.html">27. Neighborhood analysis</a></li>
<li class="toctree-l1"><a class="reference internal" href="spatial/domains.html">28. Spatial domains</a></li>
<li class="toctree-l1"><a class="reference internal" href="spatial/spatially_variable_genes.html">29. Spatially variable genes</a></li>
<li class="toctree-l1"><a class="reference internal" href="spatial/deconvolution.html">30. Spatial deconvolution</a></li>
<li class="toctree-l1"><a class="reference internal" href="spatial/imputation.html">31. Imputation</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Surface protein</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="surface_protein/quality_control.html">32. Quality control</a></li>
<li class="toctree-l1"><a class="reference internal" href="surface_protein/normalization.html">33. Normalization</a></li>
<li class="toctree-l1"><a class="reference internal" href="surface_protein/doublet_detection.html">34. Doublet detection</a></li>
<li class="toctree-l1"><a class="reference internal" href="surface_protein/dimensionality_reduction.html">35. Dimensionality Reduction</a></li>
<li class="toctree-l1"><a class="reference internal" href="surface_protein/batch_correction.html">36. Batch correction</a></li>
<li class="toctree-l1"><a class="reference internal" href="surface_protein/annotation.html">37. Annotation</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Adaptive immune receptor repertoire</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="air_repertoire/ir_profiling.html">38. Immune Receptor Profiling</a></li>
<li class="toctree-l1"><a class="reference internal" href="air_repertoire/clonotype.html">39. Clonotype analysis</a></li>



<li class="toctree-l1"><a class="reference internal" href="air_repertoire/specificity.html">43. Specificity analysis</a></li>
<li class="toctree-l1"><a class="reference internal" href="air_repertoire/multimodal_integration.html">44. Integrating AIR and transcriptomics</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Multimodal integration</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="multimodal_integration/paired_integration.html">45. Paired integration</a></li>
<li class="toctree-l1"><a class="reference internal" href="multimodal_integration/advanced_integration.html">46. Advanced integration</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Reproducibility</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="reproducibility/introduction.html">47. Reproducibility</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Outlook</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="outlook.html">48. Outlook</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Acknowledgements</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="acknowledgements.html">49. Acknowledgements</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Glossary</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="glossary.html">50. Glossary</a></li>
</ul>

    </div>
</nav></div>
    </div>
  
  
  <div class="sidebar-primary-items__end sidebar-primary__section">
  </div>
  
  <div id="rtd-footer-container"></div>


      </div>
      
      <main id="main-content" class="bd-main" role="main">
        
        

<div class="sbt-scroll-pixel-helper"></div>

          <div class="bd-content">
            <div class="bd-article-container">
              
              <div class="bd-header-article d-print-none">
<div class="header-article-items header-article__inner">
  
    <div class="header-article-items__start">
      
        <div class="header-article-item"><button class="sidebar-toggle primary-toggle btn btn-sm" title="Toggle primary sidebar" data-bs-placement="bottom" data-bs-toggle="tooltip">
  <span class="fa-solid fa-bars"></span>
</button></div>
      
    </div>
  
  
    <div class="header-article-items__end">
      
        <div class="header-article-item">

<div class="article-header-buttons">





<div class="dropdown dropdown-source-buttons">
  <button class="btn dropdown-toggle" type="button" data-bs-toggle="dropdown" aria-expanded="false" aria-label="Source repositories">
    <i class="fab fa-github"></i>
  </button>
  <ul class="dropdown-menu">
      
      
      
      <li><a href="https://github.com/theislab/single-cell-best-practices" target="_blank"
   class="btn btn-sm btn-source-repository-button dropdown-item"
   title="Source repository"
   data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fab fa-github"></i>
  </span>
<span class="btn__text-container">Repository</span>
</a>
</li>
      
      
      
      
      <li><a href="https://github.com/theislab/single-cell-best-practices/edit/master/jupyter-book/preamble.md" target="_blank"
   class="btn btn-sm btn-source-edit-button dropdown-item"
   title="Suggest edit"
   data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-pencil-alt"></i>
  </span>
<span class="btn__text-container">Suggest edit</span>
</a>
</li>
      
      
      
      
      <li><a href="https://github.com/theislab/single-cell-best-practices/issues/new?title=Issue%20on%20page%20%2Fpreamble.html&body=Your%20issue%20content%20here." target="_blank"
   class="btn btn-sm btn-source-issues-button dropdown-item"
   title="Open an issue"
   data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-lightbulb"></i>
  </span>
<span class="btn__text-container">Open issue</span>
</a>
</li>
      
  </ul>
</div>






<div class="dropdown dropdown-download-buttons">
  <button class="btn dropdown-toggle" type="button" data-bs-toggle="dropdown" aria-expanded="false" aria-label="Download this page">
    <i class="fas fa-download"></i>
  </button>
  <ul class="dropdown-menu">
      
      
      
      <li><a href="_sources/preamble.md" target="_blank"
   class="btn btn-sm btn-download-source-button dropdown-item"
   title="Download source file"
   data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-file"></i>
  </span>
<span class="btn__text-container">.md</span>
</a>
</li>
      
      
      
      
      <li>
<button onclick="window.print()"
  class="btn btn-sm btn-download-pdf-button dropdown-item"
  title="Print to PDF"
  data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-file-pdf"></i>
  </span>
<span class="btn__text-container">.pdf</span>
</button>
</li>
      
  </ul>
</div>




<button onclick="toggleFullScreen()"
  class="btn btn-sm btn-fullscreen-button"
  title="Fullscreen mode"
  data-bs-placement="bottom" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-expand"></i>
  </span>

</button>



<script>
document.write(`
  <button class="btn btn-sm nav-link pst-navbar-icon theme-switch-button" title="light/dark" aria-label="light/dark" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="theme-switch fa-solid fa-sun fa-lg" data-mode="light"></i>
    <i class="theme-switch fa-solid fa-moon fa-lg" data-mode="dark"></i>
    <i class="theme-switch fa-solid fa-circle-half-stroke fa-lg" data-mode="auto"></i>
  </button>
`);
</script>


<script>
document.write(`
  <button class="btn btn-sm pst-navbar-icon search-button search-button__button" title="Search" aria-label="Search" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="fa-solid fa-magnifying-glass fa-lg"></i>
  </button>
`);
</script>
<button class="sidebar-toggle secondary-toggle btn btn-sm" title="Toggle secondary sidebar" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <span class="fa-solid fa-list"></span>
</button>
</div></div>
      
    </div>
  
</div>
</div>
              
              

<div id="jb-print-docs-body" class="onlyprint">
    <h1>Single-cell best practices</h1>
    <!-- Table of contents -->
    <div id="print-main-content">
        <div id="jb-print-toc">
            
            <div>
                <h2> Contents </h2>
            </div>
            <nav aria-label="Page">
                <ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#introduction">Introduction</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#what-this-book-covers">What this book covers</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#what-this-book-does-not-cover">What this book does not cover</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#who-should-read-this-book">Who should read this book</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#structure-of-the-book">Structure of the book</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#prerequisites">Prerequisites</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#peer-review">Peer-review</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#citation">Citation</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#contributing">Contributing</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#alternative-formats">Alternative formats</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#contact-us">Contact us</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#license">License</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#references">References</a></li>
</ul>
            </nav>
        </div>
    </div>
</div>

              
                
<div id="searchbox"></div>
                <article class="bd-article">
                  
  <p><img alt="alt text" src="_images/title.png" /></p>
<div style="page-break-after: always;"></div>
<section class="tex2jax_ignore mathjax_ignore" id="single-cell-best-practices">
<h1>Single-cell best practices<a class="headerlink" href="#single-cell-best-practices" title="Link to this heading">#</a></h1>
<div class="admonition warning">
<p class="admonition-title">Warning</p>
<p>This project is still work in progress and the content is not final yet.
If you have early feedback feel free to open issues and to get in touch with us.
Interested in contributing? Please have a look at <a class="github reference external" href="https://github.com/theislab/single-cell-best-practices/blob/development/CONTRIBUTING.md">theislab/single-cell-best-practices</a></p>
</div>
<section id="introduction">
<h2>Introduction<a class="headerlink" href="#introduction" title="Link to this heading">#</a></h2>
<p>The human body is a complex machine that heavily relies on the basic units of life - cells.
These cells exhibit remarkable diversity, varying in type and function, and can undergo significant transitions during development, in response to disease, or during regenerative processes.
This cellular heterogeneity manifests in their structure, function, and gene expression profiles.
Disruptions to this delicate balance can lead to systemic dysregulation, contributing to severe conditions such as cancer <span id="id1">[<a class="reference internal" href="#id4" title="Iain C. Macaulay, Chris P. Ponting, and Thierry Voet. Single-cell multiomics: multiple measurements from single cells. Trends in genetics : TIG, 33(2):155-168, Feb 2017. S0168-9525(16)30169-X[PII]. URL: https://doi.org/10.1016/j.tig.2016.12.003, doi:10.1016/j.tig.2016.12.003.">Macaulay <em>et al.</em>, 2017</a>]</span>.
Therefore, understanding how cells behave under normal and perturbed states is critical for deepening our understanding of entire cellular systems.</p>
<p>To address this challenge, researchers employ a variety of strategies, with one of the most promising being profiling cells at the individual level.
Traditionally, each cells’ transcriptome was primarily examined in a process known as single-cell RNA sequencing. However, recent advancements in single-cell genomics now enable the integration of transcriptome data with spatial, chromatin accessibility, or protein-level information.
These developments not only enhance our understanding of complex regulatory mechanisms but also introduce additional challenges in data analysis.</p>
<p>Currently, analysts are faced with an overwhelming array of computational tools - over 1,700 methods dedicated to single-cell RNA-seq alone <span id="id2">[<a class="reference internal" href="#id5" title="Luke Zappia and Fabian J. Theis. Over 1000 tools reveal trends in the single-cell rna-seq analysis landscape. Genome Biology, 22(1):301, Oct 2021. URL: https://doi.org/10.1186/s13059-021-02519-4, doi:10.1186/s13059-021-02519-4.">Zappia and Theis, 2021</a>]</span>.
Navigating this extensive landscape to produce reliable, cutting-edge results poses a significant challenge.</p>
</section>
<section id="what-this-book-covers">
<h2>What this book covers<a class="headerlink" href="#what-this-book-covers" title="Link to this heading">#</a></h2>
<p>This book aims to guide both beginners and experienced professionals in <strong>best practices</strong> of single-cell sequencing analysis.
It provides a comprehensive overview of essential analytical steps, from preprocessing to visualization and statistical evaluation and beyond.
By following this book, you will acquire the skills to independently analyze unimodal and multimodal single-cell sequencing data.</p>
<p>The recommendations presented are grounded in external benchmarks and reviews wherever possible, ensuring the approaches taught are both effective and reliable.
Additionally, this book is meant to be a living resource, constantly updating to keep up with new discoveries and the latest best practices.</p>
</section>
<section id="what-this-book-does-not-cover">
<h2>What this book does not cover<a class="headerlink" href="#what-this-book-does-not-cover" title="Link to this heading">#</a></h2>
<p>This book does not cover fundamental concepts of biology or computer science, including basic programming skills.
Nor does it serve as an exhaustive catalog of all available tools for specific tasks.
Instead, it emphasizes well-validated methods that have been externally benchmarked or recognized as community standards.
When such external validation is unavailable, our recommendations are only based on our extensive practical experience.</p>
</section>
<section id="who-should-read-this-book">
<h2>Who should read this book<a class="headerlink" href="#who-should-read-this-book" title="Link to this heading">#</a></h2>
<p>This book is designed for a diverse audience with an interest in single-cell data analysis, including <strong>biologists</strong> looking to gain practical data analysis skills, <strong>computer scientists</strong> seeking to explore the biological underpinnings and computational challenges of single-cell data, and <strong>bioinformaticians</strong> aiming to refine their knowledge or stay updated with best practices.
Whether you’re a beginner looking to grasp the basics or an experienced analyst aiming to refine your skills and adopt efficient workflows, this book will guide you through the complexities of single-cell analysis.</p>
</section>
<section id="structure-of-the-book">
<h2>Structure of the book<a class="headerlink" href="#structure-of-the-book" title="Link to this heading">#</a></h2>
<p>Each chapter corresponds to a distinct phase of a typical single-cell data analysis project.
While an analysis workflow should generally follow the order of the chapters, flexibility is encouraged depending on specific downstream analysis objectives.
Each chapter is supplemented with extensive references, and readers are encouraged to consult these primary sources for a deeper understanding.
Although we strive to provide comprehensive background information, our summaries and the <a class="reference internal" href="glossary.html"><span class="std std-doc">glossary</span></a> may not capture the full scope of reasoning behind every recommendation.</p>
</section>
<section id="prerequisites">
<h2>Prerequisites<a class="headerlink" href="#prerequisites" title="Link to this heading">#</a></h2>
<p>Bioinformatics is inherently multidisciplinary, requiring knowledge of both biology and computer science.
Single-cell analysis is particularly demanding, as it integrates multiple subfields and often involves large datasets.
While this book cannot cover all the necessary foundational knowledge for computational single-cell analysis, we recommend the following resources to enhance your learning experience:</p>
<ul class="simple">
<li><p><strong>Basic Python programming</strong>: Familiarity with control flow (e.g., loops, conditional statements), basic data structures (e.g., lists, dictionaries, sets), and key libraries such as Pandas and Numpy is essential.
Newcomers can benefit from the free book <a class="reference external" href="https://automatetheboringstuff.com/">Automate the boring stuff with Python</a>.</p></li>
<li><p><strong>AnnData and Scanpy</strong>: While prior experience with these tools is helpful, it is not strictly required.
This book provides a detailed introduction to AnnData and outlines the workflow for working with Scanpy.
However, it does not cover the full range of Scanpy’s functionalities. To deepen your understanding, we recommend exploring the <a class="reference external" href="https://scanpy.readthedocs.io/en/stable/tutorials.html">scanpy tutorials</a> and referring to the <a class="reference external" href="https://scanpy.readthedocs.io/en/stable/api.html">scanpy API reference</a> as needed.</p></li>
<li><p><strong>Multimodal Data Analysis</strong>: For readers who have interest in multimodal data analysis, understanding tools like <a class="reference internal" href="glossary.html#term-muon"><span class="xref std std-term">muon</span></a> and MuData is beneficial.
The <a class="reference external" href="https://muon-tutorials.readthedocs.io/en/latest/">muon tutorials</a> provide a solid introduction to this field.</p></li>
<li><p><strong>Basic R Programming</strong>: Knowledge of control flow and basic data structures suffices.
New learners can refer to <a class="reference external" href="https://r4ds.had.co.nz/">R for data science</a> for a comprehensive introduction.</p></li>
<li><p><strong>Basic Biology</strong>: While this book offers a rough overview of data generation, it does not cover fundamental topics like <a class="reference internal" href="glossary.html#term-DNA"><span class="xref std std-term">DNA</span></a>, RNA, and proteins.
<em>Molecular Biology of the Cell</em> by Bruce Alberts et al. is a recommended resource for those new to molecular biology.</p></li>
</ul>
</section>
<section id="peer-review">
<h2>Peer-review<a class="headerlink" href="#peer-review" title="Link to this heading">#</a></h2>
<p>Although the content has been reviewed by multiple authors, editors, and external experts, this book has not undergone formal peer review.
We kindly encourage you to provide constructive feedback to help refine and improve the material.
To share your thoughts:</p>
<ol class="arabic simple">
<li><p><strong>Open an Issue</strong>: You can open an issue on <a class="reference external" href="https://github.com/theislab/single-cell-best-practices">our GitHub repository</a> for any suggestions, questions, or clarifications you have.</p></li>
<li><p><strong>Be Specific</strong>: When providing feedback, please be as specific as possible.
For example, if you have a suggestion for improving a section, point out the exact part and explain why you think it can be enhanced.</p></li>
<li><p><strong>Provide Sources</strong>: If applicable, we encourage you to provide sources or references to support your feedback or suggestions.
This helps ensure the material stays accurate and up to date.</p></li>
<li><p><strong>Discuss with the Community</strong>: Feel free to comment on existing issues to join ongoing discussions.
This can help us refine the content based on community input.</p></li>
</ol>
</section>
<section id="citation">
<h2>Citation<a class="headerlink" href="#citation" title="Link to this heading">#</a></h2>
<p>If you found our content helpful for your research, please cite it as:</p>
<blockquote>
<div><p>Heumos, L., Schaar, A.C., Lance, C. et al. Best practices for single-cell analysis across modalities. Nat Rev Genet (2023). <a class="reference external" href="https://doi.org/10.1038/s41576-023-00586-w">https://doi.org/10.1038/s41576-023-00586-w</a></p>
</div></blockquote>
</section>
<section id="contributing">
<h2>Contributing<a class="headerlink" href="#contributing" title="Link to this heading">#</a></h2>
<p>We invite the community to contribute to the ongoing improvement of this tutorial and teaching material.
Please read <a class="reference external" href="https://github.com/theislab/single-cell-best-practices/blob/development/CONTRIBUTING.md">contributing</a> for further instructions.</p>
<p>In case of questions or problems, please get in touch by posting an issue in this repository.</p>
</section>
<section id="alternative-formats">
<h2>Alternative formats<a class="headerlink" href="#alternative-formats" title="Link to this heading">#</a></h2>
<p>PDF versions of this book are available on our <a class="reference external" href="https://github.com/theislab/single-cell-best-practices/releases">releases page</a>.</p>
</section>
<section id="contact-us">
<h2>Contact us<a class="headerlink" href="#contact-us" title="Link to this heading">#</a></h2>
<p>You can report issues and requests in our <a class="reference external" href="https://github.com/theislab/single-cell-best-practices/issues">issue tracker</a>.</p>
<p>For inquiries, speaking engagements, or collaboration opportunities, please email:</p>
<ul class="simple">
<li><p>Anna Schaar: <a class="reference external" href="mailto:anna&#46;schaar&#37;&#52;&#48;helmholtz-munich&#46;de">anna<span>&#46;</span>schaar<span>&#64;</span>helmholtz-munich<span>&#46;</span>de</a></p></li>
<li><p>Lukas Heumos: <a class="reference external" href="mailto:lukas&#46;heumos&#37;&#52;&#48;helmholtz-munich&#46;de">lukas<span>&#46;</span>heumos<span>&#64;</span>helmholtz-munich<span>&#46;</span>de</a></p></li>
</ul>
</section>
<section id="license">
<h2>License<a class="headerlink" href="#license" title="Link to this heading">#</a></h2>
<p>This book is licensed under the <a class="reference external" href="https://github.com/theislab/single-cell-best-practices/blob/development/LICENSE">Apache 2.0 license</a>.</p>
</section>
<section id="references">
<h2>References<a class="headerlink" href="#references" title="Link to this heading">#</a></h2>
<div class="docutils container" id="id3">
<div role="list" class="citation-list">
<div class="citation" id="id4" role="doc-biblioentry">
<span class="label"><span class="fn-bracket">[</span><a role="doc-backlink" href="#id1">MPV17</a><span class="fn-bracket">]</span></span>
<p>Iain C. Macaulay, Chris P. Ponting, and Thierry Voet. Single-cell multiomics: multiple measurements from single cells. <em>Trends in genetics : TIG</em>, 33(2):155–168, Feb 2017. S0168-9525(16)30169-X[PII]. URL: <a class="reference external" href="https://doi.org/10.1016/j.tig.2016.12.003">https://doi.org/10.1016/j.tig.2016.12.003</a>, <a class="reference external" href="https://doi.org/10.1016/j.tig.2016.12.003">doi:10.1016/j.tig.2016.12.003</a>.</p>
</div>
<div class="citation" id="id5" role="doc-biblioentry">
<span class="label"><span class="fn-bracket">[</span><a role="doc-backlink" href="#id2">ZT21</a><span class="fn-bracket">]</span></span>
<p>Luke Zappia and Fabian J. Theis. Over 1000 tools reveal trends in the single-cell rna-seq analysis landscape. <em>Genome Biology</em>, 22(1):301, Oct 2021. URL: <a class="reference external" href="https://doi.org/10.1186/s13059-021-02519-4">https://doi.org/10.1186/s13059-021-02519-4</a>, <a class="reference external" href="https://doi.org/10.1186/s13059-021-02519-4">doi:10.1186/s13059-021-02519-4</a>.</p>
</div>
</div>
</div>
</section>
<div class="toctree-wrapper compound">
</div>
<div class="toctree-wrapper compound">
</div>
<div class="toctree-wrapper compound">
</div>
<div class="toctree-wrapper compound">
</div>
<div class="toctree-wrapper compound">
</div>
<div class="toctree-wrapper compound">
</div>
<div class="toctree-wrapper compound">
</div>
<div class="toctree-wrapper compound">
</div>
<div class="toctree-wrapper compound">
</div>
<div class="toctree-wrapper compound">
</div>
<div class="toctree-wrapper compound">
</div>
<div class="toctree-wrapper compound">
</div>
<div class="toctree-wrapper compound">
</div>
<div class="toctree-wrapper compound">
</div>
<div class="toctree-wrapper compound">
</div>
<div class="toctree-wrapper compound">
</div>
</section>

    <script type="text/x-thebe-config">
    {
        requestKernel: true,
        binderOptions: {
            repo: "binder-examples/jupyter-stacks-datascience",
            ref: "master",
        },
        codeMirrorConfig: {
            theme: "abcdef",
            mode: "python"
        },
        kernelOptions: {
            name: "python3",
            path: "./."
        },
        predefinedOutput: true
    }
    </script>
    <script>kernelName = 'python3'</script>

                </article>
              

              
              
              
              
                <footer class="prev-next-footer d-print-none">
                  
<div class="prev-next-area">
    <a class="right-next"
       href="introduction/prior_art.html"
       title="next page">
      <div class="prev-next-info">
        <p class="prev-next-subtitle">next</p>
        <p class="prev-next-title"><span class="section-number">1. </span>Prior art</p>
      </div>
      <i class="fa-solid fa-angle-right"></i>
    </a>
</div>
                </footer>
              
            </div>
            
            
              
                <div class="bd-sidebar-secondary bd-toc"><div class="sidebar-secondary-items sidebar-secondary__inner">


  <div class="sidebar-secondary-item">
  <div class="page-toc tocsection onthispage">
    <i class="fa-solid fa-list"></i> Contents
  </div>
  <nav class="bd-toc-nav page-toc">
    <ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#introduction">Introduction</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#what-this-book-covers">What this book covers</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#what-this-book-does-not-cover">What this book does not cover</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#who-should-read-this-book">Who should read this book</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#structure-of-the-book">Structure of the book</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#prerequisites">Prerequisites</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#peer-review">Peer-review</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#citation">Citation</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#contributing">Contributing</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#alternative-formats">Alternative formats</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#contact-us">Contact us</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#license">License</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#references">References</a></li>
</ul>
  </nav></div>

</div></div>
              
            
          </div>
          <footer class="bd-footer-content">
            
<div class="bd-footer-content__inner container">
  
  <div class="footer-item">
    
<p class="component-author">
By Lukas Heumos, Anna Schaar, single-cell best practices consortium
</p>

  </div>
  
  <div class="footer-item">
    

  <p class="copyright">
    
      © Copyright 2023.
      <br/>
    
  </p>

  </div>
  
  <div class="footer-item">
    
  </div>
  
  <div class="footer-item">
    
<div class="extra_footer">
  <div>
  Brought to you by
  <a href="https://www.github.com/theislab">Theislab</a>, with many thanks to the single-cell community as a whole!
</div>

</div>
  </div>
  
</div>
          </footer>
        

      </main>
    </div>
  </div>
  
  <!-- Scripts loaded after <body> so the DOM is not blocked -->
  <script src="_static/scripts/bootstrap.js?digest=dfe6caa3a7d634c4db9b"></script>
<script src="_static/scripts/pydata-sphinx-theme.js?digest=dfe6caa3a7d634c4db9b"></script>

  <footer class="bd-footer">
  </footer>
  </body>
</html>