
<!DOCTYPE html>


<html lang="en" data-content_root="../" >

  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="viewport" content="width=device-width, initial-scale=1" />

    <title>5. Interoperability &#8212; Single-cell best practices</title>
  
  
  
  <script data-cfasync="false">
    document.documentElement.dataset.mode = localStorage.getItem("mode") || "";
    document.documentElement.dataset.theme = localStorage.getItem("theme") || "";
  </script>
  
  <!-- Loaded before other Sphinx assets -->
  <link href="../_static/styles/theme.css?digest=dfe6caa3a7d634c4db9b" rel="stylesheet" />
<link href="../_static/styles/bootstrap.css?digest=dfe6caa3a7d634c4db9b" rel="stylesheet" />
<link href="../_static/styles/pydata-sphinx-theme.css?digest=dfe6caa3a7d634c4db9b" rel="stylesheet" />

  
  <link href="../_static/vendor/fontawesome/6.5.2/css/all.min.css?digest=dfe6caa3a7d634c4db9b" rel="stylesheet" />
  <link rel="preload" as="font" type="font/woff2" crossorigin href="../_static/vendor/fontawesome/6.5.2/webfonts/fa-solid-900.woff2" />
<link rel="preload" as="font" type="font/woff2" crossorigin href="../_static/vendor/fontawesome/6.5.2/webfonts/fa-brands-400.woff2" />
<link rel="preload" as="font" type="font/woff2" crossorigin href="../_static/vendor/fontawesome/6.5.2/webfonts/fa-regular-400.woff2" />

    <link rel="stylesheet" type="text/css" href="../_static/pygments.css?v=fa44fd50" />
    <link rel="stylesheet" type="text/css" href="../_static/styles/sphinx-book-theme.css?v=a3416100" />
    <link rel="stylesheet" type="text/css" href="../_static/togglebutton.css?v=13237357" />
    <link rel="stylesheet" type="text/css" href="../_static/copybutton.css?v=76b2166b" />
    <link rel="stylesheet" type="text/css" href="../_static/mystnb.4510f1fc1dee50b3e5859aac5469c37c29e427902b24a333a5f9fcb2f0b3ac41.css" />
    <link rel="stylesheet" type="text/css" href="../_static/sphinx-thebe.css?v=4fa983c6" />
    <link rel="stylesheet" type="text/css" href="../_static/sphinx-design.min.css?v=87e54e7c" />
    <link rel="stylesheet" type="text/css" href="../_static/book.css?v=148c11e2" />
  
  <!-- Pre-loaded scripts that we'll load fully later -->
  <link rel="preload" as="script" href="../_static/scripts/bootstrap.js?digest=dfe6caa3a7d634c4db9b" />
<link rel="preload" as="script" href="../_static/scripts/pydata-sphinx-theme.js?digest=dfe6caa3a7d634c4db9b" />
  <script src="../_static/vendor/fontawesome/6.5.2/js/all.min.js?digest=dfe6caa3a7d634c4db9b"></script>

    <script src="../_static/documentation_options.js?v=9eb32ce0"></script>
    <script src="../_static/doctools.js?v=9a2dae69"></script>
    <script src="../_static/sphinx_highlight.js?v=dc90522c"></script>
    <script src="../_static/clipboard.min.js?v=a7894cd8"></script>
    <script src="../_static/copybutton.js?v=f281be69"></script>
    <script src="../_static/scripts/sphinx-book-theme.js?v=887ef09a"></script>
    <script>let toggleHintShow = 'Click to show';</script>
    <script>let toggleHintHide = 'Click to hide';</script>
    <script>let toggleOpenOnPrint = 'true';</script>
    <script src="../_static/togglebutton.js?v=4a39c7ea"></script>
    <script>var togglebuttonSelector = '.toggle, .admonition.dropdown';</script>
    <script src="../_static/design-tabs.js?v=f930bc37"></script>
    <script>const THEBE_JS_URL = "https://unpkg.com/thebe@0.8.2/lib/index.js"; const thebe_selector = ".thebe,.cell"; const thebe_selector_input = "pre"; const thebe_selector_output = ".output, .cell_output"</script>
    <script async="async" src="../_static/sphinx-thebe.js?v=c100c467"></script>
    <script>var togglebuttonSelector = '.toggle, .admonition.dropdown';</script>
    <script>const THEBE_JS_URL = "https://unpkg.com/thebe@0.8.2/lib/index.js"; const thebe_selector = ".thebe,.cell"; const thebe_selector_input = "pre"; const thebe_selector_output = ".output, .cell_output"</script>
    <script>DOCUMENTATION_OPTIONS.pagename = 'introduction/interoperability';</script>
    <script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
    <script src="../_static/book.js?v=dcd3d8bf"></script>
    <link rel="icon" href="../_static/favicon.ico"/>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="6. Quality Control" href="../preprocessing_visualization/quality_control.html" />
    <link rel="prev" title="4. Analysis frameworks and tools" href="analysis_tools.html" />
  <meta name="viewport" content="width=device-width, initial-scale=1"/>
  <meta name="docsearch:language" content="en"/>
  </head>
  
  
  <body data-bs-spy="scroll" data-bs-target=".bd-toc-nav" data-offset="180" data-bs-root-margin="0px 0px -60%" data-default-mode="">

  
  
  <div id="pst-skip-link" class="skip-link d-print-none"><a href="#main-content">Skip to main content</a></div>
  
  <div id="pst-scroll-pixel-helper"></div>
  
  <button type="button" class="btn rounded-pill" id="pst-back-to-top">
    <i class="fa-solid fa-arrow-up"></i>Back to top</button>

  
  <input type="checkbox"
          class="sidebar-toggle"
          id="pst-primary-sidebar-checkbox"/>
  <label class="overlay overlay-primary" for="pst-primary-sidebar-checkbox"></label>
  
  <input type="checkbox"
          class="sidebar-toggle"
          id="pst-secondary-sidebar-checkbox"/>
  <label class="overlay overlay-secondary" for="pst-secondary-sidebar-checkbox"></label>
  
  <div class="search-button__wrapper">
    <div class="search-button__overlay"></div>
    <div class="search-button__search-container">
<form class="bd-search d-flex align-items-center"
      action="../search.html"
      method="get">
  <i class="fa-solid fa-magnifying-glass"></i>
  <input type="search"
         class="form-control"
         name="q"
         id="search-input"
         placeholder="Search this book..."
         aria-label="Search this book..."
         autocomplete="off"
         autocorrect="off"
         autocapitalize="off"
         spellcheck="false"/>
  <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd>K</kbd></span>
</form></div>
  </div>

  <div class="pst-async-banner-revealer d-none">
  <aside id="bd-header-version-warning" class="d-none d-print-none" aria-label="Version warning"></aside>
</div>

  
    <header class="bd-header navbar navbar-expand-lg bd-navbar d-print-none">
    </header>
  

  <div class="bd-container">
    <div class="bd-container__inner bd-page-width">
      
      
      
      <div class="bd-sidebar-primary bd-sidebar">
        

  
  <div class="sidebar-header-items sidebar-primary__section">
    
    
    
    
  </div>
  
    <div class="sidebar-primary-items__start sidebar-primary__section">
        <div class="sidebar-primary-item">

  
    
  

<a class="navbar-brand logo" href="../preamble.html">
  
  
  
  
  
    
    
      
    
    
    <img src="../_static/logo.png" class="logo__image only-light" alt="Single-cell best practices - Home"/>
    <script>document.write(`<img src="../_static/logo.png" class="logo__image only-dark" alt="Single-cell best practices - Home"/>`);</script>
  
  
</a></div>
        <div class="sidebar-primary-item">

 <script>
 document.write(`
   <button class="btn search-button-field search-button__button" title="Search" aria-label="Search" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="fa-solid fa-magnifying-glass"></i>
    <span class="search-button__default-text">Search</span>
    <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd class="kbd-shortcut__modifier">K</kbd></span>
   </button>
 `);
 </script></div>
        <div class="sidebar-primary-item"><nav class="bd-links bd-docs-nav" aria-label="Main">
    <div class="bd-toc-item navbar-nav active">
        <p aria-level="2" class="caption" role="heading"><span class="caption-text">Introduction</span></p>
<ul class="current nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="prior_art.html">1. Prior art</a></li>
<li class="toctree-l1"><a class="reference internal" href="scrna_seq.html">2. Single-cell RNA sequencing</a></li>
<li class="toctree-l1"><a class="reference internal" href="raw_data_processing.html">3. Raw data processing</a></li>
<li class="toctree-l1"><a class="reference internal" href="analysis_tools.html">4. Analysis frameworks and tools</a></li>
<li class="toctree-l1 current active"><a class="current reference internal" href="#">5. Interoperability</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Preprocessing and visualization</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../preprocessing_visualization/quality_control.html">6. Quality Control</a></li>
<li class="toctree-l1"><a class="reference internal" href="../preprocessing_visualization/normalization.html">7. Normalization</a></li>
<li class="toctree-l1"><a class="reference internal" href="../preprocessing_visualization/feature_selection.html">8. Feature selection</a></li>
<li class="toctree-l1"><a class="reference internal" href="../preprocessing_visualization/dimensionality_reduction.html">9. Dimensionality Reduction</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Identifying cellular structure</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../cellular_structure/clustering.html">10. Clustering</a></li>
<li class="toctree-l1"><a class="reference internal" href="../cellular_structure/annotation.html">11. Annotation</a></li>
<li class="toctree-l1"><a class="reference internal" href="../cellular_structure/integration.html">12. Data integration</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Inferring trajectories</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../trajectories/pseudotemporal.html">13. Pseudotemporal ordering</a></li>
<li class="toctree-l1"><a class="reference internal" href="../trajectories/rna_velocity.html">14. RNA velocity</a></li>
<li class="toctree-l1"><a class="reference internal" href="../trajectories/lineage_tracing.html">15. Lineage tracing</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Dealing with conditions</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../conditions/differential_gene_expression.html">16. Differential gene expression analysis</a></li>
<li class="toctree-l1"><a class="reference internal" href="../conditions/compositional.html">17. Compositional analysis</a></li>
<li class="toctree-l1"><a class="reference internal" href="../conditions/gsea_pathway.html">18. Gene set enrichment and pathway analysis</a></li>
<li class="toctree-l1"><a class="reference internal" href="../conditions/perturbation_modeling.html">19. Perturbation modeling</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Modeling mechanisms</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../mechanisms/gene_regulatory_networks.html">20. Gene regulatory networks</a></li>
<li class="toctree-l1"><a class="reference internal" href="../mechanisms/cell_cell_communication.html">21. Cell-cell communication</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Deconvolution</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../deconvolution/bulk_deconvolution.html">22. Bulk deconvolution</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Chromatin Accessibility</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../chromatin_accessibility/introduction.html">23. Single-cell ATAC sequencing</a></li>
<li class="toctree-l1"><a class="reference internal" href="../chromatin_accessibility/quality_control.html">24. Quality Control</a></li>
<li class="toctree-l1"><a class="reference internal" href="../chromatin_accessibility/gene_regulatory_networks_atac.html">25. Gene regulatory networks using chromatin accessibility</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Spatial omics</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../spatial/introduction.html">26. Single-cell data resolved in space</a></li>
<li class="toctree-l1"><a class="reference internal" href="../spatial/neighborhood.html">27. Neighborhood analysis</a></li>
<li class="toctree-l1"><a class="reference internal" href="../spatial/domains.html">28. Spatial domains</a></li>
<li class="toctree-l1"><a class="reference internal" href="../spatial/spatially_variable_genes.html">29. Spatially variable genes</a></li>
<li class="toctree-l1"><a class="reference internal" href="../spatial/deconvolution.html">30. Spatial deconvolution</a></li>
<li class="toctree-l1"><a class="reference internal" href="../spatial/imputation.html">31. Imputation</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Surface protein</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../surface_protein/quality_control.html">32. Quality control</a></li>
<li class="toctree-l1"><a class="reference internal" href="../surface_protein/normalization.html">33. Normalization</a></li>
<li class="toctree-l1"><a class="reference internal" href="../surface_protein/doublet_detection.html">34. Doublet detection</a></li>
<li class="toctree-l1"><a class="reference internal" href="../surface_protein/dimensionality_reduction.html">35. Dimensionality Reduction</a></li>
<li class="toctree-l1"><a class="reference internal" href="../surface_protein/batch_correction.html">36. Batch correction</a></li>
<li class="toctree-l1"><a class="reference internal" href="../surface_protein/annotation.html">37. Annotation</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Adaptive immune receptor repertoire</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../air_repertoire/ir_profiling.html">38. Immune Receptor Profiling</a></li>
<li class="toctree-l1"><a class="reference internal" href="../air_repertoire/clonotype.html">39. Clonotype analysis</a></li>



<li class="toctree-l1"><a class="reference internal" href="../air_repertoire/specificity.html">43. Specificity analysis</a></li>
<li class="toctree-l1"><a class="reference internal" href="../air_repertoire/multimodal_integration.html">44. Integrating AIR and transcriptomics</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Multimodal integration</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../multimodal_integration/paired_integration.html">45. Paired integration</a></li>
<li class="toctree-l1"><a class="reference internal" href="../multimodal_integration/advanced_integration.html">46. Advanced integration</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Reproducibility</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../reproducibility/introduction.html">47. Reproducibility</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Outlook</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../outlook.html">48. Outlook</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Acknowledgements</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../acknowledgements.html">49. Acknowledgements</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Glossary</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../glossary.html">50. Glossary</a></li>
</ul>

    </div>
</nav></div>
    </div>
  
  
  <div class="sidebar-primary-items__end sidebar-primary__section">
  </div>
  
  <div id="rtd-footer-container"></div>


      </div>
      
      <main id="main-content" class="bd-main" role="main">
        
        

<div class="sbt-scroll-pixel-helper"></div>

          <div class="bd-content">
            <div class="bd-article-container">
              
              <div class="bd-header-article d-print-none">
<div class="header-article-items header-article__inner">
  
    <div class="header-article-items__start">
      
        <div class="header-article-item"><button class="sidebar-toggle primary-toggle btn btn-sm" title="Toggle primary sidebar" data-bs-placement="bottom" data-bs-toggle="tooltip">
  <span class="fa-solid fa-bars"></span>
</button></div>
      
    </div>
  
  
    <div class="header-article-items__end">
      
        <div class="header-article-item">

<div class="article-header-buttons">





<div class="dropdown dropdown-source-buttons">
  <button class="btn dropdown-toggle" type="button" data-bs-toggle="dropdown" aria-expanded="false" aria-label="Source repositories">
    <i class="fab fa-github"></i>
  </button>
  <ul class="dropdown-menu">
      
      
      
      <li><a href="https://github.com/theislab/single-cell-best-practices" target="_blank"
   class="btn btn-sm btn-source-repository-button dropdown-item"
   title="Source repository"
   data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fab fa-github"></i>
  </span>
<span class="btn__text-container">Repository</span>
</a>
</li>
      
      
      
      
      <li><a href="https://github.com/theislab/single-cell-best-practices/edit/master/jupyter-book/introduction/interoperability.ipynb" target="_blank"
   class="btn btn-sm btn-source-edit-button dropdown-item"
   title="Suggest edit"
   data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-pencil-alt"></i>
  </span>
<span class="btn__text-container">Suggest edit</span>
</a>
</li>
      
      
      
      
      <li><a href="https://github.com/theislab/single-cell-best-practices/issues/new?title=Issue%20on%20page%20%2Fintroduction/interoperability.html&body=Your%20issue%20content%20here." target="_blank"
   class="btn btn-sm btn-source-issues-button dropdown-item"
   title="Open an issue"
   data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-lightbulb"></i>
  </span>
<span class="btn__text-container">Open issue</span>
</a>
</li>
      
  </ul>
</div>






<div class="dropdown dropdown-download-buttons">
  <button class="btn dropdown-toggle" type="button" data-bs-toggle="dropdown" aria-expanded="false" aria-label="Download this page">
    <i class="fas fa-download"></i>
  </button>
  <ul class="dropdown-menu">
      
      
      
      <li><a href="../_sources/introduction/interoperability.ipynb" target="_blank"
   class="btn btn-sm btn-download-source-button dropdown-item"
   title="Download source file"
   data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-file"></i>
  </span>
<span class="btn__text-container">.ipynb</span>
</a>
</li>
      
      
      
      
      <li>
<button onclick="window.print()"
  class="btn btn-sm btn-download-pdf-button dropdown-item"
  title="Print to PDF"
  data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-file-pdf"></i>
  </span>
<span class="btn__text-container">.pdf</span>
</button>
</li>
      
  </ul>
</div>




<button onclick="toggleFullScreen()"
  class="btn btn-sm btn-fullscreen-button"
  title="Fullscreen mode"
  data-bs-placement="bottom" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-expand"></i>
  </span>

</button>



<script>
document.write(`
  <button class="btn btn-sm nav-link pst-navbar-icon theme-switch-button" title="light/dark" aria-label="light/dark" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="theme-switch fa-solid fa-sun fa-lg" data-mode="light"></i>
    <i class="theme-switch fa-solid fa-moon fa-lg" data-mode="dark"></i>
    <i class="theme-switch fa-solid fa-circle-half-stroke fa-lg" data-mode="auto"></i>
  </button>
`);
</script>


<script>
document.write(`
  <button class="btn btn-sm pst-navbar-icon search-button search-button__button" title="Search" aria-label="Search" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="fa-solid fa-magnifying-glass fa-lg"></i>
  </button>
`);
</script>
<button class="sidebar-toggle secondary-toggle btn btn-sm" title="Toggle secondary sidebar" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <span class="fa-solid fa-list"></span>
</button>
</div></div>
      
    </div>
  
</div>
</div>
              
              

<div id="jb-print-docs-body" class="onlyprint">
    <h1>Interoperability</h1>
    <!-- Table of contents -->
    <div id="print-main-content">
        <div id="jb-print-toc">
            
            <div>
                <h2> Contents </h2>
            </div>
            <nav aria-label="Page">
                <ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#motivation">5.1. Motivation</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#nomenclature">5.2. Nomenclature</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#disk-based-interoperability">5.3. Disk-based interoperability</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#simple-formats">5.3.1. Simple formats</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#hdf5-based-formats">5.3.2. HDF5-based formats</a><ul class="nav section-nav flex-column">
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#h5ad">5.3.2.1. H5AD</a><ul class="nav section-nav flex-column">
<li class="toc-h5 nav-item toc-entry"><a class="reference internal nav-link" href="#reading-writing-h5ad-with-bioconductor">5.3.2.1.1. Reading/writing H5AD with Bioconductor</a></li>
<li class="toc-h5 nav-item toc-entry"><a class="reference internal nav-link" href="#reading-writing-h5ad-with-seurat">5.3.2.1.2. Reading/writing H5AD with <strong>{Seurat}</strong></a></li>
<li class="toc-h5 nav-item toc-entry"><a class="reference internal nav-link" href="#reading-writing-h5ad-with-anndata">5.3.2.1.3. Reading/writing H5AD with <strong>{anndata}</strong></a></li>
</ul>
</li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#loom">5.3.2.2. Loom</a></li>
</ul>
</li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#rds-files">5.3.3. RDS files</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#new-on-disk-formats">5.3.4. New on-disk formats</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#in-memory-interoperability">5.4. In-memory interoperability</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#interoperability-between-r-ecosystems">5.4.1. Interoperability between R ecosystems</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#accessing-r-from-python">5.4.2. Accessing R from Python</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#accessing-python-from-r">5.4.3. Accessing Python from R</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#interoperability-for-multimodal-data">5.5. Interoperability for multimodal data</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#python">5.5.1. Python</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#r">5.5.2. R</a><ul class="nav section-nav flex-column">
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#bioconductor">5.5.2.1. Bioconductor</a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#seurat">5.5.2.2. Seurat</a></li>
</ul>
</li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#interoperability-with-other-languages">5.6. Interoperability with other languages</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#julia">5.6.1. Julia</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#javascript">5.6.2. JavaScript</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#rust">5.6.3. Rust</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#session-information">5.7. Session information</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#id1">5.8. Python</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#id2">5.9. R</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#references">5.10. References</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#contributors">5.11. Contributors</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#authors">5.11.1. Authors</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#reviewers">5.11.2. Reviewers</a></li>
</ul>
</li>
</ul>
            </nav>
        </div>
    </div>
</div>

              
                
<div id="searchbox"></div>
                <article class="bd-article">
                  
  <section class="tex2jax_ignore mathjax_ignore" id="interoperability">
<h1><span class="section-number">5. </span>Interoperability<a class="headerlink" href="#interoperability" title="Link to this heading">#</a></h1>
<div class="admonition-summary admonition">
<p class="admonition-title">Summary</p>
<ul class="simple">
<li><p>Interoperabilty between languages allows analysts to take advantage of the strengths of different ecosystems</p></li>
<li><p>On-disk interoperability uses standard file formats to transfer data and is typically more reliable</p></li>
<li><p>In-memory interoperabilty transfers data directly between parallel sessions and is convenient for interactive analysis</p></li>
<li><p>While interoperability is currently possible developers continue to improve the experience</p></li>
</ul>
</div>
<section id="motivation">
<h2><span class="section-number">5.1. </span>Motivation<a class="headerlink" href="#motivation" title="Link to this heading">#</a></h2>
<p>As we have discussed in the <a class="reference internal" href="analysis_tools.html#introduction-analysis-frameworks"><span class="std std-ref">analysis frameworks and tools chapter</span></a> there are three main ecosystems for single-cell analysis, the <a class="reference external" href="https://bioconductor.org/">Bioconductor</a> and <a class="reference external" href="https://satijalab.org/seurat/index.html">Seurat</a> ecosystems in R and the Python-based <a class="reference external" href="https://scverse.org/">scverse</a> ecosystem. A common question from new analysts is which ecosystem they should focus on learning and using? While it makes sense to focus on one to start with, and a successful standard analysis can be performed in any ecosystem, we promote the idea that competent analysts should be familiar with all three ecosystems and comfortable moving between them. This approach allows analysts to use the best-performing tools and methods regardless of how they were implemented. When analysts are not comfortable moving between ecosystems they often tend to use packages that are easy to access, even when they have been shown to have shortcomings compared to packages in another ecosystem. The ability of analysts to move between ecosystems allows developers to take advantage of the different strengths of programming languages. For example, R has strong inbuilt support for complex statistical modelling while the majority of deep learning libraries are focused on Python. By supporting common on-disk data formats and in-memory data structures developers can be confident that analysts can access their package and can use the platform the platform that is most appropriate for their method. Another motivation for being comfortable with multiple ecosystems is the accessibility and availability of data, results and documentation. Often data or results are only made available in one format and analysts will need to be familiar with that format in order to access it. A basic understanding of other ecosystems is also necessary to understand package documentation and tutorials when deciding which methods to use.</p>
<p>While we encourage analysts to be comfortable with all the major ecosystems, moving between them is only possible when they are interoperable. Thankfully, lots of work has been done in this area and it is now relatively simple in most cases using standard packages. In this chapter, we discuss the various ways data can be moved between ecosystems via disk or in-memory, the differences between them and their advantages. We focus on single-modality data and moving between R and Python as these are the most common cases but we also touch on multimodal data and other languages.</p>
</section>
<section id="nomenclature">
<h2><span class="section-number">5.2. </span>Nomenclature<a class="headerlink" href="#nomenclature" title="Link to this heading">#</a></h2>
<p>Because talking about different languages can get confusing we try to use the following conventions:</p>
<ul class="simple">
<li><p><strong>{package}</strong> - An R package</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">package::function()</span></code> - A function in an R package</p></li>
<li><p><strong>package</strong> - A Python package</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">package.function()</span></code> - A function in a Python package</p></li>
<li><p><strong>Emphasised</strong> - Some other important concept</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">code</span></code> - Other parts of code including objects, variables etc. This is also used for files or directories.</p></li>
</ul>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">anndata</span>
<span class="kn">import</span> <span class="nn">numpy</span>
<span class="kn">import</span> <span class="nn">scanpy</span>
<span class="kn">import</span> <span class="nn">mudata</span>
<span class="kn">import</span> <span class="nn">tempfile</span>
<span class="kn">import</span> <span class="nn">os</span>
<span class="kn">import</span> <span class="nn">rpy2.robjects</span>
<span class="kn">import</span> <span class="nn">anndata2ri</span>
<span class="kn">from</span> <span class="nn">scipy.sparse</span> <span class="kn">import</span> <span class="n">csr_matrix</span>

<span class="n">anndata2ri</span><span class="o">.</span><span class="n">activate</span><span class="p">()</span>
<span class="o">%</span><span class="k">load_ext</span> rpy2.ipython
</pre></div>
</div>
</div>
</div>
</section>
<section id="disk-based-interoperability">
<h2><span class="section-number">5.3. </span>Disk-based interoperability<a class="headerlink" href="#disk-based-interoperability" title="Link to this heading">#</a></h2>
<p>The first approach to moving between languages is via disk-based interoperability. This involves writing a file to disk in one language and then reading that file into a second language. In many cases, this approach is simpler, more reliable and scalable than in-memory interoperability (which we discuss below) but it comes at the cost of greater storage requirements and reduced interactivity. Disk-based interoperability tends to work particularly well when there are established processes for each stage of analysis and you want to pass objects from one to the next (especially as part of a pipeline developed using a workflow manager such as <a class="reference external" href="https://www.nextflow.io/index.html">NextFlow</a> or <a class="reference external" href="https://snakemake.readthedocs.io/en/stable/">snakemake</a>). However, disk-based interoperability is less convenient for interactive steps such as data exploration or experimenting with methods as you need to write a new file whenever you want to move between languages.</p>
<section id="simple-formats">
<h3><span class="section-number">5.3.1. </span>Simple formats<a class="headerlink" href="#simple-formats" title="Link to this heading">#</a></h3>
<p>Before discussing file formats specifically developed for single-cell data we want to briefly mention that common simple text file formats (such as CSV, TSV, JSON etc.) can often be the answer to transferring data between languages. They work well in cases where some analysis has been performed and what you want to transfer is a subset of the information about an experiment. For example, you may want to transfer only the cell metadata but do not require the feature metadata, expression matrices etc. The advantage of using simple text formats is that they are well supported by almost any language and do not require single-cell specific packages. However, they can quickly become impractical as what you want to transfer becomes more complex.</p>
</section>
<section id="hdf5-based-formats">
<h3><span class="section-number">5.3.2. </span>HDF5-based formats<a class="headerlink" href="#hdf5-based-formats" title="Link to this heading">#</a></h3>
<p>The most common disk formats for single-cell data are based on <a class="reference external" href="https://www.hdfgroup.org/solutions/hdf5/">Hierarchical Data Format version 5</a> or HDF5. This is an open-source file format designed for storing large, complex and heterogeneous data. It has a file directory type structure (similar to how files and folders are organised on your computer) which allows many different kinds of data to be stored in a single file with an arbitrarily complex hierarchy. While this format is very flexible, to properly interact with it you need to know where and how the different information is stored. For this reason, standard specifications for storing single-cell data in HDF5 files have been developed.</p>
<section id="h5ad">
<h4><span class="section-number">5.3.2.1. </span>H5AD<a class="headerlink" href="#h5ad" title="Link to this heading">#</a></h4>
<p>The H5AD format is the HDF5 disk representation of the <code class="docutils literal notranslate"><span class="pre">AnnData</span></code> object used by scverse packages and is commonly used to share single-cell datasets. As it is part of the scverse ecosystem, reading and writing these files from Python is well-supported and is part of the core functionality of the <a class="reference external" href="https://anndata.readthedocs.io/en/latest/index.html"><strong>anndata</strong> package</a> (read more about the format <a class="reference external" href="https://anndata.readthedocs.io/en/latest/fileformat-prose.html">here</a>).</p>
<p>To demonstrate interoperability we will use a small, randomly generated dataset that has gone through some of the steps of a standard analysis workflow to populate the various slots.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Create a randomly generated AnnData object to use as an example</span>
<span class="n">counts</span> <span class="o">=</span> <span class="n">csr_matrix</span><span class="p">(</span><span class="n">numpy</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">poisson</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="n">size</span><span class="o">=</span><span class="p">(</span><span class="mi">100</span><span class="p">,</span> <span class="mi">2000</span><span class="p">)),</span> <span class="n">dtype</span><span class="o">=</span><span class="n">numpy</span><span class="o">.</span><span class="n">float32</span><span class="p">)</span>
<span class="n">adata</span> <span class="o">=</span> <span class="n">anndata</span><span class="o">.</span><span class="n">AnnData</span><span class="p">(</span><span class="n">counts</span><span class="p">)</span>
<span class="n">adata</span><span class="o">.</span><span class="n">obs_names</span> <span class="o">=</span> <span class="p">[</span><span class="sa">f</span><span class="s2">&quot;Cell_</span><span class="si">{</span><span class="n">i</span><span class="si">:</span><span class="s2">d</span><span class="si">}</span><span class="s2">&quot;</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">adata</span><span class="o">.</span><span class="n">n_obs</span><span class="p">)]</span>
<span class="n">adata</span><span class="o">.</span><span class="n">var_names</span> <span class="o">=</span> <span class="p">[</span><span class="sa">f</span><span class="s2">&quot;Gene_</span><span class="si">{</span><span class="n">i</span><span class="si">:</span><span class="s2">d</span><span class="si">}</span><span class="s2">&quot;</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">adata</span><span class="o">.</span><span class="n">n_vars</span><span class="p">)]</span>
<span class="c1"># Do some standard processing to populate the object</span>
<span class="n">scanpy</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">calculate_qc_metrics</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">inplace</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="n">adata</span><span class="o">.</span><span class="n">layers</span><span class="p">[</span><span class="s2">&quot;counts&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">adata</span><span class="o">.</span><span class="n">X</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
<span class="n">scanpy</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">normalize_total</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">inplace</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="n">scanpy</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">log1p</span><span class="p">(</span><span class="n">adata</span><span class="p">)</span>
<span class="n">scanpy</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">highly_variable_genes</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">inplace</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="n">scanpy</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">pca</span><span class="p">(</span><span class="n">adata</span><span class="p">)</span>
<span class="n">scanpy</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">neighbors</span><span class="p">(</span><span class="n">adata</span><span class="p">)</span>
<span class="n">scanpy</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">umap</span><span class="p">(</span><span class="n">adata</span><span class="p">)</span>
<span class="n">adata</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>AnnData object with n_obs × n_vars = 100 × 2000
    obs: &#39;n_genes_by_counts&#39;, &#39;log1p_n_genes_by_counts&#39;, &#39;total_counts&#39;, &#39;log1p_total_counts&#39;, &#39;pct_counts_in_top_50_genes&#39;, &#39;pct_counts_in_top_100_genes&#39;, &#39;pct_counts_in_top_200_genes&#39;, &#39;pct_counts_in_top_500_genes&#39;
    var: &#39;n_cells_by_counts&#39;, &#39;mean_counts&#39;, &#39;log1p_mean_counts&#39;, &#39;pct_dropout_by_counts&#39;, &#39;total_counts&#39;, &#39;log1p_total_counts&#39;, &#39;highly_variable&#39;, &#39;means&#39;, &#39;dispersions&#39;, &#39;dispersions_norm&#39;
    uns: &#39;log1p&#39;, &#39;hvg&#39;, &#39;pca&#39;, &#39;neighbors&#39;, &#39;umap&#39;
    obsm: &#39;X_pca&#39;, &#39;X_umap&#39;
    varm: &#39;PCs&#39;
    layers: &#39;counts&#39;
    obsp: &#39;distances&#39;, &#39;connectivities&#39;
</pre></div>
</div>
</div>
</div>
<p>We will write this mock object to disk as a H5AD file to demonstrate how those files can be read from R.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">temp_dir</span> <span class="o">=</span> <span class="n">tempfile</span><span class="o">.</span><span class="n">TemporaryDirectory</span><span class="p">()</span>
<span class="n">h5ad_file</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">temp_dir</span><span class="o">.</span><span class="n">name</span><span class="p">,</span> <span class="s2">&quot;example.h5ad&quot;</span><span class="p">)</span>

<span class="n">adata</span><span class="o">.</span><span class="n">write_h5ad</span><span class="p">(</span><span class="n">h5ad_file</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<p>Several packages exist for reading and writing H5AD files from R. While they result in a file on disk these packages usually rely on wrapping the Python <strong>anndata</strong> package to handle the actual reading and writing of files with an in-memory conversion step to convert between R and Python.</p>
<section id="reading-writing-h5ad-with-bioconductor">
<h5><span class="section-number">5.3.2.1.1. </span>Reading/writing H5AD with Bioconductor<a class="headerlink" href="#reading-writing-h5ad-with-bioconductor" title="Link to this heading">#</a></h5>
<p>The <a class="reference external" href="https://bioconductor.org/packages/zellkonverter/">Bioconductor <strong>{zellkonverter}</strong> package</a> helps makes this easier by using the <a class="reference external" href="https://bioconductor.org/packages/basilisk/"><strong>{basilisk}</strong> package</a> to manage creating an appropriate Python environment. If that all sounds a bit technical, the end result is that Bioconductor users can read and write H5AD files using commands like below without requiring any knowledge of Python.</p>
<p>Unfortunately, because of the way this book is made, we are unable to run the code directly here. Instead, we will show the code and what the output looks like when run in an R session:</p>
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="n">sce</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">zellkonverter</span><span class="o">::</span><span class="nf">readH5AD</span><span class="p">(</span><span class="n">h5ad_file</span><span class="p">,</span><span class="w"> </span><span class="n">verbose</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="kc">TRUE</span><span class="p">)</span>
</pre></div>
</div>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>ℹ Using the Python reader
ℹ Using anndata version 0.8.0
✔ Read /.../luke.zappia/Downloads/example.h5ad [113ms]
✔ uns$hvg$flavor converted [17ms]
✔ uns$hvg converted [50ms]
✔ uns$log1p converted [25ms]
✔ uns$neighbors converted [18ms]
✔ uns$pca$params$use_highly_variable converted [16ms]
✔ uns$pca$params$zero_center converted [16ms]
✔ uns$pca$params converted [80ms]
✔ uns$pca$variance converted [17ms]
✔ uns$pca$variance_ratio converted [16ms]
✔ uns$pca converted [184ms]
✔ uns$umap$params$a converted [16ms]
✔ uns$umap$params$b converted [16ms]
✔ uns$umap$params converted [80ms]
✔ uns$umap converted [112ms]
✔ uns converted [490ms]
✔ Converting uns to metadata ... done
✔ X matrix converted to assay [29ms]
✔ layers$counts converted [27ms]
✔ Converting layers to assays ... done
✔ var converted to rowData [25ms]
✔ obs converted to colData [24ms]
✔ varm$PCs converted [18ms]
✔ varm converted [47ms]
✔ Converting varm to rowData$varm ... done
✔ obsm$X_pca converted [15ms]
✔ obsm$X_umap converted [16ms]
✔ obsm converted [80ms]
✔ Converting obsm to reducedDims ... done
ℹ varp is empty and was skipped
✔ obsp$connectivities converted [22ms]
✔ obsp$distances converted [23ms]
✔ obsp converted [92ms]
✔ Converting obsp to colPairs ... done
✔ SingleCellExperiment constructed [164ms]
ℹ Skipping conversion of raw
✔ Converting AnnData to SingleCellExperiment ... done
</pre></div>
</div>
<p>Because we have turned on the verbose output you can see how <strong>{zellkonverter}</strong> reads the file using Python and converts each part of the <code class="docutils literal notranslate"><span class="pre">AnnData</span></code> object to a Bioconductor <code class="docutils literal notranslate"><span class="pre">SingleCellExperiment</span></code> object. We can see what the result looks like:</p>
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="n">sce</span>
</pre></div>
</div>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">class</span><span class="p">:</span> <span class="n">SingleCellExperiment</span>
<span class="n">dim</span><span class="p">:</span> <span class="mi">2000</span> <span class="mi">100</span>
<span class="n">metadata</span><span class="p">(</span><span class="mi">5</span><span class="p">):</span> <span class="n">hvg</span> <span class="n">log1p</span> <span class="n">neighbors</span> <span class="n">pca</span> <span class="n">umap</span>
<span class="n">assays</span><span class="p">(</span><span class="mi">2</span><span class="p">):</span> <span class="n">X</span> <span class="n">counts</span>
<span class="n">rownames</span><span class="p">(</span><span class="mi">2000</span><span class="p">):</span> <span class="n">Gene_0</span> <span class="n">Gene_1</span> <span class="o">...</span> <span class="n">Gene_1998</span> <span class="n">Gene_1999</span>
<span class="n">rowData</span> <span class="n">names</span><span class="p">(</span><span class="mi">11</span><span class="p">):</span> <span class="n">n_cells_by_counts</span> <span class="n">mean_counts</span> <span class="o">...</span> <span class="n">dispersions_norm</span>
  <span class="n">varm</span>
<span class="n">colnames</span><span class="p">(</span><span class="mi">100</span><span class="p">):</span> <span class="n">Cell_0</span> <span class="n">Cell_1</span> <span class="o">...</span> <span class="n">Cell_98</span> <span class="n">Cell_99</span>
<span class="n">colData</span> <span class="n">names</span><span class="p">(</span><span class="mi">8</span><span class="p">):</span> <span class="n">n_genes_by_counts</span> <span class="n">log1p_n_genes_by_counts</span> <span class="o">...</span>
  <span class="n">pct_counts_in_top_200_genes</span> <span class="n">pct_counts_in_top_500_genes</span>
<span class="n">reducedDimNames</span><span class="p">(</span><span class="mi">2</span><span class="p">):</span> <span class="n">X_pca</span> <span class="n">X_umap</span>
<span class="n">mainExpName</span><span class="p">:</span> <span class="n">NULL</span>
<span class="n">altExpNames</span><span class="p">(</span><span class="mi">0</span><span class="p">):</span>
</pre></div>
</div>
<p>This object can then be used as normal by any Bioconductor package. If we want to write a new H5AD file we can use the <code class="docutils literal notranslate"><span class="pre">writeH5AD()</span></code> function:</p>
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="n">zellkonverter_h5ad_file</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">tempfile</span><span class="p">(</span><span class="n">fileext</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;.h5ad&quot;</span><span class="p">)</span>
<span class="n">zellkonverter</span><span class="o">::</span><span class="nf">writeH5AD</span><span class="p">(</span><span class="n">sce</span><span class="p">,</span><span class="w"> </span><span class="n">zellkonverter_h5ad_file</span><span class="p">,</span><span class="w"> </span><span class="n">verbose</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="kc">TRUE</span><span class="p">)</span>
</pre></div>
</div>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>ℹ Using anndata version 0.8.0
ℹ Using the &#39;X&#39; assay as the X matrix
✔ Selected X matrix [29ms]
✔ assays$X converted to X matrix [50ms]
✔ additional assays converted to layers [30ms]
✔ rowData$varm converted to varm [28ms]
✔ reducedDims converted to obsm [68ms]
✔ metadata converted to uns [24ms]
ℹ rowPairs is empty and was skipped
✔ Converting AnnData to SingleCellExperiment ... done
✔ Wrote &#39;/.../.../rj/.../T/.../file102cfa97cc51.h5ad &#39; [133ms]
</pre></div>
</div>
<p>We can then read this file in Python:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">scanpy</span><span class="o">.</span><span class="n">read_h5ad</span><span class="p">(</span><span class="n">zellkonverter_h5ad_file</span><span class="p">)</span>
</pre></div>
</div>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>AnnData object with n_obs × n_vars = 100 × 2000
    obs: &#39;n_genes_by_counts&#39;, &#39;log1p_n_genes_by_counts&#39;, &#39;total_counts&#39;, &#39;log1p_total_counts&#39;, &#39;pct_counts_in_top_50_genes&#39;, &#39;pct_counts_in_top_100_genes&#39;, &#39;pct_counts_in_top_200_genes&#39;, &#39;pct_counts_in_top_500_genes&#39;
    var: &#39;n_cells_by_counts&#39;, &#39;mean_counts&#39;, &#39;log1p_mean_counts&#39;, &#39;pct_dropout_by_counts&#39;, &#39;total_counts&#39;, &#39;log1p_total_counts&#39;, &#39;highly_variable&#39;, &#39;means&#39;, &#39;dispersions&#39;, &#39;dispersions_norm&#39;
    uns: &#39;X_name&#39;, &#39;hvg&#39;, &#39;log1p&#39;, &#39;neighbors&#39;, &#39;pca&#39;, &#39;umap&#39;
    obsm: &#39;X_pca&#39;, &#39;X_umap&#39;
    varm: &#39;PCs&#39;
    layers: &#39;counts&#39;
    obsp: &#39;connectivities&#39;, &#39;distances&#39;
</pre></div>
</div>
<p>If this the first time that you run a <strong>{zellkonverter}</strong> function you will see that it first creates a special conda environment to use (which can take a while). Once that environment exists it will be re-used by following function calls. <strong>{zellkonverter}</strong> has additional options such as allowing you to selectively read or write parts for an object. Please refer to the package documentation for more details. Similar functionality for writing a <code class="docutils literal notranslate"><span class="pre">SingleCellExperimentObject</span></code> to an H5AD file can be found in the <a class="reference external" href="https://github.com/cellgeni/sceasy"><strong>{sceasy}</strong> package</a>. While these packages are effective, wrapping Python requires some overhead which should be addressed by native R H5AD writers/readers in the future.</p>
</section>
<section id="reading-writing-h5ad-with-seurat">
<h5><span class="section-number">5.3.2.1.2. </span>Reading/writing H5AD with <strong>{Seurat}</strong><a class="headerlink" href="#reading-writing-h5ad-with-seurat" title="Link to this heading">#</a></h5>
<p>Converting between a <code class="docutils literal notranslate"><span class="pre">Seurat</span></code> object and an H5AD file is a two-step process <a class="reference external" href="https://mojaveazure.github.io/seurat-disk/articles/convert-anndata.html">as suggested by this tutorial</a>. Firstly H5AD file is converted to a H5Seurat file (a custom HDF5 format for <code class="docutils literal notranslate"><span class="pre">Seurat</span></code> objects) using the <a class="reference external" href="https://mojaveazure.github.io/seurat-disk/"><strong>{SeuratDisk}</strong> package</a> and then this file is read as a <code class="docutils literal notranslate"><span class="pre">Seurat</span></code> object.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="o">%%</span><span class="k">R</span> -i h5ad_file

message(&quot;Converting H5AD to H5Seurat...&quot;)
SeuratDisk::Convert(h5ad_file, dest = &quot;h5seurat&quot;, overwrite = TRUE)
message(&quot;Reading H5Seurat...&quot;)
h5seurat_file &lt;- gsub(&quot;.h5ad&quot;, &quot;.h5seurat&quot;, h5ad_file)
seurat &lt;- SeuratDisk::LoadH5Seurat(h5seurat_file, assays = &quot;RNA&quot;)
message(&quot;Read Seurat object:&quot;)
seurat
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>R[write to console]: Converting H5AD to H5Seurat...
</pre></div>
</div>
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>R[write to console]: The legacy packages maptools, rgdal, and rgeos, underpinning the sp package,
which was just loaded, will retire in October 2023.
Please refer to R-spatial evolution reports for details, especially
https://r-spatial.org/r/2023/05/15/evolution4.html.
It may be desirable to make the sf package available;
package maintainers should consider adding sf to Suggests:.
The sp package is now running under evolution status 2
     (status 2 uses the sf package in place of rgdal)
</pre></div>
</div>
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>    WARNING: The R package &quot;reticulate&quot; only fixed recently
    an issue that caused a segfault when used with rpy2:
    https://github.com/rstudio/reticulate/pull/1188
    Make sure that you use a version of that package that includes
    the fix.
    
</pre></div>
</div>
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>R[write to console]: Registered S3 method overwritten by &#39;SeuratDisk&#39;:
  method            from  
  as.sparse.H5Group Seurat

R[write to console]: Warnung:
R[write to console]:  Unknown file type: h5ad

R[write to console]: Warnung:
R[write to console]:  &#39;assay&#39; not set, setting to &#39;RNA&#39;

R[write to console]: Creating h5Seurat file for version 3.1.5.9900

R[write to console]: Adding X as data

R[write to console]: Adding X as counts

R[write to console]: Adding meta.features from var

R[write to console]: Adding X_pca as cell embeddings for pca

R[write to console]: Adding X_umap as cell embeddings for umap

R[write to console]: Adding PCs as feature loadings fpr pca

R[write to console]: Adding miscellaneous information for pca

R[write to console]: Adding standard deviations for pca

R[write to console]: Adding miscellaneous information for umap

R[write to console]: Adding hvg to miscellaneous data

R[write to console]: Adding log1p to miscellaneous data

R[write to console]: Adding layer counts as data in assay counts

R[write to console]: Adding layer counts as counts in assay counts

R[write to console]: Reading H5Seurat...

R[write to console]: Validating h5Seurat file

R[write to console]: Warnung:
R[write to console]:  Feature names cannot have underscores (&#39;_&#39;), replacing with dashes (&#39;-&#39;)

R[write to console]: Initializing RNA with data

R[write to console]: Adding counts for RNA

R[write to console]: Adding feature-level metadata for RNA

R[write to console]: Adding reduction pca

R[write to console]: Adding cell embeddings for pca

R[write to console]: Adding feature loadings for pca

R[write to console]: Adding miscellaneous information for pca

R[write to console]: Adding reduction umap

R[write to console]: Adding cell embeddings for umap

R[write to console]: Adding miscellaneous information for umap

R[write to console]: Adding command information

R[write to console]: Adding cell-level metadata

R[write to console]: Read Seurat object:
</pre></div>
</div>
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>An object of class Seurat 
2000 features across 100 samples within 1 assay 
Active assay: RNA (2000 features, 0 variable features)
 2 dimensional reductions calculated: pca, umap
</pre></div>
</div>
</div>
</div>
<p>Note that because the structure of a <code class="docutils literal notranslate"><span class="pre">Seurat</span></code> object is quite different from <code class="docutils literal notranslate"><span class="pre">AnnData</span></code> and <code class="docutils literal notranslate"><span class="pre">SingleCellExperiment</span></code> objects the conversion process is more complex. See the <a class="reference external" href="https://mojaveazure.github.io/seurat-disk/reference/Convert.html">documentation of the conversion function</a> for more details on how this is done.</p>
<p>The <strong>{sceasy}</strong> package also provides a function for reading H5AD files as <code class="docutils literal notranslate"><span class="pre">Seurat</span></code> or <code class="docutils literal notranslate"><span class="pre">SingleCellExperiment</span></code> objects in a single step. <strong>{sceasy}</strong> also wraps Python functions but unlike <strong>{zellkonverter}</strong> it doesn’t use a special Python environment. This means you need to be responsible for setting up the environment, making sure that R can find it and that the correct packages are installed (again, this code is not run here).</p>
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="n">sceasy_seurat</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">sceasy</span><span class="o">::</span><span class="nf">convertFormat</span><span class="p">(</span><span class="n">h5ad_file</span><span class="p">,</span><span class="w"> </span><span class="n">from</span><span class="o">=</span><span class="s">&quot;anndata&quot;</span><span class="p">,</span><span class="w"> </span><span class="n">to</span><span class="o">=</span><span class="s">&quot;seurat&quot;</span><span class="p">)</span>
<span class="n">sceasy_seurat</span>
</pre></div>
</div>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="ne">Warning</span><span class="p">:</span> <span class="n">Feature</span> <span class="n">names</span> <span class="n">cannot</span> <span class="n">have</span> <span class="n">underscores</span> <span class="p">(</span><span class="s1">&#39;_&#39;</span><span class="p">),</span> <span class="n">replacing</span> <span class="k">with</span> <span class="n">dashes</span> <span class="p">(</span><span class="s1">&#39;-&#39;</span><span class="p">)</span>
<span class="n">X</span> <span class="o">-&gt;</span> <span class="n">counts</span>
<span class="n">An</span> <span class="nb">object</span> <span class="n">of</span> <span class="k">class</span> <span class="nc">Seurat</span>
<span class="mi">2000</span> <span class="n">features</span> <span class="n">across</span> <span class="mi">100</span> <span class="n">samples</span> <span class="n">within</span> <span class="mi">1</span> <span class="n">assay</span>
<span class="n">Active</span> <span class="n">assay</span><span class="p">:</span> <span class="n">RNA</span> <span class="p">(</span><span class="mi">2000</span> <span class="n">features</span><span class="p">,</span> <span class="mi">0</span> <span class="n">variable</span> <span class="n">features</span><span class="p">)</span>
 <span class="mi">2</span> <span class="n">dimensional</span> <span class="n">reductions</span> <span class="n">calculated</span><span class="p">:</span> <span class="n">pca</span><span class="p">,</span> <span class="n">umap</span>
</pre></div>
</div>
</section>
<section id="reading-writing-h5ad-with-anndata">
<h5><span class="section-number">5.3.2.1.3. </span>Reading/writing H5AD with <strong>{anndata}</strong><a class="headerlink" href="#reading-writing-h5ad-with-anndata" title="Link to this heading">#</a></h5>
<p>The R <a class="reference external" href="https://anndata.dynverse.org/index.html"><strong>{anndata}</strong> package</a> can also be used to read H5AD files. However, unlike the packages above it does not convert to a native R object. Instead it provides an R interface to the Python object. This is useful for accessing the data but few analysis packages will accept this as input so further in-memory conversion is usually required.</p>
</section>
</section>
<section id="loom">
<h4><span class="section-number">5.3.2.2. </span>Loom<a class="headerlink" href="#loom" title="Link to this heading">#</a></h4>
<p>The <a class="reference external" href="http://loompy.org/">Loom file format</a> is an older HDF5 specification for omics data. Unlike H5AD, it is not linked to a specific analysis ecosystem, although the structure is similar to <code class="docutils literal notranslate"><span class="pre">AnnData</span></code> and <code class="docutils literal notranslate"><span class="pre">SingleCellExperiment</span></code> objects. Packages implementing the Loom format exist for both <a class="reference external" href="https://github.com/mojaveazure/loomR">R</a> and <a class="reference external" href="https://pypi.org/project/loompy/">Python</a> as well as a <a class="reference external" href="https://bioconductor.org/packages/LoomExperiment/">Bioconductor package</a> for writing Loom files. However, it is often more convenient to use the higher-level interfaces provided by the core ecosystem packages. Apart from sharing datasets another common place Loom files are encountered is when spliced/unspliced reads are quantified using <a class="reference external" href="http://velocyto.org/">velocycto</a> for <a class="reference internal" href="../trajectories/rna_velocity.html#trajectories-rna-velocity"><span class="std std-ref">RNA velocity analysis</span></a>.</p>
</section>
</section>
<section id="rds-files">
<h3><span class="section-number">5.3.3. </span>RDS files<a class="headerlink" href="#rds-files" title="Link to this heading">#</a></h3>
<p>Another file format you may see used to share single-cell datasets is the RDS format. This is a binary format used to serialise arbitrary R objects (similar to Python Pickle files). As <code class="docutils literal notranslate"><span class="pre">SingleCellExperiment</span></code> and <code class="docutils literal notranslate"><span class="pre">Seurat</span></code> objects did not always have matching on-disk representations RDS files are sometimes used to share the results from R analyses. While this is ok within an analysis project we discourage its use for sharing data publicly or with collaborators due to the lack of interoperability with other ecosystems. Instead, we recommend using one of the HDF5 formats mentioned above that can be read from multiple languages.</p>
</section>
<section id="new-on-disk-formats">
<h3><span class="section-number">5.3.4. </span>New on-disk formats<a class="headerlink" href="#new-on-disk-formats" title="Link to this heading">#</a></h3>
<p>While HDF5-based formats are currently the standard for on-disk representations of single-cell data other newer technologies such as <a class="reference external" href="https://zarr.dev/">Zarr</a> and <a class="reference external" href="https://tiledb.com/">TileDB</a> have some advantages, particularly for very large datasets and other modalities. We expect specifications to be developed for these formats in the future which may be adopted by the community (<strong>anndata</strong> already provides support for Zarr files).</p>
</section>
</section>
<section id="in-memory-interoperability">
<h2><span class="section-number">5.4. </span>In-memory interoperability<a class="headerlink" href="#in-memory-interoperability" title="Link to this heading">#</a></h2>
<p>The second approach to interoperability is to work on in-memory representations of an object. This approach involves active sessions from two programming languages running at the same time and either accessing the same object from both or converting between them as needed. Usually, one language acts as the main environment and there is an interface to the other language. This can be very useful for interactive analysis as it allows an analyst to work in two languages simultaneously. It is also often used when creating documents that use multiple languages (such as this book). However, in-memory interoperability has some drawbacks as it requires the analyst to be familiar with setting up and using both environments, more complex objects are often not supported by both languages and there is a greater memory overhead as data can easily become duplicated (making it difficult to use for larger datasets).</p>
<section id="interoperability-between-r-ecosystems">
<h3><span class="section-number">5.4.1. </span>Interoperability between R ecosystems<a class="headerlink" href="#interoperability-between-r-ecosystems" title="Link to this heading">#</a></h3>
<p>Before we look at in-memory interoperability between R and Python first let’s consider the simpler case of converting between the two R ecosystems. The <strong>{Seurat}</strong> package provides functions for performing this conversion <a class="reference external" href="https://satijalab.org/seurat/articles/conversion_vignette.html">as described in this vignette</a>.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="o">%%</span><span class="k">R</span>
sce_from_seurat &lt;- Seurat::as.SingleCellExperiment(seurat)
sce_from_seurat
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>class: SingleCellExperiment 
dim: 2000 100 
metadata(0):
assays(2): X logcounts
rownames(2000): Gene-0 Gene-1 ... Gene-1998 Gene-1999
rowData names(0):
colnames(100): Cell_0 Cell_1 ... Cell_98 Cell_99
colData names(9): n_genes_by_counts log1p_n_genes_by_counts ...
  pct_counts_in_top_500_genes ident
reducedDimNames(2): PCA UMAP
mainExpName: NULL
altExpNames(0):
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="o">%%</span><span class="k">R</span>
seurat_from_sce &lt;- Seurat::as.Seurat(sce_from_seurat)
seurat_from_sce
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>An object of class Seurat 
2000 features across 100 samples within 1 assay 
Active assay: RNA (2000 features, 0 variable features)
 2 dimensional reductions calculated: PCA, UMAP
</pre></div>
</div>
</div>
</div>
<p>The difficult part here is due to the differences between the structures of the two objects. It is important to make sure the arguments are set correctly so that the conversion functions know which information to convert and where to place it.</p>
<p>In many cases it may not be necessary to convert a <code class="docutils literal notranslate"><span class="pre">Seurat</span></code> object to a <code class="docutils literal notranslate"><span class="pre">SingleCellExperiment</span></code>. This is because many of the core Bioconductor packages for single-cell analysis have been designed to also accept a matrix as input.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="o">%%</span><span class="k">R</span>
# Calculate Counts Per Million using the Bioconductor scuttle package
# with a matrix in a Seurat object
cpm &lt;- scuttle::calculateCPM(Seurat::GetAssayData(seurat, slot = &quot;counts&quot;))
cpm[1:10, 1:10]
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>10 x 10 sparse Matrix of class &quot;dgCMatrix&quot;
                                                                       
 [1,] 602.1263 622.1264  600.5326 1416.439   .         .       965.8600
 [2,] 602.1263   .         .       613.435 618.2562  943.8910  609.1107
 [3,] 602.1263 982.8946 1202.6879  969.506 618.2562  943.8910 1219.1005
 [4,]   .        .       600.5326  613.435 976.3175  594.3451    .     
 [5,] 602.1263 622.1264    .      1221.384 618.2562  594.3451  609.1107
 [6,] 954.8394 982.8946 1202.6879  613.435 618.2562  943.8910 1219.1005
 [7,] 954.8394 622.1264  952.6379  613.435 618.2562    .       965.8600
 [8,]   .      982.8946    .         .     618.2562  594.3451    .     
 [9,] 954.8394   .       600.5326  613.435   .      1192.4227 1219.1005
[10,] 954.8394 622.1264  952.6379  613.435 618.2562  943.8910  609.1107
                                  
 [1,]  958.4081 599.5090  610.1969
 [2,]  958.4081 952.2526  610.1969
 [3,]    .        .       965.9120
 [4,]    .        .         .     
 [5,]  605.0013 952.2526    .     
 [6,]    .        .         .     
 [7,]  605.0013 599.5090  965.9120
 [8,] 1209.0169   .       965.9120
 [9,]    .      599.5090 1413.3168
[10,] 1209.0169   .       610.1969
</pre></div>
</div>
</div>
</div>
<p>However, it is important to be sure you are accessing the right information and storing any results in the correct place if needed.</p>
</section>
<section id="accessing-r-from-python">
<h3><span class="section-number">5.4.2. </span>Accessing R from Python<a class="headerlink" href="#accessing-r-from-python" title="Link to this heading">#</a></h3>
<p>The Python interface to R is provided by the <a class="reference external" href="https://rpy2.github.io/doc/latest/html/index.html"><strong>rpy2</strong> package</a>. This allows you to access R functions and objects from Python. For example:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">counts_mat</span> <span class="o">=</span> <span class="n">adata</span><span class="o">.</span><span class="n">layers</span><span class="p">[</span><span class="s2">&quot;counts&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">T</span>
<span class="n">rpy2</span><span class="o">.</span><span class="n">robjects</span><span class="o">.</span><span class="n">globalenv</span><span class="p">[</span><span class="s2">&quot;counts_mat&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">counts_mat</span>
<span class="n">cpm</span> <span class="o">=</span> <span class="n">rpy2</span><span class="o">.</span><span class="n">robjects</span><span class="o">.</span><span class="n">r</span><span class="p">(</span><span class="s2">&quot;scuttle::calculateCPM(counts_mat)&quot;</span><span class="p">)</span>
<span class="n">cpm</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>&lt;2000x100 sparse matrix of type &#39;&lt;class &#39;numpy.float64&#39;&gt;&#39;
	with 126146 stored elements in Compressed Sparse Column format&gt;
</pre></div>
</div>
</div>
</div>
<p>Common Python objects (lists, matrices, <code class="docutils literal notranslate"><span class="pre">DataFrame</span></code>s etc.) can also be passed to R.</p>
<p>If you are using a Jupyter notebook (as we are for this book) you can use the IPython magic interface to create cells with native R code (passing objects as required). For example, starting a cell with <code class="docutils literal notranslate"><span class="pre">%%R</span> <span class="pre">-i</span> <span class="pre">input</span> <span class="pre">-o</span> <span class="pre">output</span></code> says to take <code class="docutils literal notranslate"><span class="pre">input</span></code> as input, run R code and then return <code class="docutils literal notranslate"><span class="pre">output</span></code> as output.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="o">%%</span><span class="k">R</span> -i counts_mat -o magic_cpm
# R code running using IPython magic
magic_cpm &lt;- scuttle::calculateCPM(counts_mat)
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Python code accessing the results</span>
<span class="n">magic_cpm</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>&lt;2000x100 sparse matrix of type &#39;&lt;class &#39;numpy.float64&#39;&gt;&#39;
	with 126146 stored elements in Compressed Sparse Column format&gt;
</pre></div>
</div>
</div>
</div>
<p>This is the approach you will most commonly see in later chapters. For more information about using <strong>rpy2</strong> please refer to <a class="reference external" href="https://rpy2.github.io/doc/latest/html/index.html">the documentation</a>.</p>
<p>To work with single-cell data in this way the <a class="reference external" href="https://icb-anndata2ri.readthedocs-hosted.com/en/latest/"><strong>anndata2ri</strong> package</a> is especially useful. This is an extension to <strong>rpy2</strong> which allows R to see <code class="docutils literal notranslate"><span class="pre">AnnData</span></code> objects as <code class="docutils literal notranslate"><span class="pre">SingleCellExperiment</span></code> objects. This avoids unnecessary conversion and makes it easy to run R code on a Python object. It also enables the conversion of sparse <strong>scipy</strong> matrices like we saw above.</p>
<p>In this example, we pass an <code class="docutils literal notranslate"><span class="pre">AnnData</span></code> object in the Python session to R which views it as a <code class="docutils literal notranslate"><span class="pre">SingleCellExperiment</span></code> that can be used by R functions.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="o">%%</span><span class="k">R</span> -i adata
qc &lt;- scuttle::perCellQCMetrics(adata)
head(qc)
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>/Users/luke.zappia/miniconda3/envs/interoperability2/lib/python3.9/functools.py:888: NotConvertedWarning: Conversion &#39;py2rpy&#39; not defined for objects of type &#39;&lt;class &#39;NoneType&#39;&gt;&#39;
  return dispatch(args[0].__class__)(*args, **kw)
</pre></div>
</div>
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>        sum detected total
Cell_0 2005     1274  2005
Cell_1 1941     1233  1941
Cell_2 2011     1270  2011
Cell_3 1947     1268  1947
Cell_4 1933     1265  1933
Cell_5 2031     1289  2031
</pre></div>
</div>
</div>
</div>
<p>Note that you will still run into issues if an object (or part of it) cannot be interfaced correctly (for example if there is an unsupported data type). In that case, you may need to modify your object first before it can be accessed.</p>
</section>
<section id="accessing-python-from-r">
<h3><span class="section-number">5.4.3. </span>Accessing Python from R<a class="headerlink" href="#accessing-python-from-r" title="Link to this heading">#</a></h3>
<p>Accessing Python from an R session is similar to accessing R from Python but here the interface is provided by the <a class="reference external" href="https://rstudio.github.io/reticulate/"><strong>{reticulate}</strong> package</a>. Once it is loaded we can access Python functions and objects from R.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="o">%%</span><span class="k">R</span>
reticulate_list &lt;- reticulate::r_to_py(LETTERS)
print(reticulate_list)
py_builtins &lt;- reticulate::import_builtins()
py_builtins$zip(letters, LETTERS)
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>List (26 items)
&lt;zip object at 0x18a243040&gt;
</pre></div>
</div>
</div>
</div>
<p>If you are working in an <a class="reference external" href="https://rmarkdown.rstudio.com/">RMarkdown</a> or <a class="reference external" href="https://quarto.org/">Quarto</a> document you can also write native Python chunks using the <strong>{reticulate}</strong> Python engine. When we do this we can use the magic <code class="docutils literal notranslate"><span class="pre">r</span></code> and <code class="docutils literal notranslate"><span class="pre">py</span></code> variables to access objects in the other language (the following code is an example that is not run).</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>```{r}
# An R chunk that accesses a Python object
print(py$py_object)
```

```{python}
# A Python chunk that accesses an R object
print(r$r_object)
```
</pre></div>
</div>
<p>Unlike <strong>anndata2ri</strong>, there are no R packages that provide a direct interface for Python to view <code class="docutils literal notranslate"><span class="pre">SingleCellExperiment</span></code> or <code class="docutils literal notranslate"><span class="pre">Seurat</span></code> objects as <code class="docutils literal notranslate"><span class="pre">AnnData</span></code> objects.  However, we can still access most parts of an <code class="docutils literal notranslate"><span class="pre">AnnData</span></code> using <strong>{reticulate}</strong> (this code is not run).</p>
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="c1"># Print an AnnData object in a Python environment</span>
<span class="n">py</span><span class="o">$</span><span class="n">adata</span>
</pre></div>
</div>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>AnnData object with n_obs × n_vars = 100 × 2000
    obs: &#39;n_genes_by_counts&#39;, &#39;log1p_n_genes_by_counts&#39;, &#39;total_counts&#39;, &#39;log1p_total_counts&#39;, &#39;pct_counts_in_top_50_genes&#39;, &#39;pct_counts_in_top_100_genes&#39;, &#39;pct_counts_in_top_200_genes&#39;, &#39;pct_counts_in_top_500_genes&#39;
    var: &#39;n_cells_by_counts&#39;, &#39;mean_counts&#39;, &#39;log1p_mean_counts&#39;, &#39;pct_dropout_by_counts&#39;, &#39;total_counts&#39;, &#39;log1p_total_counts&#39;, &#39;highly_variable&#39;, &#39;means&#39;, &#39;dispersions&#39;, &#39;dispersions_norm&#39;
    uns: &#39;hvg&#39;, &#39;log1p&#39;, &#39;neighbors&#39;, &#39;pca&#39;, &#39;umap&#39;
    obsm: &#39;X_pca&#39;, &#39;X_umap&#39;
    varm: &#39;PCs&#39;
    layers: &#39;counts&#39;
    obsp: &#39;connectivities&#39;, &#39;distances&#39;
</pre></div>
</div>
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="c1"># Alternatively use the Python anndata package to read a H5AD file</span>
<span class="n">anndata</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">reticulate</span><span class="o">::</span><span class="nf">import</span><span class="p">(</span><span class="s">&quot;anndata&quot;</span><span class="p">)</span>
<span class="n">anndata</span><span class="o">$</span><span class="nf">read_h5ad</span><span class="p">(</span><span class="n">h5ad_file</span><span class="p">)</span>
</pre></div>
</div>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>AnnData object with n_obs × n_vars = 100 × 2000
    obs: &#39;n_genes_by_counts&#39;, &#39;log1p_n_genes_by_counts&#39;, &#39;total_counts&#39;, &#39;log1p_total_counts&#39;, &#39;pct_counts_in_top_50_genes&#39;, &#39;pct_counts_in_top_100_genes&#39;, &#39;pct_counts_in_top_200_genes&#39;, &#39;pct_counts_in_top_500_genes&#39;
    var: &#39;n_cells_by_counts&#39;, &#39;mean_counts&#39;, &#39;log1p_mean_counts&#39;, &#39;pct_dropout_by_counts&#39;, &#39;total_counts&#39;, &#39;log1p_total_counts&#39;, &#39;highly_variable&#39;, &#39;means&#39;, &#39;dispersions&#39;, &#39;dispersions_norm&#39;
    uns: &#39;hvg&#39;, &#39;log1p&#39;, &#39;neighbors&#39;, &#39;pca&#39;, &#39;umap&#39;
    obsm: &#39;X_pca&#39;, &#39;X_umap&#39;
    varm: &#39;PCs&#39;
    layers: &#39;counts&#39;
    obsp: &#39;connectivities&#39;, &#39;distances&#39;
</pre></div>
</div>
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="c1"># Access the obs slot, pandas DataFrames are automatically converted to R data.frames</span>
<span class="nf">head</span><span class="p">(</span><span class="n">adata</span><span class="o">$</span><span class="n">obs</span><span class="p">)</span>
</pre></div>
</div>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>       <span class="n">n_genes_by_counts</span> <span class="n">log1p_n_genes_by_counts</span> <span class="n">total_counts</span>
<span class="n">Cell_0</span>              <span class="mi">1246</span>                <span class="mf">7.128496</span>         <span class="mi">1965</span>
<span class="n">Cell_1</span>              <span class="mi">1262</span>                <span class="mf">7.141245</span>         <span class="mi">2006</span>
<span class="n">Cell_2</span>              <span class="mi">1262</span>                <span class="mf">7.141245</span>         <span class="mi">1958</span>
<span class="n">Cell_3</span>              <span class="mi">1240</span>                <span class="mf">7.123673</span>         <span class="mi">1960</span>
<span class="n">Cell_4</span>              <span class="mi">1296</span>                <span class="mf">7.167809</span>         <span class="mi">2027</span>
<span class="n">Cell_5</span>              <span class="mi">1231</span>                <span class="mf">7.116394</span>         <span class="mi">1898</span>
       <span class="n">log1p_total_counts</span> <span class="n">pct_counts_in_top_50_genes</span>
<span class="n">Cell_0</span>           <span class="mf">7.583756</span>                  <span class="mf">10.025445</span>
<span class="n">Cell_1</span>           <span class="mf">7.604396</span>                   <span class="mf">9.521436</span>
<span class="n">Cell_2</span>           <span class="mf">7.580189</span>                   <span class="mf">9.959142</span>
<span class="n">Cell_3</span>           <span class="mf">7.581210</span>                   <span class="mf">9.183673</span>
<span class="n">Cell_4</span>           <span class="mf">7.614805</span>                   <span class="mf">9.718796</span>
<span class="n">Cell_5</span>           <span class="mf">7.549083</span>                  <span class="mf">10.168599</span>
       <span class="n">pct_counts_in_top_100_genes</span> <span class="n">pct_counts_in_top_200_genes</span>
<span class="n">Cell_0</span>                    <span class="mf">17.65903</span>                    <span class="mf">30.89059</span>
<span class="n">Cell_1</span>                    <span class="mf">16.99900</span>                    <span class="mf">29.71087</span>
<span class="n">Cell_2</span>                    <span class="mf">17.62002</span>                    <span class="mf">30.28601</span>
<span class="n">Cell_3</span>                    <span class="mf">16.83673</span>                    <span class="mf">30.45918</span>
<span class="n">Cell_4</span>                    <span class="mf">17.11889</span>                    <span class="mf">30.04440</span>
<span class="n">Cell_5</span>                    <span class="mf">18.07165</span>                    <span class="mf">30.29505</span>
       <span class="n">pct_counts_in_top_500_genes</span>
<span class="n">Cell_0</span>                    <span class="mf">61.42494</span>
<span class="n">Cell_1</span>                    <span class="mf">59.62114</span>
<span class="n">Cell_2</span>                    <span class="mf">60.92952</span>
<span class="n">Cell_3</span>                    <span class="mf">61.07143</span>
<span class="n">Cell_4</span>                    <span class="mf">59.64480</span>
<span class="n">Cell_5</span>                    <span class="mf">61.48577</span>
</pre></div>
</div>
<p>As mentioned above the R <strong>{anndata}</strong> package provides an R interface for <code class="docutils literal notranslate"><span class="pre">AnnData</span></code> objects but it is not currently used by many analysis packages.</p>
<p>For more complex analysis that requires a whole object to work on it may be necessary to completely convert an object from R to Python (or the opposite). This is not memory efficient as it creates a duplicate of the data but it does provide access to a greater range of packages. The <strong>{zellkonverter}</strong> package provides a function for doing this conversion (note that, unlike the function for reading H5AD files, this uses the normal Python environment rather than a specially created one) (code not run).</p>
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="c1"># Convert an AnnData to a SingleCellExperiment</span>
<span class="n">sce</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">zellkonverter</span><span class="o">::</span><span class="nf">AnnData2SCE</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span><span class="w"> </span><span class="n">verbose</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="kc">TRUE</span><span class="p">)</span>
<span class="n">sce</span>
</pre></div>
</div>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>✔ uns$hvg$flavor converted [21ms]
✔ uns$hvg converted [62ms]
✔ uns$log1p converted [22ms]
✔ uns$neighbors converted [21ms]
✔ uns$pca$params$use_highly_variable converted [22ms]
✔ uns$pca$params$zero_center converted [31ms]
✔ uns$pca$params converted [118ms]
✔ uns$pca$variance converted [17ms]
✔ uns$pca$variance_ratio converted [17ms]
✔ uns$pca converted [224ms]
✔ uns$umap$params$a converted [15ms]
✔ uns$umap$params$b converted [17ms]
✔ uns$umap$params converted [80ms]
✔ uns$umap converted [115ms]
✔ uns converted [582ms]
✔ Converting uns to metadata ... done
✔ X matrix converted to assay [44ms]
✔ layers$counts converted [29ms]
✔ Converting layers to assays ... done
✔ var converted to rowData [37ms]
✔ obs converted to colData [23ms]
✔ varm$PCs converted [18ms]
✔ varm converted [49ms]
✔ Converting varm to rowData$varm ... done
✔ obsm$X_pca converted [17ms]
✔ obsm$X_umap converted [17ms]
✔ obsm converted [80ms]
✔ Converting obsm to reducedDims ... done
ℹ varp is empty and was skipped
✔ obsp$connectivities converted [21ms]
✔ obsp$distances converted [22ms]
✔ obsp converted [89ms]
✔ Converting obsp to colPairs ... done
✔ SingleCellExperiment constructed [241ms]
ℹ Skipping conversion of raw
✔ Converting AnnData to SingleCellExperiment ... done
class: SingleCellExperiment
dim: 2000 100
metadata(5): hvg log1p neighbors pca umap
assays(2): X counts
rownames(2000): Gene_0 Gene_1 ... Gene_1998 Gene_1999
rowData names(11): n_cells_by_counts mean_counts ... dispersions_norm
  varm
colnames(100): Cell_0 Cell_1 ... Cell_98 Cell_99
colData names(8): n_genes_by_counts log1p_n_genes_by_counts ...
  pct_counts_in_top_200_genes pct_counts_in_top_500_genes
reducedDimNames(2): X_pca X_umap
mainExpName: NULL
altExpNames(0):
</pre></div>
</div>
<p>The same can also be done in reverse:</p>
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="n">adata2</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">zellkonverter</span><span class="o">::</span><span class="nf">SCE2AnnData</span><span class="p">(</span><span class="n">sce</span><span class="p">,</span><span class="w"> </span><span class="n">verbose</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="kc">TRUE</span><span class="p">)</span>
<span class="n">adata2</span>
</pre></div>
</div>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>ℹ Using the &#39;X&#39; assay as the X matrix
✔ Selected X matrix [27ms]
✔ assays$X converted to X matrix [38ms]
✔ additional assays converted to layers [31ms]
✔ rowData$varm converted to varm [15ms]
✔ reducedDims converted to obsm [63ms]
✔ metadata converted to uns [23ms]
ℹ rowPairs is empty and was skipped
✔ Converting AnnData to SingleCellExperiment ... done
AnnData object with n_obs × n_vars = 100 × 2000
    obs: &#39;n_genes_by_counts&#39;, &#39;log1p_n_genes_by_counts&#39;, &#39;total_counts&#39;, &#39;log1p_total_counts&#39;, &#39;pct_counts_in_top_50_genes&#39;, &#39;pct_counts_in_top_100_genes&#39;, &#39;pct_counts_in_top_200_genes&#39;, &#39;pct_counts_in_top_500_genes&#39;
    var: &#39;n_cells_by_counts&#39;, &#39;mean_counts&#39;, &#39;log1p_mean_counts&#39;, &#39;pct_dropout_by_counts&#39;, &#39;total_counts&#39;, &#39;log1p_total_counts&#39;, &#39;highly_variable&#39;, &#39;means&#39;, &#39;dispersions&#39;, &#39;dispersions_norm&#39;
    uns: &#39;X_name&#39;, &#39;hvg&#39;, &#39;log1p&#39;, &#39;neighbors&#39;, &#39;pca&#39;, &#39;umap&#39;
    obsm: &#39;X_pca&#39;, &#39;X_umap&#39;
    varm: &#39;PCs&#39;
    layers: &#39;counts&#39;
    obsp: &#39;connectivities&#39;, &#39;distances&#39;
</pre></div>
</div>
</section>
</section>
<section id="interoperability-for-multimodal-data">
<h2><span class="section-number">5.5. </span>Interoperability for multimodal data<a class="headerlink" href="#interoperability-for-multimodal-data" title="Link to this heading">#</a></h2>
<p>The complexity of multimodal data presents additional challenges for interoperability. Both the <code class="docutils literal notranslate"><span class="pre">SingleCellExperiment</span></code> (through “alternative experiments”, which must match in the column dimension (cells)) and <code class="docutils literal notranslate"><span class="pre">Seurat</span></code> (using “assays”) objects can store multiple modalities but the <code class="docutils literal notranslate"><span class="pre">AnnData</span></code> object is restricted to unimodal data.</p>
<p>To address this limitation, the <code class="docutils literal notranslate"><span class="pre">MuData</span></code> object (introduced in the [analysis frameworks and tools chapter](<a class="reference internal" href="analysis_tools.html#introduction-analysis-frameworks"><span class="std std-ref">analysis frameworks and tools chapter</span></a>) was developed as as an extension of <code class="docutils literal notranslate"><span class="pre">AnnData</span></code> for multimodal datasets. The developers have considered interoperability in their design. While the main platform for MuData is Python, the authors have provided the <a class="reference external" href="https://pmbio.github.io/MuDataSeurat/">MuDataSeurat R package</a> for reading the on-disk H5MU format as <code class="docutils literal notranslate"><span class="pre">Seurat</span></code> objects and the <a class="reference external" href="https://bioconductor.org/packages/MuData/">MuData R package</a> for doing the same with Bioconductor <code class="docutils literal notranslate"><span class="pre">MultiAssayExperiment</span></code> objects. This official support is very useful but there are still some inconsistencies due to differences between the objects. The MuData authors also provide a <a class="reference external" href="https://docs.juliahub.com/Muon/QfqCh/0.1.1/objects/">Julia implementation</a> of <code class="docutils literal notranslate"><span class="pre">AnnData</span></code> and <code class="docutils literal notranslate"><span class="pre">MuData</span></code>.</p>
<p>Below is an example of reading and writing a small example <code class="docutils literal notranslate"><span class="pre">MuData</span></code> dataset using the Python and R packages.</p>
<section id="python">
<h3><span class="section-number">5.5.1. </span>Python<a class="headerlink" href="#python" title="Link to this heading">#</a></h3>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Read file</span>
<span class="n">mdata</span> <span class="o">=</span> <span class="n">mudata</span><span class="o">.</span><span class="n">read_h5mu</span><span class="p">(</span><span class="s2">&quot;../../datasets/original.h5mu&quot;</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="n">mdata</span><span class="p">)</span>

<span class="c1"># Write new file</span>
<span class="n">python_h5mu_file</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">temp_dir</span><span class="o">.</span><span class="n">name</span><span class="p">,</span> <span class="s2">&quot;python.h5mu&quot;</span><span class="p">)</span>
<span class="n">mdata</span><span class="o">.</span><span class="n">write_h5mu</span><span class="p">(</span><span class="n">python_h5mu_file</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>MuData object with n_obs × n_vars = 411 × 56
  obs:	&#39;louvain&#39;, &#39;leiden&#39;, &#39;leiden_wnn&#39;, &#39;celltype&#39;
  var:	&#39;gene_ids&#39;, &#39;feature_types&#39;, &#39;highly_variable&#39;
  obsm:	&#39;X_mofa&#39;, &#39;X_mofa_umap&#39;, &#39;X_umap&#39;, &#39;X_wnn_umap&#39;
  varm:	&#39;LFs&#39;
  obsp:	&#39;connectivities&#39;, &#39;distances&#39;, &#39;wnn_connectivities&#39;, &#39;wnn_distances&#39;
  2 modalities
    prot:	411 x 29
      var:	&#39;gene_ids&#39;, &#39;feature_types&#39;, &#39;highly_variable&#39;
      uns:	&#39;neighbors&#39;, &#39;pca&#39;, &#39;umap&#39;
      obsm:	&#39;X_pca&#39;, &#39;X_umap&#39;
      varm:	&#39;PCs&#39;
      layers:	&#39;counts&#39;
      obsp:	&#39;connectivities&#39;, &#39;distances&#39;
    rna:	411 x 27
      obs:	&#39;n_genes_by_counts&#39;, &#39;total_counts&#39;, &#39;total_counts_mt&#39;, &#39;pct_counts_mt&#39;, &#39;leiden&#39;, &#39;celltype&#39;
      var:	&#39;gene_ids&#39;, &#39;feature_types&#39;, &#39;mt&#39;, &#39;n_cells_by_counts&#39;, &#39;mean_counts&#39;, &#39;pct_dropout_by_counts&#39;, &#39;total_counts&#39;, &#39;highly_variable&#39;, &#39;means&#39;, &#39;dispersions&#39;, &#39;dispersions_norm&#39;, &#39;mean&#39;, &#39;std&#39;
      uns:	&#39;celltype_colors&#39;, &#39;hvg&#39;, &#39;leiden&#39;, &#39;leiden_colors&#39;, &#39;neighbors&#39;, &#39;pca&#39;, &#39;rank_genes_groups&#39;, &#39;umap&#39;
      obsm:	&#39;X_pca&#39;, &#39;X_umap&#39;
      varm:	&#39;PCs&#39;
      obsp:	&#39;connectivities&#39;, &#39;distances&#39;
</pre></div>
</div>
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>/Users/luke.zappia/miniconda3/envs/interoperability2/lib/python3.9/site-packages/anndata/_core/anndata.py:1230: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df[key] = c
</pre></div>
</div>
</div>
</div>
</section>
<section id="r">
<h3><span class="section-number">5.5.2. </span>R<a class="headerlink" href="#r" title="Link to this heading">#</a></h3>
<section id="bioconductor">
<h4><span class="section-number">5.5.2.1. </span>Bioconductor<a class="headerlink" href="#bioconductor" title="Link to this heading">#</a></h4>
<p>Read/write from/to a <code class="docutils literal notranslate"><span class="pre">MultiAssayExperiment</span></code> object</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="o">%%</span><span class="k">R</span>
mae &lt;- MuData::readH5MU(&quot;../../datasets/original.h5mu&quot;)
print(mae)

bioc_h5mu_file &lt;- tempfile(fileext = &quot;.h5mu&quot;)
MuData::writeH5MU(mae, bioc_h5mu_file)
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>A MultiAssayExperiment object of 2 listed
 experiments with user-defined names and respective classes.
 Containing an ExperimentList class object of length 2:
 [1] prot: SingleCellExperiment with 29 rows and 411 columns
 [2] rna: SingleCellExperiment with 27 rows and 411 columns
Functionality:
 experiments() - obtain the ExperimentList instance
 colData() - the primary/phenotype DataFrame
 sampleMap() - the sample coordination DataFrame
 `$`, `[`, `[[` - extract colData columns, subset, or experiment
 *Format() - convert into a long or wide DataFrame
 assays() - convert ExperimentList to a SimpleList of matrices
 exportClass() - save data to flat files
</pre></div>
</div>
</div>
</div>
</section>
<section id="seurat">
<h4><span class="section-number">5.5.2.2. </span>Seurat<a class="headerlink" href="#seurat" title="Link to this heading">#</a></h4>
<p>Read/write from/to a <code class="docutils literal notranslate"><span class="pre">Seurat</span></code> object</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="o">%%</span><span class="k">R</span>
seurat &lt;- MuDataSeurat::ReadH5MU(&quot;../../datasets/original.h5mu&quot;)
print(seurat)

seurat_h5mu_file &lt;- tempfile(fileext = &quot;.h5mu&quot;)
MuDataSeurat::WriteH5MU(seurat, seurat_h5mu_file)
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>R[write to console]: The legacy packages maptools, rgdal, and rgeos, underpinning the sp package,
which was just loaded, will retire in October 2023.
Please refer to R-spatial evolution reports for details, especially
https://r-spatial.org/r/2023/05/15/evolution4.html.
It may be desirable to make the sf package available;
package maintainers should consider adding sf to Suggests:.
The sp package is now running under evolution status 2
     (status 2 uses the sf package in place of rgdal)
</pre></div>
</div>
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>    WARNING: The R package &quot;reticulate&quot; only fixed recently
    an issue that caused a segfault when used with rpy2:
    https://github.com/rstudio/reticulate/pull/1188
    Make sure that you use a version of that package that includes
    the fix.
    
</pre></div>
</div>
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>R[write to console]: Warnung:
R[write to console]:  Keys should be one or more alphanumeric characters followed by an underscore, setting key from prot to prot_

R[write to console]: Warnung:
R[write to console]:  No columnames present in cell embeddings, setting to &#39;MOFA_1:30&#39;

R[write to console]: Warnung:
R[write to console]:  Keys should be one or more alphanumeric characters followed by an underscore, setting key from MOFA_UMAP_ to MOFAUMAP_

R[write to console]: Warnung:
R[write to console]:  All keys should be one or more alphanumeric characters followed by an underscore &#39;_&#39;, setting key to MOFAUMAP_

R[write to console]: Warnung:
R[write to console]:  No columnames present in cell embeddings, setting to &#39;MOFAUMAP_1:2&#39;

R[write to console]: Warnung:
R[write to console]:  No columnames present in cell embeddings, setting to &#39;UMAP_1:2&#39;

R[write to console]: Warnung:
R[write to console]:  Keys should be one or more alphanumeric characters followed by an underscore, setting key from WNN_UMAP_ to WNNUMAP_

R[write to console]: Warnung:
R[write to console]:  All keys should be one or more alphanumeric characters followed by an underscore &#39;_&#39;, setting key to WNNUMAP_

R[write to console]: Warnung:
R[write to console]:  No columnames present in cell embeddings, setting to &#39;WNNUMAP_1:2&#39;

R[write to console]: Warnung:
R[write to console]:  No columnames present in cell embeddings, setting to &#39;protPCA_1:31&#39;

R[write to console]: Warnung:
R[write to console]:  No columnames present in cell embeddings, setting to &#39;protUMAP_1:2&#39;

R[write to console]: Warnung:
R[write to console]:  No columnames present in cell embeddings, setting to &#39;rnaPCA_1:50&#39;

R[write to console]: Warnung:
R[write to console]:  No columnames present in cell embeddings, setting to &#39;rnaUMAP_1:2&#39;
</pre></div>
</div>
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>An object of class Seurat 
56 features across 411 samples within 2 assays 
Active assay: prot (29 features, 29 variable features)
 1 other assay present: rna
 8 dimensional reductions calculated: MOFA, MOFA_UMAP, UMAP, WNN_UMAP, protPCA, protUMAP, rnaPCA, rnaUMAP
</pre></div>
</div>
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>R[write to console]: Added .var[&#39;highly_variable&#39;] with highly variable features

R[write to console]: Added .var[&#39;highly_variable&#39;] with highly variable features
</pre></div>
</div>
</div>
</div>
</section>
</section>
</section>
<section id="interoperability-with-other-languages">
<h2><span class="section-number">5.6. </span>Interoperability with other languages<a class="headerlink" href="#interoperability-with-other-languages" title="Link to this heading">#</a></h2>
<p>Here we briefly list some resources and tools for the interoperability of single-cell data with languages other than R and Python.</p>
<section id="julia">
<h3><span class="section-number">5.6.1. </span>Julia<a class="headerlink" href="#julia" title="Link to this heading">#</a></h3>
<ul class="simple">
<li><p><a class="reference external" href="https://docs.juliahub.com/Muon/QfqCh/0.1.1/objects/">Muon.jl</a> provides Julia implementations of <code class="docutils literal notranslate"><span class="pre">AnnData</span></code> and <code class="docutils literal notranslate"><span class="pre">MuData</span></code> objects, as well as IO for the H5AD and H5MU formats</p></li>
<li><p><a class="reference external" href="https://github.com/maren-ha/scVI.jl">scVI.jl</a> provides a Julia implementation of <code class="docutils literal notranslate"><span class="pre">AnnData</span></code> as well as IO for the H5AD format</p></li>
</ul>
</section>
<section id="javascript">
<h3><span class="section-number">5.6.2. </span>JavaScript<a class="headerlink" href="#javascript" title="Link to this heading">#</a></h3>
<ul class="simple">
<li><p><a class="reference external" href="http://vitessce.io/">Vitessce</a> contains loaders from <code class="docutils literal notranslate"><span class="pre">AnnData</span></code> objects stored using the Zarr format</p></li>
<li><p>The <a class="reference external" href="https://github.com/jkanche/kana">kana family</a> supports reading H5AD files and <code class="docutils literal notranslate"><span class="pre">SingleCellExperiment</span></code> objects saved as RDS files</p></li>
</ul>
</section>
<section id="rust">
<h3><span class="section-number">5.6.3. </span>Rust<a class="headerlink" href="#rust" title="Link to this heading">#</a></h3>
<ul class="simple">
<li><p><a class="reference external" href="https://github.com/kaizhang/anndata-rs">anndata-rs</a> provides a Rust implementation of AnnData as well as advanced IO support for the H5AD format</p></li>
</ul>
</section>
</section>
<section id="session-information">
<h2><span class="section-number">5.7. </span>Session information<a class="headerlink" href="#session-information" title="Link to this heading">#</a></h2>
</section>
<section id="id1">
<h2><span class="section-number">5.8. </span>Python<a class="headerlink" href="#id1" title="Link to this heading">#</a></h2>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">session_info</span>

<span class="n">session_info</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><details>
<summary>Click to view session information</summary>
<pre>
-----
anndata             0.9.2
anndata2ri          1.2
mudata              0.2.3
numpy               1.24.4
rpy2                3.5.11
scanpy              1.9.3
scipy               1.9.3
session_info        1.0.0
-----
</pre>
<details>
<summary>Click to view modules imported as dependencies</summary>
<pre>
CoreFoundation              NA
Foundation                  NA
PIL                         10.0.0
PyObjCTools                 NA
anyio                       NA
appnope                     0.1.3
argcomplete                 NA
arrow                       1.2.3
asttokens                   NA
attr                        23.1.0
attrs                       23.1.0
babel                       2.12.1
backcall                    0.2.0
beta_ufunc                  NA
binom_ufunc                 NA
brotli                      1.0.9
certifi                     2023.07.22
cffi                        1.15.1
charset_normalizer          3.2.0
colorama                    0.4.6
comm                        0.1.4
cycler                      0.10.0
cython_runtime              NA
dateutil                    2.8.2
debugpy                     1.6.8
decorator                   5.1.1
defusedxml                  0.7.1
executing                   1.2.0
fastjsonschema              NA
fqdn                        NA
h5py                        3.9.0
hypergeom_ufunc             NA
idna                        3.4
importlib_metadata          NA
importlib_resources         NA
ipykernel                   6.25.0
ipython_genutils            0.2.0
ipywidgets                  8.1.0
isoduration                 NA
jedi                        0.19.0
jinja2                      3.1.2
joblib                      1.3.0
json5                       NA
jsonpointer                 2.0
jsonschema                  4.18.6
jsonschema_specifications   NA
jupyter_events              0.7.0
jupyter_server              2.7.0
jupyterlab_server           2.24.0
kiwisolver                  1.4.4
llvmlite                    0.40.1
markupsafe                  2.1.3
matplotlib                  3.7.2
mpl_toolkits                NA
natsort                     8.4.0
nbformat                    5.9.2
nbinom_ufunc                NA
ncf_ufunc                   NA
numba                       0.57.1
objc                        9.2
overrides                   NA
packaging                   23.1
pandas                      2.0.3
parso                       0.8.3
pexpect                     4.8.0
pickleshare                 0.7.5
pkg_resources               NA
platformdirs                3.10.0
prometheus_client           NA
prompt_toolkit              3.0.39
psutil                      5.9.5
ptyprocess                  0.7.0
pure_eval                   0.2.2
pydev_ipython               NA
pydevconsole                NA
pydevd                      2.9.5
pydevd_file_utils           NA
pydevd_plugins              NA
pydevd_tracing              NA
pygments                    2.15.1
pyparsing                   3.0.9
pythonjsonlogger            NA
pytz                        2023.3
referencing                 NA
requests                    2.31.0
rfc3339_validator           0.1.4
rfc3986_validator           0.1.1
rpds                        NA
send2trash                  NA
six                         1.16.0
sklearn                     1.3.0
sniffio                     1.3.0
socks                       1.7.1
stack_data                  0.6.2
threadpoolctl               3.2.0
tornado                     6.3.2
traitlets                   5.9.0
typing_extensions           NA
tzlocal                     NA
uri_template                NA
urllib3                     2.0.4
wcwidth                     0.2.6
webcolors                   1.13
websocket                   1.6.1
yaml                        6.0
zipp                        NA
zmq                         25.1.0
zoneinfo                    NA
</pre>
</details> <!-- seems like this ends pre, so might as well be explicit -->
<pre>
-----
IPython             8.14.0
jupyter_client      8.3.0
jupyter_core        5.3.1
jupyterlab          3.6.3
notebook            6.5.4
-----
Python 3.9.16 | packaged by conda-forge | (main, Feb  1 2023, 21:42:20) [Clang 14.0.6 ]
macOS-13.4.1-x86_64-i386-64bit
-----
Session information updated at 2023-11-15 15:48
</pre>
</details></div></div>
</div>
</section>
<section id="id2">
<h2><span class="section-number">5.9. </span>R<a class="headerlink" href="#id2" title="Link to this heading">#</a></h2>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="o">%%</span><span class="k">R</span>
sessioninfo::session_info()
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>─ Session info ───────────────────────────────────────────────────────────────
 setting  value
 version  R version 4.2.3 (2023-03-15)
 os       macOS Ventura 13.4.1
 system   x86_64, darwin13.4.0
 ui       unknown
 language (EN)
 collate  C
 ctype    UTF-8
 tz       Europe/Berlin
 date     2023-11-15
 pandoc   2.19.2 @ /Users/luke.zappia/miniconda3/envs/interoperability2/bin/pandoc

─ Packages ───────────────────────────────────────────────────────────────────
 package              * version    date (UTC) lib source
 abind                  1.4-5      2016-07-21 [1] CRAN (R 4.2.3)
 Biobase                2.58.0     2022-11-01 [1] Bioconductor
 BiocGenerics           0.44.0     2022-11-01 [1] Bioconductor
 bit                    4.0.5      2022-11-15 [1] CRAN (R 4.2.3)
 bit64                  4.0.5      2020-08-30 [1] CRAN (R 4.2.3)
 bitops                 1.0-7      2021-04-24 [1] CRAN (R 4.2.3)
 cli                    3.6.1      2023-03-23 [1] CRAN (R 4.2.3)
 cluster                2.1.4      2022-08-22 [1] CRAN (R 4.2.3)
 codetools              0.2-19     2023-02-01 [1] CRAN (R 4.2.3)
 colorspace             2.1-0      2023-01-23 [1] CRAN (R 4.2.3)
 cowplot                1.1.1      2020-12-30 [1] CRAN (R 4.2.3)
 data.table             1.14.8     2023-02-17 [1] CRAN (R 4.2.3)
 DelayedArray           0.24.0     2022-11-01 [1] Bioconductor
 deldir                 1.0-9      2023-05-17 [1] CRAN (R 4.2.3)
 digest                 0.6.33     2023-07-07 [1] CRAN (R 4.2.3)
 dplyr                  1.1.2      2023-04-20 [1] CRAN (R 4.2.3)
 ellipsis               0.3.2      2021-04-29 [1] CRAN (R 4.2.3)
 fansi                  1.0.4      2023-01-22 [1] CRAN (R 4.2.3)
 fastmap                1.1.1      2023-02-24 [1] CRAN (R 4.2.3)
 fitdistrplus           1.1-11     2023-04-25 [1] CRAN (R 4.2.3)
 future                 1.33.0     2023-07-01 [1] CRAN (R 4.2.3)
 future.apply           1.11.0     2023-05-21 [1] CRAN (R 4.2.3)
 generics               0.1.3      2022-07-05 [1] CRAN (R 4.2.3)
 GenomeInfoDb           1.34.9     2023-02-02 [1] Bioconductor
 GenomeInfoDbData       1.2.9      2023-11-10 [1] Bioconductor
 GenomicRanges          1.50.0     2022-11-01 [1] Bioconductor
 ggplot2                3.4.2      2023-04-03 [1] CRAN (R 4.2.3)
 ggrepel                0.9.3      2023-02-03 [1] CRAN (R 4.2.3)
 ggridges               0.5.4      2022-09-26 [1] CRAN (R 4.2.3)
 globals                0.16.2     2022-11-21 [1] CRAN (R 4.2.3)
 glue                   1.6.2      2022-02-24 [1] CRAN (R 4.2.3)
 goftest                1.2-3      2021-10-07 [1] CRAN (R 4.2.3)
 gridExtra              2.3        2017-09-09 [1] CRAN (R 4.2.3)
 gtable                 0.3.3      2023-03-21 [1] CRAN (R 4.2.3)
 hdf5r                  1.3.8      2023-01-21 [1] CRAN (R 4.2.3)
 htmltools              0.5.5      2023-03-23 [1] CRAN (R 4.2.3)
 htmlwidgets            1.6.2      2023-03-17 [1] CRAN (R 4.2.3)
 httpuv                 1.6.11     2023-05-11 [1] CRAN (R 4.2.3)
 httr                   1.4.6      2023-05-08 [1] CRAN (R 4.2.3)
 ica                    1.0-3      2022-07-08 [1] CRAN (R 4.2.3)
 igraph                 1.5.0.1    2023-07-23 [1] CRAN (R 4.2.3)
 IRanges                2.32.0     2022-11-01 [1] Bioconductor
 irlba                  2.3.5.1    2022-10-03 [1] CRAN (R 4.2.3)
 jsonlite               1.8.7      2023-06-29 [1] CRAN (R 4.2.3)
 KernSmooth             2.23-22    2023-07-10 [1] CRAN (R 4.2.3)
 later                  1.3.1      2023-05-02 [1] CRAN (R 4.2.3)
 lattice                0.21-8     2023-04-05 [1] CRAN (R 4.2.3)
 lazyeval               0.2.2      2019-03-15 [1] CRAN (R 4.2.3)
 leiden                 0.4.3      2022-09-10 [1] CRAN (R 4.2.3)
 lifecycle              1.0.3      2022-10-07 [1] CRAN (R 4.2.3)
 listenv                0.9.0      2022-12-16 [1] CRAN (R 4.2.3)
 lmtest                 0.9-40     2022-03-21 [1] CRAN (R 4.2.3)
 magrittr               2.0.3      2022-03-30 [1] CRAN (R 4.2.3)
 MASS                   7.3-60     2023-05-04 [1] CRAN (R 4.2.3)
 Matrix                 1.6-0      2023-07-08 [1] CRAN (R 4.2.3)
 MatrixGenerics         1.10.0     2022-11-01 [1] Bioconductor
 matrixStats            1.0.0      2023-06-02 [1] CRAN (R 4.2.3)
 mime                   0.12       2021-09-28 [1] CRAN (R 4.2.3)
 miniUI                 0.1.1.1    2018-05-18 [1] CRAN (R 4.2.3)
 MuData                 1.2.0      2022-11-01 [1] Bioconductor
 MuDataSeurat           0.0.0.9000 2023-11-15 [1] Github (PMBio/MuDataSeurat@e34e908)
 MultiAssayExperiment   1.24.0     2022-11-01 [1] Bioconductor
 munsell                0.5.0      2018-06-12 [1] CRAN (R 4.2.3)
 nlme                   3.1-162    2023-01-31 [1] CRAN (R 4.2.3)
 parallelly             1.36.0     2023-05-26 [1] CRAN (R 4.2.3)
 patchwork              1.1.2      2022-08-19 [1] CRAN (R 4.2.3)
 pbapply                1.7-2      2023-06-27 [1] CRAN (R 4.2.3)
 pillar                 1.9.0      2023-03-22 [1] CRAN (R 4.2.3)
 pkgconfig              2.0.3      2019-09-22 [1] CRAN (R 4.2.3)
 plotly                 4.10.2     2023-06-03 [1] CRAN (R 4.2.3)
 plyr                   1.8.8      2022-11-11 [1] CRAN (R 4.2.3)
 png                    0.1-8      2022-11-29 [1] CRAN (R 4.2.3)
 polyclip               1.10-4     2022-10-20 [1] CRAN (R 4.2.3)
 progressr              0.13.0     2023-01-10 [1] CRAN (R 4.2.3)
 promises               1.2.0.1    2021-02-11 [1] CRAN (R 4.2.3)
 purrr                  1.0.1      2023-01-10 [1] CRAN (R 4.2.3)
 R6                     2.5.1      2021-08-19 [1] CRAN (R 4.2.3)
 RANN                   2.6.1      2019-01-08 [1] CRAN (R 4.2.3)
 RColorBrewer           1.1-3      2022-04-03 [1] CRAN (R 4.2.3)
 Rcpp                   1.0.11     2023-07-06 [1] CRAN (R 4.2.3)
 RcppAnnoy              0.0.21     2023-07-02 [1] CRAN (R 4.2.3)
 RCurl                  1.98-1.12  2023-03-27 [1] CRAN (R 4.2.3)
 reshape2               1.4.4      2020-04-09 [1] CRAN (R 4.2.3)
 reticulate             1.30       2023-06-09 [1] CRAN (R 4.2.3)
 rhdf5                  2.42.0     2022-11-01 [1] Bioconductor
 rhdf5filters           1.10.0     2022-11-01 [1] Bioconductor
 Rhdf5lib               1.20.0     2022-11-01 [1] Bioconductor
 rlang                  1.1.1      2023-04-28 [1] CRAN (R 4.2.3)
 ROCR                   1.0-11     2020-05-02 [1] CRAN (R 4.2.3)
 Rtsne                  0.16       2022-04-17 [1] CRAN (R 4.2.3)
 S4Vectors              0.36.0     2022-11-01 [1] Bioconductor
 scales                 1.2.1      2022-08-20 [1] CRAN (R 4.2.3)
 scattermore            1.2        2023-06-12 [1] CRAN (R 4.2.3)
 sctransform            0.3.5      2022-09-21 [1] CRAN (R 4.2.3)
 sessioninfo            1.2.2      2021-12-06 [1] CRAN (R 4.2.3)
 Seurat                 4.3.0.1    2023-06-22 [1] CRAN (R 4.2.3)
 SeuratObject           4.1.3      2022-11-07 [1] CRAN (R 4.2.3)
 shiny                  1.7.4.1    2023-07-06 [1] CRAN (R 4.2.3)
 SingleCellExperiment   1.20.0     2022-11-01 [1] Bioconductor
 sp                     2.0-0      2023-06-22 [1] CRAN (R 4.2.3)
 spatstat.data          3.0-1      2023-03-12 [1] CRAN (R 4.2.3)
 spatstat.explore       3.2-1      2023-05-13 [1] CRAN (R 4.2.3)
 spatstat.geom          3.2-4      2023-07-20 [1] CRAN (R 4.2.3)
 spatstat.random        3.1-5      2023-05-11 [1] CRAN (R 4.2.3)
 spatstat.sparse        3.0-2      2023-06-25 [1] CRAN (R 4.2.3)
 spatstat.utils         3.0-3      2023-05-09 [1] CRAN (R 4.2.3)
 stringi                1.7.12     2023-01-11 [1] CRAN (R 4.2.3)
 stringr                1.5.0      2022-12-02 [1] CRAN (R 4.2.3)
 SummarizedExperiment   1.28.0     2022-11-01 [1] Bioconductor
 survival               3.5-5      2023-03-12 [1] CRAN (R 4.2.3)
 tensor                 1.5        2012-05-05 [1] CRAN (R 4.2.3)
 tibble                 3.2.1      2023-03-20 [1] CRAN (R 4.2.3)
 tidyr                  1.3.0      2023-01-24 [1] CRAN (R 4.2.3)
 tidyselect             1.2.0      2022-10-10 [1] CRAN (R 4.2.3)
 utf8                   1.2.3      2023-01-31 [1] CRAN (R 4.2.3)
 uwot                   0.1.16     2023-06-29 [1] CRAN (R 4.2.3)
 vctrs                  0.6.3      2023-06-14 [1] CRAN (R 4.2.3)
 viridisLite            0.4.2      2023-05-02 [1] CRAN (R 4.2.3)
 xtable                 1.8-4      2019-04-21 [1] CRAN (R 4.2.3)
 XVector                0.38.0     2022-11-01 [1] Bioconductor
 zlibbioc               1.44.0     2022-11-01 [1] Bioconductor
 zoo                    1.8-12     2023-04-13 [1] CRAN (R 4.2.3)

 [1] /Users/luke.zappia/miniconda3/envs/interoperability2/lib/R/library

──────────────────────────────────────────────────────────────────────────────
</pre></div>
</div>
</div>
</div>
</section>
<section id="references">
<h2><span class="section-number">5.10. </span>References<a class="headerlink" href="#references" title="Link to this heading">#</a></h2>
<span class="target" id="id3"></span></section>
<section id="contributors">
<h2><span class="section-number">5.11. </span>Contributors<a class="headerlink" href="#contributors" title="Link to this heading">#</a></h2>
<p>We gratefully acknowledge the contributions of:</p>
<section id="authors">
<h3><span class="section-number">5.11.1. </span>Authors<a class="headerlink" href="#authors" title="Link to this heading">#</a></h3>
<ul class="simple">
<li><p>Luke Zappia</p></li>
</ul>
</section>
<section id="reviewers">
<h3><span class="section-number">5.11.2. </span>Reviewers<a class="headerlink" href="#reviewers" title="Link to this heading">#</a></h3>
<ul class="simple">
<li><p>Lukas Heumos</p></li>
<li><p>Isaac Virshup</p></li>
<li><p>Anastasia Litinetskaya</p></li>
<li><p>Ludwig Geistlinger</p></li>
<li><p>Peter Hickey</p></li>
</ul>
</section>
</section>
</section>

    <script type="text/x-thebe-config">
    {
        requestKernel: true,
        binderOptions: {
            repo: "binder-examples/jupyter-stacks-datascience",
            ref: "master",
        },
        codeMirrorConfig: {
            theme: "abcdef",
            mode: "python"
        },
        kernelOptions: {
            name: "python3",
            path: "./introduction"
        },
        predefinedOutput: true
    }
    </script>
    <script>kernelName = 'python3'</script>

                </article>
              

              
              
              
              
                <footer class="prev-next-footer d-print-none">
                  
<div class="prev-next-area">
    <a class="left-prev"
       href="analysis_tools.html"
       title="previous page">
      <i class="fa-solid fa-angle-left"></i>
      <div class="prev-next-info">
        <p class="prev-next-subtitle">previous</p>
        <p class="prev-next-title"><span class="section-number">4. </span>Analysis frameworks and tools</p>
      </div>
    </a>
    <a class="right-next"
       href="../preprocessing_visualization/quality_control.html"
       title="next page">
      <div class="prev-next-info">
        <p class="prev-next-subtitle">next</p>
        <p class="prev-next-title"><span class="section-number">6. </span>Quality Control</p>
      </div>
      <i class="fa-solid fa-angle-right"></i>
    </a>
</div>
                </footer>
              
            </div>
            
            
              
                <div class="bd-sidebar-secondary bd-toc"><div class="sidebar-secondary-items sidebar-secondary__inner">


  <div class="sidebar-secondary-item">
  <div class="page-toc tocsection onthispage">
    <i class="fa-solid fa-list"></i> Contents
  </div>
  <nav class="bd-toc-nav page-toc">
    <ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#motivation">5.1. Motivation</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#nomenclature">5.2. Nomenclature</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#disk-based-interoperability">5.3. Disk-based interoperability</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#simple-formats">5.3.1. Simple formats</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#hdf5-based-formats">5.3.2. HDF5-based formats</a><ul class="nav section-nav flex-column">
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#h5ad">5.3.2.1. H5AD</a><ul class="nav section-nav flex-column">
<li class="toc-h5 nav-item toc-entry"><a class="reference internal nav-link" href="#reading-writing-h5ad-with-bioconductor">5.3.2.1.1. Reading/writing H5AD with Bioconductor</a></li>
<li class="toc-h5 nav-item toc-entry"><a class="reference internal nav-link" href="#reading-writing-h5ad-with-seurat">5.3.2.1.2. Reading/writing H5AD with <strong>{Seurat}</strong></a></li>
<li class="toc-h5 nav-item toc-entry"><a class="reference internal nav-link" href="#reading-writing-h5ad-with-anndata">5.3.2.1.3. Reading/writing H5AD with <strong>{anndata}</strong></a></li>
</ul>
</li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#loom">5.3.2.2. Loom</a></li>
</ul>
</li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#rds-files">5.3.3. RDS files</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#new-on-disk-formats">5.3.4. New on-disk formats</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#in-memory-interoperability">5.4. In-memory interoperability</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#interoperability-between-r-ecosystems">5.4.1. Interoperability between R ecosystems</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#accessing-r-from-python">5.4.2. Accessing R from Python</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#accessing-python-from-r">5.4.3. Accessing Python from R</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#interoperability-for-multimodal-data">5.5. Interoperability for multimodal data</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#python">5.5.1. Python</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#r">5.5.2. R</a><ul class="nav section-nav flex-column">
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#bioconductor">5.5.2.1. Bioconductor</a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#seurat">5.5.2.2. Seurat</a></li>
</ul>
</li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#interoperability-with-other-languages">5.6. Interoperability with other languages</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#julia">5.6.1. Julia</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#javascript">5.6.2. JavaScript</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#rust">5.6.3. Rust</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#session-information">5.7. Session information</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#id1">5.8. Python</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#id2">5.9. R</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#references">5.10. References</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#contributors">5.11. Contributors</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#authors">5.11.1. Authors</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#reviewers">5.11.2. Reviewers</a></li>
</ul>
</li>
</ul>
  </nav></div>

</div></div>
              
            
          </div>
          <footer class="bd-footer-content">
            
<div class="bd-footer-content__inner container">
  
  <div class="footer-item">
    
<p class="component-author">
By Lukas Heumos, Anna Schaar, single-cell best practices consortium
</p>

  </div>
  
  <div class="footer-item">
    

  <p class="copyright">
    
      © Copyright 2023.
      <br/>
    
  </p>

  </div>
  
  <div class="footer-item">
    
  </div>
  
  <div class="footer-item">
    
<div class="extra_footer">
  <div>
  Brought to you by
  <a href="https://www.github.com/theislab">Theislab</a>, with many thanks to the single-cell community as a whole!
</div>

</div>
  </div>
  
</div>
          </footer>
        

      </main>
    </div>
  </div>
  
  <!-- Scripts loaded after <body> so the DOM is not blocked -->
  <script src="../_static/scripts/bootstrap.js?digest=dfe6caa3a7d634c4db9b"></script>
<script src="../_static/scripts/pydata-sphinx-theme.js?digest=dfe6caa3a7d634c4db9b"></script>

  <footer class="bd-footer">
  </footer>
  </body>
</html>