
<!DOCTYPE html>

<html lang="en">
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="generator" content="Docutils 0.17.1: http://docutils.sourceforge.net/" />

    <title>2. Single-cell RNA sequencing &#8212; Single-cell best practices</title>
    
  <!-- Loaded before other Sphinx assets -->
  <link href="../_static/styles/theme.css?digest=1999514e3f237ded88cf" rel="stylesheet">
<link href="../_static/styles/pydata-sphinx-theme.css?digest=1999514e3f237ded88cf" rel="stylesheet">

    
  <link rel="stylesheet"
    href="../_static/vendor/fontawesome/5.13.0/css/all.min.css">
  <link rel="preload" as="font" type="font/woff2" crossorigin
    href="../_static/vendor/fontawesome/5.13.0/webfonts/fa-solid-900.woff2">
  <link rel="preload" as="font" type="font/woff2" crossorigin
    href="../_static/vendor/fontawesome/5.13.0/webfonts/fa-brands-400.woff2">

    <link rel="stylesheet" type="text/css" href="../_static/pygments.css" />
    <link rel="stylesheet" href="../_static/styles/sphinx-book-theme.css?digest=5115cc725059bd94278eecd172e13a965bf8f5a9" type="text/css" />
    <link rel="stylesheet" type="text/css" href="../_static/togglebutton.css" />
    <link rel="stylesheet" type="text/css" href="../_static/copybutton.css" />
    <link rel="stylesheet" type="text/css" href="../_static/mystnb.css" />
    <link rel="stylesheet" type="text/css" href="../_static/sphinx-thebe.css" />
    <link rel="stylesheet" type="text/css" href="../_static/book.css" />
    <link rel="stylesheet" type="text/css" href="../_static/design-style.b7bb847fb20b106c3d81b95245e65545.min.css" />
    
  <!-- Pre-loaded scripts that we'll load fully later -->
  <link rel="preload" as="script" href="../_static/scripts/pydata-sphinx-theme.js?digest=1999514e3f237ded88cf">

    <script data-url_root="../" id="documentation_options" src="../_static/documentation_options.js"></script>
    <script src="../_static/jquery.js"></script>
    <script src="../_static/underscore.js"></script>
    <script src="../_static/doctools.js"></script>
    <script src="../_static/clipboard.min.js"></script>
    <script src="../_static/copybutton.js"></script>
    <script src="../_static/scripts/sphinx-book-theme.js?digest=9c920249402e914e316237a7dbc6769907cce411"></script>
    <script>let toggleHintShow = 'Click to show';</script>
    <script>let toggleHintHide = 'Click to hide';</script>
    <script>let toggleOpenOnPrint = 'true';</script>
    <script src="../_static/togglebutton.js"></script>
    <script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
    <script src="../_static/book.js"></script>
    <script>var togglebuttonSelector = '.toggle, .admonition.dropdown, .tag_hide_input div.cell_input, .tag_hide-input div.cell_input, .tag_hide_output div.cell_output, .tag_hide-output div.cell_output, .tag_hide_cell.cell, .tag_hide-cell.cell';</script>
    <script src="../_static/design-tabs.js"></script>
    <script>const THEBE_JS_URL = "https://unpkg.com/thebe@0.8.2/lib/index.js"
const thebe_selector = ".thebe,.cell"
const thebe_selector_input = "pre"
const thebe_selector_output = ".output, .cell_output"
</script>
    <script async="async" src="../_static/sphinx-thebe.js"></script>
    <link rel="shortcut icon" href="../_static/favicon.ico"/>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="3. Raw data processing" href="raw_data_processing.html" />
    <link rel="prev" title="1. Prior art" href="prior_art.html" />
    <meta name="viewport" content="width=device-width, initial-scale=1" />
    <meta name="docsearch:language" content="en">
    

    <!-- Google Analytics -->
    
  </head>
  <body data-spy="scroll" data-target="#bd-toc-nav" data-offset="60">
<!-- Checkboxes to toggle the left sidebar -->
<input type="checkbox" class="sidebar-toggle" name="__navigation" id="__navigation" aria-label="Toggle navigation sidebar">
<label class="overlay overlay-navbar" for="__navigation">
    <div class="visually-hidden">Toggle navigation sidebar</div>
</label>
<!-- Checkboxes to toggle the in-page toc -->
<input type="checkbox" class="sidebar-toggle" name="__page-toc" id="__page-toc" aria-label="Toggle in-page Table of Contents">
<label class="overlay overlay-pagetoc" for="__page-toc">
    <div class="visually-hidden">Toggle in-page Table of Contents</div>
</label>
<!-- Headers at the top -->
<div class="announcement header-item noprint"></div>
<div class="header header-item noprint"></div>

    
    <div class="container-fluid" id="banner"></div>

    

    <div class="container-xl">
      <div class="row">
          
<!-- Sidebar -->
<div class="bd-sidebar noprint" id="site-navigation">
    <div class="bd-sidebar__content">
        <div class="bd-sidebar__top"><div class="navbar-brand-box">
    <a class="navbar-brand text-wrap" href="../index.html">
      
        <!-- `logo` is deprecated in Sphinx 4.0, so remove this when we stop supporting 3 -->
        
      
      
      <img src="../_static/logo.png" class="logo" alt="logo">
      
      
      <h1 class="site-logo" id="site-title">Single-cell best practices</h1>
      
    </a>
</div><form class="bd-search d-flex align-items-center" action="../search.html" method="get">
  <i class="icon fas fa-search"></i>
  <input type="search" class="form-control" name="q" id="search-input" placeholder="Search this book..." aria-label="Search this book..." autocomplete="off" >
</form><nav class="bd-links" id="bd-docs-nav" aria-label="Main">
    <div class="bd-toc-item active">
        <p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Introduction
 </span>
</p>
<ul class="current nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="prior_art.html">
   1. Prior art
  </a>
 </li>
 <li class="toctree-l1 current active">
  <a class="current reference internal" href="#">
   2. Single-cell RNA sequencing
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="raw_data_processing.html">
   3. Raw data processing
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="analysis_tools.html">
   4. Analysis frameworks and tools
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="interoperability.html">
   5. Interoperability
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Preprocessing and visualization
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../preprocessing_visualization/quality_control.html">
   6. Quality Control
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../preprocessing_visualization/normalization.html">
   7. Normalization
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../preprocessing_visualization/feature_selection.html">
   8. Feature selection
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../preprocessing_visualization/dimensionality_reduction.html">
   9. Dimensionality Reduction
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Identifying cellular structure
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../cellular_structure/clustering.html">
   10. Clustering
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../cellular_structure/annotation.html">
   11. Annotation
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../cellular_structure/integration.html">
   12. Data integration
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Inferring trajectories
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../trajectories/pseudotemporal.html">
   13. Pseudotemporal ordering
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../trajectories/rna_velocity.html">
   14. RNA velocity
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../trajectories/lineage_tracing.html">
   15. Lineage tracing
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Dealing with conditions
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../conditions/differential_gene_expression.html">
   16. Differential gene expression analysis
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../conditions/compositional.html">
   17. Compositional analysis
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../conditions/gsea_pathway.html">
   18. Gene set enrichment and pathway analysis
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../conditions/perturbation_modeling.html">
   19. Perturbation modeling
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Modeling mechanisms
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../mechanisms/gene_regulatory_networks.html">
   20. Gene regulatory networks
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../mechanisms/cell_cell_communication.html">
   21. Cell-cell communication
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Deconvolution
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../deconvolution/bulk_deconvolution.html">
   22. Bulk deconvolution
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Chromatin Accessibility
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../chromatin_accessibility/introduction.html">
   23. Single-cell ATAC sequencing
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../chromatin_accessibility/quality_control.html">
   24. Quality Control
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../chromatin_accessibility/gene_regulatory_networks_atac.html">
   25. Gene regulatory networks using chromatin accessibility
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Spatial omics
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../spatial/introduction.html">
   26. Single-cell data resolved in space
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../spatial/neighborhood.html">
   27. Neighborhood analysis
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../spatial/domains.html">
   28. Spatial domains
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../spatial/spatially_variable_genes.html">
   29. Spatially variable genes
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../spatial/deconvolution.html">
   30. Spatial deconvolution
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../spatial/imputation.html">
   31. Imputation
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Surface protein
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../surface_protein/quality_control.html">
   35. Quality control
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../surface_protein/normalization.html">
   36. Normalization
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../surface_protein/doublet_detection.html">
   37. Doublet detection
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../surface_protein/dimensionality_reduction.html">
   38. Dimensionality Reduction
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../surface_protein/batch_correction.html">
   39. Batch correction
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../surface_protein/annotation.html">
   40. Annotation
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Adaptive immune receptor repertoire
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../air_repertoire/ir_profiling.html">
   41. Immune Receptor Profiling
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../air_repertoire/clonotype.html">
   42. Clonotype analysis
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../air_repertoire/specificity.html">
   46. Specificity analysis
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../air_repertoire/multimodal_integration.html">
   47. Integrating AIR and transcriptomics
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Multimodal integration
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../multimodal_integration/paired_integration.html">
   48. Paired integration
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../multimodal_integration/advanced_integration.html">
   49. Advanced integration
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Reproducibility
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../reproducibility/introduction.html">
   50. Reproducibility
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Outlook
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../outlook.html">
   51. Outlook
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Acknowledgements
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../acknowledgements.html">
   52. Acknowledgements
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Glossary
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../glossary.html">
   53. Glossary
  </a>
 </li>
</ul>

    </div>
</nav></div>
        <div class="bd-sidebar__bottom">
             <!-- To handle the deprecated key -->
            
            <div class="navbar_extra_footer">
            Powered by <a href="https://jupyterbook.org">Jupyter Book</a>
            </div>
            
        </div>
    </div>
    <div id="rtd-footer-container"></div>
</div>


          


          
<!-- A tiny helper pixel to detect if we've scrolled -->
<div class="sbt-scroll-pixel-helper"></div>
<!-- Main content -->
<div class="col py-0 content-container">
    
    <div class="header-article row sticky-top noprint">
        



<div class="col py-1 d-flex header-article-main">
    <div class="header-article__left">
        
        <label for="__navigation"
  class="headerbtn"
  data-toggle="tooltip"
data-placement="right"
title="Toggle navigation"
>
  

<span class="headerbtn__icon-container">
  <i class="fas fa-bars"></i>
  </span>

</label>

        
    </div>
    <div class="header-article__right">
<button onclick="toggleFullScreen()"
  class="headerbtn"
  data-toggle="tooltip"
data-placement="bottom"
title="Fullscreen mode"
>
  

<span class="headerbtn__icon-container">
  <i class="fas fa-expand"></i>
  </span>

</button>

<div class="menu-dropdown menu-dropdown-repository-buttons">
  <button class="headerbtn menu-dropdown__trigger"
      aria-label="Source repositories">
      <i class="fab fa-github"></i>
  </button>
  <div class="menu-dropdown__content">
    <ul>
      <li>
        <a href="https://github.com/theislab/single-cell-best-practices"
   class="headerbtn"
   data-toggle="tooltip"
data-placement="left"
title="Source repository"
>
  

<span class="headerbtn__icon-container">
  <i class="fab fa-github"></i>
  </span>
<span class="headerbtn__text-container">repository</span>
</a>

      </li>
      
      <li>
        <a href="https://github.com/theislab/single-cell-best-practices/issues/new?title=Issue%20on%20page%20%2Fintroduction/scrna_seq.html&body=Your%20issue%20content%20here."
   class="headerbtn"
   data-toggle="tooltip"
data-placement="left"
title="Open an issue"
>
  

<span class="headerbtn__icon-container">
  <i class="fas fa-lightbulb"></i>
  </span>
<span class="headerbtn__text-container">open issue</span>
</a>

      </li>
      
      <li>
        <a href="https://github.com/theislab/single-cell-best-practices/edit/master/jupyter-book/introduction/scrna_seq.md"
   class="headerbtn"
   data-toggle="tooltip"
data-placement="left"
title="Edit this page"
>
  

<span class="headerbtn__icon-container">
  <i class="fas fa-pencil-alt"></i>
  </span>
<span class="headerbtn__text-container">suggest edit</span>
</a>

      </li>
      
    </ul>
  </div>
</div>

<div class="menu-dropdown menu-dropdown-download-buttons">
  <button class="headerbtn menu-dropdown__trigger"
      aria-label="Download this page">
      <i class="fas fa-download"></i>
  </button>
  <div class="menu-dropdown__content">
    <ul>
      <li>
        <a href="../_sources/introduction/scrna_seq.md"
   class="headerbtn"
   data-toggle="tooltip"
data-placement="left"
title="Download source file"
>
  

<span class="headerbtn__icon-container">
  <i class="fas fa-file"></i>
  </span>
<span class="headerbtn__text-container">.md</span>
</a>

      </li>
      
      <li>
        
<button onclick="printPdf(this)"
  class="headerbtn"
  data-toggle="tooltip"
data-placement="left"
title="Print to PDF"
>
  

<span class="headerbtn__icon-container">
  <i class="fas fa-file-pdf"></i>
  </span>
<span class="headerbtn__text-container">.pdf</span>
</button>

      </li>
      
    </ul>
  </div>
</div>
<label for="__page-toc"
  class="headerbtn headerbtn-page-toc"
  
>
  

<span class="headerbtn__icon-container">
  <i class="fas fa-list"></i>
  </span>

</label>

    </div>
</div>

<!-- Table of contents -->
<div class="col-md-3 bd-toc show noprint">
    <div class="tocsection onthispage pt-5 pb-3">
        <i class="fas fa-list"></i> Contents
    </div>
    <nav id="bd-toc-nav" aria-label="Page">
        <ul class="visible nav section-nav flex-column">
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#the-building-block-of-life">
   2.1. The building block of life
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#a-brief-history-of-sequencing">
   2.2. A brief history of sequencing
  </a>
  <ul class="nav section-nav flex-column">
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#first-generation-sequencing">
     2.2.1. First generation sequencing
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#second-generation-sequencing">
     2.2.2. Second generation sequencing
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#third-generation-sequencing">
     2.2.3. Third generation sequencing
    </a>
   </li>
  </ul>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#overview-of-the-ngs-process">
   2.3. Overview of the NGS process
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#rna-sequencing">
   2.4. RNA sequencing
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#id4">
   2.5. Single-cell RNA sequencing
  </a>
  <ul class="nav section-nav flex-column">
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#overview">
     2.5.1. Overview
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#transcript-quantification">
     2.5.2. Transcript quantification
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#single-cell-sequencing-protocols">
     2.5.3. Single-cell sequencing protocols
    </a>
    <ul class="nav section-nav flex-column">
     <li class="toc-h4 nav-item toc-entry">
      <a class="reference internal nav-link" href="#microfluidic-device-based-protocols">
       2.5.3.1. Microfluidic device based protocols
      </a>
     </li>
     <li class="toc-h4 nav-item toc-entry">
      <a class="reference internal nav-link" href="#plate-based">
       2.5.3.2. Plate based
      </a>
     </li>
     <li class="toc-h4 nav-item toc-entry">
      <a class="reference internal nav-link" href="#fluidigm-c1">
       2.5.3.3. Fluidigm C1
      </a>
     </li>
     <li class="toc-h4 nav-item toc-entry">
      <a class="reference internal nav-link" href="#nanopore-single-cell-transcriptome-sequencing">
       2.5.3.4. Nanopore single-cell transcriptome sequencing
      </a>
     </li>
     <li class="toc-h4 nav-item toc-entry">
      <a class="reference internal nav-link" href="#summary">
       2.5.3.5. Summary
      </a>
     </li>
    </ul>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#single-cell-vs-single-nuclei">
     2.5.4. single-cell vs single-nuclei
    </a>
   </li>
  </ul>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#recommended-reading">
   2.6. Recommended reading
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#references">
   2.7. References
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#contributors">
   2.8. Contributors
  </a>
  <ul class="nav section-nav flex-column">
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#authors">
     2.8.1. Authors
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#reviewers">
     2.8.2. Reviewers
    </a>
   </li>
  </ul>
 </li>
</ul>

    </nav>
</div>
    </div>
    <div class="article row">
        <div class="col pl-md-3 pl-lg-5 content-container">
            <!-- Table of contents that is only displayed when printing the page -->
            <div id="jb-print-docs-body" class="onlyprint">
                <h1>Single-cell RNA sequencing</h1>
                <!-- Table of contents -->
                <div id="print-main-content">
                    <div id="jb-print-toc">
                        
                        <div>
                            <h2> Contents </h2>
                        </div>
                        <nav aria-label="Page">
                            <ul class="visible nav section-nav flex-column">
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#the-building-block-of-life">
   2.1. The building block of life
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#a-brief-history-of-sequencing">
   2.2. A brief history of sequencing
  </a>
  <ul class="nav section-nav flex-column">
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#first-generation-sequencing">
     2.2.1. First generation sequencing
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#second-generation-sequencing">
     2.2.2. Second generation sequencing
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#third-generation-sequencing">
     2.2.3. Third generation sequencing
    </a>
   </li>
  </ul>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#overview-of-the-ngs-process">
   2.3. Overview of the NGS process
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#rna-sequencing">
   2.4. RNA sequencing
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#id4">
   2.5. Single-cell RNA sequencing
  </a>
  <ul class="nav section-nav flex-column">
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#overview">
     2.5.1. Overview
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#transcript-quantification">
     2.5.2. Transcript quantification
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#single-cell-sequencing-protocols">
     2.5.3. Single-cell sequencing protocols
    </a>
    <ul class="nav section-nav flex-column">
     <li class="toc-h4 nav-item toc-entry">
      <a class="reference internal nav-link" href="#microfluidic-device-based-protocols">
       2.5.3.1. Microfluidic device based protocols
      </a>
     </li>
     <li class="toc-h4 nav-item toc-entry">
      <a class="reference internal nav-link" href="#plate-based">
       2.5.3.2. Plate based
      </a>
     </li>
     <li class="toc-h4 nav-item toc-entry">
      <a class="reference internal nav-link" href="#fluidigm-c1">
       2.5.3.3. Fluidigm C1
      </a>
     </li>
     <li class="toc-h4 nav-item toc-entry">
      <a class="reference internal nav-link" href="#nanopore-single-cell-transcriptome-sequencing">
       2.5.3.4. Nanopore single-cell transcriptome sequencing
      </a>
     </li>
     <li class="toc-h4 nav-item toc-entry">
      <a class="reference internal nav-link" href="#summary">
       2.5.3.5. Summary
      </a>
     </li>
    </ul>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#single-cell-vs-single-nuclei">
     2.5.4. single-cell vs single-nuclei
    </a>
   </li>
  </ul>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#recommended-reading">
   2.6. Recommended reading
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#references">
   2.7. References
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#contributors">
   2.8. Contributors
  </a>
  <ul class="nav section-nav flex-column">
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#authors">
     2.8.1. Authors
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#reviewers">
     2.8.2. Reviewers
    </a>
   </li>
  </ul>
 </li>
</ul>

                        </nav>
                    </div>
                </div>
            </div>
            <main id="main-content" role="main">
                
              <div>
                
  <section class="tex2jax_ignore mathjax_ignore" id="single-cell-rna-sequencing">
<h1><span class="section-number">2. </span>Single-cell RNA sequencing<a class="headerlink" href="#single-cell-rna-sequencing" title="Permalink to this headline">#</a></h1>
<p>This chapter provides a short introduction to the most widely used single-cell ribonucleic acid (<a class="reference internal" href="../glossary.html#term-RNA"><span class="xref std std-term">RNA</span></a>) sequencing assays and associated basic molecular biology concepts. Multimodal or spatial assays are not covered here, but are introduced in the respective advanced chapters. All sequencing assays have individual strengths and limitations which must be known by data analysis to be aware of possible biases in the data.</p>
<section id="the-building-block-of-life">
<h2><span class="section-number">2.1. </span>The building block of life<a class="headerlink" href="#the-building-block-of-life" title="Permalink to this headline">#</a></h2>
<p>Life, as we know it, is the characteristic that distinguishes living from dead or inanimate entities. Most definitions of the term life share a common entity - cells.
Cells form open systems which maintain homeostasis, have a metabolism, grow, adapt to their environment, reproduce, respond to stimuli, and organize themselves.
Therefore, cells are the fundamental building block of life which were first discovered in 1665 by the British scientist Robert Hooke. Hooke investigated a thin slice of cork with a very rudimentary microscope, and to his surprise noticed that the slice appeared to resemble a honeycomb. He named these tiny units ‘cells’.</p>
<figure class="align-default" id="markdown-fig">
<a class="bg-primary mb-1 reference internal image-reference" href="../_images/hooke_cork.jpg"><img alt="Robert Hook cell" class="bg-primary mb-1" src="../_images/hooke_cork.jpg" style="width: 800px;" /></a>
<figcaption>
<p><span class="caption-number">Fig. 2.1 </span><span class="caption-text">Robert Hooke’s drawing of cork cells. Image obtained from Micrographia.</span><a class="headerlink" href="#markdown-fig" title="Permalink to this image">#</a></p>
</figcaption>
</figure>
<p>In 1839, Matthias Jakob Schleiden and Theodor Schwann first described Cell Theory. It describes that all living organisms are made up of cells. Cells act as functional units that by themselves originate from other cells, making them the basic units of reproduction.</p>
<p>Since the early definition of cell theory, researchers discovered that there exists an energy flow within cells, that heredity information is passed from one cell to another in the form of <a class="reference internal" href="../glossary.html#term-DNA"><span class="xref std std-term">DNA</span></a> and that all cells have almost the same chemical composition. Two general types of cells exist, eukaryotes and prokaryotes.
Eukaryotic cells contain a nucleus, where the nuclear membrane encapsulates the chromosomes; while prokaryotic cells only have a nucleoid region, but no nucleus.
The nucleus hosts the cells’ genomic deoxyribonucleic acid <a class="reference internal" href="../glossary.html#term-DNA"><span class="xref std std-term">DNA</span></a> and is the reason for the eukaryotes’ name: <em>Nucleus</em> is Latin for kernel or seed. Eukaryotes are organisms composed of a single cell (unicellular) or multiple cells (multicellular), whereas prokaryotes are single-celled organisms. Eukaryotic cells are further distinguished from prokaryotic cells by their high degree of compartmentalization, i.e. membrane-bound organelles are carrying out highly specialized functions and providing crucial support for cells.</p>
<p>Compared to prokaryotic cells, eukaryotic cells have on average about 10,000 the volume with a rich mix of organelles and a cytoskeleton constituted of microtubules, microfilaments, and intermediate filaments.
The DNA replication machinery reads the hereditary information that is stored in the DNA in the nucleus to replicate themselves and keep the life cycle going. The eukaryotic DNA is divided into several linear bundles called chromosomes, which are separated by the microtubular spindle during nuclear division.
Understanding the hereditary information hidden in DNA is key to understanding many evolutionary and disease-related processes.
<em>sequencing</em> is the process of deciphering the order of DNA nucleotides and is primarily used to unveil the genetic information that is carried by a specific DNA segment, a complete genome, or even a complex microbiome.
DNA <em>sequencing</em> allows researchers to identify the location and function of genes and regulatory elements in the DNA molecule and the genome, and uncovers genetic features such as open reading frames (ORFs) or CpG islands, which indicate promotor regions. Another very common application area is evolutionary analysis, where homologous DNA sequences from different organisms are compared. DNA sequencing can additionally be applied for the associations between mutations and diseases or sometimes even disease resistance, deeming one of the most useful applications.</p>
<p>A very popular example is sickle cell disease, a group of blood disorders, which results from an abnormality in the oxygen-carrying protein hemoglobin in red blood cells.
This leads to serious health issues including pain, anemia, swelling in the hands and feet, bacterial infections and strokes. The cause of sickle cell disease is the inheritance of two abnormal copies of the β-globin gene (HBB) that makes hemoglobin, one from each parent.
The gene defect is caused by a single nucleotide mutation where a GAG codon changes to a GTG codon of the β-globin gene. This results in the amino acid glutamate being substituted by valine at position 6 (E6V substitution) and henceforth the above-mentioned disease.
It is unfortunately not always possible to find such “simple” associations between single nucleotide mutations and diseases, due to most diseases being caused by, for example, complex regulatory processes.</p>
</section>
<section id="a-brief-history-of-sequencing">
<h2><span class="section-number">2.2. </span>A brief history of sequencing<a class="headerlink" href="#a-brief-history-of-sequencing" title="Permalink to this headline">#</a></h2>
<section id="first-generation-sequencing">
<h3><span class="section-number">2.2.1. </span>First generation sequencing<a class="headerlink" href="#first-generation-sequencing" title="Permalink to this headline">#</a></h3>
<p>Although DNA was already first isolated in 1869 by Friedrich Mietscher, it took the scientific community more than 100 years to develop high throughput sequencing technologies.
In 1953, Watson, Crick and Franklin discovered the structure of DNA; and in 1965 Robert Holley sequenced the first tRNA.
Seven years later, in 1972, Walter Fiers was the first to sequence a complete gene (the coat protein of bacteriophage MS2) using RNAses to digest the virus RNA, isolate oligonucleotides and finally separate them with electrophoresis and chromatography<span id="id1">[<a class="reference internal" href="#id100" title="W. MIN JOU, G. HAEGEMAN, M. YSEBAERT, and W. FIERS. Nucleotide sequence of the gene coding for the bacteriophage ms2 coat protein. Nature, 237(5350):82-88, May 1972. URL: https://doi.org/10.1038/237082a0, doi:10.1038/237082a0.">JOU <em>et al.</em>, 1972</a>]</span>.
In parallel, Friedrich Sanger developed a DNA sequencing method using radiolabeled, partially digested fragments termed “chain termination method”, which is more commonly known as “Sanger Sequencing”. Although Sanger Sequencing is still used even today, it suffered from several shortcomings, including lack of automation and time-consuming.
In 1987, Leroy Hood and Michael Hunkapiller developed the ABI 370, an instrument that automates the Sanger Sequencing process. Its most important innovative accomplishment was the automatic labeling of DNA fragments with fluorescent dyes instead of radioactive molecules. This change not only made the method safer to perform, but also allowed for computers to analyze the acquired data<span id="id2">[<a class="reference internal" href="#id101" title="L E Hood, M W Hunkapiller, and L M Smith. Automated term`DNA` sequencing and analysis of the human genome. Genomics, 1(3):201–212, November 1987.">Hood <em>et al.</em>, 1987</a>]</span>.</p>
<p>Strengths:</p>
<ul class="simple">
<li><p>Sanger sequencing is simple and affordable.</p></li>
<li><p>If done correctly, the error rate is very low (&lt;0.001%).</p></li>
</ul>
<p>Limitations:</p>
<ul class="simple">
<li><p>Sanger methods can only sequence short pieces of DNA of about 300 to 1000 base pairs (bp).</p></li>
<li><p>The quality of a Sanger sequence is often not very good in the first 15 to 40 bases, because this is where the primers bind.</p></li>
<li><p>Sequencing degrades after 700 to 900 bases.</p></li>
<li><p>If the sequenced DNA fragment has been cloned, some of the cloning vector sequence may find its way into the final sequence.</p></li>
<li><p>Sanger sequencing is more expensive than second or third generation sequencing per sequenced base.</p></li>
</ul>
</section>
<section id="second-generation-sequencing">
<h3><span class="section-number">2.2.2. </span>Second generation sequencing<a class="headerlink" href="#second-generation-sequencing" title="Permalink to this headline">#</a></h3>
<p>Nine years later, in 1996, Mostafa Ronaghi, Mathias Uhlen, and Pȧl Nyŕen introduced a new DNA sequencing technique called pyrosequencing, introducing the age of second generation sequencing.
Second generation sequencing, also known as next-generation sequencing (NGS), was primarily made possible by further automation in the lab, the usage of computers, and the miniaturization of reactions.
Pyrosequencing measures luminescence that is generated by pyrophosphate synthesis during sequencing.
This process is also commonly known as “sequencing-by-synthesis”.
Two years later, Shankar Balasubramanian and David Klenerman, developed and adapted the sequencing-by-synthesis process for a new method which utilizes fluorescent dyes at the company Solexa. Solexa’s technology also forms the basis of Illumina’s sequencers, which dominate the market today.
The Roche 454 sequencer developed in 2005, was the first sequencer to fully automate the pyrosequencing process in a single, automated machine.
Many other platforms were introduced such as SOLiD systems’ “sequencing-by-ligation” (2007) and Life Technologies’ Ion Torrent (2011) that uses “sequencing-by-synthesis” to detect hydrogen ions when new DNA is synthesized.</p>
<p>Strengths:</p>
<ul class="simple">
<li><p>Second generation generation sequencing is often the cheapest option with respect to required chemicals.</p></li>
<li><p>Sparse material can still be used as input.</p></li>
<li><p>High sensitivity to detect low-frequency variants and comprehensive genome coverage.</p></li>
<li><p>High capacity with sample multiplexing.</p></li>
<li><p>Ability to sequence thousands of genes simultaneously,</p></li>
</ul>
<p>Limitations:</p>
<ul class="simple">
<li><p>The sequencing machines are expensive and often need to be shared with colleagues.</p></li>
<li><p>Second generation sequencers are big, stationary machines and not designed for field work.</p></li>
<li><p>Generally, second generation sequencing results in many short sequencing fragments (reads) which are hard to use for novel genomes.</p></li>
<li><p>The quality of sequencing result is dependent on the reference genome</p></li>
</ul>
</section>
<section id="third-generation-sequencing">
<h3><span class="section-number">2.2.3. </span>Third generation sequencing<a class="headerlink" href="#third-generation-sequencing" title="Permalink to this headline">#</a></h3>
<p>The third generation of sequencing, nowadays also known as next-generation sequencing, brought two innovations to the market.
First, long-read sequencing, which describes the ability to obtain nucleotide fragments of longer lengths than the usual Illumina short-read sequencers generate (order of 75 to 300 base pairs depending on the sequencer). This is especially important for the assembly of novel genomes without an available reference genome. Second, the ability to sequence in real time is another major advancement in third generation sequencing. Combined with portable sequencers, which are small in size and do not require further complex machines for the chemistry, sequencing is now “field-ready” and can be used even far away from laboratory facilities to collect samples.</p>
<p>Pacific Biosciences’ (PacBio) introduced zero-mode waveguide (ZMW) sequencing in 2010, which uses so-called nanoholes containing a single DNA polymerase. This allows incorporation of any single nucleotide to be directly observed by detectors attached below the nanoholes. Each type of nucleotides is labeled with a specific fluorescent dye that emits fluorescent signals during the incorporation process, which are subsequently measured as sequence readout. Reads obtained from PacBio sequencers are usually of 8 to 15 kilobases (kb), with possibilities for up to 70kb.</p>
<p>Oxford Nanopore Technologies’ introduced the GridION in 2012. The GridION and its successors MinION and Flongle are portable sequencers for DNA and RNA sequencing which produce reads of more than 2 Mb. Notably, such a sequencing device even fits into a single human hand. Oxford Nanopore sequencers observe changes in the electrical current that occur when nucleic acids pass through protein nanopores, to identify the nucleotide sequence<span id="id3">[<a class="reference internal" href="#id102" title="Miten Jain, Hugh E. Olsen, Benedict Paten, and Mark Akeson. The oxford nanopore minion: delivery of nanopore sequencing to the genomics community. Genome Biology, 17(1):239, Nov 2016. URL: https://doi.org/10.1186/s13059-016-1103-0, doi:10.1186/s13059-016-1103-0.">Jain <em>et al.</em>, 2016</a>]</span>.</p>
<p>Strengths:</p>
<ul class="simple">
<li><p>Long reads will allow for the assembly of large novel genomes.</p></li>
<li><p>Sequencers are portable, allowing for field work.</p></li>
<li><p>Possibility to directly detect epigenetic modifications of DNA and RNA sequences.</p></li>
<li><p>Speed. Third generation sequencers are fast.</p></li>
</ul>
<p>Limitations:</p>
<ul class="simple">
<li><p>Some third generation sequencers exhibit higher error rates than second generation sequencers.</p></li>
<li><p>The reagents are generally more expensive than second generation sequencing.</p></li>
</ul>
</section>
</section>
<section id="overview-of-the-ngs-process">
<h2><span class="section-number">2.3. </span>Overview of the NGS process<a class="headerlink" href="#overview-of-the-ngs-process" title="Permalink to this headline">#</a></h2>
<p>Even though a variety of NGS technologies exist, the general steps to sequence DNA (and therefore reverse transcribed RNA) are largely the same. The differences lie primarily in the chemistry of the respective sequencing technologies.</p>
<ol class="simple">
<li><p><strong>Sample and library preparation</strong>: As a first step, a so-called library is prepared by fragmenting the DNA samples and ligating them with adapter molecules. They act in the hybridisation of the library fragments to the matrix and provide a priming site.</p></li>
<li><p><strong>Amplification and sequencing</strong>: In the second step, the library gets converted into single strand molecules. During an amplification step (such as a polymerase chain reaction), clusters of DNA molecules are being created. All of the clusters perform individual reactions during a single sequencing run.</p></li>
<li><p><strong>Data output and analysis</strong>: The output of a sequencing experiment depends on the sequencing technology and chemistry. Some sequencers generate fluorescence signals which are stored in specific output files, and others may generate electric signals which are stored in corresponding file formats. Generally, the amount of generated data, the raw data, is very large. Such data requires complex and computationally heavy processing. This is further discussed in the raw data processing chapter.</p></li>
</ol>
</section>
<section id="rna-sequencing">
<span id="exp-data-rna-sequencing"></span><h2><span class="section-number">2.4. </span>RNA sequencing<a class="headerlink" href="#rna-sequencing" title="Permalink to this headline">#</a></h2>
<p>So far, we have only introduced sequencing with the unmentioned assumption that the DNA is being sequenced. However, knowing the DNA sequence of an organism and the positions of its regulatory elements tells us very little about the dynamic and real-time operations of a cell.
For example, by combining different mRNA splicing sites and exons from the same mRNA precursor, one gene can code for multiple proteins. This alternative splicing event is naturally occurring and commonly seen in eukaryotes; however, a variant could potentially result in a non-functional enzyme and an induced disease state. This is where RNA sequencing (RNA-Seq) comes into play.
RNA-Seq largely follows the DNA sequencing protocols, but includes a reverse transcription step where complementary DNA (cDNA) is synthesized from the RNA template. Sequencing RNA allows scientists to obtain snapshots of cells, tissues or organisms at the time of sequencing in the form of expression profiles of genes.
This information can be used to detect changes in disease states in response to therapeutics, under different environmental conditions, when comparing genotypes and other experimental designs.
Modern RNA sequencing allows for an unbiased sampling of transcripts in contrast to for example microarray based assays or <a class="reference internal" href="../glossary.html#term-RT-qPCR"><span class="xref std std-term">RT-qPCR</span></a>, which require probe design to specifically target the regions of interest.
The obtained gene expression profiles further enable the detection of gene isoforms, gene fusions, single nucleotide variants, and many other interesting properties.
Modern RNA sequencing is not limited by prior knowledge and allows for the capture of both known and novel features, resulting in rich data sets that can be used for exploratory data analysis.</p>
</section>
<section id="id4">
<h2><span class="section-number">2.5. </span>Single-cell RNA sequencing<a class="headerlink" href="#id4" title="Permalink to this headline">#</a></h2>
<section id="overview">
<h3><span class="section-number">2.5.1. </span>Overview<a class="headerlink" href="#overview" title="Permalink to this headline">#</a></h3>
<p>Sequencing of RNA can be mainly conducted in two ways: Either by sequencing the mixed RNA from the source of interest across cells (bulk sequencing) or by sequencing the transcriptomes of the cells individually (single-cell sequencing).
Mixing the RNA of all cells is in most cases cheaper and easier than experimentally complex single-cell sequencing. Bulk RNA-Seq results in cell-averaged expression profiles, which are generally easier to analyze, but also hide some of the complexity such as cell expression profile heterogeneity, which may help answer the question of interest.
Some drugs or perturbations may affect only specific cell types or interactions between cell types. For example, in oncology, it is possible to have rare drug resistant tumor cells causing relapse, which is difficult to identify by simple bulk RNA-seq even on cultured cells.</p>
<p>To uncover such relationships, it is vital to examine gene expression on a single-cell level.
Single-cell RNA-Seq (scRNA-Seq) does, however, come with several caveats.
First, single-cell experiments are generally more expensive and more difficult to properly conduct. Second, the downstream analysis becomes more complex due to the increased resolution, and it is easier to draw false conclusions.</p>
<p>A single-cell experiment generally speaking, follows the same steps as a bulk RNA-Seq experiment (see above), but requires several adaptations.
Just like bulk sequencing, single-cell sequencing requires lysis, reverse transcription, amplification, and the eventual sequencing.
In addition, single-cell sequencing requires cell isolation and a physical separation into smaller reaction chambers or another form of cell labeling to be able to map the obtained transcriptomes back to the cells of origin later on.
Hence, these are also the steps where most single-cell assays differ: single-cell isolation, transcript amplification, and, depending on the sequencing machine, sequencing. Before explaining how the different approaches to sequencing work, we will now discuss transcript quantification more closely.</p>
</section>
<section id="transcript-quantification">
<span id="exp-data-transcript-quantification"></span><h3><span class="section-number">2.5.2. </span>Transcript quantification<a class="headerlink" href="#transcript-quantification" title="Permalink to this headline">#</a></h3>
<p>Transcript quantification is the process of counting the hits of the sequenced transcripts against the gene sequences. These counted hits eventually make it into the count table. More details on this computational process will be described in the next chapter.
There are two major approaches to transcript quantification: full-length and tag-based.
Full-length protocols try to cover the whole transcript uniformly with sequencing reads, whereas tag-based protocols only capture the 5’ or 3’ ends.
The transcript quantification method has strong implications on the captured genes, and analysts must therefore be aware of the used quantification process.
Full-length sequencing is restricted to plate-based protocols (see below) and the library preparation is comparable to bulk RNA-seq sequencing approaches.
An even coverage of transcripts is not always achieved with full-length protocols and therefore specific regions across the gene body may still be biased.
A major advantage of full-length protocols is that they allow for the detection of splice variants.
Tag-based protocols only sequence the 3’ or 5’ ends of the transcripts. This comes at the cost of not (necessarily) covering the full gene length, making it difficult to unambiguously align reads to a transcript and distinguishing between different isoforms<span id="id5">[<a class="reference internal" href="#id123" title="Nathan Archer, Mark D. Walsh, Vahid Shahrezaei, and Daniel Hebenstreit. Modeling enzyme processivity reveals that rna-seq libraries are biased in characteristic and correctable ways. Cell Systems, 3(5):467-479.e12, 2016. URL: https://www.sciencedirect.com/science/article/pii/S2405471216303313, doi:https://doi.org/10.1016/j.cels.2016.10.012.">Archer <em>et al.</em>, 2016</a>]</span>. However, it allows for the usage of unique molecular identifiers (UMIs), which are useful to resolve biases in the transcript amplification process.
The transcript amplification process is a critical step in any RNA-seq sequencing run, to ensure that the transcripts are abundant enough for quality control and sequencing. During this process, which is typically conducted with polymerase chain reaction (PCR), copies are made from identical fragments of the original molecule. Since the copies and the original molecules are indistinguishable, determining the original number of molecules in samples becomes challenging. The usage of UMIs is a common solution to quantify the original, non-duplicated molecules.
The UMIs serve as molecular barcodes and are also sometimes referred to as random barcodes. These ‘barcodes’ consist of short random nucleotide sequences that are added to every molecule in the sample as a unique tag. UMIs must be added during library generation before the amplification step. The ability to accurately identify PCR duplicates is important for downstream analysis to rule out - or be aware of amplification biases<span id="id6">[<a class="reference internal" href="#id103" title="Daniel Aird, Michael G. Ross, Wei-Sheng Chen, Maxwell Danielsson, Timothy Fennell, Carsten Russ, David B. Jaffe, Chad Nusbaum, and Andreas Gnirke. Analyzing and minimizing pcr amplification bias in illumina sequencing libraries. Genome Biology, 12(2):R18, Feb 2011. URL: https://doi.org/10.1186/gb-2011-12-2-r18, doi:10.1186/gb-2011-12-2-r18.">Aird <em>et al.</em>, 2011</a>]</span>.
Amplification bias is a term for the RNA/cDNA sequences which are preferentially amplified and will therefore be sequenced more often, resulting in higher counts. It can have a detrimental effect on any gene expression analysis, because the not-very-active genes may suddenly appear to be highly expressed. This is especially true for sequences which are amplified at a later stage of the PCR step, where the error rate may already be comparably higher than earlier PCR stages.
Although it is computationally possible to detect and remove such sequences by removing reads with identical alignment coordinates, it is generally advised to always design the experiment with UMIs, if possible.
The usage of UMIs further allows for normalization of gene counts to be performed without a loss of accuracy<span id="id7">[<a class="reference internal" href="#id104" title="Teemu Kivioja, Anna Vähärautio, Kasper Karlsson, Martin Bonke, Martin Enge, Sten Linnarsson, and Jussi Taipale. Counting absolute numbers of molecules using unique molecular identifiers. Nature Methods, 9(1):72-74, Jan 2012. URL: https://doi.org/10.1038/nmeth.1778, doi:10.1038/nmeth.1778.">Kivioja <em>et al.</em>, 2012</a>]</span>.</p>
<p>ADD A FIGURE HERE.</p>
</section>
<section id="single-cell-sequencing-protocols">
<h3><span class="section-number">2.5.3. </span>Single-cell sequencing protocols<a class="headerlink" href="#single-cell-sequencing-protocols" title="Permalink to this headline">#</a></h3>
<p>Currently, three types of single-cell sequencing protocols exist, which are grouped primarily by their cell isolation protocols: Microfluidic device-based strategies where cells are encapsulated into hydrogel droplets; well plate based protocols where cells are physically separated into wells and finally, the commercial Fluidigm C1 microfluidic chip based solution which loads and separates cells into small reaction chambers. These three approaches differ in their ability to recover transcripts, the number of sequenced cells, and many other aspects.
In the following subsections, we will briefly discuss how they work, their strengths and weaknesses, and possible biases that data analysts should be aware of regarding the respective protocols.</p>
<section id="microfluidic-device-based-protocols">
<h4><span class="section-number">2.5.3.1. </span>Microfluidic device based protocols<a class="headerlink" href="#microfluidic-device-based-protocols" title="Permalink to this headline">#</a></h4>
<p>Microfluidic device based single-cell strategies trap cells inside hydrogel droplets allowing for compartmentalisation into single-cell reaction chambers. The most widely used protocols inDrop<span id="id8">[<a class="reference internal" href="#id110" title="Allon M. Klein, Linas Mazutis, Ilke Akartuna, Naren Tallapragada, Adrian Veres, Victor Li, Leonid Peshkin, David A. Weitz, and Marc W. Kirschner. Droplet barcoding for single-cell transcriptomics applied to embryonic stem cells. Cell, 161(5):1187-1201, May 2015. PMC4441768[pmcid]. URL: https://doi.org/10.1016/j.cell.2015.04.044, doi:10.1016/j.cell.2015.04.044.">Klein <em>et al.</em>, 2015</a>]</span>, Drop-seq<span id="id9">[<a class="reference internal" href="#id111" title="Evan Z. Macosko, Anindita Basu, Rahul Satija, James Nemesh, Karthik Shekhar, Melissa Goldman, Itay Tirosh, Allison R. Bialas, Nolan Kamitaki, Emily M. Martersteck, John J. Trombetta, David A. Weitz, Joshua R. Sanes, Alex K. Shalek, Aviv Regev, and Steven A. McCarroll. Highly parallel genome-wide expression profiling of individual cells using nanoliter droplets. Cell, 161(5):1202-1214, May 2015. URL: https://doi.org/10.1016/j.cell.2015.05.002, doi:10.1016/j.cell.2015.05.002.">Macosko <em>et al.</em>, 2015</a>]</span> and the commercially available 10x Genomics Chromium<span id="id10">[<a class="reference internal" href="#id112" title="Grace X. Y. Zheng, Jessica M. Terry, Phillip Belgrader, Paul Ryvkin, Zachary W. Bent, Ryan Wilson, Solongo B. Ziraldo, Tobias D. Wheeler, Geoff P. McDermott, Junjie Zhu, Mark T. Gregory, Joe Shuga, Luz Montesclaros, Jason G. Underwood, Donald A. Masquelier, Stefanie Y. Nishimura, Michael Schnall-Levin, Paul W. Wyatt, Christopher M. Hindson, Rajiv Bharadwaj, Alexander Wong, Kevin D. Ness, Lan W. Beppu, H. Joachim Deeg, Christopher McFarland, Keith R. Loeb, William J. Valente, Nolan G. Ericson, Emily A. Stevens, Jerald P. Radich, Tarjei S. Mikkelsen, Benjamin J. Hindson, and Jason H. Bielas. Massively parallel digital transcriptional profiling of single cells. Nature Communications, 8(1):14049, Jan 2017. URL: https://doi.org/10.1038/ncomms14049, doi:10.1038/ncomms14049.">Zheng <em>et al.</em>, 2017</a>]</span> are able to generate such droplets several thousand times per second.
This massively parallel process generates very high numbers of droplets for a relatively low cost. Although all three protocols differ in details, nanoliter-sized droplets containing encapsulated cells are always designed to capture beads and cells simultaneously.
The encapsulation process is conducted with specialized microbeads with on-bead primers containing a PCR handle, a cell barcode and a 4-8b bp-long unique molecular identifier (UMI - see below) and a poly-T tail.
Upon lysis the cell’s mRNA is instantaneously released and captured by the barcoded oligonucleotides that are attached on the beads.
Next, the droplets are collected and broken to release single-cell transcriptomes attached to microparticles (STAMPs).
This is followed by PCR and reverse transcription to capture and amplify the transcripts.
Finally, tagmentation takes place where the transcripts are randomly cut and sequencing adaptors get attached. This process results in sequencing libraries that are ready for sequencing as described above. In microfluidic based protocols only about 10% of the transcripts of the cell are recovered<span id="id11">[<a class="reference internal" href="#id113" title="Saiful Islam, Amit Zeisel, Simon Joost, Gioele La Manno, Pawel Zajac, Maria Kasper, Peter Lönnerberg, and Sten Linnarsson. Quantitative single-cell term`rna`-seq with unique molecular identifiers. Nature Methods, 11(2):163-166, Feb 2014. URL: https://doi.org/10.1038/nmeth.2772, doi:10.1038/nmeth.2772.">Islam <em>et al.</em>, 2014</a>]</span>. Notably, this low sequencing is sufficient for robust identification of cell types.</p>
<p>All three microfluidic device-based methods result in characteristic biases. The material of the used beads differs between the protocols. Drop-seq uses brittle resin for the beads and therefore the beads are encapsulated with a Poisson distribution, whereas the InDrop and 10X Genomics beads are deformable resulting in bead occupancies of over 80%<span id="id12">[<a class="reference internal" href="#id114" title="Xiannian Zhang, Tianqi Li, Feng Liu, Yaqi Chen, Jiacheng Yao, Zeyao Li, Yanyi Huang, and Jianbin Wang. Comparative analysis of droplet-based ultra-high-throughput single-cell term`rna`-seq systems. Molecular Cell, 73(1):130-142.e5, Jan 2019. URL: https://doi.org/10.1016/j.molcel.2018.10.020, doi:10.1016/j.molcel.2018.10.020.">Zhang <em>et al.</em>, 2019</a>]</span>.
Moreover, capture efficiency is likely influenced by the use of surface-tethered primers in Drop-Seq. InDrop uses primers which are released with photocleavage and 10X genomics dissolves the beads. This disparity also affects the location of the reverse transcription process. In Drop-seq, reverse transcription occurs after the beads are released from the droplets, while reverse transcription takes place inside the droplets for the InDrop and 10X genomics protocols<span id="id13">[<a class="reference internal" href="#id114" title="Xiannian Zhang, Tianqi Li, Feng Liu, Yaqi Chen, Jiacheng Yao, Zeyao Li, Yanyi Huang, and Jianbin Wang. Comparative analysis of droplet-based ultra-high-throughput single-cell term`rna`-seq systems. Molecular Cell, 73(1):130-142.e5, Jan 2019. URL: https://doi.org/10.1016/j.molcel.2018.10.020, doi:10.1016/j.molcel.2018.10.020.">Zhang <em>et al.</em>, 2019</a>]</span>.</p>
<p>A comparison from Zhang et al. in 2019 uncovered that inDrop and Drop-seq are outperformed by 10X Genomics with respect to bead quality, as the cell barcodes in the former two systems contained obvious mismatches. Moreover, the proportion of reads originating from valid barcodes was 75% for 10X Genomics, compared to only 25% for InDrop and 30% for Drop-seq.</p>
<p>Similar advantages were demonstrated for 10X Genomics regarding sensitivity. During their comparison, 10X Genomics captured about 17000 transcripts from 3000 genes on average, compared to 8000 transcripts from 2500 genes for Drop-seq and 2700 transcripts from 1250 genes for InDrop. Technical noise was the lowest for 10X Genomics, followed by Drop-seq and InDrop<span id="id14">[<a class="reference internal" href="#id114" title="Xiannian Zhang, Tianqi Li, Feng Liu, Yaqi Chen, Jiacheng Yao, Zeyao Li, Yanyi Huang, and Jianbin Wang. Comparative analysis of droplet-based ultra-high-throughput single-cell term`rna`-seq systems. Molecular Cell, 73(1):130-142.e5, Jan 2019. URL: https://doi.org/10.1016/j.molcel.2018.10.020, doi:10.1016/j.molcel.2018.10.020.">Zhang <em>et al.</em>, 2019</a>]</span>.</p>
<p>The actual generated data demonstrated large protocol biases. 10X Genomics favored the capture and amplification of shorter genes and genes with higher GC content, while Drop-seq in comparison preferred genes with lower GC content.
Although 10X Genomics was shown to outperform the other protocols in various aspects, it is also about twice as expensive per cell. Moreover, except the beads, Drop-seq is open-source and the protocol can more easily be adapted if required. InDrop is completely open-sourced, where even the beads can be manufactured and modified in labs. Hence, InDrop is the most flexible of the three protocols.</p>
<p>Strengths:</p>
<ul class="simple">
<li><p>Allows for the cost-efficient sequencing of cells in large quantities, to identify the overall composition of a tissue and characterize rare cell types.</p></li>
<li><p>UMIs can be incorporated.</p></li>
</ul>
<p>Limitations:</p>
<ul class="simple">
<li><p>Low detection rates of transcripts compared to other methods.</p></li>
<li><p>Captures 3’ only and not full transcripts, because the cell barcodes and PCR handles are only added to the end of the transcript.</p></li>
</ul>
</section>
<section id="plate-based">
<h4><span class="section-number">2.5.3.2. </span>Plate based<a class="headerlink" href="#plate-based" title="Permalink to this headline">#</a></h4>
<p>Plate based protocols typically separate the cells physically into microwell plates. The first step entails cell sorting by, for example, fluorescent-activated cell sorting (FACS), where cells are sorted according to specific cell surface markers; or by micro pipetting.
The selected cells are then placed into individual wells containing cell lysis buffers, where subsequently reverse transcription is carried out. This allows for several hundreds of cells to be analyzed in a single experiment with 5000 to 10000 captured genes each.
Plate based sequencing protocols include, but are not limited to, SMART-seq2, MARS-seq, QUARTZ-seq and SRCB-seq. Generally speaking, the protocols differ in their multiplexing ability. For example, MARS-seq allows for three barcode levels, namely molecular, cellular and plate-level tags, for robust multiplexing capabilities. SMART-seq2 on the contrary, does not allow for early multiplexing limiting cell numbers. A systematic comparison of protocols by Mereu et al in 2020 revealed that QUARTZ-seq2 is able to capture more genes than SMART-seq2, MARS-seq or SRCB-seq per cell<span id="id15">[<a class="reference internal" href="#id115" title="Elisabetta Mereu, Atefeh Lafzi, Catia Moutinho, Christoph Ziegenhain, Davis J. McCarthy, Adrián Álvarez-Varela, Eduard Batlle, Sagar, Dominic Grün, Julia K. Lau, Stéphane C. Boutet, Chad Sanada, Aik Ooi, Robert C. Jones, Kelly Kaihara, Chris Brampton, Yasha Talaga, Yohei Sasagawa, Kaori Tanaka, Tetsutaro Hayashi, Caroline Braeuning, Cornelius Fischer, Sascha Sauer, Timo Trefzer, Christian Conrad, Xian Adiconis, Lan T. Nguyen, Aviv Regev, Joshua Z. Levin, Swati Parekh, Aleksandar Janjic, Lucas E. Wange, Johannes W. Bagnoli, Wolfgang Enard, Marta Gut, Rickard Sandberg, Itoshi Nikaido, Ivo Gut, Oliver Stegle, and Holger Heyn. Benchmarking single-cell term`rna`-sequencing protocols for cell atlas projects. Nature Biotechnology, 38(6):747-755, Jun 2020. URL: https://doi.org/10.1038/s41587-020-0469-4, doi:10.1038/s41587-020-0469-4.">Mereu <em>et al.</em>, 2020</a>]</span>, which means QUARTZ-seq2 is able to capture cell-type specific marker genes well, allowing for confident cell type annotation.</p>
<p>Strengths:</p>
<ul class="simple">
<li><p>Recovers many genes per cell, allowing for a deep characterization.</p></li>
<li><p>Possible to gather information before the library preparation e.g. through FACS sorting to associate information such as cell size and the intensity of any used labels with well coordinates.</p></li>
<li><p>Allows for full-length transcript recovery.</p></li>
</ul>
<p>Limitations:</p>
<ul class="simple">
<li><p>The scale of plate-based experiments is limited by the lower throughput of their individual processing units.</p></li>
<li><p>Fragmentation step eliminates strand-specific information <span id="id16">[<a class="reference internal" href="#id122" title="Radmila Hrdlickova, Masoud Toloue, and Bin Tian. Rna-seq methods for transcriptome analysis. WIREs RNA, 8(1):e1364, 2017. URL: https://wires.onlinelibrary.wiley.com/doi/abs/10.1002/wrna.1364, arXiv:https://wires.onlinelibrary.wiley.com/doi/pdf/10.1002/wrna.1364, doi:https://doi.org/10.1002/wrna.1364.">Hrdlickova <em>et al.</em>, 2017</a>]</span>.</p></li>
<li><p>Depending on the protocol, plate based protocols might be labor-intensive with many required pipetting steps, leading to potential technical noise and batch effects.</p></li>
</ul>
</section>
<section id="fluidigm-c1">
<h4><span class="section-number">2.5.3.3. </span>Fluidigm C1<a class="headerlink" href="#fluidigm-c1" title="Permalink to this headline">#</a></h4>
<p>The commercial Fluidigm C1 system is a microfluidic chip, which loads and separates cells into small reaction chambers in an automated manner. The CEL-seq2 and SMART-seq (version 1) protocols are using the Fluidigm C1 chips in their workflow, allowing the RNA extraction and library preparation steps to be conducted together, thereby decreasing the required manual labor. However, the Fluidigm C1 requires rather homogeneous cell mixtures, since the cells will reach different locations on the microfluidic chip based on their size, which could introduce potential location bias.
Since the amplification step is carried out in individual wells, full-length sequencing is possible, effectively reducing the 3’ bias of many other single-cell RNA-seq sequencing protocols. The protocol is generally also more expensive and is therefore primarily useful for an extensive examination of a specific cell population.</p>
<p>Strengths:</p>
<ul class="simple">
<li><p>Allows for full-length transcript coverage.</p></li>
<li><p>Splicing variants and T/B cell receptor repertoire diversity can be recovered.</p></li>
</ul>
<p>Limitations:</p>
<ul class="simple">
<li><p>Only allows for the sequencing of up to 800 cells<span id="id17">[<a class="reference internal" href="#id116" title="Fluidigm. Single-cell analysis with microfluidics. https://www.fluidigm.com/area-of-interest/single-cell-analysis/single-cell-analysis-with-microfluidics, 2022. Accessed: 2022-05-07.">Fluidigm, 2022</a>]</span>.</p></li>
<li><p>More expensive per cell than other protocols.</p></li>
<li><p>Only about 10% of the extracted cells are captured, which makes this protocol unsuitable for rare cell types or low input.</p></li>
<li><p>The used arrays only capture specific cell sizes, which may bias the captured transcripts.</p></li>
</ul>
</section>
<section id="nanopore-single-cell-transcriptome-sequencing">
<h4><span class="section-number">2.5.3.4. </span>Nanopore single-cell transcriptome sequencing<a class="headerlink" href="#nanopore-single-cell-transcriptome-sequencing" title="Permalink to this headline">#</a></h4>
<p>Long-read single-cell sequencing approaches rarely used UMI <span id="id18">[<a class="reference internal" href="#id120" title="Mandeep Singh, Ghamdan Al-Eryani, Shaun Carswell, James M. Ferguson, James Blackburn, Kirston Barton, Daniel Roden, Fabio Luciani, Tri Giang Phan, Simon Junankar, Katherine Jackson, Christopher C. Goodnow, Martin A. Smith, and Alexander Swarbrick. High-throughput targeted long-read single cell sequencing reveals the clonal and transcriptional landscape of lymphocytes. Nature Communications, 10(1):3120, Jul 2019. URL: https://doi.org/10.1038/s41467-019-11049-4, doi:10.1038/s41467-019-11049-4.">Singh <em>et al.</em>, 2019</a>]</span> or did not perform UMI correction <span id="id19">[<a class="reference internal" href="#id119" title="Ishaan Gupta, Paul G. Collier, Bettina Haase, Ahmed Mahfouz, Anoushka Joglekar, Taylor Floyd, Frank Koopmans, Ben Barres, August B. Smit, Steven A. Sloan, Wenjie Luo, Olivier Fedrigo, M. Elizabeth Ross, and Hagen U. Tilgner. Single-cell isoform term`rna` sequencing characterizes isoforms in thousands of cerebellar cells. Nature Biotechnology, 36(12):1197-1202, Dec 2018. URL: https://doi.org/10.1038/nbt.4259, doi:10.1038/nbt.4259.">Gupta <em>et al.</em>, 2018</a>]</span> and therefore assigned novel UMI reads to novel UMIs. Due to the higher sequencing error rate of long-read sequencers this causes serious issues <span id="id20">[<a class="reference internal" href="#id121" title="Kevin Lebrigand, Virginie Magnone, Pascal Barbry, and Rainer Waldmann. High throughput error corrected nanopore single cell transcriptome sequencing. Nature Communications, 11(1):4025, Aug 2020. URL: https://doi.org/10.1038/s41467-020-17800-6, doi:10.1038/s41467-020-17800-6.">Lebrigand <em>et al.</em>, 2020</a>]</span>. Lebigrand et al. introduced ScNaUmi-seq (Single-cell Nanopore sequencing with UMIs) which combines Nanopore sequencing with cell barcode and UMI assignment. The barcode assignment is guided with Illumina data by comparing the cell bar code sequences found in the Nanopore reads with those recovered from the Illumina reads for the same region or gene <span id="id21">[<a class="reference internal" href="#id121" title="Kevin Lebrigand, Virginie Magnone, Pascal Barbry, and Rainer Waldmann. High throughput error corrected nanopore single cell transcriptome sequencing. Nature Communications, 11(1):4025, Aug 2020. URL: https://doi.org/10.1038/s41467-020-17800-6, doi:10.1038/s41467-020-17800-6.">Lebrigand <em>et al.</em>, 2020</a>]</span>. However, this effectively requires two single-cell libraries. scCOLOR-seq computationally identifies barcodes without errors using nucleotide pair complementary across the full length of the barcode. These barcodes are then used as guides to correct the remaining erroneous barcodes <span id="id22">[<a class="reference internal" href="#id124" title="Martin Philpott, Jonathan Watson, Anjan Thakurta, Tom Brown, Udo Oppermann, and Adam P. Cribbs. Nanopore sequencing of single-cell transcriptomes with sccolor-seq. Nature Biotechnology, 39(12):1517-1520, Dec 2021. URL: https://doi.org/10.1038/s41587-021-00965-w, doi:10.1038/s41587-021-00965-w.">Philpott <em>et al.</em>, 2021</a>]</span>. A modified UMI-tools directional network based method corrects for UMI sequence duplication.</p>
<p>Strengths:</p>
<ul class="simple">
<li><p>Recovers splicing and sequence heterogeneity information</p></li>
</ul>
<p>Weaknesses:</p>
<ul class="simple">
<li><p>Nanopore reagents are expensive.</p></li>
<li><p>High cell barcode recovery error rates.</p></li>
<li><p>Depending on the protocol, barcode assignment is guided with Illumina data requiring two sequencing assays.</p></li>
</ul>
</section>
<section id="summary">
<h4><span class="section-number">2.5.3.5. </span>Summary<a class="headerlink" href="#summary" title="Permalink to this headline">#</a></h4>
<p>In summary, we strongly recommend that wet lab and dry lab scientists select the sequencing protocol based on the aim of the study. Is a deep characterization of a specific cell type population desired? In this case one of the plate-based methods may be more suitable. On the contrary, droplet based assays will capture heterogeneous mixtures better, allowing for a more broad characterization of the sequenced cells.
Moreover, if the budget is a limiting factor, the protocol of choice should be more cost-effective and robust. When analyzing the data, be aware of the sequencing assay specific biases.
For an extensive comparison of all single-cell sequencing protocols, we recommend the “Benchmarking single-cell RNA-sequencing protocols for cell atlas projects” paper by Mereu et al<span id="id23">[<a class="reference internal" href="#id115" title="Elisabetta Mereu, Atefeh Lafzi, Catia Moutinho, Christoph Ziegenhain, Davis J. McCarthy, Adrián Álvarez-Varela, Eduard Batlle, Sagar, Dominic Grün, Julia K. Lau, Stéphane C. Boutet, Chad Sanada, Aik Ooi, Robert C. Jones, Kelly Kaihara, Chris Brampton, Yasha Talaga, Yohei Sasagawa, Kaori Tanaka, Tetsutaro Hayashi, Caroline Braeuning, Cornelius Fischer, Sascha Sauer, Timo Trefzer, Christian Conrad, Xian Adiconis, Lan T. Nguyen, Aviv Regev, Joshua Z. Levin, Swati Parekh, Aleksandar Janjic, Lucas E. Wange, Johannes W. Bagnoli, Wolfgang Enard, Marta Gut, Rickard Sandberg, Itoshi Nikaido, Ivo Gut, Oliver Stegle, and Holger Heyn. Benchmarking single-cell term`rna`-sequencing protocols for cell atlas projects. Nature Biotechnology, 38(6):747-755, Jun 2020. URL: https://doi.org/10.1038/s41587-020-0469-4, doi:10.1038/s41587-020-0469-4.">Mereu <em>et al.</em>, 2020</a>]</span>.</p>
</section>
</section>
<section id="single-cell-vs-single-nuclei">
<h3><span class="section-number">2.5.4. </span>single-cell vs single-nuclei<a class="headerlink" href="#single-cell-vs-single-nuclei" title="Permalink to this headline">#</a></h3>
<p>So far we have only been discussing single-cell assays, but it is also possible to only sequence the nuclei of the cells.
Single-cell profiling does not always provide an unbiased view on cell types for specific tissues or organs, such as, for example, the brain. During the tissue dissociation process, some cell types are more vulnerable and therefore difficult to capture. For example, fast-spiking parvalbumin-positive interneurons and subcortically projecting glutamatergic neurons were observed in lower proportions than expected in mouse neocortex<span id="id24">[<a class="reference internal" href="#id105" title="Bosiljka Tasic, Zizhen Yao, Lucas T. Graybuck, Kimberly A. Smith, Thuc Nghi Nguyen, Darren Bertagnolli, Jeff Goldy, Emma Garren, Michael N. Economo, Sarada Viswanathan, Osnat Penn, Trygve Bakken, Vilas Menon, Jeremy Miller, Olivia Fong, Karla E. Hirokawa, Kanan Lathia, Christine Rimorin, Michael Tieu, Rachael Larsen, Tamara Casper, Eliza Barkan, Matthew Kroll, Sheana Parry, Nadiya V. Shapovalova, Daniel Hirschstein, Julie Pendergraft, Heather A. Sullivan, Tae Kyung Kim, Aaron Szafer, Nick Dee, Peter Groblewski, Ian Wickersham, Ali Cetin, Julie A. Harris, Boaz P. Levi, Susan M. Sunkin, Linda Madisen, Tanya L. Daigle, Loren Looger, Amy Bernard, John Phillips, Ed Lein, Michael Hawrylycz, Karel Svoboda, Allan R. Jones, Christof Koch, and Hongkui Zeng. Shared and distinct transcriptomic cell types across neocortical areas. Nature, 563(7729):72-78, Nov 2018. URL: https://doi.org/10.1038/s41586-018-0654-5, doi:10.1038/s41586-018-0654-5.">Tasic <em>et al.</em>, 2018</a>]</span>. On the contrary, non-neuronal cells survive dissociation better than neurons and are overrepresented in single-cell suspensions in the adult human neocortex<span id="id25">[<a class="reference internal" href="#id106" title="Spyros Darmanis, Steven A. Sloan, Ye Zhang, Martin Enge, Christine Caneda, Lawrence M. Shuer, Melanie G. Hayden Gephart, Ben A. Barres, and Stephen R. Quake. A survey of human brain transcriptome diversity at the single cell level. Proceedings of the National Academy of Sciences, 112(23):7285-7290, 2015. URL: https://www.pnas.org/doi/abs/10.1073/pnas.1507125112, arXiv:https://www.pnas.org/doi/pdf/10.1073/pnas.1507125112, doi:10.1073/pnas.1507125112.">Darmanis <em>et al.</em>, 2015</a>]</span>. Moreover, single-cell sequencing highly relies on fresh tissue, making it difficult to make use of tissue biobanks.
On the other hand, the nuclei are more resistant to mechanical force, and can be safely isolated from frozen tissue without the use of tissue dissociation enzymes<span id="id26">[<a class="reference internal" href="#id108" title="Suguna Rani Krishnaswami, Rashel V. Grindberg, Mark Novotny, Pratap Venepally, Benjamin Lacar, Kunal Bhutani, Sara B. Linker, Son Pham, Jennifer A. Erwin, Jeremy A. Miller, Rebecca Hodge, James K. McCarthy, Martijn Kelder, Jamison McCorrison, Brian D. Aevermann, Francisco Diez Fuertes, Richard H. Scheuermann, Jun Lee, Ed S. Lein, Nicholas Schork, Michael J. McConnell, Fred H. Gage, and Roger S. Lasken. Using single nuclei for term`rna`-seq to capture the transcriptome of postmortem neurons. Nature Protocols, 11(3):499-524, Mar 2016. URL: https://doi.org/10.1038/nprot.2016.015, doi:10.1038/nprot.2016.015.">Krishnaswami <em>et al.</em>, 2016</a>]</span>. Both options have varying applicability across tissues and sample types, and the resulting biases and uncertainties are still not fully uncovered. It has been shown already that nuclei accurately reflect all transcriptional patterns of cells<span id="id27">[<a class="reference internal" href="#id109" title="Jiarui Ding, Xian Adiconis, Sean K. Simmons, Monika S. Kowalczyk, Cynthia C. Hession, Nemanja D. Marjanovic, Travis K. Hughes, Marc H. Wadsworth, Tyler Burks, Lan T. Nguyen, John Y. H. Kwon, Boaz Barak, William Ge, Amanda J. Kedaigle, Shaina Carroll, Shuqiang Li, Nir Hacohen, Orit Rozenblatt-Rosen, Alex K. Shalek, Alexandra-Chloé Villani, Aviv Regev, and Joshua Z. Levin. Systematic comparison of single-cell and single-nucleus term`rna`-sequencing methods. Nature Biotechnology, 38(6):737-746, Jun 2020. URL: https://doi.org/10.1038/s41587-020-0465-8, doi:10.1038/s41587-020-0465-8.">Ding <em>et al.</em>, 2020</a>]</span>.
The choice of single-cell versus single-nuclei in the experimental design is mostly driven by the type of tissue sample. Data analysis however should be aware of the fact that dissociation ability will have a strong effect on the potentially observable cell types. Therefore, we strongly encourage discussions between wet lab and dry lab scientists concerning the experimental design.</p>
</section>
</section>
<section id="recommended-reading">
<h2><span class="section-number">2.6. </span>Recommended reading<a class="headerlink" href="#recommended-reading" title="Permalink to this headline">#</a></h2>
<p>To get a more elaborate understanding of the experimental assays we recommend the following papers:</p>
<ul class="simple">
<li><p>Comparative Analysis of Single-Cell RNA Sequencing Methods<span id="id28">[<a class="reference internal" href="#id98" title="Christoph Ziegenhain, Beate Vieth, Swati Parekh, Björn Reinius, Amy Guillaumet-Adkins, Martha Smets, Heinrich Leonhardt, Holger Heyn, Ines Hellmann, and Wolfgang Enard. Comparative analysis of Single-Cell term`RNA` sequencing methods. Mol Cell, 65(4):631–643.e4, February 2017.">Ziegenhain <em>et al.</em>, 2017</a>]</span></p></li>
<li><p>Power analysis of single-cell RNA-sequencing experiments<span id="id29">[<a class="reference internal" href="#id99" title="Valentine Svensson, Kedar Nath Natarajan, Lam-Ha Ly, Ricardo J. Miragaia, Charlotte Labalette, Iain C. Macaulay, Ana Cvejic, and Sarah A. Teichmann. Power analysis of single-cell term`rna`-sequencing experiments. Nature Methods, 14(4):381-387, Apr 2017. URL: https://doi.org/10.1038/nmeth.4220, doi:10.1038/nmeth.4220.">Svensson <em>et al.</em>, 2017</a>]</span></p></li>
<li><p>Single-nucleus and single-cell transcriptomes compared in matched cortical cell types<span id="id30">[<a class="reference internal" href="#id107" title="Trygve E. Bakken, Rebecca D. Hodge, Jeremy A. Miller, Zizhen Yao, Thuc Nghi Nguyen, Brian Aevermann, Eliza Barkan, Darren Bertagnolli, Tamara Casper, Nick Dee, Emma Garren, Jeff Goldy, Lucas T. Graybuck, Matthew Kroll, Roger S. Lasken, Kanan Lathia, Sheana Parry, Christine Rimorin, Richard H. Scheuermann, Nicholas J. Schork, Soraya I. Shehata, Michael Tieu, John W. Phillips, Amy Bernard, Kimberly A. Smith, Hongkui Zeng, Ed S. Lein, and Bosiljka Tasic. Single-nucleus and single-cell transcriptomes compared in matched cortical cell types. PLOS ONE, 13(12):1-24, 12 2018. URL: https://doi.org/10.1371/journal.pone.0209648, doi:10.1371/journal.pone.0209648.">Bakken <em>et al.</em>, 2018</a>]</span></p></li>
<li><p>Guidelines for the experimental design of single-cell RNA sequencing studies<span id="id31">[<a class="reference internal" href="#id117" title="Atefeh Lafzi, Catia Moutinho, Simone Picelli, and Holger Heyn. Tutorial: guidelines for the experimental design of single-cell term`rna` sequencing studies. Nature Protocols, 13(12):2742-2757, Dec 2018. URL: https://doi.org/10.1038/s41596-018-0073-y, doi:10.1038/s41596-018-0073-y.">Lafzi <em>et al.</em>, 2018</a>]</span></p></li>
<li><p>Benchmarking single-cell RNA-sequencing protocols for cell atlas projects<span id="id32">[<a class="reference internal" href="#id115" title="Elisabetta Mereu, Atefeh Lafzi, Catia Moutinho, Christoph Ziegenhain, Davis J. McCarthy, Adrián Álvarez-Varela, Eduard Batlle, Sagar, Dominic Grün, Julia K. Lau, Stéphane C. Boutet, Chad Sanada, Aik Ooi, Robert C. Jones, Kelly Kaihara, Chris Brampton, Yasha Talaga, Yohei Sasagawa, Kaori Tanaka, Tetsutaro Hayashi, Caroline Braeuning, Cornelius Fischer, Sascha Sauer, Timo Trefzer, Christian Conrad, Xian Adiconis, Lan T. Nguyen, Aviv Regev, Joshua Z. Levin, Swati Parekh, Aleksandar Janjic, Lucas E. Wange, Johannes W. Bagnoli, Wolfgang Enard, Marta Gut, Rickard Sandberg, Itoshi Nikaido, Ivo Gut, Oliver Stegle, and Holger Heyn. Benchmarking single-cell term`rna`-sequencing protocols for cell atlas projects. Nature Biotechnology, 38(6):747-755, Jun 2020. URL: https://doi.org/10.1038/s41587-020-0469-4, doi:10.1038/s41587-020-0469-4.">Mereu <em>et al.</em>, 2020</a>]</span></p></li>
<li><p>Direct Comparative Analyses of 10X Genomics Chromium and Smart-seq2<span id="id33">[<a class="reference internal" href="#id118" title="Xiliang Wang, Yao He, Qiming Zhang, Xianwen Ren, and Zemin Zhang. Direct comparative analyses of 10x genomics chromium and smart-seq2. Genomics, Proteomics &amp; Bioinformatics, 19(2):253-266, 2021. Single-cell Omics Analysis. URL: https://www.sciencedirect.com/science/article/pii/S1672022921000486, doi:https://doi.org/10.1016/j.gpb.2020.02.005.">Wang <em>et al.</em>, 2021</a>]</span></p></li>
</ul>
</section>
<section id="references">
<h2><span class="section-number">2.7. </span>References<a class="headerlink" href="#references" title="Permalink to this headline">#</a></h2>
<div class="docutils container" id="id34">
<dl class="citation">
<dt class="label" id="id103"><span class="brackets"><a class="fn-backref" href="#id6">expARC+11</a></span></dt>
<dd><p>Daniel Aird, Michael G. Ross, Wei-Sheng Chen, Maxwell Danielsson, Timothy Fennell, Carsten Russ, David B. Jaffe, Chad Nusbaum, and Andreas Gnirke. Analyzing and minimizing pcr amplification bias in illumina sequencing libraries. <em>Genome Biology</em>, 12(2):R18, Feb 2011. URL: <a class="reference external" href="https://doi.org/10.1186/gb-2011-12-2-r18">https://doi.org/10.1186/gb-2011-12-2-r18</a>, <a class="reference external" href="https://doi.org/10.1186/gb-2011-12-2-r18">doi:10.1186/gb-2011-12-2-r18</a>.</p>
</dd>
<dt class="label" id="id123"><span class="brackets"><a class="fn-backref" href="#id5">expAWSH16</a></span></dt>
<dd><p>Nathan Archer, Mark D. Walsh, Vahid Shahrezaei, and Daniel Hebenstreit. Modeling enzyme processivity reveals that rna-seq libraries are biased in characteristic and correctable ways. <em>Cell Systems</em>, 3(5):467–479.e12, 2016. URL: <a class="reference external" href="https://www.sciencedirect.com/science/article/pii/S2405471216303313">https://www.sciencedirect.com/science/article/pii/S2405471216303313</a>, <a class="reference external" href="https://doi.org/https://doi.org/10.1016/j.cels.2016.10.012">doi:https://doi.org/10.1016/j.cels.2016.10.012</a>.</p>
</dd>
<dt class="label" id="id107"><span class="brackets"><a class="fn-backref" href="#id30">expBHM+18</a></span></dt>
<dd><p>Trygve E. Bakken, Rebecca D. Hodge, Jeremy A. Miller, Zizhen Yao, Thuc Nghi Nguyen, Brian Aevermann, Eliza Barkan, Darren Bertagnolli, Tamara Casper, Nick Dee, Emma Garren, Jeff Goldy, Lucas T. Graybuck, Matthew Kroll, Roger S. Lasken, Kanan Lathia, Sheana Parry, Christine Rimorin, Richard H. Scheuermann, Nicholas J. Schork, Soraya I. Shehata, Michael Tieu, John W. Phillips, Amy Bernard, Kimberly A. Smith, Hongkui Zeng, Ed S. Lein, and Bosiljka Tasic. Single-nucleus and single-cell transcriptomes compared in matched cortical cell types. <em>PLOS ONE</em>, 13(12):1–24, 12 2018. URL: <a class="reference external" href="https://doi.org/10.1371/journal.pone.0209648">https://doi.org/10.1371/journal.pone.0209648</a>, <a class="reference external" href="https://doi.org/10.1371/journal.pone.0209648">doi:10.1371/journal.pone.0209648</a>.</p>
</dd>
<dt class="label" id="id106"><span class="brackets"><a class="fn-backref" href="#id25">expDSZ+15</a></span></dt>
<dd><p>Spyros Darmanis, Steven A. Sloan, Ye Zhang, Martin Enge, Christine Caneda, Lawrence M. Shuer, Melanie G. Hayden Gephart, Ben A. Barres, and Stephen R. Quake. A survey of human brain transcriptome diversity at the single cell level. <em>Proceedings of the National Academy of Sciences</em>, 112(23):7285–7290, 2015. URL: <a class="reference external" href="https://www.pnas.org/doi/abs/10.1073/pnas.1507125112">https://www.pnas.org/doi/abs/10.1073/pnas.1507125112</a>, <a class="reference external" href="https://arxiv.org/abs/https://www.pnas.org/doi/pdf/10.1073/pnas.1507125112">arXiv:https://www.pnas.org/doi/pdf/10.1073/pnas.1507125112</a>, <a class="reference external" href="https://doi.org/10.1073/pnas.1507125112">doi:10.1073/pnas.1507125112</a>.</p>
</dd>
<dt class="label" id="id109"><span class="brackets"><a class="fn-backref" href="#id27">expDAS+20</a></span></dt>
<dd><p>Jiarui Ding, Xian Adiconis, Sean K. Simmons, Monika S. Kowalczyk, Cynthia C. Hession, Nemanja D. Marjanovic, Travis K. Hughes, Marc H. Wadsworth, Tyler Burks, Lan T. Nguyen, John Y. H. Kwon, Boaz Barak, William Ge, Amanda J. Kedaigle, Shaina Carroll, Shuqiang Li, Nir Hacohen, Orit Rozenblatt-Rosen, Alex K. Shalek, Alexandra-Chloé Villani, Aviv Regev, and Joshua Z. Levin. Systematic comparison of single-cell and single-nucleus term`rna`-sequencing methods. <em>Nature Biotechnology</em>, 38(6):737–746, Jun 2020. URL: <a class="reference external" href="https://doi.org/10.1038/s41587-020-0465-8">https://doi.org/10.1038/s41587-020-0465-8</a>, <a class="reference external" href="https://doi.org/10.1038/s41587-020-0465-8">doi:10.1038/s41587-020-0465-8</a>.</p>
</dd>
<dt class="label" id="id116"><span class="brackets"><a class="fn-backref" href="#id17">expFlu22</a></span></dt>
<dd><p>Fluidigm. Single-cell analysis with microfluidics. <a class="reference external" href="https://www.fluidigm.com/area-of-interest/single-cell-analysis/single-cell-analysis-with-microfluidics">https://www.fluidigm.com/area-of-interest/single-cell-analysis/single-cell-analysis-with-microfluidics</a>, 2022. Accessed: 2022-05-07.</p>
</dd>
<dt class="label" id="id119"><span class="brackets"><a class="fn-backref" href="#id19">expGCH+18</a></span></dt>
<dd><p>Ishaan Gupta, Paul G. Collier, Bettina Haase, Ahmed Mahfouz, Anoushka Joglekar, Taylor Floyd, Frank Koopmans, Ben Barres, August B. Smit, Steven A. Sloan, Wenjie Luo, Olivier Fedrigo, M. Elizabeth Ross, and Hagen U. Tilgner. Single-cell isoform term`rna` sequencing characterizes isoforms in thousands of cerebellar cells. <em>Nature Biotechnology</em>, 36(12):1197–1202, Dec 2018. URL: <a class="reference external" href="https://doi.org/10.1038/nbt.4259">https://doi.org/10.1038/nbt.4259</a>, <a class="reference external" href="https://doi.org/10.1038/nbt.4259">doi:10.1038/nbt.4259</a>.</p>
</dd>
<dt class="label" id="id101"><span class="brackets"><a class="fn-backref" href="#id2">expHHS87</a></span></dt>
<dd><p>L E Hood, M W Hunkapiller, and L M Smith. Automated term`DNA` sequencing and analysis of the human genome. <em>Genomics</em>, 1(3):201–212, November 1987.</p>
</dd>
<dt class="label" id="id122"><span class="brackets"><a class="fn-backref" href="#id16">expHTT17</a></span></dt>
<dd><p>Radmila Hrdlickova, Masoud Toloue, and Bin Tian. Rna-seq methods for transcriptome analysis. <em>WIREs RNA</em>, 8(1):e1364, 2017. URL: <a class="reference external" href="https://wires.onlinelibrary.wiley.com/doi/abs/10.1002/wrna.1364">https://wires.onlinelibrary.wiley.com/doi/abs/10.1002/wrna.1364</a>, <a class="reference external" href="https://arxiv.org/abs/https://wires.onlinelibrary.wiley.com/doi/pdf/10.1002/wrna.1364">arXiv:https://wires.onlinelibrary.wiley.com/doi/pdf/10.1002/wrna.1364</a>, <a class="reference external" href="https://doi.org/https://doi.org/10.1002/wrna.1364">doi:https://doi.org/10.1002/wrna.1364</a>.</p>
</dd>
<dt class="label" id="id113"><span class="brackets"><a class="fn-backref" href="#id11">expIZJ+14</a></span></dt>
<dd><p>Saiful Islam, Amit Zeisel, Simon Joost, Gioele La Manno, Pawel Zajac, Maria Kasper, Peter Lönnerberg, and Sten Linnarsson. Quantitative single-cell term`rna`-seq with unique molecular identifiers. <em>Nature Methods</em>, 11(2):163–166, Feb 2014. URL: <a class="reference external" href="https://doi.org/10.1038/nmeth.2772">https://doi.org/10.1038/nmeth.2772</a>, <a class="reference external" href="https://doi.org/10.1038/nmeth.2772">doi:10.1038/nmeth.2772</a>.</p>
</dd>
<dt class="label" id="id102"><span class="brackets"><a class="fn-backref" href="#id3">expJOPA16</a></span></dt>
<dd><p>Miten Jain, Hugh E. Olsen, Benedict Paten, and Mark Akeson. The oxford nanopore minion: delivery of nanopore sequencing to the genomics community. <em>Genome Biology</em>, 17(1):239, Nov 2016. URL: <a class="reference external" href="https://doi.org/10.1186/s13059-016-1103-0">https://doi.org/10.1186/s13059-016-1103-0</a>, <a class="reference external" href="https://doi.org/10.1186/s13059-016-1103-0">doi:10.1186/s13059-016-1103-0</a>.</p>
</dd>
<dt class="label" id="id100"><span class="brackets"><a class="fn-backref" href="#id1">expJHYF72</a></span></dt>
<dd><p>W. MIN JOU, G. HAEGEMAN, M. YSEBAERT, and W. FIERS. Nucleotide sequence of the gene coding for the bacteriophage ms2 coat protein. <em>Nature</em>, 237(5350):82–88, May 1972. URL: <a class="reference external" href="https://doi.org/10.1038/237082a0">https://doi.org/10.1038/237082a0</a>, <a class="reference external" href="https://doi.org/10.1038/237082a0">doi:10.1038/237082a0</a>.</p>
</dd>
<dt class="label" id="id104"><span class="brackets"><a class="fn-backref" href="#id7">expKVaharautioK+12</a></span></dt>
<dd><p>Teemu Kivioja, Anna Vähärautio, Kasper Karlsson, Martin Bonke, Martin Enge, Sten Linnarsson, and Jussi Taipale. Counting absolute numbers of molecules using unique molecular identifiers. <em>Nature Methods</em>, 9(1):72–74, Jan 2012. URL: <a class="reference external" href="https://doi.org/10.1038/nmeth.1778">https://doi.org/10.1038/nmeth.1778</a>, <a class="reference external" href="https://doi.org/10.1038/nmeth.1778">doi:10.1038/nmeth.1778</a>.</p>
</dd>
<dt class="label" id="id110"><span class="brackets"><a class="fn-backref" href="#id8">expKMA+15</a></span></dt>
<dd><p>Allon M. Klein, Linas Mazutis, Ilke Akartuna, Naren Tallapragada, Adrian Veres, Victor Li, Leonid Peshkin, David A. Weitz, and Marc W. Kirschner. Droplet barcoding for single-cell transcriptomics applied to embryonic stem cells. <em>Cell</em>, 161(5):1187–1201, May 2015. PMC4441768[pmcid]. URL: <a class="reference external" href="https://doi.org/10.1016/j.cell.2015.04.044">https://doi.org/10.1016/j.cell.2015.04.044</a>, <a class="reference external" href="https://doi.org/10.1016/j.cell.2015.04.044">doi:10.1016/j.cell.2015.04.044</a>.</p>
</dd>
<dt class="label" id="id108"><span class="brackets"><a class="fn-backref" href="#id26">expKGN+16</a></span></dt>
<dd><p>Suguna Rani Krishnaswami, Rashel V. Grindberg, Mark Novotny, Pratap Venepally, Benjamin Lacar, Kunal Bhutani, Sara B. Linker, Son Pham, Jennifer A. Erwin, Jeremy A. Miller, Rebecca Hodge, James K. McCarthy, Martijn Kelder, Jamison McCorrison, Brian D. Aevermann, Francisco Diez Fuertes, Richard H. Scheuermann, Jun Lee, Ed S. Lein, Nicholas Schork, Michael J. McConnell, Fred H. Gage, and Roger S. Lasken. Using single nuclei for term`rna`-seq to capture the transcriptome of postmortem neurons. <em>Nature Protocols</em>, 11(3):499–524, Mar 2016. URL: <a class="reference external" href="https://doi.org/10.1038/nprot.2016.015">https://doi.org/10.1038/nprot.2016.015</a>, <a class="reference external" href="https://doi.org/10.1038/nprot.2016.015">doi:10.1038/nprot.2016.015</a>.</p>
</dd>
<dt class="label" id="id117"><span class="brackets"><a class="fn-backref" href="#id31">expLMPH18</a></span></dt>
<dd><p>Atefeh Lafzi, Catia Moutinho, Simone Picelli, and Holger Heyn. Tutorial: guidelines for the experimental design of single-cell term`rna` sequencing studies. <em>Nature Protocols</em>, 13(12):2742–2757, Dec 2018. URL: <a class="reference external" href="https://doi.org/10.1038/s41596-018-0073-y">https://doi.org/10.1038/s41596-018-0073-y</a>, <a class="reference external" href="https://doi.org/10.1038/s41596-018-0073-y">doi:10.1038/s41596-018-0073-y</a>.</p>
</dd>
<dt class="label" id="id121"><span class="brackets">expLMBW20</span><span class="fn-backref">(<a href="#id20">1</a>,<a href="#id21">2</a>)</span></dt>
<dd><p>Kevin Lebrigand, Virginie Magnone, Pascal Barbry, and Rainer Waldmann. High throughput error corrected nanopore single cell transcriptome sequencing. <em>Nature Communications</em>, 11(1):4025, Aug 2020. URL: <a class="reference external" href="https://doi.org/10.1038/s41467-020-17800-6">https://doi.org/10.1038/s41467-020-17800-6</a>, <a class="reference external" href="https://doi.org/10.1038/s41467-020-17800-6">doi:10.1038/s41467-020-17800-6</a>.</p>
</dd>
<dt class="label" id="id111"><span class="brackets"><a class="fn-backref" href="#id9">expMBS+15</a></span></dt>
<dd><p>Evan Z. Macosko, Anindita Basu, Rahul Satija, James Nemesh, Karthik Shekhar, Melissa Goldman, Itay Tirosh, Allison R. Bialas, Nolan Kamitaki, Emily M. Martersteck, John J. Trombetta, David A. Weitz, Joshua R. Sanes, Alex K. Shalek, Aviv Regev, and Steven A. McCarroll. Highly parallel genome-wide expression profiling of individual cells using nanoliter droplets. <em>Cell</em>, 161(5):1202–1214, May 2015. URL: <a class="reference external" href="https://doi.org/10.1016/j.cell.2015.05.002">https://doi.org/10.1016/j.cell.2015.05.002</a>, <a class="reference external" href="https://doi.org/10.1016/j.cell.2015.05.002">doi:10.1016/j.cell.2015.05.002</a>.</p>
</dd>
<dt class="label" id="id115"><span class="brackets">expMLM+20</span><span class="fn-backref">(<a href="#id15">1</a>,<a href="#id23">2</a>,<a href="#id32">3</a>)</span></dt>
<dd><p>Elisabetta Mereu, Atefeh Lafzi, Catia Moutinho, Christoph Ziegenhain, Davis J. McCarthy, Adrián Álvarez-Varela, Eduard Batlle, Sagar, Dominic Grün, Julia K. Lau, Stéphane C. Boutet, Chad Sanada, Aik Ooi, Robert C. Jones, Kelly Kaihara, Chris Brampton, Yasha Talaga, Yohei Sasagawa, Kaori Tanaka, Tetsutaro Hayashi, Caroline Braeuning, Cornelius Fischer, Sascha Sauer, Timo Trefzer, Christian Conrad, Xian Adiconis, Lan T. Nguyen, Aviv Regev, Joshua Z. Levin, Swati Parekh, Aleksandar Janjic, Lucas E. Wange, Johannes W. Bagnoli, Wolfgang Enard, Marta Gut, Rickard Sandberg, Itoshi Nikaido, Ivo Gut, Oliver Stegle, and Holger Heyn. Benchmarking single-cell term`rna`-sequencing protocols for cell atlas projects. <em>Nature Biotechnology</em>, 38(6):747–755, Jun 2020. URL: <a class="reference external" href="https://doi.org/10.1038/s41587-020-0469-4">https://doi.org/10.1038/s41587-020-0469-4</a>, <a class="reference external" href="https://doi.org/10.1038/s41587-020-0469-4">doi:10.1038/s41587-020-0469-4</a>.</p>
</dd>
<dt class="label" id="id124"><span class="brackets"><a class="fn-backref" href="#id22">expPWT+21</a></span></dt>
<dd><p>Martin Philpott, Jonathan Watson, Anjan Thakurta, Tom Brown, Udo Oppermann, and Adam P. Cribbs. Nanopore sequencing of single-cell transcriptomes with sccolor-seq. <em>Nature Biotechnology</em>, 39(12):1517–1520, Dec 2021. URL: <a class="reference external" href="https://doi.org/10.1038/s41587-021-00965-w">https://doi.org/10.1038/s41587-021-00965-w</a>, <a class="reference external" href="https://doi.org/10.1038/s41587-021-00965-w">doi:10.1038/s41587-021-00965-w</a>.</p>
</dd>
<dt class="label" id="id120"><span class="brackets"><a class="fn-backref" href="#id18">expSAEC+19</a></span></dt>
<dd><p>Mandeep Singh, Ghamdan Al-Eryani, Shaun Carswell, James M. Ferguson, James Blackburn, Kirston Barton, Daniel Roden, Fabio Luciani, Tri Giang Phan, Simon Junankar, Katherine Jackson, Christopher C. Goodnow, Martin A. Smith, and Alexander Swarbrick. High-throughput targeted long-read single cell sequencing reveals the clonal and transcriptional landscape of lymphocytes. <em>Nature Communications</em>, 10(1):3120, Jul 2019. URL: <a class="reference external" href="https://doi.org/10.1038/s41467-019-11049-4">https://doi.org/10.1038/s41467-019-11049-4</a>, <a class="reference external" href="https://doi.org/10.1038/s41467-019-11049-4">doi:10.1038/s41467-019-11049-4</a>.</p>
</dd>
<dt class="label" id="id99"><span class="brackets"><a class="fn-backref" href="#id29">expSNL+17</a></span></dt>
<dd><p>Valentine Svensson, Kedar Nath Natarajan, Lam-Ha Ly, Ricardo J. Miragaia, Charlotte Labalette, Iain C. Macaulay, Ana Cvejic, and Sarah A. Teichmann. Power analysis of single-cell term`rna`-sequencing experiments. <em>Nature Methods</em>, 14(4):381–387, Apr 2017. URL: <a class="reference external" href="https://doi.org/10.1038/nmeth.4220">https://doi.org/10.1038/nmeth.4220</a>, <a class="reference external" href="https://doi.org/10.1038/nmeth.4220">doi:10.1038/nmeth.4220</a>.</p>
</dd>
<dt class="label" id="id105"><span class="brackets"><a class="fn-backref" href="#id24">expTYG+18</a></span></dt>
<dd><p>Bosiljka Tasic, Zizhen Yao, Lucas T. Graybuck, Kimberly A. Smith, Thuc Nghi Nguyen, Darren Bertagnolli, Jeff Goldy, Emma Garren, Michael N. Economo, Sarada Viswanathan, Osnat Penn, Trygve Bakken, Vilas Menon, Jeremy Miller, Olivia Fong, Karla E. Hirokawa, Kanan Lathia, Christine Rimorin, Michael Tieu, Rachael Larsen, Tamara Casper, Eliza Barkan, Matthew Kroll, Sheana Parry, Nadiya V. Shapovalova, Daniel Hirschstein, Julie Pendergraft, Heather A. Sullivan, Tae Kyung Kim, Aaron Szafer, Nick Dee, Peter Groblewski, Ian Wickersham, Ali Cetin, Julie A. Harris, Boaz P. Levi, Susan M. Sunkin, Linda Madisen, Tanya L. Daigle, Loren Looger, Amy Bernard, John Phillips, Ed Lein, Michael Hawrylycz, Karel Svoboda, Allan R. Jones, Christof Koch, and Hongkui Zeng. Shared and distinct transcriptomic cell types across neocortical areas. <em>Nature</em>, 563(7729):72–78, Nov 2018. URL: <a class="reference external" href="https://doi.org/10.1038/s41586-018-0654-5">https://doi.org/10.1038/s41586-018-0654-5</a>, <a class="reference external" href="https://doi.org/10.1038/s41586-018-0654-5">doi:10.1038/s41586-018-0654-5</a>.</p>
</dd>
<dt class="label" id="id118"><span class="brackets"><a class="fn-backref" href="#id33">expWHZ+21</a></span></dt>
<dd><p>Xiliang Wang, Yao He, Qiming Zhang, Xianwen Ren, and Zemin Zhang. Direct comparative analyses of 10x genomics chromium and smart-seq2. <em>Genomics, Proteomics &amp; Bioinformatics</em>, 19(2):253–266, 2021. Single-cell Omics Analysis. URL: <a class="reference external" href="https://www.sciencedirect.com/science/article/pii/S1672022921000486">https://www.sciencedirect.com/science/article/pii/S1672022921000486</a>, <a class="reference external" href="https://doi.org/https://doi.org/10.1016/j.gpb.2020.02.005">doi:https://doi.org/10.1016/j.gpb.2020.02.005</a>.</p>
</dd>
<dt class="label" id="id114"><span class="brackets">expZLL+19</span><span class="fn-backref">(<a href="#id12">1</a>,<a href="#id13">2</a>,<a href="#id14">3</a>)</span></dt>
<dd><p>Xiannian Zhang, Tianqi Li, Feng Liu, Yaqi Chen, Jiacheng Yao, Zeyao Li, Yanyi Huang, and Jianbin Wang. Comparative analysis of droplet-based ultra-high-throughput single-cell term`rna`-seq systems. <em>Molecular Cell</em>, 73(1):130–142.e5, Jan 2019. URL: <a class="reference external" href="https://doi.org/10.1016/j.molcel.2018.10.020">https://doi.org/10.1016/j.molcel.2018.10.020</a>, <a class="reference external" href="https://doi.org/10.1016/j.molcel.2018.10.020">doi:10.1016/j.molcel.2018.10.020</a>.</p>
</dd>
<dt class="label" id="id112"><span class="brackets"><a class="fn-backref" href="#id10">expZTB+17</a></span></dt>
<dd><p>Grace X. Y. Zheng, Jessica M. Terry, Phillip Belgrader, Paul Ryvkin, Zachary W. Bent, Ryan Wilson, Solongo B. Ziraldo, Tobias D. Wheeler, Geoff P. McDermott, Junjie Zhu, Mark T. Gregory, Joe Shuga, Luz Montesclaros, Jason G. Underwood, Donald A. Masquelier, Stefanie Y. Nishimura, Michael Schnall-Levin, Paul W. Wyatt, Christopher M. Hindson, Rajiv Bharadwaj, Alexander Wong, Kevin D. Ness, Lan W. Beppu, H. Joachim Deeg, Christopher McFarland, Keith R. Loeb, William J. Valente, Nolan G. Ericson, Emily A. Stevens, Jerald P. Radich, Tarjei S. Mikkelsen, Benjamin J. Hindson, and Jason H. Bielas. Massively parallel digital transcriptional profiling of single cells. <em>Nature Communications</em>, 8(1):14049, Jan 2017. URL: <a class="reference external" href="https://doi.org/10.1038/ncomms14049">https://doi.org/10.1038/ncomms14049</a>, <a class="reference external" href="https://doi.org/10.1038/ncomms14049">doi:10.1038/ncomms14049</a>.</p>
</dd>
<dt class="label" id="id98"><span class="brackets"><a class="fn-backref" href="#id28">expZVP+17</a></span></dt>
<dd><p>Christoph Ziegenhain, Beate Vieth, Swati Parekh, Björn Reinius, Amy Guillaumet-Adkins, Martha Smets, Heinrich Leonhardt, Holger Heyn, Ines Hellmann, and Wolfgang Enard. Comparative analysis of Single-Cell term`RNA` sequencing methods. <em>Mol Cell</em>, 65(4):631–643.e4, February 2017.</p>
</dd>
</dl>
</div>
</section>
<section id="contributors">
<h2><span class="section-number">2.8. </span>Contributors<a class="headerlink" href="#contributors" title="Permalink to this headline">#</a></h2>
<p>We gratefully acknowledge the contributions of:</p>
<section id="authors">
<h3><span class="section-number">2.8.1. </span>Authors<a class="headerlink" href="#authors" title="Permalink to this headline">#</a></h3>
<ul class="simple">
<li><p>Lukas Heumos</p></li>
</ul>
</section>
<section id="reviewers">
<h3><span class="section-number">2.8.2. </span>Reviewers<a class="headerlink" href="#reviewers" title="Permalink to this headline">#</a></h3>
<ul class="simple">
<li><p>Yuexin Chen</p></li>
</ul>
</section>
</section>
</section>

    <script type="text/x-thebe-config">
    {
        requestKernel: true,
        binderOptions: {
            repo: "binder-examples/jupyter-stacks-datascience",
            ref: "master",
        },
        codeMirrorConfig: {
            theme: "abcdef",
            mode: "python"
        },
        kernelOptions: {
            name: "python3",
            path: "./introduction"
        },
        predefinedOutput: true
    }
    </script>
    <script>kernelName = 'python3'</script>

              </div>
              
            </main>
            <footer class="footer-article noprint">
                
    <!-- Previous / next buttons -->
<div class='prev-next-area'>
    <a class='left-prev' id="prev-link" href="prior_art.html" title="previous page">
        <i class="fas fa-angle-left"></i>
        <div class="prev-next-info">
            <p class="prev-next-subtitle">previous</p>
            <p class="prev-next-title"><span class="section-number">1. </span>Prior art</p>
        </div>
    </a>
    <a class='right-next' id="next-link" href="raw_data_processing.html" title="next page">
    <div class="prev-next-info">
        <p class="prev-next-subtitle">next</p>
        <p class="prev-next-title"><span class="section-number">3. </span>Raw data processing</p>
    </div>
    <i class="fas fa-angle-right"></i>
    </a>
</div>
            </footer>
        </div>
    </div>
    <div class="footer-content row">
        <footer class="col footer"><p>
  
    By Lukas Heumos, Anna Schaar, single-cell best practices consortium<br/>
  
      &copy; Copyright 2022.<br/>
    <div class="extra_footer">
      <div>
  Brought to you by
  <a href="https://www.github.com/theislab">Theislab</a>, with many thanks to the single-cell community as a whole!
</div>

    </div>
</p>
        </footer>
    </div>
    
</div>


      </div>
    </div>
  
  <!-- Scripts loaded after <body> so the DOM is not blocked -->
  <script src="../_static/scripts/pydata-sphinx-theme.js?digest=1999514e3f237ded88cf"></script>


  </body>
</html>