{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "4fbdf9dd",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[92m→\u001b[0m connected lamindb: theislab/sc-best-practices\n"
     ]
    }
   ],
   "source": [
    "import lamindb as ln"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "64cf310a",
   "metadata": {},
   "outputs": [],
   "source": [
    "assert ln.setup.settings.instance.slug == \"theislab/sc-best-practices\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "ae141f0d",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[92m→\u001b[0m loaded Transform('6F2qDVGagkKP0000', key='advanced_data_structures_and_frameworks_dataset.ipynb'), re-started Run('e6RXLrfbvxN0SSyF') at 2026-02-10 07:15:34 UTC\n",
      "\u001b[92m→\u001b[0m notebook imports: anndata==0.12.7 lamindb==2.0.1 mudatasets==0.0.2 numpy==2.3.5 spatialdata==0.7.1.post1\n",
      "\u001b[94m•\u001b[0m recommendation: to identify the notebook across renames, pass the uid: ln.track(\"6F2qDVGagkKP\")\n"
     ]
    }
   ],
   "source": [
    "ln.track()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c6e9e80b",
   "metadata": {},
   "source": [
    "Import Libraries"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "91630882",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/luisheinzlmeier/miniforge3/envs/lamindb/lib/python3.14/site-packages/xarray_schema/__init__.py:1: UserWarning: pkg_resources is deprecated as an API. See https://setuptools.pypa.io/en/latest/pkg_resources.html. The pkg_resources package is slated for removal as early as 2025-11-30. Refrain from using this package or pin to Setuptools<81.\n",
      "  from pkg_resources import DistributionNotFound, get_distribution\n",
      "/Users/luisheinzlmeier/miniforge3/envs/lamindb/lib/python3.14/site-packages/tqdm/auto.py:21: TqdmWarning: IProgress not found. Please update jupyter and ipywidgets. See https://ipywidgets.readthedocs.io/en/stable/user_install.html\n",
      "  from .autonotebook import tqdm as notebook_tqdm\n"
     ]
    }
   ],
   "source": [
    "import anndata as ad\n",
    "import mudatasets as mds\n",
    "import numpy as np\n",
    "import spatialdata as sd"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "cde99586",
   "metadata": {},
   "source": [
    "Mudata dataset"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "407e4ab1",
   "metadata": {},
   "outputs": [],
   "source": [
    "n, d, k = 1000, 100, 10\n",
    "\n",
    "z = np.random.default_rng().normal(\n",
    "    loc=np.arange(k), scale=np.arange(k) * 2, size=(n, k)\n",
    ")\n",
    "\n",
    "w = np.random.default_rng().normal(size=(d, k))\n",
    "y = np.dot(z, w.T)\n",
    "\n",
    "d2 = 50\n",
    "w2 = np.random.default_rng().normal(size=(d2, k))\n",
    "y2 = np.dot(z, w2.T)\n",
    "\n",
    "adata = ad.AnnData(y)\n",
    "adata.obs_names = [f\"obs_{i+1}\" for i in range(n)]\n",
    "adata.var_names = [f\"var_{j+1}\" for j in range(d)]\n",
    "\n",
    "adata2 = ad.AnnData(y2)\n",
    "adata2.obs_names = [f\"obs_{i+1}\" for i in range(n)]\n",
    "adata2.var_names = [f\"var2_{j+1}\" for j in range(d2)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "70928f0f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "AnnData object with n_obs × n_vars = 1000 × 100"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "0c18f37d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "AnnData object with n_obs × n_vars = 1000 × 50"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "451d2d21",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata.write(\"advanced_data_structures_and_frameworks_mudata_1.h5ad\")\n",
    "adata.write(\"advanced_data_structures_and_frameworks_mudata_2.h5ad\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "9335ebec",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[92m→\u001b[0m creating new artifact version for key 'introduction/advanced_data_structures_and_frameworks_mudata_1.h5ad' in storage 's3://lamin-eu-central-1/VPwcjx3CDAa2'\n",
      "... uploading advanced_data_structures_and_frameworks_mudata_1.h5ad: 100.0%\n",
      "\u001b[94m•\u001b[0m replacing the existing cache path /Users/luisheinzlmeier/Library/Caches/lamindb/lamin-eu-central-1/VPwcjx3CDAa2/introduction/advanced_data_structures_and_frameworks_mudata_1.h5ad\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "Artifact(uid='NUgcctOuwgBF5c8l0005', version_tag=None, is_latest=True, key='introduction/advanced_data_structures_and_frameworks_mudata_1.h5ad', description='first anndata for the mudata section', suffix='.h5ad', kind=None, otype='AnnData', size=866040, hash='joqm0BQOhBb25FUHLAa_Bw', n_files=None, n_observations=None, branch_id=1, space_id=1, storage_id=1, run_id=36, schema_id=None, created_by_id=6, created_at=2026-02-10 07:15:45 UTC, is_locked=False)"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "af = ln.Artifact(\n",
    "    \"advanced_data_structures_and_frameworks_mudata_1.h5ad\",\n",
    "    key=\"introduction/advanced_data_structures_and_frameworks_mudata_1.h5ad\",\n",
    "    description=\"first anndata for the mudata section\",\n",
    ").save()\n",
    "af"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "8d4331e1",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[92m→\u001b[0m returning artifact with same hash: Artifact(uid='NUgcctOuwgBF5c8l0005', version_tag=None, is_latest=True, key='introduction/advanced_data_structures_and_frameworks_mudata_1.h5ad', description='first anndata for the mudata section', suffix='.h5ad', kind=None, otype='AnnData', size=866040, hash='joqm0BQOhBb25FUHLAa_Bw', n_files=None, n_observations=None, branch_id=1, space_id=1, storage_id=1, run_id=36, schema_id=None, created_by_id=6, created_at=2026-02-10 07:15:45 UTC, is_locked=False); to track this artifact as an input, use: ln.Artifact.get()\n",
      "\u001b[93m!\u001b[0m key introduction/advanced_data_structures_and_frameworks_mudata_1.h5ad on existing artifact differs from passed key introduction/advanced_data_structures_and_frameworks_mudata_2.h5ad, keeping original key; update manually if needed or pass skip_hash_lookup if you want to duplicate the artifact\n",
      "\u001b[93m!\u001b[0m updated description from first anndata for the mudata section to second anndata for the mudata section\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "Artifact(uid='NUgcctOuwgBF5c8l0005', version_tag=None, is_latest=True, key='introduction/advanced_data_structures_and_frameworks_mudata_1.h5ad', description='second anndata for the mudata section', suffix='.h5ad', kind=None, otype='AnnData', size=866040, hash='joqm0BQOhBb25FUHLAa_Bw', n_files=None, n_observations=None, branch_id=1, space_id=1, storage_id=1, run_id=36, schema_id=None, created_by_id=6, created_at=2026-02-10 07:15:45 UTC, is_locked=False)"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "af = ln.Artifact(\n",
    "    \"advanced_data_structures_and_frameworks_mudata_2.h5ad\",\n",
    "    key=\"introduction/advanced_data_structures_and_frameworks_mudata_2.h5ad\",\n",
    "    description=\"second anndata for the mudata section\",\n",
    ").save()\n",
    "af"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e1b28672",
   "metadata": {},
   "source": [
    "Muon dataset"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "b7cb5fe5",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "■ File filtered_feature_bc_matrix.h5 from pbmc10k_multiome has been found at /Users/luisheinzlmeier/mudatasets/pbmc10k_multiome/filtered_feature_bc_matrix.h5\n",
      "■ Checksum is validated (md5) for filtered_feature_bc_matrix.h5\n",
      "■ File atac_fragments.tsv.gz from pbmc10k_multiome has been found at /Users/luisheinzlmeier/mudatasets/pbmc10k_multiome/atac_fragments.tsv.gz\n",
      "■ Checksum is validated (md5) for atac_fragments.tsv.gz\n",
      "■ File atac_fragments.tsv.gz.tbi from pbmc10k_multiome has been found at /Users/luisheinzlmeier/mudatasets/pbmc10k_multiome/atac_fragments.tsv.gz.tbi\n",
      "■ Checksum is validated (md5) for atac_fragments.tsv.gz.tbi\n",
      "■ File atac_peaks.bed from pbmc10k_multiome has been found at /Users/luisheinzlmeier/mudatasets/pbmc10k_multiome/atac_peaks.bed\n",
      "■ Checksum is validated (md5) for atac_peaks.bed\n",
      "■ File atac_peak_annotation.tsv from pbmc10k_multiome has been found at /Users/luisheinzlmeier/mudatasets/pbmc10k_multiome/atac_peak_annotation.tsv\n",
      "■ Checksum is validated (md5) for atac_peak_annotation.tsv\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/luisheinzlmeier/miniforge3/envs/lamindb/lib/python3.14/site-packages/mudatasets/core.py:203: UserWarning: Dataset is in the 10X .h5 format and can't be loaded as backed.\n",
      "  warn(\"Dataset is in the 10X .h5 format and can't be loaded as backed.\")\n",
      "/Users/luisheinzlmeier/miniforge3/envs/lamindb/lib/python3.14/site-packages/muon/_core/preproc.py:31: FutureWarning: `__version__` is deprecated, use `importlib.metadata.version('scanpy')` instead\n",
      "  if Version(scanpy.__version__) < Version(\"1.10\"):\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "■ Loading filtered_feature_bc_matrix.h5...\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/luisheinzlmeier/miniforge3/envs/lamindb/lib/python3.14/site-packages/anndata/_core/anndata.py:1808: UserWarning: Variable names are not unique. To make them unique, call `.var_names_make_unique`.\n",
      "  utils.warn_names_duplicates(\"var\")\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Added `interval` annotation for features from /Users/luisheinzlmeier/mudatasets/pbmc10k_multiome/filtered_feature_bc_matrix.h5\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/luisheinzlmeier/miniforge3/envs/lamindb/lib/python3.14/site-packages/anndata/_core/anndata.py:1808: UserWarning: Variable names are not unique. To make them unique, call `.var_names_make_unique`.\n",
      "  utils.warn_names_duplicates(\"var\")\n",
      "/Users/luisheinzlmeier/miniforge3/envs/lamindb/lib/python3.14/site-packages/mudata/_core/mudata.py:1403: FutureWarning: From 0.4 .update() will not pull obs/var columns from individual modalities by default anymore. Set mudata.set_options(pull_on_update=False) to adopt the new behaviour, which will become the default. Use new pull_obs/pull_var and push_obs/push_var methods for more flexibility.\n",
      "  self._update_attr(\"var\", axis=0, join_common=join_common)\n",
      "/Users/luisheinzlmeier/miniforge3/envs/lamindb/lib/python3.14/site-packages/mudata/_core/mudata.py:571: UserWarning: var_names are not unique. To make them unique, call `.var_names_make_unique`.\n",
      "  self._update_attr_legacy(attr, axis, join_common, **kwargs)\n",
      "/Users/luisheinzlmeier/miniforge3/envs/lamindb/lib/python3.14/site-packages/mudata/_core/mudata.py:1275: FutureWarning: From 0.4 .update() will not pull obs/var columns from individual modalities by default anymore. Set mudata.set_options(pull_on_update=False) to adopt the new behaviour, which will become the default. Use new pull_obs/pull_var and push_obs/push_var methods for more flexibility.\n",
      "  self._update_attr(\"obs\", axis=1, join_common=join_common)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Added peak annotation from /Users/luisheinzlmeier/mudatasets/pbmc10k_multiome/atac_peak_annotation.tsv to .uns['atac']['peak_annotation']\n",
      "Added gene names to peak annotation in .uns['atac']['peak_annotation']\n",
      "Located fragments file: /Users/luisheinzlmeier/mudatasets/pbmc10k_multiome/atac_fragments.tsv.gz\n",
      "pysam is not available. It is required to work with the fragments file.                 Install pysam from PyPI (`pip install pysam`)                 or from GitHub (`pip install git+https://github.com/pysam-developers/pysam`)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<pre>MuData object with n_obs × n_vars = 11909 × 144978\n",
       "  var:\t&#x27;gene_ids&#x27;, &#x27;feature_types&#x27;, &#x27;genome&#x27;, &#x27;interval&#x27;\n",
       "  2 modalities\n",
       "    rna:\t11909 x 36601\n",
       "      var:\t&#x27;gene_ids&#x27;, &#x27;feature_types&#x27;, &#x27;genome&#x27;, &#x27;interval&#x27;\n",
       "    atac:\t11909 x 108377\n",
       "      var:\t&#x27;gene_ids&#x27;, &#x27;feature_types&#x27;, &#x27;genome&#x27;, &#x27;interval&#x27;\n",
       "      uns:\t&#x27;atac&#x27;</pre>"
      ],
      "text/plain": [
       "MuData object with n_obs × n_vars = 11909 × 144978\n",
       "  var:\t'gene_ids', 'feature_types', 'genome', 'interval'\n",
       "  2 modalities\n",
       "    rna:\t11909 x 36601\n",
       "      var:\t'gene_ids', 'feature_types', 'genome', 'interval'\n",
       "    atac:\t11909 x 108377\n",
       "      var:\t'gene_ids', 'feature_types', 'genome', 'interval'\n",
       "      uns:\t'atac'"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mdata, _ = mds.load(\"pbmc10k_multiome\", with_info=True, full=True)\n",
    "mdata.var_names_make_unique()\n",
    "mdata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "8484f861",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/luisheinzlmeier/miniforge3/envs/lamindb/lib/python3.14/site-packages/mudata/_core/mudata.py:1403: FutureWarning: From 0.4 .update() will not pull obs/var columns from individual modalities by default anymore. Set mudata.set_options(pull_on_update=False) to adopt the new behaviour, which will become the default. Use new pull_obs/pull_var and push_obs/push_var methods for more flexibility.\n",
      "  self._update_attr(\"var\", axis=0, join_common=join_common)\n",
      "/Users/luisheinzlmeier/miniforge3/envs/lamindb/lib/python3.14/site-packages/mudata/_core/mudata.py:1275: FutureWarning: From 0.4 .update() will not pull obs/var columns from individual modalities by default anymore. Set mudata.set_options(pull_on_update=False) to adopt the new behaviour, which will become the default. Use new pull_obs/pull_var and push_obs/push_var methods for more flexibility.\n",
      "  self._update_attr(\"obs\", axis=1, join_common=join_common)\n",
      "/Users/luisheinzlmeier/miniforge3/envs/lamindb/lib/python3.14/site-packages/mudata/_core/mudata.py:1403: FutureWarning: From 0.4 .update() will not pull obs/var columns from individual modalities by default anymore. Set mudata.set_options(pull_on_update=False) to adopt the new behaviour, which will become the default. Use new pull_obs/pull_var and push_obs/push_var methods for more flexibility.\n",
      "  self._update_attr(\"var\", axis=0, join_common=join_common)\n",
      "/Users/luisheinzlmeier/miniforge3/envs/lamindb/lib/python3.14/site-packages/mudata/_core/mudata.py:1275: FutureWarning: From 0.4 .update() will not pull obs/var columns from individual modalities by default anymore. Set mudata.set_options(pull_on_update=False) to adopt the new behaviour, which will become the default. Use new pull_obs/pull_var and push_obs/push_var methods for more flexibility.\n",
      "  self._update_attr(\"obs\", axis=1, join_common=join_common)\n"
     ]
    }
   ],
   "source": [
    "mdata_copy = mdata.copy()\n",
    "mdata_copy.write(\"advanced_data_structures_and_frameworks_muon.h5mu\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "8c2c757e",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[92m→\u001b[0m returning artifact with same hash: Artifact(uid='Dpo4hyqWYqKcO8oQ0000', version_tag=None, is_latest=True, key='introduction/advanced_data_structures_and_frameworks_muon.h5mu', description='mudata for the muon section', suffix='.h5mu', kind=None, otype='MuData', size=946523202, hash='Qcw9XaglnXSo3JiSKX2Eit', n_files=None, n_observations=None, branch_id=1, space_id=1, storage_id=1, run_id=36, schema_id=None, created_by_id=6, created_at=2026-02-08 07:46:15 UTC, is_locked=False); to track this artifact as an input, use: ln.Artifact.get()\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "Artifact(uid='Dpo4hyqWYqKcO8oQ0000', version_tag=None, is_latest=True, key='introduction/advanced_data_structures_and_frameworks_muon.h5mu', description='mudata for the muon section', suffix='.h5mu', kind=None, otype='MuData', size=946523202, hash='Qcw9XaglnXSo3JiSKX2Eit', n_files=None, n_observations=None, branch_id=1, space_id=1, storage_id=1, run_id=36, schema_id=None, created_by_id=6, created_at=2026-02-08 07:46:15 UTC, is_locked=False)"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "af = ln.Artifact(\n",
    "    \"advanced_data_structures_and_frameworks_muon.h5mu\",\n",
    "    key=\"introduction/advanced_data_structures_and_frameworks_muon.h5mu\",\n",
    "    description=\"mudata for the muon section\",\n",
    ").save()\n",
    "af"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0acfd8aa",
   "metadata": {},
   "source": [
    "Spatialdata dataset (https://s3.embl.de/spatialdata/spatialdata-sandbox/mouse_liver.zip)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "264f1232",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/dy/0ztxj7hd3h3cz5ghg9zsd3bh0000gn/T/ipykernel_24947/1799439451.py:2: UserWarning: SpatialData is not stored in the most current format. If you want to use Zarr v3, please write the store to a new location using `sdata.write()`.\n",
      "  sdata = sd.read_zarr(data_path)\n",
      "/Users/luisheinzlmeier/miniforge3/envs/lamindb/lib/python3.14/site-packages/zarr/core/group.py:3535: ZarrUserWarning: Object at .DS_Store is not recognized as a component of a Zarr hierarchy.\n",
      "  warnings.warn(\n",
      "no parent found for <ome_zarr.reader.Label object at 0x3629d9a90>: None\n",
      "/Users/luisheinzlmeier/miniforge3/envs/lamindb/lib/python3.14/site-packages/zarr/core/group.py:3535: ZarrUserWarning: Object at .DS_Store is not recognized as a component of a Zarr hierarchy.\n",
      "  warnings.warn(\n",
      "/Users/luisheinzlmeier/miniforge3/envs/lamindb/lib/python3.14/site-packages/zarr/core/group.py:3535: ZarrUserWarning: Object at zmetadata is not recognized as a component of a Zarr hierarchy.\n",
      "  warnings.warn(\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "SpatialData object, with associated Zarr store: /Users/luisheinzlmeier/Desktop/single-cell-best-practices/scripts/introduction/data.zarr\n",
       "├── Images\n",
       "│     └── 'raw_image': DataTree[cyx] (1, 6432, 6432), (1, 1608, 1608)\n",
       "├── Labels\n",
       "│     └── 'segmentation_mask': DataArray[yx] (6432, 6432)\n",
       "├── Points\n",
       "│     └── 'transcripts': DataFrame with shape: (<Delayed>, 3) (2D points)\n",
       "├── Shapes\n",
       "│     └── 'nucleus_boundaries': GeoDataFrame shape: (3375, 1) (2D shapes)\n",
       "└── Tables\n",
       "      └── 'table': AnnData (3375, 99)\n",
       "with coordinate systems:\n",
       "    ▸ 'global', with elements:\n",
       "        raw_image (Images), segmentation_mask (Labels), transcripts (Points), nucleus_boundaries (Shapes)"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data_path = \"data.zarr\"\n",
    "sdata = sd.read_zarr(data_path)\n",
    "sdata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "5c0b0d21",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[92m→\u001b[0m returning artifact with same hash: Artifact(uid='95no3ylux2ltRF760001', version_tag=None, is_latest=True, key='introduction/data_structures_and_frameworks_advanced_spatialdata.zarr', description='zarr for the spatialdata section', suffix='.zarr', kind=None, otype='SpatialData', size=72675549, hash='XJKmPFknKuMfz7UIj8mKfg', n_files=836, n_observations=None, branch_id=1, space_id=1, storage_id=1, run_id=36, schema_id=None, created_by_id=6, created_at=2026-02-08 07:48:54 UTC, is_locked=False); to track this artifact as an input, use: ln.Artifact.get()\n",
      "\u001b[93m!\u001b[0m key introduction/data_structures_and_frameworks_advanced_spatialdata.zarr on existing artifact differs from passed key introduction/advanced_data_structures_and_frameworks_spatialdata.zarr, keeping original key; update manually if needed or pass skip_hash_lookup if you want to duplicate the artifact\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "Artifact(uid='95no3ylux2ltRF760001', version_tag=None, is_latest=True, key='introduction/data_structures_and_frameworks_advanced_spatialdata.zarr', description='zarr for the spatialdata section', suffix='.zarr', kind=None, otype='SpatialData', size=72675549, hash='XJKmPFknKuMfz7UIj8mKfg', n_files=836, n_observations=None, branch_id=1, space_id=1, storage_id=1, run_id=36, schema_id=None, created_by_id=6, created_at=2026-02-08 07:48:54 UTC, is_locked=False)"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "af = ln.Artifact(\n",
    "    \"data.zarr\",\n",
    "    key=\"introduction/advanced_data_structures_and_frameworks_spatialdata.zarr\",\n",
    "    description=\"zarr for the spatialdata section\",\n",
    ").save()\n",
    "af"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f3f87958",
   "metadata": {},
   "source": [
    "Squidpy dataset (https://www.10xgenomics.com/datasets/preview-data-ffpe-human-lung-cancer-with-xenium-multimodal-cell-segmentation-1-standard)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "2364d6da",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/dy/0ztxj7hd3h3cz5ghg9zsd3bh0000gn/T/ipykernel_24947/3643516331.py:2: UserWarning: SpatialData is not stored in the most current format. If you want to use Zarr v3, please write the store to a new location using `sdata.write()`.\n",
      "  sdata = sd.read_zarr(data_path)\n",
      "/Users/luisheinzlmeier/miniforge3/envs/lamindb/lib/python3.14/site-packages/zarr/core/group.py:3535: ZarrUserWarning: Object at .DS_Store is not recognized as a component of a Zarr hierarchy.\n",
      "  warnings.warn(\n",
      "/Users/luisheinzlmeier/miniforge3/envs/lamindb/lib/python3.14/functools.py:982: ImplicitModificationWarning: Transforming to str index.\n",
      "  return dispatch(args[0].__class__)(*args, **kw)\n",
      "/Users/luisheinzlmeier/miniforge3/envs/lamindb/lib/python3.14/site-packages/zarr/core/group.py:3535: ZarrUserWarning: Object at .DS_Store is not recognized as a component of a Zarr hierarchy.\n",
      "  warnings.warn(\n",
      "/Users/luisheinzlmeier/miniforge3/envs/lamindb/lib/python3.14/site-packages/zarr/core/group.py:3535: ZarrUserWarning: Object at zmetadata is not recognized as a component of a Zarr hierarchy.\n",
      "  warnings.warn(\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "SpatialData object, with associated Zarr store: /Users/luisheinzlmeier/Desktop/single-cell-best-practices/scripts/introduction/xenium.zarr\n",
       "├── Images\n",
       "│     ├── 'morphology_focus': DataTree[cyx] (1, 25778, 35416), (1, 12889, 17708), (1, 6444, 8854), (1, 3222, 4427), (1, 1611, 2213)\n",
       "│     └── 'morphology_mip': DataTree[cyx] (1, 25778, 35416), (1, 12889, 17708), (1, 6444, 8854), (1, 3222, 4427), (1, 1611, 2213)\n",
       "├── Points\n",
       "│     └── 'transcripts': DataFrame with shape: (<Delayed>, 8) (3D points)\n",
       "├── Shapes\n",
       "│     ├── 'cell_boundaries': GeoDataFrame shape: (167780, 1) (2D shapes)\n",
       "│     └── 'cell_circles': GeoDataFrame shape: (167780, 2) (2D shapes)\n",
       "└── Tables\n",
       "      └── 'table': AnnData (167780, 313)\n",
       "with coordinate systems:\n",
       "    ▸ 'global', with elements:\n",
       "        morphology_focus (Images), morphology_mip (Images), transcripts (Points), cell_boundaries (Shapes), cell_circles (Shapes)"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data_path = \"xenium.zarr\"\n",
    "sdata = sd.read_zarr(data_path)\n",
    "sdata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "7c4dfb1d",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[92m→\u001b[0m returning artifact with same hash: Artifact(uid='b1RDFMMW8H5hVbzP0000', version_tag=None, is_latest=True, key='introduction/data_structures_and_frameworks_advanced_squidpy.zarr', description='zarr for the squidpy section', suffix='.zarr', kind=None, otype='SpatialData', size=3956752851, hash='n9UcLZoCI-wRgv6EE99ykw', n_files=464, n_observations=None, branch_id=1, space_id=1, storage_id=1, run_id=15, schema_id=None, created_by_id=6, created_at=2025-05-24 20:02:06 UTC, is_locked=False); to track this artifact as an input, use: ln.Artifact.get()\n",
      "\u001b[93m!\u001b[0m key introduction/data_structures_and_frameworks_advanced_squidpy.zarr on existing artifact differs from passed key introduction/advanced_data_structures_and_frameworks_squidpy.zarr, keeping original key; update manually if needed or pass skip_hash_lookup if you want to duplicate the artifact\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "Artifact(uid='b1RDFMMW8H5hVbzP0000', version_tag=None, is_latest=True, key='introduction/data_structures_and_frameworks_advanced_squidpy.zarr', description='zarr for the squidpy section', suffix='.zarr', kind=None, otype='SpatialData', size=3956752851, hash='n9UcLZoCI-wRgv6EE99ykw', n_files=464, n_observations=None, branch_id=1, space_id=1, storage_id=1, run_id=15, schema_id=None, created_by_id=6, created_at=2025-05-24 20:02:06 UTC, is_locked=False)"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "af = ln.Artifact(\n",
    "    \"xenium.zarr\",\n",
    "    key=\"introduction/advanced_data_structures_and_frameworks_squidpy.zarr\",\n",
    "    description=\"zarr for the squidpy section\",\n",
    ").save()\n",
    "af"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "f48b95e6",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[94m•\u001b[0m please hit CMD + s to save the notebook in your editor .... still waiting .....\n",
      "....................\u001b[93m!\u001b[0m please hit CMD + s to save the notebook in your editor and re-run finish()\n"
     ]
    }
   ],
   "source": [
    "ln.finish()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "lamindb",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.14.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
