
<!DOCTYPE html>

<html lang="en">
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="generator" content="Docutils 0.17.1: http://docutils.sourceforge.net/" />

    <title>23. Single-cell data resolved in space &#8212; Single-cell best practices</title>
    
  <!-- Loaded before other Sphinx assets -->
  <link href="../_static/styles/theme.css?digest=1999514e3f237ded88cf" rel="stylesheet">
<link href="../_static/styles/pydata-sphinx-theme.css?digest=1999514e3f237ded88cf" rel="stylesheet">

    
  <link rel="stylesheet"
    href="../_static/vendor/fontawesome/5.13.0/css/all.min.css">
  <link rel="preload" as="font" type="font/woff2" crossorigin
    href="../_static/vendor/fontawesome/5.13.0/webfonts/fa-solid-900.woff2">
  <link rel="preload" as="font" type="font/woff2" crossorigin
    href="../_static/vendor/fontawesome/5.13.0/webfonts/fa-brands-400.woff2">

    <link rel="stylesheet" type="text/css" href="../_static/pygments.css" />
    <link rel="stylesheet" href="../_static/styles/sphinx-book-theme.css?digest=5115cc725059bd94278eecd172e13a965bf8f5a9" type="text/css" />
    <link rel="stylesheet" type="text/css" href="../_static/togglebutton.css" />
    <link rel="stylesheet" type="text/css" href="../_static/copybutton.css" />
    <link rel="stylesheet" type="text/css" href="../_static/mystnb.css" />
    <link rel="stylesheet" type="text/css" href="../_static/sphinx-thebe.css" />
    <link rel="stylesheet" type="text/css" href="../_static/book.css" />
    <link rel="stylesheet" type="text/css" href="../_static/design-style.b7bb847fb20b106c3d81b95245e65545.min.css" />
    
  <!-- Pre-loaded scripts that we'll load fully later -->
  <link rel="preload" as="script" href="../_static/scripts/pydata-sphinx-theme.js?digest=1999514e3f237ded88cf">

    <script data-url_root="../" id="documentation_options" src="../_static/documentation_options.js"></script>
    <script src="../_static/jquery.js"></script>
    <script src="../_static/underscore.js"></script>
    <script src="../_static/doctools.js"></script>
    <script src="../_static/clipboard.min.js"></script>
    <script src="../_static/copybutton.js"></script>
    <script src="../_static/scripts/sphinx-book-theme.js?digest=9c920249402e914e316237a7dbc6769907cce411"></script>
    <script>let toggleHintShow = 'Click to show';</script>
    <script>let toggleHintHide = 'Click to hide';</script>
    <script>let toggleOpenOnPrint = 'true';</script>
    <script src="../_static/togglebutton.js"></script>
    <script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
    <script src="../_static/book.js"></script>
    <script>var togglebuttonSelector = '.toggle, .admonition.dropdown, .tag_hide_input div.cell_input, .tag_hide-input div.cell_input, .tag_hide_output div.cell_output, .tag_hide-output div.cell_output, .tag_hide_cell.cell, .tag_hide-cell.cell';</script>
    <script src="../_static/design-tabs.js"></script>
    <script>const THEBE_JS_URL = "https://unpkg.com/thebe@0.8.2/lib/index.js"
const thebe_selector = ".thebe,.cell"
const thebe_selector_input = "pre"
const thebe_selector_output = ".output, .cell_output"
</script>
    <script async="async" src="../_static/sphinx-thebe.js"></script>
    <link rel="shortcut icon" href="../_static/favicon.ico"/>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="24. Neighborhood analysis" href="neighborhood.html" />
    <link rel="prev" title="22. Bulk deconvolution" href="../deconvolution/bulk_deconvolution.html" />
    <meta name="viewport" content="width=device-width, initial-scale=1" />
    <meta name="docsearch:language" content="en">
    

    <!-- Google Analytics -->
    
  </head>
  <body data-spy="scroll" data-target="#bd-toc-nav" data-offset="60">
<!-- Checkboxes to toggle the left sidebar -->
<input type="checkbox" class="sidebar-toggle" name="__navigation" id="__navigation" aria-label="Toggle navigation sidebar">
<label class="overlay overlay-navbar" for="__navigation">
    <div class="visually-hidden">Toggle navigation sidebar</div>
</label>
<!-- Checkboxes to toggle the in-page toc -->
<input type="checkbox" class="sidebar-toggle" name="__page-toc" id="__page-toc" aria-label="Toggle in-page Table of Contents">
<label class="overlay overlay-pagetoc" for="__page-toc">
    <div class="visually-hidden">Toggle in-page Table of Contents</div>
</label>
<!-- Headers at the top -->
<div class="announcement header-item noprint"></div>
<div class="header header-item noprint"></div>

    
    <div class="container-fluid" id="banner"></div>

    

    <div class="container-xl">
      <div class="row">
          
<!-- Sidebar -->
<div class="bd-sidebar noprint" id="site-navigation">
    <div class="bd-sidebar__content">
        <div class="bd-sidebar__top"><div class="navbar-brand-box">
    <a class="navbar-brand text-wrap" href="../index.html">
      
        <!-- `logo` is deprecated in Sphinx 4.0, so remove this when we stop supporting 3 -->
        
      
      
      <img src="../_static/logo.png" class="logo" alt="logo">
      
      
      <h1 class="site-logo" id="site-title">Single-cell best practices</h1>
      
    </a>
</div><form class="bd-search d-flex align-items-center" action="../search.html" method="get">
  <i class="icon fas fa-search"></i>
  <input type="search" class="form-control" name="q" id="search-input" placeholder="Search this book..." aria-label="Search this book..." autocomplete="off" >
</form><nav class="bd-links" id="bd-docs-nav" aria-label="Main">
    <div class="bd-toc-item active">
        <p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Introduction
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../introduction/prior_art.html">
   1. Prior art
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../introduction/scrna_seq.html">
   2. Single-cell RNA sequencing
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../introduction/raw_data_processing.html">
   3. Raw data processing
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../introduction/analysis_tools.html">
   4. Analysis frameworks and tools
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../introduction/interoperability.html">
   5. Interoperability
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Preprocessing and visualization
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../preprocessing_visualization/quality_control.html">
   6. Quality Control
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../preprocessing_visualization/normalization.html">
   7. Normalization
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../preprocessing_visualization/feature_selection.html">
   8. Feature selection
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../preprocessing_visualization/dimensionality_reduction.html">
   9. Dimensionality Reduction
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Identifying cellular structure
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../cellular_structure/clustering.html">
   10. Clustering
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../cellular_structure/annotation.html">
   11. Annotation
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../cellular_structure/integration.html">
   12. Data integration
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Inferring trajectories
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../trajectories/pseudotemporal.html">
   13. Pseudotemporal ordering
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../trajectories/rna_velocity.html">
   14. RNA velocity
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../trajectories/lineage_tracing.html">
   15. Lineage tracing
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Dealing with conditions
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../conditions/differential_gene_expression.html">
   16. Differential gene expression analysis
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../conditions/compositional.html">
   17. Compositional analysis
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../conditions/gsea_pathway.html">
   18. Gene set enrichment and pathway analysis
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../conditions/perturbation_modeling.html">
   19. Perturbation modeling
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Modeling mechanisms
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../mechanisms/grns.html">
   20. Gene regulatory networks
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../mechanisms/cell_cell_communication.html">
   21. Cell-cell communication
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Deconvolution
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../deconvolution/bulk_deconvolution.html">
   22. Bulk deconvolution
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Spatial omics
 </span>
</p>
<ul class="current nav bd-sidenav">
 <li class="toctree-l1 current active">
  <a class="current reference internal" href="#">
   23. Single-cell data resolved in space
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="neighborhood.html">
   24. Neighborhood analysis
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="domains.html">
   25. Spatial domains
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="spatially_variable_genes.html">
   26. Spatially variable genes
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="deconvolution.html">
   27. Spatial deconvolution
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="imputation.html">
   28. Imputation
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Surface protein
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../surface_protein/quality_control.html">
   32. Quality control
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../surface_protein/normalization.html">
   33. Normalization
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../surface_protein/doublet_detection.html">
   34. Doublet detection
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../surface_protein/dimensionality_reduction.html">
   35. Dimensionality Reduction
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../surface_protein/batch_correction.html">
   36. Batch correction
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../surface_protein/annotation.html">
   37. Annotation
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  AIR repertoire
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../air_repertoire/ir_profiling.html">
   38. Immune Receptor Profiling
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../air_repertoire/clonotype.html">
   39. Clonotype analysis
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../air_repertoire/specificity.html">
   43. Specificity analysis
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../air_repertoire/multimodal_integration.html">
   44. Integrating AIR and transcriptomics
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Multimodal integration
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../multimodal_integration/paired_integration.html">
   45. Paired integration
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../multimodal_integration/advanced_integration.html">
   46. Advanced integration
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Reproducibility
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../reproducibility/introduction.html">
   47. Reproducibility
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Outlook
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../outlook.html">
   48. Outlook
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Acknowledgements
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../acknowledgements.html">
   49. Acknowledgements
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Glossary
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../glossary.html">
   50. Glossary
  </a>
 </li>
</ul>

    </div>
</nav></div>
        <div class="bd-sidebar__bottom">
             <!-- To handle the deprecated key -->
            
            <div class="navbar_extra_footer">
            Powered by <a href="https://jupyterbook.org">Jupyter Book</a>
            </div>
            
        </div>
    </div>
    <div id="rtd-footer-container"></div>
</div>


          


          
<!-- A tiny helper pixel to detect if we've scrolled -->
<div class="sbt-scroll-pixel-helper"></div>
<!-- Main content -->
<div class="col py-0 content-container">
    
    <div class="header-article row sticky-top noprint">
        



<div class="col py-1 d-flex header-article-main">
    <div class="header-article__left">
        
        <label for="__navigation"
  class="headerbtn"
  data-toggle="tooltip"
data-placement="right"
title="Toggle navigation"
>
  

<span class="headerbtn__icon-container">
  <i class="fas fa-bars"></i>
  </span>

</label>

        
    </div>
    <div class="header-article__right">
<div class="menu-dropdown menu-dropdown-launch-buttons">
  <button class="headerbtn menu-dropdown__trigger"
      aria-label="Launch interactive content">
      <i class="fas fa-rocket"></i>
  </button>
  <div class="menu-dropdown__content">
    <ul>
      <li>
        <a href="https://mybinder.org/v2/gh/theislab/single-cell-best-practices/master?urlpath=tree/spatial/introduction.ipynb"
   class="headerbtn"
   data-toggle="tooltip"
data-placement="left"
title="Launch on Binder"
>
  

<span class="headerbtn__icon-container">
  
    <img src="../_static/images/logo_binder.svg">
  </span>
<span class="headerbtn__text-container">Binder</span>
</a>

      </li>
      
    </ul>
  </div>
</div>

<button onclick="toggleFullScreen()"
  class="headerbtn"
  data-toggle="tooltip"
data-placement="bottom"
title="Fullscreen mode"
>
  

<span class="headerbtn__icon-container">
  <i class="fas fa-expand"></i>
  </span>

</button>

<div class="menu-dropdown menu-dropdown-repository-buttons">
  <button class="headerbtn menu-dropdown__trigger"
      aria-label="Source repositories">
      <i class="fab fa-github"></i>
  </button>
  <div class="menu-dropdown__content">
    <ul>
      <li>
        <a href="https://github.com/theislab/single-cell-best-practices"
   class="headerbtn"
   data-toggle="tooltip"
data-placement="left"
title="Source repository"
>
  

<span class="headerbtn__icon-container">
  <i class="fab fa-github"></i>
  </span>
<span class="headerbtn__text-container">repository</span>
</a>

      </li>
      
      <li>
        <a href="https://github.com/theislab/single-cell-best-practices/issues/new?title=Issue%20on%20page%20%2Fspatial/introduction.html&body=Your%20issue%20content%20here."
   class="headerbtn"
   data-toggle="tooltip"
data-placement="left"
title="Open an issue"
>
  

<span class="headerbtn__icon-container">
  <i class="fas fa-lightbulb"></i>
  </span>
<span class="headerbtn__text-container">open issue</span>
</a>

      </li>
      
      <li>
        <a href="https://github.com/theislab/single-cell-best-practices/edit/master/spatial/introduction.ipynb"
   class="headerbtn"
   data-toggle="tooltip"
data-placement="left"
title="Edit this page"
>
  

<span class="headerbtn__icon-container">
  <i class="fas fa-pencil-alt"></i>
  </span>
<span class="headerbtn__text-container">suggest edit</span>
</a>

      </li>
      
    </ul>
  </div>
</div>

<div class="menu-dropdown menu-dropdown-download-buttons">
  <button class="headerbtn menu-dropdown__trigger"
      aria-label="Download this page">
      <i class="fas fa-download"></i>
  </button>
  <div class="menu-dropdown__content">
    <ul>
      <li>
        <a href="../_sources/spatial/introduction.ipynb"
   class="headerbtn"
   data-toggle="tooltip"
data-placement="left"
title="Download source file"
>
  

<span class="headerbtn__icon-container">
  <i class="fas fa-file"></i>
  </span>
<span class="headerbtn__text-container">.ipynb</span>
</a>

      </li>
      
      <li>
        
<button onclick="printPdf(this)"
  class="headerbtn"
  data-toggle="tooltip"
data-placement="left"
title="Print to PDF"
>
  

<span class="headerbtn__icon-container">
  <i class="fas fa-file-pdf"></i>
  </span>
<span class="headerbtn__text-container">.pdf</span>
</button>

      </li>
      
    </ul>
  </div>
</div>
<label for="__page-toc"
  class="headerbtn headerbtn-page-toc"
  
>
  

<span class="headerbtn__icon-container">
  <i class="fas fa-list"></i>
  </span>

</label>

    </div>
</div>

<!-- Table of contents -->
<div class="col-md-3 bd-toc show noprint">
    <div class="tocsection onthispage pt-5 pb-3">
        <i class="fas fa-list"></i> Contents
    </div>
    <nav id="bd-toc-nav" aria-label="Page">
        <ul class="visible nav section-nav flex-column">
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#motivation">
   23.1. Motivation
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#overview-of-spatial-profiling-measurements">
   23.2. Overview of spatial profiling measurements
  </a>
  <ul class="nav section-nav flex-column">
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#multi-cell-resolution">
     23.2.1. Multi-cell resolution
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#single-cell-resolution">
     23.2.2. Single-cell resolution
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#sub-cellular-resolution">
     23.2.3. Sub-cellular resolution
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#recommended-reading">
     23.2.4. Recommended reading
    </a>
   </li>
  </ul>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#references">
   23.3. References
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#contributors">
   23.4. Contributors
  </a>
  <ul class="nav section-nav flex-column">
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#authors">
     23.4.1. Authors
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#reviewers">
     23.4.2. Reviewers
    </a>
   </li>
  </ul>
 </li>
</ul>

    </nav>
</div>
    </div>
    <div class="article row">
        <div class="col pl-md-3 pl-lg-5 content-container">
            <!-- Table of contents that is only displayed when printing the page -->
            <div id="jb-print-docs-body" class="onlyprint">
                <h1>Single-cell data resolved in space</h1>
                <!-- Table of contents -->
                <div id="print-main-content">
                    <div id="jb-print-toc">
                        
                        <div>
                            <h2> Contents </h2>
                        </div>
                        <nav aria-label="Page">
                            <ul class="visible nav section-nav flex-column">
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#motivation">
   23.1. Motivation
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#overview-of-spatial-profiling-measurements">
   23.2. Overview of spatial profiling measurements
  </a>
  <ul class="nav section-nav flex-column">
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#multi-cell-resolution">
     23.2.1. Multi-cell resolution
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#single-cell-resolution">
     23.2.2. Single-cell resolution
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#sub-cellular-resolution">
     23.2.3. Sub-cellular resolution
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#recommended-reading">
     23.2.4. Recommended reading
    </a>
   </li>
  </ul>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#references">
   23.3. References
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#contributors">
   23.4. Contributors
  </a>
  <ul class="nav section-nav flex-column">
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#authors">
     23.4.1. Authors
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#reviewers">
     23.4.2. Reviewers
    </a>
   </li>
  </ul>
 </li>
</ul>

                        </nav>
                    </div>
                </div>
            </div>
            <main id="main-content" role="main">
                
              <div>
                
  <section class="tex2jax_ignore mathjax_ignore" id="single-cell-data-resolved-in-space">
<h1><span class="section-number">23. </span>Single-cell data resolved in space<a class="headerlink" href="#single-cell-data-resolved-in-space" title="Permalink to this headline">#</a></h1>
<section id="motivation">
<h2><span class="section-number">23.1. </span>Motivation<a class="headerlink" href="#motivation" title="Permalink to this headline">#</a></h2>
<p>Bulk and single-cell genomic technologies enabled us to characterize and understand cellular identities and their dependencies on genome scale. Up to this point, all single-cell technologies described in this work characterized dissociated cells that were removed from the sample of interest. However, by removing the spatial context of the cells and their molecules, one also loses the spatial context which is an essential component for many biological questions. Spatially resolved genomics resolve this issue by measuring both genome-scale omics and preserving the spatial information.</p>
</section>
<section id="overview-of-spatial-profiling-measurements">
<h2><span class="section-number">23.2. </span>Overview of spatial profiling measurements<a class="headerlink" href="#overview-of-spatial-profiling-measurements" title="Permalink to this headline">#</a></h2>
<p>Spatially resolved genomics can be measured with various technologies that quantify the transcriptome, proteome or chromatin accessibility. However, these technologies vary in terms of scale, resolution, sensitivity, multiplexing and applicability. As the landscape of spatially resolved genomics is developing fast and is expected to change massively in the next few years, we are introducing only three overall groups of spatial omics technologies based on their capturing resolution. Additional information can be obtained through the linked reviews and papers following this introduction. <span id="id1">[<a class="reference internal" href="#id483" title="Jeffrey R. Moffitt, Emma Lundberg, and Holger Heyn. The emerging landscape of spatial profiling technologies. Nature Reviews Genetics, 23(12):741–759, December 2022. URL: https://doi.org/10.1038/s41576-022-00515-3, doi:10.1038/s41576-022-00515-3.">Moffitt <em>et al.</em>, 2022</a>]</span> <span id="id2">[<a class="reference internal" href="spatially_variable_genes.html#id484" title="Benjamin L. Walker, Zixuan Cang, Honglei Ren, Eric Bourgain-Chang, and Qing Nie. Deciphering tissue structure and function using spatial transcriptomics. Communications Biology, 5(1):220, March 2022. URL: https://doi.org/10.1038/s42003-022-03175-5, doi:10.1038/s42003-022-03175-5.">Walker <em>et al.</em>, 2022</a>]</span> <span id="id3">[<a class="reference internal" href="#id482" title="Liangchen Yue, Feng Liu, Jiongsong Hu, Pin Yang, Yuxiang Wang, Junguo Dong, Wenjie Shu, Xingxu Huang, and Shengqi Wang. A guidebook of spatial transcriptomic technologies, data resources and analysis approaches. Computational and Structural Biotechnology Journal, 21:940–955, 2023. URL: https://www.sciencedirect.com/science/article/pii/S2001037023000156, doi:https://doi.org/10.1016/j.csbj.2023.01.016.">Yue <em>et al.</em>, 2023</a>]</span></p>
<figure class="align-default" id="resolutions">
<a class="bg-primary mb-1 reference internal image-reference" href="../_images/introduction_1.jpeg"><img alt="Spatial Omics Resolutions Overview" class="bg-primary mb-1" src="../_images/introduction_1.jpeg" style="width: 800px;" /></a>
<figcaption>
<p><span class="caption-number">Fig. 23.1 </span><span class="caption-text">Spatially resolved genomics data varies in resolution from multi-cell, single-cell to sub-cellular resolution.</span><a class="headerlink" href="#resolutions" title="Permalink to this image">#</a></p>
</figcaption>
</figure>
<p>Broadly spoken, one can differentiate spatially resolved genomics into technologies that measure at multi-cell, single-cell and sub-cellular resolution. We will shortly motivate each of those scales and highlight their advantages and challenges and introduce the reader to a few technologies that fall into the respective category.</p>
<section id="multi-cell-resolution">
<h3><span class="section-number">23.2.1. </span>Multi-cell resolution<a class="headerlink" href="#multi-cell-resolution" title="Permalink to this headline">#</a></h3>
<p>Spatial omics data obtained at multi-cell resolution typically captures omics measurements among several cells. So, each datapoint contains information from a varying number of cells and also potentially different cell types. Multi-cell resolution data can be decomposed with deconvolution methods to obtain proportions of different cells or cell types per spot.</p>
<p>Multi-cell resolution data typically capture transcriptome-wide gene expression profiles at varying resolution. The obtained resolution for spot-based technologies varies between 55um (Visium) to 10 um (slide-seq). A widely known, commercially available and successful technology is Visium provided by 10x Genomics. We will showcase how to analyze and deconvolve Visium data in the tutorials.</p>
</section>
<section id="single-cell-resolution">
<h3><span class="section-number">23.2.2. </span>Single-cell resolution<a class="headerlink" href="#single-cell-resolution" title="Permalink to this headline">#</a></h3>
<p>Spatial omics data obtained at single-cell resolution either directly capture single cells at their exact position or capture spots on the scale of single-cell. Examples for spot-based methods are HDST, slide-seqV2 or stero-seq. These methods capture the whole transcriptome but still have a low capture efficiency.</p>
<p>Targeted methods provide an alternative for measuring cells at their exact position. Common examples are MERFISH, seqFISH+, IMC or multiplexed IHC (e.g. cyCIF and CODEX). These technologies are usually expensive and only measure a limited features space. These methods do not capture spots at a predefined location or grid, but measure individual transcript or cellular locations.</p>
</section>
<section id="sub-cellular-resolution">
<h3><span class="section-number">23.2.3. </span>Sub-cellular resolution<a class="headerlink" href="#sub-cellular-resolution" title="Permalink to this headline">#</a></h3>
<p>Spatial omics data at sub-cellular resolution captures the position of individual RNA molecules. Those positions can either be captured through single-molecule imaging or via spatial barcoding with spot-sizes smaller than single-cells. By performing cell segmentation on sub-cellular data one can obtain single-cell resolution data where expression is aggregated to cell-wise measurements which than can be processed in spatially-aware analysis pipelines. We will show how to perform this analysis for a MERFISH dataset.</p>
</section>
<section id="recommended-reading">
<h3><span class="section-number">23.2.4. </span>Recommended reading<a class="headerlink" href="#recommended-reading" title="Permalink to this headline">#</a></h3>
<p>To get a more elaborate understanding of the differences between the spatial experimental assays we recommend the following papers:</p>
<ul class="simple">
<li><p>The emerging landscape of spatial profiling technologies <span id="id4">[<a class="reference internal" href="#id483" title="Jeffrey R. Moffitt, Emma Lundberg, and Holger Heyn. The emerging landscape of spatial profiling technologies. Nature Reviews Genetics, 23(12):741–759, December 2022. URL: https://doi.org/10.1038/s41576-022-00515-3, doi:10.1038/s41576-022-00515-3.">Moffitt <em>et al.</em>, 2022</a>]</span></p></li>
<li><p>Deciphering tissue structure and function using spatial transcriptomics <span id="id5">[<a class="reference internal" href="spatially_variable_genes.html#id484" title="Benjamin L. Walker, Zixuan Cang, Honglei Ren, Eric Bourgain-Chang, and Qing Nie. Deciphering tissue structure and function using spatial transcriptomics. Communications Biology, 5(1):220, March 2022. URL: https://doi.org/10.1038/s42003-022-03175-5, doi:10.1038/s42003-022-03175-5.">Walker <em>et al.</em>, 2022</a>]</span></p></li>
<li><p>A guidebook of spatial transcriptomic technologies, data resources and analysis approaches <span id="id6">[<a class="reference internal" href="#id482" title="Liangchen Yue, Feng Liu, Jiongsong Hu, Pin Yang, Yuxiang Wang, Junguo Dong, Wenjie Shu, Xingxu Huang, and Shengqi Wang. A guidebook of spatial transcriptomic technologies, data resources and analysis approaches. Computational and Structural Biotechnology Journal, 21:940–955, 2023. URL: https://www.sciencedirect.com/science/article/pii/S2001037023000156, doi:https://doi.org/10.1016/j.csbj.2023.01.016.">Yue <em>et al.</em>, 2023</a>]</span></p></li>
</ul>
</section>
</section>
<section id="references">
<h2><span class="section-number">23.3. </span>References<a class="headerlink" href="#references" title="Permalink to this headline">#</a></h2>
<div class="docutils container" id="id7">
<dl class="citation">
<dt class="label" id="id483"><span class="brackets">spatialMLH22</span><span class="fn-backref">(<a href="#id1">1</a>,<a href="#id4">2</a>)</span></dt>
<dd><p>Jeffrey R. Moffitt, Emma Lundberg, and Holger Heyn. The emerging landscape of spatial profiling technologies. <em>Nature Reviews Genetics</em>, 23(12):741–759, December 2022. URL: <a class="reference external" href="https://doi.org/10.1038/s41576-022-00515-3">https://doi.org/10.1038/s41576-022-00515-3</a>, <a class="reference external" href="https://doi.org/10.1038/s41576-022-00515-3">doi:10.1038/s41576-022-00515-3</a>.</p>
</dd>
<dt class="label" id="id481"><span class="brackets">spatialWCR+22</span><span class="fn-backref">(<a href="#id2">1</a>,<a href="#id5">2</a>)</span></dt>
<dd><p>Benjamin L. Walker, Zixuan Cang, Honglei Ren, Eric Bourgain-Chang, and Qing Nie. Deciphering tissue structure and function using spatial transcriptomics. <em>Communications Biology</em>, 5(1):220, March 2022. URL: <a class="reference external" href="https://doi.org/10.1038/s42003-022-03175-5">https://doi.org/10.1038/s42003-022-03175-5</a>, <a class="reference external" href="https://doi.org/10.1038/s42003-022-03175-5">doi:10.1038/s42003-022-03175-5</a>.</p>
</dd>
<dt class="label" id="id482"><span class="brackets">spatialYLH+23</span><span class="fn-backref">(<a href="#id3">1</a>,<a href="#id6">2</a>)</span></dt>
<dd><p>Liangchen Yue, Feng Liu, Jiongsong Hu, Pin Yang, Yuxiang Wang, Junguo Dong, Wenjie Shu, Xingxu Huang, and Shengqi Wang. A guidebook of spatial transcriptomic technologies, data resources and analysis approaches. <em>Computational and Structural Biotechnology Journal</em>, 21:940–955, 2023. URL: <a class="reference external" href="https://www.sciencedirect.com/science/article/pii/S2001037023000156">https://www.sciencedirect.com/science/article/pii/S2001037023000156</a>, <a class="reference external" href="https://doi.org/https://doi.org/10.1016/j.csbj.2023.01.016">doi:https://doi.org/10.1016/j.csbj.2023.01.016</a>.</p>
</dd>
</dl>
</div>
</section>
<section id="contributors">
<h2><span class="section-number">23.4. </span>Contributors<a class="headerlink" href="#contributors" title="Permalink to this headline">#</a></h2>
<section id="authors">
<h3><span class="section-number">23.4.1. </span>Authors<a class="headerlink" href="#authors" title="Permalink to this headline">#</a></h3>
<ul class="simple">
<li><p>Anna Schaar</p></li>
</ul>
</section>
<section id="reviewers">
<h3><span class="section-number">23.4.2. </span>Reviewers<a class="headerlink" href="#reviewers" title="Permalink to this headline">#</a></h3>
<ul class="simple">
<li><p>Lukas Heumos</p></li>
</ul>
</section>
</section>
</section>

    <script type="text/x-thebe-config">
    {
        requestKernel: true,
        binderOptions: {
            repo: "binder-examples/jupyter-stacks-datascience",
            ref: "master",
        },
        codeMirrorConfig: {
            theme: "abcdef",
            mode: "cell2loc_env"
        },
        kernelOptions: {
            name: "cell2loc_env",
            path: "./spatial"
        },
        predefinedOutput: true
    }
    </script>
    <script>kernelName = 'cell2loc_env'</script>

              </div>
              
            </main>
            <footer class="footer-article noprint">
                
    <!-- Previous / next buttons -->
<div class='prev-next-area'>
    <a class='left-prev' id="prev-link" href="../deconvolution/bulk_deconvolution.html" title="previous page">
        <i class="fas fa-angle-left"></i>
        <div class="prev-next-info">
            <p class="prev-next-subtitle">previous</p>
            <p class="prev-next-title"><span class="section-number">22. </span>Bulk deconvolution</p>
        </div>
    </a>
    <a class='right-next' id="next-link" href="neighborhood.html" title="next page">
    <div class="prev-next-info">
        <p class="prev-next-subtitle">next</p>
        <p class="prev-next-title"><span class="section-number">24. </span>Neighborhood analysis</p>
    </div>
    <i class="fas fa-angle-right"></i>
    </a>
</div>
            </footer>
        </div>
    </div>
    <div class="footer-content row">
        <footer class="col footer"><p>
  
    By Lukas Heumos, Anna Schaar<br/>
  
      &copy; Copyright 2022.<br/>
    <div class="extra_footer">
      <div>
  Brought to you by
  <a href="https://www.github.com/theislab">Theislab</a>, with many thanks to the single-cell community as a whole!
</div>

    </div>
</p>
        </footer>
    </div>
    
</div>


      </div>
    </div>
  
  <!-- Scripts loaded after <body> so the DOM is not blocked -->
  <script src="../_static/scripts/pydata-sphinx-theme.js?digest=1999514e3f237ded88cf"></script>


  </body>
</html>