
<!DOCTYPE html>

<html lang="en">
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="generator" content="Docutils 0.17.1: http://docutils.sourceforge.net/" />

    <title>27. Neighborhood analysis &#8212; Single-cell best practices</title>
    
  <!-- Loaded before other Sphinx assets -->
  <link href="../_static/styles/theme.css?digest=1999514e3f237ded88cf" rel="stylesheet">
<link href="../_static/styles/pydata-sphinx-theme.css?digest=1999514e3f237ded88cf" rel="stylesheet">

    
  <link rel="stylesheet"
    href="../_static/vendor/fontawesome/5.13.0/css/all.min.css">
  <link rel="preload" as="font" type="font/woff2" crossorigin
    href="../_static/vendor/fontawesome/5.13.0/webfonts/fa-solid-900.woff2">
  <link rel="preload" as="font" type="font/woff2" crossorigin
    href="../_static/vendor/fontawesome/5.13.0/webfonts/fa-brands-400.woff2">

    <link rel="stylesheet" type="text/css" href="../_static/pygments.css" />
    <link rel="stylesheet" href="../_static/styles/sphinx-book-theme.css?digest=5115cc725059bd94278eecd172e13a965bf8f5a9" type="text/css" />
    <link rel="stylesheet" type="text/css" href="../_static/togglebutton.css" />
    <link rel="stylesheet" type="text/css" href="../_static/copybutton.css" />
    <link rel="stylesheet" type="text/css" href="../_static/mystnb.css" />
    <link rel="stylesheet" type="text/css" href="../_static/sphinx-thebe.css" />
    <link rel="stylesheet" type="text/css" href="../_static/book.css" />
    <link rel="stylesheet" type="text/css" href="../_static/design-style.b7bb847fb20b106c3d81b95245e65545.min.css" />
    
  <!-- Pre-loaded scripts that we'll load fully later -->
  <link rel="preload" as="script" href="../_static/scripts/pydata-sphinx-theme.js?digest=1999514e3f237ded88cf">

    <script data-url_root="../" id="documentation_options" src="../_static/documentation_options.js"></script>
    <script src="../_static/jquery.js"></script>
    <script src="../_static/underscore.js"></script>
    <script src="../_static/doctools.js"></script>
    <script src="../_static/clipboard.min.js"></script>
    <script src="../_static/copybutton.js"></script>
    <script src="../_static/scripts/sphinx-book-theme.js?digest=9c920249402e914e316237a7dbc6769907cce411"></script>
    <script>let toggleHintShow = 'Click to show';</script>
    <script>let toggleHintHide = 'Click to hide';</script>
    <script>let toggleOpenOnPrint = 'true';</script>
    <script src="../_static/togglebutton.js"></script>
    <script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
    <script src="../_static/book.js"></script>
    <script>var togglebuttonSelector = '.toggle, .admonition.dropdown, .tag_hide_input div.cell_input, .tag_hide-input div.cell_input, .tag_hide_output div.cell_output, .tag_hide-output div.cell_output, .tag_hide_cell.cell, .tag_hide-cell.cell';</script>
    <script src="../_static/design-tabs.js"></script>
    <script>const THEBE_JS_URL = "https://unpkg.com/thebe@0.8.2/lib/index.js"
const thebe_selector = ".thebe,.cell"
const thebe_selector_input = "pre"
const thebe_selector_output = ".output, .cell_output"
</script>
    <script async="async" src="../_static/sphinx-thebe.js"></script>
    <script>window.MathJax = {"options": {"processHtmlClass": "tex2jax_process|mathjax_process|math|output_area"}}</script>
    <script defer="defer" src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-mml-chtml.js"></script>
    <link rel="shortcut icon" href="../_static/favicon.ico"/>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="28. Spatial domains" href="domains.html" />
    <link rel="prev" title="26. Single-cell data resolved in space" href="introduction.html" />
    <meta name="viewport" content="width=device-width, initial-scale=1" />
    <meta name="docsearch:language" content="en">
    

    <!-- Google Analytics -->
    
  </head>
  <body data-spy="scroll" data-target="#bd-toc-nav" data-offset="60">
<!-- Checkboxes to toggle the left sidebar -->
<input type="checkbox" class="sidebar-toggle" name="__navigation" id="__navigation" aria-label="Toggle navigation sidebar">
<label class="overlay overlay-navbar" for="__navigation">
    <div class="visually-hidden">Toggle navigation sidebar</div>
</label>
<!-- Checkboxes to toggle the in-page toc -->
<input type="checkbox" class="sidebar-toggle" name="__page-toc" id="__page-toc" aria-label="Toggle in-page Table of Contents">
<label class="overlay overlay-pagetoc" for="__page-toc">
    <div class="visually-hidden">Toggle in-page Table of Contents</div>
</label>
<!-- Headers at the top -->
<div class="announcement header-item noprint"></div>
<div class="header header-item noprint"></div>

    
    <div class="container-fluid" id="banner"></div>

    

    <div class="container-xl">
      <div class="row">
          
<!-- Sidebar -->
<div class="bd-sidebar noprint" id="site-navigation">
    <div class="bd-sidebar__content">
        <div class="bd-sidebar__top"><div class="navbar-brand-box">
    <a class="navbar-brand text-wrap" href="../index.html">
      
        <!-- `logo` is deprecated in Sphinx 4.0, so remove this when we stop supporting 3 -->
        
      
      
      <img src="../_static/logo.png" class="logo" alt="logo">
      
      
      <h1 class="site-logo" id="site-title">Single-cell best practices</h1>
      
    </a>
</div><form class="bd-search d-flex align-items-center" action="../search.html" method="get">
  <i class="icon fas fa-search"></i>
  <input type="search" class="form-control" name="q" id="search-input" placeholder="Search this book..." aria-label="Search this book..." autocomplete="off" >
</form><nav class="bd-links" id="bd-docs-nav" aria-label="Main">
    <div class="bd-toc-item active">
        <p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Introduction
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../introduction/prior_art.html">
   1. Prior art
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../introduction/scrna_seq.html">
   2. Single-cell RNA sequencing
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../introduction/raw_data_processing.html">
   3. Raw data processing
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../introduction/analysis_tools.html">
   4. Analysis frameworks and tools
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../introduction/interoperability.html">
   5. Interoperability
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Preprocessing and visualization
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../preprocessing_visualization/quality_control.html">
   6. Quality Control
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../preprocessing_visualization/normalization.html">
   7. Normalization
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../preprocessing_visualization/feature_selection.html">
   8. Feature selection
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../preprocessing_visualization/dimensionality_reduction.html">
   9. Dimensionality Reduction
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Identifying cellular structure
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../cellular_structure/clustering.html">
   10. Clustering
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../cellular_structure/annotation.html">
   11. Annotation
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../cellular_structure/integration.html">
   12. Data integration
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Inferring trajectories
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../trajectories/pseudotemporal.html">
   13. Pseudotemporal ordering
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../trajectories/rna_velocity.html">
   14. RNA velocity
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../trajectories/lineage_tracing.html">
   15. Lineage tracing
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Dealing with conditions
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../conditions/differential_gene_expression.html">
   16. Differential gene expression analysis
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../conditions/compositional.html">
   17. Compositional analysis
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../conditions/gsea_pathway.html">
   18. Gene set enrichment and pathway analysis
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../conditions/perturbation_modeling.html">
   19. Perturbation modeling
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Modeling mechanisms
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../mechanisms/gene_regulatory_networks.html">
   20. Gene regulatory networks
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../mechanisms/cell_cell_communication.html">
   21. Cell-cell communication
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Deconvolution
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../deconvolution/bulk_deconvolution.html">
   22. Bulk deconvolution
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Chromatin Accessibility
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../chromatin_accessibility/introduction.html">
   23. Single-cell ATAC sequencing
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../chromatin_accessibility/quality_control.html">
   24. Quality Control
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../chromatin_accessibility/gene_regulatory_networks_atac.html">
   25. Gene regulatory networks using chromatin accessibility
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Spatial omics
 </span>
</p>
<ul class="current nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="introduction.html">
   26. Single-cell data resolved in space
  </a>
 </li>
 <li class="toctree-l1 current active">
  <a class="current reference internal" href="#">
   27. Neighborhood analysis
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="domains.html">
   28. Spatial domains
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="spatially_variable_genes.html">
   29. Spatially variable genes
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="deconvolution.html">
   30. Spatial deconvolution
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="imputation.html">
   31. Imputation
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Surface protein
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../surface_protein/quality_control.html">
   32. Quality control
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../surface_protein/normalization.html">
   33. Normalization
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../surface_protein/doublet_detection.html">
   34. Doublet detection
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../surface_protein/dimensionality_reduction.html">
   35. Dimensionality Reduction
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../surface_protein/batch_correction.html">
   36. Batch correction
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../surface_protein/annotation.html">
   37. Annotation
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Adaptive immune receptor repertoire
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../air_repertoire/ir_profiling.html">
   38. Immune Receptor Profiling
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../air_repertoire/clonotype.html">
   39. Clonotype analysis
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../air_repertoire/specificity.html">
   43. Specificity analysis
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../air_repertoire/multimodal_integration.html">
   44. Integrating AIR and transcriptomics
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Multimodal integration
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../multimodal_integration/paired_integration.html">
   45. Paired integration
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../multimodal_integration/advanced_integration.html">
   46. Advanced integration
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Reproducibility
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../reproducibility/introduction.html">
   47. Reproducibility
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Outlook
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../outlook.html">
   48. Outlook
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Acknowledgements
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../acknowledgements.html">
   49. Acknowledgements
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Glossary
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../glossary.html">
   50. Glossary
  </a>
 </li>
</ul>

    </div>
</nav></div>
        <div class="bd-sidebar__bottom">
             <!-- To handle the deprecated key -->
            
            <div class="navbar_extra_footer">
            Powered by <a href="https://jupyterbook.org">Jupyter Book</a>
            </div>
            
        </div>
    </div>
    <div id="rtd-footer-container"></div>
</div>


          


          
<!-- A tiny helper pixel to detect if we've scrolled -->
<div class="sbt-scroll-pixel-helper"></div>
<!-- Main content -->
<div class="col py-0 content-container">
    
    <div class="header-article row sticky-top noprint">
        



<div class="col py-1 d-flex header-article-main">
    <div class="header-article__left">
        
        <label for="__navigation"
  class="headerbtn"
  data-toggle="tooltip"
data-placement="right"
title="Toggle navigation"
>
  

<span class="headerbtn__icon-container">
  <i class="fas fa-bars"></i>
  </span>

</label>

        
    </div>
    <div class="header-article__right">
<div class="menu-dropdown menu-dropdown-launch-buttons">
  <button class="headerbtn menu-dropdown__trigger"
      aria-label="Launch interactive content">
      <i class="fas fa-rocket"></i>
  </button>
  <div class="menu-dropdown__content">
    <ul>
      <li>
        <a href="https://mybinder.org/v2/gh/theislab/single-cell-best-practices/master?urlpath=tree/jupyter-book/spatial/neighborhood.ipynb"
   class="headerbtn"
   data-toggle="tooltip"
data-placement="left"
title="Launch on Binder"
>
  

<span class="headerbtn__icon-container">
  
    <img src="../_static/images/logo_binder.svg">
  </span>
<span class="headerbtn__text-container">Binder</span>
</a>

      </li>
      
    </ul>
  </div>
</div>

<button onclick="toggleFullScreen()"
  class="headerbtn"
  data-toggle="tooltip"
data-placement="bottom"
title="Fullscreen mode"
>
  

<span class="headerbtn__icon-container">
  <i class="fas fa-expand"></i>
  </span>

</button>

<div class="menu-dropdown menu-dropdown-repository-buttons">
  <button class="headerbtn menu-dropdown__trigger"
      aria-label="Source repositories">
      <i class="fab fa-github"></i>
  </button>
  <div class="menu-dropdown__content">
    <ul>
      <li>
        <a href="https://github.com/theislab/single-cell-best-practices"
   class="headerbtn"
   data-toggle="tooltip"
data-placement="left"
title="Source repository"
>
  

<span class="headerbtn__icon-container">
  <i class="fab fa-github"></i>
  </span>
<span class="headerbtn__text-container">repository</span>
</a>

      </li>
      
      <li>
        <a href="https://github.com/theislab/single-cell-best-practices/issues/new?title=Issue%20on%20page%20%2Fspatial/neighborhood.html&body=Your%20issue%20content%20here."
   class="headerbtn"
   data-toggle="tooltip"
data-placement="left"
title="Open an issue"
>
  

<span class="headerbtn__icon-container">
  <i class="fas fa-lightbulb"></i>
  </span>
<span class="headerbtn__text-container">open issue</span>
</a>

      </li>
      
      <li>
        <a href="https://github.com/theislab/single-cell-best-practices/edit/master/jupyter-book/spatial/neighborhood.ipynb"
   class="headerbtn"
   data-toggle="tooltip"
data-placement="left"
title="Edit this page"
>
  

<span class="headerbtn__icon-container">
  <i class="fas fa-pencil-alt"></i>
  </span>
<span class="headerbtn__text-container">suggest edit</span>
</a>

      </li>
      
    </ul>
  </div>
</div>

<div class="menu-dropdown menu-dropdown-download-buttons">
  <button class="headerbtn menu-dropdown__trigger"
      aria-label="Download this page">
      <i class="fas fa-download"></i>
  </button>
  <div class="menu-dropdown__content">
    <ul>
      <li>
        <a href="../_sources/spatial/neighborhood.ipynb"
   class="headerbtn"
   data-toggle="tooltip"
data-placement="left"
title="Download source file"
>
  

<span class="headerbtn__icon-container">
  <i class="fas fa-file"></i>
  </span>
<span class="headerbtn__text-container">.ipynb</span>
</a>

      </li>
      
      <li>
        
<button onclick="printPdf(this)"
  class="headerbtn"
  data-toggle="tooltip"
data-placement="left"
title="Print to PDF"
>
  

<span class="headerbtn__icon-container">
  <i class="fas fa-file-pdf"></i>
  </span>
<span class="headerbtn__text-container">.pdf</span>
</button>

      </li>
      
    </ul>
  </div>
</div>
<label for="__page-toc"
  class="headerbtn headerbtn-page-toc"
  
>
  

<span class="headerbtn__icon-container">
  <i class="fas fa-list"></i>
  </span>

</label>

    </div>
</div>

<!-- Table of contents -->
<div class="col-md-3 bd-toc show noprint">
    <div class="tocsection onthispage pt-5 pb-3">
        <i class="fas fa-list"></i> Contents
    </div>
    <nav id="bd-toc-nav" aria-label="Page">
        <ul class="visible nav section-nav flex-column">
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#motivation">
   27.1. Motivation
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#environment-setup-and-data">
   27.2. Environment setup and data
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#identifying-interactions-between-spatial-communities">
   27.3. Identifying interactions between spatial communities
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#co-occurrence-across-spatial-dimensions">
   27.4. Co-occurrence across spatial dimensions
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#key-takeaways">
   27.5. Key takeaways
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#references">
   27.6. References
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#contributors">
   27.7. Contributors
  </a>
  <ul class="nav section-nav flex-column">
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#authors">
     27.7.1. Authors
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#reviewers">
     27.7.2. Reviewers
    </a>
   </li>
  </ul>
 </li>
</ul>

    </nav>
</div>
    </div>
    <div class="article row">
        <div class="col pl-md-3 pl-lg-5 content-container">
            <!-- Table of contents that is only displayed when printing the page -->
            <div id="jb-print-docs-body" class="onlyprint">
                <h1>Neighborhood analysis</h1>
                <!-- Table of contents -->
                <div id="print-main-content">
                    <div id="jb-print-toc">
                        
                        <div>
                            <h2> Contents </h2>
                        </div>
                        <nav aria-label="Page">
                            <ul class="visible nav section-nav flex-column">
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#motivation">
   27.1. Motivation
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#environment-setup-and-data">
   27.2. Environment setup and data
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#identifying-interactions-between-spatial-communities">
   27.3. Identifying interactions between spatial communities
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#co-occurrence-across-spatial-dimensions">
   27.4. Co-occurrence across spatial dimensions
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#key-takeaways">
   27.5. Key takeaways
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#references">
   27.6. References
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#contributors">
   27.7. Contributors
  </a>
  <ul class="nav section-nav flex-column">
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#authors">
     27.7.1. Authors
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#reviewers">
     27.7.2. Reviewers
    </a>
   </li>
  </ul>
 </li>
</ul>

                        </nav>
                    </div>
                </div>
            </div>
            <main id="main-content" role="main">
                
              <div>
                
  <section class="tex2jax_ignore mathjax_ignore" id="neighborhood-analysis">
<h1><span class="section-number">27. </span>Neighborhood analysis<a class="headerlink" href="#neighborhood-analysis" title="Permalink to this headline">#</a></h1>
<section id="motivation">
<h2><span class="section-number">27.1. </span>Motivation<a class="headerlink" href="#motivation" title="Permalink to this headline">#</a></h2>
<p>After annotating cell types or cell states in the dataset (or spots, according to the technology at end), we can quantify whether such annotations are spatially enriched and analyze cellular neighborhoods across the tissue.</p>
<p>Cellular neighborhood analysis is a good starting point for various downstream tasks as it can help to understand the cellular composition of the tissue and identify candidates for more in-depth analysis. For example, it can help to find candidates for cell-cell communication based on spatial proximity, or spatial regions and clusters for identification of spatially variable genes.</p>
<p>Neighborhood analysis is often performed through <strong>spatial statistics</strong> <span id="id1">[<a class="reference internal" href="#id528" title="A.E. Gelfand, M. Fuentes, P. Guttorp, and P. Diggle. Handbook of Spatial Statistics. Chapman &amp; Hall/CRC Handbooks of Modern Statistical Methods. Taylor &amp; Francis, 2010. ISBN 9781420072877. URL: http://books.google.com/books?id=EFbbcMFZ2mMC.">Gelfand <em>et al.</em>, 2010</a>]</span>, which are quantitative scores that can be used to identify spatial neighborhoods in the tissue. Here, we’ll take a look at various spatial statistics implemented in Squidpy <span id="id2">[<a class="reference internal" href="spatially_variable_genes.html#id105" title="Giovanni Palla, Hannah Spitzer, Michal Klein, David Fischer, Anna Christina Schaar, Louis Benedikt Kuemmerle, Sergei Rybakov, Ignacio L. Ibarra, Olle Holmberg, Isaac Virshup, Mohammad Lotfollahi, Sabrina Richter, and Fabian J. Theis. Squidpy: a scalable framework for spatial omics analysis. Nature Methods, 19(2):171-178, Feb 2022. URL: https://doi.org/10.1038/s41592-021-01358-2, doi:10.1038/s41592-021-01358-2.">Palla <em>et al.</em>, 2022</a>]</span>.</p>
</section>
<section id="environment-setup-and-data">
<h2><span class="section-number">27.2. </span>Environment setup and data<a class="headerlink" href="#environment-setup-and-data" title="Permalink to this headline">#</a></h2>
<p>We first load the respective packages needed in this tutorial and the dataset.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">scanpy</span> <span class="k">as</span> <span class="nn">sc</span>
<span class="kn">import</span> <span class="nn">squidpy</span> <span class="k">as</span> <span class="nn">sq</span>

<span class="n">sc</span><span class="o">.</span><span class="n">settings</span><span class="o">.</span><span class="n">verbosity</span> <span class="o">=</span> <span class="mi">3</span>
<span class="n">sc</span><span class="o">.</span><span class="n">settings</span><span class="o">.</span><span class="n">set_figure_params</span><span class="p">(</span><span class="n">dpi</span><span class="o">=</span><span class="mi">80</span><span class="p">,</span> <span class="n">facecolor</span><span class="o">=</span><span class="s2">&quot;white&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<p>The dataset used in this tutorial consists of 1 tissue slides from 1 mouse and is provided by <a class="reference external" href="https://support.10xgenomics.com/spatial-gene-expression/datasets/1.1.0/V1_Adult_Mouse_Brain">10x Genomics Space Ranger 1.1.0</a>. The dataset was pre-processed in Squidpy, which provides a loading function for this dataset.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">adata</span> <span class="o">=</span> <span class="n">sq</span><span class="o">.</span><span class="n">datasets</span><span class="o">.</span><span class="n">visium_hne_adata</span><span class="p">()</span>
</pre></div>
</div>
</div>
</div>
</section>
<section id="identifying-interactions-between-spatial-communities">
<h2><span class="section-number">27.3. </span>Identifying interactions between spatial communities<a class="headerlink" href="#identifying-interactions-between-spatial-communities" title="Permalink to this headline">#</a></h2>
<p>After annotating cell types or cell states in the dataset (or spots, according to the technology at end), we can quantify whether such annotations are spatially enriched. To this end, computing a neighborhood enrichment can help us identify clusters that are neighbors in the tissue of interest. In short, it’s an enrichment score on spatial proximity of clusters: if observations (cells or spots) belonging to a cluster are often close to observations belonging to another cluster, then they will have a high score and will appear to be enriched. On the other hand, if they are far apart, and therefore are seldom neighbors, the score will be low and they can be defined as depleted. This score is based on a permutation-based test, and you can set the number of permutations with the n_perms argument (default is 1000).</p>
<p>Since the function works on a spatial connectivity matrix (spatial graph), we need to compute that as well. This can be done with <code class="docutils literal notranslate"><span class="pre">squidpy.gr.spatial_neighbors()</span></code>.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">sq</span><span class="o">.</span><span class="n">gr</span><span class="o">.</span><span class="n">spatial_neighbors</span><span class="p">(</span><span class="n">adata</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Creating graph using `grid` coordinates and `None` transform and `1` libraries.
Adding `adata.obsp[&#39;spatial_connectivities&#39;]`
       `adata.obsp[&#39;spatial_distances&#39;]`
       `adata.uns[&#39;spatial_neighbors&#39;]`
Finish (0:00:01)
</pre></div>
</div>
</div>
</div>
<p>We can now run the neighborhood enrichment test by providing the annotation key in <code class="docutils literal notranslate"><span class="pre">adata.obs</span></code>.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">sq</span><span class="o">.</span><span class="n">gr</span><span class="o">.</span><span class="n">nhood_enrichment</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">cluster_key</span><span class="o">=</span><span class="s2">&quot;cluster&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Calculating neighborhood enrichment using `1` core(s)
</pre></div>
</div>
<script type="application/vnd.jupyter.widget-view+json">
{"model_id": "e85c84fa9a914c9bb8b15ece58c88cdf", "version_major": 2, "version_minor": 0}
</script><div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Adding `adata.uns[&#39;cluster_nhood_enrichment&#39;]`
Finish (0:00:00)
</pre></div>
</div>
</div>
</div>
<p>The method added <code class="docutils literal notranslate"><span class="pre">adata.uns['cluster_nhood_enrichment']</span></code> to our AnnData object.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">adata</span><span class="o">.</span><span class="n">uns</span><span class="p">[</span><span class="s2">&quot;cluster_nhood_enrichment&quot;</span><span class="p">]</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>{&#39;zscore&#39;: array([[ 70.22155762, -13.6041789 ,  -0.4258425 ,   2.94739732,
          -9.03708733, -12.44368289,  -8.94699794, -12.42056733,
          -9.76328672,  -8.53570662,  -5.34106034,  -6.81758862,
         -10.46793569, -13.79518632, -11.85984283],
        [-13.6041789 ,  74.34650092,  -6.37460975,  -9.11260235,
           1.92787693, -12.0461501 , -11.90907953, -11.42070949,
          -8.79949063,  -7.03693412,  -5.27623436,  -6.33541649,
          -3.39758871, -13.09498141, -11.02100047],
        [ -0.4258425 ,  -6.37460975,  71.02312621, -10.2557408 ,
          -8.38252823,  -3.56904737, -11.60339076, -11.17301353,
          -8.63097371,  -8.04109503,  -4.88912876,  -6.29768401,
          -7.40871581, -12.6590886 , -10.5828295 ],
        [  2.94739732,  -9.11260235, -10.2557408 ,  57.83928709,
          14.41213846,  -9.54187862,  -9.20031921,  -9.36357782,
          -7.19965825,  -6.49172687,  -4.07412227,  -5.14772134,
          -7.55767703, -10.34784232,  -8.81768536],
        [ -9.03708733,   1.92787693,  -8.38252823,  14.41213846,
          36.34406673,  -8.27754053,  -7.55077944,  -8.38725173,
          -6.28914454,  -1.78180094,  -3.50081508,  -4.46780626,
          -6.80187795,  -9.09205304,  -7.97723212],
        [-12.44368289, -12.0461501 ,  -3.56904737,  -9.54187862,
          -8.27754053,  56.46317761,  -1.65912586,  -9.83917564,
           3.34500357,   4.76287001,  -4.4421212 ,  -6.07070432,
          -4.06673556,  -9.01905311,  -7.63068773],
        [ -8.94699794, -11.90907953, -11.60339076,  -9.20031921,
          -7.55077944,  -1.65912586,  56.52983503, -10.53203586,
          -6.95861812,  -2.59374816,  26.16381548,  13.58097471,
          -8.86853162, -11.81470655, -10.82483187],
        [-12.42056733, -11.42070949, -11.17301353,  -9.36357782,
          -8.38725173,  -9.83917564, -10.53203586,  78.38442071,
          -3.86826297,  -4.37445824,  -4.5991668 ,  -5.83739543,
          -8.6848676 , -11.53420802,  -6.35636122],
        [ -9.76328672,  -8.79949063,  -8.63097371,  -7.19965825,
          -6.28914454,   3.34500357,  -6.95861812,  -3.86826297,
          61.62040415,  -4.40694372,  -3.54475788,  -2.20267766,
          -6.88126696,  -1.30472155,  -2.3125057 ],
        [ -8.53570662,  -7.03693412,  -8.04109503,  -6.49172687,
          -1.78180094,   4.76287001,  -2.59374816,  -4.37445824,
          -4.40694372,  42.4700539 ,  -3.1814847 ,  -4.0503999 ,
           4.9198482 ,  -8.01018986,  -4.18579912],
        [ -5.34106034,  -5.27623436,  -4.88912876,  -4.07412227,
          -3.50081508,  -4.4421212 ,  26.16381548,  -4.5991668 ,
          -3.54475788,  -3.1814847 ,  45.15843296,  -1.28213058,
          -3.87010332,  -4.99305736,  -4.42082271],
        [ -6.81758862,  -6.33541649,  -6.29768401,  -5.14772134,
          -4.46780626,  -6.07070432,  13.58097471,  -5.83739543,
          -2.20267766,  -4.0503999 ,  -1.28213058,  64.52229328,
          -4.87520289,  -5.13157661,  -5.494849  ],
        [-10.46793569,  -3.39758871,  -7.40871581,  -7.55767703,
          -6.80187795,  -4.06673556,  -8.86853162,  -8.6848676 ,
          -6.88126696,   4.9198482 ,  -3.87010332,  -4.87520289,
          72.72449982, -10.10115791,  -8.53496568],
        [-13.79518632, -13.09498141, -12.6590886 , -10.34784232,
          -9.09205304,  -9.01905311, -11.81470655, -11.53420802,
          -1.30472155,  -8.01018986,  -4.99305736,  -5.13157661,
         -10.10115791,  80.27730927,  -3.61634144],
        [-11.85984283, -11.02100047, -10.5828295 ,  -8.81768536,
          -7.97723212,  -7.63068773, -10.82483187,  -6.35636122,
          -2.3125057 ,  -4.18579912,  -4.42082271,  -5.494849  ,
          -8.53496568,  -3.61634144,  67.62254826]]),
 &#39;count&#39;: array([[1352,    1,  145,  128,    4,    4,   40,    0,    0,    0,    0,
            0,    0,    0,    0],
        [   1, 1290,   67,    9,   87,    0,    0,    0,    0,    7,    0,
            0,   55,    0,    0],
        [ 145,   67, 1140,    0,    2,   83,    3,    0,    0,    0,    0,
            0,   18,    0,    0],
        [ 128,    9,    0,  678,  140,    0,    0,    0,    0,    0,    0,
            0,    0,    0,    0],
        [   4,   87,    2,  140,  330,    0,    6,    0,    0,   20,    0,
            0,    0,    0,    0],
        [   4,    0,   83,    0,    0,  878,   92,    7,   90,   84,    1,
            0,   43,   35,   24],
        [  40,    0,    3,    0,    6,   92,  872,    0,   13,   34,  128,
          104,    0,    3,    0],
        [   0,    0,    0,    0,    0,    7,    0, 1080,   31,   19,    0,
            0,    0,    0,   29],
        [   0,    0,    0,    0,    0,   90,   13,   31,  544,    7,    0,
           10,    0,   64,   39],
        [   0,    7,    0,    0,   20,   84,   34,   19,    7,  314,    0,
            0,   63,    1,   17],
        [   0,    0,    0,    0,    0,    1,  128,    0,    0,    0,  120,
            3,    0,    0,    0],
        [   0,    0,    0,    0,    0,    0,  104,    0,   10,    0,    3,
          280,    0,    9,    0],
        [   0,   55,   18,    0,    0,   43,    0,    0,    0,   63,    0,
            0,  736,    0,    0],
        [   0,    0,    0,    0,    0,   35,    3,    0,   64,    1,    0,
            9,    0, 1382,   72],
        [   0,    0,    0,    0,    0,   24,    0,   29,   39,   17,    0,
            0,    0,   72,  902]], dtype=uint32)}
</pre></div>
</div>
</div>
</div>
<p>The added object contains two arrays. The first stored under <code class="docutils literal notranslate"><span class="pre">zscore</span></code> contains the enrichment z-score for each cell-cell interaction. The second is stored under <code class="docutils literal notranslate"><span class="pre">count</span></code> and represents the enrichment count.</p>
<p>Finally, we’ll directly visualize the results with <code class="docutils literal notranslate"><span class="pre">squidpy.pl.nhood_enrichment()</span></code>.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">sq</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">nhood_enrichment</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">cluster_key</span><span class="o">=</span><span class="s2">&quot;cluster&quot;</span><span class="p">,</span> <span class="n">method</span><span class="o">=</span><span class="s2">&quot;average&quot;</span><span class="p">,</span> <span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">5</span><span class="p">,</span> <span class="mi">5</span><span class="p">))</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="../_images/neighborhood_13_0.png" src="../_images/neighborhood_13_0.png" />
</div>
</div>
<p>From the above plot, we can see that there seems to be an enrichment for clusters of the <code class="docutils literal notranslate"><span class="pre">Pyramidal_layer</span></code> and <code class="docutils literal notranslate"><span class="pre">Dentate_gyrus</span></code>. By looking at the spatial scatterplot above, we can confirm that these clusters are indeed “neighbors” as their members are often close.</p>
<p>A similar approach to such problem is computing what we call an <em>interaction matrix</em>, that is, the sum of all connecting observations between clusters in tissue. The approach is related to the neighborhood enrichment analysis yet it is not a test, but should be viewed as a simple summary statistics of the spatial graph.
Let’s take a look at how the interaction matrix looks like for the dataset.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">sq</span><span class="o">.</span><span class="n">gr</span><span class="o">.</span><span class="n">interaction_matrix</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">cluster_key</span><span class="o">=</span><span class="s2">&quot;cluster&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Adding `adata.uns[&#39;cluster_interactions&#39;]`
</pre></div>
</div>
</div>
</div>
<p>The function added <code class="docutils literal notranslate"><span class="pre">adata.uns['cluster_interactions']</span></code> to our AnnData object, which contains the number of interactions between two clusters with respect to the provided spatial connectivities graph.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">adata</span><span class="o">.</span><span class="n">uns</span><span class="p">[</span><span class="s2">&quot;cluster_interactions&quot;</span><span class="p">]</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>array([[1.352e+03, 1.000e+00, 1.450e+02, 1.280e+02, 4.000e+00, 4.000e+00,
        4.000e+01, 0.000e+00, 0.000e+00, 0.000e+00, 0.000e+00, 0.000e+00,
        0.000e+00, 0.000e+00, 0.000e+00],
       [1.000e+00, 1.290e+03, 6.700e+01, 9.000e+00, 8.700e+01, 0.000e+00,
        0.000e+00, 0.000e+00, 0.000e+00, 7.000e+00, 0.000e+00, 0.000e+00,
        5.500e+01, 0.000e+00, 0.000e+00],
       [1.450e+02, 6.700e+01, 1.140e+03, 0.000e+00, 2.000e+00, 8.300e+01,
        3.000e+00, 0.000e+00, 0.000e+00, 0.000e+00, 0.000e+00, 0.000e+00,
        1.800e+01, 0.000e+00, 0.000e+00],
       [1.280e+02, 9.000e+00, 0.000e+00, 6.780e+02, 1.400e+02, 0.000e+00,
        0.000e+00, 0.000e+00, 0.000e+00, 0.000e+00, 0.000e+00, 0.000e+00,
        0.000e+00, 0.000e+00, 0.000e+00],
       [4.000e+00, 8.700e+01, 2.000e+00, 1.400e+02, 3.300e+02, 0.000e+00,
        6.000e+00, 0.000e+00, 0.000e+00, 2.000e+01, 0.000e+00, 0.000e+00,
        0.000e+00, 0.000e+00, 0.000e+00],
       [4.000e+00, 0.000e+00, 8.300e+01, 0.000e+00, 0.000e+00, 8.780e+02,
        9.200e+01, 7.000e+00, 9.000e+01, 8.400e+01, 1.000e+00, 0.000e+00,
        4.300e+01, 3.500e+01, 2.400e+01],
       [4.000e+01, 0.000e+00, 3.000e+00, 0.000e+00, 6.000e+00, 9.200e+01,
        8.720e+02, 0.000e+00, 1.300e+01, 3.400e+01, 1.280e+02, 1.040e+02,
        0.000e+00, 3.000e+00, 0.000e+00],
       [0.000e+00, 0.000e+00, 0.000e+00, 0.000e+00, 0.000e+00, 7.000e+00,
        0.000e+00, 1.080e+03, 3.100e+01, 1.900e+01, 0.000e+00, 0.000e+00,
        0.000e+00, 0.000e+00, 2.900e+01],
       [0.000e+00, 0.000e+00, 0.000e+00, 0.000e+00, 0.000e+00, 9.000e+01,
        1.300e+01, 3.100e+01, 5.440e+02, 7.000e+00, 0.000e+00, 1.000e+01,
        0.000e+00, 6.400e+01, 3.900e+01],
       [0.000e+00, 7.000e+00, 0.000e+00, 0.000e+00, 2.000e+01, 8.400e+01,
        3.400e+01, 1.900e+01, 7.000e+00, 3.140e+02, 0.000e+00, 0.000e+00,
        6.300e+01, 1.000e+00, 1.700e+01],
       [0.000e+00, 0.000e+00, 0.000e+00, 0.000e+00, 0.000e+00, 1.000e+00,
        1.280e+02, 0.000e+00, 0.000e+00, 0.000e+00, 1.200e+02, 3.000e+00,
        0.000e+00, 0.000e+00, 0.000e+00],
       [0.000e+00, 0.000e+00, 0.000e+00, 0.000e+00, 0.000e+00, 0.000e+00,
        1.040e+02, 0.000e+00, 1.000e+01, 0.000e+00, 3.000e+00, 2.800e+02,
        0.000e+00, 9.000e+00, 0.000e+00],
       [0.000e+00, 5.500e+01, 1.800e+01, 0.000e+00, 0.000e+00, 4.300e+01,
        0.000e+00, 0.000e+00, 0.000e+00, 6.300e+01, 0.000e+00, 0.000e+00,
        7.360e+02, 0.000e+00, 0.000e+00],
       [0.000e+00, 0.000e+00, 0.000e+00, 0.000e+00, 0.000e+00, 3.500e+01,
        3.000e+00, 0.000e+00, 6.400e+01, 1.000e+00, 0.000e+00, 9.000e+00,
        0.000e+00, 1.382e+03, 7.200e+01],
       [0.000e+00, 0.000e+00, 0.000e+00, 0.000e+00, 0.000e+00, 2.400e+01,
        0.000e+00, 2.900e+01, 3.900e+01, 1.700e+01, 0.000e+00, 0.000e+00,
        0.000e+00, 7.200e+01, 9.020e+02]])
</pre></div>
</div>
</div>
</div>
<p>We can visualize the results with <code class="docutils literal notranslate"><span class="pre">squidpy.pl.interaction_matrix()</span></code>.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">sq</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">interaction_matrix</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">cluster_key</span><span class="o">=</span><span class="s2">&quot;cluster&quot;</span><span class="p">,</span> <span class="n">method</span><span class="o">=</span><span class="s2">&quot;average&quot;</span><span class="p">,</span> <span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">5</span><span class="p">,</span> <span class="mi">5</span><span class="p">))</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="../_images/neighborhood_20_0.png" src="../_images/neighborhood_20_0.png" />
</div>
</div>
<p>For the dataset, we roughly recapitulate the neighborhood enrichment test, yet we seem to not observe a particularly strong interaction between the <code class="docutils literal notranslate"><span class="pre">Pyramidal_layer</span></code> and <code class="docutils literal notranslate"><span class="pre">Dentate_gyrus</span></code> clusters.
One explanation for such a result is that the number of observations of such clusters is low, hence the low number of interactions.</p>
</section>
<section id="co-occurrence-across-spatial-dimensions">
<h2><span class="section-number">27.4. </span>Co-occurrence across spatial dimensions<a class="headerlink" href="#co-occurrence-across-spatial-dimensions" title="Permalink to this headline">#</a></h2>
<p>Another spatial statistic that can be computed on cell type annotations in spatial coordinates is what we call the co-occurrence score <span id="id3">[<a class="reference internal" href="spatially_variable_genes.html#id105" title="Giovanni Palla, Hannah Spitzer, Michal Klein, David Fischer, Anna Christina Schaar, Louis Benedikt Kuemmerle, Sergei Rybakov, Ignacio L. Ibarra, Olle Holmberg, Isaac Virshup, Mohammad Lotfollahi, Sabrina Richter, and Fabian J. Theis. Squidpy: a scalable framework for spatial omics analysis. Nature Methods, 19(2):171-178, Feb 2022. URL: https://doi.org/10.1038/s41592-021-01358-2, doi:10.1038/s41592-021-01358-2.">Palla <em>et al.</em>, 2022</a>, <a class="reference internal" href="#id550" title="Luca Tosti, Yan Hang, Olivia Debnath, Sebastian Tiesmeyer, Timo Trefzer, Katja Steiger, Foo Wei Ten, Sören Lukassen, Simone Ballke, Anja A. Kühl, Simone Spieckermann, Rita Bottino, Naveed Ishaque, Wilko Weichert, Seung K. Kim, Roland Eils, and Christian Conrad. Single-nucleus and in situ rna–sequencing reveal cell topographies in the human pancreas. Gastroenterology, 160(4):1330-1344.e11, 2021. URL: https://www.sciencedirect.com/science/article/pii/S0016508520353993, doi:https://doi.org/10.1053/j.gastro.2020.11.010.">Tosti <em>et al.</em>, 2021</a>]</span>. The co-occurrence score gives us an indication on whether clusters co-occur with each other at increasing distances across the tissue.
The co-occurrence score is defined as:</p>
<p><span class="math notranslate nohighlight">\(\frac{p(exp|cond)}{p(exp)}\)</span></p>
<p>where <span class="math notranslate nohighlight">\(p(exp|cond)\)</span> is the conditional probability of observing a cluster <span class="math notranslate nohighlight">\(exp\)</span> conditioned
on the presence of a cluster <span class="math notranslate nohighlight">\(cond\)</span> whereas <span class="math notranslate nohighlight">\(exp\)</span> is the probability of observing <span class="math notranslate nohighlight">\(exp\)</span>
in the radius size of interest. The score is computed across increasing radii size around each observation (i.e. spots here) in the tissue.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">sq</span><span class="o">.</span><span class="n">gr</span><span class="o">.</span><span class="n">co_occurrence</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">cluster_key</span><span class="o">=</span><span class="s2">&quot;cluster&quot;</span><span class="p">)</span>
<span class="n">sq</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">co_occurrence</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">cluster_key</span><span class="o">=</span><span class="s2">&quot;cluster&quot;</span><span class="p">,</span> <span class="n">clusters</span><span class="o">=</span><span class="s2">&quot;Cortex_1&quot;</span><span class="p">,</span> <span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">8</span><span class="p">,</span> <span class="mi">5</span><span class="p">))</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Calculating co-occurrence probabilities for `50` intervals `1` split combinations using `1` core(s)
</pre></div>
</div>
<script type="application/vnd.jupyter.widget-view+json">
{"model_id": "0d9d3480b8d04b2a8a1f5da796495036", "version_major": 2, "version_minor": 0}
</script><div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Adding `adata.uns[&#39;cluster_co_occurrence&#39;]`
Finish (0:00:00)
</pre></div>
</div>
<img alt="../_images/neighborhood_23_3.png" src="../_images/neighborhood_23_3.png" />
</div>
</div>
<p>Here, we selected to visualize the cluster <code class="docutils literal notranslate"><span class="pre">Cortex_1</span></code> to visualize how at close distances, the cluster co-occur with the other <code class="docutils literal notranslate"><span class="pre">Cortex</span></code> clusters, as expected.</p>
</section>
<section id="key-takeaways">
<h2><span class="section-number">27.5. </span>Key takeaways<a class="headerlink" href="#key-takeaways" title="Permalink to this headline">#</a></h2>
<ul class="simple">
<li><p>Spatial statistics and neighborhood analysis can be a good first starting point for analyzing spatial omics data</p></li>
<li><p>Analysis tools like Squidpy provide several spatial statistics that help to understand the neighborhood structure in spatial omics datasets</p></li>
</ul>
</section>
<section id="references">
<h2><span class="section-number">27.6. </span>References<a class="headerlink" href="#references" title="Permalink to this headline">#</a></h2>
<div class="docutils container" id="id4">
<dl class="citation">
<dt class="label" id="id528"><span class="brackets"><a class="fn-backref" href="#id1">spatialGFGD10</a></span></dt>
<dd><p>A.E. Gelfand, M. Fuentes, P. Guttorp, and P. Diggle. <em>Handbook of Spatial Statistics</em>. Chapman &amp; Hall/CRC Handbooks of Modern Statistical Methods. Taylor &amp; Francis, 2010. ISBN 9781420072877. URL: <a class="reference external" href="http://books.google.com/books?id=EFbbcMFZ2mMC">http://books.google.com/books?id=EFbbcMFZ2mMC</a>.</p>
</dd>
<dt class="label" id="id99"><span class="brackets">spatialPSK+22</span><span class="fn-backref">(<a href="#id2">1</a>,<a href="#id3">2</a>)</span></dt>
<dd><p>Giovanni Palla, Hannah Spitzer, Michal Klein, David Fischer, Anna Christina Schaar, Louis Benedikt Kuemmerle, Sergei Rybakov, Ignacio L. Ibarra, Olle Holmberg, Isaac Virshup, Mohammad Lotfollahi, Sabrina Richter, and Fabian J. Theis. Squidpy: a scalable framework for spatial omics analysis. <em>Nature Methods</em>, 19(2):171–178, Feb 2022. URL: <a class="reference external" href="https://doi.org/10.1038/s41592-021-01358-2">https://doi.org/10.1038/s41592-021-01358-2</a>, <a class="reference external" href="https://doi.org/10.1038/s41592-021-01358-2">doi:10.1038/s41592-021-01358-2</a>.</p>
</dd>
<dt class="label" id="id550"><span class="brackets"><a class="fn-backref" href="#id3">spatialTHD+21</a></span></dt>
<dd><p>Luca Tosti, Yan Hang, Olivia Debnath, Sebastian Tiesmeyer, Timo Trefzer, Katja Steiger, Foo Wei Ten, Sören Lukassen, Simone Ballke, Anja A. Kühl, Simone Spieckermann, Rita Bottino, Naveed Ishaque, Wilko Weichert, Seung K. Kim, Roland Eils, and Christian Conrad. Single-nucleus and in situ rna–sequencing reveal cell topographies in the human pancreas. <em>Gastroenterology</em>, 160(4):1330–1344.e11, 2021. URL: <a class="reference external" href="https://www.sciencedirect.com/science/article/pii/S0016508520353993">https://www.sciencedirect.com/science/article/pii/S0016508520353993</a>, <a class="reference external" href="https://doi.org/https://doi.org/10.1053/j.gastro.2020.11.010">doi:https://doi.org/10.1053/j.gastro.2020.11.010</a>.</p>
</dd>
</dl>
</div>
</section>
<section id="contributors">
<h2><span class="section-number">27.7. </span>Contributors<a class="headerlink" href="#contributors" title="Permalink to this headline">#</a></h2>
<section id="authors">
<h3><span class="section-number">27.7.1. </span>Authors<a class="headerlink" href="#authors" title="Permalink to this headline">#</a></h3>
<ul class="simple">
<li><p>Giovanni Palla</p></li>
</ul>
</section>
<section id="reviewers">
<h3><span class="section-number">27.7.2. </span>Reviewers<a class="headerlink" href="#reviewers" title="Permalink to this headline">#</a></h3>
<ul class="simple">
<li><p>Anna Schaar</p></li>
<li><p>Lukas Heumos</p></li>
</ul>
</section>
</section>
</section>

    <script type="text/x-thebe-config">
    {
        requestKernel: true,
        binderOptions: {
            repo: "binder-examples/jupyter-stacks-datascience",
            ref: "master",
        },
        codeMirrorConfig: {
            theme: "abcdef",
            mode: "spatial-book"
        },
        kernelOptions: {
            name: "spatial-book",
            path: "./spatial"
        },
        predefinedOutput: true
    }
    </script>
    <script>kernelName = 'spatial-book'</script>

              </div>
              
            </main>
            <footer class="footer-article noprint">
                
    <!-- Previous / next buttons -->
<div class='prev-next-area'>
    <a class='left-prev' id="prev-link" href="introduction.html" title="previous page">
        <i class="fas fa-angle-left"></i>
        <div class="prev-next-info">
            <p class="prev-next-subtitle">previous</p>
            <p class="prev-next-title"><span class="section-number">26. </span>Single-cell data resolved in space</p>
        </div>
    </a>
    <a class='right-next' id="next-link" href="domains.html" title="next page">
    <div class="prev-next-info">
        <p class="prev-next-subtitle">next</p>
        <p class="prev-next-title"><span class="section-number">28. </span>Spatial domains</p>
    </div>
    <i class="fas fa-angle-right"></i>
    </a>
</div>
            </footer>
        </div>
    </div>
    <div class="footer-content row">
        <footer class="col footer"><p>
  
    By Lukas Heumos, Anna Schaar, single-cell best practices consortium<br/>
  
      &copy; Copyright 2022.<br/>
    <div class="extra_footer">
      <div>
  Brought to you by
  <a href="https://www.github.com/theislab">Theislab</a>, with many thanks to the single-cell community as a whole!
</div>

    </div>
</p>
        </footer>
    </div>
    
</div>


      </div>
    </div>
  
  <!-- Scripts loaded after <body> so the DOM is not blocked -->
  <script src="../_static/scripts/pydata-sphinx-theme.js?digest=1999514e3f237ded88cf"></script>


  </body>
</html>