
<!DOCTYPE html>


<html lang="en" data-content_root="../" >

  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="viewport" content="width=device-width, initial-scale=1" />

    <title>43. Specificity analysis &#8212; Single-cell best practices</title>
  
  
  
  <script data-cfasync="false">
    document.documentElement.dataset.mode = localStorage.getItem("mode") || "";
    document.documentElement.dataset.theme = localStorage.getItem("theme") || "";
  </script>
  
  <!-- Loaded before other Sphinx assets -->
  <link href="../_static/styles/theme.css?digest=dfe6caa3a7d634c4db9b" rel="stylesheet" />
<link href="../_static/styles/bootstrap.css?digest=dfe6caa3a7d634c4db9b" rel="stylesheet" />
<link href="../_static/styles/pydata-sphinx-theme.css?digest=dfe6caa3a7d634c4db9b" rel="stylesheet" />

  
  <link href="../_static/vendor/fontawesome/6.5.2/css/all.min.css?digest=dfe6caa3a7d634c4db9b" rel="stylesheet" />
  <link rel="preload" as="font" type="font/woff2" crossorigin href="../_static/vendor/fontawesome/6.5.2/webfonts/fa-solid-900.woff2" />
<link rel="preload" as="font" type="font/woff2" crossorigin href="../_static/vendor/fontawesome/6.5.2/webfonts/fa-brands-400.woff2" />
<link rel="preload" as="font" type="font/woff2" crossorigin href="../_static/vendor/fontawesome/6.5.2/webfonts/fa-regular-400.woff2" />

    <link rel="stylesheet" type="text/css" href="../_static/pygments.css?v=03e43079" />
    <link rel="stylesheet" type="text/css" href="../_static/styles/sphinx-book-theme.css?v=eba8b062" />
    <link rel="stylesheet" type="text/css" href="../_static/togglebutton.css?v=13237357" />
    <link rel="stylesheet" type="text/css" href="../_static/copybutton.css?v=76b2166b" />
    <link rel="stylesheet" type="text/css" href="../_static/mystnb.4510f1fc1dee50b3e5859aac5469c37c29e427902b24a333a5f9fcb2f0b3ac41.css" />
    <link rel="stylesheet" type="text/css" href="../_static/sphinx-thebe.css?v=4fa983c6" />
    <link rel="stylesheet" type="text/css" href="../_static/sphinx-design.min.css?v=95c83b7e" />
    <link rel="stylesheet" type="text/css" href="../_static/book.css?v=148c11e2" />
  
  <!-- Pre-loaded scripts that we'll load fully later -->
  <link rel="preload" as="script" href="../_static/scripts/bootstrap.js?digest=dfe6caa3a7d634c4db9b" />
<link rel="preload" as="script" href="../_static/scripts/pydata-sphinx-theme.js?digest=dfe6caa3a7d634c4db9b" />
  <script src="../_static/vendor/fontawesome/6.5.2/js/all.min.js?digest=dfe6caa3a7d634c4db9b"></script>

    <script src="../_static/documentation_options.js?v=9eb32ce0"></script>
    <script src="../_static/doctools.js?v=9a2dae69"></script>
    <script src="../_static/sphinx_highlight.js?v=dc90522c"></script>
    <script src="../_static/clipboard.min.js?v=a7894cd8"></script>
    <script src="../_static/copybutton.js?v=f281be69"></script>
    <script src="../_static/scripts/sphinx-book-theme.js?v=887ef09a"></script>
    <script>let toggleHintShow = 'Click to show';</script>
    <script>let toggleHintHide = 'Click to hide';</script>
    <script>let toggleOpenOnPrint = 'true';</script>
    <script src="../_static/togglebutton.js?v=4a39c7ea"></script>
    <script>var togglebuttonSelector = '.toggle, .admonition.dropdown';</script>
    <script src="../_static/design-tabs.js?v=f930bc37"></script>
    <script>const THEBE_JS_URL = "https://unpkg.com/thebe@0.8.2/lib/index.js"; const thebe_selector = ".thebe,.cell"; const thebe_selector_input = "pre"; const thebe_selector_output = ".output, .cell_output"</script>
    <script async="async" src="../_static/sphinx-thebe.js?v=c100c467"></script>
    <script>var togglebuttonSelector = '.toggle, .admonition.dropdown';</script>
    <script>const THEBE_JS_URL = "https://unpkg.com/thebe@0.8.2/lib/index.js"; const thebe_selector = ".thebe,.cell"; const thebe_selector_input = "pre"; const thebe_selector_output = ".output, .cell_output"</script>
    <script>DOCUMENTATION_OPTIONS.pagename = 'air_repertoire/specificity';</script>
    <script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
    <script src="../_static/book.js?v=dcd3d8bf"></script>
    <link rel="icon" href="../_static/favicon.ico"/>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="44. Integrating AIR and transcriptomics" href="multimodal_integration.html" />
    <link rel="prev" title="39. Clonotype analysis" href="clonotype.html" />
  <meta name="viewport" content="width=device-width, initial-scale=1"/>
  <meta name="docsearch:language" content="en"/>
  </head>
  
  
  <body data-bs-spy="scroll" data-bs-target=".bd-toc-nav" data-offset="180" data-bs-root-margin="0px 0px -60%" data-default-mode="">

  
  
  <div id="pst-skip-link" class="skip-link d-print-none"><a href="#main-content">Skip to main content</a></div>
  
  <div id="pst-scroll-pixel-helper"></div>
  
  <button type="button" class="btn rounded-pill" id="pst-back-to-top">
    <i class="fa-solid fa-arrow-up"></i>Back to top</button>

  
  <input type="checkbox"
          class="sidebar-toggle"
          id="pst-primary-sidebar-checkbox"/>
  <label class="overlay overlay-primary" for="pst-primary-sidebar-checkbox"></label>
  
  <input type="checkbox"
          class="sidebar-toggle"
          id="pst-secondary-sidebar-checkbox"/>
  <label class="overlay overlay-secondary" for="pst-secondary-sidebar-checkbox"></label>
  
  <div class="search-button__wrapper">
    <div class="search-button__overlay"></div>
    <div class="search-button__search-container">
<form class="bd-search d-flex align-items-center"
      action="../search.html"
      method="get">
  <i class="fa-solid fa-magnifying-glass"></i>
  <input type="search"
         class="form-control"
         name="q"
         id="search-input"
         placeholder="Search this book..."
         aria-label="Search this book..."
         autocomplete="off"
         autocorrect="off"
         autocapitalize="off"
         spellcheck="false"/>
  <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd>K</kbd></span>
</form></div>
  </div>

  <div class="pst-async-banner-revealer d-none">
  <aside id="bd-header-version-warning" class="d-none d-print-none" aria-label="Version warning"></aside>
</div>

  
    <header class="bd-header navbar navbar-expand-lg bd-navbar d-print-none">
    </header>
  

  <div class="bd-container">
    <div class="bd-container__inner bd-page-width">
      
      
      
      <div class="bd-sidebar-primary bd-sidebar">
        

  
  <div class="sidebar-header-items sidebar-primary__section">
    
    
    
    
  </div>
  
    <div class="sidebar-primary-items__start sidebar-primary__section">
        <div class="sidebar-primary-item">

  
    
  

<a class="navbar-brand logo" href="../preamble.html">
  
  
  
  
  
    
    
      
    
    
    <img src="../_static/logo.png" class="logo__image only-light" alt="Single-cell best practices - Home"/>
    <script>document.write(`<img src="../_static/logo.png" class="logo__image only-dark" alt="Single-cell best practices - Home"/>`);</script>
  
  
</a></div>
        <div class="sidebar-primary-item">

 <script>
 document.write(`
   <button class="btn search-button-field search-button__button" title="Search" aria-label="Search" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="fa-solid fa-magnifying-glass"></i>
    <span class="search-button__default-text">Search</span>
    <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd class="kbd-shortcut__modifier">K</kbd></span>
   </button>
 `);
 </script></div>
        <div class="sidebar-primary-item"><nav class="bd-links bd-docs-nav" aria-label="Main">
    <div class="bd-toc-item navbar-nav active">
        <p aria-level="2" class="caption" role="heading"><span class="caption-text">Introduction</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../introduction/prior_art.html">1. Prior art</a></li>
<li class="toctree-l1"><a class="reference internal" href="../introduction/scrna_seq.html">2. Single-cell RNA sequencing</a></li>
<li class="toctree-l1"><a class="reference internal" href="../introduction/raw_data_processing.html">3. Raw data processing</a></li>
<li class="toctree-l1"><a class="reference internal" href="../introduction/analysis_tools.html">4. Analysis frameworks and tools</a></li>
<li class="toctree-l1"><a class="reference internal" href="../introduction/interoperability.html">5. Interoperability</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Preprocessing and visualization</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../preprocessing_visualization/quality_control.html">6. Quality Control</a></li>
<li class="toctree-l1"><a class="reference internal" href="../preprocessing_visualization/normalization.html">7. Normalization</a></li>
<li class="toctree-l1"><a class="reference internal" href="../preprocessing_visualization/feature_selection.html">8. Feature selection</a></li>
<li class="toctree-l1"><a class="reference internal" href="../preprocessing_visualization/dimensionality_reduction.html">9. Dimensionality Reduction</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Identifying cellular structure</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../cellular_structure/clustering.html">10. Clustering</a></li>
<li class="toctree-l1"><a class="reference internal" href="../cellular_structure/annotation.html">11. Annotation</a></li>
<li class="toctree-l1"><a class="reference internal" href="../cellular_structure/integration.html">12. Data integration</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Inferring trajectories</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../trajectories/pseudotemporal.html">13. Pseudotemporal ordering</a></li>
<li class="toctree-l1"><a class="reference internal" href="../trajectories/rna_velocity.html">14. RNA velocity</a></li>
<li class="toctree-l1"><a class="reference internal" href="../trajectories/lineage_tracing.html">15. Lineage tracing</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Dealing with conditions</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../conditions/differential_gene_expression.html">16. Differential gene expression analysis</a></li>
<li class="toctree-l1"><a class="reference internal" href="../conditions/compositional.html">17. Compositional analysis</a></li>
<li class="toctree-l1"><a class="reference internal" href="../conditions/gsea_pathway.html">18. Gene set enrichment and pathway analysis</a></li>
<li class="toctree-l1"><a class="reference internal" href="../conditions/perturbation_modeling.html">19. Perturbation modeling</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Modeling mechanisms</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../mechanisms/gene_regulatory_networks.html">20. Gene regulatory networks</a></li>
<li class="toctree-l1"><a class="reference internal" href="../mechanisms/cell_cell_communication.html">21. Cell-cell communication</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Deconvolution</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../deconvolution/bulk_deconvolution.html">22. Bulk deconvolution</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Chromatin Accessibility</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../chromatin_accessibility/introduction.html">23. Single-cell ATAC sequencing</a></li>
<li class="toctree-l1"><a class="reference internal" href="../chromatin_accessibility/quality_control.html">24. Quality Control</a></li>
<li class="toctree-l1"><a class="reference internal" href="../chromatin_accessibility/gene_regulatory_networks_atac.html">25. Gene regulatory networks using chromatin accessibility</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Spatial omics</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../spatial/introduction.html">26. Single-cell data resolved in space</a></li>
<li class="toctree-l1"><a class="reference internal" href="../spatial/neighborhood.html">27. Neighborhood analysis</a></li>
<li class="toctree-l1"><a class="reference internal" href="../spatial/domains.html">28. Spatial domains</a></li>
<li class="toctree-l1"><a class="reference internal" href="../spatial/spatially_variable_genes.html">29. Spatially variable genes</a></li>
<li class="toctree-l1"><a class="reference internal" href="../spatial/deconvolution.html">30. Spatial deconvolution</a></li>
<li class="toctree-l1"><a class="reference internal" href="../spatial/imputation.html">31. Imputation</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Surface protein</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../surface_protein/quality_control.html">32. Quality control</a></li>
<li class="toctree-l1"><a class="reference internal" href="../surface_protein/normalization.html">33. Normalization</a></li>
<li class="toctree-l1"><a class="reference internal" href="../surface_protein/doublet_detection.html">34. Doublet detection</a></li>
<li class="toctree-l1"><a class="reference internal" href="../surface_protein/dimensionality_reduction.html">35. Dimensionality Reduction</a></li>
<li class="toctree-l1"><a class="reference internal" href="../surface_protein/batch_correction.html">36. Batch correction</a></li>
<li class="toctree-l1"><a class="reference internal" href="../surface_protein/annotation.html">37. Annotation</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Adaptive immune receptor repertoire</span></p>
<ul class="current nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="ir_profiling.html">38. Immune Receptor Profiling</a></li>
<li class="toctree-l1"><a class="reference internal" href="clonotype.html">39. Clonotype analysis</a></li>



<li class="toctree-l1 current active"><a class="current reference internal" href="#">43. Specificity analysis</a></li>
<li class="toctree-l1"><a class="reference internal" href="multimodal_integration.html">44. Integrating AIR and transcriptomics</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Multimodal integration</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../multimodal_integration/paired_integration.html">45. Paired integration</a></li>
<li class="toctree-l1"><a class="reference internal" href="../multimodal_integration/advanced_integration.html">46. Advanced integration</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Reproducibility</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../reproducibility/introduction.html">47. Reproducibility</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Outlook</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../outlook.html">48. Outlook</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Acknowledgements</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../acknowledgements.html">49. Acknowledgements</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Glossary</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../glossary.html">50. Glossary</a></li>
</ul>

    </div>
</nav></div>
    </div>
  
  
  <div class="sidebar-primary-items__end sidebar-primary__section">
  </div>
  
  <div id="rtd-footer-container"></div>


      </div>
      
      <main id="main-content" class="bd-main" role="main">
        
        

<div class="sbt-scroll-pixel-helper"></div>

          <div class="bd-content">
            <div class="bd-article-container">
              
              <div class="bd-header-article d-print-none">
<div class="header-article-items header-article__inner">
  
    <div class="header-article-items__start">
      
        <div class="header-article-item"><button class="sidebar-toggle primary-toggle btn btn-sm" title="Toggle primary sidebar" data-bs-placement="bottom" data-bs-toggle="tooltip">
  <span class="fa-solid fa-bars"></span>
</button></div>
      
    </div>
  
  
    <div class="header-article-items__end">
      
        <div class="header-article-item">

<div class="article-header-buttons">





<div class="dropdown dropdown-source-buttons">
  <button class="btn dropdown-toggle" type="button" data-bs-toggle="dropdown" aria-expanded="false" aria-label="Source repositories">
    <i class="fab fa-github"></i>
  </button>
  <ul class="dropdown-menu">
      
      
      
      <li><a href="https://github.com/theislab/single-cell-best-practices" target="_blank"
   class="btn btn-sm btn-source-repository-button dropdown-item"
   title="Source repository"
   data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fab fa-github"></i>
  </span>
<span class="btn__text-container">Repository</span>
</a>
</li>
      
      
      
      
      <li><a href="https://github.com/theislab/single-cell-best-practices/edit/master/jupyter-book/air_repertoire/specificity.ipynb" target="_blank"
   class="btn btn-sm btn-source-edit-button dropdown-item"
   title="Suggest edit"
   data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-pencil-alt"></i>
  </span>
<span class="btn__text-container">Suggest edit</span>
</a>
</li>
      
      
      
      
      <li><a href="https://github.com/theislab/single-cell-best-practices/issues/new?title=Issue%20on%20page%20%2Fair_repertoire/specificity.html&body=Your%20issue%20content%20here." target="_blank"
   class="btn btn-sm btn-source-issues-button dropdown-item"
   title="Open an issue"
   data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-lightbulb"></i>
  </span>
<span class="btn__text-container">Open issue</span>
</a>
</li>
      
  </ul>
</div>






<div class="dropdown dropdown-download-buttons">
  <button class="btn dropdown-toggle" type="button" data-bs-toggle="dropdown" aria-expanded="false" aria-label="Download this page">
    <i class="fas fa-download"></i>
  </button>
  <ul class="dropdown-menu">
      
      
      
      <li><a href="../_sources/air_repertoire/specificity.ipynb" target="_blank"
   class="btn btn-sm btn-download-source-button dropdown-item"
   title="Download source file"
   data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-file"></i>
  </span>
<span class="btn__text-container">.ipynb</span>
</a>
</li>
      
      
      
      
      <li>
<button onclick="window.print()"
  class="btn btn-sm btn-download-pdf-button dropdown-item"
  title="Print to PDF"
  data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-file-pdf"></i>
  </span>
<span class="btn__text-container">.pdf</span>
</button>
</li>
      
  </ul>
</div>




<button onclick="toggleFullScreen()"
  class="btn btn-sm btn-fullscreen-button"
  title="Fullscreen mode"
  data-bs-placement="bottom" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-expand"></i>
  </span>

</button>



<script>
document.write(`
  <button class="btn btn-sm nav-link pst-navbar-icon theme-switch-button" title="light/dark" aria-label="light/dark" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="theme-switch fa-solid fa-sun fa-lg" data-mode="light"></i>
    <i class="theme-switch fa-solid fa-moon fa-lg" data-mode="dark"></i>
    <i class="theme-switch fa-solid fa-circle-half-stroke fa-lg" data-mode="auto"></i>
  </button>
`);
</script>


<script>
document.write(`
  <button class="btn btn-sm pst-navbar-icon search-button search-button__button" title="Search" aria-label="Search" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="fa-solid fa-magnifying-glass fa-lg"></i>
  </button>
`);
</script>
<button class="sidebar-toggle secondary-toggle btn btn-sm" title="Toggle secondary sidebar" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <span class="fa-solid fa-list"></span>
</button>
</div></div>
      
    </div>
  
</div>
</div>
              
              

<div id="jb-print-docs-body" class="onlyprint">
    <h1>Specificity analysis</h1>
    <!-- Table of contents -->
    <div id="print-main-content">
        <div id="jb-print-toc">
            
            <div>
                <h2> Contents </h2>
            </div>
            <nav aria-label="Page">
                <ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#introduction">43.1. Introduction</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#tcr-specificity-analysis">43.2. TCR specificity analysis</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#database-queries">43.2.1. Database Queries</a><ul class="nav section-nav flex-column">
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#manual-query">43.2.1.1. Manual Query</a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#query-with-various-strictness">43.2.1.2. Query with various Strictness</a></li>
</ul>
</li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#distances">43.2.2. Distances</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#tcrdist">43.2.3. TCRdist</a><ul class="nav section-nav flex-column">
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#tcrmatch">43.2.3.1. TCRmatch</a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#database-query-via-distance-metrics">43.2.3.2. Database query via distance metrics</a></li>
</ul>
</li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#epitope-prediction">43.2.4. Epitope Prediction</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#bcr-specificity-analysis">43.3. BCR specificity analysis</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#id13">43.3.1. Database Queries</a><ul class="nav section-nav flex-column">
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#database-preparation">43.3.1.1. Database Preparation</a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#identical-matches">43.3.1.2. Identical matches</a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#query-via-hamming-distance">43.3.1.3. Query via Hamming Distance</a></li>
</ul>
</li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#distance-measurements">43.3.2. Distance Measurements</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#prediction">43.3.3. Prediction</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#quiz">43.4. Quiz</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#references">43.5. References</a></li>
</ul>
            </nav>
        </div>
    </div>
</div>

              
                
<div id="searchbox"></div>
                <article class="bd-article">
                  
  <section class="tex2jax_ignore mathjax_ignore" id="specificity-analysis">
<h1><span class="section-number">43. </span>Specificity analysis<a class="headerlink" href="#specificity-analysis" title="Link to this heading">#</a></h1>
<details class="sd-sphinx-override sd-dropdown sd-card sd-mb-3">
<summary class="sd-summary-title sd-card-header">
<span class="sd-summary-text"><i class="fas fa-brain"></i>   Key takeaways</span><span class="sd-summary-state-marker sd-summary-chevron-right"><svg version="1.1" width="1.5em" height="1.5em" class="sd-octicon sd-octicon-chevron-right" viewBox="0 0 24 24" aria-hidden="true"><path d="M8.72 18.78a.75.75 0 0 1 0-1.06L14.44 12 8.72 6.28a.751.751 0 0 1 .018-1.042.751.751 0 0 1 1.042-.018l6.25 6.25a.75.75 0 0 1 0 1.06l-6.25 6.25a.75.75 0 0 1-1.06 0Z"></path></svg></span></summary><div class="sd-summary-content sd-card-body docutils">
<div class="sd-card sd-sphinx-override sd-mb-3 sd-shadow-sm sd-card-hover docutils">
<div class="sd-card-body docutils">
<p class="sd-card-text">The AIR sequence determines the epitope-specificity of the cell.
Cells with similar AIR sequence, bind to the same antigen.</p>
</div>
<a class="sd-stretched-link sd-hide-link-text reference internal" href="#air-repertoire-specificity-key-takeaway-1"><span class="std std-ref">Epitope Prediction</span></a></div>
<div class="sd-card sd-sphinx-override sd-mb-3 sd-shadow-sm sd-card-hover docutils">
<div class="sd-card-body docutils">
<p class="sd-card-text">Specificity can be inferred via Database queries, AIR comparison, or prediction.</p>
</div>
<a class="sd-stretched-link sd-hide-link-text reference internal" href="#air-repertoire-specificity-key-takeaway-2"><span class="std std-ref">Database Queries</span></a></div>
<div class="sd-card sd-sphinx-override sd-mb-3 sd-shadow-sm sd-card-hover docutils">
<div class="sd-card-body docutils">
<p class="sd-card-text">Most approaches are not independently benchmarked and should be used with some caution and additional validation.</p>
</div>
<a class="sd-stretched-link sd-hide-link-text reference internal" href="#air-repertoire-specificity-key-takeaway-3"><span class="std std-ref">Epitope Prediction</span></a></div>
</div>
</details><details class="sd-sphinx-override sd-dropdown sd-card sd-mb-3">
<summary class="sd-summary-title sd-card-header">
<span class="sd-summary-text"><i class="fa-solid fa-gear"></i>   Environment setup</span><span class="sd-summary-state-marker sd-summary-chevron-right"><svg version="1.1" width="1.5em" height="1.5em" class="sd-octicon sd-octicon-chevron-right" viewBox="0 0 24 24" aria-hidden="true"><path d="M8.72 18.78a.75.75 0 0 1 0-1.06L14.44 12 8.72 6.28a.751.751 0 0 1 .018-1.042.751.751 0 0 1 1.042-.018l6.25 6.25a.75.75 0 0 1 0 1.06l-6.25 6.25a.75.75 0 0 1-1.06 0Z"></path></svg></span></summary><div class="sd-summary-content sd-card-body docutils">
<div class="sd-tab-set docutils">
<input checked="checked" id="sd-tab-item-0" name="sd-tab-set-0" type="radio">
<label class="sd-tab-label" for="sd-tab-item-0">
Steps</label><div class="sd-tab-content docutils">
<ol class="arabic">
<li><p class="sd-card-text"><strong>Install conda</strong>:</p>
<ul class="simple">
<li><p class="sd-card-text">Before creating the environment, ensure that <a class="reference external" href="https://docs.conda.io/projects/conda/en/latest/user-guide/install/index.html">conda</a> is installed on your system.</p></li>
</ul>
</li>
<li><p class="sd-card-text"><strong>Save the yml content</strong>:</p>
<ul class="simple">
<li><p class="sd-card-text">Copy the content from the yml tab into a file named <code class="docutils literal notranslate"><span class="pre">environment.yml</span></code>.</p></li>
</ul>
</li>
<li><p class="sd-card-text"><strong>Create the environment</strong>:</p>
<ul>
<li><p class="sd-card-text">Open a terminal or command prompt.</p></li>
<li><p class="sd-card-text">Run the following command:</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>conda<span class="w"> </span>env<span class="w"> </span>create<span class="w"> </span>-f<span class="w"> </span>environment.yml
</pre></div>
</div>
</li>
</ul>
</li>
<li><p class="sd-card-text"><strong>Activate the environment</strong>:</p>
<ul>
<li><p class="sd-card-text">After the environment is created, activate it using:</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>conda<span class="w"> </span>activate<span class="w"> </span>&lt;environment_name&gt;
</pre></div>
</div>
</li>
<li><p class="sd-card-text">Replace <code class="docutils literal notranslate"><span class="pre">&lt;environment_name&gt;</span></code> with the name specified in the <code class="docutils literal notranslate"><span class="pre">environment.yml</span></code> file. In the yml file it will look like this:</p>
<div class="highlight-yaml notranslate"><div class="highlight"><pre><span></span><span class="nt">name</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">&lt;environment_name&gt;</span>
</pre></div>
</div>
</li>
</ul>
</li>
<li><p class="sd-card-text"><strong>Verify the installation</strong>:</p>
<ul>
<li><p class="sd-card-text">Check that the environment was created successfully by running:</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>conda<span class="w"> </span>env<span class="w"> </span>list
</pre></div>
</div>
</li>
</ul>
</li>
</ol>
</div>
<input id="sd-tab-item-1" name="sd-tab-set-0" type="radio">
<label class="sd-tab-label" for="sd-tab-item-1">
yml</label><div class="sd-tab-content docutils">
</div>
</div>
</div>
</details><section id="introduction">
<h2><span class="section-number">43.1. </span>Introduction<a class="headerlink" href="#introduction" title="Link to this heading">#</a></h2>
<p>B- and T-cells recognize their targets via their AIR. Their specificity is determined via the amino acid sequence. Previous studies provided evidence, that the most influencing factor of AIR-target interaction is the CDR3 sequence of the VDJ-chain and to a lesser degree of the CDR3-VJ chain <span id="id1">[<a class="reference internal" href="#id487" title="Ido Springer, Nili Tickotsky, and Yoram Louzoun. Contribution of t cell receptor alpha and beta cdr3, mhc typing, v and j genes to peptide binding prediction. Frontiers in immunology, 2021.">Springer <em>et al.</em>, 2021</a>]</span>.</p>
<p>In many studies, determining the individual cell’s specificity is of key interest: researcher’s can then select the cells relevant to their study and observe their behavior. Specificity screening can be performed via pMHC multimers (TCRs) or antigens (BCRs) tagged with feature barcodes. However, this adds another level of complexity onto study design and further increases cost.
Computational approaches for inferring specificity can provide this layer of information.</p>
<p>Here, we will introduce the following approaches:</p>
<ul class="simple">
<li><p><strong>Database Query</strong>: IR sequences and their targets from various studies were collected in multiple databases. We can use these to find matches to our single-cell study.</p></li>
<li><p><strong>Clustering and distances</strong>: <span id="id2">[<a class="reference internal" href="#id499" title="Jacob Glanville, Huang Huang, Allison Nau, Olivia Hatton, Lisa E Wagar, Florian Rubelt, Xuhuai Ji, Arnold Han, Sheri M Krams, Christina Pettus, and others. Identifying specificity groups in the t cell receptor repertoire. Nature, 547(7661):94–98, 2017.">Glanville <em>et al.</em>, 2017</a>]</span> showed, that IRs with similar receptors have common specificity. This property has been used in multiple approaches for comparing AIRs with distance metrics and unsupervised clustering.</p></li>
<li><p><strong>Epitope prediction</strong>: Recently, several machine-learning methods were developed that directly predict binding between AIRs and a target. In theory, these methods could be used to directly assign specificity to the AIRs involved in single-cell studies.</p></li>
</ul>
<p>However, all three approaches have major pitfalls. The amount of samples in the public databases is severely biased towards diseases and use cases that are commonly researched. Examplatory, this leads to known bindings for only several 100 epitopes sequences for TCRs in the major public databases. Further, a majority amount of samples in these databases does not provide the full AIR sequence (V-, (D-,) and J-genes and CDRs for both chains), but rather focuses on the CDR3, while often only reporting VJ or VDJ sequences.</p>
</section>
<section id="tcr-specificity-analysis">
<h2><span class="section-number">43.2. </span>TCR specificity analysis<a class="headerlink" href="#tcr-specificity-analysis" title="Link to this heading">#</a></h2>
<p>In the following, we will showcase these approaches for TCRs. Previous studies <span id="id3">[<a class="reference internal" href="#id499" title="Jacob Glanville, Huang Huang, Allison Nau, Olivia Hatton, Lisa E Wagar, Florian Rubelt, Xuhuai Ji, Arnold Han, Sheri M Krams, Christina Pettus, and others. Identifying specificity groups in the t cell receptor repertoire. Nature, 547(7661):94–98, 2017.">Glanville <em>et al.</em>, 2017</a>, <a class="reference internal" href="#id500" title="Mark M Davis and Pamela J Bjorkman. T-cell antigen receptor genes and t-cell recognition. Nature, 334(6181):395–402, 1988.">Davis and Bjorkman, 1988</a>, <a class="reference internal" href="#id501" title="Markus G Rudolph, Robyn L Stanfield, and Ian A Wilson. How tcrs bind mhcs, peptides, and coreceptors. Annual review of immunology, 24(1):419–466, 2006.">Rudolph <em>et al.</em>, 2006</a>]</span> showed that the TCR is in close contact with the pMHC at the CDR3, specifically, of the β-chain. This is in line with the regions of high variability of the TCR sequence. In <span id="id4">[<a class="reference internal" href="#id487" title="Ido Springer, Nili Tickotsky, and Yoram Louzoun. Contribution of t cell receptor alpha and beta cdr3, mhc typing, v and j genes to peptide binding prediction. Frontiers in immunology, 2021.">Springer <em>et al.</em>, 2021</a>]</span> by Springer et al., the importance of different information (elements of the TCR and MHC type) was reported for training a sequence-based classifier, which are mostly in line with these findings:</p>
<p>CDR3β &gt; V- and J-gene &gt; CDR3α &gt; MHC &gt; cell type</p>
<p>We can assume that this importance is similar not only for prediction, but also for querying, clustering, and distance calculation.</p>
<div class="admonition warning">
<p class="admonition-title">Warning</p>
<p>Scirpy changed the format of <a class="reference external" href="https://scirpy.scverse.org/en/latest/data-structure.html#storing-airr-rearrangement-data-in-anndata">its datastructure</a>
with v0.13. While the overall analysis workflow has not changed, some outputs shown in this chapter might not be accurate anymore.</p>
<p>See <a class="reference external" href="https://scirpy.scverse.org/en/latest/changelog.html#v0-13-0-new-data-structure-based-on-awkward-arrays">the scirpy release notes</a> for more details about this change.
Until we update this chapter, please also refer to the <a class="reference external" href="https://scirpy.scverse.org">official scirpy documentation</a>.</p>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span><span class="w"> </span><span class="nn">warnings</span>

<span class="n">warnings</span><span class="o">.</span><span class="n">filterwarnings</span><span class="p">(</span>
    <span class="s2">&quot;ignore&quot;</span><span class="p">,</span>
    <span class="s2">&quot;.*IProgress not found*&quot;</span><span class="p">,</span>
<span class="p">)</span>

<span class="kn">import</span><span class="w"> </span><span class="nn">matplotlib.pyplot</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">plt</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">pandas</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">pd</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">scanpy</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">sc</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">scirpy</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">ir</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">seaborn</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">sb</span>
</pre></div>
</div>
</div>
</div>
<p>Let’s load the data from our previous notebooks.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">path_data</span> <span class="o">=</span> <span class="s2">&quot;data&quot;</span>
<span class="n">path_tcr</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">path_data</span><span class="si">}</span><span class="s2">/TCR_01_preprocessed.h5ad&quot;</span>
<span class="n">adata_tcr</span> <span class="o">=</span> <span class="n">sc</span><span class="o">.</span><span class="n">read</span><span class="p">(</span><span class="n">path_tcr</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># TODO final decision sampling: prosponed until remaining chapters are fixed</span>
<span class="n">adata_tcr</span> <span class="o">=</span> <span class="n">adata_tcr</span><span class="p">[</span><span class="n">adata_tcr</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s2">&quot;patient_id&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">isin</span><span class="p">([</span><span class="s2">&quot;CV0902&quot;</span><span class="p">,</span> <span class="s2">&quot;AP6&quot;</span><span class="p">])]</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
</pre></div>
</div>
</div>
</div>
<section id="database-queries">
<span id="air-repertoire-specificity-key-takeaway-2"></span><h3><span class="section-number">43.2.1. </span>Database Queries<a class="headerlink" href="#database-queries" title="Link to this heading">#</a></h3>
<p>Here, we will search for TCRs with specificity annotation from previous studies. Common large-scale databases with TCR-epitope pairs are:</p>
<ul class="simple">
<li><p><a class="reference external" href="https://www.iedb.org/">IEDB</a> <span id="id5">[<a class="reference internal" href="#id490" title="Ward Fleri, Sinu Paul, Sandeep Kumar Dhanda, Swapnil Mahajan, Xiaojun Xu, Bjoern Peters, and Alessandro Sette. The immune epitope database and analysis resource in epitope discovery and synthetic vaccine design. Frontiers in immunology, 8:278, 2017.">Fleri <em>et al.</em>, 2017</a>]</span></p></li>
<li><p><a class="reference external" href="https://vdjdb.cdr3.net/">vdjDB</a> <span id="id6">[<a class="reference internal" href="#id489" title="Mikhail Shugay, Dmitriy V Bagaev, Ivan V Zvyagin, Renske M Vroomans, Jeremy Chase Crawford, Garry Dolton, Ekaterina A Komech, Anastasiya L Sycheva, Anna E Koneva, Evgeniy S Egorov, and others. Vdjdb: a curated database of t-cell receptor sequences with known antigen specificity. Nucleic acids research, 46(D1):D419–D427, 2018.">Shugay <em>et al.</em>, 2018</a>]</span></p></li>
<li><p><a class="reference external" href="http://friedmanlab.weizmann.ac.il/McPAS-TCR/">McPAS-TCR</a> <span id="id7">[<a class="reference internal" href="#id488" title="Nili Tickotsky, Tal Sagiv, Jaime Prilusky, Eric Shifrut, and Nir Friedman. Mcpas-tcr: a manually curated catalogue of pathology-associated t cell receptor sequences. Bioinformatics, 33(18):2924–2929, 2017.">Tickotsky <em>et al.</em>, 2017</a>]</span></p></li>
<li><p><a class="reference external" href="https://db.cngb.org/pird/">PIRD</a> <span id="id8">[<a class="reference internal" href="#id491" title="Wei Zhang, Longlong Wang, Ke Liu, Xiaofeng Wei, Kai Yang, Wensi Du, Shiyu Wang, Nannan Guo, Chuanchuan Ma, Lihua Luo, and others. Pird: pan immune repertoire database. Bioinformatics, 36(3):897–903, 2020.">Zhang <em>et al.</em>, 2020</a>]</span></p></li>
<li><p><a class="reference external" href="https://clients.adaptivebiotech.com/pub/covid-2020">immuneCode</a><sup>TM</sup> for covid specific TCRs <span id="id9">[<a class="reference internal" href="#id495" title="Sean Nolan, Marissa Vignali, Mark Klinger, Jennifer N Dines, Ian M Kaplan, Emily Svejnoha, Tracy Craft, Katie Boland, Mitch Pesesky, Rachel M Gittelman, and others. A large-scale database of t-cell receptor beta (tcrβ) sequences and binding associations from natural and synthetic exposure to sars-cov-2. Research square, 2020.">Nolan <em>et al.</em>, 2020</a>]</span></p></li>
</ul>
<p>The first four general-purpose databases collect TCR-peptide pairs from a wide variety of diseases, while immuneCode focuses on TCRs related to SARS-CoV-2. However, these databases underlie severe bias towards which epitopes are represented. You will therefore need to check, which database contains sufficient disease-specific TCRs for your study. Depending on your use case, databases or study data with specific content might be better suited.</p>
<p>Overall, there is a trade-off between precision and recall depending on the strictness of our query depending on which information we use for defining clonotypes (see clonotype definition). Unfortunately, there are no general guidelines regarding what information should be taken into account. We will demonstrate various degrees of strictness when querying our data against the VDJdb, since it contains the largest variety of different SARS-CoV-2 epitopes in a general-purpose database.</p>
<p>This database can be conveniently accessed via Scirpy. Information on TCR, epitope, and the experimental setup is stored in <code class="docutils literal notranslate"><span class="pre">vdjdb.obs</span></code> following common Scirpy format.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">vdjdb</span> <span class="o">=</span> <span class="n">ir</span><span class="o">.</span><span class="n">datasets</span><span class="o">.</span><span class="n">vdjdb</span><span class="p">()</span>
<span class="n">vdjdb</span><span class="o">.</span><span class="n">obs</span><span class="o">.</span><span class="n">head</span><span class="p">(</span><span class="mi">5</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>multi_chain</th>
      <th>species</th>
      <th>mhc.a</th>
      <th>mhc.b</th>
      <th>mhc.class</th>
      <th>antigen.epitope</th>
      <th>antigen.gene</th>
      <th>antigen.species</th>
      <th>reference.id</th>
      <th>method.identification</th>
      <th>...</th>
      <th>IR_VDJ_2_sequence_id</th>
      <th>IR_VJ_1_v_call</th>
      <th>IR_VJ_2_v_call</th>
      <th>IR_VDJ_1_v_call</th>
      <th>IR_VDJ_2_v_call</th>
      <th>IR_VJ_1_v_cigar</th>
      <th>IR_VJ_2_v_cigar</th>
      <th>IR_VDJ_1_v_cigar</th>
      <th>IR_VDJ_2_v_cigar</th>
      <th>has_ir</th>
    </tr>
    <tr>
      <th>cell_id</th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>False</td>
      <td>HomoSapiens</td>
      <td>HLA-B*08</td>
      <td>B2M</td>
      <td>MHCI</td>
      <td>FLKEKGGL</td>
      <td>Nef</td>
      <td>HIV-1</td>
      <td>PMID:15596521</td>
      <td>tetramer-sort</td>
      <td>...</td>
      <td>NaN</td>
      <td>TRAV26-1*01</td>
      <td>NaN</td>
      <td>TRBV13*01</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>True</td>
    </tr>
    <tr>
      <th>1</th>
      <td>False</td>
      <td>HomoSapiens</td>
      <td>HLA-B*08</td>
      <td>B2M</td>
      <td>MHCI</td>
      <td>FLKEKGGL</td>
      <td>Nef</td>
      <td>HIV-1</td>
      <td>PMID:15596521</td>
      <td>tetramer-sort</td>
      <td>...</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>TRBV13*01</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>True</td>
    </tr>
    <tr>
      <th>2</th>
      <td>False</td>
      <td>HomoSapiens</td>
      <td>HLA-B*08</td>
      <td>B2M</td>
      <td>MHCI</td>
      <td>FLKEKGGL</td>
      <td>Nef</td>
      <td>HIV-1</td>
      <td>PMID:15596521</td>
      <td>tetramer-sort</td>
      <td>...</td>
      <td>NaN</td>
      <td>TRAV20*01</td>
      <td>NaN</td>
      <td>TRBV13*01</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>True</td>
    </tr>
    <tr>
      <th>3</th>
      <td>False</td>
      <td>HomoSapiens</td>
      <td>HLA-B*08</td>
      <td>B2M</td>
      <td>MHCI</td>
      <td>FLKEKGGL</td>
      <td>Nef</td>
      <td>HIV-1</td>
      <td>PMID:15596521</td>
      <td>tetramer-sort</td>
      <td>...</td>
      <td>NaN</td>
      <td>TRAV2*01</td>
      <td>NaN</td>
      <td>TRBV13*01</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>True</td>
    </tr>
    <tr>
      <th>4</th>
      <td>False</td>
      <td>HomoSapiens</td>
      <td>HLA-B*08</td>
      <td>B2M</td>
      <td>MHCI</td>
      <td>FLKEKGGL</td>
      <td>Nef</td>
      <td>HIV-1</td>
      <td>PMID:15596521</td>
      <td>tetramer-sort</td>
      <td>...</td>
      <td>NaN</td>
      <td>TRAV38-2/DV8*01</td>
      <td>NaN</td>
      <td>TRBV14*01</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>True</td>
    </tr>
  </tbody>
</table>
<p>5 rows × 92 columns</p>
</div></div></div>
</div>
<p>VDJdb provides us with various additional information of TCR-epitope pairs, such as the method of identification, or the associated disease. Let’s show what epitopes and diseases are contained in this database.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">vdjdb</span><span class="p">[</span><span class="n">vdjdb</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s2">&quot;antigen.species&quot;</span><span class="p">]</span> <span class="o">==</span> <span class="s2">&quot;SARS-CoV-2&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span>
    <span class="s2">&quot;antigen.epitope&quot;</span>
<span class="p">]</span><span class="o">.</span><span class="n">value_counts</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>YLQPRTFLL     1326
SPRWYFYYL      309
TTDPSFLGRY     254
RLQSLQTYV      150
LTDEMIAQY      135
              ... 
LQAENVTGL        1
LPSYAAFAT        1
LPPVYTNSF        1
LPPANTNSF        1
YYTSNPTTF        1
Name: antigen.epitope, Length: 667, dtype: int64
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">vdjdb</span><span class="p">[</span><span class="n">vdjdb</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s2">&quot;antigen.species&quot;</span><span class="p">]</span> <span class="o">==</span> <span class="s2">&quot;SARS-CoV-2&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span>
    <span class="s2">&quot;antigen.species&quot;</span>
<span class="p">]</span><span class="o">.</span><span class="n">value_counts</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>SARS-CoV-2    4566
Name: antigen.species, dtype: int64
</pre></div>
</div>
</div>
</div>
<p>We can easily observe, that a large fraction of sample points stem from the SARS-CoV-2 derived peptide YLQPRTFLL. This is then also reflected in the distribution of diseases within the database.</p>
<section id="manual-query">
<h4><span class="section-number">43.2.1.1. </span>Manual Query<a class="headerlink" href="#manual-query" title="Link to this heading">#</a></h4>
<p>Queries between two datasets are already integrated in Scirpy and other toolkits such as scRepertoire. To query the single-cell data to a new database it is therefore most convenient, to convert the database to the toolkits format and use the toolkit’s functions. However, since they are not available in every toolkit, and we might want to add more constraints (e.g. MHC type), we will show an example how to manually perform a basic query, which can be adapted for other databases, toolkits and constraints.</p>
<p>First, we will reduce the vdjdb to samples that have overlapping CDR3β sequence to our data. This will greatly reduce the size of the database and thereby reduce computational time.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Amount of samples in VDJDB: </span><span class="si">{</span><span class="nb">len</span><span class="p">(</span><span class="n">vdjdb</span><span class="p">)</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
<span class="n">vdjdb_overlap</span> <span class="o">=</span> <span class="n">vdjdb</span><span class="p">[</span>
    <span class="n">vdjdb</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s2">&quot;IR_VDJ_1_junction_aa&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">isin</span><span class="p">(</span><span class="n">adata_tcr</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s2">&quot;IR_VDJ_1_junction_aa&quot;</span><span class="p">])</span>
<span class="p">]</span><span class="o">.</span><span class="n">obs</span>
<span class="n">vdjdb_overlap</span> <span class="o">=</span> <span class="n">vdjdb_overlap</span><span class="p">[[</span><span class="s2">&quot;IR_VDJ_1_junction_aa&quot;</span><span class="p">,</span> <span class="s2">&quot;antigen.species&quot;</span><span class="p">]]</span>
<span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Amount of overlapping samples in VDJDB: </span><span class="si">{</span><span class="nb">len</span><span class="p">(</span><span class="n">vdjdb_overlap</span><span class="p">)</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Amount of samples in VDJDB:  60055
Amount of overlapping samples in VDJDB:  7761
</pre></div>
</div>
</div>
</div>
<p>Second, we will annotate the columns in our query data, that have an overlap to the database. We will later use this annotation to work only on the data, where there exists an overlap.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">adata_tcr</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s2">&quot;has_vdjdb_overlap&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">adata_tcr</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s2">&quot;IR_VDJ_1_junction_aa&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">isin</span><span class="p">(</span>
    <span class="n">vdjdb_overlap</span><span class="p">[</span><span class="s2">&quot;IR_VDJ_1_junction_aa&quot;</span><span class="p">]</span>
<span class="p">)</span>
<span class="n">adata_tcr</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s2">&quot;has_vdjdb_overlap&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">value_counts</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>False    11522
True      1293
Name: has_vdjdb_overlap, dtype: int64
</pre></div>
</div>
</div>
</div>
<p>Next, we need to define a function that returns the desired annotation based on the input of various features. Here, we will query for identical CDR3β chains, but in theory, other constraints can be added here as well. Based on this sequence, all entries of VDJdb with the same sequence are selected. From here, we collect all diseases associated with these elements. If several diseases are associated we will return ‘ambiguous’, otherwise the disease.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="k">def</span><span class="w"> </span><span class="nf">assign_disease</span><span class="p">(</span><span class="n">cdr3beta</span><span class="p">):</span>
    <span class="n">matching_rows</span> <span class="o">=</span> <span class="n">vdjdb_overlap</span><span class="p">[</span><span class="n">vdjdb_overlap</span><span class="p">[</span><span class="s2">&quot;IR_VDJ_1_junction_aa&quot;</span><span class="p">]</span> <span class="o">==</span> <span class="n">cdr3beta</span><span class="p">]</span>
    <span class="n">diseases</span> <span class="o">=</span> <span class="n">matching_rows</span><span class="p">[</span><span class="s2">&quot;antigen.species&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">values</span><span class="o">.</span><span class="n">tolist</span><span class="p">()</span>
    <span class="n">diseases</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="nb">set</span><span class="p">(</span><span class="n">diseases</span><span class="p">))</span>
    <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">diseases</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">1</span><span class="p">:</span>
        <span class="k">return</span> <span class="s2">&quot;ambiguous&quot;</span>
    <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">diseases</span><span class="p">)</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
        <span class="k">return</span> <span class="s2">&quot;no entry&quot;</span>
    <span class="k">return</span> <span class="n">diseases</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
</pre></div>
</div>
</div>
</div>
<p>Finally, we apply the function to our query data. The resulting column provides us with a per cell annotation of disease linked to that TCR.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">adata_tcr</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s2">&quot;antigen.species_manual&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="p">(</span>
    <span class="n">adata_tcr</span><span class="p">[</span><span class="n">adata_tcr</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s2">&quot;has_vdjdb_overlap&quot;</span><span class="p">]]</span>
    <span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s2">&quot;IR_VDJ_1_junction_aa&quot;</span><span class="p">]</span>
    <span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="n">assign_disease</span><span class="p">)</span>
<span class="p">)</span>
<span class="n">adata_tcr</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s2">&quot;antigen.species_manual&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">value_counts</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>CMV                 601
EBV                  83
HIV-1                70
ambiguous            55
SARS-CoV-2           12
HomoSapiens           8
InfluenzaA            7
HCV                   4
TriticumAestivum      1
YFV                   1
DENV2                 1
MCMV                  1
SIV                   1
Name: antigen.species_manual, dtype: int64
</pre></div>
</div>
</div>
</div>
<p>We observe several matches towards different diseases. These annotations must be viewed carefully, since they depend on the abundance of the disease in the databases, wrong annotation in the databases, false matches due to incomplete information, and possible MHC restrictions of epitopes. However, from the great abundance of disease-specific TCRs, we could deduct, that one patient has a latent infection towards the Cytomegalovirus (CMV) and Epstein–Barr virus (EBV), which are common in population. However, the third most common disease association is toward the less frequent HIV-1. # todo: might change when different donors are chosen</p>
</section>
<section id="query-with-various-strictness">
<h4><span class="section-number">43.2.1.2. </span>Query with various Strictness<a class="headerlink" href="#query-with-various-strictness" title="Link to this heading">#</a></h4>
<p>The same query can be performed via toolkits such as Scirpy. First, we calculate the overlap between the TCRs of Query (our single-cell data) and database (here: vdjDB). The resulting matrices for VJ and VDJ receptor (n_ours x n_vdjdb) will contain the entry 1 where the TCRs match between both datasets and a 0 otherwise.</p>
<p>We provide the following additional information to the function call:</p>
<ul class="simple">
<li><p><strong>metric=’identity’</strong>: For now, we only consider exact sequence matches. We will later introduce queries for similar sequences as well.</p></li>
<li><p><strong>sequence=’aa’</strong>: TCR sequences are compared on an amino-acid level instead of nucleobases, since specificity depends on the protein structure and this information is provided by the VDJdb.</p></li>
</ul>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">metric</span> <span class="o">=</span> <span class="s2">&quot;identity&quot;</span>
<span class="n">sequence</span> <span class="o">=</span> <span class="s2">&quot;aa&quot;</span>
<span class="n">ir</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">ir_dist</span><span class="p">(</span><span class="n">adata_tcr</span><span class="p">,</span> <span class="n">vdjdb</span><span class="p">,</span> <span class="n">metric</span><span class="o">=</span><span class="n">metric</span><span class="p">,</span> <span class="n">sequence</span><span class="o">=</span><span class="n">sequence</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<p>Next, we will match all cells between Query and database based on the condition we provide:</p>
<ul class="simple">
<li><p><strong>metric=’identity’</strong>, <strong>sequence=’aa’</strong>: we need to provide the same values as used during distance calculation</p></li>
<li><p><strong>receptor_arms=’VDJ’</strong>: Compare TCRs based on CDR3β. Other options are ‘VJ’ (α-chain), ‘all’ (α- and β-chain), and ‘any’ (either α- or β-chain).</p></li>
<li><p><strong>dual_ir=’primary_only’</strong>: as discussed in the first chapter, T cells can carry a secondary TCR. This parameter determines on which receptor the query is conducted. Other options are ‘any’ (primary or secondary receptor), and ‘all’ (primary and secondary receptor).</p></li>
</ul>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">ir</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">ir_query</span><span class="p">(</span>
    <span class="n">adata_tcr</span><span class="p">,</span>
    <span class="n">vdjdb</span><span class="p">,</span>
    <span class="n">metric</span><span class="o">=</span><span class="n">metric</span><span class="p">,</span>
    <span class="n">sequence</span><span class="o">=</span><span class="n">sequence</span><span class="p">,</span>
    <span class="n">receptor_arms</span><span class="o">=</span><span class="s2">&quot;VDJ&quot;</span><span class="p">,</span>
    <span class="n">dual_ir</span><span class="o">=</span><span class="s2">&quot;primary_only&quot;</span><span class="p">,</span>
<span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>100%|█████████████████████████████████████| 6191/6191 [00:04&lt;00:00, 1292.12it/s]
</pre></div>
</div>
</div>
</div>
<p>Finally, we can annotate the TCRs in the single-cell data with the annotation of matching entries of the database.</p>
<ul class="simple">
<li><p><strong>metric=’identity’</strong>, <strong>sequence=’aa’</strong>: Use the same values as provided earlier.</p></li>
<li><p><strong>include_ref_cols</strong>: list of columns from the reference database, that will be added to the query database. Here, we will add the species of the antigen, and the epitope sequence.</p></li>
<li><p><strong>suffix</strong>: Here, we mark the different queries we will conduct with a suffix.</p></li>
</ul>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">ir</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">ir_query_annotate</span><span class="p">(</span>
    <span class="n">adata_tcr</span><span class="p">,</span>
    <span class="n">vdjdb</span><span class="p">,</span>
    <span class="n">metric</span><span class="o">=</span><span class="n">metric</span><span class="p">,</span>
    <span class="n">sequence</span><span class="o">=</span><span class="n">sequence</span><span class="p">,</span>
    <span class="n">include_ref_cols</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;antigen.species&quot;</span><span class="p">,</span> <span class="s2">&quot;antigen.epitope&quot;</span><span class="p">],</span>
<span class="p">)</span>
<span class="n">adata_tcr</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s2">&quot;antigen.species&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">value_counts</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>100%|█████████████████████████████████████| 1690/1690 [00:00&lt;00:00, 5387.03it/s]
</pre></div>
</div>
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>CMV                 601
EBV                  83
HIV-1                70
ambiguous            55
SARS-CoV-2           12
HomoSapiens           8
InfluenzaA            7
HCV                   4
TriticumAestivum      1
YFV                   1
DENV2                 1
MCMV                  1
SIV                   1
Name: antigen.species, dtype: int64
</pre></div>
</div>
</div>
</div>
<p>We here obtain the same results as our manual query above. The advantage is, that it required less custom code. Also, distance metrics are already implemented, when not querying for exact matches.</p>
<p>Next we perform a similar query, but regarding only the α-chain. We will indicate this assignment by using the suffix ‘_VJ’. Since we already calculated the matching matrix earlier, we only need to perform the annotation step.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">ir</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">ir_query</span><span class="p">(</span>
    <span class="n">adata_tcr</span><span class="p">,</span>
    <span class="n">vdjdb</span><span class="p">,</span>
    <span class="n">metric</span><span class="o">=</span><span class="n">metric</span><span class="p">,</span>
    <span class="n">sequence</span><span class="o">=</span><span class="n">sequence</span><span class="p">,</span>
    <span class="n">receptor_arms</span><span class="o">=</span><span class="s2">&quot;VJ&quot;</span><span class="p">,</span>
    <span class="n">dual_ir</span><span class="o">=</span><span class="s2">&quot;primary_only&quot;</span><span class="p">,</span>
<span class="p">)</span>
<span class="n">ir</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">ir_query_annotate</span><span class="p">(</span>
    <span class="n">adata_tcr</span><span class="p">,</span>
    <span class="n">vdjdb</span><span class="p">,</span>
    <span class="n">metric</span><span class="o">=</span><span class="n">metric</span><span class="p">,</span>
    <span class="n">sequence</span><span class="o">=</span><span class="n">sequence</span><span class="p">,</span>
    <span class="n">include_ref_cols</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;antigen.species&quot;</span><span class="p">,</span> <span class="s2">&quot;antigen.epitope&quot;</span><span class="p">],</span>
    <span class="n">suffix</span><span class="o">=</span><span class="s2">&quot;_VJ&quot;</span><span class="p">,</span>
<span class="p">)</span>
<span class="n">adata_tcr</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s2">&quot;antigen.species_VJ&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">value_counts</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>100%|█████████████████████████████████████| 5209/5209 [00:03&lt;00:00, 1536.95it/s]
100%|█████████████████████████████████████| 4064/4064 [00:00&lt;00:00, 5683.50it/s]
</pre></div>
</div>
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>ambiguous       927
CMV             466
InfluenzaA      349
MCMV            143
SARS-CoV-2       55
EBV              43
HomoSapiens      27
HCV              11
HIV-1             6
YFV               3
HSV-2             1
Homo sapiens      1
Name: antigen.species_VJ, dtype: int64
</pre></div>
</div>
</div>
</div>
<p>Note, that we receive more matches of the α-chain, since it is less variable than the β-chain. This results in many TCRs, that cannot be uniquely assigned to a specific disease and are therefore ‘ambiguous’. As expected, we can still observe a many matches towards CMV, reinforcing the probability of an infection. However, there is only limited indication for HIV-specific TCRs.</p>
<p>Finally, we query identical sequences on both receptors:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">ir</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">ir_query</span><span class="p">(</span>
    <span class="n">adata_tcr</span><span class="p">,</span>
    <span class="n">vdjdb</span><span class="p">,</span>
    <span class="n">metric</span><span class="o">=</span><span class="n">metric</span><span class="p">,</span>
    <span class="n">sequence</span><span class="o">=</span><span class="n">sequence</span><span class="p">,</span>
    <span class="n">receptor_arms</span><span class="o">=</span><span class="s2">&quot;all&quot;</span><span class="p">,</span>
    <span class="n">dual_ir</span><span class="o">=</span><span class="s2">&quot;primary_only&quot;</span><span class="p">,</span>
<span class="p">)</span>
<span class="n">ir</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">ir_query_annotate</span><span class="p">(</span>
    <span class="n">adata_tcr</span><span class="p">,</span>
    <span class="n">vdjdb</span><span class="p">,</span>
    <span class="n">metric</span><span class="o">=</span><span class="n">metric</span><span class="p">,</span>
    <span class="n">sequence</span><span class="o">=</span><span class="n">sequence</span><span class="p">,</span>
    <span class="n">include_ref_cols</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;antigen.species&quot;</span><span class="p">,</span> <span class="s2">&quot;antigen.epitope&quot;</span><span class="p">],</span>
    <span class="n">suffix</span><span class="o">=</span><span class="s2">&quot;_fullIR&quot;</span><span class="p">,</span>
<span class="p">)</span>
<span class="n">adata_tcr</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s2">&quot;antigen.species_fullIR&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">value_counts</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>100%|██████████████████████████████████████| 6722/6722 [00:07&lt;00:00, 850.87it/s]
100%|███████████████████████████████████████| 328/328 [00:00&lt;00:00, 5088.80it/s]
</pre></div>
</div>
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>CMV            87
InfluenzaA     26
EBV            21
ambiguous      16
HomoSapiens     5
HIV-1           4
YFV             3
SIV             1
SARS-CoV-2      1
Name: antigen.species_fullIR, dtype: int64
</pre></div>
</div>
</div>
</div>
<p>Since this search was more constrained, we receive less matches. However, these matches are likely to be of higher quality.</p>
</section>
</section>
<section id="distances">
<h3><span class="section-number">43.2.2. </span>Distances<a class="headerlink" href="#distances" title="Link to this heading">#</a></h3>
<p>To detect cells with shared specificity, we can calculate pairwise sequence distances between their TCRs. This can be either used to group cells within a dataset or to increase the amount of hits, we receive from database queries. For sequence distances, there are generally three different approaches:</p>
<ul class="simple">
<li><p><strong>Edit distances</strong>: calculate the cost of transforming the first into the second sequence.</p></li>
<li><p><strong>k-mer matching</strong>: compares the occurrence of short motifs of length k between two sequences.</p></li>
<li><p><strong>Embeddings</strong>: the sequence is embedded into a numeric representation (e.g. via deep learning).</p></li>
</ul>
<p>Note, that these approaches have not been independently benchmarked. We will therefore focus here on two selected distance metrics:</p>
<ul class="simple">
<li><p><strong>TCRdist</strong>: this commonly used metric uses the sequences of all CDRs and compares them via transformation cost and gap penalties <span id="id10">[<a class="reference internal" href="#id496" title="Pradyot Dash, Andrew J Fiore-Gartland, Tomer Hertz, George C Wang, Shalini Sharma, Aisha Souquette, Jeremy Chase Crawford, E Bridie Clemens, Thi HO Nguyen, Katherine Kedzierska, and others. Quantifiable predictive features define epitope-specific t cell receptor repertoires. Nature, 547(7661):89–93, 2017.">Dash <em>et al.</em>, 2017</a>]</span>. The costs are based on the BLOSUM matrix, which indicates the probabilities of substituting one amino acid against another <span id="id11">[<a class="reference internal" href="#id503" title="Steven Henikoff and Jorja G Henikoff. Amino acid substitution matrices from protein blocks. Proceedings of the National Academy of Sciences, 89(22):10915–10919, 1992.">Henikoff and Henikoff, 1992</a>]</span>. By incorporating the full sequence, accuracy is most likely increased as compared to other approaches, but limits its use, when only a subset of information is provided.</p></li>
<li><p><strong>TCRmatch</strong>: this novel metric uses all k-mers to compare the overlap in motifs between two TCRs based on their CDR3β sequences <span id="id12">[<a class="reference internal" href="#id497" title="William D Chronister, Austin Crinklaw, Swapnil Mahajan, Randi Vita, Zeynep Koşaloğlu-Yalçın, Zhen Yan, Jason A Greenbaum, Leon E Jessen, Morten Nielsen, Scott Christley, and others. Tcrmatch: predicting t-cell receptor specificity based on sequence similarity to previously characterized receptors. Frontiers in immunology, 12:673, 2021.">Chronister <em>et al.</em>, 2021</a>]</span>. It can therefore, also be utilized on most databases, that mainly contain this information. It is also conveniently integrated into IEDB.</p></li>
</ul>
<p>These metrics use the AIR sequence as input. Cells with the same receptors will therefore have identical output. To reduce computational cost, we therefore first extract the unique AIRs from the AnnData object into a DataFrame.</p>
</section>
<section id="tcrdist">
<h3><span class="section-number">43.2.3. </span>TCRdist<a class="headerlink" href="#tcrdist" title="Link to this heading">#</a></h3>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span><span class="w"> </span><span class="nn">tcrdist.repertoire</span><span class="w"> </span><span class="kn">import</span> <span class="n">TCRrep</span>
</pre></div>
</div>
</div>
</div>
<p>TCRdist requires complete information of CDR3 and V-gene of both chains. We therefore filter all cells missing this information. To avoid problems with categorical entries in the DataFrame, we change the type of all columns to ‘str’. Finally, we remove duplicates (i.e., we only consider clonotypes), to reduce computational cost.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">adata_tcrdist</span> <span class="o">=</span> <span class="n">adata_tcr</span><span class="p">[</span>
    <span class="o">~</span><span class="n">adata_tcr</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s2">&quot;chain_pairing&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">isin</span><span class="p">([</span><span class="s2">&quot;No IR&quot;</span><span class="p">,</span> <span class="s2">&quot;orphan VDJ&quot;</span><span class="p">,</span> <span class="s2">&quot;orphan VJ&quot;</span><span class="p">])</span>
<span class="p">]</span>

<span class="k">for</span> <span class="n">col</span> <span class="ow">in</span> <span class="p">[</span><span class="s2">&quot;IR_VJ_1_v_call&quot;</span><span class="p">,</span> <span class="s2">&quot;IR_VDJ_1_v_call&quot;</span><span class="p">]:</span>
    <span class="n">adata_tcrdist</span> <span class="o">=</span> <span class="n">adata_tcrdist</span><span class="p">[</span><span class="o">~</span><span class="n">adata_tcrdist</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="n">col</span><span class="p">]</span><span class="o">.</span><span class="n">isna</span><span class="p">()]</span>
<span class="n">adata_tcrdist</span> <span class="o">=</span> <span class="n">adata_tcrdist</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>


<span class="n">df_tcrdist</span> <span class="o">=</span> <span class="n">adata_tcrdist</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span>
    <span class="p">[</span>
        <span class="s2">&quot;IR_VJ_1_junction_aa&quot;</span><span class="p">,</span>
        <span class="s2">&quot;IR_VJ_1_v_call&quot;</span><span class="p">,</span>
        <span class="s2">&quot;IR_VJ_1_j_call&quot;</span><span class="p">,</span>
        <span class="s2">&quot;IR_VDJ_1_junction_aa&quot;</span><span class="p">,</span>
        <span class="s2">&quot;IR_VDJ_1_v_call&quot;</span><span class="p">,</span>
        <span class="s2">&quot;IR_VDJ_1_j_call&quot;</span><span class="p">,</span>
        <span class="s2">&quot;chain_pairing&quot;</span><span class="p">,</span>
        <span class="s2">&quot;patient_id&quot;</span><span class="p">,</span>
        <span class="s2">&quot;antigen.species&quot;</span><span class="p">,</span>
        <span class="s2">&quot;initial_clustering&quot;</span><span class="p">,</span>
    <span class="p">]</span>
<span class="p">]</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>

<span class="k">for</span> <span class="n">col</span> <span class="ow">in</span> <span class="n">df_tcrdist</span><span class="o">.</span><span class="n">columns</span><span class="p">:</span>
    <span class="n">df_tcrdist</span><span class="p">[</span><span class="n">col</span><span class="p">]</span> <span class="o">=</span> <span class="n">df_tcrdist</span><span class="p">[</span><span class="n">col</span><span class="p">]</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="nb">str</span><span class="p">)</span>
<span class="n">df_tcrdist</span> <span class="o">=</span> <span class="n">df_tcrdist</span><span class="o">.</span><span class="n">drop_duplicates</span><span class="p">()</span>
<span class="n">df_tcrdist</span> <span class="o">=</span> <span class="n">df_tcrdist</span><span class="o">.</span><span class="n">reset_index</span><span class="p">(</span><span class="n">drop</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<p>We need to adjust our DataFrame to the format required by TCRdist. We therefore rename the columns. Further, TCRdist requires the V-gene subtype, which is not provided in our data. Assuming high similarity between the subtypes, we assign the subtype ‘*01’ to all V-genes. Caution: TCRdist internally re-orders the output DataFrame. We therefore add the index as a column, to later restore the original order.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">dict_rename_tcrdist</span> <span class="o">=</span> <span class="p">{</span>
    <span class="s2">&quot;IR_VJ_1_junction_aa&quot;</span><span class="p">:</span> <span class="s2">&quot;cdr3_a_aa&quot;</span><span class="p">,</span>
    <span class="s2">&quot;IR_VJ_1_v_call&quot;</span><span class="p">:</span> <span class="s2">&quot;v_a_gene&quot;</span><span class="p">,</span>
    <span class="s2">&quot;IR_VDJ_1_junction_aa&quot;</span><span class="p">:</span> <span class="s2">&quot;cdr3_b_aa&quot;</span><span class="p">,</span>
    <span class="s2">&quot;IR_VDJ_1_v_call&quot;</span><span class="p">:</span> <span class="s2">&quot;v_b_gene&quot;</span><span class="p">,</span>
<span class="p">}</span>

<span class="n">df_tcrdist</span> <span class="o">=</span> <span class="n">df_tcrdist</span><span class="o">.</span><span class="n">rename</span><span class="p">(</span><span class="n">columns</span><span class="o">=</span><span class="n">dict_rename_tcrdist</span><span class="p">)</span>

<span class="n">df_tcrdist</span><span class="p">[</span><span class="s2">&quot;v_a_gene&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">df_tcrdist</span><span class="p">[</span><span class="s2">&quot;v_a_gene&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="nb">str</span><span class="p">)</span> <span class="o">+</span> <span class="s2">&quot;*01&quot;</span>
<span class="n">df_tcrdist</span><span class="p">[</span><span class="s2">&quot;v_b_gene&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">df_tcrdist</span><span class="p">[</span><span class="s2">&quot;v_b_gene&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="nb">str</span><span class="p">)</span> <span class="o">+</span> <span class="s2">&quot;*01&quot;</span>

<span class="n">df_tcrdist</span><span class="p">[</span><span class="s2">&quot;count&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="mi">1</span>  <span class="c1"># todo</span>

<span class="n">df_tcrdist</span><span class="p">[</span><span class="s2">&quot;index&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">df_tcrdist</span><span class="o">.</span><span class="n">index</span>

<span class="n">df_tcrdist</span><span class="o">.</span><span class="n">head</span><span class="p">(</span><span class="mi">5</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>cdr3_a_aa</th>
      <th>v_a_gene</th>
      <th>IR_VJ_1_j_call</th>
      <th>cdr3_b_aa</th>
      <th>v_b_gene</th>
      <th>IR_VDJ_1_j_call</th>
      <th>chain_pairing</th>
      <th>patient_id</th>
      <th>antigen.species</th>
      <th>initial_clustering</th>
      <th>counts</th>
      <th>index</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>CAVKRGNNARLMF</td>
      <td>TRAV21*01</td>
      <td>TRAJ31</td>
      <td>CASSQTGGQPQHF</td>
      <td>TRBV5-1*01</td>
      <td>TRBJ1-5</td>
      <td>single pair</td>
      <td>CV0902</td>
      <td>nan</td>
      <td>CD8</td>
      <td>1</td>
      <td>0</td>
    </tr>
    <tr>
      <th>1</th>
      <td>CAVGAPGDDKIIF</td>
      <td>TRAV8-3*01</td>
      <td>TRAJ30</td>
      <td>CASRPSGLNTDTQYF</td>
      <td>TRBV7-9*01</td>
      <td>TRBJ2-3</td>
      <td>single pair</td>
      <td>CV0902</td>
      <td>nan</td>
      <td>CD8</td>
      <td>1</td>
      <td>1</td>
    </tr>
    <tr>
      <th>2</th>
      <td>CAVNSPGGYQKVTF</td>
      <td>TRAV8-1*01</td>
      <td>TRAJ13</td>
      <td>CAISASGGGSSGNTIYF</td>
      <td>TRBV10-3*01</td>
      <td>TRBJ1-3</td>
      <td>single pair</td>
      <td>CV0902</td>
      <td>nan</td>
      <td>CD4</td>
      <td>1</td>
      <td>2</td>
    </tr>
    <tr>
      <th>3</th>
      <td>CAVSPFNAGGGNKLTF</td>
      <td>TRAV3*01</td>
      <td>TRAJ10</td>
      <td>CASSQTSGGTDTQYF</td>
      <td>TRBV18*01</td>
      <td>TRBJ2-3</td>
      <td>single pair</td>
      <td>CV0902</td>
      <td>CMV</td>
      <td>CD4</td>
      <td>1</td>
      <td>3</td>
    </tr>
    <tr>
      <th>4</th>
      <td>CAEAGRDDKIIF</td>
      <td>TRAV5*01</td>
      <td>TRAJ30</td>
      <td>CASSLSGNSYEQYF</td>
      <td>TRBV5-1*01</td>
      <td>TRBJ2-7</td>
      <td>single pair</td>
      <td>CV0902</td>
      <td>nan</td>
      <td>CD8</td>
      <td>1</td>
      <td>4</td>
    </tr>
  </tbody>
</table>
</div></div></div>
</div>
<p>We create a TCR repertoire object, that will contain the pairwise distances between all TCRs. We need to specify the TCRs organism (‘human’ or ‘mouse’) and the TCR chains ([‘alpha’, ‘beta’]).</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">tr</span> <span class="o">=</span> <span class="n">TCRrep</span><span class="p">(</span><span class="n">cell_df</span><span class="o">=</span><span class="n">df_tcrdist</span><span class="p">,</span> <span class="n">organism</span><span class="o">=</span><span class="s2">&quot;human&quot;</span><span class="p">,</span> <span class="n">chains</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;alpha&quot;</span><span class="p">,</span> <span class="s2">&quot;beta&quot;</span><span class="p">])</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>/home/icb/felix.drost/miniconda/envs/bestPractice/lib/python3.8/site-packages/tcrdist/repertoire.py:159: UserWarning: cell_df needs a counts column to track clonal number of frequency

  self._validate_cell_df()
/home/icb/felix.drost/miniconda/envs/bestPractice/lib/python3.8/site-packages/tcrdist/repertoire.py:791: UserWarning: No &#39;count&#39; column provided; count column set to 1
  warnings.warn(&quot;No &#39;count&#39; column provided; count column set to 1&quot;)
</pre></div>
</div>
</div>
</div>
<p>The TCR repertoire object contains the pairwise distance of α- and β-chains, which we extract to a DataFrame using our original index.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">dist_total</span> <span class="o">=</span> <span class="n">tr</span><span class="o">.</span><span class="n">pw_alpha</span> <span class="o">+</span> <span class="n">tr</span><span class="o">.</span><span class="n">pw_beta</span>
<span class="n">columns</span> <span class="o">=</span> <span class="n">tr</span><span class="o">.</span><span class="n">clone_df</span><span class="p">[</span><span class="s2">&quot;index&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="nb">float</span><span class="p">)</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="nb">int</span><span class="p">)</span>
<span class="n">df_dist</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">dist_total</span><span class="p">,</span> <span class="n">columns</span><span class="o">=</span><span class="n">columns</span><span class="p">,</span> <span class="n">index</span><span class="o">=</span><span class="n">columns</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<p>Finally, we plot the pairwise distances into a heatmap plot. Using hierarchical clustering, we group clones with similar TCRs (low TCRdist value).</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span><span class="w"> </span><span class="nn">scipy.cluster.hierarchy</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">hc</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">scipy.spatial</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">sp</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">matplotlib</span><span class="w"> </span><span class="kn">import</span> <span class="n">rcParams</span>

<span class="n">rcParams</span><span class="p">[</span><span class="s2">&quot;figure.figsize&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="p">(</span><span class="mi">20</span><span class="p">,</span> <span class="mi">20</span><span class="p">)</span>

<span class="n">linkage</span> <span class="o">=</span> <span class="n">hc</span><span class="o">.</span><span class="n">linkage</span><span class="p">(</span><span class="n">sp</span><span class="o">.</span><span class="n">distance</span><span class="o">.</span><span class="n">squareform</span><span class="p">(</span><span class="n">df_dist</span><span class="p">),</span> <span class="n">method</span><span class="o">=</span><span class="s2">&quot;average&quot;</span><span class="p">)</span>
<span class="n">plot</span> <span class="o">=</span> <span class="n">sb</span><span class="o">.</span><span class="n">clustermap</span><span class="p">(</span>
    <span class="n">df_dist</span><span class="p">,</span>
    <span class="n">row_linkage</span><span class="o">=</span><span class="n">linkage</span><span class="p">,</span>
    <span class="n">col_linkage</span><span class="o">=</span><span class="n">linkage</span><span class="p">,</span>
    <span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">40</span><span class="p">,</span> <span class="mi">40</span><span class="p">),</span>
    <span class="n">yticklabels</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
    <span class="n">xticklabels</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
<span class="p">)</span>
<span class="n">plot</span><span class="o">.</span><span class="n">ax_row_dendrogram</span><span class="o">.</span><span class="n">set_visible</span><span class="p">(</span><span class="kc">False</span><span class="p">)</span>
<span class="n">plot</span><span class="o">.</span><span class="n">ax_col_dendrogram</span><span class="o">.</span><span class="n">set_visible</span><span class="p">(</span><span class="kc">False</span><span class="p">)</span>

<span class="n">plot</span><span class="o">.</span><span class="n">fig</span><span class="o">.</span><span class="n">suptitle</span><span class="p">(</span><span class="s2">&quot;TCRdist&quot;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">tight_layout</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="../_images/5eb228c0f62900a47e5ed95ab7f6271248bb374fbfd54c4442a39f82562dc20d.png" src="../_images/5eb228c0f62900a47e5ed95ab7f6271248bb374fbfd54c4442a39f82562dc20d.png" />
</div>
</div>
<p>It can be observed, that several clusters of similar receptors form, which can be assumed to have the same epitope specificity. From this pairwise-distances, we can create groups of receptors which are likely to bind to the same epitopes. For this, we switch back to the Scirpy framework, which incorporates tools to create and visualize clonotype frameworks, based on distances of α- and β-chains.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">df_tcrdist_alpha</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span>
    <span class="n">tr</span><span class="o">.</span><span class="n">pw_alpha</span><span class="p">,</span> <span class="n">columns</span><span class="o">=</span><span class="n">tr</span><span class="o">.</span><span class="n">clone_df</span><span class="p">[</span><span class="s2">&quot;cdr3_a_aa&quot;</span><span class="p">],</span> <span class="n">index</span><span class="o">=</span><span class="n">tr</span><span class="o">.</span><span class="n">clone_df</span><span class="p">[</span><span class="s2">&quot;cdr3_a_aa&quot;</span><span class="p">]</span>
<span class="p">)</span>
<span class="n">df_tcrdist_beta</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span>
    <span class="n">tr</span><span class="o">.</span><span class="n">pw_beta</span><span class="p">,</span> <span class="n">columns</span><span class="o">=</span><span class="n">tr</span><span class="o">.</span><span class="n">clone_df</span><span class="p">[</span><span class="s2">&quot;cdr3_b_aa&quot;</span><span class="p">],</span> <span class="n">index</span><span class="o">=</span><span class="n">tr</span><span class="o">.</span><span class="n">clone_df</span><span class="p">[</span><span class="s2">&quot;cdr3_b_aa&quot;</span><span class="p">]</span>
<span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<p>We will wrap the addition of pairwise distances to the Scirpy object into a function. Overall, it stores the name of distance, the sequences, and the pairwise distances to <code class="docutils literal notranslate"><span class="pre">adata.uns</span></code>. To store the pairwise distances efficiently in a sparse matrix, all entries above a user-defined cutoff (here: 60 per chain, since a commonly used TCRdist cutoff is 120) will be set to 0, and all other distances are shifted by 1 (i.e. perfect matches will have a distance of 1).</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span><span class="w"> </span><span class="nn">scipy.sparse</span><span class="w"> </span><span class="kn">import</span> <span class="n">csr_matrix</span>


<span class="k">def</span><span class="w"> </span><span class="nf">add_dists</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">df_dist_alpha</span><span class="p">,</span> <span class="n">df_dist_beta</span><span class="p">,</span> <span class="n">name</span><span class="p">,</span> <span class="n">cutoff</span><span class="p">):</span>
    <span class="n">adata</span><span class="o">.</span><span class="n">uns</span><span class="p">[</span><span class="sa">f</span><span class="s2">&quot;ir_dist_aa_</span><span class="si">{</span><span class="n">name</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="p">{</span>
        <span class="s2">&quot;params&quot;</span><span class="p">:</span> <span class="p">{</span><span class="s2">&quot;metric&quot;</span><span class="p">:</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">name</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">,</span> <span class="s2">&quot;sequence&quot;</span><span class="p">:</span> <span class="s2">&quot;aa&quot;</span><span class="p">,</span> <span class="s2">&quot;cutoff&quot;</span><span class="p">:</span> <span class="n">cutoff</span><span class="p">}</span>
    <span class="p">}</span>

    <span class="k">for</span> <span class="n">chain</span><span class="p">,</span> <span class="n">dists</span> <span class="ow">in</span> <span class="p">[(</span><span class="s2">&quot;VJ&quot;</span><span class="p">,</span> <span class="n">df_dist_alpha</span><span class="p">),</span> <span class="p">(</span><span class="s2">&quot;VDJ&quot;</span><span class="p">,</span> <span class="n">df_dist_beta</span><span class="p">)]:</span>
        <span class="k">if</span> <span class="n">dists</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
            <span class="k">continue</span>
        <span class="n">dist_values</span> <span class="o">=</span> <span class="n">dists</span><span class="o">.</span><span class="n">values</span> <span class="o">+</span> <span class="mi">1</span>
        <span class="n">dist_values</span><span class="p">[</span><span class="n">dist_values</span> <span class="o">&gt;</span> <span class="p">(</span><span class="n">cutoff</span> <span class="o">+</span> <span class="mi">1</span><span class="p">)]</span> <span class="o">=</span> <span class="mi">0</span>
        <span class="n">dist_values</span> <span class="o">=</span> <span class="n">csr_matrix</span><span class="p">(</span><span class="n">dist_values</span><span class="p">)</span>
        <span class="n">adata</span><span class="o">.</span><span class="n">uns</span><span class="p">[</span><span class="sa">f</span><span class="s2">&quot;ir_dist_aa_</span><span class="si">{</span><span class="n">name</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">][</span><span class="n">chain</span><span class="p">]</span> <span class="o">=</span> <span class="p">{</span>
            <span class="s2">&quot;seqs&quot;</span><span class="p">:</span> <span class="n">dists</span><span class="o">.</span><span class="n">index</span><span class="o">.</span><span class="n">tolist</span><span class="p">(),</span>
            <span class="s2">&quot;distances&quot;</span><span class="p">:</span> <span class="n">dist_values</span><span class="p">,</span>
        <span class="p">}</span>


<span class="n">add_dists</span><span class="p">(</span><span class="n">adata_tcrdist</span><span class="p">,</span> <span class="n">df_tcrdist_alpha</span><span class="p">,</span> <span class="n">df_tcrdist_beta</span><span class="p">,</span> <span class="s2">&quot;tcrdist&quot;</span><span class="p">,</span> <span class="mi">60</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<p>Starting from these pairwise distances, we can define clonotype clusters, with the same parameters as for the database queries.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">ir</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">define_clonotype_clusters</span><span class="p">(</span>
    <span class="n">adata_tcrdist</span><span class="p">,</span>
    <span class="n">sequence</span><span class="o">=</span><span class="s2">&quot;aa&quot;</span><span class="p">,</span>
    <span class="n">metric</span><span class="o">=</span><span class="s2">&quot;tcrdist&quot;</span><span class="p">,</span>
    <span class="n">receptor_arms</span><span class="o">=</span><span class="s2">&quot;all&quot;</span><span class="p">,</span>
    <span class="n">dual_ir</span><span class="o">=</span><span class="s2">&quot;primary_only&quot;</span><span class="p">,</span>
<span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>100%|██████████████████████████████████████| 5325/5325 [00:12&lt;00:00, 432.66it/s]
</pre></div>
</div>
</div>
</div>
<p>We can now form and visualize clonotype networks, which contain cells of the different clonotype clusters. For visualization purposes, we will show only clusters with more than 2 cells and more than 2 clonotypes.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">adata_tcrdist</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s2">&quot;antigen.species&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">adata_tcrdist</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s2">&quot;antigen.species&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="nb">str</span><span class="p">)</span>

<span class="n">ir</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">clonotype_network</span><span class="p">(</span>
    <span class="n">adata_tcrdist</span><span class="p">,</span> <span class="n">min_cells</span><span class="o">=</span><span class="mi">3</span><span class="p">,</span> <span class="n">min_nodes</span><span class="o">=</span><span class="mi">3</span><span class="p">,</span> <span class="n">sequence</span><span class="o">=</span><span class="s2">&quot;aa&quot;</span><span class="p">,</span> <span class="n">metric</span><span class="o">=</span><span class="s2">&quot;tcrdist&quot;</span>
<span class="p">)</span>
<span class="n">ir</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">clonotype_network</span><span class="p">(</span>
    <span class="n">adata_tcrdist</span><span class="p">,</span>
    <span class="n">color</span><span class="o">=</span><span class="s2">&quot;antigen.species&quot;</span><span class="p">,</span>
    <span class="n">label_fontsize</span><span class="o">=</span><span class="mi">9</span><span class="p">,</span>
    <span class="n">panel_size</span><span class="o">=</span><span class="p">(</span><span class="mi">14</span><span class="p">,</span> <span class="mi">7</span><span class="p">),</span>
    <span class="n">base_size</span><span class="o">=</span><span class="mi">10</span><span class="p">,</span>
    <span class="n">size_power</span><span class="o">=</span><span class="mf">0.75</span><span class="p">,</span>
<span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>&lt;AxesSubplot:&gt;
</pre></div>
</div>
<img alt="../_images/15c5257b26de38f0f1743a9ca4e239fab7f7c140056f67b26fdb2b3e7745a6a9.png" src="../_images/15c5257b26de38f0f1743a9ca4e239fab7f7c140056f67b26fdb2b3e7745a6a9.png" />
</div>
</div>
<p>We derived a clonotype network graph: all clonotypes within a cluster are likely to have similar epitope specificity. The dot size corresponds to the clonal expansion of this cell. The clonotypes are colored by their antigen specificity assigned by the previous database query. These clusters can be used e.g. as groups for DEG analysis on transcriptome data, or for detecting enriched motifs and preserved residues (see chapter Sequence Analysis).</p>
<section id="tcrmatch">
<h4><span class="section-number">43.2.3.1. </span>TCRmatch<a class="headerlink" href="#tcrmatch" title="Link to this heading">#</a></h4>
<p>TCRmatch requires only the CDR3β of the TCR. It can therefore use a wider array of data, since especially less recent studies focused mainly on measuring the β-chain. We will therefore filter all cells that are missing this information and doublet cells. As before, we reduce the sample points to unique sequences to avoid unnecessary computation.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">adata_tcrmatch</span> <span class="o">=</span> <span class="n">adata_tcr</span><span class="p">[</span>
    <span class="o">~</span><span class="n">adata_tcr</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s2">&quot;chain_pairing&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">isin</span><span class="p">([</span><span class="s2">&quot;No IR&quot;</span><span class="p">,</span> <span class="s2">&quot;orphan VJ&quot;</span><span class="p">,</span> <span class="s2">&quot;multi_chain&quot;</span><span class="p">])</span>
<span class="p">]</span>
<span class="n">adata_tcrmatch</span> <span class="o">=</span> <span class="n">adata_tcrmatch</span><span class="p">[</span>
    <span class="o">~</span><span class="n">adata_tcrmatch</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s2">&quot;IR_VDJ_1_junction_aa&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">isna</span><span class="p">()</span>
<span class="p">]</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>

<span class="n">df_tcrmatch</span> <span class="o">=</span> <span class="n">adata_tcrmatch</span><span class="o">.</span><span class="n">obs</span><span class="p">[[</span><span class="s2">&quot;IR_VDJ_1_junction_aa&quot;</span><span class="p">,</span> <span class="s2">&quot;antigen.species&quot;</span><span class="p">]]</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
<span class="n">df_tcrmatch</span> <span class="o">=</span> <span class="n">df_tcrmatch</span><span class="o">.</span><span class="n">drop_duplicates</span><span class="p">()</span>
<span class="n">df_tcrmatch</span> <span class="o">=</span> <span class="n">df_tcrmatch</span><span class="o">.</span><span class="n">reset_index</span><span class="p">(</span><span class="n">drop</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="nb">len</span><span class="p">(</span><span class="n">df_tcrmatch</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>6190
</pre></div>
</div>
</div>
</div>
<p>As defined in the documentation, the method uses a Text-File as input, containing only the β-chain, without leading C and ending F or W as provided by the Cell Ranger output. We therefore need to adjust our data accordingly.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">df_tcrmatch</span><span class="p">[</span><span class="s2">&quot;CDR3b_trimmed&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">df_tcrmatch</span><span class="p">[</span><span class="s2">&quot;IR_VDJ_1_junction_aa&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">str</span><span class="p">[</span><span class="mi">1</span><span class="p">:</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span>
<span class="n">df_tcrmatch</span><span class="p">[[</span><span class="s2">&quot;CDR3b_trimmed&quot;</span><span class="p">]]</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span>
    <span class="s2">&quot;tmp/tcrmatch_input.csv&quot;</span><span class="p">,</span> <span class="n">header</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">index</span><span class="o">=</span><span class="kc">False</span>
<span class="p">)</span>
<span class="n">df_tcrmatch</span><span class="p">[[</span><span class="s2">&quot;CDR3b_trimmed&quot;</span><span class="p">]]</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span>
    <span class="s2">&quot;tmp/tcrmatch_input.csv&quot;</span><span class="p">,</span> <span class="n">header</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">index</span><span class="o">=</span><span class="kc">False</span>
<span class="p">)</span>
<span class="n">df_tcrmatch</span><span class="o">.</span><span class="n">head</span><span class="p">(</span><span class="mi">5</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>IR_VDJ_1_junction_aa</th>
      <th>antigen.species</th>
      <th>CDR3b_trimmed</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>CASSDSSTDTQYF</td>
      <td>EBV</td>
      <td>ASSDSSTDTQY</td>
    </tr>
    <tr>
      <th>1</th>
      <td>CASSQTGGQPQHF</td>
      <td>NaN</td>
      <td>ASSQTGGQPQH</td>
    </tr>
    <tr>
      <th>2</th>
      <td>CASRPSGLNTDTQYF</td>
      <td>NaN</td>
      <td>ASRPSGLNTDTQY</td>
    </tr>
    <tr>
      <th>3</th>
      <td>CASSYGGYNEQFF</td>
      <td>CMV</td>
      <td>ASSYGGYNEQF</td>
    </tr>
    <tr>
      <th>4</th>
      <td>CAISASGGGSSGNTIYF</td>
      <td>NaN</td>
      <td>AISASGGGSSGNTIY</td>
    </tr>
  </tbody>
</table>
</div></div></div>
</div>
<p>We will call TCRmatch via the command line, with the following arguments:</p>
<ul class="simple">
<li><p><strong>-i</strong>: the query data, here: our input file</p></li>
<li><p><strong>-t</strong>: amount of cores used for calculation</p></li>
<li><p><strong>-d</strong>: the reference data, either a database or our input file (for pairwise matches)- <strong>-s</strong>: threshold for considering a match, where 0 is no similarity and 1 is perfect match. Here we use the threshold of medium confidence 0.9. Alternative, you can use the more stricter threshold of 0.97 for high confidence binding. Note, that less stringent cutoffs also result in higher computation times.</p></li>
</ul>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">cmd_tcrmatch</span> <span class="o">=</span> <span class="s2">&quot;cd TCRMatch/ &amp;&amp;&quot;</span>
<span class="n">cmd_tcrmatch</span> <span class="o">+=</span> <span class="s2">&quot;./tcrmatch -i ../tmp/tcrmatch_input.csv -t 1 -d ../tmp/tcrmatch_input.csv -s 0.9 &gt; ../tmp/tcrmatch_output.csv&quot;</span>
</pre></div>
</div>
</div>
</div>
<p>We can call the run command via the bash shell:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="o">!</span><span class="nv">$cmd_tcrmatch</span>
</pre></div>
</div>
</div>
</div>
<p>The output file contains all pairwise distances above the specified threshold in long format: i.e., each row contains the two sequences (input_sequence and matching_sequence) and the similarity score. We will transform this DataFrame into wide format, where the rows and the axis represent the matching sequences. To counter numeric instabilities we further ensure symmetry of the matrix. TCRdist provides a similarity with high values representing likely binders. To make it compliant with Scirpy’s distance framework we subtract all pairwise distances from 1 and fill all elements for which no threshold exists with 1 (these combinations were formerly below the distance cutoff).</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">dist_tcrmatch</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="s2">&quot;tmp/tcrmatch_output.csv&quot;</span><span class="p">,</span> <span class="n">sep</span><span class="o">=</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">)</span>
<span class="n">dist_tcrmatch</span> <span class="o">=</span> <span class="n">dist_tcrmatch</span><span class="p">[[</span><span class="s2">&quot;input_sequence&quot;</span><span class="p">,</span> <span class="s2">&quot;match_sequence&quot;</span><span class="p">,</span> <span class="s2">&quot;score&quot;</span><span class="p">]]</span>
<span class="n">dist_tcrmatch</span> <span class="o">=</span> <span class="n">dist_tcrmatch</span><span class="o">.</span><span class="n">pivot</span><span class="p">(</span><span class="s2">&quot;input_sequence&quot;</span><span class="p">,</span> <span class="s2">&quot;match_sequence&quot;</span><span class="p">,</span> <span class="s2">&quot;score&quot;</span><span class="p">)</span>
<span class="n">values</span> <span class="o">=</span> <span class="mi">1</span> <span class="o">-</span> <span class="p">(</span><span class="n">dist_tcrmatch</span><span class="o">.</span><span class="n">values</span> <span class="o">+</span> <span class="n">dist_tcrmatch</span><span class="o">.</span><span class="n">values</span><span class="o">.</span><span class="n">transpose</span><span class="p">())</span> <span class="o">/</span> <span class="mi">2</span>

<span class="n">trimmed_2_full</span> <span class="o">=</span> <span class="nb">dict</span><span class="p">(</span><span class="n">df_tcrmatch</span><span class="p">[[</span><span class="s2">&quot;CDR3b_trimmed&quot;</span><span class="p">,</span> <span class="s2">&quot;IR_VDJ_1_junction_aa&quot;</span><span class="p">]]</span><span class="o">.</span><span class="n">values</span><span class="p">)</span>
<span class="n">columns</span> <span class="o">=</span> <span class="n">dist_tcrmatch</span><span class="o">.</span><span class="n">index</span><span class="o">.</span><span class="n">map</span><span class="p">(</span><span class="n">trimmed_2_full</span><span class="p">)</span>
<span class="n">dist_tcrmatch</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">index</span><span class="o">=</span><span class="n">columns</span><span class="p">,</span> <span class="n">columns</span><span class="o">=</span><span class="n">columns</span><span class="p">,</span> <span class="n">data</span><span class="o">=</span><span class="n">values</span><span class="p">)</span>
<span class="n">dist_tcrmatch</span> <span class="o">=</span> <span class="n">dist_tcrmatch</span><span class="o">.</span><span class="n">fillna</span><span class="p">(</span><span class="mi">1</span><span class="p">)</span>
<span class="n">dist_tcrmatch</span><span class="o">.</span><span class="n">index</span><span class="o">.</span><span class="n">name</span> <span class="o">=</span> <span class="kc">None</span>
<span class="n">dist_tcrmatch</span><span class="o">.</span><span class="n">head</span><span class="p">(</span><span class="mi">5</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>CAAGTDYEQYF</th>
      <th>CAAQGSSMETQYF</th>
      <th>CAARTVNTGELFF</th>
      <th>CAAVQGPSYEQYF</th>
      <th>CACGAGEDTGELFF</th>
      <th>CACGKGGGSLRYTF</th>
      <th>CACGPGGPSTDTQYF</th>
      <th>CACPSTSGINTGELFF</th>
      <th>CACRGLAGYEQYF</th>
      <th>CADARSSWDTQYF</th>
      <th>...</th>
      <th>CSVWTGEGYTF</th>
      <th>CSVYTGTSAYEQYF</th>
      <th>CSWLAGQETQYF</th>
      <th>CTSRMDSNYGYTF</th>
      <th>CVFGFRGDTQYF</th>
      <th>CVSRDKYEQYF</th>
      <th>CVTEGSSYNEQFF</th>
      <th>CVTGLAENTQYF</th>
      <th>CVTRETGGGGYTF</th>
      <th>CVTRYSYEQYF</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>CAAGTDYEQYF</th>
      <td>0.0</td>
      <td>1.0</td>
      <td>1.0</td>
      <td>1.0</td>
      <td>1.0</td>
      <td>1.0</td>
      <td>1.0</td>
      <td>1.0</td>
      <td>1.0</td>
      <td>1.0</td>
      <td>...</td>
      <td>1.0</td>
      <td>1.0</td>
      <td>1.0</td>
      <td>1.0</td>
      <td>1.0</td>
      <td>1.0</td>
      <td>1.0</td>
      <td>1.0</td>
      <td>1.0</td>
      <td>1.0</td>
    </tr>
    <tr>
      <th>CAAQGSSMETQYF</th>
      <td>1.0</td>
      <td>0.0</td>
      <td>1.0</td>
      <td>1.0</td>
      <td>1.0</td>
      <td>1.0</td>
      <td>1.0</td>
      <td>1.0</td>
      <td>1.0</td>
      <td>1.0</td>
      <td>...</td>
      <td>1.0</td>
      <td>1.0</td>
      <td>1.0</td>
      <td>1.0</td>
      <td>1.0</td>
      <td>1.0</td>
      <td>1.0</td>
      <td>1.0</td>
      <td>1.0</td>
      <td>1.0</td>
    </tr>
    <tr>
      <th>CAARTVNTGELFF</th>
      <td>1.0</td>
      <td>1.0</td>
      <td>0.0</td>
      <td>1.0</td>
      <td>1.0</td>
      <td>1.0</td>
      <td>1.0</td>
      <td>1.0</td>
      <td>1.0</td>
      <td>1.0</td>
      <td>...</td>
      <td>1.0</td>
      <td>1.0</td>
      <td>1.0</td>
      <td>1.0</td>
      <td>1.0</td>
      <td>1.0</td>
      <td>1.0</td>
      <td>1.0</td>
      <td>1.0</td>
      <td>1.0</td>
    </tr>
    <tr>
      <th>CAAVQGPSYEQYF</th>
      <td>1.0</td>
      <td>1.0</td>
      <td>1.0</td>
      <td>0.0</td>
      <td>1.0</td>
      <td>1.0</td>
      <td>1.0</td>
      <td>1.0</td>
      <td>1.0</td>
      <td>1.0</td>
      <td>...</td>
      <td>1.0</td>
      <td>1.0</td>
      <td>1.0</td>
      <td>1.0</td>
      <td>1.0</td>
      <td>1.0</td>
      <td>1.0</td>
      <td>1.0</td>
      <td>1.0</td>
      <td>1.0</td>
    </tr>
    <tr>
      <th>CACGAGEDTGELFF</th>
      <td>1.0</td>
      <td>1.0</td>
      <td>1.0</td>
      <td>1.0</td>
      <td>0.0</td>
      <td>1.0</td>
      <td>1.0</td>
      <td>1.0</td>
      <td>1.0</td>
      <td>1.0</td>
      <td>...</td>
      <td>1.0</td>
      <td>1.0</td>
      <td>1.0</td>
      <td>1.0</td>
      <td>1.0</td>
      <td>1.0</td>
      <td>1.0</td>
      <td>1.0</td>
      <td>1.0</td>
      <td>1.0</td>
    </tr>
  </tbody>
</table>
<p>5 rows × 6190 columns</p>
</div></div></div>
</div>
<p>Again, we can demonstrate the pairwise distances in a heatmap plot. You will note, that the majority of entries is below the cutoff threshold.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">linkage</span> <span class="o">=</span> <span class="n">hc</span><span class="o">.</span><span class="n">linkage</span><span class="p">(</span><span class="n">sp</span><span class="o">.</span><span class="n">distance</span><span class="o">.</span><span class="n">squareform</span><span class="p">(</span><span class="n">dist_tcrmatch</span><span class="p">),</span> <span class="n">method</span><span class="o">=</span><span class="s2">&quot;average&quot;</span><span class="p">)</span>
<span class="n">plot</span> <span class="o">=</span> <span class="n">sb</span><span class="o">.</span><span class="n">clustermap</span><span class="p">(</span>
    <span class="n">dist_tcrmatch</span><span class="p">,</span>
    <span class="n">row_linkage</span><span class="o">=</span><span class="n">linkage</span><span class="p">,</span>
    <span class="n">col_linkage</span><span class="o">=</span><span class="n">linkage</span><span class="p">,</span>
    <span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">40</span><span class="p">,</span> <span class="mi">40</span><span class="p">),</span>
    <span class="n">yticklabels</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
    <span class="n">xticklabels</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
<span class="p">)</span>
<span class="n">plot</span><span class="o">.</span><span class="n">ax_row_dendrogram</span><span class="o">.</span><span class="n">set_visible</span><span class="p">(</span><span class="kc">False</span><span class="p">)</span>
<span class="n">plot</span><span class="o">.</span><span class="n">ax_col_dendrogram</span><span class="o">.</span><span class="n">set_visible</span><span class="p">(</span><span class="kc">False</span><span class="p">)</span>
<span class="n">plot</span><span class="o">.</span><span class="n">fig</span><span class="o">.</span><span class="n">suptitle</span><span class="p">(</span><span class="s2">&quot;TCRmatch&quot;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">tight_layout</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="../_images/abf8d4f11f6655fbaa5cc55a9177745eaf9845d0ff1689140a2a68c2fcfd1f28.png" src="../_images/abf8d4f11f6655fbaa5cc55a9177745eaf9845d0ff1689140a2a68c2fcfd1f28.png" />
</div>
</div>
<p>Using the function we defined earlier, we can now add this distance metric to the AnnData object in Scirpy-Format and plot the corresponding clonotype networks.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">add_dists</span><span class="p">(</span><span class="n">adata_tcrmatch</span><span class="p">,</span> <span class="kc">None</span><span class="p">,</span> <span class="n">dist_tcrmatch</span><span class="p">,</span> <span class="s2">&quot;tcrmatch&quot;</span><span class="p">,</span> <span class="mf">0.05</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">ir</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">define_clonotype_clusters</span><span class="p">(</span>
    <span class="n">adata_tcrmatch</span><span class="p">,</span>
    <span class="n">sequence</span><span class="o">=</span><span class="s2">&quot;aa&quot;</span><span class="p">,</span>
    <span class="n">metric</span><span class="o">=</span><span class="s2">&quot;tcrmatch&quot;</span><span class="p">,</span>
    <span class="n">receptor_arms</span><span class="o">=</span><span class="s2">&quot;VDJ&quot;</span><span class="p">,</span>
    <span class="n">dual_ir</span><span class="o">=</span><span class="s2">&quot;primary_only&quot;</span><span class="p">,</span>
<span class="p">)</span>
<span class="n">ir</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">clonotype_network</span><span class="p">(</span>
    <span class="n">adata_tcrmatch</span><span class="p">,</span> <span class="n">min_cells</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span> <span class="n">min_nodes</span><span class="o">=</span><span class="mi">3</span><span class="p">,</span> <span class="n">sequence</span><span class="o">=</span><span class="s2">&quot;aa&quot;</span><span class="p">,</span> <span class="n">metric</span><span class="o">=</span><span class="s2">&quot;tcrmatch&quot;</span>
<span class="p">)</span>

<span class="n">adata_tcrmatch</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s2">&quot;antigen.species&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">adata_tcrmatch</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s2">&quot;antigen.species&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span>
    <span class="nb">str</span>
<span class="p">)</span>
<span class="n">ir</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">clonotype_network</span><span class="p">(</span>
    <span class="n">adata_tcrmatch</span><span class="p">,</span>
    <span class="n">color</span><span class="o">=</span><span class="s2">&quot;antigen.species&quot;</span><span class="p">,</span>
    <span class="n">label_fontsize</span><span class="o">=</span><span class="mi">9</span><span class="p">,</span>
    <span class="n">panel_size</span><span class="o">=</span><span class="p">(</span><span class="mi">14</span><span class="p">,</span> <span class="mi">7</span><span class="p">),</span>
    <span class="n">base_size</span><span class="o">=</span><span class="mi">20</span><span class="p">,</span>
    <span class="n">size_power</span><span class="o">=</span><span class="mf">0.5</span><span class="p">,</span>
<span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>100%|█████████████████████████████████████| 6190/6190 [00:02&lt;00:00, 2370.78it/s]
</pre></div>
</div>
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>&lt;AxesSubplot:&gt;
</pre></div>
</div>
<img alt="../_images/dbdfbd5c301f510e6e1a6268a9bb51807868cd359f4daeb7b086e8c2046c18a8.png" src="../_images/dbdfbd5c301f510e6e1a6268a9bb51807868cd359f4daeb7b086e8c2046c18a8.png" />
</div>
</div>
<p>Similarly as for TCRdist, we derived a clonotype network graph, where clones of a subgraph are likely to bind to the same epitopes. Differing, clonotypes are defined via identical CRD3β sequence, which makes this metric applicable for more cells in the dataset or when comparing to databases.</p>
</section>
<section id="database-query-via-distance-metrics">
<h4><span class="section-number">43.2.3.2. </span>Database query via distance metrics<a class="headerlink" href="#database-query-via-distance-metrics" title="Link to this heading">#</a></h4>
<p>To increase the amount of annotated cells, we can also query the databases using distance metrics to match similar TCRs. This will lead to more hits, but at the same time also increase the amount of false positives due to inaccuracies of the distance metrics. Here, we show a database query using the Scirpy distance <code class="docutils literal notranslate"><span class="pre">alignment</span></code>, which is similar to TCRdist, but only applied on the CDR3 sequences.</p>
<p>Since this operation is computationally expensive, we will further downsample the data to one donor.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">adata_tcr_align</span> <span class="o">=</span> <span class="n">adata_tcr</span><span class="p">[</span><span class="n">adata_tcr</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s2">&quot;patient_id&quot;</span><span class="p">]</span> <span class="o">==</span> <span class="s2">&quot;AP6&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
</pre></div>
</div>
</div>
</div>
<p>Using the alignment metrics and a cutoff of 10, we can now call the previously mentioned functions.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">metric</span> <span class="o">=</span> <span class="s2">&quot;alignment&quot;</span>
<span class="n">sequence</span> <span class="o">=</span> <span class="s2">&quot;aa&quot;</span>
<span class="n">ir</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">ir_dist</span><span class="p">(</span><span class="n">adata_tcr_align</span><span class="p">,</span> <span class="n">vdjdb</span><span class="p">,</span> <span class="n">metric</span><span class="o">=</span><span class="n">metric</span><span class="p">,</span> <span class="n">sequence</span><span class="o">=</span><span class="n">sequence</span><span class="p">,</span> <span class="n">cutoff</span><span class="o">=</span><span class="mi">10</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>100%|███████████████████████████████████████| 9468/9468 [02:46&lt;00:00, 57.00it/s]
100%|█████████████████████████████████████| 13702/13702 [04:08&lt;00:00, 55.13it/s]
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">ir</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">ir_query</span><span class="p">(</span>
    <span class="n">adata_tcr_align</span><span class="p">,</span>
    <span class="n">vdjdb</span><span class="p">,</span>
    <span class="n">metric</span><span class="o">=</span><span class="n">metric</span><span class="p">,</span>
    <span class="n">sequence</span><span class="o">=</span><span class="n">sequence</span><span class="p">,</span>
    <span class="n">receptor_arms</span><span class="o">=</span><span class="s2">&quot;all&quot;</span><span class="p">,</span>
    <span class="n">dual_ir</span><span class="o">=</span><span class="s2">&quot;primary_only&quot;</span><span class="p">,</span>
<span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>100%|██████████████████████████████████████| 1012/1012 [00:01&lt;00:00, 714.86it/s]
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">ir</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">ir_query_annotate</span><span class="p">(</span>
    <span class="n">adata_tcr_align</span><span class="p">,</span>
    <span class="n">vdjdb</span><span class="p">,</span>
    <span class="n">metric</span><span class="o">=</span><span class="n">metric</span><span class="p">,</span>
    <span class="n">sequence</span><span class="o">=</span><span class="n">sequence</span><span class="p">,</span>
    <span class="n">include_ref_cols</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;antigen.species&quot;</span><span class="p">,</span> <span class="s2">&quot;antigen.epitope&quot;</span><span class="p">],</span>
    <span class="n">suffix</span><span class="o">=</span><span class="s2">&quot;_alignment&quot;</span><span class="p">,</span>
<span class="p">)</span>
<span class="n">adata_tcr_align</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s2">&quot;antigen.species_alignment&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">value_counts</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>100%|█████████████████████████████████████| 2202/2202 [00:00&lt;00:00, 5430.96it/s]
</pre></div>
</div>
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>CMV            840
ambiguous      182
HIV-1           30
EBV             20
InfluenzaA      18
SARS-CoV-2       6
HomoSapiens      4
HCV              1
Name: antigen.species_alignment, dtype: int64
</pre></div>
</div>
</div>
</div>
<p>We can see, that we receive more matches, as compared when querying for direct matches. On the same hand, the amount of receptors that have ambiguous annotation increases as well.</p>
</section>
</section>
<section id="epitope-prediction">
<span id="air-repertoire-specificity-key-takeaway-3"></span><span id="air-repertoire-specificity-key-takeaway-1"></span><h3><span class="section-number">43.2.4. </span>Epitope Prediction<a class="headerlink" href="#epitope-prediction" title="Link to this heading">#</a></h3>
<p>Instead of comparing TCR sequences, binding is directly predicted in this approach from the TCR sequence. This can be helpful, when there is no sufficient amount of TCRs for a specific disease present in a database. These approaches train machine learning models on TCR- and epitope-sequences, which biases these models towards the high-abundant epitopes within the public databases. Generally, two approaches can be observed:</p>
<ul class="simple">
<li><p><strong>Categorical models</strong>: The target epitope is used as category. Therefore, the model predicts binding to a fixed set of epitopes from the TCR sequence.</p></li>
<li><p><strong>Epitope-Sequence models</strong>: The model uses the TCR- and epitope-sequence to predict binding, which has the advantage, that arbitrary epitopes can be used. However, it has been shown that performance is greatly reduced for unknown epitopes.</p></li>
</ul>
<p>We therefore advise to check the training data of the model, whether it contains the epitopes of interest. Still, these tools should be used with caution until their performance is further evaluated in a standardized benchmark for their individual strengths and weaknesses. However, we expect a great increase in performance of similar tools in the near future, providing us with a high-quality prediction of specificity.</p>
<p>Out of several tools, we choose ERGO-II {cite}’springer2021contribution’. Since performance comparison still largely remains unknown, we choose a tool which incorporates the different AIR information in a flexible manner allowing a degree of missing data.</p>
<p>First, we will filter our TCR DataFrame to include only samples with a β chain, since this input is required by ERGO-II. Additionally, we also exclude doublet cells (‘multi_chains’). ERGO-II allows missing values for all information except the CDR3β, but other tools might not. So check, what information is required for the specific tool.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">adata_ergo</span> <span class="o">=</span> <span class="n">adata_tcr</span><span class="p">[</span>
    <span class="o">~</span><span class="n">adata_tcr</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s2">&quot;chain_pairing&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">isin</span><span class="p">([</span><span class="s2">&quot;orphan VJ&quot;</span><span class="p">,</span> <span class="s2">&quot;no IR&quot;</span><span class="p">,</span> <span class="s2">&quot;multi_chains&quot;</span><span class="p">])</span>
<span class="p">]</span>
<span class="n">df_ergo</span> <span class="o">=</span> <span class="n">adata_ergo</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span>
    <span class="p">[</span>
        <span class="s2">&quot;IR_VJ_1_junction_aa&quot;</span><span class="p">,</span>
        <span class="s2">&quot;IR_VJ_1_v_call&quot;</span><span class="p">,</span>
        <span class="s2">&quot;IR_VJ_1_j_call&quot;</span><span class="p">,</span>
        <span class="s2">&quot;IR_VDJ_1_junction_aa&quot;</span><span class="p">,</span>
        <span class="s2">&quot;IR_VDJ_1_v_call&quot;</span><span class="p">,</span>
        <span class="s2">&quot;IR_VDJ_1_j_call&quot;</span><span class="p">,</span>
    <span class="p">]</span>
<span class="p">]</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
</pre></div>
</div>
</div>
</div>
<p>Most tools are not using the Scirpy naming convention, we therefore need to rename the columns to match the convention specified in the documentation.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">dict_rename</span> <span class="o">=</span> <span class="p">{</span>
    <span class="s2">&quot;IR_VJ_1_junction_aa&quot;</span><span class="p">:</span> <span class="s2">&quot;TRA&quot;</span><span class="p">,</span>
    <span class="s2">&quot;IR_VJ_1_v_call&quot;</span><span class="p">:</span> <span class="s2">&quot;TRAV&quot;</span><span class="p">,</span>
    <span class="s2">&quot;IR_VJ_1_j_call&quot;</span><span class="p">:</span> <span class="s2">&quot;TRAJ&quot;</span><span class="p">,</span>
    <span class="s2">&quot;IR_VDJ_1_junction_aa&quot;</span><span class="p">:</span> <span class="s2">&quot;TRB&quot;</span><span class="p">,</span>
    <span class="s2">&quot;IR_VDJ_1_v_call&quot;</span><span class="p">:</span> <span class="s2">&quot;TRBV&quot;</span><span class="p">,</span>
    <span class="s2">&quot;IR_VDJ_1_j_call&quot;</span><span class="p">:</span> <span class="s2">&quot;TRBJ&quot;</span><span class="p">,</span>
<span class="p">}</span>
<span class="n">df_ergo</span> <span class="o">=</span> <span class="n">df_ergo</span><span class="o">.</span><span class="n">rename</span><span class="p">(</span><span class="n">columns</span><span class="o">=</span><span class="n">dict_rename</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<p>Next, we add the remaining information. For the epitope sequence, we choose the epitope ‘KLGGALQAK’ from the CMV, since we identified an underlying infection earlier. Since T-cell type information only contributed negligible, we set it to ‘None’ for all rows. We do not have annotation of MHC type, so we also set this column to ‘None’. Finally, we save the models input as ‘.csv’-File.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">df_ergo</span><span class="p">[</span><span class="s2">&quot;Peptide&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="s2">&quot;KLGGALQAK&quot;</span>  <span class="c1">#&#39;YLQPRTFLL&#39;</span>
<span class="n">df_ergo</span><span class="p">[</span><span class="s2">&quot;T-Cell-Type&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span>
<span class="n">df_ergo</span><span class="p">[</span><span class="s2">&quot;MHC&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span>

<span class="n">df_ergo</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="s2">&quot;tmp/ergo_input.csv&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<p>The model is called via the command line. We activate the environment and call the tool by specifying the database it was trained on (‘vdjdb’) and providing the path to the input file. Note: we uncommented line 105 of ‘<a class="reference external" href="http://Predict.py">Predict.py</a>’ to save the output to file. There are several reported bugs, when setting up this model (see Github issues), which you will need to fix before running the model. Alternatively, you can use the web interface as described on the GitHub-page.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">cmd_ergo</span> <span class="o">=</span> <span class="s2">&quot;source ~/.bashrc &amp;&amp;&quot;</span>
<span class="n">cmd_ergo</span> <span class="o">+=</span> <span class="s2">&quot;conda activate ergo &amp;&amp;&quot;</span>
<span class="n">cmd_ergo</span> <span class="o">+=</span> <span class="s2">&quot;cd ERGO-II &amp;&amp;&quot;</span>
<span class="n">cmd_ergo</span> <span class="o">+=</span> <span class="s2">&quot;python Predict.py vdjdb ../tmp/ergo_input.csv&quot;</span>
<span class="c1"># cmd_ergo += &#39;cd ..&#39;</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="o">!</span><span class="nv">$cmd_ergo</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>/home/icb/felix.drost/miniconda/envs/ergo/lib/python3.8/site-packages/pytorch_lightning/core/decorators.py:13: UserWarning: data_loader decorator deprecated in 0.7.0. Will remove 0.9.0
  warnings.warn(w)
                         cell_id              TRA  ... MHC     Score
0      BGCV01_AAACCTGAGACCGGAT-1              NaN  ... NaN  0.000002
1      BGCV01_AAACCTGAGGAACTGC-1    CAVKRGNNARLMF  ... NaN  0.629516
2      BGCV01_AAACCTGAGGGATCTG-1    CAVGAPGDDKIIF  ... NaN  0.635792
3      BGCV01_AAACCTGCAACGCACC-1              NaN  ... NaN  0.000017
4      BGCV01_AAACCTGCACGTGAGA-1   CAVNSPGGYQKVTF  ... NaN  0.625245
...                          ...              ...  ...  ..       ...
12362     S12_TTTGTCAGTAGAAAGG-1      CARNTGNQFYF  ... NaN  0.614335
12363     S12_TTTGTCAGTCCAACTA-1              NaN  ... NaN  0.000023
12364     S12_TTTGTCAGTGAGTGAC-1      CARNTGNQFYF  ... NaN  0.614335
12365     S12_TTTGTCATCCACTCCA-1    CAVSELGSEKLVF  ... NaN  0.685439
12366     S12_TTTGTCATCGCATGAT-1  CVVRAPWGSARQLTF  ... NaN  0.673902

[12367 rows x 11 columns]
</pre></div>
</div>
</div>
</div>
<p>After calling the command via the bash shell, we can read the output file.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">df_tcr_ergo</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="s2">&quot;ERGO-II/results.csv&quot;</span><span class="p">,</span> <span class="n">index_col</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>
<span class="n">df_tcr_ergo</span><span class="o">.</span><span class="n">head</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>TRA</th>
      <th>TRAV</th>
      <th>TRAJ</th>
      <th>TRB</th>
      <th>TRBV</th>
      <th>TRBJ</th>
      <th>Peptide</th>
      <th>T-Cell-Type</th>
      <th>MHC</th>
      <th>Score</th>
    </tr>
    <tr>
      <th>cell_id</th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>BGCV01_AAACCTGAGACCGGAT-1</th>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>CASSDSSTDTQYF</td>
      <td>TRBV6-4</td>
      <td>TRBJ2-3</td>
      <td>KLGGALQAK</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>0.000002</td>
    </tr>
    <tr>
      <th>BGCV01_AAACCTGAGGAACTGC-1</th>
      <td>CAVKRGNNARLMF</td>
      <td>TRAV21</td>
      <td>TRAJ31</td>
      <td>CASSQTGGQPQHF</td>
      <td>TRBV5-1</td>
      <td>TRBJ1-5</td>
      <td>KLGGALQAK</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>0.629516</td>
    </tr>
    <tr>
      <th>BGCV01_AAACCTGAGGGATCTG-1</th>
      <td>CAVGAPGDDKIIF</td>
      <td>TRAV8-3</td>
      <td>TRAJ30</td>
      <td>CASRPSGLNTDTQYF</td>
      <td>TRBV7-9</td>
      <td>TRBJ2-3</td>
      <td>KLGGALQAK</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>0.635792</td>
    </tr>
    <tr>
      <th>BGCV01_AAACCTGCAACGCACC-1</th>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>CASSYGGYNEQFF</td>
      <td>TRBV6-5</td>
      <td>TRBJ2-1</td>
      <td>KLGGALQAK</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>0.000017</td>
    </tr>
    <tr>
      <th>BGCV01_AAACCTGCACGTGAGA-1</th>
      <td>CAVNSPGGYQKVTF</td>
      <td>TRAV8-1</td>
      <td>TRAJ13</td>
      <td>CAISASGGGSSGNTIYF</td>
      <td>TRBV10-3</td>
      <td>TRBJ1-3</td>
      <td>KLGGALQAK</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>0.625245</td>
    </tr>
  </tbody>
</table>
</div></div></div>
</div>
<p>We can now visualize the distribution of predicted binding score.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">sb</span><span class="o">.</span><span class="n">distplot</span><span class="p">(</span><span class="n">df_tcr_ergo</span><span class="p">[</span><span class="s2">&quot;Score&quot;</span><span class="p">])</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>/home/icb/felix.drost/miniconda/envs/bestPractice/lib/python3.8/site-packages/seaborn/distributions.py:2619: FutureWarning: `distplot` is a deprecated function and will be removed in a future version. Please adapt your code to use either `displot` (a figure-level function with similar flexibility) or `histplot` (an axes-level function for histograms).
  warnings.warn(msg, FutureWarning)
</pre></div>
</div>
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>&lt;AxesSubplot:xlabel=&#39;Score&#39;, ylabel=&#39;Density&#39;&gt;
</pre></div>
</div>
<img alt="../_images/c89950f43ada264e07a9466a398e14aba887558a54b85314d81d5947e8a142ce.png" src="../_images/c89950f43ada264e07a9466a398e14aba887558a54b85314d81d5947e8a142ce.png" />
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span><span class="w"> </span><span class="nn">numpy</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">np</span>

<span class="n">np</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">df_tcr_ergo</span><span class="p">[</span><span class="s2">&quot;Score&quot;</span><span class="p">]</span> <span class="o">&gt;=</span> <span class="mf">0.9</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>1
</pre></div>
</div>
</div>
</div>
<p>TODO: Interpretation once donors fixed.</p>
</section>
</section>
<section id="bcr-specificity-analysis">
<h2><span class="section-number">43.3. </span>BCR specificity analysis<a class="headerlink" href="#bcr-specificity-analysis" title="Link to this heading">#</a></h2>
<p>Generally, approaches for inferring specificity of BCRs are similar to TCRs. Following, we will stick to the same structure as above with limited explanation, when the topic already has been introduced.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">path_data</span> <span class="o">=</span> <span class="s2">&quot;data&quot;</span>
<span class="n">path_bcr</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">path_data</span><span class="si">}</span><span class="s2">/BCR_01_preprocessed.h5ad&quot;</span>
<span class="n">adata_bcr</span> <span class="o">=</span> <span class="n">sc</span><span class="o">.</span><span class="n">read</span><span class="p">(</span><span class="n">path_bcr</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">adata_bcr</span> <span class="o">=</span> <span class="n">adata_bcr</span><span class="p">[</span>
    <span class="n">adata_bcr</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s2">&quot;patient_id&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">isin</span><span class="p">(</span>
        <span class="p">[</span><span class="s2">&quot;COVID-030&quot;</span><span class="p">,</span> <span class="s2">&quot;IVLPS-6&quot;</span><span class="p">,</span> <span class="s2">&quot;COVID-064&quot;</span><span class="p">,</span> <span class="s2">&quot;COVID-014&quot;</span><span class="p">,</span> <span class="s2">&quot;COVID-027&quot;</span><span class="p">,</span> <span class="s2">&quot;COVID-024&quot;</span><span class="p">]</span>
    <span class="p">)</span>
<span class="p">]</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
</pre></div>
</div>
</div>
</div>
<section id="id13">
<h3><span class="section-number">43.3.1. </span>Database Queries<a class="headerlink" href="#id13" title="Link to this heading">#</a></h3>
<p>Here, we will search for BCRs with specificity annotatation from previous studies. Common large-scale databases with BCR-epitope pairs are:</p>
<ul class="simple">
<li><p><a class="reference external" href="https://www.iedb.org/">IEDB</a> <span id="id14">[<a class="reference internal" href="#id490" title="Ward Fleri, Sinu Paul, Sandeep Kumar Dhanda, Swapnil Mahajan, Xiaojun Xu, Bjoern Peters, and Alessandro Sette. The immune epitope database and analysis resource in epitope discovery and synthetic vaccine design. Frontiers in immunology, 8:278, 2017.">Fleri <em>et al.</em>, 2017</a>]</span></p></li>
<li><p><a class="reference external" href="https://db.cngb.org/pird/">PIRD</a> <span id="id15">[<a class="reference internal" href="#id491" title="Wei Zhang, Longlong Wang, Ke Liu, Xiaofeng Wei, Kai Yang, Wensi Du, Shiyu Wang, Nannan Guo, Chuanchuan Ma, Lihua Luo, and others. Pird: pan immune repertoire database. Bioinformatics, 36(3):897–903, 2020.">Zhang <em>et al.</em>, 2020</a>]</span></p></li>
<li><p><a class="reference external" href="http://opig.stats.ox.ac.uk/webapps/covabdab/">CoV-AbDab</a> for COVID-19 specific BCRs <span id="id16">[<a class="reference internal" href="#id498" title="Matthew I. J. Raybould, Aleksandr Kovaltsuk, Claire Marks, and Charlotte M. Deane. Cov-abdab: the coronavirus antibody database. Bioinformatics, 37(5):734–735, 2021. URL: https://academic.oup.com/bioinformatics/advance-article/doi/10.1093/bioinformatics/btaa739/5893556, doi:10.1093/bioinformatics/btaa739.">Raybould <em>et al.</em>, 2021</a>]</span></p></li>
</ul>
<p>As for TCRs, it highly depends, which database is applicable for your study design. While the first two databases, can be used for a wider variety of epitopes and diseases, the latter contains only epitopes in the context of SARS-CoV.</p>
<section id="database-preparation">
<h4><span class="section-number">43.3.1.1. </span>Database Preparation<a class="headerlink" href="#database-preparation" title="Link to this heading">#</a></h4>
<p>We will demonstrate, how to construct an AnnData object from a Database using CoV-AbDab. You can download the database via:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="o">!</span>wget<span class="w"> </span>-O<span class="w"> </span>tmp/CoV-AbDab.csv<span class="w"> </span>http://opig.stats.ox.ac.uk/webapps/covabdab/static/downloads/CoV-AbDab_200422.csv
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>--2022-06-07 22:46:19--  http://opig.stats.ox.ac.uk/webapps/covabdab/static/downloads/CoV-AbDab_200422.csv
Resolving opig.stats.ox.ac.uk (opig.stats.ox.ac.uk)... 163.1.32.58
Connecting to opig.stats.ox.ac.uk (opig.stats.ox.ac.uk)|163.1.32.58|:80... connected.
HTTP request sent, awaiting response... 200 OK
Length: 3397624 (3.2M) [text/csv]
Saving to: ‘tmp/CoV-AbDab.csv’

100%[======================================&gt;] 3,397,624   10.1MB/s   in 0.3s   

2022-06-07 22:46:19 (10.1 MB/s) - ‘tmp/CoV-AbDab.csv’ saved [3397624/3397624]
</pre></div>
</div>
</div>
</div>
<p>Let’s read the database and filter for antibodies.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">cov_abdab</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="s2">&quot;tmp/CoV-AbDab.csv&quot;</span><span class="p">)</span>
<span class="n">cov_abdab</span> <span class="o">=</span> <span class="n">cov_abdab</span><span class="p">[</span><span class="n">cov_abdab</span><span class="p">[</span><span class="s2">&quot;Ab or Nb&quot;</span><span class="p">]</span> <span class="o">==</span> <span class="s2">&quot;Ab&quot;</span><span class="p">]</span>
<span class="n">cov_abdab</span><span class="o">.</span><span class="n">head</span><span class="p">(</span><span class="mi">5</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>Name</th>
      <th>Ab or Nb</th>
      <th>Binds to</th>
      <th>Doesn't Bind to</th>
      <th>Neutralising Vs</th>
      <th>Not Neutralising Vs</th>
      <th>Protein + Epitope</th>
      <th>Origin</th>
      <th>VH or VHH</th>
      <th>VL</th>
      <th>...</th>
      <th>Light J Gene</th>
      <th>CDRH3</th>
      <th>CDRL3</th>
      <th>Structures</th>
      <th>ABB Homology Model (if no structure)</th>
      <th>Sources</th>
      <th>Date Added</th>
      <th>Last Updated</th>
      <th>Update Description</th>
      <th>Notes/Following Up?</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>T6</td>
      <td>Ab</td>
      <td>SARS-CoV2_WT;SARS-CoV2_Alpha;SARS-CoV2_Beta;SA...</td>
      <td>NaN</td>
      <td>SARS-CoV2_WT;SARS-CoV2_Alpha;SARS-CoV2_Beta;SA...</td>
      <td>NaN</td>
      <td>S; RBD</td>
      <td>B-cells; SARS-CoV2_WT RBD Vaccinee</td>
      <td>QVQLQQPGTELVNPGASLKMSCKTSGYRFTSYIIHWVKQTPGQGLE...</td>
      <td>QIVLTQSPSSLAVSVGEKVTLSCKSSQSLLYSNNQKNYLAWYQQKS...</td>
      <td>...</td>
      <td>IGKJ1 (Mouse)</td>
      <td>ARDGENVLDY</td>
      <td>QQYYTYPWT</td>
      <td>https://www.rcsb.org/structure/7FJO;Expected (...</td>
      <td>NaN</td>
      <td>Qingtai Liang et al., 2022 (https://www.cell.c...</td>
      <td>Apr 20, 2022</td>
      <td>Apr 20, 2022</td>
      <td>NaN</td>
      <td>Complete</td>
    </tr>
    <tr>
      <th>1</th>
      <td>Hauser_Ab15</td>
      <td>Ab</td>
      <td>SARS-CoV2_WT (weak);SARS-CoV2_Beta (weak);SARS...</td>
      <td>NaN</td>
      <td>SARS-CoV2_WT (weak);SARS-CoV2_Gamma (weak);RaTG13</td>
      <td>SARS-CoV2_Beta;SHC014</td>
      <td>S; RBD</td>
      <td>Immunised Mice (rational immunogen)</td>
      <td>EVQLQQSGPELVKPGASVKISCKASGYSFTGYYMNWVKQSPEKSLE...</td>
      <td>DILMTQSPSSMSVSLGDTVSITCHASQGISSNIGWLQQKPGKSFKG...</td>
      <td>...</td>
      <td>IGKJ2 (Mouse)</td>
      <td>ARYYGNLYAMDY</td>
      <td>VQYTHFPYT</td>
      <td>ND</td>
      <td>Coronavirus%20Binding%20Antibody%20Sequences%2...</td>
      <td>Blake Hauser et al., 2022 (https://www.cell.co...</td>
      <td>Apr 20, 2022</td>
      <td>Apr 20, 2022</td>
      <td>NaN</td>
      <td>Complete</td>
    </tr>
    <tr>
      <th>2</th>
      <td>Hauser_Ab16</td>
      <td>Ab</td>
      <td>SARS-CoV2_WT;SARS-CoV2_Beta;SARS-CoV1;RaTG13;S...</td>
      <td>NaN</td>
      <td>SARS-CoV2_WT;SARS-CoV2_Beta (weak);SARS-CoV2_G...</td>
      <td>NaN</td>
      <td>S; RBD</td>
      <td>Immunised Mice (rational immunogen)</td>
      <td>EVQLQQSGPELVKPGASVKISCKASGYSFNNYYMNWVKQSPEKSLE...</td>
      <td>DILMTQSPSSMSVSLGDTVSITCHASQGIGSNIGWLQQKPGKSFKG...</td>
      <td>...</td>
      <td>IGKJ2 (Mouse)</td>
      <td>ARYFGNLFAMDF</td>
      <td>VQYVQFPYT</td>
      <td>ND</td>
      <td>Coronavirus%20Binding%20Antibody%20Sequences%2...</td>
      <td>Blake Hauser et al., 2022 (https://www.cell.co...</td>
      <td>Apr 20, 2022</td>
      <td>Apr 20, 2022</td>
      <td>NaN</td>
      <td>Complete</td>
    </tr>
    <tr>
      <th>3</th>
      <td>Hauser_Ab17</td>
      <td>Ab</td>
      <td>SARS-CoV2_WT;SARS-CoV2_Beta;SARS-CoV1;RaTG13;S...</td>
      <td>NaN</td>
      <td>SHC014 (weak);WIV-1 (weak);RaTG13 (weak)</td>
      <td>SARS-CoV2_WT;SARS-CoV2_Beta;SARS-CoV2_Gamma;SA...</td>
      <td>S; RBD</td>
      <td>Immunised Mice (rational immunogen)</td>
      <td>EVQLQQSGPELVKPGASVKISCKASGYSFTDYYMNWVKQSPEKSLE...</td>
      <td>DILMTQSPSSMSVSLGDTVSITCHASQGISSNIGWLQQKPGKSFKG...</td>
      <td>...</td>
      <td>IGKJ2 (Mouse)</td>
      <td>ARYYGNLYAMDY</td>
      <td>VQYVQFPYT</td>
      <td>https://www.rcsb.org/structure/7TE1</td>
      <td>NaN</td>
      <td>Blake Hauser et al., 2022 (https://www.cell.co...</td>
      <td>Apr 20, 2022</td>
      <td>Apr 20, 2022</td>
      <td>NaN</td>
      <td>Complete</td>
    </tr>
    <tr>
      <th>4</th>
      <td>Hauser_Ab19</td>
      <td>Ab</td>
      <td>SARS-CoV2_WT (weak);SARS-CoV2_Beta (weak)</td>
      <td>SARS-CoV1;WIV-1;RaTG13;SHC014</td>
      <td>RaTG13 (weak)</td>
      <td>SARS-CoV2_WT;SARS-CoV2_Gamma</td>
      <td>S; RBD</td>
      <td>Immunised Mice (rational immunogen)</td>
      <td>QVQLQQSGAELARPGASVKLSCKASGYPFTSYGINWVKQRTGQGLE...</td>
      <td>DIVMTQSHKFMSTSIGDRVSITCKASHDVSTAVAWYQQKPGQSPKL...</td>
      <td>...</td>
      <td>IGKJ2 (Mouse)</td>
      <td>ARSWNSNYGEYYFDY</td>
      <td>QQHYSTPYT</td>
      <td>ND</td>
      <td>Coronavirus%20Binding%20Antibody%20Sequences%2...</td>
      <td>Blake Hauser et al., 2022 (https://www.cell.co...</td>
      <td>Apr 20, 2022</td>
      <td>Apr 20, 2022</td>
      <td>NaN</td>
      <td>Complete</td>
    </tr>
  </tbody>
</table>
<p>5 rows × 23 columns</p>
</div></div></div>
</div>
<p>To use the toolkit’s function, we need to re-name the database columns to Scirpy format.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">dict_rename_cov_abdab</span> <span class="o">=</span> <span class="p">{</span>
    <span class="s2">&quot;Heavy V Gene&quot;</span><span class="p">:</span> <span class="s2">&quot;IR_VDJ_1_v_call&quot;</span><span class="p">,</span>
    <span class="s2">&quot;Heavy J Gene&quot;</span><span class="p">:</span> <span class="s2">&quot;IR_VDJ_1_j_call&quot;</span><span class="p">,</span>
    <span class="s2">&quot;Light V Gene&quot;</span><span class="p">:</span> <span class="s2">&quot;IR_VJ_1_v_call&quot;</span><span class="p">,</span>
    <span class="s2">&quot;Light J Gene&quot;</span><span class="p">:</span> <span class="s2">&quot;IR_VJ_1_h_call&quot;</span><span class="p">,</span>
    <span class="s2">&quot;CDRH3&quot;</span><span class="p">:</span> <span class="s2">&quot;IR_VDJ_1_junction_aa&quot;</span><span class="p">,</span>
    <span class="s2">&quot;CDRL3&quot;</span><span class="p">:</span> <span class="s2">&quot;IR_VJ_1_junction_aa&quot;</span><span class="p">,</span>
    <span class="s2">&quot;Binds to&quot;</span><span class="p">:</span> <span class="s2">&quot;Binding&quot;</span><span class="p">,</span>
<span class="p">}</span>
<span class="n">cov_abdab</span> <span class="o">=</span> <span class="n">cov_abdab</span><span class="o">.</span><span class="n">rename</span><span class="p">(</span><span class="n">columns</span><span class="o">=</span><span class="n">dict_rename_cov_abdab</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<p>The Scirpy-Format requires additional columns, so we add constant values.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">cov_abdab</span><span class="p">[</span><span class="s2">&quot;has_ir&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="s2">&quot;True&quot;</span>
<span class="n">cov_abdab</span><span class="p">[</span><span class="s2">&quot;IR_VJ_2_junction_aa&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span>
<span class="n">cov_abdab</span><span class="p">[</span><span class="s2">&quot;IR_VDJ_2_junction_aa&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span>
</pre></div>
</div>
</div>
</div>
<p>All AIRs in the database are tested against SARS-CoV-2 epitopes. To make the result more clearer to view, we will annotate the binding to include only SARS-CoV-2 and None. This is however heavily dependent on your specific research question.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">cov_abdab</span><span class="p">[</span><span class="s2">&quot;Binding&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">cov_abdab</span><span class="p">[</span><span class="s2">&quot;Binding&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span>
    <span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="s2">&quot;SARS-CoV-2&quot;</span> <span class="k">if</span> <span class="s2">&quot;SARS-CoV2&quot;</span> <span class="ow">in</span> <span class="n">x</span> <span class="k">else</span> <span class="s2">&quot;None&quot;</span>
<span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<p>Warning: many databases follow diverging definition of the CDR3 region. While the cellranger output contains leading ‘C’ and ending ‘F’ and ‘W’, many databases do not. We therefore need to add them to the database entries, in order to obtain correct matches.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">cov_abdab</span><span class="p">[</span><span class="s2">&quot;IR_VJ_1_junction_aa&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="s2">&quot;C&quot;</span> <span class="o">+</span> <span class="n">cov_abdab</span><span class="p">[</span><span class="s2">&quot;IR_VJ_1_junction_aa&quot;</span><span class="p">]</span> <span class="o">+</span> <span class="s2">&quot;F&quot;</span>
<span class="n">cov_abdab</span><span class="p">[</span><span class="s2">&quot;IR_VDJ_1_junction_aa&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="s2">&quot;C&quot;</span> <span class="o">+</span> <span class="n">cov_abdab</span><span class="p">[</span><span class="s2">&quot;IR_VDJ_1_junction_aa&quot;</span><span class="p">]</span> <span class="o">+</span> <span class="s2">&quot;W&quot;</span>
</pre></div>
</div>
</div>
</div>
<p>Finally, we can create an AnnData object, where AnnData.obs stores the BCR- and binding-information. Additionally, we need to add the name of the Database to the object, for the query functions to work.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">cov_abdab</span> <span class="o">=</span> <span class="n">sc</span><span class="o">.</span><span class="n">AnnData</span><span class="p">(</span><span class="n">obs</span><span class="o">=</span><span class="n">cov_abdab</span><span class="p">)</span>
<span class="n">cov_abdab</span><span class="o">.</span><span class="n">uns</span><span class="p">[</span><span class="s2">&quot;DB&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="p">{}</span>
<span class="n">cov_abdab</span><span class="o">.</span><span class="n">uns</span><span class="p">[</span><span class="s2">&quot;DB&quot;</span><span class="p">][</span><span class="s2">&quot;name&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="s2">&quot;CoV-AbDab&quot;</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>/home/icb/felix.drost/miniconda/envs/bestPractice/lib/python3.8/site-packages/anndata/_core/anndata.py:121: ImplicitModificationWarning: Transforming to str index.
  warnings.warn(&quot;Transforming to str index.&quot;, ImplicitModificationWarning)
</pre></div>
</div>
</div>
</div>
</section>
<section id="identical-matches">
<h4><span class="section-number">43.3.1.2. </span>Identical matches<a class="headerlink" href="#identical-matches" title="Link to this heading">#</a></h4>
<p>Let us first query for identical Heavy-chain.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">metric</span> <span class="o">=</span> <span class="s2">&quot;identity&quot;</span>
<span class="n">sequence</span> <span class="o">=</span> <span class="s2">&quot;aa&quot;</span>

<span class="n">ir</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">ir_dist</span><span class="p">(</span><span class="n">adata_bcr</span><span class="p">,</span> <span class="n">cov_abdab</span><span class="p">,</span> <span class="n">metric</span><span class="o">=</span><span class="n">metric</span><span class="p">,</span> <span class="n">sequence</span><span class="o">=</span><span class="n">sequence</span><span class="p">)</span>
<span class="n">ir</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">ir_query</span><span class="p">(</span>
    <span class="n">adata_bcr</span><span class="p">,</span>
    <span class="n">cov_abdab</span><span class="p">,</span>
    <span class="n">metric</span><span class="o">=</span><span class="n">metric</span><span class="p">,</span>
    <span class="n">sequence</span><span class="o">=</span><span class="n">sequence</span><span class="p">,</span>
    <span class="n">receptor_arms</span><span class="o">=</span><span class="s2">&quot;VDJ&quot;</span><span class="p">,</span>
    <span class="n">dual_ir</span><span class="o">=</span><span class="s2">&quot;primary_only&quot;</span><span class="p">,</span>
<span class="p">)</span>
<span class="n">ir</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">ir_query_annotate</span><span class="p">(</span>
    <span class="n">adata_bcr</span><span class="p">,</span> <span class="n">cov_abdab</span><span class="p">,</span> <span class="n">metric</span><span class="o">=</span><span class="n">metric</span><span class="p">,</span> <span class="n">sequence</span><span class="o">=</span><span class="n">sequence</span><span class="p">,</span> <span class="n">include_ref_cols</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;Binding&quot;</span><span class="p">]</span>
<span class="p">)</span>
<span class="n">adata_bcr</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s2">&quot;Binding&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">value_counts</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>100%|███████████████████████████████████| 15443/15443 [00:04&lt;00:00, 3441.29it/s]
100%|█████████████████████████████████████████| 26/26 [00:00&lt;00:00, 3350.08it/s]
</pre></div>
</div>
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>SARS-CoV-2    26
Name: Binding, dtype: int64
</pre></div>
</div>
</div>
</div>
<p>The result of the query shows 26 cells annotated with direct matches to the database in their Heavy-Chain.</p>
<p>In a next step, we will query for matches on Heavy- and Light-Chain:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">ir</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">ir_query</span><span class="p">(</span>
    <span class="n">adata_bcr</span><span class="p">,</span>
    <span class="n">cov_abdab</span><span class="p">,</span>
    <span class="n">metric</span><span class="o">=</span><span class="n">metric</span><span class="p">,</span>
    <span class="n">sequence</span><span class="o">=</span><span class="n">sequence</span><span class="p">,</span>
    <span class="n">receptor_arms</span><span class="o">=</span><span class="s2">&quot;all&quot;</span><span class="p">,</span>
    <span class="n">dual_ir</span><span class="o">=</span><span class="s2">&quot;primary_only&quot;</span><span class="p">,</span>
<span class="p">)</span>
<span class="n">ir</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">ir_query_annotate</span><span class="p">(</span>
    <span class="n">adata_bcr</span><span class="p">,</span>
    <span class="n">cov_abdab</span><span class="p">,</span>
    <span class="n">metric</span><span class="o">=</span><span class="n">metric</span><span class="p">,</span>
    <span class="n">sequence</span><span class="o">=</span><span class="n">sequence</span><span class="p">,</span>
    <span class="n">include_ref_cols</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;Binding&quot;</span><span class="p">],</span>
    <span class="n">suffix</span><span class="o">=</span><span class="s2">&quot;_fullIR&quot;</span><span class="p">,</span>
<span class="p">)</span>
<span class="n">adata_bcr</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s2">&quot;Binding_fullIR&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">value_counts</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>100%|███████████████████████████████████| 19402/19402 [00:09&lt;00:00, 2025.33it/s]
100%|█████████████████████████████████████████| 21/21 [00:00&lt;00:00, 2945.83it/s]
</pre></div>
</div>
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>SARS-CoV-2    21
Name: Binding_fullIR, dtype: int64
</pre></div>
</div>
</div>
</div>
<p>Again, this reduces the amount of hits, but ensures higher hit quality.</p>
</section>
<section id="query-via-hamming-distance">
<h4><span class="section-number">43.3.1.3. </span>Query via Hamming Distance<a class="headerlink" href="#query-via-hamming-distance" title="Link to this heading">#</a></h4>
<p>Due to Somatic Hypermutation, BCRs of the same lineage often differ with mutations on one position, with deletion and addition of amino acids being more unlikely. We will therefore query the database using a Hamming distance, which will mark CDR3 with a single mutation. You can set the threshold for calculating matches during distance calculation. However, we advise using conservative thresholds to limit the amount of false positives.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">metric</span> <span class="o">=</span> <span class="s2">&quot;hamming&quot;</span>
<span class="n">sequence</span> <span class="o">=</span> <span class="s2">&quot;aa&quot;</span>

<span class="n">ir</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">ir_dist</span><span class="p">(</span><span class="n">adata_bcr</span><span class="p">,</span> <span class="n">cov_abdab</span><span class="p">,</span> <span class="n">metric</span><span class="o">=</span><span class="n">metric</span><span class="p">,</span> <span class="n">sequence</span><span class="o">=</span><span class="n">sequence</span><span class="p">)</span>
<span class="n">ir</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">ir_query</span><span class="p">(</span>
    <span class="n">adata_bcr</span><span class="p">,</span>
    <span class="n">cov_abdab</span><span class="p">,</span>
    <span class="n">metric</span><span class="o">=</span><span class="n">metric</span><span class="p">,</span>
    <span class="n">sequence</span><span class="o">=</span><span class="n">seqeunce</span><span class="p">,</span>
    <span class="n">receptor_arms</span><span class="o">=</span><span class="s2">&quot;all&quot;</span><span class="p">,</span>
    <span class="n">dual_ir</span><span class="o">=</span><span class="s2">&quot;primary_only&quot;</span><span class="p">,</span>
<span class="p">)</span>
<span class="n">ir</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">ir_query_annotate</span><span class="p">(</span>
    <span class="n">adata_bcr</span><span class="p">,</span>
    <span class="n">cov_abdab</span><span class="p">,</span>
    <span class="n">metric</span><span class="o">=</span><span class="n">metric</span><span class="p">,</span>
    <span class="n">sequence</span><span class="o">=</span><span class="n">sequence</span><span class="p">,</span>
    <span class="n">include_ref_cols</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;Binding&quot;</span><span class="p">],</span>
    <span class="n">suffix</span><span class="o">=</span><span class="s2">&quot;_Hamming&quot;</span><span class="p">,</span>
<span class="p">)</span>
<span class="n">adata_bcr</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s2">&quot;Binding_Hamming&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">value_counts</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>100%|███████████████████████████████████| 10561/10561 [00:02&lt;00:00, 5105.46it/s]
100%|███████████████████████████████████| 24885/24885 [00:04&lt;00:00, 5575.50it/s]
100%|████████████████████████████████████| 19402/19402 [00:23&lt;00:00, 835.62it/s]
100%|█████████████████████████████████████████| 44/44 [00:00&lt;00:00, 4068.01it/s]
</pre></div>
</div>
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>SARS-CoV-2    44
Name: Binding_Hamming, dtype: int64
</pre></div>
</div>
</div>
</div>
<p>We can see, that we are able to assign more cells by its specificity. With a hamming distance of 1 for both chains, we are able to increase the amount of annotated cells even compared to querying direct matches of the heavy chain.</p>
<p>Such grouping can be used, e.g. when incorporating other modalities such as transcriptome in order to observe the reaction of SARS-CoV-2-specific B-cells within your single-cell study.</p>
</section>
</section>
<section id="distance-measurements">
<h3><span class="section-number">43.3.2. </span>Distance Measurements<a class="headerlink" href="#distance-measurements" title="Link to this heading">#</a></h3>
<p>Contrary to TCRs, a BCR clonotype may contain different sequences due mutations stemming from somatic hypermutation (see chapter 02_clonotypes). The BCRs within this clonotype often target the same epitope with varying strength due to affinity maturation. We therefore already explained distance based clustering in the chapter <LINK-CLONOTYPE-DEFINITION>.</p>
<p>However, several, different clonotype lineages can share their specificity. While these clonotypes are not ancestrally related, they might be related by similar BCR sequences. Recently, methods were developed to compare Antibody sequences based on shared specificity, which can also be applied to BCRs. Since they often rely on structural information (often from prediction) applying these methods is not feasible for large single-cell studies and are, therefore, not included in this tutorial.</p>
</section>
<section id="prediction">
<h3><span class="section-number">43.3.3. </span>Prediction<a class="headerlink" href="#prediction" title="Link to this heading">#</a></h3>
<p>While TCRs bind linear peptides constrained by their binding to the MHC, BCRs can bind to linear or non-continoues antigens formed from proteins and polysaccharides. This unconstrained binding is highly dependent on the three-dimensional structure of BCRs and antigens. There are several prediction tools for antibodies/BCRs focusing on identifying paratopes (binding residues in the AB) or epitopes (binding residues in the antigen). Further, deep learning is used for AB structure prediction, design, optimization, and docking prediction, which often rely on (inferred) spatial structure implying high computational costs. However, these models rather focus on the application of AB development for therapeutics than the analysis of large-scale single cell studies and are therefore out of the scope for this tutorial.</p>
</section>
</section>
<section id="quiz">
<h2><span class="section-number">43.4. </span>Quiz<a class="headerlink" href="#quiz" title="Link to this heading">#</a></h2>
<div class="cell tag_remove_input docutils container">
<div class="cell_output docutils container">
<div class="output text_html">
    <style>
        .output.text_html {
            --pst-color-text-base: transparent !important;
        }
        .flip-card-q1 {
            background-color: transparent;
            width: 350px;
            height: 200px;
            perspective: 1000px;
            display: inline-block;
            margin: 10px;
            text-align: center;
            padding: 10px;
        }
        .flip-card-inner-q1 {
            position: relative;
            width: 100%;
            height: 100%;
            text-align: center;
            transition: transform 0.6s;
            transform-style: preserve-3d;
            transform-origin: center;
        }
        .flip-card-q1:hover .flip-card-inner-q1 {
            transform: rotateY(180deg);
        }
        .flip-card-front-q1, .flip-card-back-q1 {
            position: absolute;
            width: 100%;
            height: 100%;
            min-height: 200px;
            backface-visibility: hidden;
            display: flex;
            align-items: center;
            justify-content: center;
            font-weight: bold;
            color: white;
            padding: 20px;
            border-radius: 15px;
            box-sizing: border-box;
            overflow: hidden;
            text-align: center;
        }
        .flip-card-front-q1 {
            background-color: #3965a3;
            font-size: 20px;
        }
        .flip-card-back-q1 {
            background-color: #a8d480;
            transform: rotateY(180deg);
            font-size: 20px;
        }_b
    </style>

    <div class="flip-card-q1">
        <div class="flip-card-inner-q1">
            <div class="flip-card-front-q1">
                What is assumed to be the most important AIR information, when considering specificity?
            </div>
            <div class="flip-card-back-q1">
                CDR3 of the VDJ-chain, V- and J-gene of the VJ-chain, etc.
            </div>
        </div>
    </div>
    <noscript>
        <p><strong>Q:</strong> What is assumed to be the most important AIR information, when considering specificity?</p>
        <p><strong>Answer:</strong> CDR3 of the VDJ-chain, V- and J-gene of the VJ-chain, etc.</p>
    </noscript>
    </div><div class="output text_html">
    <style>
        .output.text_html {
            --pst-color-text-base: transparent !important;
        }
        .flip-card-q2 {
            background-color: transparent;
            width: 350px;
            height: 200px;
            perspective: 1000px;
            display: inline-block;
            margin: 10px;
            text-align: center;
            padding: 10px;
        }
        .flip-card-inner-q2 {
            position: relative;
            width: 100%;
            height: 100%;
            text-align: center;
            transition: transform 0.6s;
            transform-style: preserve-3d;
            transform-origin: center;
        }
        .flip-card-q2:hover .flip-card-inner-q2 {
            transform: rotateY(180deg);
        }
        .flip-card-front-q2, .flip-card-back-q2 {
            position: absolute;
            width: 100%;
            height: 100%;
            min-height: 200px;
            backface-visibility: hidden;
            display: flex;
            align-items: center;
            justify-content: center;
            font-weight: bold;
            color: white;
            padding: 20px;
            border-radius: 15px;
            box-sizing: border-box;
            overflow: hidden;
            text-align: center;
        }
        .flip-card-front-q2 {
            background-color: #3965a3;
            font-size: 20px;
        }
        .flip-card-back-q2 {
            background-color: #a8d480;
            transform: rotateY(180deg);
            font-size: 20px;
        }_b
    </style>

    <div class="flip-card-q2">
        <div class="flip-card-inner-q2">
            <div class="flip-card-front-q2">
                How can you computationally infer specificity of AIR cells?
            </div>
            <div class="flip-card-back-q2">
                Transforming the epitope Sequence to CDR3-format, querying appropriate databases, etc.
            </div>
        </div>
    </div>
    <noscript>
        <p><strong>Q:</strong> How can you computationally infer specificity of AIR cells?</p>
        <p><strong>Answer:</strong> Transforming the epitope Sequence to CDR3-format, querying appropriate databases, etc.</p>
    </noscript>
    </div><div class="output text_html">
    <style>
        .output.text_html {
            --pst-color-text-base: transparent !important;
        }
        .flip-card-q3 {
            background-color: transparent;
            width: 350px;
            height: 200px;
            perspective: 1000px;
            display: inline-block;
            margin: 10px;
            text-align: center;
            padding: 10px;
        }
        .flip-card-inner-q3 {
            position: relative;
            width: 100%;
            height: 100%;
            text-align: center;
            transition: transform 0.6s;
            transform-style: preserve-3d;
            transform-origin: center;
        }
        .flip-card-q3:hover .flip-card-inner-q3 {
            transform: rotateY(180deg);
        }
        .flip-card-front-q3, .flip-card-back-q3 {
            position: absolute;
            width: 100%;
            height: 100%;
            min-height: 200px;
            backface-visibility: hidden;
            display: flex;
            align-items: center;
            justify-content: center;
            font-weight: bold;
            color: white;
            padding: 20px;
            border-radius: 15px;
            box-sizing: border-box;
            overflow: hidden;
            text-align: center;
        }
        .flip-card-front-q3 {
            background-color: #3965a3;
            font-size: 20px;
        }
        .flip-card-back-q3 {
            background-color: #a8d480;
            transform: rotateY(180deg);
            font-size: 20px;
        }_b
    </style>

    <div class="flip-card-q3">
        <div class="flip-card-inner-q3">
            <div class="flip-card-front-q3">
                Why should you be careful when using database queries?
            </div>
            <div class="flip-card-back-q3">
                Most databases only include low-throughput samples, databases are biased towoard well researched diseases or epitopes, etc.
            </div>
        </div>
    </div>
    <noscript>
        <p><strong>Q:</strong> Why should you be careful when using database queries?</p>
        <p><strong>Answer:</strong> Most databases only include low-throughput samples, databases are biased towoard well researched diseases or epitopes, etc.</p>
    </noscript>
    </div></div>
</div>
</section>
<section id="references">
<h2><span class="section-number">43.5. </span>References<a class="headerlink" href="#references" title="Link to this heading">#</a></h2>
<div class="docutils container" id="id17">
<div role="list" class="citation-list">
<div class="citation" id="id497" role="doc-biblioentry">
<span class="label"><span class="fn-bracket">[</span><a role="doc-backlink" href="#id12">CCM+21</a><span class="fn-bracket">]</span></span>
<p>William D Chronister, Austin Crinklaw, Swapnil Mahajan, Randi Vita, Zeynep Koşaloğlu-Yalçın, Zhen Yan, Jason A Greenbaum, Leon E Jessen, Morten Nielsen, Scott Christley, and others. Tcrmatch: predicting t-cell receptor specificity based on sequence similarity to previously characterized receptors. <em>Frontiers in immunology</em>, 12:673, 2021.</p>
</div>
<div class="citation" id="id496" role="doc-biblioentry">
<span class="label"><span class="fn-bracket">[</span><a role="doc-backlink" href="#id10">DFGH+17</a><span class="fn-bracket">]</span></span>
<p>Pradyot Dash, Andrew J Fiore-Gartland, Tomer Hertz, George C Wang, Shalini Sharma, Aisha Souquette, Jeremy Chase Crawford, E Bridie Clemens, Thi HO Nguyen, Katherine Kedzierska, and others. Quantifiable predictive features define epitope-specific t cell receptor repertoires. <em>Nature</em>, 547(7661):89–93, 2017.</p>
</div>
<div class="citation" id="id500" role="doc-biblioentry">
<span class="label"><span class="fn-bracket">[</span><a role="doc-backlink" href="#id3">DB88</a><span class="fn-bracket">]</span></span>
<p>Mark M Davis and Pamela J Bjorkman. T-cell antigen receptor genes and t-cell recognition. <em>Nature</em>, 334(6181):395–402, 1988.</p>
</div>
<div class="citation" id="id490" role="doc-biblioentry">
<span class="label"><span class="fn-bracket">[</span>FPD+17<span class="fn-bracket">]</span></span>
<span class="backrefs">(<a role="doc-backlink" href="#id5">1</a>,<a role="doc-backlink" href="#id14">2</a>)</span>
<p>Ward Fleri, Sinu Paul, Sandeep Kumar Dhanda, Swapnil Mahajan, Xiaojun Xu, Bjoern Peters, and Alessandro Sette. The immune epitope database and analysis resource in epitope discovery and synthetic vaccine design. <em>Frontiers in immunology</em>, 8:278, 2017.</p>
</div>
<div class="citation" id="id499" role="doc-biblioentry">
<span class="label"><span class="fn-bracket">[</span>GHN+17<span class="fn-bracket">]</span></span>
<span class="backrefs">(<a role="doc-backlink" href="#id2">1</a>,<a role="doc-backlink" href="#id3">2</a>)</span>
<p>Jacob Glanville, Huang Huang, Allison Nau, Olivia Hatton, Lisa E Wagar, Florian Rubelt, Xuhuai Ji, Arnold Han, Sheri M Krams, Christina Pettus, and others. Identifying specificity groups in the t cell receptor repertoire. <em>Nature</em>, 547(7661):94–98, 2017.</p>
</div>
<div class="citation" id="id503" role="doc-biblioentry">
<span class="label"><span class="fn-bracket">[</span><a role="doc-backlink" href="#id11">HH92</a><span class="fn-bracket">]</span></span>
<p>Steven Henikoff and Jorja G Henikoff. Amino acid substitution matrices from protein blocks. <em>Proceedings of the National Academy of Sciences</em>, 89(22):10915–10919, 1992.</p>
</div>
<div class="citation" id="id495" role="doc-biblioentry">
<span class="label"><span class="fn-bracket">[</span><a role="doc-backlink" href="#id9">NVK+20</a><span class="fn-bracket">]</span></span>
<p>Sean Nolan, Marissa Vignali, Mark Klinger, Jennifer N Dines, Ian M Kaplan, Emily Svejnoha, Tracy Craft, Katie Boland, Mitch Pesesky, Rachel M Gittelman, and others. A large-scale database of t-cell receptor beta (tcrβ) sequences and binding associations from natural and synthetic exposure to sars-cov-2. <em>Research square</em>, 2020.</p>
</div>
<div class="citation" id="id498" role="doc-biblioentry">
<span class="label"><span class="fn-bracket">[</span><a role="doc-backlink" href="#id16">RKMD21</a><span class="fn-bracket">]</span></span>
<p>Matthew I. J. Raybould, Aleksandr Kovaltsuk, Claire Marks, and Charlotte M. Deane. Cov-abdab: the coronavirus antibody database. <em>Bioinformatics</em>, 37(5):734–735, 2021. URL: <a class="reference external" href="https://academic.oup.com/bioinformatics/advance-article/doi/10.1093/bioinformatics/btaa739/5893556">https://academic.oup.com/bioinformatics/advance-article/doi/10.1093/bioinformatics/btaa739/5893556</a>, <a class="reference external" href="https://doi.org/10.1093/bioinformatics/btaa739">doi:10.1093/bioinformatics/btaa739</a>.</p>
</div>
<div class="citation" id="id501" role="doc-biblioentry">
<span class="label"><span class="fn-bracket">[</span><a role="doc-backlink" href="#id3">RSW06</a><span class="fn-bracket">]</span></span>
<p>Markus G Rudolph, Robyn L Stanfield, and Ian A Wilson. How tcrs bind mhcs, peptides, and coreceptors. <em>Annual review of immunology</em>, 24(1):419–466, 2006.</p>
</div>
<div class="citation" id="id489" role="doc-biblioentry">
<span class="label"><span class="fn-bracket">[</span><a role="doc-backlink" href="#id6">SBZ+18</a><span class="fn-bracket">]</span></span>
<p>Mikhail Shugay, Dmitriy V Bagaev, Ivan V Zvyagin, Renske M Vroomans, Jeremy Chase Crawford, Garry Dolton, Ekaterina A Komech, Anastasiya L Sycheva, Anna E Koneva, Evgeniy S Egorov, and others. Vdjdb: a curated database of t-cell receptor sequences with known antigen specificity. <em>Nucleic acids research</em>, 46(D1):D419–D427, 2018.</p>
</div>
<div class="citation" id="id487" role="doc-biblioentry">
<span class="label"><span class="fn-bracket">[</span>STL21<span class="fn-bracket">]</span></span>
<span class="backrefs">(<a role="doc-backlink" href="#id1">1</a>,<a role="doc-backlink" href="#id4">2</a>)</span>
<p>Ido Springer, Nili Tickotsky, and Yoram Louzoun. Contribution of t cell receptor alpha and beta cdr3, mhc typing, v and j genes to peptide binding prediction. <em>Frontiers in immunology</em>, 2021.</p>
</div>
<div class="citation" id="id488" role="doc-biblioentry">
<span class="label"><span class="fn-bracket">[</span><a role="doc-backlink" href="#id7">TSP+17</a><span class="fn-bracket">]</span></span>
<p>Nili Tickotsky, Tal Sagiv, Jaime Prilusky, Eric Shifrut, and Nir Friedman. Mcpas-tcr: a manually curated catalogue of pathology-associated t cell receptor sequences. <em>Bioinformatics</em>, 33(18):2924–2929, 2017.</p>
</div>
<div class="citation" id="id491" role="doc-biblioentry">
<span class="label"><span class="fn-bracket">[</span>ZWL+20<span class="fn-bracket">]</span></span>
<span class="backrefs">(<a role="doc-backlink" href="#id8">1</a>,<a role="doc-backlink" href="#id15">2</a>)</span>
<p>Wei Zhang, Longlong Wang, Ke Liu, Xiaofeng Wei, Kai Yang, Wensi Du, Shiyu Wang, Nannan Guo, Chuanchuan Ma, Lihua Luo, and others. Pird: pan immune repertoire database. <em>Bioinformatics</em>, 36(3):897–903, 2020.</p>
</div>
</div>
</div>
</section>
</section>

    <script type="text/x-thebe-config">
    {
        requestKernel: true,
        binderOptions: {
            repo: "binder-examples/jupyter-stacks-datascience",
            ref: "master",
        },
        codeMirrorConfig: {
            theme: "abcdef",
            mode: "python"
        },
        kernelOptions: {
            name: "python3",
            path: "./air_repertoire"
        },
        predefinedOutput: true
    }
    </script>
    <script>kernelName = 'python3'</script>

                </article>
              

              
              
              
              
                <footer class="prev-next-footer d-print-none">
                  
<div class="prev-next-area">
    <a class="left-prev"
       href="clonotype.html"
       title="previous page">
      <i class="fa-solid fa-angle-left"></i>
      <div class="prev-next-info">
        <p class="prev-next-subtitle">previous</p>
        <p class="prev-next-title"><span class="section-number">39. </span>Clonotype analysis</p>
      </div>
    </a>
    <a class="right-next"
       href="multimodal_integration.html"
       title="next page">
      <div class="prev-next-info">
        <p class="prev-next-subtitle">next</p>
        <p class="prev-next-title"><span class="section-number">44. </span>Integrating AIR and transcriptomics</p>
      </div>
      <i class="fa-solid fa-angle-right"></i>
    </a>
</div>
                </footer>
              
            </div>
            
            
              
                <div class="bd-sidebar-secondary bd-toc"><div class="sidebar-secondary-items sidebar-secondary__inner">


  <div class="sidebar-secondary-item">
  <div class="page-toc tocsection onthispage">
    <i class="fa-solid fa-list"></i> Contents
  </div>
  <nav class="bd-toc-nav page-toc">
    <ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#introduction">43.1. Introduction</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#tcr-specificity-analysis">43.2. TCR specificity analysis</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#database-queries">43.2.1. Database Queries</a><ul class="nav section-nav flex-column">
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#manual-query">43.2.1.1. Manual Query</a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#query-with-various-strictness">43.2.1.2. Query with various Strictness</a></li>
</ul>
</li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#distances">43.2.2. Distances</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#tcrdist">43.2.3. TCRdist</a><ul class="nav section-nav flex-column">
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#tcrmatch">43.2.3.1. TCRmatch</a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#database-query-via-distance-metrics">43.2.3.2. Database query via distance metrics</a></li>
</ul>
</li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#epitope-prediction">43.2.4. Epitope Prediction</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#bcr-specificity-analysis">43.3. BCR specificity analysis</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#id13">43.3.1. Database Queries</a><ul class="nav section-nav flex-column">
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#database-preparation">43.3.1.1. Database Preparation</a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#identical-matches">43.3.1.2. Identical matches</a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#query-via-hamming-distance">43.3.1.3. Query via Hamming Distance</a></li>
</ul>
</li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#distance-measurements">43.3.2. Distance Measurements</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#prediction">43.3.3. Prediction</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#quiz">43.4. Quiz</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#references">43.5. References</a></li>
</ul>
  </nav></div>

</div></div>
              
            
          </div>
          <footer class="bd-footer-content">
            
<div class="bd-footer-content__inner container">
  
  <div class="footer-item">
    
<p class="component-author">
By Lukas Heumos, Anna Schaar, single-cell best practices consortium
</p>

  </div>
  
  <div class="footer-item">
    

  <p class="copyright">
    
      © Copyright 2023.
      <br/>
    
  </p>

  </div>
  
  <div class="footer-item">
    
  </div>
  
  <div class="footer-item">
    
<div class="extra_footer">
  <div>
  Brought to you by
  <a href="https://www.github.com/theislab">Theislab</a>, with many thanks to the single-cell community as a whole!
</div>

</div>
  </div>
  
</div>
          </footer>
        

      </main>
    </div>
  </div>
  
  <!-- Scripts loaded after <body> so the DOM is not blocked -->
  <script src="../_static/scripts/bootstrap.js?digest=dfe6caa3a7d634c4db9b"></script>
<script src="../_static/scripts/pydata-sphinx-theme.js?digest=dfe6caa3a7d634c4db9b"></script>

  <footer class="bd-footer">
  </footer>
  </body>
</html>