
<!DOCTYPE html>


<html lang="en" data-content_root="../" >

  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="viewport" content="width=device-width, initial-scale=1" />

    <title>3. Raw data processing &#8212; Single-cell best practices</title>
  
  
  
  <script data-cfasync="false">
    document.documentElement.dataset.mode = localStorage.getItem("mode") || "";
    document.documentElement.dataset.theme = localStorage.getItem("theme") || "";
  </script>
  
  <!-- Loaded before other Sphinx assets -->
  <link href="../_static/styles/theme.css?digest=dfe6caa3a7d634c4db9b" rel="stylesheet" />
<link href="../_static/styles/bootstrap.css?digest=dfe6caa3a7d634c4db9b" rel="stylesheet" />
<link href="../_static/styles/pydata-sphinx-theme.css?digest=dfe6caa3a7d634c4db9b" rel="stylesheet" />

  
  <link href="../_static/vendor/fontawesome/6.5.2/css/all.min.css?digest=dfe6caa3a7d634c4db9b" rel="stylesheet" />
  <link rel="preload" as="font" type="font/woff2" crossorigin href="../_static/vendor/fontawesome/6.5.2/webfonts/fa-solid-900.woff2" />
<link rel="preload" as="font" type="font/woff2" crossorigin href="../_static/vendor/fontawesome/6.5.2/webfonts/fa-brands-400.woff2" />
<link rel="preload" as="font" type="font/woff2" crossorigin href="../_static/vendor/fontawesome/6.5.2/webfonts/fa-regular-400.woff2" />

    <link rel="stylesheet" type="text/css" href="../_static/pygments.css?v=03e43079" />
    <link rel="stylesheet" type="text/css" href="../_static/styles/sphinx-book-theme.css?v=eba8b062" />
    <link rel="stylesheet" type="text/css" href="../_static/togglebutton.css?v=13237357" />
    <link rel="stylesheet" type="text/css" href="../_static/copybutton.css?v=76b2166b" />
    <link rel="stylesheet" type="text/css" href="../_static/mystnb.4510f1fc1dee50b3e5859aac5469c37c29e427902b24a333a5f9fcb2f0b3ac41.css" />
    <link rel="stylesheet" type="text/css" href="../_static/sphinx-thebe.css?v=4fa983c6" />
    <link rel="stylesheet" type="text/css" href="../_static/sphinx-design.min.css?v=95c83b7e" />
    <link rel="stylesheet" type="text/css" href="../_static/book.css?v=148c11e2" />
  
  <!-- Pre-loaded scripts that we'll load fully later -->
  <link rel="preload" as="script" href="../_static/scripts/bootstrap.js?digest=dfe6caa3a7d634c4db9b" />
<link rel="preload" as="script" href="../_static/scripts/pydata-sphinx-theme.js?digest=dfe6caa3a7d634c4db9b" />
  <script src="../_static/vendor/fontawesome/6.5.2/js/all.min.js?digest=dfe6caa3a7d634c4db9b"></script>

    <script src="../_static/documentation_options.js?v=9eb32ce0"></script>
    <script src="../_static/doctools.js?v=9a2dae69"></script>
    <script src="../_static/sphinx_highlight.js?v=dc90522c"></script>
    <script src="../_static/clipboard.min.js?v=a7894cd8"></script>
    <script src="../_static/copybutton.js?v=f281be69"></script>
    <script src="../_static/scripts/sphinx-book-theme.js?v=887ef09a"></script>
    <script>let toggleHintShow = 'Click to show';</script>
    <script>let toggleHintHide = 'Click to hide';</script>
    <script>let toggleOpenOnPrint = 'true';</script>
    <script src="../_static/togglebutton.js?v=4a39c7ea"></script>
    <script>var togglebuttonSelector = '.toggle, .admonition.dropdown';</script>
    <script src="../_static/design-tabs.js?v=f930bc37"></script>
    <script>const THEBE_JS_URL = "https://unpkg.com/thebe@0.8.2/lib/index.js"; const thebe_selector = ".thebe,.cell"; const thebe_selector_input = "pre"; const thebe_selector_output = ".output, .cell_output"</script>
    <script async="async" src="../_static/sphinx-thebe.js?v=c100c467"></script>
    <script>var togglebuttonSelector = '.toggle, .admonition.dropdown';</script>
    <script>const THEBE_JS_URL = "https://unpkg.com/thebe@0.8.2/lib/index.js"; const thebe_selector = ".thebe,.cell"; const thebe_selector_input = "pre"; const thebe_selector_output = ".output, .cell_output"</script>
    <script>window.MathJax = {"options": {"processHtmlClass": "tex2jax_process|mathjax_process|math|output_area"}}</script>
    <script defer="defer" src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-mml-chtml.js"></script>
    <script>DOCUMENTATION_OPTIONS.pagename = 'introduction/raw_data_processing';</script>
    <script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
    <script src="../_static/book.js?v=57d67cd3"></script>
    <link rel="icon" href="../_static/favicon.ico"/>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="4. Analysis frameworks and tools" href="analysis_tools.html" />
    <link rel="prev" title="2. Single-cell RNA sequencing" href="scrna_seq.html" />
  <meta name="viewport" content="width=device-width, initial-scale=1"/>
  <meta name="docsearch:language" content="en"/>
  </head>
  
  
  <body data-bs-spy="scroll" data-bs-target=".bd-toc-nav" data-offset="180" data-bs-root-margin="0px 0px -60%" data-default-mode="">

  
  
  <div id="pst-skip-link" class="skip-link d-print-none"><a href="#main-content">Skip to main content</a></div>
  
  <div id="pst-scroll-pixel-helper"></div>
  
  <button type="button" class="btn rounded-pill" id="pst-back-to-top">
    <i class="fa-solid fa-arrow-up"></i>Back to top</button>

  
  <input type="checkbox"
          class="sidebar-toggle"
          id="pst-primary-sidebar-checkbox"/>
  <label class="overlay overlay-primary" for="pst-primary-sidebar-checkbox"></label>
  
  <input type="checkbox"
          class="sidebar-toggle"
          id="pst-secondary-sidebar-checkbox"/>
  <label class="overlay overlay-secondary" for="pst-secondary-sidebar-checkbox"></label>
  
  <div class="search-button__wrapper">
    <div class="search-button__overlay"></div>
    <div class="search-button__search-container">
<form class="bd-search d-flex align-items-center"
      action="../search.html"
      method="get">
  <i class="fa-solid fa-magnifying-glass"></i>
  <input type="search"
         class="form-control"
         name="q"
         id="search-input"
         placeholder="Search this book..."
         aria-label="Search this book..."
         autocomplete="off"
         autocorrect="off"
         autocapitalize="off"
         spellcheck="false"/>
  <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd>K</kbd></span>
</form></div>
  </div>

  <div class="pst-async-banner-revealer d-none">
  <aside id="bd-header-version-warning" class="d-none d-print-none" aria-label="Version warning"></aside>
</div>

  
    <header class="bd-header navbar navbar-expand-lg bd-navbar d-print-none">
    </header>
  

  <div class="bd-container">
    <div class="bd-container__inner bd-page-width">
      
      
      
      <div class="bd-sidebar-primary bd-sidebar">
        

  
  <div class="sidebar-header-items sidebar-primary__section">
    
    
    
    
  </div>
  
    <div class="sidebar-primary-items__start sidebar-primary__section">
        <div class="sidebar-primary-item">

  
    
  

<a class="navbar-brand logo" href="../preamble.html">
  
  
  
  
  
    
    
      
    
    
    <img src="../_static/logo.png" class="logo__image only-light" alt="Single-cell best practices - Home"/>
    <script>document.write(`<img src="../_static/logo.png" class="logo__image only-dark" alt="Single-cell best practices - Home"/>`);</script>
  
  
</a></div>
        <div class="sidebar-primary-item">

 <script>
 document.write(`
   <button class="btn search-button-field search-button__button" title="Search" aria-label="Search" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="fa-solid fa-magnifying-glass"></i>
    <span class="search-button__default-text">Search</span>
    <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd class="kbd-shortcut__modifier">K</kbd></span>
   </button>
 `);
 </script></div>
        <div class="sidebar-primary-item"><nav class="bd-links bd-docs-nav" aria-label="Main">
    <div class="bd-toc-item navbar-nav active">
        <p aria-level="2" class="caption" role="heading"><span class="caption-text">Introduction</span></p>
<ul class="current nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="prior_art.html">1. Prior art</a></li>
<li class="toctree-l1"><a class="reference internal" href="scrna_seq.html">2. Single-cell RNA sequencing</a></li>
<li class="toctree-l1 current active"><a class="current reference internal" href="#">3. Raw data processing</a></li>
<li class="toctree-l1"><a class="reference internal" href="analysis_tools.html">4. Analysis frameworks and tools</a></li>
<li class="toctree-l1"><a class="reference internal" href="interoperability.html">5. Interoperability</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Preprocessing and visualization</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../preprocessing_visualization/quality_control.html">6. Quality Control</a></li>
<li class="toctree-l1"><a class="reference internal" href="../preprocessing_visualization/normalization.html">7. Normalization</a></li>
<li class="toctree-l1"><a class="reference internal" href="../preprocessing_visualization/feature_selection.html">8. Feature selection</a></li>
<li class="toctree-l1"><a class="reference internal" href="../preprocessing_visualization/dimensionality_reduction.html">9. Dimensionality Reduction</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Identifying cellular structure</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../cellular_structure/clustering.html">10. Clustering</a></li>
<li class="toctree-l1"><a class="reference internal" href="../cellular_structure/annotation.html">11. Annotation</a></li>
<li class="toctree-l1"><a class="reference internal" href="../cellular_structure/integration.html">12. Data integration</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Inferring trajectories</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../trajectories/pseudotemporal.html">13. Pseudotemporal ordering</a></li>
<li class="toctree-l1"><a class="reference internal" href="../trajectories/rna_velocity.html">14. RNA velocity</a></li>
<li class="toctree-l1"><a class="reference internal" href="../trajectories/lineage_tracing.html">15. Lineage tracing</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Dealing with conditions</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../conditions/differential_gene_expression.html">16. Differential gene expression analysis</a></li>
<li class="toctree-l1"><a class="reference internal" href="../conditions/compositional.html">17. Compositional analysis</a></li>
<li class="toctree-l1"><a class="reference internal" href="../conditions/gsea_pathway.html">18. Gene set enrichment and pathway analysis</a></li>
<li class="toctree-l1"><a class="reference internal" href="../conditions/perturbation_modeling.html">19. Perturbation modeling</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Modeling mechanisms</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../mechanisms/gene_regulatory_networks.html">20. Gene regulatory networks</a></li>
<li class="toctree-l1"><a class="reference internal" href="../mechanisms/cell_cell_communication.html">21. Cell-cell communication</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Deconvolution</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../deconvolution/bulk_deconvolution.html">22. Bulk deconvolution</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Chromatin Accessibility</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../chromatin_accessibility/introduction.html">23. Single-cell ATAC sequencing</a></li>
<li class="toctree-l1"><a class="reference internal" href="../chromatin_accessibility/quality_control.html">24. Quality Control</a></li>
<li class="toctree-l1"><a class="reference internal" href="../chromatin_accessibility/gene_regulatory_networks_atac.html">25. Gene regulatory networks using chromatin accessibility</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Spatial omics</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../spatial/introduction.html">26. Single-cell data resolved in space</a></li>
<li class="toctree-l1"><a class="reference internal" href="../spatial/neighborhood.html">27. Neighborhood analysis</a></li>
<li class="toctree-l1"><a class="reference internal" href="../spatial/domains.html">28. Spatial domains</a></li>
<li class="toctree-l1"><a class="reference internal" href="../spatial/spatially_variable_genes.html">29. Spatially variable genes</a></li>
<li class="toctree-l1"><a class="reference internal" href="../spatial/deconvolution.html">30. Spatial deconvolution</a></li>
<li class="toctree-l1"><a class="reference internal" href="../spatial/imputation.html">31. Imputation</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Surface protein</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../surface_protein/quality_control.html">32. Quality control</a></li>
<li class="toctree-l1"><a class="reference internal" href="../surface_protein/normalization.html">33. Normalization</a></li>
<li class="toctree-l1"><a class="reference internal" href="../surface_protein/doublet_detection.html">34. Doublet detection</a></li>
<li class="toctree-l1"><a class="reference internal" href="../surface_protein/dimensionality_reduction.html">35. Dimensionality Reduction</a></li>
<li class="toctree-l1"><a class="reference internal" href="../surface_protein/batch_correction.html">36. Batch correction</a></li>
<li class="toctree-l1"><a class="reference internal" href="../surface_protein/annotation.html">37. Annotation</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Adaptive immune receptor repertoire</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../air_repertoire/ir_profiling.html">38. Immune Receptor Profiling</a></li>
<li class="toctree-l1"><a class="reference internal" href="../air_repertoire/clonotype.html">39. Clonotype analysis</a></li>



<li class="toctree-l1"><a class="reference internal" href="../air_repertoire/specificity.html">43. Specificity analysis</a></li>
<li class="toctree-l1"><a class="reference internal" href="../air_repertoire/multimodal_integration.html">44. Integrating AIR and transcriptomics</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Multimodal integration</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../multimodal_integration/paired_integration.html">45. Paired integration</a></li>
<li class="toctree-l1"><a class="reference internal" href="../multimodal_integration/advanced_integration.html">46. Advanced integration</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Outlook</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../outlook.html">47. Outlook</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Acknowledgements</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../acknowledgements.html">48. Acknowledgements</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Glossary</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../glossary.html">49. Glossary</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Changelog</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../CHANGELOG.html">Changelog</a></li>
</ul>

    </div>
</nav></div>
    </div>
  
  
  <div class="sidebar-primary-items__end sidebar-primary__section">
  </div>
  
  <div id="rtd-footer-container"></div>


      </div>
      
      <main id="main-content" class="bd-main" role="main">
        
        

<div class="sbt-scroll-pixel-helper"></div>

          <div class="bd-content">
            <div class="bd-article-container">
              
              <div class="bd-header-article d-print-none">
<div class="header-article-items header-article__inner">
  
    <div class="header-article-items__start">
      
        <div class="header-article-item"><button class="sidebar-toggle primary-toggle btn btn-sm" title="Toggle primary sidebar" data-bs-placement="bottom" data-bs-toggle="tooltip">
  <span class="fa-solid fa-bars"></span>
</button></div>
      
    </div>
  
  
    <div class="header-article-items__end">
      
        <div class="header-article-item">

<div class="article-header-buttons">





<div class="dropdown dropdown-source-buttons">
  <button class="btn dropdown-toggle" type="button" data-bs-toggle="dropdown" aria-expanded="false" aria-label="Source repositories">
    <i class="fab fa-github"></i>
  </button>
  <ul class="dropdown-menu">
      
      
      
      <li><a href="https://github.com/theislab/single-cell-best-practices" target="_blank"
   class="btn btn-sm btn-source-repository-button dropdown-item"
   title="Source repository"
   data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fab fa-github"></i>
  </span>
<span class="btn__text-container">Repository</span>
</a>
</li>
      
      
      
      
      <li><a href="https://github.com/theislab/single-cell-best-practices/edit/master/jupyter-book/introduction/raw_data_processing.md" target="_blank"
   class="btn btn-sm btn-source-edit-button dropdown-item"
   title="Suggest edit"
   data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-pencil-alt"></i>
  </span>
<span class="btn__text-container">Suggest edit</span>
</a>
</li>
      
      
      
      
      <li><a href="https://github.com/theislab/single-cell-best-practices/issues/new?title=Issue%20on%20page%20%2Fintroduction/raw_data_processing.html&body=Your%20issue%20content%20here." target="_blank"
   class="btn btn-sm btn-source-issues-button dropdown-item"
   title="Open an issue"
   data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-lightbulb"></i>
  </span>
<span class="btn__text-container">Open issue</span>
</a>
</li>
      
  </ul>
</div>






<div class="dropdown dropdown-download-buttons">
  <button class="btn dropdown-toggle" type="button" data-bs-toggle="dropdown" aria-expanded="false" aria-label="Download this page">
    <i class="fas fa-download"></i>
  </button>
  <ul class="dropdown-menu">
      
      
      
      <li><a href="../_sources/introduction/raw_data_processing.md" target="_blank"
   class="btn btn-sm btn-download-source-button dropdown-item"
   title="Download source file"
   data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-file"></i>
  </span>
<span class="btn__text-container">.md</span>
</a>
</li>
      
      
      
      
      <li>
<button onclick="window.print()"
  class="btn btn-sm btn-download-pdf-button dropdown-item"
  title="Print to PDF"
  data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-file-pdf"></i>
  </span>
<span class="btn__text-container">.pdf</span>
</button>
</li>
      
  </ul>
</div>




<button onclick="toggleFullScreen()"
  class="btn btn-sm btn-fullscreen-button"
  title="Fullscreen mode"
  data-bs-placement="bottom" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-expand"></i>
  </span>

</button>



<script>
document.write(`
  <button class="btn btn-sm nav-link pst-navbar-icon theme-switch-button" title="light/dark" aria-label="light/dark" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="theme-switch fa-solid fa-sun fa-lg" data-mode="light"></i>
    <i class="theme-switch fa-solid fa-moon fa-lg" data-mode="dark"></i>
    <i class="theme-switch fa-solid fa-circle-half-stroke fa-lg" data-mode="auto"></i>
  </button>
`);
</script>


<script>
document.write(`
  <button class="btn btn-sm pst-navbar-icon search-button search-button__button" title="Search" aria-label="Search" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="fa-solid fa-magnifying-glass fa-lg"></i>
  </button>
`);
</script>
<button class="sidebar-toggle secondary-toggle btn btn-sm" title="Toggle secondary sidebar" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <span class="fa-solid fa-list"></span>
</button>
</div></div>
      
    </div>
  
</div>
</div>
              
              

<div id="jb-print-docs-body" class="onlyprint">
    <h1>Raw data processing</h1>
    <!-- Table of contents -->
    <div id="print-main-content">
        <div id="jb-print-toc">
            
            <div>
                <h2> Contents </h2>
            </div>
            <nav aria-label="Page">
                <ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#motivation">3.1. Motivation</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#raw-data-quality-control">3.2. Raw data quality control</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#alignment-and-mapping">3.3. Alignment and mapping</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#types-of-mapping">3.3.1. Types of mapping</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#mapping-against-different-reference-sequences">3.3.2. Mapping against different reference sequences</a><ul class="nav section-nav flex-column">
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#mapping-to-the-full-genome">3.3.2.1. Mapping to the full genome</a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#mapping-to-the-spliced-transcriptome">3.3.2.2. Mapping to the spliced transcriptome</a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#mapping-to-an-augmented-transcriptome">3.3.2.3. Mapping to an augmented transcriptome</a></li>
</ul>
</li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#cell-barcode-correction">3.4. Cell barcode correction</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#type-of-errors-in-barcoding">3.4.1. Type of errors in barcoding</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#umi-resolution">3.5. UMI resolution</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#the-need-for-umi-resolution">3.5.1. The need for UMI resolution</a><ul class="nav section-nav flex-column">
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#quantification">3.5.1.1. Quantification</a></li>
</ul>
</li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#count-matrix-quality-control">3.6. Count matrix quality control</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#empty-droplet-detection">3.6.1. Empty droplet detection</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#doublet-detection">3.6.2. Doublet detection</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#count-data-representation">3.7. Count data representation</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#brief-discussion">3.8. Brief discussion</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#a-real-world-example">3.9. A real-world example</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#preparation">3.9.1. Preparation</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#simplified-raw-data-processing-pipeline">3.9.2. Simplified raw data processing pipeline</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#the-complete-alevin-fry-pipeline">3.9.3. The complete alevin-fry pipeline</a><ul class="nav section-nav flex-column">
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#building-the-index">3.9.3.1. Building the index</a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#mapping-and-quantification">3.9.3.2. Mapping and quantification</a></li>
</ul>
</li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#useful-links">3.10. Useful links</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#references">3.11. References</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#contributors">3.12. Contributors</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#authors">3.12.1. Authors</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#reviewers">3.12.2. Reviewers</a></li>
</ul>
</li>
</ul>
            </nav>
        </div>
    </div>
</div>

              
                
<div id="searchbox"></div>
                <article class="bd-article">
                  
  <section class="tex2jax_ignore mathjax_ignore" id="raw-data-processing">
<span id="raw-proc"></span><h1><span class="section-number">3. </span>Raw data processing<a class="headerlink" href="#raw-data-processing" title="Link to this heading">#</a></h1>
<details class="sd-sphinx-override sd-dropdown sd-card sd-mb-3">
<summary class="sd-summary-title sd-card-header">
<span class="sd-summary-text"><i class="fas fa-brain"></i>   Key takeaways</span><span class="sd-summary-state-marker sd-summary-chevron-right"><svg version="1.1" width="1.5em" height="1.5em" class="sd-octicon sd-octicon-chevron-right" viewBox="0 0 24 24" aria-hidden="true"><path d="M8.72 18.78a.75.75 0 0 1 0-1.06L14.44 12 8.72 6.28a.751.751 0 0 1 .018-1.042.751.751 0 0 1 1.042-.018l6.25 6.25a.75.75 0 0 1 0 1.06l-6.25 6.25a.75.75 0 0 1-1.06 0Z"></path></svg></span></summary><div class="sd-summary-content sd-card-body docutils">
<div class="sd-card sd-sphinx-override sd-mb-3 sd-shadow-sm sd-card-hover docutils">
<div class="sd-card-body docutils">
<p class="sd-card-text">Accurate raw data processing transforms FASTQ files into a count matrix, enabling reliable single-cell sequencing analyses through read alignment, barcode correction, and UMI counting.</p>
</div>
<a class="sd-stretched-link sd-hide-link-text reference internal" href="#introduction-raw-data-processing-key-takeaway-1"><span class="std std-ref">Motivation</span></a></div>
</div>
</details><details class="sd-sphinx-override sd-dropdown sd-card sd-mb-3">
<summary class="sd-summary-title sd-card-header">
<span class="sd-summary-text"><i class="fa-solid fa-gear"></i>   Environment setup</span><span class="sd-summary-state-marker sd-summary-chevron-right"><svg version="1.1" width="1.5em" height="1.5em" class="sd-octicon sd-octicon-chevron-right" viewBox="0 0 24 24" aria-hidden="true"><path d="M8.72 18.78a.75.75 0 0 1 0-1.06L14.44 12 8.72 6.28a.751.751 0 0 1 .018-1.042.751.751 0 0 1 1.042-.018l6.25 6.25a.75.75 0 0 1 0 1.06l-6.25 6.25a.75.75 0 0 1-1.06 0Z"></path></svg></span></summary><div class="sd-summary-content sd-card-body docutils">
<div class="sd-tab-set docutils">
<input checked="checked" id="sd-tab-item-0" name="sd-tab-set-0" type="radio">
<label class="sd-tab-label" for="sd-tab-item-0">
Steps</label><div class="sd-tab-content docutils">
<ol class="arabic">
<li><p class="sd-card-text"><strong>Install conda</strong>:</p>
<ul class="simple">
<li><p class="sd-card-text">Before creating the environment, ensure that <a class="reference external" href="https://docs.conda.io/projects/conda/en/latest/user-guide/install/index.html">conda</a> is installed on your system.</p></li>
</ul>
</li>
<li><p class="sd-card-text"><strong>Save the yml content</strong>:</p>
<ul class="simple">
<li><p class="sd-card-text">Copy the content from the yml tab into a file named <code class="docutils literal notranslate"><span class="pre">environment.yml</span></code>.</p></li>
</ul>
</li>
<li><p class="sd-card-text"><strong>Create the environment</strong>:</p>
<ul>
<li><p class="sd-card-text">Open a terminal or command prompt.</p></li>
<li><p class="sd-card-text">Run the following command:</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>conda<span class="w"> </span>env<span class="w"> </span>create<span class="w"> </span>-f<span class="w"> </span>environment.yml
</pre></div>
</div>
</li>
</ul>
</li>
<li><p class="sd-card-text"><strong>Activate the environment</strong>:</p>
<ul>
<li><p class="sd-card-text">After the environment is created, activate it using:</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>conda<span class="w"> </span>activate<span class="w"> </span>&lt;environment_name&gt;
</pre></div>
</div>
</li>
<li><p class="sd-card-text">Replace <code class="docutils literal notranslate"><span class="pre">&lt;environment_name&gt;</span></code> with the name specified in the <code class="docutils literal notranslate"><span class="pre">environment.yml</span></code> file. In the yml file it will look like this:</p>
<div class="highlight-yaml notranslate"><div class="highlight"><pre><span></span><span class="nt">name</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">&lt;environment_name&gt;</span>
</pre></div>
</div>
</li>
</ul>
</li>
<li><p class="sd-card-text"><strong>Verify the installation</strong>:</p>
<ul>
<li><p class="sd-card-text">Check that the environment was created successfully by running:</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>conda<span class="w"> </span>env<span class="w"> </span>list
</pre></div>
</div>
</li>
</ul>
</li>
</ol>
</div>
<input id="sd-tab-item-1" name="sd-tab-set-0" type="radio">
<label class="sd-tab-label" for="sd-tab-item-1">
yml</label><div class="sd-tab-content docutils">
</div>
</div>
</div>
</details><!-- END DROPDOWNS -->
<section id="motivation">
<span id="introduction-raw-data-processing-key-takeaway-1"></span><h2><span class="section-number">3.1. </span>Motivation<a class="headerlink" href="#motivation" title="Link to this heading">#</a></h2>
<p>Raw data processing in single-cell sequencing converts sequencing machine output (so-called lane-demultiplexed <a class="reference internal" href="../glossary.html#term-FASTQ"><span class="xref std std-term">FASTQ</span></a> files) into readily analyzable representations such as a count matrix.
This matrix represents the estimated number of distinct molecules derived from each gene per quantified cell, sometimes categorized by the inferred splicing status of each molecule (<a class="reference internal" href="#raw-proc-fig-overview"><span class="std std-numref">Fig. 3.1</span></a>).</p>
<figure class="align-default" id="raw-proc-fig-overview">
<a class="bg-primary mb-1 reference internal image-reference" href="../_images/overview_raw_data_processing.jpg"><img alt="Chapter Overview" class="bg-primary mb-1" src="../_images/overview_raw_data_processing.jpg" style="width: 800px;" />
</a>
<figcaption>
<p><span class="caption-number">Fig. 3.1 </span><span class="caption-text">An overview of the topics discussed in this chapter. In the plot, “txome” stands for transcriptome.</span><a class="headerlink" href="#raw-proc-fig-overview" title="Link to this image">#</a></p>
</figcaption>
</figure>
<p>The count matrix is the foundation for a wide range of scRNA-seq analyses <span id="id1">[<a class="reference internal" href="#id87" title="Luke Zappia and Fabian J. Theis. Over 1000 tools reveal trends in the single-cell rna-seq analysis landscape. Genome Biology, 22(1):301, Oct 2021. URL: https://doi.org/10.1186/s13059-021-02519-4, doi:10.1186/s13059-021-02519-4.">Zappia and Theis, 2021</a>]</span>, including cell type identification or developmental trajectory inference.
A robust and accurate count matrix is essential for reliable <a class="reference internal" href="../glossary.html#term-Downstream-analysis"><span class="xref std std-term">downstream analyses</span></a>.
Errors at this stage can lead to invalid conclusions and discoveries based on missed insights, or distorted signals in the data.
Despite the straightforward nature of the input (FASTQ files) and the desired output (count matrix), raw data processing presents several technical challenges.</p>
<p>In this section, we focus on key steps of raw data processing:</p>
<ol class="arabic simple">
<li><p>Read alignment/mapping</p></li>
<li><p>Cell barcode (CB) identification and correction</p></li>
<li><p>Estimation of molecule counts through <a class="reference internal" href="../glossary.html#term-Unique-Molecular-Identifier-UMI"><span class="xref std std-term">unique molecular identifiers (UMIs)</span></a></p></li>
</ol>
<p>We also discuss the challenges and trade-offs involved in each step.</p>
<div class="admonition-a-note-on-preceding-steps admonition">
<p class="admonition-title">A note on preceding steps</p>
<p>The starting point for raw data processing is somewhat arbitrary. For this discussion, we treat lane-demultiplexed FASTQ files as the <em>raw</em> input.
However, these files are derived from earlier steps, such as base calling and base quality estimation, which can influence downstream processing.
For example, base-calling errors and index hopping <span id="id2">[<a class="reference internal" href="#id101" title="Rick Farouni, Haig Djambazian, Lorenzo E Ferri, Jiannis Ragoussis, and Hamed S Najafabadi. Model-based analysis of sample index hopping reveals its widespread artifacts in multiplexed single-cell term`RNA`-sequencing. Nature communications, 11(1):1–8, 2020.">Farouni <em>et al.</em>, 2020</a>]</span> can introduce inaccuracies in FASTQ data.
These issues can be mitigated with computational approaches <span id="id3">[<a class="reference internal" href="#id101" title="Rick Farouni, Haig Djambazian, Lorenzo E Ferri, Jiannis Ragoussis, and Hamed S Najafabadi. Model-based analysis of sample index hopping reveals its widespread artifacts in multiplexed single-cell term`RNA`-sequencing. Nature communications, 11(1):1–8, 2020.">Farouni <em>et al.</em>, 2020</a>]</span> or experimental enhancements like <a class="reference external" href="https://www.10xgenomics.com/blog/sequence-with-confidence-understand-index-hopping-and-how-to-resolve-it">dual indexing</a>.</p>
<p>Here, we do not delve into the upstream processes, but consider the FASTQ files, derived from, e.g., BCL files via <a class="reference external" href="https://support.10xgenomics.com/single-cell-gene-expression/software/pipelines/latest/using/bcl2fastq-direct">appropriate tools</a>, as the raw input under consideration.</p>
</div>
</section>
<section id="raw-data-quality-control">
<h2><span class="section-number">3.2. </span>Raw data quality control<a class="headerlink" href="#raw-data-quality-control" title="Link to this heading">#</a></h2>
<p>After obtaining raw FASTQ files, it is important to evaluate the quality of the sequencing reads.
A quick and effective way to perform this is by using quality control (QC) tools like <code class="docutils literal notranslate"><span class="pre">FastQC</span></code>.
<code class="docutils literal notranslate"><span class="pre">FastQC</span></code> generates a detailed report for each FASTQ file, summarizing key metrics such as quality scores, base content, and other statistics that help identify potential issues arising from library preparation or sequencing.</p>
<p>While many modern single-cell data processing tools include some built-in quality checks—such as evaluating the N content of sequences or the fraction of mapped reads - it is still good practice to run an independent QC check.</p>
<p>For readers interested in what a typical <code class="docutils literal notranslate"><span class="pre">FastQC</span></code> report looks like, in the following toggle content, example reports for both <a class="reference external" href="https://www.bioinformatics.babraham.ac.uk/projects/fastqc/good_sequence_short_fastqc.html">high-quality</a> and <a class="reference external" href="https://www.bioinformatics.babraham.ac.uk/projects/fastqc/bad_sequence_fastqc.html">low-quality</a> Illumina data provided by the <code class="docutils literal notranslate"><span class="pre">FastQC</span></code> <a class="reference external" href="https://www.bioinformatics.babraham.ac.uk/projects/fastqc/">manual webpage</a>, along with the tutorials and descriptions from <a class="reference external" href="https://rtsf.natsci.msu.edu/genomics/technical-documents/fastqc-tutorial-and-faq.aspx">the RTSF at MSU</a>, <a class="reference external" href="https://hbctraining.github.io/Intro-to-rnaseq-hpc-salmon/lessons/qc_fastqc_assessment.html">the HBC training program</a>, and <a class="reference external" href="https://sequencing.qcfail.com/software/fastqc/">the QC Fail website</a> are used to demonstrate the modules in the <code class="docutils literal notranslate"><span class="pre">FastQC</span></code> report.
Although these tutorials are not explicitly made for single-cell data, many of the results are still relevant for single-cell data, with a few caveats described below.</p>
<p>In the toggle section, all graphs, except specifically mentioned, are taken from the example reports on the <code class="docutils literal notranslate"><span class="pre">FastQC</span></code> <a class="reference external" href="https://www.bioinformatics.babraham.ac.uk/projects/fastqc/">manual webpage</a>.</p>
<p>It is important to note that many QC metrics in FastQC reports are most meaningful only for biological reads—those derived from gene transcripts.
For single-cell datasets, such as 10x Chromium v2 and v3, this typically corresponds to read 2 (the files containing <code class="docutils literal notranslate"><span class="pre">R2</span></code> in their filename), which contain transcript-derived sequences.
In contrast, technical reads, which contain barcode and UMI sequences, often do not exhibit biologically typical sequence or GC content.
However, certain metrics, like the fraction of <code class="docutils literal notranslate"><span class="pre">N</span></code> base calls, are still relevant for all reads.</p>
<details class="sd-sphinx-override sd-dropdown sd-card sd-mb-3">
<summary class="sd-summary-title sd-card-header">
<span class="sd-summary-text">Example FastQC Reports and Tutorials</span><span class="sd-summary-state-marker sd-summary-chevron-right"><svg version="1.1" width="1.5em" height="1.5em" class="sd-octicon sd-octicon-chevron-right" viewBox="0 0 24 24" aria-hidden="true"><path d="M8.72 18.78a.75.75 0 0 1 0-1.06L14.44 12 8.72 6.28a.751.751 0 0 1 .018-1.042.751.751 0 0 1 1.042-.018l6.25 6.25a.75.75 0 0 1 0 1.06l-6.25 6.25a.75.75 0 0 1-1.06 0Z"></path></svg></span></summary><div class="sd-summary-content sd-card-body docutils">
<p class="sd-card-text"><strong>0. Summary</strong></p>
<p class="sd-card-text">The summary panel on the left side of the HTML report displays the module names along with symbols that provide a quick assessment of the module results.
However, <code class="docutils literal notranslate"><span class="pre">FastQC</span></code> applies uniform thresholds across all sequencing platforms and biological materials.
As a result, warnings (orange exclamation marks) or failures (red crosses) may appear for high-quality data, while questionable data might receive passes (green ticks).
Therefore, each module should be carefully reviewed before drawing conclusions about data quality.</p>
<figure class="align-default" id="raw-proc-fig-fastqc-summary">
<a class="bg-primary mb-1 reference internal image-reference" href="../_images/summary.jpg"><img alt="Summary" class="bg-primary mb-1" src="../_images/summary.jpg" style="width: 300px;" />
</a>
<figcaption>
<p><span class="caption-number">Fig. 3.2 </span><span class="caption-text">The summary panel of a bad example.</span><a class="headerlink" href="#raw-proc-fig-fastqc-summary" title="Link to this image">#</a></p>
</figcaption>
</figure>
<p class="sd-card-text"><strong>1. Basic statistics</strong></p>
<p class="sd-card-text">The basic statistics module provides an overview of key information and statistics for the input FASTQ file, including the filename, total number of sequences, number of poor-quality sequences, sequence length, and the overall GC content (%GC) across all bases in all sequences.
High-quality single-cell data typically have very few poor-quality sequences and exhibit a uniform sequence length.
Additionally, the GC content should align with the expected GC content of the genome or transcriptome of the sequenced species.</p>
<figure class="align-default" id="raw-proc-fig-fastqc-basic-statistics">
<a class="bg-primary mb-1 reference internal image-reference" href="../_images/basic_statistics.jpg"><img alt="Basic Statistics" class="bg-primary mb-1" src="../_images/basic_statistics.jpg" style="width: 800px;" />
</a>
<figcaption>
<p><span class="caption-number">Fig. 3.3 </span><span class="caption-text">A good basic statistics report example.</span><a class="headerlink" href="#raw-proc-fig-fastqc-basic-statistics" title="Link to this image">#</a></p>
</figcaption>
</figure>
<p class="sd-card-text"><strong>2. Per base sequence quality</strong></p>
<p class="sd-card-text">The per-base sequence quality view displays a box-and-whisker plot for each position in the read.
The x-axis represents the positions within the read, while the y-axis shows the quality scores.</p>
<p class="sd-card-text">For high-quality single-cell data, the yellow boxes—representing the interquartile range of quality scores—should fall within the green area (indicating good quality calls).
Similarly, the whiskers, which represent the 10th and 90th percentiles of the distribution, should also remain within the green area.
It is common to observe a gradual drop in quality scores along the length of the read, with some base calls at the last positions falling into the orange area (reasonable quality) due to a decreasing <a class="reference internal" href="../glossary.html#term-signal-to-noise-ratio"><span class="xref std std-term">signal-to-noise ratio</span></a>, a characteristic of sequencing-by-synthesis methods.
However, the boxes should not extend into the red area (poor quality calls).</p>
<p class="sd-card-text">If poor-quality calls are observed, quality trimming may be necessary. <a class="reference external" href="https://hbctraining.github.io/Intro-to-rnaseq-hpc-salmon/lessons/qc_fastqc_assessment.html">A more detailed explanation</a> of sequencing error profiles can be found in the <a class="reference external" href="https://hbctraining.github.io/main/">HBC training program</a>.</p>
<figure class="align-default" id="raw-proc-fig-fastqc-per-read-sequence-quality">
<a class="bg-primary mb-1 reference internal image-reference" href="../_images/per_read_sequence_quality.jpg"><img alt="per read sequence quality" class="bg-primary mb-1" src="../_images/per_read_sequence_quality.jpg" style="width: 800px;" />
</a>
<figcaption>
<p><span class="caption-number">Fig. 3.4 </span><span class="caption-text">A good (left) and a bad (right) per-read sequence quality graph.</span><a class="headerlink" href="#raw-proc-fig-fastqc-per-read-sequence-quality" title="Link to this image">#</a></p>
</figcaption>
</figure>
<p class="sd-card-text"><strong>3. Per tile sequence quality</strong></p>
<p class="sd-card-text">Using an Illumina library, the per-tile sequence quality plot highlights deviations from the average quality for reads across each <a class="reference internal" href="../glossary.html#term-Flowcell"><span class="xref std std-term"> </span></a> <a class="reference external" href="https://www.biostars.org/p/9461090/">tile</a>(miniature imaging areas of the <a class="reference internal" href="../glossary.html#term-Flowcell"><span class="xref std std-term">flowcell</span></a>).
The plot uses a color gradient to represent deviations, where warmer colors indicate larger deviations.
High-quality data typically display a uniform blue color across the plot, indicating consistent quality across all tiles of the flowcell.</p>
<p class="sd-card-text">If warm colors appear in certain areas, it suggests that only part of the flowcell experienced poor quality.
This could result from transient issues during sequencing, such as bubbles passing through the flowcell or smudges and debris within the flowcell lane.
For further investigation, consult resources like <a class="reference external" href="https://sequencing.qcfail.com/articles/position-specific-failures-of-flowcells/">QC Fail</a> and the <a class="reference external" href="https://www.bioinformatics.babraham.ac.uk/projects/fastqc/Help/3%20Analysis%20Modules/12%20Per%20Tile%20Sequence%20Quality.html">common reasons for warnings</a> provided in the <code class="docutils literal notranslate"><span class="pre">FastQC</span></code> manual.</p>
<figure class="align-default" id="raw-proc-fig-fastqc-per-tile-sequence-quality">
<a class="bg-primary mb-1 reference internal image-reference" href="../_images/per_tile_sequence_quality.jpg"><img alt="per tile sequence quality" class="bg-primary mb-1" src="../_images/per_tile_sequence_quality.jpg" style="width: 800px;" />
</a>
<figcaption>
<p><span class="caption-number">Fig. 3.5 </span><span class="caption-text">A good (left) and a bad (right) per tile sequence quality view.</span><a class="headerlink" href="#raw-proc-fig-fastqc-per-tile-sequence-quality" title="Link to this image">#</a></p>
</figcaption>
</figure>
<p class="sd-card-text"><strong>4. Per sequence quality scores</strong></p>
<p class="sd-card-text">The per-sequence quality score plot displays the distribution of average quality scores for each read in the file.
The x-axis represents the average quality scores, while the y-axis shows the frequency of each score.
For high-quality data, the plot should have a single peak near the high-quality end of the scale.
If additional peaks appear, it may indicate a subset of reads with quality issues.</p>
<figure class="align-default" id="raw-proc-fig-fastqc-per-sequence-quality-scores">
<a class="bg-primary mb-1 reference internal image-reference" href="../_images/per_sequence_quality_scores.jpg"><img alt="per sequence quality scores" class="bg-primary mb-1" src="../_images/per_sequence_quality_scores.jpg" style="width: 800px;" />
</a>
<figcaption>
<p><span class="caption-number">Fig. 3.6 </span><span class="caption-text">A good (left) and a bad (right) per sequence quality score plot.</span><a class="headerlink" href="#raw-proc-fig-fastqc-per-sequence-quality-scores" title="Link to this image">#</a></p>
</figcaption>
</figure>
<p class="sd-card-text"><strong>5. Per base sequence content</strong></p>
<p class="sd-card-text">The per-base sequence content plot shows the percentage of each nucleotide (A, T, G, and C) called at each base position across all reads in the file.
For single-cell data, it is common to observe fluctuations at the start of the reads.
This occurs because the initial bases represent the sequence of the priming sites, which are often not perfectly random.
This is a frequent occurrence in RNA-seq libraries, even though <code class="docutils literal notranslate"><span class="pre">FastQC</span></code> may flag it with a warning or failure, as noted on the <a class="reference external" href="https://sequencing.qcfail.com/articles/positional-sequence-bias-in-random-primed-libraries/">QC Fail website</a>.</p>
<figure class="align-default" id="raw-proc-fig-fastqc-per-base-sequence-content">
<a class="bg-primary mb-1 reference internal image-reference" href="../_images/per_base_sequence_content.jpg"><img alt="per base sequence content" class="bg-primary mb-1" src="../_images/per_base_sequence_content.jpg" style="width: 800px;" />
</a>
<figcaption>
<p><span class="caption-number">Fig. 3.7 </span><span class="caption-text">A good (left) and bad (right) per base sequence content plot.</span><a class="headerlink" href="#raw-proc-fig-fastqc-per-base-sequence-content" title="Link to this image">#</a></p>
</figcaption>
</figure>
<p class="sd-card-text"><strong>6. Per sequence GC content</strong></p>
<p class="sd-card-text">The per-sequence GC content plot displays the GC content distribution across all reads (in red) compared to a theoretical distribution (in blue).
The central peak of the observed distribution should align with the overall GC content of the transcriptome.
However, the observed distribution may appear wider or narrower than the theoretical one due to differences between the transcriptome’s GC content and the genome’s expected GC distribution.
Such variations are common and may trigger a warning or failure in <code class="docutils literal notranslate"><span class="pre">FastQC</span></code>, even if the data is acceptable.</p>
<p class="sd-card-text">A complex or irregular distribution in this plot, however, often indicates contamination in the library.
It is also important to note that interpreting GC content in transcriptomics can be challenging.
The expected GC distribution depends not only on the sequence composition of the transcriptome but also on gene expression levels in the sample, which are typically unknown beforehand.
As a result, some deviation from the theoretical distribution is not unusual in RNA-seq data.</p>
<figure class="align-default" id="raw-proc-fig-fastqc-per-sequence-gc-content">
<a class="bg-primary mb-1 reference internal image-reference" href="../_images/per_sequence_gc_content.jpg"><img alt="Per Sequence GC Content" class="bg-primary mb-1" src="../_images/per_sequence_gc_content.jpg" style="width: 800px;" />
</a>
<figcaption>
<p><span class="caption-number">Fig. 3.8 </span><span class="caption-text">A good (left) and a bad (right) per sequence GC content plot.
The plot on the left is from <a class="reference external" href="https://rtsf.natsci.msu.edu/genomics/technical-documents/fastqc-tutorial-and-faq.aspx">the RTSF at MSU</a>.
The plot on the right is taken from <a class="reference external" href="https://hbctraining.github.io/Intro-to-rnaseq-hpc-salmon/lessons/qc_fastqc_assessment.html">the HBC training program</a>.</span><a class="headerlink" href="#raw-proc-fig-fastqc-per-sequence-gc-content" title="Link to this image">#</a></p>
</figcaption>
</figure>
<p class="sd-card-text"><strong>7. Per base N content</strong></p>
<p class="sd-card-text">The per-base N content plot displays the percentage of bases at each position that were called as <code class="docutils literal notranslate"><span class="pre">N</span></code>, indicating that the sequencer lacked sufficient confidence to assign a specific nucleotide.
In a high-quality library, the <code class="docutils literal notranslate"><span class="pre">N</span></code> content should remain consistently at or near zero across the entire length of the reads.
Any noticeable non-zero <code class="docutils literal notranslate"><span class="pre">N</span></code> content may indicate issues with sequencing quality or library preparation.</p>
<figure class="align-default" id="raw-proc-fig-fastqc-per-base-n-content">
<a class="bg-primary mb-1 reference internal image-reference" href="../_images/per_base_n_content.jpg"><img alt="Per Base N Content" class="bg-primary mb-1" src="../_images/per_base_n_content.jpg" style="width: 800px;" />
</a>
<figcaption>
<p><span class="caption-number">Fig. 3.9 </span><span class="caption-text">A good (left) and a bad (right) per base N content plot.</span><a class="headerlink" href="#raw-proc-fig-fastqc-per-base-n-content" title="Link to this image">#</a></p>
</figcaption>
</figure>
<p class="sd-card-text"><strong>8. Sequence length distribution</strong></p>
<p class="sd-card-text">The sequence length distribution graph displays the distribution of read lengths across all sequences in the file.
For most single-cell sequencing chemistries, all reads are expected to have the same length, resulting in a single peak in the graph.
However, if quality trimming was applied before the quality assessment, some variation in read lengths may be observed.
Small differences in read lengths due to trimming are normal and should not be a cause for concern if expected.</p>
<figure class="align-default" id="raw-proc-fig-fastqc-sequence-length-distribution">
<a class="bg-primary mb-1 reference internal image-reference" href="../_images/sequence_length_distribution.jpg"><img alt="Sequence Length Distribution" class="bg-primary mb-1" src="../_images/sequence_length_distribution.jpg" style="width: 800px;" />
</a>
<figcaption>
<p><span class="caption-number">Fig. 3.10 </span><span class="caption-text">A good (left) and a bad (right) sequence length distribution plot.</span><a class="headerlink" href="#raw-proc-fig-fastqc-sequence-length-distribution" title="Link to this image">#</a></p>
</figcaption>
</figure>
<p class="sd-card-text"><strong>9. Sequence duplication levels</strong></p>
<p class="sd-card-text">The sequence duplication level plot illustrates the distribution of duplication levels for read sequences, represented by the blue line, both before and after deduplication.
In single-cell platforms, multiple rounds of <a class="reference internal" href="../glossary.html#term-PCR"><span class="xref std std-term">PCR</span></a> are typically required, and highly expressed genes naturally produce a large number of transcripts.
Additionally, since <code class="docutils literal notranslate"><span class="pre">FastQC</span></code> is not UMI-aware (i.e., it does not account for unique molecular identifiers), it is common for a small subset of sequences to show high duplication levels.</p>
<p class="sd-card-text">While this may trigger a warning or failure in this module, it does not necessarily indicate a quality issue with the data.
However, the majority of sequences should still exhibit low duplication levels, reflecting a diverse and well-prepared library.</p>
<figure class="align-default" id="raw-proc-fig-fastqc-sequence-duplication-levels">
<a class="bg-primary mb-1 reference internal image-reference" href="../_images/sequence_duplication_levels.jpg"><img alt="Sequence Duplication Levels" class="bg-primary mb-1" src="../_images/sequence_duplication_levels.jpg" style="width: 800px;" />
</a>
<figcaption>
<p><span class="caption-number">Fig. 3.11 </span><span class="caption-text">A good (left) and a bad (right) per sequence duplication levels plot.</span><a class="headerlink" href="#raw-proc-fig-fastqc-sequence-duplication-levels" title="Link to this image">#</a></p>
</figcaption>
</figure>
<p class="sd-card-text"><strong>10. Overrepresented sequences</strong></p>
<p class="sd-card-text">The overrepresented sequences module identifies read sequences that constitute more than 0.1% of the total reads.
In single-cell sequencing, some overrepresented sequences may arise from highly expressed genes amplified during PCR.
However, the majority of sequences should not be overrepresented.</p>
<p class="sd-card-text">If the source of an overrepresented sequence is identified (i.e., not listed as “No Hit”), it could indicate potential contamination in the library from the corresponding source.
Such cases warrant further investigation to ensure data quality.</p>
<figure class="align-default" id="raw-proc-fig-fastqc-overrepresented-sequences">
<a class="bg-primary mb-1 reference internal image-reference" href="../_images/overrepresented_sequences.jpg"><img alt="Overrepresented Sequences" class="bg-primary mb-1" src="../_images/overrepresented_sequences.jpg" style="width: 800px;" />
</a>
<figcaption>
<p><span class="caption-number">Fig. 3.12 </span><span class="caption-text">An overrepresented sequence table.</span><a class="headerlink" href="#raw-proc-fig-fastqc-overrepresented-sequences" title="Link to this image">#</a></p>
</figcaption>
</figure>
<p class="sd-card-text"><strong>11. Adapter content</strong></p>
<p class="sd-card-text">The adapter content module displays the cumulative percentage of reads containing <a class="reference internal" href="../glossary.html#term-Adapter-sequences"><span class="xref std std-term">adapter sequences</span></a> at each base position.
High levels of adapter sequences indicate incomplete removal of adapters during library preparation, which can interfere with downstream analyses.
Ideally, no significant adapter content should be present in the data.
If adapter sequences are abundant, additional trimming may be necessary to improve data quality.</p>
<figure class="align-default" id="raw-proc-fig-fastqc-adapter-content">
<a class="bg-primary mb-1 reference internal image-reference" href="../_images/adapter_content.jpg"><img alt="Adapter Content" class="bg-primary mb-1" src="../_images/adapter_content.jpg" style="width: 800px;" />
</a>
<figcaption>
<p><span class="caption-number">Fig. 3.13 </span><span class="caption-text">A good (left) and a bad (right) per sequence quality score plot. The plot on the right is from <a class="reference external" href="https://sequencing.qcfail.com/articles/read-through-adapters-can-appear-at-the-ends-of-sequencing-reads/">the QC Fail website</a>.</span><a class="headerlink" href="#raw-proc-fig-fastqc-adapter-content" title="Link to this image">#</a></p>
</figcaption>
</figure>
</div>
</details><p>Multiple FastQC reports can be combined into a single report using the tool <a class="reference external" href="https://multiqc.info"><code class="docutils literal notranslate"><span class="pre">MultiQC</span></code></a>.</p>
</section>
<section id="alignment-and-mapping">
<span id="raw-proc-aln-map"></span><h2><span class="section-number">3.3. </span>Alignment and mapping<a class="headerlink" href="#alignment-and-mapping" title="Link to this heading">#</a></h2>
<p>Mapping or Alignment is a critical step in single-cell raw data processing.
It involves determining the potential <a class="reference internal" href="../glossary.html#term-Locus"><span class="xref std std-term">loci</span></a> of origin for each sequenced fragment, such as the genomic or transcriptomic locations that closely match the read sequence.
This step is essential for correctly assigning reads to their source regions.</p>
<p>In single-cell sequencing protocols, the raw sequence files typically include:</p>
<ul class="simple">
<li><p>Cell <a class="reference internal" href="../glossary.html#term-Barcode"><span class="xref std std-term">Barcodes</span></a> (CB): Unique identifiers for individual cells.</p></li>
<li><p>Unique Molecular Identifiers (UMIs): Tags that distinguish individual molecules to account for amplification bias.</p></li>
<li><p>Raw <a class="reference internal" href="../glossary.html#term-Complementary-DNA-cDNA"><span class="xref std std-term">cDNA</span></a> Sequences: The actual read sequences generated from the molecules.</p></li>
</ul>
<p>As the first step (<a class="reference internal" href="#raw-proc-fig-overview"><span class="std std-numref">Fig. 3.1</span></a>), accurate mapping or alignment is crucial for reliable downstream analyses.
Errors during this step, such as incorrect mapping of reads to transcripts or genes, can result in inaccurate or misleading count matrices.</p>
<p>While mapping read sequences to reference sequences <em>far</em> predates the development of scRNA-seq, the sheer scale of modern scRNA-seq datasets—often involving hundreds of millions to billions of reads—makes this step particularly computationally intensive.
Many existing RNA-seq aligners are protocol-agnostic and do not inherently account for features specific to scRNA-seq, such as cell barcodes, UMIs, or their positions and lengths.
As a result, additional tools are often required for steps like demultiplexing and UMI resolution <span id="id4">[<a class="reference internal" href="#id102" title="Tom Smith, Andreas Heger, and Ian Sudbery. UMI-tools: modeling sequencing errors in Unique Molecular Identifiers to improve quantification accuracy. Genome research, 27(3):491–499, 2017.">Smith <em>et al.</em>, 2017</a>]</span>.</p>
<p>To address the challenges of aligning and mapping scRNA-seq data, several specialized tools have been developed that handle the additional processing requirements automatically or internally.
These tools include:</p>
<ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">Cell</span> <span class="pre">Ranger</span></code> (commercial software from 10x Genomics) <span id="id5">[<a class="reference internal" href="#id143" title="Grace X. Y. Zheng, Jessica M. Terry, Phillip Belgrader, Paul Ryvkin, Zachary W. Bent, Ryan Wilson, Solongo B. Ziraldo, Tobias D. Wheeler, Geoff P. McDermott, Junjie Zhu, Mark T. Gregory, Joe Shuga, Luz Montesclaros, Jason G. Underwood, Donald A. Masquelier, Stefanie Y. Nishimura, Michael Schnall-Levin, Paul W. Wyatt, Christopher M. Hindson, Rajiv Bharadwaj, Alexander Wong, Kevin D. Ness, Lan W. Beppu, H. Joachim Deeg, Christopher McFarland, Keith R. Loeb, William J. Valente, Nolan G. Ericson, Emily A. Stevens, Jerald P. Radich, Tarjei S. Mikkelsen, Benjamin J. Hindson, and Jason H. Bielas. Massively parallel digital transcriptional profiling of single cells. Nature Communications, 8(1):14049, Jan 2017. URL: https://doi.org/10.1038/ncomms14049, doi:10.1038/ncomms14049.">Zheng <em>et al.</em>, 2017</a>]</span></p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">zUMIs</span></code> <span id="id6">[<a class="reference internal" href="#id113" title="Swati Parekh, Christoph Ziegenhain, Beate Vieth, Wolfgang Enard, and Ines Hellmann. zUMIs - a fast and flexible pipeline to process term`RNA` sequencing data with UMIs. GigaScience, May 2018. URL: https://doi.org/10.1093/gigascience/giy059, doi:10.1093/gigascience/giy059.">Parekh <em>et al.</em>, 2018</a>]</span></p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">alevin</span></code> <span id="id7">[<a class="reference internal" href="#id103" title="Avi Srivastava, Laraib Malik, Tom Smith, Ian Sudbery, and Rob Patro. Alevin efficiently estimates accurate gene abundances from dscRNA-seq data. Genome biology, 20(1):1–16, 2019.">Srivastava <em>et al.</em>, 2019</a>]</span></p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">RainDrop</span></code> <span id="id8">[<a class="reference internal" href="#id90" title="Stefan Niebler, André Müller, Thomas Hankeln, and Bertil Schmidt. RainDrop: Rapid activation matrix computation for droplet-based single-cell term`RNA`-seq reads. BMC bioinformatics, 21(1):1–14, 2020.">Niebler <em>et al.</em>, 2020</a>]</span></p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">kallisto|bustools</span></code> <span id="id9">[<a class="reference internal" href="#id108" title="Páll Melsted, A. Sina Booeshaghi, Lauren Liu, Fan Gao, Lambda Lu, Kyung Hoi Min, Eduardo da Veiga Beltrame, Kristján Eldjárn Hjörleifsson, Jase Gehring, and Lior Pachter. Modular, efficient and constant-memory single-cell rna-seq preprocessing. Nature Biotechnology, 39(7):813–818, April 2021. URL: https://doi.org/10.1038/s41587-021-00870-2, doi:10.1038/s41587-021-00870-2.">Melsted <em>et al.</em>, 2021</a>]</span></p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">STARsolo</span></code> <span id="id10">[<a class="reference internal" href="#id107" title="Benjamin Kaminow, Dinar Yunusov, and Alexander Dobin. STARsolo: accurate, fast and versatile mapping/quantification of single-cell and single-nucleus term`RNA`-seq data. bioRxiv, 2021.">Kaminow <em>et al.</em>, 2021</a>]</span></p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">alevin-fry</span></code> <span id="id11">[<a class="reference internal" href="#id110" title="Dongze He, Mohsen Zakeri, Hirak Sarkar, Charlotte Soneson, Avi Srivastava, and Rob Patro. Alevin-fry unlocks rapid, accurate and memory-frugal quantification of single-cell RNA-seq data. Nature Methods, 19(3):316–322, 2022.">He <em>et al.</em>, 2022</a>]</span></p></li>
</ul>
<p>These tools provide specialized capabilities for aligning scRNA-seq reads, parsing technical read content (e.g., cell barcodes and UMIs), demultiplexing, and UMI resolution.
Although they offer simplified user interfaces, their internal methodologies differ significantly.
Some tools generate traditional intermediate files, such as <a class="reference internal" href="../glossary.html#term-BAM"><span class="xref std std-term">BAM</span></a> files, which are processed further, while others operate entirely in memory or use compact intermediate representations to minimize input/output operations and reduce computational overhead.</p>
<p>While these tools vary in their specific algorithms, data structures, and trade-offs in time and space complexity, their approaches can generally be categorized along two axes:</p>
<ol class="arabic simple">
<li><p><strong>The type of mapping they perform</strong>, and</p></li>
<li><p><strong>The type of reference sequence against which they map reads</strong>.</p></li>
</ol>
<section id="types-of-mapping">
<span id="raw-proc-types-of-mapping"></span><h3><span class="section-number">3.3.1. </span>Types of mapping<a class="headerlink" href="#types-of-mapping" title="Link to this heading">#</a></h3>
<p>We focus on three main types of mapping algorithms commonly used for mapping sc/snRNA-seq data: spliced alignment, contiguous alignment, and variations of lightweight mapping.</p>
<p>First, we distinguish between alignment-based approaches and lightweight mapping-based approaches (<a class="reference internal" href="#raw-proc-fig-alignment-mapping"><span class="std std-numref">Fig. 3.14</span></a>).
Alignment-based methods use various heuristics to identify potential loci from which reads may originate and then score the best nucleotide-level alignment between the read and reference, typically using dynamic programming algorithms.</p>
<p><a class="reference external" href="https://en.wikipedia.org/wiki/Needleman%E2%80%93Wunsch_algorithm">global alignment</a> aligns the entirety of the query and reference sequences, while <a class="reference external" href="https://en.wikipedia.org/wiki/Smith%E2%80%93Waterman_algorithm">local alignment</a> focuses on aligning subsequences.
Short-read alignment often employs a semi-global approach, also known as “fitting” alignment, where most of the query aligns to a substring of the reference.
Additionally, “soft-clipping” may be used to reduce penalties for mismatches, insertions, or deletions at the start or end of the read, achieved through <a class="reference external" href="https://github.com/smarco/WFA2-lib#-33-alignment-span">“extension” alignment</a>.
While these variations modify the rules of the dynamic programming recurrence and traceback, they do not fundamentally alter its overall complexity.</p>
<p>Several sophisticated modifications and heuristics have been developed to enhance the practical efficiency of aligning genomic sequencing reads.
For example, <code class="docutils literal notranslate"><span class="pre">banded</span> <span class="pre">alignment</span></code> <span id="id12">[<a class="reference internal" href="#id88" title="Kun-Mao Chao, William R. Pearson, and Webb Miller. Aligning two sequences within a specified diagonal band. Bioinformatics, 8(5):481–487, 1992. URL: https://doi.org/10.1093/bioinformatics/8.5.481, doi:10.1093/bioinformatics/8.5.481.">Chao <em>et al.</em>, 1992</a>]</span> is a popular heuristic used by many tools to avoid computing large portions of the dynamic programming table when alignment scores below a threshold are not of interest.
Other heuristics, like X-drop <span id="id13">[<a class="reference internal" href="#id100" title="Zheng Zhang, Scott Schwartz, Lukas Wagner, and Webb Miller. A greedy algorithm for aligning term`DNA` sequences. Journal of Computational Biology, 7(1-2):203–214, February 2000. URL: https://doi.org/10.1089/10665270050081478, doi:10.1089/10665270050081478.">Zhang <em>et al.</em>, 2000</a>]</span> and Z-drop <span id="id14">[<a class="reference internal" href="#id89" title="Heng Li. Minimap2: pairwise alignment for nucleotide sequences. Bioinformatics, 34(18):3094–3100, May 2018. URL: https://doi.org/10.1093/bioinformatics/bty191, doi:10.1093/bioinformatics/bty191.">Li, 2018</a>]</span>, efficiently prune unpromising alignments early in the process.
Recent advances, such as wavefront alignment <span id="id15">[<a class="reference internal" href="#id97" title="Santiago Marco-Sola, Juan Carlos Moure, Miquel Moreto, and Antonio Espinosa. Fast gap-affine pairwise alignment using the wavefront algorithm. Bioinformatics, September 2020. URL: https://doi.org/10.1093/bioinformatics/btaa777, doi:10.1093/bioinformatics/btaa777.">Marco-Sola <em>et al.</em>, 2020</a>]</span>, marco2022optimal, enable the determination of optimal alignments in significantly reduced time and space, particularly when high-scoring alignments are present.
Additionally, much work has focused on optimizing data layout and computation to leverage instruction-level parallelism <span id="id16">[<a class="reference internal" href="#id92" title="Michael Farrar. Striped Smith–Waterman speeds database searches six times over other SIMD implementations. Bioinformatics, 23(2):156–161, 2007.">Farrar, 2007</a>, <a class="reference internal" href="#id93" title="Torbjørn Rognes and Erling Seeberg. Six-fold speed-up of Smith–Waterman sequence database searches using parallel processing on common microprocessors. Bioinformatics, 16(8):699–706, 2000.">Rognes and Seeberg, 2000</a>, <a class="reference internal" href="#id91" title="Andrzej Wozniak. Using video-oriented instructions to speed up sequence comparison. Bioinformatics, 13(2):145–150, 1997.">Wozniak, 1997</a>]</span>, and expressing dynamic programming recurrences in ways that facilitate data parallelism and vectorization, such as through difference encoding <span id="id17">Suzuki and Kasahara [<a class="reference internal" href="#id98" title="Hajime Suzuki and Masahiro Kasahara. Introducing difference recurrence relations for faster semi-global alignment of long sequences. BMC Bioinformatics, February 2018. URL: https://doi.org/10.1186/s12859-018-2014-8, doi:10.1186/s12859-018-2014-8.">2018</a>]</span>.
Most widely-used alignment tools incorporate these highly optimized, vectorized implementations.</p>
<p>In addition to the alignment score, the backtrace of the actual alignment that produces this score is often encoded as a <code class="docutils literal notranslate"><span class="pre">CIGAR</span></code> string (short for “Concise Idiosyncratic Gapped Alignment Report”).
This alphanumeric representation is typically stored in the SAM or BAM file output.
For example, the <code class="docutils literal notranslate"><span class="pre">CIGAR</span></code> string <code class="docutils literal notranslate"><span class="pre">3M2D4M</span></code> indicates that the alignment has three matches or mismatches, followed by a deletion of length two (representing bases present in the reference but not the read), and then four more matches or mismatches.
Extended <code class="docutils literal notranslate"><span class="pre">CIGAR</span></code> strings can provide additional details, such as distinguishing between matches, mismatches, or insertions.
For instance, <code class="docutils literal notranslate"><span class="pre">3=2D2=2X</span></code> encodes the same alignment as the previous example but specifies that the three bases before the deletion are matches, followed by two matched bases and two mismatched bases after the deletion.
A detailed description of the <code class="docutils literal notranslate"><span class="pre">CIGAR</span></code> string format can be found in <a class="reference external" href="https://samtools.github.io/hts-specs/SAMv1.pdf">the SAMtools manual</a> or <a class="reference external" href="https://genome.sph.umich.edu/wiki/SAM#What_is_a_CIGAR.3F">the SAM wiki page of UMICH</a>.</p>
<p>Alignment-based approaches, though computationally expensive, provide a quality score for each potential mapping of a read.
This score allows them to distinguish between high-quality alignments and low-complexity or “spurious” matches between the read and reference.
These approaches include traditional “full-alignment” methods, such as those implemented in tools like <code class="docutils literal notranslate"><span class="pre">STAR</span></code> <span id="id18">[<a class="reference internal" href="#id141" title="Alexander Dobin, Carrie A Davis, Felix Schlesinger, Jorg Drenkow, Chris Zaleski, Sonali Jha, Philippe Batut, Mark Chaisson, and Thomas R Gingeras. Star: ultrafast universal rna-seq aligner. Bioinformatics, 29(1):15–21, 2013.">Dobin <em>et al.</em>, 2013</a>]</span> and <code class="docutils literal notranslate"><span class="pre">STARsolo</span></code> <span id="id19">[<a class="reference internal" href="#id107" title="Benjamin Kaminow, Dinar Yunusov, and Alexander Dobin. STARsolo: accurate, fast and versatile mapping/quantification of single-cell and single-nucleus term`RNA`-seq data. bioRxiv, 2021.">Kaminow <em>et al.</em>, 2021</a>]</span>, as well as <em>selective-alignment</em> methods, like those in <code class="docutils literal notranslate"><span class="pre">salmon</span></code> <span id="id20">[<a class="reference internal" href="#id131" title="Avi Srivastava, Laraib Malik, Hirak Sarkar, Mohsen Zakeri, Fatemeh Almodaresi, Charlotte Soneson, Michael I Love, Carl Kingsford, and Rob Patro. Alignment and mapping methodology influence transcript abundance estimation. Genome Biology, 21(1):1–29, 2020.">Srivastava <em>et al.</em>, 2020</a>]</span> and <code class="docutils literal notranslate"><span class="pre">alevin</span></code> <span id="id21">[<a class="reference internal" href="#id103" title="Avi Srivastava, Laraib Malik, Tom Smith, Ian Sudbery, and Rob Patro. Alevin efficiently estimates accurate gene abundances from dscRNA-seq data. Genome biology, 20(1):1–16, 2019.">Srivastava <em>et al.</em>, 2019</a>]</span>, which score mappings but skip the computation of the optimal alignment’s backtrace.</p>
<figure class="align-default" id="raw-proc-fig-alignment-mapping">
<a class="bg-primary mb-1 reference internal image-reference" href="../_images/alignment_vs_mapping.png"><img alt="Alignment vs Mapping" class="bg-primary mb-1" src="../_images/alignment_vs_mapping.png" style="width: 800px;" />
</a>
<figcaption>
<p><span class="caption-number">Fig. 3.14 </span><span class="caption-text">An abstract overview of the alignment-based method and lightweight mapping-based method.</span><a class="headerlink" href="#raw-proc-fig-alignment-mapping" title="Link to this image">#</a></p>
</figcaption>
</figure>
<p>Alignment-based approaches can be categorized into spliced-alignment and contiguous-alignment methods.</p>
<details class="sd-sphinx-override sd-dropdown sd-card sd-mb-3">
<summary class="sd-summary-title sd-card-header">
<span class="sd-summary-text">Spliced-alignment methods</span><span class="sd-summary-state-marker sd-summary-chevron-right"><svg version="1.1" width="1.5em" height="1.5em" class="sd-octicon sd-octicon-chevron-right" viewBox="0 0 24 24" aria-hidden="true"><path d="M8.72 18.78a.75.75 0 0 1 0-1.06L14.44 12 8.72 6.28a.751.751 0 0 1 .018-1.042.751.751 0 0 1 1.042-.018l6.25 6.25a.75.75 0 0 1 0 1.06l-6.25 6.25a.75.75 0 0 1-1.06 0Z"></path></svg></span></summary><div class="sd-summary-content sd-card-body docutils">
<p class="sd-card-text">Spliced-alignment methods allow a sequence read to align across multiple distinct segments of a reference, allowing potentially large gaps between aligned regions.
These approaches are particularly useful for aligning RNA-seq reads to the genome, where reads may span <a class="reference internal" href="../glossary.html#term-Splice-Junctions"><span class="xref std std-term">splice junctions</span></a>.
In such cases, a contiguous sequence in the read may be separated by intron and exon subsequence in the reference, potentially spanning kilobases of sequence.
Spliced alignment is especially challenging when only a small portion of a read overlaps a splice junction, as limited sequence information is available to accurately place the overhanging segment.</p>
</div>
</details><details class="sd-sphinx-override sd-dropdown sd-card sd-mb-3">
<summary class="sd-summary-title sd-card-header">
<span class="sd-summary-text">Contiguous-alignment methods</span><span class="sd-summary-state-marker sd-summary-chevron-right"><svg version="1.1" width="1.5em" height="1.5em" class="sd-octicon sd-octicon-chevron-right" viewBox="0 0 24 24" aria-hidden="true"><path d="M8.72 18.78a.75.75 0 0 1 0-1.06L14.44 12 8.72 6.28a.751.751 0 0 1 .018-1.042.751.751 0 0 1 1.042-.018l6.25 6.25a.75.75 0 0 1 0 1.06l-6.25 6.25a.75.75 0 0 1-1.06 0Z"></path></svg></span></summary><div class="sd-summary-content sd-card-body docutils">
<p class="sd-card-text">Contiguous-alignment methods require a continuous substring of the reference to align well with the read.
While small insertions and deletions may be tolerated, large gaps—such as those in spliced alignments—are generally not allowed.</p>
</div>
</details><p>Alignment-based methods, such as spliced and contiguous alignment, can be distinguished from <strong>lightweight-mapping methods</strong>, which include approaches like <strong>pseudoalignment</strong> <span id="id22">[<a class="reference internal" href="#id127" title="Nicolas L Bray, Harold Pimentel, Páll Melsted, and Lior Pachter. Near-optimal probabilistic term`RNA`-seq quantification. Nature biotechnology, 34(5):525–527, 2016.">Bray <em>et al.</em>, 2016</a>]</span>, <strong>quasi-mapping</strong> <span id="id23">[<a class="reference internal" href="#id126" title="Avi Srivastava, Hirak Sarkar, Nitish Gupta, and Rob Patro. Rapmap: a rapid, sensitive and accurate tool for mapping term`rna`-seq reads to transcriptomes. Bioinformatics, 32(12):i192–i200, 2016.">Srivastava <em>et al.</em>, 2016</a>]</span>, and <strong>pseudoalignment with structural constraints</strong> <span id="id24">[<a class="reference internal" href="#id110" title="Dongze He, Mohsen Zakeri, Hirak Sarkar, Charlotte Soneson, Avi Srivastava, and Rob Patro. Alevin-fry unlocks rapid, accurate and memory-frugal quantification of single-cell RNA-seq data. Nature Methods, 19(3):316–322, 2022.">He <em>et al.</em>, 2022</a>]</span>.</p>
<p>Lightweight-mapping methods achieve significantly higher speed.
However, they do not provide easily-interpretable score-based assessments to determine the quality of a match, making it more difficult to assess alignment confidence.</p>
</section>
<section id="mapping-against-different-reference-sequences">
<span id="raw-proc-mapping-references"></span><h3><span class="section-number">3.3.2. </span>Mapping against different reference sequences<a class="headerlink" href="#mapping-against-different-reference-sequences" title="Link to this heading">#</a></h3>
<p>In addition to selecting a mapping algorithm, choices can <em>also</em> be made regarding the reference sequence against which the reads are mapped.
There are three main categories of reference sequences:</p>
<ul class="simple">
<li><p>Full reference genome (typically annotated)</p></li>
<li><p>Annotated transcriptome</p></li>
<li><p>Augmented transcriptome</p></li>
</ul>
<p>Currently, not all combinations of mapping algorithms and reference sequences are possible.
For instance, lightweight-mapping algorithms do not yet support spliced mapping of reads against a reference genome.</p>
<section id="mapping-to-the-full-genome">
<span id="raw-proc-genome-mapping"></span><h4><span class="section-number">3.3.2.1. </span>Mapping to the full genome<a class="headerlink" href="#mapping-to-the-full-genome" title="Link to this heading">#</a></h4>
<p>The first type of reference used for mapping is the <strong>entire genome</strong> of the target organism, typically with annotated transcripts considered during mapping.
Tools such as <code class="docutils literal notranslate"><span class="pre">zUMIs</span></code> <span id="id25">[<a class="reference internal" href="#id113" title="Swati Parekh, Christoph Ziegenhain, Beate Vieth, Wolfgang Enard, and Ines Hellmann. zUMIs - a fast and flexible pipeline to process term`RNA` sequencing data with UMIs. GigaScience, May 2018. URL: https://doi.org/10.1093/gigascience/giy059, doi:10.1093/gigascience/giy059.">Parekh <em>et al.</em>, 2018</a>]</span>, <code class="docutils literal notranslate"><span class="pre">Cell</span> <span class="pre">Ranger</span></code> <span id="id26">[<a class="reference internal" href="#id143" title="Grace X. Y. Zheng, Jessica M. Terry, Phillip Belgrader, Paul Ryvkin, Zachary W. Bent, Ryan Wilson, Solongo B. Ziraldo, Tobias D. Wheeler, Geoff P. McDermott, Junjie Zhu, Mark T. Gregory, Joe Shuga, Luz Montesclaros, Jason G. Underwood, Donald A. Masquelier, Stefanie Y. Nishimura, Michael Schnall-Levin, Paul W. Wyatt, Christopher M. Hindson, Rajiv Bharadwaj, Alexander Wong, Kevin D. Ness, Lan W. Beppu, H. Joachim Deeg, Christopher McFarland, Keith R. Loeb, William J. Valente, Nolan G. Ericson, Emily A. Stevens, Jerald P. Radich, Tarjei S. Mikkelsen, Benjamin J. Hindson, and Jason H. Bielas. Massively parallel digital transcriptional profiling of single cells. Nature Communications, 8(1):14049, Jan 2017. URL: https://doi.org/10.1038/ncomms14049, doi:10.1038/ncomms14049.">Zheng <em>et al.</em>, 2017</a>]</span>, and <code class="docutils literal notranslate"><span class="pre">STARsolo</span></code> <span id="id27">[<a class="reference internal" href="#id107" title="Benjamin Kaminow, Dinar Yunusov, and Alexander Dobin. STARsolo: accurate, fast and versatile mapping/quantification of single-cell and single-nucleus term`RNA`-seq data. bioRxiv, 2021.">Kaminow <em>et al.</em>, 2021</a>]</span> follow this approach.
Since many reads originate from <strong>spliced transcripts</strong>, this method requires a <strong>splice-aware alignment algorithm</strong> capable of splitting alignments across one or more splice junctions.</p>
<p>A key advantage of this approach is that it accounts for reads arising from any location in the genome, not just those from annotated transcripts.
Additionally, because a <strong>genome-wide index</strong> is constructed, there is minimal additional cost in reporting not only reads that map to known spliced transcripts but also those that overlap introns or align within non-coding regions, making this method equally effective for <strong>single-cell</strong> and <strong>single-nucleus</strong> data.
Another benefit is that even reads mapping outside annotated transcripts, exons, or introns can still be accounted for, enabling <strong><em>post hoc</em> augmentation</strong> of the quantified loci.</p>
</section>
<section id="mapping-to-the-spliced-transcriptome">
<span id="raw-proc-txome-mapping"></span><h4><span class="section-number">3.3.2.2. </span>Mapping to the spliced transcriptome<a class="headerlink" href="#mapping-to-the-spliced-transcriptome" title="Link to this heading">#</a></h4>
<p>To reduce the computational overhead of spliced alignment to a genome, a widely adopted alternative is to use only the annotated transcript sequences as the reference.
Since most single-cell experiments are conducted on model organisms like mouse or human, which have well-annotated transcriptomes, transcriptome-based quantification can achieve similar read coverage to genome-based methods.</p>
<p>Compared to the genome, transcriptome sequences are much smaller, significantly reducing the computational resources needed for mapping.
Additionally, because splicing patterns are already represented in transcript sequences, this approach eliminates the need for complex spliced alignment.
Instead, one can simply search for contiguous alignments or mappings for the read.
Alternatively, reads can be mapped using contiguous alignments, making both alignment-based and lightweight-mapping techniques suitable for transcriptome references.</p>
<p>While these approaches significantly reduce the memory and time required for alignment and mapping, they fail to capture reads that arise from outside the spliced transcriptome.
As a result, they are not suitable for processing single-nucleus data.
Even in single-cell experiments, reads arising from outside of the spliced transcriptome can constitute a substantial fraction of all data, and there is growing evidence that such reads should be incorporated into subsequent analysis <span id="id28">[<a class="reference internal" href="#id134" title="Allan-Hermann Pool, Helen Poldsam, Sisi Chen, Matt Thomson, and Yuki Oka. Enhanced recovery of single-cell term`RNA`-sequencing reads for missing gene expression data. bioRxiv, 2022. URL: https://www.biorxiv.org/content/early/2022/04/27/2022.04.26.489449, arXiv:https://www.biorxiv.org/content/early/2022/04/27/2022.04.26.489449.full.pdf, doi:10.1101/2022.04.26.489449.">Pool <em>et al.</em>, 2022</a>, <a class="reference internal" href="#id104" title="10x Genomics. Technical note - interpreting intronic and antisense reads in 10x genomics single cell gene expression data. https://www.10xgenomics.com/support/single-cell-gene-expression/documentation/steps/sequencing/interpreting-intronic-and-antisense-reads-in-10-x-genomics-single-cell-gene-expression-data, August 2021.">10x Genomics, 2021</a>]</span>.
Even in single-cell experiments, a substantial fraction of reads may arise from regions outside the spliced transcriptome, and increasing evidence suggests that incorporating these reads into downstream analyses can be beneficial <span id="id29">[<a class="reference internal" href="#id134" title="Allan-Hermann Pool, Helen Poldsam, Sisi Chen, Matt Thomson, and Yuki Oka. Enhanced recovery of single-cell term`RNA`-sequencing reads for missing gene expression data. bioRxiv, 2022. URL: https://www.biorxiv.org/content/early/2022/04/27/2022.04.26.489449, arXiv:https://www.biorxiv.org/content/early/2022/04/27/2022.04.26.489449.full.pdf, doi:10.1101/2022.04.26.489449.">Pool <em>et al.</em>, 2022</a>, <a class="reference internal" href="#id104" title="10x Genomics. Technical note - interpreting intronic and antisense reads in 10x genomics single cell gene expression data. https://www.10xgenomics.com/support/single-cell-gene-expression/documentation/steps/sequencing/interpreting-intronic-and-antisense-reads-in-10-x-genomics-single-cell-gene-expression-data, August 2021.">10x Genomics, 2021</a>]</span>.
Additionally, when paired with lightweight-mapping methods, short sequences shared between the spliced transcriptome and the actual genomic regions that generated a read can lead to spurious mappings.
This, in turn, may result in misleading and even biologically implausible gene expression estimates <span id="id30">[<a class="reference internal" href="#id133" title="Ralf Schulze Brüning, Lukas Tombor, Marcel H Schulz, Stefanie Dimmeler, and David John. Comparative analysis of common alignment tools for single-cell term`RNA` sequencing. GigaScience, 2022.">Brüning <em>et al.</em>, 2022</a>, <a class="reference internal" href="#id110" title="Dongze He, Mohsen Zakeri, Hirak Sarkar, Charlotte Soneson, Avi Srivastava, and Rob Patro. Alevin-fry unlocks rapid, accurate and memory-frugal quantification of single-cell RNA-seq data. Nature Methods, 19(3):316–322, 2022.">He <em>et al.</em>, 2022</a>, <a class="reference internal" href="#id107" title="Benjamin Kaminow, Dinar Yunusov, and Alexander Dobin. STARsolo: accurate, fast and versatile mapping/quantification of single-cell and single-nucleus term`RNA`-seq data. bioRxiv, 2021.">Kaminow <em>et al.</em>, 2021</a>]</span>.</p>
</section>
<section id="mapping-to-an-augmented-transcriptome">
<span id="raw-proc-aug-txome-mapping"></span><h4><span class="section-number">3.3.2.3. </span>Mapping to an augmented transcriptome<a class="headerlink" href="#mapping-to-an-augmented-transcriptome" title="Link to this heading">#</a></h4>
<p>To account for reads originating outside spliced transcripts, the spliced transcript sequences can be augmented with additional reference sequences, such as full-length unspliced transcripts or excised intronic sequences.
This enables better, faster, and more memory-efficient mapping compared to full-genome alignment, while still capturing many reads that would otherwise be missed.
More reads can be confidently assigned compared to using only the spliced transcriptome, and when combined with lightweight mapping approaches, spurious mappings can be significantly reduced <span id="id31">[<a class="reference internal" href="#id110" title="Dongze He, Mohsen Zakeri, Hirak Sarkar, Charlotte Soneson, Avi Srivastava, and Rob Patro. Alevin-fry unlocks rapid, accurate and memory-frugal quantification of single-cell RNA-seq data. Nature Methods, 19(3):316–322, 2022.">He <em>et al.</em>, 2022</a>]</span>.
Augmented transcriptomes are widely used in methods that do not map to the full genome, particularly for single-nucleus data processing and <a class="reference internal" href="../glossary.html#term-RNA-velocity"><span class="xref std std-term">RNA velocity</span></a> analysis <span id="id32">[<a class="reference internal" href="#id109" title="Charlotte Soneson, Avi Srivastava, Rob Patro, and Michael B Stadler. Preprocessing choices affect RNA velocity results for droplet scRNA-seq data. PLoS computational biology, 17(1):e1008585, 2021.">Soneson <em>et al.</em>, 2021</a>]</span> (see <a class="reference internal" href="../trajectories/rna_velocity.html"><span class="doc">RNA velocity</span></a>).
These augmented references can be constructed for all common methods that do not rely on spliced alignment to the full genome <span id="id33">[<a class="reference internal" href="#id110" title="Dongze He, Mohsen Zakeri, Hirak Sarkar, Charlotte Soneson, Avi Srivastava, and Rob Patro. Alevin-fry unlocks rapid, accurate and memory-frugal quantification of single-cell RNA-seq data. Nature Methods, 19(3):316–322, 2022.">He <em>et al.</em>, 2022</a>, <a class="reference internal" href="#id108" title="Páll Melsted, A. Sina Booeshaghi, Lauren Liu, Fan Gao, Lambda Lu, Kyung Hoi Min, Eduardo da Veiga Beltrame, Kristján Eldjárn Hjörleifsson, Jase Gehring, and Lior Pachter. Modular, efficient and constant-memory single-cell rna-seq preprocessing. Nature Biotechnology, 39(7):813–818, April 2021. URL: https://doi.org/10.1038/s41587-021-00870-2, doi:10.1038/s41587-021-00870-2.">Melsted <em>et al.</em>, 2021</a>, <a class="reference internal" href="#id103" title="Avi Srivastava, Laraib Malik, Tom Smith, Ian Sudbery, and Rob Patro. Alevin efficiently estimates accurate gene abundances from dscRNA-seq data. Genome biology, 20(1):1–16, 2019.">Srivastava <em>et al.</em>, 2019</a>]</span>.</p>
</section>
</section>
</section>
<section id="cell-barcode-correction">
<span id="raw-proc-cb-correction"></span><h2><span class="section-number">3.4. </span>Cell barcode correction<a class="headerlink" href="#cell-barcode-correction" title="Link to this heading">#</a></h2>
<p>Droplet-based single-cell segregation systems, such as those provided by 10x Genomics, have become an important tool for studying the cause and consequences of cellular heterogeneity.
In this segregation system, the RNA material of each captured cell is extracted within a water-based droplet encapsulation along with a <strong>barcoded bead</strong>.
These beads tag the RNA content of individual cells with unique oligonucleotides, called cell barcodes (CBs), that are later sequenced along with the fragments of the cDNAs that are reversely transcribed from the RNA content.
The beads contain high-diversity DNA barcodes, allowing for parallel barcoding of a cell’s molecular content and <em>in silico</em> demultiplexing of sequencing reads into individual cellular bins.</p>
<div class="admonition-a-note-on-alignment-orientation admonition">
<p class="admonition-title">A note on alignment orientation</p>
<p>Depending on the sample chemistry and user-defined processing options, not all sequenced fragments that align to the reference are necessarily considered for quantification and barcode correction.
One commonly-applied criterion for filtering is alignment orientation.
Specifically, certain chemistries specify protocols such that the aligned reads should only derive from (i.e. map back to) the underlying transcripts in a specific orientation.
For example, in 10x Genomics 3’ Chromium chemistries, we expect the biological read to align to the underlying transcript’s forward strand, though anti-sense reads do exist <span id="id34">[<a class="reference internal" href="#id104" title="10x Genomics. Technical note - interpreting intronic and antisense reads in 10x genomics single cell gene expression data. https://www.10xgenomics.com/support/single-cell-gene-expression/documentation/steps/sequencing/interpreting-intronic-and-antisense-reads-in-10-x-genomics-single-cell-gene-expression-data, August 2021.">10x Genomics, 2021</a>]</span>.
As a result, reads mapped in the reverse-complement orientation to the reference sequences may be ignored or filtered out based on user-defined settings.
If a chemistry follows such a so-called “stranded” protocol, this should be documented.</p>
</div>
<section id="type-of-errors-in-barcoding">
<h3><span class="section-number">3.4.1. </span>Type of errors in barcoding<a class="headerlink" href="#type-of-errors-in-barcoding" title="Link to this heading">#</a></h3>
<p>The tag, sequence, and demultiplexing method used for single-cell profiling is generally effective.
However, in droplet-based libraries, the number of observed cell barcodes (CBs) can differ significantly—often by several fold—from the number of originally encapsulated cells.
This discrepancy arises from several key sources of error:</p>
<ul class="simple">
<li><p>Doublets/multiplets: A single barcode may be associated with multiple cells, leading to an undercounting of cells.</p></li>
<li><p>Empty droplets: Some droplets contain no encapsulated cells, and ambient RNA can become tagged with a barcode and sequenced, resulting in overcounting of cells.</p></li>
<li><p>Sequence errors: Errors introduced during PCR amplification or sequencing can distort barcode counts, contributing to both under- and over-counting.</p></li>
</ul>
<p>To address these issues, computational tools for demultiplexing RNA-seq reads into cell-specific bins use various diagnostic indicators to filter out artefactual or low-quality data.
Numerous methods exist for removing ambient RNA contamination <span id="id35">[<a class="reference internal" href="#id115" title="Aaron TL Lun, Samantha Riesenfeld, Tallulah Andrews, Tomas Gomes, John C Marioni, and others. EmptyDrops: distinguishing cells from empty droplets in droplet-based single-cell term`RNA` sequencing data. Genome biology, 20(1):1–9, 2019.">Lun <em>et al.</em>, 2019</a>, <a class="reference internal" href="#id118" title="Walter Muskovic and Joseph E Powell. DropletQC: improved identification of empty droplets and damaged cells in single-cell term`RNA`-seq data. Genome Biology, 22(1):1–9, 2021.">Muskovic and Powell, 2021</a>, <a class="reference internal" href="#id120" title="Matthew D Young and Sam Behjati. SoupX removes ambient term`RNA` contamination from droplet-based single-cell term`RNA` sequencing data. GigaScience, December 2020. URL: https://doi.org/10.1093/gigascience/giaa151, doi:10.1093/gigascience/giaa151.">Young and Behjati, 2020</a>]</span>, detecting doublets <span id="id36">[<a class="reference internal" href="#id124" title="Abha S Bais and Dennis Kostka. Scds: computational annotation of doublets in single-cell term`RNA` sequencing data. Bioinformatics, 36(4):1150–1158, September 2019. URL: https://doi.org/10.1093/bioinformatics/btz698, doi:10.1093/bioinformatics/btz698.">Bais and Kostka, 2019</a>, <a class="reference internal" href="#id121" title="Erica A.K. DePasquale, Daniel J. Schnell, Pieter-Jan Van Camp, Íñigo Valiente-Aland\'ı, Burns C. Blaxall, H. Leighton Grimes, Harinder Singh, and Nathan Salomonis. DoubletDecon: deconvoluting doublets from single-cell term`RNA`-sequencing data. Cell Reports, 29(6):1718–1727.e8, November 2019. URL: https://doi.org/10.1016/j.celrep.2019.09.082, doi:10.1016/j.celrep.2019.09.082.">DePasquale <em>et al.</em>, 2019</a>, <a class="reference internal" href="#id122" title="Christopher S. McGinnis, Lyndsay M. Murrow, and Zev J. Gartner. DoubletFinder: doublet detection in single-cell term`RNA` sequencing data using artificial nearest neighbors. Cell Systems, 8(4):329–337.e4, April 2019. URL: https://doi.org/10.1016/j.cels.2019.03.003, doi:10.1016/j.cels.2019.03.003.">McGinnis <em>et al.</em>, 2019</a>, <a class="reference internal" href="#id123" title="Samuel L. Wolock, Romain Lopez, and Allon M. Klein. Scrublet: Computational Identification of Cell Doublets in Single-Cell Transcriptomic Data. Cell Systems, 8(4):281–291.e9, April 2019. URL: https://doi.org/10.1016/j.cels.2018.11.005, doi:10.1016/j.cels.2018.11.005.">Wolock <em>et al.</em>, 2019</a>]</span>, and correcting cell barcode errors based on nucleotide sequence similarity.</p>
<p>Several common strategies are used for cell barcode identification and correction.</p>
<ol class="arabic simple">
<li><p><strong>Correction against a known list of <em>potential</em> barcodes</strong>:
Certain chemistries, such as 10x Chromium, draw CBs from a known pool of potential barcode sequences.
Thus, the set of barcodes observed in any sample is expected to be a subset of this known list, often called a “whitelist”.
In this case, the standard approach assumes that:</p></li>
</ol>
<ul class="simple">
<li><p>Any barcode matching an entry in the known list is correct.</p></li>
<li><p>Any barcode not in the list is corrected by finding the closest match from the permit list, typically using <a class="reference internal" href="../glossary.html#term-Hamming-distance"><span class="xref std std-term">Hamming distance</span></a> or <a class="reference internal" href="../glossary.html#term-Edit-distance"><span class="xref std std-term">edit distance</span></a>.
This strategy allows for efficient barcode correction but has limitations.
If a corrupted barcode closely resembles multiple barcodes in the permit list, its correction becomes ambiguous.
For example, for a barcode taken from the <a class="reference external" href="https://teichlab.github.io/scg_lib_structs/data/10X-Genomics/3M-february-2018.txt.gz">10x Chromium v3 permit list</a> and mutated at a single position to a barcode not in the list, there is an <span class="math notranslate nohighlight">\(\sim 81\%\)</span> probability that it sits at hamming distance <span class="math notranslate nohighlight">\(1\)</span> from two or more barcodes in the permit list.
The probability of such collisions can be reduced by considering correcting <em>only</em> against barcodes from the known permit list, which, themselves, occur exactly in the given sample (or even only those that occur exactly in the given sample above some nominal frequency threshold).
Also, information such as the base quality at the “corrected” position can be used to potentially break ties in the case of ambiguous corrections.
Yet, as the number of assayed cells increases, insufficient sequence diversity in the set of potential cell barcodes increases the frequency of ambiguous corrections, and reads tagged with barcodes having ambiguous corrections are most commonly discarded.</p></li>
</ul>
<ol class="arabic simple" start="2">
<li><p><strong>Knee or elbow-based methods</strong>:
If a set of potential barcodes is unknown - or even if it is known, but one wishes to correct directly from the observed data itself without consulting an external list - one can use a method based on the observation that high-quality barcodes are those associated with the highest number of reads in the sample.
To achieve this, one can construct a cumulative frequency plot where barcodes are sorted in descending order based on the number of distinct reads or UMIs they are associated with.
Often, this ranked cumulative frequency plot will contain a “knee” or “elbow” – an inflection point that can be used to characterize frequently occurring barcodes from infrequent (and therefore likely erroneous) barcodes.
Many methods exist for attempting to identify such an inflection point <span id="id37">[<a class="reference internal" href="#id110" title="Dongze He, Mohsen Zakeri, Hirak Sarkar, Charlotte Soneson, Avi Srivastava, and Rob Patro. Alevin-fry unlocks rapid, accurate and memory-frugal quantification of single-cell RNA-seq data. Nature Methods, 19(3):316–322, 2022.">He <em>et al.</em>, 2022</a>, <a class="reference internal" href="#id115" title="Aaron TL Lun, Samantha Riesenfeld, Tallulah Andrews, Tomas Gomes, John C Marioni, and others. EmptyDrops: distinguishing cells from empty droplets in droplet-based single-cell term`RNA` sequencing data. Genome biology, 20(1):1–9, 2019.">Lun <em>et al.</em>, 2019</a>, <a class="reference internal" href="#id102" title="Tom Smith, Andreas Heger, and Ian Sudbery. UMI-tools: modeling sequencing errors in Unique Molecular Identifiers to improve quantification accuracy. Genome research, 27(3):491–499, 2017.">Smith <em>et al.</em>, 2017</a>]</span> as a likely point of discrimination between properly captured cells and empty droplets.
Subsequently, the set of barcodes that appear “above” the knee can be treated as a permit list against which the rest of the barcodes may be corrected, as in the first method list above.
Such an approach is flexible as it can be applied in chemistries that have an external permit list and those that don’t.
Further parameters of the knee-finding algorithms can be altered to yield more or less restrictive selected barcode sets.
Yet, such an approach can have certain drawbacks, like a tendency to be overly conservative and sometimes failing to work robustly in samples where no clear knee is present.</p></li>
<li><p><strong>Filtering and correction based on an expected cell count</strong>:
When barcode frequency distributions lack a clear knee or show bimodal patterns due to technical artifacts, barcode correction can be guided by a user-provided expected cell count.
In such an approach, the user provides an estimate of the expected number of assayed cells.
Then, the barcodes are ordered by descending frequency, the frequency <span class="math notranslate nohighlight">\(f\)</span> at a robust quantile index near the expected cell count is obtained, and all cells having a frequency within a small constant fraction of <span class="math notranslate nohighlight">\(f\)</span> (e.g., <span class="math notranslate nohighlight">\(\ge \frac{f}{10}\)</span>) are considered as valid barcodes.
Again, the remaining barcodes are corrected against this valid list by attempting to correct uniquely to one of these valid barcodes based on sequence similarity.</p></li>
<li><p><strong>Filtering based on a forced number of valid cells</strong>:
The simplest approach, although potentially problematic, is for the user to manually specify the number of valid barcodes.</p></li>
</ol>
<ul class="simple">
<li><p>The user chooses an index in the sorted barcode frequency list.</p></li>
<li><p>All barcodes above this threshold are considered valid.</p></li>
<li><p>Remaining barcodes are corrected against this list using standard similarity-based correction methods.
While this guarantees selection of at least n cells, it assumes that the chosen threshold accurately reflects the number of real cells.
It is only reasonable if the user has a good reason to believe that the threshold frequency should be set around the provided index.</p></li>
</ul>
</section>
</section>
<section id="umi-resolution">
<span id="raw-proc-umi-resolution"></span><h2><span class="section-number">3.5. </span>UMI resolution<a class="headerlink" href="#umi-resolution" title="Link to this heading">#</a></h2>
<p>After cell barcode (CB) correction, reads have either been discarded or assigned to a corrected CB.
Subsequently, we wish to quantify the abundance of each gene within each corrected CB.</p>
<p>Because of the <a class="reference internal" href="../glossary.html#term-Amplification-bias"><span class="xref std std-term">amplification bias</span></a> as discussed in <a class="reference internal" href="scrna_seq.html#exp-data-transcript-quantification"><span class="std std-ref">Transcript quantification</span></a>, reads must be deduplicated, based upon their UMI, to assess the true count of sampled molecules. Additionally, several other complicating factors present challenges when attempting to perform this estimation.</p>
<p>The UMI deduplication step aims to identify the set of reads and UMIs derived from each original, pre-PCR molecule in each cell captured and sequenced in the experiment.
The result of this process is to allocate a molecule count to each gene in each cell, which is subsequently used in the downstream analysis as the raw expression estimate for this gene.
We refer to this process of looking at the collection of observed UMIs and their associated mapped reads and attempting to infer the original number of observed molecules arising from each gene as the process of <em>UMI resolution</em>.</p>
<p>To simplify the explanation, reads that map to a reference (e.g., a genomic locus of a gene) are referred to as the reads of that reference, and their UMI tags are called the UMIs of that reference.
The set of reads associated with a specific UMI is referred to as the reads of that UMI.</p>
<p>A read can be tagged by only one UMI but may belong to multiple references if it maps to more than one.
Additionally, since molecule barcoding in scRNA-seq is typically isolated and independent for each cell (aside from the previously discussed challenges in resolving cell barcodes), <em>UMI resolution</em> will be explained for a single cell without loss of generality.
This same procedure is generally applied to all cells independently.</p>
<section id="the-need-for-umi-resolution">
<span id="raw-proc-need-for-umi-resolution"></span><h3><span class="section-number">3.5.1. </span>The need for UMI resolution<a class="headerlink" href="#the-need-for-umi-resolution" title="Link to this heading">#</a></h3>
<p>In the ideal case, where the correct (unaltered) UMIs tag reads, the reads of each UMI uniquely map to a common reference gene, and there is a bijection between UMIs and pre-PCR molecules.
Consequently, the UMI deduplication procedure is conceptually straightforward: the reads of a UMI are the PCR duplicates from a single pre-PCR molecule.
The number of captured and sequenced molecules of each gene is the number of distinct UMIs observed for this gene.</p>
<p>However, the problems encountered in practice make the simple rules described above insufficient for identifying the gene origin of UMIs in general and necessitate the development of more sophisticated models:</p>
<ul class="simple">
<li><p><strong>Errors in UMIs</strong>:
These occur when the sequenced UMI tag of reads contains errors introduced during PCR or the sequencing process.
Common UMI errors include nucleotide substitutions during PCR and read errors during sequencing.
Failing to address such UMI errors can inflate the estimated number of molecules <span id="id38">[<a class="reference internal" href="#id102" title="Tom Smith, Andreas Heger, and Ian Sudbery. UMI-tools: modeling sequencing errors in Unique Molecular Identifiers to improve quantification accuracy. Genome research, 27(3):491–499, 2017.">Smith <em>et al.</em>, 2017</a>, <a class="reference internal" href="#id135" title="Christoph Ziegenhain, Gert-Jan Hendriks, Michael Hagemann-Jensen, and Rickard Sandberg. Molecular spikes: a gold standard for single-cell term`RNA` counting. Nature Methods, 19(5):560–566, 2022.">Ziegenhain <em>et al.</em>, 2022</a>]</span>.</p></li>
<li><p><strong>Multimapping</strong>:
This issue arises in cases where a read or UMI belongs to multiple references (e.g., multi-gene reads/UMIs).
This happens when different reads of a UMI map to different genes, when a read maps to multiple genes, or both.
The consequence of this issue is that the gene origin of the multi-gene reads/UMIs is ambiguous, which results in uncertainty about the sampled pre-PCR molecule count of those genes.
Simply discarding multi-gene reads/UMIs can lead to a loss of data or a biased estimate among genes that tend to produce multimapping reads, such as sequence-similar gene families <span id="id39">[<a class="reference internal" href="#id103" title="Avi Srivastava, Laraib Malik, Tom Smith, Ian Sudbery, and Rob Patro. Alevin efficiently estimates accurate gene abundances from dscRNA-seq data. Genome biology, 20(1):1–16, 2019.">Srivastava <em>et al.</em>, 2019</a>]</span>.</p></li>
</ul>
<div class="admonition-a-note-on-umi-errors admonition">
<p class="admonition-title">A Note on UMI Errors</p>
<p>UMI errors, especially those due to nucleotide substitutions and miscallings, are prevalent in single-cell experiments.
<span id="id40">Smith <em>et al.</em> [<a class="reference internal" href="#id102" title="Tom Smith, Andreas Heger, and Ian Sudbery. UMI-tools: modeling sequencing errors in Unique Molecular Identifiers to improve quantification accuracy. Genome research, 27(3):491–499, 2017.">2017</a>]</span> establish that the average number of bases different (edit distance) between the observed UMI sequences in the tested single-cell experiments is lower than randomly sampled UMI sequences, and the enrichment of low edit distances is well correlated with the degree of PCR amplification.
Multimapping also exists in single-cell data and, depending upon the gene being considered, can occur at a non-trivial rate.
<span id="id41">Srivastava <em>et al.</em> [<a class="reference internal" href="#id103" title="Avi Srivastava, Laraib Malik, Tom Smith, Ian Sudbery, and Rob Patro. Alevin efficiently estimates accurate gene abundances from dscRNA-seq data. Genome biology, 20(1):1–16, 2019.">2019</a>]</span> show that discarding the multimapping reads can negatively bias the predicted molecule counts.</p>
</div>
<p>There exist other challenges that we do not focus upon here, such as “convergent” and “divergent” UMI collisions.
We consider the case where the same UMI is used to tag two different pre-PCR molecules arising from the same gene, in the same cell, as a convergent collision.
When two or more distinct UMIs arise from the same pre-PCR molecule, e.g., due to the sampling of multiple priming sites from this molecule, we consider this a divergent collision.
We expect convergent UMI collisions to be rare and, therefore, their effect typically small.
Further, transcript-level mapping information can sometimes be used to resolve such collisions <span id="id42">[<a class="reference internal" href="#id103" title="Avi Srivastava, Laraib Malik, Tom Smith, Ian Sudbery, and Rob Patro. Alevin efficiently estimates accurate gene abundances from dscRNA-seq data. Genome biology, 20(1):1–16, 2019.">Srivastava <em>et al.</em>, 2019</a>]</span>.
Divergent UMI collisions occur primarily among introns of unspliced transcripts <span id="id43">[<a class="reference internal" href="#id104" title="10x Genomics. Technical note - interpreting intronic and antisense reads in 10x genomics single cell gene expression data. https://www.10xgenomics.com/support/single-cell-gene-expression/documentation/steps/sequencing/interpreting-intronic-and-antisense-reads-in-10-x-genomics-single-cell-gene-expression-data, August 2021.">10x Genomics, 2021</a>]</span>, and approaches to addressing the issues they raise are an area of active research <span id="id44">[<a class="reference internal" href="#id136" title="Gennady Gorin and Lior Pachter. Length Biases in Single-Cell term`RNA` Sequencing of pre-mRNA. bioRxiv, 2021. URL: https://www.biorxiv.org/content/early/2021/07/31/2021.07.30.454514, arXiv:https://www.biorxiv.org/content/early/2021/07/31/2021.07.30.454514.full.pdf, doi:10.1101/2021.07.30.454514.">Gorin and Pachter, 2021</a>, <a class="reference internal" href="#id104" title="10x Genomics. Technical note - interpreting intronic and antisense reads in 10x genomics single cell gene expression data. https://www.10xgenomics.com/support/single-cell-gene-expression/documentation/steps/sequencing/interpreting-intronic-and-antisense-reads-in-10-x-genomics-single-cell-gene-expression-data, August 2021.">10x Genomics, 2021</a>]</span>.</p>
<p>Given that the use of UMIs is near ubiquitous in high-throughput scRNA-seq protocols and the fact that addressing these errors improves the estimation of gene abundances, there has been much attention paid to the problem of UMI resolution in recent literature <span id="id45">[<a class="reference internal" href="#id106" title="Sayantan Bose, Zhenmao Wan, Ambrose Carr, Abbas H. Rizvi, Gregory Vieira, Dana Pe'er, and Peter A. Sims. Scalable microfluidics for single-cell term`RNA` printing and sequencing. Genome Biology, June 2015. URL: https://doi.org/10.1186/s13059-015-0684-3, doi:10.1186/s13059-015-0684-3.">Bose <em>et al.</em>, 2015</a>, <a class="reference internal" href="#id110" title="Dongze He, Mohsen Zakeri, Hirak Sarkar, Charlotte Soneson, Avi Srivastava, and Rob Patro. Alevin-fry unlocks rapid, accurate and memory-frugal quantification of single-cell RNA-seq data. Nature Methods, 19(3):316–322, 2022.">He <em>et al.</em>, 2022</a>, <a class="reference internal" href="#id105" title="Saiful Islam, Amit Zeisel, Simon Joost, Gioele La Manno, Pawel Zajac, Maria Kasper, Peter Lönnerberg, and Sten Linnarsson. Quantitative single-cell term`RNA`-seq with unique molecular identifiers. Nature Methods, 11(2):163–166, December 2013. URL: https://doi.org/10.1038/nmeth.2772, doi:10.1038/nmeth.2772.">Islam <em>et al.</em>, 2013</a>, <a class="reference internal" href="#id107" title="Benjamin Kaminow, Dinar Yunusov, and Alexander Dobin. STARsolo: accurate, fast and versatile mapping/quantification of single-cell and single-nucleus term`RNA`-seq data. bioRxiv, 2021.">Kaminow <em>et al.</em>, 2021</a>, <a class="reference internal" href="#id142" title="Evan Z. Macosko, Anindita Basu, Rahul Satija, James Nemesh, Karthik Shekhar, Melissa Goldman, Itay Tirosh, Allison R. Bialas, Nolan Kamitaki, Emily M. Martersteck, John J. Trombetta, David A. Weitz, Joshua R. Sanes, Alex K. Shalek, Aviv Regev, and Steven A. McCarroll. Highly parallel genome-wide expression profiling of individual cells using nanoliter droplets. Cell, 161(5):1202-1214, May 2015. URL: https://doi.org/10.1016/j.cell.2015.05.002, doi:10.1016/j.cell.2015.05.002.">Macosko <em>et al.</em>, 2015</a>, <a class="reference internal" href="#id108" title="Páll Melsted, A. Sina Booeshaghi, Lauren Liu, Fan Gao, Lambda Lu, Kyung Hoi Min, Eduardo da Veiga Beltrame, Kristján Eldjárn Hjörleifsson, Jase Gehring, and Lior Pachter. Modular, efficient and constant-memory single-cell rna-seq preprocessing. Nature Biotechnology, 39(7):813–818, April 2021. URL: https://doi.org/10.1038/s41587-021-00870-2, doi:10.1038/s41587-021-00870-2.">Melsted <em>et al.</em>, 2021</a>, <a class="reference internal" href="#id111" title="Baraa Orabi, Emre Erhan, Brian McConeghy, Stanislav V Volik, Stephane Le Bihan, Robert Bell, Colin C Collins, Cedric Chauve, and Faraz Hach. Alignment-free clustering of UMI tagged term`DNA` molecules. Bioinformatics, 35(11):1829–1836, October 2018. URL: https://doi.org/10.1093/bioinformatics/bty888, doi:10.1093/bioinformatics/bty888.">Orabi <em>et al.</em>, 2018</a>, <a class="reference internal" href="#id113" title="Swati Parekh, Christoph Ziegenhain, Beate Vieth, Wolfgang Enard, and Ines Hellmann. zUMIs - a fast and flexible pipeline to process term`RNA` sequencing data with UMIs. GigaScience, May 2018. URL: https://doi.org/10.1093/gigascience/giy059, doi:10.1093/gigascience/giy059.">Parekh <em>et al.</em>, 2018</a>, <a class="reference internal" href="#id102" title="Tom Smith, Andreas Heger, and Ian Sudbery. UMI-tools: modeling sequencing errors in Unique Molecular Identifiers to improve quantification accuracy. Genome research, 27(3):491–499, 2017.">Smith <em>et al.</em>, 2017</a>, <a class="reference internal" href="#id103" title="Avi Srivastava, Laraib Malik, Tom Smith, Ian Sudbery, and Rob Patro. Alevin efficiently estimates accurate gene abundances from dscRNA-seq data. Genome biology, 20(1):1–16, 2019.">Srivastava <em>et al.</em>, 2019</a>, <a class="reference internal" href="#id112" title="Maria Tsagiopoulou, Maria Christina Maniou, Nikolaos Pechlivanis, Anastasis Togkousidis, Michaela Kotrová, Tobias Hutzenlaub, Ilias Kappas, Anastasia Chatzidimitriou, and Fotis Psomopoulos. UMIc: a preprocessing method for UMI deduplication and reads correction. Frontiers in Genetics, May 2021. URL: https://doi.org/10.3389/fgene.2021.660366, doi:10.3389/fgene.2021.660366.">Tsagiopoulou <em>et al.</em>, 2021</a>]</span>.</p>
<details class="sd-sphinx-override sd-dropdown sd-card sd-mb-3">
<summary class="sd-summary-title sd-card-header">
<span class="sd-summary-text">Graph-based UMI resolution</span><span class="sd-summary-state-marker sd-summary-chevron-right"><svg version="1.1" width="1.5em" height="1.5em" class="sd-octicon sd-octicon-chevron-right" viewBox="0 0 24 24" aria-hidden="true"><path d="M8.72 18.78a.75.75 0 0 1 0-1.06L14.44 12 8.72 6.28a.751.751 0 0 1 .018-1.042.751.751 0 0 1 1.042-.018l6.25 6.25a.75.75 0 0 1 0 1.06l-6.25 6.25a.75.75 0 0 1-1.06 0Z"></path></svg></span></summary><div class="sd-summary-content sd-card-body docutils">
<h3 class="rubric" id="graph-based-umi-resolution"><span id="raw-proc-graph-based-umi-resolution"></span>Graph-based UMI resolution</h3>
<p class="sd-card-text">As a result of the problems that ariOther UMI resolution approaches exist, for example, the reference-free model <span id="id46">[<a class="reference internal" href="#id112" title="Maria Tsagiopoulou, Maria Christina Maniou, Nikolaos Pechlivanis, Anastasis Togkousidis, Michaela Kotrová, Tobias Hutzenlaub, Ilias Kappas, Anastasia Chatzidimitriou, and Fotis Psomopoulos. UMIc: a preprocessing method for UMI deduplication and reads correction. Frontiers in Genetics, May 2021. URL: https://doi.org/10.3389/fgene.2021.660366, doi:10.3389/fgene.2021.660366.">Tsagiopoulou <em>et al.</em>, 2021</a>]</span> and the method of moments <span id="id47">[<a class="reference internal" href="#id108" title="Páll Melsted, A. Sina Booeshaghi, Lauren Liu, Fan Gao, Lambda Lu, Kyung Hoi Min, Eduardo da Veiga Beltrame, Kristján Eldjárn Hjörleifsson, Jase Gehring, and Lior Pachter. Modular, efficient and constant-memory single-cell rna-seq preprocessing. Nature Biotechnology, 39(7):813–818, April 2021. URL: https://doi.org/10.1038/s41587-021-00870-2, doi:10.1038/s41587-021-00870-2.">Melsted <em>et al.</em>, 2021</a>]</span>, but they may not be easily represented in this framework and are not discussed in further detail <a class="reference external" href="http://here.se">here.se</a> when attempting to resolve UMIs, many methods have been developed to address the problem of UMI resolution.
While there are a host of different approaches for UMI resolution, we will focus on a framework for representing problem instances, modified from a framework initially proposed by <span id="id48">Smith <em>et al.</em> [<a class="reference internal" href="#id102" title="Tom Smith, Andreas Heger, and Ian Sudbery. UMI-tools: modeling sequencing errors in Unique Molecular Identifiers to improve quantification accuracy. Genome research, 27(3):491–499, 2017.">2017</a>]</span>, that relies upon the notion of a <em>UMI graph</em>.
Each connected component of this graph represents a sub-problem wherein certain subsets of UMIs are collapsed (i.e., resolved as evidence of the same pre-PCR molecule).
Many popular UMI resolution approaches can be interpreted in this framework by simply modifying precisely how the graph is refined and how the collapse or resolution procedure carried out over this graph works.</p>
<p class="sd-card-text">In the context of single-cell data, a UMI graph <span class="math notranslate nohighlight">\(G(V,E)\)</span> is a <a class="reference internal" href="../glossary.html#term-directed-graph"><span class="xref std std-term">directed graph</span></a> with a node set <span class="math notranslate nohighlight">\(V\)</span> and an edge set <span class="math notranslate nohighlight">\(E\)</span>.
Each node <span class="math notranslate nohighlight">\(v_i \in V\)</span> represents an equivalence class (EC) of reads, and the edge set <span class="math notranslate nohighlight">\(E\)</span> encodes the relationship between the ECs.
The equivalence relation <span class="math notranslate nohighlight">\(\sim_r\)</span> defined on reads is based on their UMI and mapping information.
We say reads <span class="math notranslate nohighlight">\(r_x\)</span> and <span class="math notranslate nohighlight">\(r_y\)</span> are equivalent, <span class="math notranslate nohighlight">\(r_x \sim_r r_y\)</span>, if and only if they have identical UMI tags and map to the same set of references.
UMI resolution approaches may define a “reference” as a genomic locus <span id="id49">[<a class="reference internal" href="#id102" title="Tom Smith, Andreas Heger, and Ian Sudbery. UMI-tools: modeling sequencing errors in Unique Molecular Identifiers to improve quantification accuracy. Genome research, 27(3):491–499, 2017.">Smith <em>et al.</em>, 2017</a>]</span>, transcript <span id="id50">[<a class="reference internal" href="#id110" title="Dongze He, Mohsen Zakeri, Hirak Sarkar, Charlotte Soneson, Avi Srivastava, and Rob Patro. Alevin-fry unlocks rapid, accurate and memory-frugal quantification of single-cell RNA-seq data. Nature Methods, 19(3):316–322, 2022.">He <em>et al.</em>, 2022</a>, <a class="reference internal" href="#id103" title="Avi Srivastava, Laraib Malik, Tom Smith, Ian Sudbery, and Rob Patro. Alevin efficiently estimates accurate gene abundances from dscRNA-seq data. Genome biology, 20(1):1–16, 2019.">Srivastava <em>et al.</em>, 2019</a>]</span> or gene <span id="id51">[<a class="reference internal" href="#id107" title="Benjamin Kaminow, Dinar Yunusov, and Alexander Dobin. STARsolo: accurate, fast and versatile mapping/quantification of single-cell and single-nucleus term`RNA`-seq data. bioRxiv, 2021.">Kaminow <em>et al.</em>, 2021</a>, <a class="reference internal" href="#id143" title="Grace X. Y. Zheng, Jessica M. Terry, Phillip Belgrader, Paul Ryvkin, Zachary W. Bent, Ryan Wilson, Solongo B. Ziraldo, Tobias D. Wheeler, Geoff P. McDermott, Junjie Zhu, Mark T. Gregory, Joe Shuga, Luz Montesclaros, Jason G. Underwood, Donald A. Masquelier, Stefanie Y. Nishimura, Michael Schnall-Levin, Paul W. Wyatt, Christopher M. Hindson, Rajiv Bharadwaj, Alexander Wong, Kevin D. Ness, Lan W. Beppu, H. Joachim Deeg, Christopher McFarland, Keith R. Loeb, William J. Valente, Nolan G. Ericson, Emily A. Stevens, Jerald P. Radich, Tarjei S. Mikkelsen, Benjamin J. Hindson, and Jason H. Bielas. Massively parallel digital transcriptional profiling of single cells. Nature Communications, 8(1):14049, Jan 2017. URL: https://doi.org/10.1038/ncomms14049, doi:10.1038/ncomms14049.">Zheng <em>et al.</em>, 2017</a>]</span>.</p>
<p class="sd-card-text">In the UMI graph framework, a UMI resolution approach can be divided into three major steps:
<strong>defining nodes</strong>, <strong>defining adjacency relationships</strong>, and <strong>resolving components</strong>.
Each of these steps has different options that can be modularly composed by different approaches.
Additionally, these steps may sometimes be preceded (and/or followed) by filtering steps designed to discard or heuristically assign (by modifying the set of reference mappings reported) reads and UMIs exhibiting certain types of mapping ambiguity.</p>
<h4 class="rubric" id="defining-nodes"><span id="raw-proc-umi-graph-node-def"></span>Defining nodes</h4>
<p class="sd-card-text">As described above, a node <span class="math notranslate nohighlight">\(v_i \in V\)</span> is an equivalence class of reads.
Therefore, <span class="math notranslate nohighlight">\(V\)</span> can be defined based on the full or filtered set of mapped reads and their associated <em>uncorrected</em> UMIs.
All reads that satisfy the equivalence relation <span class="math notranslate nohighlight">\(\sim_r\)</span> based on their reference set and UMI tag are associated with the same vertex <span class="math notranslate nohighlight">\(v \in V\)</span>.
An EC is a multi-gene EC if its UMI is a multi-gene UMI.
Some approaches will avoid the creation of such ECs by filtering or heuristically assigning reads prior to node creation, while other approaches will retain and process these ambiguous vertices and attempt and resolve their gene origin via parsimony, probabilistic assignment, or based on a related rule or model <span id="id52">[<a class="reference internal" href="#id110" title="Dongze He, Mohsen Zakeri, Hirak Sarkar, Charlotte Soneson, Avi Srivastava, and Rob Patro. Alevin-fry unlocks rapid, accurate and memory-frugal quantification of single-cell RNA-seq data. Nature Methods, 19(3):316–322, 2022.">He <em>et al.</em>, 2022</a>, <a class="reference internal" href="#id107" title="Benjamin Kaminow, Dinar Yunusov, and Alexander Dobin. STARsolo: accurate, fast and versatile mapping/quantification of single-cell and single-nucleus term`RNA`-seq data. bioRxiv, 2021.">Kaminow <em>et al.</em>, 2021</a>, <a class="reference internal" href="#id103" title="Avi Srivastava, Laraib Malik, Tom Smith, Ian Sudbery, and Rob Patro. Alevin efficiently estimates accurate gene abundances from dscRNA-seq data. Genome biology, 20(1):1–16, 2019.">Srivastava <em>et al.</em>, 2019</a>]</span>.</p>
<h4 class="rubric" id="defining-the-adjacency-relationship"><span id="raw-proc-umi-graph-edge-def"></span>Defining the adjacency relationship</h4>
<p class="sd-card-text">After creating the node set <span class="math notranslate nohighlight">\(V\)</span> of a UMI graph, the adjacency of nodes in <span class="math notranslate nohighlight">\(V\)</span> is defined based on the distance, typically the Hamming or edit distance, between their UMI sequences and, optionally, the content of their associated reference sets.</p>
<p class="sd-card-text">Here we define the following functions on the node <span class="math notranslate nohighlight">\(v_i \in V\)</span>:</p>
<ul class="simple">
<li><p class="sd-card-text"><span class="math notranslate nohighlight">\(u(v_i)\)</span> is the UMI tag of <span class="math notranslate nohighlight">\(v_i\)</span>.</p></li>
<li><p class="sd-card-text"><span class="math notranslate nohighlight">\(c(v_i) = |v_i|\)</span> is the cardinality of <span class="math notranslate nohighlight">\(v_i\)</span>, i.e., the number of reads associated with <span class="math notranslate nohighlight">\(v_i\)</span> that are equivalent under <span class="math notranslate nohighlight">\(\sim_r\)</span>.</p></li>
<li><p class="sd-card-text"><span class="math notranslate nohighlight">\(m(v_i)\)</span> is the reference set encoded in the mapping information, for <span class="math notranslate nohighlight">\(v_i\)</span>.</p></li>
<li><p class="sd-card-text"><span class="math notranslate nohighlight">\(D(v_i, v_j)\)</span> is the distance between <span class="math notranslate nohighlight">\(u(v_i)\)</span> and <span class="math notranslate nohighlight">\(u(v_j)\)</span>, where <span class="math notranslate nohighlight">\(v_j \in V\)</span>.</p></li>
</ul>
<p class="sd-card-text">Given these function definitions, any two nodes <span class="math notranslate nohighlight">\(v_i, v_j \in V\)</span> will be incident with a bi-directed edge if and only if <span class="math notranslate nohighlight">\(m(v_i) \cap m(v_j) \ne \emptyset\)</span> and <span class="math notranslate nohighlight">\(D(v_i,v_j) \le \theta\)</span>, where <span class="math notranslate nohighlight">\(\theta\)</span> is a distance threshold and is often set as <span class="math notranslate nohighlight">\(\theta=1\)</span> <span id="id53">[<a class="reference internal" href="#id107" title="Benjamin Kaminow, Dinar Yunusov, and Alexander Dobin. STARsolo: accurate, fast and versatile mapping/quantification of single-cell and single-nucleus term`RNA`-seq data. bioRxiv, 2021.">Kaminow <em>et al.</em>, 2021</a>, <a class="reference internal" href="#id102" title="Tom Smith, Andreas Heger, and Ian Sudbery. UMI-tools: modeling sequencing errors in Unique Molecular Identifiers to improve quantification accuracy. Genome research, 27(3):491–499, 2017.">Smith <em>et al.</em>, 2017</a>, <a class="reference internal" href="#id103" title="Avi Srivastava, Laraib Malik, Tom Smith, Ian Sudbery, and Rob Patro. Alevin efficiently estimates accurate gene abundances from dscRNA-seq data. Genome biology, 20(1):1–16, 2019.">Srivastava <em>et al.</em>, 2019</a>]</span>.
Additionally, the bi-directed edge might be replaced by a directed edge incident from <span class="math notranslate nohighlight">\(v_i\)</span> to <span class="math notranslate nohighlight">\(v_j\)</span> if <span class="math notranslate nohighlight">\(c(v_i) \ge 2c(v_j) -1\)</span> or vice versa <span id="id54">[<a class="reference internal" href="#id102" title="Tom Smith, Andreas Heger, and Ian Sudbery. UMI-tools: modeling sequencing errors in Unique Molecular Identifiers to improve quantification accuracy. Genome research, 27(3):491–499, 2017.">Smith <em>et al.</em>, 2017</a>, <a class="reference internal" href="#id103" title="Avi Srivastava, Laraib Malik, Tom Smith, Ian Sudbery, and Rob Patro. Alevin efficiently estimates accurate gene abundances from dscRNA-seq data. Genome biology, 20(1):1–16, 2019.">Srivastava <em>et al.</em>, 2019</a>]</span>.
Though these edge definitions are among the most common, others are possible, so long as they are completely defined by the <span class="math notranslate nohighlight">\(u\)</span>, <span class="math notranslate nohighlight">\(c\)</span>, <span class="math notranslate nohighlight">\(m\)</span>, and <span class="math notranslate nohighlight">\(D\)</span> functions. With <span class="math notranslate nohighlight">\(V\)</span> and <span class="math notranslate nohighlight">\(E\)</span> in hand, the UMI graph <span class="math notranslate nohighlight">\(G = (V,E)\)</span> is now defined.</p>
<h4 class="rubric" id="defining-the-graph-resolution-approach"><span id="raw-proc-umi-graph-resolution-def"></span>Defining the graph resolution approach</h4>
<p class="sd-card-text">Given the defined UMI graph, many different resolution approaches may be applied.
A resolution method may be as simple as finding the set of connected components, clustering the graph, greedily collapsing nodes or contracting edges <span id="id55">[<a class="reference internal" href="#id102" title="Tom Smith, Andreas Heger, and Ian Sudbery. UMI-tools: modeling sequencing errors in Unique Molecular Identifiers to improve quantification accuracy. Genome research, 27(3):491–499, 2017.">Smith <em>et al.</em>, 2017</a>]</span>, or searching for a cover of the graph by structures following certain rules (e.g., monochromatic arboresences <span id="id56">[<a class="reference internal" href="#id103" title="Avi Srivastava, Laraib Malik, Tom Smith, Ian Sudbery, and Rob Patro. Alevin efficiently estimates accurate gene abundances from dscRNA-seq data. Genome biology, 20(1):1–16, 2019.">Srivastava <em>et al.</em>, 2019</a>]</span>) to reduce the graph.
As a result, each node in the reduced UMI graph, or each element in the cover in the case that the graph is not modified dynamically, represents a pre-PCR molecule.
The collapsed nodes or covering sets are regarded as the PCR duplicates of that molecule.</p>
<p class="sd-card-text">Different rules for defining the adjacency relationship and different approaches for graph resolution itself can seek to preserve different properties and can define a wide variety of distinct overall UMI resolution approaches.
For approaches that probabilistically resolve ambiguity caused by multimapping, the resolved UMI graph may contain multi-gene equivalence classes (ECs), with their gene origins determined in the next step.</p>
</div>
</details><section id="quantification">
<span id="raw-proc-umi-graph-quantification"></span><h4><span class="section-number">3.5.1.1. </span>Quantification<a class="headerlink" href="#quantification" title="Link to this heading">#</a></h4>
<p>The last step in UMI resolution is quantifying the abundance of each gene using the resolved UMI graph.
For approaches that discard multi-gene ECs, the molecule count vector for the genes in the current cell being processed (or count vector for short) is generated by counting the number of ECs labeled with each gene.
On the other hand, approaches that process, rather than discard, multi-gene ECs usually resolve the ambiguity by applying some statistical inference procedure.
For example, <span id="id57">Srivastava <em>et al.</em> [<a class="reference internal" href="#id103" title="Avi Srivastava, Laraib Malik, Tom Smith, Ian Sudbery, and Rob Patro. Alevin efficiently estimates accurate gene abundances from dscRNA-seq data. Genome biology, 20(1):1–16, 2019.">2019</a>]</span> introduce an expectation-maximization (EM) approach for probabilistically assigning multi-gene UMIs, and related EM algorithms have also been introduced as optional steps in subsequent tools <span id="id58">[<a class="reference internal" href="#id110" title="Dongze He, Mohsen Zakeri, Hirak Sarkar, Charlotte Soneson, Avi Srivastava, and Rob Patro. Alevin-fry unlocks rapid, accurate and memory-frugal quantification of single-cell RNA-seq data. Nature Methods, 19(3):316–322, 2022.">He <em>et al.</em>, 2022</a>, <a class="reference internal" href="#id107" title="Benjamin Kaminow, Dinar Yunusov, and Alexander Dobin. STARsolo: accurate, fast and versatile mapping/quantification of single-cell and single-nucleus term`RNA`-seq data. bioRxiv, 2021.">Kaminow <em>et al.</em>, 2021</a>, <a class="reference internal" href="#id108" title="Páll Melsted, A. Sina Booeshaghi, Lauren Liu, Fan Gao, Lambda Lu, Kyung Hoi Min, Eduardo da Veiga Beltrame, Kristján Eldjárn Hjörleifsson, Jase Gehring, and Lior Pachter. Modular, efficient and constant-memory single-cell rna-seq preprocessing. Nature Biotechnology, 39(7):813–818, April 2021. URL: https://doi.org/10.1038/s41587-021-00870-2, doi:10.1038/s41587-021-00870-2.">Melsted <em>et al.</em>, 2021</a>]</span>.
In this model, the collapsed-EC-to-gene assignments are latent variables, and the deduplicated molecule count of genes are the main parameters.
Intuitively, evidence from gene-unique ECs will be used to help probabilistically apportion the multi-gene ECs.
The EM algorithm seeks the parameters that together have the (locally) highest likelihood of generating the observed ECs.</p>
<p>Usually, the UMI resolution and quantification process described above will be performed separately for each cell, represented by a corrected CB, to create a complete count matrix for all genes in all cells.
However, the relative paucity of per-cell information in high-throughput single-cell samples limits the evidence available when performing UMI resolution, which in turn limits the potential efficacy of model-based solutions like the statistical inference procedure described above.</p>
</section>
</section>
</section>
<section id="count-matrix-quality-control">
<span id="raw-proc-count-qc"></span><h2><span class="section-number">3.6. </span>Count matrix quality control<a class="headerlink" href="#count-matrix-quality-control" title="Link to this heading">#</a></h2>
<p>Once a count matrix has been generated, it is important to perform a quality control (QC) assessment.
There are several distinct assessments that generally fall under the rubric of quality control.
Basic global metrics are often recorded and reported to help assess the overall quality of the sequencing measurement itself.
These metrics consist of quantities such as the total fraction of mapped reads, the distribution of distinct UMIs observed per cell, the distribution of UMI deduplication rates, the distribution of detected genes per cell, etc.
These and similar metrics are often recorded by the quantification tools themselves <span id="id59">[<a class="reference internal" href="#id110" title="Dongze He, Mohsen Zakeri, Hirak Sarkar, Charlotte Soneson, Avi Srivastava, and Rob Patro. Alevin-fry unlocks rapid, accurate and memory-frugal quantification of single-cell RNA-seq data. Nature Methods, 19(3):316–322, 2022.">He <em>et al.</em>, 2022</a>, <a class="reference internal" href="#id107" title="Benjamin Kaminow, Dinar Yunusov, and Alexander Dobin. STARsolo: accurate, fast and versatile mapping/quantification of single-cell and single-nucleus term`RNA`-seq data. bioRxiv, 2021.">Kaminow <em>et al.</em>, 2021</a>, <a class="reference internal" href="#id108" title="Páll Melsted, A. Sina Booeshaghi, Lauren Liu, Fan Gao, Lambda Lu, Kyung Hoi Min, Eduardo da Veiga Beltrame, Kristján Eldjárn Hjörleifsson, Jase Gehring, and Lior Pachter. Modular, efficient and constant-memory single-cell rna-seq preprocessing. Nature Biotechnology, 39(7):813–818, April 2021. URL: https://doi.org/10.1038/s41587-021-00870-2, doi:10.1038/s41587-021-00870-2.">Melsted <em>et al.</em>, 2021</a>, <a class="reference internal" href="#id143" title="Grace X. Y. Zheng, Jessica M. Terry, Phillip Belgrader, Paul Ryvkin, Zachary W. Bent, Ryan Wilson, Solongo B. Ziraldo, Tobias D. Wheeler, Geoff P. McDermott, Junjie Zhu, Mark T. Gregory, Joe Shuga, Luz Montesclaros, Jason G. Underwood, Donald A. Masquelier, Stefanie Y. Nishimura, Michael Schnall-Levin, Paul W. Wyatt, Christopher M. Hindson, Rajiv Bharadwaj, Alexander Wong, Kevin D. Ness, Lan W. Beppu, H. Joachim Deeg, Christopher McFarland, Keith R. Loeb, William J. Valente, Nolan G. Ericson, Emily A. Stevens, Jerald P. Radich, Tarjei S. Mikkelsen, Benjamin J. Hindson, and Jason H. Bielas. Massively parallel digital transcriptional profiling of single cells. Nature Communications, 8(1):14049, Jan 2017. URL: https://doi.org/10.1038/ncomms14049, doi:10.1038/ncomms14049.">Zheng <em>et al.</em>, 2017</a>]</span> since they arise naturally and can be computed during the process of read mapping, cell barcode correction, and UMI resolution.
Likewise, there exist several tools to help organize and visualize these basic metrics, such as the <a class="reference external" href="https://support.10xgenomics.com/single-cell-gene-expression/software/visualization/latest/what-is-loupe-cell-browser">Loupe browser</a>, <a class="reference external" href="https://github.com/csoneson/alevinQC">alevinQC</a>, or a <a class="reference external" href="https://github.com/pachterlab/kb_python">kb_python report</a>, depending upon the quantification pipeline being used.
Beyond these basic global metrics, at this stage of analysis, QC metrics are designed primarily to help determine which cells (CBs) have been sequenced “successfully”, and which exhibit artifacts that warrant filtering or correction.</p>
<p>In the following toggle section, we discuss an example alevinQC report taken from the <code class="docutils literal notranslate"><span class="pre">alevinQC</span></code> <a class="reference external" href="https://github.com/csoneson/alevinQC">manual webpage</a>.</p>
<div class="toggle docutils container">
<p>Once <code class="docutils literal notranslate"><span class="pre">alevin</span></code> or <code class="docutils literal notranslate"><span class="pre">alevin-fry</span></code> quantifies the single-cell data, the quality of the data can be assessed through the R package <a class="reference external" href="https://github.com/csoneson/alevinQC"><code class="docutils literal notranslate"><span class="pre">alevinQC</span></code></a>.
The alevinQC report can be generated in PDF format or as R/Shiny applications, which summarizes various components of the single-cell library, such as reads, CBs, and UMIs.</p>
<p><strong>1. Metadata and summary tables</strong></p>
<figure class="align-default" id="raw-proc-fig-alevinqc-summary">
<a class="bg-primary mb-1 reference internal image-reference" href="../_images/alevinQC_summary.png"><img alt="AlevinQC Summary" class="bg-primary mb-1" src="../_images/alevinQC_summary.png" style="width: 800px;" />
</a>
<figcaption>
<p><span class="caption-number">Fig. 3.15 </span><span class="caption-text">An example of the summary section of an alevinQC report.</span><a class="headerlink" href="#raw-proc-fig-alevinqc-summary" title="Link to this image">#</a></p>
</figcaption>
</figure>
<p>The first section of an alevinQC report shows a summary of the input files and the processing result, among which, the top left table displays the metadata provided by <code class="docutils literal notranslate"><span class="pre">alevin</span></code> (or <code class="docutils literal notranslate"><span class="pre">alevin-fry</span></code>) for the quantification results.
For example, this includes the time of the run, the version of the tool, and the path to the input FASTQ and index files.
The top right summary table provides the summary statistics for various components of the single-cell library, for example, the number of sequencing reads, the number of selected cell barcodes at various levels of filtering, and the total number of deduplicated UMIs.</p>
<p><strong>2. Knee plot, initial whitelist determination</strong></p>
<figure class="align-default" id="raw-proc-fig-alevinqc-plots">
<a class="bg-primary mb-1 reference internal image-reference" href="../_images/alevinQC_plots.png"><img alt="AlevinQC Plots" class="bg-primary mb-1" src="../_images/alevinQC_plots.png" style="width: 800px;" />
</a>
<figcaption>
<p><span class="caption-number">Fig. 3.16 </span><span class="caption-text">The figure shows the plots in the alevinQC report of an example single-cell dataset, of which the cells are filtered using the “knee” finding method.
Each dot represents a corrected cell barcode with its corrected profile.</span><a class="headerlink" href="#raw-proc-fig-alevinqc-plots" title="Link to this image">#</a></p>
</figcaption>
</figure>
<p>The first (top left) view in <a class="reference internal" href="#raw-proc-fig-alevinqc-plots"><span class="std std-numref">Fig. 3.16</span></a> shows the distribution of cell barcode frequency in decreasing order.
In all plots shown above, each point represents a corrected cell barcode, with its x-coordinate corresponding to its cell barcode frequency rank.
In the top left plot, the y-coordinate corresponds to the observed frequency of the corrected barcode.
Generally, this plot shows a “knee”-like pattern, which can be used to identify the initial list of high-quality barcodes.
The red dots in the plot represent the cell barcodes selected as the high-quality cell barcodes in the case that “knee”-based filtering was applied.
In other words, these cell barcodes contain a sufficient number of reads to be deemed high-quality and likely derived from truly present cells.
Suppose an external permit list is passed in the CB correction step, which implies no internal algorithm was used to distinguish high-quality cell barcodes.
In that case, all dots in the plot will be colored red, as all these corrected cell barcodes are processed throughout the raw data processing pipeline and reported in the gene count matrix.
One should be skeptical of the data quality if the frequency is consistently low across all cell barcodes.</p>
<p><strong>3. Barcode collapsing</strong></p>
<p>After identification of the barcodes that will be processed, either through an internal threshold (e.g., from the “knee”-based method) or through external whitelisting, <code class="docutils literal notranslate"><span class="pre">alevin</span></code> (or <code class="docutils literal notranslate"><span class="pre">alevin-fry</span></code>) performs cell barcode sequence correction.
The barcode collapsing plot, the upper middle plot in the <a class="reference internal" href="#raw-proc-fig-alevinqc-plots"><span class="std std-numref">Fig. 3.16</span></a>, shows the number of reads assigned to a cell barcode after sequence correction of the cell barcodes versus prior to correction.
Generally, we would see that all points fall close to the line representing <span class="math notranslate nohighlight">\(x = y\)</span>, which means that the reassignments in CB correction usually do not drastically change the profile of the cell barcodes.</p>
<p><strong>4. Knee Plot, number of genes per cell</strong></p>
<p>The upper right plot in <a class="reference internal" href="#raw-proc-fig-alevinqc-plots"><span class="std std-numref">Fig. 3.16</span></a> shows the distribution of the number of observed genes of all processed cell barcodes.
Generally, a mean of <span class="math notranslate nohighlight">\(2,000\)</span> genes per cell is considered modest but reasonable for the downstream analyses.
One should double-check the quality of the data if all cells have a low number of observed genes.</p>
<p><strong>5. Quantification summary</strong></p>
<p>Finally, a series of quantification summary plots, the bottom plots in <a class="reference internal" href="#raw-proc-fig-alevinqc-plots"><span class="std std-numref">Fig. 3.16</span></a>, compare the cell barcode frequency, the total number of UMIs after deduplication and the total number of non-zero genes using scatter plots.
In general, in each plot, the plotted data should demonstrate a positive correlation, and, if high-quality filtering (e.g., knee filtering) has been performed, the high-quality cell barcodes should be well separated from the rest.
Moreover, one should expect all three plots to convey similar trends.
If using an external permit list, all the dots in the plots will be colored red, as all these cell barcodes are processed and reported in the gene count matrix.
Still, we should see the correlation between the plots and the separation of the dots representing high-quality cells to others.
If all of these metrics are consistently low across cells or if these plots convey substantially different trends, then one should be concerned about the data quality.</p>
</div>
<section id="empty-droplet-detection">
<h3><span class="section-number">3.6.1. </span>Empty droplet detection<a class="headerlink" href="#empty-droplet-detection" title="Link to this heading">#</a></h3>
<p>One of the first QC steps is determining which cell barcodes correspond to “high-confidence” sequenced cells.
It is common in droplet-based protocols <span id="id60">[<a class="reference internal" href="#id142" title="Evan Z. Macosko, Anindita Basu, Rahul Satija, James Nemesh, Karthik Shekhar, Melissa Goldman, Itay Tirosh, Allison R. Bialas, Nolan Kamitaki, Emily M. Martersteck, John J. Trombetta, David A. Weitz, Joshua R. Sanes, Alex K. Shalek, Aviv Regev, and Steven A. McCarroll. Highly parallel genome-wide expression profiling of individual cells using nanoliter droplets. Cell, 161(5):1202-1214, May 2015. URL: https://doi.org/10.1016/j.cell.2015.05.002, doi:10.1016/j.cell.2015.05.002.">Macosko <em>et al.</em>, 2015</a>]</span> that certain barcodes are associated with ambient <a class="reference internal" href="../glossary.html#term-RNA"><span class="xref std std-term">RNA</span></a> instead of the RNA of a captured cell.
This happens when droplets fail to capture a cell.
These empty droplets still tend to produce sequenced reads, although the characteristics of these reads look markedly different from those associated with barcodes corresponding to properly captured cells.
Many approaches exist to assess whether a barcode likely corresponds to an empty droplet or not.
One simple method is to examine the cumulative frequency plot of the barcodes, in which barcodes are sorted in descending order of the number of distinct UMIs with which they are associated.
This plot often contains a “knee” that can be identified as a likely point of discrimination between properly captured cells and empty droplets <span id="id61">[<a class="reference internal" href="#id110" title="Dongze He, Mohsen Zakeri, Hirak Sarkar, Charlotte Soneson, Avi Srivastava, and Rob Patro. Alevin-fry unlocks rapid, accurate and memory-frugal quantification of single-cell RNA-seq data. Nature Methods, 19(3):316–322, 2022.">He <em>et al.</em>, 2022</a>, <a class="reference internal" href="#id102" title="Tom Smith, Andreas Heger, and Ian Sudbery. UMI-tools: modeling sequencing errors in Unique Molecular Identifiers to improve quantification accuracy. Genome research, 27(3):491–499, 2017.">Smith <em>et al.</em>, 2017</a>]</span>.
While this “knee” method is intuitive and can often estimate a reasonable threshold, it has several drawbacks.
For example, not all cumulative histograms display an obvious knee, and it is notoriously difficult to design algorithms that can robustly and automatically detect such knees.
Finally, the total UMI count associated with a barcode may not, alone, be the best signal to determine if the barcode was associated with an empty or damaged cell.</p>
<p>This led to the development of several tools specifically designed to detect empty or damaged droplets, or cells generally deemed to be of “low quality” <span id="id62">[<a class="reference internal" href="#id119" title="Marcus Alvarez, Elior Rahmani, Brandon Jew, Kristina M. Garske, Zong Miao, Jihane N. Benhammou, Chun Jimmie Ye, Joseph R. Pisegna, Kirsi H. Pietiläinen, Eran Halperin, and Päivi Pajukanta. Enhancing droplet-based single-nucleus term`RNA`-seq resolution using the semi-supervised machine learning classifier DIEM. Scientific Reports, July 2020. URL: https://doi.org/10.1038/s41598-020-67513-5, doi:10.1038/s41598-020-67513-5.">Alvarez <em>et al.</em>, 2020</a>, <a class="reference internal" href="#id116" title="Cody N. Heiser, Victoria M. Wang, Bob Chen, Jacob J. Hughey, and Ken S. Lau. Automated quality control and cell identification of droplet-based single-cell data using dropkick. Genome Research, 31(10):1742–1752, April 2021. URL: https://doi.org/10.1101/gr.271908.120, doi:10.1101/gr.271908.120.">Heiser <em>et al.</em>, 2021</a>, <a class="reference internal" href="#id117" title="Ariel A Hippen, Matias M Falco, Lukas M Weber, Erdogan Pekcan Erkan, Kaiyang Zhang, Jennifer Anne Doherty, Anna Vähärautio, Casey S Greene, and Stephanie C Hicks. miQC: An adaptive probabilistic framework for quality control of single-cell term`RNA`-sequencing data. PLoS computational biology, 17(8):e1009290, 2021.">Hippen <em>et al.</em>, 2021</a>, <a class="reference internal" href="#id115" title="Aaron TL Lun, Samantha Riesenfeld, Tallulah Andrews, Tomas Gomes, John C Marioni, and others. EmptyDrops: distinguishing cells from empty droplets in droplet-based single-cell term`RNA` sequencing data. Genome biology, 20(1):1–9, 2019.">Lun <em>et al.</em>, 2019</a>, <a class="reference internal" href="#id118" title="Walter Muskovic and Joseph E Powell. DropletQC: improved identification of empty droplets and damaged cells in single-cell term`RNA`-seq data. Genome Biology, 22(1):1–9, 2021.">Muskovic and Powell, 2021</a>, <a class="reference internal" href="#id120" title="Matthew D Young and Sam Behjati. SoupX removes ambient term`RNA` contamination from droplet-based single-cell term`RNA` sequencing data. GigaScience, December 2020. URL: https://doi.org/10.1093/gigascience/giaa151, doi:10.1093/gigascience/giaa151.">Young and Behjati, 2020</a>]</span>.
These tools incorporate a variety of different measures of cell quality, including the frequencies of distinct UMIs, the number of detected genes, and the fraction of mitochondrial RNA, and typically work by applying a statistical model to these features to classify high-quality cells from putative empty droplets or damaged cells.
This means that cells can typically be scored, and a final filtering can be selected based on an estimated posterior probability that cells are not empty or compromised.
While these models generally work well for single-cell <a class="reference internal" href="../glossary.html#term-RNA"><span class="xref std std-term">RNA</span></a>-seq data, one may have to apply several additional filters or heuristics to obtain robust filtering in single-nucleus <a class="reference internal" href="../glossary.html#term-RNA"><span class="xref std std-term">RNA</span></a>-seq data <span id="id63">[<a class="reference internal" href="#id110" title="Dongze He, Mohsen Zakeri, Hirak Sarkar, Charlotte Soneson, Avi Srivastava, and Rob Patro. Alevin-fry unlocks rapid, accurate and memory-frugal quantification of single-cell RNA-seq data. Nature Methods, 19(3):316–322, 2022.">He <em>et al.</em>, 2022</a>, <a class="reference internal" href="#id107" title="Benjamin Kaminow, Dinar Yunusov, and Alexander Dobin. STARsolo: accurate, fast and versatile mapping/quantification of single-cell and single-nucleus term`RNA`-seq data. bioRxiv, 2021.">Kaminow <em>et al.</em>, 2021</a>]</span>, like those exposed in the <a class="reference external" href="https://github.com/MarioniLab/DropletUtils/blob/master/R/emptyDropsCellRanger.R"><code class="docutils literal notranslate"><span class="pre">emptyDropsCellRanger</span></code></a> function of <code class="docutils literal notranslate"><span class="pre">DropletUtils</span></code> <span id="id64">[<a class="reference internal" href="#id115" title="Aaron TL Lun, Samantha Riesenfeld, Tallulah Andrews, Tomas Gomes, John C Marioni, and others. EmptyDrops: distinguishing cells from empty droplets in droplet-based single-cell term`RNA` sequencing data. Genome biology, 20(1):1–9, 2019.">Lun <em>et al.</em>, 2019</a>]</span>.</p>
</section>
<section id="doublet-detection">
<h3><span class="section-number">3.6.2. </span>Doublet detection<a class="headerlink" href="#doublet-detection" title="Link to this heading">#</a></h3>
<p>In addition to determining which cell barcodes correspond to empty droplets or damaged cells, one may also wish to identify those cell barcodes that correspond to doublets or multiplets.
When a given droplet captures two (doublets) or more (multiplets) cells, this can result in a skewed distribution for these cell barcodes in terms of quantities like the number of reads and UMIs they represent, as well as gene expression profiles they display.
Many tools have also been developed to predict the doublet status of cell barcodes <span id="id65">[<a class="reference internal" href="#id124" title="Abha S Bais and Dennis Kostka. Scds: computational annotation of doublets in single-cell term`RNA` sequencing data. Bioinformatics, 36(4):1150–1158, September 2019. URL: https://doi.org/10.1093/bioinformatics/btz698, doi:10.1093/bioinformatics/btz698.">Bais and Kostka, 2019</a>, <a class="reference internal" href="#id137" title="Nicholas J. Bernstein, Nicole L. Fong, Irene Lam, Margaret A. Roy, David G. Hendrickson, and David R. Kelley. Solo: Doublet Identification in Single-Cell term`RNA`-Seq via Semi-Supervised Deep Learning. Cell Systems, 11(1):95–101.e5, July 2020. URL: https://doi.org/10.1016/j.cels.2020.05.010, doi:10.1016/j.cels.2020.05.010.">Bernstein <em>et al.</em>, 2020</a>, <a class="reference internal" href="#id121" title="Erica A.K. DePasquale, Daniel J. Schnell, Pieter-Jan Van Camp, Íñigo Valiente-Aland\'ı, Burns C. Blaxall, H. Leighton Grimes, Harinder Singh, and Nathan Salomonis. DoubletDecon: deconvoluting doublets from single-cell term`RNA`-sequencing data. Cell Reports, 29(6):1718–1727.e8, November 2019. URL: https://doi.org/10.1016/j.celrep.2019.09.082, doi:10.1016/j.celrep.2019.09.082.">DePasquale <em>et al.</em>, 2019</a>, <a class="reference internal" href="#id122" title="Christopher S. McGinnis, Lyndsay M. Murrow, and Zev J. Gartner. DoubletFinder: doublet detection in single-cell term`RNA` sequencing data using artificial nearest neighbors. Cell Systems, 8(4):329–337.e4, April 2019. URL: https://doi.org/10.1016/j.cels.2019.03.003, doi:10.1016/j.cels.2019.03.003.">McGinnis <em>et al.</em>, 2019</a>, <a class="reference internal" href="#id123" title="Samuel L. Wolock, Romain Lopez, and Allon M. Klein. Scrublet: Computational Identification of Cell Doublets in Single-Cell Transcriptomic Data. Cell Systems, 8(4):281–291.e9, April 2019. URL: https://doi.org/10.1016/j.cels.2018.11.005, doi:10.1016/j.cels.2018.11.005.">Wolock <em>et al.</em>, 2019</a>]</span>.
Once detected, cells determined to likely be doublets and multiplets can be removed or otherwise adjusted for in the subsequent analysis.</p>
</section>
</section>
<section id="count-data-representation">
<span id="raw-proc-output-representation"></span><h2><span class="section-number">3.7. </span>Count data representation<a class="headerlink" href="#count-data-representation" title="Link to this heading">#</a></h2>
<p>As one completes initial raw data processing and quality control and moves on to subsequent analyses, it is important to acknowledge and remember that the cell-by-gene count matrix is, at best, an approximation of the sequenced molecules in the original sample.
At several stages of the raw data processing pipeline, heuristics are applied, and simplifications are made to enable the generation of this count matrix.
For example, read mapping is imperfect, as is cell barcode correction.
Accurately resolving UMIs is particularly challenging, and issues related to UMIs attached to multimapping reads are often overlooked.
Additionally, multiple priming sites, especially in unspliced molecules, can violate the commonly assumed one molecule-to-one UMI relationship.</p>
</section>
<section id="brief-discussion">
<h2><span class="section-number">3.8. </span>Brief discussion<a class="headerlink" href="#brief-discussion" title="Link to this heading">#</a></h2>
<p>To close this chapter, we convey some observations and suggestions that have arisen from recent benchmarking and review studies surrounding some of the common preprocessing tools described above <span id="id66">[<a class="reference internal" href="#id95" title="Ralf Schulze Brüning, Lukas Tombor, Marcel H Schulz, Stefanie Dimmeler, and David John. Comparative analysis of common alignment tools for single-cell term`RNA` sequencing. GigaScience, 2022. URL: https://doi.org/10.1093%2Fgigascience%2Fgiac001, doi:10.1093/gigascience/giac001.">Brüning <em>et al.</em>, 2022</a>, <a class="reference internal" href="#id96" title="Yue You, Luyi Tian, Shian Su, Xueyi Dong, Jafar S. Jabbari, Peter F. Hickey, and Matthew E. Ritchie. Benchmarking UMI-based single-cell term`RNA`-seq preprocessing workflows. Genome Biology, dec 2021. URL: https://doi.org/10.1186%2Fs13059-021-02552-3, doi:10.1186/s13059-021-02552-3.">You <em>et al.</em>, 2021</a>]</span>.
It is, of course, important to note that the development of methods and tools for single-cell and single-nucleus RNA-seq raw data processing, as well as the continual evaluation of such methods, is an ongoing community effort.
It is therefore often useful and reasonable, when performing your own analyses, to experiment with several different tools.</p>
<p>At the coarsest level, the most common tools can process data robustly and accurately.
It has been suggested that with many common downstream analyses like clustering, and the methods used to perform them, the choice of preprocessing tool typically makes less difference than other steps in the analysis process <span id="id67">[<a class="reference internal" href="#id96" title="Yue You, Luyi Tian, Shian Su, Xueyi Dong, Jafar S. Jabbari, Peter F. Hickey, and Matthew E. Ritchie. Benchmarking UMI-based single-cell term`RNA`-seq preprocessing workflows. Genome Biology, dec 2021. URL: https://doi.org/10.1186%2Fs13059-021-02552-3, doi:10.1186/s13059-021-02552-3.">You <em>et al.</em>, 2021</a>]</span>.
Nonetheless, it has also been observed that applying lightweight mapping restricted to the spliced transcriptome can increase the probability of spurious mapping and gene expression <span id="id68">[<a class="reference internal" href="#id95" title="Ralf Schulze Brüning, Lukas Tombor, Marcel H Schulz, Stefanie Dimmeler, and David John. Comparative analysis of common alignment tools for single-cell term`RNA` sequencing. GigaScience, 2022. URL: https://doi.org/10.1093%2Fgigascience%2Fgiac001, doi:10.1093/gigascience/giac001.">Brüning <em>et al.</em>, 2022</a>]</span>.</p>
<p>Ultimately, the choice of a specific tool largely depends on the task at hand, and the constraints on available computational resources.
If performing a standard single-cell analysis, lightweight mapping-based methods are a good choice since they are faster (often considerably so) and more memory frugal than existing alignment-based tools.
If performing single-nucleus RNA-seq analysis, <code class="docutils literal notranslate"><span class="pre">alevin-fry</span></code> is an attractive option in particular, as it remains memory frugal and its index remains relatively small even as the transcriptome reference is expanded to include unspliced reference sequence.
On the other hand, alignment-based methods are recommended when recovering reads that map outside the (extended) transcriptome is important or when genomic mapping sites are required for downstream analyses.
This is particularly relevant for tasks such as differential transcript usage analysis using tools like <code class="docutils literal notranslate"><span class="pre">sierra</span></code> <span id="id69">[<a class="reference internal" href="#id94" title="Ralph Patrick, David T. Humphreys, Vaibhao Janbandhu, Alicia Oshlack, Joshua W.K. Ho, Richard P. Harvey, and Kitty K. Lo. Sierra: discovery of differential transcript usage from polyA-captured single-cell term`RNA`-seq data. Genome Biology, jul 2020. URL: https://doi.org/10.1186%2Fs13059-020-02071-7, doi:10.1186/s13059-020-02071-7.">Patrick <em>et al.</em>, 2020</a>]</span>.
Among the alignment-based pipelines, according to <span id="id70">Brüning <em>et al.</em> [<a class="reference internal" href="#id95" title="Ralf Schulze Brüning, Lukas Tombor, Marcel H Schulz, Stefanie Dimmeler, and David John. Comparative analysis of common alignment tools for single-cell term`RNA` sequencing. GigaScience, 2022. URL: https://doi.org/10.1093%2Fgigascience%2Fgiac001, doi:10.1093/gigascience/giac001.">2022</a>]</span>, <code class="docutils literal notranslate"><span class="pre">STARsolo</span></code> should be favored over <code class="docutils literal notranslate"><span class="pre">Cell</span> <span class="pre">Ranger</span></code> because the former is much faster than the latter, and requires less memory, while it is also capable of producing almost identical results.</p>
</section>
<section id="a-real-world-example">
<span id="raw-proc-example-workflow"></span><h2><span class="section-number">3.9. </span>A real-world example<a class="headerlink" href="#a-real-world-example" title="Link to this heading">#</a></h2>
<p>Given that we have covered the concepts underlying various approaches for raw data processing, we now turn our attention to demonstrating how a specific tool (in this case, <code class="docutils literal notranslate"><span class="pre">alevin-fry</span></code>) can be used to process a small example dataset.
To start, we need the sequenced reads from a single-cell experiment in <a class="reference external" href="https://en.wikipedia.org/wiki/FASTQ_format">FASTQ format</a> and the reference (e.g., transcriptome) against which the reads will be mapped.
Usually, a reference includes the genome sequences and the corresponding gene annotations of the sequenced species in the <a class="reference external" href="https://en.wikipedia.org/wiki/FASTA_format">FASTA</a> and <a class="reference external" href="https://useast.ensembl.org/info/website/upload/gff.html">GTF</a> format, respectively.</p>
<p>In this example, we will use <em>chromosome 5</em> of the human genome and its related gene annotations as the reference, which is a subset of the Human reference, <a class="reference external" href="https://support.10xgenomics.com/single-cell-gene-expression/software/release-notes/build#GRCh38_2020A">GRCh38 (GENCODE v32/Ensembl 98) reference</a> from the 10x Genomics reference build.
Correspondingly, we extract the subset of reads that map to the generated reference from a <a class="reference external" href="https://www.10xgenomics.com/resources/datasets/200-sorted-cells-from-human-glioblastoma-multiforme-3-lt-v-3-1-3-1-low-6-0-0">human brain tumor dataset</a> from 10x Genomics.</p>
<p><a class="reference external" href="https://alevin-fry.readthedocs.io/en/latest/"><code class="docutils literal notranslate"><span class="pre">Alevin-fry</span></code></a> <span id="id71">[<a class="reference internal" href="#id110" title="Dongze He, Mohsen Zakeri, Hirak Sarkar, Charlotte Soneson, Avi Srivastava, and Rob Patro. Alevin-fry unlocks rapid, accurate and memory-frugal quantification of single-cell RNA-seq data. Nature Methods, 19(3):316–322, 2022.">He <em>et al.</em>, 2022</a>]</span> is a fast, accurate, and memory-frugal single-cell and single-nucleus data processing tool.
<a class="reference external" href="https://github.com/COMBINE-lab/simpleaf">Simpleaf</a> is a program, written in <a class="reference external" href="https://www.rust-lang.org/">rust</a>, that exposes a unified and simplified interface for processing some of the most common protocols and data types using the <code class="docutils literal notranslate"><span class="pre">alevin-fry</span></code> pipeline.
A nextflow-based <a class="reference external" href="https://github.com/COMBINE-lab/quantaf">workflow</a> tool also exists to process extensive collections of single-cell data.
Here we will first show how to process single-cell raw data using two <code class="docutils literal notranslate"><span class="pre">simpleaf</span></code> commands. Then, we describe the complete set of <code class="docutils literal notranslate"><span class="pre">salmon</span> <span class="pre">alevin</span></code> and <code class="docutils literal notranslate"><span class="pre">alevin-fry</span></code> commands to which these <code class="docutils literal notranslate"><span class="pre">simpleaf</span></code> commands correspond, to outline where the steps described in this section occur and to convey the possible different processing options.
These commands will be run from the command line, and <a class="reference external" href="https://docs.conda.io/en/latest/"><code class="docutils literal notranslate"><span class="pre">conda</span></code></a> will be used for installing all of the software required for running this example.</p>
<section id="preparation">
<span id="raw-proc-example-prep"></span><h3><span class="section-number">3.9.1. </span>Preparation<a class="headerlink" href="#preparation" title="Link to this heading">#</a></h3>
<p>Before we start, we create a conda environment in the terminal and install the required package.
<code class="docutils literal notranslate"><span class="pre">Simpleaf</span></code> depends on <a class="reference external" href="https://alevin-fry.readthedocs.io/en/latest/"><code class="docutils literal notranslate"><span class="pre">alevin-fry</span></code></a>, <a class="reference external" href="https://salmon.readthedocs.io/en/latest/"><code class="docutils literal notranslate"><span class="pre">salmon</span></code></a> and <a class="reference external" href="https://github.com/COMBINE-lab/pyroe"><code class="docutils literal notranslate"><span class="pre">pyroe</span></code></a>.
They are all available on <code class="docutils literal notranslate"><span class="pre">bioconda</span></code> and will be automatically installed when installing <code class="docutils literal notranslate"><span class="pre">simpleaf</span></code>.</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>conda<span class="w"> </span>create<span class="w"> </span>-n<span class="w"> </span>af<span class="w"> </span>-y<span class="w"> </span>-c<span class="w"> </span>bioconda<span class="w"> </span>simpleaf
conda<span class="w"> </span>activate<span class="w"> </span>af
</pre></div>
</div>
<div class="admonition-note-on-using-an-apple-silicon-based-device admonition">
<p class="admonition-title">Note on using an Apple silicon-based device</p>
<p>Conda does not currently build most packages natively for Apple silicon.
Therefore, if you are using a non-Intel-based Apple computer (e.g., with an M1 (Pro/Max/Ultra) or M2 chip),
you should make sure to specify that your environment uses the Rosetta2 translation layer.
To do this, you can replace the above commands with the following (instructions adopted
from <a class="reference external" href="https://github.com/Haydnspass/miniforge#rosetta-on-mac-with-apple-silicon-hardware">here</a>):</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span><span class="nv">CONDA_SUBDIR</span><span class="o">=</span>osx-64<span class="w"> </span>conda<span class="w"> </span>create<span class="w"> </span>-n<span class="w"> </span>af<span class="w"> </span>-y<span class="w"> </span>-c<span class="w"> </span>bioconda<span class="w"> </span>simpleaf<span class="w">   </span><span class="c1"># create a new environment</span>
conda<span class="w"> </span>activate<span class="w"> </span>af
conda<span class="w"> </span>env<span class="w"> </span>config<span class="w"> </span>vars<span class="w"> </span><span class="nb">set</span><span class="w"> </span><span class="nv">CONDA_SUBDIR</span><span class="o">=</span>osx-64<span class="w">  </span><span class="c1"># subsequent commands use intel packages</span>
</pre></div>
</div>
</div>
<p>Next, we create a working directory, <code class="docutils literal notranslate"><span class="pre">af_xmpl_run</span></code>, and download and uncompress the example dataset from a remote host.</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span><span class="c1"># Create a working dir and go to the working directory</span>
<span class="c1">## The &amp;&amp; operator helps execute two commands using a single line of code.</span>
mkdir<span class="w"> </span>af_xmpl_run<span class="w"> </span><span class="o">&amp;&amp;</span><span class="w"> </span><span class="nb">cd</span><span class="w"> </span>af_xmpl_run

<span class="c1"># Fetch the example dataset and CB permit list and decompress them</span>
<span class="c1">## The pipe operator (|) passes the output of the wget command to the tar command.</span>
<span class="c1">## The dash operator (-) after `tar xzf` captures the output of the first command.</span>
<span class="c1">## - example dataset</span>
wget<span class="w"> </span>-qO-<span class="w"> </span>https://umd.box.com/shared/static/lx2xownlrhz3us8496tyu9c4dgade814.gz<span class="w"> </span><span class="p">|</span><span class="w"> </span>tar<span class="w"> </span>xzf<span class="w"> </span>-<span class="w"> </span>--strip-components<span class="o">=</span><span class="m">1</span><span class="w"> </span>-C<span class="w"> </span>.
<span class="c1">## The fetched folder containing the fastq files is called toy_read_fastq.</span>
<span class="nv">fastq_dir</span><span class="o">=</span><span class="s2">&quot;toy_read_fastq&quot;</span>
<span class="c1">## The fetched folder containing the human ref files is called toy_human_ref.</span>
<span class="nv">ref_dir</span><span class="o">=</span><span class="s2">&quot;toy_human_ref&quot;</span>

<span class="c1"># Fetch CB permit list</span>
<span class="c1">## the right chevron (&gt;) redirects the STDOUT to a file.</span>
wget<span class="w"> </span>-qO-<span class="w"> </span>https://github.com/f0t1h/3M-february-2018/raw/master/3M-february-2018.txt.gz<span class="w"> </span><span class="p">|</span><span class="w"> </span>gunzip<span class="w"> </span>-<span class="w"> </span>&gt;<span class="w"> </span>3M-february-2018.txt
</pre></div>
</div>
<p>With the reference files (the genome FASTA file and the gene annotation GTF file) and read records (the FASTQ files) ready, we can now apply the raw data processing pipeline discussed above to generate the gene count matrix.</p>
</section>
<section id="simplified-raw-data-processing-pipeline">
<span id="raw-proc-example-simpleaf"></span><h3><span class="section-number">3.9.2. </span>Simplified raw data processing pipeline<a class="headerlink" href="#simplified-raw-data-processing-pipeline" title="Link to this heading">#</a></h3>
<p><a class="reference external" href="https://github.com/COMBINE-lab/simpleaf">Simpleaf</a> is designed to simplify the <code class="docutils literal notranslate"><span class="pre">alevin-fry</span></code> interface for single-cell and nucleus raw data processing. It encapsulates the whole processing pipeline into two steps:</p>
<ol class="arabic simple">
<li><p><a class="reference external" href="https://simpleaf.readthedocs.io/en/latest/index-command.html"><code class="docutils literal notranslate"><span class="pre">simpleaf</span> <span class="pre">index</span></code></a> indexes the provided reference or makes a <em>splici</em> reference (<u>splic</u>ed transcripts + <u>i</u>ntrons) and index it.</p></li>
<li><p><a class="reference external" href="https://simpleaf.readthedocs.io/en/latest/quant-command.html"><code class="docutils literal notranslate"><span class="pre">simpleaf</span> <span class="pre">quant</span></code></a> maps the sequencing reads against the indexed reference and quantifies the mapping records to generate a gene count matrix.</p></li>
</ol>
<p>More advanced usages and options for mapping with <code class="docutils literal notranslate"><span class="pre">simpleaf</span></code> can be found <a class="reference external" href="https://simpleaf.readthedocs.io/en/latest/">here</a>.</p>
<p>When running <code class="docutils literal notranslate"><span class="pre">simpleaf</span> <span class="pre">index</span></code>, if a genome FASTA file (<code class="docutils literal notranslate"><span class="pre">-f</span></code>) and a gene annotation GTF file(<code class="docutils literal notranslate"><span class="pre">-g</span></code>) are provided, it will generate a <em>splici</em> reference and index it; if only a transcriptome FASTA file is provided (<code class="docutils literal notranslate"><span class="pre">--refseq</span></code>), it will directly index it. Currently, we recommend the <em>splici</em> index.</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span><span class="c1"># simpleaf needs the environment variable ALEVIN_FRY_HOME to store configuration and data.</span>
<span class="c1"># For example, the paths to the underlying programs it uses and the CB permit list</span>
mkdir<span class="w"> </span>alevin_fry_home<span class="w"> </span><span class="o">&amp;&amp;</span><span class="w"> </span><span class="nb">export</span><span class="w"> </span><span class="nv">ALEVIN_FRY_HOME</span><span class="o">=</span><span class="s1">&#39;alevin_fry_home&#39;</span>

<span class="c1"># the simpleaf set-paths command finds the path to the required tools and writes a configuration JSON file in the ALEVIN_FRY_HOME folder.</span>
simpleaf<span class="w"> </span>set-paths

<span class="c1"># simpleaf index</span>
<span class="c1"># Usage: simpleaf index -o out_dir [-f genome_fasta -g gene_annotation_GTF|--refseq transcriptome_fasta] -r read_length -t number_of_threads</span>
<span class="c1">## The -r read_lengh is the number of sequencing cycles performed by the sequencer to generate biological reads (read2 in Illumina).</span>
<span class="c1">## Publicly available datasets usually have the read length in the description. Sometimes they are called the number of cycles.</span>
simpleaf<span class="w"> </span>index<span class="w"> </span><span class="se">\</span>
-o<span class="w"> </span>simpleaf_index<span class="w"> </span><span class="se">\</span>
-f<span class="w"> </span>toy_human_ref/fasta/genome.fa<span class="w"> </span><span class="se">\</span>
-g<span class="w"> </span>toy_human_ref/genes/genes.gtf<span class="w"> </span><span class="se">\</span>
-r<span class="w"> </span><span class="m">90</span><span class="w"> </span><span class="se">\</span>
-t<span class="w"> </span><span class="m">8</span>
</pre></div>
</div>
<p>In the output directory <code class="docutils literal notranslate"><span class="pre">simpleaf_index</span></code>, the <code class="docutils literal notranslate"><span class="pre">ref</span></code> folder contains the <em>splici</em> reference; The <code class="docutils literal notranslate"><span class="pre">index</span></code> folder contains the salmon index built upon the <em>splici</em> reference.</p>
<p>The next step, <code class="docutils literal notranslate"><span class="pre">simpleaf</span> <span class="pre">quant</span></code>, consumes an index directory and the mapping record FASTQ files to generate a gene count matrix. This command encapsulates all the major steps discussed in this section, including mapping, cell barcode correction, and UMI resolution.</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span><span class="c1"># Collecting sequencing read files</span>
<span class="c1">## The reads1 and reads2 variables are defined by finding the filenames with the pattern &quot;_R1_&quot; and &quot;_R2_&quot; from the toy_read_fastq directory.</span>
<span class="nv">reads1_pat</span><span class="o">=</span><span class="s2">&quot;_R1_&quot;</span>
<span class="nv">reads2_pat</span><span class="o">=</span><span class="s2">&quot;_R2_&quot;</span>

<span class="c1">## The read files must be sorted and separated by a comma.</span>
<span class="c1">### The find command finds the files in the fastq_dir with the name pattern</span>
<span class="c1">### The sort command sorts the file names</span>
<span class="c1">### The awk command and the paste command together convert the file names into a comma-separated string.</span>
<span class="nv">reads1</span><span class="o">=</span><span class="s2">&quot;</span><span class="k">$(</span>find<span class="w"> </span>-L<span class="w"> </span><span class="si">${</span><span class="nv">fastq_dir</span><span class="si">}</span><span class="w"> </span>-name<span class="w"> </span><span class="s2">&quot;*</span><span class="nv">$reads1_pat</span><span class="s2">*&quot;</span><span class="w"> </span>-type<span class="w"> </span>f<span class="w"> </span><span class="p">|</span><span class="w"> </span>sort<span class="w"> </span><span class="p">|</span><span class="w"> </span>awk<span class="w"> </span>-v<span class="w"> </span><span class="nv">OFS</span><span class="o">=</span>,<span class="w"> </span><span class="s1">&#39;{$1=$1;print}&#39;</span><span class="w"> </span><span class="p">|</span><span class="w"> </span>paste<span class="w"> </span>-sd,<span class="k">)</span><span class="s2">&quot;</span>
<span class="nv">reads2</span><span class="o">=</span><span class="s2">&quot;</span><span class="k">$(</span>find<span class="w"> </span>-L<span class="w"> </span><span class="si">${</span><span class="nv">fastq_dir</span><span class="si">}</span><span class="w"> </span>-name<span class="w"> </span><span class="s2">&quot;*</span><span class="nv">$reads2_pat</span><span class="s2">*&quot;</span><span class="w"> </span>-type<span class="w"> </span>f<span class="w"> </span><span class="p">|</span><span class="w"> </span>sort<span class="w"> </span><span class="p">|</span><span class="w"> </span>awk<span class="w"> </span>-v<span class="w"> </span><span class="nv">OFS</span><span class="o">=</span>,<span class="w"> </span><span class="s1">&#39;{$1=$1;print}&#39;</span><span class="w"> </span><span class="p">|</span><span class="w"> </span>paste<span class="w"> </span>-sd,<span class="k">)</span><span class="s2">&quot;</span>

<span class="c1"># simpleaf quant</span>
<span class="c1">## Usage: simpleaf quant -c chemistry -t threads -1 reads1 -2 reads2 -i index -u [unspliced permit list] -r resolution -m t2g_3col -o output_dir</span>
simpleaf<span class="w"> </span>quant<span class="w"> </span><span class="se">\</span>
-c<span class="w"> </span>10xv3<span class="w"> </span>-t<span class="w"> </span><span class="m">8</span><span class="w"> </span><span class="se">\</span>
-1<span class="w"> </span><span class="nv">$reads1</span><span class="w"> </span>-2<span class="w"> </span><span class="nv">$reads2</span><span class="w"> </span><span class="se">\</span>
-i<span class="w"> </span>simpleaf_index/index<span class="w"> </span><span class="se">\</span>
-u<span class="w"> </span>-r<span class="w"> </span>cr-like<span class="w"> </span><span class="se">\</span>
-m<span class="w"> </span>simpleaf_index/index/t2g_3col.tsv<span class="w"> </span><span class="se">\</span>
-o<span class="w"> </span>simpleaf_quant
</pre></div>
</div>
<p>After running these commands, the resulting quantification information can be found in the <code class="docutils literal notranslate"><span class="pre">simpleaf_quant/af_quant/alevin</span></code> folder.
Within this directory, there are three files: <code class="docutils literal notranslate"><span class="pre">quants_mat.mtx</span></code>, <code class="docutils literal notranslate"><span class="pre">quants_mat_cols.txt</span></code>, and <code class="docutils literal notranslate"><span class="pre">quants_mat_rows.txt</span></code>, which correspond, respectively, to the count matrix, the gene names for each column of this matrix, and the corrected, filtered cell barcodes for each row of this matrix. The tail lines of these files are shown below.
Of note here is the fact that <code class="docutils literal notranslate"><span class="pre">alevin-fry</span></code> was run in the USA-mode (<u>u</u>nspliced, <u>s</u>pliced, and <u>a</u>mbiguous mode), and so quantification was performed for both the spliced and unspliced status of each gene — the resulting <code class="docutils literal notranslate"><span class="pre">quants_mat_cols.txt</span></code> file will then have a number of rows equal to 3 times the number of annotated genes which correspond, to the names used for the spliced (S), unspliced (U), and splicing-ambiguous variants (A) of each gene.</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span><span class="c1"># Each line in `quants_mat.mtx` represents</span>
<span class="c1"># a non-zero entry in the format row column entry</span>
$<span class="w"> </span>tail<span class="w"> </span>-3<span class="w"> </span>simpleaf_quant/af_quant/alevin/quants_mat.mtx
<span class="m">138</span><span class="w"> </span><span class="m">58</span><span class="w"> </span><span class="m">1</span>
<span class="m">139</span><span class="w"> </span><span class="m">9</span><span class="w"> </span><span class="m">1</span>
<span class="m">139</span><span class="w"> </span><span class="m">37</span><span class="w"> </span><span class="m">1</span>

<span class="c1"># Each line in `quants_mat_cols.txt` is a splice status</span>
<span class="c1"># of a gene in the format (gene name)-(splice status)</span>
$<span class="w"> </span>tail<span class="w"> </span>-3<span class="w"> </span>simpleaf_quant/af_quant/alevin/quants_mat_cols.txt
ENSG00000120705-A
ENSG00000198961-A
ENSG00000245526-A

<span class="c1"># Each line in `quants_mat_rows.txt` is a corrected</span>
<span class="c1"># (and, potentially, filtered) cell barcode</span>
$<span class="w"> </span>tail<span class="w"> </span>-3<span class="w"> </span>simpleaf_quant/af_quant/alevin/quants_mat_rows.txt
TTCGATTTCTGAATCG
TGCTCGTGTTCGAAGG
ACTGTGAAGAAATTGC
</pre></div>
</div>
<p>We can load the count matrix into Python as an <a class="reference external" href="https://anndata.readthedocs.io/en/latest/"><code class="docutils literal notranslate"><span class="pre">AnnData</span></code></a> object using the <code class="docutils literal notranslate"><span class="pre">load_fry</span></code> function from <a class="reference external" href="https://github.com/COMBINE-lab/pyroe"><code class="docutils literal notranslate"><span class="pre">pyroe</span></code></a>.
A similar function, <a class="reference external" href="https://rdrr.io/github/mikelove/fishpond/man/loadFry.html">loadFry</a>, has been implemented in the <a class="reference external" href="https://github.com/mikelove/fishpond"><code class="docutils literal notranslate"><span class="pre">fishpond</span></code></a> R package.</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span><span class="w"> </span><span class="nn">pyroe</span>

<span class="n">quant_dir</span> <span class="o">=</span> <span class="s1">&#39;simpleaf_quant/af_quant&#39;</span>
<span class="n">adata_sa</span> <span class="o">=</span> <span class="n">pyroe</span><span class="o">.</span><span class="n">load_fry</span><span class="p">(</span><span class="n">quant_dir</span><span class="p">)</span>
</pre></div>
</div>
<p>The default behavior loads the <code class="docutils literal notranslate"><span class="pre">X</span></code> layer of the <code class="docutils literal notranslate"><span class="pre">Anndata</span></code> object as the sum of the spliced and ambiguous counts for each gene.
However, recent work <span id="id72">[<a class="reference internal" href="#id134" title="Allan-Hermann Pool, Helen Poldsam, Sisi Chen, Matt Thomson, and Yuki Oka. Enhanced recovery of single-cell term`RNA`-sequencing reads for missing gene expression data. bioRxiv, 2022. URL: https://www.biorxiv.org/content/early/2022/04/27/2022.04.26.489449, arXiv:https://www.biorxiv.org/content/early/2022/04/27/2022.04.26.489449.full.pdf, doi:10.1101/2022.04.26.489449.">Pool <em>et al.</em>, 2022</a>]</span> and <a class="reference external" href="https://support.10xgenomics.com/single-cell-gene-expression/software/pipelines/latest/release-notes">updated practices</a> suggest that the inclusion of intronic counts, even in single-cell RNA-seq data, may increase sensitivity and benefit downstream analyses.
While the best way to make use of this information is the subject of ongoing research, since <code class="docutils literal notranslate"><span class="pre">alevin-fry</span></code> automatically quantifies spliced, unspliced, and ambiguous reads in each sample, the count matrix containing the total counts for each gene can be simply obtained as follows:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span><span class="w"> </span><span class="nn">pyroe</span>

<span class="n">quant_dir</span> <span class="o">=</span> <span class="s1">&#39;simpleaf_quant/af_quant&#39;</span>
<span class="n">adata_usa</span> <span class="o">=</span> <span class="n">pyroe</span><span class="o">.</span><span class="n">load_fry</span><span class="p">(</span><span class="n">quant_dir</span><span class="p">,</span> <span class="n">output_format</span><span class="o">=</span><span class="p">{</span><span class="s1">&#39;X&#39;</span> <span class="p">:</span> <span class="p">[</span><span class="s1">&#39;U&#39;</span><span class="p">,</span><span class="s1">&#39;S&#39;</span><span class="p">,</span><span class="s1">&#39;A&#39;</span><span class="p">]})</span>
</pre></div>
</div>
</section>
<section id="the-complete-alevin-fry-pipeline">
<span id="raw-proc-example-map"></span><h3><span class="section-number">3.9.3. </span>The complete alevin-fry pipeline<a class="headerlink" href="#the-complete-alevin-fry-pipeline" title="Link to this heading">#</a></h3>
<p><code class="docutils literal notranslate"><span class="pre">Simpleaf</span></code> makes it possible to process single-cell raw data in the “standard” way with a few commands.
Next, we will show how to generate the identical quantification result by explicitly calling the <code class="docutils literal notranslate"><span class="pre">pyroe</span></code>, <code class="docutils literal notranslate"><span class="pre">salmon</span></code>, and <code class="docutils literal notranslate"><span class="pre">alevin-fry</span></code> commands.
On top of the pedagogical value, knowing the exact command of each step will be helpful if only a part of the pipeline needs to be rerun or if some parameters not currently exposed by <code class="docutils literal notranslate"><span class="pre">simpleaf</span></code> need to be specified.</p>
<p>Please note that the commands in the <a class="reference internal" href="#raw-proc-example-prep"><span class="std std-ref">Preparation</span></a> section should be executed in advance.
All the tools called in the following commands, <code class="docutils literal notranslate"><span class="pre">pyroe</span></code>, <code class="docutils literal notranslate"><span class="pre">salmon</span></code>, and <code class="docutils literal notranslate"><span class="pre">alevin-fry</span></code>, have already been installed when installing <code class="docutils literal notranslate"><span class="pre">simpleaf</span></code>.</p>
<section id="building-the-index">
<h4><span class="section-number">3.9.3.1. </span>Building the index<a class="headerlink" href="#building-the-index" title="Link to this heading">#</a></h4>
<p>First, we process the genome FASTA file and gene annotation GTF file to obtain the <em>splici</em> index.
The commands in the following code chunk are analogous to the <code class="docutils literal notranslate"><span class="pre">simpleaf</span> <span class="pre">index</span></code> command discussed above. This includes two steps:</p>
<ol class="arabic simple">
<li><p>Building the <em>splici</em> reference (<u>splic</u>ed transcripts + <u>i</u>ntrons) by calling <code class="docutils literal notranslate"><span class="pre">pyroe</span> <span class="pre">make-splici</span></code>, using the genome and gene annotation file</p></li>
<li><p>Indexing the <em>splici</em> reference by calling <code class="docutils literal notranslate"><span class="pre">salmon</span> <span class="pre">index</span></code></p></li>
</ol>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span><span class="c1"># make splici reference</span>
<span class="c1">## Usage: pyroe make-splici genome_file gtf_file read_length out_dir</span>
<span class="c1">## The read_lengh is the number of sequencing cycles performed by the sequencer. Ask your technician if you are not sure about it.</span>
<span class="c1">## Publicly available datasets usually have the read length in the description.</span>
pyroe<span class="w"> </span>make-splici<span class="w"> </span><span class="se">\</span>
<span class="si">${</span><span class="nv">ref_dir</span><span class="si">}</span>/fasta/genome.fa<span class="w"> </span><span class="se">\</span>
<span class="si">${</span><span class="nv">ref_dir</span><span class="si">}</span>/genes/genes.gtf<span class="w"> </span><span class="se">\</span>
<span class="m">90</span><span class="w"> </span><span class="se">\</span>
splici_rl90_ref

<span class="c1"># Index the reference</span>
<span class="c1">## Usage: salmon index -t extend_txome.fa -i idx_out_dir -p num_threads</span>
<span class="c1">## The $() expression runs the command inside and puts the output in place.</span>
<span class="c1">## Please ensure that only one file ends with &quot;.fa&quot; in the `splici_ref` folder.</span>
salmon<span class="w"> </span>index<span class="w"> </span><span class="se">\</span>
-t<span class="w"> </span><span class="k">$(</span>ls<span class="w"> </span>splici_rl90_ref/*<span class="se">\.</span>fa<span class="k">)</span><span class="w"> </span><span class="se">\</span>
-i<span class="w"> </span>salmon_index<span class="w"> </span><span class="se">\</span>
-p<span class="w"> </span><span class="m">8</span>
</pre></div>
</div>
<p>The <em>splici</em> index can be found in the <code class="docutils literal notranslate"><span class="pre">salmon_index</span></code> directory.</p>
</section>
<section id="mapping-and-quantification">
<span id="raw-proc-example-quant"></span><h4><span class="section-number">3.9.3.2. </span>Mapping and quantification<a class="headerlink" href="#mapping-and-quantification" title="Link to this heading">#</a></h4>
<p>Next, we will map the sequencing reads recorded against the <em>splici</em> index by calling <a class="reference external" href="https://salmon.readthedocs.io/en/latest/alevin.html"><code class="docutils literal notranslate"><span class="pre">salmon</span> <span class="pre">alevin</span></code></a>. This will produce an output folder called <code class="docutils literal notranslate"><span class="pre">salmon_alevin</span></code> that contains all the information we need to process the mapped reads using <code class="docutils literal notranslate"><span class="pre">alevin-fry</span></code>.</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span><span class="c1"># Collect FASTQ files</span>
<span class="c1">## The filenames are sorted and separated by space.</span>
<span class="nv">reads1</span><span class="o">=</span><span class="s2">&quot;</span><span class="k">$(</span>find<span class="w"> </span>-L<span class="w"> </span><span class="nv">$fastq_dir</span><span class="w"> </span>-name<span class="w"> </span><span class="s2">&quot;*</span><span class="nv">$reads1_pat</span><span class="s2">*&quot;</span><span class="w"> </span>-type<span class="w"> </span>f<span class="w"> </span><span class="p">|</span><span class="w"> </span>sort<span class="w"> </span><span class="p">|</span><span class="w"> </span>awk<span class="w"> </span><span class="s1">&#39;{$1=$1;print}&#39;</span><span class="w"> </span><span class="p">|</span><span class="w"> </span>paste<span class="w"> </span>-sd<span class="s1">&#39; &#39;</span><span class="k">)</span><span class="s2">&quot;</span>
<span class="nv">reads2</span><span class="o">=</span><span class="s2">&quot;</span><span class="k">$(</span>find<span class="w"> </span>-L<span class="w"> </span><span class="nv">$fastq_dir</span><span class="w"> </span>-name<span class="w"> </span><span class="s2">&quot;*</span><span class="nv">$reads2_pat</span><span class="s2">*&quot;</span><span class="w"> </span>-type<span class="w"> </span>f<span class="w"> </span><span class="p">|</span><span class="w"> </span>sort<span class="w"> </span><span class="p">|</span><span class="w"> </span>awk<span class="w"> </span><span class="s1">&#39;{$1=$1;print}&#39;</span><span class="w"> </span><span class="p">|</span><span class="w"> </span>paste<span class="w"> </span>-sd<span class="s1">&#39; &#39;</span><span class="k">)</span><span class="s2">&quot;</span>

<span class="c1"># Mapping</span>
<span class="c1">## Usage: salmon alevin -i index_dir -l library_type -1 reads1_files -2 reads2_files -p num_threads -o output_dir</span>
<span class="c1">## The variable reads1 and reads2 defined above are passed in using ${}.</span>
salmon<span class="w"> </span>alevin<span class="w"> </span><span class="se">\</span>
-i<span class="w"> </span>salmon_index<span class="w"> </span><span class="se">\</span>
-l<span class="w"> </span>ISR<span class="w"> </span><span class="se">\</span>
-1<span class="w"> </span><span class="si">${</span><span class="nv">reads1</span><span class="si">}</span><span class="w"> </span><span class="se">\</span>
-2<span class="w"> </span><span class="si">${</span><span class="nv">reads2</span><span class="si">}</span><span class="w"> </span><span class="se">\</span>
-p<span class="w"> </span><span class="m">8</span><span class="w"> </span><span class="se">\</span>
-o<span class="w"> </span>salmon_alevin<span class="w"> </span><span class="se">\</span>
--chromiumV3<span class="w"> </span><span class="se">\</span>
--sketch
</pre></div>
</div>
<p>Then, we execute the cell barcode correction and UMI resolution step using <code class="docutils literal notranslate"><span class="pre">alevin-fry</span></code>. This procedure involves three <code class="docutils literal notranslate"><span class="pre">alevin-fry</span></code> commands:</p>
<ol class="arabic simple">
<li><p>The <a class="reference external" href="https://alevin-fry.readthedocs.io/en/latest/generate_permit_list.html"><code class="docutils literal notranslate"><span class="pre">generate-permit-list</span></code></a> command is used for cell barcode correction.</p></li>
<li><p>The <a class="reference external" href="https://alevin-fry.readthedocs.io/en/latest/collate.html"><code class="docutils literal notranslate"><span class="pre">collate</span></code></a> command filters out invalid mapping records, corrects cell barcodes and collates mapping records originating from the same corrected cell barcode.</p></li>
<li><p>The <a class="reference external" href="https://alevin-fry.readthedocs.io/en/latest/quant.html"><code class="docutils literal notranslate"><span class="pre">quant</span></code></a> command performs UMI resolution and quantification.</p></li>
</ol>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span><span class="c1"># Cell barcode correction</span>
<span class="c1">## Usage: alevin-fry generate-permit-list -u CB_permit_list -d expected_orientation -o gpl_out_dir</span>
<span class="c1">## Here, the reads that map to the reverse complement strand of transcripts are filtered out by specifying `-d fw`.</span>
alevin-fry<span class="w"> </span>generate-permit-list<span class="w"> </span><span class="se">\</span>
-u<span class="w"> </span>3M-february-2018.txt<span class="w"> </span><span class="se">\</span>
-d<span class="w"> </span>fw<span class="w"> </span><span class="se">\</span>
-i<span class="w"> </span>salmon_alevin<span class="w"> </span><span class="se">\</span>
-o<span class="w"> </span>alevin_fry_gpl

<span class="c1"># Filter mapping information</span>
<span class="c1">## Usage: alevin-fry collate -i gpl_out_dir -r alevin_map_dir -t num_threads</span>
alevin-fry<span class="w"> </span>collate<span class="w"> </span><span class="se">\</span>
-i<span class="w"> </span>alevin_fry_gpl<span class="w"> </span><span class="se">\</span>
-r<span class="w"> </span>salmon_alevin<span class="w"> </span><span class="se">\</span>
-t<span class="w"> </span><span class="m">8</span>

<span class="c1"># UMI resolution + quantification</span>
<span class="c1">## Usage: alevin-fry quant -r resolution -m txp_to_gene_mapping -i gpl_out_dir -o quant_out_dir -t num_threads</span>
<span class="c1">## The file ends with `3col.tsv` in the splici_ref folder will be passed to the -m argument.</span>
<span class="c1">## Please ensure that there is only one such file in the `splici_ref` folder.</span>
alevin-fry<span class="w"> </span>quant<span class="w"> </span>-r<span class="w"> </span>cr-like<span class="w"> </span><span class="se">\</span>
-m<span class="w"> </span><span class="k">$(</span>ls<span class="w"> </span>splici_rl90_ref/*3col.tsv<span class="k">)</span><span class="w"> </span><span class="se">\</span>
-i<span class="w"> </span>alevin_fry_gpl<span class="w"> </span><span class="se">\</span>
-o<span class="w"> </span>alevin_fry_quant<span class="w"> </span><span class="se">\</span>
-t<span class="w"> </span><span class="m">8</span>
</pre></div>
</div>
<p>After running these commands, the resulting quantification information can be found in <code class="docutils literal notranslate"><span class="pre">alevin_fry_quant/alevin</span></code>.
Other relevant information concerning the mapping, CB correction, and UMI resolution steps can be found in the <code class="docutils literal notranslate"><span class="pre">salmon_alevin</span></code>, <code class="docutils literal notranslate"><span class="pre">alevin_fry_gpl</span></code>, and <code class="docutils literal notranslate"><span class="pre">alevin_fry_quant</span></code> folders, respectively.</p>
<p>In the example given here, we demonstrate using <code class="docutils literal notranslate"><span class="pre">simpleaf</span></code> and <code class="docutils literal notranslate"><span class="pre">alevin-fry</span></code> to process a 10x Chromium 3’ v3 dataset.
<code class="docutils literal notranslate"><span class="pre">Alevin-fry</span></code> and <code class="docutils literal notranslate"><span class="pre">simpleaf</span></code> provide many other options for processing different single-cell protocols, including but not limited to Dropseq <span id="id73">[<a class="reference internal" href="#id142" title="Evan Z. Macosko, Anindita Basu, Rahul Satija, James Nemesh, Karthik Shekhar, Melissa Goldman, Itay Tirosh, Allison R. Bialas, Nolan Kamitaki, Emily M. Martersteck, John J. Trombetta, David A. Weitz, Joshua R. Sanes, Alex K. Shalek, Aviv Regev, and Steven A. McCarroll. Highly parallel genome-wide expression profiling of individual cells using nanoliter droplets. Cell, 161(5):1202-1214, May 2015. URL: https://doi.org/10.1016/j.cell.2015.05.002, doi:10.1016/j.cell.2015.05.002.">Macosko <em>et al.</em>, 2015</a>]</span>, sci-RNA-seq3 <span id="id74">[<a class="reference internal" href="#id140" title="Junyue Cao, Malte Spielmann, Xiaojie Qiu, Xingfan Huang, Daniel M. Ibrahim, Andrew J. Hill, Fan Zhang, Stefan Mundlos, Lena Christiansen, Frank J. Steemers, Cole Trapnell, and Jay Shendure. The single-cell transcriptional landscape of mammalian organogenesis. Nature, 566(7745):496–502, February 2019. URL: https://doi.org/10.1038/s41586-019-0969-x, doi:10.1038/s41586-019-0969-x.">Cao <em>et al.</em>, 2019</a>]</span> and other 10x Chromium platforms.
A more comprehensive list and description of available options for different stages of processing can be found in the <a class="reference external" href="https://alevin-fry.readthedocs.io/en/latest/"><code class="docutils literal notranslate"><span class="pre">alevin-fry</span></code></a> and <a class="reference external" href="https://github.com/COMBINE-lab/simpleaf"><code class="docutils literal notranslate"><span class="pre">simpleaf</span></code></a> documentation.
<code class="docutils literal notranslate"><span class="pre">alevin-fry</span></code> also provides a <a class="reference external" href="https://www.nextflow.io/docs/latest/">nextflow</a>-based workflow, called <a class="reference external" href="https://github.com/COMBINE-lab/quantaf">quantaf</a>, for conveniently processing many samples from a simply-defined sample sheet.</p>
<p>Of course, similar resources exist for many of the other raw data processing tools referenced and described throughout this section, including <a class="reference external" href="https://github.com/sdparekh/zUMIs/wiki"><code class="docutils literal notranslate"><span class="pre">zUMIs</span></code></a> <span id="id75">[<a class="reference internal" href="#id113" title="Swati Parekh, Christoph Ziegenhain, Beate Vieth, Wolfgang Enard, and Ines Hellmann. zUMIs - a fast and flexible pipeline to process term`RNA` sequencing data with UMIs. GigaScience, May 2018. URL: https://doi.org/10.1093/gigascience/giy059, doi:10.1093/gigascience/giy059.">Parekh <em>et al.</em>, 2018</a>]</span>, <a class="reference external" href="https://salmon.readthedocs.io/en/latest/alevin.html"><code class="docutils literal notranslate"><span class="pre">alevin</span></code></a> <span id="id76">[<a class="reference internal" href="#id103" title="Avi Srivastava, Laraib Malik, Tom Smith, Ian Sudbery, and Rob Patro. Alevin efficiently estimates accurate gene abundances from dscRNA-seq data. Genome biology, 20(1):1–16, 2019.">Srivastava <em>et al.</em>, 2019</a>]</span>, <a class="reference external" href="https://www.kallistobus.tools/"><code class="docutils literal notranslate"><span class="pre">kallisto|bustools</span></code></a> <span id="id77">[<a class="reference internal" href="#id108" title="Páll Melsted, A. Sina Booeshaghi, Lauren Liu, Fan Gao, Lambda Lu, Kyung Hoi Min, Eduardo da Veiga Beltrame, Kristján Eldjárn Hjörleifsson, Jase Gehring, and Lior Pachter. Modular, efficient and constant-memory single-cell rna-seq preprocessing. Nature Biotechnology, 39(7):813–818, April 2021. URL: https://doi.org/10.1038/s41587-021-00870-2, doi:10.1038/s41587-021-00870-2.">Melsted <em>et al.</em>, 2021</a>]</span>, <a class="reference external" href="https://github.com/alexdobin/STAR/blob/master/docs/STARsolo.md"><code class="docutils literal notranslate"><span class="pre">STARsolo</span></code></a> <span id="id78">[<a class="reference internal" href="#id107" title="Benjamin Kaminow, Dinar Yunusov, and Alexander Dobin. STARsolo: accurate, fast and versatile mapping/quantification of single-cell and single-nucleus term`RNA`-seq data. bioRxiv, 2021.">Kaminow <em>et al.</em>, 2021</a>]</span> and <a class="reference external" href="https://support.10xgenomics.com/single-cell-gene-expression/software/pipelines/latest/what-is-cell-ranger"><code class="docutils literal notranslate"><span class="pre">CellRanger</span></code></a>.
The <a class="reference external" href="https://nf-co.re/scrnaseq"><code class="docutils literal notranslate"><span class="pre">scrnaseq</span></code></a> pipeline from <a class="reference external" href="https://nf-co.re/"><code class="docutils literal notranslate"><span class="pre">nf-core</span></code></a> also provides a nextflow-based pipeline for processing single-cell RNA-seq data generated using a range of different chemistries and integrates several of the tools described in this section.</p>
</section>
</section>
</section>
<section id="useful-links">
<span id="raw-proc-useful-links"></span><h2><span class="section-number">3.10. </span>Useful links<a class="headerlink" href="#useful-links" title="Link to this heading">#</a></h2>
<p><a class="reference external" href="https://combine-lab.github.io/alevin-fry-tutorials/">Alevin-fry tutorials</a> provide tutorials for processing different types of data.</p>
<p><a class="reference external" href="https://github.com/COMBINE-lab/pyroe"><code class="docutils literal notranslate"><span class="pre">Pyroe</span></code></a> in python and <a class="reference external" href="https://github.com/COMBINE-lab/roe"><code class="docutils literal notranslate"><span class="pre">roe</span></code></a> in R provide helper functions for processing <code class="docutils literal notranslate"><span class="pre">alevin-fry</span></code> quantification information. They also provide an interface to the preprocessed datasets in <a class="reference external" href="https://combine-lab.github.io/quantaf"><code class="docutils literal notranslate"><span class="pre">quantaf</span></code></a>.</p>
<p><a class="reference external" href="https://github.com/COMBINE-lab/quantaf"><code class="docutils literal notranslate"><span class="pre">Quantaf</span></code></a> is a nextflow-based workflow of the <code class="docutils literal notranslate"><span class="pre">alevin-fry</span></code> pipeline for conveniently processing a large number of single-cell and single-nucleus data based on the input sheets. The preprocessed quantification information of publicly available single-cell datasets is available on its <a class="reference external" href="https://combine-lab.github.io/quantaf">webpage</a>.</p>
<p><a class="reference external" href="https://github.com/COMBINE-lab/simpleaf"><code class="docutils literal notranslate"><span class="pre">Simpleaf</span></code></a> is a wrapper of the alevin-fry workflow that allows executing the whole pipeline, from making <em>splici</em> reference to quantification as shown in the above example, using only two commands.</p>
<p>Tutorials for processing scRNA-seq raw data from <a class="reference external" href="https://galaxyproject.org/">the galaxy project</a> can be found at <a class="reference external" href="https://training.galaxyproject.org/training-material/topics/transcriptomics/tutorials/scrna-preprocessing-tenx/tutorial.html">here</a> and <a class="reference external" href="https://training.galaxyproject.org/training-material/topics/transcriptomics/tutorials/scrna-preprocessing/tutorial.html">here</a>.</p>
<p>Tutorials for explaining and evaluating FastQC report are available from <a class="reference external" href="https://rtsf.natsci.msu.edu/genomics/technical-documents/fastqc-tutorial-and-faq.aspx">MSU</a>, <a class="reference external" href="https://hbctraining.github.io/Intro-to-rnaseq-hpc-salmon/lessons/qc_fastqc_assessment.html">the HBC training program</a>, <a class="reference external" href="https://training.galaxyproject.org/training-material/topics/sequence-analysis/tutorials/quality-control/tutorial.html">Galaxy Training</a> and <a class="reference external" href="https://sequencing.qcfail.com/software/fastqc/">the QC Fail website</a>.</p>
</section>
<section id="references">
<span id="raw-proc-references"></span><h2><span class="section-number">3.11. </span>References<a class="headerlink" href="#references" title="Link to this heading">#</a></h2>
<div class="docutils container" id="id79">
<div role="list" class="citation-list">
<div class="citation" id="id119" role="doc-biblioentry">
<span class="label"><span class="fn-bracket">[</span><a role="doc-backlink" href="#id62">rawARJ+20</a><span class="fn-bracket">]</span></span>
<p>Marcus Alvarez, Elior Rahmani, Brandon Jew, Kristina M. Garske, Zong Miao, Jihane N. Benhammou, Chun Jimmie Ye, Joseph R. Pisegna, Kirsi H. Pietiläinen, Eran Halperin, and Päivi Pajukanta. Enhancing droplet-based single-nucleus term`RNA`-seq resolution using the semi-supervised machine learning classifier DIEM. <em>Scientific Reports</em>, July 2020. URL: <a class="reference external" href="https://doi.org/10.1038/s41598-020-67513-5">https://doi.org/10.1038/s41598-020-67513-5</a>, <a class="reference external" href="https://doi.org/10.1038/s41598-020-67513-5">doi:10.1038/s41598-020-67513-5</a>.</p>
</div>
<div class="citation" id="id124" role="doc-biblioentry">
<span class="label"><span class="fn-bracket">[</span>rawBK19<span class="fn-bracket">]</span></span>
<span class="backrefs">(<a role="doc-backlink" href="#id36">1</a>,<a role="doc-backlink" href="#id65">2</a>)</span>
<p>Abha S Bais and Dennis Kostka. Scds: computational annotation of doublets in single-cell term`RNA` sequencing data. <em>Bioinformatics</em>, 36(4):1150–1158, September 2019. URL: <a class="reference external" href="https://doi.org/10.1093/bioinformatics/btz698">https://doi.org/10.1093/bioinformatics/btz698</a>, <a class="reference external" href="https://doi.org/10.1093/bioinformatics/btz698">doi:10.1093/bioinformatics/btz698</a>.</p>
</div>
<div class="citation" id="id137" role="doc-biblioentry">
<span class="label"><span class="fn-bracket">[</span><a role="doc-backlink" href="#id65">rawBFL+20</a><span class="fn-bracket">]</span></span>
<p>Nicholas J. Bernstein, Nicole L. Fong, Irene Lam, Margaret A. Roy, David G. Hendrickson, and David R. Kelley. Solo: Doublet Identification in Single-Cell term`RNA`-Seq via Semi-Supervised Deep Learning. <em>Cell Systems</em>, 11(1):95–101.e5, July 2020. URL: <a class="reference external" href="https://doi.org/10.1016/j.cels.2020.05.010">https://doi.org/10.1016/j.cels.2020.05.010</a>, <a class="reference external" href="https://doi.org/10.1016/j.cels.2020.05.010">doi:10.1016/j.cels.2020.05.010</a>.</p>
</div>
<div class="citation" id="id106" role="doc-biblioentry">
<span class="label"><span class="fn-bracket">[</span><a role="doc-backlink" href="#id45">rawBWC+15</a><span class="fn-bracket">]</span></span>
<p>Sayantan Bose, Zhenmao Wan, Ambrose Carr, Abbas H. Rizvi, Gregory Vieira, Dana Pe'er, and Peter A. Sims. Scalable microfluidics for single-cell term`RNA` printing and sequencing. <em>Genome Biology</em>, June 2015. URL: <a class="reference external" href="https://doi.org/10.1186/s13059-015-0684-3">https://doi.org/10.1186/s13059-015-0684-3</a>, <a class="reference external" href="https://doi.org/10.1186/s13059-015-0684-3">doi:10.1186/s13059-015-0684-3</a>.</p>
</div>
<div class="citation" id="id127" role="doc-biblioentry">
<span class="label"><span class="fn-bracket">[</span><a role="doc-backlink" href="#id22">rawBPMP16</a><span class="fn-bracket">]</span></span>
<p>Nicolas L Bray, Harold Pimentel, Páll Melsted, and Lior Pachter. Near-optimal probabilistic term`RNA`-seq quantification. <em>Nature biotechnology</em>, 34(5):525–527, 2016.</p>
</div>
<div class="citation" id="id133" role="doc-biblioentry">
<span class="label"><span class="fn-bracket">[</span><a role="doc-backlink" href="#id30">rawBruningTS+22</a><span class="fn-bracket">]</span></span>
<p>Ralf Schulze Brüning, Lukas Tombor, Marcel H Schulz, Stefanie Dimmeler, and David John. Comparative analysis of common alignment tools for single-cell term`RNA` sequencing. <em>GigaScience</em>, 2022.</p>
</div>
<div class="citation" id="id95" role="doc-biblioentry">
<span class="label"><span class="fn-bracket">[</span>rawBTS+22<span class="fn-bracket">]</span></span>
<span class="backrefs">(<a role="doc-backlink" href="#id66">1</a>,<a role="doc-backlink" href="#id68">2</a>,<a role="doc-backlink" href="#id70">3</a>)</span>
<p>Ralf Schulze Brüning, Lukas Tombor, Marcel H Schulz, Stefanie Dimmeler, and David John. Comparative analysis of common alignment tools for single-cell term`RNA` sequencing. <em>GigaScience</em>, 2022. URL: <a class="reference external" href="https://doi.org/10.1093%2Fgigascience%2Fgiac001">https://doi.org/10.1093%2Fgigascience%2Fgiac001</a>, <a class="reference external" href="https://doi.org/10.1093/gigascience/giac001">doi:10.1093/gigascience/giac001</a>.</p>
</div>
<div class="citation" id="id140" role="doc-biblioentry">
<span class="label"><span class="fn-bracket">[</span><a role="doc-backlink" href="#id74">rawCSQ+19</a><span class="fn-bracket">]</span></span>
<p>Junyue Cao, Malte Spielmann, Xiaojie Qiu, Xingfan Huang, Daniel M. Ibrahim, Andrew J. Hill, Fan Zhang, Stefan Mundlos, Lena Christiansen, Frank J. Steemers, Cole Trapnell, and Jay Shendure. The single-cell transcriptional landscape of mammalian organogenesis. <em>Nature</em>, 566(7745):496–502, February 2019. URL: <a class="reference external" href="https://doi.org/10.1038/s41586-019-0969-x">https://doi.org/10.1038/s41586-019-0969-x</a>, <a class="reference external" href="https://doi.org/10.1038/s41586-019-0969-x">doi:10.1038/s41586-019-0969-x</a>.</p>
</div>
<div class="citation" id="id88" role="doc-biblioentry">
<span class="label"><span class="fn-bracket">[</span><a role="doc-backlink" href="#id12">rawCPM92</a><span class="fn-bracket">]</span></span>
<p>Kun-Mao Chao, William R. Pearson, and Webb Miller. Aligning two sequences within a specified diagonal band. <em>Bioinformatics</em>, 8(5):481–487, 1992. URL: <a class="reference external" href="https://doi.org/10.1093/bioinformatics/8.5.481">https://doi.org/10.1093/bioinformatics/8.5.481</a>, <a class="reference external" href="https://doi.org/10.1093/bioinformatics/8.5.481">doi:10.1093/bioinformatics/8.5.481</a>.</p>
</div>
<div class="citation" id="id121" role="doc-biblioentry">
<span class="label"><span class="fn-bracket">[</span>rawDSC+19<span class="fn-bracket">]</span></span>
<span class="backrefs">(<a role="doc-backlink" href="#id36">1</a>,<a role="doc-backlink" href="#id65">2</a>)</span>
<p>Erica A.K. DePasquale, Daniel J. Schnell, Pieter-Jan Van Camp, Íñigo Valiente-Aland\'ı, Burns C. Blaxall, H. Leighton Grimes, Harinder Singh, and Nathan Salomonis. DoubletDecon: deconvoluting doublets from single-cell term`RNA`-sequencing data. <em>Cell Reports</em>, 29(6):1718–1727.e8, November 2019. URL: <a class="reference external" href="https://doi.org/10.1016/j.celrep.2019.09.082">https://doi.org/10.1016/j.celrep.2019.09.082</a>, <a class="reference external" href="https://doi.org/10.1016/j.celrep.2019.09.082">doi:10.1016/j.celrep.2019.09.082</a>.</p>
</div>
<div class="citation" id="id141" role="doc-biblioentry">
<span class="label"><span class="fn-bracket">[</span><a role="doc-backlink" href="#id18">rawDDS+13</a><span class="fn-bracket">]</span></span>
<p>Alexander Dobin, Carrie A Davis, Felix Schlesinger, Jorg Drenkow, Chris Zaleski, Sonali Jha, Philippe Batut, Mark Chaisson, and Thomas R Gingeras. Star: ultrafast universal rna-seq aligner. <em>Bioinformatics</em>, 29(1):15–21, 2013.</p>
</div>
<div class="citation" id="id101" role="doc-biblioentry">
<span class="label"><span class="fn-bracket">[</span>rawFDF+20<span class="fn-bracket">]</span></span>
<span class="backrefs">(<a role="doc-backlink" href="#id2">1</a>,<a role="doc-backlink" href="#id3">2</a>)</span>
<p>Rick Farouni, Haig Djambazian, Lorenzo E Ferri, Jiannis Ragoussis, and Hamed S Najafabadi. Model-based analysis of sample index hopping reveals its widespread artifacts in multiplexed single-cell term`RNA`-sequencing. <em>Nature communications</em>, 11(1):1–8, 2020.</p>
</div>
<div class="citation" id="id92" role="doc-biblioentry">
<span class="label"><span class="fn-bracket">[</span><a role="doc-backlink" href="#id16">rawFar07</a><span class="fn-bracket">]</span></span>
<p>Michael Farrar. Striped Smith–Waterman speeds database searches six times over other SIMD implementations. <em>Bioinformatics</em>, 23(2):156–161, 2007.</p>
</div>
<div class="citation" id="id136" role="doc-biblioentry">
<span class="label"><span class="fn-bracket">[</span><a role="doc-backlink" href="#id44">rawGP21</a><span class="fn-bracket">]</span></span>
<p>Gennady Gorin and Lior Pachter. Length Biases in Single-Cell term`RNA` Sequencing of pre-mRNA. <em>bioRxiv</em>, 2021. URL: <a class="reference external" href="https://www.biorxiv.org/content/early/2021/07/31/2021.07.30.454514">https://www.biorxiv.org/content/early/2021/07/31/2021.07.30.454514</a>, <a class="reference external" href="https://arxiv.org/abs/https://www.biorxiv.org/content/early/2021/07/31/2021.07.30.454514.full.pdf">arXiv:https://www.biorxiv.org/content/early/2021/07/31/2021.07.30.454514.full.pdf</a>, <a class="reference external" href="https://doi.org/10.1101/2021.07.30.454514">doi:10.1101/2021.07.30.454514</a>.</p>
</div>
<div class="citation" id="id110" role="doc-biblioentry">
<span class="label"><span class="fn-bracket">[</span>rawHZS+22<span class="fn-bracket">]</span></span>
<span class="backrefs">(<a role="doc-backlink" href="#id11">1</a>,<a role="doc-backlink" href="#id24">2</a>,<a role="doc-backlink" href="#id30">3</a>,<a role="doc-backlink" href="#id31">4</a>,<a role="doc-backlink" href="#id33">5</a>,<a role="doc-backlink" href="#id37">6</a>,<a role="doc-backlink" href="#id45">7</a>,<a role="doc-backlink" href="#id50">8</a>,<a role="doc-backlink" href="#id52">9</a>,<a role="doc-backlink" href="#id58">10</a>,<a role="doc-backlink" href="#id59">11</a>,<a role="doc-backlink" href="#id61">12</a>,<a role="doc-backlink" href="#id63">13</a>,<a role="doc-backlink" href="#id71">14</a>)</span>
<p>Dongze He, Mohsen Zakeri, Hirak Sarkar, Charlotte Soneson, Avi Srivastava, and Rob Patro. Alevin-fry unlocks rapid, accurate and memory-frugal quantification of single-cell RNA-seq data. <em>Nature Methods</em>, 19(3):316–322, 2022.</p>
</div>
<div class="citation" id="id116" role="doc-biblioentry">
<span class="label"><span class="fn-bracket">[</span><a role="doc-backlink" href="#id62">rawHWC+21</a><span class="fn-bracket">]</span></span>
<p>Cody N. Heiser, Victoria M. Wang, Bob Chen, Jacob J. Hughey, and Ken S. Lau. Automated quality control and cell identification of droplet-based single-cell data using dropkick. <em>Genome Research</em>, 31(10):1742–1752, April 2021. URL: <a class="reference external" href="https://doi.org/10.1101/gr.271908.120">https://doi.org/10.1101/gr.271908.120</a>, <a class="reference external" href="https://doi.org/10.1101/gr.271908.120">doi:10.1101/gr.271908.120</a>.</p>
</div>
<div class="citation" id="id117" role="doc-biblioentry">
<span class="label"><span class="fn-bracket">[</span><a role="doc-backlink" href="#id62">rawHFW+21</a><span class="fn-bracket">]</span></span>
<p>Ariel A Hippen, Matias M Falco, Lukas M Weber, Erdogan Pekcan Erkan, Kaiyang Zhang, Jennifer Anne Doherty, Anna Vähärautio, Casey S Greene, and Stephanie C Hicks. miQC: An adaptive probabilistic framework for quality control of single-cell term`RNA`-sequencing data. <em>PLoS computational biology</em>, 17(8):e1009290, 2021.</p>
</div>
<div class="citation" id="id105" role="doc-biblioentry">
<span class="label"><span class="fn-bracket">[</span><a role="doc-backlink" href="#id45">rawIZJ+13</a><span class="fn-bracket">]</span></span>
<p>Saiful Islam, Amit Zeisel, Simon Joost, Gioele La Manno, Pawel Zajac, Maria Kasper, Peter Lönnerberg, and Sten Linnarsson. Quantitative single-cell term`RNA`-seq with unique molecular identifiers. <em>Nature Methods</em>, 11(2):163–166, December 2013. URL: <a class="reference external" href="https://doi.org/10.1038/nmeth.2772">https://doi.org/10.1038/nmeth.2772</a>, <a class="reference external" href="https://doi.org/10.1038/nmeth.2772">doi:10.1038/nmeth.2772</a>.</p>
</div>
<div class="citation" id="id107" role="doc-biblioentry">
<span class="label"><span class="fn-bracket">[</span>rawKYD21<span class="fn-bracket">]</span></span>
<span class="backrefs">(<a role="doc-backlink" href="#id10">1</a>,<a role="doc-backlink" href="#id19">2</a>,<a role="doc-backlink" href="#id27">3</a>,<a role="doc-backlink" href="#id30">4</a>,<a role="doc-backlink" href="#id45">5</a>,<a role="doc-backlink" href="#id51">6</a>,<a role="doc-backlink" href="#id52">7</a>,<a role="doc-backlink" href="#id53">8</a>,<a role="doc-backlink" href="#id58">9</a>,<a role="doc-backlink" href="#id59">10</a>,<a role="doc-backlink" href="#id63">11</a>,<a role="doc-backlink" href="#id78">12</a>)</span>
<p>Benjamin Kaminow, Dinar Yunusov, and Alexander Dobin. STARsolo: accurate, fast and versatile mapping/quantification of single-cell and single-nucleus term`RNA`-seq data. <em>bioRxiv</em>, 2021.</p>
</div>
<div class="citation" id="id89" role="doc-biblioentry">
<span class="label"><span class="fn-bracket">[</span><a role="doc-backlink" href="#id14">rawLi18</a><span class="fn-bracket">]</span></span>
<p>Heng Li. Minimap2: pairwise alignment for nucleotide sequences. <em>Bioinformatics</em>, 34(18):3094–3100, May 2018. URL: <a class="reference external" href="https://doi.org/10.1093/bioinformatics/bty191">https://doi.org/10.1093/bioinformatics/bty191</a>, <a class="reference external" href="https://doi.org/10.1093/bioinformatics/bty191">doi:10.1093/bioinformatics/bty191</a>.</p>
</div>
<div class="citation" id="id115" role="doc-biblioentry">
<span class="label"><span class="fn-bracket">[</span>rawLRA+19<span class="fn-bracket">]</span></span>
<span class="backrefs">(<a role="doc-backlink" href="#id35">1</a>,<a role="doc-backlink" href="#id37">2</a>,<a role="doc-backlink" href="#id62">3</a>,<a role="doc-backlink" href="#id64">4</a>)</span>
<p>Aaron TL Lun, Samantha Riesenfeld, Tallulah Andrews, Tomas Gomes, John C Marioni, and others. EmptyDrops: distinguishing cells from empty droplets in droplet-based single-cell term`RNA` sequencing data. <em>Genome biology</em>, 20(1):1–9, 2019.</p>
</div>
<div class="citation" id="id142" role="doc-biblioentry">
<span class="label"><span class="fn-bracket">[</span>rawMBS+15<span class="fn-bracket">]</span></span>
<span class="backrefs">(<a role="doc-backlink" href="#id45">1</a>,<a role="doc-backlink" href="#id60">2</a>,<a role="doc-backlink" href="#id73">3</a>)</span>
<p>Evan Z. Macosko, Anindita Basu, Rahul Satija, James Nemesh, Karthik Shekhar, Melissa Goldman, Itay Tirosh, Allison R. Bialas, Nolan Kamitaki, Emily M. Martersteck, John J. Trombetta, David A. Weitz, Joshua R. Sanes, Alex K. Shalek, Aviv Regev, and Steven A. McCarroll. Highly parallel genome-wide expression profiling of individual cells using nanoliter droplets. <em>Cell</em>, 161(5):1202–1214, May 2015. URL: <a class="reference external" href="https://doi.org/10.1016/j.cell.2015.05.002">https://doi.org/10.1016/j.cell.2015.05.002</a>, <a class="reference external" href="https://doi.org/10.1016/j.cell.2015.05.002">doi:10.1016/j.cell.2015.05.002</a>.</p>
</div>
<div class="citation" id="id97" role="doc-biblioentry">
<span class="label"><span class="fn-bracket">[</span><a role="doc-backlink" href="#id15">rawMSMME20</a><span class="fn-bracket">]</span></span>
<p>Santiago Marco-Sola, Juan Carlos Moure, Miquel Moreto, and Antonio Espinosa. Fast gap-affine pairwise alignment using the wavefront algorithm. <em>Bioinformatics</em>, September 2020. URL: <a class="reference external" href="https://doi.org/10.1093/bioinformatics/btaa777">https://doi.org/10.1093/bioinformatics/btaa777</a>, <a class="reference external" href="https://doi.org/10.1093/bioinformatics/btaa777">doi:10.1093/bioinformatics/btaa777</a>.</p>
</div>
<div class="citation" id="id122" role="doc-biblioentry">
<span class="label"><span class="fn-bracket">[</span>rawMMG19<span class="fn-bracket">]</span></span>
<span class="backrefs">(<a role="doc-backlink" href="#id36">1</a>,<a role="doc-backlink" href="#id65">2</a>)</span>
<p>Christopher S. McGinnis, Lyndsay M. Murrow, and Zev J. Gartner. DoubletFinder: doublet detection in single-cell term`RNA` sequencing data using artificial nearest neighbors. <em>Cell Systems</em>, 8(4):329–337.e4, April 2019. URL: <a class="reference external" href="https://doi.org/10.1016/j.cels.2019.03.003">https://doi.org/10.1016/j.cels.2019.03.003</a>, <a class="reference external" href="https://doi.org/10.1016/j.cels.2019.03.003">doi:10.1016/j.cels.2019.03.003</a>.</p>
</div>
<div class="citation" id="id108" role="doc-biblioentry">
<span class="label"><span class="fn-bracket">[</span>rawMBL+21<span class="fn-bracket">]</span></span>
<span class="backrefs">(<a role="doc-backlink" href="#id9">1</a>,<a role="doc-backlink" href="#id33">2</a>,<a role="doc-backlink" href="#id45">3</a>,<a role="doc-backlink" href="#id47">4</a>,<a role="doc-backlink" href="#id58">5</a>,<a role="doc-backlink" href="#id59">6</a>,<a role="doc-backlink" href="#id77">7</a>)</span>
<p>Páll Melsted, A. Sina Booeshaghi, Lauren Liu, Fan Gao, Lambda Lu, Kyung Hoi Min, Eduardo da Veiga Beltrame, Kristján Eldjárn Hjörleifsson, Jase Gehring, and Lior Pachter. Modular, efficient and constant-memory single-cell rna-seq preprocessing. <em>Nature Biotechnology</em>, 39(7):813–818, April 2021. URL: <a class="reference external" href="https://doi.org/10.1038/s41587-021-00870-2">https://doi.org/10.1038/s41587-021-00870-2</a>, <a class="reference external" href="https://doi.org/10.1038/s41587-021-00870-2">doi:10.1038/s41587-021-00870-2</a>.</p>
</div>
<div class="citation" id="id118" role="doc-biblioentry">
<span class="label"><span class="fn-bracket">[</span>rawMP21<span class="fn-bracket">]</span></span>
<span class="backrefs">(<a role="doc-backlink" href="#id35">1</a>,<a role="doc-backlink" href="#id62">2</a>)</span>
<p>Walter Muskovic and Joseph E Powell. DropletQC: improved identification of empty droplets and damaged cells in single-cell term`RNA`-seq data. <em>Genome Biology</em>, 22(1):1–9, 2021.</p>
</div>
<div class="citation" id="id90" role="doc-biblioentry">
<span class="label"><span class="fn-bracket">[</span><a role="doc-backlink" href="#id8">rawNMullerHS20</a><span class="fn-bracket">]</span></span>
<p>Stefan Niebler, André Müller, Thomas Hankeln, and Bertil Schmidt. RainDrop: Rapid activation matrix computation for droplet-based single-cell term`RNA`-seq reads. <em>BMC bioinformatics</em>, 21(1):1–14, 2020.</p>
</div>
<div class="citation" id="id111" role="doc-biblioentry">
<span class="label"><span class="fn-bracket">[</span><a role="doc-backlink" href="#id45">rawOEM+18</a><span class="fn-bracket">]</span></span>
<p>Baraa Orabi, Emre Erhan, Brian McConeghy, Stanislav V Volik, Stephane Le Bihan, Robert Bell, Colin C Collins, Cedric Chauve, and Faraz Hach. Alignment-free clustering of UMI tagged term`DNA` molecules. <em>Bioinformatics</em>, 35(11):1829–1836, October 2018. URL: <a class="reference external" href="https://doi.org/10.1093/bioinformatics/bty888">https://doi.org/10.1093/bioinformatics/bty888</a>, <a class="reference external" href="https://doi.org/10.1093/bioinformatics/bty888">doi:10.1093/bioinformatics/bty888</a>.</p>
</div>
<div class="citation" id="id113" role="doc-biblioentry">
<span class="label"><span class="fn-bracket">[</span>rawPZV+18<span class="fn-bracket">]</span></span>
<span class="backrefs">(<a role="doc-backlink" href="#id6">1</a>,<a role="doc-backlink" href="#id25">2</a>,<a role="doc-backlink" href="#id45">3</a>,<a role="doc-backlink" href="#id75">4</a>)</span>
<p>Swati Parekh, Christoph Ziegenhain, Beate Vieth, Wolfgang Enard, and Ines Hellmann. zUMIs - a fast and flexible pipeline to process term`RNA` sequencing data with UMIs. <em>GigaScience</em>, May 2018. URL: <a class="reference external" href="https://doi.org/10.1093/gigascience/giy059">https://doi.org/10.1093/gigascience/giy059</a>, <a class="reference external" href="https://doi.org/10.1093/gigascience/giy059">doi:10.1093/gigascience/giy059</a>.</p>
</div>
<div class="citation" id="id94" role="doc-biblioentry">
<span class="label"><span class="fn-bracket">[</span><a role="doc-backlink" href="#id69">rawPHJ+20</a><span class="fn-bracket">]</span></span>
<p>Ralph Patrick, David T. Humphreys, Vaibhao Janbandhu, Alicia Oshlack, Joshua W.K. Ho, Richard P. Harvey, and Kitty K. Lo. Sierra: discovery of differential transcript usage from polyA-captured single-cell term`RNA`-seq data. <em>Genome Biology</em>, jul 2020. URL: <a class="reference external" href="https://doi.org/10.1186%2Fs13059-020-02071-7">https://doi.org/10.1186%2Fs13059-020-02071-7</a>, <a class="reference external" href="https://doi.org/10.1186/s13059-020-02071-7">doi:10.1186/s13059-020-02071-7</a>.</p>
</div>
<div class="citation" id="id134" role="doc-biblioentry">
<span class="label"><span class="fn-bracket">[</span>rawPPC+22<span class="fn-bracket">]</span></span>
<span class="backrefs">(<a role="doc-backlink" href="#id28">1</a>,<a role="doc-backlink" href="#id29">2</a>,<a role="doc-backlink" href="#id72">3</a>)</span>
<p>Allan-Hermann Pool, Helen Poldsam, Sisi Chen, Matt Thomson, and Yuki Oka. Enhanced recovery of single-cell term`RNA`-sequencing reads for missing gene expression data. <em>bioRxiv</em>, 2022. URL: <a class="reference external" href="https://www.biorxiv.org/content/early/2022/04/27/2022.04.26.489449">https://www.biorxiv.org/content/early/2022/04/27/2022.04.26.489449</a>, <a class="reference external" href="https://arxiv.org/abs/https://www.biorxiv.org/content/early/2022/04/27/2022.04.26.489449.full.pdf">arXiv:https://www.biorxiv.org/content/early/2022/04/27/2022.04.26.489449.full.pdf</a>, <a class="reference external" href="https://doi.org/10.1101/2022.04.26.489449">doi:10.1101/2022.04.26.489449</a>.</p>
</div>
<div class="citation" id="id93" role="doc-biblioentry">
<span class="label"><span class="fn-bracket">[</span><a role="doc-backlink" href="#id16">rawRS00</a><span class="fn-bracket">]</span></span>
<p>Torbjørn Rognes and Erling Seeberg. Six-fold speed-up of Smith–Waterman sequence database searches using parallel processing on common microprocessors. <em>Bioinformatics</em>, 16(8):699–706, 2000.</p>
</div>
<div class="citation" id="id102" role="doc-biblioentry">
<span class="label"><span class="fn-bracket">[</span>rawSHS17<span class="fn-bracket">]</span></span>
<span class="backrefs">(<a role="doc-backlink" href="#id4">1</a>,<a role="doc-backlink" href="#id37">2</a>,<a role="doc-backlink" href="#id38">3</a>,<a role="doc-backlink" href="#id40">4</a>,<a role="doc-backlink" href="#id45">5</a>,<a role="doc-backlink" href="#id48">6</a>,<a role="doc-backlink" href="#id49">7</a>,<a role="doc-backlink" href="#id53">8</a>,<a role="doc-backlink" href="#id54">9</a>,<a role="doc-backlink" href="#id55">10</a>,<a role="doc-backlink" href="#id61">11</a>)</span>
<p>Tom Smith, Andreas Heger, and Ian Sudbery. UMI-tools: modeling sequencing errors in Unique Molecular Identifiers to improve quantification accuracy. <em>Genome research</em>, 27(3):491–499, 2017.</p>
</div>
<div class="citation" id="id109" role="doc-biblioentry">
<span class="label"><span class="fn-bracket">[</span><a role="doc-backlink" href="#id32">rawSSPS21</a><span class="fn-bracket">]</span></span>
<p>Charlotte Soneson, Avi Srivastava, Rob Patro, and Michael B Stadler. Preprocessing choices affect RNA velocity results for droplet scRNA-seq data. <em>PLoS computational biology</em>, 17(1):e1008585, 2021.</p>
</div>
<div class="citation" id="id131" role="doc-biblioentry">
<span class="label"><span class="fn-bracket">[</span><a role="doc-backlink" href="#id20">rawSMS+20</a><span class="fn-bracket">]</span></span>
<p>Avi Srivastava, Laraib Malik, Hirak Sarkar, Mohsen Zakeri, Fatemeh Almodaresi, Charlotte Soneson, Michael I Love, Carl Kingsford, and Rob Patro. Alignment and mapping methodology influence transcript abundance estimation. <em>Genome Biology</em>, 21(1):1–29, 2020.</p>
</div>
<div class="citation" id="id103" role="doc-biblioentry">
<span class="label"><span class="fn-bracket">[</span>rawSMS+19<span class="fn-bracket">]</span></span>
<span class="backrefs">(<a role="doc-backlink" href="#id7">1</a>,<a role="doc-backlink" href="#id21">2</a>,<a role="doc-backlink" href="#id33">3</a>,<a role="doc-backlink" href="#id39">4</a>,<a role="doc-backlink" href="#id41">5</a>,<a role="doc-backlink" href="#id42">6</a>,<a role="doc-backlink" href="#id45">7</a>,<a role="doc-backlink" href="#id50">8</a>,<a role="doc-backlink" href="#id52">9</a>,<a role="doc-backlink" href="#id53">10</a>,<a role="doc-backlink" href="#id54">11</a>,<a role="doc-backlink" href="#id56">12</a>,<a role="doc-backlink" href="#id57">13</a>,<a role="doc-backlink" href="#id76">14</a>)</span>
<p>Avi Srivastava, Laraib Malik, Tom Smith, Ian Sudbery, and Rob Patro. Alevin efficiently estimates accurate gene abundances from dscRNA-seq data. <em>Genome biology</em>, 20(1):1–16, 2019.</p>
</div>
<div class="citation" id="id126" role="doc-biblioentry">
<span class="label"><span class="fn-bracket">[</span><a role="doc-backlink" href="#id23">rawSSGP16</a><span class="fn-bracket">]</span></span>
<p>Avi Srivastava, Hirak Sarkar, Nitish Gupta, and Rob Patro. Rapmap: a rapid, sensitive and accurate tool for mapping term`rna`-seq reads to transcriptomes. <em>Bioinformatics</em>, 32(12):i192–i200, 2016.</p>
</div>
<div class="citation" id="id98" role="doc-biblioentry">
<span class="label"><span class="fn-bracket">[</span><a role="doc-backlink" href="#id17">rawSK18</a><span class="fn-bracket">]</span></span>
<p>Hajime Suzuki and Masahiro Kasahara. Introducing difference recurrence relations for faster semi-global alignment of long sequences. <em>BMC Bioinformatics</em>, February 2018. URL: <a class="reference external" href="https://doi.org/10.1186/s12859-018-2014-8">https://doi.org/10.1186/s12859-018-2014-8</a>, <a class="reference external" href="https://doi.org/10.1186/s12859-018-2014-8">doi:10.1186/s12859-018-2014-8</a>.</p>
</div>
<div class="citation" id="id112" role="doc-biblioentry">
<span class="label"><span class="fn-bracket">[</span>rawTMP+21<span class="fn-bracket">]</span></span>
<span class="backrefs">(<a role="doc-backlink" href="#id45">1</a>,<a role="doc-backlink" href="#id46">2</a>)</span>
<p>Maria Tsagiopoulou, Maria Christina Maniou, Nikolaos Pechlivanis, Anastasis Togkousidis, Michaela Kotrová, Tobias Hutzenlaub, Ilias Kappas, Anastasia Chatzidimitriou, and Fotis Psomopoulos. UMIc: a preprocessing method for UMI deduplication and reads correction. <em>Frontiers in Genetics</em>, May 2021. URL: <a class="reference external" href="https://doi.org/10.3389/fgene.2021.660366">https://doi.org/10.3389/fgene.2021.660366</a>, <a class="reference external" href="https://doi.org/10.3389/fgene.2021.660366">doi:10.3389/fgene.2021.660366</a>.</p>
</div>
<div class="citation" id="id123" role="doc-biblioentry">
<span class="label"><span class="fn-bracket">[</span>rawWLK19<span class="fn-bracket">]</span></span>
<span class="backrefs">(<a role="doc-backlink" href="#id36">1</a>,<a role="doc-backlink" href="#id65">2</a>)</span>
<p>Samuel L. Wolock, Romain Lopez, and Allon M. Klein. Scrublet: Computational Identification of Cell Doublets in Single-Cell Transcriptomic Data. <em>Cell Systems</em>, 8(4):281–291.e9, April 2019. URL: <a class="reference external" href="https://doi.org/10.1016/j.cels.2018.11.005">https://doi.org/10.1016/j.cels.2018.11.005</a>, <a class="reference external" href="https://doi.org/10.1016/j.cels.2018.11.005">doi:10.1016/j.cels.2018.11.005</a>.</p>
</div>
<div class="citation" id="id91" role="doc-biblioentry">
<span class="label"><span class="fn-bracket">[</span><a role="doc-backlink" href="#id16">rawWoz97</a><span class="fn-bracket">]</span></span>
<p>Andrzej Wozniak. Using video-oriented instructions to speed up sequence comparison. <em>Bioinformatics</em>, 13(2):145–150, 1997.</p>
</div>
<div class="citation" id="id96" role="doc-biblioentry">
<span class="label"><span class="fn-bracket">[</span>rawYTS+21<span class="fn-bracket">]</span></span>
<span class="backrefs">(<a role="doc-backlink" href="#id66">1</a>,<a role="doc-backlink" href="#id67">2</a>)</span>
<p>Yue You, Luyi Tian, Shian Su, Xueyi Dong, Jafar S. Jabbari, Peter F. Hickey, and Matthew E. Ritchie. Benchmarking UMI-based single-cell term`RNA`-seq preprocessing workflows. <em>Genome Biology</em>, dec 2021. URL: <a class="reference external" href="https://doi.org/10.1186%2Fs13059-021-02552-3">https://doi.org/10.1186%2Fs13059-021-02552-3</a>, <a class="reference external" href="https://doi.org/10.1186/s13059-021-02552-3">doi:10.1186/s13059-021-02552-3</a>.</p>
</div>
<div class="citation" id="id120" role="doc-biblioentry">
<span class="label"><span class="fn-bracket">[</span>rawYB20<span class="fn-bracket">]</span></span>
<span class="backrefs">(<a role="doc-backlink" href="#id35">1</a>,<a role="doc-backlink" href="#id62">2</a>)</span>
<p>Matthew D Young and Sam Behjati. SoupX removes ambient term`RNA` contamination from droplet-based single-cell term`RNA` sequencing data. <em>GigaScience</em>, December 2020. URL: <a class="reference external" href="https://doi.org/10.1093/gigascience/giaa151">https://doi.org/10.1093/gigascience/giaa151</a>, <a class="reference external" href="https://doi.org/10.1093/gigascience/giaa151">doi:10.1093/gigascience/giaa151</a>.</p>
</div>
<div class="citation" id="id87" role="doc-biblioentry">
<span class="label"><span class="fn-bracket">[</span><a role="doc-backlink" href="#id1">rawZT21</a><span class="fn-bracket">]</span></span>
<p>Luke Zappia and Fabian J. Theis. Over 1000 tools reveal trends in the single-cell rna-seq analysis landscape. <em>Genome Biology</em>, 22(1):301, Oct 2021. URL: <a class="reference external" href="https://doi.org/10.1186/s13059-021-02519-4">https://doi.org/10.1186/s13059-021-02519-4</a>, <a class="reference external" href="https://doi.org/10.1186/s13059-021-02519-4">doi:10.1186/s13059-021-02519-4</a>.</p>
</div>
<div class="citation" id="id100" role="doc-biblioentry">
<span class="label"><span class="fn-bracket">[</span><a role="doc-backlink" href="#id13">rawZSWM00</a><span class="fn-bracket">]</span></span>
<p>Zheng Zhang, Scott Schwartz, Lukas Wagner, and Webb Miller. A greedy algorithm for aligning term`DNA` sequences. <em>Journal of Computational Biology</em>, 7(1-2):203–214, February 2000. URL: <a class="reference external" href="https://doi.org/10.1089/10665270050081478">https://doi.org/10.1089/10665270050081478</a>, <a class="reference external" href="https://doi.org/10.1089/10665270050081478">doi:10.1089/10665270050081478</a>.</p>
</div>
<div class="citation" id="id143" role="doc-biblioentry">
<span class="label"><span class="fn-bracket">[</span>rawZTB+17<span class="fn-bracket">]</span></span>
<span class="backrefs">(<a role="doc-backlink" href="#id5">1</a>,<a role="doc-backlink" href="#id26">2</a>,<a role="doc-backlink" href="#id51">3</a>,<a role="doc-backlink" href="#id59">4</a>)</span>
<p>Grace X. Y. Zheng, Jessica M. Terry, Phillip Belgrader, Paul Ryvkin, Zachary W. Bent, Ryan Wilson, Solongo B. Ziraldo, Tobias D. Wheeler, Geoff P. McDermott, Junjie Zhu, Mark T. Gregory, Joe Shuga, Luz Montesclaros, Jason G. Underwood, Donald A. Masquelier, Stefanie Y. Nishimura, Michael Schnall-Levin, Paul W. Wyatt, Christopher M. Hindson, Rajiv Bharadwaj, Alexander Wong, Kevin D. Ness, Lan W. Beppu, H. Joachim Deeg, Christopher McFarland, Keith R. Loeb, William J. Valente, Nolan G. Ericson, Emily A. Stevens, Jerald P. Radich, Tarjei S. Mikkelsen, Benjamin J. Hindson, and Jason H. Bielas. Massively parallel digital transcriptional profiling of single cells. <em>Nature Communications</em>, 8(1):14049, Jan 2017. URL: <a class="reference external" href="https://doi.org/10.1038/ncomms14049">https://doi.org/10.1038/ncomms14049</a>, <a class="reference external" href="https://doi.org/10.1038/ncomms14049">doi:10.1038/ncomms14049</a>.</p>
</div>
<div class="citation" id="id135" role="doc-biblioentry">
<span class="label"><span class="fn-bracket">[</span><a role="doc-backlink" href="#id38">rawZHHJS22</a><span class="fn-bracket">]</span></span>
<p>Christoph Ziegenhain, Gert-Jan Hendriks, Michael Hagemann-Jensen, and Rickard Sandberg. Molecular spikes: a gold standard for single-cell term`RNA` counting. <em>Nature Methods</em>, 19(5):560–566, 2022.</p>
</div>
<div class="citation" id="id104" role="doc-biblioentry">
<span class="label"><span class="fn-bracket">[</span>raw10xGenomics21<span class="fn-bracket">]</span></span>
<span class="backrefs">(<a role="doc-backlink" href="#id28">1</a>,<a role="doc-backlink" href="#id29">2</a>,<a role="doc-backlink" href="#id34">3</a>,<a role="doc-backlink" href="#id43">4</a>,<a role="doc-backlink" href="#id44">5</a>)</span>
<p>10x Genomics. Technical note - interpreting intronic and antisense reads in 10x genomics single cell gene expression data. <a class="reference external" href="https://www.10xgenomics.com/support/single-cell-gene-expression/documentation/steps/sequencing/interpreting-intronic-and-antisense-reads-in-10-x-genomics-single-cell-gene-expression-data">https://www.10xgenomics.com/support/single-cell-gene-expression/documentation/steps/sequencing/interpreting-intronic-and-antisense-reads-in-10-x-genomics-single-cell-gene-expression-data</a>, August 2021.</p>
</div>
</div>
</div>
</section>
<section id="contributors">
<h2><span class="section-number">3.12. </span>Contributors<a class="headerlink" href="#contributors" title="Link to this heading">#</a></h2>
<p>We gratefully acknowledge the contributions of:</p>
<section id="authors">
<h3><span class="section-number">3.12.1. </span>Authors<a class="headerlink" href="#authors" title="Link to this heading">#</a></h3>
<ul class="simple">
<li><p>Dongze He</p></li>
<li><p>Avi Srivastava</p></li>
<li><p>Hirak Sarkar</p></li>
<li><p>Rob Patro</p></li>
<li><p>Seo H. Kim</p></li>
</ul>
</section>
<section id="reviewers">
<h3><span class="section-number">3.12.2. </span>Reviewers<a class="headerlink" href="#reviewers" title="Link to this heading">#</a></h3>
<ul class="simple">
<li><p>Lukas Heumos</p></li>
</ul>
</section>
</section>
</section>

    <script type="text/x-thebe-config">
    {
        requestKernel: true,
        binderOptions: {
            repo: "binder-examples/jupyter-stacks-datascience",
            ref: "master",
        },
        codeMirrorConfig: {
            theme: "abcdef",
            mode: "python"
        },
        kernelOptions: {
            name: "python3",
            path: "./introduction"
        },
        predefinedOutput: true
    }
    </script>
    <script>kernelName = 'python3'</script>

                </article>
              

              
              
              
              
                <footer class="prev-next-footer d-print-none">
                  
<div class="prev-next-area">
    <a class="left-prev"
       href="scrna_seq.html"
       title="previous page">
      <i class="fa-solid fa-angle-left"></i>
      <div class="prev-next-info">
        <p class="prev-next-subtitle">previous</p>
        <p class="prev-next-title"><span class="section-number">2. </span>Single-cell RNA sequencing</p>
      </div>
    </a>
    <a class="right-next"
       href="analysis_tools.html"
       title="next page">
      <div class="prev-next-info">
        <p class="prev-next-subtitle">next</p>
        <p class="prev-next-title"><span class="section-number">4. </span>Analysis frameworks and tools</p>
      </div>
      <i class="fa-solid fa-angle-right"></i>
    </a>
</div>
                </footer>
              
            </div>
            
            
              
                <div class="bd-sidebar-secondary bd-toc"><div class="sidebar-secondary-items sidebar-secondary__inner">


  <div class="sidebar-secondary-item">
  <div class="page-toc tocsection onthispage">
    <i class="fa-solid fa-list"></i> Contents
  </div>
  <nav class="bd-toc-nav page-toc">
    <ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#motivation">3.1. Motivation</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#raw-data-quality-control">3.2. Raw data quality control</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#alignment-and-mapping">3.3. Alignment and mapping</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#types-of-mapping">3.3.1. Types of mapping</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#mapping-against-different-reference-sequences">3.3.2. Mapping against different reference sequences</a><ul class="nav section-nav flex-column">
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#mapping-to-the-full-genome">3.3.2.1. Mapping to the full genome</a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#mapping-to-the-spliced-transcriptome">3.3.2.2. Mapping to the spliced transcriptome</a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#mapping-to-an-augmented-transcriptome">3.3.2.3. Mapping to an augmented transcriptome</a></li>
</ul>
</li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#cell-barcode-correction">3.4. Cell barcode correction</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#type-of-errors-in-barcoding">3.4.1. Type of errors in barcoding</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#umi-resolution">3.5. UMI resolution</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#the-need-for-umi-resolution">3.5.1. The need for UMI resolution</a><ul class="nav section-nav flex-column">
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#quantification">3.5.1.1. Quantification</a></li>
</ul>
</li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#count-matrix-quality-control">3.6. Count matrix quality control</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#empty-droplet-detection">3.6.1. Empty droplet detection</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#doublet-detection">3.6.2. Doublet detection</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#count-data-representation">3.7. Count data representation</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#brief-discussion">3.8. Brief discussion</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#a-real-world-example">3.9. A real-world example</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#preparation">3.9.1. Preparation</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#simplified-raw-data-processing-pipeline">3.9.2. Simplified raw data processing pipeline</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#the-complete-alevin-fry-pipeline">3.9.3. The complete alevin-fry pipeline</a><ul class="nav section-nav flex-column">
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#building-the-index">3.9.3.1. Building the index</a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#mapping-and-quantification">3.9.3.2. Mapping and quantification</a></li>
</ul>
</li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#useful-links">3.10. Useful links</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#references">3.11. References</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#contributors">3.12. Contributors</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#authors">3.12.1. Authors</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#reviewers">3.12.2. Reviewers</a></li>
</ul>
</li>
</ul>
  </nav></div>

</div></div>
              
            
          </div>
          <footer class="bd-footer-content">
            
<div class="bd-footer-content__inner container">
  
  <div class="footer-item">
    
<p class="component-author">
By Lukas Heumos, Anna Schaar, single-cell best practices consortium
</p>

  </div>
  
  <div class="footer-item">
    

  <p class="copyright">
    
      © Copyright 2023.
      <br/>
    
  </p>

  </div>
  
  <div class="footer-item">
    
  </div>
  
  <div class="footer-item">
    
<div class="extra_footer">
  <div>
  Brought to you by
  <a href="https://www.github.com/theislab">Theislab</a>, with many thanks to the single-cell community as a whole!
</div>

</div>
  </div>
  
</div>
          </footer>
        

      </main>
    </div>
  </div>
  
  <!-- Scripts loaded after <body> so the DOM is not blocked -->
  <script src="../_static/scripts/bootstrap.js?digest=dfe6caa3a7d634c4db9b"></script>
<script src="../_static/scripts/pydata-sphinx-theme.js?digest=dfe6caa3a7d634c4db9b"></script>

  <footer class="bd-footer">
  </footer>
  </body>
</html>