
<!DOCTYPE html>


<html lang="en" data-content_root="../" >

  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="viewport" content="width=device-width, initial-scale=1" />

    <title>21. Cell-cell communication &#8212; Single-cell best practices</title>
  
  
  
  <script data-cfasync="false">
    document.documentElement.dataset.mode = localStorage.getItem("mode") || "";
    document.documentElement.dataset.theme = localStorage.getItem("theme") || "";
  </script>
  <!--
    this give us a css class that will be invisible only if js is disabled
  -->
  <noscript>
    <style>
      .pst-js-only { display: none !important; }

    </style>
  </noscript>
  
  <!-- Loaded before other Sphinx assets -->
  <link href="../_static/styles/theme.css?digest=8878045cc6db502f8baf" rel="stylesheet" />
<link href="../_static/styles/pydata-sphinx-theme.css?digest=8878045cc6db502f8baf" rel="stylesheet" />

    <link rel="stylesheet" type="text/css" href="../_static/pygments.css?v=03e43079" />
    <link rel="stylesheet" type="text/css" href="../_static/styles/sphinx-book-theme.css?v=a3416100" />
    <link rel="stylesheet" type="text/css" href="../_static/togglebutton.css?v=13237357" />
    <link rel="stylesheet" type="text/css" href="../_static/copybutton.css?v=76b2166b" />
    <link rel="stylesheet" type="text/css" href="../_static/mystnb.4510f1fc1dee50b3e5859aac5469c37c29e427902b24a333a5f9fcb2f0b3ac41.css" />
    <link rel="stylesheet" type="text/css" href="../_static/sphinx-thebe.css?v=4fa983c6" />
    <link rel="stylesheet" type="text/css" href="../_static/sphinx-design.min.css?v=95c83b7e" />
    <link rel="stylesheet" type="text/css" href="../_static/book.css?v=148c11e2" />
  
  <!-- So that users can add custom icons -->
  <script src="../_static/scripts/fontawesome.js?digest=8878045cc6db502f8baf"></script>
  <!-- Pre-loaded scripts that we'll load fully later -->
  <link rel="preload" as="script" href="../_static/scripts/bootstrap.js?digest=8878045cc6db502f8baf" />
<link rel="preload" as="script" href="../_static/scripts/pydata-sphinx-theme.js?digest=8878045cc6db502f8baf" />

    <script src="../_static/documentation_options.js?v=9eb32ce0"></script>
    <script src="../_static/doctools.js?v=9a2dae69"></script>
    <script src="../_static/sphinx_highlight.js?v=dc90522c"></script>
    <script src="../_static/clipboard.min.js?v=a7894cd8"></script>
    <script src="../_static/copybutton.js?v=f281be69"></script>
    <script src="../_static/scripts/sphinx-book-theme.js?v=887ef09a"></script>
    <script>let toggleHintShow = 'Click to show';</script>
    <script>let toggleHintHide = 'Click to hide';</script>
    <script>let toggleOpenOnPrint = 'true';</script>
    <script src="../_static/togglebutton.js?v=4a39c7ea"></script>
    <script>var togglebuttonSelector = '.toggle, .admonition.dropdown';</script>
    <script src="../_static/design-tabs.js?v=f930bc37"></script>
    <script>const THEBE_JS_URL = "https://unpkg.com/thebe@0.8.2/lib/index.js"; const thebe_selector = ".thebe,.cell"; const thebe_selector_input = "pre"; const thebe_selector_output = ".output, .cell_output"</script>
    <script async="async" src="../_static/sphinx-thebe.js?v=c100c467"></script>
    <script>var togglebuttonSelector = '.toggle, .admonition.dropdown';</script>
    <script>const THEBE_JS_URL = "https://unpkg.com/thebe@0.8.2/lib/index.js"; const thebe_selector = ".thebe,.cell"; const thebe_selector_input = "pre"; const thebe_selector_output = ".output, .cell_output"</script>
    <script>DOCUMENTATION_OPTIONS.pagename = 'mechanisms/cell_cell_communication';</script>
    <script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
    <script src="../_static/book.js?v=dcd3d8bf"></script>
    <link rel="icon" href="../_static/favicon.ico"/>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="22. Bulk deconvolution" href="../deconvolution/bulk_deconvolution.html" />
    <link rel="prev" title="20. Gene regulatory networks" href="gene_regulatory_networks.html" />
  <meta name="viewport" content="width=device-width, initial-scale=1"/>
  <meta name="docsearch:language" content="en"/>
  <meta name="docsearch:version" content="" />
  </head>
  
  
  <body data-bs-spy="scroll" data-bs-target=".bd-toc-nav" data-offset="180" data-bs-root-margin="0px 0px -60%" data-default-mode="">

  
  
  <div id="pst-skip-link" class="skip-link d-print-none"><a href="#main-content">Skip to main content</a></div>
  
  <div id="pst-scroll-pixel-helper"></div>
  
  <button type="button" class="btn rounded-pill" id="pst-back-to-top">
    <i class="fa-solid fa-arrow-up"></i>Back to top</button>

  
  <dialog id="pst-search-dialog">
    
<form class="bd-search d-flex align-items-center"
      action="../search.html"
      method="get">
  <i class="fa-solid fa-magnifying-glass"></i>
  <input type="search"
         class="form-control"
         name="q"
         placeholder="Search this book..."
         aria-label="Search this book..."
         autocomplete="off"
         autocorrect="off"
         autocapitalize="off"
         spellcheck="false"/>
  <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd>K</kbd></span>
</form>
  </dialog>

  <div class="pst-async-banner-revealer d-none">
  <aside id="bd-header-version-warning" class="d-none d-print-none" aria-label="Version warning"></aside>
</div>

  
    <header class="bd-header navbar navbar-expand-lg bd-navbar d-print-none">
    </header>
  

  <div class="bd-container">
    <div class="bd-container__inner bd-page-width">
      
      
      
      <dialog id="pst-primary-sidebar-modal"></dialog>
      <div id="pst-primary-sidebar" class="bd-sidebar-primary bd-sidebar">
        

  
  <div class="sidebar-header-items sidebar-primary__section">
    
    
    
    
  </div>
  
    <div class="sidebar-primary-items__start sidebar-primary__section">
        <div class="sidebar-primary-item">

  
    
  

<a class="navbar-brand logo" href="../preamble.html">
  
  
  
  
  
    
    
      
    
    
    <img src="../_static/logo.png" class="logo__image only-light" alt="Single-cell best practices - Home"/>
    <img src="../_static/logo.png" class="logo__image only-dark pst-js-only" alt="Single-cell best practices - Home"/>
  
  
</a></div>
        <div class="sidebar-primary-item">

<button class="btn search-button-field search-button__button pst-js-only" title="Search" aria-label="Search" data-bs-placement="bottom" data-bs-toggle="tooltip">
 <i class="fa-solid fa-magnifying-glass"></i>
 <span class="search-button__default-text">Search</span>
 <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd class="kbd-shortcut__modifier">K</kbd></span>
</button></div>
        <div class="sidebar-primary-item"><nav class="bd-links bd-docs-nav" aria-label="Main">
    <div class="bd-toc-item navbar-nav active">
        <p aria-level="2" class="caption" role="heading"><span class="caption-text">Introduction</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../introduction/prior_art.html">1. Prior art</a></li>
<li class="toctree-l1"><a class="reference internal" href="../introduction/scrna_seq.html">2. Single-cell RNA sequencing</a></li>
<li class="toctree-l1"><a class="reference internal" href="../introduction/raw_data_processing.html">3. Raw data processing</a></li>
<li class="toctree-l1"><a class="reference internal" href="../introduction/analysis_tools.html">4. Analysis frameworks and tools</a></li>
<li class="toctree-l1"><a class="reference internal" href="../introduction/interoperability.html">5. Interoperability</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Preprocessing and visualization</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../preprocessing_visualization/quality_control.html">6. Quality Control</a></li>
<li class="toctree-l1"><a class="reference internal" href="../preprocessing_visualization/normalization.html">7. Normalization</a></li>
<li class="toctree-l1"><a class="reference internal" href="../preprocessing_visualization/feature_selection.html">8. Feature selection</a></li>
<li class="toctree-l1"><a class="reference internal" href="../preprocessing_visualization/dimensionality_reduction.html">9. Dimensionality Reduction</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Identifying cellular structure</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../cellular_structure/clustering.html">10. Clustering</a></li>
<li class="toctree-l1"><a class="reference internal" href="../cellular_structure/annotation.html">11. Annotation</a></li>
<li class="toctree-l1"><a class="reference internal" href="../cellular_structure/integration.html">12. Data integration</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Inferring trajectories</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../trajectories/pseudotemporal.html">13. Pseudotemporal ordering</a></li>
<li class="toctree-l1"><a class="reference internal" href="../trajectories/rna_velocity.html">14. RNA velocity</a></li>
<li class="toctree-l1"><a class="reference internal" href="../trajectories/lineage_tracing.html">15. Lineage tracing</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Dealing with conditions</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../conditions/differential_gene_expression.html">16. Differential gene expression analysis</a></li>
<li class="toctree-l1"><a class="reference internal" href="../conditions/compositional.html">17. Compositional analysis</a></li>
<li class="toctree-l1"><a class="reference internal" href="../conditions/gsea_pathway.html">18. Gene set enrichment and pathway analysis</a></li>
<li class="toctree-l1"><a class="reference internal" href="../conditions/perturbation_modeling.html">19. Perturbation modeling</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Modeling mechanisms</span></p>
<ul class="current nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="gene_regulatory_networks.html">20. Gene regulatory networks</a></li>
<li class="toctree-l1 current active"><a class="current reference internal" href="#">21. Cell-cell communication</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Deconvolution</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../deconvolution/bulk_deconvolution.html">22. Bulk deconvolution</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Chromatin Accessibility</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../chromatin_accessibility/introduction.html">23. Single-cell ATAC sequencing</a></li>
<li class="toctree-l1"><a class="reference internal" href="../chromatin_accessibility/quality_control.html">24. Quality Control</a></li>
<li class="toctree-l1"><a class="reference internal" href="../chromatin_accessibility/gene_regulatory_networks_atac.html">25. Gene regulatory networks using chromatin accessibility</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Spatial omics</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../spatial/introduction.html">26. Single-cell data resolved in space</a></li>
<li class="toctree-l1"><a class="reference internal" href="../spatial/neighborhood.html">27. Neighborhood analysis</a></li>
<li class="toctree-l1"><a class="reference internal" href="../spatial/domains.html">28. Spatial domains</a></li>
<li class="toctree-l1"><a class="reference internal" href="../spatial/spatially_variable_genes.html">29. Spatially variable genes</a></li>
<li class="toctree-l1"><a class="reference internal" href="../spatial/deconvolution.html">30. Spatial deconvolution</a></li>
<li class="toctree-l1"><a class="reference internal" href="../spatial/imputation.html">31. Imputation</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Surface protein</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../surface_protein/quality_control.html">32. Quality control</a></li>
<li class="toctree-l1"><a class="reference internal" href="../surface_protein/normalization.html">33. Normalization</a></li>
<li class="toctree-l1"><a class="reference internal" href="../surface_protein/doublet_detection.html">34. Doublet detection</a></li>
<li class="toctree-l1"><a class="reference internal" href="../surface_protein/dimensionality_reduction.html">35. Dimensionality Reduction</a></li>
<li class="toctree-l1"><a class="reference internal" href="../surface_protein/batch_correction.html">36. Batch correction</a></li>
<li class="toctree-l1"><a class="reference internal" href="../surface_protein/annotation.html">37. Annotation</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Adaptive immune receptor repertoire</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../air_repertoire/ir_profiling.html">38. Immune Receptor Profiling</a></li>
<li class="toctree-l1"><a class="reference internal" href="../air_repertoire/clonotype.html">39. Clonotype analysis</a></li>



<li class="toctree-l1"><a class="reference internal" href="../air_repertoire/specificity.html">43. Specificity analysis</a></li>
<li class="toctree-l1"><a class="reference internal" href="../air_repertoire/multimodal_integration.html">44. Integrating AIR and transcriptomics</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Multimodal integration</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../multimodal_integration/paired_integration.html">45. Paired integration</a></li>
<li class="toctree-l1"><a class="reference internal" href="../multimodal_integration/advanced_integration.html">46. Advanced integration</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Reproducibility</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../reproducibility/introduction.html">47. Reproducibility</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Outlook</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../outlook.html">48. Outlook</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Acknowledgements</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../acknowledgements.html">49. Acknowledgements</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Glossary</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../glossary.html">50. Glossary</a></li>
</ul>

    </div>
</nav></div>
    </div>
  
  
  <div class="sidebar-primary-items__end sidebar-primary__section">
      <div class="sidebar-primary-item">
<div id="ethical-ad-placement"
      class="flat"
      data-ea-publisher="readthedocs"
      data-ea-type="readthedocs-sidebar"
      data-ea-manual="true">
</div></div>
  </div>


      </div>
      
      <main id="main-content" class="bd-main" role="main">
        
        

<div class="sbt-scroll-pixel-helper"></div>

          <div class="bd-content">
            <div class="bd-article-container">
              
              <div class="bd-header-article d-print-none">
<div class="header-article-items header-article__inner">
  
    <div class="header-article-items__start">
      
        <div class="header-article-item"><button class="sidebar-toggle primary-toggle btn btn-sm" title="Toggle primary sidebar" data-bs-placement="bottom" data-bs-toggle="tooltip">
  <span class="fa-solid fa-bars"></span>
</button></div>
      
    </div>
  
  
    <div class="header-article-items__end">
      
        <div class="header-article-item">

<div class="article-header-buttons">





<div class="dropdown dropdown-source-buttons">
  <button class="btn dropdown-toggle" type="button" data-bs-toggle="dropdown" aria-expanded="false" aria-label="Source repositories">
    <i class="fab fa-github"></i>
  </button>
  <ul class="dropdown-menu">
      
      
      
      <li><a href="https://github.com/theislab/single-cell-best-practices" target="_blank"
   class="btn btn-sm btn-source-repository-button dropdown-item"
   title="Source repository"
   data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fab fa-github"></i>
  </span>
<span class="btn__text-container">Repository</span>
</a>
</li>
      
      
      
      
      <li><a href="https://github.com/theislab/single-cell-best-practices/edit/master/jupyter-book/mechanisms/cell_cell_communication.ipynb" target="_blank"
   class="btn btn-sm btn-source-edit-button dropdown-item"
   title="Suggest edit"
   data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-pencil-alt"></i>
  </span>
<span class="btn__text-container">Suggest edit</span>
</a>
</li>
      
      
      
      
      <li><a href="https://github.com/theislab/single-cell-best-practices/issues/new?title=Issue%20on%20page%20%2Fmechanisms/cell_cell_communication.html&body=Your%20issue%20content%20here." target="_blank"
   class="btn btn-sm btn-source-issues-button dropdown-item"
   title="Open an issue"
   data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-lightbulb"></i>
  </span>
<span class="btn__text-container">Open issue</span>
</a>
</li>
      
  </ul>
</div>






<div class="dropdown dropdown-download-buttons">
  <button class="btn dropdown-toggle" type="button" data-bs-toggle="dropdown" aria-expanded="false" aria-label="Download this page">
    <i class="fas fa-download"></i>
  </button>
  <ul class="dropdown-menu">
      
      
      
      <li><a href="../_sources/mechanisms/cell_cell_communication.ipynb" target="_blank"
   class="btn btn-sm btn-download-source-button dropdown-item"
   title="Download source file"
   data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-file"></i>
  </span>
<span class="btn__text-container">.ipynb</span>
</a>
</li>
      
      
      
      
      <li>
<button onclick="window.print()"
  class="btn btn-sm btn-download-pdf-button dropdown-item"
  title="Print to PDF"
  data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-file-pdf"></i>
  </span>
<span class="btn__text-container">.pdf</span>
</button>
</li>
      
  </ul>
</div>




<button onclick="toggleFullScreen()"
  class="btn btn-sm btn-fullscreen-button"
  title="Fullscreen mode"
  data-bs-placement="bottom" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-expand"></i>
  </span>

</button>



<button class="btn btn-sm nav-link pst-navbar-icon theme-switch-button pst-js-only" aria-label="Color mode" data-bs-title="Color mode"  data-bs-placement="bottom" data-bs-toggle="tooltip">
  <i class="theme-switch fa-solid fa-sun                fa-lg" data-mode="light" title="Light"></i>
  <i class="theme-switch fa-solid fa-moon               fa-lg" data-mode="dark"  title="Dark"></i>
  <i class="theme-switch fa-solid fa-circle-half-stroke fa-lg" data-mode="auto"  title="System Settings"></i>
</button>


<button class="btn btn-sm pst-navbar-icon search-button search-button__button pst-js-only" title="Search" aria-label="Search" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="fa-solid fa-magnifying-glass fa-lg"></i>
</button>
<button class="sidebar-toggle secondary-toggle btn btn-sm" title="Toggle secondary sidebar" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <span class="fa-solid fa-list"></span>
</button>
</div></div>
      
    </div>
  
</div>
</div>
              
              

<div id="jb-print-docs-body" class="onlyprint">
    <h1>Cell-cell communication</h1>
    <!-- Table of contents -->
    <div id="print-main-content">
        <div id="jb-print-toc">
            
            <div>
                <h2> Contents </h2>
            </div>
            <nav aria-label="Page">
                <ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#motivation">21.1. Motivation</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#approaches">21.2. Approaches</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#environment-setup">21.3. Environment setup</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#case-study">21.4. Case Study</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#ligand-receptor-inference">21.4.1. Ligand-receptor inference</a><ul class="nav section-nav flex-column">
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#visual-exploration">21.4.1.1. Visual exploration</a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#generating-a-ligand-receptor-consensus-with-liana">21.4.1.2. Generating a Ligand-Receptor consensus with LIANA</a></li>
</ul>
</li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#modelling-differential-intercellular-signalling-with-nichenet">21.4.2. Modelling differential intercellular signalling with NicheNet</a><ul class="nav section-nav flex-column">
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#load-nichenet-prior-knowledge">21.4.2.1. Load NicheNet Prior-Knowledge</a><ul class="nav section-nav flex-column">
<li class="toc-h5 nav-item toc-entry"><a class="reference internal nav-link" href="#step-1-define-cell-types-of-interest-to-be-considered-as-senders-sources-and-receiver-targets-of-ccc-interactions">21.4.2.1.1. Step 1. Define cell types of interest to be considered as senders/sources and receiver/targets of CCC interactions</a></li>
<li class="toc-h5 nav-item toc-entry"><a class="reference internal nav-link" href="#step-2-define-a-set-of-ligands-that-can-potentially-affect-receiver-cell-types">21.4.2.1.2. Step 2. Define a set of ligands that can <strong>potentially</strong> affect receiver cell types</a></li>
<li class="toc-h5 nav-item toc-entry"><a class="reference internal nav-link" href="#step-3-define-a-gene-set-of-interest-in-receiver-cell-type-s">21.4.2.1.3. Step 3. Define a gene set of interest in receiver cell type(s)</a></li>
<li class="toc-h5 nav-item toc-entry"><a class="reference internal nav-link" href="#step-4-nichenet-ligand-activity-estimation">21.4.2.1.4. Step 4. NicheNet ligand activity estimation</a></li>
<li class="toc-h5 nav-item toc-entry"><a class="reference internal nav-link" href="#step-5-infer-visualize-top-predicted-target-genes-for-top-ligands">21.4.2.1.5. Step 5. Infer &amp; Visualize top-predicted target genes for top ligands</a></li>
<li class="toc-h5 nav-item toc-entry"><a class="reference internal nav-link" href="#visualize-top-ligands-regulatory-targets">21.4.2.1.6. Visualize top ligands &amp; regulatory targets</a></li>
</ul>
</li>
</ul>
</li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#combining-nichenet-output-with-ligand-receptor-inference">21.4.3. Combining NicheNet output with ligand-receptor inference</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#key-takeaways">21.5. Key takeaways</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#assumptions-limitations">21.5.1. Assumptions &amp; Limitations</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#summary-outlook">21.5.2. Summary &amp; Outlook:</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#quiz">21.6. Quiz</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#contributions">21.7. Contributions</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#authors">21.7.1. Authors:</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#reviewers">21.7.2. Reviewers:</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#session-info">21.8. Session Info</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#references">21.9. References</a></li>
</ul>
            </nav>
        </div>
    </div>
</div>

              
                
<div id="searchbox"></div>
                <article class="bd-article">
                  
  <div class="dropdown admonition">
<p class="admonition-title">How do I set up an environment with the yml file used in this chapter?</p>
<div class="sd-tab-set docutils">
<input checked="checked" id="sd-tab-item-0" name="sd-tab-set-0" type="radio">
<label class="sd-tab-label" for="sd-tab-item-0">
Steps</label><div class="sd-tab-content docutils">
<ol class="arabic">
<li><p><strong>Install conda</strong>:</p>
<ul class="simple">
<li><p>Before creating the environment, ensure that <a class="reference external" href="https://docs.conda.io/projects/conda/en/latest/user-guide/install/index.html">conda</a> is installed on your system.</p></li>
</ul>
</li>
<li><p><strong>Save the yml content</strong>:</p>
<ul class="simple">
<li><p>Copy the content from the yml tab into a file named <code class="docutils literal notranslate"><span class="pre">environment.yml</span></code>.</p></li>
</ul>
</li>
<li><p><strong>Create the environment</strong>:</p>
<ul>
<li><p>Open a terminal or command prompt.</p></li>
<li><p>Run the following command:</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>conda<span class="w"> </span>env<span class="w"> </span>create<span class="w"> </span>-f<span class="w"> </span>environment.yml
</pre></div>
</div>
</li>
</ul>
</li>
<li><p><strong>Activate the environment</strong>:</p>
<ul>
<li><p>After the environment is created, activate it using:</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>conda<span class="w"> </span>activate<span class="w"> </span>&lt;environment_name&gt;
</pre></div>
</div>
</li>
<li><p>Replace <code class="docutils literal notranslate"><span class="pre">&lt;environment_name&gt;</span></code> with the name specified in the <code class="docutils literal notranslate"><span class="pre">environment.yml</span></code> file. In the yml file it will look like this:</p>
<div class="highlight-yaml notranslate"><div class="highlight"><pre><span></span><span class="nt">name</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">&lt;environment_name&gt;</span>
</pre></div>
</div>
</li>
</ul>
</li>
<li><p><strong>Verify the installation</strong>:</p>
<ul>
<li><p>Check that the environment was created successfully by running:</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>conda<span class="w"> </span>env<span class="w"> </span>list
</pre></div>
</div>
</li>
</ul>
</li>
</ol>
</div>
<input id="sd-tab-item-1" name="sd-tab-set-0" type="radio">
<label class="sd-tab-label" for="sd-tab-item-1">
yml</label><div class="sd-tab-content docutils">
<div class="highlight-yaml notranslate"><div class="highlight"><pre><span></span><span class="nt">name</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">cellcell</span>
<span class="nt">channels</span><span class="p">:</span>
<span class="w">  </span><span class="p p-Indicator">-</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">conda-forge</span>
<span class="w">  </span><span class="p p-Indicator">-</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">bioconda</span>
<span class="w">  </span><span class="p p-Indicator">-</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">r</span>
<span class="w">  </span><span class="p p-Indicator">-</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">defaults</span>
<span class="nt">dependencies</span><span class="p">:</span>
<span class="w">  </span><span class="p p-Indicator">-</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">conda-forge::r-base=4.2.2</span>
<span class="w">  </span><span class="p p-Indicator">-</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">conda-forge::pip</span>
<span class="w">  </span><span class="p p-Indicator">-</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">conda-forge::rpy2=3.5.1</span>
<span class="w">  </span><span class="p p-Indicator">-</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">conda-forge::scanpy=1.9.2</span>
<span class="w">  </span><span class="p p-Indicator">-</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">conda-forge::anndata=0.8.0</span>
<span class="w">  </span><span class="p p-Indicator">-</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">conda-forge::ipykernel</span>
<span class="w">  </span><span class="p p-Indicator">-</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">conda-forge::session-info</span>
<span class="w">  </span><span class="p p-Indicator">-</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">conda-forge::umap-learn</span>
<span class="w">  </span><span class="p p-Indicator">-</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">conda-forge::scikit-misc</span>
<span class="w">  </span><span class="p p-Indicator">-</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">bioconda::anndata2ri==1.1</span>
<span class="w">  </span><span class="p p-Indicator">-</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">bioconda::bioconductor-complexheatmap</span>
<span class="w">  </span><span class="p p-Indicator">-</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">bioconda::bioconductor-limma</span>
<span class="w">  </span><span class="p p-Indicator">-</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">r-seurat</span>
<span class="w">  </span><span class="p p-Indicator">-</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">r-dplyr</span>
<span class="w">  </span><span class="p p-Indicator">-</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">r-tidyr</span>
<span class="w">  </span><span class="p p-Indicator">-</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">r-ggplot2</span>
<span class="w">  </span><span class="p p-Indicator">-</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">r-purrr</span>
<span class="w">  </span><span class="p p-Indicator">-</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">r-diagrammer</span>
<span class="w">  </span><span class="p p-Indicator">-</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">r-mlrmbo</span>
<span class="w">  </span><span class="p p-Indicator">-</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">r-dicekriging</span>
<span class="w">  </span><span class="p p-Indicator">-</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">r-hmisc</span>
<span class="w">  </span><span class="p p-Indicator">-</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">r-remotes</span>
<span class="w">  </span><span class="p p-Indicator">-</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">r-randomforest</span>
<span class="w">  </span><span class="p p-Indicator">-</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">r-diffobj</span>
<span class="w">  </span><span class="p p-Indicator">-</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">r-rstatix</span>
<span class="w">  </span><span class="p p-Indicator">-</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">r-ggpubr</span>
<span class="w">  </span><span class="p p-Indicator">-</span><span class="w"> </span><span class="nt">pip</span><span class="p">:</span>
<span class="w">      </span><span class="p p-Indicator">-</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">adjusttext==0.7.3</span>
<span class="w">      </span><span class="p p-Indicator">-</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">decoupler==1.3.3</span>
<span class="w">      </span><span class="p p-Indicator">-</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">liana==0.1.6</span>
<span class="w">      </span><span class="p p-Indicator">-</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">mizani==0.8.1</span>
<span class="w">      </span><span class="p p-Indicator">-</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">palettable==3.3.0</span>
<span class="w">      </span><span class="p p-Indicator">-</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">plotnine==0.10.1</span>
</pre></div>
</div>
</div>
</div>
</div>
<section class="tex2jax_ignore mathjax_ignore" id="cell-cell-communication">
<span id="cell-cell"></span><h1><span class="section-number">21. </span>Cell-cell communication<a class="headerlink" href="#cell-cell-communication" title="Link to this heading">#</a></h1>
<p><em>TL;DR This chapter serves as a brief overview of basic concepts and assumptions of cell-cell communication inference from single-cell transcriptomics data. We provide examples of the two most common approaches for CCC inference: those that focus on ligand-receptor interactions and those that include downstream response.</em></p>
<section id="motivation">
<h2><span class="section-number">21.1. </span>Motivation<a class="headerlink" href="#motivation" title="Link to this heading">#</a></h2>
<p>Cell communication is a process by which cells react to stimuli from their environment and also from themselves. In multicellular organisms, the dynamic coordination of cells, also called cell-cell communication (CCC), is involved in many biological processes, such as apoptosis and cell migration, and is consequently essential in homeostasis and disease. CCC commonly focuses on protein-mediated interactions, most typically perceived as a secreted ligand binding to its corresponding plasma membrane receptor. However, this picture can be broadened to include secreted enzymes, extra-cellular matrix proteins, transporters, and interactions that require the physical contact between cells, such as cell-cell adhesion proteins and gap junctions <span id="id1">[<a class="reference internal" href="#id413" title="Erick Armingol, Adam Officer, Olivier Harismendy, and Nathan E Lewis. Deciphering cell-cell interactions and communication from gene expression. Nature Reviews. Genetics, 22(2):71-88, 2021. URL: http://www.nature.com/articles/s41576-020-00292-x (visited on 2021-10-05), doi:10.1038/s41576-020-00292-x.">Armingol <em>et al.</em>, 2021</a>]</span>. Cell communication is further not independent of other processes, but the contrary, as external stimuli commonly elicit a downstream response. In the case of CCC, this is typically perceived as the induction of canonical pathways and downstream transcription factors in the cells receiving the signal, or receiver cells. Ultimately these external stimuli alter the function of receiver cells, and this alteration is further propagated via the subsequent interaction of these cells with their microenvironment.
Traditionally, the study of CCC required specialized in-situ biochemical assays, such proximity labelling proteomics, co-immunoprecipitation, and yeast two-hybrid screening <span id="id2">[<a class="reference internal" href="#id413" title="Erick Armingol, Adam Officer, Olivier Harismendy, and Nathan E Lewis. Deciphering cell-cell interactions and communication from gene expression. Nature Reviews. Genetics, 22(2):71-88, 2021. URL: http://www.nature.com/articles/s41576-020-00292-x (visited on 2021-10-05), doi:10.1038/s41576-020-00292-x.">Armingol <em>et al.</em>, 2021</a>]</span>. Yet, the rapid developments and dropping costs of transcriptomics data generation has enabled a paradigm shift away from focusing on which types of cells are present, but rather on the relationships between them <span id="id3">[<a class="reference internal" href="#id416" title="Axel A Almet, Zixuan Cang, Suoqin Jin, and Qing Nie. The landscape of cell-cell communication through single-cell transcriptomics. Current Opinion in Systems Biology, 26:12-23, jun 2021. URL: https://linkinghub.elsevier.com/retrieve/pii/S2452310021000081 (visited on 2021-09-06), doi:10.1016/j.coisb.2021.03.007.">Almet <em>et al.</em>, 2021</a>]</span>. As a consequence, CCC inference from single-cell data is now becoming a routine approach, capable of providing a system-level hypotheses of intercellular crosstalk in vivo.</p>
</section>
<section id="approaches">
<h2><span class="section-number">21.2. </span>Approaches<a class="headerlink" href="#approaches" title="Link to this heading">#</a></h2>
<p>As a result of this increased interest, a number of computational tools for CCC inference from single-cell transcriptomics have emerged that can be classified as those that predict CCC interactions alone, commonly referred to as ligand-receptor inference methods (e.g. <span id="id4">[<a class="reference internal" href="#id412" title="Mirjana Efremova, Miquel Vento-Tormo, Sarah A Teichmann, and Roser Vento-Tormo. CellPhoneDB: inferring cell-cell communication from combined expression of multi-subunit ligand-receptor complexes. Nature Protocols, 15(4):1484-1506, apr 2020. URL: http://www.nature.com/articles/s41596-020-0292-x (visited on 2020-12-04), doi:10.1038/s41596-020-0292-x.">Efremova <em>et al.</em>, 2020</a>, <a class="reference internal" href="#id429" title="Rui Hou, Elena Denisenko, Huan Ting Ong, Jordan A Ramilowski, and Alistair RR Forrest. Predicting cell-to-cell communication networks using natmi. Nature communications, 11(1):1–11, 2020.">Hou <em>et al.</em>, 2020</a>, <a class="reference internal" href="#id415" title="Suoqin Jin, Christian F Guerrero-Juarez, Lihua Zhang, Ivan Chang, Raul Ramos, Chen-Hsiang Kuan, Peggy Myung, Maksim V Plikus, and Qing Nie. Inference and analysis of cell-cell communication using CellChat. Nature Communications, 12(1):1088, feb 2021. URL: http://dx.doi.org/10.1038/s41467-021-21246-9 (visited on 2021-10-05), doi:10.1038/s41467-021-21246-9.">Jin <em>et al.</em>, 2021</a>, <a class="reference internal" href="#id422" title="Micha Sam Brickman Raredon, Junchen Yang, James Garritano, Meng Wang, Dan Kushnir, Jonas Christian Schupp, Taylor S Adams, Allison M Greaney, Katherine L Leiby, Naftali Kaminski, Yuval Kluger, Andre Levchenko, and Laura E Niklason. Computation and visualization of cell-cell signaling topologies in single-cell systems data using connectome. Scientific Reports, 12(1):4187, mar 2022. URL: http://dx.doi.org/10.1038/s41598-022-07959-x (visited on 2021-01-25), doi:10.1038/s41598-022-07959-x.">Raredon <em>et al.</em>, 2022</a>]</span>), and those that additionally estimate intracellular activities induced by CCC (e.g. <span id="id5">[<a class="reference internal" href="#id411" title="Robin Browaeys, Wouter Saelens, and Yvan Saeys. NicheNet: modeling intercellular communication by linking ligands to target genes. Nature Methods, 17(2):159-162, feb 2020. URL: http://www.nature.com/articles/s41592-019-0667-5 (visited on 2021-10-05), doi:10.1038/s41592-019-0667-5.">Browaeys <em>et al.</em>, 2020</a>, <a class="reference internal" href="#id418" title="Yuxuan Hu, Tao Peng, Lin Gao, and Kai Tan. CytoTalk: de novo construction of signal transduction networks using single-cell transcriptomic data. Science Advances, apr 2021. URL: https://advances.sciencemag.org/lookup/doi/10.1126/sciadv.abf1356 (visited on 2021-10-11), doi:10.1126/sciadv.abf1356.">Hu <em>et al.</em>, 2021</a>, <a class="reference internal" href="#id410" title="Shuxiong Wang, Matthew Karikomi, Adam L MacLean, and Qing Nie. Cell lineage and communication network inference via optimization for single-cell transcriptomics. Nucleic Acids Research, 47(11):e66, jun 2019. URL: https://academic.oup.com/nar/article/47/11/e66/5421812 (visited on 2019-05-01), doi:10.1093/nar/gkz204.">Wang <em>et al.</em>, 2019</a>]</span>). Both categories of tools use gene expression information as a proxy of protein abundance, and typically require the clustering of cells into biologically-meaningful groups (See Annotation tutorial). These CCC tools infer intercellular crosstalk between pairs of cell groups, one group being the source and the other the receiver of a CCC event. CCC events are thus commonly represented as interactions between proteins, expressed by the source and receiver cell clusters.</p>
<p>The information about the interacting proteins is commonly extracted from prior knowledge resources. In the case of ligand-receptor methods, the interactions can also be represented by heteromeric protein complexes, as different subunit combinations can induce distinct responses and the inclusion of protein complex information has been shown to reduce false positive rates <span id="id6">[<a class="reference internal" href="#id412" title="Mirjana Efremova, Miquel Vento-Tormo, Sarah A Teichmann, and Roser Vento-Tormo. CellPhoneDB: inferring cell-cell communication from combined expression of multi-subunit ligand-receptor complexes. Nature Protocols, 15(4):1484-1506, apr 2020. URL: http://www.nature.com/articles/s41596-020-0292-x (visited on 2020-12-04), doi:10.1038/s41596-020-0292-x.">Efremova <em>et al.</em>, 2020</a>, <a class="reference internal" href="#id415" title="Suoqin Jin, Christian F Guerrero-Juarez, Lihua Zhang, Ivan Chang, Raul Ramos, Chen-Hsiang Kuan, Peggy Myung, Maksim V Plikus, and Qing Nie. Inference and analysis of cell-cell communication using CellChat. Nature Communications, 12(1):1088, feb 2021. URL: http://dx.doi.org/10.1038/s41467-021-21246-9 (visited on 2021-10-05), doi:10.1038/s41467-021-21246-9.">Jin <em>et al.</em>, 2021</a>, <a class="reference internal" href="#id426" title="Zhaoyang Liu, Dongqing Sun, and Chenfei Wang. Evaluation of cell-cell interaction methods by integrating single-cell RNA sequencing data with spatial information. Genome Biology, 23(1):218, oct 2022. URL: http://dx.doi.org/10.1186/s13059-022-02783-y (visited on 2022-10-20), doi:10.1186/s13059-022-02783-y.">Liu <em>et al.</em>, 2022</a>]</span>. On the other hand, the approaches that model intracellular signalling also leverage the functional information in receiver cell types, and thus require additional information such as intracellular protein-protein interaction network and/or gene regulatory interactions.</p>
<p>Recent work has highlighted that the choice of method and/or resource leads to limited consensus in inferred predictions when using different tools <span id="id7">[<a class="reference internal" href="#id423" title="Daniel Dimitrov, Dénes Türei, Martin Garrido-Rodriguez, Paul L Burmedi, James S Nagai, Charlotte Boys, Ricardo O Ramirez Flores, Hyojin Kim, Bence Szalai, Ivan G Costa, Alberto Valdeolivas, Aurélien Dugourd, and Julio Saez-Rodriguez. Comparison of methods and resources for cell-cell communication inference from single-cell RNA-seq data. Nature Communications, 13(1):3224, jun 2022. URL: https://www.nature.com/articles/s41467-022-30755-0 (visited on 2021-05-31), doi:10.1038/s41467-022-30755-0.">Dimitrov <em>et al.</em>, 2022</a>, <a class="reference internal" href="#id426" title="Zhaoyang Liu, Dongqing Sun, and Chenfei Wang. Evaluation of cell-cell interaction methods by integrating single-cell RNA sequencing data with spatial information. Genome Biology, 23(1):218, oct 2022. URL: http://dx.doi.org/10.1186/s13059-022-02783-y (visited on 2022-10-20), doi:10.1186/s13059-022-02783-y.">Liu <em>et al.</em>, 2022</a>, <a class="reference internal" href="#id428" title="Saidi Wang, Hansi Zheng, James S Choi, Jae K Lee, Xiaoman Li, and Haiyan Hu. A systematic evaluation of the computational tools for ligand-receptor-based cell-cell interaction inference. Briefings in functional genomics, 21(5):339-356, sep 2022. URL: http://dx.doi.org/10.1093/bfgp/elac019 (visited on 2022-04-27), doi:10.1093/bfgp/elac019.">Wang <em>et al.</em>, 2022</a>]</span>, thus prompting caution when interpreting their output. The CCC field is further plagued by the lack of ground truth <span id="id8">[<a class="reference internal" href="#id416" title="Axel A Almet, Zixuan Cang, Suoqin Jin, and Qing Nie. The landscape of cell-cell communication through single-cell transcriptomics. Current Opinion in Systems Biology, 26:12-23, jun 2021. URL: https://linkinghub.elsevier.com/retrieve/pii/S2452310021000081 (visited on 2021-09-06), doi:10.1016/j.coisb.2021.03.007.">Almet <em>et al.</em>, 2021</a>, <a class="reference internal" href="#id413" title="Erick Armingol, Adam Officer, Olivier Harismendy, and Nathan E Lewis. Deciphering cell-cell interactions and communication from gene expression. Nature Reviews. Genetics, 22(2):71-88, 2021. URL: http://www.nature.com/articles/s41576-020-00292-x (visited on 2021-10-05), doi:10.1038/s41576-020-00292-x.">Armingol <em>et al.</em>, 2021</a>]</span>, capable of capturing the complex and dynamic interplay between large numbers of cells and molecules. Nevertheless, independent evaluations have shown that CCC methods are fairly robust to the introduction of noise <span id="id9">[<a class="reference internal" href="#id423" title="Daniel Dimitrov, Dénes Türei, Martin Garrido-Rodriguez, Paul L Burmedi, James S Nagai, Charlotte Boys, Ricardo O Ramirez Flores, Hyojin Kim, Bence Szalai, Ivan G Costa, Alberto Valdeolivas, Aurélien Dugourd, and Julio Saez-Rodriguez. Comparison of methods and resources for cell-cell communication inference from single-cell RNA-seq data. Nature Communications, 13(1):3224, jun 2022. URL: https://www.nature.com/articles/s41467-022-30755-0 (visited on 2021-05-31), doi:10.1038/s41467-022-30755-0.">Dimitrov <em>et al.</em>, 2022</a>, <a class="reference internal" href="#id426" title="Zhaoyang Liu, Dongqing Sun, and Chenfei Wang. Evaluation of cell-cell interaction methods by integrating single-cell RNA sequencing data with spatial information. Genome Biology, 23(1):218, oct 2022. URL: http://dx.doi.org/10.1186/s13059-022-02783-y (visited on 2022-10-20), doi:10.1186/s13059-022-02783-y.">Liu <em>et al.</em>, 2022</a>, <a class="reference internal" href="#id428" title="Saidi Wang, Hansi Zheng, James S Choi, Jae K Lee, Xiaoman Li, and Haiyan Hu. A systematic evaluation of the computational tools for ligand-receptor-based cell-cell interaction inference. Briefings in functional genomics, 21(5):339-356, sep 2022. URL: http://dx.doi.org/10.1093/bfgp/elac019 (visited on 2022-04-27), doi:10.1093/bfgp/elac019.">Wang <em>et al.</em>, 2022</a>]</span>, and are largely concordant with alternative data modalities such as intracellular signalling and spatial information <span id="id10">[<a class="reference internal" href="#id423" title="Daniel Dimitrov, Dénes Türei, Martin Garrido-Rodriguez, Paul L Burmedi, James S Nagai, Charlotte Boys, Ricardo O Ramirez Flores, Hyojin Kim, Bence Szalai, Ivan G Costa, Alberto Valdeolivas, Aurélien Dugourd, and Julio Saez-Rodriguez. Comparison of methods and resources for cell-cell communication inference from single-cell RNA-seq data. Nature Communications, 13(1):3224, jun 2022. URL: https://www.nature.com/articles/s41467-022-30755-0 (visited on 2021-05-31), doi:10.1038/s41467-022-30755-0.">Dimitrov <em>et al.</em>, 2022</a>, <a class="reference internal" href="#id426" title="Zhaoyang Liu, Dongqing Sun, and Chenfei Wang. Evaluation of cell-cell interaction methods by integrating single-cell RNA sequencing data with spatial information. Genome Biology, 23(1):218, oct 2022. URL: http://dx.doi.org/10.1186/s13059-022-02783-y (visited on 2022-10-20), doi:10.1186/s13059-022-02783-y.">Liu <em>et al.</em>, 2022</a>]</span>.</p>
<p>In this chapter, we first provide an introduction and example to perhaps the most common and simplest CCC approaches, i.e. ligand-receptor inference with CellPhoneDB <span id="id11">[<a class="reference internal" href="#id412" title="Mirjana Efremova, Miquel Vento-Tormo, Sarah A Teichmann, and Roser Vento-Tormo. CellPhoneDB: inferring cell-cell communication from combined expression of multi-subunit ligand-receptor complexes. Nature Protocols, 15(4):1484-1506, apr 2020. URL: http://www.nature.com/articles/s41596-020-0292-x (visited on 2020-12-04), doi:10.1038/s41596-020-0292-x.">Efremova <em>et al.</em>, 2020</a>]</span> and LIANA <span id="id12">[<a class="reference internal" href="#id423" title="Daniel Dimitrov, Dénes Türei, Martin Garrido-Rodriguez, Paul L Burmedi, James S Nagai, Charlotte Boys, Ricardo O Ramirez Flores, Hyojin Kim, Bence Szalai, Ivan G Costa, Alberto Valdeolivas, Aurélien Dugourd, and Julio Saez-Rodriguez. Comparison of methods and resources for cell-cell communication inference from single-cell RNA-seq data. Nature Communications, 13(1):3224, jun 2022. URL: https://www.nature.com/articles/s41467-022-30755-0 (visited on 2021-05-31), doi:10.1038/s41467-022-30755-0.">Dimitrov <em>et al.</em>, 2022</a>]</span>. Then we showcase NicheNet as an example of a CCC inference method with a focus on intracellular activities downstream of CCC events <span id="id13">[<a class="reference internal" href="#id411" title="Robin Browaeys, Wouter Saelens, and Yvan Saeys. NicheNet: modeling intercellular communication by linking ligands to target genes. Nature Methods, 17(2):159-162, feb 2020. URL: http://www.nature.com/articles/s41592-019-0667-5 (visited on 2021-10-05), doi:10.1038/s41592-019-0667-5.">Browaeys <em>et al.</em>, 2020</a>]</span>. Finally, we highlight the common assumptions and limitations of CCC inference from single-cell transcriptomics data as well as suggestions how to improve confidence in intercellular communication predictions. Note that here we generalize for the sake of simplicity, however, there are a plethora of different and newly emerging CCC approaches. We highlight some of those in the <strong>Outlook</strong> section.</p>
<a class="bg-primary mb-1 reference internal image-reference" href="../_images/cell_cell_communication.png"><img alt="CCC Overview" class="bg-primary mb-1" src="../_images/cell_cell_communication.png" style="width: 800px;" /></a>
<p><strong>FIGURE 22.1</strong>. Cell-Cell Communication Overview</p>
</section>
<section id="environment-setup">
<h2><span class="section-number">21.3. </span>Environment setup<a class="headerlink" href="#environment-setup" title="Link to this heading">#</a></h2>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># python libs</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">decoupler</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">dc</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">liana</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">li</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">matplotlib.pyplot</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">plt</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">numpy</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">np</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">pandas</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">pd</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">scanpy</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">sc</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">seaborn</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">sns</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">session_info</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Setting up R dependencies</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">anndata2ri</span>

<span class="n">anndata2ri</span><span class="o">.</span><span class="n">activate</span><span class="p">()</span>

<span class="o">%</span><span class="k">load_ext</span> rpy2.ipython
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="o">%%</span><span class="k">R</span>
suppressPackageStartupMessages({
    library(reticulate)
    library(ggplot2)
    library(tidyr)
    library(dplyr)
    library(purrr)
    library(tibble)
})
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># figure settings</span>
<span class="n">sc</span><span class="o">.</span><span class="n">settings</span><span class="o">.</span><span class="n">set_figure_params</span><span class="p">(</span><span class="n">dpi</span><span class="o">=</span><span class="mi">200</span><span class="p">,</span> <span class="n">frameon</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
<span class="n">sc</span><span class="o">.</span><span class="n">set_figure_params</span><span class="p">(</span><span class="n">dpi</span><span class="o">=</span><span class="mi">200</span><span class="p">,</span> <span class="n">facecolor</span><span class="o">=</span><span class="s2">&quot;white&quot;</span><span class="p">)</span>
<span class="n">sc</span><span class="o">.</span><span class="n">set_figure_params</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">5</span><span class="p">,</span> <span class="mi">5</span><span class="p">))</span>
</pre></div>
</div>
</div>
</div>
</section>
<section id="case-study">
<h2><span class="section-number">21.4. </span>Case Study<a class="headerlink" href="#case-study" title="Link to this heading">#</a></h2>
<p>As a simple example, we will look at ~25k PBMCs from 8 lupus patients, each before and after IFN-β stimulation <span id="id14">[<a class="reference internal" href="#id432" title="Hyun Min Kang, Meena Subramaniam, Sasha Targ, Michelle Nguyen, Lenka Maliskova, Elizabeth McCarthy, Eunice Wan, Simon Wong, Lauren Byrnes, Cristina M Lanata, and others. Multiplexed droplet single-cell rna-sequencing using natural genetic variation. Nature biotechnology, 36(1):89–94, 2018.">Kang <em>et al.</em>, 2018</a>]</span>. Note that by focusing on PBMCs, for the purpose of this tutorial, we assume that coordinated events occur among them.</p>
<p>So, let’s first download the pre-processed data.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Read in</span>
<span class="n">adata</span> <span class="o">=</span> <span class="n">sc</span><span class="o">.</span><span class="n">read</span><span class="p">(</span>
    <span class="s2">&quot;kang_counts_25k.h5ad&quot;</span><span class="p">,</span> <span class="n">backup_url</span><span class="o">=</span><span class="s2">&quot;https://figshare.com/ndownloader/files/34464122&quot;</span>
<span class="p">)</span>

<span class="c1"># Store the counts for later use</span>
<span class="n">adata</span><span class="o">.</span><span class="n">layers</span><span class="p">[</span><span class="s2">&quot;counts&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">adata</span><span class="o">.</span><span class="n">X</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
</pre></div>
</div>
</div>
</div>
<p>Apply basic quality control steps to remove any low quality cells and lowly expressed genes. We refer the user to the <a class="reference internal" href="../preprocessing_visualization/quality_control.html#quality-control"><span class="std std-ref">Quality Control</span></a> chapter for more extensive QC steps.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">sc</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">filter_cells</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">min_genes</span><span class="o">=</span><span class="mi">200</span><span class="p">)</span>
<span class="n">sc</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">filter_genes</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">min_cells</span><span class="o">=</span><span class="mi">3</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Store the counts for later use</span>
<span class="n">adata</span><span class="o">.</span><span class="n">layers</span><span class="p">[</span><span class="s2">&quot;counts&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">adata</span><span class="o">.</span><span class="n">X</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
<span class="c1"># Rename label to condition, replicate to patient</span>
<span class="n">adata</span><span class="o">.</span><span class="n">obs</span> <span class="o">=</span> <span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="o">.</span><span class="n">rename</span><span class="p">({</span><span class="s2">&quot;label&quot;</span><span class="p">:</span> <span class="s2">&quot;condition&quot;</span><span class="p">,</span> <span class="s2">&quot;replicate&quot;</span><span class="p">:</span> <span class="s2">&quot;patient&quot;</span><span class="p">},</span> <span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
<span class="c1"># assign sample</span>
<span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s2">&quot;sample&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="p">(</span>
    <span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s2">&quot;condition&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="s2">&quot;str&quot;</span><span class="p">)</span> <span class="o">+</span> <span class="s2">&quot;&amp;&quot;</span> <span class="o">+</span> <span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s2">&quot;patient&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="s2">&quot;str&quot;</span><span class="p">)</span>
<span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<p>We will also normalize the data to ensure that the count depth is equalized for all cells, given that we need the gene expression values to be comparable across the cell types. We refer the user to the <a class="reference internal" href="../preprocessing_visualization/normalization.html#normalization"><span class="std std-ref">Normalization</span></a> chapter for more information and alternative normalization approaches that might fit their data better.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># log1p normalize the data</span>
<span class="n">sc</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">normalize_total</span><span class="p">(</span><span class="n">adata</span><span class="p">)</span>
<span class="n">sc</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">log1p</span><span class="p">(</span><span class="n">adata</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<p>In this case study, we will assume that cell types such as B cells and CD4 T cells carry out a signal mediator role, while others, such as CD8 T cells and Natural Killer cells, are composed of the cells that carry out the response. In other words, we will treat B and CD4 T cells as the sources of CCC signalling, while the latter are the receivers of CCC stimuli. This is of course an oversimplification as signalling sources and receivers are expected to be dynamic and multi-directional, thus the cell types that we treat as which category depends on the hypothesis in mind.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s2">&quot;cell_type&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">cat</span><span class="o">.</span><span class="n">categories</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Index([&#39;CD4 T cells&#39;, &#39;CD14+ Monocytes&#39;, &#39;B cells&#39;, &#39;NK cells&#39;, &#39;CD8 T cells&#39;,
       &#39;FCGR3A+ Monocytes&#39;, &#39;Dendritic cells&#39;, &#39;Megakaryocytes&#39;],
      dtype=&#39;object&#39;)
</pre></div>
</div>
</div>
</div>
<p>Show pre-computed UMAP, just to showcase the data</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">sc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">umap</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;condition&quot;</span><span class="p">,</span> <span class="s2">&quot;cell_type&quot;</span><span class="p">],</span> <span class="n">frameon</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>/home/dbdimitrov/anaconda3/envs/cellcell/lib/python3.10/site-packages/scanpy/plotting/_tools/scatterplots.py:364: UserWarning: No data for colormapping provided via &#39;c&#39;. Parameters &#39;cmap&#39; will be ignored
/home/dbdimitrov/anaconda3/envs/cellcell/lib/python3.10/site-packages/scanpy/plotting/_tools/scatterplots.py:364: UserWarning: No data for colormapping provided via &#39;c&#39;. Parameters &#39;cmap&#39; will be ignored
</pre></div>
</div>
<img alt="../_images/a4b490b024600783118bb420c29789ab0b1e8245edbf45a8afa421e35f12fd0b.png" src="../_images/a4b490b024600783118bb420c29789ab0b1e8245edbf45a8afa421e35f12fd0b.png" />
</div>
</div>
<section id="ligand-receptor-inference">
<h3><span class="section-number">21.4.1. </span>Ligand-receptor inference<a class="headerlink" href="#ligand-receptor-inference" title="Link to this heading">#</a></h3>
<p>First, let’s use the CellPhoneDB (v2) ligand-receptor method <span id="id15">[<a class="reference internal" href="#id412" title="Mirjana Efremova, Miquel Vento-Tormo, Sarah A Teichmann, and Roser Vento-Tormo. CellPhoneDB: inferring cell-cell communication from combined expression of multi-subunit ligand-receptor complexes. Nature Protocols, 15(4):1484-1506, apr 2020. URL: http://www.nature.com/articles/s41596-020-0292-x (visited on 2020-12-04), doi:10.1038/s41596-020-0292-x.">Efremova <em>et al.</em>, 2020</a>]</span>.</p>
<p>We will run CellPhoneDB on the data post IFN-beta stimulation alone, as such methods were initially designed for the inference of CCC events in “steady-state” data, or in other words, they are meant to be used not across samples or conditions, but rather on a single condition or sample at a time. Note that, nevertheless, certain approaches exist to apply ligand-receptor methods across conditions but these are out of scope for this tutorial, and instead we refer to them in the <strong>Outlook</strong> section.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">adata_stim</span> <span class="o">=</span> <span class="n">adata</span><span class="p">[</span><span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s2">&quot;condition&quot;</span><span class="p">]</span> <span class="o">==</span> <span class="s2">&quot;stim&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
<span class="n">adata_stim</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>AnnData object with n_obs × n_vars = 12301 × 15701
    obs: &#39;nCount_RNA&#39;, &#39;nFeature_RNA&#39;, &#39;tsne1&#39;, &#39;tsne2&#39;, &#39;condition&#39;, &#39;cluster&#39;, &#39;cell_type&#39;, &#39;patient&#39;, &#39;nCount_SCT&#39;, &#39;nFeature_SCT&#39;, &#39;integrated_snn_res.0.4&#39;, &#39;seurat_clusters&#39;, &#39;n_genes&#39;, &#39;sample&#39;
    var: &#39;name&#39;, &#39;n_cells&#39;
    uns: &#39;log1p&#39;, &#39;condition_colors&#39;, &#39;cell_type_colors&#39;
    obsm: &#39;X_pca&#39;, &#39;X_umap&#39;
    layers: &#39;counts&#39;
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># import cellphonedb method via liana</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">liana.method</span><span class="w"> </span><span class="kn">import</span> <span class="n">cellphonedb</span>
</pre></div>
</div>
</div>
</div>
<p>CellPhoneDB is one of the most commonly-used CCC tools, it represents intercellular communication events as the average gene expression of the proteins involved in the interaction. The proteins involved can also be can also take the form of heteromeric complexes, and in that case the minimum gene expression of the subunits is considered. In addition to the expression average, interaction significance is determined against a null distribution, generated by shuffling the cell group labels.</p>
<p>Note that we are grouping by cell type and that the CCC statistics that we get will reflect the cell types that were previously pre-defined.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">cellphonedb</span><span class="p">(</span>
    <span class="n">adata_stim</span><span class="p">,</span> <span class="n">groupby</span><span class="o">=</span><span class="s2">&quot;cell_type&quot;</span><span class="p">,</span> <span class="n">use_raw</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">return_all_lrs</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">verbose</span><span class="o">=</span><span class="kc">True</span>
<span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Using `.X`!
Converting mat to CSR format
227 features of mat are empty, they will be removed.
</pre></div>
</div>
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>/home/dbdimitrov/anaconda3/envs/cellcell/lib/python3.10/site-packages/pandas/core/indexing.py:1728: ImplicitModificationWarning: Trying to modify attribute `.obs` of view, initializing view as actual.
</pre></div>
</div>
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>0.46 of entities in the resource are missing from the data.
Generating ligand-receptor stats for 12301 samples and 15474 features
</pre></div>
</div>
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>100%|██████████| 1000/1000 [00:22&lt;00:00, 45.07it/s]
</pre></div>
</div>
</div>
</div>
<p>By default, the results are written in place within the anndata object, more specifically in <code class="docutils literal notranslate"><span class="pre">.uns['liana_res']</span></code>.</p>
<p>Let’s examine the output from the CellPhoneDB method:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">adata_stim</span><span class="o">.</span><span class="n">uns</span><span class="p">[</span><span class="s2">&quot;liana_res&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">head</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>ligand</th>
      <th>ligand_complex</th>
      <th>ligand_means</th>
      <th>ligand_props</th>
      <th>receptor</th>
      <th>receptor_complex</th>
      <th>receptor_means</th>
      <th>receptor_props</th>
      <th>source</th>
      <th>target</th>
      <th>lrs_to_keep</th>
      <th>lr_means</th>
      <th>cellphone_pvals</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>LGALS9</td>
      <td>LGALS9</td>
      <td>0.072739</td>
      <td>0.101973</td>
      <td>PTPRC</td>
      <td>PTPRC</td>
      <td>0.457824</td>
      <td>0.501761</td>
      <td>CD4 T cells</td>
      <td>CD4 T cells</td>
      <td>True</td>
      <td>0.265281</td>
      <td>1.0</td>
    </tr>
    <tr>
      <th>9</th>
      <td>LGALS9</td>
      <td>LGALS9</td>
      <td>0.072739</td>
      <td>0.101973</td>
      <td>CD44</td>
      <td>CD44</td>
      <td>0.304042</td>
      <td>0.364741</td>
      <td>CD4 T cells</td>
      <td>CD4 T cells</td>
      <td>True</td>
      <td>0.188391</td>
      <td>1.0</td>
    </tr>
    <tr>
      <th>37</th>
      <td>VIM</td>
      <td>VIM</td>
      <td>0.507059</td>
      <td>0.519725</td>
      <td>CD44</td>
      <td>CD44</td>
      <td>0.304042</td>
      <td>0.364741</td>
      <td>CD4 T cells</td>
      <td>CD4 T cells</td>
      <td>True</td>
      <td>0.405550</td>
      <td>1.0</td>
    </tr>
    <tr>
      <th>38</th>
      <td>PKM</td>
      <td>PKM</td>
      <td>0.230863</td>
      <td>0.278267</td>
      <td>CD44</td>
      <td>CD44</td>
      <td>0.304042</td>
      <td>0.364741</td>
      <td>CD4 T cells</td>
      <td>CD4 T cells</td>
      <td>True</td>
      <td>0.267453</td>
      <td>1.0</td>
    </tr>
    <tr>
      <th>66</th>
      <td>LGALS9</td>
      <td>LGALS9</td>
      <td>0.072739</td>
      <td>0.101973</td>
      <td>CD47</td>
      <td>CD47</td>
      <td>0.213795</td>
      <td>0.272807</td>
      <td>CD4 T cells</td>
      <td>CD4 T cells</td>
      <td>True</td>
      <td>0.143267</td>
      <td>1.0</td>
    </tr>
  </tbody>
</table>
</div></div></div>
</div>
<p>Here, we see that stats are provided for both ligand and receptor entities, more specifically: - <code class="docutils literal notranslate"><span class="pre">ligand</span></code> and <code class="docutils literal notranslate"><span class="pre">receptor</span></code> are typically the two entities that interact. As a reminder, CCC events are not limited to secreted signalling, but we refer to them as <code class="docutils literal notranslate"><span class="pre">ligand</span></code> and <code class="docutils literal notranslate"><span class="pre">receptor</span></code> for simplicity.</p>
<p>Also, in the case of heteromeric complexes, the <code class="docutils literal notranslate"><span class="pre">ligand</span></code> and <code class="docutils literal notranslate"><span class="pre">receptor</span></code> columns represent the subunit with minimum expression, while <code class="docutils literal notranslate"><span class="pre">*_complex</span></code> corresponds to the actual complex, with subunits being separated by <code class="docutils literal notranslate"><span class="pre">_</span></code>.</p>
<ul>
<li><p><code class="docutils literal notranslate"><span class="pre">source</span></code> and <code class="docutils literal notranslate"><span class="pre">target</span></code> columns represent the source/sender and target/receiver cell identity for each interaction, respectively</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">*_props</span></code>: represents the proportion of cells that express the entity.</p>
<p>By default, in CellPhoneDB and LIANA, any interactions in which either entity is not expressed in above 10% of cells per cell type is considered as a false positive,
under the assumption that since CCC occurs between cell types, a sufficient proportion of cells within should express the genes.</p>
</li>
<li><p><code class="docutils literal notranslate"><span class="pre">*_means</span></code>: entity expression mean per cell type</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">lr_means</span></code>: mean ligand-receptor expression, as a measure of ligand-receptor interaction <strong>magnitude</strong></p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">cellphone_pvals</span></code>: permutation-based p-values, as a measure of interaction <strong>specificity</strong></p></li>
</ul>
<p>Note that <code class="docutils literal notranslate"><span class="pre">ligand</span></code>, <code class="docutils literal notranslate"><span class="pre">receptor</span></code>, <code class="docutils literal notranslate"><span class="pre">source</span></code>, and <code class="docutils literal notranslate"><span class="pre">target</span></code> columns are returned by every ligand-receptor method, while the rest of the columns can vary across the ligand-receptor methods, as each method relies on different assumptions and scoring functions, and hence each returns different ligand-receptor scores.
Nevertheless, typically most methods use a pair of scoring functions - where one often corresponds to the <strong>magnitude</strong> (strength) of interaction and the other reflects the <strong>specificity</strong> of a given interaction to a pair of cell identities.</p>
<section id="visual-exploration">
<h4><span class="section-number">21.4.1.1. </span>Visual exploration<a class="headerlink" href="#visual-exploration" title="Link to this heading">#</a></h4>
<p>We can now visualize the results that we just obtained as a dotplot, in which rows represent the prioritized interactions between source/sender (top) and target/receiver (bottom) cell types.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">li</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">dotplot</span><span class="p">(</span>
    <span class="n">adata</span><span class="o">=</span><span class="n">adata_stim</span><span class="p">,</span>
    <span class="n">colour</span><span class="o">=</span><span class="s2">&quot;lr_means&quot;</span><span class="p">,</span>
    <span class="n">size</span><span class="o">=</span><span class="s2">&quot;cellphone_pvals&quot;</span><span class="p">,</span>
    <span class="n">inverse_size</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>  <span class="c1"># we inverse sign since we want small p-values to have large sizes</span>
    <span class="c1"># We choose only the cell types which we wish to plot</span>
    <span class="n">source_labels</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;CD4 T cells&quot;</span><span class="p">,</span> <span class="s2">&quot;B cells&quot;</span><span class="p">,</span> <span class="s2">&quot;FCGR3A+ Monocytes&quot;</span><span class="p">],</span>
    <span class="n">target_labels</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;CD8 T cells&quot;</span><span class="p">,</span> <span class="s2">&quot;CD14+ Monocytes&quot;</span><span class="p">,</span> <span class="s2">&quot;NK cells&quot;</span><span class="p">],</span>
    <span class="c1"># since cpdbv2 suggests using a filter to FPs</span>
    <span class="c1"># we can filter the interactions according to p-values &lt;= 0.01</span>
    <span class="n">filterby</span><span class="o">=</span><span class="s2">&quot;cellphone_pvals&quot;</span><span class="p">,</span>
    <span class="n">filter_lambda</span><span class="o">=</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">x</span> <span class="o">&lt;=</span> <span class="mf">0.01</span><span class="p">,</span>
    <span class="c1"># as this type of methods tends to result in large numbers</span>
    <span class="c1"># of predictions, we can also further order according to</span>
    <span class="c1"># expression magnitude</span>
    <span class="n">orderby</span><span class="o">=</span><span class="s2">&quot;lr_means&quot;</span><span class="p">,</span>
    <span class="n">orderby_ascending</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>  <span class="c1"># we want to prioritize those with highest expression</span>
    <span class="n">top_n</span><span class="o">=</span><span class="mi">20</span><span class="p">,</span>  <span class="c1"># and we want to keep only the top 20 interactions</span>
    <span class="n">figure_size</span><span class="o">=</span><span class="p">(</span><span class="mi">9</span><span class="p">,</span> <span class="mi">5</span><span class="p">),</span>
    <span class="n">size_range</span><span class="o">=</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="mi">6</span><span class="p">),</span>
<span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="../_images/d4cc30eec339500ccbe3f9c1a6f755db78ed259eea95cfa8820c971ad9653089.png" src="../_images/d4cc30eec339500ccbe3f9c1a6f755db78ed259eea95cfa8820c971ad9653089.png" />
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>&lt;ggplot: (8783502753003)&gt;
</pre></div>
</div>
</div>
</div>
<p>Great, we get a number of interactions potentially linked to IFN-beta stimulation.</p>
<p>We can also see that both the magnitude (expression strength) and specificity of the interactions are cell-type dependent.
For example, the potential binding of HLA-B binding to CD8A/B logically occurs only when the receiver cells are CD8 T cells.</p>
</section>
<section id="generating-a-ligand-receptor-consensus-with-liana">
<h4><span class="section-number">21.4.1.2. </span>Generating a Ligand-Receptor consensus with LIANA<a class="headerlink" href="#generating-a-ligand-receptor-consensus-with-liana" title="Link to this heading">#</a></h4>
<p>Given the reported limited agreement between the interactions inferred by different ligand-receptor methods, as a way to further increase the confidence in a potential interaction of interest, one could check if this interaction is predicted as relevant by more than a single method. In the same manner, one could also use multiple methods and focus on their consensus, or in other words focus on interactions consistently predicted as relevant. To this end, we will run the <code class="docutils literal notranslate"><span class="pre">rank_aggregate</span></code> method of liana <span id="id16">[<a class="reference internal" href="#id423" title="Daniel Dimitrov, Dénes Türei, Martin Garrido-Rodriguez, Paul L Burmedi, James S Nagai, Charlotte Boys, Ricardo O Ramirez Flores, Hyojin Kim, Bence Szalai, Ivan G Costa, Alberto Valdeolivas, Aurélien Dugourd, and Julio Saez-Rodriguez. Comparison of methods and resources for cell-cell communication inference from single-cell RNA-seq data. Nature Communications, 13(1):3224, jun 2022. URL: https://www.nature.com/articles/s41467-022-30755-0 (visited on 2021-05-31), doi:10.1038/s41467-022-30755-0.">Dimitrov <em>et al.</em>, 2022</a>]</span>, which generates a probability distribution of highly ranked interactions across the methods.</p>
<p>Let’s first examine the ligand-receptor methods in LIANA:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">li</span><span class="o">.</span><span class="n">method</span><span class="o">.</span><span class="n">show_methods</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>Method Name</th>
      <th>Magnitude Score</th>
      <th>Specificity Score</th>
      <th>Reference</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>CellPhoneDB</td>
      <td>lr_means</td>
      <td>cellphone_pvals</td>
      <td>Efremova, M., Vento-Tormo, M., Teichmann, S.A....</td>
    </tr>
    <tr>
      <th>0</th>
      <td>Connectome</td>
      <td>expr_prod</td>
      <td>scaled_weight</td>
      <td>Raredon, M.S.B., Yang, J., Garritano, J., Wang...</td>
    </tr>
    <tr>
      <th>0</th>
      <td>log2FC</td>
      <td>None</td>
      <td>lr_logfc</td>
      <td>Dimitrov, D., Türei, D., Garrido-Rodriguez, M....</td>
    </tr>
    <tr>
      <th>0</th>
      <td>NATMI</td>
      <td>expr_prod</td>
      <td>spec_weight</td>
      <td>Hou, R., Denisenko, E., Ong, H.T., Ramilowski,...</td>
    </tr>
    <tr>
      <th>0</th>
      <td>SingleCellSignalR</td>
      <td>lrscore</td>
      <td>None</td>
      <td>Cabello-Aguilar, S., Alame, M., Kon-Sun-Tack, ...</td>
    </tr>
    <tr>
      <th>0</th>
      <td>CellChat</td>
      <td>lr_probs</td>
      <td>cellchat_pvals</td>
      <td>Jin, S., Guerrero-Juarez, C.F., Zhang, L., Cha...</td>
    </tr>
    <tr>
      <th>0</th>
      <td>Rank_Aggregate</td>
      <td>magnitude_rank</td>
      <td>specificity_rank</td>
      <td>Dimitrov, D., Türei, D., Garrido-Rodriguez, M....</td>
    </tr>
    <tr>
      <th>0</th>
      <td>Geometric Mean</td>
      <td>lr_gmeans</td>
      <td>gmean_pvals</td>
      <td>CellPhoneDBv2's permutation approach applied t...</td>
    </tr>
  </tbody>
</table>
</div></div></div>
</div>
<p>Let’s now run the Rank_Aggregate method, which will essentially run the other methods in the background and then generate a consensus.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span><span class="w"> </span><span class="nn">liana.method</span><span class="w"> </span><span class="kn">import</span> <span class="n">rank_aggregate</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">rank_aggregate</span><span class="p">(</span>
    <span class="n">adata_stim</span><span class="p">,</span> <span class="n">groupby</span><span class="o">=</span><span class="s2">&quot;cell_type&quot;</span><span class="p">,</span> <span class="n">return_all_lrs</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">use_raw</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">verbose</span><span class="o">=</span><span class="kc">True</span>
<span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Using `.X`!
Converting mat to CSR format
227 features of mat are empty, they will be removed.
</pre></div>
</div>
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>/home/dbdimitrov/anaconda3/envs/cellcell/lib/python3.10/site-packages/pandas/core/indexing.py:1728: ImplicitModificationWarning: Trying to modify attribute `.obs` of view, initializing view as actual.
</pre></div>
</div>
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>0.46 of entities in the resource are missing from the data.
Generating ligand-receptor stats for 12301 samples and 15474 features
Assuming that counts were `natural` log-normalized!
Running CellPhoneDB
</pre></div>
</div>
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>100%|██████████| 1000/1000 [00:05&lt;00:00, 171.04it/s]
</pre></div>
</div>
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Running Connectome
Running log2FC
Running NATMI
Running SingleCellSignalR
Running CellChat
</pre></div>
</div>
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>100%|██████████| 1000/1000 [01:59&lt;00:00,  8.35it/s]
</pre></div>
</div>
</div>
</div>
<p>Let’s now check how the output of liana’s rank_aggregate:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">adata_stim</span><span class="o">.</span><span class="n">uns</span><span class="p">[</span><span class="s2">&quot;liana_res&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">drop_duplicates</span><span class="p">(</span>
    <span class="p">[</span><span class="s2">&quot;ligand_complex&quot;</span><span class="p">,</span> <span class="s2">&quot;receptor_complex&quot;</span><span class="p">]</span>
<span class="p">)</span><span class="o">.</span><span class="n">head</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>source</th>
      <th>target</th>
      <th>ligand_complex</th>
      <th>receptor_complex</th>
      <th>lr_means</th>
      <th>cellphone_pvals</th>
      <th>expr_prod</th>
      <th>scaled_weight</th>
      <th>lr_logfc</th>
      <th>spec_weight</th>
      <th>lrscore</th>
      <th>lr_probs</th>
      <th>cellchat_pvals</th>
      <th>steady_rank</th>
      <th>specificity_rank</th>
      <th>magnitude_rank</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>1129</th>
      <td>CD8 T cells</td>
      <td>CD8 T cells</td>
      <td>B2M</td>
      <td>CD3D</td>
      <td>2.562213</td>
      <td>0.0</td>
      <td>3.070147</td>
      <td>1.070524</td>
      <td>0.690730</td>
      <td>0.062383</td>
      <td>0.982125</td>
      <td>0.123504</td>
      <td>0.0</td>
      <td>2.092188e-11</td>
      <td>1.415831e-09</td>
      <td>9.840017e-13</td>
    </tr>
    <tr>
      <th>831</th>
      <td>CD8 T cells</td>
      <td>NK cells</td>
      <td>B2M</td>
      <td>KLRD1</td>
      <td>2.511733</td>
      <td>0.0</td>
      <td>2.622731</td>
      <td>1.541763</td>
      <td>0.858930</td>
      <td>0.080626</td>
      <td>0.980689</td>
      <td>0.110393</td>
      <td>0.0</td>
      <td>2.092188e-11</td>
      <td>1.415831e-09</td>
      <td>8.965786e-11</td>
    </tr>
    <tr>
      <th>468</th>
      <td>FCGR3A+ Monocytes</td>
      <td>CD14+ Monocytes</td>
      <td>TIMP1</td>
      <td>CD63</td>
      <td>1.877567</td>
      <td>0.0</td>
      <td>3.374182</td>
      <td>1.478816</td>
      <td>1.808744</td>
      <td>0.123245</td>
      <td>0.982935</td>
      <td>0.136754</td>
      <td>0.0</td>
      <td>2.092188e-11</td>
      <td>1.415831e-09</td>
      <td>9.348292e-09</td>
    </tr>
    <tr>
      <th>1513</th>
      <td>CD8 T cells</td>
      <td>FCGR3A+ Monocytes</td>
      <td>B2M</td>
      <td>LILRB2</td>
      <td>2.402974</td>
      <td>0.0</td>
      <td>1.658766</td>
      <td>1.466587</td>
      <td>0.604138</td>
      <td>0.078486</td>
      <td>0.975838</td>
      <td>0.063762</td>
      <td>0.0</td>
      <td>2.092188e-11</td>
      <td>1.415831e-09</td>
      <td>1.935320e-08</td>
    </tr>
    <tr>
      <th>836</th>
      <td>CD8 T cells</td>
      <td>NK cells</td>
      <td>B2M</td>
      <td>CD247</td>
      <td>2.414776</td>
      <td>0.0</td>
      <td>1.763377</td>
      <td>1.156763</td>
      <td>0.644404</td>
      <td>0.074642</td>
      <td>0.976548</td>
      <td>0.041130</td>
      <td>0.0</td>
      <td>2.092188e-11</td>
      <td>1.415831e-09</td>
      <td>1.228218e-07</td>
    </tr>
  </tbody>
</table>
</div></div></div>
</div>
<p>Here, we can see the output of the scoring functions of all methods (refer to the table above, if interested to map which score belong to which method). More importantly, we also get the <code class="docutils literal notranslate"><span class="pre">magnitude_rank</span></code> and <code class="docutils literal notranslate"><span class="pre">specificity_rank</span></code> for each interaction, which represent the consensus interaction <strong>magnitude</strong> (strength of expression) and <strong>specificity</strong> (across all cell type pairs), respectively. For example, going back to CellPhoneDB, <code class="docutils literal notranslate"><span class="pre">lr_mean</span></code> and <code class="docutils literal notranslate"><span class="pre">cellphone_pvals</span></code> will be correspondingly aggregated into those.</p>
<p>Let’s generate the same plot, but now using the aggregate of the methods:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">li</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">dotplot</span><span class="p">(</span>
    <span class="n">adata</span><span class="o">=</span><span class="n">adata_stim</span><span class="p">,</span>
    <span class="n">colour</span><span class="o">=</span><span class="s2">&quot;magnitude_rank&quot;</span><span class="p">,</span>
    <span class="n">size</span><span class="o">=</span><span class="s2">&quot;specificity_rank&quot;</span><span class="p">,</span>
    <span class="n">inverse_colour</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>  <span class="c1"># we inverse sign since we want small p-values to have large sizes</span>
    <span class="n">inverse_size</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
    <span class="c1"># We choose only the cell types which we wish to plot</span>
    <span class="n">source_labels</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;CD4 T cells&quot;</span><span class="p">,</span> <span class="s2">&quot;B cells&quot;</span><span class="p">,</span> <span class="s2">&quot;FCGR3A+ Monocytes&quot;</span><span class="p">],</span>
    <span class="n">target_labels</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;CD8 T cells&quot;</span><span class="p">,</span> <span class="s2">&quot;CD14+ Monocytes&quot;</span><span class="p">,</span> <span class="s2">&quot;NK cells&quot;</span><span class="p">],</span>
    <span class="c1"># since the rank_aggregate can also be interpreted as a probability distribution</span>
    <span class="c1"># we can again filter them according to their specificity significance</span>
    <span class="c1"># yet here the interactions are filtered according to</span>
    <span class="c1"># how consistently highly-ranked is their specificity across the methods</span>
    <span class="n">filterby</span><span class="o">=</span><span class="s2">&quot;specificity_rank&quot;</span><span class="p">,</span>
    <span class="n">filter_lambda</span><span class="o">=</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">x</span> <span class="o">&lt;=</span> <span class="mf">0.05</span><span class="p">,</span>
    <span class="c1"># again, we can also further order according to magnitude</span>
    <span class="n">orderby</span><span class="o">=</span><span class="s2">&quot;magnitude_rank&quot;</span><span class="p">,</span>
    <span class="n">orderby_ascending</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>  <span class="c1"># prioritize those with lowest values</span>
    <span class="n">top_n</span><span class="o">=</span><span class="mi">20</span><span class="p">,</span>  <span class="c1"># and we want to keep only the top 20 interactions</span>
    <span class="n">figure_size</span><span class="o">=</span><span class="p">(</span><span class="mi">9</span><span class="p">,</span> <span class="mi">5</span><span class="p">),</span>
    <span class="n">size_range</span><span class="o">=</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="mi">6</span><span class="p">),</span>
<span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="../_images/29c2f9cdf51dcd872e3d6f2c509aab86eeadbd618ed5cf592579f578c7ef07f1.png" src="../_images/29c2f9cdf51dcd872e3d6f2c509aab86eeadbd618ed5cf592579f578c7ef07f1.png" />
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>&lt;ggplot: (8783489520283)&gt;
</pre></div>
</div>
</div>
</div>
<p>Although, the prioritized interactions by both CellPhoneDB and LIANA seem biologically-plausible and potentially relevant to the treatment, it is challenging to ascertain their relevance. In particular, the advantage of these methods to generate systems-level insights in a hypothesis-free manner happens to also be one of their major disadvantages. Specifically because ligand-receptor tools return all plausible ligand-receptor interactions for every pair of cell types, thus we end up with huge lists of interactions, and choosing targets for subsequent experimental validation can be challenging.</p>
<p>Thus, prior to experimental validation, we suggest that any potential interaction hypotheses should be supported with additional prior knowledge. This may include domain knowledge for the specific condition (e.g. cell types or receptors of interest), as well as orthogonal modalities, such as protein abundance, spatial co-localization, or downstream signalling. To this end, we also further refer the reader to the <a class="reference internal" href="../conditions/gsea_pathway.html#enrichment-analysis"><span class="std std-ref">Gene set enrichment and pathway analysis</span></a> and <strong>spatial CCC</strong> (to be added) chapters.</p>
</section>
</section>
<section id="modelling-differential-intercellular-signalling-with-nichenet">
<h3><span class="section-number">21.4.2. </span>Modelling differential intercellular signalling with NicheNet<a class="headerlink" href="#modelling-differential-intercellular-signalling-with-nichenet" title="Link to this heading">#</a></h3>
<p>NicheNet is another type of CCC method that considers the <strong>intra</strong>cellular signaling effects triggered by <strong>inter</strong>cellular interactions <span id="id17">[<a class="reference internal" href="#id411" title="Robin Browaeys, Wouter Saelens, and Yvan Saeys. NicheNet: modeling intercellular communication by linking ligands to target genes. Nature Methods, 17(2):159-162, feb 2020. URL: http://www.nature.com/articles/s41592-019-0667-5 (visited on 2021-10-05), doi:10.1038/s41592-019-0667-5.">Browaeys <em>et al.</em>, 2020</a>]</span>.</p>
<p>In short, NicheNet infers associations between ligands and the downstream targets that they potentially modulate <span id="id18">[<a class="reference internal" href="#id411" title="Robin Browaeys, Wouter Saelens, and Yvan Saeys. NicheNet: modeling intercellular communication by linking ligands to target genes. Nature Methods, 17(2):159-162, feb 2020. URL: http://www.nature.com/articles/s41592-019-0667-5 (visited on 2021-10-05), doi:10.1038/s41592-019-0667-5.">Browaeys <em>et al.</em>, 2020</a>]</span>. Or in other words, NicheNet assumes that a certain sender/source cell type produces a ligand, and the binding of that ligand to a specific receiver/target cell type(s) leads to a signal propagation that affects master gene regulators, or transcription factors, and subsequently their targets. Therefore, predictions on which target genes are modulated by which ligand-receptor pairs may provide interesting hypotheses concerning the ongoing CCC events. Furthermore, enrichment of target genes of a specific ligand-receptor pair in a receiver cell type can also indicate that this ligand-receptor pair is functionally active. In summary, a NicheNet analysis can thus be performed to 1) infer potential target genes of expressed ligand-receptor interactions and 2) prioritize ligand-receptor interactions based on their target gene enrichment in the receiver. This enrichment is called “ligand activity”, in analogy to transcription factor activity <a class="reference internal" href="../conditions/gsea_pathway.html#enrichment-analysis"><span class="std std-ref">Gene set enrichment and pathway analysis</span></a>.</p>
<p>To perform these tasks, NicheNet makes use of prior knowledge about ligand-target associations. However, contrary to ligand-receptor databases, comprehensive ligand-target databases do not exist. Therefore, NicheNet predicts ligand-target associations by integrating  three layers of prior knowledge covering ligand-receptor, intracellular signaling, and gene regulatory interactions. Using these three layers of information, NicheNet calculates a regulatory potential score for each ligand-target link. This regulatory potential denotes how well prior knowledge supports that a ligand may regulate the expression of a target gene. To calculate regulatory potential, NicheNet first employs a network diffusion algorithm, known as Personalized PageRank (PPR), on the integrated signaling network to estimate the probability that a given ligand may signal to a particular regulator.  Specifically, the PPR implementation in NicheNet considers a given ligand as the seed node of interest. Hereby, nodes that are closer to the ligand in the signaling network get higher scores than more distant nodes under the assumption that regulators in the network vicinity of the ligand are more likely modulated by the ligand than more distant nodes. The application of PPR for all ligands in the database thus results in a ligand-regulator matrix of signalling probabilities that are then multiplied with a weight matrix of regulators to target genes in order to obtain ligand-target regulatory potential scores {cite}browaeys_2020. Conceptually, this implies that a ligand–target pair receives a high regulatory potential score if the ligand can signal to regulators of that target gene. These prior knowledge-derived scores are then used in conjunction with the expression data of interacting cells to prioritize ligand-receptor interactions and predict their target genes.</p>
<p>Because NicheNet focuses on how ligands affect gene expression in potentially interacting cells, one needs to be able to define which gene expression changes may (partially) be caused by CCC processes.  To rank the ligands according to their potential ligand activity in a receiver cell type, it requires a set of genes assumed to be affected by CCC events in the receiver cell types. The authors of NicheNet recommend ideally defining those sets of genes when working with conditions, e.g. treatment  vs control. To this end, we will apply NicheNet in exactly that manner and compare the expression of the patients in our data before and after stimulation with INF-beta.</p>
<p>For a more detailed description of the NicheNet method, we specifically refer the reader to the footprint part of the <a class="reference internal" href="../conditions/gsea_pathway.html#enrichment-analysis"><span class="std std-ref">Gene set enrichment and pathway analysis</span></a> chapter as well as NicheNet’s <a class="reference external" href="https://github.com/saeyslab/nichenetr">tutorials</a> and manuscript <span id="id19">[<a class="reference internal" href="#id411" title="Robin Browaeys, Wouter Saelens, and Yvan Saeys. NicheNet: modeling intercellular communication by linking ligands to target genes. Nature Methods, 17(2):159-162, feb 2020. URL: http://www.nature.com/articles/s41592-019-0667-5 (visited on 2021-10-05), doi:10.1038/s41592-019-0667-5.">Browaeys <em>et al.</em>, 2020</a>]</span>.</p>
<section id="load-nichenet-prior-knowledge">
<h4><span class="section-number">21.4.2.1. </span>Load NicheNet Prior-Knowledge<a class="headerlink" href="#load-nichenet-prior-knowledge" title="Link to this heading">#</a></h4>
<p>As mentioned above, NicheNet requires  prior knowledge about ligand-receptor interactions and ligand-target links:</p>
<ul class="simple">
<li><p>a ligand_target_matrix - denotes the potential that a ligand might regulate the expression of a target genes. The weights in this matrix are based on prior knowledge and are required to prioritize possible ligand-receptor interactions and affected target genes.</p></li>
<li><p>lr_network - the database of ligand-receptor interactions needed to define expressed ligands, receptors and their interactions.</p></li>
</ul>
<p>We will load each of those from Zenodo <a class="reference external" href="https://doi.org/10.5281/zenodo.7074291"><img alt="DOI" src="https://zenodo.org/badge/DOI/10.5281/zenodo.7074291.svg" /></a> (this might take a couple of minutes).</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="o">%%</span><span class="k">R</span>
# load NicheNet (NicheNet is only available on GitHub)
suppressPackageStartupMessages({
    if(!require(nichenetr)) remotes::install_github(&quot;saeyslab/nichenetr&quot;, upgrade = &quot;never&quot;)
})
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="o">%%</span><span class="k">R</span>
# Increase timeout threshold
options(timeout=600)

# Load PK
ligand_target_matrix &lt;- readRDS(url(&quot;https://zenodo.org/record/7074291/files/ligand_target_matrix_nsga2r_final.rds&quot;))
lr_network &lt;- readRDS(url(&quot;https://zenodo.org/record/7074291/files/lr_network_human_21122021.rds&quot;))
</pre></div>
</div>
</div>
</div>
<p>Moreover, one may also tailor the NicheNet prior knowledge and contextualize the networks according to the data at hand using the <a class="reference external" href="https://github.com/saezlab/OmnipathR">OmnipathR</a> package.</p>
<section id="step-1-define-cell-types-of-interest-to-be-considered-as-senders-sources-and-receiver-targets-of-ccc-interactions">
<h5><span class="section-number">21.4.2.1.1. </span>Step 1. Define cell types of interest to be considered as senders/sources and receiver/targets of CCC interactions<a class="headerlink" href="#step-1-define-cell-types-of-interest-to-be-considered-as-senders-sources-and-receiver-targets-of-ccc-interactions" title="Link to this heading">#</a></h5>
<p>Here, we will assume that multiple cell types are able to affect the receiver cell type</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">sender_celltypes</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;CD4 T cells&quot;</span><span class="p">,</span> <span class="s2">&quot;B cells&quot;</span><span class="p">,</span> <span class="s2">&quot;FCGR3A+ Monocytes&quot;</span><span class="p">]</span>
<span class="n">receiver_celltypes</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;CD8 T cells&quot;</span><span class="p">]</span>
</pre></div>
</div>
</div>
</div>
</section>
<section id="step-2-define-a-set-of-ligands-that-can-potentially-affect-receiver-cell-types">
<h5><span class="section-number">21.4.2.1.2. </span>Step 2. Define a set of ligands that can <strong>potentially</strong> affect receiver cell types<a class="headerlink" href="#step-2-define-a-set-of-ligands-that-can-potentially-affect-receiver-cell-types" title="Link to this heading">#</a></h5>
<p>Similarly to the ligand-receptor methods above, here we are only interested in the potential interactions that involve sufficiently expressed genes in each cell type. So, we will assume that e.g. 10% of the cells is a good threshold to reflect genes as expressed within a cell type.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Helper function to obtain sufficiently expressed genes</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">functools</span><span class="w"> </span><span class="kn">import</span> <span class="n">reduce</span>


<span class="k">def</span><span class="w"> </span><span class="nf">get_expressed_genes</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">cell_type</span><span class="p">,</span> <span class="n">expr_prop</span><span class="p">):</span>
    <span class="c1"># calculate proportions</span>
    <span class="n">temp</span> <span class="o">=</span> <span class="n">adata</span><span class="p">[</span><span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s2">&quot;cell_type&quot;</span><span class="p">]</span> <span class="o">==</span> <span class="n">cell_type</span><span class="p">,</span> <span class="p">:]</span>
    <span class="n">a</span> <span class="o">=</span> <span class="n">temp</span><span class="o">.</span><span class="n">X</span><span class="o">.</span><span class="n">getnnz</span><span class="p">(</span><span class="n">axis</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span> <span class="o">/</span> <span class="n">temp</span><span class="o">.</span><span class="n">X</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
    <span class="n">stats</span> <span class="o">=</span> <span class="p">(</span>
        <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">({</span><span class="s2">&quot;genes&quot;</span><span class="p">:</span> <span class="n">temp</span><span class="o">.</span><span class="n">var_names</span><span class="p">,</span> <span class="s2">&quot;props&quot;</span><span class="p">:</span> <span class="n">a</span><span class="p">})</span>
        <span class="o">.</span><span class="n">assign</span><span class="p">(</span><span class="n">cell_type</span><span class="o">=</span><span class="n">cell_type</span><span class="p">)</span>
        <span class="o">.</span><span class="n">sort_values</span><span class="p">(</span><span class="s2">&quot;genes&quot;</span><span class="p">)</span>
    <span class="p">)</span>

    <span class="c1"># obtain expressed genes</span>
    <span class="n">stats</span> <span class="o">=</span> <span class="n">stats</span><span class="p">[</span><span class="n">stats</span><span class="p">[</span><span class="s2">&quot;props&quot;</span><span class="p">]</span> <span class="o">&gt;=</span> <span class="n">expr_prop</span><span class="p">]</span>
    <span class="n">expressed_genes</span> <span class="o">=</span> <span class="n">stats</span><span class="p">[</span><span class="s2">&quot;genes&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">values</span>

    <span class="k">return</span> <span class="n">expressed_genes</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">sender_expressed</span> <span class="o">=</span> <span class="n">reduce</span><span class="p">(</span>
    <span class="n">np</span><span class="o">.</span><span class="n">union1d</span><span class="p">,</span>
    <span class="p">[</span>
        <span class="n">get_expressed_genes</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">cell_type</span><span class="o">=</span><span class="n">cell_type</span><span class="p">,</span> <span class="n">expr_prop</span><span class="o">=</span><span class="mf">0.1</span><span class="p">)</span>
        <span class="k">for</span> <span class="n">cell_type</span> <span class="ow">in</span> <span class="n">sender_celltypes</span>
    <span class="p">],</span>
<span class="p">)</span>
<span class="n">receiver_expressed</span> <span class="o">=</span> <span class="n">reduce</span><span class="p">(</span>
    <span class="n">np</span><span class="o">.</span><span class="n">union1d</span><span class="p">,</span>
    <span class="p">[</span>
        <span class="n">get_expressed_genes</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">cell_type</span><span class="o">=</span><span class="n">cell_type</span><span class="p">,</span> <span class="n">expr_prop</span><span class="o">=</span><span class="mf">0.1</span><span class="p">)</span>
        <span class="k">for</span> <span class="n">cell_type</span> <span class="ow">in</span> <span class="n">receiver_celltypes</span>
    <span class="p">],</span>
<span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<p>Then use this information to keep only ligand-receptor pairs in the NicheNet network that are expressed.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="o">%%</span><span class="k">R</span> -i sender_expressed -i receiver_expressed
# get ligands and receptors in the resource
ligands &lt;- lr_network %&gt;% pull(from) %&gt;% unique()
receptors &lt;- lr_network %&gt;% pull(to) %&gt;% unique()

# only keep the intersect between the resource and the data
expressed_ligands &lt;- intersect(ligands, sender_expressed)
expressed_receptors &lt;- intersect(receptors, receiver_expressed)

# filter the network to only include ligands for which both the ligand and receptor are expressed
potential_ligands &lt;- lr_network %&gt;% 
  filter(from %in% expressed_ligands &amp; to %in% expressed_receptors) %&gt;%
  pull(from) %&gt;% unique()
</pre></div>
</div>
</div>
</div>
</section>
<section id="step-3-define-a-gene-set-of-interest-in-receiver-cell-type-s">
<h5><span class="section-number">21.4.2.1.3. </span>Step 3. Define a gene set of interest in receiver cell type(s)<a class="headerlink" href="#step-3-define-a-gene-set-of-interest-in-receiver-cell-type-s" title="Link to this heading">#</a></h5>
<p>This step is the most critical one in a NicheNet analysis. Here, one defines which genes are potentially modulated by cell-cell communication, i.e. the genes thought to be affected by ligand signalling. For example, one can assume that differentially expressed genes between conditions in receiver cell type(s) are driven by ligands from one of more interacting sender cell populations. Another example would be differentially expressed genes between a differentiated cell population and a progenitor population in case the differentiation is likely induced through interactions with other cell types.</p>
<p>So, we will now use <code class="docutils literal notranslate"><span class="pre">decoupler</span></code> to generate pseudobulk profiles per cell type and sample, and then perform a differential expression analysis on those. A crucial step of pseudo-bulking is filtering out genes that are not expressed across most cells and samples, since they are very noisy and can result in unstable log-fold changes. To get robust profiles, genes are thus filtered out if again they are not expressed sufficiently per sample (min_prop) and and in not enough samples. We refer the reader to the {ref}normalization and {ref}differential-analysis chapters for more info on differential expression analysis and pseudobulk profiles.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Get pseudo-bulk profile</span>
<span class="n">pdata</span> <span class="o">=</span> <span class="n">dc</span><span class="o">.</span><span class="n">get_pseudobulk</span><span class="p">(</span>
    <span class="n">adata</span><span class="p">,</span>
    <span class="n">sample_col</span><span class="o">=</span><span class="s2">&quot;sample&quot;</span><span class="p">,</span>
    <span class="n">groups_col</span><span class="o">=</span><span class="s2">&quot;cell_type&quot;</span><span class="p">,</span>
    <span class="n">min_prop</span><span class="o">=</span><span class="mf">0.1</span><span class="p">,</span>
    <span class="n">min_smpls</span><span class="o">=</span><span class="mi">3</span><span class="p">,</span>
    <span class="n">layer</span><span class="o">=</span><span class="s2">&quot;counts&quot;</span><span class="p">,</span>
<span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<p>Normalize the pseudobulk counts</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Storing the raw counts</span>
<span class="n">pdata</span><span class="o">.</span><span class="n">layers</span><span class="p">[</span><span class="s2">&quot;counts&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">pdata</span><span class="o">.</span><span class="n">X</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>

<span class="c1"># Does PC1 captures a meaningful biological or technical fact?</span>
<span class="n">pdata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s2">&quot;lib_size&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">pdata</span><span class="o">.</span><span class="n">X</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="mi">1</span><span class="p">)</span>

<span class="c1"># Normalize</span>
<span class="n">sc</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">normalize_total</span><span class="p">(</span><span class="n">pdata</span><span class="p">,</span> <span class="n">target_sum</span><span class="o">=</span><span class="mf">1e4</span><span class="p">)</span>
<span class="n">sc</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">log1p</span><span class="p">(</span><span class="n">pdata</span><span class="p">)</span>
<span class="c1"># check how this looks like</span>
<span class="n">pdata</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>AnnData object with n_obs × n_vars = 108 × 4412
    obs: &#39;condition&#39;, &#39;cell_type&#39;, &#39;patient&#39;, &#39;sample&#39;, &#39;lib_size&#39;
    uns: &#39;log1p&#39;
    layers: &#39;counts&#39;
</pre></div>
</div>
</div>
</div>
<p>Then we perform a very simple differential analysis contrast. For this example we will use t-test as is implemented in <code class="docutils literal notranslate"><span class="pre">scanpy</span></code> but we could use any other.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">logFCs</span><span class="p">,</span> <span class="n">pvals</span> <span class="o">=</span> <span class="n">dc</span><span class="o">.</span><span class="n">get_contrast</span><span class="p">(</span>
    <span class="n">pdata</span><span class="p">,</span>
    <span class="n">group_col</span><span class="o">=</span><span class="s2">&quot;cell_type&quot;</span><span class="p">,</span>
    <span class="n">condition_col</span><span class="o">=</span><span class="s2">&quot;condition&quot;</span><span class="p">,</span>
    <span class="n">condition</span><span class="o">=</span><span class="s2">&quot;stim&quot;</span><span class="p">,</span>
    <span class="n">reference</span><span class="o">=</span><span class="s2">&quot;ctrl&quot;</span><span class="p">,</span>
    <span class="n">method</span><span class="o">=</span><span class="s2">&quot;t-test&quot;</span><span class="p">,</span>
<span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<p>Then keep only the positive significant differentially expressed genes in the receiver cell type(s)</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Visualize those for e.g. CD14+ Monocytes</span>
<span class="n">dc</span><span class="o">.</span><span class="n">plot_volcano</span><span class="p">(</span><span class="n">logFCs</span><span class="p">,</span> <span class="n">pvals</span><span class="p">,</span> <span class="s2">&quot;CD14+ Monocytes&quot;</span><span class="p">,</span> <span class="n">top</span><span class="o">=</span><span class="mi">15</span><span class="p">,</span> <span class="n">sign_thr</span><span class="o">=</span><span class="mf">0.05</span><span class="p">,</span> <span class="n">lFCs_thr</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="../_images/7322920c98729b3097ad0152c4d53c6f31d88fe980bd5ee7eb52f8eb50e2836d.png" src="../_images/7322920c98729b3097ad0152c4d53c6f31d88fe980bd5ee7eb52f8eb50e2836d.png" />
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># format results</span>
<span class="n">deg</span> <span class="o">=</span> <span class="n">dc</span><span class="o">.</span><span class="n">format_contrast_results</span><span class="p">(</span><span class="n">logFCs</span><span class="p">,</span> <span class="n">pvals</span><span class="p">)</span>
<span class="c1"># only keep the receiver cell type(s)</span>
<span class="n">deg</span> <span class="o">=</span> <span class="n">deg</span><span class="p">[</span><span class="n">np</span><span class="o">.</span><span class="n">isin</span><span class="p">(</span><span class="n">deg</span><span class="p">[</span><span class="s2">&quot;contrast&quot;</span><span class="p">],</span> <span class="n">receiver_celltypes</span><span class="p">)]</span>
<span class="n">deg</span><span class="o">.</span><span class="n">head</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>contrast</th>
      <th>name</th>
      <th>logFCs</th>
      <th>pvals</th>
      <th>adj_pvals</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>13236</th>
      <td>CD8 T cells</td>
      <td>SAMD9L</td>
      <td>3.592979</td>
      <td>1.818029e-10</td>
      <td>0.000001</td>
    </tr>
    <tr>
      <th>13237</th>
      <td>CD8 T cells</td>
      <td>IFIT3</td>
      <td>6.486142</td>
      <td>6.843872e-10</td>
      <td>0.000002</td>
    </tr>
    <tr>
      <th>13238</th>
      <td>CD8 T cells</td>
      <td>IFI6</td>
      <td>5.555549</td>
      <td>1.835813e-09</td>
      <td>0.000003</td>
    </tr>
    <tr>
      <th>13239</th>
      <td>CD8 T cells</td>
      <td>OAS1</td>
      <td>5.479982</td>
      <td>7.165866e-09</td>
      <td>0.000008</td>
    </tr>
    <tr>
      <th>13240</th>
      <td>CD8 T cells</td>
      <td>IFIT1</td>
      <td>6.617856</td>
      <td>1.149598e-08</td>
      <td>0.00001</td>
    </tr>
  </tbody>
</table>
</div></div></div>
</div>
<p>Now that we have the DE stats for the receiver cell type, we can use those to define the background and geneset of interest for ligand activity analysis with NicheNet.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># define background of sufficiently expressed genes</span>
<span class="n">background_genes</span> <span class="o">=</span> <span class="n">deg</span><span class="p">[</span><span class="s2">&quot;name&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">values</span>

<span class="c1"># only keep significant and positive DE genes</span>
<span class="n">deg</span> <span class="o">=</span> <span class="n">deg</span><span class="p">[(</span><span class="n">deg</span><span class="p">[</span><span class="s2">&quot;pvals&quot;</span><span class="p">]</span> <span class="o">&lt;=</span> <span class="mf">0.05</span><span class="p">)</span> <span class="o">&amp;</span> <span class="p">(</span><span class="n">deg</span><span class="p">[</span><span class="s2">&quot;logFCs&quot;</span><span class="p">]</span> <span class="o">&gt;</span> <span class="mi">1</span><span class="p">)]</span>
<span class="c1"># get geneset of interest</span>
<span class="n">geneset_oi</span> <span class="o">=</span> <span class="n">deg</span><span class="p">[</span><span class="s2">&quot;name&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">values</span>
</pre></div>
</div>
</div>
</div>
</section>
<section id="step-4-nichenet-ligand-activity-estimation">
<h5><span class="section-number">21.4.2.1.4. </span>Step 4. NicheNet ligand activity estimation<a class="headerlink" href="#step-4-nichenet-ligand-activity-estimation" title="Link to this heading">#</a></h5>
<p>To estimate ligand activity, NicheNet uses the regulatory potential of genes (based on prior knowledge) to predict which ligand best predicts the gene set of interest.
Or in other words, it assesses whether the genes with high regulatory potential in regards to a specific ligand, are more likely to belong to the geneset of interest that we derive for the receiver cell type(s).
Conceptually, this is not too dissimilar from standard <a class="reference internal" href="../conditions/gsea_pathway.html#enrichment-analysis"><span class="std std-ref">Gene set enrichment and pathway analysis</span></a> and NicheNet proposes different ways to estimate ligand activity, such as the area under the receiver operating characteristic curve, or Pearson correlation <span id="id20">[<a class="reference internal" href="#id411" title="Robin Browaeys, Wouter Saelens, and Yvan Saeys. NicheNet: modeling intercellular communication by linking ligands to target genes. Nature Methods, 17(2):159-162, feb 2020. URL: http://www.nature.com/articles/s41592-019-0667-5 (visited on 2021-10-05), doi:10.1038/s41592-019-0667-5.">Browaeys <em>et al.</em>, 2020</a>]</span>.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="o">%%</span><span class="k">R</span> -i geneset_oi -i background_genes -o ligand_activities

ligand_activities &lt;- predict_ligand_activities(geneset = geneset_oi, 
                                               background_expressed_genes = background_genes,
                                               ligand_target_matrix = ligand_target_matrix,
                                               potential_ligands = potential_ligands)

ligand_activities &lt;- ligand_activities %&gt;% 
  arrange(-aupr) %&gt;% 
  mutate(rank = rank(desc(aupr)))

# show top10 ligand activities
head(ligand_activities, n=10)
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span># A tibble: 10 × 6
   test_ligand auroc   aupr aupr_corrected pearson  rank
   &lt;chr&gt;       &lt;dbl&gt;  &lt;dbl&gt;          &lt;dbl&gt;   &lt;dbl&gt; &lt;dbl&gt;
 1 PTPRC       0.801 0.116          0.0882  0.168      1
 2 HLA-F       0.781 0.104          0.0765  0.188      2
 3 HLA-A       0.776 0.0941         0.0662  0.180      3
 4 HLA-B       0.766 0.0864         0.0586  0.162      4
 5 HLA-E       0.767 0.0856         0.0578  0.129      5
 6 CCL5        0.762 0.0812         0.0534  0.0527     6
 7 CD48        0.762 0.0786         0.0508  0.106      7
 8 B2M         0.753 0.0755         0.0477  0.121      8
 9 HLA-DRA     0.730 0.0694         0.0416  0.111      9
10 CXCL10      0.736 0.0678         0.0400  0.0803    10
</pre></div>
</div>
</div>
</div>
</section>
<section id="step-5-infer-visualize-top-predicted-target-genes-for-top-ligands">
<h5><span class="section-number">21.4.2.1.5. </span>Step 5. Infer &amp; Visualize top-predicted target genes for top ligands<a class="headerlink" href="#step-5-infer-visualize-top-predicted-target-genes-for-top-ligands" title="Link to this heading">#</a></h5>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="o">%%</span><span class="k">R</span> -o vis_ligand_target
top_ligands &lt;- ligand_activities %&gt;%
  top_n(15, aupr) %&gt;% 
  arrange(-aupr) %&gt;%
  pull(test_ligand) %&gt;%
  unique()

# get regulatory potentials
ligand_target_potential &lt;- map(top_ligands,
                               ~get_weighted_ligand_target_links(.x,
                                                                 geneset = geneset_oi,
                                                                 ligand_target_matrix = ligand_target_matrix,
                                                                 n = 500)
                              ) %&gt;%
    bind_rows() %&gt;% 
    drop_na()
    
# prep for visualization
active_ligand_target_links &lt;- 
  prepare_ligand_target_visualization(ligand_target_df = ligand_target_potential, 
                                      ligand_target_matrix = ligand_target_matrix)

# order ligands &amp; targets
order_ligands &lt;- intersect(top_ligands,
                           colnames(active_ligand_target_links)) %&gt;% rev() %&gt;% make.names()
order_targets &lt;- ligand_target_potential$target %&gt;%
  unique() %&gt;% 
  intersect(rownames(active_ligand_target_links)) %&gt;%
  make.names()
rownames(active_ligand_target_links) &lt;- rownames(active_ligand_target_links) %&gt;%
  make.names() # make.names() for heatmap visualization of genes like H2-T23
colnames(active_ligand_target_links) &lt;- colnames(active_ligand_target_links) %&gt;%
  make.names() # make.names() for heatmap visualization of genes like H2-T23

vis_ligand_target &lt;- active_ligand_target_links[order_targets, order_ligands] %&gt;%
  t()
    
# convert to dataframe, and then it&#39;s returned to py
vis_ligand_target &lt;- vis_ligand_target %&gt;%
    as.data.frame() %&gt;%
    rownames_to_column(&quot;ligand&quot;) %&gt;%
    as_tibble()
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># convert dot to underscore and set ligand as index</span>
<span class="n">vis_ligand_target</span><span class="p">[</span><span class="s2">&quot;ligand&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">vis_ligand_target</span><span class="p">[</span><span class="s2">&quot;ligand&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span>
    <span class="sa">r</span><span class="s2">&quot;\.&quot;</span><span class="p">,</span> <span class="s2">&quot;_&quot;</span><span class="p">,</span> <span class="n">regex</span><span class="o">=</span><span class="kc">True</span>
<span class="p">)</span>
<span class="n">vis_ligand_target</span><span class="o">.</span><span class="n">set_index</span><span class="p">(</span><span class="s2">&quot;ligand&quot;</span><span class="p">,</span> <span class="n">inplace</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="c1"># keep only columns where at least one gene has a regulatory potential &gt;= 0.05</span>
<span class="n">vis_ligand_target</span> <span class="o">=</span> <span class="n">vis_ligand_target</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span>
    <span class="p">:,</span> <span class="n">vis_ligand_target</span><span class="p">[</span><span class="n">vis_ligand_target</span> <span class="o">&gt;=</span> <span class="mf">0.05</span><span class="p">]</span><span class="o">.</span><span class="n">any</span><span class="p">()</span>
<span class="p">]</span>
<span class="n">vis_ligand_target</span><span class="o">.</span><span class="n">head</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>IFI16</th>
      <th>IFIH1</th>
      <th>IFIT2</th>
      <th>IRF7</th>
      <th>MT2A</th>
      <th>MX1</th>
      <th>NMI</th>
      <th>PARP14</th>
      <th>PARP9</th>
      <th>SOCS1</th>
      <th>TNFSF10</th>
      <th>EPSTI1</th>
      <th>HERC6</th>
      <th>LAP3</th>
      <th>PARP12</th>
      <th>TMEM140</th>
    </tr>
    <tr>
      <th>ligand</th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>ICAM1</th>
      <td>0.010111</td>
      <td>0.008649</td>
      <td>0.010065</td>
      <td>0.013067</td>
      <td>0.007855</td>
      <td>0.007761</td>
      <td>0.000000</td>
      <td>0.007268</td>
      <td>0.0000</td>
      <td>0.008907</td>
      <td>0.012580</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.00000</td>
    </tr>
    <tr>
      <th>CXCL11</th>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.0000</td>
      <td>0.000000</td>
      <td>0.007285</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.00000</td>
    </tr>
    <tr>
      <th>HLA_DQB1</th>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.007094</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.0000</td>
      <td>0.000000</td>
      <td>0.007733</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.00000</td>
    </tr>
    <tr>
      <th>HMGB1</th>
      <td>0.158681</td>
      <td>0.077152</td>
      <td>0.151255</td>
      <td>0.091884</td>
      <td>0.075502</td>
      <td>0.074791</td>
      <td>0.147717</td>
      <td>0.075350</td>
      <td>0.0745</td>
      <td>0.078945</td>
      <td>0.082315</td>
      <td>0.074556</td>
      <td>0.073812</td>
      <td>0.145847</td>
      <td>0.147686</td>
      <td>0.07226</td>
    </tr>
    <tr>
      <th>ITGB2</th>
      <td>0.007564</td>
      <td>0.000000</td>
      <td>0.006784</td>
      <td>0.009015</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.0000</td>
      <td>0.000000</td>
      <td>0.008748</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.00000</td>
    </tr>
  </tbody>
</table>
</div></div></div>
</div>
</section>
<section id="visualize-top-ligands-regulatory-targets">
<h5><span class="section-number">21.4.2.1.6. </span>Visualize top ligands &amp; regulatory targets<a class="headerlink" href="#visualize-top-ligands-regulatory-targets" title="Link to this heading">#</a></h5>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">fig</span><span class="p">,</span> <span class="n">ax</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">subplots</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">15</span><span class="p">,</span> <span class="mi">5</span><span class="p">))</span>
<span class="n">sns</span><span class="o">.</span><span class="n">heatmap</span><span class="p">(</span><span class="n">vis_ligand_target</span><span class="p">,</span> <span class="n">xticklabels</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">ax</span><span class="o">=</span><span class="n">ax</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="../_images/a415fe05827a4cc0301a2d510334979c258cac6824054db8bf11d76fcb6e2003.png" src="../_images/a415fe05827a4cc0301a2d510334979c258cac6824054db8bf11d76fcb6e2003.png" />
</div>
</div>
<p>Perfect, we end up with ligands that are most probable to affect downstream signalling in the receiver cell type(s), as well as their most likely targets.</p>
</section>
</section>
</section>
<section id="combining-nichenet-output-with-ligand-receptor-inference">
<h3><span class="section-number">21.4.3. </span>Combining NicheNet output with ligand-receptor inference<a class="headerlink" href="#combining-nichenet-output-with-ligand-receptor-inference" title="Link to this heading">#</a></h3>
<p>NicheNet and ligand-receptor methods are not exclusive, but rather complimentary, as they address different questions and in different ways. Ligand-receptor methods, such as the ones that we show above, infer ligand-receptor interactions using the expression of interacting ligands and receptors, and typically work on “steady-state” data. On the contrary, NicheNet predicts which of these inferred ligand-receptor links are possibly the most functional based on gene expression changes that are induced in the target cell type(s), in a process largely dependent on prior knowledge alone. Thus, while the choice of tool depends on the research question, one could also see ligand-receptor (e.g. CellPhoneDB or LIANA) methods and intracellular signalling CCC methods (e.g. NicheNet) as complementary.</p>
<p>For example, given the top 3 prioritized ligands, we can focus on ligand-receptor pairs that include those and the receiver cell type that we used to infer the active ligands.</p>
<p>So, let’s see which ligand-receptor interactions that involve the top 5 potentially active ligands from NicheNet were prioritized as consistently relevant by the different methods in LIANA:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">ligand_oi</span> <span class="o">=</span> <span class="n">ligand_activities</span><span class="o">.</span><span class="n">head</span><span class="p">(</span><span class="mi">3</span><span class="p">)[</span><span class="s2">&quot;test_ligand&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">values</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">ligand_oi</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>array([&#39;PTPRC&#39;, &#39;HLA-F&#39;, &#39;HLA-A&#39;], dtype=object)
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">li</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">dotplot</span><span class="p">(</span>
    <span class="n">adata</span><span class="o">=</span><span class="n">adata_stim</span><span class="p">,</span>
    <span class="n">colour</span><span class="o">=</span><span class="s2">&quot;lr_means&quot;</span><span class="p">,</span>
    <span class="n">size</span><span class="o">=</span><span class="s2">&quot;cellphone_pvals&quot;</span><span class="p">,</span>
    <span class="n">inverse_size</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>  <span class="c1"># we inverse sign since we want small p-values to have large sizes</span>
    <span class="c1"># We choose only the cell types which we wish to plot</span>
    <span class="n">source_labels</span><span class="o">=</span><span class="n">sender_celltypes</span><span class="p">,</span>
    <span class="n">target_labels</span><span class="o">=</span><span class="n">receiver_celltypes</span><span class="p">,</span>
    <span class="c1"># keep only those ligands</span>
    <span class="n">filterby</span><span class="o">=</span><span class="s2">&quot;ligand_complex&quot;</span><span class="p">,</span>
    <span class="n">filter_lambda</span><span class="o">=</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">np</span><span class="o">.</span><span class="n">isin</span><span class="p">(</span><span class="n">x</span><span class="p">,</span> <span class="n">ligand_oi</span><span class="p">),</span>
    <span class="c1"># as this type of methods tends to result in large numbers</span>
    <span class="c1"># of predictions, we can also further order according to</span>
    <span class="c1"># expression magnitude</span>
    <span class="n">orderby</span><span class="o">=</span><span class="s2">&quot;magnitude_rank&quot;</span><span class="p">,</span>
    <span class="n">orderby_ascending</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>  <span class="c1"># we want to prioritize those with highest expression</span>
    <span class="n">top_n</span><span class="o">=</span><span class="mi">25</span><span class="p">,</span>  <span class="c1"># and we want to keep only the top 25 interactions</span>
    <span class="n">figure_size</span><span class="o">=</span><span class="p">(</span><span class="mi">9</span><span class="p">,</span> <span class="mi">9</span><span class="p">),</span>
    <span class="n">size_range</span><span class="o">=</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="mi">6</span><span class="p">),</span>
<span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="../_images/08d6728e2637196d24c27f3d192962b9a7dc94ac83dc94be09080a241194f02d.png" src="../_images/08d6728e2637196d24c27f3d192962b9a7dc94ac83dc94be09080a241194f02d.png" />
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>&lt;ggplot: (8783461218238)&gt;
</pre></div>
</div>
</div>
</div>
<p>From the results, above we can now see the specific ligand-receptor interactions potentially associated with the inferred ligand activities by NicheNet. Moreover, we can hypothesize that the ligand activities inferred by NicheNet, are perhaps affecting the cell types via ligand-receptor interactions specific to some cell type pairs (e.g. <em>HLA-A -&gt; CD3G</em>). Or in other words, while the same ligands may be active as a consequence of IFN-β stimulation, they potentially induce cell-type specific signal transduction changes.</p>
<p>While we use NicheNet and LIANA in this case, some recent CCC tool developments provide various combinations of the ideas and assumptions in these two categories of CCC inference methods (e.g. <span id="id21">[<a class="reference internal" href="#id434" title="Giacomo Baruzzo, Giulia Cesaro, and Barbara Di Camillo. Identify, quantify and characterize cellular communication from single-cell rna sequencing data with scseqcomm. Bioinformatics, 38(7):1920–1929, 2022.">Baruzzo <em>et al.</em>, 2022</a>, <a class="reference internal" href="#id433" title="Yang Zhang, Tianyuan Liu, Xuesong Hu, Mei Wang, Jing Wang, Bohao Zou, Puwen Tan, Tianyu Cui, Yiying Dou, Lin Ning, and others. Cellcall: integrating paired ligand–receptor and transcription factor activities for cell–cell communication. Nucleic Acids Research, 49(15):8520–8534, 2021.">Zhang <em>et al.</em>, 2021</a>, <a class="reference internal" href="#id435" title="Yang Zhang, Tianyuan Liu, Jing Wang, Bohao Zou, Le Li, Linhui Yao, Kechen Chen, Lin Ning, Bingyi Wu, Xiaoyang Zhao, and others. Cellinker: a platform of ligand–receptor interactions for intercellular communication analysis. Bioinformatics, 37(14):2025–2032, 2021.">Zhang <em>et al.</em>, 2021</a>]</span>).</p>
</section>
</section>
<section id="key-takeaways">
<h2><span class="section-number">21.5. </span>Key takeaways<a class="headerlink" href="#key-takeaways" title="Link to this heading">#</a></h2>
<section id="assumptions-limitations">
<h3><span class="section-number">21.5.1. </span>Assumptions &amp; Limitations<a class="headerlink" href="#assumptions-limitations" title="Link to this heading">#</a></h3>
<p>The shared purpose of the methods considered in this work is to predict the most relevant intercellular interactions between different cell types using single-cell transcriptomics data. Thus, all methods assume that the gene expression of different cell types is an informative proxy of the CCC events that occur within the sampled tissue. While single-cell transcriptomics enables the inference of CCC events at a previously unprecedented scale, some assumptions and limitations should be kept in mind.
Starting with the assumption that protein co-expressions reflect intercellular interactions, and consequently they are also reflective of any events preceding the interaction, including protein translation and processing, secretion, and diffusion <span id="id22">[<a class="reference internal" href="#id413" title="Erick Armingol, Adam Officer, Olivier Harismendy, and Nathan E Lewis. Deciphering cell-cell interactions and communication from gene expression. Nature Reviews. Genetics, 22(2):71-88, 2021. URL: http://www.nature.com/articles/s41576-020-00292-x (visited on 2021-10-05), doi:10.1038/s41576-020-00292-x.">Armingol <em>et al.</em>, 2021</a>, <a class="reference internal" href="#id423" title="Daniel Dimitrov, Dénes Türei, Martin Garrido-Rodriguez, Paul L Burmedi, James S Nagai, Charlotte Boys, Ricardo O Ramirez Flores, Hyojin Kim, Bence Szalai, Ivan G Costa, Alberto Valdeolivas, Aurélien Dugourd, and Julio Saez-Rodriguez. Comparison of methods and resources for cell-cell communication inference from single-cell RNA-seq data. Nature Communications, 13(1):3224, jun 2022. URL: https://www.nature.com/articles/s41467-022-30755-0 (visited on 2021-05-31), doi:10.1038/s41467-022-30755-0.">Dimitrov <em>et al.</em>, 2022</a>]</span> (<strong>Figure 22.2</strong>). Furthermore, if cell communication within an organism is conceptualized as the combination of CCC events that occur at different ‘length scales’ or ranges <span id="id23">[<a class="reference internal" href="#id431" title="Giovanni Palla, David S Fischer, Aviv Regev, and Fabian J Theis. Spatial components of molecular tissue biology. Nature Biotechnology, 40(3):308–318, 2022.">Palla <em>et al.</em>, 2022</a>]</span>, then the CCC events that can be inferred from single-cell transcriptomics are potentially limited to only the protein-mediated events that occur at ‘local’ ranges, i.e. between the cell types were sampled. As a result, long-range signalling and CCC driven by other molecules, such as endocrine signalling and system gradients like Calcium and Oxygen concentrations, are likely not captured <span id="id24">[<a class="reference internal" href="#id423" title="Daniel Dimitrov, Dénes Türei, Martin Garrido-Rodriguez, Paul L Burmedi, James S Nagai, Charlotte Boys, Ricardo O Ramirez Flores, Hyojin Kim, Bence Szalai, Ivan G Costa, Alberto Valdeolivas, Aurélien Dugourd, and Julio Saez-Rodriguez. Comparison of methods and resources for cell-cell communication inference from single-cell RNA-seq data. Nature Communications, 13(1):3224, jun 2022. URL: https://www.nature.com/articles/s41467-022-30755-0 (visited on 2021-05-31), doi:10.1038/s41467-022-30755-0.">Dimitrov <em>et al.</em>, 2022</a>]</span>.</p>
<p>In addition, while the grouping of cell types according to their lineage is common practice to structure our data, when considering the tissue as the place where communication takes place, interactions do not necessarily occur between cell types but rather individual cells <span id="id25">[<a class="reference internal" href="#id420" title="Aaron J Wilk, Alex K Shalek, Susan Holmes, and Catherine A Blish. Comparative analysis of cell-cell communication at single-cell resolution. BioRxiv, mar 2022. URL: http://biorxiv.org/lookup/doi/10.1101/2022.02.04.479209 (visited on 2022-02-22), doi:10.1101/2022.02.04.479209.">Wilk <em>et al.</em>, 2022</a>]</span>. Also, the intercellular interactions are typically presented as one-to-one events between cell types and/or proteins. Therefore, the assumption that the potential co-expression events across cell types might not necessarily reflect true signalling as in order to fundamentally capture the events within a local community, one must identify the magnitude, directionality and biological relevance of the messages passed between the cells within that community <span id="id26">[<a class="reference internal" href="#id413" title="Erick Armingol, Adam Officer, Olivier Harismendy, and Nathan E Lewis. Deciphering cell-cell interactions and communication from gene expression. Nature Reviews. Genetics, 22(2):71-88, 2021. URL: http://www.nature.com/articles/s41576-020-00292-x (visited on 2021-10-05), doi:10.1038/s41576-020-00292-x.">Armingol <em>et al.</em>, 2021</a>]</span>.</p>
</section>
<section id="summary-outlook">
<h3><span class="section-number">21.5.2. </span>Summary &amp; Outlook:<a class="headerlink" href="#summary-outlook" title="Link to this heading">#</a></h3>
<p>In this chapter we presented two applications of CCC inference methods, namely we used CellPhoneDB and LIANA to predict relevant ligand-receptor interactions from single-context (or steady-state) data and NicheNet to infer potentially active ligands and their targets in a differential-expression context.
As the focus of the single-cell field moves further away from the definition of linages, and into characterizing changes within cell types between conditions, approaches to disentangle CCC insights across contexts are becoming essential. Thus, in addition to NicheNet, we refer the user to other approaches that enable cross-condition comparisons, such as NATMI’s differential cell-connectivity analysis <span id="id27">[<a class="reference internal" href="#id429" title="Rui Hou, Elena Denisenko, Huan Ting Ong, Jordan A Ramilowski, and Alistair RR Forrest. Predicting cell-to-cell communication networks using natmi. Nature communications, 11(1):1–11, 2020.">Hou <em>et al.</em>, 2020</a>]</span>, Crosstalker’s network topological measures <span id="id28">[<a class="reference internal" href="#id430" title="James S Nagai, Nils B Leimkühler, Michael T Schaub, Rebekka K Schneider, and Ivan G Costa. Crosstalker: analysis and visualization of ligand–receptorne tworks. Bioinformatics, 37(22):4263–4265, 2021.">Nagai <em>et al.</em>, 2021</a>]</span>, CellChat’s pathway-focused manifold learning <span id="id29">[<a class="reference internal" href="#id415" title="Suoqin Jin, Christian F Guerrero-Juarez, Lihua Zhang, Ivan Chang, Raul Ramos, Chen-Hsiang Kuan, Peggy Myung, Maksim V Plikus, and Qing Nie. Inference and analysis of cell-cell communication using CellChat. Nature Communications, 12(1):1088, feb 2021. URL: http://dx.doi.org/10.1038/s41467-021-21246-9 (visited on 2021-10-05), doi:10.1038/s41467-021-21246-9.">Jin <em>et al.</em>, 2021</a>]</span>, as well as Tensor-cell2cell’s untargeted factorization approach to infer CCC patterns across contexts <span id="id30">[<a class="reference internal" href="#id425" title="Erick Armingol, Hratch M Baghdassarian, Cameron Martino, Araceli Perez-Lopez, Caitlin Aamodt, Rob Knight, and Nathan E Lewis. Context-aware deconvolution of cell-cell communication with tensor-cell2cell. Nature Communications, 13(1):3665, jun 2022. URL: https://www.nature.com/articles/s41467-022-31369-2 (visited on 2021-09-24), doi:10.1038/s41467-022-31369-2.">Armingol <em>et al.</em>, 2022</a>]</span>.</p>
<p>As a consequence of the ongoing developments within the single-cell and the cell-cell communication field specifically, there are is an ever-growing number of methods, some of which propose alternative ways to predict CCC events, such as those that work at the single-cell resolution <span id="id31">[<a class="reference internal" href="#id427" title="Micha Sam Brickman Raredon, Junchen Yang, Neeharika Kothapalli, Wesley Lewis, Naftali Kaminski, Laura E Niklason, and Yuval Kluger. Comprehensive visualization of cell-cell interactions in single-cell and spatial transcriptomics with NICHES. Bioinformatics, jan 2023. URL: http://dx.doi.org/10.1093/bioinformatics/btac775 (visited on 2022-01-27), doi:10.1093/bioinformatics/btac775.">Raredon <em>et al.</em>, 2023</a>, <a class="reference internal" href="#id410" title="Shuxiong Wang, Matthew Karikomi, Adam L MacLean, and Qing Nie. Cell lineage and communication network inference via optimization for single-cell transcriptomics. Nucleic Acids Research, 47(11):e66, jun 2019. URL: https://academic.oup.com/nar/article/47/11/e66/5421812 (visited on 2019-05-01), doi:10.1093/nar/gkz204.">Wang <em>et al.</em>, 2019</a>, <a class="reference internal" href="#id420" title="Aaron J Wilk, Alex K Shalek, Susan Holmes, and Catherine A Blish. Comparative analysis of cell-cell communication at single-cell resolution. BioRxiv, mar 2022. URL: http://biorxiv.org/lookup/doi/10.1101/2022.02.04.479209 (visited on 2022-02-22), doi:10.1101/2022.02.04.479209.">Wilk <em>et al.</em>, 2022</a>]</span>. While others attempt to address some of the limitations above, e.g. by including the inference of interactions mediated by metabolites or small molecules <span id="id32">[<a class="reference internal" href="#id424" title="Luz Garcia-Alonso, Valentina Lorenzi, Cecilia Icoresi Mazzeo, João Pedro Alves-Lopes, Kenny Roberts, Carmen Sancho-Serra, Justin Engelbert, Magda Marečková, Wolfram H Gruhn, Rachel A Botting, Tong Li, Berta Crespo, Stijn van Dongen, Vladimir Yu Kiselev, Elena Prigmore, Mary Herbert, Ashley Moffett, Alain Chédotal, Omer Ali Bayraktar, Azim Surani, Muzlifah Haniffa, and Roser Vento-Tormo. Single-cell roadmap of human gonadal development. Nature, 607(7919):540-547, jul 2022. URL: https://www.nature.com/articles/s41586-022-04918-4 (visited on 2022-07-07), doi:10.1038/s41586-022-04918-4.">Garcia-Alonso <em>et al.</em>, 2022</a>, <a class="reference internal" href="#id417" title="Yang Zhang, Tianyuan Liu, Jing Wang, Bohao Zou, Le Li, Linhui Yao, Kechen Chen, Lin Ning, Bingyi Wu, Xiaoyang Zhao, and Dong Wang. Cellinker: a platform of ligand-receptor interactions for intercellular communication analysis. Bioinformatics, jan 2021. URL: http://dx.doi.org/10.1093/bioinformatics/btab036 (visited on 2021-09-12), doi:10.1093/bioinformatics/btab036.">Zhang <em>et al.</em>, 2021</a>, <a class="reference internal" href="#id421" title="Rongbin Zheng, Yang Zhang, Tadataka Tsuji, Lili Zhang, Yu-Hua Tseng, and Kaifu Chen. MEBOCOST: metabolic cell-cell communication modeling by single cell transcriptome. BioRxiv, may 2022. URL: http://biorxiv.org/lookup/doi/10.1101/2022.05.30.494067 (visited on 2022-06-02), doi:10.1101/2022.05.30.494067.">Zheng <em>et al.</em>, 2022</a>]</span>.</p>
<p>Albeit, this chapter does not capture the diversity and all current developments of the CCC field in dissociated single-cell data, it serves to highlight some of the general ideas and limitations. Thus, we intend it as a starting point upon which the reader can expand on by combining their knowledge from other chapters, and further reading within the CCC field itself.</p>
<a class="bg-primary mb-1 reference internal image-reference" href="../_images/ccc_limitations.png"><img alt="CCC Overview" class="bg-primary mb-1" src="../_images/ccc_limitations.png" style="width: 800px;" /></a>
<p><strong>FIGURE 22.2</strong>. Assumptions and Limitations of Cell-Cell Communication from single-cell transcriptomics data</p>
</section>
</section>
<section id="quiz">
<h2><span class="section-number">21.6. </span>Quiz<a class="headerlink" href="#quiz" title="Link to this heading">#</a></h2>
<div class="cell tag_remove_input docutils container">
<div class="cell_output docutils container">
<div class="output text_html">
    <style>
        /* to remove the background color in dark mode */
        div#flip-card .output.text_html {
            background-color: transparent;
        }
        .flip-card-q1 {
            background-color: transparent;
            width: 350px;
            height: 200px;
            perspective: 1000px;
            display: inline-block;
            margin: 10px;
            text-align: center;
            padding: 10px;
        }
        .flip-card-inner-q1 {
            position: relative;
            width: 100%;
            height: 100%;
            text-align: center;
            transition: transform 0.6s;
            transform-style: preserve-3d;
            transform-origin: center;
        }
        .flip-card-q1:hover .flip-card-inner-q1 {
            transform: rotateY(180deg);
        }
        .flip-card-front-q1, .flip-card-back-q1 {
            position: absolute;
            width: 100%;
            height: 100%;
            min-height: 200px;
            backface-visibility: hidden;
            display: flex;
            align-items: center;
            justify-content: center;
            font-weight: bold;
            color: white;
            padding: 20px;
            border-radius: 15px;
            box-sizing: border-box;
            overflow: hidden;
            text-align: center;
        }
        .flip-card-front-q1 {
            background-color: #3965a3;
            font-size: 20px;
        }
        .flip-card-back-q1 {
            background-color: #a8d480;
            transform: rotateY(180deg);
            font-size: 20px;
        }
    </style>

    <div class="flip-card-q1">
        <div class="flip-card-inner-q1">
            <div class="flip-card-front-q1">
                What are the first three limitations that come to mind when inferring CCC from single-cell transcriptomics data?
            </div>
            <div class="flip-card-back-q1">
                Spatial context absence, post-transcriptional modifications, temporal dynamics
            </div>
        </div>
    </div>
    <noscript>
        <p><strong>Q:</strong> What are the first three limitations that come to mind when inferring CCC from single-cell transcriptomics data?</p>
        <p><strong>Answer:</strong> Spatial context absence, post-transcriptional modifications, temporal dynamics</p>
    </noscript>
    </div><div class="output text_html">
    <style>
        /* to remove the background color in dark mode */
        div#flip-card .output.text_html {
            background-color: transparent;
        }
        .flip-card-q2 {
            background-color: transparent;
            width: 350px;
            height: 200px;
            perspective: 1000px;
            display: inline-block;
            margin: 10px;
            text-align: center;
            padding: 10px;
        }
        .flip-card-inner-q2 {
            position: relative;
            width: 100%;
            height: 100%;
            text-align: center;
            transition: transform 0.6s;
            transform-style: preserve-3d;
            transform-origin: center;
        }
        .flip-card-q2:hover .flip-card-inner-q2 {
            transform: rotateY(180deg);
        }
        .flip-card-front-q2, .flip-card-back-q2 {
            position: absolute;
            width: 100%;
            height: 100%;
            min-height: 200px;
            backface-visibility: hidden;
            display: flex;
            align-items: center;
            justify-content: center;
            font-weight: bold;
            color: white;
            padding: 20px;
            border-radius: 15px;
            box-sizing: border-box;
            overflow: hidden;
            text-align: center;
        }
        .flip-card-front-q2 {
            background-color: #3965a3;
            font-size: 20px;
        }
        .flip-card-back-q2 {
            background-color: #a8d480;
            transform: rotateY(180deg);
            font-size: 20px;
        }
    </style>

    <div class="flip-card-q2">
        <div class="flip-card-inner-q2">
            <div class="flip-card-front-q2">
                What is the role of heteromeric complexes in ligand-receptor inference?
            </div>
            <div class="flip-card-back-q2">
                Heteromeric complexes, composed of multiple protein subunits, function collectively as a single receptor or ligand.
            </div>
        </div>
    </div>
    <noscript>
        <p><strong>Q:</strong> What is the role of heteromeric complexes in ligand-receptor inference?</p>
        <p><strong>Answer:</strong> Heteromeric complexes, composed of multiple protein subunits, function collectively as a single receptor or ligand.</p>
    </noscript>
    </div><div class="output text_html">
    <style>
        /* to remove the background color in dark mode */
        div#flip-card .output.text_html {
            background-color: transparent;
        }
        .flip-card-q3 {
            background-color: transparent;
            width: 350px;
            height: 200px;
            perspective: 1000px;
            display: inline-block;
            margin: 10px;
            text-align: center;
            padding: 10px;
        }
        .flip-card-inner-q3 {
            position: relative;
            width: 100%;
            height: 100%;
            text-align: center;
            transition: transform 0.6s;
            transform-style: preserve-3d;
            transform-origin: center;
        }
        .flip-card-q3:hover .flip-card-inner-q3 {
            transform: rotateY(180deg);
        }
        .flip-card-front-q3, .flip-card-back-q3 {
            position: absolute;
            width: 100%;
            height: 100%;
            min-height: 200px;
            backface-visibility: hidden;
            display: flex;
            align-items: center;
            justify-content: center;
            font-weight: bold;
            color: white;
            padding: 20px;
            border-radius: 15px;
            box-sizing: border-box;
            overflow: hidden;
            text-align: center;
        }
        .flip-card-front-q3 {
            background-color: #3965a3;
            font-size: 20px;
        }
        .flip-card-back-q3 {
            background-color: #a8d480;
            transform: rotateY(180deg);
            font-size: 13px;
        }
    </style>

    <div class="flip-card-q3">
        <div class="flip-card-inner-q3">
            <div class="flip-card-front-q3">
                Can you think of an advantage and a disadvantage of inference based on prior-knowledge?
            </div>
            <div class="flip-card-back-q3">
                Advantage: Utilizing established databases of ligand-receptor pairs can enhance the accuracy of inferred interactions by providing a validated framework. Disadvantage: Reliance on prior knowledge may bias analyses toward well-characterized interactions, potentially overlooking novel or context-specific communications not documented in existing databases.
            </div>
        </div>
    </div>
    <noscript>
        <p><strong>Q:</strong> Can you think of an advantage and a disadvantage of inference based on prior-knowledge?</p>
        <p><strong>Answer:</strong> Advantage: Utilizing established databases of ligand-receptor pairs can enhance the accuracy of inferred interactions by providing a validated framework. Disadvantage: Reliance on prior knowledge may bias analyses toward well-characterized interactions, potentially overlooking novel or context-specific communications not documented in existing databases.</p>
    </noscript>
    </div><div class="output text_html">
    <style>
        /* to remove the background color in dark mode */
        div#flip-card .output.text_html {
            background-color: transparent;
        }
        .flip-card-q4 {
            background-color: transparent;
            width: 350px;
            height: 200px;
            perspective: 1000px;
            display: inline-block;
            margin: 10px;
            text-align: center;
            padding: 10px;
        }
        .flip-card-inner-q4 {
            position: relative;
            width: 100%;
            height: 100%;
            text-align: center;
            transition: transform 0.6s;
            transform-style: preserve-3d;
            transform-origin: center;
        }
        .flip-card-q4:hover .flip-card-inner-q4 {
            transform: rotateY(180deg);
        }
        .flip-card-front-q4, .flip-card-back-q4 {
            position: absolute;
            width: 100%;
            height: 100%;
            min-height: 200px;
            backface-visibility: hidden;
            display: flex;
            align-items: center;
            justify-content: center;
            font-weight: bold;
            color: white;
            padding: 20px;
            border-radius: 15px;
            box-sizing: border-box;
            overflow: hidden;
            text-align: center;
        }
        .flip-card-front-q4 {
            background-color: #3965a3;
            font-size: 15px;
        }
        .flip-card-back-q4 {
            background-color: #a8d480;
            transform: rotateY(180deg);
            font-size: 13px;
        }
    </style>

    <div class="flip-card-q4">
        <div class="flip-card-inner-q4">
            <div class="flip-card-front-q4">
                Can you think of an important limitation we should consider here regarding the definition of the geneset of interest for the NicheNet analysis on the IFN-beta stimulated PBMC dataset?
            </div>
            <div class="flip-card-back-q4">
                An important limitation is ensuring that the gene set of interest accurately reflects the biological response to IFN-beta stimulation. If the selected genes are not truly responsive to IFN-beta, the analysis may yield misleading insights into the signaling pathways and interactions involved.
            </div>
        </div>
    </div>
    <noscript>
        <p><strong>Q:</strong> Can you think of an important limitation we should consider here regarding the definition of the geneset of interest for the NicheNet analysis on the IFN-beta stimulated PBMC dataset?</p>
        <p><strong>Answer:</strong> An important limitation is ensuring that the gene set of interest accurately reflects the biological response to IFN-beta stimulation. If the selected genes are not truly responsive to IFN-beta, the analysis may yield misleading insights into the signaling pathways and interactions involved.</p>
    </noscript>
    </div></div>
</div>
</section>
<section id="contributions">
<h2><span class="section-number">21.7. </span>Contributions<a class="headerlink" href="#contributions" title="Link to this heading">#</a></h2>
<section id="authors">
<h3><span class="section-number">21.7.1. </span>Authors:<a class="headerlink" href="#authors" title="Link to this heading">#</a></h3>
<ul class="simple">
<li><p>Daniel Dimitrov</p></li>
</ul>
</section>
<section id="reviewers">
<h3><span class="section-number">21.7.2. </span>Reviewers:<a class="headerlink" href="#reviewers" title="Link to this heading">#</a></h3>
<ul class="simple">
<li><p>Lukas Heumos</p></li>
<li><p>Gregor Sturm</p></li>
<li><p>Robin Browaeys</p></li>
</ul>
</section>
</section>
<section id="session-info">
<h2><span class="section-number">21.8. </span>Session Info<a class="headerlink" href="#session-info" title="Link to this heading">#</a></h2>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="o">%%</span><span class="k">R</span>
sessionInfo()
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>R version 4.2.2 (2022-10-31)
Platform: x86_64-conda-linux-gnu (64-bit)
Running under: Ubuntu 20.04.5 LTS

Matrix products: default
BLAS/LAPACK: /home/dbdimitrov/anaconda3/envs/cellcell/lib/libopenblasp-r0.3.21.so

locale:
 [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              
 [3] LC_TIME=de_DE.UTF-8        LC_COLLATE=en_US.UTF-8    
 [5] LC_MONETARY=de_DE.UTF-8    LC_MESSAGES=en_US.UTF-8   
 [7] LC_PAPER=de_DE.UTF-8       LC_NAME=C                 
 [9] LC_ADDRESS=C               LC_TELEPHONE=C            
[11] LC_MEASUREMENT=de_DE.UTF-8 LC_IDENTIFICATION=C       

attached base packages:
[1] tools     stats     graphics  grDevices utils     datasets  methods  
[8] base     

other attached packages:
[1] nichenetr_1.1.1 tibble_3.1.8    purrr_1.0.1     dplyr_1.0.9    
[5] tidyr_1.2.0     ggplot2_3.4.0   reticulate_1.25

loaded via a namespace (and not attached):
  [1] backports_1.4.1        circlize_0.4.15        Hmisc_4.7-2           
  [4] plyr_1.8.8             igraph_1.3.5           lazyeval_0.2.2        
  [7] sp_1.6-0               splines_4.2.2          listenv_0.9.0         
 [10] scattermore_0.8        digest_0.6.31          foreach_1.5.2         
 [13] htmltools_0.5.4        fansi_1.0.3            checkmate_2.1.0       
 [16] magrittr_2.0.3         tensor_1.5             cluster_2.1.4         
 [19] doParallel_1.0.17      ROCR_1.0-11            limma_3.54.0          
 [22] tzdb_0.3.0             recipes_1.0.4          ComplexHeatmap_2.14.0 
 [25] globals_0.16.2         readr_2.1.2            gower_1.0.1           
 [28] matrixStats_0.63.0     hardhat_1.2.0          timechange_0.2.0      
 [31] spatstat.sparse_3.0-0  jpeg_0.1-10            colorspace_2.0-3      
 [34] ggrepel_0.9.2          xfun_0.36              crayon_1.5.2          
 [37] jsonlite_1.8.4         progressr_0.13.0       spatstat.data_3.0-0   
 [40] survival_3.5-0         zoo_1.8-11             iterators_1.0.14      
 [43] glue_1.6.2             polyclip_1.10-4        gtable_0.3.1          
 [46] ipred_0.9-13           leiden_0.4.3           GetoptLong_1.0.5      
 [49] future.apply_1.10.0    shape_1.4.6            BiocGenerics_0.44.0   
 [52] abind_1.4-5            scales_1.2.1           spatstat.random_3.0-1 
 [55] miniUI_0.1.1.1         Rcpp_1.0.9             htmlTable_2.4.1       
 [58] viridisLite_0.4.1      xtable_1.8-4           clue_0.3-63           
 [61] foreign_0.8-84         proxy_0.4-27           Formula_1.2-4         
 [64] stats4_4.2.2           lava_1.7.1             prodlim_2019.11.13    
 [67] htmlwidgets_1.6.1      httr_1.4.4             DiagrammeR_1.0.9      
 [70] RColorBrewer_1.1-3     ellipsis_0.3.2         Seurat_4.3.0          
 [73] ica_1.0-3              pkgconfig_2.0.3        nnet_7.3-18           
 [76] uwot_0.1.14            deldir_1.0-6           utf8_1.2.2            
 [79] caret_6.0-93           tidyselect_1.2.0       rlang_1.0.6           
 [82] reshape2_1.4.4         later_1.3.0            visNetwork_2.1.0      
 [85] munsell_0.5.0          cli_3.6.0              generics_0.1.3        
 [88] ggridges_0.5.4         fdrtool_1.2.17         stringr_1.5.0         
 [91] fastmap_1.1.0          goftest_1.2-3          knitr_1.41            
 [94] ModelMetrics_1.2.2.2   fitdistrplus_1.1-8     caTools_1.18.2        
 [97] randomForest_4.7-1.1   RANN_2.6.1             pbapply_1.7-0         
[100] future_1.30.0          nlme_3.1-161           mime_0.12             
[103] rstudioapi_0.13        compiler_4.2.2         plotly_4.10.1         
[106] png_0.1-8              e1071_1.7-12           spatstat.utils_3.0-1  
[109] stringi_1.7.12         lattice_0.20-45        Matrix_1.5-3          
[112] vctrs_0.5.1            pillar_1.8.1           lifecycle_1.0.3       
[115] spatstat.geom_3.0-3    lmtest_0.9-40          GlobalOptions_0.1.2   
[118] RcppAnnoy_0.0.20       bitops_1.0-7           data.table_1.14.6     
[121] cowplot_1.1.1          irlba_2.3.5.1          httpuv_1.6.8          
[124] patchwork_1.1.2        latticeExtra_0.6-30    R6_2.5.1              
[127] promises_1.2.0.1       KernSmooth_2.23-20     gridExtra_2.3         
[130] IRanges_2.32.0         parallelly_1.34.0      codetools_0.2-18      
[133] MASS_7.3-58.1          rjson_0.2.21           withr_2.5.0           
[136] SeuratObject_4.1.3     sctransform_0.3.5      S4Vectors_0.36.0      
[139] parallel_4.2.2         hms_1.1.1              grid_4.2.2            
[142] rpart_4.1.19           timeDate_4022.108      class_7.3-20          
[145] Rtsne_0.16             spatstat.explore_3.0-5 pROC_1.18.0           
[148] base64enc_0.1-3        shiny_1.7.4            lubridate_1.9.0       
[151] interp_1.1-3          
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">session_info</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><details>
<summary>Click to view session information</summary>
<pre>
-----
anndata             0.8.0
anndata2ri          1.1
decoupler           1.3.3
liana               0.1.6
matplotlib          3.6.3
numpy               1.23.5
pandas              1.5.3
plotnine            0.10.1
rpy2                3.5.7
scanpy              1.7.2
seaborn             0.12.2
session_info        1.0.0
-----
</pre>
<details>
<summary>Click to view modules imported as dependencies</summary>
<pre>
PIL                         9.4.0
adjustText                  NA
asttokens                   NA
backcall                    0.2.0
beta_ufunc                  NA
binom_ufunc                 NA
cffi                        1.15.1
colorama                    0.4.6
cycler                      0.10.0
cython_runtime              NA
dateutil                    2.8.2
debugpy                     1.5.1
decorator                   5.1.1
defusedxml                  0.7.1
dunamai                     1.15.0
entrypoints                 0.4
executing                   0.8.3
fontTools                   4.38.0
get_version                 3.5.4
h5py                        3.7.0
hypergeom_ufunc             NA
invgauss_ufunc              NA
ipykernel                   6.9.1
ipython_genutils            0.2.0
ipywidgets                  7.6.5
jedi                        0.18.1
jinja2                      3.1.2
joblib                      1.2.0
kiwisolver                  1.4.4
legacy_api_wrap             0.0.0
llvmlite                    0.39.1
markupsafe                  2.1.2
matplotlib_inline           NA
mizani                      0.8.1
mpl_toolkits                NA
natsort                     8.2.0
nbinom_ufunc                NA
ncf_ufunc                   NA
nct_ufunc                   NA
ncx2_ufunc                  NA
numba                       0.56.4
numexpr                     2.8.3
packaging                   23.0
palettable                  3.3.0
parso                       0.8.3
patsy                       0.5.3
pexpect                     4.8.0
pickleshare                 0.7.5
pkg_resources               NA
prompt_toolkit              3.0.20
ptyprocess                  0.7.0
pure_eval                   0.2.2
pycparser                   2.21
pydev_ipython               NA
pydevconsole                NA
pydevd                      2.6.0
pydevd_concurrency_analyser NA
pydevd_file_utils           NA
pydevd_plugins              NA
pydevd_tracing              NA
pygments                    2.11.2
pyparsing                   3.0.9
pytz                        2022.7.1
pytz_deprecation_shim       NA
scipy                       1.10.0
setuptools                  66.1.1
setuptools_scm              NA
sinfo                       0.3.1
six                         1.16.0
skewnorm_ufunc              NA
sklearn                     1.2.0
stack_data                  0.2.0
statsmodels                 0.13.5
tables                      3.7.0
threadpoolctl               3.1.0
tornado                     6.1
tqdm                        4.64.1
traitlets                   5.1.1
typing_extensions           NA
tzlocal                     NA
wcwidth                     0.2.5
zmq                         23.2.0
zoneinfo                    NA
</pre>
</details> <!-- seems like this ends pre, so might as well be explicit -->
<pre>
-----
IPython             8.4.0
jupyter_client      7.2.2
jupyter_core        4.10.0
notebook            6.4.12
-----
Python 3.10.8 | packaged by conda-forge | (main, Nov 22 2022, 08:26:04) [GCC 10.4.0]
Linux-5.15.0-60-generic-x86_64-with-glibc2.31
-----
Session information updated at 2023-02-19 13:46
</pre>
</details></div></div>
</div>
</section>
<section id="references">
<h2><span class="section-number">21.9. </span>References<a class="headerlink" href="#references" title="Link to this heading">#</a></h2>
<div class="docutils container" id="id33">
<div role="list" class="citation-list">
<div class="citation" id="id416" role="doc-biblioentry">
<span class="label"><span class="fn-bracket">[</span>ACJN21<span class="fn-bracket">]</span></span>
<span class="backrefs">(<a role="doc-backlink" href="#id3">1</a>,<a role="doc-backlink" href="#id8">2</a>)</span>
<p>Axel A Almet, Zixuan Cang, Suoqin Jin, and Qing Nie. The landscape of cell-cell communication through single-cell transcriptomics. <em>Current Opinion in Systems Biology</em>, 26:12–23, jun 2021. URL: <a class="reference external" href="https://linkinghub.elsevier.com/retrieve/pii/S2452310021000081">https://linkinghub.elsevier.com/retrieve/pii/S2452310021000081</a> (visited on 2021-09-06), <a class="reference external" href="https://doi.org/10.1016/j.coisb.2021.03.007">doi:10.1016/j.coisb.2021.03.007</a>.</p>
</div>
<div class="citation" id="id425" role="doc-biblioentry">
<span class="label"><span class="fn-bracket">[</span><a role="doc-backlink" href="#id30">ABM+22</a><span class="fn-bracket">]</span></span>
<p>Erick Armingol, Hratch M Baghdassarian, Cameron Martino, Araceli Perez-Lopez, Caitlin Aamodt, Rob Knight, and Nathan E Lewis. Context-aware deconvolution of cell-cell communication with tensor-cell2cell. <em>Nature Communications</em>, 13(1):3665, jun 2022. URL: <a class="reference external" href="https://www.nature.com/articles/s41467-022-31369-2">https://www.nature.com/articles/s41467-022-31369-2</a> (visited on 2021-09-24), <a class="reference external" href="https://doi.org/10.1038/s41467-022-31369-2">doi:10.1038/s41467-022-31369-2</a>.</p>
</div>
<div class="citation" id="id413" role="doc-biblioentry">
<span class="label"><span class="fn-bracket">[</span>AOHL21<span class="fn-bracket">]</span></span>
<span class="backrefs">(<a role="doc-backlink" href="#id1">1</a>,<a role="doc-backlink" href="#id2">2</a>,<a role="doc-backlink" href="#id8">3</a>,<a role="doc-backlink" href="#id22">4</a>,<a role="doc-backlink" href="#id26">5</a>)</span>
<p>Erick Armingol, Adam Officer, Olivier Harismendy, and Nathan E Lewis. Deciphering cell-cell interactions and communication from gene expression. <em>Nature Reviews. Genetics</em>, 22(2):71–88, 2021. URL: <a class="reference external" href="http://www.nature.com/articles/s41576-020-00292-x">http://www.nature.com/articles/s41576-020-00292-x</a> (visited on 2021-10-05), <a class="reference external" href="https://doi.org/10.1038/s41576-020-00292-x">doi:10.1038/s41576-020-00292-x</a>.</p>
</div>
<div class="citation" id="id434" role="doc-biblioentry">
<span class="label"><span class="fn-bracket">[</span><a role="doc-backlink" href="#id21">BCDC22</a><span class="fn-bracket">]</span></span>
<p>Giacomo Baruzzo, Giulia Cesaro, and Barbara Di Camillo. Identify, quantify and characterize cellular communication from single-cell rna sequencing data with scseqcomm. <em>Bioinformatics</em>, 38(7):1920–1929, 2022.</p>
</div>
<div class="citation" id="id411" role="doc-biblioentry">
<span class="label"><span class="fn-bracket">[</span>BSS20<span class="fn-bracket">]</span></span>
<span class="backrefs">(<a role="doc-backlink" href="#id5">1</a>,<a role="doc-backlink" href="#id13">2</a>,<a role="doc-backlink" href="#id17">3</a>,<a role="doc-backlink" href="#id18">4</a>,<a role="doc-backlink" href="#id19">5</a>,<a role="doc-backlink" href="#id20">6</a>)</span>
<p>Robin Browaeys, Wouter Saelens, and Yvan Saeys. NicheNet: modeling intercellular communication by linking ligands to target genes. <em>Nature Methods</em>, 17(2):159–162, feb 2020. URL: <a class="reference external" href="http://www.nature.com/articles/s41592-019-0667-5">http://www.nature.com/articles/s41592-019-0667-5</a> (visited on 2021-10-05), <a class="reference external" href="https://doi.org/10.1038/s41592-019-0667-5">doi:10.1038/s41592-019-0667-5</a>.</p>
</div>
<div class="citation" id="id423" role="doc-biblioentry">
<span class="label"><span class="fn-bracket">[</span>DTGR+22<span class="fn-bracket">]</span></span>
<span class="backrefs">(<a role="doc-backlink" href="#id7">1</a>,<a role="doc-backlink" href="#id9">2</a>,<a role="doc-backlink" href="#id10">3</a>,<a role="doc-backlink" href="#id12">4</a>,<a role="doc-backlink" href="#id16">5</a>,<a role="doc-backlink" href="#id22">6</a>,<a role="doc-backlink" href="#id24">7</a>)</span>
<p>Daniel Dimitrov, Dénes Türei, Martin Garrido-Rodriguez, Paul L Burmedi, James S Nagai, Charlotte Boys, Ricardo O Ramirez Flores, Hyojin Kim, Bence Szalai, Ivan G Costa, Alberto Valdeolivas, Aurélien Dugourd, and Julio Saez-Rodriguez. Comparison of methods and resources for cell-cell communication inference from single-cell RNA-seq data. <em>Nature Communications</em>, 13(1):3224, jun 2022. URL: <a class="reference external" href="https://www.nature.com/articles/s41467-022-30755-0">https://www.nature.com/articles/s41467-022-30755-0</a> (visited on 2021-05-31), <a class="reference external" href="https://doi.org/10.1038/s41467-022-30755-0">doi:10.1038/s41467-022-30755-0</a>.</p>
</div>
<div class="citation" id="id412" role="doc-biblioentry">
<span class="label"><span class="fn-bracket">[</span>EVTTVT20<span class="fn-bracket">]</span></span>
<span class="backrefs">(<a role="doc-backlink" href="#id4">1</a>,<a role="doc-backlink" href="#id6">2</a>,<a role="doc-backlink" href="#id11">3</a>,<a role="doc-backlink" href="#id15">4</a>)</span>
<p>Mirjana Efremova, Miquel Vento-Tormo, Sarah A Teichmann, and Roser Vento-Tormo. CellPhoneDB: inferring cell-cell communication from combined expression of multi-subunit ligand-receptor complexes. <em>Nature Protocols</em>, 15(4):1484–1506, apr 2020. URL: <a class="reference external" href="http://www.nature.com/articles/s41596-020-0292-x">http://www.nature.com/articles/s41596-020-0292-x</a> (visited on 2020-12-04), <a class="reference external" href="https://doi.org/10.1038/s41596-020-0292-x">doi:10.1038/s41596-020-0292-x</a>.</p>
</div>
<div class="citation" id="id424" role="doc-biblioentry">
<span class="label"><span class="fn-bracket">[</span><a role="doc-backlink" href="#id32">GALM+22</a><span class="fn-bracket">]</span></span>
<p>Luz Garcia-Alonso, Valentina Lorenzi, Cecilia Icoresi Mazzeo, João Pedro Alves-Lopes, Kenny Roberts, Carmen Sancho-Serra, Justin Engelbert, Magda Marečková, Wolfram H Gruhn, Rachel A Botting, Tong Li, Berta Crespo, Stijn van Dongen, Vladimir Yu Kiselev, Elena Prigmore, Mary Herbert, Ashley Moffett, Alain Chédotal, Omer Ali Bayraktar, Azim Surani, Muzlifah Haniffa, and Roser Vento-Tormo. Single-cell roadmap of human gonadal development. <em>Nature</em>, 607(7919):540–547, jul 2022. URL: <a class="reference external" href="https://www.nature.com/articles/s41586-022-04918-4">https://www.nature.com/articles/s41586-022-04918-4</a> (visited on 2022-07-07), <a class="reference external" href="https://doi.org/10.1038/s41586-022-04918-4">doi:10.1038/s41586-022-04918-4</a>.</p>
</div>
<div class="citation" id="id429" role="doc-biblioentry">
<span class="label"><span class="fn-bracket">[</span>HDO+20<span class="fn-bracket">]</span></span>
<span class="backrefs">(<a role="doc-backlink" href="#id4">1</a>,<a role="doc-backlink" href="#id27">2</a>)</span>
<p>Rui Hou, Elena Denisenko, Huan Ting Ong, Jordan A Ramilowski, and Alistair RR Forrest. Predicting cell-to-cell communication networks using natmi. <em>Nature communications</em>, 11(1):1–11, 2020.</p>
</div>
<div class="citation" id="id418" role="doc-biblioentry">
<span class="label"><span class="fn-bracket">[</span><a role="doc-backlink" href="#id5">HPGT21</a><span class="fn-bracket">]</span></span>
<p>Yuxuan Hu, Tao Peng, Lin Gao, and Kai Tan. CytoTalk: de novo construction of signal transduction networks using single-cell transcriptomic data. <em>Science Advances</em>, apr 2021. URL: <a class="reference external" href="https://advances.sciencemag.org/lookup/doi/10.1126/sciadv.abf1356">https://advances.sciencemag.org/lookup/doi/10.1126/sciadv.abf1356</a> (visited on 2021-10-11), <a class="reference external" href="https://doi.org/10.1126/sciadv.abf1356">doi:10.1126/sciadv.abf1356</a>.</p>
</div>
<div class="citation" id="id415" role="doc-biblioentry">
<span class="label"><span class="fn-bracket">[</span>JGJZ+21<span class="fn-bracket">]</span></span>
<span class="backrefs">(<a role="doc-backlink" href="#id4">1</a>,<a role="doc-backlink" href="#id6">2</a>,<a role="doc-backlink" href="#id29">3</a>)</span>
<p>Suoqin Jin, Christian F Guerrero-Juarez, Lihua Zhang, Ivan Chang, Raul Ramos, Chen-Hsiang Kuan, Peggy Myung, Maksim V Plikus, and Qing Nie. Inference and analysis of cell-cell communication using CellChat. <em>Nature Communications</em>, 12(1):1088, feb 2021. URL: <a class="reference external" href="http://dx.doi.org/10.1038/s41467-021-21246-9">http://dx.doi.org/10.1038/s41467-021-21246-9</a> (visited on 2021-10-05), <a class="reference external" href="https://doi.org/10.1038/s41467-021-21246-9">doi:10.1038/s41467-021-21246-9</a>.</p>
</div>
<div class="citation" id="id432" role="doc-biblioentry">
<span class="label"><span class="fn-bracket">[</span><a role="doc-backlink" href="#id14">KST+18</a><span class="fn-bracket">]</span></span>
<p>Hyun Min Kang, Meena Subramaniam, Sasha Targ, Michelle Nguyen, Lenka Maliskova, Elizabeth McCarthy, Eunice Wan, Simon Wong, Lauren Byrnes, Cristina M Lanata, and others. Multiplexed droplet single-cell rna-sequencing using natural genetic variation. <em>Nature biotechnology</em>, 36(1):89–94, 2018.</p>
</div>
<div class="citation" id="id426" role="doc-biblioentry">
<span class="label"><span class="fn-bracket">[</span>LSW22<span class="fn-bracket">]</span></span>
<span class="backrefs">(<a role="doc-backlink" href="#id6">1</a>,<a role="doc-backlink" href="#id7">2</a>,<a role="doc-backlink" href="#id9">3</a>,<a role="doc-backlink" href="#id10">4</a>)</span>
<p>Zhaoyang Liu, Dongqing Sun, and Chenfei Wang. Evaluation of cell-cell interaction methods by integrating single-cell RNA sequencing data with spatial information. <em>Genome Biology</em>, 23(1):218, oct 2022. URL: <a class="reference external" href="http://dx.doi.org/10.1186/s13059-022-02783-y">http://dx.doi.org/10.1186/s13059-022-02783-y</a> (visited on 2022-10-20), <a class="reference external" href="https://doi.org/10.1186/s13059-022-02783-y">doi:10.1186/s13059-022-02783-y</a>.</p>
</div>
<div class="citation" id="id430" role="doc-biblioentry">
<span class="label"><span class="fn-bracket">[</span><a role="doc-backlink" href="#id28">NLeimkuhlerS+21</a><span class="fn-bracket">]</span></span>
<p>James S Nagai, Nils B Leimkühler, Michael T Schaub, Rebekka K Schneider, and Ivan G Costa. Crosstalker: analysis and visualization of ligand–receptorne tworks. <em>Bioinformatics</em>, 37(22):4263–4265, 2021.</p>
</div>
<div class="citation" id="id431" role="doc-biblioentry">
<span class="label"><span class="fn-bracket">[</span><a role="doc-backlink" href="#id23">PFRT22</a><span class="fn-bracket">]</span></span>
<p>Giovanni Palla, David S Fischer, Aviv Regev, and Fabian J Theis. Spatial components of molecular tissue biology. <em>Nature Biotechnology</em>, 40(3):308–318, 2022.</p>
</div>
<div class="citation" id="id422" role="doc-biblioentry">
<span class="label"><span class="fn-bracket">[</span><a role="doc-backlink" href="#id4">RYG+22</a><span class="fn-bracket">]</span></span>
<p>Micha Sam Brickman Raredon, Junchen Yang, James Garritano, Meng Wang, Dan Kushnir, Jonas Christian Schupp, Taylor S Adams, Allison M Greaney, Katherine L Leiby, Naftali Kaminski, Yuval Kluger, Andre Levchenko, and Laura E Niklason. Computation and visualization of cell-cell signaling topologies in single-cell systems data using connectome. <em>Scientific Reports</em>, 12(1):4187, mar 2022. URL: <a class="reference external" href="http://dx.doi.org/10.1038/s41598-022-07959-x">http://dx.doi.org/10.1038/s41598-022-07959-x</a> (visited on 2021-01-25), <a class="reference external" href="https://doi.org/10.1038/s41598-022-07959-x">doi:10.1038/s41598-022-07959-x</a>.</p>
</div>
<div class="citation" id="id427" role="doc-biblioentry">
<span class="label"><span class="fn-bracket">[</span><a role="doc-backlink" href="#id31">RYK+23</a><span class="fn-bracket">]</span></span>
<p>Micha Sam Brickman Raredon, Junchen Yang, Neeharika Kothapalli, Wesley Lewis, Naftali Kaminski, Laura E Niklason, and Yuval Kluger. Comprehensive visualization of cell-cell interactions in single-cell and spatial transcriptomics with NICHES. <em>Bioinformatics</em>, jan 2023. URL: <a class="reference external" href="http://dx.doi.org/10.1093/bioinformatics/btac775">http://dx.doi.org/10.1093/bioinformatics/btac775</a> (visited on 2022-01-27), <a class="reference external" href="https://doi.org/10.1093/bioinformatics/btac775">doi:10.1093/bioinformatics/btac775</a>.</p>
</div>
<div class="citation" id="id428" role="doc-biblioentry">
<span class="label"><span class="fn-bracket">[</span>WZC+22<span class="fn-bracket">]</span></span>
<span class="backrefs">(<a role="doc-backlink" href="#id7">1</a>,<a role="doc-backlink" href="#id9">2</a>)</span>
<p>Saidi Wang, Hansi Zheng, James S Choi, Jae K Lee, Xiaoman Li, and Haiyan Hu. A systematic evaluation of the computational tools for ligand-receptor-based cell-cell interaction inference. <em>Briefings in functional genomics</em>, 21(5):339–356, sep 2022. URL: <a class="reference external" href="http://dx.doi.org/10.1093/bfgp/elac019">http://dx.doi.org/10.1093/bfgp/elac019</a> (visited on 2022-04-27), <a class="reference external" href="https://doi.org/10.1093/bfgp/elac019">doi:10.1093/bfgp/elac019</a>.</p>
</div>
<div class="citation" id="id410" role="doc-biblioentry">
<span class="label"><span class="fn-bracket">[</span>WKMacLeanN19<span class="fn-bracket">]</span></span>
<span class="backrefs">(<a role="doc-backlink" href="#id5">1</a>,<a role="doc-backlink" href="#id31">2</a>)</span>
<p>Shuxiong Wang, Matthew Karikomi, Adam L MacLean, and Qing Nie. Cell lineage and communication network inference via optimization for single-cell transcriptomics. <em>Nucleic Acids Research</em>, 47(11):e66, jun 2019. URL: <a class="reference external" href="https://academic.oup.com/nar/article/47/11/e66/5421812">https://academic.oup.com/nar/article/47/11/e66/5421812</a> (visited on 2019-05-01), <a class="reference external" href="https://doi.org/10.1093/nar/gkz204">doi:10.1093/nar/gkz204</a>.</p>
</div>
<div class="citation" id="id420" role="doc-biblioentry">
<span class="label"><span class="fn-bracket">[</span>WSHB22<span class="fn-bracket">]</span></span>
<span class="backrefs">(<a role="doc-backlink" href="#id25">1</a>,<a role="doc-backlink" href="#id31">2</a>)</span>
<p>Aaron J Wilk, Alex K Shalek, Susan Holmes, and Catherine A Blish. Comparative analysis of cell-cell communication at single-cell resolution. <em>BioRxiv</em>, mar 2022. URL: <a class="reference external" href="http://biorxiv.org/lookup/doi/10.1101/2022.02.04.479209">http://biorxiv.org/lookup/doi/10.1101/2022.02.04.479209</a> (visited on 2022-02-22), <a class="reference external" href="https://doi.org/10.1101/2022.02.04.479209">doi:10.1101/2022.02.04.479209</a>.</p>
</div>
<div class="citation" id="id433" role="doc-biblioentry">
<span class="label"><span class="fn-bracket">[</span><a role="doc-backlink" href="#id21">ZLH+21</a><span class="fn-bracket">]</span></span>
<p>Yang Zhang, Tianyuan Liu, Xuesong Hu, Mei Wang, Jing Wang, Bohao Zou, Puwen Tan, Tianyu Cui, Yiying Dou, Lin Ning, and others. Cellcall: integrating paired ligand–receptor and transcription factor activities for cell–cell communication. <em>Nucleic Acids Research</em>, 49(15):8520–8534, 2021.</p>
</div>
<div class="citation" id="id435" role="doc-biblioentry">
<span class="label"><span class="fn-bracket">[</span><a role="doc-backlink" href="#id21">ZLW+21a</a><span class="fn-bracket">]</span></span>
<p>Yang Zhang, Tianyuan Liu, Jing Wang, Bohao Zou, Le Li, Linhui Yao, Kechen Chen, Lin Ning, Bingyi Wu, Xiaoyang Zhao, and others. Cellinker: a platform of ligand–receptor interactions for intercellular communication analysis. <em>Bioinformatics</em>, 37(14):2025–2032, 2021.</p>
</div>
<div class="citation" id="id417" role="doc-biblioentry">
<span class="label"><span class="fn-bracket">[</span><a role="doc-backlink" href="#id32">ZLW+21b</a><span class="fn-bracket">]</span></span>
<p>Yang Zhang, Tianyuan Liu, Jing Wang, Bohao Zou, Le Li, Linhui Yao, Kechen Chen, Lin Ning, Bingyi Wu, Xiaoyang Zhao, and Dong Wang. Cellinker: a platform of ligand-receptor interactions for intercellular communication analysis. <em>Bioinformatics</em>, jan 2021. URL: <a class="reference external" href="http://dx.doi.org/10.1093/bioinformatics/btab036">http://dx.doi.org/10.1093/bioinformatics/btab036</a> (visited on 2021-09-12), <a class="reference external" href="https://doi.org/10.1093/bioinformatics/btab036">doi:10.1093/bioinformatics/btab036</a>.</p>
</div>
<div class="citation" id="id421" role="doc-biblioentry">
<span class="label"><span class="fn-bracket">[</span><a role="doc-backlink" href="#id32">ZZT+22</a><span class="fn-bracket">]</span></span>
<p>Rongbin Zheng, Yang Zhang, Tadataka Tsuji, Lili Zhang, Yu-Hua Tseng, and Kaifu Chen. MEBOCOST: metabolic cell-cell communication modeling by single cell transcriptome. <em>BioRxiv</em>, may 2022. URL: <a class="reference external" href="http://biorxiv.org/lookup/doi/10.1101/2022.05.30.494067">http://biorxiv.org/lookup/doi/10.1101/2022.05.30.494067</a> (visited on 2022-06-02), <a class="reference external" href="https://doi.org/10.1101/2022.05.30.494067">doi:10.1101/2022.05.30.494067</a>.</p>
</div>
</div>
</div>
</section>
</section>

    <script type="text/x-thebe-config">
    {
        requestKernel: true,
        binderOptions: {
            repo: "binder-examples/jupyter-stacks-datascience",
            ref: "master",
        },
        codeMirrorConfig: {
            theme: "abcdef",
            mode: "python"
        },
        kernelOptions: {
            name: "python3",
            path: "./mechanisms"
        },
        predefinedOutput: true
    }
    </script>
    <script>kernelName = 'python3'</script>

                </article>
              

              
              
              
              
                <footer class="prev-next-footer d-print-none">
                  
<div class="prev-next-area">
    <a class="left-prev"
       href="gene_regulatory_networks.html"
       title="previous page">
      <i class="fa-solid fa-angle-left"></i>
      <div class="prev-next-info">
        <p class="prev-next-subtitle">previous</p>
        <p class="prev-next-title"><span class="section-number">20. </span>Gene regulatory networks</p>
      </div>
    </a>
    <a class="right-next"
       href="../deconvolution/bulk_deconvolution.html"
       title="next page">
      <div class="prev-next-info">
        <p class="prev-next-subtitle">next</p>
        <p class="prev-next-title"><span class="section-number">22. </span>Bulk deconvolution</p>
      </div>
      <i class="fa-solid fa-angle-right"></i>
    </a>
</div>
                </footer>
              
            </div>
            
            
              
                <dialog id="pst-secondary-sidebar-modal"></dialog>
                <div id="pst-secondary-sidebar" class="bd-sidebar-secondary bd-toc"><div class="sidebar-secondary-items sidebar-secondary__inner">


  <div class="sidebar-secondary-item">
  <div class="page-toc tocsection onthispage">
    <i class="fa-solid fa-list"></i> Contents
  </div>
  <nav class="bd-toc-nav page-toc">
    <ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#motivation">21.1. Motivation</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#approaches">21.2. Approaches</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#environment-setup">21.3. Environment setup</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#case-study">21.4. Case Study</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#ligand-receptor-inference">21.4.1. Ligand-receptor inference</a><ul class="nav section-nav flex-column">
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#visual-exploration">21.4.1.1. Visual exploration</a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#generating-a-ligand-receptor-consensus-with-liana">21.4.1.2. Generating a Ligand-Receptor consensus with LIANA</a></li>
</ul>
</li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#modelling-differential-intercellular-signalling-with-nichenet">21.4.2. Modelling differential intercellular signalling with NicheNet</a><ul class="nav section-nav flex-column">
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#load-nichenet-prior-knowledge">21.4.2.1. Load NicheNet Prior-Knowledge</a><ul class="nav section-nav flex-column">
<li class="toc-h5 nav-item toc-entry"><a class="reference internal nav-link" href="#step-1-define-cell-types-of-interest-to-be-considered-as-senders-sources-and-receiver-targets-of-ccc-interactions">21.4.2.1.1. Step 1. Define cell types of interest to be considered as senders/sources and receiver/targets of CCC interactions</a></li>
<li class="toc-h5 nav-item toc-entry"><a class="reference internal nav-link" href="#step-2-define-a-set-of-ligands-that-can-potentially-affect-receiver-cell-types">21.4.2.1.2. Step 2. Define a set of ligands that can <strong>potentially</strong> affect receiver cell types</a></li>
<li class="toc-h5 nav-item toc-entry"><a class="reference internal nav-link" href="#step-3-define-a-gene-set-of-interest-in-receiver-cell-type-s">21.4.2.1.3. Step 3. Define a gene set of interest in receiver cell type(s)</a></li>
<li class="toc-h5 nav-item toc-entry"><a class="reference internal nav-link" href="#step-4-nichenet-ligand-activity-estimation">21.4.2.1.4. Step 4. NicheNet ligand activity estimation</a></li>
<li class="toc-h5 nav-item toc-entry"><a class="reference internal nav-link" href="#step-5-infer-visualize-top-predicted-target-genes-for-top-ligands">21.4.2.1.5. Step 5. Infer &amp; Visualize top-predicted target genes for top ligands</a></li>
<li class="toc-h5 nav-item toc-entry"><a class="reference internal nav-link" href="#visualize-top-ligands-regulatory-targets">21.4.2.1.6. Visualize top ligands &amp; regulatory targets</a></li>
</ul>
</li>
</ul>
</li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#combining-nichenet-output-with-ligand-receptor-inference">21.4.3. Combining NicheNet output with ligand-receptor inference</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#key-takeaways">21.5. Key takeaways</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#assumptions-limitations">21.5.1. Assumptions &amp; Limitations</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#summary-outlook">21.5.2. Summary &amp; Outlook:</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#quiz">21.6. Quiz</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#contributions">21.7. Contributions</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#authors">21.7.1. Authors:</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#reviewers">21.7.2. Reviewers:</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#session-info">21.8. Session Info</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#references">21.9. References</a></li>
</ul>
  </nav></div>

</div></div>
              
            
          </div>
          <footer class="bd-footer-content">
            
<div class="bd-footer-content__inner container">
  
  <div class="footer-item">
    
<p class="component-author">
By Lukas Heumos, Anna Schaar, single-cell best practices consortium
</p>

  </div>
  
  <div class="footer-item">
    

  <p class="copyright">
    
      © Copyright 2023.
      <br/>
    
  </p>

  </div>
  
  <div class="footer-item">
    
  </div>
  
  <div class="footer-item">
    
<div class="extra_footer">
  <div>
  Brought to you by
  <a href="https://www.github.com/theislab">Theislab</a>, with many thanks to the single-cell community as a whole!
</div>

</div>
  </div>
  
</div>
          </footer>
        

      </main>
    </div>
  </div>
  
  <!-- Scripts loaded after <body> so the DOM is not blocked -->
  <script defer src="../_static/scripts/bootstrap.js?digest=8878045cc6db502f8baf"></script>
<script defer src="../_static/scripts/pydata-sphinx-theme.js?digest=8878045cc6db502f8baf"></script>

  <footer class="bd-footer">
  </footer>
  </body>
</html>